// @cdktf/provider-google/lib/data-google-dataproc-metastore-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleDataprocMetastoreServiceTelemetryConfigToHclTerraform = exports.dataGoogleDataprocMetastoreServiceTelemetryConfigToTerraform = exports.DataGoogleDataprocMetastoreServiceScheduledBackupList = exports.DataGoogleDataprocMetastoreServiceScheduledBackupOutputReference = exports.dataGoogleDataprocMetastoreServiceScheduledBackupToHclTerraform = exports.dataGoogleDataprocMetastoreServiceScheduledBackupToTerraform = exports.DataGoogleDataprocMetastoreServiceScalingConfigList = exports.DataGoogleDataprocMetastoreServiceScalingConfigOutputReference = exports.dataGoogleDataprocMetastoreServiceScalingConfigToHclTerraform = exports.dataGoogleDataprocMetastoreServiceScalingConfigToTerraform = exports.DataGoogleDataprocMetastoreServiceNetworkConfigList = exports.DataGoogleDataprocMetastoreServiceNetworkConfigOutputReference = exports.dataGoogleDataprocMetastoreServiceNetworkConfigToHclTerraform = exports.dataGoogleDataprocMetastoreServiceNetworkConfigToTerraform = exports.DataGoogleDataprocMetastoreServiceNetworkConfigConsumersList = exports.DataGoogleDataprocMetastoreServiceNetworkConfigConsumersOutputReference = exports.dataGoogleDataprocMetastoreServiceNetworkConfigConsumersToHclTerraform = exports.dataGoogleDataprocMetastoreServiceNetworkConfigConsumersToTerraform = exports.DataGoogleDataprocMetastoreServiceMetadataIntegrationList = exports.DataGoogleDataprocMetastoreServiceMetadataIntegrationOutputReference = exports.dataGoogleDataprocMetastoreServiceMetadataIntegrationToHclTerraform = exports.dataGoogleDataprocMetastoreServiceMetadataIntegrationToTerraform = exports.DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigList = exports.DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigOutputReference = exports.dataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigToHclTerraform = exports.dataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigToTerraform = exports.DataGoogleDataprocMetastoreServiceMaintenanceWindowList = exports.DataGoogleDataprocMetastoreServiceMaintenanceWindowOutputReference = exports.dataGoogleDataprocMetastoreServiceMaintenanceWindowToHclTerraform = exports.dataGoogleDataprocMetastoreServiceMaintenanceWindowToTerraform = exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigList = exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigOutputReference = exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigToHclTerraform = exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigToTerraform = exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigList = exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigOutputReference = exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigToHclTerraform = exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigToTerraform = exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabList = exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabOutputReference = exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabToHclTerraform = exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabToTerraform = exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsList = exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsOutputReference = exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsToHclTerraform = exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsToTerraform = exports.DataGoogleDataprocMetastoreServiceEncryptionConfigList = exports.DataGoogleDataprocMetastoreServiceEncryptionConfigOutputReference = exports.dataGoogleDataprocMetastoreServiceEncryptionConfigToHclTerraform = exports.dataGoogleDataprocMetastoreServiceEncryptionConfigToTerraform = void 0;
exports.DataGoogleDataprocMetastoreService = exports.DataGoogleDataprocMetastoreServiceTelemetryConfigList = exports.DataGoogleDataprocMetastoreServiceTelemetryConfigOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleDataprocMetastoreServiceEncryptionConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceEncryptionConfigToTerraform = dataGoogleDataprocMetastoreServiceEncryptionConfigToTerraform;
function dataGoogleDataprocMetastoreServiceEncryptionConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceEncryptionConfigToHclTerraform = dataGoogleDataprocMetastoreServiceEncryptionConfigToHclTerraform;
class DataGoogleDataprocMetastoreServiceEncryptionConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key - computed: true, optional: false, required: false
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
}
exports.DataGoogleDataprocMetastoreServiceEncryptionConfigOutputReference = DataGoogleDataprocMetastoreServiceEncryptionConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceEncryptionConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceEncryptionConfigOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceEncryptionConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceEncryptionConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceEncryptionConfigList = DataGoogleDataprocMetastoreServiceEncryptionConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceEncryptionConfigList[_b] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceEncryptionConfigList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsToTerraform = dataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsToTerraform;
function dataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsToHclTerraform = dataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsToHclTerraform;
class DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // config_overrides - computed: true, optional: false, required: false
        this._configOverrides = new cdktf.StringMap(this, "config_overrides");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get configOverrides() {
        return this._configOverrides;
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsOutputReference = DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsList = DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsList[_d] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabToTerraform = dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabToTerraform;
function dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabToHclTerraform = dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabToHclTerraform;
class DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cloud_secret - computed: true, optional: false, required: false
    get cloudSecret() {
        return this.getStringAttribute('cloud_secret');
    }
}
exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabOutputReference = DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabList = DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabList[_f] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigToTerraform = dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigToTerraform;
function dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigToHclTerraform = dataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigToHclTerraform;
class DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // keytab - computed: true, optional: false, required: false
        this._keytab = new DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigKeytabList(this, "keytab", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get keytab() {
        return this._keytab;
    }
    // krb5_config_gcs_uri - computed: true, optional: false, required: false
    get krb5ConfigGcsUri() {
        return this.getStringAttribute('krb5_config_gcs_uri');
    }
    // principal - computed: true, optional: false, required: false
    get principal() {
        return this.getStringAttribute('principal');
    }
}
exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigOutputReference = DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigList = DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigList[_h] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceHiveMetastoreConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigToTerraform = dataGoogleDataprocMetastoreServiceHiveMetastoreConfigToTerraform;
function dataGoogleDataprocMetastoreServiceHiveMetastoreConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceHiveMetastoreConfigToHclTerraform = dataGoogleDataprocMetastoreServiceHiveMetastoreConfigToHclTerraform;
class DataGoogleDataprocMetastoreServiceHiveMetastoreConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // auxiliary_versions - computed: true, optional: false, required: false
        this._auxiliaryVersions = new DataGoogleDataprocMetastoreServiceHiveMetastoreConfigAuxiliaryVersionsList(this, "auxiliary_versions", true);
        // config_overrides - computed: true, optional: false, required: false
        this._configOverrides = new cdktf.StringMap(this, "config_overrides");
        // kerberos_config - computed: true, optional: false, required: false
        this._kerberosConfig = new DataGoogleDataprocMetastoreServiceHiveMetastoreConfigKerberosConfigList(this, "kerberos_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get auxiliaryVersions() {
        return this._auxiliaryVersions;
    }
    get configOverrides() {
        return this._configOverrides;
    }
    // endpoint_protocol - computed: true, optional: false, required: false
    get endpointProtocol() {
        return this.getStringAttribute('endpoint_protocol');
    }
    get kerberosConfig() {
        return this._kerberosConfig;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigOutputReference = DataGoogleDataprocMetastoreServiceHiveMetastoreConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceHiveMetastoreConfigOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceHiveMetastoreConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceHiveMetastoreConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigList = DataGoogleDataprocMetastoreServiceHiveMetastoreConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceHiveMetastoreConfigList[_k] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceHiveMetastoreConfigList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceMaintenanceWindowToTerraform = dataGoogleDataprocMetastoreServiceMaintenanceWindowToTerraform;
function dataGoogleDataprocMetastoreServiceMaintenanceWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceMaintenanceWindowToHclTerraform = dataGoogleDataprocMetastoreServiceMaintenanceWindowToHclTerraform;
class DataGoogleDataprocMetastoreServiceMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day_of_week - computed: true, optional: false, required: false
    get dayOfWeek() {
        return this.getStringAttribute('day_of_week');
    }
    // hour_of_day - computed: true, optional: false, required: false
    get hourOfDay() {
        return this.getNumberAttribute('hour_of_day');
    }
}
exports.DataGoogleDataprocMetastoreServiceMaintenanceWindowOutputReference = DataGoogleDataprocMetastoreServiceMaintenanceWindowOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceMaintenanceWindowOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceMaintenanceWindowOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceMaintenanceWindowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceMaintenanceWindowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceMaintenanceWindowList = DataGoogleDataprocMetastoreServiceMaintenanceWindowList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceMaintenanceWindowList[_m] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceMaintenanceWindowList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigToTerraform = dataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigToTerraform;
function dataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigToHclTerraform = dataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigToHclTerraform;
class DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigOutputReference = DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigList = DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigList[_p] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceMetadataIntegrationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceMetadataIntegrationToTerraform = dataGoogleDataprocMetastoreServiceMetadataIntegrationToTerraform;
function dataGoogleDataprocMetastoreServiceMetadataIntegrationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceMetadataIntegrationToHclTerraform = dataGoogleDataprocMetastoreServiceMetadataIntegrationToHclTerraform;
class DataGoogleDataprocMetastoreServiceMetadataIntegrationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // data_catalog_config - computed: true, optional: false, required: false
        this._dataCatalogConfig = new DataGoogleDataprocMetastoreServiceMetadataIntegrationDataCatalogConfigList(this, "data_catalog_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get dataCatalogConfig() {
        return this._dataCatalogConfig;
    }
}
exports.DataGoogleDataprocMetastoreServiceMetadataIntegrationOutputReference = DataGoogleDataprocMetastoreServiceMetadataIntegrationOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceMetadataIntegrationOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceMetadataIntegrationOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceMetadataIntegrationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceMetadataIntegrationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceMetadataIntegrationList = DataGoogleDataprocMetastoreServiceMetadataIntegrationList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceMetadataIntegrationList[_r] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceMetadataIntegrationList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceNetworkConfigConsumersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceNetworkConfigConsumersToTerraform = dataGoogleDataprocMetastoreServiceNetworkConfigConsumersToTerraform;
function dataGoogleDataprocMetastoreServiceNetworkConfigConsumersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceNetworkConfigConsumersToHclTerraform = dataGoogleDataprocMetastoreServiceNetworkConfigConsumersToHclTerraform;
class DataGoogleDataprocMetastoreServiceNetworkConfigConsumersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // endpoint_uri - computed: true, optional: false, required: false
    get endpointUri() {
        return this.getStringAttribute('endpoint_uri');
    }
    // subnetwork - computed: true, optional: false, required: false
    get subnetwork() {
        return this.getStringAttribute('subnetwork');
    }
}
exports.DataGoogleDataprocMetastoreServiceNetworkConfigConsumersOutputReference = DataGoogleDataprocMetastoreServiceNetworkConfigConsumersOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceNetworkConfigConsumersOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceNetworkConfigConsumersOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceNetworkConfigConsumersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceNetworkConfigConsumersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceNetworkConfigConsumersList = DataGoogleDataprocMetastoreServiceNetworkConfigConsumersList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceNetworkConfigConsumersList[_t] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceNetworkConfigConsumersList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceNetworkConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceNetworkConfigToTerraform = dataGoogleDataprocMetastoreServiceNetworkConfigToTerraform;
function dataGoogleDataprocMetastoreServiceNetworkConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceNetworkConfigToHclTerraform = dataGoogleDataprocMetastoreServiceNetworkConfigToHclTerraform;
class DataGoogleDataprocMetastoreServiceNetworkConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // consumers - computed: true, optional: false, required: false
        this._consumers = new DataGoogleDataprocMetastoreServiceNetworkConfigConsumersList(this, "consumers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get consumers() {
        return this._consumers;
    }
}
exports.DataGoogleDataprocMetastoreServiceNetworkConfigOutputReference = DataGoogleDataprocMetastoreServiceNetworkConfigOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceNetworkConfigOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceNetworkConfigOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceNetworkConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceNetworkConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceNetworkConfigList = DataGoogleDataprocMetastoreServiceNetworkConfigList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceNetworkConfigList[_v] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceNetworkConfigList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceScalingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceScalingConfigToTerraform = dataGoogleDataprocMetastoreServiceScalingConfigToTerraform;
function dataGoogleDataprocMetastoreServiceScalingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceScalingConfigToHclTerraform = dataGoogleDataprocMetastoreServiceScalingConfigToHclTerraform;
class DataGoogleDataprocMetastoreServiceScalingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // instance_size - computed: true, optional: false, required: false
    get instanceSize() {
        return this.getStringAttribute('instance_size');
    }
    // scaling_factor - computed: true, optional: false, required: false
    get scalingFactor() {
        return this.getNumberAttribute('scaling_factor');
    }
}
exports.DataGoogleDataprocMetastoreServiceScalingConfigOutputReference = DataGoogleDataprocMetastoreServiceScalingConfigOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceScalingConfigOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceScalingConfigOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceScalingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceScalingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceScalingConfigList = DataGoogleDataprocMetastoreServiceScalingConfigList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceScalingConfigList[_x] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceScalingConfigList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceScheduledBackupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceScheduledBackupToTerraform = dataGoogleDataprocMetastoreServiceScheduledBackupToTerraform;
function dataGoogleDataprocMetastoreServiceScheduledBackupToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceScheduledBackupToHclTerraform = dataGoogleDataprocMetastoreServiceScheduledBackupToHclTerraform;
class DataGoogleDataprocMetastoreServiceScheduledBackupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // backup_location - computed: true, optional: false, required: false
    get backupLocation() {
        return this.getStringAttribute('backup_location');
    }
    // cron_schedule - computed: true, optional: false, required: false
    get cronSchedule() {
        return this.getStringAttribute('cron_schedule');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // time_zone - computed: true, optional: false, required: false
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
}
exports.DataGoogleDataprocMetastoreServiceScheduledBackupOutputReference = DataGoogleDataprocMetastoreServiceScheduledBackupOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceScheduledBackupOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceScheduledBackupOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceScheduledBackupList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceScheduledBackupOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceScheduledBackupList = DataGoogleDataprocMetastoreServiceScheduledBackupList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceScheduledBackupList[_z] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceScheduledBackupList", version: "14.13.1" };
function dataGoogleDataprocMetastoreServiceTelemetryConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDataprocMetastoreServiceTelemetryConfigToTerraform = dataGoogleDataprocMetastoreServiceTelemetryConfigToTerraform;
function dataGoogleDataprocMetastoreServiceTelemetryConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDataprocMetastoreServiceTelemetryConfigToHclTerraform = dataGoogleDataprocMetastoreServiceTelemetryConfigToHclTerraform;
class DataGoogleDataprocMetastoreServiceTelemetryConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // log_format - computed: true, optional: false, required: false
    get logFormat() {
        return this.getStringAttribute('log_format');
    }
}
exports.DataGoogleDataprocMetastoreServiceTelemetryConfigOutputReference = DataGoogleDataprocMetastoreServiceTelemetryConfigOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceTelemetryConfigOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceTelemetryConfigOutputReference", version: "14.13.1" };
class DataGoogleDataprocMetastoreServiceTelemetryConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDataprocMetastoreServiceTelemetryConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDataprocMetastoreServiceTelemetryConfigList = DataGoogleDataprocMetastoreServiceTelemetryConfigList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreServiceTelemetryConfigList[_1] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreServiceTelemetryConfigList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/dataproc_metastore_service google_dataproc_metastore_service}
*/
class DataGoogleDataprocMetastoreService extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleDataprocMetastoreService resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleDataprocMetastoreService to import
    * @param importFromId The id of the existing DataGoogleDataprocMetastoreService that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/dataproc_metastore_service#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleDataprocMetastoreService to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_dataproc_metastore_service", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/dataproc_metastore_service google_dataproc_metastore_service} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleDataprocMetastoreServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_dataproc_metastore_service',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // encryption_config - computed: true, optional: false, required: false
        this._encryptionConfig = new DataGoogleDataprocMetastoreServiceEncryptionConfigList(this, "encryption_config", false);
        // hive_metastore_config - computed: true, optional: false, required: false
        this._hiveMetastoreConfig = new DataGoogleDataprocMetastoreServiceHiveMetastoreConfigList(this, "hive_metastore_config", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // maintenance_window - computed: true, optional: false, required: false
        this._maintenanceWindow = new DataGoogleDataprocMetastoreServiceMaintenanceWindowList(this, "maintenance_window", false);
        // metadata_integration - computed: true, optional: false, required: false
        this._metadataIntegration = new DataGoogleDataprocMetastoreServiceMetadataIntegrationList(this, "metadata_integration", false);
        // network_config - computed: true, optional: false, required: false
        this._networkConfig = new DataGoogleDataprocMetastoreServiceNetworkConfigList(this, "network_config", false);
        // scaling_config - computed: true, optional: false, required: false
        this._scalingConfig = new DataGoogleDataprocMetastoreServiceScalingConfigList(this, "scaling_config", false);
        // scheduled_backup - computed: true, optional: false, required: false
        this._scheduledBackup = new DataGoogleDataprocMetastoreServiceScheduledBackupList(this, "scheduled_backup", false);
        // telemetry_config - computed: true, optional: false, required: false
        this._telemetryConfig = new DataGoogleDataprocMetastoreServiceTelemetryConfigList(this, "telemetry_config", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._id = config.id;
        this._location = config.location;
        this._project = config.project;
        this._serviceId = config.serviceId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // artifact_gcs_uri - computed: true, optional: false, required: false
    get artifactGcsUri() {
        return this.getStringAttribute('artifact_gcs_uri');
    }
    // database_type - computed: true, optional: false, required: false
    get databaseType() {
        return this.getStringAttribute('database_type');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get encryptionConfig() {
        return this._encryptionConfig;
    }
    // endpoint_uri - computed: true, optional: false, required: false
    get endpointUri() {
        return this.getStringAttribute('endpoint_uri');
    }
    get hiveMetastoreConfig() {
        return this._hiveMetastoreConfig;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    get metadataIntegration() {
        return this._metadataIntegration;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    get networkConfig() {
        return this._networkConfig;
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // release_channel - computed: true, optional: false, required: false
    get releaseChannel() {
        return this.getStringAttribute('release_channel');
    }
    get scalingConfig() {
        return this._scalingConfig;
    }
    get scheduledBackup() {
        return this._scheduledBackup;
    }
    get serviceId() {
        return this.getStringAttribute('service_id');
    }
    set serviceId(value) {
        this._serviceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceIdInput() {
        return this._serviceId;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // state_message - computed: true, optional: false, required: false
    get stateMessage() {
        return this.getStringAttribute('state_message');
    }
    get telemetryConfig() {
        return this._telemetryConfig;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // tier - computed: true, optional: false, required: false
    get tier() {
        return this.getStringAttribute('tier');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            service_id: cdktf.stringToTerraform(this._serviceId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_id: {
                value: cdktf.stringToHclTerraform(this._serviceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleDataprocMetastoreService = DataGoogleDataprocMetastoreService;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleDataprocMetastoreService[_2] = { fqn: "@cdktf/provider-google.dataGoogleDataprocMetastoreService.DataGoogleDataprocMetastoreService", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleDataprocMetastoreService.tfResourceType = "google_dataproc_metastore_service";
