// @cdktf/provider-google/lib/notebooks-runtime/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotebooksRuntime = exports.NotebooksRuntimeVirtualMachineOutputReference = exports.notebooksRuntimeVirtualMachineToHclTerraform = exports.notebooksRuntimeVirtualMachineToTerraform = exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigOutputReference = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigToHclTerraform = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigToTerraform = exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigOutputReference = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToHclTerraform = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToTerraform = exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigOutputReference = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToHclTerraform = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToTerraform = exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskOutputReference = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToHclTerraform = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToTerraform = exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsOutputReference = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToHclTerraform = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToTerraform = exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesList = exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesOutputReference = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToHclTerraform = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToTerraform = exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigOutputReference = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToHclTerraform = exports.notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToTerraform = exports.NotebooksRuntimeTimeoutsOutputReference = exports.notebooksRuntimeTimeoutsToHclTerraform = exports.notebooksRuntimeTimeoutsToTerraform = exports.NotebooksRuntimeSoftwareConfigOutputReference = exports.notebooksRuntimeSoftwareConfigToHclTerraform = exports.notebooksRuntimeSoftwareConfigToTerraform = exports.NotebooksRuntimeSoftwareConfigKernelsList = exports.NotebooksRuntimeSoftwareConfigKernelsOutputReference = exports.notebooksRuntimeSoftwareConfigKernelsToHclTerraform = exports.notebooksRuntimeSoftwareConfigKernelsToTerraform = exports.NotebooksRuntimeAccessConfigOutputReference = exports.notebooksRuntimeAccessConfigToHclTerraform = exports.notebooksRuntimeAccessConfigToTerraform = exports.NotebooksRuntimeMetricsList = exports.NotebooksRuntimeMetricsOutputReference = exports.notebooksRuntimeMetricsToHclTerraform = exports.notebooksRuntimeMetricsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function notebooksRuntimeMetricsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.notebooksRuntimeMetricsToTerraform = notebooksRuntimeMetricsToTerraform;
function notebooksRuntimeMetricsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.notebooksRuntimeMetricsToHclTerraform = notebooksRuntimeMetricsToHclTerraform;
class NotebooksRuntimeMetricsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // system_metrics - computed: true, optional: false, required: false
        this._systemMetrics = new cdktf.StringMap(this, "system_metrics");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get systemMetrics() {
        return this._systemMetrics;
    }
}
exports.NotebooksRuntimeMetricsOutputReference = NotebooksRuntimeMetricsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeMetricsOutputReference[_a] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeMetricsOutputReference", version: "14.13.1" };
class NotebooksRuntimeMetricsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NotebooksRuntimeMetricsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NotebooksRuntimeMetricsList = NotebooksRuntimeMetricsList;
_b = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeMetricsList[_b] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeMetricsList", version: "14.13.1" };
function notebooksRuntimeAccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access_type: cdktf.stringToTerraform(struct.accessType),
        runtime_owner: cdktf.stringToTerraform(struct.runtimeOwner),
    };
}
exports.notebooksRuntimeAccessConfigToTerraform = notebooksRuntimeAccessConfigToTerraform;
function notebooksRuntimeAccessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        access_type: {
            value: cdktf.stringToHclTerraform(struct.accessType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        runtime_owner: {
            value: cdktf.stringToHclTerraform(struct.runtimeOwner),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeAccessConfigToHclTerraform = notebooksRuntimeAccessConfigToHclTerraform;
class NotebooksRuntimeAccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accessType !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessType = this._accessType;
        }
        if (this._runtimeOwner !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtimeOwner = this._runtimeOwner;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accessType = undefined;
            this._runtimeOwner = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accessType = value.accessType;
            this._runtimeOwner = value.runtimeOwner;
        }
    }
    get accessType() {
        return this.getStringAttribute('access_type');
    }
    set accessType(value) {
        this._accessType = value;
    }
    resetAccessType() {
        this._accessType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessTypeInput() {
        return this._accessType;
    }
    // proxy_uri - computed: true, optional: false, required: false
    get proxyUri() {
        return this.getStringAttribute('proxy_uri');
    }
    get runtimeOwner() {
        return this.getStringAttribute('runtime_owner');
    }
    set runtimeOwner(value) {
        this._runtimeOwner = value;
    }
    resetRuntimeOwner() {
        this._runtimeOwner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeOwnerInput() {
        return this._runtimeOwner;
    }
}
exports.NotebooksRuntimeAccessConfigOutputReference = NotebooksRuntimeAccessConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeAccessConfigOutputReference[_c] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeAccessConfigOutputReference", version: "14.13.1" };
function notebooksRuntimeSoftwareConfigKernelsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository: cdktf.stringToTerraform(struct.repository),
        tag: cdktf.stringToTerraform(struct.tag),
    };
}
exports.notebooksRuntimeSoftwareConfigKernelsToTerraform = notebooksRuntimeSoftwareConfigKernelsToTerraform;
function notebooksRuntimeSoftwareConfigKernelsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repository: {
            value: cdktf.stringToHclTerraform(struct.repository),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tag: {
            value: cdktf.stringToHclTerraform(struct.tag),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeSoftwareConfigKernelsToHclTerraform = notebooksRuntimeSoftwareConfigKernelsToHclTerraform;
class NotebooksRuntimeSoftwareConfigKernelsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repository !== undefined) {
            hasAnyValues = true;
            internalValueResult.repository = this._repository;
        }
        if (this._tag !== undefined) {
            hasAnyValues = true;
            internalValueResult.tag = this._tag;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._repository = undefined;
            this._tag = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._repository = value.repository;
            this._tag = value.tag;
        }
    }
    get repository() {
        return this.getStringAttribute('repository');
    }
    set repository(value) {
        this._repository = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryInput() {
        return this._repository;
    }
    get tag() {
        return this.getStringAttribute('tag');
    }
    set tag(value) {
        this._tag = value;
    }
    resetTag() {
        this._tag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagInput() {
        return this._tag;
    }
}
exports.NotebooksRuntimeSoftwareConfigKernelsOutputReference = NotebooksRuntimeSoftwareConfigKernelsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeSoftwareConfigKernelsOutputReference[_d] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeSoftwareConfigKernelsOutputReference", version: "14.13.1" };
class NotebooksRuntimeSoftwareConfigKernelsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NotebooksRuntimeSoftwareConfigKernelsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NotebooksRuntimeSoftwareConfigKernelsList = NotebooksRuntimeSoftwareConfigKernelsList;
_e = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeSoftwareConfigKernelsList[_e] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeSoftwareConfigKernelsList", version: "14.13.1" };
function notebooksRuntimeSoftwareConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        custom_gpu_driver_path: cdktf.stringToTerraform(struct.customGpuDriverPath),
        enable_health_monitoring: cdktf.booleanToTerraform(struct.enableHealthMonitoring),
        idle_shutdown: cdktf.booleanToTerraform(struct.idleShutdown),
        idle_shutdown_timeout: cdktf.numberToTerraform(struct.idleShutdownTimeout),
        install_gpu_driver: cdktf.booleanToTerraform(struct.installGpuDriver),
        notebook_upgrade_schedule: cdktf.stringToTerraform(struct.notebookUpgradeSchedule),
        post_startup_script: cdktf.stringToTerraform(struct.postStartupScript),
        post_startup_script_behavior: cdktf.stringToTerraform(struct.postStartupScriptBehavior),
        kernels: cdktf.listMapper(notebooksRuntimeSoftwareConfigKernelsToTerraform, true)(struct.kernels),
    };
}
exports.notebooksRuntimeSoftwareConfigToTerraform = notebooksRuntimeSoftwareConfigToTerraform;
function notebooksRuntimeSoftwareConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        custom_gpu_driver_path: {
            value: cdktf.stringToHclTerraform(struct.customGpuDriverPath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enable_health_monitoring: {
            value: cdktf.booleanToHclTerraform(struct.enableHealthMonitoring),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        idle_shutdown: {
            value: cdktf.booleanToHclTerraform(struct.idleShutdown),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        idle_shutdown_timeout: {
            value: cdktf.numberToHclTerraform(struct.idleShutdownTimeout),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        install_gpu_driver: {
            value: cdktf.booleanToHclTerraform(struct.installGpuDriver),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        notebook_upgrade_schedule: {
            value: cdktf.stringToHclTerraform(struct.notebookUpgradeSchedule),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        post_startup_script: {
            value: cdktf.stringToHclTerraform(struct.postStartupScript),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        post_startup_script_behavior: {
            value: cdktf.stringToHclTerraform(struct.postStartupScriptBehavior),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kernels: {
            value: cdktf.listMapperHcl(notebooksRuntimeSoftwareConfigKernelsToHclTerraform, true)(struct.kernels),
            isBlock: true,
            type: "list",
            storageClassType: "NotebooksRuntimeSoftwareConfigKernelsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeSoftwareConfigToHclTerraform = notebooksRuntimeSoftwareConfigToHclTerraform;
class NotebooksRuntimeSoftwareConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // kernels - computed: false, optional: true, required: false
        this._kernels = new NotebooksRuntimeSoftwareConfigKernelsList(this, "kernels", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customGpuDriverPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.customGpuDriverPath = this._customGpuDriverPath;
        }
        if (this._enableHealthMonitoring !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableHealthMonitoring = this._enableHealthMonitoring;
        }
        if (this._idleShutdown !== undefined) {
            hasAnyValues = true;
            internalValueResult.idleShutdown = this._idleShutdown;
        }
        if (this._idleShutdownTimeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.idleShutdownTimeout = this._idleShutdownTimeout;
        }
        if (this._installGpuDriver !== undefined) {
            hasAnyValues = true;
            internalValueResult.installGpuDriver = this._installGpuDriver;
        }
        if (this._notebookUpgradeSchedule !== undefined) {
            hasAnyValues = true;
            internalValueResult.notebookUpgradeSchedule = this._notebookUpgradeSchedule;
        }
        if (this._postStartupScript !== undefined) {
            hasAnyValues = true;
            internalValueResult.postStartupScript = this._postStartupScript;
        }
        if (this._postStartupScriptBehavior !== undefined) {
            hasAnyValues = true;
            internalValueResult.postStartupScriptBehavior = this._postStartupScriptBehavior;
        }
        if (this._kernels?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kernels = this._kernels?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._customGpuDriverPath = undefined;
            this._enableHealthMonitoring = undefined;
            this._idleShutdown = undefined;
            this._idleShutdownTimeout = undefined;
            this._installGpuDriver = undefined;
            this._notebookUpgradeSchedule = undefined;
            this._postStartupScript = undefined;
            this._postStartupScriptBehavior = undefined;
            this._kernels.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._customGpuDriverPath = value.customGpuDriverPath;
            this._enableHealthMonitoring = value.enableHealthMonitoring;
            this._idleShutdown = value.idleShutdown;
            this._idleShutdownTimeout = value.idleShutdownTimeout;
            this._installGpuDriver = value.installGpuDriver;
            this._notebookUpgradeSchedule = value.notebookUpgradeSchedule;
            this._postStartupScript = value.postStartupScript;
            this._postStartupScriptBehavior = value.postStartupScriptBehavior;
            this._kernels.internalValue = value.kernels;
        }
    }
    get customGpuDriverPath() {
        return this.getStringAttribute('custom_gpu_driver_path');
    }
    set customGpuDriverPath(value) {
        this._customGpuDriverPath = value;
    }
    resetCustomGpuDriverPath() {
        this._customGpuDriverPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customGpuDriverPathInput() {
        return this._customGpuDriverPath;
    }
    get enableHealthMonitoring() {
        return this.getBooleanAttribute('enable_health_monitoring');
    }
    set enableHealthMonitoring(value) {
        this._enableHealthMonitoring = value;
    }
    resetEnableHealthMonitoring() {
        this._enableHealthMonitoring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableHealthMonitoringInput() {
        return this._enableHealthMonitoring;
    }
    get idleShutdown() {
        return this.getBooleanAttribute('idle_shutdown');
    }
    set idleShutdown(value) {
        this._idleShutdown = value;
    }
    resetIdleShutdown() {
        this._idleShutdown = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleShutdownInput() {
        return this._idleShutdown;
    }
    get idleShutdownTimeout() {
        return this.getNumberAttribute('idle_shutdown_timeout');
    }
    set idleShutdownTimeout(value) {
        this._idleShutdownTimeout = value;
    }
    resetIdleShutdownTimeout() {
        this._idleShutdownTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idleShutdownTimeoutInput() {
        return this._idleShutdownTimeout;
    }
    get installGpuDriver() {
        return this.getBooleanAttribute('install_gpu_driver');
    }
    set installGpuDriver(value) {
        this._installGpuDriver = value;
    }
    resetInstallGpuDriver() {
        this._installGpuDriver = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get installGpuDriverInput() {
        return this._installGpuDriver;
    }
    get notebookUpgradeSchedule() {
        return this.getStringAttribute('notebook_upgrade_schedule');
    }
    set notebookUpgradeSchedule(value) {
        this._notebookUpgradeSchedule = value;
    }
    resetNotebookUpgradeSchedule() {
        this._notebookUpgradeSchedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notebookUpgradeScheduleInput() {
        return this._notebookUpgradeSchedule;
    }
    get postStartupScript() {
        return this.getStringAttribute('post_startup_script');
    }
    set postStartupScript(value) {
        this._postStartupScript = value;
    }
    resetPostStartupScript() {
        this._postStartupScript = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postStartupScriptInput() {
        return this._postStartupScript;
    }
    get postStartupScriptBehavior() {
        return this.getStringAttribute('post_startup_script_behavior');
    }
    set postStartupScriptBehavior(value) {
        this._postStartupScriptBehavior = value;
    }
    resetPostStartupScriptBehavior() {
        this._postStartupScriptBehavior = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postStartupScriptBehaviorInput() {
        return this._postStartupScriptBehavior;
    }
    // upgradeable - computed: true, optional: false, required: false
    get upgradeable() {
        return this.getBooleanAttribute('upgradeable');
    }
    get kernels() {
        return this._kernels;
    }
    putKernels(value) {
        this._kernels.internalValue = value;
    }
    resetKernels() {
        this._kernels.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kernelsInput() {
        return this._kernels.internalValue;
    }
}
exports.NotebooksRuntimeSoftwareConfigOutputReference = NotebooksRuntimeSoftwareConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeSoftwareConfigOutputReference[_f] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeSoftwareConfigOutputReference", version: "14.13.1" };
function notebooksRuntimeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.notebooksRuntimeTimeoutsToTerraform = notebooksRuntimeTimeoutsToTerraform;
function notebooksRuntimeTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeTimeoutsToHclTerraform = notebooksRuntimeTimeoutsToHclTerraform;
class NotebooksRuntimeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NotebooksRuntimeTimeoutsOutputReference = NotebooksRuntimeTimeoutsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeTimeoutsOutputReference[_g] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeTimeoutsOutputReference", version: "14.13.1" };
function notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        core_count: cdktf.numberToTerraform(struct.coreCount),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToTerraform;
function notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        core_count: {
            value: cdktf.numberToHclTerraform(struct.coreCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToHclTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToHclTerraform;
class NotebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coreCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.coreCount = this._coreCount;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._coreCount = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._coreCount = value.coreCount;
            this._type = value.type;
        }
    }
    get coreCount() {
        return this.getNumberAttribute('core_count');
    }
    set coreCount(value) {
        this._coreCount = value;
    }
    resetCoreCount() {
        this._coreCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreCountInput() {
        return this._coreCount;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigOutputReference = NotebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigOutputReference[_h] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigOutputReference", version: "14.13.1" };
function notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository: cdktf.stringToTerraform(struct.repository),
        tag: cdktf.stringToTerraform(struct.tag),
    };
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToTerraform;
function notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repository: {
            value: cdktf.stringToHclTerraform(struct.repository),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tag: {
            value: cdktf.stringToHclTerraform(struct.tag),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToHclTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToHclTerraform;
class NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repository !== undefined) {
            hasAnyValues = true;
            internalValueResult.repository = this._repository;
        }
        if (this._tag !== undefined) {
            hasAnyValues = true;
            internalValueResult.tag = this._tag;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._repository = undefined;
            this._tag = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._repository = value.repository;
            this._tag = value.tag;
        }
    }
    get repository() {
        return this.getStringAttribute('repository');
    }
    set repository(value) {
        this._repository = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryInput() {
        return this._repository;
    }
    get tag() {
        return this.getStringAttribute('tag');
    }
    set tag(value) {
        this._tag = value;
    }
    resetTag() {
        this._tag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagInput() {
        return this._tag;
    }
}
exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesOutputReference = NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesOutputReference;
_j = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesOutputReference[_j] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesOutputReference", version: "14.13.1" };
class NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesList = NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesList;
_k = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesList[_k] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesList", version: "14.13.1" };
function notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        disk_name: cdktf.stringToTerraform(struct.diskName),
        disk_size_gb: cdktf.numberToTerraform(struct.diskSizeGb),
        disk_type: cdktf.stringToTerraform(struct.diskType),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
    };
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToTerraform;
function notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        disk_name: {
            value: cdktf.stringToHclTerraform(struct.diskName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        disk_size_gb: {
            value: cdktf.numberToHclTerraform(struct.diskSizeGb),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        disk_type: {
            value: cdktf.stringToHclTerraform(struct.diskType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        labels: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.labels),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToHclTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToHclTerraform;
class NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._diskName !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskName = this._diskName;
        }
        if (this._diskSizeGb !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSizeGb = this._diskSizeGb;
        }
        if (this._diskType !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskType = this._diskType;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._diskName = undefined;
            this._diskSizeGb = undefined;
            this._diskType = undefined;
            this._labels = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._diskName = value.diskName;
            this._diskSizeGb = value.diskSizeGb;
            this._diskType = value.diskType;
            this._labels = value.labels;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get diskName() {
        return this.getStringAttribute('disk_name');
    }
    set diskName(value) {
        this._diskName = value;
    }
    resetDiskName() {
        this._diskName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskNameInput() {
        return this._diskName;
    }
    get diskSizeGb() {
        return this.getNumberAttribute('disk_size_gb');
    }
    set diskSizeGb(value) {
        this._diskSizeGb = value;
    }
    resetDiskSizeGb() {
        this._diskSizeGb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeGbInput() {
        return this._diskSizeGb;
    }
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    set diskType(value) {
        this._diskType = value;
    }
    resetDiskType() {
        this._diskType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeInput() {
        return this._diskType;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
}
exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsOutputReference = NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsOutputReference[_l] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsOutputReference", version: "14.13.1" };
function notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        interface: cdktf.stringToTerraform(struct.interface),
        mode: cdktf.stringToTerraform(struct.mode),
        source: cdktf.stringToTerraform(struct.source),
        type: cdktf.stringToTerraform(struct.type),
        initialize_params: notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToTerraform(struct.initializeParams),
    };
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToTerraform;
function notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        interface: {
            value: cdktf.stringToHclTerraform(struct.interface),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source: {
            value: cdktf.stringToHclTerraform(struct.source),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        initialize_params: {
            value: notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsToHclTerraform(struct.initializeParams),
            isBlock: true,
            type: "list",
            storageClassType: "NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToHclTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToHclTerraform;
class NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // initialize_params - computed: false, optional: true, required: false
        this._initializeParams = new NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskInitializeParamsOutputReference(this, "initialize_params");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._interface !== undefined) {
            hasAnyValues = true;
            internalValueResult.interface = this._interface;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._source !== undefined) {
            hasAnyValues = true;
            internalValueResult.source = this._source;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._initializeParams?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.initializeParams = this._initializeParams?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._interface = undefined;
            this._mode = undefined;
            this._source = undefined;
            this._type = undefined;
            this._initializeParams.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._interface = value.interface;
            this._mode = value.mode;
            this._source = value.source;
            this._type = value.type;
            this._initializeParams.internalValue = value.initializeParams;
        }
    }
    // auto_delete - computed: true, optional: false, required: false
    get autoDelete() {
        return this.getBooleanAttribute('auto_delete');
    }
    // boot - computed: true, optional: false, required: false
    get boot() {
        return this.getBooleanAttribute('boot');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // guest_os_features - computed: true, optional: false, required: false
    get guestOsFeatures() {
        return this.getListAttribute('guest_os_features');
    }
    // index - computed: true, optional: false, required: false
    get index() {
        return this.getNumberAttribute('index');
    }
    get interface() {
        return this.getStringAttribute('interface');
    }
    set interface(value) {
        this._interface = value;
    }
    resetInterface() {
        this._interface = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get interfaceInput() {
        return this._interface;
    }
    // kind - computed: true, optional: false, required: false
    get kind() {
        return this.getStringAttribute('kind');
    }
    // licenses - computed: true, optional: false, required: false
    get licenses() {
        return this.getListAttribute('licenses');
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get source() {
        return this.getStringAttribute('source');
    }
    set source(value) {
        this._source = value;
    }
    resetSource() {
        this._source = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get initializeParams() {
        return this._initializeParams;
    }
    putInitializeParams(value) {
        this._initializeParams.internalValue = value;
    }
    resetInitializeParams() {
        this._initializeParams.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initializeParamsInput() {
        return this._initializeParams.internalValue;
    }
}
exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskOutputReference = NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskOutputReference;
_m = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskOutputReference[_m] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskOutputReference", version: "14.13.1" };
function notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key: cdktf.stringToTerraform(struct.kmsKey),
    };
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToTerraform;
function notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key: {
            value: cdktf.stringToHclTerraform(struct.kmsKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToHclTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToHclTerraform;
class NotebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKey = this._kmsKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKey = value.kmsKey;
        }
    }
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    set kmsKey(value) {
        this._kmsKey = value;
    }
    resetKmsKey() {
        this._kmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyInput() {
        return this._kmsKey;
    }
}
exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigOutputReference = NotebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigOutputReference[_o] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigOutputReference", version: "14.13.1" };
function notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_integrity_monitoring: cdktf.booleanToTerraform(struct.enableIntegrityMonitoring),
        enable_secure_boot: cdktf.booleanToTerraform(struct.enableSecureBoot),
        enable_vtpm: cdktf.booleanToTerraform(struct.enableVtpm),
    };
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToTerraform;
function notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_integrity_monitoring: {
            value: cdktf.booleanToHclTerraform(struct.enableIntegrityMonitoring),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_secure_boot: {
            value: cdktf.booleanToHclTerraform(struct.enableSecureBoot),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_vtpm: {
            value: cdktf.booleanToHclTerraform(struct.enableVtpm),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToHclTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToHclTerraform;
class NotebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableIntegrityMonitoring !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableIntegrityMonitoring = this._enableIntegrityMonitoring;
        }
        if (this._enableSecureBoot !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableSecureBoot = this._enableSecureBoot;
        }
        if (this._enableVtpm !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableVtpm = this._enableVtpm;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableIntegrityMonitoring = undefined;
            this._enableSecureBoot = undefined;
            this._enableVtpm = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableIntegrityMonitoring = value.enableIntegrityMonitoring;
            this._enableSecureBoot = value.enableSecureBoot;
            this._enableVtpm = value.enableVtpm;
        }
    }
    get enableIntegrityMonitoring() {
        return this.getBooleanAttribute('enable_integrity_monitoring');
    }
    set enableIntegrityMonitoring(value) {
        this._enableIntegrityMonitoring = value;
    }
    resetEnableIntegrityMonitoring() {
        this._enableIntegrityMonitoring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableIntegrityMonitoringInput() {
        return this._enableIntegrityMonitoring;
    }
    get enableSecureBoot() {
        return this.getBooleanAttribute('enable_secure_boot');
    }
    set enableSecureBoot(value) {
        this._enableSecureBoot = value;
    }
    resetEnableSecureBoot() {
        this._enableSecureBoot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableSecureBootInput() {
        return this._enableSecureBoot;
    }
    get enableVtpm() {
        return this.getBooleanAttribute('enable_vtpm');
    }
    set enableVtpm(value) {
        this._enableVtpm = value;
    }
    resetEnableVtpm() {
        this._enableVtpm = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableVtpmInput() {
        return this._enableVtpm;
    }
}
exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigOutputReference = NotebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigOutputReference;
_p = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigOutputReference[_p] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigOutputReference", version: "14.13.1" };
function notebooksRuntimeVirtualMachineVirtualMachineConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        internal_ip_only: cdktf.booleanToTerraform(struct.internalIpOnly),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        machine_type: cdktf.stringToTerraform(struct.machineType),
        metadata: cdktf.hashMapper(cdktf.stringToTerraform)(struct.metadata),
        network: cdktf.stringToTerraform(struct.network),
        nic_type: cdktf.stringToTerraform(struct.nicType),
        reserved_ip_range: cdktf.stringToTerraform(struct.reservedIpRange),
        subnet: cdktf.stringToTerraform(struct.subnet),
        tags: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tags),
        accelerator_config: notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToTerraform(struct.acceleratorConfig),
        container_images: cdktf.listMapper(notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToTerraform, true)(struct.containerImages),
        data_disk: notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToTerraform(struct.dataDisk),
        encryption_config: notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToTerraform(struct.encryptionConfig),
        shielded_instance_config: notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToTerraform(struct.shieldedInstanceConfig),
    };
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigToTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigToTerraform;
function notebooksRuntimeVirtualMachineVirtualMachineConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        internal_ip_only: {
            value: cdktf.booleanToHclTerraform(struct.internalIpOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        labels: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.labels),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        machine_type: {
            value: cdktf.stringToHclTerraform(struct.machineType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        metadata: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.metadata),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        network: {
            value: cdktf.stringToHclTerraform(struct.network),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        nic_type: {
            value: cdktf.stringToHclTerraform(struct.nicType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        reserved_ip_range: {
            value: cdktf.stringToHclTerraform(struct.reservedIpRange),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subnet: {
            value: cdktf.stringToHclTerraform(struct.subnet),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tags: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.tags),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        accelerator_config: {
            value: notebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigToHclTerraform(struct.acceleratorConfig),
            isBlock: true,
            type: "list",
            storageClassType: "NotebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigList",
        },
        container_images: {
            value: cdktf.listMapperHcl(notebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesToHclTerraform, true)(struct.containerImages),
            isBlock: true,
            type: "list",
            storageClassType: "NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesList",
        },
        data_disk: {
            value: notebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskToHclTerraform(struct.dataDisk),
            isBlock: true,
            type: "list",
            storageClassType: "NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskList",
        },
        encryption_config: {
            value: notebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigToHclTerraform(struct.encryptionConfig),
            isBlock: true,
            type: "list",
            storageClassType: "NotebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigList",
        },
        shielded_instance_config: {
            value: notebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigToHclTerraform(struct.shieldedInstanceConfig),
            isBlock: true,
            type: "list",
            storageClassType: "NotebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeVirtualMachineVirtualMachineConfigToHclTerraform = notebooksRuntimeVirtualMachineVirtualMachineConfigToHclTerraform;
class NotebooksRuntimeVirtualMachineVirtualMachineConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // guest_attributes - computed: true, optional: false, required: false
        this._guestAttributes = new cdktf.StringMap(this, "guest_attributes");
        // accelerator_config - computed: false, optional: true, required: false
        this._acceleratorConfig = new NotebooksRuntimeVirtualMachineVirtualMachineConfigAcceleratorConfigOutputReference(this, "accelerator_config");
        // container_images - computed: false, optional: true, required: false
        this._containerImages = new NotebooksRuntimeVirtualMachineVirtualMachineConfigContainerImagesList(this, "container_images", false);
        // data_disk - computed: false, optional: false, required: true
        this._dataDisk = new NotebooksRuntimeVirtualMachineVirtualMachineConfigDataDiskOutputReference(this, "data_disk");
        // encryption_config - computed: false, optional: true, required: false
        this._encryptionConfig = new NotebooksRuntimeVirtualMachineVirtualMachineConfigEncryptionConfigOutputReference(this, "encryption_config");
        // shielded_instance_config - computed: false, optional: true, required: false
        this._shieldedInstanceConfig = new NotebooksRuntimeVirtualMachineVirtualMachineConfigShieldedInstanceConfigOutputReference(this, "shielded_instance_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._internalIpOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.internalIpOnly = this._internalIpOnly;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._machineType !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineType = this._machineType;
        }
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        if (this._network !== undefined) {
            hasAnyValues = true;
            internalValueResult.network = this._network;
        }
        if (this._nicType !== undefined) {
            hasAnyValues = true;
            internalValueResult.nicType = this._nicType;
        }
        if (this._reservedIpRange !== undefined) {
            hasAnyValues = true;
            internalValueResult.reservedIpRange = this._reservedIpRange;
        }
        if (this._subnet !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnet = this._subnet;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        if (this._acceleratorConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.acceleratorConfig = this._acceleratorConfig?.internalValue;
        }
        if (this._containerImages?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerImages = this._containerImages?.internalValue;
        }
        if (this._dataDisk?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataDisk = this._dataDisk?.internalValue;
        }
        if (this._encryptionConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionConfig = this._encryptionConfig?.internalValue;
        }
        if (this._shieldedInstanceConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.shieldedInstanceConfig = this._shieldedInstanceConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._internalIpOnly = undefined;
            this._labels = undefined;
            this._machineType = undefined;
            this._metadata = undefined;
            this._network = undefined;
            this._nicType = undefined;
            this._reservedIpRange = undefined;
            this._subnet = undefined;
            this._tags = undefined;
            this._acceleratorConfig.internalValue = undefined;
            this._containerImages.internalValue = undefined;
            this._dataDisk.internalValue = undefined;
            this._encryptionConfig.internalValue = undefined;
            this._shieldedInstanceConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._internalIpOnly = value.internalIpOnly;
            this._labels = value.labels;
            this._machineType = value.machineType;
            this._metadata = value.metadata;
            this._network = value.network;
            this._nicType = value.nicType;
            this._reservedIpRange = value.reservedIpRange;
            this._subnet = value.subnet;
            this._tags = value.tags;
            this._acceleratorConfig.internalValue = value.acceleratorConfig;
            this._containerImages.internalValue = value.containerImages;
            this._dataDisk.internalValue = value.dataDisk;
            this._encryptionConfig.internalValue = value.encryptionConfig;
            this._shieldedInstanceConfig.internalValue = value.shieldedInstanceConfig;
        }
    }
    get guestAttributes() {
        return this._guestAttributes;
    }
    get internalIpOnly() {
        return this.getBooleanAttribute('internal_ip_only');
    }
    set internalIpOnly(value) {
        this._internalIpOnly = value;
    }
    resetInternalIpOnly() {
        this._internalIpOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get internalIpOnlyInput() {
        return this._internalIpOnly;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    set machineType(value) {
        this._machineType = value;
    }
    // Temporarily expose input value. Use with caution.
    get machineTypeInput() {
        return this._machineType;
    }
    get metadata() {
        return this.getStringMapAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    resetNetwork() {
        this._network = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get nicType() {
        return this.getStringAttribute('nic_type');
    }
    set nicType(value) {
        this._nicType = value;
    }
    resetNicType() {
        this._nicType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nicTypeInput() {
        return this._nicType;
    }
    get reservedIpRange() {
        return this.getStringAttribute('reserved_ip_range');
    }
    set reservedIpRange(value) {
        this._reservedIpRange = value;
    }
    resetReservedIpRange() {
        this._reservedIpRange = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reservedIpRangeInput() {
        return this._reservedIpRange;
    }
    get subnet() {
        return this.getStringAttribute('subnet');
    }
    set subnet(value) {
        this._subnet = value;
    }
    resetSubnet() {
        this._subnet = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetInput() {
        return this._subnet;
    }
    get tags() {
        return this.getListAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
    get acceleratorConfig() {
        return this._acceleratorConfig;
    }
    putAcceleratorConfig(value) {
        this._acceleratorConfig.internalValue = value;
    }
    resetAcceleratorConfig() {
        this._acceleratorConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acceleratorConfigInput() {
        return this._acceleratorConfig.internalValue;
    }
    get containerImages() {
        return this._containerImages;
    }
    putContainerImages(value) {
        this._containerImages.internalValue = value;
    }
    resetContainerImages() {
        this._containerImages.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerImagesInput() {
        return this._containerImages.internalValue;
    }
    get dataDisk() {
        return this._dataDisk;
    }
    putDataDisk(value) {
        this._dataDisk.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataDiskInput() {
        return this._dataDisk.internalValue;
    }
    get encryptionConfig() {
        return this._encryptionConfig;
    }
    putEncryptionConfig(value) {
        this._encryptionConfig.internalValue = value;
    }
    resetEncryptionConfig() {
        this._encryptionConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionConfigInput() {
        return this._encryptionConfig.internalValue;
    }
    get shieldedInstanceConfig() {
        return this._shieldedInstanceConfig;
    }
    putShieldedInstanceConfig(value) {
        this._shieldedInstanceConfig.internalValue = value;
    }
    resetShieldedInstanceConfig() {
        this._shieldedInstanceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shieldedInstanceConfigInput() {
        return this._shieldedInstanceConfig.internalValue;
    }
}
exports.NotebooksRuntimeVirtualMachineVirtualMachineConfigOutputReference = NotebooksRuntimeVirtualMachineVirtualMachineConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineVirtualMachineConfigOutputReference[_q] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineVirtualMachineConfigOutputReference", version: "14.13.1" };
function notebooksRuntimeVirtualMachineToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        virtual_machine_config: notebooksRuntimeVirtualMachineVirtualMachineConfigToTerraform(struct.virtualMachineConfig),
    };
}
exports.notebooksRuntimeVirtualMachineToTerraform = notebooksRuntimeVirtualMachineToTerraform;
function notebooksRuntimeVirtualMachineToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        virtual_machine_config: {
            value: notebooksRuntimeVirtualMachineVirtualMachineConfigToHclTerraform(struct.virtualMachineConfig),
            isBlock: true,
            type: "list",
            storageClassType: "NotebooksRuntimeVirtualMachineVirtualMachineConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.notebooksRuntimeVirtualMachineToHclTerraform = notebooksRuntimeVirtualMachineToHclTerraform;
class NotebooksRuntimeVirtualMachineOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // virtual_machine_config - computed: false, optional: true, required: false
        this._virtualMachineConfig = new NotebooksRuntimeVirtualMachineVirtualMachineConfigOutputReference(this, "virtual_machine_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._virtualMachineConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualMachineConfig = this._virtualMachineConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._virtualMachineConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._virtualMachineConfig.internalValue = value.virtualMachineConfig;
        }
    }
    // instance_id - computed: true, optional: false, required: false
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    // instance_name - computed: true, optional: false, required: false
    get instanceName() {
        return this.getStringAttribute('instance_name');
    }
    get virtualMachineConfig() {
        return this._virtualMachineConfig;
    }
    putVirtualMachineConfig(value) {
        this._virtualMachineConfig.internalValue = value;
    }
    resetVirtualMachineConfig() {
        this._virtualMachineConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get virtualMachineConfigInput() {
        return this._virtualMachineConfig.internalValue;
    }
}
exports.NotebooksRuntimeVirtualMachineOutputReference = NotebooksRuntimeVirtualMachineOutputReference;
_r = JSII_RTTI_SYMBOL_1;
NotebooksRuntimeVirtualMachineOutputReference[_r] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntimeVirtualMachineOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/notebooks_runtime google_notebooks_runtime}
*/
class NotebooksRuntime extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NotebooksRuntime resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NotebooksRuntime to import
    * @param importFromId The id of the existing NotebooksRuntime that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/notebooks_runtime#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NotebooksRuntime to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_notebooks_runtime", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/notebooks_runtime google_notebooks_runtime} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NotebooksRuntimeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_notebooks_runtime',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // metrics - computed: true, optional: false, required: false
        this._metrics = new NotebooksRuntimeMetricsList(this, "metrics", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // access_config - computed: false, optional: true, required: false
        this._accessConfig = new NotebooksRuntimeAccessConfigOutputReference(this, "access_config");
        // software_config - computed: false, optional: true, required: false
        this._softwareConfig = new NotebooksRuntimeSoftwareConfigOutputReference(this, "software_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NotebooksRuntimeTimeoutsOutputReference(this, "timeouts");
        // virtual_machine - computed: false, optional: true, required: false
        this._virtualMachine = new NotebooksRuntimeVirtualMachineOutputReference(this, "virtual_machine");
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._accessConfig.internalValue = config.accessConfig;
        this._softwareConfig.internalValue = config.softwareConfig;
        this._timeouts.internalValue = config.timeouts;
        this._virtualMachine.internalValue = config.virtualMachine;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // health_state - computed: true, optional: false, required: false
    get healthState() {
        return this.getStringAttribute('health_state');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get metrics() {
        return this._metrics;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    get accessConfig() {
        return this._accessConfig;
    }
    putAccessConfig(value) {
        this._accessConfig.internalValue = value;
    }
    resetAccessConfig() {
        this._accessConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessConfigInput() {
        return this._accessConfig.internalValue;
    }
    get softwareConfig() {
        return this._softwareConfig;
    }
    putSoftwareConfig(value) {
        this._softwareConfig.internalValue = value;
    }
    resetSoftwareConfig() {
        this._softwareConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get softwareConfigInput() {
        return this._softwareConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get virtualMachine() {
        return this._virtualMachine;
    }
    putVirtualMachine(value) {
        this._virtualMachine.internalValue = value;
    }
    resetVirtualMachine() {
        this._virtualMachine.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get virtualMachineInput() {
        return this._virtualMachine.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            access_config: notebooksRuntimeAccessConfigToTerraform(this._accessConfig.internalValue),
            software_config: notebooksRuntimeSoftwareConfigToTerraform(this._softwareConfig.internalValue),
            timeouts: notebooksRuntimeTimeoutsToTerraform(this._timeouts.internalValue),
            virtual_machine: notebooksRuntimeVirtualMachineToTerraform(this._virtualMachine.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            access_config: {
                value: notebooksRuntimeAccessConfigToHclTerraform(this._accessConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NotebooksRuntimeAccessConfigList",
            },
            software_config: {
                value: notebooksRuntimeSoftwareConfigToHclTerraform(this._softwareConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NotebooksRuntimeSoftwareConfigList",
            },
            timeouts: {
                value: notebooksRuntimeTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NotebooksRuntimeTimeouts",
            },
            virtual_machine: {
                value: notebooksRuntimeVirtualMachineToHclTerraform(this._virtualMachine.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NotebooksRuntimeVirtualMachineList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NotebooksRuntime = NotebooksRuntime;
_s = JSII_RTTI_SYMBOL_1;
NotebooksRuntime[_s] = { fqn: "@cdktf/provider-google.notebooksRuntime.NotebooksRuntime", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NotebooksRuntime.tfResourceType = "google_notebooks_runtime";
