// @cdktf/provider-google/lib/redis-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9;
Object.defineProperty(exports, "__esModule", { value: true });
exports.redisClusterMaintenancePolicyWeeklyMaintenanceWindowToTerraform = exports.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeOutputReference = exports.redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToHclTerraform = exports.redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToTerraform = exports.RedisClusterCrossClusterReplicationConfigOutputReference = exports.redisClusterCrossClusterReplicationConfigToHclTerraform = exports.redisClusterCrossClusterReplicationConfigToTerraform = exports.RedisClusterCrossClusterReplicationConfigSecondaryClustersList = exports.RedisClusterCrossClusterReplicationConfigSecondaryClustersOutputReference = exports.redisClusterCrossClusterReplicationConfigSecondaryClustersToHclTerraform = exports.redisClusterCrossClusterReplicationConfigSecondaryClustersToTerraform = exports.RedisClusterCrossClusterReplicationConfigPrimaryClusterOutputReference = exports.redisClusterCrossClusterReplicationConfigPrimaryClusterToHclTerraform = exports.redisClusterCrossClusterReplicationConfigPrimaryClusterToTerraform = exports.RedisClusterCrossClusterReplicationConfigMembershipList = exports.RedisClusterCrossClusterReplicationConfigMembershipOutputReference = exports.redisClusterCrossClusterReplicationConfigMembershipToHclTerraform = exports.redisClusterCrossClusterReplicationConfigMembershipToTerraform = exports.RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersList = exports.RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersOutputReference = exports.redisClusterCrossClusterReplicationConfigMembershipSecondaryClustersToHclTerraform = exports.redisClusterCrossClusterReplicationConfigMembershipSecondaryClustersToTerraform = exports.RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterList = exports.RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterOutputReference = exports.redisClusterCrossClusterReplicationConfigMembershipPrimaryClusterToHclTerraform = exports.redisClusterCrossClusterReplicationConfigMembershipPrimaryClusterToTerraform = exports.RedisClusterStateInfoList = exports.RedisClusterStateInfoOutputReference = exports.redisClusterStateInfoToHclTerraform = exports.redisClusterStateInfoToTerraform = exports.RedisClusterStateInfoUpdateInfoList = exports.RedisClusterStateInfoUpdateInfoOutputReference = exports.redisClusterStateInfoUpdateInfoToHclTerraform = exports.redisClusterStateInfoUpdateInfoToTerraform = exports.RedisClusterPscConnectionsList = exports.RedisClusterPscConnectionsOutputReference = exports.redisClusterPscConnectionsToHclTerraform = exports.redisClusterPscConnectionsToTerraform = exports.RedisClusterMaintenanceScheduleList = exports.RedisClusterMaintenanceScheduleOutputReference = exports.redisClusterMaintenanceScheduleToHclTerraform = exports.redisClusterMaintenanceScheduleToTerraform = exports.RedisClusterDiscoveryEndpointsList = exports.RedisClusterDiscoveryEndpointsOutputReference = exports.redisClusterDiscoveryEndpointsToHclTerraform = exports.redisClusterDiscoveryEndpointsToTerraform = exports.RedisClusterDiscoveryEndpointsPscConfigList = exports.RedisClusterDiscoveryEndpointsPscConfigOutputReference = exports.redisClusterDiscoveryEndpointsPscConfigToHclTerraform = exports.redisClusterDiscoveryEndpointsPscConfigToTerraform = void 0;
exports.RedisCluster = exports.RedisClusterZoneDistributionConfigOutputReference = exports.redisClusterZoneDistributionConfigToHclTerraform = exports.redisClusterZoneDistributionConfigToTerraform = exports.RedisClusterTimeoutsOutputReference = exports.redisClusterTimeoutsToHclTerraform = exports.redisClusterTimeoutsToTerraform = exports.RedisClusterPscConfigsList = exports.RedisClusterPscConfigsOutputReference = exports.redisClusterPscConfigsToHclTerraform = exports.redisClusterPscConfigsToTerraform = exports.RedisClusterPersistenceConfigOutputReference = exports.redisClusterPersistenceConfigToHclTerraform = exports.redisClusterPersistenceConfigToTerraform = exports.RedisClusterPersistenceConfigRdbConfigOutputReference = exports.redisClusterPersistenceConfigRdbConfigToHclTerraform = exports.redisClusterPersistenceConfigRdbConfigToTerraform = exports.RedisClusterPersistenceConfigAofConfigOutputReference = exports.redisClusterPersistenceConfigAofConfigToHclTerraform = exports.redisClusterPersistenceConfigAofConfigToTerraform = exports.RedisClusterMaintenancePolicyOutputReference = exports.redisClusterMaintenancePolicyToHclTerraform = exports.redisClusterMaintenancePolicyToTerraform = exports.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowList = exports.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowOutputReference = exports.redisClusterMaintenancePolicyWeeklyMaintenanceWindowToHclTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function redisClusterDiscoveryEndpointsPscConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterDiscoveryEndpointsPscConfigToTerraform = redisClusterDiscoveryEndpointsPscConfigToTerraform;
function redisClusterDiscoveryEndpointsPscConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterDiscoveryEndpointsPscConfigToHclTerraform = redisClusterDiscoveryEndpointsPscConfigToHclTerraform;
class RedisClusterDiscoveryEndpointsPscConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
}
exports.RedisClusterDiscoveryEndpointsPscConfigOutputReference = RedisClusterDiscoveryEndpointsPscConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
RedisClusterDiscoveryEndpointsPscConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterDiscoveryEndpointsPscConfigOutputReference", version: "14.13.1" };
class RedisClusterDiscoveryEndpointsPscConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterDiscoveryEndpointsPscConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterDiscoveryEndpointsPscConfigList = RedisClusterDiscoveryEndpointsPscConfigList;
_b = JSII_RTTI_SYMBOL_1;
RedisClusterDiscoveryEndpointsPscConfigList[_b] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterDiscoveryEndpointsPscConfigList", version: "14.13.1" };
function redisClusterDiscoveryEndpointsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterDiscoveryEndpointsToTerraform = redisClusterDiscoveryEndpointsToTerraform;
function redisClusterDiscoveryEndpointsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterDiscoveryEndpointsToHclTerraform = redisClusterDiscoveryEndpointsToHclTerraform;
class RedisClusterDiscoveryEndpointsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // psc_config - computed: true, optional: false, required: false
        this._pscConfig = new RedisClusterDiscoveryEndpointsPscConfigList(this, "psc_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // address - computed: true, optional: false, required: false
    get address() {
        return this.getStringAttribute('address');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    get pscConfig() {
        return this._pscConfig;
    }
}
exports.RedisClusterDiscoveryEndpointsOutputReference = RedisClusterDiscoveryEndpointsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
RedisClusterDiscoveryEndpointsOutputReference[_c] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterDiscoveryEndpointsOutputReference", version: "14.13.1" };
class RedisClusterDiscoveryEndpointsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterDiscoveryEndpointsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterDiscoveryEndpointsList = RedisClusterDiscoveryEndpointsList;
_d = JSII_RTTI_SYMBOL_1;
RedisClusterDiscoveryEndpointsList[_d] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterDiscoveryEndpointsList", version: "14.13.1" };
function redisClusterMaintenanceScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterMaintenanceScheduleToTerraform = redisClusterMaintenanceScheduleToTerraform;
function redisClusterMaintenanceScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterMaintenanceScheduleToHclTerraform = redisClusterMaintenanceScheduleToHclTerraform;
class RedisClusterMaintenanceScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // end_time - computed: true, optional: false, required: false
    get endTime() {
        return this.getStringAttribute('end_time');
    }
    // schedule_deadline_time - computed: true, optional: false, required: false
    get scheduleDeadlineTime() {
        return this.getStringAttribute('schedule_deadline_time');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
}
exports.RedisClusterMaintenanceScheduleOutputReference = RedisClusterMaintenanceScheduleOutputReference;
_e = JSII_RTTI_SYMBOL_1;
RedisClusterMaintenanceScheduleOutputReference[_e] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterMaintenanceScheduleOutputReference", version: "14.13.1" };
class RedisClusterMaintenanceScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterMaintenanceScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterMaintenanceScheduleList = RedisClusterMaintenanceScheduleList;
_f = JSII_RTTI_SYMBOL_1;
RedisClusterMaintenanceScheduleList[_f] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterMaintenanceScheduleList", version: "14.13.1" };
function redisClusterPscConnectionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterPscConnectionsToTerraform = redisClusterPscConnectionsToTerraform;
function redisClusterPscConnectionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterPscConnectionsToHclTerraform = redisClusterPscConnectionsToHclTerraform;
class RedisClusterPscConnectionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // address - computed: true, optional: false, required: false
    get address() {
        return this.getStringAttribute('address');
    }
    // forwarding_rule - computed: true, optional: false, required: false
    get forwardingRule() {
        return this.getStringAttribute('forwarding_rule');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    // project_id - computed: true, optional: false, required: false
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    // psc_connection_id - computed: true, optional: false, required: false
    get pscConnectionId() {
        return this.getStringAttribute('psc_connection_id');
    }
}
exports.RedisClusterPscConnectionsOutputReference = RedisClusterPscConnectionsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
RedisClusterPscConnectionsOutputReference[_g] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterPscConnectionsOutputReference", version: "14.13.1" };
class RedisClusterPscConnectionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterPscConnectionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterPscConnectionsList = RedisClusterPscConnectionsList;
_h = JSII_RTTI_SYMBOL_1;
RedisClusterPscConnectionsList[_h] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterPscConnectionsList", version: "14.13.1" };
function redisClusterStateInfoUpdateInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterStateInfoUpdateInfoToTerraform = redisClusterStateInfoUpdateInfoToTerraform;
function redisClusterStateInfoUpdateInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterStateInfoUpdateInfoToHclTerraform = redisClusterStateInfoUpdateInfoToHclTerraform;
class RedisClusterStateInfoUpdateInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // target_replica_count - computed: true, optional: false, required: false
    get targetReplicaCount() {
        return this.getNumberAttribute('target_replica_count');
    }
    // target_shard_count - computed: true, optional: false, required: false
    get targetShardCount() {
        return this.getNumberAttribute('target_shard_count');
    }
}
exports.RedisClusterStateInfoUpdateInfoOutputReference = RedisClusterStateInfoUpdateInfoOutputReference;
_j = JSII_RTTI_SYMBOL_1;
RedisClusterStateInfoUpdateInfoOutputReference[_j] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterStateInfoUpdateInfoOutputReference", version: "14.13.1" };
class RedisClusterStateInfoUpdateInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterStateInfoUpdateInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterStateInfoUpdateInfoList = RedisClusterStateInfoUpdateInfoList;
_k = JSII_RTTI_SYMBOL_1;
RedisClusterStateInfoUpdateInfoList[_k] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterStateInfoUpdateInfoList", version: "14.13.1" };
function redisClusterStateInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterStateInfoToTerraform = redisClusterStateInfoToTerraform;
function redisClusterStateInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterStateInfoToHclTerraform = redisClusterStateInfoToHclTerraform;
class RedisClusterStateInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // update_info - computed: true, optional: false, required: false
        this._updateInfo = new RedisClusterStateInfoUpdateInfoList(this, "update_info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get updateInfo() {
        return this._updateInfo;
    }
}
exports.RedisClusterStateInfoOutputReference = RedisClusterStateInfoOutputReference;
_l = JSII_RTTI_SYMBOL_1;
RedisClusterStateInfoOutputReference[_l] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterStateInfoOutputReference", version: "14.13.1" };
class RedisClusterStateInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterStateInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterStateInfoList = RedisClusterStateInfoList;
_m = JSII_RTTI_SYMBOL_1;
RedisClusterStateInfoList[_m] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterStateInfoList", version: "14.13.1" };
function redisClusterCrossClusterReplicationConfigMembershipPrimaryClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterCrossClusterReplicationConfigMembershipPrimaryClusterToTerraform = redisClusterCrossClusterReplicationConfigMembershipPrimaryClusterToTerraform;
function redisClusterCrossClusterReplicationConfigMembershipPrimaryClusterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterCrossClusterReplicationConfigMembershipPrimaryClusterToHclTerraform = redisClusterCrossClusterReplicationConfigMembershipPrimaryClusterToHclTerraform;
class RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cluster - computed: true, optional: false, required: false
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
}
exports.RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterOutputReference = RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterOutputReference;
_o = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterOutputReference[_o] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterOutputReference", version: "14.13.1" };
class RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterList = RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterList;
_p = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterList[_p] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterList", version: "14.13.1" };
function redisClusterCrossClusterReplicationConfigMembershipSecondaryClustersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterCrossClusterReplicationConfigMembershipSecondaryClustersToTerraform = redisClusterCrossClusterReplicationConfigMembershipSecondaryClustersToTerraform;
function redisClusterCrossClusterReplicationConfigMembershipSecondaryClustersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterCrossClusterReplicationConfigMembershipSecondaryClustersToHclTerraform = redisClusterCrossClusterReplicationConfigMembershipSecondaryClustersToHclTerraform;
class RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cluster - computed: true, optional: false, required: false
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
}
exports.RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersOutputReference = RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersOutputReference;
_q = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersOutputReference[_q] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersOutputReference", version: "14.13.1" };
class RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersList = RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersList;
_r = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersList[_r] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersList", version: "14.13.1" };
function redisClusterCrossClusterReplicationConfigMembershipToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.redisClusterCrossClusterReplicationConfigMembershipToTerraform = redisClusterCrossClusterReplicationConfigMembershipToTerraform;
function redisClusterCrossClusterReplicationConfigMembershipToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.redisClusterCrossClusterReplicationConfigMembershipToHclTerraform = redisClusterCrossClusterReplicationConfigMembershipToHclTerraform;
class RedisClusterCrossClusterReplicationConfigMembershipOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // primary_cluster - computed: true, optional: false, required: false
        this._primaryCluster = new RedisClusterCrossClusterReplicationConfigMembershipPrimaryClusterList(this, "primary_cluster", false);
        // secondary_clusters - computed: true, optional: false, required: false
        this._secondaryClusters = new RedisClusterCrossClusterReplicationConfigMembershipSecondaryClustersList(this, "secondary_clusters", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get primaryCluster() {
        return this._primaryCluster;
    }
    get secondaryClusters() {
        return this._secondaryClusters;
    }
}
exports.RedisClusterCrossClusterReplicationConfigMembershipOutputReference = RedisClusterCrossClusterReplicationConfigMembershipOutputReference;
_s = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigMembershipOutputReference[_s] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigMembershipOutputReference", version: "14.13.1" };
class RedisClusterCrossClusterReplicationConfigMembershipList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterCrossClusterReplicationConfigMembershipOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterCrossClusterReplicationConfigMembershipList = RedisClusterCrossClusterReplicationConfigMembershipList;
_t = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigMembershipList[_t] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigMembershipList", version: "14.13.1" };
function redisClusterCrossClusterReplicationConfigPrimaryClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cluster: cdktf.stringToTerraform(struct.cluster),
    };
}
exports.redisClusterCrossClusterReplicationConfigPrimaryClusterToTerraform = redisClusterCrossClusterReplicationConfigPrimaryClusterToTerraform;
function redisClusterCrossClusterReplicationConfigPrimaryClusterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cluster: {
            value: cdktf.stringToHclTerraform(struct.cluster),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterCrossClusterReplicationConfigPrimaryClusterToHclTerraform = redisClusterCrossClusterReplicationConfigPrimaryClusterToHclTerraform;
class RedisClusterCrossClusterReplicationConfigPrimaryClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cluster !== undefined) {
            hasAnyValues = true;
            internalValueResult.cluster = this._cluster;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cluster = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cluster = value.cluster;
        }
    }
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    set cluster(value) {
        this._cluster = value;
    }
    resetCluster() {
        this._cluster = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterInput() {
        return this._cluster;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
}
exports.RedisClusterCrossClusterReplicationConfigPrimaryClusterOutputReference = RedisClusterCrossClusterReplicationConfigPrimaryClusterOutputReference;
_u = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigPrimaryClusterOutputReference[_u] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigPrimaryClusterOutputReference", version: "14.13.1" };
function redisClusterCrossClusterReplicationConfigSecondaryClustersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cluster: cdktf.stringToTerraform(struct.cluster),
    };
}
exports.redisClusterCrossClusterReplicationConfigSecondaryClustersToTerraform = redisClusterCrossClusterReplicationConfigSecondaryClustersToTerraform;
function redisClusterCrossClusterReplicationConfigSecondaryClustersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cluster: {
            value: cdktf.stringToHclTerraform(struct.cluster),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterCrossClusterReplicationConfigSecondaryClustersToHclTerraform = redisClusterCrossClusterReplicationConfigSecondaryClustersToHclTerraform;
class RedisClusterCrossClusterReplicationConfigSecondaryClustersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cluster !== undefined) {
            hasAnyValues = true;
            internalValueResult.cluster = this._cluster;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cluster = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cluster = value.cluster;
        }
    }
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    set cluster(value) {
        this._cluster = value;
    }
    resetCluster() {
        this._cluster = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterInput() {
        return this._cluster;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
}
exports.RedisClusterCrossClusterReplicationConfigSecondaryClustersOutputReference = RedisClusterCrossClusterReplicationConfigSecondaryClustersOutputReference;
_v = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigSecondaryClustersOutputReference[_v] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigSecondaryClustersOutputReference", version: "14.13.1" };
class RedisClusterCrossClusterReplicationConfigSecondaryClustersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterCrossClusterReplicationConfigSecondaryClustersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterCrossClusterReplicationConfigSecondaryClustersList = RedisClusterCrossClusterReplicationConfigSecondaryClustersList;
_w = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigSecondaryClustersList[_w] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigSecondaryClustersList", version: "14.13.1" };
function redisClusterCrossClusterReplicationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cluster_role: cdktf.stringToTerraform(struct.clusterRole),
        primary_cluster: redisClusterCrossClusterReplicationConfigPrimaryClusterToTerraform(struct.primaryCluster),
        secondary_clusters: cdktf.listMapper(redisClusterCrossClusterReplicationConfigSecondaryClustersToTerraform, true)(struct.secondaryClusters),
    };
}
exports.redisClusterCrossClusterReplicationConfigToTerraform = redisClusterCrossClusterReplicationConfigToTerraform;
function redisClusterCrossClusterReplicationConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cluster_role: {
            value: cdktf.stringToHclTerraform(struct.clusterRole),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        primary_cluster: {
            value: redisClusterCrossClusterReplicationConfigPrimaryClusterToHclTerraform(struct.primaryCluster),
            isBlock: true,
            type: "list",
            storageClassType: "RedisClusterCrossClusterReplicationConfigPrimaryClusterList",
        },
        secondary_clusters: {
            value: cdktf.listMapperHcl(redisClusterCrossClusterReplicationConfigSecondaryClustersToHclTerraform, true)(struct.secondaryClusters),
            isBlock: true,
            type: "list",
            storageClassType: "RedisClusterCrossClusterReplicationConfigSecondaryClustersList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterCrossClusterReplicationConfigToHclTerraform = redisClusterCrossClusterReplicationConfigToHclTerraform;
class RedisClusterCrossClusterReplicationConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // membership - computed: true, optional: false, required: false
        this._membership = new RedisClusterCrossClusterReplicationConfigMembershipList(this, "membership", false);
        // primary_cluster - computed: false, optional: true, required: false
        this._primaryCluster = new RedisClusterCrossClusterReplicationConfigPrimaryClusterOutputReference(this, "primary_cluster");
        // secondary_clusters - computed: false, optional: true, required: false
        this._secondaryClusters = new RedisClusterCrossClusterReplicationConfigSecondaryClustersList(this, "secondary_clusters", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clusterRole !== undefined) {
            hasAnyValues = true;
            internalValueResult.clusterRole = this._clusterRole;
        }
        if (this._primaryCluster?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.primaryCluster = this._primaryCluster?.internalValue;
        }
        if (this._secondaryClusters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secondaryClusters = this._secondaryClusters?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clusterRole = undefined;
            this._primaryCluster.internalValue = undefined;
            this._secondaryClusters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clusterRole = value.clusterRole;
            this._primaryCluster.internalValue = value.primaryCluster;
            this._secondaryClusters.internalValue = value.secondaryClusters;
        }
    }
    get clusterRole() {
        return this.getStringAttribute('cluster_role');
    }
    set clusterRole(value) {
        this._clusterRole = value;
    }
    resetClusterRole() {
        this._clusterRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterRoleInput() {
        return this._clusterRole;
    }
    get membership() {
        return this._membership;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get primaryCluster() {
        return this._primaryCluster;
    }
    putPrimaryCluster(value) {
        this._primaryCluster.internalValue = value;
    }
    resetPrimaryCluster() {
        this._primaryCluster.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get primaryClusterInput() {
        return this._primaryCluster.internalValue;
    }
    get secondaryClusters() {
        return this._secondaryClusters;
    }
    putSecondaryClusters(value) {
        this._secondaryClusters.internalValue = value;
    }
    resetSecondaryClusters() {
        this._secondaryClusters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secondaryClustersInput() {
        return this._secondaryClusters.internalValue;
    }
}
exports.RedisClusterCrossClusterReplicationConfigOutputReference = RedisClusterCrossClusterReplicationConfigOutputReference;
_x = JSII_RTTI_SYMBOL_1;
RedisClusterCrossClusterReplicationConfigOutputReference[_x] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterCrossClusterReplicationConfigOutputReference", version: "14.13.1" };
function redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hours: cdktf.numberToTerraform(struct.hours),
        minutes: cdktf.numberToTerraform(struct.minutes),
        nanos: cdktf.numberToTerraform(struct.nanos),
        seconds: cdktf.numberToTerraform(struct.seconds),
    };
}
exports.redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToTerraform = redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToTerraform;
function redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        hours: {
            value: cdktf.numberToHclTerraform(struct.hours),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minutes: {
            value: cdktf.numberToHclTerraform(struct.minutes),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        nanos: {
            value: cdktf.numberToHclTerraform(struct.nanos),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        seconds: {
            value: cdktf.numberToHclTerraform(struct.seconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToHclTerraform = redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToHclTerraform;
class RedisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hours !== undefined) {
            hasAnyValues = true;
            internalValueResult.hours = this._hours;
        }
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        if (this._nanos !== undefined) {
            hasAnyValues = true;
            internalValueResult.nanos = this._nanos;
        }
        if (this._seconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.seconds = this._seconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hours = undefined;
            this._minutes = undefined;
            this._nanos = undefined;
            this._seconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hours = value.hours;
            this._minutes = value.minutes;
            this._nanos = value.nanos;
            this._seconds = value.seconds;
        }
    }
    get hours() {
        return this.getNumberAttribute('hours');
    }
    set hours(value) {
        this._hours = value;
    }
    resetHours() {
        this._hours = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hoursInput() {
        return this._hours;
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    resetMinutes() {
        this._minutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    set nanos(value) {
        this._nanos = value;
    }
    resetNanos() {
        this._nanos = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nanosInput() {
        return this._nanos;
    }
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
    set seconds(value) {
        this._seconds = value;
    }
    resetSeconds() {
        this._seconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secondsInput() {
        return this._seconds;
    }
}
exports.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeOutputReference = RedisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeOutputReference;
_y = JSII_RTTI_SYMBOL_1;
RedisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeOutputReference[_y] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeOutputReference", version: "14.13.1" };
function redisClusterMaintenancePolicyWeeklyMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        start_time: redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToTerraform(struct.startTime),
    };
}
exports.redisClusterMaintenancePolicyWeeklyMaintenanceWindowToTerraform = redisClusterMaintenancePolicyWeeklyMaintenanceWindowToTerraform;
function redisClusterMaintenancePolicyWeeklyMaintenanceWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        day: {
            value: cdktf.stringToHclTerraform(struct.day),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start_time: {
            value: redisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeToHclTerraform(struct.startTime),
            isBlock: true,
            type: "list",
            storageClassType: "RedisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterMaintenancePolicyWeeklyMaintenanceWindowToHclTerraform = redisClusterMaintenancePolicyWeeklyMaintenanceWindowToHclTerraform;
class RedisClusterMaintenancePolicyWeeklyMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // start_time - computed: false, optional: false, required: true
        this._startTime = new RedisClusterMaintenancePolicyWeeklyMaintenanceWindowStartTimeOutputReference(this, "start_time");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._startTime?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._day = undefined;
            this._startTime.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._day = value.day;
            this._startTime.internalValue = value.startTime;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    // duration - computed: true, optional: false, required: false
    get duration() {
        return this.getStringAttribute('duration');
    }
    get startTime() {
        return this._startTime;
    }
    putStartTime(value) {
        this._startTime.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime.internalValue;
    }
}
exports.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowOutputReference = RedisClusterMaintenancePolicyWeeklyMaintenanceWindowOutputReference;
_z = JSII_RTTI_SYMBOL_1;
RedisClusterMaintenancePolicyWeeklyMaintenanceWindowOutputReference[_z] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowOutputReference", version: "14.13.1" };
class RedisClusterMaintenancePolicyWeeklyMaintenanceWindowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterMaintenancePolicyWeeklyMaintenanceWindowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowList = RedisClusterMaintenancePolicyWeeklyMaintenanceWindowList;
_0 = JSII_RTTI_SYMBOL_1;
RedisClusterMaintenancePolicyWeeklyMaintenanceWindowList[_0] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterMaintenancePolicyWeeklyMaintenanceWindowList", version: "14.13.1" };
function redisClusterMaintenancePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        weekly_maintenance_window: cdktf.listMapper(redisClusterMaintenancePolicyWeeklyMaintenanceWindowToTerraform, true)(struct.weeklyMaintenanceWindow),
    };
}
exports.redisClusterMaintenancePolicyToTerraform = redisClusterMaintenancePolicyToTerraform;
function redisClusterMaintenancePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        weekly_maintenance_window: {
            value: cdktf.listMapperHcl(redisClusterMaintenancePolicyWeeklyMaintenanceWindowToHclTerraform, true)(struct.weeklyMaintenanceWindow),
            isBlock: true,
            type: "list",
            storageClassType: "RedisClusterMaintenancePolicyWeeklyMaintenanceWindowList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterMaintenancePolicyToHclTerraform = redisClusterMaintenancePolicyToHclTerraform;
class RedisClusterMaintenancePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // weekly_maintenance_window - computed: false, optional: true, required: false
        this._weeklyMaintenanceWindow = new RedisClusterMaintenancePolicyWeeklyMaintenanceWindowList(this, "weekly_maintenance_window", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._weeklyMaintenanceWindow?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.weeklyMaintenanceWindow = this._weeklyMaintenanceWindow?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._weeklyMaintenanceWindow.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._weeklyMaintenanceWindow.internalValue = value.weeklyMaintenanceWindow;
        }
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get weeklyMaintenanceWindow() {
        return this._weeklyMaintenanceWindow;
    }
    putWeeklyMaintenanceWindow(value) {
        this._weeklyMaintenanceWindow.internalValue = value;
    }
    resetWeeklyMaintenanceWindow() {
        this._weeklyMaintenanceWindow.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weeklyMaintenanceWindowInput() {
        return this._weeklyMaintenanceWindow.internalValue;
    }
}
exports.RedisClusterMaintenancePolicyOutputReference = RedisClusterMaintenancePolicyOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
RedisClusterMaintenancePolicyOutputReference[_1] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterMaintenancePolicyOutputReference", version: "14.13.1" };
function redisClusterPersistenceConfigAofConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        append_fsync: cdktf.stringToTerraform(struct.appendFsync),
    };
}
exports.redisClusterPersistenceConfigAofConfigToTerraform = redisClusterPersistenceConfigAofConfigToTerraform;
function redisClusterPersistenceConfigAofConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        append_fsync: {
            value: cdktf.stringToHclTerraform(struct.appendFsync),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterPersistenceConfigAofConfigToHclTerraform = redisClusterPersistenceConfigAofConfigToHclTerraform;
class RedisClusterPersistenceConfigAofConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._appendFsync !== undefined) {
            hasAnyValues = true;
            internalValueResult.appendFsync = this._appendFsync;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._appendFsync = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._appendFsync = value.appendFsync;
        }
    }
    get appendFsync() {
        return this.getStringAttribute('append_fsync');
    }
    set appendFsync(value) {
        this._appendFsync = value;
    }
    resetAppendFsync() {
        this._appendFsync = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get appendFsyncInput() {
        return this._appendFsync;
    }
}
exports.RedisClusterPersistenceConfigAofConfigOutputReference = RedisClusterPersistenceConfigAofConfigOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
RedisClusterPersistenceConfigAofConfigOutputReference[_2] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterPersistenceConfigAofConfigOutputReference", version: "14.13.1" };
function redisClusterPersistenceConfigRdbConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        rdb_snapshot_period: cdktf.stringToTerraform(struct.rdbSnapshotPeriod),
        rdb_snapshot_start_time: cdktf.stringToTerraform(struct.rdbSnapshotStartTime),
    };
}
exports.redisClusterPersistenceConfigRdbConfigToTerraform = redisClusterPersistenceConfigRdbConfigToTerraform;
function redisClusterPersistenceConfigRdbConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        rdb_snapshot_period: {
            value: cdktf.stringToHclTerraform(struct.rdbSnapshotPeriod),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rdb_snapshot_start_time: {
            value: cdktf.stringToHclTerraform(struct.rdbSnapshotStartTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterPersistenceConfigRdbConfigToHclTerraform = redisClusterPersistenceConfigRdbConfigToHclTerraform;
class RedisClusterPersistenceConfigRdbConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rdbSnapshotPeriod !== undefined) {
            hasAnyValues = true;
            internalValueResult.rdbSnapshotPeriod = this._rdbSnapshotPeriod;
        }
        if (this._rdbSnapshotStartTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.rdbSnapshotStartTime = this._rdbSnapshotStartTime;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rdbSnapshotPeriod = undefined;
            this._rdbSnapshotStartTime = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rdbSnapshotPeriod = value.rdbSnapshotPeriod;
            this._rdbSnapshotStartTime = value.rdbSnapshotStartTime;
        }
    }
    get rdbSnapshotPeriod() {
        return this.getStringAttribute('rdb_snapshot_period');
    }
    set rdbSnapshotPeriod(value) {
        this._rdbSnapshotPeriod = value;
    }
    resetRdbSnapshotPeriod() {
        this._rdbSnapshotPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rdbSnapshotPeriodInput() {
        return this._rdbSnapshotPeriod;
    }
    get rdbSnapshotStartTime() {
        return this.getStringAttribute('rdb_snapshot_start_time');
    }
    set rdbSnapshotStartTime(value) {
        this._rdbSnapshotStartTime = value;
    }
    resetRdbSnapshotStartTime() {
        this._rdbSnapshotStartTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rdbSnapshotStartTimeInput() {
        return this._rdbSnapshotStartTime;
    }
}
exports.RedisClusterPersistenceConfigRdbConfigOutputReference = RedisClusterPersistenceConfigRdbConfigOutputReference;
_3 = JSII_RTTI_SYMBOL_1;
RedisClusterPersistenceConfigRdbConfigOutputReference[_3] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterPersistenceConfigRdbConfigOutputReference", version: "14.13.1" };
function redisClusterPersistenceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
        aof_config: redisClusterPersistenceConfigAofConfigToTerraform(struct.aofConfig),
        rdb_config: redisClusterPersistenceConfigRdbConfigToTerraform(struct.rdbConfig),
    };
}
exports.redisClusterPersistenceConfigToTerraform = redisClusterPersistenceConfigToTerraform;
function redisClusterPersistenceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        aof_config: {
            value: redisClusterPersistenceConfigAofConfigToHclTerraform(struct.aofConfig),
            isBlock: true,
            type: "list",
            storageClassType: "RedisClusterPersistenceConfigAofConfigList",
        },
        rdb_config: {
            value: redisClusterPersistenceConfigRdbConfigToHclTerraform(struct.rdbConfig),
            isBlock: true,
            type: "list",
            storageClassType: "RedisClusterPersistenceConfigRdbConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterPersistenceConfigToHclTerraform = redisClusterPersistenceConfigToHclTerraform;
class RedisClusterPersistenceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // aof_config - computed: false, optional: true, required: false
        this._aofConfig = new RedisClusterPersistenceConfigAofConfigOutputReference(this, "aof_config");
        // rdb_config - computed: false, optional: true, required: false
        this._rdbConfig = new RedisClusterPersistenceConfigRdbConfigOutputReference(this, "rdb_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._aofConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.aofConfig = this._aofConfig?.internalValue;
        }
        if (this._rdbConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rdbConfig = this._rdbConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mode = undefined;
            this._aofConfig.internalValue = undefined;
            this._rdbConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mode = value.mode;
            this._aofConfig.internalValue = value.aofConfig;
            this._rdbConfig.internalValue = value.rdbConfig;
        }
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get aofConfig() {
        return this._aofConfig;
    }
    putAofConfig(value) {
        this._aofConfig.internalValue = value;
    }
    resetAofConfig() {
        this._aofConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get aofConfigInput() {
        return this._aofConfig.internalValue;
    }
    get rdbConfig() {
        return this._rdbConfig;
    }
    putRdbConfig(value) {
        this._rdbConfig.internalValue = value;
    }
    resetRdbConfig() {
        this._rdbConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rdbConfigInput() {
        return this._rdbConfig.internalValue;
    }
}
exports.RedisClusterPersistenceConfigOutputReference = RedisClusterPersistenceConfigOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
RedisClusterPersistenceConfigOutputReference[_4] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterPersistenceConfigOutputReference", version: "14.13.1" };
function redisClusterPscConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        network: cdktf.stringToTerraform(struct.network),
    };
}
exports.redisClusterPscConfigsToTerraform = redisClusterPscConfigsToTerraform;
function redisClusterPscConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        network: {
            value: cdktf.stringToHclTerraform(struct.network),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterPscConfigsToHclTerraform = redisClusterPscConfigsToHclTerraform;
class RedisClusterPscConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._network !== undefined) {
            hasAnyValues = true;
            internalValueResult.network = this._network;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._network = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._network = value.network;
        }
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
}
exports.RedisClusterPscConfigsOutputReference = RedisClusterPscConfigsOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
RedisClusterPscConfigsOutputReference[_5] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterPscConfigsOutputReference", version: "14.13.1" };
class RedisClusterPscConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new RedisClusterPscConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.RedisClusterPscConfigsList = RedisClusterPscConfigsList;
_6 = JSII_RTTI_SYMBOL_1;
RedisClusterPscConfigsList[_6] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterPscConfigsList", version: "14.13.1" };
function redisClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.redisClusterTimeoutsToTerraform = redisClusterTimeoutsToTerraform;
function redisClusterTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterTimeoutsToHclTerraform = redisClusterTimeoutsToHclTerraform;
class RedisClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.RedisClusterTimeoutsOutputReference = RedisClusterTimeoutsOutputReference;
_7 = JSII_RTTI_SYMBOL_1;
RedisClusterTimeoutsOutputReference[_7] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterTimeoutsOutputReference", version: "14.13.1" };
function redisClusterZoneDistributionConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
        zone: cdktf.stringToTerraform(struct.zone),
    };
}
exports.redisClusterZoneDistributionConfigToTerraform = redisClusterZoneDistributionConfigToTerraform;
function redisClusterZoneDistributionConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        zone: {
            value: cdktf.stringToHclTerraform(struct.zone),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.redisClusterZoneDistributionConfigToHclTerraform = redisClusterZoneDistributionConfigToHclTerraform;
class RedisClusterZoneDistributionConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._zone !== undefined) {
            hasAnyValues = true;
            internalValueResult.zone = this._zone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mode = undefined;
            this._zone = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mode = value.mode;
            this._zone = value.zone;
        }
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
}
exports.RedisClusterZoneDistributionConfigOutputReference = RedisClusterZoneDistributionConfigOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
RedisClusterZoneDistributionConfigOutputReference[_8] = { fqn: "@cdktf/provider-google.redisCluster.RedisClusterZoneDistributionConfigOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/redis_cluster google_redis_cluster}
*/
class RedisCluster extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a RedisCluster resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the RedisCluster to import
    * @param importFromId The id of the existing RedisCluster that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/redis_cluster#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the RedisCluster to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_redis_cluster", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/redis_cluster google_redis_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options RedisClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_redis_cluster',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // discovery_endpoints - computed: true, optional: false, required: false
        this._discoveryEndpoints = new RedisClusterDiscoveryEndpointsList(this, "discovery_endpoints", false);
        // maintenance_schedule - computed: true, optional: false, required: false
        this._maintenanceSchedule = new RedisClusterMaintenanceScheduleList(this, "maintenance_schedule", false);
        // psc_connections - computed: true, optional: false, required: false
        this._pscConnections = new RedisClusterPscConnectionsList(this, "psc_connections", false);
        // state_info - computed: true, optional: false, required: false
        this._stateInfo = new RedisClusterStateInfoList(this, "state_info", false);
        // cross_cluster_replication_config - computed: false, optional: true, required: false
        this._crossClusterReplicationConfig = new RedisClusterCrossClusterReplicationConfigOutputReference(this, "cross_cluster_replication_config");
        // maintenance_policy - computed: false, optional: true, required: false
        this._maintenancePolicy = new RedisClusterMaintenancePolicyOutputReference(this, "maintenance_policy");
        // persistence_config - computed: false, optional: true, required: false
        this._persistenceConfig = new RedisClusterPersistenceConfigOutputReference(this, "persistence_config");
        // psc_configs - computed: false, optional: false, required: true
        this._pscConfigs = new RedisClusterPscConfigsList(this, "psc_configs", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new RedisClusterTimeoutsOutputReference(this, "timeouts");
        // zone_distribution_config - computed: false, optional: true, required: false
        this._zoneDistributionConfig = new RedisClusterZoneDistributionConfigOutputReference(this, "zone_distribution_config");
        this._authorizationMode = config.authorizationMode;
        this._deletionProtectionEnabled = config.deletionProtectionEnabled;
        this._id = config.id;
        this._name = config.name;
        this._nodeType = config.nodeType;
        this._project = config.project;
        this._redisConfigs = config.redisConfigs;
        this._region = config.region;
        this._replicaCount = config.replicaCount;
        this._shardCount = config.shardCount;
        this._transitEncryptionMode = config.transitEncryptionMode;
        this._crossClusterReplicationConfig.internalValue = config.crossClusterReplicationConfig;
        this._maintenancePolicy.internalValue = config.maintenancePolicy;
        this._persistenceConfig.internalValue = config.persistenceConfig;
        this._pscConfigs.internalValue = config.pscConfigs;
        this._timeouts.internalValue = config.timeouts;
        this._zoneDistributionConfig.internalValue = config.zoneDistributionConfig;
    }
    get authorizationMode() {
        return this.getStringAttribute('authorization_mode');
    }
    set authorizationMode(value) {
        this._authorizationMode = value;
    }
    resetAuthorizationMode() {
        this._authorizationMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationModeInput() {
        return this._authorizationMode;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get deletionProtectionEnabled() {
        return this.getBooleanAttribute('deletion_protection_enabled');
    }
    set deletionProtectionEnabled(value) {
        this._deletionProtectionEnabled = value;
    }
    resetDeletionProtectionEnabled() {
        this._deletionProtectionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionEnabledInput() {
        return this._deletionProtectionEnabled;
    }
    get discoveryEndpoints() {
        return this._discoveryEndpoints;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get maintenanceSchedule() {
        return this._maintenanceSchedule;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get nodeType() {
        return this.getStringAttribute('node_type');
    }
    set nodeType(value) {
        this._nodeType = value;
    }
    resetNodeType() {
        this._nodeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeTypeInput() {
        return this._nodeType;
    }
    // precise_size_gb - computed: true, optional: false, required: false
    get preciseSizeGb() {
        return this.getNumberAttribute('precise_size_gb');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get pscConnections() {
        return this._pscConnections;
    }
    get redisConfigs() {
        return this.getStringMapAttribute('redis_configs');
    }
    set redisConfigs(value) {
        this._redisConfigs = value;
    }
    resetRedisConfigs() {
        this._redisConfigs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get redisConfigsInput() {
        return this._redisConfigs;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get replicaCount() {
        return this.getNumberAttribute('replica_count');
    }
    set replicaCount(value) {
        this._replicaCount = value;
    }
    resetReplicaCount() {
        this._replicaCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicaCountInput() {
        return this._replicaCount;
    }
    get shardCount() {
        return this.getNumberAttribute('shard_count');
    }
    set shardCount(value) {
        this._shardCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get shardCountInput() {
        return this._shardCount;
    }
    // size_gb - computed: true, optional: false, required: false
    get sizeGb() {
        return this.getNumberAttribute('size_gb');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get stateInfo() {
        return this._stateInfo;
    }
    get transitEncryptionMode() {
        return this.getStringAttribute('transit_encryption_mode');
    }
    set transitEncryptionMode(value) {
        this._transitEncryptionMode = value;
    }
    resetTransitEncryptionMode() {
        this._transitEncryptionMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transitEncryptionModeInput() {
        return this._transitEncryptionMode;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    get crossClusterReplicationConfig() {
        return this._crossClusterReplicationConfig;
    }
    putCrossClusterReplicationConfig(value) {
        this._crossClusterReplicationConfig.internalValue = value;
    }
    resetCrossClusterReplicationConfig() {
        this._crossClusterReplicationConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get crossClusterReplicationConfigInput() {
        return this._crossClusterReplicationConfig.internalValue;
    }
    get maintenancePolicy() {
        return this._maintenancePolicy;
    }
    putMaintenancePolicy(value) {
        this._maintenancePolicy.internalValue = value;
    }
    resetMaintenancePolicy() {
        this._maintenancePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenancePolicyInput() {
        return this._maintenancePolicy.internalValue;
    }
    get persistenceConfig() {
        return this._persistenceConfig;
    }
    putPersistenceConfig(value) {
        this._persistenceConfig.internalValue = value;
    }
    resetPersistenceConfig() {
        this._persistenceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get persistenceConfigInput() {
        return this._persistenceConfig.internalValue;
    }
    get pscConfigs() {
        return this._pscConfigs;
    }
    putPscConfigs(value) {
        this._pscConfigs.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get pscConfigsInput() {
        return this._pscConfigs.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get zoneDistributionConfig() {
        return this._zoneDistributionConfig;
    }
    putZoneDistributionConfig(value) {
        this._zoneDistributionConfig.internalValue = value;
    }
    resetZoneDistributionConfig() {
        this._zoneDistributionConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneDistributionConfigInput() {
        return this._zoneDistributionConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            authorization_mode: cdktf.stringToTerraform(this._authorizationMode),
            deletion_protection_enabled: cdktf.booleanToTerraform(this._deletionProtectionEnabled),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            node_type: cdktf.stringToTerraform(this._nodeType),
            project: cdktf.stringToTerraform(this._project),
            redis_configs: cdktf.hashMapper(cdktf.stringToTerraform)(this._redisConfigs),
            region: cdktf.stringToTerraform(this._region),
            replica_count: cdktf.numberToTerraform(this._replicaCount),
            shard_count: cdktf.numberToTerraform(this._shardCount),
            transit_encryption_mode: cdktf.stringToTerraform(this._transitEncryptionMode),
            cross_cluster_replication_config: redisClusterCrossClusterReplicationConfigToTerraform(this._crossClusterReplicationConfig.internalValue),
            maintenance_policy: redisClusterMaintenancePolicyToTerraform(this._maintenancePolicy.internalValue),
            persistence_config: redisClusterPersistenceConfigToTerraform(this._persistenceConfig.internalValue),
            psc_configs: cdktf.listMapper(redisClusterPscConfigsToTerraform, true)(this._pscConfigs.internalValue),
            timeouts: redisClusterTimeoutsToTerraform(this._timeouts.internalValue),
            zone_distribution_config: redisClusterZoneDistributionConfigToTerraform(this._zoneDistributionConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            authorization_mode: {
                value: cdktf.stringToHclTerraform(this._authorizationMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            deletion_protection_enabled: {
                value: cdktf.booleanToHclTerraform(this._deletionProtectionEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            node_type: {
                value: cdktf.stringToHclTerraform(this._nodeType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            redis_configs: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._redisConfigs),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            replica_count: {
                value: cdktf.numberToHclTerraform(this._replicaCount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            shard_count: {
                value: cdktf.numberToHclTerraform(this._shardCount),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            transit_encryption_mode: {
                value: cdktf.stringToHclTerraform(this._transitEncryptionMode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cross_cluster_replication_config: {
                value: redisClusterCrossClusterReplicationConfigToHclTerraform(this._crossClusterReplicationConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "RedisClusterCrossClusterReplicationConfigList",
            },
            maintenance_policy: {
                value: redisClusterMaintenancePolicyToHclTerraform(this._maintenancePolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "RedisClusterMaintenancePolicyList",
            },
            persistence_config: {
                value: redisClusterPersistenceConfigToHclTerraform(this._persistenceConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "RedisClusterPersistenceConfigList",
            },
            psc_configs: {
                value: cdktf.listMapperHcl(redisClusterPscConfigsToHclTerraform, true)(this._pscConfigs.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "RedisClusterPscConfigsList",
            },
            timeouts: {
                value: redisClusterTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "RedisClusterTimeouts",
            },
            zone_distribution_config: {
                value: redisClusterZoneDistributionConfigToHclTerraform(this._zoneDistributionConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "RedisClusterZoneDistributionConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.RedisCluster = RedisCluster;
_9 = JSII_RTTI_SYMBOL_1;
RedisCluster[_9] = { fqn: "@cdktf/provider-google.redisCluster.RedisCluster", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
RedisCluster.tfResourceType = "google_redis_cluster";
