// @cdktf/provider-google/lib/netapp-storage-pool/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetappStoragePool = exports.NetappStoragePoolTimeoutsOutputReference = exports.netappStoragePoolTimeoutsToHclTerraform = exports.netappStoragePoolTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function netappStoragePoolTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.netappStoragePoolTimeoutsToTerraform = netappStoragePoolTimeoutsToTerraform;
function netappStoragePoolTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappStoragePoolTimeoutsToHclTerraform = netappStoragePoolTimeoutsToHclTerraform;
class NetappStoragePoolTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetappStoragePoolTimeoutsOutputReference = NetappStoragePoolTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetappStoragePoolTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-google.netappStoragePool.NetappStoragePoolTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/netapp_storage_pool google_netapp_storage_pool}
*/
class NetappStoragePool extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetappStoragePool resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetappStoragePool to import
    * @param importFromId The id of the existing NetappStoragePool that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/netapp_storage_pool#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetappStoragePool to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_netapp_storage_pool", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/netapp_storage_pool google_netapp_storage_pool} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetappStoragePoolConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_netapp_storage_pool',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetappStoragePoolTimeoutsOutputReference(this, "timeouts");
        this._activeDirectory = config.activeDirectory;
        this._allowAutoTiering = config.allowAutoTiering;
        this._capacityGib = config.capacityGib;
        this._description = config.description;
        this._id = config.id;
        this._kmsConfig = config.kmsConfig;
        this._labels = config.labels;
        this._ldapEnabled = config.ldapEnabled;
        this._location = config.location;
        this._name = config.name;
        this._network = config.network;
        this._project = config.project;
        this._replicaZone = config.replicaZone;
        this._serviceLevel = config.serviceLevel;
        this._zone = config.zone;
        this._timeouts.internalValue = config.timeouts;
    }
    get activeDirectory() {
        return this.getStringAttribute('active_directory');
    }
    set activeDirectory(value) {
        this._activeDirectory = value;
    }
    resetActiveDirectory() {
        this._activeDirectory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get activeDirectoryInput() {
        return this._activeDirectory;
    }
    get allowAutoTiering() {
        return this.getBooleanAttribute('allow_auto_tiering');
    }
    set allowAutoTiering(value) {
        this._allowAutoTiering = value;
    }
    resetAllowAutoTiering() {
        this._allowAutoTiering = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowAutoTieringInput() {
        return this._allowAutoTiering;
    }
    get capacityGib() {
        return this.getStringAttribute('capacity_gib');
    }
    set capacityGib(value) {
        this._capacityGib = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityGibInput() {
        return this._capacityGib;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // encryption_type - computed: true, optional: false, required: false
    get encryptionType() {
        return this.getStringAttribute('encryption_type');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsConfig() {
        return this.getStringAttribute('kms_config');
    }
    set kmsConfig(value) {
        this._kmsConfig = value;
    }
    resetKmsConfig() {
        this._kmsConfig = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsConfigInput() {
        return this._kmsConfig;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get ldapEnabled() {
        return this.getBooleanAttribute('ldap_enabled');
    }
    set ldapEnabled(value) {
        this._ldapEnabled = value;
    }
    resetLdapEnabled() {
        this._ldapEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ldapEnabledInput() {
        return this._ldapEnabled;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get replicaZone() {
        return this.getStringAttribute('replica_zone');
    }
    set replicaZone(value) {
        this._replicaZone = value;
    }
    resetReplicaZone() {
        this._replicaZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicaZoneInput() {
        return this._replicaZone;
    }
    get serviceLevel() {
        return this.getStringAttribute('service_level');
    }
    set serviceLevel(value) {
        this._serviceLevel = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceLevelInput() {
        return this._serviceLevel;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // volume_capacity_gib - computed: true, optional: false, required: false
    get volumeCapacityGib() {
        return this.getStringAttribute('volume_capacity_gib');
    }
    // volume_count - computed: true, optional: false, required: false
    get volumeCount() {
        return this.getNumberAttribute('volume_count');
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            active_directory: cdktf.stringToTerraform(this._activeDirectory),
            allow_auto_tiering: cdktf.booleanToTerraform(this._allowAutoTiering),
            capacity_gib: cdktf.stringToTerraform(this._capacityGib),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            kms_config: cdktf.stringToTerraform(this._kmsConfig),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            ldap_enabled: cdktf.booleanToTerraform(this._ldapEnabled),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            network: cdktf.stringToTerraform(this._network),
            project: cdktf.stringToTerraform(this._project),
            replica_zone: cdktf.stringToTerraform(this._replicaZone),
            service_level: cdktf.stringToTerraform(this._serviceLevel),
            zone: cdktf.stringToTerraform(this._zone),
            timeouts: netappStoragePoolTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            active_directory: {
                value: cdktf.stringToHclTerraform(this._activeDirectory),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            allow_auto_tiering: {
                value: cdktf.booleanToHclTerraform(this._allowAutoTiering),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            capacity_gib: {
                value: cdktf.stringToHclTerraform(this._capacityGib),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_config: {
                value: cdktf.stringToHclTerraform(this._kmsConfig),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            ldap_enabled: {
                value: cdktf.booleanToHclTerraform(this._ldapEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            network: {
                value: cdktf.stringToHclTerraform(this._network),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            replica_zone: {
                value: cdktf.stringToHclTerraform(this._replicaZone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_level: {
                value: cdktf.stringToHclTerraform(this._serviceLevel),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            zone: {
                value: cdktf.stringToHclTerraform(this._zone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: netappStoragePoolTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetappStoragePoolTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetappStoragePool = NetappStoragePool;
_b = JSII_RTTI_SYMBOL_1;
NetappStoragePool[_b] = { fqn: "@cdktf/provider-google.netappStoragePool.NetappStoragePool", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetappStoragePool.tfResourceType = "google_netapp_storage_pool";
