// @cdktf/cli-core/src/lib/dependencies/cdktf-config-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CdktfConfigManager = void 0;
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const cdktf_config_1 = require("../cdktf-config");
const dependency_manager_1 = require("./dependency-manager");
/**
 * facilitates adding provider dependencies to cdktf.json
 */
class CdktfConfigManager {
    constructor() {
        this.config = cdktf_config_1.CdktfConfig.read();
    }
    async hasProvider(constraint) {
        return this.config.terraformProviders.some((provider) => dependency_manager_1.ProviderConstraint.fromConfigEntry(provider).source ===
            constraint.source);
    }
    async setProvider(constraint) {
        // excluding the constraint to be added (if it already existed)
        const currentProviders = this.config.terraformProviders.filter((provider) => dependency_manager_1.ProviderConstraint.fromConfigEntry(provider).source !==
            constraint.source);
        const simplifiedSource = constraint.source.replace("registry.terraform.io/", "");
        const provider = simplifiedSource + (constraint.version ? `@${constraint.version}` : "");
        currentProviders.push(provider);
        this.config.writeTerraformProviders(currentProviders);
    }
    async addProvider(constraint) {
        await this.setProvider(constraint);
    }
    async updateProvider(constraint) {
        await this.setProvider(constraint);
    }
}
exports.CdktfConfigManager = CdktfConfigManager;
