// @cdktf/provider-google/lib/iam-access-boundary-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamAccessBoundaryPolicy = exports.IamAccessBoundaryPolicyTimeoutsOutputReference = exports.iamAccessBoundaryPolicyTimeoutsToHclTerraform = exports.iamAccessBoundaryPolicyTimeoutsToTerraform = exports.IamAccessBoundaryPolicyRulesList = exports.IamAccessBoundaryPolicyRulesOutputReference = exports.iamAccessBoundaryPolicyRulesToHclTerraform = exports.iamAccessBoundaryPolicyRulesToTerraform = exports.IamAccessBoundaryPolicyRulesAccessBoundaryRuleOutputReference = exports.iamAccessBoundaryPolicyRulesAccessBoundaryRuleToHclTerraform = exports.iamAccessBoundaryPolicyRulesAccessBoundaryRuleToTerraform = exports.IamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionOutputReference = exports.iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToHclTerraform = exports.iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        expression: cdktf.stringToTerraform(struct.expression),
        location: cdktf.stringToTerraform(struct.location),
        title: cdktf.stringToTerraform(struct.title),
    };
}
exports.iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToTerraform = iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToTerraform;
function iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        location: {
            value: cdktf.stringToHclTerraform(struct.location),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        title: {
            value: cdktf.stringToHclTerraform(struct.title),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToHclTerraform = iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToHclTerraform;
class IamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._title !== undefined) {
            hasAnyValues = true;
            internalValueResult.title = this._title;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._expression = undefined;
            this._location = undefined;
            this._title = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._expression = value.expression;
            this._location = value.location;
            this._title = value.title;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get title() {
        return this.getStringAttribute('title');
    }
    set title(value) {
        this._title = value;
    }
    resetTitle() {
        this._title = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get titleInput() {
        return this._title;
    }
}
exports.IamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionOutputReference = IamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionOutputReference[_a] = { fqn: "@cdktf/provider-google.iamAccessBoundaryPolicy.IamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionOutputReference", version: "14.13.1" };
function iamAccessBoundaryPolicyRulesAccessBoundaryRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        available_permissions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.availablePermissions),
        available_resource: cdktf.stringToTerraform(struct.availableResource),
        availability_condition: iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToTerraform(struct.availabilityCondition),
    };
}
exports.iamAccessBoundaryPolicyRulesAccessBoundaryRuleToTerraform = iamAccessBoundaryPolicyRulesAccessBoundaryRuleToTerraform;
function iamAccessBoundaryPolicyRulesAccessBoundaryRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        available_permissions: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.availablePermissions),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        available_resource: {
            value: cdktf.stringToHclTerraform(struct.availableResource),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        availability_condition: {
            value: iamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionToHclTerraform(struct.availabilityCondition),
            isBlock: true,
            type: "list",
            storageClassType: "IamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamAccessBoundaryPolicyRulesAccessBoundaryRuleToHclTerraform = iamAccessBoundaryPolicyRulesAccessBoundaryRuleToHclTerraform;
class IamAccessBoundaryPolicyRulesAccessBoundaryRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // availability_condition - computed: false, optional: true, required: false
        this._availabilityCondition = new IamAccessBoundaryPolicyRulesAccessBoundaryRuleAvailabilityConditionOutputReference(this, "availability_condition");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._availablePermissions !== undefined) {
            hasAnyValues = true;
            internalValueResult.availablePermissions = this._availablePermissions;
        }
        if (this._availableResource !== undefined) {
            hasAnyValues = true;
            internalValueResult.availableResource = this._availableResource;
        }
        if (this._availabilityCondition?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.availabilityCondition = this._availabilityCondition?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._availablePermissions = undefined;
            this._availableResource = undefined;
            this._availabilityCondition.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._availablePermissions = value.availablePermissions;
            this._availableResource = value.availableResource;
            this._availabilityCondition.internalValue = value.availabilityCondition;
        }
    }
    get availablePermissions() {
        return this.getListAttribute('available_permissions');
    }
    set availablePermissions(value) {
        this._availablePermissions = value;
    }
    resetAvailablePermissions() {
        this._availablePermissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availablePermissionsInput() {
        return this._availablePermissions;
    }
    get availableResource() {
        return this.getStringAttribute('available_resource');
    }
    set availableResource(value) {
        this._availableResource = value;
    }
    resetAvailableResource() {
        this._availableResource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availableResourceInput() {
        return this._availableResource;
    }
    get availabilityCondition() {
        return this._availabilityCondition;
    }
    putAvailabilityCondition(value) {
        this._availabilityCondition.internalValue = value;
    }
    resetAvailabilityCondition() {
        this._availabilityCondition.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get availabilityConditionInput() {
        return this._availabilityCondition.internalValue;
    }
}
exports.IamAccessBoundaryPolicyRulesAccessBoundaryRuleOutputReference = IamAccessBoundaryPolicyRulesAccessBoundaryRuleOutputReference;
_b = JSII_RTTI_SYMBOL_1;
IamAccessBoundaryPolicyRulesAccessBoundaryRuleOutputReference[_b] = { fqn: "@cdktf/provider-google.iamAccessBoundaryPolicy.IamAccessBoundaryPolicyRulesAccessBoundaryRuleOutputReference", version: "14.13.1" };
function iamAccessBoundaryPolicyRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        access_boundary_rule: iamAccessBoundaryPolicyRulesAccessBoundaryRuleToTerraform(struct.accessBoundaryRule),
    };
}
exports.iamAccessBoundaryPolicyRulesToTerraform = iamAccessBoundaryPolicyRulesToTerraform;
function iamAccessBoundaryPolicyRulesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        access_boundary_rule: {
            value: iamAccessBoundaryPolicyRulesAccessBoundaryRuleToHclTerraform(struct.accessBoundaryRule),
            isBlock: true,
            type: "list",
            storageClassType: "IamAccessBoundaryPolicyRulesAccessBoundaryRuleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamAccessBoundaryPolicyRulesToHclTerraform = iamAccessBoundaryPolicyRulesToHclTerraform;
class IamAccessBoundaryPolicyRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // access_boundary_rule - computed: false, optional: true, required: false
        this._accessBoundaryRule = new IamAccessBoundaryPolicyRulesAccessBoundaryRuleOutputReference(this, "access_boundary_rule");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._accessBoundaryRule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessBoundaryRule = this._accessBoundaryRule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._accessBoundaryRule.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._accessBoundaryRule.internalValue = value.accessBoundaryRule;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get accessBoundaryRule() {
        return this._accessBoundaryRule;
    }
    putAccessBoundaryRule(value) {
        this._accessBoundaryRule.internalValue = value;
    }
    resetAccessBoundaryRule() {
        this._accessBoundaryRule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessBoundaryRuleInput() {
        return this._accessBoundaryRule.internalValue;
    }
}
exports.IamAccessBoundaryPolicyRulesOutputReference = IamAccessBoundaryPolicyRulesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IamAccessBoundaryPolicyRulesOutputReference[_c] = { fqn: "@cdktf/provider-google.iamAccessBoundaryPolicy.IamAccessBoundaryPolicyRulesOutputReference", version: "14.13.1" };
class IamAccessBoundaryPolicyRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IamAccessBoundaryPolicyRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IamAccessBoundaryPolicyRulesList = IamAccessBoundaryPolicyRulesList;
_d = JSII_RTTI_SYMBOL_1;
IamAccessBoundaryPolicyRulesList[_d] = { fqn: "@cdktf/provider-google.iamAccessBoundaryPolicy.IamAccessBoundaryPolicyRulesList", version: "14.13.1" };
function iamAccessBoundaryPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.iamAccessBoundaryPolicyTimeoutsToTerraform = iamAccessBoundaryPolicyTimeoutsToTerraform;
function iamAccessBoundaryPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamAccessBoundaryPolicyTimeoutsToHclTerraform = iamAccessBoundaryPolicyTimeoutsToHclTerraform;
class IamAccessBoundaryPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.IamAccessBoundaryPolicyTimeoutsOutputReference = IamAccessBoundaryPolicyTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
IamAccessBoundaryPolicyTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-google.iamAccessBoundaryPolicy.IamAccessBoundaryPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_access_boundary_policy google_iam_access_boundary_policy}
*/
class IamAccessBoundaryPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IamAccessBoundaryPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IamAccessBoundaryPolicy to import
    * @param importFromId The id of the existing IamAccessBoundaryPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_access_boundary_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IamAccessBoundaryPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_iam_access_boundary_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_access_boundary_policy google_iam_access_boundary_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamAccessBoundaryPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_iam_access_boundary_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // rules - computed: false, optional: false, required: true
        this._rules = new IamAccessBoundaryPolicyRulesList(this, "rules", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new IamAccessBoundaryPolicyTimeoutsOutputReference(this, "timeouts");
        this._displayName = config.displayName;
        this._id = config.id;
        this._name = config.name;
        this._parent = config.parent;
        this._rules.internalValue = config.rules;
        this._timeouts.internalValue = config.timeouts;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    get rules() {
        return this._rules;
    }
    putRules(value) {
        this._rules.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get rulesInput() {
        return this._rules.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            parent: cdktf.stringToTerraform(this._parent),
            rules: cdktf.listMapper(iamAccessBoundaryPolicyRulesToTerraform, true)(this._rules.internalValue),
            timeouts: iamAccessBoundaryPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent: {
                value: cdktf.stringToHclTerraform(this._parent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rules: {
                value: cdktf.listMapperHcl(iamAccessBoundaryPolicyRulesToHclTerraform, true)(this._rules.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IamAccessBoundaryPolicyRulesList",
            },
            timeouts: {
                value: iamAccessBoundaryPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "IamAccessBoundaryPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IamAccessBoundaryPolicy = IamAccessBoundaryPolicy;
_f = JSII_RTTI_SYMBOL_1;
IamAccessBoundaryPolicy[_f] = { fqn: "@cdktf/provider-google.iamAccessBoundaryPolicy.IamAccessBoundaryPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
IamAccessBoundaryPolicy.tfResourceType = "google_iam_access_boundary_policy";
