// @cdktf/provider-google/lib/network-connectivity-spoke/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkConnectivitySpoke = exports.NetworkConnectivitySpokeTimeoutsOutputReference = exports.networkConnectivitySpokeTimeoutsToHclTerraform = exports.networkConnectivitySpokeTimeoutsToTerraform = exports.NetworkConnectivitySpokeLinkedVpnTunnelsOutputReference = exports.networkConnectivitySpokeLinkedVpnTunnelsToHclTerraform = exports.networkConnectivitySpokeLinkedVpnTunnelsToTerraform = exports.NetworkConnectivitySpokeLinkedVpcNetworkOutputReference = exports.networkConnectivitySpokeLinkedVpcNetworkToHclTerraform = exports.networkConnectivitySpokeLinkedVpcNetworkToTerraform = exports.NetworkConnectivitySpokeLinkedRouterApplianceInstancesOutputReference = exports.networkConnectivitySpokeLinkedRouterApplianceInstancesToHclTerraform = exports.networkConnectivitySpokeLinkedRouterApplianceInstancesToTerraform = exports.NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesList = exports.NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesOutputReference = exports.networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToHclTerraform = exports.networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToTerraform = exports.NetworkConnectivitySpokeLinkedProducerVpcNetworkOutputReference = exports.networkConnectivitySpokeLinkedProducerVpcNetworkToHclTerraform = exports.networkConnectivitySpokeLinkedProducerVpcNetworkToTerraform = exports.NetworkConnectivitySpokeLinkedInterconnectAttachmentsOutputReference = exports.networkConnectivitySpokeLinkedInterconnectAttachmentsToHclTerraform = exports.networkConnectivitySpokeLinkedInterconnectAttachmentsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkConnectivitySpokeLinkedInterconnectAttachmentsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        include_import_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includeImportRanges),
        site_to_site_data_transfer: cdktf.booleanToTerraform(struct.siteToSiteDataTransfer),
        uris: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.uris),
    };
}
exports.networkConnectivitySpokeLinkedInterconnectAttachmentsToTerraform = networkConnectivitySpokeLinkedInterconnectAttachmentsToTerraform;
function networkConnectivitySpokeLinkedInterconnectAttachmentsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        include_import_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includeImportRanges),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        site_to_site_data_transfer: {
            value: cdktf.booleanToHclTerraform(struct.siteToSiteDataTransfer),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        uris: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.uris),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivitySpokeLinkedInterconnectAttachmentsToHclTerraform = networkConnectivitySpokeLinkedInterconnectAttachmentsToHclTerraform;
class NetworkConnectivitySpokeLinkedInterconnectAttachmentsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._includeImportRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeImportRanges = this._includeImportRanges;
        }
        if (this._siteToSiteDataTransfer !== undefined) {
            hasAnyValues = true;
            internalValueResult.siteToSiteDataTransfer = this._siteToSiteDataTransfer;
        }
        if (this._uris !== undefined) {
            hasAnyValues = true;
            internalValueResult.uris = this._uris;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._includeImportRanges = undefined;
            this._siteToSiteDataTransfer = undefined;
            this._uris = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._includeImportRanges = value.includeImportRanges;
            this._siteToSiteDataTransfer = value.siteToSiteDataTransfer;
            this._uris = value.uris;
        }
    }
    get includeImportRanges() {
        return this.getListAttribute('include_import_ranges');
    }
    set includeImportRanges(value) {
        this._includeImportRanges = value;
    }
    resetIncludeImportRanges() {
        this._includeImportRanges = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeImportRangesInput() {
        return this._includeImportRanges;
    }
    get siteToSiteDataTransfer() {
        return this.getBooleanAttribute('site_to_site_data_transfer');
    }
    set siteToSiteDataTransfer(value) {
        this._siteToSiteDataTransfer = value;
    }
    // Temporarily expose input value. Use with caution.
    get siteToSiteDataTransferInput() {
        return this._siteToSiteDataTransfer;
    }
    get uris() {
        return this.getListAttribute('uris');
    }
    set uris(value) {
        this._uris = value;
    }
    // Temporarily expose input value. Use with caution.
    get urisInput() {
        return this._uris;
    }
}
exports.NetworkConnectivitySpokeLinkedInterconnectAttachmentsOutputReference = NetworkConnectivitySpokeLinkedInterconnectAttachmentsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpokeLinkedInterconnectAttachmentsOutputReference[_a] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpokeLinkedInterconnectAttachmentsOutputReference", version: "14.13.1" };
function networkConnectivitySpokeLinkedProducerVpcNetworkToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exclude_export_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludeExportRanges),
        include_export_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includeExportRanges),
        network: cdktf.stringToTerraform(struct.network),
        peering: cdktf.stringToTerraform(struct.peering),
    };
}
exports.networkConnectivitySpokeLinkedProducerVpcNetworkToTerraform = networkConnectivitySpokeLinkedProducerVpcNetworkToTerraform;
function networkConnectivitySpokeLinkedProducerVpcNetworkToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        exclude_export_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludeExportRanges),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        include_export_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includeExportRanges),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        network: {
            value: cdktf.stringToHclTerraform(struct.network),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        peering: {
            value: cdktf.stringToHclTerraform(struct.peering),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivitySpokeLinkedProducerVpcNetworkToHclTerraform = networkConnectivitySpokeLinkedProducerVpcNetworkToHclTerraform;
class NetworkConnectivitySpokeLinkedProducerVpcNetworkOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludeExportRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludeExportRanges = this._excludeExportRanges;
        }
        if (this._includeExportRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeExportRanges = this._includeExportRanges;
        }
        if (this._network !== undefined) {
            hasAnyValues = true;
            internalValueResult.network = this._network;
        }
        if (this._peering !== undefined) {
            hasAnyValues = true;
            internalValueResult.peering = this._peering;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludeExportRanges = undefined;
            this._includeExportRanges = undefined;
            this._network = undefined;
            this._peering = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludeExportRanges = value.excludeExportRanges;
            this._includeExportRanges = value.includeExportRanges;
            this._network = value.network;
            this._peering = value.peering;
        }
    }
    get excludeExportRanges() {
        return this.getListAttribute('exclude_export_ranges');
    }
    set excludeExportRanges(value) {
        this._excludeExportRanges = value;
    }
    resetExcludeExportRanges() {
        this._excludeExportRanges = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludeExportRangesInput() {
        return this._excludeExportRanges;
    }
    get includeExportRanges() {
        return this.getListAttribute('include_export_ranges');
    }
    set includeExportRanges(value) {
        this._includeExportRanges = value;
    }
    resetIncludeExportRanges() {
        this._includeExportRanges = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeExportRangesInput() {
        return this._includeExportRanges;
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get peering() {
        return this.getStringAttribute('peering');
    }
    set peering(value) {
        this._peering = value;
    }
    // Temporarily expose input value. Use with caution.
    get peeringInput() {
        return this._peering;
    }
    // producer_network - computed: true, optional: false, required: false
    get producerNetwork() {
        return this.getStringAttribute('producer_network');
    }
}
exports.NetworkConnectivitySpokeLinkedProducerVpcNetworkOutputReference = NetworkConnectivitySpokeLinkedProducerVpcNetworkOutputReference;
_b = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpokeLinkedProducerVpcNetworkOutputReference[_b] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpokeLinkedProducerVpcNetworkOutputReference", version: "14.13.1" };
function networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ip_address: cdktf.stringToTerraform(struct.ipAddress),
        virtual_machine: cdktf.stringToTerraform(struct.virtualMachine),
    };
}
exports.networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToTerraform = networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToTerraform;
function networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ip_address: {
            value: cdktf.stringToHclTerraform(struct.ipAddress),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        virtual_machine: {
            value: cdktf.stringToHclTerraform(struct.virtualMachine),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToHclTerraform = networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToHclTerraform;
class NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ipAddress !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipAddress = this._ipAddress;
        }
        if (this._virtualMachine !== undefined) {
            hasAnyValues = true;
            internalValueResult.virtualMachine = this._virtualMachine;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._ipAddress = undefined;
            this._virtualMachine = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._ipAddress = value.ipAddress;
            this._virtualMachine = value.virtualMachine;
        }
    }
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    set ipAddress(value) {
        this._ipAddress = value;
    }
    // Temporarily expose input value. Use with caution.
    get ipAddressInput() {
        return this._ipAddress;
    }
    get virtualMachine() {
        return this.getStringAttribute('virtual_machine');
    }
    set virtualMachine(value) {
        this._virtualMachine = value;
    }
    // Temporarily expose input value. Use with caution.
    get virtualMachineInput() {
        return this._virtualMachine;
    }
}
exports.NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesOutputReference = NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesOutputReference[_c] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesOutputReference", version: "14.13.1" };
class NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesList = NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesList;
_d = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesList[_d] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesList", version: "14.13.1" };
function networkConnectivitySpokeLinkedRouterApplianceInstancesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        include_import_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includeImportRanges),
        site_to_site_data_transfer: cdktf.booleanToTerraform(struct.siteToSiteDataTransfer),
        instances: cdktf.listMapper(networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToTerraform, true)(struct.instances),
    };
}
exports.networkConnectivitySpokeLinkedRouterApplianceInstancesToTerraform = networkConnectivitySpokeLinkedRouterApplianceInstancesToTerraform;
function networkConnectivitySpokeLinkedRouterApplianceInstancesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        include_import_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includeImportRanges),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        site_to_site_data_transfer: {
            value: cdktf.booleanToHclTerraform(struct.siteToSiteDataTransfer),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        instances: {
            value: cdktf.listMapperHcl(networkConnectivitySpokeLinkedRouterApplianceInstancesInstancesToHclTerraform, true)(struct.instances),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivitySpokeLinkedRouterApplianceInstancesToHclTerraform = networkConnectivitySpokeLinkedRouterApplianceInstancesToHclTerraform;
class NetworkConnectivitySpokeLinkedRouterApplianceInstancesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // instances - computed: false, optional: false, required: true
        this._instances = new NetworkConnectivitySpokeLinkedRouterApplianceInstancesInstancesList(this, "instances", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._includeImportRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeImportRanges = this._includeImportRanges;
        }
        if (this._siteToSiteDataTransfer !== undefined) {
            hasAnyValues = true;
            internalValueResult.siteToSiteDataTransfer = this._siteToSiteDataTransfer;
        }
        if (this._instances?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.instances = this._instances?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._includeImportRanges = undefined;
            this._siteToSiteDataTransfer = undefined;
            this._instances.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._includeImportRanges = value.includeImportRanges;
            this._siteToSiteDataTransfer = value.siteToSiteDataTransfer;
            this._instances.internalValue = value.instances;
        }
    }
    get includeImportRanges() {
        return this.getListAttribute('include_import_ranges');
    }
    set includeImportRanges(value) {
        this._includeImportRanges = value;
    }
    resetIncludeImportRanges() {
        this._includeImportRanges = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeImportRangesInput() {
        return this._includeImportRanges;
    }
    get siteToSiteDataTransfer() {
        return this.getBooleanAttribute('site_to_site_data_transfer');
    }
    set siteToSiteDataTransfer(value) {
        this._siteToSiteDataTransfer = value;
    }
    // Temporarily expose input value. Use with caution.
    get siteToSiteDataTransferInput() {
        return this._siteToSiteDataTransfer;
    }
    get instances() {
        return this._instances;
    }
    putInstances(value) {
        this._instances.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get instancesInput() {
        return this._instances.internalValue;
    }
}
exports.NetworkConnectivitySpokeLinkedRouterApplianceInstancesOutputReference = NetworkConnectivitySpokeLinkedRouterApplianceInstancesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpokeLinkedRouterApplianceInstancesOutputReference[_e] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpokeLinkedRouterApplianceInstancesOutputReference", version: "14.13.1" };
function networkConnectivitySpokeLinkedVpcNetworkToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exclude_export_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludeExportRanges),
        include_export_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includeExportRanges),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.networkConnectivitySpokeLinkedVpcNetworkToTerraform = networkConnectivitySpokeLinkedVpcNetworkToTerraform;
function networkConnectivitySpokeLinkedVpcNetworkToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        exclude_export_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludeExportRanges),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        include_export_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includeExportRanges),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivitySpokeLinkedVpcNetworkToHclTerraform = networkConnectivitySpokeLinkedVpcNetworkToHclTerraform;
class NetworkConnectivitySpokeLinkedVpcNetworkOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludeExportRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludeExportRanges = this._excludeExportRanges;
        }
        if (this._includeExportRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeExportRanges = this._includeExportRanges;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludeExportRanges = undefined;
            this._includeExportRanges = undefined;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludeExportRanges = value.excludeExportRanges;
            this._includeExportRanges = value.includeExportRanges;
            this._uri = value.uri;
        }
    }
    get excludeExportRanges() {
        return this.getListAttribute('exclude_export_ranges');
    }
    set excludeExportRanges(value) {
        this._excludeExportRanges = value;
    }
    resetExcludeExportRanges() {
        this._excludeExportRanges = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludeExportRangesInput() {
        return this._excludeExportRanges;
    }
    get includeExportRanges() {
        return this.getListAttribute('include_export_ranges');
    }
    set includeExportRanges(value) {
        this._includeExportRanges = value;
    }
    resetIncludeExportRanges() {
        this._includeExportRanges = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeExportRangesInput() {
        return this._includeExportRanges;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.NetworkConnectivitySpokeLinkedVpcNetworkOutputReference = NetworkConnectivitySpokeLinkedVpcNetworkOutputReference;
_f = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpokeLinkedVpcNetworkOutputReference[_f] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpokeLinkedVpcNetworkOutputReference", version: "14.13.1" };
function networkConnectivitySpokeLinkedVpnTunnelsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        include_import_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includeImportRanges),
        site_to_site_data_transfer: cdktf.booleanToTerraform(struct.siteToSiteDataTransfer),
        uris: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.uris),
    };
}
exports.networkConnectivitySpokeLinkedVpnTunnelsToTerraform = networkConnectivitySpokeLinkedVpnTunnelsToTerraform;
function networkConnectivitySpokeLinkedVpnTunnelsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        include_import_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includeImportRanges),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        site_to_site_data_transfer: {
            value: cdktf.booleanToHclTerraform(struct.siteToSiteDataTransfer),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        uris: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.uris),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivitySpokeLinkedVpnTunnelsToHclTerraform = networkConnectivitySpokeLinkedVpnTunnelsToHclTerraform;
class NetworkConnectivitySpokeLinkedVpnTunnelsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._includeImportRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeImportRanges = this._includeImportRanges;
        }
        if (this._siteToSiteDataTransfer !== undefined) {
            hasAnyValues = true;
            internalValueResult.siteToSiteDataTransfer = this._siteToSiteDataTransfer;
        }
        if (this._uris !== undefined) {
            hasAnyValues = true;
            internalValueResult.uris = this._uris;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._includeImportRanges = undefined;
            this._siteToSiteDataTransfer = undefined;
            this._uris = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._includeImportRanges = value.includeImportRanges;
            this._siteToSiteDataTransfer = value.siteToSiteDataTransfer;
            this._uris = value.uris;
        }
    }
    get includeImportRanges() {
        return this.getListAttribute('include_import_ranges');
    }
    set includeImportRanges(value) {
        this._includeImportRanges = value;
    }
    resetIncludeImportRanges() {
        this._includeImportRanges = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeImportRangesInput() {
        return this._includeImportRanges;
    }
    get siteToSiteDataTransfer() {
        return this.getBooleanAttribute('site_to_site_data_transfer');
    }
    set siteToSiteDataTransfer(value) {
        this._siteToSiteDataTransfer = value;
    }
    // Temporarily expose input value. Use with caution.
    get siteToSiteDataTransferInput() {
        return this._siteToSiteDataTransfer;
    }
    get uris() {
        return this.getListAttribute('uris');
    }
    set uris(value) {
        this._uris = value;
    }
    // Temporarily expose input value. Use with caution.
    get urisInput() {
        return this._uris;
    }
}
exports.NetworkConnectivitySpokeLinkedVpnTunnelsOutputReference = NetworkConnectivitySpokeLinkedVpnTunnelsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpokeLinkedVpnTunnelsOutputReference[_g] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpokeLinkedVpnTunnelsOutputReference", version: "14.13.1" };
function networkConnectivitySpokeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkConnectivitySpokeTimeoutsToTerraform = networkConnectivitySpokeTimeoutsToTerraform;
function networkConnectivitySpokeTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivitySpokeTimeoutsToHclTerraform = networkConnectivitySpokeTimeoutsToHclTerraform;
class NetworkConnectivitySpokeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkConnectivitySpokeTimeoutsOutputReference = NetworkConnectivitySpokeTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpokeTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpokeTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_spoke google_network_connectivity_spoke}
*/
class NetworkConnectivitySpoke extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkConnectivitySpoke resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkConnectivitySpoke to import
    * @param importFromId The id of the existing NetworkConnectivitySpoke that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_spoke#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkConnectivitySpoke to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_network_connectivity_spoke", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_spoke google_network_connectivity_spoke} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkConnectivitySpokeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_network_connectivity_spoke',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // linked_interconnect_attachments - computed: false, optional: true, required: false
        this._linkedInterconnectAttachments = new NetworkConnectivitySpokeLinkedInterconnectAttachmentsOutputReference(this, "linked_interconnect_attachments");
        // linked_producer_vpc_network - computed: false, optional: true, required: false
        this._linkedProducerVpcNetwork = new NetworkConnectivitySpokeLinkedProducerVpcNetworkOutputReference(this, "linked_producer_vpc_network");
        // linked_router_appliance_instances - computed: false, optional: true, required: false
        this._linkedRouterApplianceInstances = new NetworkConnectivitySpokeLinkedRouterApplianceInstancesOutputReference(this, "linked_router_appliance_instances");
        // linked_vpc_network - computed: false, optional: true, required: false
        this._linkedVpcNetwork = new NetworkConnectivitySpokeLinkedVpcNetworkOutputReference(this, "linked_vpc_network");
        // linked_vpn_tunnels - computed: false, optional: true, required: false
        this._linkedVpnTunnels = new NetworkConnectivitySpokeLinkedVpnTunnelsOutputReference(this, "linked_vpn_tunnels");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkConnectivitySpokeTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._hub = config.hub;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._linkedInterconnectAttachments.internalValue = config.linkedInterconnectAttachments;
        this._linkedProducerVpcNetwork.internalValue = config.linkedProducerVpcNetwork;
        this._linkedRouterApplianceInstances.internalValue = config.linkedRouterApplianceInstances;
        this._linkedVpcNetwork.internalValue = config.linkedVpcNetwork;
        this._linkedVpnTunnels.internalValue = config.linkedVpnTunnels;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get hub() {
        return this.getStringAttribute('hub');
    }
    set hub(value) {
        this._hub = value;
    }
    // Temporarily expose input value. Use with caution.
    get hubInput() {
        return this._hub;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // unique_id - computed: true, optional: false, required: false
    get uniqueId() {
        return this.getStringAttribute('unique_id');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get linkedInterconnectAttachments() {
        return this._linkedInterconnectAttachments;
    }
    putLinkedInterconnectAttachments(value) {
        this._linkedInterconnectAttachments.internalValue = value;
    }
    resetLinkedInterconnectAttachments() {
        this._linkedInterconnectAttachments.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get linkedInterconnectAttachmentsInput() {
        return this._linkedInterconnectAttachments.internalValue;
    }
    get linkedProducerVpcNetwork() {
        return this._linkedProducerVpcNetwork;
    }
    putLinkedProducerVpcNetwork(value) {
        this._linkedProducerVpcNetwork.internalValue = value;
    }
    resetLinkedProducerVpcNetwork() {
        this._linkedProducerVpcNetwork.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get linkedProducerVpcNetworkInput() {
        return this._linkedProducerVpcNetwork.internalValue;
    }
    get linkedRouterApplianceInstances() {
        return this._linkedRouterApplianceInstances;
    }
    putLinkedRouterApplianceInstances(value) {
        this._linkedRouterApplianceInstances.internalValue = value;
    }
    resetLinkedRouterApplianceInstances() {
        this._linkedRouterApplianceInstances.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get linkedRouterApplianceInstancesInput() {
        return this._linkedRouterApplianceInstances.internalValue;
    }
    get linkedVpcNetwork() {
        return this._linkedVpcNetwork;
    }
    putLinkedVpcNetwork(value) {
        this._linkedVpcNetwork.internalValue = value;
    }
    resetLinkedVpcNetwork() {
        this._linkedVpcNetwork.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get linkedVpcNetworkInput() {
        return this._linkedVpcNetwork.internalValue;
    }
    get linkedVpnTunnels() {
        return this._linkedVpnTunnels;
    }
    putLinkedVpnTunnels(value) {
        this._linkedVpnTunnels.internalValue = value;
    }
    resetLinkedVpnTunnels() {
        this._linkedVpnTunnels.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get linkedVpnTunnelsInput() {
        return this._linkedVpnTunnels.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            hub: cdktf.stringToTerraform(this._hub),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            linked_interconnect_attachments: networkConnectivitySpokeLinkedInterconnectAttachmentsToTerraform(this._linkedInterconnectAttachments.internalValue),
            linked_producer_vpc_network: networkConnectivitySpokeLinkedProducerVpcNetworkToTerraform(this._linkedProducerVpcNetwork.internalValue),
            linked_router_appliance_instances: networkConnectivitySpokeLinkedRouterApplianceInstancesToTerraform(this._linkedRouterApplianceInstances.internalValue),
            linked_vpc_network: networkConnectivitySpokeLinkedVpcNetworkToTerraform(this._linkedVpcNetwork.internalValue),
            linked_vpn_tunnels: networkConnectivitySpokeLinkedVpnTunnelsToTerraform(this._linkedVpnTunnels.internalValue),
            timeouts: networkConnectivitySpokeTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            hub: {
                value: cdktf.stringToHclTerraform(this._hub),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            linked_interconnect_attachments: {
                value: networkConnectivitySpokeLinkedInterconnectAttachmentsToHclTerraform(this._linkedInterconnectAttachments.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivitySpokeLinkedInterconnectAttachmentsList",
            },
            linked_producer_vpc_network: {
                value: networkConnectivitySpokeLinkedProducerVpcNetworkToHclTerraform(this._linkedProducerVpcNetwork.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivitySpokeLinkedProducerVpcNetworkList",
            },
            linked_router_appliance_instances: {
                value: networkConnectivitySpokeLinkedRouterApplianceInstancesToHclTerraform(this._linkedRouterApplianceInstances.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivitySpokeLinkedRouterApplianceInstancesList",
            },
            linked_vpc_network: {
                value: networkConnectivitySpokeLinkedVpcNetworkToHclTerraform(this._linkedVpcNetwork.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivitySpokeLinkedVpcNetworkList",
            },
            linked_vpn_tunnels: {
                value: networkConnectivitySpokeLinkedVpnTunnelsToHclTerraform(this._linkedVpnTunnels.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivitySpokeLinkedVpnTunnelsList",
            },
            timeouts: {
                value: networkConnectivitySpokeTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkConnectivitySpokeTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkConnectivitySpoke = NetworkConnectivitySpoke;
_j = JSII_RTTI_SYMBOL_1;
NetworkConnectivitySpoke[_j] = { fqn: "@cdktf/provider-google.networkConnectivitySpoke.NetworkConnectivitySpoke", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetworkConnectivitySpoke.tfResourceType = "google_network_connectivity_spoke";
