// @cdktf/provider-google/lib/transcoder-job-template/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18;
Object.defineProperty(exports, "__esModule", { value: true });
exports.transcoderJobTemplateConfigEncryptionsSampleAesToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsSampleAesToTerraform = exports.TranscoderJobTemplateConfigEncryptionsMpegCencOutputReference = exports.transcoderJobTemplateConfigEncryptionsMpegCencToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsMpegCencToTerraform = exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsOutputReference = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsToTerraform = exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsWidevineOutputReference = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToTerraform = exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyOutputReference = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToTerraform = exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsFairplayOutputReference = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToTerraform = exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsClearkeyOutputReference = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToTerraform = exports.TranscoderJobTemplateConfigEncryptionsAes128OutputReference = exports.transcoderJobTemplateConfigEncryptionsAes128ToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsAes128ToTerraform = exports.TranscoderJobTemplateConfigElementaryStreamsList = exports.TranscoderJobTemplateConfigElementaryStreamsOutputReference = exports.transcoderJobTemplateConfigElementaryStreamsToHclTerraform = exports.transcoderJobTemplateConfigElementaryStreamsToTerraform = exports.TranscoderJobTemplateConfigElementaryStreamsVideoStreamOutputReference = exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamToHclTerraform = exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamToTerraform = exports.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264OutputReference = exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToHclTerraform = exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToTerraform = exports.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrOutputReference = exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToHclTerraform = exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToTerraform = exports.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgOutputReference = exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToHclTerraform = exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToTerraform = exports.TranscoderJobTemplateConfigElementaryStreamsAudioStreamOutputReference = exports.transcoderJobTemplateConfigElementaryStreamsAudioStreamToHclTerraform = exports.transcoderJobTemplateConfigElementaryStreamsAudioStreamToTerraform = exports.TranscoderJobTemplateConfigEditListStructList = exports.TranscoderJobTemplateConfigEditListStructOutputReference = exports.transcoderJobTemplateConfigEditListStructToHclTerraform = exports.transcoderJobTemplateConfigEditListStructToTerraform = exports.TranscoderJobTemplateConfigAdBreaksList = exports.TranscoderJobTemplateConfigAdBreaksOutputReference = exports.transcoderJobTemplateConfigAdBreaksToHclTerraform = exports.transcoderJobTemplateConfigAdBreaksToTerraform = void 0;
exports.transcoderJobTemplateTimeoutsToTerraform = exports.TranscoderJobTemplateConfigAOutputReference = exports.transcoderJobTemplateConfigAToHclTerraform = exports.transcoderJobTemplateConfigAToTerraform = exports.TranscoderJobTemplateConfigPubsubDestinationOutputReference = exports.transcoderJobTemplateConfigPubsubDestinationToHclTerraform = exports.transcoderJobTemplateConfigPubsubDestinationToTerraform = exports.TranscoderJobTemplateConfigOverlaysList = exports.TranscoderJobTemplateConfigOverlaysOutputReference = exports.transcoderJobTemplateConfigOverlaysToHclTerraform = exports.transcoderJobTemplateConfigOverlaysToTerraform = exports.TranscoderJobTemplateConfigOverlaysImageOutputReference = exports.transcoderJobTemplateConfigOverlaysImageToHclTerraform = exports.transcoderJobTemplateConfigOverlaysImageToTerraform = exports.TranscoderJobTemplateConfigOverlaysAnimationsList = exports.TranscoderJobTemplateConfigOverlaysAnimationsOutputReference = exports.transcoderJobTemplateConfigOverlaysAnimationsToHclTerraform = exports.transcoderJobTemplateConfigOverlaysAnimationsToTerraform = exports.TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeOutputReference = exports.transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToHclTerraform = exports.transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToTerraform = exports.TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyOutputReference = exports.transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToHclTerraform = exports.transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToTerraform = exports.TranscoderJobTemplateConfigOutputOutputReference = exports.transcoderJobTemplateConfigOutputToHclTerraform = exports.transcoderJobTemplateConfigOutputToTerraform = exports.TranscoderJobTemplateConfigMuxStreamsList = exports.TranscoderJobTemplateConfigMuxStreamsOutputReference = exports.transcoderJobTemplateConfigMuxStreamsToHclTerraform = exports.transcoderJobTemplateConfigMuxStreamsToTerraform = exports.TranscoderJobTemplateConfigMuxStreamsSegmentSettingsOutputReference = exports.transcoderJobTemplateConfigMuxStreamsSegmentSettingsToHclTerraform = exports.transcoderJobTemplateConfigMuxStreamsSegmentSettingsToTerraform = exports.TranscoderJobTemplateConfigManifestsList = exports.TranscoderJobTemplateConfigManifestsOutputReference = exports.transcoderJobTemplateConfigManifestsToHclTerraform = exports.transcoderJobTemplateConfigManifestsToTerraform = exports.TranscoderJobTemplateConfigInputsList = exports.TranscoderJobTemplateConfigInputsOutputReference = exports.transcoderJobTemplateConfigInputsToHclTerraform = exports.transcoderJobTemplateConfigInputsToTerraform = exports.TranscoderJobTemplateConfigEncryptionsList = exports.TranscoderJobTemplateConfigEncryptionsOutputReference = exports.transcoderJobTemplateConfigEncryptionsToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsToTerraform = exports.TranscoderJobTemplateConfigEncryptionsSecretManagerKeySourceOutputReference = exports.transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToHclTerraform = exports.transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToTerraform = exports.TranscoderJobTemplateConfigEncryptionsSampleAesOutputReference = void 0;
exports.TranscoderJobTemplate = exports.TranscoderJobTemplateTimeoutsOutputReference = exports.transcoderJobTemplateTimeoutsToHclTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function transcoderJobTemplateConfigAdBreaksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        start_time_offset: cdktf.stringToTerraform(struct.startTimeOffset),
    };
}
exports.transcoderJobTemplateConfigAdBreaksToTerraform = transcoderJobTemplateConfigAdBreaksToTerraform;
function transcoderJobTemplateConfigAdBreaksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        start_time_offset: {
            value: cdktf.stringToHclTerraform(struct.startTimeOffset),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigAdBreaksToHclTerraform = transcoderJobTemplateConfigAdBreaksToHclTerraform;
class TranscoderJobTemplateConfigAdBreaksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._startTimeOffset !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTimeOffset = this._startTimeOffset;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._startTimeOffset = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._startTimeOffset = value.startTimeOffset;
        }
    }
    get startTimeOffset() {
        return this.getStringAttribute('start_time_offset');
    }
    set startTimeOffset(value) {
        this._startTimeOffset = value;
    }
    resetStartTimeOffset() {
        this._startTimeOffset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeOffsetInput() {
        return this._startTimeOffset;
    }
}
exports.TranscoderJobTemplateConfigAdBreaksOutputReference = TranscoderJobTemplateConfigAdBreaksOutputReference;
_a = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigAdBreaksOutputReference[_a] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigAdBreaksOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigAdBreaksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigAdBreaksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigAdBreaksList = TranscoderJobTemplateConfigAdBreaksList;
_b = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigAdBreaksList[_b] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigAdBreaksList", version: "14.13.1" };
function transcoderJobTemplateConfigEditListStructToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        inputs: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.inputs),
        key: cdktf.stringToTerraform(struct.key),
        start_time_offset: cdktf.stringToTerraform(struct.startTimeOffset),
    };
}
exports.transcoderJobTemplateConfigEditListStructToTerraform = transcoderJobTemplateConfigEditListStructToTerraform;
function transcoderJobTemplateConfigEditListStructToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        inputs: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.inputs),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start_time_offset: {
            value: cdktf.stringToHclTerraform(struct.startTimeOffset),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigEditListStructToHclTerraform = transcoderJobTemplateConfigEditListStructToHclTerraform;
class TranscoderJobTemplateConfigEditListStructOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._inputs !== undefined) {
            hasAnyValues = true;
            internalValueResult.inputs = this._inputs;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._startTimeOffset !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTimeOffset = this._startTimeOffset;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._inputs = undefined;
            this._key = undefined;
            this._startTimeOffset = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._inputs = value.inputs;
            this._key = value.key;
            this._startTimeOffset = value.startTimeOffset;
        }
    }
    get inputs() {
        return this.getListAttribute('inputs');
    }
    set inputs(value) {
        this._inputs = value;
    }
    resetInputs() {
        this._inputs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputsInput() {
        return this._inputs;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get startTimeOffset() {
        return this.getStringAttribute('start_time_offset');
    }
    set startTimeOffset(value) {
        this._startTimeOffset = value;
    }
    resetStartTimeOffset() {
        this._startTimeOffset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeOffsetInput() {
        return this._startTimeOffset;
    }
}
exports.TranscoderJobTemplateConfigEditListStructOutputReference = TranscoderJobTemplateConfigEditListStructOutputReference;
_c = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEditListStructOutputReference[_c] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEditListStructOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigEditListStructList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigEditListStructOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigEditListStructList = TranscoderJobTemplateConfigEditListStructList;
_d = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEditListStructList[_d] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEditListStructList", version: "14.13.1" };
function transcoderJobTemplateConfigElementaryStreamsAudioStreamToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bitrate_bps: cdktf.numberToTerraform(struct.bitrateBps),
        channel_count: cdktf.numberToTerraform(struct.channelCount),
        channel_layout: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.channelLayout),
        codec: cdktf.stringToTerraform(struct.codec),
        sample_rate_hertz: cdktf.numberToTerraform(struct.sampleRateHertz),
    };
}
exports.transcoderJobTemplateConfigElementaryStreamsAudioStreamToTerraform = transcoderJobTemplateConfigElementaryStreamsAudioStreamToTerraform;
function transcoderJobTemplateConfigElementaryStreamsAudioStreamToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bitrate_bps: {
            value: cdktf.numberToHclTerraform(struct.bitrateBps),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        channel_count: {
            value: cdktf.numberToHclTerraform(struct.channelCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        channel_layout: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.channelLayout),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        codec: {
            value: cdktf.stringToHclTerraform(struct.codec),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        sample_rate_hertz: {
            value: cdktf.numberToHclTerraform(struct.sampleRateHertz),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigElementaryStreamsAudioStreamToHclTerraform = transcoderJobTemplateConfigElementaryStreamsAudioStreamToHclTerraform;
class TranscoderJobTemplateConfigElementaryStreamsAudioStreamOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bitrateBps !== undefined) {
            hasAnyValues = true;
            internalValueResult.bitrateBps = this._bitrateBps;
        }
        if (this._channelCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.channelCount = this._channelCount;
        }
        if (this._channelLayout !== undefined) {
            hasAnyValues = true;
            internalValueResult.channelLayout = this._channelLayout;
        }
        if (this._codec !== undefined) {
            hasAnyValues = true;
            internalValueResult.codec = this._codec;
        }
        if (this._sampleRateHertz !== undefined) {
            hasAnyValues = true;
            internalValueResult.sampleRateHertz = this._sampleRateHertz;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bitrateBps = undefined;
            this._channelCount = undefined;
            this._channelLayout = undefined;
            this._codec = undefined;
            this._sampleRateHertz = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bitrateBps = value.bitrateBps;
            this._channelCount = value.channelCount;
            this._channelLayout = value.channelLayout;
            this._codec = value.codec;
            this._sampleRateHertz = value.sampleRateHertz;
        }
    }
    get bitrateBps() {
        return this.getNumberAttribute('bitrate_bps');
    }
    set bitrateBps(value) {
        this._bitrateBps = value;
    }
    // Temporarily expose input value. Use with caution.
    get bitrateBpsInput() {
        return this._bitrateBps;
    }
    get channelCount() {
        return this.getNumberAttribute('channel_count');
    }
    set channelCount(value) {
        this._channelCount = value;
    }
    resetChannelCount() {
        this._channelCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get channelCountInput() {
        return this._channelCount;
    }
    get channelLayout() {
        return this.getListAttribute('channel_layout');
    }
    set channelLayout(value) {
        this._channelLayout = value;
    }
    resetChannelLayout() {
        this._channelLayout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get channelLayoutInput() {
        return this._channelLayout;
    }
    get codec() {
        return this.getStringAttribute('codec');
    }
    set codec(value) {
        this._codec = value;
    }
    resetCodec() {
        this._codec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get codecInput() {
        return this._codec;
    }
    get sampleRateHertz() {
        return this.getNumberAttribute('sample_rate_hertz');
    }
    set sampleRateHertz(value) {
        this._sampleRateHertz = value;
    }
    resetSampleRateHertz() {
        this._sampleRateHertz = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sampleRateHertzInput() {
        return this._sampleRateHertz;
    }
}
exports.TranscoderJobTemplateConfigElementaryStreamsAudioStreamOutputReference = TranscoderJobTemplateConfigElementaryStreamsAudioStreamOutputReference;
_e = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigElementaryStreamsAudioStreamOutputReference[_e] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigElementaryStreamsAudioStreamOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToTerraform = transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToTerraform;
function transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToHclTerraform = transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToHclTerraform;
class TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgOutputReference = TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgOutputReference;
_f = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgOutputReference[_f] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToTerraform = transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToTerraform;
function transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToHclTerraform = transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToHclTerraform;
class TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrOutputReference = TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrOutputReference;
_g = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrOutputReference[_g] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bitrate_bps: cdktf.numberToTerraform(struct.bitrateBps),
        crf_level: cdktf.numberToTerraform(struct.crfLevel),
        entropy_coder: cdktf.stringToTerraform(struct.entropyCoder),
        frame_rate: cdktf.numberToTerraform(struct.frameRate),
        gop_duration: cdktf.stringToTerraform(struct.gopDuration),
        height_pixels: cdktf.numberToTerraform(struct.heightPixels),
        pixel_format: cdktf.stringToTerraform(struct.pixelFormat),
        preset: cdktf.stringToTerraform(struct.preset),
        profile: cdktf.stringToTerraform(struct.profile),
        rate_control_mode: cdktf.stringToTerraform(struct.rateControlMode),
        vbv_fullness_bits: cdktf.numberToTerraform(struct.vbvFullnessBits),
        vbv_size_bits: cdktf.numberToTerraform(struct.vbvSizeBits),
        width_pixels: cdktf.numberToTerraform(struct.widthPixels),
        hlg: transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToTerraform(struct.hlg),
        sdr: transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToTerraform(struct.sdr),
    };
}
exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToTerraform = transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToTerraform;
function transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bitrate_bps: {
            value: cdktf.numberToHclTerraform(struct.bitrateBps),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        crf_level: {
            value: cdktf.numberToHclTerraform(struct.crfLevel),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        entropy_coder: {
            value: cdktf.stringToHclTerraform(struct.entropyCoder),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        frame_rate: {
            value: cdktf.numberToHclTerraform(struct.frameRate),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        gop_duration: {
            value: cdktf.stringToHclTerraform(struct.gopDuration),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        height_pixels: {
            value: cdktf.numberToHclTerraform(struct.heightPixels),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        pixel_format: {
            value: cdktf.stringToHclTerraform(struct.pixelFormat),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        preset: {
            value: cdktf.stringToHclTerraform(struct.preset),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        profile: {
            value: cdktf.stringToHclTerraform(struct.profile),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rate_control_mode: {
            value: cdktf.stringToHclTerraform(struct.rateControlMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        vbv_fullness_bits: {
            value: cdktf.numberToHclTerraform(struct.vbvFullnessBits),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        vbv_size_bits: {
            value: cdktf.numberToHclTerraform(struct.vbvSizeBits),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        width_pixels: {
            value: cdktf.numberToHclTerraform(struct.widthPixels),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        hlg: {
            value: transcoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgToHclTerraform(struct.hlg),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgList",
        },
        sdr: {
            value: transcoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrToHclTerraform(struct.sdr),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToHclTerraform = transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToHclTerraform;
class TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // hlg - computed: false, optional: true, required: false
        this._hlg = new TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264HlgOutputReference(this, "hlg");
        // sdr - computed: false, optional: true, required: false
        this._sdr = new TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264SdrOutputReference(this, "sdr");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bitrateBps !== undefined) {
            hasAnyValues = true;
            internalValueResult.bitrateBps = this._bitrateBps;
        }
        if (this._crfLevel !== undefined) {
            hasAnyValues = true;
            internalValueResult.crfLevel = this._crfLevel;
        }
        if (this._entropyCoder !== undefined) {
            hasAnyValues = true;
            internalValueResult.entropyCoder = this._entropyCoder;
        }
        if (this._frameRate !== undefined) {
            hasAnyValues = true;
            internalValueResult.frameRate = this._frameRate;
        }
        if (this._gopDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.gopDuration = this._gopDuration;
        }
        if (this._heightPixels !== undefined) {
            hasAnyValues = true;
            internalValueResult.heightPixels = this._heightPixels;
        }
        if (this._pixelFormat !== undefined) {
            hasAnyValues = true;
            internalValueResult.pixelFormat = this._pixelFormat;
        }
        if (this._preset !== undefined) {
            hasAnyValues = true;
            internalValueResult.preset = this._preset;
        }
        if (this._profile !== undefined) {
            hasAnyValues = true;
            internalValueResult.profile = this._profile;
        }
        if (this._rateControlMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.rateControlMode = this._rateControlMode;
        }
        if (this._vbvFullnessBits !== undefined) {
            hasAnyValues = true;
            internalValueResult.vbvFullnessBits = this._vbvFullnessBits;
        }
        if (this._vbvSizeBits !== undefined) {
            hasAnyValues = true;
            internalValueResult.vbvSizeBits = this._vbvSizeBits;
        }
        if (this._widthPixels !== undefined) {
            hasAnyValues = true;
            internalValueResult.widthPixels = this._widthPixels;
        }
        if (this._hlg?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hlg = this._hlg?.internalValue;
        }
        if (this._sdr?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sdr = this._sdr?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bitrateBps = undefined;
            this._crfLevel = undefined;
            this._entropyCoder = undefined;
            this._frameRate = undefined;
            this._gopDuration = undefined;
            this._heightPixels = undefined;
            this._pixelFormat = undefined;
            this._preset = undefined;
            this._profile = undefined;
            this._rateControlMode = undefined;
            this._vbvFullnessBits = undefined;
            this._vbvSizeBits = undefined;
            this._widthPixels = undefined;
            this._hlg.internalValue = undefined;
            this._sdr.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bitrateBps = value.bitrateBps;
            this._crfLevel = value.crfLevel;
            this._entropyCoder = value.entropyCoder;
            this._frameRate = value.frameRate;
            this._gopDuration = value.gopDuration;
            this._heightPixels = value.heightPixels;
            this._pixelFormat = value.pixelFormat;
            this._preset = value.preset;
            this._profile = value.profile;
            this._rateControlMode = value.rateControlMode;
            this._vbvFullnessBits = value.vbvFullnessBits;
            this._vbvSizeBits = value.vbvSizeBits;
            this._widthPixels = value.widthPixels;
            this._hlg.internalValue = value.hlg;
            this._sdr.internalValue = value.sdr;
        }
    }
    get bitrateBps() {
        return this.getNumberAttribute('bitrate_bps');
    }
    set bitrateBps(value) {
        this._bitrateBps = value;
    }
    // Temporarily expose input value. Use with caution.
    get bitrateBpsInput() {
        return this._bitrateBps;
    }
    get crfLevel() {
        return this.getNumberAttribute('crf_level');
    }
    set crfLevel(value) {
        this._crfLevel = value;
    }
    resetCrfLevel() {
        this._crfLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get crfLevelInput() {
        return this._crfLevel;
    }
    get entropyCoder() {
        return this.getStringAttribute('entropy_coder');
    }
    set entropyCoder(value) {
        this._entropyCoder = value;
    }
    resetEntropyCoder() {
        this._entropyCoder = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get entropyCoderInput() {
        return this._entropyCoder;
    }
    get frameRate() {
        return this.getNumberAttribute('frame_rate');
    }
    set frameRate(value) {
        this._frameRate = value;
    }
    // Temporarily expose input value. Use with caution.
    get frameRateInput() {
        return this._frameRate;
    }
    get gopDuration() {
        return this.getStringAttribute('gop_duration');
    }
    set gopDuration(value) {
        this._gopDuration = value;
    }
    resetGopDuration() {
        this._gopDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gopDurationInput() {
        return this._gopDuration;
    }
    get heightPixels() {
        return this.getNumberAttribute('height_pixels');
    }
    set heightPixels(value) {
        this._heightPixels = value;
    }
    resetHeightPixels() {
        this._heightPixels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get heightPixelsInput() {
        return this._heightPixels;
    }
    get pixelFormat() {
        return this.getStringAttribute('pixel_format');
    }
    set pixelFormat(value) {
        this._pixelFormat = value;
    }
    resetPixelFormat() {
        this._pixelFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pixelFormatInput() {
        return this._pixelFormat;
    }
    get preset() {
        return this.getStringAttribute('preset');
    }
    set preset(value) {
        this._preset = value;
    }
    resetPreset() {
        this._preset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get presetInput() {
        return this._preset;
    }
    get profile() {
        return this.getStringAttribute('profile');
    }
    set profile(value) {
        this._profile = value;
    }
    resetProfile() {
        this._profile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get profileInput() {
        return this._profile;
    }
    get rateControlMode() {
        return this.getStringAttribute('rate_control_mode');
    }
    set rateControlMode(value) {
        this._rateControlMode = value;
    }
    resetRateControlMode() {
        this._rateControlMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rateControlModeInput() {
        return this._rateControlMode;
    }
    get vbvFullnessBits() {
        return this.getNumberAttribute('vbv_fullness_bits');
    }
    set vbvFullnessBits(value) {
        this._vbvFullnessBits = value;
    }
    resetVbvFullnessBits() {
        this._vbvFullnessBits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vbvFullnessBitsInput() {
        return this._vbvFullnessBits;
    }
    get vbvSizeBits() {
        return this.getNumberAttribute('vbv_size_bits');
    }
    set vbvSizeBits(value) {
        this._vbvSizeBits = value;
    }
    resetVbvSizeBits() {
        this._vbvSizeBits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vbvSizeBitsInput() {
        return this._vbvSizeBits;
    }
    get widthPixels() {
        return this.getNumberAttribute('width_pixels');
    }
    set widthPixels(value) {
        this._widthPixels = value;
    }
    resetWidthPixels() {
        this._widthPixels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get widthPixelsInput() {
        return this._widthPixels;
    }
    get hlg() {
        return this._hlg;
    }
    putHlg(value) {
        this._hlg.internalValue = value;
    }
    resetHlg() {
        this._hlg.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hlgInput() {
        return this._hlg.internalValue;
    }
    get sdr() {
        return this._sdr;
    }
    putSdr(value) {
        this._sdr.internalValue = value;
    }
    resetSdr() {
        this._sdr.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sdrInput() {
        return this._sdr.internalValue;
    }
}
exports.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264OutputReference = TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264OutputReference;
_h = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264OutputReference[_h] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264OutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigElementaryStreamsVideoStreamToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        h264: transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToTerraform(struct.h264),
    };
}
exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamToTerraform = transcoderJobTemplateConfigElementaryStreamsVideoStreamToTerraform;
function transcoderJobTemplateConfigElementaryStreamsVideoStreamToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        h264: {
            value: transcoderJobTemplateConfigElementaryStreamsVideoStreamH264ToHclTerraform(struct.h264),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264List",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigElementaryStreamsVideoStreamToHclTerraform = transcoderJobTemplateConfigElementaryStreamsVideoStreamToHclTerraform;
class TranscoderJobTemplateConfigElementaryStreamsVideoStreamOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // h264 - computed: false, optional: true, required: false
        this._h264 = new TranscoderJobTemplateConfigElementaryStreamsVideoStreamH264OutputReference(this, "h264");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._h264?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.h264 = this._h264?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._h264.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._h264.internalValue = value.h264;
        }
    }
    get h264() {
        return this._h264;
    }
    putH264(value) {
        this._h264.internalValue = value;
    }
    resetH264() {
        this._h264.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get h264Input() {
        return this._h264.internalValue;
    }
}
exports.TranscoderJobTemplateConfigElementaryStreamsVideoStreamOutputReference = TranscoderJobTemplateConfigElementaryStreamsVideoStreamOutputReference;
_j = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigElementaryStreamsVideoStreamOutputReference[_j] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigElementaryStreamsVideoStreamOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigElementaryStreamsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        audio_stream: transcoderJobTemplateConfigElementaryStreamsAudioStreamToTerraform(struct.audioStream),
        video_stream: transcoderJobTemplateConfigElementaryStreamsVideoStreamToTerraform(struct.videoStream),
    };
}
exports.transcoderJobTemplateConfigElementaryStreamsToTerraform = transcoderJobTemplateConfigElementaryStreamsToTerraform;
function transcoderJobTemplateConfigElementaryStreamsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        audio_stream: {
            value: transcoderJobTemplateConfigElementaryStreamsAudioStreamToHclTerraform(struct.audioStream),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigElementaryStreamsAudioStreamList",
        },
        video_stream: {
            value: transcoderJobTemplateConfigElementaryStreamsVideoStreamToHclTerraform(struct.videoStream),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigElementaryStreamsVideoStreamList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigElementaryStreamsToHclTerraform = transcoderJobTemplateConfigElementaryStreamsToHclTerraform;
class TranscoderJobTemplateConfigElementaryStreamsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // audio_stream - computed: false, optional: true, required: false
        this._audioStream = new TranscoderJobTemplateConfigElementaryStreamsAudioStreamOutputReference(this, "audio_stream");
        // video_stream - computed: false, optional: true, required: false
        this._videoStream = new TranscoderJobTemplateConfigElementaryStreamsVideoStreamOutputReference(this, "video_stream");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._audioStream?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.audioStream = this._audioStream?.internalValue;
        }
        if (this._videoStream?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.videoStream = this._videoStream?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._audioStream.internalValue = undefined;
            this._videoStream.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._audioStream.internalValue = value.audioStream;
            this._videoStream.internalValue = value.videoStream;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get audioStream() {
        return this._audioStream;
    }
    putAudioStream(value) {
        this._audioStream.internalValue = value;
    }
    resetAudioStream() {
        this._audioStream.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get audioStreamInput() {
        return this._audioStream.internalValue;
    }
    get videoStream() {
        return this._videoStream;
    }
    putVideoStream(value) {
        this._videoStream.internalValue = value;
    }
    resetVideoStream() {
        this._videoStream.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get videoStreamInput() {
        return this._videoStream.internalValue;
    }
}
exports.TranscoderJobTemplateConfigElementaryStreamsOutputReference = TranscoderJobTemplateConfigElementaryStreamsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigElementaryStreamsOutputReference[_k] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigElementaryStreamsOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigElementaryStreamsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigElementaryStreamsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigElementaryStreamsList = TranscoderJobTemplateConfigElementaryStreamsList;
_l = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigElementaryStreamsList[_l] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigElementaryStreamsList", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsAes128ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.transcoderJobTemplateConfigEncryptionsAes128ToTerraform = transcoderJobTemplateConfigEncryptionsAes128ToTerraform;
function transcoderJobTemplateConfigEncryptionsAes128ToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.transcoderJobTemplateConfigEncryptionsAes128ToHclTerraform = transcoderJobTemplateConfigEncryptionsAes128ToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsAes128OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.TranscoderJobTemplateConfigEncryptionsAes128OutputReference = TranscoderJobTemplateConfigEncryptionsAes128OutputReference;
_m = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsAes128OutputReference[_m] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsAes128OutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToTerraform;
function transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToHclTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsDrmSystemsClearkeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsClearkeyOutputReference = TranscoderJobTemplateConfigEncryptionsDrmSystemsClearkeyOutputReference;
_o = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsDrmSystemsClearkeyOutputReference[_o] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsDrmSystemsClearkeyOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToTerraform;
function transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToHclTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsDrmSystemsFairplayOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsFairplayOutputReference = TranscoderJobTemplateConfigEncryptionsDrmSystemsFairplayOutputReference;
_p = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsDrmSystemsFairplayOutputReference[_p] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsDrmSystemsFairplayOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToTerraform;
function transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToHclTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyOutputReference = TranscoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyOutputReference;
_q = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyOutputReference[_q] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToTerraform;
function transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToHclTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsDrmSystemsWidevineOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsWidevineOutputReference = TranscoderJobTemplateConfigEncryptionsDrmSystemsWidevineOutputReference;
_r = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsDrmSystemsWidevineOutputReference[_r] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsDrmSystemsWidevineOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsDrmSystemsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        clearkey: transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToTerraform(struct.clearkey),
        fairplay: transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToTerraform(struct.fairplay),
        playready: transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToTerraform(struct.playready),
        widevine: transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToTerraform(struct.widevine),
    };
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsToTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsToTerraform;
function transcoderJobTemplateConfigEncryptionsDrmSystemsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        clearkey: {
            value: transcoderJobTemplateConfigEncryptionsDrmSystemsClearkeyToHclTerraform(struct.clearkey),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsDrmSystemsClearkeyList",
        },
        fairplay: {
            value: transcoderJobTemplateConfigEncryptionsDrmSystemsFairplayToHclTerraform(struct.fairplay),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsDrmSystemsFairplayList",
        },
        playready: {
            value: transcoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyToHclTerraform(struct.playready),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyList",
        },
        widevine: {
            value: transcoderJobTemplateConfigEncryptionsDrmSystemsWidevineToHclTerraform(struct.widevine),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsDrmSystemsWidevineList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigEncryptionsDrmSystemsToHclTerraform = transcoderJobTemplateConfigEncryptionsDrmSystemsToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsDrmSystemsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // clearkey - computed: false, optional: true, required: false
        this._clearkey = new TranscoderJobTemplateConfigEncryptionsDrmSystemsClearkeyOutputReference(this, "clearkey");
        // fairplay - computed: false, optional: true, required: false
        this._fairplay = new TranscoderJobTemplateConfigEncryptionsDrmSystemsFairplayOutputReference(this, "fairplay");
        // playready - computed: false, optional: true, required: false
        this._playready = new TranscoderJobTemplateConfigEncryptionsDrmSystemsPlayreadyOutputReference(this, "playready");
        // widevine - computed: false, optional: true, required: false
        this._widevine = new TranscoderJobTemplateConfigEncryptionsDrmSystemsWidevineOutputReference(this, "widevine");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clearkey?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.clearkey = this._clearkey?.internalValue;
        }
        if (this._fairplay?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.fairplay = this._fairplay?.internalValue;
        }
        if (this._playready?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.playready = this._playready?.internalValue;
        }
        if (this._widevine?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.widevine = this._widevine?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clearkey.internalValue = undefined;
            this._fairplay.internalValue = undefined;
            this._playready.internalValue = undefined;
            this._widevine.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clearkey.internalValue = value.clearkey;
            this._fairplay.internalValue = value.fairplay;
            this._playready.internalValue = value.playready;
            this._widevine.internalValue = value.widevine;
        }
    }
    get clearkey() {
        return this._clearkey;
    }
    putClearkey(value) {
        this._clearkey.internalValue = value;
    }
    resetClearkey() {
        this._clearkey.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clearkeyInput() {
        return this._clearkey.internalValue;
    }
    get fairplay() {
        return this._fairplay;
    }
    putFairplay(value) {
        this._fairplay.internalValue = value;
    }
    resetFairplay() {
        this._fairplay.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fairplayInput() {
        return this._fairplay.internalValue;
    }
    get playready() {
        return this._playready;
    }
    putPlayready(value) {
        this._playready.internalValue = value;
    }
    resetPlayready() {
        this._playready.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get playreadyInput() {
        return this._playready.internalValue;
    }
    get widevine() {
        return this._widevine;
    }
    putWidevine(value) {
        this._widevine.internalValue = value;
    }
    resetWidevine() {
        this._widevine.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get widevineInput() {
        return this._widevine.internalValue;
    }
}
exports.TranscoderJobTemplateConfigEncryptionsDrmSystemsOutputReference = TranscoderJobTemplateConfigEncryptionsDrmSystemsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsDrmSystemsOutputReference[_s] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsDrmSystemsOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsMpegCencToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        scheme: cdktf.stringToTerraform(struct.scheme),
    };
}
exports.transcoderJobTemplateConfigEncryptionsMpegCencToTerraform = transcoderJobTemplateConfigEncryptionsMpegCencToTerraform;
function transcoderJobTemplateConfigEncryptionsMpegCencToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        scheme: {
            value: cdktf.stringToHclTerraform(struct.scheme),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigEncryptionsMpegCencToHclTerraform = transcoderJobTemplateConfigEncryptionsMpegCencToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsMpegCencOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._scheme !== undefined) {
            hasAnyValues = true;
            internalValueResult.scheme = this._scheme;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._scheme = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._scheme = value.scheme;
        }
    }
    get scheme() {
        return this.getStringAttribute('scheme');
    }
    set scheme(value) {
        this._scheme = value;
    }
    // Temporarily expose input value. Use with caution.
    get schemeInput() {
        return this._scheme;
    }
}
exports.TranscoderJobTemplateConfigEncryptionsMpegCencOutputReference = TranscoderJobTemplateConfigEncryptionsMpegCencOutputReference;
_t = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsMpegCencOutputReference[_t] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsMpegCencOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsSampleAesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.transcoderJobTemplateConfigEncryptionsSampleAesToTerraform = transcoderJobTemplateConfigEncryptionsSampleAesToTerraform;
function transcoderJobTemplateConfigEncryptionsSampleAesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.transcoderJobTemplateConfigEncryptionsSampleAesToHclTerraform = transcoderJobTemplateConfigEncryptionsSampleAesToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsSampleAesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.TranscoderJobTemplateConfigEncryptionsSampleAesOutputReference = TranscoderJobTemplateConfigEncryptionsSampleAesOutputReference;
_u = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsSampleAesOutputReference[_u] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsSampleAesOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_version: cdktf.stringToTerraform(struct.secretVersion),
    };
}
exports.transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToTerraform = transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToTerraform;
function transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        secret_version: {
            value: cdktf.stringToHclTerraform(struct.secretVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToHclTerraform = transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsSecretManagerKeySourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretVersion = this._secretVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretVersion = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretVersion = value.secretVersion;
        }
    }
    get secretVersion() {
        return this.getStringAttribute('secret_version');
    }
    set secretVersion(value) {
        this._secretVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretVersionInput() {
        return this._secretVersion;
    }
}
exports.TranscoderJobTemplateConfigEncryptionsSecretManagerKeySourceOutputReference = TranscoderJobTemplateConfigEncryptionsSecretManagerKeySourceOutputReference;
_v = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsSecretManagerKeySourceOutputReference[_v] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsSecretManagerKeySourceOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigEncryptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        aes128: transcoderJobTemplateConfigEncryptionsAes128ToTerraform(struct.aes128),
        drm_systems: transcoderJobTemplateConfigEncryptionsDrmSystemsToTerraform(struct.drmSystems),
        mpeg_cenc: transcoderJobTemplateConfigEncryptionsMpegCencToTerraform(struct.mpegCenc),
        sample_aes: transcoderJobTemplateConfigEncryptionsSampleAesToTerraform(struct.sampleAes),
        secret_manager_key_source: transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToTerraform(struct.secretManagerKeySource),
    };
}
exports.transcoderJobTemplateConfigEncryptionsToTerraform = transcoderJobTemplateConfigEncryptionsToTerraform;
function transcoderJobTemplateConfigEncryptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        aes128: {
            value: transcoderJobTemplateConfigEncryptionsAes128ToHclTerraform(struct.aes128),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsAes128List",
        },
        drm_systems: {
            value: transcoderJobTemplateConfigEncryptionsDrmSystemsToHclTerraform(struct.drmSystems),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsDrmSystemsList",
        },
        mpeg_cenc: {
            value: transcoderJobTemplateConfigEncryptionsMpegCencToHclTerraform(struct.mpegCenc),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsMpegCencList",
        },
        sample_aes: {
            value: transcoderJobTemplateConfigEncryptionsSampleAesToHclTerraform(struct.sampleAes),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsSampleAesList",
        },
        secret_manager_key_source: {
            value: transcoderJobTemplateConfigEncryptionsSecretManagerKeySourceToHclTerraform(struct.secretManagerKeySource),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsSecretManagerKeySourceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigEncryptionsToHclTerraform = transcoderJobTemplateConfigEncryptionsToHclTerraform;
class TranscoderJobTemplateConfigEncryptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // aes128 - computed: false, optional: true, required: false
        this._aes128 = new TranscoderJobTemplateConfigEncryptionsAes128OutputReference(this, "aes128");
        // drm_systems - computed: false, optional: true, required: false
        this._drmSystems = new TranscoderJobTemplateConfigEncryptionsDrmSystemsOutputReference(this, "drm_systems");
        // mpeg_cenc - computed: false, optional: true, required: false
        this._mpegCenc = new TranscoderJobTemplateConfigEncryptionsMpegCencOutputReference(this, "mpeg_cenc");
        // sample_aes - computed: false, optional: true, required: false
        this._sampleAes = new TranscoderJobTemplateConfigEncryptionsSampleAesOutputReference(this, "sample_aes");
        // secret_manager_key_source - computed: false, optional: true, required: false
        this._secretManagerKeySource = new TranscoderJobTemplateConfigEncryptionsSecretManagerKeySourceOutputReference(this, "secret_manager_key_source");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._aes128?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.aes128 = this._aes128?.internalValue;
        }
        if (this._drmSystems?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.drmSystems = this._drmSystems?.internalValue;
        }
        if (this._mpegCenc?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.mpegCenc = this._mpegCenc?.internalValue;
        }
        if (this._sampleAes?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sampleAes = this._sampleAes?.internalValue;
        }
        if (this._secretManagerKeySource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretManagerKeySource = this._secretManagerKeySource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._aes128.internalValue = undefined;
            this._drmSystems.internalValue = undefined;
            this._mpegCenc.internalValue = undefined;
            this._sampleAes.internalValue = undefined;
            this._secretManagerKeySource.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._aes128.internalValue = value.aes128;
            this._drmSystems.internalValue = value.drmSystems;
            this._mpegCenc.internalValue = value.mpegCenc;
            this._sampleAes.internalValue = value.sampleAes;
            this._secretManagerKeySource.internalValue = value.secretManagerKeySource;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get aes128() {
        return this._aes128;
    }
    putAes128(value) {
        this._aes128.internalValue = value;
    }
    resetAes128() {
        this._aes128.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get aes128Input() {
        return this._aes128.internalValue;
    }
    get drmSystems() {
        return this._drmSystems;
    }
    putDrmSystems(value) {
        this._drmSystems.internalValue = value;
    }
    resetDrmSystems() {
        this._drmSystems.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get drmSystemsInput() {
        return this._drmSystems.internalValue;
    }
    get mpegCenc() {
        return this._mpegCenc;
    }
    putMpegCenc(value) {
        this._mpegCenc.internalValue = value;
    }
    resetMpegCenc() {
        this._mpegCenc.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mpegCencInput() {
        return this._mpegCenc.internalValue;
    }
    get sampleAes() {
        return this._sampleAes;
    }
    putSampleAes(value) {
        this._sampleAes.internalValue = value;
    }
    resetSampleAes() {
        this._sampleAes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sampleAesInput() {
        return this._sampleAes.internalValue;
    }
    get secretManagerKeySource() {
        return this._secretManagerKeySource;
    }
    putSecretManagerKeySource(value) {
        this._secretManagerKeySource.internalValue = value;
    }
    resetSecretManagerKeySource() {
        this._secretManagerKeySource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretManagerKeySourceInput() {
        return this._secretManagerKeySource.internalValue;
    }
}
exports.TranscoderJobTemplateConfigEncryptionsOutputReference = TranscoderJobTemplateConfigEncryptionsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsOutputReference[_w] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigEncryptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigEncryptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigEncryptionsList = TranscoderJobTemplateConfigEncryptionsList;
_z = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigEncryptionsList[_z] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigEncryptionsList", version: "14.13.1" };
function transcoderJobTemplateConfigInputsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.transcoderJobTemplateConfigInputsToTerraform = transcoderJobTemplateConfigInputsToTerraform;
function transcoderJobTemplateConfigInputsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigInputsToHclTerraform = transcoderJobTemplateConfigInputsToHclTerraform;
class TranscoderJobTemplateConfigInputsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._uri = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._uri = value.uri;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    resetUri() {
        this._uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.TranscoderJobTemplateConfigInputsOutputReference = TranscoderJobTemplateConfigInputsOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigInputsOutputReference[_0] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigInputsOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigInputsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigInputsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigInputsList = TranscoderJobTemplateConfigInputsList;
_1 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigInputsList[_1] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigInputsList", version: "14.13.1" };
function transcoderJobTemplateConfigManifestsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        file_name: cdktf.stringToTerraform(struct.fileName),
        mux_streams: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.muxStreams),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.transcoderJobTemplateConfigManifestsToTerraform = transcoderJobTemplateConfigManifestsToTerraform;
function transcoderJobTemplateConfigManifestsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        file_name: {
            value: cdktf.stringToHclTerraform(struct.fileName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        mux_streams: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.muxStreams),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigManifestsToHclTerraform = transcoderJobTemplateConfigManifestsToHclTerraform;
class TranscoderJobTemplateConfigManifestsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fileName !== undefined) {
            hasAnyValues = true;
            internalValueResult.fileName = this._fileName;
        }
        if (this._muxStreams !== undefined) {
            hasAnyValues = true;
            internalValueResult.muxStreams = this._muxStreams;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._fileName = undefined;
            this._muxStreams = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._fileName = value.fileName;
            this._muxStreams = value.muxStreams;
            this._type = value.type;
        }
    }
    get fileName() {
        return this.getStringAttribute('file_name');
    }
    set fileName(value) {
        this._fileName = value;
    }
    resetFileName() {
        this._fileName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileNameInput() {
        return this._fileName;
    }
    get muxStreams() {
        return this.getListAttribute('mux_streams');
    }
    set muxStreams(value) {
        this._muxStreams = value;
    }
    resetMuxStreams() {
        this._muxStreams = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get muxStreamsInput() {
        return this._muxStreams;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.TranscoderJobTemplateConfigManifestsOutputReference = TranscoderJobTemplateConfigManifestsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigManifestsOutputReference[_2] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigManifestsOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigManifestsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigManifestsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigManifestsList = TranscoderJobTemplateConfigManifestsList;
_3 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigManifestsList[_3] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigManifestsList", version: "14.13.1" };
function transcoderJobTemplateConfigMuxStreamsSegmentSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        segment_duration: cdktf.stringToTerraform(struct.segmentDuration),
    };
}
exports.transcoderJobTemplateConfigMuxStreamsSegmentSettingsToTerraform = transcoderJobTemplateConfigMuxStreamsSegmentSettingsToTerraform;
function transcoderJobTemplateConfigMuxStreamsSegmentSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        segment_duration: {
            value: cdktf.stringToHclTerraform(struct.segmentDuration),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigMuxStreamsSegmentSettingsToHclTerraform = transcoderJobTemplateConfigMuxStreamsSegmentSettingsToHclTerraform;
class TranscoderJobTemplateConfigMuxStreamsSegmentSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._segmentDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.segmentDuration = this._segmentDuration;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._segmentDuration = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._segmentDuration = value.segmentDuration;
        }
    }
    get segmentDuration() {
        return this.getStringAttribute('segment_duration');
    }
    set segmentDuration(value) {
        this._segmentDuration = value;
    }
    resetSegmentDuration() {
        this._segmentDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get segmentDurationInput() {
        return this._segmentDuration;
    }
}
exports.TranscoderJobTemplateConfigMuxStreamsSegmentSettingsOutputReference = TranscoderJobTemplateConfigMuxStreamsSegmentSettingsOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigMuxStreamsSegmentSettingsOutputReference[_4] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigMuxStreamsSegmentSettingsOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigMuxStreamsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container: cdktf.stringToTerraform(struct.container),
        elementary_streams: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.elementaryStreams),
        encryption_id: cdktf.stringToTerraform(struct.encryptionId),
        file_name: cdktf.stringToTerraform(struct.fileName),
        key: cdktf.stringToTerraform(struct.key),
        segment_settings: transcoderJobTemplateConfigMuxStreamsSegmentSettingsToTerraform(struct.segmentSettings),
    };
}
exports.transcoderJobTemplateConfigMuxStreamsToTerraform = transcoderJobTemplateConfigMuxStreamsToTerraform;
function transcoderJobTemplateConfigMuxStreamsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        container: {
            value: cdktf.stringToHclTerraform(struct.container),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        elementary_streams: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.elementaryStreams),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        encryption_id: {
            value: cdktf.stringToHclTerraform(struct.encryptionId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        file_name: {
            value: cdktf.stringToHclTerraform(struct.fileName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        segment_settings: {
            value: transcoderJobTemplateConfigMuxStreamsSegmentSettingsToHclTerraform(struct.segmentSettings),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigMuxStreamsSegmentSettingsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigMuxStreamsToHclTerraform = transcoderJobTemplateConfigMuxStreamsToHclTerraform;
class TranscoderJobTemplateConfigMuxStreamsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // segment_settings - computed: false, optional: true, required: false
        this._segmentSettings = new TranscoderJobTemplateConfigMuxStreamsSegmentSettingsOutputReference(this, "segment_settings");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._container !== undefined) {
            hasAnyValues = true;
            internalValueResult.container = this._container;
        }
        if (this._elementaryStreams !== undefined) {
            hasAnyValues = true;
            internalValueResult.elementaryStreams = this._elementaryStreams;
        }
        if (this._encryptionId !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionId = this._encryptionId;
        }
        if (this._fileName !== undefined) {
            hasAnyValues = true;
            internalValueResult.fileName = this._fileName;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._segmentSettings?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.segmentSettings = this._segmentSettings?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._container = undefined;
            this._elementaryStreams = undefined;
            this._encryptionId = undefined;
            this._fileName = undefined;
            this._key = undefined;
            this._segmentSettings.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._container = value.container;
            this._elementaryStreams = value.elementaryStreams;
            this._encryptionId = value.encryptionId;
            this._fileName = value.fileName;
            this._key = value.key;
            this._segmentSettings.internalValue = value.segmentSettings;
        }
    }
    get container() {
        return this.getStringAttribute('container');
    }
    set container(value) {
        this._container = value;
    }
    resetContainer() {
        this._container = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerInput() {
        return this._container;
    }
    get elementaryStreams() {
        return this.getListAttribute('elementary_streams');
    }
    set elementaryStreams(value) {
        this._elementaryStreams = value;
    }
    resetElementaryStreams() {
        this._elementaryStreams = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get elementaryStreamsInput() {
        return this._elementaryStreams;
    }
    get encryptionId() {
        return this.getStringAttribute('encryption_id');
    }
    set encryptionId(value) {
        this._encryptionId = value;
    }
    resetEncryptionId() {
        this._encryptionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionIdInput() {
        return this._encryptionId;
    }
    get fileName() {
        return this.getStringAttribute('file_name');
    }
    set fileName(value) {
        this._fileName = value;
    }
    resetFileName() {
        this._fileName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileNameInput() {
        return this._fileName;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get segmentSettings() {
        return this._segmentSettings;
    }
    putSegmentSettings(value) {
        this._segmentSettings.internalValue = value;
    }
    resetSegmentSettings() {
        this._segmentSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get segmentSettingsInput() {
        return this._segmentSettings.internalValue;
    }
}
exports.TranscoderJobTemplateConfigMuxStreamsOutputReference = TranscoderJobTemplateConfigMuxStreamsOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigMuxStreamsOutputReference[_5] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigMuxStreamsOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigMuxStreamsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigMuxStreamsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigMuxStreamsList = TranscoderJobTemplateConfigMuxStreamsList;
_6 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigMuxStreamsList[_6] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigMuxStreamsList", version: "14.13.1" };
function transcoderJobTemplateConfigOutputToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.transcoderJobTemplateConfigOutputToTerraform = transcoderJobTemplateConfigOutputToTerraform;
function transcoderJobTemplateConfigOutputToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigOutputToHclTerraform = transcoderJobTemplateConfigOutputToHclTerraform;
class TranscoderJobTemplateConfigOutputOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._uri = value.uri;
        }
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    resetUri() {
        this._uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.TranscoderJobTemplateConfigOutputOutputReference = TranscoderJobTemplateConfigOutputOutputReference;
_7 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigOutputOutputReference[_7] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigOutputOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        x: cdktf.numberToTerraform(struct.x),
        y: cdktf.numberToTerraform(struct.y),
    };
}
exports.transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToTerraform = transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToTerraform;
function transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        x: {
            value: cdktf.numberToHclTerraform(struct.x),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        y: {
            value: cdktf.numberToHclTerraform(struct.y),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToHclTerraform = transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToHclTerraform;
class TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._x !== undefined) {
            hasAnyValues = true;
            internalValueResult.x = this._x;
        }
        if (this._y !== undefined) {
            hasAnyValues = true;
            internalValueResult.y = this._y;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._x = undefined;
            this._y = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._x = value.x;
            this._y = value.y;
        }
    }
    get x() {
        return this.getNumberAttribute('x');
    }
    set x(value) {
        this._x = value;
    }
    resetX() {
        this._x = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get xInput() {
        return this._x;
    }
    get y() {
        return this.getNumberAttribute('y');
    }
    set y(value) {
        this._y = value;
    }
    resetY() {
        this._y = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get yInput() {
        return this._y;
    }
}
exports.TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyOutputReference = TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyOutputReference[_8] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        end_time_offset: cdktf.stringToTerraform(struct.endTimeOffset),
        fade_type: cdktf.stringToTerraform(struct.fadeType),
        start_time_offset: cdktf.stringToTerraform(struct.startTimeOffset),
        xy: transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToTerraform(struct.xy),
    };
}
exports.transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToTerraform = transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToTerraform;
function transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        end_time_offset: {
            value: cdktf.stringToHclTerraform(struct.endTimeOffset),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        fade_type: {
            value: cdktf.stringToHclTerraform(struct.fadeType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start_time_offset: {
            value: cdktf.stringToHclTerraform(struct.startTimeOffset),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        xy: {
            value: transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyToHclTerraform(struct.xy),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToHclTerraform = transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToHclTerraform;
class TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // xy - computed: false, optional: true, required: false
        this._xy = new TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeXyOutputReference(this, "xy");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endTimeOffset !== undefined) {
            hasAnyValues = true;
            internalValueResult.endTimeOffset = this._endTimeOffset;
        }
        if (this._fadeType !== undefined) {
            hasAnyValues = true;
            internalValueResult.fadeType = this._fadeType;
        }
        if (this._startTimeOffset !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTimeOffset = this._startTimeOffset;
        }
        if (this._xy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.xy = this._xy?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endTimeOffset = undefined;
            this._fadeType = undefined;
            this._startTimeOffset = undefined;
            this._xy.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endTimeOffset = value.endTimeOffset;
            this._fadeType = value.fadeType;
            this._startTimeOffset = value.startTimeOffset;
            this._xy.internalValue = value.xy;
        }
    }
    get endTimeOffset() {
        return this.getStringAttribute('end_time_offset');
    }
    set endTimeOffset(value) {
        this._endTimeOffset = value;
    }
    resetEndTimeOffset() {
        this._endTimeOffset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endTimeOffsetInput() {
        return this._endTimeOffset;
    }
    get fadeType() {
        return this.getStringAttribute('fade_type');
    }
    set fadeType(value) {
        this._fadeType = value;
    }
    // Temporarily expose input value. Use with caution.
    get fadeTypeInput() {
        return this._fadeType;
    }
    get startTimeOffset() {
        return this.getStringAttribute('start_time_offset');
    }
    set startTimeOffset(value) {
        this._startTimeOffset = value;
    }
    resetStartTimeOffset() {
        this._startTimeOffset = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeOffsetInput() {
        return this._startTimeOffset;
    }
    get xy() {
        return this._xy;
    }
    putXy(value) {
        this._xy.internalValue = value;
    }
    resetXy() {
        this._xy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get xyInput() {
        return this._xy.internalValue;
    }
}
exports.TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeOutputReference = TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeOutputReference;
_9 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeOutputReference[_9] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigOverlaysAnimationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        animation_fade: transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToTerraform(struct.animationFade),
    };
}
exports.transcoderJobTemplateConfigOverlaysAnimationsToTerraform = transcoderJobTemplateConfigOverlaysAnimationsToTerraform;
function transcoderJobTemplateConfigOverlaysAnimationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        animation_fade: {
            value: transcoderJobTemplateConfigOverlaysAnimationsAnimationFadeToHclTerraform(struct.animationFade),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigOverlaysAnimationsToHclTerraform = transcoderJobTemplateConfigOverlaysAnimationsToHclTerraform;
class TranscoderJobTemplateConfigOverlaysAnimationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // animation_fade - computed: false, optional: true, required: false
        this._animationFade = new TranscoderJobTemplateConfigOverlaysAnimationsAnimationFadeOutputReference(this, "animation_fade");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._animationFade?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.animationFade = this._animationFade?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._animationFade.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._animationFade.internalValue = value.animationFade;
        }
    }
    get animationFade() {
        return this._animationFade;
    }
    putAnimationFade(value) {
        this._animationFade.internalValue = value;
    }
    resetAnimationFade() {
        this._animationFade.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get animationFadeInput() {
        return this._animationFade.internalValue;
    }
}
exports.TranscoderJobTemplateConfigOverlaysAnimationsOutputReference = TranscoderJobTemplateConfigOverlaysAnimationsOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigOverlaysAnimationsOutputReference[_10] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigOverlaysAnimationsOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigOverlaysAnimationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigOverlaysAnimationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigOverlaysAnimationsList = TranscoderJobTemplateConfigOverlaysAnimationsList;
_11 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigOverlaysAnimationsList[_11] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigOverlaysAnimationsList", version: "14.13.1" };
function transcoderJobTemplateConfigOverlaysImageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.transcoderJobTemplateConfigOverlaysImageToTerraform = transcoderJobTemplateConfigOverlaysImageToTerraform;
function transcoderJobTemplateConfigOverlaysImageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigOverlaysImageToHclTerraform = transcoderJobTemplateConfigOverlaysImageToHclTerraform;
class TranscoderJobTemplateConfigOverlaysImageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._uri = value.uri;
        }
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.TranscoderJobTemplateConfigOverlaysImageOutputReference = TranscoderJobTemplateConfigOverlaysImageOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigOverlaysImageOutputReference[_12] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigOverlaysImageOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigOverlaysToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        animations: cdktf.listMapper(transcoderJobTemplateConfigOverlaysAnimationsToTerraform, true)(struct.animations),
        image: transcoderJobTemplateConfigOverlaysImageToTerraform(struct.image),
    };
}
exports.transcoderJobTemplateConfigOverlaysToTerraform = transcoderJobTemplateConfigOverlaysToTerraform;
function transcoderJobTemplateConfigOverlaysToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        animations: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigOverlaysAnimationsToHclTerraform, true)(struct.animations),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigOverlaysAnimationsList",
        },
        image: {
            value: transcoderJobTemplateConfigOverlaysImageToHclTerraform(struct.image),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigOverlaysImageList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigOverlaysToHclTerraform = transcoderJobTemplateConfigOverlaysToHclTerraform;
class TranscoderJobTemplateConfigOverlaysOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // animations - computed: false, optional: true, required: false
        this._animations = new TranscoderJobTemplateConfigOverlaysAnimationsList(this, "animations", false);
        // image - computed: false, optional: true, required: false
        this._image = new TranscoderJobTemplateConfigOverlaysImageOutputReference(this, "image");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._animations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.animations = this._animations?.internalValue;
        }
        if (this._image?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.image = this._image?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._animations.internalValue = undefined;
            this._image.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._animations.internalValue = value.animations;
            this._image.internalValue = value.image;
        }
    }
    get animations() {
        return this._animations;
    }
    putAnimations(value) {
        this._animations.internalValue = value;
    }
    resetAnimations() {
        this._animations.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get animationsInput() {
        return this._animations.internalValue;
    }
    get image() {
        return this._image;
    }
    putImage(value) {
        this._image.internalValue = value;
    }
    resetImage() {
        this._image.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageInput() {
        return this._image.internalValue;
    }
}
exports.TranscoderJobTemplateConfigOverlaysOutputReference = TranscoderJobTemplateConfigOverlaysOutputReference;
_13 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigOverlaysOutputReference[_13] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigOverlaysOutputReference", version: "14.13.1" };
class TranscoderJobTemplateConfigOverlaysList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new TranscoderJobTemplateConfigOverlaysOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.TranscoderJobTemplateConfigOverlaysList = TranscoderJobTemplateConfigOverlaysList;
_14 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigOverlaysList[_14] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigOverlaysList", version: "14.13.1" };
function transcoderJobTemplateConfigPubsubDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        topic: cdktf.stringToTerraform(struct.topic),
    };
}
exports.transcoderJobTemplateConfigPubsubDestinationToTerraform = transcoderJobTemplateConfigPubsubDestinationToTerraform;
function transcoderJobTemplateConfigPubsubDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        topic: {
            value: cdktf.stringToHclTerraform(struct.topic),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigPubsubDestinationToHclTerraform = transcoderJobTemplateConfigPubsubDestinationToHclTerraform;
class TranscoderJobTemplateConfigPubsubDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._topic !== undefined) {
            hasAnyValues = true;
            internalValueResult.topic = this._topic;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._topic = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._topic = value.topic;
        }
    }
    get topic() {
        return this.getStringAttribute('topic');
    }
    set topic(value) {
        this._topic = value;
    }
    resetTopic() {
        this._topic = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get topicInput() {
        return this._topic;
    }
}
exports.TranscoderJobTemplateConfigPubsubDestinationOutputReference = TranscoderJobTemplateConfigPubsubDestinationOutputReference;
_15 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigPubsubDestinationOutputReference[_15] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigPubsubDestinationOutputReference", version: "14.13.1" };
function transcoderJobTemplateConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ad_breaks: cdktf.listMapper(transcoderJobTemplateConfigAdBreaksToTerraform, true)(struct.adBreaks),
        edit_list: cdktf.listMapper(transcoderJobTemplateConfigEditListStructToTerraform, true)(struct.editList),
        elementary_streams: cdktf.listMapper(transcoderJobTemplateConfigElementaryStreamsToTerraform, true)(struct.elementaryStreams),
        encryptions: cdktf.listMapper(transcoderJobTemplateConfigEncryptionsToTerraform, true)(struct.encryptions),
        inputs: cdktf.listMapper(transcoderJobTemplateConfigInputsToTerraform, true)(struct.inputs),
        manifests: cdktf.listMapper(transcoderJobTemplateConfigManifestsToTerraform, true)(struct.manifests),
        mux_streams: cdktf.listMapper(transcoderJobTemplateConfigMuxStreamsToTerraform, true)(struct.muxStreams),
        output: transcoderJobTemplateConfigOutputToTerraform(struct.output),
        overlays: cdktf.listMapper(transcoderJobTemplateConfigOverlaysToTerraform, true)(struct.overlays),
        pubsub_destination: transcoderJobTemplateConfigPubsubDestinationToTerraform(struct.pubsubDestination),
    };
}
exports.transcoderJobTemplateConfigAToTerraform = transcoderJobTemplateConfigAToTerraform;
function transcoderJobTemplateConfigAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ad_breaks: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigAdBreaksToHclTerraform, true)(struct.adBreaks),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigAdBreaksList",
        },
        edit_list: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigEditListStructToHclTerraform, true)(struct.editList),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEditListStructList",
        },
        elementary_streams: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigElementaryStreamsToHclTerraform, true)(struct.elementaryStreams),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigElementaryStreamsList",
        },
        encryptions: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigEncryptionsToHclTerraform, true)(struct.encryptions),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigEncryptionsList",
        },
        inputs: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigInputsToHclTerraform, true)(struct.inputs),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigInputsList",
        },
        manifests: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigManifestsToHclTerraform, true)(struct.manifests),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigManifestsList",
        },
        mux_streams: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigMuxStreamsToHclTerraform, true)(struct.muxStreams),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigMuxStreamsList",
        },
        output: {
            value: transcoderJobTemplateConfigOutputToHclTerraform(struct.output),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigOutputList",
        },
        overlays: {
            value: cdktf.listMapperHcl(transcoderJobTemplateConfigOverlaysToHclTerraform, true)(struct.overlays),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigOverlaysList",
        },
        pubsub_destination: {
            value: transcoderJobTemplateConfigPubsubDestinationToHclTerraform(struct.pubsubDestination),
            isBlock: true,
            type: "list",
            storageClassType: "TranscoderJobTemplateConfigPubsubDestinationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateConfigAToHclTerraform = transcoderJobTemplateConfigAToHclTerraform;
class TranscoderJobTemplateConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // ad_breaks - computed: false, optional: true, required: false
        this._adBreaks = new TranscoderJobTemplateConfigAdBreaksList(this, "ad_breaks", false);
        // edit_list - computed: false, optional: true, required: false
        this._editList = new TranscoderJobTemplateConfigEditListStructList(this, "edit_list", false);
        // elementary_streams - computed: false, optional: true, required: false
        this._elementaryStreams = new TranscoderJobTemplateConfigElementaryStreamsList(this, "elementary_streams", false);
        // encryptions - computed: false, optional: true, required: false
        this._encryptions = new TranscoderJobTemplateConfigEncryptionsList(this, "encryptions", false);
        // inputs - computed: false, optional: true, required: false
        this._inputs = new TranscoderJobTemplateConfigInputsList(this, "inputs", false);
        // manifests - computed: false, optional: true, required: false
        this._manifests = new TranscoderJobTemplateConfigManifestsList(this, "manifests", false);
        // mux_streams - computed: false, optional: true, required: false
        this._muxStreams = new TranscoderJobTemplateConfigMuxStreamsList(this, "mux_streams", false);
        // output - computed: false, optional: true, required: false
        this._output = new TranscoderJobTemplateConfigOutputOutputReference(this, "output");
        // overlays - computed: false, optional: true, required: false
        this._overlays = new TranscoderJobTemplateConfigOverlaysList(this, "overlays", false);
        // pubsub_destination - computed: false, optional: true, required: false
        this._pubsubDestination = new TranscoderJobTemplateConfigPubsubDestinationOutputReference(this, "pubsub_destination");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._adBreaks?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.adBreaks = this._adBreaks?.internalValue;
        }
        if (this._editList?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.editList = this._editList?.internalValue;
        }
        if (this._elementaryStreams?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.elementaryStreams = this._elementaryStreams?.internalValue;
        }
        if (this._encryptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptions = this._encryptions?.internalValue;
        }
        if (this._inputs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.inputs = this._inputs?.internalValue;
        }
        if (this._manifests?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.manifests = this._manifests?.internalValue;
        }
        if (this._muxStreams?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.muxStreams = this._muxStreams?.internalValue;
        }
        if (this._output?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.output = this._output?.internalValue;
        }
        if (this._overlays?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.overlays = this._overlays?.internalValue;
        }
        if (this._pubsubDestination?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.pubsubDestination = this._pubsubDestination?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._adBreaks.internalValue = undefined;
            this._editList.internalValue = undefined;
            this._elementaryStreams.internalValue = undefined;
            this._encryptions.internalValue = undefined;
            this._inputs.internalValue = undefined;
            this._manifests.internalValue = undefined;
            this._muxStreams.internalValue = undefined;
            this._output.internalValue = undefined;
            this._overlays.internalValue = undefined;
            this._pubsubDestination.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._adBreaks.internalValue = value.adBreaks;
            this._editList.internalValue = value.editList;
            this._elementaryStreams.internalValue = value.elementaryStreams;
            this._encryptions.internalValue = value.encryptions;
            this._inputs.internalValue = value.inputs;
            this._manifests.internalValue = value.manifests;
            this._muxStreams.internalValue = value.muxStreams;
            this._output.internalValue = value.output;
            this._overlays.internalValue = value.overlays;
            this._pubsubDestination.internalValue = value.pubsubDestination;
        }
    }
    get adBreaks() {
        return this._adBreaks;
    }
    putAdBreaks(value) {
        this._adBreaks.internalValue = value;
    }
    resetAdBreaks() {
        this._adBreaks.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adBreaksInput() {
        return this._adBreaks.internalValue;
    }
    get editList() {
        return this._editList;
    }
    putEditList(value) {
        this._editList.internalValue = value;
    }
    resetEditList() {
        this._editList.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get editListInput() {
        return this._editList.internalValue;
    }
    get elementaryStreams() {
        return this._elementaryStreams;
    }
    putElementaryStreams(value) {
        this._elementaryStreams.internalValue = value;
    }
    resetElementaryStreams() {
        this._elementaryStreams.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get elementaryStreamsInput() {
        return this._elementaryStreams.internalValue;
    }
    get encryptions() {
        return this._encryptions;
    }
    putEncryptions(value) {
        this._encryptions.internalValue = value;
    }
    resetEncryptions() {
        this._encryptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionsInput() {
        return this._encryptions.internalValue;
    }
    get inputs() {
        return this._inputs;
    }
    putInputs(value) {
        this._inputs.internalValue = value;
    }
    resetInputs() {
        this._inputs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputsInput() {
        return this._inputs.internalValue;
    }
    get manifests() {
        return this._manifests;
    }
    putManifests(value) {
        this._manifests.internalValue = value;
    }
    resetManifests() {
        this._manifests.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get manifestsInput() {
        return this._manifests.internalValue;
    }
    get muxStreams() {
        return this._muxStreams;
    }
    putMuxStreams(value) {
        this._muxStreams.internalValue = value;
    }
    resetMuxStreams() {
        this._muxStreams.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get muxStreamsInput() {
        return this._muxStreams.internalValue;
    }
    get output() {
        return this._output;
    }
    putOutput(value) {
        this._output.internalValue = value;
    }
    resetOutput() {
        this._output.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputInput() {
        return this._output.internalValue;
    }
    get overlays() {
        return this._overlays;
    }
    putOverlays(value) {
        this._overlays.internalValue = value;
    }
    resetOverlays() {
        this._overlays.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overlaysInput() {
        return this._overlays.internalValue;
    }
    get pubsubDestination() {
        return this._pubsubDestination;
    }
    putPubsubDestination(value) {
        this._pubsubDestination.internalValue = value;
    }
    resetPubsubDestination() {
        this._pubsubDestination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pubsubDestinationInput() {
        return this._pubsubDestination.internalValue;
    }
}
exports.TranscoderJobTemplateConfigAOutputReference = TranscoderJobTemplateConfigAOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateConfigAOutputReference[_16] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateConfigAOutputReference", version: "14.13.1" };
function transcoderJobTemplateTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.transcoderJobTemplateTimeoutsToTerraform = transcoderJobTemplateTimeoutsToTerraform;
function transcoderJobTemplateTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.transcoderJobTemplateTimeoutsToHclTerraform = transcoderJobTemplateTimeoutsToHclTerraform;
class TranscoderJobTemplateTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.TranscoderJobTemplateTimeoutsOutputReference = TranscoderJobTemplateTimeoutsOutputReference;
_17 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplateTimeoutsOutputReference[_17] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplateTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/transcoder_job_template google_transcoder_job_template}
*/
class TranscoderJobTemplate extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a TranscoderJobTemplate resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the TranscoderJobTemplate to import
    * @param importFromId The id of the existing TranscoderJobTemplate that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/transcoder_job_template#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the TranscoderJobTemplate to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_transcoder_job_template", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/transcoder_job_template google_transcoder_job_template} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options TranscoderJobTemplateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_transcoder_job_template',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // config - computed: false, optional: true, required: false
        this._config = new TranscoderJobTemplateConfigAOutputReference(this, "config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new TranscoderJobTemplateTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._jobTemplateId = config.jobTemplateId;
        this._labels = config.labels;
        this._location = config.location;
        this._project = config.project;
        this._config.internalValue = config.config;
        this._timeouts.internalValue = config.timeouts;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get jobTemplateId() {
        return this.getStringAttribute('job_template_id');
    }
    set jobTemplateId(value) {
        this._jobTemplateId = value;
    }
    // Temporarily expose input value. Use with caution.
    get jobTemplateIdInput() {
        return this._jobTemplateId;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    get config() {
        return this._config;
    }
    putConfig(value) {
        this._config.internalValue = value;
    }
    resetConfig() {
        this._config.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configInput() {
        return this._config.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            job_template_id: cdktf.stringToTerraform(this._jobTemplateId),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            config: transcoderJobTemplateConfigAToTerraform(this._config.internalValue),
            timeouts: transcoderJobTemplateTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            job_template_id: {
                value: cdktf.stringToHclTerraform(this._jobTemplateId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            config: {
                value: transcoderJobTemplateConfigAToHclTerraform(this._config.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "TranscoderJobTemplateConfigAList",
            },
            timeouts: {
                value: transcoderJobTemplateTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "TranscoderJobTemplateTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.TranscoderJobTemplate = TranscoderJobTemplate;
_18 = JSII_RTTI_SYMBOL_1;
TranscoderJobTemplate[_18] = { fqn: "@cdktf/provider-google.transcoderJobTemplate.TranscoderJobTemplate", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
TranscoderJobTemplate.tfResourceType = "google_transcoder_job_template";
