// @cdktf/provider-google/lib/compute-resize-request/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14;
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoToHclTerraform = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoToTerraform = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageList = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageOutputReference = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageToHclTerraform = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageToTerraform = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpList = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpOutputReference = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpToHclTerraform = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpToTerraform = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksList = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksOutputReference = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksToHclTerraform = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksToTerraform = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoList = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoOutputReference = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoToHclTerraform = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoToTerraform = exports.ComputeResizeRequestStatusErrorList = exports.ComputeResizeRequestStatusErrorOutputReference = exports.computeResizeRequestStatusErrorToHclTerraform = exports.computeResizeRequestStatusErrorToTerraform = exports.ComputeResizeRequestStatusErrorErrorsList = exports.ComputeResizeRequestStatusErrorErrorsOutputReference = exports.computeResizeRequestStatusErrorErrorsToHclTerraform = exports.computeResizeRequestStatusErrorErrorsToTerraform = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsList = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsOutputReference = exports.computeResizeRequestStatusErrorErrorsErrorDetailsToHclTerraform = exports.computeResizeRequestStatusErrorErrorsErrorDetailsToTerraform = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoList = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoOutputReference = exports.computeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoToHclTerraform = exports.computeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoToTerraform = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageList = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageOutputReference = exports.computeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageToHclTerraform = exports.computeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageToTerraform = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpList = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpOutputReference = exports.computeResizeRequestStatusErrorErrorsErrorDetailsHelpToHclTerraform = exports.computeResizeRequestStatusErrorErrorsErrorDetailsHelpToTerraform = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksList = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksOutputReference = exports.computeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksToHclTerraform = exports.computeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksToTerraform = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoList = exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoOutputReference = exports.computeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoToHclTerraform = exports.computeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoToTerraform = void 0;
exports.ComputeResizeRequest = exports.ComputeResizeRequestTimeoutsOutputReference = exports.computeResizeRequestTimeoutsToHclTerraform = exports.computeResizeRequestTimeoutsToTerraform = exports.ComputeResizeRequestRequestedRunDurationOutputReference = exports.computeResizeRequestRequestedRunDurationToHclTerraform = exports.computeResizeRequestRequestedRunDurationToTerraform = exports.ComputeResizeRequestStatusList = exports.ComputeResizeRequestStatusOutputReference = exports.computeResizeRequestStatusToHclTerraform = exports.computeResizeRequestStatusToTerraform = exports.ComputeResizeRequestStatusLastAttemptList = exports.ComputeResizeRequestStatusLastAttemptOutputReference = exports.computeResizeRequestStatusLastAttemptToHclTerraform = exports.computeResizeRequestStatusLastAttemptToTerraform = exports.ComputeResizeRequestStatusLastAttemptErrorList = exports.ComputeResizeRequestStatusLastAttemptErrorOutputReference = exports.computeResizeRequestStatusLastAttemptErrorToHclTerraform = exports.computeResizeRequestStatusLastAttemptErrorToTerraform = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsList = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsOutputReference = exports.computeResizeRequestStatusLastAttemptErrorErrorsToHclTerraform = exports.computeResizeRequestStatusLastAttemptErrorErrorsToTerraform = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsList = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsOutputReference = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsToHclTerraform = exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsToTerraform = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoList = exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoToTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoToTerraform;
function computeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoToHclTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoToHclTerraform;
class ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // metadatas - computed: true, optional: false, required: false
        this._metadatas = new cdktf.StringMap(this, "metadatas");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
    get metadatas() {
        return this._metadatas;
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoOutputReference = ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoOutputReference[_a] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoList = ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoList;
_b = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoList[_b] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoList", version: "14.13.1" };
function computeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksToTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksToTerraform;
function computeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksToHclTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksToHclTerraform;
class ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksOutputReference = ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksOutputReference[_c] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksList = ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksList;
_d = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksList[_d] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksList", version: "14.13.1" };
function computeResizeRequestStatusErrorErrorsErrorDetailsHelpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsHelpToTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsHelpToTerraform;
function computeResizeRequestStatusErrorErrorsErrorDetailsHelpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsHelpToHclTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsHelpToHclTerraform;
class ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // links - computed: true, optional: false, required: false
        this._links = new ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpLinksList(this, "links", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get links() {
        return this._links;
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpOutputReference = ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpOutputReference[_e] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpList = ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpList;
_f = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpList[_f] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpList", version: "14.13.1" };
function computeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageToTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageToTerraform;
function computeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageToHclTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageToHclTerraform;
class ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // locale - computed: true, optional: false, required: false
    get locale() {
        return this.getStringAttribute('locale');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageOutputReference = ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageOutputReference[_g] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageList = ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageList;
_h = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageList[_h] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageList", version: "14.13.1" };
function computeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoToTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoToTerraform;
function computeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoToHclTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoToHclTerraform;
class ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // dimensions - computed: true, optional: false, required: false
        this._dimensions = new cdktf.StringMap(this, "dimensions");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get dimensions() {
        return this._dimensions;
    }
    // future_limit - computed: true, optional: false, required: false
    get futureLimit() {
        return this.getNumberAttribute('future_limit');
    }
    // limit - computed: true, optional: false, required: false
    get limit() {
        return this.getNumberAttribute('limit');
    }
    // limit_name - computed: true, optional: false, required: false
    get limitName() {
        return this.getStringAttribute('limit_name');
    }
    // metric_name - computed: true, optional: false, required: false
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    // rollout_status - computed: true, optional: false, required: false
    get rolloutStatus() {
        return this.getStringAttribute('rollout_status');
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoOutputReference = ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoOutputReference[_j] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoList = ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoList;
_k = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoList[_k] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoList", version: "14.13.1" };
function computeResizeRequestStatusErrorErrorsErrorDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsToTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsToTerraform;
function computeResizeRequestStatusErrorErrorsErrorDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusErrorErrorsErrorDetailsToHclTerraform = computeResizeRequestStatusErrorErrorsErrorDetailsToHclTerraform;
class ComputeResizeRequestStatusErrorErrorsErrorDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // error_info - computed: true, optional: false, required: false
        this._errorInfo = new ComputeResizeRequestStatusErrorErrorsErrorDetailsErrorInfoList(this, "error_info", false);
        // help - computed: true, optional: false, required: false
        this._help = new ComputeResizeRequestStatusErrorErrorsErrorDetailsHelpList(this, "help", false);
        // localized_message - computed: true, optional: false, required: false
        this._localizedMessage = new ComputeResizeRequestStatusErrorErrorsErrorDetailsLocalizedMessageList(this, "localized_message", false);
        // quota_info - computed: true, optional: false, required: false
        this._quotaInfo = new ComputeResizeRequestStatusErrorErrorsErrorDetailsQuotaInfoList(this, "quota_info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get errorInfo() {
        return this._errorInfo;
    }
    get help() {
        return this._help;
    }
    get localizedMessage() {
        return this._localizedMessage;
    }
    get quotaInfo() {
        return this._quotaInfo;
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsOutputReference = ComputeResizeRequestStatusErrorErrorsErrorDetailsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsOutputReference[_l] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusErrorErrorsErrorDetailsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusErrorErrorsErrorDetailsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusErrorErrorsErrorDetailsList = ComputeResizeRequestStatusErrorErrorsErrorDetailsList;
_m = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsErrorDetailsList[_m] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsErrorDetailsList", version: "14.13.1" };
function computeResizeRequestStatusErrorErrorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusErrorErrorsToTerraform = computeResizeRequestStatusErrorErrorsToTerraform;
function computeResizeRequestStatusErrorErrorsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusErrorErrorsToHclTerraform = computeResizeRequestStatusErrorErrorsToHclTerraform;
class ComputeResizeRequestStatusErrorErrorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // error_details - computed: true, optional: false, required: false
        this._errorDetails = new ComputeResizeRequestStatusErrorErrorsErrorDetailsList(this, "error_details", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // code - computed: true, optional: false, required: false
    get code() {
        return this.getStringAttribute('code');
    }
    get errorDetails() {
        return this._errorDetails;
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
}
exports.ComputeResizeRequestStatusErrorErrorsOutputReference = ComputeResizeRequestStatusErrorErrorsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsOutputReference[_o] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusErrorErrorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusErrorErrorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusErrorErrorsList = ComputeResizeRequestStatusErrorErrorsList;
_p = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorErrorsList[_p] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorErrorsList", version: "14.13.1" };
function computeResizeRequestStatusErrorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusErrorToTerraform = computeResizeRequestStatusErrorToTerraform;
function computeResizeRequestStatusErrorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusErrorToHclTerraform = computeResizeRequestStatusErrorToHclTerraform;
class ComputeResizeRequestStatusErrorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // errors - computed: true, optional: false, required: false
        this._errors = new ComputeResizeRequestStatusErrorErrorsList(this, "errors", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get errors() {
        return this._errors;
    }
}
exports.ComputeResizeRequestStatusErrorOutputReference = ComputeResizeRequestStatusErrorOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorOutputReference[_q] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusErrorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusErrorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusErrorList = ComputeResizeRequestStatusErrorList;
_r = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusErrorList[_r] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusErrorList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoToTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoToTerraform;
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoToHclTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoToHclTerraform;
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // metadatas - computed: true, optional: false, required: false
        this._metadatas = new cdktf.StringMap(this, "metadatas");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
    get metadatas() {
        return this._metadatas;
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoOutputReference = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoOutputReference;
_s = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoOutputReference[_s] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoList = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoList;
_t = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoList[_t] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksToTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksToTerraform;
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksToHclTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksToHclTerraform;
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksOutputReference = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksOutputReference;
_u = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksOutputReference[_u] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksList = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksList;
_v = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksList[_v] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpToTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpToTerraform;
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpToHclTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpToHclTerraform;
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // links - computed: true, optional: false, required: false
        this._links = new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpLinksList(this, "links", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get links() {
        return this._links;
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpOutputReference = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpOutputReference;
_w = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpOutputReference[_w] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpList = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpList;
_x = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpList[_x] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageToTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageToTerraform;
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageToHclTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageToHclTerraform;
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // locale - computed: true, optional: false, required: false
    get locale() {
        return this.getStringAttribute('locale');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageOutputReference = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageOutputReference;
_y = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageOutputReference[_y] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageList = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageList;
_z = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageList[_z] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoToTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoToTerraform;
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoToHclTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoToHclTerraform;
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // dimensions - computed: true, optional: false, required: false
        this._dimensions = new cdktf.StringMap(this, "dimensions");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get dimensions() {
        return this._dimensions;
    }
    // future_limit - computed: true, optional: false, required: false
    get futureLimit() {
        return this.getNumberAttribute('future_limit');
    }
    // limit - computed: true, optional: false, required: false
    get limit() {
        return this.getNumberAttribute('limit');
    }
    // limit_name - computed: true, optional: false, required: false
    get limitName() {
        return this.getStringAttribute('limit_name');
    }
    // metric_name - computed: true, optional: false, required: false
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    // rollout_status - computed: true, optional: false, required: false
    get rolloutStatus() {
        return this.getStringAttribute('rollout_status');
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoOutputReference = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoOutputReference[_0] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoList = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoList;
_1 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoList[_1] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsToTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsToTerraform;
function computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsToHclTerraform = computeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsToHclTerraform;
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // error_info - computed: true, optional: false, required: false
        this._errorInfo = new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsErrorInfoList(this, "error_info", false);
        // help - computed: true, optional: false, required: false
        this._help = new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsHelpList(this, "help", false);
        // localized_message - computed: true, optional: false, required: false
        this._localizedMessage = new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsLocalizedMessageList(this, "localized_message", false);
        // quota_info - computed: true, optional: false, required: false
        this._quotaInfo = new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsQuotaInfoList(this, "quota_info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get errorInfo() {
        return this._errorInfo;
    }
    get help() {
        return this._help;
    }
    get localizedMessage() {
        return this._localizedMessage;
    }
    get quotaInfo() {
        return this._quotaInfo;
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsOutputReference = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsOutputReference[_2] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsList = ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsList;
_3 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsList[_3] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptErrorErrorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsToTerraform = computeResizeRequestStatusLastAttemptErrorErrorsToTerraform;
function computeResizeRequestStatusLastAttemptErrorErrorsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptErrorErrorsToHclTerraform = computeResizeRequestStatusLastAttemptErrorErrorsToHclTerraform;
class ComputeResizeRequestStatusLastAttemptErrorErrorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // error_details - computed: true, optional: false, required: false
        this._errorDetails = new ComputeResizeRequestStatusLastAttemptErrorErrorsErrorDetailsList(this, "error_details", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // code - computed: true, optional: false, required: false
    get code() {
        return this.getStringAttribute('code');
    }
    get errorDetails() {
        return this._errorDetails;
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsOutputReference = ComputeResizeRequestStatusLastAttemptErrorErrorsOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsOutputReference[_4] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptErrorErrorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptErrorErrorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorErrorsList = ComputeResizeRequestStatusLastAttemptErrorErrorsList;
_5 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorErrorsList[_5] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorErrorsList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptErrorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptErrorToTerraform = computeResizeRequestStatusLastAttemptErrorToTerraform;
function computeResizeRequestStatusLastAttemptErrorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptErrorToHclTerraform = computeResizeRequestStatusLastAttemptErrorToHclTerraform;
class ComputeResizeRequestStatusLastAttemptErrorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // errors - computed: true, optional: false, required: false
        this._errors = new ComputeResizeRequestStatusLastAttemptErrorErrorsList(this, "errors", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get errors() {
        return this._errors;
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorOutputReference = ComputeResizeRequestStatusLastAttemptErrorOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorOutputReference[_6] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptErrorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptErrorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptErrorList = ComputeResizeRequestStatusLastAttemptErrorList;
_7 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptErrorList[_7] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptErrorList", version: "14.13.1" };
function computeResizeRequestStatusLastAttemptToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusLastAttemptToTerraform = computeResizeRequestStatusLastAttemptToTerraform;
function computeResizeRequestStatusLastAttemptToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusLastAttemptToHclTerraform = computeResizeRequestStatusLastAttemptToHclTerraform;
class ComputeResizeRequestStatusLastAttemptOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // error - computed: true, optional: false, required: false
        this._error = new ComputeResizeRequestStatusLastAttemptErrorList(this, "error", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get error() {
        return this._error;
    }
}
exports.ComputeResizeRequestStatusLastAttemptOutputReference = ComputeResizeRequestStatusLastAttemptOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptOutputReference[_8] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusLastAttemptList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusLastAttemptOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusLastAttemptList = ComputeResizeRequestStatusLastAttemptList;
_9 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusLastAttemptList[_9] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusLastAttemptList", version: "14.13.1" };
function computeResizeRequestStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeResizeRequestStatusToTerraform = computeResizeRequestStatusToTerraform;
function computeResizeRequestStatusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeResizeRequestStatusToHclTerraform = computeResizeRequestStatusToHclTerraform;
class ComputeResizeRequestStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // error - computed: true, optional: false, required: false
        this._error = new ComputeResizeRequestStatusErrorList(this, "error", false);
        // last_attempt - computed: true, optional: false, required: false
        this._lastAttempt = new ComputeResizeRequestStatusLastAttemptList(this, "last_attempt", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get error() {
        return this._error;
    }
    get lastAttempt() {
        return this._lastAttempt;
    }
}
exports.ComputeResizeRequestStatusOutputReference = ComputeResizeRequestStatusOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusOutputReference[_10] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusOutputReference", version: "14.13.1" };
class ComputeResizeRequestStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeResizeRequestStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeResizeRequestStatusList = ComputeResizeRequestStatusList;
_11 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestStatusList[_11] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestStatusList", version: "14.13.1" };
function computeResizeRequestRequestedRunDurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        nanos: cdktf.numberToTerraform(struct.nanos),
        seconds: cdktf.stringToTerraform(struct.seconds),
    };
}
exports.computeResizeRequestRequestedRunDurationToTerraform = computeResizeRequestRequestedRunDurationToTerraform;
function computeResizeRequestRequestedRunDurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        nanos: {
            value: cdktf.numberToHclTerraform(struct.nanos),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        seconds: {
            value: cdktf.stringToHclTerraform(struct.seconds),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeResizeRequestRequestedRunDurationToHclTerraform = computeResizeRequestRequestedRunDurationToHclTerraform;
class ComputeResizeRequestRequestedRunDurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._nanos !== undefined) {
            hasAnyValues = true;
            internalValueResult.nanos = this._nanos;
        }
        if (this._seconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.seconds = this._seconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._nanos = undefined;
            this._seconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._nanos = value.nanos;
            this._seconds = value.seconds;
        }
    }
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    set nanos(value) {
        this._nanos = value;
    }
    resetNanos() {
        this._nanos = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nanosInput() {
        return this._nanos;
    }
    get seconds() {
        return this.getStringAttribute('seconds');
    }
    set seconds(value) {
        this._seconds = value;
    }
    // Temporarily expose input value. Use with caution.
    get secondsInput() {
        return this._seconds;
    }
}
exports.ComputeResizeRequestRequestedRunDurationOutputReference = ComputeResizeRequestRequestedRunDurationOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestRequestedRunDurationOutputReference[_12] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestRequestedRunDurationOutputReference", version: "14.13.1" };
function computeResizeRequestTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.computeResizeRequestTimeoutsToTerraform = computeResizeRequestTimeoutsToTerraform;
function computeResizeRequestTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeResizeRequestTimeoutsToHclTerraform = computeResizeRequestTimeoutsToHclTerraform;
class ComputeResizeRequestTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.ComputeResizeRequestTimeoutsOutputReference = ComputeResizeRequestTimeoutsOutputReference;
_13 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequestTimeoutsOutputReference[_13] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequestTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_resize_request google_compute_resize_request}
*/
class ComputeResizeRequest extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ComputeResizeRequest resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ComputeResizeRequest to import
    * @param importFromId The id of the existing ComputeResizeRequest that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_resize_request#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ComputeResizeRequest to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_resize_request", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_resize_request google_compute_resize_request} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeResizeRequestConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_resize_request',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // status - computed: true, optional: false, required: false
        this._status = new ComputeResizeRequestStatusList(this, "status", false);
        // requested_run_duration - computed: false, optional: true, required: false
        this._requestedRunDuration = new ComputeResizeRequestRequestedRunDurationOutputReference(this, "requested_run_duration");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeResizeRequestTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._instanceGroupManager = config.instanceGroupManager;
        this._name = config.name;
        this._project = config.project;
        this._resizeBy = config.resizeBy;
        this._zone = config.zone;
        this._requestedRunDuration.internalValue = config.requestedRunDuration;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceGroupManager() {
        return this.getStringAttribute('instance_group_manager');
    }
    set instanceGroupManager(value) {
        this._instanceGroupManager = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceGroupManagerInput() {
        return this._instanceGroupManager;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get resizeBy() {
        return this.getNumberAttribute('resize_by');
    }
    set resizeBy(value) {
        this._resizeBy = value;
    }
    // Temporarily expose input value. Use with caution.
    get resizeByInput() {
        return this._resizeBy;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get status() {
        return this._status;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    get requestedRunDuration() {
        return this._requestedRunDuration;
    }
    putRequestedRunDuration(value) {
        this._requestedRunDuration.internalValue = value;
    }
    resetRequestedRunDuration() {
        this._requestedRunDuration.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestedRunDurationInput() {
        return this._requestedRunDuration.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            instance_group_manager: cdktf.stringToTerraform(this._instanceGroupManager),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            resize_by: cdktf.numberToTerraform(this._resizeBy),
            zone: cdktf.stringToTerraform(this._zone),
            requested_run_duration: computeResizeRequestRequestedRunDurationToTerraform(this._requestedRunDuration.internalValue),
            timeouts: computeResizeRequestTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_group_manager: {
                value: cdktf.stringToHclTerraform(this._instanceGroupManager),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            resize_by: {
                value: cdktf.numberToHclTerraform(this._resizeBy),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            zone: {
                value: cdktf.stringToHclTerraform(this._zone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            requested_run_duration: {
                value: computeResizeRequestRequestedRunDurationToHclTerraform(this._requestedRunDuration.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeResizeRequestRequestedRunDurationList",
            },
            timeouts: {
                value: computeResizeRequestTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ComputeResizeRequestTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ComputeResizeRequest = ComputeResizeRequest;
_14 = JSII_RTTI_SYMBOL_1;
ComputeResizeRequest[_14] = { fqn: "@cdktf/provider-google.computeResizeRequest.ComputeResizeRequest", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ComputeResizeRequest.tfResourceType = "google_compute_resize_request";
