// @cdktf/provider-generator/lib/get/generator/models/scope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scope = void 0;
class Scope {
    constructor(props) {
        var _a, _b, _c, _d, _e;
        this.name = props.name;
        this.parent = props.parent;
        this.isProvider = props.isProvider;
        this.isComputed = (_a = props.isComputed) !== null && _a !== void 0 ? _a : false;
        this.isOptional = (_b = props.isOptional) !== null && _b !== void 0 ? _b : true;
        this.isRequired = (_c = props.isRequired) !== null && _c !== void 0 ? _c : false;
        this.inBlockType = (_d = props.inBlockType) !== null && _d !== void 0 ? _d : false;
        this.isNestedType = (_e = props.isNestedType) !== null && _e !== void 0 ? _e : false;
    }
    fullName(attributeName) {
        return `${this.baseName}.${attributeName}`;
    }
    get baseName() {
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        let scope = this;
        let name = "";
        while (scope) {
            if (name === "") {
                name = scope.name;
            }
            else {
                name = `${scope.name}.${name}`;
            }
            scope = scope.parent;
        }
        return name;
    }
}
exports.Scope = Scope;
