// @cdktf/provider-google/lib/netapp-volume/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetappVolume = exports.NetappVolumeTimeoutsOutputReference = exports.netappVolumeTimeoutsToHclTerraform = exports.netappVolumeTimeoutsToTerraform = exports.NetappVolumeTieringPolicyOutputReference = exports.netappVolumeTieringPolicyToHclTerraform = exports.netappVolumeTieringPolicyToTerraform = exports.NetappVolumeSnapshotPolicyOutputReference = exports.netappVolumeSnapshotPolicyToHclTerraform = exports.netappVolumeSnapshotPolicyToTerraform = exports.NetappVolumeSnapshotPolicyWeeklyScheduleOutputReference = exports.netappVolumeSnapshotPolicyWeeklyScheduleToHclTerraform = exports.netappVolumeSnapshotPolicyWeeklyScheduleToTerraform = exports.NetappVolumeSnapshotPolicyMonthlyScheduleOutputReference = exports.netappVolumeSnapshotPolicyMonthlyScheduleToHclTerraform = exports.netappVolumeSnapshotPolicyMonthlyScheduleToTerraform = exports.NetappVolumeSnapshotPolicyHourlyScheduleOutputReference = exports.netappVolumeSnapshotPolicyHourlyScheduleToHclTerraform = exports.netappVolumeSnapshotPolicyHourlyScheduleToTerraform = exports.NetappVolumeSnapshotPolicyDailyScheduleOutputReference = exports.netappVolumeSnapshotPolicyDailyScheduleToHclTerraform = exports.netappVolumeSnapshotPolicyDailyScheduleToTerraform = exports.NetappVolumeRestoreParametersOutputReference = exports.netappVolumeRestoreParametersToHclTerraform = exports.netappVolumeRestoreParametersToTerraform = exports.NetappVolumeExportPolicyOutputReference = exports.netappVolumeExportPolicyToHclTerraform = exports.netappVolumeExportPolicyToTerraform = exports.NetappVolumeExportPolicyRulesList = exports.NetappVolumeExportPolicyRulesOutputReference = exports.netappVolumeExportPolicyRulesToHclTerraform = exports.netappVolumeExportPolicyRulesToTerraform = exports.NetappVolumeBackupConfigOutputReference = exports.netappVolumeBackupConfigToHclTerraform = exports.netappVolumeBackupConfigToTerraform = exports.NetappVolumeMountOptionsList = exports.NetappVolumeMountOptionsOutputReference = exports.netappVolumeMountOptionsToHclTerraform = exports.netappVolumeMountOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function netappVolumeMountOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.netappVolumeMountOptionsToTerraform = netappVolumeMountOptionsToTerraform;
function netappVolumeMountOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.netappVolumeMountOptionsToHclTerraform = netappVolumeMountOptionsToHclTerraform;
class NetappVolumeMountOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // export - computed: true, optional: false, required: false
    get export() {
        return this.getStringAttribute('export');
    }
    // export_full - computed: true, optional: false, required: false
    get exportFull() {
        return this.getStringAttribute('export_full');
    }
    // instructions - computed: true, optional: false, required: false
    get instructions() {
        return this.getStringAttribute('instructions');
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
}
exports.NetappVolumeMountOptionsOutputReference = NetappVolumeMountOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetappVolumeMountOptionsOutputReference[_a] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeMountOptionsOutputReference", version: "14.13.1" };
class NetappVolumeMountOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetappVolumeMountOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetappVolumeMountOptionsList = NetappVolumeMountOptionsList;
_b = JSII_RTTI_SYMBOL_1;
NetappVolumeMountOptionsList[_b] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeMountOptionsList", version: "14.13.1" };
function netappVolumeBackupConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        backup_policies: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.backupPolicies),
        backup_vault: cdktf.stringToTerraform(struct.backupVault),
        scheduled_backup_enabled: cdktf.booleanToTerraform(struct.scheduledBackupEnabled),
    };
}
exports.netappVolumeBackupConfigToTerraform = netappVolumeBackupConfigToTerraform;
function netappVolumeBackupConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        backup_policies: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.backupPolicies),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        backup_vault: {
            value: cdktf.stringToHclTerraform(struct.backupVault),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        scheduled_backup_enabled: {
            value: cdktf.booleanToHclTerraform(struct.scheduledBackupEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeBackupConfigToHclTerraform = netappVolumeBackupConfigToHclTerraform;
class NetappVolumeBackupConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._backupPolicies !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupPolicies = this._backupPolicies;
        }
        if (this._backupVault !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupVault = this._backupVault;
        }
        if (this._scheduledBackupEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.scheduledBackupEnabled = this._scheduledBackupEnabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._backupPolicies = undefined;
            this._backupVault = undefined;
            this._scheduledBackupEnabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._backupPolicies = value.backupPolicies;
            this._backupVault = value.backupVault;
            this._scheduledBackupEnabled = value.scheduledBackupEnabled;
        }
    }
    get backupPolicies() {
        return this.getListAttribute('backup_policies');
    }
    set backupPolicies(value) {
        this._backupPolicies = value;
    }
    resetBackupPolicies() {
        this._backupPolicies = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupPoliciesInput() {
        return this._backupPolicies;
    }
    get backupVault() {
        return this.getStringAttribute('backup_vault');
    }
    set backupVault(value) {
        this._backupVault = value;
    }
    resetBackupVault() {
        this._backupVault = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupVaultInput() {
        return this._backupVault;
    }
    get scheduledBackupEnabled() {
        return this.getBooleanAttribute('scheduled_backup_enabled');
    }
    set scheduledBackupEnabled(value) {
        this._scheduledBackupEnabled = value;
    }
    resetScheduledBackupEnabled() {
        this._scheduledBackupEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduledBackupEnabledInput() {
        return this._scheduledBackupEnabled;
    }
}
exports.NetappVolumeBackupConfigOutputReference = NetappVolumeBackupConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetappVolumeBackupConfigOutputReference[_c] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeBackupConfigOutputReference", version: "14.13.1" };
function netappVolumeExportPolicyRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access_type: cdktf.stringToTerraform(struct.accessType),
        allowed_clients: cdktf.stringToTerraform(struct.allowedClients),
        has_root_access: cdktf.stringToTerraform(struct.hasRootAccess),
        kerberos5_read_only: cdktf.booleanToTerraform(struct.kerberos5ReadOnly),
        kerberos5_read_write: cdktf.booleanToTerraform(struct.kerberos5ReadWrite),
        kerberos5i_read_only: cdktf.booleanToTerraform(struct.kerberos5IReadOnly),
        kerberos5i_read_write: cdktf.booleanToTerraform(struct.kerberos5IReadWrite),
        kerberos5p_read_only: cdktf.booleanToTerraform(struct.kerberos5PReadOnly),
        kerberos5p_read_write: cdktf.booleanToTerraform(struct.kerberos5PReadWrite),
        nfsv3: cdktf.booleanToTerraform(struct.nfsv3),
        nfsv4: cdktf.booleanToTerraform(struct.nfsv4),
    };
}
exports.netappVolumeExportPolicyRulesToTerraform = netappVolumeExportPolicyRulesToTerraform;
function netappVolumeExportPolicyRulesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        access_type: {
            value: cdktf.stringToHclTerraform(struct.accessType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        allowed_clients: {
            value: cdktf.stringToHclTerraform(struct.allowedClients),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        has_root_access: {
            value: cdktf.stringToHclTerraform(struct.hasRootAccess),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kerberos5_read_only: {
            value: cdktf.booleanToHclTerraform(struct.kerberos5ReadOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        kerberos5_read_write: {
            value: cdktf.booleanToHclTerraform(struct.kerberos5ReadWrite),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        kerberos5i_read_only: {
            value: cdktf.booleanToHclTerraform(struct.kerberos5IReadOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        kerberos5i_read_write: {
            value: cdktf.booleanToHclTerraform(struct.kerberos5IReadWrite),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        kerberos5p_read_only: {
            value: cdktf.booleanToHclTerraform(struct.kerberos5PReadOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        kerberos5p_read_write: {
            value: cdktf.booleanToHclTerraform(struct.kerberos5PReadWrite),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        nfsv3: {
            value: cdktf.booleanToHclTerraform(struct.nfsv3),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        nfsv4: {
            value: cdktf.booleanToHclTerraform(struct.nfsv4),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeExportPolicyRulesToHclTerraform = netappVolumeExportPolicyRulesToHclTerraform;
class NetappVolumeExportPolicyRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accessType !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessType = this._accessType;
        }
        if (this._allowedClients !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowedClients = this._allowedClients;
        }
        if (this._hasRootAccess !== undefined) {
            hasAnyValues = true;
            internalValueResult.hasRootAccess = this._hasRootAccess;
        }
        if (this._kerberos5ReadOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.kerberos5ReadOnly = this._kerberos5ReadOnly;
        }
        if (this._kerberos5ReadWrite !== undefined) {
            hasAnyValues = true;
            internalValueResult.kerberos5ReadWrite = this._kerberos5ReadWrite;
        }
        if (this._kerberos5IReadOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.kerberos5IReadOnly = this._kerberos5IReadOnly;
        }
        if (this._kerberos5IReadWrite !== undefined) {
            hasAnyValues = true;
            internalValueResult.kerberos5IReadWrite = this._kerberos5IReadWrite;
        }
        if (this._kerberos5PReadOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.kerberos5PReadOnly = this._kerberos5PReadOnly;
        }
        if (this._kerberos5PReadWrite !== undefined) {
            hasAnyValues = true;
            internalValueResult.kerberos5PReadWrite = this._kerberos5PReadWrite;
        }
        if (this._nfsv3 !== undefined) {
            hasAnyValues = true;
            internalValueResult.nfsv3 = this._nfsv3;
        }
        if (this._nfsv4 !== undefined) {
            hasAnyValues = true;
            internalValueResult.nfsv4 = this._nfsv4;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._accessType = undefined;
            this._allowedClients = undefined;
            this._hasRootAccess = undefined;
            this._kerberos5ReadOnly = undefined;
            this._kerberos5ReadWrite = undefined;
            this._kerberos5IReadOnly = undefined;
            this._kerberos5IReadWrite = undefined;
            this._kerberos5PReadOnly = undefined;
            this._kerberos5PReadWrite = undefined;
            this._nfsv3 = undefined;
            this._nfsv4 = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._accessType = value.accessType;
            this._allowedClients = value.allowedClients;
            this._hasRootAccess = value.hasRootAccess;
            this._kerberos5ReadOnly = value.kerberos5ReadOnly;
            this._kerberos5ReadWrite = value.kerberos5ReadWrite;
            this._kerberos5IReadOnly = value.kerberos5IReadOnly;
            this._kerberos5IReadWrite = value.kerberos5IReadWrite;
            this._kerberos5PReadOnly = value.kerberos5PReadOnly;
            this._kerberos5PReadWrite = value.kerberos5PReadWrite;
            this._nfsv3 = value.nfsv3;
            this._nfsv4 = value.nfsv4;
        }
    }
    get accessType() {
        return this.getStringAttribute('access_type');
    }
    set accessType(value) {
        this._accessType = value;
    }
    resetAccessType() {
        this._accessType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessTypeInput() {
        return this._accessType;
    }
    get allowedClients() {
        return this.getStringAttribute('allowed_clients');
    }
    set allowedClients(value) {
        this._allowedClients = value;
    }
    resetAllowedClients() {
        this._allowedClients = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowedClientsInput() {
        return this._allowedClients;
    }
    get hasRootAccess() {
        return this.getStringAttribute('has_root_access');
    }
    set hasRootAccess(value) {
        this._hasRootAccess = value;
    }
    resetHasRootAccess() {
        this._hasRootAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hasRootAccessInput() {
        return this._hasRootAccess;
    }
    get kerberos5ReadOnly() {
        return this.getBooleanAttribute('kerberos5_read_only');
    }
    set kerberos5ReadOnly(value) {
        this._kerberos5ReadOnly = value;
    }
    resetKerberos5ReadOnly() {
        this._kerberos5ReadOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberos5ReadOnlyInput() {
        return this._kerberos5ReadOnly;
    }
    get kerberos5ReadWrite() {
        return this.getBooleanAttribute('kerberos5_read_write');
    }
    set kerberos5ReadWrite(value) {
        this._kerberos5ReadWrite = value;
    }
    resetKerberos5ReadWrite() {
        this._kerberos5ReadWrite = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberos5ReadWriteInput() {
        return this._kerberos5ReadWrite;
    }
    get kerberos5IReadOnly() {
        return this.getBooleanAttribute('kerberos5i_read_only');
    }
    set kerberos5IReadOnly(value) {
        this._kerberos5IReadOnly = value;
    }
    resetKerberos5IReadOnly() {
        this._kerberos5IReadOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberos5IReadOnlyInput() {
        return this._kerberos5IReadOnly;
    }
    get kerberos5IReadWrite() {
        return this.getBooleanAttribute('kerberos5i_read_write');
    }
    set kerberos5IReadWrite(value) {
        this._kerberos5IReadWrite = value;
    }
    resetKerberos5IReadWrite() {
        this._kerberos5IReadWrite = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberos5IReadWriteInput() {
        return this._kerberos5IReadWrite;
    }
    get kerberos5PReadOnly() {
        return this.getBooleanAttribute('kerberos5p_read_only');
    }
    set kerberos5PReadOnly(value) {
        this._kerberos5PReadOnly = value;
    }
    resetKerberos5PReadOnly() {
        this._kerberos5PReadOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberos5PReadOnlyInput() {
        return this._kerberos5PReadOnly;
    }
    get kerberos5PReadWrite() {
        return this.getBooleanAttribute('kerberos5p_read_write');
    }
    set kerberos5PReadWrite(value) {
        this._kerberos5PReadWrite = value;
    }
    resetKerberos5PReadWrite() {
        this._kerberos5PReadWrite = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberos5PReadWriteInput() {
        return this._kerberos5PReadWrite;
    }
    get nfsv3() {
        return this.getBooleanAttribute('nfsv3');
    }
    set nfsv3(value) {
        this._nfsv3 = value;
    }
    resetNfsv3() {
        this._nfsv3 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nfsv3Input() {
        return this._nfsv3;
    }
    get nfsv4() {
        return this.getBooleanAttribute('nfsv4');
    }
    set nfsv4(value) {
        this._nfsv4 = value;
    }
    resetNfsv4() {
        this._nfsv4 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nfsv4Input() {
        return this._nfsv4;
    }
}
exports.NetappVolumeExportPolicyRulesOutputReference = NetappVolumeExportPolicyRulesOutputReference;
_d = JSII_RTTI_SYMBOL_1;
NetappVolumeExportPolicyRulesOutputReference[_d] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeExportPolicyRulesOutputReference", version: "14.13.1" };
class NetappVolumeExportPolicyRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetappVolumeExportPolicyRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetappVolumeExportPolicyRulesList = NetappVolumeExportPolicyRulesList;
_e = JSII_RTTI_SYMBOL_1;
NetappVolumeExportPolicyRulesList[_e] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeExportPolicyRulesList", version: "14.13.1" };
function netappVolumeExportPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        rules: cdktf.listMapper(netappVolumeExportPolicyRulesToTerraform, true)(struct.rules),
    };
}
exports.netappVolumeExportPolicyToTerraform = netappVolumeExportPolicyToTerraform;
function netappVolumeExportPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        rules: {
            value: cdktf.listMapperHcl(netappVolumeExportPolicyRulesToHclTerraform, true)(struct.rules),
            isBlock: true,
            type: "list",
            storageClassType: "NetappVolumeExportPolicyRulesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeExportPolicyToHclTerraform = netappVolumeExportPolicyToHclTerraform;
class NetappVolumeExportPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // rules - computed: false, optional: false, required: true
        this._rules = new NetappVolumeExportPolicyRulesList(this, "rules", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rules?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rules = this._rules?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rules.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rules.internalValue = value.rules;
        }
    }
    get rules() {
        return this._rules;
    }
    putRules(value) {
        this._rules.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get rulesInput() {
        return this._rules.internalValue;
    }
}
exports.NetappVolumeExportPolicyOutputReference = NetappVolumeExportPolicyOutputReference;
_f = JSII_RTTI_SYMBOL_1;
NetappVolumeExportPolicyOutputReference[_f] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeExportPolicyOutputReference", version: "14.13.1" };
function netappVolumeRestoreParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        source_backup: cdktf.stringToTerraform(struct.sourceBackup),
        source_snapshot: cdktf.stringToTerraform(struct.sourceSnapshot),
    };
}
exports.netappVolumeRestoreParametersToTerraform = netappVolumeRestoreParametersToTerraform;
function netappVolumeRestoreParametersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        source_backup: {
            value: cdktf.stringToHclTerraform(struct.sourceBackup),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_snapshot: {
            value: cdktf.stringToHclTerraform(struct.sourceSnapshot),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeRestoreParametersToHclTerraform = netappVolumeRestoreParametersToHclTerraform;
class NetappVolumeRestoreParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sourceBackup !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceBackup = this._sourceBackup;
        }
        if (this._sourceSnapshot !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceSnapshot = this._sourceSnapshot;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sourceBackup = undefined;
            this._sourceSnapshot = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sourceBackup = value.sourceBackup;
            this._sourceSnapshot = value.sourceSnapshot;
        }
    }
    get sourceBackup() {
        return this.getStringAttribute('source_backup');
    }
    set sourceBackup(value) {
        this._sourceBackup = value;
    }
    resetSourceBackup() {
        this._sourceBackup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceBackupInput() {
        return this._sourceBackup;
    }
    get sourceSnapshot() {
        return this.getStringAttribute('source_snapshot');
    }
    set sourceSnapshot(value) {
        this._sourceSnapshot = value;
    }
    resetSourceSnapshot() {
        this._sourceSnapshot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceSnapshotInput() {
        return this._sourceSnapshot;
    }
}
exports.NetappVolumeRestoreParametersOutputReference = NetappVolumeRestoreParametersOutputReference;
_g = JSII_RTTI_SYMBOL_1;
NetappVolumeRestoreParametersOutputReference[_g] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeRestoreParametersOutputReference", version: "14.13.1" };
function netappVolumeSnapshotPolicyDailyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hour: cdktf.numberToTerraform(struct.hour),
        minute: cdktf.numberToTerraform(struct.minute),
        snapshots_to_keep: cdktf.numberToTerraform(struct.snapshotsToKeep),
    };
}
exports.netappVolumeSnapshotPolicyDailyScheduleToTerraform = netappVolumeSnapshotPolicyDailyScheduleToTerraform;
function netappVolumeSnapshotPolicyDailyScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        hour: {
            value: cdktf.numberToHclTerraform(struct.hour),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minute: {
            value: cdktf.numberToHclTerraform(struct.minute),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        snapshots_to_keep: {
            value: cdktf.numberToHclTerraform(struct.snapshotsToKeep),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeSnapshotPolicyDailyScheduleToHclTerraform = netappVolumeSnapshotPolicyDailyScheduleToHclTerraform;
class NetappVolumeSnapshotPolicyDailyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hour !== undefined) {
            hasAnyValues = true;
            internalValueResult.hour = this._hour;
        }
        if (this._minute !== undefined) {
            hasAnyValues = true;
            internalValueResult.minute = this._minute;
        }
        if (this._snapshotsToKeep !== undefined) {
            hasAnyValues = true;
            internalValueResult.snapshotsToKeep = this._snapshotsToKeep;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hour = undefined;
            this._minute = undefined;
            this._snapshotsToKeep = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hour = value.hour;
            this._minute = value.minute;
            this._snapshotsToKeep = value.snapshotsToKeep;
        }
    }
    get hour() {
        return this.getNumberAttribute('hour');
    }
    set hour(value) {
        this._hour = value;
    }
    resetHour() {
        this._hour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourInput() {
        return this._hour;
    }
    get minute() {
        return this.getNumberAttribute('minute');
    }
    set minute(value) {
        this._minute = value;
    }
    resetMinute() {
        this._minute = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minuteInput() {
        return this._minute;
    }
    get snapshotsToKeep() {
        return this.getNumberAttribute('snapshots_to_keep');
    }
    set snapshotsToKeep(value) {
        this._snapshotsToKeep = value;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotsToKeepInput() {
        return this._snapshotsToKeep;
    }
}
exports.NetappVolumeSnapshotPolicyDailyScheduleOutputReference = NetappVolumeSnapshotPolicyDailyScheduleOutputReference;
_h = JSII_RTTI_SYMBOL_1;
NetappVolumeSnapshotPolicyDailyScheduleOutputReference[_h] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeSnapshotPolicyDailyScheduleOutputReference", version: "14.13.1" };
function netappVolumeSnapshotPolicyHourlyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        minute: cdktf.numberToTerraform(struct.minute),
        snapshots_to_keep: cdktf.numberToTerraform(struct.snapshotsToKeep),
    };
}
exports.netappVolumeSnapshotPolicyHourlyScheduleToTerraform = netappVolumeSnapshotPolicyHourlyScheduleToTerraform;
function netappVolumeSnapshotPolicyHourlyScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        minute: {
            value: cdktf.numberToHclTerraform(struct.minute),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        snapshots_to_keep: {
            value: cdktf.numberToHclTerraform(struct.snapshotsToKeep),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeSnapshotPolicyHourlyScheduleToHclTerraform = netappVolumeSnapshotPolicyHourlyScheduleToHclTerraform;
class NetappVolumeSnapshotPolicyHourlyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._minute !== undefined) {
            hasAnyValues = true;
            internalValueResult.minute = this._minute;
        }
        if (this._snapshotsToKeep !== undefined) {
            hasAnyValues = true;
            internalValueResult.snapshotsToKeep = this._snapshotsToKeep;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._minute = undefined;
            this._snapshotsToKeep = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._minute = value.minute;
            this._snapshotsToKeep = value.snapshotsToKeep;
        }
    }
    get minute() {
        return this.getNumberAttribute('minute');
    }
    set minute(value) {
        this._minute = value;
    }
    resetMinute() {
        this._minute = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minuteInput() {
        return this._minute;
    }
    get snapshotsToKeep() {
        return this.getNumberAttribute('snapshots_to_keep');
    }
    set snapshotsToKeep(value) {
        this._snapshotsToKeep = value;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotsToKeepInput() {
        return this._snapshotsToKeep;
    }
}
exports.NetappVolumeSnapshotPolicyHourlyScheduleOutputReference = NetappVolumeSnapshotPolicyHourlyScheduleOutputReference;
_j = JSII_RTTI_SYMBOL_1;
NetappVolumeSnapshotPolicyHourlyScheduleOutputReference[_j] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeSnapshotPolicyHourlyScheduleOutputReference", version: "14.13.1" };
function netappVolumeSnapshotPolicyMonthlyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        days_of_month: cdktf.stringToTerraform(struct.daysOfMonth),
        hour: cdktf.numberToTerraform(struct.hour),
        minute: cdktf.numberToTerraform(struct.minute),
        snapshots_to_keep: cdktf.numberToTerraform(struct.snapshotsToKeep),
    };
}
exports.netappVolumeSnapshotPolicyMonthlyScheduleToTerraform = netappVolumeSnapshotPolicyMonthlyScheduleToTerraform;
function netappVolumeSnapshotPolicyMonthlyScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        days_of_month: {
            value: cdktf.stringToHclTerraform(struct.daysOfMonth),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        hour: {
            value: cdktf.numberToHclTerraform(struct.hour),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minute: {
            value: cdktf.numberToHclTerraform(struct.minute),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        snapshots_to_keep: {
            value: cdktf.numberToHclTerraform(struct.snapshotsToKeep),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeSnapshotPolicyMonthlyScheduleToHclTerraform = netappVolumeSnapshotPolicyMonthlyScheduleToHclTerraform;
class NetappVolumeSnapshotPolicyMonthlyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._daysOfMonth !== undefined) {
            hasAnyValues = true;
            internalValueResult.daysOfMonth = this._daysOfMonth;
        }
        if (this._hour !== undefined) {
            hasAnyValues = true;
            internalValueResult.hour = this._hour;
        }
        if (this._minute !== undefined) {
            hasAnyValues = true;
            internalValueResult.minute = this._minute;
        }
        if (this._snapshotsToKeep !== undefined) {
            hasAnyValues = true;
            internalValueResult.snapshotsToKeep = this._snapshotsToKeep;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._daysOfMonth = undefined;
            this._hour = undefined;
            this._minute = undefined;
            this._snapshotsToKeep = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._daysOfMonth = value.daysOfMonth;
            this._hour = value.hour;
            this._minute = value.minute;
            this._snapshotsToKeep = value.snapshotsToKeep;
        }
    }
    get daysOfMonth() {
        return this.getStringAttribute('days_of_month');
    }
    set daysOfMonth(value) {
        this._daysOfMonth = value;
    }
    resetDaysOfMonth() {
        this._daysOfMonth = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get daysOfMonthInput() {
        return this._daysOfMonth;
    }
    get hour() {
        return this.getNumberAttribute('hour');
    }
    set hour(value) {
        this._hour = value;
    }
    resetHour() {
        this._hour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourInput() {
        return this._hour;
    }
    get minute() {
        return this.getNumberAttribute('minute');
    }
    set minute(value) {
        this._minute = value;
    }
    resetMinute() {
        this._minute = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minuteInput() {
        return this._minute;
    }
    get snapshotsToKeep() {
        return this.getNumberAttribute('snapshots_to_keep');
    }
    set snapshotsToKeep(value) {
        this._snapshotsToKeep = value;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotsToKeepInput() {
        return this._snapshotsToKeep;
    }
}
exports.NetappVolumeSnapshotPolicyMonthlyScheduleOutputReference = NetappVolumeSnapshotPolicyMonthlyScheduleOutputReference;
_k = JSII_RTTI_SYMBOL_1;
NetappVolumeSnapshotPolicyMonthlyScheduleOutputReference[_k] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeSnapshotPolicyMonthlyScheduleOutputReference", version: "14.13.1" };
function netappVolumeSnapshotPolicyWeeklyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        hour: cdktf.numberToTerraform(struct.hour),
        minute: cdktf.numberToTerraform(struct.minute),
        snapshots_to_keep: cdktf.numberToTerraform(struct.snapshotsToKeep),
    };
}
exports.netappVolumeSnapshotPolicyWeeklyScheduleToTerraform = netappVolumeSnapshotPolicyWeeklyScheduleToTerraform;
function netappVolumeSnapshotPolicyWeeklyScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        day: {
            value: cdktf.stringToHclTerraform(struct.day),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        hour: {
            value: cdktf.numberToHclTerraform(struct.hour),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minute: {
            value: cdktf.numberToHclTerraform(struct.minute),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        snapshots_to_keep: {
            value: cdktf.numberToHclTerraform(struct.snapshotsToKeep),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeSnapshotPolicyWeeklyScheduleToHclTerraform = netappVolumeSnapshotPolicyWeeklyScheduleToHclTerraform;
class NetappVolumeSnapshotPolicyWeeklyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._hour !== undefined) {
            hasAnyValues = true;
            internalValueResult.hour = this._hour;
        }
        if (this._minute !== undefined) {
            hasAnyValues = true;
            internalValueResult.minute = this._minute;
        }
        if (this._snapshotsToKeep !== undefined) {
            hasAnyValues = true;
            internalValueResult.snapshotsToKeep = this._snapshotsToKeep;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._hour = undefined;
            this._minute = undefined;
            this._snapshotsToKeep = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._hour = value.hour;
            this._minute = value.minute;
            this._snapshotsToKeep = value.snapshotsToKeep;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    resetDay() {
        this._day = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get hour() {
        return this.getNumberAttribute('hour');
    }
    set hour(value) {
        this._hour = value;
    }
    resetHour() {
        this._hour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourInput() {
        return this._hour;
    }
    get minute() {
        return this.getNumberAttribute('minute');
    }
    set minute(value) {
        this._minute = value;
    }
    resetMinute() {
        this._minute = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minuteInput() {
        return this._minute;
    }
    get snapshotsToKeep() {
        return this.getNumberAttribute('snapshots_to_keep');
    }
    set snapshotsToKeep(value) {
        this._snapshotsToKeep = value;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotsToKeepInput() {
        return this._snapshotsToKeep;
    }
}
exports.NetappVolumeSnapshotPolicyWeeklyScheduleOutputReference = NetappVolumeSnapshotPolicyWeeklyScheduleOutputReference;
_l = JSII_RTTI_SYMBOL_1;
NetappVolumeSnapshotPolicyWeeklyScheduleOutputReference[_l] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeSnapshotPolicyWeeklyScheduleOutputReference", version: "14.13.1" };
function netappVolumeSnapshotPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        daily_schedule: netappVolumeSnapshotPolicyDailyScheduleToTerraform(struct.dailySchedule),
        hourly_schedule: netappVolumeSnapshotPolicyHourlyScheduleToTerraform(struct.hourlySchedule),
        monthly_schedule: netappVolumeSnapshotPolicyMonthlyScheduleToTerraform(struct.monthlySchedule),
        weekly_schedule: netappVolumeSnapshotPolicyWeeklyScheduleToTerraform(struct.weeklySchedule),
    };
}
exports.netappVolumeSnapshotPolicyToTerraform = netappVolumeSnapshotPolicyToTerraform;
function netappVolumeSnapshotPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        daily_schedule: {
            value: netappVolumeSnapshotPolicyDailyScheduleToHclTerraform(struct.dailySchedule),
            isBlock: true,
            type: "list",
            storageClassType: "NetappVolumeSnapshotPolicyDailyScheduleList",
        },
        hourly_schedule: {
            value: netappVolumeSnapshotPolicyHourlyScheduleToHclTerraform(struct.hourlySchedule),
            isBlock: true,
            type: "list",
            storageClassType: "NetappVolumeSnapshotPolicyHourlyScheduleList",
        },
        monthly_schedule: {
            value: netappVolumeSnapshotPolicyMonthlyScheduleToHclTerraform(struct.monthlySchedule),
            isBlock: true,
            type: "list",
            storageClassType: "NetappVolumeSnapshotPolicyMonthlyScheduleList",
        },
        weekly_schedule: {
            value: netappVolumeSnapshotPolicyWeeklyScheduleToHclTerraform(struct.weeklySchedule),
            isBlock: true,
            type: "list",
            storageClassType: "NetappVolumeSnapshotPolicyWeeklyScheduleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeSnapshotPolicyToHclTerraform = netappVolumeSnapshotPolicyToHclTerraform;
class NetappVolumeSnapshotPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // daily_schedule - computed: false, optional: true, required: false
        this._dailySchedule = new NetappVolumeSnapshotPolicyDailyScheduleOutputReference(this, "daily_schedule");
        // hourly_schedule - computed: false, optional: true, required: false
        this._hourlySchedule = new NetappVolumeSnapshotPolicyHourlyScheduleOutputReference(this, "hourly_schedule");
        // monthly_schedule - computed: false, optional: true, required: false
        this._monthlySchedule = new NetappVolumeSnapshotPolicyMonthlyScheduleOutputReference(this, "monthly_schedule");
        // weekly_schedule - computed: false, optional: true, required: false
        this._weeklySchedule = new NetappVolumeSnapshotPolicyWeeklyScheduleOutputReference(this, "weekly_schedule");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._dailySchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dailySchedule = this._dailySchedule?.internalValue;
        }
        if (this._hourlySchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hourlySchedule = this._hourlySchedule?.internalValue;
        }
        if (this._monthlySchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.monthlySchedule = this._monthlySchedule?.internalValue;
        }
        if (this._weeklySchedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.weeklySchedule = this._weeklySchedule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._dailySchedule.internalValue = undefined;
            this._hourlySchedule.internalValue = undefined;
            this._monthlySchedule.internalValue = undefined;
            this._weeklySchedule.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._dailySchedule.internalValue = value.dailySchedule;
            this._hourlySchedule.internalValue = value.hourlySchedule;
            this._monthlySchedule.internalValue = value.monthlySchedule;
            this._weeklySchedule.internalValue = value.weeklySchedule;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get dailySchedule() {
        return this._dailySchedule;
    }
    putDailySchedule(value) {
        this._dailySchedule.internalValue = value;
    }
    resetDailySchedule() {
        this._dailySchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dailyScheduleInput() {
        return this._dailySchedule.internalValue;
    }
    get hourlySchedule() {
        return this._hourlySchedule;
    }
    putHourlySchedule(value) {
        this._hourlySchedule.internalValue = value;
    }
    resetHourlySchedule() {
        this._hourlySchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourlyScheduleInput() {
        return this._hourlySchedule.internalValue;
    }
    get monthlySchedule() {
        return this._monthlySchedule;
    }
    putMonthlySchedule(value) {
        this._monthlySchedule.internalValue = value;
    }
    resetMonthlySchedule() {
        this._monthlySchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get monthlyScheduleInput() {
        return this._monthlySchedule.internalValue;
    }
    get weeklySchedule() {
        return this._weeklySchedule;
    }
    putWeeklySchedule(value) {
        this._weeklySchedule.internalValue = value;
    }
    resetWeeklySchedule() {
        this._weeklySchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get weeklyScheduleInput() {
        return this._weeklySchedule.internalValue;
    }
}
exports.NetappVolumeSnapshotPolicyOutputReference = NetappVolumeSnapshotPolicyOutputReference;
_m = JSII_RTTI_SYMBOL_1;
NetappVolumeSnapshotPolicyOutputReference[_m] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeSnapshotPolicyOutputReference", version: "14.13.1" };
function netappVolumeTieringPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cooling_threshold_days: cdktf.numberToTerraform(struct.coolingThresholdDays),
        tier_action: cdktf.stringToTerraform(struct.tierAction),
    };
}
exports.netappVolumeTieringPolicyToTerraform = netappVolumeTieringPolicyToTerraform;
function netappVolumeTieringPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cooling_threshold_days: {
            value: cdktf.numberToHclTerraform(struct.coolingThresholdDays),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        tier_action: {
            value: cdktf.stringToHclTerraform(struct.tierAction),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeTieringPolicyToHclTerraform = netappVolumeTieringPolicyToHclTerraform;
class NetappVolumeTieringPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coolingThresholdDays !== undefined) {
            hasAnyValues = true;
            internalValueResult.coolingThresholdDays = this._coolingThresholdDays;
        }
        if (this._tierAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.tierAction = this._tierAction;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._coolingThresholdDays = undefined;
            this._tierAction = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._coolingThresholdDays = value.coolingThresholdDays;
            this._tierAction = value.tierAction;
        }
    }
    get coolingThresholdDays() {
        return this.getNumberAttribute('cooling_threshold_days');
    }
    set coolingThresholdDays(value) {
        this._coolingThresholdDays = value;
    }
    resetCoolingThresholdDays() {
        this._coolingThresholdDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coolingThresholdDaysInput() {
        return this._coolingThresholdDays;
    }
    get tierAction() {
        return this.getStringAttribute('tier_action');
    }
    set tierAction(value) {
        this._tierAction = value;
    }
    resetTierAction() {
        this._tierAction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tierActionInput() {
        return this._tierAction;
    }
}
exports.NetappVolumeTieringPolicyOutputReference = NetappVolumeTieringPolicyOutputReference;
_o = JSII_RTTI_SYMBOL_1;
NetappVolumeTieringPolicyOutputReference[_o] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeTieringPolicyOutputReference", version: "14.13.1" };
function netappVolumeTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.netappVolumeTimeoutsToTerraform = netappVolumeTimeoutsToTerraform;
function netappVolumeTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.netappVolumeTimeoutsToHclTerraform = netappVolumeTimeoutsToHclTerraform;
class NetappVolumeTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetappVolumeTimeoutsOutputReference = NetappVolumeTimeoutsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
NetappVolumeTimeoutsOutputReference[_p] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolumeTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/netapp_volume google_netapp_volume}
*/
class NetappVolume extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetappVolume resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetappVolume to import
    * @param importFromId The id of the existing NetappVolume that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/netapp_volume#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetappVolume to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_netapp_volume", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/netapp_volume google_netapp_volume} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetappVolumeConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_netapp_volume',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // mount_options - computed: true, optional: false, required: false
        this._mountOptions = new NetappVolumeMountOptionsList(this, "mount_options", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // backup_config - computed: false, optional: true, required: false
        this._backupConfig = new NetappVolumeBackupConfigOutputReference(this, "backup_config");
        // export_policy - computed: false, optional: true, required: false
        this._exportPolicy = new NetappVolumeExportPolicyOutputReference(this, "export_policy");
        // restore_parameters - computed: false, optional: true, required: false
        this._restoreParameters = new NetappVolumeRestoreParametersOutputReference(this, "restore_parameters");
        // snapshot_policy - computed: false, optional: true, required: false
        this._snapshotPolicy = new NetappVolumeSnapshotPolicyOutputReference(this, "snapshot_policy");
        // tiering_policy - computed: false, optional: true, required: false
        this._tieringPolicy = new NetappVolumeTieringPolicyOutputReference(this, "tiering_policy");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetappVolumeTimeoutsOutputReference(this, "timeouts");
        this._capacityGib = config.capacityGib;
        this._deletionPolicy = config.deletionPolicy;
        this._description = config.description;
        this._id = config.id;
        this._kerberosEnabled = config.kerberosEnabled;
        this._labels = config.labels;
        this._largeCapacity = config.largeCapacity;
        this._location = config.location;
        this._multipleEndpoints = config.multipleEndpoints;
        this._name = config.name;
        this._project = config.project;
        this._protocols = config.protocols;
        this._restrictedActions = config.restrictedActions;
        this._securityStyle = config.securityStyle;
        this._shareName = config.shareName;
        this._smbSettings = config.smbSettings;
        this._snapshotDirectory = config.snapshotDirectory;
        this._storagePool = config.storagePool;
        this._unixPermissions = config.unixPermissions;
        this._backupConfig.internalValue = config.backupConfig;
        this._exportPolicy.internalValue = config.exportPolicy;
        this._restoreParameters.internalValue = config.restoreParameters;
        this._snapshotPolicy.internalValue = config.snapshotPolicy;
        this._tieringPolicy.internalValue = config.tieringPolicy;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // active_directory - computed: true, optional: false, required: false
    get activeDirectory() {
        return this.getStringAttribute('active_directory');
    }
    get capacityGib() {
        return this.getStringAttribute('capacity_gib');
    }
    set capacityGib(value) {
        this._capacityGib = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityGibInput() {
        return this._capacityGib;
    }
    // cold_tier_size_gib - computed: true, optional: false, required: false
    get coldTierSizeGib() {
        return this.getStringAttribute('cold_tier_size_gib');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get deletionPolicy() {
        return this.getStringAttribute('deletion_policy');
    }
    set deletionPolicy(value) {
        this._deletionPolicy = value;
    }
    resetDeletionPolicy() {
        this._deletionPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionPolicyInput() {
        return this._deletionPolicy;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // encryption_type - computed: true, optional: false, required: false
    get encryptionType() {
        return this.getStringAttribute('encryption_type');
    }
    // has_replication - computed: true, optional: false, required: false
    get hasReplication() {
        return this.getBooleanAttribute('has_replication');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kerberosEnabled() {
        return this.getBooleanAttribute('kerberos_enabled');
    }
    set kerberosEnabled(value) {
        this._kerberosEnabled = value;
    }
    resetKerberosEnabled() {
        this._kerberosEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kerberosEnabledInput() {
        return this._kerberosEnabled;
    }
    // kms_config - computed: true, optional: false, required: false
    get kmsConfig() {
        return this.getStringAttribute('kms_config');
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get largeCapacity() {
        return this.getBooleanAttribute('large_capacity');
    }
    set largeCapacity(value) {
        this._largeCapacity = value;
    }
    resetLargeCapacity() {
        this._largeCapacity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get largeCapacityInput() {
        return this._largeCapacity;
    }
    // ldap_enabled - computed: true, optional: false, required: false
    get ldapEnabled() {
        return this.getBooleanAttribute('ldap_enabled');
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get mountOptions() {
        return this._mountOptions;
    }
    get multipleEndpoints() {
        return this.getBooleanAttribute('multiple_endpoints');
    }
    set multipleEndpoints(value) {
        this._multipleEndpoints = value;
    }
    resetMultipleEndpoints() {
        this._multipleEndpoints = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get multipleEndpointsInput() {
        return this._multipleEndpoints;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get protocols() {
        return this.getListAttribute('protocols');
    }
    set protocols(value) {
        this._protocols = value;
    }
    // Temporarily expose input value. Use with caution.
    get protocolsInput() {
        return this._protocols;
    }
    // psa_range - computed: true, optional: false, required: false
    get psaRange() {
        return this.getStringAttribute('psa_range');
    }
    // replica_zone - computed: true, optional: false, required: false
    get replicaZone() {
        return this.getStringAttribute('replica_zone');
    }
    get restrictedActions() {
        return this.getListAttribute('restricted_actions');
    }
    set restrictedActions(value) {
        this._restrictedActions = value;
    }
    resetRestrictedActions() {
        this._restrictedActions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get restrictedActionsInput() {
        return this._restrictedActions;
    }
    get securityStyle() {
        return this.getStringAttribute('security_style');
    }
    set securityStyle(value) {
        this._securityStyle = value;
    }
    resetSecurityStyle() {
        this._securityStyle = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityStyleInput() {
        return this._securityStyle;
    }
    // service_level - computed: true, optional: false, required: false
    get serviceLevel() {
        return this.getStringAttribute('service_level');
    }
    get shareName() {
        return this.getStringAttribute('share_name');
    }
    set shareName(value) {
        this._shareName = value;
    }
    // Temporarily expose input value. Use with caution.
    get shareNameInput() {
        return this._shareName;
    }
    get smbSettings() {
        return this.getListAttribute('smb_settings');
    }
    set smbSettings(value) {
        this._smbSettings = value;
    }
    resetSmbSettings() {
        this._smbSettings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get smbSettingsInput() {
        return this._smbSettings;
    }
    get snapshotDirectory() {
        return this.getBooleanAttribute('snapshot_directory');
    }
    set snapshotDirectory(value) {
        this._snapshotDirectory = value;
    }
    resetSnapshotDirectory() {
        this._snapshotDirectory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotDirectoryInput() {
        return this._snapshotDirectory;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // state_details - computed: true, optional: false, required: false
    get stateDetails() {
        return this.getStringAttribute('state_details');
    }
    get storagePool() {
        return this.getStringAttribute('storage_pool');
    }
    set storagePool(value) {
        this._storagePool = value;
    }
    // Temporarily expose input value. Use with caution.
    get storagePoolInput() {
        return this._storagePool;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    get unixPermissions() {
        return this.getStringAttribute('unix_permissions');
    }
    set unixPermissions(value) {
        this._unixPermissions = value;
    }
    resetUnixPermissions() {
        this._unixPermissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get unixPermissionsInput() {
        return this._unixPermissions;
    }
    // used_gib - computed: true, optional: false, required: false
    get usedGib() {
        return this.getStringAttribute('used_gib');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
    get backupConfig() {
        return this._backupConfig;
    }
    putBackupConfig(value) {
        this._backupConfig.internalValue = value;
    }
    resetBackupConfig() {
        this._backupConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupConfigInput() {
        return this._backupConfig.internalValue;
    }
    get exportPolicy() {
        return this._exportPolicy;
    }
    putExportPolicy(value) {
        this._exportPolicy.internalValue = value;
    }
    resetExportPolicy() {
        this._exportPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exportPolicyInput() {
        return this._exportPolicy.internalValue;
    }
    get restoreParameters() {
        return this._restoreParameters;
    }
    putRestoreParameters(value) {
        this._restoreParameters.internalValue = value;
    }
    resetRestoreParameters() {
        this._restoreParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get restoreParametersInput() {
        return this._restoreParameters.internalValue;
    }
    get snapshotPolicy() {
        return this._snapshotPolicy;
    }
    putSnapshotPolicy(value) {
        this._snapshotPolicy.internalValue = value;
    }
    resetSnapshotPolicy() {
        this._snapshotPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotPolicyInput() {
        return this._snapshotPolicy.internalValue;
    }
    get tieringPolicy() {
        return this._tieringPolicy;
    }
    putTieringPolicy(value) {
        this._tieringPolicy.internalValue = value;
    }
    resetTieringPolicy() {
        this._tieringPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tieringPolicyInput() {
        return this._tieringPolicy.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            capacity_gib: cdktf.stringToTerraform(this._capacityGib),
            deletion_policy: cdktf.stringToTerraform(this._deletionPolicy),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            kerberos_enabled: cdktf.booleanToTerraform(this._kerberosEnabled),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            large_capacity: cdktf.booleanToTerraform(this._largeCapacity),
            location: cdktf.stringToTerraform(this._location),
            multiple_endpoints: cdktf.booleanToTerraform(this._multipleEndpoints),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            protocols: cdktf.listMapper(cdktf.stringToTerraform, false)(this._protocols),
            restricted_actions: cdktf.listMapper(cdktf.stringToTerraform, false)(this._restrictedActions),
            security_style: cdktf.stringToTerraform(this._securityStyle),
            share_name: cdktf.stringToTerraform(this._shareName),
            smb_settings: cdktf.listMapper(cdktf.stringToTerraform, false)(this._smbSettings),
            snapshot_directory: cdktf.booleanToTerraform(this._snapshotDirectory),
            storage_pool: cdktf.stringToTerraform(this._storagePool),
            unix_permissions: cdktf.stringToTerraform(this._unixPermissions),
            backup_config: netappVolumeBackupConfigToTerraform(this._backupConfig.internalValue),
            export_policy: netappVolumeExportPolicyToTerraform(this._exportPolicy.internalValue),
            restore_parameters: netappVolumeRestoreParametersToTerraform(this._restoreParameters.internalValue),
            snapshot_policy: netappVolumeSnapshotPolicyToTerraform(this._snapshotPolicy.internalValue),
            tiering_policy: netappVolumeTieringPolicyToTerraform(this._tieringPolicy.internalValue),
            timeouts: netappVolumeTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            capacity_gib: {
                value: cdktf.stringToHclTerraform(this._capacityGib),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            deletion_policy: {
                value: cdktf.stringToHclTerraform(this._deletionPolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kerberos_enabled: {
                value: cdktf.booleanToHclTerraform(this._kerberosEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            large_capacity: {
                value: cdktf.booleanToHclTerraform(this._largeCapacity),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            multiple_endpoints: {
                value: cdktf.booleanToHclTerraform(this._multipleEndpoints),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            protocols: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._protocols),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            restricted_actions: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._restrictedActions),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            security_style: {
                value: cdktf.stringToHclTerraform(this._securityStyle),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            share_name: {
                value: cdktf.stringToHclTerraform(this._shareName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            smb_settings: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._smbSettings),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            snapshot_directory: {
                value: cdktf.booleanToHclTerraform(this._snapshotDirectory),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            storage_pool: {
                value: cdktf.stringToHclTerraform(this._storagePool),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            unix_permissions: {
                value: cdktf.stringToHclTerraform(this._unixPermissions),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            backup_config: {
                value: netappVolumeBackupConfigToHclTerraform(this._backupConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetappVolumeBackupConfigList",
            },
            export_policy: {
                value: netappVolumeExportPolicyToHclTerraform(this._exportPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetappVolumeExportPolicyList",
            },
            restore_parameters: {
                value: netappVolumeRestoreParametersToHclTerraform(this._restoreParameters.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetappVolumeRestoreParametersList",
            },
            snapshot_policy: {
                value: netappVolumeSnapshotPolicyToHclTerraform(this._snapshotPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetappVolumeSnapshotPolicyList",
            },
            tiering_policy: {
                value: netappVolumeTieringPolicyToHclTerraform(this._tieringPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetappVolumeTieringPolicyList",
            },
            timeouts: {
                value: netappVolumeTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetappVolumeTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetappVolume = NetappVolume;
_q = JSII_RTTI_SYMBOL_1;
NetappVolume[_q] = { fqn: "@cdktf/provider-google.netappVolume.NetappVolume", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetappVolume.tfResourceType = "google_netapp_volume";
