// @cdktf/provider-google/lib/dataproc-gdc-service-instance/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataprocGdcServiceInstance = exports.DataprocGdcServiceInstanceTimeoutsOutputReference = exports.dataprocGdcServiceInstanceTimeoutsToHclTerraform = exports.dataprocGdcServiceInstanceTimeoutsToTerraform = exports.DataprocGdcServiceInstanceSparkServiceInstanceConfigOutputReference = exports.dataprocGdcServiceInstanceSparkServiceInstanceConfigToHclTerraform = exports.dataprocGdcServiceInstanceSparkServiceInstanceConfigToTerraform = exports.DataprocGdcServiceInstanceGdceClusterOutputReference = exports.dataprocGdcServiceInstanceGdceClusterToHclTerraform = exports.dataprocGdcServiceInstanceGdceClusterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataprocGdcServiceInstanceGdceClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        gdce_cluster: cdktf.stringToTerraform(struct.gdceCluster),
    };
}
exports.dataprocGdcServiceInstanceGdceClusterToTerraform = dataprocGdcServiceInstanceGdceClusterToTerraform;
function dataprocGdcServiceInstanceGdceClusterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        gdce_cluster: {
            value: cdktf.stringToHclTerraform(struct.gdceCluster),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataprocGdcServiceInstanceGdceClusterToHclTerraform = dataprocGdcServiceInstanceGdceClusterToHclTerraform;
class DataprocGdcServiceInstanceGdceClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._gdceCluster !== undefined) {
            hasAnyValues = true;
            internalValueResult.gdceCluster = this._gdceCluster;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._gdceCluster = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._gdceCluster = value.gdceCluster;
        }
    }
    get gdceCluster() {
        return this.getStringAttribute('gdce_cluster');
    }
    set gdceCluster(value) {
        this._gdceCluster = value;
    }
    // Temporarily expose input value. Use with caution.
    get gdceClusterInput() {
        return this._gdceCluster;
    }
}
exports.DataprocGdcServiceInstanceGdceClusterOutputReference = DataprocGdcServiceInstanceGdceClusterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataprocGdcServiceInstanceGdceClusterOutputReference[_a] = { fqn: "@cdktf/provider-google.dataprocGdcServiceInstance.DataprocGdcServiceInstanceGdceClusterOutputReference", version: "14.13.1" };
function dataprocGdcServiceInstanceSparkServiceInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataprocGdcServiceInstanceSparkServiceInstanceConfigToTerraform = dataprocGdcServiceInstanceSparkServiceInstanceConfigToTerraform;
function dataprocGdcServiceInstanceSparkServiceInstanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataprocGdcServiceInstanceSparkServiceInstanceConfigToHclTerraform = dataprocGdcServiceInstanceSparkServiceInstanceConfigToHclTerraform;
class DataprocGdcServiceInstanceSparkServiceInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataprocGdcServiceInstanceSparkServiceInstanceConfigOutputReference = DataprocGdcServiceInstanceSparkServiceInstanceConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DataprocGdcServiceInstanceSparkServiceInstanceConfigOutputReference[_b] = { fqn: "@cdktf/provider-google.dataprocGdcServiceInstance.DataprocGdcServiceInstanceSparkServiceInstanceConfigOutputReference", version: "14.13.1" };
function dataprocGdcServiceInstanceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dataprocGdcServiceInstanceTimeoutsToTerraform = dataprocGdcServiceInstanceTimeoutsToTerraform;
function dataprocGdcServiceInstanceTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataprocGdcServiceInstanceTimeoutsToHclTerraform = dataprocGdcServiceInstanceTimeoutsToHclTerraform;
class DataprocGdcServiceInstanceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DataprocGdcServiceInstanceTimeoutsOutputReference = DataprocGdcServiceInstanceTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataprocGdcServiceInstanceTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-google.dataprocGdcServiceInstance.DataprocGdcServiceInstanceTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataproc_gdc_service_instance google_dataproc_gdc_service_instance}
*/
class DataprocGdcServiceInstance extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataprocGdcServiceInstance resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataprocGdcServiceInstance to import
    * @param importFromId The id of the existing DataprocGdcServiceInstance that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataproc_gdc_service_instance#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataprocGdcServiceInstance to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_dataproc_gdc_service_instance", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataproc_gdc_service_instance google_dataproc_gdc_service_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataprocGdcServiceInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_dataproc_gdc_service_instance',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // gdce_cluster - computed: false, optional: true, required: false
        this._gdceCluster = new DataprocGdcServiceInstanceGdceClusterOutputReference(this, "gdce_cluster");
        // spark_service_instance_config - computed: false, optional: true, required: false
        this._sparkServiceInstanceConfig = new DataprocGdcServiceInstanceSparkServiceInstanceConfigOutputReference(this, "spark_service_instance_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataprocGdcServiceInstanceTimeoutsOutputReference(this, "timeouts");
        this._displayName = config.displayName;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._project = config.project;
        this._serviceAccount = config.serviceAccount;
        this._serviceInstanceId = config.serviceInstanceId;
        this._gdceCluster.internalValue = config.gdceCluster;
        this._sparkServiceInstanceConfig.internalValue = config.sparkServiceInstanceConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // effective_service_account - computed: true, optional: false, required: false
    get effectiveServiceAccount() {
        return this.getStringAttribute('effective_service_account');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // reconciling - computed: true, optional: false, required: false
    get reconciling() {
        return this.getBooleanAttribute('reconciling');
    }
    // requested_state - computed: true, optional: false, required: false
    get requestedState() {
        return this.getStringAttribute('requested_state');
    }
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    set serviceAccount(value) {
        this._serviceAccount = value;
    }
    resetServiceAccount() {
        this._serviceAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountInput() {
        return this._serviceAccount;
    }
    get serviceInstanceId() {
        return this.getStringAttribute('service_instance_id');
    }
    set serviceInstanceId(value) {
        this._serviceInstanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceInstanceIdInput() {
        return this._serviceInstanceId;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // state_message - computed: true, optional: false, required: false
    get stateMessage() {
        return this.getStringAttribute('state_message');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get gdceCluster() {
        return this._gdceCluster;
    }
    putGdceCluster(value) {
        this._gdceCluster.internalValue = value;
    }
    resetGdceCluster() {
        this._gdceCluster.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gdceClusterInput() {
        return this._gdceCluster.internalValue;
    }
    get sparkServiceInstanceConfig() {
        return this._sparkServiceInstanceConfig;
    }
    putSparkServiceInstanceConfig(value) {
        this._sparkServiceInstanceConfig.internalValue = value;
    }
    resetSparkServiceInstanceConfig() {
        this._sparkServiceInstanceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sparkServiceInstanceConfigInput() {
        return this._sparkServiceInstanceConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            service_account: cdktf.stringToTerraform(this._serviceAccount),
            service_instance_id: cdktf.stringToTerraform(this._serviceInstanceId),
            gdce_cluster: dataprocGdcServiceInstanceGdceClusterToTerraform(this._gdceCluster.internalValue),
            spark_service_instance_config: dataprocGdcServiceInstanceSparkServiceInstanceConfigToTerraform(this._sparkServiceInstanceConfig.internalValue),
            timeouts: dataprocGdcServiceInstanceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_account: {
                value: cdktf.stringToHclTerraform(this._serviceAccount),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_instance_id: {
                value: cdktf.stringToHclTerraform(this._serviceInstanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            gdce_cluster: {
                value: dataprocGdcServiceInstanceGdceClusterToHclTerraform(this._gdceCluster.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataprocGdcServiceInstanceGdceClusterList",
            },
            spark_service_instance_config: {
                value: dataprocGdcServiceInstanceSparkServiceInstanceConfigToHclTerraform(this._sparkServiceInstanceConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataprocGdcServiceInstanceSparkServiceInstanceConfigList",
            },
            timeouts: {
                value: dataprocGdcServiceInstanceTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataprocGdcServiceInstanceTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataprocGdcServiceInstance = DataprocGdcServiceInstance;
_d = JSII_RTTI_SYMBOL_1;
DataprocGdcServiceInstance[_d] = { fqn: "@cdktf/provider-google.dataprocGdcServiceInstance.DataprocGdcServiceInstance", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataprocGdcServiceInstance.tfResourceType = "google_dataproc_gdc_service_instance";
