// @cdktf/provider-google/lib/data-google-folder-organization-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleFolderOrganizationPolicy = exports.DataGoogleFolderOrganizationPolicyRestorePolicyList = exports.DataGoogleFolderOrganizationPolicyRestorePolicyOutputReference = exports.dataGoogleFolderOrganizationPolicyRestorePolicyToHclTerraform = exports.dataGoogleFolderOrganizationPolicyRestorePolicyToTerraform = exports.DataGoogleFolderOrganizationPolicyListPolicyList = exports.DataGoogleFolderOrganizationPolicyListPolicyOutputReference = exports.dataGoogleFolderOrganizationPolicyListPolicyToHclTerraform = exports.dataGoogleFolderOrganizationPolicyListPolicyToTerraform = exports.DataGoogleFolderOrganizationPolicyListPolicyDenyList = exports.DataGoogleFolderOrganizationPolicyListPolicyDenyOutputReference = exports.dataGoogleFolderOrganizationPolicyListPolicyDenyToHclTerraform = exports.dataGoogleFolderOrganizationPolicyListPolicyDenyToTerraform = exports.DataGoogleFolderOrganizationPolicyListPolicyAllowList = exports.DataGoogleFolderOrganizationPolicyListPolicyAllowOutputReference = exports.dataGoogleFolderOrganizationPolicyListPolicyAllowToHclTerraform = exports.dataGoogleFolderOrganizationPolicyListPolicyAllowToTerraform = exports.DataGoogleFolderOrganizationPolicyBooleanPolicyList = exports.DataGoogleFolderOrganizationPolicyBooleanPolicyOutputReference = exports.dataGoogleFolderOrganizationPolicyBooleanPolicyToHclTerraform = exports.dataGoogleFolderOrganizationPolicyBooleanPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleFolderOrganizationPolicyBooleanPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleFolderOrganizationPolicyBooleanPolicyToTerraform = dataGoogleFolderOrganizationPolicyBooleanPolicyToTerraform;
function dataGoogleFolderOrganizationPolicyBooleanPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleFolderOrganizationPolicyBooleanPolicyToHclTerraform = dataGoogleFolderOrganizationPolicyBooleanPolicyToHclTerraform;
class DataGoogleFolderOrganizationPolicyBooleanPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enforced - computed: true, optional: false, required: false
    get enforced() {
        return this.getBooleanAttribute('enforced');
    }
}
exports.DataGoogleFolderOrganizationPolicyBooleanPolicyOutputReference = DataGoogleFolderOrganizationPolicyBooleanPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyBooleanPolicyOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyBooleanPolicyOutputReference", version: "14.13.1" };
class DataGoogleFolderOrganizationPolicyBooleanPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleFolderOrganizationPolicyBooleanPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleFolderOrganizationPolicyBooleanPolicyList = DataGoogleFolderOrganizationPolicyBooleanPolicyList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyBooleanPolicyList[_b] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyBooleanPolicyList", version: "14.13.1" };
function dataGoogleFolderOrganizationPolicyListPolicyAllowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleFolderOrganizationPolicyListPolicyAllowToTerraform = dataGoogleFolderOrganizationPolicyListPolicyAllowToTerraform;
function dataGoogleFolderOrganizationPolicyListPolicyAllowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleFolderOrganizationPolicyListPolicyAllowToHclTerraform = dataGoogleFolderOrganizationPolicyListPolicyAllowToHclTerraform;
class DataGoogleFolderOrganizationPolicyListPolicyAllowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all - computed: true, optional: false, required: false
    get all() {
        return this.getBooleanAttribute('all');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataGoogleFolderOrganizationPolicyListPolicyAllowOutputReference = DataGoogleFolderOrganizationPolicyListPolicyAllowOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyListPolicyAllowOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyListPolicyAllowOutputReference", version: "14.13.1" };
class DataGoogleFolderOrganizationPolicyListPolicyAllowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleFolderOrganizationPolicyListPolicyAllowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleFolderOrganizationPolicyListPolicyAllowList = DataGoogleFolderOrganizationPolicyListPolicyAllowList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyListPolicyAllowList[_d] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyListPolicyAllowList", version: "14.13.1" };
function dataGoogleFolderOrganizationPolicyListPolicyDenyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleFolderOrganizationPolicyListPolicyDenyToTerraform = dataGoogleFolderOrganizationPolicyListPolicyDenyToTerraform;
function dataGoogleFolderOrganizationPolicyListPolicyDenyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleFolderOrganizationPolicyListPolicyDenyToHclTerraform = dataGoogleFolderOrganizationPolicyListPolicyDenyToHclTerraform;
class DataGoogleFolderOrganizationPolicyListPolicyDenyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all - computed: true, optional: false, required: false
    get all() {
        return this.getBooleanAttribute('all');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataGoogleFolderOrganizationPolicyListPolicyDenyOutputReference = DataGoogleFolderOrganizationPolicyListPolicyDenyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyListPolicyDenyOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyListPolicyDenyOutputReference", version: "14.13.1" };
class DataGoogleFolderOrganizationPolicyListPolicyDenyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleFolderOrganizationPolicyListPolicyDenyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleFolderOrganizationPolicyListPolicyDenyList = DataGoogleFolderOrganizationPolicyListPolicyDenyList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyListPolicyDenyList[_f] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyListPolicyDenyList", version: "14.13.1" };
function dataGoogleFolderOrganizationPolicyListPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleFolderOrganizationPolicyListPolicyToTerraform = dataGoogleFolderOrganizationPolicyListPolicyToTerraform;
function dataGoogleFolderOrganizationPolicyListPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleFolderOrganizationPolicyListPolicyToHclTerraform = dataGoogleFolderOrganizationPolicyListPolicyToHclTerraform;
class DataGoogleFolderOrganizationPolicyListPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // allow - computed: true, optional: false, required: false
        this._allow = new DataGoogleFolderOrganizationPolicyListPolicyAllowList(this, "allow", false);
        // deny - computed: true, optional: false, required: false
        this._deny = new DataGoogleFolderOrganizationPolicyListPolicyDenyList(this, "deny", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get allow() {
        return this._allow;
    }
    get deny() {
        return this._deny;
    }
    // inherit_from_parent - computed: true, optional: false, required: false
    get inheritFromParent() {
        return this.getBooleanAttribute('inherit_from_parent');
    }
    // suggested_value - computed: true, optional: false, required: false
    get suggestedValue() {
        return this.getStringAttribute('suggested_value');
    }
}
exports.DataGoogleFolderOrganizationPolicyListPolicyOutputReference = DataGoogleFolderOrganizationPolicyListPolicyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyListPolicyOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyListPolicyOutputReference", version: "14.13.1" };
class DataGoogleFolderOrganizationPolicyListPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleFolderOrganizationPolicyListPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleFolderOrganizationPolicyListPolicyList = DataGoogleFolderOrganizationPolicyListPolicyList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyListPolicyList[_h] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyListPolicyList", version: "14.13.1" };
function dataGoogleFolderOrganizationPolicyRestorePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleFolderOrganizationPolicyRestorePolicyToTerraform = dataGoogleFolderOrganizationPolicyRestorePolicyToTerraform;
function dataGoogleFolderOrganizationPolicyRestorePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleFolderOrganizationPolicyRestorePolicyToHclTerraform = dataGoogleFolderOrganizationPolicyRestorePolicyToHclTerraform;
class DataGoogleFolderOrganizationPolicyRestorePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // default - computed: true, optional: false, required: false
    get default() {
        return this.getBooleanAttribute('default');
    }
}
exports.DataGoogleFolderOrganizationPolicyRestorePolicyOutputReference = DataGoogleFolderOrganizationPolicyRestorePolicyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyRestorePolicyOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyRestorePolicyOutputReference", version: "14.13.1" };
class DataGoogleFolderOrganizationPolicyRestorePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleFolderOrganizationPolicyRestorePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleFolderOrganizationPolicyRestorePolicyList = DataGoogleFolderOrganizationPolicyRestorePolicyList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicyRestorePolicyList[_k] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicyRestorePolicyList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/folder_organization_policy google_folder_organization_policy}
*/
class DataGoogleFolderOrganizationPolicy extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleFolderOrganizationPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleFolderOrganizationPolicy to import
    * @param importFromId The id of the existing DataGoogleFolderOrganizationPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/folder_organization_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleFolderOrganizationPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_folder_organization_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/folder_organization_policy google_folder_organization_policy} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleFolderOrganizationPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_folder_organization_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // boolean_policy - computed: true, optional: false, required: false
        this._booleanPolicy = new DataGoogleFolderOrganizationPolicyBooleanPolicyList(this, "boolean_policy", false);
        // list_policy - computed: true, optional: false, required: false
        this._listPolicy = new DataGoogleFolderOrganizationPolicyListPolicyList(this, "list_policy", false);
        // restore_policy - computed: true, optional: false, required: false
        this._restorePolicy = new DataGoogleFolderOrganizationPolicyRestorePolicyList(this, "restore_policy", false);
        this._constraint = config.constraint;
        this._folder = config.folder;
        this._id = config.id;
    }
    get booleanPolicy() {
        return this._booleanPolicy;
    }
    get constraint() {
        return this.getStringAttribute('constraint');
    }
    set constraint(value) {
        this._constraint = value;
    }
    // Temporarily expose input value. Use with caution.
    get constraintInput() {
        return this._constraint;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get folder() {
        return this.getStringAttribute('folder');
    }
    set folder(value) {
        this._folder = value;
    }
    // Temporarily expose input value. Use with caution.
    get folderInput() {
        return this._folder;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get listPolicy() {
        return this._listPolicy;
    }
    get restorePolicy() {
        return this._restorePolicy;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            constraint: cdktf.stringToTerraform(this._constraint),
            folder: cdktf.stringToTerraform(this._folder),
            id: cdktf.stringToTerraform(this._id),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            constraint: {
                value: cdktf.stringToHclTerraform(this._constraint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            folder: {
                value: cdktf.stringToHclTerraform(this._folder),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleFolderOrganizationPolicy = DataGoogleFolderOrganizationPolicy;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleFolderOrganizationPolicy[_l] = { fqn: "@cdktf/provider-google.dataGoogleFolderOrganizationPolicy.DataGoogleFolderOrganizationPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleFolderOrganizationPolicy.tfResourceType = "google_folder_organization_policy";
