// @cdktf/provider-google/lib/data-google-privateca-certificate-authority/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdList = exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigList = exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameList = exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectList = exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectToTerraform = exports.DataGooglePrivatecaCertificateAuthorityAccessUrlsList = exports.DataGooglePrivatecaCertificateAuthorityAccessUrlsOutputReference = exports.dataGooglePrivatecaCertificateAuthorityAccessUrlsToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityAccessUrlsToTerraform = void 0;
exports.DataGooglePrivatecaCertificateAuthority = exports.DataGooglePrivatecaCertificateAuthoritySubordinateConfigList = exports.DataGooglePrivatecaCertificateAuthoritySubordinateConfigOutputReference = exports.dataGooglePrivatecaCertificateAuthoritySubordinateConfigToHclTerraform = exports.dataGooglePrivatecaCertificateAuthoritySubordinateConfigToTerraform = exports.DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainList = exports.DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainOutputReference = exports.dataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainToHclTerraform = exports.dataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainToTerraform = exports.DataGooglePrivatecaCertificateAuthorityKeySpecList = exports.DataGooglePrivatecaCertificateAuthorityKeySpecOutputReference = exports.dataGooglePrivatecaCertificateAuthorityKeySpecToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityKeySpecToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigAList = exports.DataGooglePrivatecaCertificateAuthorityConfigAOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigAToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigAToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsOutputReference = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsToHclTerraform = exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsToTerraform = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsList = exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGooglePrivatecaCertificateAuthorityAccessUrlsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityAccessUrlsToTerraform = dataGooglePrivatecaCertificateAuthorityAccessUrlsToTerraform;
function dataGooglePrivatecaCertificateAuthorityAccessUrlsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityAccessUrlsToHclTerraform = dataGooglePrivatecaCertificateAuthorityAccessUrlsToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityAccessUrlsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ca_certificate_access_url - computed: true, optional: false, required: false
    get caCertificateAccessUrl() {
        return this.getStringAttribute('ca_certificate_access_url');
    }
    // crl_access_urls - computed: true, optional: false, required: false
    get crlAccessUrls() {
        return this.getListAttribute('crl_access_urls');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityAccessUrlsOutputReference = DataGooglePrivatecaCertificateAuthorityAccessUrlsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityAccessUrlsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityAccessUrlsOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityAccessUrlsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityAccessUrlsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityAccessUrlsList = DataGooglePrivatecaCertificateAuthorityAccessUrlsList;
_b = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityAccessUrlsList[_b] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityAccessUrlsList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectToTerraform = dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // common_name - computed: true, optional: false, required: false
    get commonName() {
        return this.getStringAttribute('common_name');
    }
    // country_code - computed: true, optional: false, required: false
    get countryCode() {
        return this.getStringAttribute('country_code');
    }
    // locality - computed: true, optional: false, required: false
    get locality() {
        return this.getStringAttribute('locality');
    }
    // organization - computed: true, optional: false, required: false
    get organization() {
        return this.getStringAttribute('organization');
    }
    // organizational_unit - computed: true, optional: false, required: false
    get organizationalUnit() {
        return this.getStringAttribute('organizational_unit');
    }
    // postal_code - computed: true, optional: false, required: false
    get postalCode() {
        return this.getStringAttribute('postal_code');
    }
    // province - computed: true, optional: false, required: false
    get province() {
        return this.getStringAttribute('province');
    }
    // street_address - computed: true, optional: false, required: false
    get streetAddress() {
        return this.getStringAttribute('street_address');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectOutputReference = DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectList = DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectList;
_d = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectList[_d] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameToTerraform = dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_names - computed: true, optional: false, required: false
    get dnsNames() {
        return this.getListAttribute('dns_names');
    }
    // email_addresses - computed: true, optional: false, required: false
    get emailAddresses() {
        return this.getListAttribute('email_addresses');
    }
    // ip_addresses - computed: true, optional: false, required: false
    get ipAddresses() {
        return this.getListAttribute('ip_addresses');
    }
    // uris - computed: true, optional: false, required: false
    get uris() {
        return this.getListAttribute('uris');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameOutputReference = DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameList = DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameList;
_f = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameList[_f] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigToTerraform = dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigSubjectConfigToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // subject - computed: true, optional: false, required: false
        this._subject = new DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectList(this, "subject", false);
        // subject_alt_name - computed: true, optional: false, required: false
        this._subjectAltName = new DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigSubjectAltNameList(this, "subject_alt_name", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get subject() {
        return this._subject;
    }
    get subjectAltName() {
        return this._subjectAltName;
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigOutputReference = DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigList = DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigList[_h] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdToTerraform = dataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key_id - computed: true, optional: false, required: false
    get keyId() {
        return this.getStringAttribute('key_id');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdOutputReference = DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdList = DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdList;
_k = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdList[_k] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // object_id_path - computed: true, optional: false, required: false
    get objectIdPath() {
        return this.getNumberListAttribute('object_id_path');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdList;
_m = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdList[_m] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // object_id - computed: true, optional: false, required: false
        this._objectId = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsObjectIdList(this, "object_id", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // critical - computed: true, optional: false, required: false
    get critical() {
        return this.getBooleanAttribute('critical');
    }
    get objectId() {
        return this._objectId;
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsList;
_p = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsList[_p] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_ca - computed: true, optional: false, required: false
    get isCa() {
        return this.getBooleanAttribute('is_ca');
    }
    // max_issuer_path_length - computed: true, optional: false, required: false
    get maxIssuerPathLength() {
        return this.getNumberAttribute('max_issuer_path_length');
    }
    // non_ca - computed: true, optional: false, required: false
    get nonCa() {
        return this.getBooleanAttribute('non_ca');
    }
    // zero_max_issuer_path_length - computed: true, optional: false, required: false
    get zeroMaxIssuerPathLength() {
        return this.getBooleanAttribute('zero_max_issuer_path_length');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsList;
_r = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsList[_r] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cert_sign - computed: true, optional: false, required: false
    get certSign() {
        return this.getBooleanAttribute('cert_sign');
    }
    // content_commitment - computed: true, optional: false, required: false
    get contentCommitment() {
        return this.getBooleanAttribute('content_commitment');
    }
    // crl_sign - computed: true, optional: false, required: false
    get crlSign() {
        return this.getBooleanAttribute('crl_sign');
    }
    // data_encipherment - computed: true, optional: false, required: false
    get dataEncipherment() {
        return this.getBooleanAttribute('data_encipherment');
    }
    // decipher_only - computed: true, optional: false, required: false
    get decipherOnly() {
        return this.getBooleanAttribute('decipher_only');
    }
    // digital_signature - computed: true, optional: false, required: false
    get digitalSignature() {
        return this.getBooleanAttribute('digital_signature');
    }
    // encipher_only - computed: true, optional: false, required: false
    get encipherOnly() {
        return this.getBooleanAttribute('encipher_only');
    }
    // key_agreement - computed: true, optional: false, required: false
    get keyAgreement() {
        return this.getBooleanAttribute('key_agreement');
    }
    // key_encipherment - computed: true, optional: false, required: false
    get keyEncipherment() {
        return this.getBooleanAttribute('key_encipherment');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageList;
_t = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageList[_t] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // client_auth - computed: true, optional: false, required: false
    get clientAuth() {
        return this.getBooleanAttribute('client_auth');
    }
    // code_signing - computed: true, optional: false, required: false
    get codeSigning() {
        return this.getBooleanAttribute('code_signing');
    }
    // email_protection - computed: true, optional: false, required: false
    get emailProtection() {
        return this.getBooleanAttribute('email_protection');
    }
    // ocsp_signing - computed: true, optional: false, required: false
    get ocspSigning() {
        return this.getBooleanAttribute('ocsp_signing');
    }
    // server_auth - computed: true, optional: false, required: false
    get serverAuth() {
        return this.getBooleanAttribute('server_auth');
    }
    // time_stamping - computed: true, optional: false, required: false
    get timeStamping() {
        return this.getBooleanAttribute('time_stamping');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageList;
_v = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageList[_v] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // object_id_path - computed: true, optional: false, required: false
    get objectIdPath() {
        return this.getNumberListAttribute('object_id_path');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesList;
_x = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesList[_x] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // base_key_usage - computed: true, optional: false, required: false
        this._baseKeyUsage = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageBaseKeyUsageList(this, "base_key_usage", false);
        // extended_key_usage - computed: true, optional: false, required: false
        this._extendedKeyUsage = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageExtendedKeyUsageList(this, "extended_key_usage", false);
        // unknown_extended_key_usages - computed: true, optional: false, required: false
        this._unknownExtendedKeyUsages = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageUnknownExtendedKeyUsagesList(this, "unknown_extended_key_usages", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get baseKeyUsage() {
        return this._baseKeyUsage;
    }
    get extendedKeyUsage() {
        return this._extendedKeyUsage;
    }
    get unknownExtendedKeyUsages() {
        return this._unknownExtendedKeyUsages;
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageList;
_z = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageList[_z] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // critical - computed: true, optional: false, required: false
    get critical() {
        return this.getBooleanAttribute('critical');
    }
    // excluded_dns_names - computed: true, optional: false, required: false
    get excludedDnsNames() {
        return this.getListAttribute('excluded_dns_names');
    }
    // excluded_email_addresses - computed: true, optional: false, required: false
    get excludedEmailAddresses() {
        return this.getListAttribute('excluded_email_addresses');
    }
    // excluded_ip_ranges - computed: true, optional: false, required: false
    get excludedIpRanges() {
        return this.getListAttribute('excluded_ip_ranges');
    }
    // excluded_uris - computed: true, optional: false, required: false
    get excludedUris() {
        return this.getListAttribute('excluded_uris');
    }
    // permitted_dns_names - computed: true, optional: false, required: false
    get permittedDnsNames() {
        return this.getListAttribute('permitted_dns_names');
    }
    // permitted_email_addresses - computed: true, optional: false, required: false
    get permittedEmailAddresses() {
        return this.getListAttribute('permitted_email_addresses');
    }
    // permitted_ip_ranges - computed: true, optional: false, required: false
    get permittedIpRanges() {
        return this.getListAttribute('permitted_ip_ranges');
    }
    // permitted_uris - computed: true, optional: false, required: false
    get permittedUris() {
        return this.getListAttribute('permitted_uris');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsList;
_1 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsList[_1] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // object_id_path - computed: true, optional: false, required: false
    get objectIdPath() {
        return this.getNumberListAttribute('object_id_path');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsList;
_3 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsList[_3] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigToTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigX509ConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigX509ConfigToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigX509ConfigToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // additional_extensions - computed: true, optional: false, required: false
        this._additionalExtensions = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigAdditionalExtensionsList(this, "additional_extensions", false);
        // ca_options - computed: true, optional: false, required: false
        this._caOptions = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigCaOptionsList(this, "ca_options", false);
        // key_usage - computed: true, optional: false, required: false
        this._keyUsage = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigKeyUsageList(this, "key_usage", false);
        // name_constraints - computed: true, optional: false, required: false
        this._nameConstraints = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigNameConstraintsList(this, "name_constraints", false);
        // policy_ids - computed: true, optional: false, required: false
        this._policyIds = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigPolicyIdsList(this, "policy_ids", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get additionalExtensions() {
        return this._additionalExtensions;
    }
    // aia_ocsp_servers - computed: true, optional: false, required: false
    get aiaOcspServers() {
        return this.getListAttribute('aia_ocsp_servers');
    }
    get caOptions() {
        return this._caOptions;
    }
    get keyUsage() {
        return this._keyUsage;
    }
    get nameConstraints() {
        return this._nameConstraints;
    }
    get policyIds() {
        return this._policyIds;
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigOutputReference = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigX509ConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigList = DataGooglePrivatecaCertificateAuthorityConfigX509ConfigList;
_5 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigX509ConfigList[_5] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigX509ConfigList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityConfigAToTerraform = dataGooglePrivatecaCertificateAuthorityConfigAToTerraform;
function dataGooglePrivatecaCertificateAuthorityConfigAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityConfigAToHclTerraform = dataGooglePrivatecaCertificateAuthorityConfigAToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // subject_config - computed: true, optional: false, required: false
        this._subjectConfig = new DataGooglePrivatecaCertificateAuthorityConfigSubjectConfigList(this, "subject_config", false);
        // subject_key_id - computed: true, optional: false, required: false
        this._subjectKeyId = new DataGooglePrivatecaCertificateAuthorityConfigSubjectKeyIdList(this, "subject_key_id", false);
        // x509_config - computed: true, optional: false, required: false
        this._x509Config = new DataGooglePrivatecaCertificateAuthorityConfigX509ConfigList(this, "x509_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get subjectConfig() {
        return this._subjectConfig;
    }
    get subjectKeyId() {
        return this._subjectKeyId;
    }
    get x509Config() {
        return this._x509Config;
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigAOutputReference = DataGooglePrivatecaCertificateAuthorityConfigAOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigAOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigAOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityConfigAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityConfigAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityConfigAList = DataGooglePrivatecaCertificateAuthorityConfigAList;
_7 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityConfigAList[_7] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityConfigAList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthorityKeySpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthorityKeySpecToTerraform = dataGooglePrivatecaCertificateAuthorityKeySpecToTerraform;
function dataGooglePrivatecaCertificateAuthorityKeySpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthorityKeySpecToHclTerraform = dataGooglePrivatecaCertificateAuthorityKeySpecToHclTerraform;
class DataGooglePrivatecaCertificateAuthorityKeySpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // algorithm - computed: true, optional: false, required: false
    get algorithm() {
        return this.getStringAttribute('algorithm');
    }
    // cloud_kms_key_version - computed: true, optional: false, required: false
    get cloudKmsKeyVersion() {
        return this.getStringAttribute('cloud_kms_key_version');
    }
}
exports.DataGooglePrivatecaCertificateAuthorityKeySpecOutputReference = DataGooglePrivatecaCertificateAuthorityKeySpecOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityKeySpecOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityKeySpecOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthorityKeySpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthorityKeySpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthorityKeySpecList = DataGooglePrivatecaCertificateAuthorityKeySpecList;
_9 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthorityKeySpecList[_9] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthorityKeySpecList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainToTerraform = dataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainToTerraform;
function dataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainToHclTerraform = dataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainToHclTerraform;
class DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // pem_certificates - computed: true, optional: false, required: false
    get pemCertificates() {
        return this.getListAttribute('pem_certificates');
    }
}
exports.DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainOutputReference = DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainOutputReference[_10] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainList = DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainList;
_11 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainList[_11] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainList", version: "14.13.1" };
function dataGooglePrivatecaCertificateAuthoritySubordinateConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivatecaCertificateAuthoritySubordinateConfigToTerraform = dataGooglePrivatecaCertificateAuthoritySubordinateConfigToTerraform;
function dataGooglePrivatecaCertificateAuthoritySubordinateConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivatecaCertificateAuthoritySubordinateConfigToHclTerraform = dataGooglePrivatecaCertificateAuthoritySubordinateConfigToHclTerraform;
class DataGooglePrivatecaCertificateAuthoritySubordinateConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // pem_issuer_chain - computed: true, optional: false, required: false
        this._pemIssuerChain = new DataGooglePrivatecaCertificateAuthoritySubordinateConfigPemIssuerChainList(this, "pem_issuer_chain", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // certificate_authority - computed: true, optional: false, required: false
    get certificateAuthority() {
        return this.getStringAttribute('certificate_authority');
    }
    get pemIssuerChain() {
        return this._pemIssuerChain;
    }
}
exports.DataGooglePrivatecaCertificateAuthoritySubordinateConfigOutputReference = DataGooglePrivatecaCertificateAuthoritySubordinateConfigOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthoritySubordinateConfigOutputReference[_12] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthoritySubordinateConfigOutputReference", version: "14.13.1" };
class DataGooglePrivatecaCertificateAuthoritySubordinateConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivatecaCertificateAuthoritySubordinateConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivatecaCertificateAuthoritySubordinateConfigList = DataGooglePrivatecaCertificateAuthoritySubordinateConfigList;
_13 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthoritySubordinateConfigList[_13] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthoritySubordinateConfigList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/privateca_certificate_authority google_privateca_certificate_authority}
*/
class DataGooglePrivatecaCertificateAuthority extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGooglePrivatecaCertificateAuthority resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGooglePrivatecaCertificateAuthority to import
    * @param importFromId The id of the existing DataGooglePrivatecaCertificateAuthority that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/privateca_certificate_authority#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGooglePrivatecaCertificateAuthority to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_privateca_certificate_authority", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/privateca_certificate_authority google_privateca_certificate_authority} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGooglePrivatecaCertificateAuthorityConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_privateca_certificate_authority',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // access_urls - computed: true, optional: false, required: false
        this._accessUrls = new DataGooglePrivatecaCertificateAuthorityAccessUrlsList(this, "access_urls", false);
        // config - computed: true, optional: false, required: false
        this._config = new DataGooglePrivatecaCertificateAuthorityConfigAList(this, "config", false);
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // key_spec - computed: true, optional: false, required: false
        this._keySpec = new DataGooglePrivatecaCertificateAuthorityKeySpecList(this, "key_spec", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // subordinate_config - computed: true, optional: false, required: false
        this._subordinateConfig = new DataGooglePrivatecaCertificateAuthoritySubordinateConfigList(this, "subordinate_config", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._certificateAuthorityId = config.certificateAuthorityId;
        this._id = config.id;
        this._location = config.location;
        this._pool = config.pool;
        this._project = config.project;
    }
    get accessUrls() {
        return this._accessUrls;
    }
    get certificateAuthorityId() {
        return this.getStringAttribute('certificate_authority_id');
    }
    set certificateAuthorityId(value) {
        this._certificateAuthorityId = value;
    }
    resetCertificateAuthorityId() {
        this._certificateAuthorityId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get certificateAuthorityIdInput() {
        return this._certificateAuthorityId;
    }
    get config() {
        return this._config;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    // desired_state - computed: true, optional: false, required: false
    get desiredState() {
        return this.getStringAttribute('desired_state');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // gcs_bucket - computed: true, optional: false, required: false
    get gcsBucket() {
        return this.getStringAttribute('gcs_bucket');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // ignore_active_certificates_on_deletion - computed: true, optional: false, required: false
    get ignoreActiveCertificatesOnDeletion() {
        return this.getBooleanAttribute('ignore_active_certificates_on_deletion');
    }
    get keySpec() {
        return this._keySpec;
    }
    get labels() {
        return this._labels;
    }
    // lifetime - computed: true, optional: false, required: false
    get lifetime() {
        return this.getStringAttribute('lifetime');
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // pem_ca_certificate - computed: true, optional: false, required: false
    get pemCaCertificate() {
        return this.getStringAttribute('pem_ca_certificate');
    }
    // pem_ca_certificates - computed: true, optional: false, required: false
    get pemCaCertificates() {
        return this.getListAttribute('pem_ca_certificates');
    }
    // pem_csr - computed: true, optional: false, required: false
    get pemCsr() {
        return this.getStringAttribute('pem_csr');
    }
    get pool() {
        return this.getStringAttribute('pool');
    }
    set pool(value) {
        this._pool = value;
    }
    resetPool() {
        this._pool = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get poolInput() {
        return this._pool;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // skip_grace_period - computed: true, optional: false, required: false
    get skipGracePeriod() {
        return this.getBooleanAttribute('skip_grace_period');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get subordinateConfig() {
        return this._subordinateConfig;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            certificate_authority_id: cdktf.stringToTerraform(this._certificateAuthorityId),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            pool: cdktf.stringToTerraform(this._pool),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            certificate_authority_id: {
                value: cdktf.stringToHclTerraform(this._certificateAuthorityId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pool: {
                value: cdktf.stringToHclTerraform(this._pool),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGooglePrivatecaCertificateAuthority = DataGooglePrivatecaCertificateAuthority;
_14 = JSII_RTTI_SYMBOL_1;
DataGooglePrivatecaCertificateAuthority[_14] = { fqn: "@cdktf/provider-google.dataGooglePrivatecaCertificateAuthority.DataGooglePrivatecaCertificateAuthority", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGooglePrivatecaCertificateAuthority.tfResourceType = "google_privateca_certificate_authority";
