// @cdktf/provider-google/lib/compute-disk-async-replication/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputeDiskAsyncReplication = exports.ComputeDiskAsyncReplicationTimeoutsOutputReference = exports.computeDiskAsyncReplicationTimeoutsToHclTerraform = exports.computeDiskAsyncReplicationTimeoutsToTerraform = exports.ComputeDiskAsyncReplicationSecondaryDiskOutputReference = exports.computeDiskAsyncReplicationSecondaryDiskToHclTerraform = exports.computeDiskAsyncReplicationSecondaryDiskToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeDiskAsyncReplicationSecondaryDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk: cdktf.stringToTerraform(struct.disk),
    };
}
exports.computeDiskAsyncReplicationSecondaryDiskToTerraform = computeDiskAsyncReplicationSecondaryDiskToTerraform;
function computeDiskAsyncReplicationSecondaryDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        disk: {
            value: cdktf.stringToHclTerraform(struct.disk),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeDiskAsyncReplicationSecondaryDiskToHclTerraform = computeDiskAsyncReplicationSecondaryDiskToHclTerraform;
class ComputeDiskAsyncReplicationSecondaryDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._disk !== undefined) {
            hasAnyValues = true;
            internalValueResult.disk = this._disk;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._disk = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._disk = value.disk;
        }
    }
    get disk() {
        return this.getStringAttribute('disk');
    }
    set disk(value) {
        this._disk = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskInput() {
        return this._disk;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.ComputeDiskAsyncReplicationSecondaryDiskOutputReference = ComputeDiskAsyncReplicationSecondaryDiskOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeDiskAsyncReplicationSecondaryDiskOutputReference[_a] = { fqn: "@cdktf/provider-google.computeDiskAsyncReplication.ComputeDiskAsyncReplicationSecondaryDiskOutputReference", version: "14.13.1" };
function computeDiskAsyncReplicationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.computeDiskAsyncReplicationTimeoutsToTerraform = computeDiskAsyncReplicationTimeoutsToTerraform;
function computeDiskAsyncReplicationTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeDiskAsyncReplicationTimeoutsToHclTerraform = computeDiskAsyncReplicationTimeoutsToHclTerraform;
class ComputeDiskAsyncReplicationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.ComputeDiskAsyncReplicationTimeoutsOutputReference = ComputeDiskAsyncReplicationTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ComputeDiskAsyncReplicationTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-google.computeDiskAsyncReplication.ComputeDiskAsyncReplicationTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_disk_async_replication google_compute_disk_async_replication}
*/
class ComputeDiskAsyncReplication extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ComputeDiskAsyncReplication resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ComputeDiskAsyncReplication to import
    * @param importFromId The id of the existing ComputeDiskAsyncReplication that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_disk_async_replication#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ComputeDiskAsyncReplication to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_disk_async_replication", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_disk_async_replication google_compute_disk_async_replication} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeDiskAsyncReplicationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_disk_async_replication',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // secondary_disk - computed: false, optional: false, required: true
        this._secondaryDisk = new ComputeDiskAsyncReplicationSecondaryDiskOutputReference(this, "secondary_disk");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeDiskAsyncReplicationTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._primaryDisk = config.primaryDisk;
        this._secondaryDisk.internalValue = config.secondaryDisk;
        this._timeouts.internalValue = config.timeouts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get primaryDisk() {
        return this.getStringAttribute('primary_disk');
    }
    set primaryDisk(value) {
        this._primaryDisk = value;
    }
    // Temporarily expose input value. Use with caution.
    get primaryDiskInput() {
        return this._primaryDisk;
    }
    get secondaryDisk() {
        return this._secondaryDisk;
    }
    putSecondaryDisk(value) {
        this._secondaryDisk.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get secondaryDiskInput() {
        return this._secondaryDisk.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            primary_disk: cdktf.stringToTerraform(this._primaryDisk),
            secondary_disk: computeDiskAsyncReplicationSecondaryDiskToTerraform(this._secondaryDisk.internalValue),
            timeouts: computeDiskAsyncReplicationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            primary_disk: {
                value: cdktf.stringToHclTerraform(this._primaryDisk),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            secondary_disk: {
                value: computeDiskAsyncReplicationSecondaryDiskToHclTerraform(this._secondaryDisk.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeDiskAsyncReplicationSecondaryDiskList",
            },
            timeouts: {
                value: computeDiskAsyncReplicationTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ComputeDiskAsyncReplicationTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ComputeDiskAsyncReplication = ComputeDiskAsyncReplication;
_c = JSII_RTTI_SYMBOL_1;
ComputeDiskAsyncReplication[_c] = { fqn: "@cdktf/provider-google.computeDiskAsyncReplication.ComputeDiskAsyncReplication", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ComputeDiskAsyncReplication.tfResourceType = "google_compute_disk_async_replication";
