// @cdktf/provider-google/lib/data-google-privileged-access-manager-entitlement/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGooglePrivilegedAccessManagerEntitlement = exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigList = exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredList = exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryList = exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessList = exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessList = exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsList = exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementEligibleUsersList = exports.DataGooglePrivilegedAccessManagerEntitlementEligibleUsersOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementEligibleUsersToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementEligibleUsersToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowList = exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsList = exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsList = exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversList = exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversToTerraform = exports.DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsList = exports.DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsOutputReference = exports.dataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsToHclTerraform = exports.dataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsToTerraform = dataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // admin_email_recipients - computed: true, optional: false, required: false
    get adminEmailRecipients() {
        return cdktf.Fn.tolist(this.getListAttribute('admin_email_recipients'));
    }
    // requester_email_recipients - computed: true, optional: false, required: false
    get requesterEmailRecipients() {
        return cdktf.Fn.tolist(this.getListAttribute('requester_email_recipients'));
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsOutputReference = DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsList = DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsList;
_b = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsList[_b] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversToTerraform = dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // principals - computed: true, optional: false, required: false
    get principals() {
        return cdktf.Fn.tolist(this.getListAttribute('principals'));
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversOutputReference = DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversList = DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversList;
_d = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversList[_d] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsToTerraform = dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // approvers - computed: true, optional: false, required: false
        this._approvers = new DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsApproversList(this, "approvers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // approvals_needed - computed: true, optional: false, required: false
    get approvalsNeeded() {
        return this.getNumberAttribute('approvals_needed');
    }
    // approver_email_recipients - computed: true, optional: false, required: false
    get approverEmailRecipients() {
        return cdktf.Fn.tolist(this.getListAttribute('approver_email_recipients'));
    }
    get approvers() {
        return this._approvers;
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsOutputReference = DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsList = DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsList;
_f = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsList[_f] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsToTerraform = dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // steps - computed: true, optional: false, required: false
        this._steps = new DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsStepsList(this, "steps", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // require_approver_justification - computed: true, optional: false, required: false
    get requireApproverJustification() {
        return this.getBooleanAttribute('require_approver_justification');
    }
    get steps() {
        return this._steps;
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsOutputReference = DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsList = DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsList;
_h = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsList[_h] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowToTerraform = dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // manual_approvals - computed: true, optional: false, required: false
        this._manualApprovals = new DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowManualApprovalsList(this, "manual_approvals", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get manualApprovals() {
        return this._manualApprovals;
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowOutputReference = DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowList = DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowList;
_k = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowList[_k] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementEligibleUsersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementEligibleUsersToTerraform = dataGooglePrivilegedAccessManagerEntitlementEligibleUsersToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementEligibleUsersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementEligibleUsersToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementEligibleUsersToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementEligibleUsersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // principals - computed: true, optional: false, required: false
    get principals() {
        return cdktf.Fn.tolist(this.getListAttribute('principals'));
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementEligibleUsersOutputReference = DataGooglePrivilegedAccessManagerEntitlementEligibleUsersOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementEligibleUsersOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementEligibleUsersOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementEligibleUsersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementEligibleUsersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementEligibleUsersList = DataGooglePrivilegedAccessManagerEntitlementEligibleUsersList;
_m = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementEligibleUsersList[_m] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementEligibleUsersList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsToTerraform = dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // condition_expression - computed: true, optional: false, required: false
    get conditionExpression() {
        return this.getStringAttribute('condition_expression');
    }
    // role - computed: true, optional: false, required: false
    get role() {
        return this.getStringAttribute('role');
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsOutputReference = DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsList = DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsList;
_p = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsList[_p] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessToTerraform = dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // role_bindings - computed: true, optional: false, required: false
        this._roleBindings = new DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessRoleBindingsList(this, "role_bindings", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // resource - computed: true, optional: false, required: false
    get resource() {
        return this.getStringAttribute('resource');
    }
    // resource_type - computed: true, optional: false, required: false
    get resourceType() {
        return this.getStringAttribute('resource_type');
    }
    get roleBindings() {
        return this._roleBindings;
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessOutputReference = DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessList = DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessList;
_r = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessList[_r] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessToTerraform = dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // gcp_iam_access - computed: true, optional: false, required: false
        this._gcpIamAccess = new DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessGcpIamAccessList(this, "gcp_iam_access", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get gcpIamAccess() {
        return this._gcpIamAccess;
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessOutputReference = DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessList = DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessList;
_t = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessList[_t] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryToTerraform = dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryOutputReference = DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryList = DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryList;
_v = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryList[_v] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredToTerraform = dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredOutputReference = DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredList = DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredList;
_x = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredList[_x] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredList", version: "14.13.1" };
function dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigToTerraform = dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigToTerraform;
function dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigToHclTerraform = dataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigToHclTerraform;
class DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // not_mandatory - computed: true, optional: false, required: false
        this._notMandatory = new DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigNotMandatoryList(this, "not_mandatory", false);
        // unstructured - computed: true, optional: false, required: false
        this._unstructured = new DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigUnstructuredList(this, "unstructured", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get notMandatory() {
        return this._notMandatory;
    }
    get unstructured() {
        return this._unstructured;
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigOutputReference = DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigOutputReference", version: "14.13.1" };
class DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigList = DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigList;
_z = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigList[_z] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/privileged_access_manager_entitlement google_privileged_access_manager_entitlement}
*/
class DataGooglePrivilegedAccessManagerEntitlement extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGooglePrivilegedAccessManagerEntitlement resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGooglePrivilegedAccessManagerEntitlement to import
    * @param importFromId The id of the existing DataGooglePrivilegedAccessManagerEntitlement that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/privileged_access_manager_entitlement#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGooglePrivilegedAccessManagerEntitlement to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_privileged_access_manager_entitlement", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/privileged_access_manager_entitlement google_privileged_access_manager_entitlement} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGooglePrivilegedAccessManagerEntitlementConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_privileged_access_manager_entitlement',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // additional_notification_targets - computed: true, optional: false, required: false
        this._additionalNotificationTargets = new DataGooglePrivilegedAccessManagerEntitlementAdditionalNotificationTargetsList(this, "additional_notification_targets", false);
        // approval_workflow - computed: true, optional: false, required: false
        this._approvalWorkflow = new DataGooglePrivilegedAccessManagerEntitlementApprovalWorkflowList(this, "approval_workflow", false);
        // eligible_users - computed: true, optional: false, required: false
        this._eligibleUsers = new DataGooglePrivilegedAccessManagerEntitlementEligibleUsersList(this, "eligible_users", false);
        // privileged_access - computed: true, optional: false, required: false
        this._privilegedAccess = new DataGooglePrivilegedAccessManagerEntitlementPrivilegedAccessList(this, "privileged_access", false);
        // requester_justification_config - computed: true, optional: false, required: false
        this._requesterJustificationConfig = new DataGooglePrivilegedAccessManagerEntitlementRequesterJustificationConfigList(this, "requester_justification_config", false);
        this._entitlementId = config.entitlementId;
        this._id = config.id;
        this._location = config.location;
        this._parent = config.parent;
    }
    get additionalNotificationTargets() {
        return this._additionalNotificationTargets;
    }
    get approvalWorkflow() {
        return this._approvalWorkflow;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get eligibleUsers() {
        return this._eligibleUsers;
    }
    get entitlementId() {
        return this.getStringAttribute('entitlement_id');
    }
    set entitlementId(value) {
        this._entitlementId = value;
    }
    resetEntitlementId() {
        this._entitlementId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get entitlementIdInput() {
        return this._entitlementId;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // max_request_duration - computed: true, optional: false, required: false
    get maxRequestDuration() {
        return this.getStringAttribute('max_request_duration');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    resetParent() {
        this._parent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    get privilegedAccess() {
        return this._privilegedAccess;
    }
    get requesterJustificationConfig() {
        return this._requesterJustificationConfig;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            entitlement_id: cdktf.stringToTerraform(this._entitlementId),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            parent: cdktf.stringToTerraform(this._parent),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            entitlement_id: {
                value: cdktf.stringToHclTerraform(this._entitlementId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent: {
                value: cdktf.stringToHclTerraform(this._parent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGooglePrivilegedAccessManagerEntitlement = DataGooglePrivilegedAccessManagerEntitlement;
_0 = JSII_RTTI_SYMBOL_1;
DataGooglePrivilegedAccessManagerEntitlement[_0] = { fqn: "@cdktf/provider-google.dataGooglePrivilegedAccessManagerEntitlement.DataGooglePrivilegedAccessManagerEntitlement", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGooglePrivilegedAccessManagerEntitlement.tfResourceType = "google_privileged_access_manager_entitlement";
