// @cdktf/provider-google/lib/backup-dr-backup-vault/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupDrBackupVault = exports.BackupDrBackupVaultTimeoutsOutputReference = exports.backupDrBackupVaultTimeoutsToHclTerraform = exports.backupDrBackupVaultTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function backupDrBackupVaultTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.backupDrBackupVaultTimeoutsToTerraform = backupDrBackupVaultTimeoutsToTerraform;
function backupDrBackupVaultTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.backupDrBackupVaultTimeoutsToHclTerraform = backupDrBackupVaultTimeoutsToHclTerraform;
class BackupDrBackupVaultTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.BackupDrBackupVaultTimeoutsOutputReference = BackupDrBackupVaultTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BackupDrBackupVaultTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-google.backupDrBackupVault.BackupDrBackupVaultTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/backup_dr_backup_vault google_backup_dr_backup_vault}
*/
class BackupDrBackupVault extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a BackupDrBackupVault resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the BackupDrBackupVault to import
    * @param importFromId The id of the existing BackupDrBackupVault that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/backup_dr_backup_vault#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the BackupDrBackupVault to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_backup_dr_backup_vault", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/backup_dr_backup_vault google_backup_dr_backup_vault} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BackupDrBackupVaultConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_backup_dr_backup_vault',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_annotations - computed: true, optional: false, required: false
        this._effectiveAnnotations = new cdktf.StringMap(this, "effective_annotations");
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new BackupDrBackupVaultTimeoutsOutputReference(this, "timeouts");
        this._accessRestriction = config.accessRestriction;
        this._allowMissing = config.allowMissing;
        this._annotations = config.annotations;
        this._backupMinimumEnforcedRetentionDuration = config.backupMinimumEnforcedRetentionDuration;
        this._backupVaultId = config.backupVaultId;
        this._description = config.description;
        this._effectiveTime = config.effectiveTime;
        this._forceDelete = config.forceDelete;
        this._forceUpdate = config.forceUpdate;
        this._id = config.id;
        this._ignoreBackupPlanReferences = config.ignoreBackupPlanReferences;
        this._ignoreInactiveDatasources = config.ignoreInactiveDatasources;
        this._labels = config.labels;
        this._location = config.location;
        this._project = config.project;
        this._timeouts.internalValue = config.timeouts;
    }
    get accessRestriction() {
        return this.getStringAttribute('access_restriction');
    }
    set accessRestriction(value) {
        this._accessRestriction = value;
    }
    resetAccessRestriction() {
        this._accessRestriction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessRestrictionInput() {
        return this._accessRestriction;
    }
    get allowMissing() {
        return this.getBooleanAttribute('allow_missing');
    }
    set allowMissing(value) {
        this._allowMissing = value;
    }
    resetAllowMissing() {
        this._allowMissing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowMissingInput() {
        return this._allowMissing;
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    // backup_count - computed: true, optional: false, required: false
    get backupCount() {
        return this.getStringAttribute('backup_count');
    }
    get backupMinimumEnforcedRetentionDuration() {
        return this.getStringAttribute('backup_minimum_enforced_retention_duration');
    }
    set backupMinimumEnforcedRetentionDuration(value) {
        this._backupMinimumEnforcedRetentionDuration = value;
    }
    // Temporarily expose input value. Use with caution.
    get backupMinimumEnforcedRetentionDurationInput() {
        return this._backupMinimumEnforcedRetentionDuration;
    }
    get backupVaultId() {
        return this.getStringAttribute('backup_vault_id');
    }
    set backupVaultId(value) {
        this._backupVaultId = value;
    }
    // Temporarily expose input value. Use with caution.
    get backupVaultIdInput() {
        return this._backupVaultId;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // deletable - computed: true, optional: false, required: false
    get deletable() {
        return this.getBooleanAttribute('deletable');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveAnnotations() {
        return this._effectiveAnnotations;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get effectiveTime() {
        return this.getStringAttribute('effective_time');
    }
    set effectiveTime(value) {
        this._effectiveTime = value;
    }
    resetEffectiveTime() {
        this._effectiveTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get effectiveTimeInput() {
        return this._effectiveTime;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get forceDelete() {
        return this.getBooleanAttribute('force_delete');
    }
    set forceDelete(value) {
        this._forceDelete = value;
    }
    resetForceDelete() {
        this._forceDelete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forceDeleteInput() {
        return this._forceDelete;
    }
    get forceUpdate() {
        return this.getBooleanAttribute('force_update');
    }
    set forceUpdate(value) {
        this._forceUpdate = value;
    }
    resetForceUpdate() {
        this._forceUpdate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forceUpdateInput() {
        return this._forceUpdate;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ignoreBackupPlanReferences() {
        return this.getBooleanAttribute('ignore_backup_plan_references');
    }
    set ignoreBackupPlanReferences(value) {
        this._ignoreBackupPlanReferences = value;
    }
    resetIgnoreBackupPlanReferences() {
        this._ignoreBackupPlanReferences = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreBackupPlanReferencesInput() {
        return this._ignoreBackupPlanReferences;
    }
    get ignoreInactiveDatasources() {
        return this.getBooleanAttribute('ignore_inactive_datasources');
    }
    set ignoreInactiveDatasources(value) {
        this._ignoreInactiveDatasources = value;
    }
    resetIgnoreInactiveDatasources() {
        this._ignoreInactiveDatasources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreInactiveDatasourcesInput() {
        return this._ignoreInactiveDatasources;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // service_account - computed: true, optional: false, required: false
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // total_stored_bytes - computed: true, optional: false, required: false
    get totalStoredBytes() {
        return this.getStringAttribute('total_stored_bytes');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            access_restriction: cdktf.stringToTerraform(this._accessRestriction),
            allow_missing: cdktf.booleanToTerraform(this._allowMissing),
            annotations: cdktf.hashMapper(cdktf.stringToTerraform)(this._annotations),
            backup_minimum_enforced_retention_duration: cdktf.stringToTerraform(this._backupMinimumEnforcedRetentionDuration),
            backup_vault_id: cdktf.stringToTerraform(this._backupVaultId),
            description: cdktf.stringToTerraform(this._description),
            effective_time: cdktf.stringToTerraform(this._effectiveTime),
            force_delete: cdktf.booleanToTerraform(this._forceDelete),
            force_update: cdktf.booleanToTerraform(this._forceUpdate),
            id: cdktf.stringToTerraform(this._id),
            ignore_backup_plan_references: cdktf.booleanToTerraform(this._ignoreBackupPlanReferences),
            ignore_inactive_datasources: cdktf.booleanToTerraform(this._ignoreInactiveDatasources),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            timeouts: backupDrBackupVaultTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            access_restriction: {
                value: cdktf.stringToHclTerraform(this._accessRestriction),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            allow_missing: {
                value: cdktf.booleanToHclTerraform(this._allowMissing),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            annotations: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._annotations),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            backup_minimum_enforced_retention_duration: {
                value: cdktf.stringToHclTerraform(this._backupMinimumEnforcedRetentionDuration),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            backup_vault_id: {
                value: cdktf.stringToHclTerraform(this._backupVaultId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            effective_time: {
                value: cdktf.stringToHclTerraform(this._effectiveTime),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            force_delete: {
                value: cdktf.booleanToHclTerraform(this._forceDelete),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            force_update: {
                value: cdktf.booleanToHclTerraform(this._forceUpdate),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ignore_backup_plan_references: {
                value: cdktf.booleanToHclTerraform(this._ignoreBackupPlanReferences),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            ignore_inactive_datasources: {
                value: cdktf.booleanToHclTerraform(this._ignoreInactiveDatasources),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: backupDrBackupVaultTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "BackupDrBackupVaultTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.BackupDrBackupVault = BackupDrBackupVault;
_b = JSII_RTTI_SYMBOL_1;
BackupDrBackupVault[_b] = { fqn: "@cdktf/provider-google.backupDrBackupVault.BackupDrBackupVault", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
BackupDrBackupVault.tfResourceType = "google_backup_dr_backup_vault";
