// @cdktf/provider-google/lib/data-google-cloud-quotas-quota-info/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleCloudQuotasQuotaInfo = exports.DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityList = exports.DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityOutputReference = exports.dataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityToHclTerraform = exports.dataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityToTerraform = exports.DataGoogleCloudQuotasQuotaInfoDimensionsInfosList = exports.DataGoogleCloudQuotasQuotaInfoDimensionsInfosOutputReference = exports.dataGoogleCloudQuotasQuotaInfoDimensionsInfosToHclTerraform = exports.dataGoogleCloudQuotasQuotaInfoDimensionsInfosToTerraform = exports.DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsList = exports.DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsOutputReference = exports.dataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsToHclTerraform = exports.dataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsToTerraform = dataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsToTerraform;
function dataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsToHclTerraform = dataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsToHclTerraform;
class DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsOutputReference = DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleCloudQuotasQuotaInfo.DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsOutputReference", version: "14.13.1" };
class DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsList = DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleCloudQuotasQuotaInfo.DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsList", version: "14.13.1" };
function dataGoogleCloudQuotasQuotaInfoDimensionsInfosToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudQuotasQuotaInfoDimensionsInfosToTerraform = dataGoogleCloudQuotasQuotaInfoDimensionsInfosToTerraform;
function dataGoogleCloudQuotasQuotaInfoDimensionsInfosToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudQuotasQuotaInfoDimensionsInfosToHclTerraform = dataGoogleCloudQuotasQuotaInfoDimensionsInfosToHclTerraform;
class DataGoogleCloudQuotasQuotaInfoDimensionsInfosOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // details - computed: true, optional: false, required: false
        this._details = new DataGoogleCloudQuotasQuotaInfoDimensionsInfosDetailsList(this, "details", false);
        // dimensions - computed: true, optional: false, required: false
        this._dimensions = new cdktf.StringMap(this, "dimensions");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // applicable_locations - computed: true, optional: false, required: false
    get applicableLocations() {
        return this.getListAttribute('applicable_locations');
    }
    get details() {
        return this._details;
    }
    get dimensions() {
        return this._dimensions;
    }
}
exports.DataGoogleCloudQuotasQuotaInfoDimensionsInfosOutputReference = DataGoogleCloudQuotasQuotaInfoDimensionsInfosOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleCloudQuotasQuotaInfoDimensionsInfosOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleCloudQuotasQuotaInfo.DataGoogleCloudQuotasQuotaInfoDimensionsInfosOutputReference", version: "14.13.1" };
class DataGoogleCloudQuotasQuotaInfoDimensionsInfosList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudQuotasQuotaInfoDimensionsInfosOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudQuotasQuotaInfoDimensionsInfosList = DataGoogleCloudQuotasQuotaInfoDimensionsInfosList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleCloudQuotasQuotaInfoDimensionsInfosList[_d] = { fqn: "@cdktf/provider-google.dataGoogleCloudQuotasQuotaInfo.DataGoogleCloudQuotasQuotaInfoDimensionsInfosList", version: "14.13.1" };
function dataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityToTerraform = dataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityToTerraform;
function dataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityToHclTerraform = dataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityToHclTerraform;
class DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ineligibility_reason - computed: true, optional: false, required: false
    get ineligibilityReason() {
        return this.getStringAttribute('ineligibility_reason');
    }
    // is_eligible - computed: true, optional: false, required: false
    get isEligible() {
        return this.getBooleanAttribute('is_eligible');
    }
}
exports.DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityOutputReference = DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleCloudQuotasQuotaInfo.DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityOutputReference", version: "14.13.1" };
class DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityList = DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityList[_f] = { fqn: "@cdktf/provider-google.dataGoogleCloudQuotasQuotaInfo.DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_quotas_quota_info google_cloud_quotas_quota_info}
*/
class DataGoogleCloudQuotasQuotaInfo extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleCloudQuotasQuotaInfo resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleCloudQuotasQuotaInfo to import
    * @param importFromId The id of the existing DataGoogleCloudQuotasQuotaInfo that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_quotas_quota_info#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleCloudQuotasQuotaInfo to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_cloud_quotas_quota_info", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_quotas_quota_info google_cloud_quotas_quota_info} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleCloudQuotasQuotaInfoConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloud_quotas_quota_info',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dimensions_infos - computed: true, optional: false, required: false
        this._dimensionsInfos = new DataGoogleCloudQuotasQuotaInfoDimensionsInfosList(this, "dimensions_infos", false);
        // quota_increase_eligibility - computed: true, optional: false, required: false
        this._quotaIncreaseEligibility = new DataGoogleCloudQuotasQuotaInfoQuotaIncreaseEligibilityList(this, "quota_increase_eligibility", false);
        this._id = config.id;
        this._parent = config.parent;
        this._quotaId = config.quotaId;
        this._service = config.service;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // container_type - computed: true, optional: false, required: false
    get containerType() {
        return this.getStringAttribute('container_type');
    }
    // dimensions - computed: true, optional: false, required: false
    get dimensions() {
        return this.getListAttribute('dimensions');
    }
    get dimensionsInfos() {
        return this._dimensionsInfos;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // is_concurrent - computed: true, optional: false, required: false
    get isConcurrent() {
        return this.getBooleanAttribute('is_concurrent');
    }
    // is_fixed - computed: true, optional: false, required: false
    get isFixed() {
        return this.getBooleanAttribute('is_fixed');
    }
    // is_precise - computed: true, optional: false, required: false
    get isPrecise() {
        return this.getBooleanAttribute('is_precise');
    }
    // metric - computed: true, optional: false, required: false
    get metric() {
        return this.getStringAttribute('metric');
    }
    // metric_display_name - computed: true, optional: false, required: false
    get metricDisplayName() {
        return this.getStringAttribute('metric_display_name');
    }
    // metric_unit - computed: true, optional: false, required: false
    get metricUnit() {
        return this.getStringAttribute('metric_unit');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    // quota_display_name - computed: true, optional: false, required: false
    get quotaDisplayName() {
        return this.getStringAttribute('quota_display_name');
    }
    get quotaId() {
        return this.getStringAttribute('quota_id');
    }
    set quotaId(value) {
        this._quotaId = value;
    }
    // Temporarily expose input value. Use with caution.
    get quotaIdInput() {
        return this._quotaId;
    }
    get quotaIncreaseEligibility() {
        return this._quotaIncreaseEligibility;
    }
    // refresh_interval - computed: true, optional: false, required: false
    get refreshInterval() {
        return this.getStringAttribute('refresh_interval');
    }
    get service() {
        return this.getStringAttribute('service');
    }
    set service(value) {
        this._service = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceInput() {
        return this._service;
    }
    // service_request_quota_uri - computed: true, optional: false, required: false
    get serviceRequestQuotaUri() {
        return this.getStringAttribute('service_request_quota_uri');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            parent: cdktf.stringToTerraform(this._parent),
            quota_id: cdktf.stringToTerraform(this._quotaId),
            service: cdktf.stringToTerraform(this._service),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent: {
                value: cdktf.stringToHclTerraform(this._parent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            quota_id: {
                value: cdktf.stringToHclTerraform(this._quotaId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service: {
                value: cdktf.stringToHclTerraform(this._service),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleCloudQuotasQuotaInfo = DataGoogleCloudQuotasQuotaInfo;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleCloudQuotasQuotaInfo[_g] = { fqn: "@cdktf/provider-google.dataGoogleCloudQuotasQuotaInfo.DataGoogleCloudQuotasQuotaInfo", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleCloudQuotasQuotaInfo.tfResourceType = "google_cloud_quotas_quota_info";
