// @cdktf/provider-google/lib/apigee-keystores-aliases-self-signed-cert/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApigeeKeystoresAliasesSelfSignedCert = exports.ApigeeKeystoresAliasesSelfSignedCertTimeoutsOutputReference = exports.apigeeKeystoresAliasesSelfSignedCertTimeoutsToHclTerraform = exports.apigeeKeystoresAliasesSelfSignedCertTimeoutsToTerraform = exports.ApigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesOutputReference = exports.apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToHclTerraform = exports.apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToTerraform = exports.ApigeeKeystoresAliasesSelfSignedCertSubjectOutputReference = exports.apigeeKeystoresAliasesSelfSignedCertSubjectToHclTerraform = exports.apigeeKeystoresAliasesSelfSignedCertSubjectToTerraform = exports.ApigeeKeystoresAliasesSelfSignedCertCertsInfoList = exports.ApigeeKeystoresAliasesSelfSignedCertCertsInfoOutputReference = exports.apigeeKeystoresAliasesSelfSignedCertCertsInfoToHclTerraform = exports.apigeeKeystoresAliasesSelfSignedCertCertsInfoToTerraform = exports.ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoList = exports.ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoOutputReference = exports.apigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoToHclTerraform = exports.apigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.apigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoToTerraform = apigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoToTerraform;
function apigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.apigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoToHclTerraform = apigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoToHclTerraform;
class ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // basic_constraints - computed: true, optional: false, required: false
    get basicConstraints() {
        return this.getStringAttribute('basic_constraints');
    }
    // expiry_date - computed: true, optional: false, required: false
    get expiryDate() {
        return this.getStringAttribute('expiry_date');
    }
    // is_valid - computed: true, optional: false, required: false
    get isValid() {
        return this.getStringAttribute('is_valid');
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    // public_key - computed: true, optional: false, required: false
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
    // serial_number - computed: true, optional: false, required: false
    get serialNumber() {
        return this.getStringAttribute('serial_number');
    }
    // sig_alg_name - computed: true, optional: false, required: false
    get sigAlgName() {
        return this.getStringAttribute('sig_alg_name');
    }
    // subject - computed: true, optional: false, required: false
    get subject() {
        return this.getStringAttribute('subject');
    }
    // subject_alternative_names - computed: true, optional: false, required: false
    get subjectAlternativeNames() {
        return this.getListAttribute('subject_alternative_names');
    }
    // valid_from - computed: true, optional: false, required: false
    get validFrom() {
        return this.getStringAttribute('valid_from');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
}
exports.ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoOutputReference = ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoOutputReference[_a] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesSelfSignedCert.ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoOutputReference", version: "14.13.1" };
class ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoList = ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoList;
_b = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoList[_b] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesSelfSignedCert.ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoList", version: "14.13.1" };
function apigeeKeystoresAliasesSelfSignedCertCertsInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.apigeeKeystoresAliasesSelfSignedCertCertsInfoToTerraform = apigeeKeystoresAliasesSelfSignedCertCertsInfoToTerraform;
function apigeeKeystoresAliasesSelfSignedCertCertsInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.apigeeKeystoresAliasesSelfSignedCertCertsInfoToHclTerraform = apigeeKeystoresAliasesSelfSignedCertCertsInfoToHclTerraform;
class ApigeeKeystoresAliasesSelfSignedCertCertsInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cert_info - computed: true, optional: false, required: false
        this._certInfo = new ApigeeKeystoresAliasesSelfSignedCertCertsInfoCertInfoList(this, "cert_info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get certInfo() {
        return this._certInfo;
    }
}
exports.ApigeeKeystoresAliasesSelfSignedCertCertsInfoOutputReference = ApigeeKeystoresAliasesSelfSignedCertCertsInfoOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesSelfSignedCertCertsInfoOutputReference[_c] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesSelfSignedCert.ApigeeKeystoresAliasesSelfSignedCertCertsInfoOutputReference", version: "14.13.1" };
class ApigeeKeystoresAliasesSelfSignedCertCertsInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApigeeKeystoresAliasesSelfSignedCertCertsInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApigeeKeystoresAliasesSelfSignedCertCertsInfoList = ApigeeKeystoresAliasesSelfSignedCertCertsInfoList;
_d = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesSelfSignedCertCertsInfoList[_d] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesSelfSignedCert.ApigeeKeystoresAliasesSelfSignedCertCertsInfoList", version: "14.13.1" };
function apigeeKeystoresAliasesSelfSignedCertSubjectToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        common_name: cdktf.stringToTerraform(struct.commonName),
        country_code: cdktf.stringToTerraform(struct.countryCode),
        email: cdktf.stringToTerraform(struct.email),
        locality: cdktf.stringToTerraform(struct.locality),
        org: cdktf.stringToTerraform(struct.org),
        org_unit: cdktf.stringToTerraform(struct.orgUnit),
        state: cdktf.stringToTerraform(struct.state),
    };
}
exports.apigeeKeystoresAliasesSelfSignedCertSubjectToTerraform = apigeeKeystoresAliasesSelfSignedCertSubjectToTerraform;
function apigeeKeystoresAliasesSelfSignedCertSubjectToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        common_name: {
            value: cdktf.stringToHclTerraform(struct.commonName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        country_code: {
            value: cdktf.stringToHclTerraform(struct.countryCode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        email: {
            value: cdktf.stringToHclTerraform(struct.email),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        locality: {
            value: cdktf.stringToHclTerraform(struct.locality),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        org: {
            value: cdktf.stringToHclTerraform(struct.org),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        org_unit: {
            value: cdktf.stringToHclTerraform(struct.orgUnit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        state: {
            value: cdktf.stringToHclTerraform(struct.state),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apigeeKeystoresAliasesSelfSignedCertSubjectToHclTerraform = apigeeKeystoresAliasesSelfSignedCertSubjectToHclTerraform;
class ApigeeKeystoresAliasesSelfSignedCertSubjectOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._commonName !== undefined) {
            hasAnyValues = true;
            internalValueResult.commonName = this._commonName;
        }
        if (this._countryCode !== undefined) {
            hasAnyValues = true;
            internalValueResult.countryCode = this._countryCode;
        }
        if (this._email !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email;
        }
        if (this._locality !== undefined) {
            hasAnyValues = true;
            internalValueResult.locality = this._locality;
        }
        if (this._org !== undefined) {
            hasAnyValues = true;
            internalValueResult.org = this._org;
        }
        if (this._orgUnit !== undefined) {
            hasAnyValues = true;
            internalValueResult.orgUnit = this._orgUnit;
        }
        if (this._state !== undefined) {
            hasAnyValues = true;
            internalValueResult.state = this._state;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._commonName = undefined;
            this._countryCode = undefined;
            this._email = undefined;
            this._locality = undefined;
            this._org = undefined;
            this._orgUnit = undefined;
            this._state = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._commonName = value.commonName;
            this._countryCode = value.countryCode;
            this._email = value.email;
            this._locality = value.locality;
            this._org = value.org;
            this._orgUnit = value.orgUnit;
            this._state = value.state;
        }
    }
    get commonName() {
        return this.getStringAttribute('common_name');
    }
    set commonName(value) {
        this._commonName = value;
    }
    resetCommonName() {
        this._commonName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commonNameInput() {
        return this._commonName;
    }
    get countryCode() {
        return this.getStringAttribute('country_code');
    }
    set countryCode(value) {
        this._countryCode = value;
    }
    resetCountryCode() {
        this._countryCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countryCodeInput() {
        return this._countryCode;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    resetEmail() {
        this._email = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
    get locality() {
        return this.getStringAttribute('locality');
    }
    set locality(value) {
        this._locality = value;
    }
    resetLocality() {
        this._locality = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get localityInput() {
        return this._locality;
    }
    get org() {
        return this.getStringAttribute('org');
    }
    set org(value) {
        this._org = value;
    }
    resetOrg() {
        this._org = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get orgInput() {
        return this._org;
    }
    get orgUnit() {
        return this.getStringAttribute('org_unit');
    }
    set orgUnit(value) {
        this._orgUnit = value;
    }
    resetOrgUnit() {
        this._orgUnit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get orgUnitInput() {
        return this._orgUnit;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
}
exports.ApigeeKeystoresAliasesSelfSignedCertSubjectOutputReference = ApigeeKeystoresAliasesSelfSignedCertSubjectOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesSelfSignedCertSubjectOutputReference[_e] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesSelfSignedCert.ApigeeKeystoresAliasesSelfSignedCertSubjectOutputReference", version: "14.13.1" };
function apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subject_alternative_name: cdktf.stringToTerraform(struct.subjectAlternativeName),
    };
}
exports.apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToTerraform = apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToTerraform;
function apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        subject_alternative_name: {
            value: cdktf.stringToHclTerraform(struct.subjectAlternativeName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToHclTerraform = apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToHclTerraform;
class ApigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._subjectAlternativeName !== undefined) {
            hasAnyValues = true;
            internalValueResult.subjectAlternativeName = this._subjectAlternativeName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._subjectAlternativeName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._subjectAlternativeName = value.subjectAlternativeName;
        }
    }
    get subjectAlternativeName() {
        return this.getStringAttribute('subject_alternative_name');
    }
    set subjectAlternativeName(value) {
        this._subjectAlternativeName = value;
    }
    resetSubjectAlternativeName() {
        this._subjectAlternativeName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subjectAlternativeNameInput() {
        return this._subjectAlternativeName;
    }
}
exports.ApigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesOutputReference = ApigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesOutputReference[_f] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesSelfSignedCert.ApigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesOutputReference", version: "14.13.1" };
function apigeeKeystoresAliasesSelfSignedCertTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.apigeeKeystoresAliasesSelfSignedCertTimeoutsToTerraform = apigeeKeystoresAliasesSelfSignedCertTimeoutsToTerraform;
function apigeeKeystoresAliasesSelfSignedCertTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apigeeKeystoresAliasesSelfSignedCertTimeoutsToHclTerraform = apigeeKeystoresAliasesSelfSignedCertTimeoutsToHclTerraform;
class ApigeeKeystoresAliasesSelfSignedCertTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.ApigeeKeystoresAliasesSelfSignedCertTimeoutsOutputReference = ApigeeKeystoresAliasesSelfSignedCertTimeoutsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesSelfSignedCertTimeoutsOutputReference[_g] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesSelfSignedCert.ApigeeKeystoresAliasesSelfSignedCertTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apigee_keystores_aliases_self_signed_cert google_apigee_keystores_aliases_self_signed_cert}
*/
class ApigeeKeystoresAliasesSelfSignedCert extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ApigeeKeystoresAliasesSelfSignedCert resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ApigeeKeystoresAliasesSelfSignedCert to import
    * @param importFromId The id of the existing ApigeeKeystoresAliasesSelfSignedCert that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apigee_keystores_aliases_self_signed_cert#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ApigeeKeystoresAliasesSelfSignedCert to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_apigee_keystores_aliases_self_signed_cert", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apigee_keystores_aliases_self_signed_cert google_apigee_keystores_aliases_self_signed_cert} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApigeeKeystoresAliasesSelfSignedCertConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_apigee_keystores_aliases_self_signed_cert',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // certs_info - computed: true, optional: false, required: false
        this._certsInfo = new ApigeeKeystoresAliasesSelfSignedCertCertsInfoList(this, "certs_info", false);
        // subject - computed: false, optional: false, required: true
        this._subject = new ApigeeKeystoresAliasesSelfSignedCertSubjectOutputReference(this, "subject");
        // subject_alternative_dns_names - computed: false, optional: true, required: false
        this._subjectAlternativeDnsNames = new ApigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesOutputReference(this, "subject_alternative_dns_names");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ApigeeKeystoresAliasesSelfSignedCertTimeoutsOutputReference(this, "timeouts");
        this._alias = config.alias;
        this._certValidityInDays = config.certValidityInDays;
        this._environment = config.environment;
        this._id = config.id;
        this._keySize = config.keySize;
        this._keystore = config.keystore;
        this._orgId = config.orgId;
        this._sigAlg = config.sigAlg;
        this._subject.internalValue = config.subject;
        this._subjectAlternativeDnsNames.internalValue = config.subjectAlternativeDnsNames;
        this._timeouts.internalValue = config.timeouts;
    }
    get alias() {
        return this.getStringAttribute('alias');
    }
    set alias(value) {
        this._alias = value;
    }
    // Temporarily expose input value. Use with caution.
    get aliasInput() {
        return this._alias;
    }
    get certValidityInDays() {
        return this.getNumberAttribute('cert_validity_in_days');
    }
    set certValidityInDays(value) {
        this._certValidityInDays = value;
    }
    resetCertValidityInDays() {
        this._certValidityInDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get certValidityInDaysInput() {
        return this._certValidityInDays;
    }
    get certsInfo() {
        return this._certsInfo;
    }
    get environment() {
        return this.getStringAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keySize() {
        return this.getStringAttribute('key_size');
    }
    set keySize(value) {
        this._keySize = value;
    }
    resetKeySize() {
        this._keySize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keySizeInput() {
        return this._keySize;
    }
    get keystore() {
        return this.getStringAttribute('keystore');
    }
    set keystore(value) {
        this._keystore = value;
    }
    // Temporarily expose input value. Use with caution.
    get keystoreInput() {
        return this._keystore;
    }
    get orgId() {
        return this.getStringAttribute('org_id');
    }
    set orgId(value) {
        this._orgId = value;
    }
    // Temporarily expose input value. Use with caution.
    get orgIdInput() {
        return this._orgId;
    }
    get sigAlg() {
        return this.getStringAttribute('sig_alg');
    }
    set sigAlg(value) {
        this._sigAlg = value;
    }
    // Temporarily expose input value. Use with caution.
    get sigAlgInput() {
        return this._sigAlg;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    get subject() {
        return this._subject;
    }
    putSubject(value) {
        this._subject.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get subjectInput() {
        return this._subject.internalValue;
    }
    get subjectAlternativeDnsNames() {
        return this._subjectAlternativeDnsNames;
    }
    putSubjectAlternativeDnsNames(value) {
        this._subjectAlternativeDnsNames.internalValue = value;
    }
    resetSubjectAlternativeDnsNames() {
        this._subjectAlternativeDnsNames.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subjectAlternativeDnsNamesInput() {
        return this._subjectAlternativeDnsNames.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            alias: cdktf.stringToTerraform(this._alias),
            cert_validity_in_days: cdktf.numberToTerraform(this._certValidityInDays),
            environment: cdktf.stringToTerraform(this._environment),
            id: cdktf.stringToTerraform(this._id),
            key_size: cdktf.stringToTerraform(this._keySize),
            keystore: cdktf.stringToTerraform(this._keystore),
            org_id: cdktf.stringToTerraform(this._orgId),
            sig_alg: cdktf.stringToTerraform(this._sigAlg),
            subject: apigeeKeystoresAliasesSelfSignedCertSubjectToTerraform(this._subject.internalValue),
            subject_alternative_dns_names: apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToTerraform(this._subjectAlternativeDnsNames.internalValue),
            timeouts: apigeeKeystoresAliasesSelfSignedCertTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            alias: {
                value: cdktf.stringToHclTerraform(this._alias),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            cert_validity_in_days: {
                value: cdktf.numberToHclTerraform(this._certValidityInDays),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            environment: {
                value: cdktf.stringToHclTerraform(this._environment),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            key_size: {
                value: cdktf.stringToHclTerraform(this._keySize),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            keystore: {
                value: cdktf.stringToHclTerraform(this._keystore),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            org_id: {
                value: cdktf.stringToHclTerraform(this._orgId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            sig_alg: {
                value: cdktf.stringToHclTerraform(this._sigAlg),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subject: {
                value: apigeeKeystoresAliasesSelfSignedCertSubjectToHclTerraform(this._subject.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApigeeKeystoresAliasesSelfSignedCertSubjectList",
            },
            subject_alternative_dns_names: {
                value: apigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesToHclTerraform(this._subjectAlternativeDnsNames.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApigeeKeystoresAliasesSelfSignedCertSubjectAlternativeDnsNamesList",
            },
            timeouts: {
                value: apigeeKeystoresAliasesSelfSignedCertTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ApigeeKeystoresAliasesSelfSignedCertTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ApigeeKeystoresAliasesSelfSignedCert = ApigeeKeystoresAliasesSelfSignedCert;
_h = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesSelfSignedCert[_h] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesSelfSignedCert.ApigeeKeystoresAliasesSelfSignedCert", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ApigeeKeystoresAliasesSelfSignedCert.tfResourceType = "google_apigee_keystores_aliases_self_signed_cert";
