// @cdktf/provider-google/lib/iam-workload-identity-pool-provider/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamWorkloadIdentityPoolProvider = exports.IamWorkloadIdentityPoolProviderX509OutputReference = exports.iamWorkloadIdentityPoolProviderX509ToHclTerraform = exports.iamWorkloadIdentityPoolProviderX509ToTerraform = exports.IamWorkloadIdentityPoolProviderX509TrustStoreOutputReference = exports.iamWorkloadIdentityPoolProviderX509TrustStoreToHclTerraform = exports.iamWorkloadIdentityPoolProviderX509TrustStoreToTerraform = exports.IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsList = exports.IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsOutputReference = exports.iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToHclTerraform = exports.iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToTerraform = exports.IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasList = exports.IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasOutputReference = exports.iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToHclTerraform = exports.iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToTerraform = exports.IamWorkloadIdentityPoolProviderTimeoutsOutputReference = exports.iamWorkloadIdentityPoolProviderTimeoutsToHclTerraform = exports.iamWorkloadIdentityPoolProviderTimeoutsToTerraform = exports.IamWorkloadIdentityPoolProviderSamlOutputReference = exports.iamWorkloadIdentityPoolProviderSamlToHclTerraform = exports.iamWorkloadIdentityPoolProviderSamlToTerraform = exports.IamWorkloadIdentityPoolProviderOidcOutputReference = exports.iamWorkloadIdentityPoolProviderOidcToHclTerraform = exports.iamWorkloadIdentityPoolProviderOidcToTerraform = exports.IamWorkloadIdentityPoolProviderAwsOutputReference = exports.iamWorkloadIdentityPoolProviderAwsToHclTerraform = exports.iamWorkloadIdentityPoolProviderAwsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iamWorkloadIdentityPoolProviderAwsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        account_id: cdktf.stringToTerraform(struct.accountId),
    };
}
exports.iamWorkloadIdentityPoolProviderAwsToTerraform = iamWorkloadIdentityPoolProviderAwsToTerraform;
function iamWorkloadIdentityPoolProviderAwsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        account_id: {
            value: cdktf.stringToHclTerraform(struct.accountId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkloadIdentityPoolProviderAwsToHclTerraform = iamWorkloadIdentityPoolProviderAwsToHclTerraform;
class IamWorkloadIdentityPoolProviderAwsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accountId !== undefined) {
            hasAnyValues = true;
            internalValueResult.accountId = this._accountId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accountId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accountId = value.accountId;
        }
    }
    get accountId() {
        return this.getStringAttribute('account_id');
    }
    set accountId(value) {
        this._accountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get accountIdInput() {
        return this._accountId;
    }
}
exports.IamWorkloadIdentityPoolProviderAwsOutputReference = IamWorkloadIdentityPoolProviderAwsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderAwsOutputReference[_a] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderAwsOutputReference", version: "14.13.1" };
function iamWorkloadIdentityPoolProviderOidcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        allowed_audiences: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.allowedAudiences),
        issuer_uri: cdktf.stringToTerraform(struct.issuerUri),
        jwks_json: cdktf.stringToTerraform(struct.jwksJson),
    };
}
exports.iamWorkloadIdentityPoolProviderOidcToTerraform = iamWorkloadIdentityPoolProviderOidcToTerraform;
function iamWorkloadIdentityPoolProviderOidcToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        allowed_audiences: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.allowedAudiences),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        issuer_uri: {
            value: cdktf.stringToHclTerraform(struct.issuerUri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        jwks_json: {
            value: cdktf.stringToHclTerraform(struct.jwksJson),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkloadIdentityPoolProviderOidcToHclTerraform = iamWorkloadIdentityPoolProviderOidcToHclTerraform;
class IamWorkloadIdentityPoolProviderOidcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allowedAudiences !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowedAudiences = this._allowedAudiences;
        }
        if (this._issuerUri !== undefined) {
            hasAnyValues = true;
            internalValueResult.issuerUri = this._issuerUri;
        }
        if (this._jwksJson !== undefined) {
            hasAnyValues = true;
            internalValueResult.jwksJson = this._jwksJson;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allowedAudiences = undefined;
            this._issuerUri = undefined;
            this._jwksJson = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allowedAudiences = value.allowedAudiences;
            this._issuerUri = value.issuerUri;
            this._jwksJson = value.jwksJson;
        }
    }
    get allowedAudiences() {
        return this.getListAttribute('allowed_audiences');
    }
    set allowedAudiences(value) {
        this._allowedAudiences = value;
    }
    resetAllowedAudiences() {
        this._allowedAudiences = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowedAudiencesInput() {
        return this._allowedAudiences;
    }
    get issuerUri() {
        return this.getStringAttribute('issuer_uri');
    }
    set issuerUri(value) {
        this._issuerUri = value;
    }
    // Temporarily expose input value. Use with caution.
    get issuerUriInput() {
        return this._issuerUri;
    }
    get jwksJson() {
        return this.getStringAttribute('jwks_json');
    }
    set jwksJson(value) {
        this._jwksJson = value;
    }
    resetJwksJson() {
        this._jwksJson = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jwksJsonInput() {
        return this._jwksJson;
    }
}
exports.IamWorkloadIdentityPoolProviderOidcOutputReference = IamWorkloadIdentityPoolProviderOidcOutputReference;
_b = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderOidcOutputReference[_b] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderOidcOutputReference", version: "14.13.1" };
function iamWorkloadIdentityPoolProviderSamlToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        idp_metadata_xml: cdktf.stringToTerraform(struct.idpMetadataXml),
    };
}
exports.iamWorkloadIdentityPoolProviderSamlToTerraform = iamWorkloadIdentityPoolProviderSamlToTerraform;
function iamWorkloadIdentityPoolProviderSamlToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        idp_metadata_xml: {
            value: cdktf.stringToHclTerraform(struct.idpMetadataXml),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkloadIdentityPoolProviderSamlToHclTerraform = iamWorkloadIdentityPoolProviderSamlToHclTerraform;
class IamWorkloadIdentityPoolProviderSamlOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._idpMetadataXml !== undefined) {
            hasAnyValues = true;
            internalValueResult.idpMetadataXml = this._idpMetadataXml;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._idpMetadataXml = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._idpMetadataXml = value.idpMetadataXml;
        }
    }
    get idpMetadataXml() {
        return this.getStringAttribute('idp_metadata_xml');
    }
    set idpMetadataXml(value) {
        this._idpMetadataXml = value;
    }
    // Temporarily expose input value. Use with caution.
    get idpMetadataXmlInput() {
        return this._idpMetadataXml;
    }
}
exports.IamWorkloadIdentityPoolProviderSamlOutputReference = IamWorkloadIdentityPoolProviderSamlOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderSamlOutputReference[_c] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderSamlOutputReference", version: "14.13.1" };
function iamWorkloadIdentityPoolProviderTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.iamWorkloadIdentityPoolProviderTimeoutsToTerraform = iamWorkloadIdentityPoolProviderTimeoutsToTerraform;
function iamWorkloadIdentityPoolProviderTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkloadIdentityPoolProviderTimeoutsToHclTerraform = iamWorkloadIdentityPoolProviderTimeoutsToHclTerraform;
class IamWorkloadIdentityPoolProviderTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.IamWorkloadIdentityPoolProviderTimeoutsOutputReference = IamWorkloadIdentityPoolProviderTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderTimeoutsOutputReference", version: "14.13.1" };
function iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        pem_certificate: cdktf.stringToTerraform(struct.pemCertificate),
    };
}
exports.iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToTerraform = iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToTerraform;
function iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        pem_certificate: {
            value: cdktf.stringToHclTerraform(struct.pemCertificate),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToHclTerraform = iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToHclTerraform;
class IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._pemCertificate !== undefined) {
            hasAnyValues = true;
            internalValueResult.pemCertificate = this._pemCertificate;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._pemCertificate = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._pemCertificate = value.pemCertificate;
        }
    }
    get pemCertificate() {
        return this.getStringAttribute('pem_certificate');
    }
    set pemCertificate(value) {
        this._pemCertificate = value;
    }
    resetPemCertificate() {
        this._pemCertificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pemCertificateInput() {
        return this._pemCertificate;
    }
}
exports.IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasOutputReference = IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasOutputReference;
_e = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasOutputReference[_e] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasOutputReference", version: "14.13.1" };
class IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasList = IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasList;
_f = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasList[_f] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasList", version: "14.13.1" };
function iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        pem_certificate: cdktf.stringToTerraform(struct.pemCertificate),
    };
}
exports.iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToTerraform = iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToTerraform;
function iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        pem_certificate: {
            value: cdktf.stringToHclTerraform(struct.pemCertificate),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToHclTerraform = iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToHclTerraform;
class IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._pemCertificate !== undefined) {
            hasAnyValues = true;
            internalValueResult.pemCertificate = this._pemCertificate;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._pemCertificate = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._pemCertificate = value.pemCertificate;
        }
    }
    get pemCertificate() {
        return this.getStringAttribute('pem_certificate');
    }
    set pemCertificate(value) {
        this._pemCertificate = value;
    }
    resetPemCertificate() {
        this._pemCertificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pemCertificateInput() {
        return this._pemCertificate;
    }
}
exports.IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsOutputReference = IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsOutputReference[_g] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsOutputReference", version: "14.13.1" };
class IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsList = IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsList;
_h = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsList[_h] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsList", version: "14.13.1" };
function iamWorkloadIdentityPoolProviderX509TrustStoreToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        intermediate_cas: cdktf.listMapper(iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToTerraform, true)(struct.intermediateCas),
        trust_anchors: cdktf.listMapper(iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToTerraform, true)(struct.trustAnchors),
    };
}
exports.iamWorkloadIdentityPoolProviderX509TrustStoreToTerraform = iamWorkloadIdentityPoolProviderX509TrustStoreToTerraform;
function iamWorkloadIdentityPoolProviderX509TrustStoreToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        intermediate_cas: {
            value: cdktf.listMapperHcl(iamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasToHclTerraform, true)(struct.intermediateCas),
            isBlock: true,
            type: "list",
            storageClassType: "IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasList",
        },
        trust_anchors: {
            value: cdktf.listMapperHcl(iamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsToHclTerraform, true)(struct.trustAnchors),
            isBlock: true,
            type: "list",
            storageClassType: "IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkloadIdentityPoolProviderX509TrustStoreToHclTerraform = iamWorkloadIdentityPoolProviderX509TrustStoreToHclTerraform;
class IamWorkloadIdentityPoolProviderX509TrustStoreOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // intermediate_cas - computed: false, optional: true, required: false
        this._intermediateCas = new IamWorkloadIdentityPoolProviderX509TrustStoreIntermediateCasList(this, "intermediate_cas", false);
        // trust_anchors - computed: false, optional: false, required: true
        this._trustAnchors = new IamWorkloadIdentityPoolProviderX509TrustStoreTrustAnchorsList(this, "trust_anchors", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._intermediateCas?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.intermediateCas = this._intermediateCas?.internalValue;
        }
        if (this._trustAnchors?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.trustAnchors = this._trustAnchors?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._intermediateCas.internalValue = undefined;
            this._trustAnchors.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._intermediateCas.internalValue = value.intermediateCas;
            this._trustAnchors.internalValue = value.trustAnchors;
        }
    }
    get intermediateCas() {
        return this._intermediateCas;
    }
    putIntermediateCas(value) {
        this._intermediateCas.internalValue = value;
    }
    resetIntermediateCas() {
        this._intermediateCas.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get intermediateCasInput() {
        return this._intermediateCas.internalValue;
    }
    get trustAnchors() {
        return this._trustAnchors;
    }
    putTrustAnchors(value) {
        this._trustAnchors.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get trustAnchorsInput() {
        return this._trustAnchors.internalValue;
    }
}
exports.IamWorkloadIdentityPoolProviderX509TrustStoreOutputReference = IamWorkloadIdentityPoolProviderX509TrustStoreOutputReference;
_j = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderX509TrustStoreOutputReference[_j] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderX509TrustStoreOutputReference", version: "14.13.1" };
function iamWorkloadIdentityPoolProviderX509ToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        trust_store: iamWorkloadIdentityPoolProviderX509TrustStoreToTerraform(struct.trustStore),
    };
}
exports.iamWorkloadIdentityPoolProviderX509ToTerraform = iamWorkloadIdentityPoolProviderX509ToTerraform;
function iamWorkloadIdentityPoolProviderX509ToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        trust_store: {
            value: iamWorkloadIdentityPoolProviderX509TrustStoreToHclTerraform(struct.trustStore),
            isBlock: true,
            type: "list",
            storageClassType: "IamWorkloadIdentityPoolProviderX509TrustStoreList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkloadIdentityPoolProviderX509ToHclTerraform = iamWorkloadIdentityPoolProviderX509ToHclTerraform;
class IamWorkloadIdentityPoolProviderX509OutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // trust_store - computed: false, optional: false, required: true
        this._trustStore = new IamWorkloadIdentityPoolProviderX509TrustStoreOutputReference(this, "trust_store");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._trustStore?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.trustStore = this._trustStore?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._trustStore.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._trustStore.internalValue = value.trustStore;
        }
    }
    get trustStore() {
        return this._trustStore;
    }
    putTrustStore(value) {
        this._trustStore.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get trustStoreInput() {
        return this._trustStore.internalValue;
    }
}
exports.IamWorkloadIdentityPoolProviderX509OutputReference = IamWorkloadIdentityPoolProviderX509OutputReference;
_k = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProviderX509OutputReference[_k] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProviderX509OutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_workload_identity_pool_provider google_iam_workload_identity_pool_provider}
*/
class IamWorkloadIdentityPoolProvider extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IamWorkloadIdentityPoolProvider resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IamWorkloadIdentityPoolProvider to import
    * @param importFromId The id of the existing IamWorkloadIdentityPoolProvider that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_workload_identity_pool_provider#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IamWorkloadIdentityPoolProvider to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_iam_workload_identity_pool_provider", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_workload_identity_pool_provider google_iam_workload_identity_pool_provider} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamWorkloadIdentityPoolProviderConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_iam_workload_identity_pool_provider',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // aws - computed: false, optional: true, required: false
        this._aws = new IamWorkloadIdentityPoolProviderAwsOutputReference(this, "aws");
        // oidc - computed: false, optional: true, required: false
        this._oidc = new IamWorkloadIdentityPoolProviderOidcOutputReference(this, "oidc");
        // saml - computed: false, optional: true, required: false
        this._saml = new IamWorkloadIdentityPoolProviderSamlOutputReference(this, "saml");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new IamWorkloadIdentityPoolProviderTimeoutsOutputReference(this, "timeouts");
        // x509 - computed: false, optional: true, required: false
        this._x509 = new IamWorkloadIdentityPoolProviderX509OutputReference(this, "x509");
        this._attributeCondition = config.attributeCondition;
        this._attributeMapping = config.attributeMapping;
        this._description = config.description;
        this._disabled = config.disabled;
        this._displayName = config.displayName;
        this._id = config.id;
        this._project = config.project;
        this._workloadIdentityPoolId = config.workloadIdentityPoolId;
        this._workloadIdentityPoolProviderId = config.workloadIdentityPoolProviderId;
        this._aws.internalValue = config.aws;
        this._oidc.internalValue = config.oidc;
        this._saml.internalValue = config.saml;
        this._timeouts.internalValue = config.timeouts;
        this._x509.internalValue = config.x509;
    }
    get attributeCondition() {
        return this.getStringAttribute('attribute_condition');
    }
    set attributeCondition(value) {
        this._attributeCondition = value;
    }
    resetAttributeCondition() {
        this._attributeCondition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attributeConditionInput() {
        return this._attributeCondition;
    }
    get attributeMapping() {
        return this.getStringMapAttribute('attribute_mapping');
    }
    set attributeMapping(value) {
        this._attributeMapping = value;
    }
    resetAttributeMapping() {
        this._attributeMapping = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attributeMappingInput() {
        return this._attributeMapping;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    set disabled(value) {
        this._disabled = value;
    }
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disabledInput() {
        return this._disabled;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get workloadIdentityPoolId() {
        return this.getStringAttribute('workload_identity_pool_id');
    }
    set workloadIdentityPoolId(value) {
        this._workloadIdentityPoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get workloadIdentityPoolIdInput() {
        return this._workloadIdentityPoolId;
    }
    get workloadIdentityPoolProviderId() {
        return this.getStringAttribute('workload_identity_pool_provider_id');
    }
    set workloadIdentityPoolProviderId(value) {
        this._workloadIdentityPoolProviderId = value;
    }
    // Temporarily expose input value. Use with caution.
    get workloadIdentityPoolProviderIdInput() {
        return this._workloadIdentityPoolProviderId;
    }
    get aws() {
        return this._aws;
    }
    putAws(value) {
        this._aws.internalValue = value;
    }
    resetAws() {
        this._aws.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsInput() {
        return this._aws.internalValue;
    }
    get oidc() {
        return this._oidc;
    }
    putOidc(value) {
        this._oidc.internalValue = value;
    }
    resetOidc() {
        this._oidc.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get oidcInput() {
        return this._oidc.internalValue;
    }
    get saml() {
        return this._saml;
    }
    putSaml(value) {
        this._saml.internalValue = value;
    }
    resetSaml() {
        this._saml.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get samlInput() {
        return this._saml.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get x509() {
        return this._x509;
    }
    putX509(value) {
        this._x509.internalValue = value;
    }
    resetX509() {
        this._x509.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get x509Input() {
        return this._x509.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            attribute_condition: cdktf.stringToTerraform(this._attributeCondition),
            attribute_mapping: cdktf.hashMapper(cdktf.stringToTerraform)(this._attributeMapping),
            description: cdktf.stringToTerraform(this._description),
            disabled: cdktf.booleanToTerraform(this._disabled),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
            workload_identity_pool_id: cdktf.stringToTerraform(this._workloadIdentityPoolId),
            workload_identity_pool_provider_id: cdktf.stringToTerraform(this._workloadIdentityPoolProviderId),
            aws: iamWorkloadIdentityPoolProviderAwsToTerraform(this._aws.internalValue),
            oidc: iamWorkloadIdentityPoolProviderOidcToTerraform(this._oidc.internalValue),
            saml: iamWorkloadIdentityPoolProviderSamlToTerraform(this._saml.internalValue),
            timeouts: iamWorkloadIdentityPoolProviderTimeoutsToTerraform(this._timeouts.internalValue),
            x509: iamWorkloadIdentityPoolProviderX509ToTerraform(this._x509.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            attribute_condition: {
                value: cdktf.stringToHclTerraform(this._attributeCondition),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            attribute_mapping: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._attributeMapping),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            disabled: {
                value: cdktf.booleanToHclTerraform(this._disabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            workload_identity_pool_id: {
                value: cdktf.stringToHclTerraform(this._workloadIdentityPoolId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            workload_identity_pool_provider_id: {
                value: cdktf.stringToHclTerraform(this._workloadIdentityPoolProviderId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            aws: {
                value: iamWorkloadIdentityPoolProviderAwsToHclTerraform(this._aws.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IamWorkloadIdentityPoolProviderAwsList",
            },
            oidc: {
                value: iamWorkloadIdentityPoolProviderOidcToHclTerraform(this._oidc.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IamWorkloadIdentityPoolProviderOidcList",
            },
            saml: {
                value: iamWorkloadIdentityPoolProviderSamlToHclTerraform(this._saml.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IamWorkloadIdentityPoolProviderSamlList",
            },
            timeouts: {
                value: iamWorkloadIdentityPoolProviderTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "IamWorkloadIdentityPoolProviderTimeouts",
            },
            x509: {
                value: iamWorkloadIdentityPoolProviderX509ToHclTerraform(this._x509.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IamWorkloadIdentityPoolProviderX509List",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IamWorkloadIdentityPoolProvider = IamWorkloadIdentityPoolProvider;
_l = JSII_RTTI_SYMBOL_1;
IamWorkloadIdentityPoolProvider[_l] = { fqn: "@cdktf/provider-google.iamWorkloadIdentityPoolProvider.IamWorkloadIdentityPoolProvider", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
IamWorkloadIdentityPoolProvider.tfResourceType = "google_iam_workload_identity_pool_provider";
