// @cdktf/provider-google/lib/data-google-cloud-run-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToTerraform = exports.DataGoogleCloudRunServiceTemplateMetadataList = exports.DataGoogleCloudRunServiceTemplateMetadataOutputReference = exports.dataGoogleCloudRunServiceTemplateMetadataToHclTerraform = exports.dataGoogleCloudRunServiceTemplateMetadataToTerraform = exports.DataGoogleCloudRunServiceStatusList = exports.DataGoogleCloudRunServiceStatusOutputReference = exports.dataGoogleCloudRunServiceStatusToHclTerraform = exports.dataGoogleCloudRunServiceStatusToTerraform = exports.DataGoogleCloudRunServiceStatusTrafficList = exports.DataGoogleCloudRunServiceStatusTrafficOutputReference = exports.dataGoogleCloudRunServiceStatusTrafficToHclTerraform = exports.dataGoogleCloudRunServiceStatusTrafficToTerraform = exports.DataGoogleCloudRunServiceStatusConditionsList = exports.DataGoogleCloudRunServiceStatusConditionsOutputReference = exports.dataGoogleCloudRunServiceStatusConditionsToHclTerraform = exports.dataGoogleCloudRunServiceStatusConditionsToTerraform = exports.DataGoogleCloudRunServiceMetadataList = exports.DataGoogleCloudRunServiceMetadataOutputReference = exports.dataGoogleCloudRunServiceMetadataToHclTerraform = exports.dataGoogleCloudRunServiceMetadataToTerraform = void 0;
exports.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList = exports.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeList = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketList = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetList = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersList = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcList = exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersResourcesList = exports.DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersResourcesToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersResourcesToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersPortsList = exports.DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersPortsToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersPortsToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeList = exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetList = exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersList = exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcList = exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList = exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference = void 0;
exports.DataGoogleCloudRunService = exports.DataGoogleCloudRunServiceTrafficList = exports.DataGoogleCloudRunServiceTrafficOutputReference = exports.dataGoogleCloudRunServiceTrafficToHclTerraform = exports.dataGoogleCloudRunServiceTrafficToTerraform = exports.DataGoogleCloudRunServiceTemplateList = exports.DataGoogleCloudRunServiceTemplateOutputReference = exports.dataGoogleCloudRunServiceTemplateToHclTerraform = exports.dataGoogleCloudRunServiceTemplateToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecList = exports.DataGoogleCloudRunServiceTemplateSpecOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecVolumesToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesNfsList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesNfsOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesNfsToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecVolumesNfsToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecVolumesCsiList = exports.DataGoogleCloudRunServiceTemplateSpecVolumesCsiOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecVolumesCsiToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecVolumesCsiToTerraform = exports.DataGoogleCloudRunServiceTemplateSpecContainersList = exports.DataGoogleCloudRunServiceTemplateSpecContainersOutputReference = exports.dataGoogleCloudRunServiceTemplateSpecContainersToHclTerraform = exports.dataGoogleCloudRunServiceTemplateSpecContainersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleCloudRunServiceMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceMetadataToTerraform = dataGoogleCloudRunServiceMetadataToTerraform;
function dataGoogleCloudRunServiceMetadataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceMetadataToHclTerraform = dataGoogleCloudRunServiceMetadataToHclTerraform;
class DataGoogleCloudRunServiceMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // annotations - computed: true, optional: false, required: false
        this._annotations = new cdktf.StringMap(this, "annotations");
        // effective_annotations - computed: true, optional: false, required: false
        this._effectiveAnnotations = new cdktf.StringMap(this, "effective_annotations");
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get annotations() {
        return this._annotations;
    }
    get effectiveAnnotations() {
        return this._effectiveAnnotations;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getNumberAttribute('generation');
    }
    get labels() {
        return this._labels;
    }
    // namespace - computed: true, optional: false, required: false
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    // resource_version - computed: true, optional: false, required: false
    get resourceVersion() {
        return this.getStringAttribute('resource_version');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
}
exports.DataGoogleCloudRunServiceMetadataOutputReference = DataGoogleCloudRunServiceMetadataOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceMetadataOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceMetadataOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceMetadataList = DataGoogleCloudRunServiceMetadataList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceMetadataList[_b] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceMetadataList", version: "14.13.1" };
function dataGoogleCloudRunServiceStatusConditionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceStatusConditionsToTerraform = dataGoogleCloudRunServiceStatusConditionsToTerraform;
function dataGoogleCloudRunServiceStatusConditionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceStatusConditionsToHclTerraform = dataGoogleCloudRunServiceStatusConditionsToHclTerraform;
class DataGoogleCloudRunServiceStatusConditionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleCloudRunServiceStatusConditionsOutputReference = DataGoogleCloudRunServiceStatusConditionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusConditionsOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusConditionsOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceStatusConditionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceStatusConditionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceStatusConditionsList = DataGoogleCloudRunServiceStatusConditionsList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusConditionsList[_d] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusConditionsList", version: "14.13.1" };
function dataGoogleCloudRunServiceStatusTrafficToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceStatusTrafficToTerraform = dataGoogleCloudRunServiceStatusTrafficToTerraform;
function dataGoogleCloudRunServiceStatusTrafficToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceStatusTrafficToHclTerraform = dataGoogleCloudRunServiceStatusTrafficToHclTerraform;
class DataGoogleCloudRunServiceStatusTrafficOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // latest_revision - computed: true, optional: false, required: false
    get latestRevision() {
        return this.getBooleanAttribute('latest_revision');
    }
    // percent - computed: true, optional: false, required: false
    get percent() {
        return this.getNumberAttribute('percent');
    }
    // revision_name - computed: true, optional: false, required: false
    get revisionName() {
        return this.getStringAttribute('revision_name');
    }
    // tag - computed: true, optional: false, required: false
    get tag() {
        return this.getStringAttribute('tag');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
}
exports.DataGoogleCloudRunServiceStatusTrafficOutputReference = DataGoogleCloudRunServiceStatusTrafficOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusTrafficOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusTrafficOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceStatusTrafficList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceStatusTrafficOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceStatusTrafficList = DataGoogleCloudRunServiceStatusTrafficList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusTrafficList[_f] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusTrafficList", version: "14.13.1" };
function dataGoogleCloudRunServiceStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceStatusToTerraform = dataGoogleCloudRunServiceStatusToTerraform;
function dataGoogleCloudRunServiceStatusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceStatusToHclTerraform = dataGoogleCloudRunServiceStatusToHclTerraform;
class DataGoogleCloudRunServiceStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // conditions - computed: true, optional: false, required: false
        this._conditions = new DataGoogleCloudRunServiceStatusConditionsList(this, "conditions", false);
        // traffic - computed: true, optional: false, required: false
        this._traffic = new DataGoogleCloudRunServiceStatusTrafficList(this, "traffic", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get conditions() {
        return this._conditions;
    }
    // latest_created_revision_name - computed: true, optional: false, required: false
    get latestCreatedRevisionName() {
        return this.getStringAttribute('latest_created_revision_name');
    }
    // latest_ready_revision_name - computed: true, optional: false, required: false
    get latestReadyRevisionName() {
        return this.getStringAttribute('latest_ready_revision_name');
    }
    // observed_generation - computed: true, optional: false, required: false
    get observedGeneration() {
        return this.getNumberAttribute('observed_generation');
    }
    get traffic() {
        return this._traffic;
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
}
exports.DataGoogleCloudRunServiceStatusOutputReference = DataGoogleCloudRunServiceStatusOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceStatusList = DataGoogleCloudRunServiceStatusList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceStatusList[_h] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceStatusList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateMetadataToTerraform = dataGoogleCloudRunServiceTemplateMetadataToTerraform;
function dataGoogleCloudRunServiceTemplateMetadataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateMetadataToHclTerraform = dataGoogleCloudRunServiceTemplateMetadataToHclTerraform;
class DataGoogleCloudRunServiceTemplateMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // annotations - computed: true, optional: false, required: false
        this._annotations = new cdktf.StringMap(this, "annotations");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get annotations() {
        return this._annotations;
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getNumberAttribute('generation');
    }
    get labels() {
        return this._labels;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // namespace - computed: true, optional: false, required: false
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    // resource_version - computed: true, optional: false, required: false
    get resourceVersion() {
        return this.getStringAttribute('resource_version');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
}
exports.DataGoogleCloudRunServiceTemplateMetadataOutputReference = DataGoogleCloudRunServiceTemplateMetadataOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateMetadataOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateMetadataOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateMetadataList = DataGoogleCloudRunServiceTemplateMetadataList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateMetadataList[_k] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateMetadataList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList = DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList[_m] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // secret_key_ref - computed: true, optional: false, required: false
        this._secretKeyRef = new DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromSecretKeyRefList(this, "secret_key_ref", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get secretKeyRef() {
        return this._secretKeyRef;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList = DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList[_p] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersEnvToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // value_from - computed: true, optional: false, required: false
        this._valueFrom = new DataGoogleCloudRunServiceTemplateSpecContainersEnvValueFromList(this, "value_from", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
    get valueFrom() {
        return this._valueFrom;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvList = DataGoogleCloudRunServiceTemplateSpecContainersEnvList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvList[_r] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList[_t] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // local_object_reference - computed: true, optional: false, required: false
        this._localObjectReference = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefLocalObjectReferenceList(this, "local_object_reference", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get localObjectReference() {
        return this._localObjectReference;
    }
    // optional - computed: true, optional: false, required: false
    get optional() {
        return this.getBooleanAttribute('optional');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList[_v] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList[_x] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // local_object_reference - computed: true, optional: false, required: false
        this._localObjectReference = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefLocalObjectReferenceList(this, "local_object_reference", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get localObjectReference() {
        return this._localObjectReference;
    }
    // optional - computed: true, optional: false, required: false
    get optional() {
        return this.getBooleanAttribute('optional');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList[_z] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersEnvFromToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // config_map_ref - computed: true, optional: false, required: false
        this._configMapRef = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromConfigMapRefList(this, "config_map_ref", false);
        // secret_ref - computed: true, optional: false, required: false
        this._secretRef = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromSecretRefList(this, "secret_ref", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get configMapRef() {
        return this._configMapRef;
    }
    // prefix - computed: true, optional: false, required: false
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    get secretRef() {
        return this._secretRef;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList = DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList[_1] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // service - computed: true, optional: false, required: false
    get service() {
        return this.getStringAttribute('service');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcList = DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcList[_3] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersList = DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersList[_5] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // http_headers - computed: true, optional: false, required: false
        this._httpHeaders = new DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetHttpHeadersList(this, "http_headers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get httpHeaders() {
        return this._httpHeaders;
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetList = DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetList[_7] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // grpc - computed: true, optional: false, required: false
        this._grpc = new DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeGrpcList(this, "grpc", false);
        // http_get - computed: true, optional: false, required: false
        this._httpGet = new DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeHttpGetList(this, "http_get", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // failure_threshold - computed: true, optional: false, required: false
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
    get grpc() {
        return this._grpc;
    }
    get httpGet() {
        return this._httpGet;
    }
    // initial_delay_seconds - computed: true, optional: false, required: false
    get initialDelaySeconds() {
        return this.getNumberAttribute('initial_delay_seconds');
    }
    // period_seconds - computed: true, optional: false, required: false
    get periodSeconds() {
        return this.getNumberAttribute('period_seconds');
    }
    // timeout_seconds - computed: true, optional: false, required: false
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeList = DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeList[_9] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersPortsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersPortsToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersPortsToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersPortsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersPortsToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersPortsToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // container_port - computed: true, optional: false, required: false
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference[_10] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersPortsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersPortsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersPortsList = DataGoogleCloudRunServiceTemplateSpecContainersPortsList;
_11 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersPortsList[_11] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersPortsList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersResourcesToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersResourcesToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersResourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersResourcesToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersResourcesToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // limits - computed: true, optional: false, required: false
        this._limits = new cdktf.StringMap(this, "limits");
        // requests - computed: true, optional: false, required: false
        this._requests = new cdktf.StringMap(this, "requests");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get limits() {
        return this._limits;
    }
    get requests() {
        return this._requests;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference[_12] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersResourcesList = DataGoogleCloudRunServiceTemplateSpecContainersResourcesList;
_13 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersResourcesList[_13] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersResourcesList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    // service - computed: true, optional: false, required: false
    get service() {
        return this.getStringAttribute('service');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcOutputReference[_14] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcList = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcList;
_15 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcList[_15] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersOutputReference[_16] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersList = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersList;
_17 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersList[_17] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // http_headers - computed: true, optional: false, required: false
        this._httpHeaders = new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetHttpHeadersList(this, "http_headers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get httpHeaders() {
        return this._httpHeaders;
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetOutputReference;
_18 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetOutputReference[_18] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetList = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetList;
_19 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetList[_19] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketOutputReference;
_20 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketOutputReference[_20] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketList = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketList;
_21 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketList[_21] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersStartupProbeToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // grpc - computed: true, optional: false, required: false
        this._grpc = new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeGrpcList(this, "grpc", false);
        // http_get - computed: true, optional: false, required: false
        this._httpGet = new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeHttpGetList(this, "http_get", false);
        // tcp_socket - computed: true, optional: false, required: false
        this._tcpSocket = new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeTcpSocketList(this, "tcp_socket", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // failure_threshold - computed: true, optional: false, required: false
    get failureThreshold() {
        return this.getNumberAttribute('failure_threshold');
    }
    get grpc() {
        return this._grpc;
    }
    get httpGet() {
        return this._httpGet;
    }
    // initial_delay_seconds - computed: true, optional: false, required: false
    get initialDelaySeconds() {
        return this.getNumberAttribute('initial_delay_seconds');
    }
    // period_seconds - computed: true, optional: false, required: false
    get periodSeconds() {
        return this.getNumberAttribute('period_seconds');
    }
    get tcpSocket() {
        return this._tcpSocket;
    }
    // timeout_seconds - computed: true, optional: false, required: false
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeOutputReference;
_22 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeOutputReference[_22] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeList = DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeList;
_23 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeList[_23] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // mount_path - computed: true, optional: false, required: false
    get mountPath() {
        return this.getStringAttribute('mount_path');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference;
_24 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference[_24] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList = DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList;
_25 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList[_25] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecContainersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersToTerraform = dataGoogleCloudRunServiceTemplateSpecContainersToTerraform;
function dataGoogleCloudRunServiceTemplateSpecContainersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecContainersToHclTerraform = dataGoogleCloudRunServiceTemplateSpecContainersToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecContainersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // env - computed: true, optional: false, required: false
        this._env = new DataGoogleCloudRunServiceTemplateSpecContainersEnvList(this, "env", true);
        // env_from - computed: true, optional: false, required: false
        this._envFrom = new DataGoogleCloudRunServiceTemplateSpecContainersEnvFromList(this, "env_from", false);
        // liveness_probe - computed: true, optional: false, required: false
        this._livenessProbe = new DataGoogleCloudRunServiceTemplateSpecContainersLivenessProbeList(this, "liveness_probe", false);
        // ports - computed: true, optional: false, required: false
        this._ports = new DataGoogleCloudRunServiceTemplateSpecContainersPortsList(this, "ports", false);
        // resources - computed: true, optional: false, required: false
        this._resources = new DataGoogleCloudRunServiceTemplateSpecContainersResourcesList(this, "resources", false);
        // startup_probe - computed: true, optional: false, required: false
        this._startupProbe = new DataGoogleCloudRunServiceTemplateSpecContainersStartupProbeList(this, "startup_probe", false);
        // volume_mounts - computed: true, optional: false, required: false
        this._volumeMounts = new DataGoogleCloudRunServiceTemplateSpecContainersVolumeMountsList(this, "volume_mounts", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // args - computed: true, optional: false, required: false
    get args() {
        return this.getListAttribute('args');
    }
    // command - computed: true, optional: false, required: false
    get command() {
        return this.getListAttribute('command');
    }
    get env() {
        return this._env;
    }
    get envFrom() {
        return this._envFrom;
    }
    // image - computed: true, optional: false, required: false
    get image() {
        return this.getStringAttribute('image');
    }
    get livenessProbe() {
        return this._livenessProbe;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get ports() {
        return this._ports;
    }
    get resources() {
        return this._resources;
    }
    get startupProbe() {
        return this._startupProbe;
    }
    get volumeMounts() {
        return this._volumeMounts;
    }
    // working_dir - computed: true, optional: false, required: false
    get workingDir() {
        return this.getStringAttribute('working_dir');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersOutputReference = DataGoogleCloudRunServiceTemplateSpecContainersOutputReference;
_26 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersOutputReference[_26] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecContainersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecContainersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecContainersList = DataGoogleCloudRunServiceTemplateSpecContainersList;
_27 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecContainersList[_27] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecContainersList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesCsiToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesCsiToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesCsiToTerraform;
function dataGoogleCloudRunServiceTemplateSpecVolumesCsiToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesCsiToHclTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesCsiToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesCsiOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // volume_attributes - computed: true, optional: false, required: false
        this._volumeAttributes = new cdktf.StringMap(this, "volume_attributes");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // driver - computed: true, optional: false, required: false
    get driver() {
        return this.getStringAttribute('driver');
    }
    // read_only - computed: true, optional: false, required: false
    get readOnly() {
        return this.getBooleanAttribute('read_only');
    }
    get volumeAttributes() {
        return this._volumeAttributes;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesCsiOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesCsiOutputReference;
_28 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesCsiOutputReference[_28] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesCsiOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesCsiList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesCsiOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesCsiList = DataGoogleCloudRunServiceTemplateSpecVolumesCsiList;
_29 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesCsiList[_29] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesCsiList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirToTerraform;
function dataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirToHclTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // medium - computed: true, optional: false, required: false
    get medium() {
        return this.getStringAttribute('medium');
    }
    // size_limit - computed: true, optional: false, required: false
    get sizeLimit() {
        return this.getStringAttribute('size_limit');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirOutputReference;
_30 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirOutputReference[_30] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirList = DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirList;
_31 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirList[_31] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesNfsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesNfsToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesNfsToTerraform;
function dataGoogleCloudRunServiceTemplateSpecVolumesNfsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesNfsToHclTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesNfsToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesNfsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // read_only - computed: true, optional: false, required: false
    get readOnly() {
        return this.getBooleanAttribute('read_only');
    }
    // server - computed: true, optional: false, required: false
    get server() {
        return this.getStringAttribute('server');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesNfsOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesNfsOutputReference;
_32 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesNfsOutputReference[_32] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesNfsOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesNfsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesNfsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesNfsList = DataGoogleCloudRunServiceTemplateSpecVolumesNfsList;
_33 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesNfsList[_33] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesNfsList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToTerraform;
function dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToHclTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getNumberAttribute('mode');
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference;
_34 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference[_34] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList = DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList;
_35 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList[_35] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesSecretToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesSecretToTerraform;
function dataGoogleCloudRunServiceTemplateSpecVolumesSecretToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesSecretToHclTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesSecretToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // items - computed: true, optional: false, required: false
        this._items = new DataGoogleCloudRunServiceTemplateSpecVolumesSecretItemsList(this, "items", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // default_mode - computed: true, optional: false, required: false
    get defaultMode() {
        return this.getNumberAttribute('default_mode');
    }
    get items() {
        return this._items;
    }
    // secret_name - computed: true, optional: false, required: false
    get secretName() {
        return this.getStringAttribute('secret_name');
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference;
_36 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference[_36] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesSecretList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesSecretOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesSecretList = DataGoogleCloudRunServiceTemplateSpecVolumesSecretList;
_37 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesSecretList[_37] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesSecretList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesToTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesToTerraform;
function dataGoogleCloudRunServiceTemplateSpecVolumesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecVolumesToHclTerraform = dataGoogleCloudRunServiceTemplateSpecVolumesToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // csi - computed: true, optional: false, required: false
        this._csi = new DataGoogleCloudRunServiceTemplateSpecVolumesCsiList(this, "csi", false);
        // empty_dir - computed: true, optional: false, required: false
        this._emptyDir = new DataGoogleCloudRunServiceTemplateSpecVolumesEmptyDirList(this, "empty_dir", false);
        // nfs - computed: true, optional: false, required: false
        this._nfs = new DataGoogleCloudRunServiceTemplateSpecVolumesNfsList(this, "nfs", false);
        // secret - computed: true, optional: false, required: false
        this._secret = new DataGoogleCloudRunServiceTemplateSpecVolumesSecretList(this, "secret", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get csi() {
        return this._csi;
    }
    get emptyDir() {
        return this._emptyDir;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get nfs() {
        return this._nfs;
    }
    get secret() {
        return this._secret;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference = DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference;
_38 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference[_38] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecVolumesList = DataGoogleCloudRunServiceTemplateSpecVolumesList;
_39 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecVolumesList[_39] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecVolumesList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateSpecToTerraform = dataGoogleCloudRunServiceTemplateSpecToTerraform;
function dataGoogleCloudRunServiceTemplateSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateSpecToHclTerraform = dataGoogleCloudRunServiceTemplateSpecToHclTerraform;
class DataGoogleCloudRunServiceTemplateSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // containers - computed: true, optional: false, required: false
        this._containers = new DataGoogleCloudRunServiceTemplateSpecContainersList(this, "containers", false);
        // volumes - computed: true, optional: false, required: false
        this._volumes = new DataGoogleCloudRunServiceTemplateSpecVolumesList(this, "volumes", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // container_concurrency - computed: true, optional: false, required: false
    get containerConcurrency() {
        return this.getNumberAttribute('container_concurrency');
    }
    get containers() {
        return this._containers;
    }
    // service_account_name - computed: true, optional: false, required: false
    get serviceAccountName() {
        return this.getStringAttribute('service_account_name');
    }
    // serving_state - computed: true, optional: false, required: false
    get servingState() {
        return this.getStringAttribute('serving_state');
    }
    // timeout_seconds - computed: true, optional: false, required: false
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    get volumes() {
        return this._volumes;
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecOutputReference = DataGoogleCloudRunServiceTemplateSpecOutputReference;
_40 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecOutputReference[_40] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateSpecList = DataGoogleCloudRunServiceTemplateSpecList;
_41 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateSpecList[_41] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateSpecList", version: "14.13.1" };
function dataGoogleCloudRunServiceTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTemplateToTerraform = dataGoogleCloudRunServiceTemplateToTerraform;
function dataGoogleCloudRunServiceTemplateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTemplateToHclTerraform = dataGoogleCloudRunServiceTemplateToHclTerraform;
class DataGoogleCloudRunServiceTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // metadata - computed: true, optional: false, required: false
        this._metadata = new DataGoogleCloudRunServiceTemplateMetadataList(this, "metadata", false);
        // spec - computed: true, optional: false, required: false
        this._spec = new DataGoogleCloudRunServiceTemplateSpecList(this, "spec", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get metadata() {
        return this._metadata;
    }
    get spec() {
        return this._spec;
    }
}
exports.DataGoogleCloudRunServiceTemplateOutputReference = DataGoogleCloudRunServiceTemplateOutputReference;
_42 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateOutputReference[_42] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTemplateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTemplateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTemplateList = DataGoogleCloudRunServiceTemplateList;
_43 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTemplateList[_43] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTemplateList", version: "14.13.1" };
function dataGoogleCloudRunServiceTrafficToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudRunServiceTrafficToTerraform = dataGoogleCloudRunServiceTrafficToTerraform;
function dataGoogleCloudRunServiceTrafficToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudRunServiceTrafficToHclTerraform = dataGoogleCloudRunServiceTrafficToHclTerraform;
class DataGoogleCloudRunServiceTrafficOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // latest_revision - computed: true, optional: false, required: false
    get latestRevision() {
        return this.getBooleanAttribute('latest_revision');
    }
    // percent - computed: true, optional: false, required: false
    get percent() {
        return this.getNumberAttribute('percent');
    }
    // revision_name - computed: true, optional: false, required: false
    get revisionName() {
        return this.getStringAttribute('revision_name');
    }
    // tag - computed: true, optional: false, required: false
    get tag() {
        return this.getStringAttribute('tag');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
}
exports.DataGoogleCloudRunServiceTrafficOutputReference = DataGoogleCloudRunServiceTrafficOutputReference;
_44 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTrafficOutputReference[_44] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTrafficOutputReference", version: "14.13.1" };
class DataGoogleCloudRunServiceTrafficList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudRunServiceTrafficOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudRunServiceTrafficList = DataGoogleCloudRunServiceTrafficList;
_45 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunServiceTrafficList[_45] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunServiceTrafficList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_run_service google_cloud_run_service}
*/
class DataGoogleCloudRunService extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleCloudRunService resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleCloudRunService to import
    * @param importFromId The id of the existing DataGoogleCloudRunService that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_run_service#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleCloudRunService to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_cloud_run_service", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_run_service google_cloud_run_service} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleCloudRunServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloud_run_service',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // metadata - computed: true, optional: false, required: false
        this._metadata = new DataGoogleCloudRunServiceMetadataList(this, "metadata", false);
        // status - computed: true, optional: false, required: false
        this._status = new DataGoogleCloudRunServiceStatusList(this, "status", false);
        // template - computed: true, optional: false, required: false
        this._template = new DataGoogleCloudRunServiceTemplateList(this, "template", false);
        // traffic - computed: true, optional: false, required: false
        this._traffic = new DataGoogleCloudRunServiceTrafficList(this, "traffic", false);
        this._id = config.id;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // autogenerate_revision_name - computed: true, optional: false, required: false
    get autogenerateRevisionName() {
        return this.getBooleanAttribute('autogenerate_revision_name');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get metadata() {
        return this._metadata;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get status() {
        return this._status;
    }
    get template() {
        return this._template;
    }
    get traffic() {
        return this._traffic;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleCloudRunService = DataGoogleCloudRunService;
_46 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudRunService[_46] = { fqn: "@cdktf/provider-google.dataGoogleCloudRunService.DataGoogleCloudRunService", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleCloudRunService.tfResourceType = "google_cloud_run_service";
