// @cdktf/provider-google/lib/cloud-run-v2-job/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9;
Object.defineProperty(exports, "__esModule", { value: true });
exports.cloudRunV2JobTemplateTemplateVolumesNfsToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesGcsOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesGcsToHclTerraform = exports.cloudRunV2JobTemplateTemplateVolumesGcsToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesEmptyDirOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesEmptyDirToHclTerraform = exports.cloudRunV2JobTemplateTemplateVolumesEmptyDirToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToHclTerraform = exports.cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform = exports.CloudRunV2JobTemplateTemplateContainersList = exports.CloudRunV2JobTemplateTemplateContainersOutputReference = exports.cloudRunV2JobTemplateTemplateContainersToHclTerraform = exports.cloudRunV2JobTemplateTemplateContainersToTerraform = exports.CloudRunV2JobTemplateTemplateContainersVolumeMountsList = exports.CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference = exports.cloudRunV2JobTemplateTemplateContainersVolumeMountsToHclTerraform = exports.cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform = exports.CloudRunV2JobTemplateTemplateContainersResourcesOutputReference = exports.cloudRunV2JobTemplateTemplateContainersResourcesToHclTerraform = exports.cloudRunV2JobTemplateTemplateContainersResourcesToTerraform = exports.CloudRunV2JobTemplateTemplateContainersPortsList = exports.CloudRunV2JobTemplateTemplateContainersPortsOutputReference = exports.cloudRunV2JobTemplateTemplateContainersPortsToHclTerraform = exports.cloudRunV2JobTemplateTemplateContainersPortsToTerraform = exports.CloudRunV2JobTemplateTemplateContainersEnvList = exports.CloudRunV2JobTemplateTemplateContainersEnvOutputReference = exports.cloudRunV2JobTemplateTemplateContainersEnvToHclTerraform = exports.cloudRunV2JobTemplateTemplateContainersEnvToTerraform = exports.CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference = exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceToHclTerraform = exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform = exports.CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference = exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToHclTerraform = exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform = exports.CloudRunV2JobBinaryAuthorizationOutputReference = exports.cloudRunV2JobBinaryAuthorizationToHclTerraform = exports.cloudRunV2JobBinaryAuthorizationToTerraform = exports.CloudRunV2JobTerminalConditionList = exports.CloudRunV2JobTerminalConditionOutputReference = exports.cloudRunV2JobTerminalConditionToHclTerraform = exports.cloudRunV2JobTerminalConditionToTerraform = exports.CloudRunV2JobLatestCreatedExecutionList = exports.CloudRunV2JobLatestCreatedExecutionOutputReference = exports.cloudRunV2JobLatestCreatedExecutionToHclTerraform = exports.cloudRunV2JobLatestCreatedExecutionToTerraform = exports.CloudRunV2JobConditionsList = exports.CloudRunV2JobConditionsOutputReference = exports.cloudRunV2JobConditionsToHclTerraform = exports.cloudRunV2JobConditionsToTerraform = void 0;
exports.CloudRunV2Job = exports.CloudRunV2JobTimeoutsOutputReference = exports.cloudRunV2JobTimeoutsToHclTerraform = exports.cloudRunV2JobTimeoutsToTerraform = exports.CloudRunV2JobTemplateOutputReference = exports.cloudRunV2JobTemplateToHclTerraform = exports.cloudRunV2JobTemplateToTerraform = exports.CloudRunV2JobTemplateTemplateOutputReference = exports.cloudRunV2JobTemplateTemplateToHclTerraform = exports.cloudRunV2JobTemplateTemplateToTerraform = exports.CloudRunV2JobTemplateTemplateVpcAccessOutputReference = exports.cloudRunV2JobTemplateTemplateVpcAccessToHclTerraform = exports.cloudRunV2JobTemplateTemplateVpcAccessToTerraform = exports.CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesList = exports.CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesOutputReference = exports.cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToHclTerraform = exports.cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesList = exports.CloudRunV2JobTemplateTemplateVolumesOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesToHclTerraform = exports.cloudRunV2JobTemplateTemplateVolumesToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesSecretOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesSecretToHclTerraform = exports.cloudRunV2JobTemplateTemplateVolumesSecretToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesSecretItemsList = exports.CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesSecretItemsToHclTerraform = exports.cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform = exports.CloudRunV2JobTemplateTemplateVolumesNfsOutputReference = exports.cloudRunV2JobTemplateTemplateVolumesNfsToHclTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudRunV2JobConditionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2JobConditionsToTerraform = cloudRunV2JobConditionsToTerraform;
function cloudRunV2JobConditionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.cloudRunV2JobConditionsToHclTerraform = cloudRunV2JobConditionsToHclTerraform;
class CloudRunV2JobConditionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // execution_reason - computed: true, optional: false, required: false
    get executionReason() {
        return this.getStringAttribute('execution_reason');
    }
    // last_transition_time - computed: true, optional: false, required: false
    get lastTransitionTime() {
        return this.getStringAttribute('last_transition_time');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // revision_reason - computed: true, optional: false, required: false
    get revisionReason() {
        return this.getStringAttribute('revision_reason');
    }
    // severity - computed: true, optional: false, required: false
    get severity() {
        return this.getStringAttribute('severity');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.CloudRunV2JobConditionsOutputReference = CloudRunV2JobConditionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudRunV2JobConditionsOutputReference[_a] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobConditionsOutputReference", version: "14.13.1" };
class CloudRunV2JobConditionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobConditionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobConditionsList = CloudRunV2JobConditionsList;
_b = JSII_RTTI_SYMBOL_1;
CloudRunV2JobConditionsList[_b] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobConditionsList", version: "14.13.1" };
function cloudRunV2JobLatestCreatedExecutionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2JobLatestCreatedExecutionToTerraform = cloudRunV2JobLatestCreatedExecutionToTerraform;
function cloudRunV2JobLatestCreatedExecutionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.cloudRunV2JobLatestCreatedExecutionToHclTerraform = cloudRunV2JobLatestCreatedExecutionToHclTerraform;
class CloudRunV2JobLatestCreatedExecutionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // completion_time - computed: true, optional: false, required: false
    get completionTime() {
        return this.getStringAttribute('completion_time');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.CloudRunV2JobLatestCreatedExecutionOutputReference = CloudRunV2JobLatestCreatedExecutionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CloudRunV2JobLatestCreatedExecutionOutputReference[_c] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobLatestCreatedExecutionOutputReference", version: "14.13.1" };
class CloudRunV2JobLatestCreatedExecutionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobLatestCreatedExecutionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobLatestCreatedExecutionList = CloudRunV2JobLatestCreatedExecutionList;
_d = JSII_RTTI_SYMBOL_1;
CloudRunV2JobLatestCreatedExecutionList[_d] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobLatestCreatedExecutionList", version: "14.13.1" };
function cloudRunV2JobTerminalConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.cloudRunV2JobTerminalConditionToTerraform = cloudRunV2JobTerminalConditionToTerraform;
function cloudRunV2JobTerminalConditionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.cloudRunV2JobTerminalConditionToHclTerraform = cloudRunV2JobTerminalConditionToHclTerraform;
class CloudRunV2JobTerminalConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // execution_reason - computed: true, optional: false, required: false
    get executionReason() {
        return this.getStringAttribute('execution_reason');
    }
    // last_transition_time - computed: true, optional: false, required: false
    get lastTransitionTime() {
        return this.getStringAttribute('last_transition_time');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
    // revision_reason - computed: true, optional: false, required: false
    get revisionReason() {
        return this.getStringAttribute('revision_reason');
    }
    // severity - computed: true, optional: false, required: false
    get severity() {
        return this.getStringAttribute('severity');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.CloudRunV2JobTerminalConditionOutputReference = CloudRunV2JobTerminalConditionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTerminalConditionOutputReference[_e] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTerminalConditionOutputReference", version: "14.13.1" };
class CloudRunV2JobTerminalConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTerminalConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTerminalConditionList = CloudRunV2JobTerminalConditionList;
_f = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTerminalConditionList[_f] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTerminalConditionList", version: "14.13.1" };
function cloudRunV2JobBinaryAuthorizationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        breakglass_justification: cdktf.stringToTerraform(struct.breakglassJustification),
        policy: cdktf.stringToTerraform(struct.policy),
        use_default: cdktf.booleanToTerraform(struct.useDefault),
    };
}
exports.cloudRunV2JobBinaryAuthorizationToTerraform = cloudRunV2JobBinaryAuthorizationToTerraform;
function cloudRunV2JobBinaryAuthorizationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        breakglass_justification: {
            value: cdktf.stringToHclTerraform(struct.breakglassJustification),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        policy: {
            value: cdktf.stringToHclTerraform(struct.policy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        use_default: {
            value: cdktf.booleanToHclTerraform(struct.useDefault),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobBinaryAuthorizationToHclTerraform = cloudRunV2JobBinaryAuthorizationToHclTerraform;
class CloudRunV2JobBinaryAuthorizationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._breakglassJustification !== undefined) {
            hasAnyValues = true;
            internalValueResult.breakglassJustification = this._breakglassJustification;
        }
        if (this._policy !== undefined) {
            hasAnyValues = true;
            internalValueResult.policy = this._policy;
        }
        if (this._useDefault !== undefined) {
            hasAnyValues = true;
            internalValueResult.useDefault = this._useDefault;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._breakglassJustification = undefined;
            this._policy = undefined;
            this._useDefault = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._breakglassJustification = value.breakglassJustification;
            this._policy = value.policy;
            this._useDefault = value.useDefault;
        }
    }
    get breakglassJustification() {
        return this.getStringAttribute('breakglass_justification');
    }
    set breakglassJustification(value) {
        this._breakglassJustification = value;
    }
    resetBreakglassJustification() {
        this._breakglassJustification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get breakglassJustificationInput() {
        return this._breakglassJustification;
    }
    get policy() {
        return this.getStringAttribute('policy');
    }
    set policy(value) {
        this._policy = value;
    }
    resetPolicy() {
        this._policy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get policyInput() {
        return this._policy;
    }
    get useDefault() {
        return this.getBooleanAttribute('use_default');
    }
    set useDefault(value) {
        this._useDefault = value;
    }
    resetUseDefault() {
        this._useDefault = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useDefaultInput() {
        return this._useDefault;
    }
}
exports.CloudRunV2JobBinaryAuthorizationOutputReference = CloudRunV2JobBinaryAuthorizationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
CloudRunV2JobBinaryAuthorizationOutputReference[_g] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobBinaryAuthorizationOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret: cdktf.stringToTerraform(struct.secret),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform = cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform;
function cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        secret: {
            value: cdktf.stringToHclTerraform(struct.secret),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        version: {
            value: cdktf.stringToHclTerraform(struct.version),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToHclTerraform = cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToHclTerraform;
class CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secret !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secret = undefined;
            this._version = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secret = value.secret;
            this._version = value.version;
        }
    }
    get secret() {
        return this.getStringAttribute('secret');
    }
    set secret(value) {
        this._secret = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference = CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference;
_h = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference[_h] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_key_ref: cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToTerraform(struct.secretKeyRef),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform = cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform;
function cloudRunV2JobTemplateTemplateContainersEnvValueSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        secret_key_ref: {
            value: cloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefToHclTerraform(struct.secretKeyRef),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateContainersEnvValueSourceToHclTerraform = cloudRunV2JobTemplateTemplateContainersEnvValueSourceToHclTerraform;
class CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // secret_key_ref - computed: false, optional: true, required: false
        this._secretKeyRef = new CloudRunV2JobTemplateTemplateContainersEnvValueSourceSecretKeyRefOutputReference(this, "secret_key_ref");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretKeyRef?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretKeyRef = this._secretKeyRef?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretKeyRef.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretKeyRef.internalValue = value.secretKeyRef;
        }
    }
    get secretKeyRef() {
        return this._secretKeyRef;
    }
    putSecretKeyRef(value) {
        this._secretKeyRef.internalValue = value;
    }
    resetSecretKeyRef() {
        this._secretKeyRef.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretKeyRefInput() {
        return this._secretKeyRef.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference = CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference;
_j = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference[_j] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateContainersEnvToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
        value_source: cloudRunV2JobTemplateTemplateContainersEnvValueSourceToTerraform(struct.valueSource),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersEnvToTerraform = cloudRunV2JobTemplateTemplateContainersEnvToTerraform;
function cloudRunV2JobTemplateTemplateContainersEnvToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value_source: {
            value: cloudRunV2JobTemplateTemplateContainersEnvValueSourceToHclTerraform(struct.valueSource),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateContainersEnvValueSourceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateContainersEnvToHclTerraform = cloudRunV2JobTemplateTemplateContainersEnvToHclTerraform;
class CloudRunV2JobTemplateTemplateContainersEnvOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // value_source - computed: false, optional: true, required: false
        this._valueSource = new CloudRunV2JobTemplateTemplateContainersEnvValueSourceOutputReference(this, "value_source");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        if (this._valueSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.valueSource = this._valueSource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
            this._valueSource.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
            this._valueSource.internalValue = value.valueSource;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
    get valueSource() {
        return this._valueSource;
    }
    putValueSource(value) {
        this._valueSource.internalValue = value;
    }
    resetValueSource() {
        this._valueSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueSourceInput() {
        return this._valueSource.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersEnvOutputReference = CloudRunV2JobTemplateTemplateContainersEnvOutputReference;
_k = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersEnvOutputReference[_k] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersEnvOutputReference", version: "14.13.1" };
class CloudRunV2JobTemplateTemplateContainersEnvList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersEnvOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersEnvList = CloudRunV2JobTemplateTemplateContainersEnvList;
_l = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersEnvList[_l] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersEnvList", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateContainersPortsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_port: cdktf.numberToTerraform(struct.containerPort),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersPortsToTerraform = cloudRunV2JobTemplateTemplateContainersPortsToTerraform;
function cloudRunV2JobTemplateTemplateContainersPortsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        container_port: {
            value: cdktf.numberToHclTerraform(struct.containerPort),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateContainersPortsToHclTerraform = cloudRunV2JobTemplateTemplateContainersPortsToHclTerraform;
class CloudRunV2JobTemplateTemplateContainersPortsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerPort = this._containerPort;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._containerPort = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._containerPort = value.containerPort;
            this._name = value.name;
        }
    }
    get containerPort() {
        return this.getNumberAttribute('container_port');
    }
    set containerPort(value) {
        this._containerPort = value;
    }
    resetContainerPort() {
        this._containerPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerPortInput() {
        return this._containerPort;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersPortsOutputReference = CloudRunV2JobTemplateTemplateContainersPortsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersPortsOutputReference[_m] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersPortsOutputReference", version: "14.13.1" };
class CloudRunV2JobTemplateTemplateContainersPortsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersPortsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersPortsList = CloudRunV2JobTemplateTemplateContainersPortsList;
_o = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersPortsList[_o] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersPortsList", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateContainersResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        limits: cdktf.hashMapper(cdktf.stringToTerraform)(struct.limits),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersResourcesToTerraform = cloudRunV2JobTemplateTemplateContainersResourcesToTerraform;
function cloudRunV2JobTemplateTemplateContainersResourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        limits: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.limits),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateContainersResourcesToHclTerraform = cloudRunV2JobTemplateTemplateContainersResourcesToHclTerraform;
class CloudRunV2JobTemplateTemplateContainersResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._limits !== undefined) {
            hasAnyValues = true;
            internalValueResult.limits = this._limits;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._limits = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._limits = value.limits;
        }
    }
    get limits() {
        return this.getStringMapAttribute('limits');
    }
    set limits(value) {
        this._limits = value;
    }
    resetLimits() {
        this._limits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get limitsInput() {
        return this._limits;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersResourcesOutputReference = CloudRunV2JobTemplateTemplateContainersResourcesOutputReference;
_p = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersResourcesOutputReference[_p] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersResourcesOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mount_path: cdktf.stringToTerraform(struct.mountPath),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform = cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform;
function cloudRunV2JobTemplateTemplateContainersVolumeMountsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        mount_path: {
            value: cdktf.stringToHclTerraform(struct.mountPath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateContainersVolumeMountsToHclTerraform = cloudRunV2JobTemplateTemplateContainersVolumeMountsToHclTerraform;
class CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mountPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.mountPath = this._mountPath;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mountPath = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mountPath = value.mountPath;
            this._name = value.name;
        }
    }
    get mountPath() {
        return this.getStringAttribute('mount_path');
    }
    set mountPath(value) {
        this._mountPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get mountPathInput() {
        return this._mountPath;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference = CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference[_q] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference", version: "14.13.1" };
class CloudRunV2JobTemplateTemplateContainersVolumeMountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersVolumeMountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersVolumeMountsList = CloudRunV2JobTemplateTemplateContainersVolumeMountsList;
_r = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersVolumeMountsList[_r] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersVolumeMountsList", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateContainersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        args: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.args),
        command: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.command),
        image: cdktf.stringToTerraform(struct.image),
        name: cdktf.stringToTerraform(struct.name),
        working_dir: cdktf.stringToTerraform(struct.workingDir),
        env: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersEnvToTerraform, true)(struct.env),
        ports: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersPortsToTerraform, true)(struct.ports),
        resources: cloudRunV2JobTemplateTemplateContainersResourcesToTerraform(struct.resources),
        volume_mounts: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersVolumeMountsToTerraform, true)(struct.volumeMounts),
    };
}
exports.cloudRunV2JobTemplateTemplateContainersToTerraform = cloudRunV2JobTemplateTemplateContainersToTerraform;
function cloudRunV2JobTemplateTemplateContainersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        args: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.args),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        command: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.command),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        image: {
            value: cdktf.stringToHclTerraform(struct.image),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        working_dir: {
            value: cdktf.stringToHclTerraform(struct.workingDir),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        env: {
            value: cdktf.listMapperHcl(cloudRunV2JobTemplateTemplateContainersEnvToHclTerraform, true)(struct.env),
            isBlock: true,
            type: "set",
            storageClassType: "CloudRunV2JobTemplateTemplateContainersEnvList",
        },
        ports: {
            value: cdktf.listMapperHcl(cloudRunV2JobTemplateTemplateContainersPortsToHclTerraform, true)(struct.ports),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateContainersPortsList",
        },
        resources: {
            value: cloudRunV2JobTemplateTemplateContainersResourcesToHclTerraform(struct.resources),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateContainersResourcesList",
        },
        volume_mounts: {
            value: cdktf.listMapperHcl(cloudRunV2JobTemplateTemplateContainersVolumeMountsToHclTerraform, true)(struct.volumeMounts),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateContainersVolumeMountsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateContainersToHclTerraform = cloudRunV2JobTemplateTemplateContainersToHclTerraform;
class CloudRunV2JobTemplateTemplateContainersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // env - computed: false, optional: true, required: false
        this._env = new CloudRunV2JobTemplateTemplateContainersEnvList(this, "env", true);
        // ports - computed: false, optional: true, required: false
        this._ports = new CloudRunV2JobTemplateTemplateContainersPortsList(this, "ports", false);
        // resources - computed: false, optional: true, required: false
        this._resources = new CloudRunV2JobTemplateTemplateContainersResourcesOutputReference(this, "resources");
        // volume_mounts - computed: false, optional: true, required: false
        this._volumeMounts = new CloudRunV2JobTemplateTemplateContainersVolumeMountsList(this, "volume_mounts", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._args !== undefined) {
            hasAnyValues = true;
            internalValueResult.args = this._args;
        }
        if (this._command !== undefined) {
            hasAnyValues = true;
            internalValueResult.command = this._command;
        }
        if (this._image !== undefined) {
            hasAnyValues = true;
            internalValueResult.image = this._image;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._workingDir !== undefined) {
            hasAnyValues = true;
            internalValueResult.workingDir = this._workingDir;
        }
        if (this._env?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.env = this._env?.internalValue;
        }
        if (this._ports?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ports = this._ports?.internalValue;
        }
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        if (this._volumeMounts?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeMounts = this._volumeMounts?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._args = undefined;
            this._command = undefined;
            this._image = undefined;
            this._name = undefined;
            this._workingDir = undefined;
            this._env.internalValue = undefined;
            this._ports.internalValue = undefined;
            this._resources.internalValue = undefined;
            this._volumeMounts.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._args = value.args;
            this._command = value.command;
            this._image = value.image;
            this._name = value.name;
            this._workingDir = value.workingDir;
            this._env.internalValue = value.env;
            this._ports.internalValue = value.ports;
            this._resources.internalValue = value.resources;
            this._volumeMounts.internalValue = value.volumeMounts;
        }
    }
    get args() {
        return this.getListAttribute('args');
    }
    set args(value) {
        this._args = value;
    }
    resetArgs() {
        this._args = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argsInput() {
        return this._args;
    }
    get command() {
        return this.getListAttribute('command');
    }
    set command(value) {
        this._command = value;
    }
    resetCommand() {
        this._command = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commandInput() {
        return this._command;
    }
    get image() {
        return this.getStringAttribute('image');
    }
    set image(value) {
        this._image = value;
    }
    // Temporarily expose input value. Use with caution.
    get imageInput() {
        return this._image;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get workingDir() {
        return this.getStringAttribute('working_dir');
    }
    set workingDir(value) {
        this._workingDir = value;
    }
    resetWorkingDir() {
        this._workingDir = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workingDirInput() {
        return this._workingDir;
    }
    get env() {
        return this._env;
    }
    putEnv(value) {
        this._env.internalValue = value;
    }
    resetEnv() {
        this._env.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get envInput() {
        return this._env.internalValue;
    }
    get ports() {
        return this._ports;
    }
    putPorts(value) {
        this._ports.internalValue = value;
    }
    resetPorts() {
        this._ports.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portsInput() {
        return this._ports.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    resetResources() {
        this._resources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
    get volumeMounts() {
        return this._volumeMounts;
    }
    putVolumeMounts(value) {
        this._volumeMounts.internalValue = value;
    }
    resetVolumeMounts() {
        this._volumeMounts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeMountsInput() {
        return this._volumeMounts.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateContainersOutputReference = CloudRunV2JobTemplateTemplateContainersOutputReference;
_s = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersOutputReference[_s] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersOutputReference", version: "14.13.1" };
class CloudRunV2JobTemplateTemplateContainersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateContainersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateContainersList = CloudRunV2JobTemplateTemplateContainersList;
_t = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateContainersList[_t] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateContainersList", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instances: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.instances),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform = cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform;
function cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        instances: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.instances),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToHclTerraform = cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToHclTerraform;
class CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instances !== undefined) {
            hasAnyValues = true;
            internalValueResult.instances = this._instances;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._instances = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._instances = value.instances;
        }
    }
    get instances() {
        return this.getListAttribute('instances');
    }
    set instances(value) {
        this._instances = value;
    }
    resetInstances() {
        this._instances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instancesInput() {
        return this._instances;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference = CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference;
_u = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference[_u] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVolumesEmptyDirToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        medium: cdktf.stringToTerraform(struct.medium),
        size_limit: cdktf.stringToTerraform(struct.sizeLimit),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesEmptyDirToTerraform = cloudRunV2JobTemplateTemplateVolumesEmptyDirToTerraform;
function cloudRunV2JobTemplateTemplateVolumesEmptyDirToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        medium: {
            value: cdktf.stringToHclTerraform(struct.medium),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        size_limit: {
            value: cdktf.stringToHclTerraform(struct.sizeLimit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVolumesEmptyDirToHclTerraform = cloudRunV2JobTemplateTemplateVolumesEmptyDirToHclTerraform;
class CloudRunV2JobTemplateTemplateVolumesEmptyDirOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._medium !== undefined) {
            hasAnyValues = true;
            internalValueResult.medium = this._medium;
        }
        if (this._sizeLimit !== undefined) {
            hasAnyValues = true;
            internalValueResult.sizeLimit = this._sizeLimit;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._medium = undefined;
            this._sizeLimit = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._medium = value.medium;
            this._sizeLimit = value.sizeLimit;
        }
    }
    get medium() {
        return this.getStringAttribute('medium');
    }
    set medium(value) {
        this._medium = value;
    }
    resetMedium() {
        this._medium = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mediumInput() {
        return this._medium;
    }
    get sizeLimit() {
        return this.getStringAttribute('size_limit');
    }
    set sizeLimit(value) {
        this._sizeLimit = value;
    }
    resetSizeLimit() {
        this._sizeLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sizeLimitInput() {
        return this._sizeLimit;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesEmptyDirOutputReference = CloudRunV2JobTemplateTemplateVolumesEmptyDirOutputReference;
_v = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesEmptyDirOutputReference[_v] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesEmptyDirOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVolumesGcsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket: cdktf.stringToTerraform(struct.bucket),
        read_only: cdktf.booleanToTerraform(struct.readOnly),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesGcsToTerraform = cloudRunV2JobTemplateTemplateVolumesGcsToTerraform;
function cloudRunV2JobTemplateTemplateVolumesGcsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket: {
            value: cdktf.stringToHclTerraform(struct.bucket),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        read_only: {
            value: cdktf.booleanToHclTerraform(struct.readOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVolumesGcsToHclTerraform = cloudRunV2JobTemplateTemplateVolumesGcsToHclTerraform;
class CloudRunV2JobTemplateTemplateVolumesGcsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucket !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucket = this._bucket;
        }
        if (this._readOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.readOnly = this._readOnly;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucket = undefined;
            this._readOnly = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucket = value.bucket;
            this._readOnly = value.readOnly;
        }
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    get readOnly() {
        return this.getBooleanAttribute('read_only');
    }
    set readOnly(value) {
        this._readOnly = value;
    }
    resetReadOnly() {
        this._readOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readOnlyInput() {
        return this._readOnly;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesGcsOutputReference = CloudRunV2JobTemplateTemplateVolumesGcsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesGcsOutputReference[_w] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesGcsOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVolumesNfsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        read_only: cdktf.booleanToTerraform(struct.readOnly),
        server: cdktf.stringToTerraform(struct.server),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesNfsToTerraform = cloudRunV2JobTemplateTemplateVolumesNfsToTerraform;
function cloudRunV2JobTemplateTemplateVolumesNfsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        read_only: {
            value: cdktf.booleanToHclTerraform(struct.readOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        server: {
            value: cdktf.stringToHclTerraform(struct.server),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVolumesNfsToHclTerraform = cloudRunV2JobTemplateTemplateVolumesNfsToHclTerraform;
class CloudRunV2JobTemplateTemplateVolumesNfsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._readOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.readOnly = this._readOnly;
        }
        if (this._server !== undefined) {
            hasAnyValues = true;
            internalValueResult.server = this._server;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._readOnly = undefined;
            this._server = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._readOnly = value.readOnly;
            this._server = value.server;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get readOnly() {
        return this.getBooleanAttribute('read_only');
    }
    set readOnly(value) {
        this._readOnly = value;
    }
    resetReadOnly() {
        this._readOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readOnlyInput() {
        return this._readOnly;
    }
    get server() {
        return this.getStringAttribute('server');
    }
    set server(value) {
        this._server = value;
    }
    // Temporarily expose input value. Use with caution.
    get serverInput() {
        return this._server;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesNfsOutputReference = CloudRunV2JobTemplateTemplateVolumesNfsOutputReference;
_x = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesNfsOutputReference[_x] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesNfsOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.numberToTerraform(struct.mode),
        path: cdktf.stringToTerraform(struct.path),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform = cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform;
function cloudRunV2JobTemplateTemplateVolumesSecretItemsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        mode: {
            value: cdktf.numberToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        version: {
            value: cdktf.stringToHclTerraform(struct.version),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVolumesSecretItemsToHclTerraform = cloudRunV2JobTemplateTemplateVolumesSecretItemsToHclTerraform;
class CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._mode = undefined;
            this._path = undefined;
            this._version = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._mode = value.mode;
            this._path = value.path;
            this._version = value.version;
        }
    }
    get mode() {
        return this.getNumberAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference = CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference;
_y = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference[_y] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference", version: "14.13.1" };
class CloudRunV2JobTemplateTemplateVolumesSecretItemsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateVolumesSecretItemsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesSecretItemsList = CloudRunV2JobTemplateTemplateVolumesSecretItemsList;
_z = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesSecretItemsList[_z] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesSecretItemsList", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVolumesSecretToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_mode: cdktf.numberToTerraform(struct.defaultMode),
        secret: cdktf.stringToTerraform(struct.secret),
        items: cdktf.listMapper(cloudRunV2JobTemplateTemplateVolumesSecretItemsToTerraform, true)(struct.items),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesSecretToTerraform = cloudRunV2JobTemplateTemplateVolumesSecretToTerraform;
function cloudRunV2JobTemplateTemplateVolumesSecretToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        default_mode: {
            value: cdktf.numberToHclTerraform(struct.defaultMode),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        secret: {
            value: cdktf.stringToHclTerraform(struct.secret),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        items: {
            value: cdktf.listMapperHcl(cloudRunV2JobTemplateTemplateVolumesSecretItemsToHclTerraform, true)(struct.items),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVolumesSecretItemsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVolumesSecretToHclTerraform = cloudRunV2JobTemplateTemplateVolumesSecretToHclTerraform;
class CloudRunV2JobTemplateTemplateVolumesSecretOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // items - computed: false, optional: true, required: false
        this._items = new CloudRunV2JobTemplateTemplateVolumesSecretItemsList(this, "items", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultMode = this._defaultMode;
        }
        if (this._secret !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret;
        }
        if (this._items?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.items = this._items?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultMode = undefined;
            this._secret = undefined;
            this._items.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultMode = value.defaultMode;
            this._secret = value.secret;
            this._items.internalValue = value.items;
        }
    }
    get defaultMode() {
        return this.getNumberAttribute('default_mode');
    }
    set defaultMode(value) {
        this._defaultMode = value;
    }
    resetDefaultMode() {
        this._defaultMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultModeInput() {
        return this._defaultMode;
    }
    get secret() {
        return this.getStringAttribute('secret');
    }
    set secret(value) {
        this._secret = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret;
    }
    get items() {
        return this._items;
    }
    putItems(value) {
        this._items.internalValue = value;
    }
    resetItems() {
        this._items.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get itemsInput() {
        return this._items.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesSecretOutputReference = CloudRunV2JobTemplateTemplateVolumesSecretOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesSecretOutputReference[_0] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesSecretOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        cloud_sql_instance: cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToTerraform(struct.cloudSqlInstance),
        empty_dir: cloudRunV2JobTemplateTemplateVolumesEmptyDirToTerraform(struct.emptyDir),
        gcs: cloudRunV2JobTemplateTemplateVolumesGcsToTerraform(struct.gcs),
        nfs: cloudRunV2JobTemplateTemplateVolumesNfsToTerraform(struct.nfs),
        secret: cloudRunV2JobTemplateTemplateVolumesSecretToTerraform(struct.secret),
    };
}
exports.cloudRunV2JobTemplateTemplateVolumesToTerraform = cloudRunV2JobTemplateTemplateVolumesToTerraform;
function cloudRunV2JobTemplateTemplateVolumesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        cloud_sql_instance: {
            value: cloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceToHclTerraform(struct.cloudSqlInstance),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceList",
        },
        empty_dir: {
            value: cloudRunV2JobTemplateTemplateVolumesEmptyDirToHclTerraform(struct.emptyDir),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVolumesEmptyDirList",
        },
        gcs: {
            value: cloudRunV2JobTemplateTemplateVolumesGcsToHclTerraform(struct.gcs),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVolumesGcsList",
        },
        nfs: {
            value: cloudRunV2JobTemplateTemplateVolumesNfsToHclTerraform(struct.nfs),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVolumesNfsList",
        },
        secret: {
            value: cloudRunV2JobTemplateTemplateVolumesSecretToHclTerraform(struct.secret),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVolumesSecretList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVolumesToHclTerraform = cloudRunV2JobTemplateTemplateVolumesToHclTerraform;
class CloudRunV2JobTemplateTemplateVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cloud_sql_instance - computed: false, optional: true, required: false
        this._cloudSqlInstance = new CloudRunV2JobTemplateTemplateVolumesCloudSqlInstanceOutputReference(this, "cloud_sql_instance");
        // empty_dir - computed: false, optional: true, required: false
        this._emptyDir = new CloudRunV2JobTemplateTemplateVolumesEmptyDirOutputReference(this, "empty_dir");
        // gcs - computed: false, optional: true, required: false
        this._gcs = new CloudRunV2JobTemplateTemplateVolumesGcsOutputReference(this, "gcs");
        // nfs - computed: false, optional: true, required: false
        this._nfs = new CloudRunV2JobTemplateTemplateVolumesNfsOutputReference(this, "nfs");
        // secret - computed: false, optional: true, required: false
        this._secret = new CloudRunV2JobTemplateTemplateVolumesSecretOutputReference(this, "secret");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._cloudSqlInstance?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudSqlInstance = this._cloudSqlInstance?.internalValue;
        }
        if (this._emptyDir?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.emptyDir = this._emptyDir?.internalValue;
        }
        if (this._gcs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.gcs = this._gcs?.internalValue;
        }
        if (this._nfs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.nfs = this._nfs?.internalValue;
        }
        if (this._secret?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.secret = this._secret?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._cloudSqlInstance.internalValue = undefined;
            this._emptyDir.internalValue = undefined;
            this._gcs.internalValue = undefined;
            this._nfs.internalValue = undefined;
            this._secret.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._cloudSqlInstance.internalValue = value.cloudSqlInstance;
            this._emptyDir.internalValue = value.emptyDir;
            this._gcs.internalValue = value.gcs;
            this._nfs.internalValue = value.nfs;
            this._secret.internalValue = value.secret;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get cloudSqlInstance() {
        return this._cloudSqlInstance;
    }
    putCloudSqlInstance(value) {
        this._cloudSqlInstance.internalValue = value;
    }
    resetCloudSqlInstance() {
        this._cloudSqlInstance.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudSqlInstanceInput() {
        return this._cloudSqlInstance.internalValue;
    }
    get emptyDir() {
        return this._emptyDir;
    }
    putEmptyDir(value) {
        this._emptyDir.internalValue = value;
    }
    resetEmptyDir() {
        this._emptyDir.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emptyDirInput() {
        return this._emptyDir.internalValue;
    }
    get gcs() {
        return this._gcs;
    }
    putGcs(value) {
        this._gcs.internalValue = value;
    }
    resetGcs() {
        this._gcs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gcsInput() {
        return this._gcs.internalValue;
    }
    get nfs() {
        return this._nfs;
    }
    putNfs(value) {
        this._nfs.internalValue = value;
    }
    resetNfs() {
        this._nfs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nfsInput() {
        return this._nfs.internalValue;
    }
    get secret() {
        return this._secret;
    }
    putSecret(value) {
        this._secret.internalValue = value;
    }
    resetSecret() {
        this._secret.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secretInput() {
        return this._secret.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesOutputReference = CloudRunV2JobTemplateTemplateVolumesOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesOutputReference[_1] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesOutputReference", version: "14.13.1" };
class CloudRunV2JobTemplateTemplateVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateVolumesList = CloudRunV2JobTemplateTemplateVolumesList;
_2 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVolumesList[_2] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVolumesList", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        network: cdktf.stringToTerraform(struct.network),
        subnetwork: cdktf.stringToTerraform(struct.subnetwork),
        tags: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tags),
    };
}
exports.cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToTerraform = cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToTerraform;
function cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        network: {
            value: cdktf.stringToHclTerraform(struct.network),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subnetwork: {
            value: cdktf.stringToHclTerraform(struct.subnetwork),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tags: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.tags),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToHclTerraform = cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToHclTerraform;
class CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._network !== undefined) {
            hasAnyValues = true;
            internalValueResult.network = this._network;
        }
        if (this._subnetwork !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetwork = this._subnetwork;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._network = undefined;
            this._subnetwork = undefined;
            this._tags = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._network = value.network;
            this._subnetwork = value.subnetwork;
            this._tags = value.tags;
        }
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    resetNetwork() {
        this._network = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get subnetwork() {
        return this.getStringAttribute('subnetwork');
    }
    set subnetwork(value) {
        this._subnetwork = value;
    }
    resetSubnetwork() {
        this._subnetwork = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetworkInput() {
        return this._subnetwork;
    }
    get tags() {
        return this.getListAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
}
exports.CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesOutputReference = CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesOutputReference;
_3 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesOutputReference[_3] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesOutputReference", version: "14.13.1" };
class CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesList = CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesList;
_4 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesList[_4] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesList", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateVpcAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connector: cdktf.stringToTerraform(struct.connector),
        egress: cdktf.stringToTerraform(struct.egress),
        network_interfaces: cdktf.listMapper(cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToTerraform, true)(struct.networkInterfaces),
    };
}
exports.cloudRunV2JobTemplateTemplateVpcAccessToTerraform = cloudRunV2JobTemplateTemplateVpcAccessToTerraform;
function cloudRunV2JobTemplateTemplateVpcAccessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        connector: {
            value: cdktf.stringToHclTerraform(struct.connector),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        egress: {
            value: cdktf.stringToHclTerraform(struct.egress),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        network_interfaces: {
            value: cdktf.listMapperHcl(cloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesToHclTerraform, true)(struct.networkInterfaces),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateVpcAccessToHclTerraform = cloudRunV2JobTemplateTemplateVpcAccessToHclTerraform;
class CloudRunV2JobTemplateTemplateVpcAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // network_interfaces - computed: false, optional: true, required: false
        this._networkInterfaces = new CloudRunV2JobTemplateTemplateVpcAccessNetworkInterfacesList(this, "network_interfaces", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connector !== undefined) {
            hasAnyValues = true;
            internalValueResult.connector = this._connector;
        }
        if (this._egress !== undefined) {
            hasAnyValues = true;
            internalValueResult.egress = this._egress;
        }
        if (this._networkInterfaces?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.networkInterfaces = this._networkInterfaces?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._connector = undefined;
            this._egress = undefined;
            this._networkInterfaces.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._connector = value.connector;
            this._egress = value.egress;
            this._networkInterfaces.internalValue = value.networkInterfaces;
        }
    }
    get connector() {
        return this.getStringAttribute('connector');
    }
    set connector(value) {
        this._connector = value;
    }
    resetConnector() {
        this._connector = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectorInput() {
        return this._connector;
    }
    get egress() {
        return this.getStringAttribute('egress');
    }
    set egress(value) {
        this._egress = value;
    }
    resetEgress() {
        this._egress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get egressInput() {
        return this._egress;
    }
    get networkInterfaces() {
        return this._networkInterfaces;
    }
    putNetworkInterfaces(value) {
        this._networkInterfaces.internalValue = value;
    }
    resetNetworkInterfaces() {
        this._networkInterfaces.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInterfacesInput() {
        return this._networkInterfaces.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateVpcAccessOutputReference = CloudRunV2JobTemplateTemplateVpcAccessOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateVpcAccessOutputReference[_5] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateVpcAccessOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encryption_key: cdktf.stringToTerraform(struct.encryptionKey),
        execution_environment: cdktf.stringToTerraform(struct.executionEnvironment),
        max_retries: cdktf.numberToTerraform(struct.maxRetries),
        service_account: cdktf.stringToTerraform(struct.serviceAccount),
        timeout: cdktf.stringToTerraform(struct.timeout),
        containers: cdktf.listMapper(cloudRunV2JobTemplateTemplateContainersToTerraform, true)(struct.containers),
        volumes: cdktf.listMapper(cloudRunV2JobTemplateTemplateVolumesToTerraform, true)(struct.volumes),
        vpc_access: cloudRunV2JobTemplateTemplateVpcAccessToTerraform(struct.vpcAccess),
    };
}
exports.cloudRunV2JobTemplateTemplateToTerraform = cloudRunV2JobTemplateTemplateToTerraform;
function cloudRunV2JobTemplateTemplateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        encryption_key: {
            value: cdktf.stringToHclTerraform(struct.encryptionKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        execution_environment: {
            value: cdktf.stringToHclTerraform(struct.executionEnvironment),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        max_retries: {
            value: cdktf.numberToHclTerraform(struct.maxRetries),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        service_account: {
            value: cdktf.stringToHclTerraform(struct.serviceAccount),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        timeout: {
            value: cdktf.stringToHclTerraform(struct.timeout),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        containers: {
            value: cdktf.listMapperHcl(cloudRunV2JobTemplateTemplateContainersToHclTerraform, true)(struct.containers),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateContainersList",
        },
        volumes: {
            value: cdktf.listMapperHcl(cloudRunV2JobTemplateTemplateVolumesToHclTerraform, true)(struct.volumes),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVolumesList",
        },
        vpc_access: {
            value: cloudRunV2JobTemplateTemplateVpcAccessToHclTerraform(struct.vpcAccess),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateVpcAccessList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateTemplateToHclTerraform = cloudRunV2JobTemplateTemplateToHclTerraform;
class CloudRunV2JobTemplateTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // containers - computed: false, optional: true, required: false
        this._containers = new CloudRunV2JobTemplateTemplateContainersList(this, "containers", false);
        // volumes - computed: false, optional: true, required: false
        this._volumes = new CloudRunV2JobTemplateTemplateVolumesList(this, "volumes", false);
        // vpc_access - computed: false, optional: true, required: false
        this._vpcAccess = new CloudRunV2JobTemplateTemplateVpcAccessOutputReference(this, "vpc_access");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptionKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionKey = this._encryptionKey;
        }
        if (this._executionEnvironment !== undefined) {
            hasAnyValues = true;
            internalValueResult.executionEnvironment = this._executionEnvironment;
        }
        if (this._maxRetries !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxRetries = this._maxRetries;
        }
        if (this._serviceAccount !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccount = this._serviceAccount;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._containers?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.containers = this._containers?.internalValue;
        }
        if (this._volumes?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumes = this._volumes?.internalValue;
        }
        if (this._vpcAccess?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.vpcAccess = this._vpcAccess?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptionKey = undefined;
            this._executionEnvironment = undefined;
            this._maxRetries = undefined;
            this._serviceAccount = undefined;
            this._timeout = undefined;
            this._containers.internalValue = undefined;
            this._volumes.internalValue = undefined;
            this._vpcAccess.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptionKey = value.encryptionKey;
            this._executionEnvironment = value.executionEnvironment;
            this._maxRetries = value.maxRetries;
            this._serviceAccount = value.serviceAccount;
            this._timeout = value.timeout;
            this._containers.internalValue = value.containers;
            this._volumes.internalValue = value.volumes;
            this._vpcAccess.internalValue = value.vpcAccess;
        }
    }
    get encryptionKey() {
        return this.getStringAttribute('encryption_key');
    }
    set encryptionKey(value) {
        this._encryptionKey = value;
    }
    resetEncryptionKey() {
        this._encryptionKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionKeyInput() {
        return this._encryptionKey;
    }
    get executionEnvironment() {
        return this.getStringAttribute('execution_environment');
    }
    set executionEnvironment(value) {
        this._executionEnvironment = value;
    }
    resetExecutionEnvironment() {
        this._executionEnvironment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get executionEnvironmentInput() {
        return this._executionEnvironment;
    }
    get maxRetries() {
        return this.getNumberAttribute('max_retries');
    }
    set maxRetries(value) {
        this._maxRetries = value;
    }
    resetMaxRetries() {
        this._maxRetries = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxRetriesInput() {
        return this._maxRetries;
    }
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    set serviceAccount(value) {
        this._serviceAccount = value;
    }
    resetServiceAccount() {
        this._serviceAccount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountInput() {
        return this._serviceAccount;
    }
    get timeout() {
        return this.getStringAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get containers() {
        return this._containers;
    }
    putContainers(value) {
        this._containers.internalValue = value;
    }
    resetContainers() {
        this._containers.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containersInput() {
        return this._containers.internalValue;
    }
    get volumes() {
        return this._volumes;
    }
    putVolumes(value) {
        this._volumes.internalValue = value;
    }
    resetVolumes() {
        this._volumes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumesInput() {
        return this._volumes.internalValue;
    }
    get vpcAccess() {
        return this._vpcAccess;
    }
    putVpcAccess(value) {
        this._vpcAccess.internalValue = value;
    }
    resetVpcAccess() {
        this._vpcAccess.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vpcAccessInput() {
        return this._vpcAccess.internalValue;
    }
}
exports.CloudRunV2JobTemplateTemplateOutputReference = CloudRunV2JobTemplateTemplateOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateTemplateOutputReference[_6] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateTemplateOutputReference", version: "14.13.1" };
function cloudRunV2JobTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        annotations: cdktf.hashMapper(cdktf.stringToTerraform)(struct.annotations),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        parallelism: cdktf.numberToTerraform(struct.parallelism),
        task_count: cdktf.numberToTerraform(struct.taskCount),
        template: cloudRunV2JobTemplateTemplateToTerraform(struct.template),
    };
}
exports.cloudRunV2JobTemplateToTerraform = cloudRunV2JobTemplateToTerraform;
function cloudRunV2JobTemplateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        annotations: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.annotations),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        labels: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.labels),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        parallelism: {
            value: cdktf.numberToHclTerraform(struct.parallelism),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        task_count: {
            value: cdktf.numberToHclTerraform(struct.taskCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        template: {
            value: cloudRunV2JobTemplateTemplateToHclTerraform(struct.template),
            isBlock: true,
            type: "list",
            storageClassType: "CloudRunV2JobTemplateTemplateList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTemplateToHclTerraform = cloudRunV2JobTemplateToHclTerraform;
class CloudRunV2JobTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // template - computed: false, optional: false, required: true
        this._template = new CloudRunV2JobTemplateTemplateOutputReference(this, "template");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._annotations !== undefined) {
            hasAnyValues = true;
            internalValueResult.annotations = this._annotations;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._parallelism !== undefined) {
            hasAnyValues = true;
            internalValueResult.parallelism = this._parallelism;
        }
        if (this._taskCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.taskCount = this._taskCount;
        }
        if (this._template?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.template = this._template?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._annotations = undefined;
            this._labels = undefined;
            this._parallelism = undefined;
            this._taskCount = undefined;
            this._template.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._annotations = value.annotations;
            this._labels = value.labels;
            this._parallelism = value.parallelism;
            this._taskCount = value.taskCount;
            this._template.internalValue = value.template;
        }
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get parallelism() {
        return this.getNumberAttribute('parallelism');
    }
    set parallelism(value) {
        this._parallelism = value;
    }
    resetParallelism() {
        this._parallelism = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parallelismInput() {
        return this._parallelism;
    }
    get taskCount() {
        return this.getNumberAttribute('task_count');
    }
    set taskCount(value) {
        this._taskCount = value;
    }
    resetTaskCount() {
        this._taskCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taskCountInput() {
        return this._taskCount;
    }
    get template() {
        return this._template;
    }
    putTemplate(value) {
        this._template.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get templateInput() {
        return this._template.internalValue;
    }
}
exports.CloudRunV2JobTemplateOutputReference = CloudRunV2JobTemplateOutputReference;
_7 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTemplateOutputReference[_7] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTemplateOutputReference", version: "14.13.1" };
function cloudRunV2JobTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudRunV2JobTimeoutsToTerraform = cloudRunV2JobTimeoutsToTerraform;
function cloudRunV2JobTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudRunV2JobTimeoutsToHclTerraform = cloudRunV2JobTimeoutsToHclTerraform;
class CloudRunV2JobTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudRunV2JobTimeoutsOutputReference = CloudRunV2JobTimeoutsOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
CloudRunV2JobTimeoutsOutputReference[_8] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2JobTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/cloud_run_v2_job google_cloud_run_v2_job}
*/
class CloudRunV2Job extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CloudRunV2Job resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CloudRunV2Job to import
    * @param importFromId The id of the existing CloudRunV2Job that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/cloud_run_v2_job#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CloudRunV2Job to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_cloud_run_v2_job", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/cloud_run_v2_job google_cloud_run_v2_job} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudRunV2JobConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloud_run_v2_job',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // conditions - computed: true, optional: false, required: false
        this._conditions = new CloudRunV2JobConditionsList(this, "conditions", false);
        // effective_annotations - computed: true, optional: false, required: false
        this._effectiveAnnotations = new cdktf.StringMap(this, "effective_annotations");
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // latest_created_execution - computed: true, optional: false, required: false
        this._latestCreatedExecution = new CloudRunV2JobLatestCreatedExecutionList(this, "latest_created_execution", false);
        // terminal_condition - computed: true, optional: false, required: false
        this._terminalCondition = new CloudRunV2JobTerminalConditionList(this, "terminal_condition", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // binary_authorization - computed: false, optional: true, required: false
        this._binaryAuthorization = new CloudRunV2JobBinaryAuthorizationOutputReference(this, "binary_authorization");
        // template - computed: false, optional: false, required: true
        this._template = new CloudRunV2JobTemplateOutputReference(this, "template");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudRunV2JobTimeoutsOutputReference(this, "timeouts");
        this._annotations = config.annotations;
        this._client = config.client;
        this._clientVersion = config.clientVersion;
        this._deletionProtection = config.deletionProtection;
        this._id = config.id;
        this._labels = config.labels;
        this._launchStage = config.launchStage;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._binaryAuthorization.internalValue = config.binaryAuthorization;
        this._template.internalValue = config.template;
        this._timeouts.internalValue = config.timeouts;
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    get client() {
        return this.getStringAttribute('client');
    }
    set client(value) {
        this._client = value;
    }
    resetClient() {
        this._client = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientInput() {
        return this._client;
    }
    get clientVersion() {
        return this.getStringAttribute('client_version');
    }
    set clientVersion(value) {
        this._clientVersion = value;
    }
    resetClientVersion() {
        this._clientVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientVersionInput() {
        return this._clientVersion;
    }
    get conditions() {
        return this._conditions;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // creator - computed: true, optional: false, required: false
    get creator() {
        return this.getStringAttribute('creator');
    }
    // delete_time - computed: true, optional: false, required: false
    get deleteTime() {
        return this.getStringAttribute('delete_time');
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get effectiveAnnotations() {
        return this._effectiveAnnotations;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    // execution_count - computed: true, optional: false, required: false
    get executionCount() {
        return this.getNumberAttribute('execution_count');
    }
    // expire_time - computed: true, optional: false, required: false
    get expireTime() {
        return this.getStringAttribute('expire_time');
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getStringAttribute('generation');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    // last_modifier - computed: true, optional: false, required: false
    get lastModifier() {
        return this.getStringAttribute('last_modifier');
    }
    get latestCreatedExecution() {
        return this._latestCreatedExecution;
    }
    get launchStage() {
        return this.getStringAttribute('launch_stage');
    }
    set launchStage(value) {
        this._launchStage = value;
    }
    resetLaunchStage() {
        this._launchStage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchStageInput() {
        return this._launchStage;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // observed_generation - computed: true, optional: false, required: false
    get observedGeneration() {
        return this.getStringAttribute('observed_generation');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // reconciling - computed: true, optional: false, required: false
    get reconciling() {
        return this.getBooleanAttribute('reconciling');
    }
    get terminalCondition() {
        return this._terminalCondition;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get binaryAuthorization() {
        return this._binaryAuthorization;
    }
    putBinaryAuthorization(value) {
        this._binaryAuthorization.internalValue = value;
    }
    resetBinaryAuthorization() {
        this._binaryAuthorization.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get binaryAuthorizationInput() {
        return this._binaryAuthorization.internalValue;
    }
    get template() {
        return this._template;
    }
    putTemplate(value) {
        this._template.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get templateInput() {
        return this._template.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            annotations: cdktf.hashMapper(cdktf.stringToTerraform)(this._annotations),
            client: cdktf.stringToTerraform(this._client),
            client_version: cdktf.stringToTerraform(this._clientVersion),
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            launch_stage: cdktf.stringToTerraform(this._launchStage),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            binary_authorization: cloudRunV2JobBinaryAuthorizationToTerraform(this._binaryAuthorization.internalValue),
            template: cloudRunV2JobTemplateToTerraform(this._template.internalValue),
            timeouts: cloudRunV2JobTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            annotations: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._annotations),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            client: {
                value: cdktf.stringToHclTerraform(this._client),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            client_version: {
                value: cdktf.stringToHclTerraform(this._clientVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            deletion_protection: {
                value: cdktf.booleanToHclTerraform(this._deletionProtection),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            launch_stage: {
                value: cdktf.stringToHclTerraform(this._launchStage),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            binary_authorization: {
                value: cloudRunV2JobBinaryAuthorizationToHclTerraform(this._binaryAuthorization.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudRunV2JobBinaryAuthorizationList",
            },
            template: {
                value: cloudRunV2JobTemplateToHclTerraform(this._template.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudRunV2JobTemplateList",
            },
            timeouts: {
                value: cloudRunV2JobTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "CloudRunV2JobTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CloudRunV2Job = CloudRunV2Job;
_9 = JSII_RTTI_SYMBOL_1;
CloudRunV2Job[_9] = { fqn: "@cdktf/provider-google.cloudRunV2Job.CloudRunV2Job", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
CloudRunV2Job.tfResourceType = "google_cloud_run_v2_job";
