// @cdktf/provider-google/lib/data-google-secret-manager-secrets/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleSecretManagerSecrets = exports.DataGoogleSecretManagerSecretsSecretsList = exports.DataGoogleSecretManagerSecretsSecretsOutputReference = exports.dataGoogleSecretManagerSecretsSecretsToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsToTerraform = exports.DataGoogleSecretManagerSecretsSecretsTopicsList = exports.DataGoogleSecretManagerSecretsSecretsTopicsOutputReference = exports.dataGoogleSecretManagerSecretsSecretsTopicsToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsTopicsToTerraform = exports.DataGoogleSecretManagerSecretsSecretsRotationList = exports.DataGoogleSecretManagerSecretsSecretsRotationOutputReference = exports.dataGoogleSecretManagerSecretsSecretsRotationToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsRotationToTerraform = exports.DataGoogleSecretManagerSecretsSecretsReplicationList = exports.DataGoogleSecretManagerSecretsSecretsReplicationOutputReference = exports.dataGoogleSecretManagerSecretsSecretsReplicationToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsReplicationToTerraform = exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedList = exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedOutputReference = exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedToTerraform = exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasList = exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasOutputReference = exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasToTerraform = exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionList = exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionOutputReference = exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionToTerraform = exports.DataGoogleSecretManagerSecretsSecretsReplicationAutoList = exports.DataGoogleSecretManagerSecretsSecretsReplicationAutoOutputReference = exports.dataGoogleSecretManagerSecretsSecretsReplicationAutoToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsReplicationAutoToTerraform = exports.DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionList = exports.DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionOutputReference = exports.dataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionToHclTerraform = exports.dataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionToTerraform = dataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionToTerraform;
function dataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionToHclTerraform = dataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_name - computed: true, optional: false, required: false
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionOutputReference = DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionList = DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionList[_b] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionList", version: "14.13.1" };
function dataGoogleSecretManagerSecretsSecretsReplicationAutoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationAutoToTerraform = dataGoogleSecretManagerSecretsSecretsReplicationAutoToTerraform;
function dataGoogleSecretManagerSecretsSecretsReplicationAutoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationAutoToHclTerraform = dataGoogleSecretManagerSecretsSecretsReplicationAutoToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsReplicationAutoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // customer_managed_encryption - computed: true, optional: false, required: false
        this._customerManagedEncryption = new DataGoogleSecretManagerSecretsSecretsReplicationAutoCustomerManagedEncryptionList(this, "customer_managed_encryption", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get customerManagedEncryption() {
        return this._customerManagedEncryption;
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationAutoOutputReference = DataGoogleSecretManagerSecretsSecretsReplicationAutoOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationAutoOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationAutoOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsReplicationAutoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsReplicationAutoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationAutoList = DataGoogleSecretManagerSecretsSecretsReplicationAutoList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationAutoList[_d] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationAutoList", version: "14.13.1" };
function dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionToTerraform = dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionToTerraform;
function dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionToHclTerraform = dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_name - computed: true, optional: false, required: false
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionOutputReference = DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionList = DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionList[_f] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionList", version: "14.13.1" };
function dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasToTerraform = dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasToTerraform;
function dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasToHclTerraform = dataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // customer_managed_encryption - computed: true, optional: false, required: false
        this._customerManagedEncryption = new DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasCustomerManagedEncryptionList(this, "customer_managed_encryption", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get customerManagedEncryption() {
        return this._customerManagedEncryption;
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasOutputReference = DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasList = DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasList[_h] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasList", version: "14.13.1" };
function dataGoogleSecretManagerSecretsSecretsReplicationUserManagedToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedToTerraform = dataGoogleSecretManagerSecretsSecretsReplicationUserManagedToTerraform;
function dataGoogleSecretManagerSecretsSecretsReplicationUserManagedToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationUserManagedToHclTerraform = dataGoogleSecretManagerSecretsSecretsReplicationUserManagedToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsReplicationUserManagedOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // replicas - computed: true, optional: false, required: false
        this._replicas = new DataGoogleSecretManagerSecretsSecretsReplicationUserManagedReplicasList(this, "replicas", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get replicas() {
        return this._replicas;
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedOutputReference = DataGoogleSecretManagerSecretsSecretsReplicationUserManagedOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationUserManagedOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsReplicationUserManagedList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsReplicationUserManagedOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedList = DataGoogleSecretManagerSecretsSecretsReplicationUserManagedList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationUserManagedList[_k] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationUserManagedList", version: "14.13.1" };
function dataGoogleSecretManagerSecretsSecretsReplicationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationToTerraform = dataGoogleSecretManagerSecretsSecretsReplicationToTerraform;
function dataGoogleSecretManagerSecretsSecretsReplicationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsReplicationToHclTerraform = dataGoogleSecretManagerSecretsSecretsReplicationToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsReplicationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // auto - computed: true, optional: false, required: false
        this._auto = new DataGoogleSecretManagerSecretsSecretsReplicationAutoList(this, "auto", false);
        // user_managed - computed: true, optional: false, required: false
        this._userManaged = new DataGoogleSecretManagerSecretsSecretsReplicationUserManagedList(this, "user_managed", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get auto() {
        return this._auto;
    }
    get userManaged() {
        return this._userManaged;
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationOutputReference = DataGoogleSecretManagerSecretsSecretsReplicationOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsReplicationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsReplicationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsReplicationList = DataGoogleSecretManagerSecretsSecretsReplicationList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsReplicationList[_m] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsReplicationList", version: "14.13.1" };
function dataGoogleSecretManagerSecretsSecretsRotationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsRotationToTerraform = dataGoogleSecretManagerSecretsSecretsRotationToTerraform;
function dataGoogleSecretManagerSecretsSecretsRotationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsRotationToHclTerraform = dataGoogleSecretManagerSecretsSecretsRotationToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsRotationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // next_rotation_time - computed: true, optional: false, required: false
    get nextRotationTime() {
        return this.getStringAttribute('next_rotation_time');
    }
    // rotation_period - computed: true, optional: false, required: false
    get rotationPeriod() {
        return this.getStringAttribute('rotation_period');
    }
}
exports.DataGoogleSecretManagerSecretsSecretsRotationOutputReference = DataGoogleSecretManagerSecretsSecretsRotationOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsRotationOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsRotationOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsRotationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsRotationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsRotationList = DataGoogleSecretManagerSecretsSecretsRotationList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsRotationList[_p] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsRotationList", version: "14.13.1" };
function dataGoogleSecretManagerSecretsSecretsTopicsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsTopicsToTerraform = dataGoogleSecretManagerSecretsSecretsTopicsToTerraform;
function dataGoogleSecretManagerSecretsSecretsTopicsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsTopicsToHclTerraform = dataGoogleSecretManagerSecretsSecretsTopicsToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsTopicsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataGoogleSecretManagerSecretsSecretsTopicsOutputReference = DataGoogleSecretManagerSecretsSecretsTopicsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsTopicsOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsTopicsOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsTopicsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsTopicsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsTopicsList = DataGoogleSecretManagerSecretsSecretsTopicsList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsTopicsList[_r] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsTopicsList", version: "14.13.1" };
function dataGoogleSecretManagerSecretsSecretsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSecretManagerSecretsSecretsToTerraform = dataGoogleSecretManagerSecretsSecretsToTerraform;
function dataGoogleSecretManagerSecretsSecretsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSecretManagerSecretsSecretsToHclTerraform = dataGoogleSecretManagerSecretsSecretsToHclTerraform;
class DataGoogleSecretManagerSecretsSecretsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // annotations - computed: true, optional: false, required: false
        this._annotations = new cdktf.StringMap(this, "annotations");
        // effective_annotations - computed: true, optional: false, required: false
        this._effectiveAnnotations = new cdktf.StringMap(this, "effective_annotations");
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // replication - computed: true, optional: false, required: false
        this._replication = new DataGoogleSecretManagerSecretsSecretsReplicationList(this, "replication", false);
        // rotation - computed: true, optional: false, required: false
        this._rotation = new DataGoogleSecretManagerSecretsSecretsRotationList(this, "rotation", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // topics - computed: true, optional: false, required: false
        this._topics = new DataGoogleSecretManagerSecretsSecretsTopicsList(this, "topics", false);
        // version_aliases - computed: true, optional: false, required: false
        this._versionAliases = new cdktf.StringMap(this, "version_aliases");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get annotations() {
        return this._annotations;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get effectiveAnnotations() {
        return this._effectiveAnnotations;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // expire_time - computed: true, optional: false, required: false
    get expireTime() {
        return this.getStringAttribute('expire_time');
    }
    get labels() {
        return this._labels;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
    get replication() {
        return this._replication;
    }
    get rotation() {
        return this._rotation;
    }
    // secret_id - computed: true, optional: false, required: false
    get secretId() {
        return this.getStringAttribute('secret_id');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    get topics() {
        return this._topics;
    }
    // ttl - computed: true, optional: false, required: false
    get ttl() {
        return this.getStringAttribute('ttl');
    }
    get versionAliases() {
        return this._versionAliases;
    }
    // version_destroy_ttl - computed: true, optional: false, required: false
    get versionDestroyTtl() {
        return this.getStringAttribute('version_destroy_ttl');
    }
}
exports.DataGoogleSecretManagerSecretsSecretsOutputReference = DataGoogleSecretManagerSecretsSecretsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsOutputReference", version: "14.13.1" };
class DataGoogleSecretManagerSecretsSecretsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSecretManagerSecretsSecretsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSecretManagerSecretsSecretsList = DataGoogleSecretManagerSecretsSecretsList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecretsSecretsList[_t] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecretsSecretsList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/secret_manager_secrets google_secret_manager_secrets}
*/
class DataGoogleSecretManagerSecrets extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleSecretManagerSecrets resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleSecretManagerSecrets to import
    * @param importFromId The id of the existing DataGoogleSecretManagerSecrets that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/secret_manager_secrets#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleSecretManagerSecrets to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_secret_manager_secrets", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/secret_manager_secrets google_secret_manager_secrets} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleSecretManagerSecretsConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_secret_manager_secrets',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // secrets - computed: true, optional: false, required: false
        this._secrets = new DataGoogleSecretManagerSecretsSecretsList(this, "secrets", false);
        this._filter = config.filter;
        this._id = config.id;
        this._project = config.project;
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get secrets() {
        return this._secrets;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filter: cdktf.stringToTerraform(this._filter),
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            filter: {
                value: cdktf.stringToHclTerraform(this._filter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleSecretManagerSecrets = DataGoogleSecretManagerSecrets;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleSecretManagerSecrets[_u] = { fqn: "@cdktf/provider-google.dataGoogleSecretManagerSecrets.DataGoogleSecretManagerSecrets", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleSecretManagerSecrets.tfResourceType = "google_secret_manager_secrets";
