// @cdktf/provider-google/lib/data-google-apphub-application/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleApphubApplication = exports.DataGoogleApphubApplicationScopeList = exports.DataGoogleApphubApplicationScopeOutputReference = exports.dataGoogleApphubApplicationScopeToHclTerraform = exports.dataGoogleApphubApplicationScopeToTerraform = exports.DataGoogleApphubApplicationAttributesList = exports.DataGoogleApphubApplicationAttributesOutputReference = exports.dataGoogleApphubApplicationAttributesToHclTerraform = exports.dataGoogleApphubApplicationAttributesToTerraform = exports.DataGoogleApphubApplicationAttributesOperatorOwnersList = exports.DataGoogleApphubApplicationAttributesOperatorOwnersOutputReference = exports.dataGoogleApphubApplicationAttributesOperatorOwnersToHclTerraform = exports.dataGoogleApphubApplicationAttributesOperatorOwnersToTerraform = exports.DataGoogleApphubApplicationAttributesEnvironmentList = exports.DataGoogleApphubApplicationAttributesEnvironmentOutputReference = exports.dataGoogleApphubApplicationAttributesEnvironmentToHclTerraform = exports.dataGoogleApphubApplicationAttributesEnvironmentToTerraform = exports.DataGoogleApphubApplicationAttributesDeveloperOwnersList = exports.DataGoogleApphubApplicationAttributesDeveloperOwnersOutputReference = exports.dataGoogleApphubApplicationAttributesDeveloperOwnersToHclTerraform = exports.dataGoogleApphubApplicationAttributesDeveloperOwnersToTerraform = exports.DataGoogleApphubApplicationAttributesCriticalityList = exports.DataGoogleApphubApplicationAttributesCriticalityOutputReference = exports.dataGoogleApphubApplicationAttributesCriticalityToHclTerraform = exports.dataGoogleApphubApplicationAttributesCriticalityToTerraform = exports.DataGoogleApphubApplicationAttributesBusinessOwnersList = exports.DataGoogleApphubApplicationAttributesBusinessOwnersOutputReference = exports.dataGoogleApphubApplicationAttributesBusinessOwnersToHclTerraform = exports.dataGoogleApphubApplicationAttributesBusinessOwnersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleApphubApplicationAttributesBusinessOwnersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleApphubApplicationAttributesBusinessOwnersToTerraform = dataGoogleApphubApplicationAttributesBusinessOwnersToTerraform;
function dataGoogleApphubApplicationAttributesBusinessOwnersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleApphubApplicationAttributesBusinessOwnersToHclTerraform = dataGoogleApphubApplicationAttributesBusinessOwnersToHclTerraform;
class DataGoogleApphubApplicationAttributesBusinessOwnersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
}
exports.DataGoogleApphubApplicationAttributesBusinessOwnersOutputReference = DataGoogleApphubApplicationAttributesBusinessOwnersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesBusinessOwnersOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesBusinessOwnersOutputReference", version: "14.13.1" };
class DataGoogleApphubApplicationAttributesBusinessOwnersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleApphubApplicationAttributesBusinessOwnersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleApphubApplicationAttributesBusinessOwnersList = DataGoogleApphubApplicationAttributesBusinessOwnersList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesBusinessOwnersList[_b] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesBusinessOwnersList", version: "14.13.1" };
function dataGoogleApphubApplicationAttributesCriticalityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleApphubApplicationAttributesCriticalityToTerraform = dataGoogleApphubApplicationAttributesCriticalityToTerraform;
function dataGoogleApphubApplicationAttributesCriticalityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleApphubApplicationAttributesCriticalityToHclTerraform = dataGoogleApphubApplicationAttributesCriticalityToHclTerraform;
class DataGoogleApphubApplicationAttributesCriticalityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleApphubApplicationAttributesCriticalityOutputReference = DataGoogleApphubApplicationAttributesCriticalityOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesCriticalityOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesCriticalityOutputReference", version: "14.13.1" };
class DataGoogleApphubApplicationAttributesCriticalityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleApphubApplicationAttributesCriticalityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleApphubApplicationAttributesCriticalityList = DataGoogleApphubApplicationAttributesCriticalityList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesCriticalityList[_d] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesCriticalityList", version: "14.13.1" };
function dataGoogleApphubApplicationAttributesDeveloperOwnersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleApphubApplicationAttributesDeveloperOwnersToTerraform = dataGoogleApphubApplicationAttributesDeveloperOwnersToTerraform;
function dataGoogleApphubApplicationAttributesDeveloperOwnersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleApphubApplicationAttributesDeveloperOwnersToHclTerraform = dataGoogleApphubApplicationAttributesDeveloperOwnersToHclTerraform;
class DataGoogleApphubApplicationAttributesDeveloperOwnersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
}
exports.DataGoogleApphubApplicationAttributesDeveloperOwnersOutputReference = DataGoogleApphubApplicationAttributesDeveloperOwnersOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesDeveloperOwnersOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesDeveloperOwnersOutputReference", version: "14.13.1" };
class DataGoogleApphubApplicationAttributesDeveloperOwnersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleApphubApplicationAttributesDeveloperOwnersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleApphubApplicationAttributesDeveloperOwnersList = DataGoogleApphubApplicationAttributesDeveloperOwnersList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesDeveloperOwnersList[_f] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesDeveloperOwnersList", version: "14.13.1" };
function dataGoogleApphubApplicationAttributesEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleApphubApplicationAttributesEnvironmentToTerraform = dataGoogleApphubApplicationAttributesEnvironmentToTerraform;
function dataGoogleApphubApplicationAttributesEnvironmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleApphubApplicationAttributesEnvironmentToHclTerraform = dataGoogleApphubApplicationAttributesEnvironmentToHclTerraform;
class DataGoogleApphubApplicationAttributesEnvironmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleApphubApplicationAttributesEnvironmentOutputReference = DataGoogleApphubApplicationAttributesEnvironmentOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesEnvironmentOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesEnvironmentOutputReference", version: "14.13.1" };
class DataGoogleApphubApplicationAttributesEnvironmentList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleApphubApplicationAttributesEnvironmentOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleApphubApplicationAttributesEnvironmentList = DataGoogleApphubApplicationAttributesEnvironmentList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesEnvironmentList[_h] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesEnvironmentList", version: "14.13.1" };
function dataGoogleApphubApplicationAttributesOperatorOwnersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleApphubApplicationAttributesOperatorOwnersToTerraform = dataGoogleApphubApplicationAttributesOperatorOwnersToTerraform;
function dataGoogleApphubApplicationAttributesOperatorOwnersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleApphubApplicationAttributesOperatorOwnersToHclTerraform = dataGoogleApphubApplicationAttributesOperatorOwnersToHclTerraform;
class DataGoogleApphubApplicationAttributesOperatorOwnersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
}
exports.DataGoogleApphubApplicationAttributesOperatorOwnersOutputReference = DataGoogleApphubApplicationAttributesOperatorOwnersOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesOperatorOwnersOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesOperatorOwnersOutputReference", version: "14.13.1" };
class DataGoogleApphubApplicationAttributesOperatorOwnersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleApphubApplicationAttributesOperatorOwnersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleApphubApplicationAttributesOperatorOwnersList = DataGoogleApphubApplicationAttributesOperatorOwnersList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesOperatorOwnersList[_k] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesOperatorOwnersList", version: "14.13.1" };
function dataGoogleApphubApplicationAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleApphubApplicationAttributesToTerraform = dataGoogleApphubApplicationAttributesToTerraform;
function dataGoogleApphubApplicationAttributesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleApphubApplicationAttributesToHclTerraform = dataGoogleApphubApplicationAttributesToHclTerraform;
class DataGoogleApphubApplicationAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // business_owners - computed: true, optional: false, required: false
        this._businessOwners = new DataGoogleApphubApplicationAttributesBusinessOwnersList(this, "business_owners", false);
        // criticality - computed: true, optional: false, required: false
        this._criticality = new DataGoogleApphubApplicationAttributesCriticalityList(this, "criticality", false);
        // developer_owners - computed: true, optional: false, required: false
        this._developerOwners = new DataGoogleApphubApplicationAttributesDeveloperOwnersList(this, "developer_owners", false);
        // environment - computed: true, optional: false, required: false
        this._environment = new DataGoogleApphubApplicationAttributesEnvironmentList(this, "environment", false);
        // operator_owners - computed: true, optional: false, required: false
        this._operatorOwners = new DataGoogleApphubApplicationAttributesOperatorOwnersList(this, "operator_owners", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get businessOwners() {
        return this._businessOwners;
    }
    get criticality() {
        return this._criticality;
    }
    get developerOwners() {
        return this._developerOwners;
    }
    get environment() {
        return this._environment;
    }
    get operatorOwners() {
        return this._operatorOwners;
    }
}
exports.DataGoogleApphubApplicationAttributesOutputReference = DataGoogleApphubApplicationAttributesOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesOutputReference", version: "14.13.1" };
class DataGoogleApphubApplicationAttributesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleApphubApplicationAttributesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleApphubApplicationAttributesList = DataGoogleApphubApplicationAttributesList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationAttributesList[_m] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationAttributesList", version: "14.13.1" };
function dataGoogleApphubApplicationScopeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleApphubApplicationScopeToTerraform = dataGoogleApphubApplicationScopeToTerraform;
function dataGoogleApphubApplicationScopeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleApphubApplicationScopeToHclTerraform = dataGoogleApphubApplicationScopeToHclTerraform;
class DataGoogleApphubApplicationScopeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleApphubApplicationScopeOutputReference = DataGoogleApphubApplicationScopeOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationScopeOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationScopeOutputReference", version: "14.13.1" };
class DataGoogleApphubApplicationScopeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleApphubApplicationScopeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleApphubApplicationScopeList = DataGoogleApphubApplicationScopeList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplicationScopeList[_p] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplicationScopeList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/apphub_application google_apphub_application}
*/
class DataGoogleApphubApplication extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleApphubApplication resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleApphubApplication to import
    * @param importFromId The id of the existing DataGoogleApphubApplication that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/apphub_application#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleApphubApplication to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_apphub_application", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/apphub_application google_apphub_application} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleApphubApplicationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_apphub_application',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attributes - computed: true, optional: false, required: false
        this._attributes = new DataGoogleApphubApplicationAttributesList(this, "attributes", false);
        // scope - computed: true, optional: false, required: false
        this._scope = new DataGoogleApphubApplicationScopeList(this, "scope", false);
        this._applicationId = config.applicationId;
        this._id = config.id;
        this._location = config.location;
        this._project = config.project;
    }
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    set applicationId(value) {
        this._applicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationIdInput() {
        return this._applicationId;
    }
    get attributes() {
        return this._attributes;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get scope() {
        return this._scope;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_id: cdktf.stringToTerraform(this._applicationId),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_id: {
                value: cdktf.stringToHclTerraform(this._applicationId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleApphubApplication = DataGoogleApphubApplication;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleApphubApplication[_q] = { fqn: "@cdktf/provider-google.dataGoogleApphubApplication.DataGoogleApphubApplication", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleApphubApplication.tfResourceType = "google_apphub_application";
