// @cdktf/provider-google/lib/container-aws-node-pool/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerAwsNodePool = exports.ContainerAwsNodePoolUpdateSettingsOutputReference = exports.containerAwsNodePoolUpdateSettingsToHclTerraform = exports.containerAwsNodePoolUpdateSettingsToTerraform = exports.ContainerAwsNodePoolUpdateSettingsSurgeSettingsOutputReference = exports.containerAwsNodePoolUpdateSettingsSurgeSettingsToHclTerraform = exports.containerAwsNodePoolUpdateSettingsSurgeSettingsToTerraform = exports.ContainerAwsNodePoolTimeoutsOutputReference = exports.containerAwsNodePoolTimeoutsToHclTerraform = exports.containerAwsNodePoolTimeoutsToTerraform = exports.ContainerAwsNodePoolMaxPodsConstraintOutputReference = exports.containerAwsNodePoolMaxPodsConstraintToHclTerraform = exports.containerAwsNodePoolMaxPodsConstraintToTerraform = exports.ContainerAwsNodePoolManagementOutputReference = exports.containerAwsNodePoolManagementToHclTerraform = exports.containerAwsNodePoolManagementToTerraform = exports.ContainerAwsNodePoolKubeletConfigOutputReference = exports.containerAwsNodePoolKubeletConfigToHclTerraform = exports.containerAwsNodePoolKubeletConfigToTerraform = exports.ContainerAwsNodePoolConfigAOutputReference = exports.containerAwsNodePoolConfigAToHclTerraform = exports.containerAwsNodePoolConfigAToTerraform = exports.ContainerAwsNodePoolConfigTaintsList = exports.ContainerAwsNodePoolConfigTaintsOutputReference = exports.containerAwsNodePoolConfigTaintsToHclTerraform = exports.containerAwsNodePoolConfigTaintsToTerraform = exports.ContainerAwsNodePoolConfigSshConfigOutputReference = exports.containerAwsNodePoolConfigSshConfigToHclTerraform = exports.containerAwsNodePoolConfigSshConfigToTerraform = exports.ContainerAwsNodePoolConfigRootVolumeOutputReference = exports.containerAwsNodePoolConfigRootVolumeToHclTerraform = exports.containerAwsNodePoolConfigRootVolumeToTerraform = exports.ContainerAwsNodePoolConfigProxyConfigOutputReference = exports.containerAwsNodePoolConfigProxyConfigToHclTerraform = exports.containerAwsNodePoolConfigProxyConfigToTerraform = exports.ContainerAwsNodePoolConfigConfigEncryptionOutputReference = exports.containerAwsNodePoolConfigConfigEncryptionToHclTerraform = exports.containerAwsNodePoolConfigConfigEncryptionToTerraform = exports.ContainerAwsNodePoolConfigAutoscalingMetricsCollectionOutputReference = exports.containerAwsNodePoolConfigAutoscalingMetricsCollectionToHclTerraform = exports.containerAwsNodePoolConfigAutoscalingMetricsCollectionToTerraform = exports.ContainerAwsNodePoolAutoscalingOutputReference = exports.containerAwsNodePoolAutoscalingToHclTerraform = exports.containerAwsNodePoolAutoscalingToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function containerAwsNodePoolAutoscalingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_node_count: cdktf.numberToTerraform(struct.maxNodeCount),
        min_node_count: cdktf.numberToTerraform(struct.minNodeCount),
    };
}
exports.containerAwsNodePoolAutoscalingToTerraform = containerAwsNodePoolAutoscalingToTerraform;
function containerAwsNodePoolAutoscalingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_node_count: {
            value: cdktf.numberToHclTerraform(struct.maxNodeCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        min_node_count: {
            value: cdktf.numberToHclTerraform(struct.minNodeCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolAutoscalingToHclTerraform = containerAwsNodePoolAutoscalingToHclTerraform;
class ContainerAwsNodePoolAutoscalingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxNodeCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxNodeCount = this._maxNodeCount;
        }
        if (this._minNodeCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.minNodeCount = this._minNodeCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxNodeCount = undefined;
            this._minNodeCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxNodeCount = value.maxNodeCount;
            this._minNodeCount = value.minNodeCount;
        }
    }
    get maxNodeCount() {
        return this.getNumberAttribute('max_node_count');
    }
    set maxNodeCount(value) {
        this._maxNodeCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxNodeCountInput() {
        return this._maxNodeCount;
    }
    get minNodeCount() {
        return this.getNumberAttribute('min_node_count');
    }
    set minNodeCount(value) {
        this._minNodeCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get minNodeCountInput() {
        return this._minNodeCount;
    }
}
exports.ContainerAwsNodePoolAutoscalingOutputReference = ContainerAwsNodePoolAutoscalingOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolAutoscalingOutputReference[_a] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolAutoscalingOutputReference", version: "14.13.1" };
function containerAwsNodePoolConfigAutoscalingMetricsCollectionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        granularity: cdktf.stringToTerraform(struct.granularity),
        metrics: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.metrics),
    };
}
exports.containerAwsNodePoolConfigAutoscalingMetricsCollectionToTerraform = containerAwsNodePoolConfigAutoscalingMetricsCollectionToTerraform;
function containerAwsNodePoolConfigAutoscalingMetricsCollectionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        granularity: {
            value: cdktf.stringToHclTerraform(struct.granularity),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        metrics: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.metrics),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolConfigAutoscalingMetricsCollectionToHclTerraform = containerAwsNodePoolConfigAutoscalingMetricsCollectionToHclTerraform;
class ContainerAwsNodePoolConfigAutoscalingMetricsCollectionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._granularity !== undefined) {
            hasAnyValues = true;
            internalValueResult.granularity = this._granularity;
        }
        if (this._metrics !== undefined) {
            hasAnyValues = true;
            internalValueResult.metrics = this._metrics;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._granularity = undefined;
            this._metrics = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._granularity = value.granularity;
            this._metrics = value.metrics;
        }
    }
    get granularity() {
        return this.getStringAttribute('granularity');
    }
    set granularity(value) {
        this._granularity = value;
    }
    // Temporarily expose input value. Use with caution.
    get granularityInput() {
        return this._granularity;
    }
    get metrics() {
        return this.getListAttribute('metrics');
    }
    set metrics(value) {
        this._metrics = value;
    }
    resetMetrics() {
        this._metrics = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metricsInput() {
        return this._metrics;
    }
}
exports.ContainerAwsNodePoolConfigAutoscalingMetricsCollectionOutputReference = ContainerAwsNodePoolConfigAutoscalingMetricsCollectionOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolConfigAutoscalingMetricsCollectionOutputReference[_b] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolConfigAutoscalingMetricsCollectionOutputReference", version: "14.13.1" };
function containerAwsNodePoolConfigConfigEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_arn: cdktf.stringToTerraform(struct.kmsKeyArn),
    };
}
exports.containerAwsNodePoolConfigConfigEncryptionToTerraform = containerAwsNodePoolConfigConfigEncryptionToTerraform;
function containerAwsNodePoolConfigConfigEncryptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key_arn: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolConfigConfigEncryptionToHclTerraform = containerAwsNodePoolConfigConfigEncryptionToHclTerraform;
class ContainerAwsNodePoolConfigConfigEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyArn = this._kmsKeyArn;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyArn = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyArn = value.kmsKeyArn;
        }
    }
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
}
exports.ContainerAwsNodePoolConfigConfigEncryptionOutputReference = ContainerAwsNodePoolConfigConfigEncryptionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolConfigConfigEncryptionOutputReference[_c] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolConfigConfigEncryptionOutputReference", version: "14.13.1" };
function containerAwsNodePoolConfigProxyConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        secret_arn: cdktf.stringToTerraform(struct.secretArn),
        secret_version: cdktf.stringToTerraform(struct.secretVersion),
    };
}
exports.containerAwsNodePoolConfigProxyConfigToTerraform = containerAwsNodePoolConfigProxyConfigToTerraform;
function containerAwsNodePoolConfigProxyConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        secret_arn: {
            value: cdktf.stringToHclTerraform(struct.secretArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        secret_version: {
            value: cdktf.stringToHclTerraform(struct.secretVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolConfigProxyConfigToHclTerraform = containerAwsNodePoolConfigProxyConfigToHclTerraform;
class ContainerAwsNodePoolConfigProxyConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._secretArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretArn = this._secretArn;
        }
        if (this._secretVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretVersion = this._secretVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._secretArn = undefined;
            this._secretVersion = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._secretArn = value.secretArn;
            this._secretVersion = value.secretVersion;
        }
    }
    get secretArn() {
        return this.getStringAttribute('secret_arn');
    }
    set secretArn(value) {
        this._secretArn = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretArnInput() {
        return this._secretArn;
    }
    get secretVersion() {
        return this.getStringAttribute('secret_version');
    }
    set secretVersion(value) {
        this._secretVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretVersionInput() {
        return this._secretVersion;
    }
}
exports.ContainerAwsNodePoolConfigProxyConfigOutputReference = ContainerAwsNodePoolConfigProxyConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolConfigProxyConfigOutputReference[_d] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolConfigProxyConfigOutputReference", version: "14.13.1" };
function containerAwsNodePoolConfigRootVolumeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iops: cdktf.numberToTerraform(struct.iops),
        kms_key_arn: cdktf.stringToTerraform(struct.kmsKeyArn),
        size_gib: cdktf.numberToTerraform(struct.sizeGib),
        throughput: cdktf.numberToTerraform(struct.throughput),
        volume_type: cdktf.stringToTerraform(struct.volumeType),
    };
}
exports.containerAwsNodePoolConfigRootVolumeToTerraform = containerAwsNodePoolConfigRootVolumeToTerraform;
function containerAwsNodePoolConfigRootVolumeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        iops: {
            value: cdktf.numberToHclTerraform(struct.iops),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        kms_key_arn: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        size_gib: {
            value: cdktf.numberToHclTerraform(struct.sizeGib),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        throughput: {
            value: cdktf.numberToHclTerraform(struct.throughput),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        volume_type: {
            value: cdktf.stringToHclTerraform(struct.volumeType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolConfigRootVolumeToHclTerraform = containerAwsNodePoolConfigRootVolumeToHclTerraform;
class ContainerAwsNodePoolConfigRootVolumeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iops !== undefined) {
            hasAnyValues = true;
            internalValueResult.iops = this._iops;
        }
        if (this._kmsKeyArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyArn = this._kmsKeyArn;
        }
        if (this._sizeGib !== undefined) {
            hasAnyValues = true;
            internalValueResult.sizeGib = this._sizeGib;
        }
        if (this._throughput !== undefined) {
            hasAnyValues = true;
            internalValueResult.throughput = this._throughput;
        }
        if (this._volumeType !== undefined) {
            hasAnyValues = true;
            internalValueResult.volumeType = this._volumeType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iops = undefined;
            this._kmsKeyArn = undefined;
            this._sizeGib = undefined;
            this._throughput = undefined;
            this._volumeType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iops = value.iops;
            this._kmsKeyArn = value.kmsKeyArn;
            this._sizeGib = value.sizeGib;
            this._throughput = value.throughput;
            this._volumeType = value.volumeType;
        }
    }
    get iops() {
        return this.getNumberAttribute('iops');
    }
    set iops(value) {
        this._iops = value;
    }
    resetIops() {
        this._iops = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iopsInput() {
        return this._iops;
    }
    get kmsKeyArn() {
        return this.getStringAttribute('kms_key_arn');
    }
    set kmsKeyArn(value) {
        this._kmsKeyArn = value;
    }
    resetKmsKeyArn() {
        this._kmsKeyArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyArnInput() {
        return this._kmsKeyArn;
    }
    get sizeGib() {
        return this.getNumberAttribute('size_gib');
    }
    set sizeGib(value) {
        this._sizeGib = value;
    }
    resetSizeGib() {
        this._sizeGib = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sizeGibInput() {
        return this._sizeGib;
    }
    get throughput() {
        return this.getNumberAttribute('throughput');
    }
    set throughput(value) {
        this._throughput = value;
    }
    resetThroughput() {
        this._throughput = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get throughputInput() {
        return this._throughput;
    }
    get volumeType() {
        return this.getStringAttribute('volume_type');
    }
    set volumeType(value) {
        this._volumeType = value;
    }
    resetVolumeType() {
        this._volumeType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get volumeTypeInput() {
        return this._volumeType;
    }
}
exports.ContainerAwsNodePoolConfigRootVolumeOutputReference = ContainerAwsNodePoolConfigRootVolumeOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolConfigRootVolumeOutputReference[_e] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolConfigRootVolumeOutputReference", version: "14.13.1" };
function containerAwsNodePoolConfigSshConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ec2_key_pair: cdktf.stringToTerraform(struct.ec2KeyPair),
    };
}
exports.containerAwsNodePoolConfigSshConfigToTerraform = containerAwsNodePoolConfigSshConfigToTerraform;
function containerAwsNodePoolConfigSshConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ec2_key_pair: {
            value: cdktf.stringToHclTerraform(struct.ec2KeyPair),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolConfigSshConfigToHclTerraform = containerAwsNodePoolConfigSshConfigToHclTerraform;
class ContainerAwsNodePoolConfigSshConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ec2KeyPair !== undefined) {
            hasAnyValues = true;
            internalValueResult.ec2KeyPair = this._ec2KeyPair;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ec2KeyPair = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ec2KeyPair = value.ec2KeyPair;
        }
    }
    get ec2KeyPair() {
        return this.getStringAttribute('ec2_key_pair');
    }
    set ec2KeyPair(value) {
        this._ec2KeyPair = value;
    }
    // Temporarily expose input value. Use with caution.
    get ec2KeyPairInput() {
        return this._ec2KeyPair;
    }
}
exports.ContainerAwsNodePoolConfigSshConfigOutputReference = ContainerAwsNodePoolConfigSshConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolConfigSshConfigOutputReference[_f] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolConfigSshConfigOutputReference", version: "14.13.1" };
function containerAwsNodePoolConfigTaintsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        effect: cdktf.stringToTerraform(struct.effect),
        key: cdktf.stringToTerraform(struct.key),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.containerAwsNodePoolConfigTaintsToTerraform = containerAwsNodePoolConfigTaintsToTerraform;
function containerAwsNodePoolConfigTaintsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        effect: {
            value: cdktf.stringToHclTerraform(struct.effect),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        key: {
            value: cdktf.stringToHclTerraform(struct.key),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolConfigTaintsToHclTerraform = containerAwsNodePoolConfigTaintsToHclTerraform;
class ContainerAwsNodePoolConfigTaintsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._effect !== undefined) {
            hasAnyValues = true;
            internalValueResult.effect = this._effect;
        }
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._effect = undefined;
            this._key = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._effect = value.effect;
            this._key = value.key;
            this._value = value.value;
        }
    }
    get effect() {
        return this.getStringAttribute('effect');
    }
    set effect(value) {
        this._effect = value;
    }
    // Temporarily expose input value. Use with caution.
    get effectInput() {
        return this._effect;
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ContainerAwsNodePoolConfigTaintsOutputReference = ContainerAwsNodePoolConfigTaintsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolConfigTaintsOutputReference[_g] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolConfigTaintsOutputReference", version: "14.13.1" };
class ContainerAwsNodePoolConfigTaintsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ContainerAwsNodePoolConfigTaintsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ContainerAwsNodePoolConfigTaintsList = ContainerAwsNodePoolConfigTaintsList;
_h = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolConfigTaintsList[_h] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolConfigTaintsList", version: "14.13.1" };
function containerAwsNodePoolConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam_instance_profile: cdktf.stringToTerraform(struct.iamInstanceProfile),
        instance_type: cdktf.stringToTerraform(struct.instanceType),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        tags: cdktf.hashMapper(cdktf.stringToTerraform)(struct.tags),
        autoscaling_metrics_collection: containerAwsNodePoolConfigAutoscalingMetricsCollectionToTerraform(struct.autoscalingMetricsCollection),
        config_encryption: containerAwsNodePoolConfigConfigEncryptionToTerraform(struct.configEncryption),
        proxy_config: containerAwsNodePoolConfigProxyConfigToTerraform(struct.proxyConfig),
        root_volume: containerAwsNodePoolConfigRootVolumeToTerraform(struct.rootVolume),
        ssh_config: containerAwsNodePoolConfigSshConfigToTerraform(struct.sshConfig),
        taints: cdktf.listMapper(containerAwsNodePoolConfigTaintsToTerraform, true)(struct.taints),
    };
}
exports.containerAwsNodePoolConfigAToTerraform = containerAwsNodePoolConfigAToTerraform;
function containerAwsNodePoolConfigAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        iam_instance_profile: {
            value: cdktf.stringToHclTerraform(struct.iamInstanceProfile),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        instance_type: {
            value: cdktf.stringToHclTerraform(struct.instanceType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        labels: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.labels),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        security_group_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.securityGroupIds),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        tags: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.tags),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        autoscaling_metrics_collection: {
            value: containerAwsNodePoolConfigAutoscalingMetricsCollectionToHclTerraform(struct.autoscalingMetricsCollection),
            isBlock: true,
            type: "list",
            storageClassType: "ContainerAwsNodePoolConfigAutoscalingMetricsCollectionList",
        },
        config_encryption: {
            value: containerAwsNodePoolConfigConfigEncryptionToHclTerraform(struct.configEncryption),
            isBlock: true,
            type: "list",
            storageClassType: "ContainerAwsNodePoolConfigConfigEncryptionList",
        },
        proxy_config: {
            value: containerAwsNodePoolConfigProxyConfigToHclTerraform(struct.proxyConfig),
            isBlock: true,
            type: "list",
            storageClassType: "ContainerAwsNodePoolConfigProxyConfigList",
        },
        root_volume: {
            value: containerAwsNodePoolConfigRootVolumeToHclTerraform(struct.rootVolume),
            isBlock: true,
            type: "list",
            storageClassType: "ContainerAwsNodePoolConfigRootVolumeList",
        },
        ssh_config: {
            value: containerAwsNodePoolConfigSshConfigToHclTerraform(struct.sshConfig),
            isBlock: true,
            type: "list",
            storageClassType: "ContainerAwsNodePoolConfigSshConfigList",
        },
        taints: {
            value: cdktf.listMapperHcl(containerAwsNodePoolConfigTaintsToHclTerraform, true)(struct.taints),
            isBlock: true,
            type: "list",
            storageClassType: "ContainerAwsNodePoolConfigTaintsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolConfigAToHclTerraform = containerAwsNodePoolConfigAToHclTerraform;
class ContainerAwsNodePoolConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // autoscaling_metrics_collection - computed: false, optional: true, required: false
        this._autoscalingMetricsCollection = new ContainerAwsNodePoolConfigAutoscalingMetricsCollectionOutputReference(this, "autoscaling_metrics_collection");
        // config_encryption - computed: false, optional: false, required: true
        this._configEncryption = new ContainerAwsNodePoolConfigConfigEncryptionOutputReference(this, "config_encryption");
        // proxy_config - computed: false, optional: true, required: false
        this._proxyConfig = new ContainerAwsNodePoolConfigProxyConfigOutputReference(this, "proxy_config");
        // root_volume - computed: false, optional: true, required: false
        this._rootVolume = new ContainerAwsNodePoolConfigRootVolumeOutputReference(this, "root_volume");
        // ssh_config - computed: false, optional: true, required: false
        this._sshConfig = new ContainerAwsNodePoolConfigSshConfigOutputReference(this, "ssh_config");
        // taints - computed: false, optional: true, required: false
        this._taints = new ContainerAwsNodePoolConfigTaintsList(this, "taints", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iamInstanceProfile !== undefined) {
            hasAnyValues = true;
            internalValueResult.iamInstanceProfile = this._iamInstanceProfile;
        }
        if (this._instanceType !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceType = this._instanceType;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        if (this._autoscalingMetricsCollection?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoscalingMetricsCollection = this._autoscalingMetricsCollection?.internalValue;
        }
        if (this._configEncryption?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.configEncryption = this._configEncryption?.internalValue;
        }
        if (this._proxyConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.proxyConfig = this._proxyConfig?.internalValue;
        }
        if (this._rootVolume?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rootVolume = this._rootVolume?.internalValue;
        }
        if (this._sshConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sshConfig = this._sshConfig?.internalValue;
        }
        if (this._taints?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.taints = this._taints?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iamInstanceProfile = undefined;
            this._instanceType = undefined;
            this._labels = undefined;
            this._securityGroupIds = undefined;
            this._tags = undefined;
            this._autoscalingMetricsCollection.internalValue = undefined;
            this._configEncryption.internalValue = undefined;
            this._proxyConfig.internalValue = undefined;
            this._rootVolume.internalValue = undefined;
            this._sshConfig.internalValue = undefined;
            this._taints.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iamInstanceProfile = value.iamInstanceProfile;
            this._instanceType = value.instanceType;
            this._labels = value.labels;
            this._securityGroupIds = value.securityGroupIds;
            this._tags = value.tags;
            this._autoscalingMetricsCollection.internalValue = value.autoscalingMetricsCollection;
            this._configEncryption.internalValue = value.configEncryption;
            this._proxyConfig.internalValue = value.proxyConfig;
            this._rootVolume.internalValue = value.rootVolume;
            this._sshConfig.internalValue = value.sshConfig;
            this._taints.internalValue = value.taints;
        }
    }
    get iamInstanceProfile() {
        return this.getStringAttribute('iam_instance_profile');
    }
    set iamInstanceProfile(value) {
        this._iamInstanceProfile = value;
    }
    // Temporarily expose input value. Use with caution.
    get iamInstanceProfileInput() {
        return this._iamInstanceProfile;
    }
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    set instanceType(value) {
        this._instanceType = value;
    }
    resetInstanceType() {
        this._instanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTypeInput() {
        return this._instanceType;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get securityGroupIds() {
        return this.getListAttribute('security_group_ids');
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get tags() {
        return this.getStringMapAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get autoscalingMetricsCollection() {
        return this._autoscalingMetricsCollection;
    }
    putAutoscalingMetricsCollection(value) {
        this._autoscalingMetricsCollection.internalValue = value;
    }
    resetAutoscalingMetricsCollection() {
        this._autoscalingMetricsCollection.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingMetricsCollectionInput() {
        return this._autoscalingMetricsCollection.internalValue;
    }
    get configEncryption() {
        return this._configEncryption;
    }
    putConfigEncryption(value) {
        this._configEncryption.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get configEncryptionInput() {
        return this._configEncryption.internalValue;
    }
    get proxyConfig() {
        return this._proxyConfig;
    }
    putProxyConfig(value) {
        this._proxyConfig.internalValue = value;
    }
    resetProxyConfig() {
        this._proxyConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get proxyConfigInput() {
        return this._proxyConfig.internalValue;
    }
    get rootVolume() {
        return this._rootVolume;
    }
    putRootVolume(value) {
        this._rootVolume.internalValue = value;
    }
    resetRootVolume() {
        this._rootVolume.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rootVolumeInput() {
        return this._rootVolume.internalValue;
    }
    get sshConfig() {
        return this._sshConfig;
    }
    putSshConfig(value) {
        this._sshConfig.internalValue = value;
    }
    resetSshConfig() {
        this._sshConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sshConfigInput() {
        return this._sshConfig.internalValue;
    }
    get taints() {
        return this._taints;
    }
    putTaints(value) {
        this._taints.internalValue = value;
    }
    resetTaints() {
        this._taints.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get taintsInput() {
        return this._taints.internalValue;
    }
}
exports.ContainerAwsNodePoolConfigAOutputReference = ContainerAwsNodePoolConfigAOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolConfigAOutputReference[_j] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolConfigAOutputReference", version: "14.13.1" };
function containerAwsNodePoolKubeletConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu_cfs_quota: cdktf.booleanToTerraform(struct.cpuCfsQuota),
        cpu_cfs_quota_period: cdktf.stringToTerraform(struct.cpuCfsQuotaPeriod),
        cpu_manager_policy: cdktf.stringToTerraform(struct.cpuManagerPolicy),
        pod_pids_limit: cdktf.numberToTerraform(struct.podPidsLimit),
    };
}
exports.containerAwsNodePoolKubeletConfigToTerraform = containerAwsNodePoolKubeletConfigToTerraform;
function containerAwsNodePoolKubeletConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cpu_cfs_quota: {
            value: cdktf.booleanToHclTerraform(struct.cpuCfsQuota),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        cpu_cfs_quota_period: {
            value: cdktf.stringToHclTerraform(struct.cpuCfsQuotaPeriod),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        cpu_manager_policy: {
            value: cdktf.stringToHclTerraform(struct.cpuManagerPolicy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        pod_pids_limit: {
            value: cdktf.numberToHclTerraform(struct.podPidsLimit),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolKubeletConfigToHclTerraform = containerAwsNodePoolKubeletConfigToHclTerraform;
class ContainerAwsNodePoolKubeletConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpuCfsQuota !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuCfsQuota = this._cpuCfsQuota;
        }
        if (this._cpuCfsQuotaPeriod !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuCfsQuotaPeriod = this._cpuCfsQuotaPeriod;
        }
        if (this._cpuManagerPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuManagerPolicy = this._cpuManagerPolicy;
        }
        if (this._podPidsLimit !== undefined) {
            hasAnyValues = true;
            internalValueResult.podPidsLimit = this._podPidsLimit;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpuCfsQuota = undefined;
            this._cpuCfsQuotaPeriod = undefined;
            this._cpuManagerPolicy = undefined;
            this._podPidsLimit = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpuCfsQuota = value.cpuCfsQuota;
            this._cpuCfsQuotaPeriod = value.cpuCfsQuotaPeriod;
            this._cpuManagerPolicy = value.cpuManagerPolicy;
            this._podPidsLimit = value.podPidsLimit;
        }
    }
    get cpuCfsQuota() {
        return this.getBooleanAttribute('cpu_cfs_quota');
    }
    set cpuCfsQuota(value) {
        this._cpuCfsQuota = value;
    }
    resetCpuCfsQuota() {
        this._cpuCfsQuota = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuCfsQuotaInput() {
        return this._cpuCfsQuota;
    }
    get cpuCfsQuotaPeriod() {
        return this.getStringAttribute('cpu_cfs_quota_period');
    }
    set cpuCfsQuotaPeriod(value) {
        this._cpuCfsQuotaPeriod = value;
    }
    resetCpuCfsQuotaPeriod() {
        this._cpuCfsQuotaPeriod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuCfsQuotaPeriodInput() {
        return this._cpuCfsQuotaPeriod;
    }
    get cpuManagerPolicy() {
        return this.getStringAttribute('cpu_manager_policy');
    }
    set cpuManagerPolicy(value) {
        this._cpuManagerPolicy = value;
    }
    resetCpuManagerPolicy() {
        this._cpuManagerPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuManagerPolicyInput() {
        return this._cpuManagerPolicy;
    }
    get podPidsLimit() {
        return this.getNumberAttribute('pod_pids_limit');
    }
    set podPidsLimit(value) {
        this._podPidsLimit = value;
    }
    resetPodPidsLimit() {
        this._podPidsLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get podPidsLimitInput() {
        return this._podPidsLimit;
    }
}
exports.ContainerAwsNodePoolKubeletConfigOutputReference = ContainerAwsNodePoolKubeletConfigOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolKubeletConfigOutputReference[_k] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolKubeletConfigOutputReference", version: "14.13.1" };
function containerAwsNodePoolManagementToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_repair: cdktf.booleanToTerraform(struct.autoRepair),
    };
}
exports.containerAwsNodePoolManagementToTerraform = containerAwsNodePoolManagementToTerraform;
function containerAwsNodePoolManagementToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        auto_repair: {
            value: cdktf.booleanToHclTerraform(struct.autoRepair),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolManagementToHclTerraform = containerAwsNodePoolManagementToHclTerraform;
class ContainerAwsNodePoolManagementOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoRepair !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoRepair = this._autoRepair;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoRepair = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoRepair = value.autoRepair;
        }
    }
    get autoRepair() {
        return this.getBooleanAttribute('auto_repair');
    }
    set autoRepair(value) {
        this._autoRepair = value;
    }
    resetAutoRepair() {
        this._autoRepair = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoRepairInput() {
        return this._autoRepair;
    }
}
exports.ContainerAwsNodePoolManagementOutputReference = ContainerAwsNodePoolManagementOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolManagementOutputReference[_l] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolManagementOutputReference", version: "14.13.1" };
function containerAwsNodePoolMaxPodsConstraintToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_pods_per_node: cdktf.numberToTerraform(struct.maxPodsPerNode),
    };
}
exports.containerAwsNodePoolMaxPodsConstraintToTerraform = containerAwsNodePoolMaxPodsConstraintToTerraform;
function containerAwsNodePoolMaxPodsConstraintToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_pods_per_node: {
            value: cdktf.numberToHclTerraform(struct.maxPodsPerNode),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolMaxPodsConstraintToHclTerraform = containerAwsNodePoolMaxPodsConstraintToHclTerraform;
class ContainerAwsNodePoolMaxPodsConstraintOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxPodsPerNode !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxPodsPerNode = this._maxPodsPerNode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxPodsPerNode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxPodsPerNode = value.maxPodsPerNode;
        }
    }
    get maxPodsPerNode() {
        return this.getNumberAttribute('max_pods_per_node');
    }
    set maxPodsPerNode(value) {
        this._maxPodsPerNode = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxPodsPerNodeInput() {
        return this._maxPodsPerNode;
    }
}
exports.ContainerAwsNodePoolMaxPodsConstraintOutputReference = ContainerAwsNodePoolMaxPodsConstraintOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolMaxPodsConstraintOutputReference[_m] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolMaxPodsConstraintOutputReference", version: "14.13.1" };
function containerAwsNodePoolTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.containerAwsNodePoolTimeoutsToTerraform = containerAwsNodePoolTimeoutsToTerraform;
function containerAwsNodePoolTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolTimeoutsToHclTerraform = containerAwsNodePoolTimeoutsToHclTerraform;
class ContainerAwsNodePoolTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ContainerAwsNodePoolTimeoutsOutputReference = ContainerAwsNodePoolTimeoutsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolTimeoutsOutputReference[_o] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolTimeoutsOutputReference", version: "14.13.1" };
function containerAwsNodePoolUpdateSettingsSurgeSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_surge: cdktf.numberToTerraform(struct.maxSurge),
        max_unavailable: cdktf.numberToTerraform(struct.maxUnavailable),
    };
}
exports.containerAwsNodePoolUpdateSettingsSurgeSettingsToTerraform = containerAwsNodePoolUpdateSettingsSurgeSettingsToTerraform;
function containerAwsNodePoolUpdateSettingsSurgeSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_surge: {
            value: cdktf.numberToHclTerraform(struct.maxSurge),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        max_unavailable: {
            value: cdktf.numberToHclTerraform(struct.maxUnavailable),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolUpdateSettingsSurgeSettingsToHclTerraform = containerAwsNodePoolUpdateSettingsSurgeSettingsToHclTerraform;
class ContainerAwsNodePoolUpdateSettingsSurgeSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxSurge !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxSurge = this._maxSurge;
        }
        if (this._maxUnavailable !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxUnavailable = this._maxUnavailable;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxSurge = undefined;
            this._maxUnavailable = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxSurge = value.maxSurge;
            this._maxUnavailable = value.maxUnavailable;
        }
    }
    get maxSurge() {
        return this.getNumberAttribute('max_surge');
    }
    set maxSurge(value) {
        this._maxSurge = value;
    }
    resetMaxSurge() {
        this._maxSurge = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxSurgeInput() {
        return this._maxSurge;
    }
    get maxUnavailable() {
        return this.getNumberAttribute('max_unavailable');
    }
    set maxUnavailable(value) {
        this._maxUnavailable = value;
    }
    resetMaxUnavailable() {
        this._maxUnavailable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUnavailableInput() {
        return this._maxUnavailable;
    }
}
exports.ContainerAwsNodePoolUpdateSettingsSurgeSettingsOutputReference = ContainerAwsNodePoolUpdateSettingsSurgeSettingsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolUpdateSettingsSurgeSettingsOutputReference[_p] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolUpdateSettingsSurgeSettingsOutputReference", version: "14.13.1" };
function containerAwsNodePoolUpdateSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        surge_settings: containerAwsNodePoolUpdateSettingsSurgeSettingsToTerraform(struct.surgeSettings),
    };
}
exports.containerAwsNodePoolUpdateSettingsToTerraform = containerAwsNodePoolUpdateSettingsToTerraform;
function containerAwsNodePoolUpdateSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        surge_settings: {
            value: containerAwsNodePoolUpdateSettingsSurgeSettingsToHclTerraform(struct.surgeSettings),
            isBlock: true,
            type: "list",
            storageClassType: "ContainerAwsNodePoolUpdateSettingsSurgeSettingsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.containerAwsNodePoolUpdateSettingsToHclTerraform = containerAwsNodePoolUpdateSettingsToHclTerraform;
class ContainerAwsNodePoolUpdateSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // surge_settings - computed: false, optional: true, required: false
        this._surgeSettings = new ContainerAwsNodePoolUpdateSettingsSurgeSettingsOutputReference(this, "surge_settings");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._surgeSettings?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.surgeSettings = this._surgeSettings?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._surgeSettings.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._surgeSettings.internalValue = value.surgeSettings;
        }
    }
    get surgeSettings() {
        return this._surgeSettings;
    }
    putSurgeSettings(value) {
        this._surgeSettings.internalValue = value;
    }
    resetSurgeSettings() {
        this._surgeSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get surgeSettingsInput() {
        return this._surgeSettings.internalValue;
    }
}
exports.ContainerAwsNodePoolUpdateSettingsOutputReference = ContainerAwsNodePoolUpdateSettingsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePoolUpdateSettingsOutputReference[_q] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePoolUpdateSettingsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/container_aws_node_pool google_container_aws_node_pool}
*/
class ContainerAwsNodePool extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ContainerAwsNodePool resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ContainerAwsNodePool to import
    * @param importFromId The id of the existing ContainerAwsNodePool that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/container_aws_node_pool#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ContainerAwsNodePool to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_container_aws_node_pool", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/container_aws_node_pool google_container_aws_node_pool} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ContainerAwsNodePoolConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_container_aws_node_pool',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_annotations - computed: true, optional: false, required: false
        this._effectiveAnnotations = new cdktf.StringMap(this, "effective_annotations");
        // autoscaling - computed: false, optional: false, required: true
        this._autoscaling = new ContainerAwsNodePoolAutoscalingOutputReference(this, "autoscaling");
        // config - computed: false, optional: false, required: true
        this._config = new ContainerAwsNodePoolConfigAOutputReference(this, "config");
        // kubelet_config - computed: false, optional: true, required: false
        this._kubeletConfig = new ContainerAwsNodePoolKubeletConfigOutputReference(this, "kubelet_config");
        // management - computed: false, optional: true, required: false
        this._management = new ContainerAwsNodePoolManagementOutputReference(this, "management");
        // max_pods_constraint - computed: false, optional: false, required: true
        this._maxPodsConstraint = new ContainerAwsNodePoolMaxPodsConstraintOutputReference(this, "max_pods_constraint");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ContainerAwsNodePoolTimeoutsOutputReference(this, "timeouts");
        // update_settings - computed: false, optional: true, required: false
        this._updateSettings = new ContainerAwsNodePoolUpdateSettingsOutputReference(this, "update_settings");
        this._annotations = config.annotations;
        this._cluster = config.cluster;
        this._id = config.id;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._subnetId = config.subnetId;
        this._version = config.version;
        this._autoscaling.internalValue = config.autoscaling;
        this._config.internalValue = config.config;
        this._kubeletConfig.internalValue = config.kubeletConfig;
        this._management.internalValue = config.management;
        this._maxPodsConstraint.internalValue = config.maxPodsConstraint;
        this._timeouts.internalValue = config.timeouts;
        this._updateSettings.internalValue = config.updateSettings;
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    set cluster(value) {
        this._cluster = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterInput() {
        return this._cluster;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get effectiveAnnotations() {
        return this._effectiveAnnotations;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // reconciling - computed: true, optional: false, required: false
    get reconciling() {
        return this.getBooleanAttribute('reconciling');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get autoscaling() {
        return this._autoscaling;
    }
    putAutoscaling(value) {
        this._autoscaling.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingInput() {
        return this._autoscaling.internalValue;
    }
    get config() {
        return this._config;
    }
    putConfig(value) {
        this._config.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get configInput() {
        return this._config.internalValue;
    }
    get kubeletConfig() {
        return this._kubeletConfig;
    }
    putKubeletConfig(value) {
        this._kubeletConfig.internalValue = value;
    }
    resetKubeletConfig() {
        this._kubeletConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kubeletConfigInput() {
        return this._kubeletConfig.internalValue;
    }
    get management() {
        return this._management;
    }
    putManagement(value) {
        this._management.internalValue = value;
    }
    resetManagement() {
        this._management.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get managementInput() {
        return this._management.internalValue;
    }
    get maxPodsConstraint() {
        return this._maxPodsConstraint;
    }
    putMaxPodsConstraint(value) {
        this._maxPodsConstraint.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxPodsConstraintInput() {
        return this._maxPodsConstraint.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get updateSettings() {
        return this._updateSettings;
    }
    putUpdateSettings(value) {
        this._updateSettings.internalValue = value;
    }
    resetUpdateSettings() {
        this._updateSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateSettingsInput() {
        return this._updateSettings.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            annotations: cdktf.hashMapper(cdktf.stringToTerraform)(this._annotations),
            cluster: cdktf.stringToTerraform(this._cluster),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            subnet_id: cdktf.stringToTerraform(this._subnetId),
            version: cdktf.stringToTerraform(this._version),
            autoscaling: containerAwsNodePoolAutoscalingToTerraform(this._autoscaling.internalValue),
            config: containerAwsNodePoolConfigAToTerraform(this._config.internalValue),
            kubelet_config: containerAwsNodePoolKubeletConfigToTerraform(this._kubeletConfig.internalValue),
            management: containerAwsNodePoolManagementToTerraform(this._management.internalValue),
            max_pods_constraint: containerAwsNodePoolMaxPodsConstraintToTerraform(this._maxPodsConstraint.internalValue),
            timeouts: containerAwsNodePoolTimeoutsToTerraform(this._timeouts.internalValue),
            update_settings: containerAwsNodePoolUpdateSettingsToTerraform(this._updateSettings.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            annotations: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._annotations),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            cluster: {
                value: cdktf.stringToHclTerraform(this._cluster),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            subnet_id: {
                value: cdktf.stringToHclTerraform(this._subnetId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            version: {
                value: cdktf.stringToHclTerraform(this._version),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            autoscaling: {
                value: containerAwsNodePoolAutoscalingToHclTerraform(this._autoscaling.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ContainerAwsNodePoolAutoscalingList",
            },
            config: {
                value: containerAwsNodePoolConfigAToHclTerraform(this._config.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ContainerAwsNodePoolConfigAList",
            },
            kubelet_config: {
                value: containerAwsNodePoolKubeletConfigToHclTerraform(this._kubeletConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ContainerAwsNodePoolKubeletConfigList",
            },
            management: {
                value: containerAwsNodePoolManagementToHclTerraform(this._management.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ContainerAwsNodePoolManagementList",
            },
            max_pods_constraint: {
                value: containerAwsNodePoolMaxPodsConstraintToHclTerraform(this._maxPodsConstraint.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ContainerAwsNodePoolMaxPodsConstraintList",
            },
            timeouts: {
                value: containerAwsNodePoolTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ContainerAwsNodePoolTimeouts",
            },
            update_settings: {
                value: containerAwsNodePoolUpdateSettingsToHclTerraform(this._updateSettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ContainerAwsNodePoolUpdateSettingsList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ContainerAwsNodePool = ContainerAwsNodePool;
_r = JSII_RTTI_SYMBOL_1;
ContainerAwsNodePool[_r] = { fqn: "@cdktf/provider-google.containerAwsNodePool.ContainerAwsNodePool", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ContainerAwsNodePool.tfResourceType = "google_container_aws_node_pool";
