// @cdktf/provider-google/lib/network-security-gateway-security-policy-rule/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkSecurityGatewaySecurityPolicyRule = exports.NetworkSecurityGatewaySecurityPolicyRuleTimeoutsOutputReference = exports.networkSecurityGatewaySecurityPolicyRuleTimeoutsToHclTerraform = exports.networkSecurityGatewaySecurityPolicyRuleTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkSecurityGatewaySecurityPolicyRuleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkSecurityGatewaySecurityPolicyRuleTimeoutsToTerraform = networkSecurityGatewaySecurityPolicyRuleTimeoutsToTerraform;
function networkSecurityGatewaySecurityPolicyRuleTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityGatewaySecurityPolicyRuleTimeoutsToHclTerraform = networkSecurityGatewaySecurityPolicyRuleTimeoutsToHclTerraform;
class NetworkSecurityGatewaySecurityPolicyRuleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkSecurityGatewaySecurityPolicyRuleTimeoutsOutputReference = NetworkSecurityGatewaySecurityPolicyRuleTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkSecurityGatewaySecurityPolicyRuleTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-google.networkSecurityGatewaySecurityPolicyRule.NetworkSecurityGatewaySecurityPolicyRuleTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_gateway_security_policy_rule google_network_security_gateway_security_policy_rule}
*/
class NetworkSecurityGatewaySecurityPolicyRule extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkSecurityGatewaySecurityPolicyRule resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkSecurityGatewaySecurityPolicyRule to import
    * @param importFromId The id of the existing NetworkSecurityGatewaySecurityPolicyRule that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_gateway_security_policy_rule#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkSecurityGatewaySecurityPolicyRule to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_network_security_gateway_security_policy_rule", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_gateway_security_policy_rule google_network_security_gateway_security_policy_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkSecurityGatewaySecurityPolicyRuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_network_security_gateway_security_policy_rule',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkSecurityGatewaySecurityPolicyRuleTimeoutsOutputReference(this, "timeouts");
        this._applicationMatcher = config.applicationMatcher;
        this._basicProfile = config.basicProfile;
        this._description = config.description;
        this._enabled = config.enabled;
        this._gatewaySecurityPolicy = config.gatewaySecurityPolicy;
        this._id = config.id;
        this._location = config.location;
        this._name = config.name;
        this._priority = config.priority;
        this._project = config.project;
        this._sessionMatcher = config.sessionMatcher;
        this._tlsInspectionEnabled = config.tlsInspectionEnabled;
        this._timeouts.internalValue = config.timeouts;
    }
    get applicationMatcher() {
        return this.getStringAttribute('application_matcher');
    }
    set applicationMatcher(value) {
        this._applicationMatcher = value;
    }
    resetApplicationMatcher() {
        this._applicationMatcher = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applicationMatcherInput() {
        return this._applicationMatcher;
    }
    get basicProfile() {
        return this.getStringAttribute('basic_profile');
    }
    set basicProfile(value) {
        this._basicProfile = value;
    }
    // Temporarily expose input value. Use with caution.
    get basicProfileInput() {
        return this._basicProfile;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get gatewaySecurityPolicy() {
        return this.getStringAttribute('gateway_security_policy');
    }
    set gatewaySecurityPolicy(value) {
        this._gatewaySecurityPolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    get gatewaySecurityPolicyInput() {
        return this._gatewaySecurityPolicy;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get sessionMatcher() {
        return this.getStringAttribute('session_matcher');
    }
    set sessionMatcher(value) {
        this._sessionMatcher = value;
    }
    // Temporarily expose input value. Use with caution.
    get sessionMatcherInput() {
        return this._sessionMatcher;
    }
    get tlsInspectionEnabled() {
        return this.getBooleanAttribute('tls_inspection_enabled');
    }
    set tlsInspectionEnabled(value) {
        this._tlsInspectionEnabled = value;
    }
    resetTlsInspectionEnabled() {
        this._tlsInspectionEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsInspectionEnabledInput() {
        return this._tlsInspectionEnabled;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_matcher: cdktf.stringToTerraform(this._applicationMatcher),
            basic_profile: cdktf.stringToTerraform(this._basicProfile),
            description: cdktf.stringToTerraform(this._description),
            enabled: cdktf.booleanToTerraform(this._enabled),
            gateway_security_policy: cdktf.stringToTerraform(this._gatewaySecurityPolicy),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            priority: cdktf.numberToTerraform(this._priority),
            project: cdktf.stringToTerraform(this._project),
            session_matcher: cdktf.stringToTerraform(this._sessionMatcher),
            tls_inspection_enabled: cdktf.booleanToTerraform(this._tlsInspectionEnabled),
            timeouts: networkSecurityGatewaySecurityPolicyRuleTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_matcher: {
                value: cdktf.stringToHclTerraform(this._applicationMatcher),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            basic_profile: {
                value: cdktf.stringToHclTerraform(this._basicProfile),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            enabled: {
                value: cdktf.booleanToHclTerraform(this._enabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            gateway_security_policy: {
                value: cdktf.stringToHclTerraform(this._gatewaySecurityPolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            priority: {
                value: cdktf.numberToHclTerraform(this._priority),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            session_matcher: {
                value: cdktf.stringToHclTerraform(this._sessionMatcher),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tls_inspection_enabled: {
                value: cdktf.booleanToHclTerraform(this._tlsInspectionEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            timeouts: {
                value: networkSecurityGatewaySecurityPolicyRuleTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkSecurityGatewaySecurityPolicyRuleTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkSecurityGatewaySecurityPolicyRule = NetworkSecurityGatewaySecurityPolicyRule;
_b = JSII_RTTI_SYMBOL_1;
NetworkSecurityGatewaySecurityPolicyRule[_b] = { fqn: "@cdktf/provider-google.networkSecurityGatewaySecurityPolicyRule.NetworkSecurityGatewaySecurityPolicyRule", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetworkSecurityGatewaySecurityPolicyRule.tfResourceType = "google_network_security_gateway_security_policy_rule";
