// @cdktf/hcl2cdk/lib/function-bindings/functions.js
"use strict";
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.tsFunctionsMap = exports.functionsMap = void 0;
const functions_generated_1 = require("./functions.generated");
exports.functionsMap = {
    ...functions_generated_1.functionsMapGenerated,
    bcrypt: {
        name: "bcrypt",
        returnType: "string",
        parameters: [{ type: "string" }, { type: "number", optional: true }],
    },
    join: {
        name: "join",
        returnType: "string",
        parameters: [{ type: "string" }, { type: ["list", "string"] }],
        /**
         * Terraform supports join(separator, listA, listB)
         * wheras CDKTF only supports join(separator, list) (to make it simpler to use as JSII does not support variadic parameters)
         * and we'd need to convert this to join(separator, concat(listA, listB)) if multiple variadic args are passed
         */
        transformer: (fc) => {
            if (fc.children.length <= 2) {
                return fc; // just one child -> nothing to do
            }
            const concatFunction = {
                type: "function",
                meta: {
                    name: "concat",
                    expandedFinalArgument: fc.meta.expandedFinalArgument,
                    nameRange: {},
                    openParenRange: {},
                    closeParenRange: {},
                    argsRanges: [],
                },
                children: fc.children.slice(1),
            };
            return {
                type: "function",
                meta: {
                    name: "join",
                },
                children: [
                    fc.children[0],
                    concatFunction,
                ],
            };
        },
    },
    range: {
        name: "range",
        returnType: ["list", "string"],
        parameters: [
            { type: "number" },
            { type: "number" },
            { type: "number", optional: true },
        ],
    },
    lookup: {
        name: "lookup",
        returnType: "dynamic",
        parameters: [{ type: "dynamic" }, { type: "string" }, { type: "dynamic" }],
    },
};
exports.tsFunctionsMap = Object.entries(exports.functionsMap).reduce((acc, [name, fnInfo]) => {
    // Swap the names around
    acc[fnInfo.name] = { ...fnInfo, name: name };
    return acc;
}, {});
