// @cdktf/provider-google/lib/network-security-tls-inspection-policy/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkSecurityTlsInspectionPolicy = exports.NetworkSecurityTlsInspectionPolicyTimeoutsOutputReference = exports.networkSecurityTlsInspectionPolicyTimeoutsToHclTerraform = exports.networkSecurityTlsInspectionPolicyTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkSecurityTlsInspectionPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkSecurityTlsInspectionPolicyTimeoutsToTerraform = networkSecurityTlsInspectionPolicyTimeoutsToTerraform;
function networkSecurityTlsInspectionPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityTlsInspectionPolicyTimeoutsToHclTerraform = networkSecurityTlsInspectionPolicyTimeoutsToHclTerraform;
class NetworkSecurityTlsInspectionPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkSecurityTlsInspectionPolicyTimeoutsOutputReference = NetworkSecurityTlsInspectionPolicyTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkSecurityTlsInspectionPolicyTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-google.networkSecurityTlsInspectionPolicy.NetworkSecurityTlsInspectionPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_tls_inspection_policy google_network_security_tls_inspection_policy}
*/
class NetworkSecurityTlsInspectionPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkSecurityTlsInspectionPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkSecurityTlsInspectionPolicy to import
    * @param importFromId The id of the existing NetworkSecurityTlsInspectionPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_tls_inspection_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkSecurityTlsInspectionPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_network_security_tls_inspection_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_tls_inspection_policy google_network_security_tls_inspection_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkSecurityTlsInspectionPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_network_security_tls_inspection_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkSecurityTlsInspectionPolicyTimeoutsOutputReference(this, "timeouts");
        this._caPool = config.caPool;
        this._customTlsFeatures = config.customTlsFeatures;
        this._description = config.description;
        this._excludePublicCaSet = config.excludePublicCaSet;
        this._id = config.id;
        this._location = config.location;
        this._minTlsVersion = config.minTlsVersion;
        this._name = config.name;
        this._project = config.project;
        this._tlsFeatureProfile = config.tlsFeatureProfile;
        this._trustConfig = config.trustConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    get caPool() {
        return this.getStringAttribute('ca_pool');
    }
    set caPool(value) {
        this._caPool = value;
    }
    // Temporarily expose input value. Use with caution.
    get caPoolInput() {
        return this._caPool;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get customTlsFeatures() {
        return this.getListAttribute('custom_tls_features');
    }
    set customTlsFeatures(value) {
        this._customTlsFeatures = value;
    }
    resetCustomTlsFeatures() {
        this._customTlsFeatures = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customTlsFeaturesInput() {
        return this._customTlsFeatures;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get excludePublicCaSet() {
        return this.getBooleanAttribute('exclude_public_ca_set');
    }
    set excludePublicCaSet(value) {
        this._excludePublicCaSet = value;
    }
    resetExcludePublicCaSet() {
        this._excludePublicCaSet = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludePublicCaSetInput() {
        return this._excludePublicCaSet;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get minTlsVersion() {
        return this.getStringAttribute('min_tls_version');
    }
    set minTlsVersion(value) {
        this._minTlsVersion = value;
    }
    resetMinTlsVersion() {
        this._minTlsVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minTlsVersionInput() {
        return this._minTlsVersion;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get tlsFeatureProfile() {
        return this.getStringAttribute('tls_feature_profile');
    }
    set tlsFeatureProfile(value) {
        this._tlsFeatureProfile = value;
    }
    resetTlsFeatureProfile() {
        this._tlsFeatureProfile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsFeatureProfileInput() {
        return this._tlsFeatureProfile;
    }
    get trustConfig() {
        return this.getStringAttribute('trust_config');
    }
    set trustConfig(value) {
        this._trustConfig = value;
    }
    resetTrustConfig() {
        this._trustConfig = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trustConfigInput() {
        return this._trustConfig;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            ca_pool: cdktf.stringToTerraform(this._caPool),
            custom_tls_features: cdktf.listMapper(cdktf.stringToTerraform, false)(this._customTlsFeatures),
            description: cdktf.stringToTerraform(this._description),
            exclude_public_ca_set: cdktf.booleanToTerraform(this._excludePublicCaSet),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            min_tls_version: cdktf.stringToTerraform(this._minTlsVersion),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            tls_feature_profile: cdktf.stringToTerraform(this._tlsFeatureProfile),
            trust_config: cdktf.stringToTerraform(this._trustConfig),
            timeouts: networkSecurityTlsInspectionPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            ca_pool: {
                value: cdktf.stringToHclTerraform(this._caPool),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            custom_tls_features: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._customTlsFeatures),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            exclude_public_ca_set: {
                value: cdktf.booleanToHclTerraform(this._excludePublicCaSet),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            min_tls_version: {
                value: cdktf.stringToHclTerraform(this._minTlsVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tls_feature_profile: {
                value: cdktf.stringToHclTerraform(this._tlsFeatureProfile),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trust_config: {
                value: cdktf.stringToHclTerraform(this._trustConfig),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: networkSecurityTlsInspectionPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkSecurityTlsInspectionPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkSecurityTlsInspectionPolicy = NetworkSecurityTlsInspectionPolicy;
_b = JSII_RTTI_SYMBOL_1;
NetworkSecurityTlsInspectionPolicy[_b] = { fqn: "@cdktf/provider-google.networkSecurityTlsInspectionPolicy.NetworkSecurityTlsInspectionPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetworkSecurityTlsInspectionPolicy.tfResourceType = "google_network_security_tls_inspection_policy";
