// @cdktf/provider-google/lib/data-google-compute-instance-template/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToHclTerraform = exports.dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkPerformanceConfigList = exports.DataGoogleComputeInstanceTemplateNetworkPerformanceConfigOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkPerformanceConfigToHclTerraform = exports.dataGoogleComputeInstanceTemplateNetworkPerformanceConfigToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceList = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceToHclTerraform = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToHclTerraform = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToHclTerraform = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList = exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToHclTerraform = exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToTerraform = exports.DataGoogleComputeInstanceTemplateGuestAcceleratorList = exports.DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference = exports.dataGoogleComputeInstanceTemplateGuestAcceleratorToHclTerraform = exports.dataGoogleComputeInstanceTemplateGuestAcceleratorToTerraform = exports.DataGoogleComputeInstanceTemplateDiskList = exports.DataGoogleComputeInstanceTemplateDiskOutputReference = exports.dataGoogleComputeInstanceTemplateDiskToHclTerraform = exports.dataGoogleComputeInstanceTemplateDiskToTerraform = exports.DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyList = exports.DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference = exports.dataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyToHclTerraform = exports.dataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyToTerraform = exports.DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyList = exports.DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyOutputReference = exports.dataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyToHclTerraform = exports.dataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyToTerraform = exports.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList = exports.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference = exports.dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToHclTerraform = exports.dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToTerraform = exports.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList = exports.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference = exports.dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToHclTerraform = exports.dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToTerraform = exports.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList = exports.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference = exports.dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToHclTerraform = exports.dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToTerraform = void 0;
exports.DataGoogleComputeInstanceTemplate = exports.DataGoogleComputeInstanceTemplateShieldedInstanceConfigList = exports.DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference = exports.dataGoogleComputeInstanceTemplateShieldedInstanceConfigToHclTerraform = exports.dataGoogleComputeInstanceTemplateShieldedInstanceConfigToTerraform = exports.DataGoogleComputeInstanceTemplateServiceAccountList = exports.DataGoogleComputeInstanceTemplateServiceAccountOutputReference = exports.dataGoogleComputeInstanceTemplateServiceAccountToHclTerraform = exports.dataGoogleComputeInstanceTemplateServiceAccountToTerraform = exports.DataGoogleComputeInstanceTemplateSchedulingList = exports.DataGoogleComputeInstanceTemplateSchedulingOutputReference = exports.dataGoogleComputeInstanceTemplateSchedulingToHclTerraform = exports.dataGoogleComputeInstanceTemplateSchedulingToTerraform = exports.DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionList = exports.DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionOutputReference = exports.dataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionToHclTerraform = exports.dataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionToTerraform = exports.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList = exports.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference = exports.dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToHclTerraform = exports.dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToTerraform = exports.DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationList = exports.DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationOutputReference = exports.dataGoogleComputeInstanceTemplateSchedulingMaxRunDurationToHclTerraform = exports.dataGoogleComputeInstanceTemplateSchedulingMaxRunDurationToTerraform = exports.DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList = exports.DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference = exports.dataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToHclTerraform = exports.dataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToTerraform = exports.DataGoogleComputeInstanceTemplateReservationAffinityList = exports.DataGoogleComputeInstanceTemplateReservationAffinityOutputReference = exports.dataGoogleComputeInstanceTemplateReservationAffinityToHclTerraform = exports.dataGoogleComputeInstanceTemplateReservationAffinityToTerraform = exports.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList = exports.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToTerraform = dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToTerraform;
function dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToHclTerraform = dataGoogleComputeInstanceTemplateAdvancedMachineFeaturesToHclTerraform;
class DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_nested_virtualization - computed: true, optional: false, required: false
    get enableNestedVirtualization() {
        return this.getBooleanAttribute('enable_nested_virtualization');
    }
    // enable_uefi_networking - computed: true, optional: false, required: false
    get enableUefiNetworking() {
        return this.getBooleanAttribute('enable_uefi_networking');
    }
    // performance_monitoring_unit - computed: true, optional: false, required: false
    get performanceMonitoringUnit() {
        return this.getStringAttribute('performance_monitoring_unit');
    }
    // threads_per_core - computed: true, optional: false, required: false
    get threadsPerCore() {
        return this.getNumberAttribute('threads_per_core');
    }
    // turbo_mode - computed: true, optional: false, required: false
    get turboMode() {
        return this.getStringAttribute('turbo_mode');
    }
    // visible_core_count - computed: true, optional: false, required: false
    get visibleCoreCount() {
        return this.getNumberAttribute('visible_core_count');
    }
}
exports.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference = DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList = DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToTerraform = dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToTerraform;
function dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToHclTerraform = dataGoogleComputeInstanceTemplateConfidentialInstanceConfigToHclTerraform;
class DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // confidential_instance_type - computed: true, optional: false, required: false
    get confidentialInstanceType() {
        return this.getStringAttribute('confidential_instance_type');
    }
    // enable_confidential_compute - computed: true, optional: false, required: false
    get enableConfidentialCompute() {
        return this.getBooleanAttribute('enable_confidential_compute');
    }
}
exports.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference = DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateConfidentialInstanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList = DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToTerraform = dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToTerraform;
function dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToHclTerraform = dataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyToHclTerraform;
class DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
}
exports.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference = DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList = DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyToTerraform = dataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyToTerraform;
function dataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyToHclTerraform = dataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyToHclTerraform;
class DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // kms_key_service_account - computed: true, optional: false, required: false
    get kmsKeyServiceAccount() {
        return this.getStringAttribute('kms_key_service_account');
    }
}
exports.DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyOutputReference = DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyList = DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyToTerraform = dataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyToTerraform;
function dataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyToHclTerraform = dataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyToHclTerraform;
class DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // kms_key_service_account - computed: true, optional: false, required: false
    get kmsKeyServiceAccount() {
        return this.getStringAttribute('kms_key_service_account');
    }
}
exports.DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference = DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyList = DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateDiskToTerraform = dataGoogleComputeInstanceTemplateDiskToTerraform;
function dataGoogleComputeInstanceTemplateDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateDiskToHclTerraform = dataGoogleComputeInstanceTemplateDiskToHclTerraform;
class DataGoogleComputeInstanceTemplateDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // disk_encryption_key - computed: true, optional: false, required: false
        this._diskEncryptionKey = new DataGoogleComputeInstanceTemplateDiskDiskEncryptionKeyList(this, "disk_encryption_key", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // resource_manager_tags - computed: true, optional: false, required: false
        this._resourceManagerTags = new cdktf.StringMap(this, "resource_manager_tags");
        // source_image_encryption_key - computed: true, optional: false, required: false
        this._sourceImageEncryptionKey = new DataGoogleComputeInstanceTemplateDiskSourceImageEncryptionKeyList(this, "source_image_encryption_key", false);
        // source_snapshot_encryption_key - computed: true, optional: false, required: false
        this._sourceSnapshotEncryptionKey = new DataGoogleComputeInstanceTemplateDiskSourceSnapshotEncryptionKeyList(this, "source_snapshot_encryption_key", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // auto_delete - computed: true, optional: false, required: false
    get autoDelete() {
        return this.getBooleanAttribute('auto_delete');
    }
    // boot - computed: true, optional: false, required: false
    get boot() {
        return this.getBooleanAttribute('boot');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    get diskEncryptionKey() {
        return this._diskEncryptionKey;
    }
    // disk_name - computed: true, optional: false, required: false
    get diskName() {
        return this.getStringAttribute('disk_name');
    }
    // disk_size_gb - computed: true, optional: false, required: false
    get diskSizeGb() {
        return this.getNumberAttribute('disk_size_gb');
    }
    // disk_type - computed: true, optional: false, required: false
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    // interface - computed: true, optional: false, required: false
    get interface() {
        return this.getStringAttribute('interface');
    }
    get labels() {
        return this._labels;
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
    // provisioned_iops - computed: true, optional: false, required: false
    get provisionedIops() {
        return this.getNumberAttribute('provisioned_iops');
    }
    // provisioned_throughput - computed: true, optional: false, required: false
    get provisionedThroughput() {
        return this.getNumberAttribute('provisioned_throughput');
    }
    get resourceManagerTags() {
        return this._resourceManagerTags;
    }
    // resource_policies - computed: true, optional: false, required: false
    get resourcePolicies() {
        return this.getListAttribute('resource_policies');
    }
    // source - computed: true, optional: false, required: false
    get source() {
        return this.getStringAttribute('source');
    }
    // source_image - computed: true, optional: false, required: false
    get sourceImage() {
        return this.getStringAttribute('source_image');
    }
    get sourceImageEncryptionKey() {
        return this._sourceImageEncryptionKey;
    }
    // source_snapshot - computed: true, optional: false, required: false
    get sourceSnapshot() {
        return this.getStringAttribute('source_snapshot');
    }
    get sourceSnapshotEncryptionKey() {
        return this._sourceSnapshotEncryptionKey;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceTemplateDiskOutputReference = DataGoogleComputeInstanceTemplateDiskOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateDiskList = DataGoogleComputeInstanceTemplateDiskList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateDiskList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateDiskList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateGuestAcceleratorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateGuestAcceleratorToTerraform = dataGoogleComputeInstanceTemplateGuestAcceleratorToTerraform;
function dataGoogleComputeInstanceTemplateGuestAcceleratorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateGuestAcceleratorToHclTerraform = dataGoogleComputeInstanceTemplateGuestAcceleratorToHclTerraform;
class DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // count - computed: true, optional: false, required: false
    get count() {
        return this.getNumberAttribute('count');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference = DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateGuestAcceleratorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateGuestAcceleratorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateGuestAcceleratorList = DataGoogleComputeInstanceTemplateGuestAcceleratorList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateGuestAcceleratorList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateGuestAcceleratorList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToTerraform;
function dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToHclTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigToHclTerraform;
class DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // nat_ip - computed: true, optional: false, required: false
    get natIp() {
        return this.getStringAttribute('nat_ip');
    }
    // network_tier - computed: true, optional: false, required: false
    get networkTier() {
        return this.getStringAttribute('network_tier');
    }
    // public_ptr_domain_name - computed: true, optional: false, required: false
    get publicPtrDomainName() {
        return this.getStringAttribute('public_ptr_domain_name');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference = DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList = DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList[_r] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform;
function dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToHclTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeToHclTerraform;
class DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_cidr_range - computed: true, optional: false, required: false
    get ipCidrRange() {
        return this.getStringAttribute('ip_cidr_range');
    }
    // subnetwork_range_name - computed: true, optional: false, required: false
    get subnetworkRangeName() {
        return this.getStringAttribute('subnetwork_range_name');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference = DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList = DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList[_t] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform;
function dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToHclTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigToHclTerraform;
class DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // external_ipv6 - computed: true, optional: false, required: false
    get externalIpv6() {
        return this.getStringAttribute('external_ipv6');
    }
    // external_ipv6_prefix_length - computed: true, optional: false, required: false
    get externalIpv6PrefixLength() {
        return this.getStringAttribute('external_ipv6_prefix_length');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network_tier - computed: true, optional: false, required: false
    get networkTier() {
        return this.getStringAttribute('network_tier');
    }
    // public_ptr_domain_name - computed: true, optional: false, required: false
    get publicPtrDomainName() {
        return this.getStringAttribute('public_ptr_domain_name');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference = DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList = DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList[_v] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateNetworkInterfaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceToTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceToTerraform;
function dataGoogleComputeInstanceTemplateNetworkInterfaceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateNetworkInterfaceToHclTerraform = dataGoogleComputeInstanceTemplateNetworkInterfaceToHclTerraform;
class DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // access_config - computed: true, optional: false, required: false
        this._accessConfig = new DataGoogleComputeInstanceTemplateNetworkInterfaceAccessConfigList(this, "access_config", false);
        // alias_ip_range - computed: true, optional: false, required: false
        this._aliasIpRange = new DataGoogleComputeInstanceTemplateNetworkInterfaceAliasIpRangeList(this, "alias_ip_range", false);
        // ipv6_access_config - computed: true, optional: false, required: false
        this._ipv6AccessConfig = new DataGoogleComputeInstanceTemplateNetworkInterfaceIpv6AccessConfigList(this, "ipv6_access_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get accessConfig() {
        return this._accessConfig;
    }
    get aliasIpRange() {
        return this._aliasIpRange;
    }
    // internal_ipv6_prefix_length - computed: true, optional: false, required: false
    get internalIpv6PrefixLength() {
        return this.getNumberAttribute('internal_ipv6_prefix_length');
    }
    get ipv6AccessConfig() {
        return this._ipv6AccessConfig;
    }
    // ipv6_access_type - computed: true, optional: false, required: false
    get ipv6AccessType() {
        return this.getStringAttribute('ipv6_access_type');
    }
    // ipv6_address - computed: true, optional: false, required: false
    get ipv6Address() {
        return this.getStringAttribute('ipv6_address');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    // network_ip - computed: true, optional: false, required: false
    get networkIp() {
        return this.getStringAttribute('network_ip');
    }
    // nic_type - computed: true, optional: false, required: false
    get nicType() {
        return this.getStringAttribute('nic_type');
    }
    // queue_count - computed: true, optional: false, required: false
    get queueCount() {
        return this.getNumberAttribute('queue_count');
    }
    // stack_type - computed: true, optional: false, required: false
    get stackType() {
        return this.getStringAttribute('stack_type');
    }
    // subnetwork - computed: true, optional: false, required: false
    get subnetwork() {
        return this.getStringAttribute('subnetwork');
    }
    // subnetwork_project - computed: true, optional: false, required: false
    get subnetworkProject() {
        return this.getStringAttribute('subnetwork_project');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference = DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateNetworkInterfaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkInterfaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkInterfaceList = DataGoogleComputeInstanceTemplateNetworkInterfaceList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkInterfaceList[_x] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkInterfaceList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateNetworkPerformanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateNetworkPerformanceConfigToTerraform = dataGoogleComputeInstanceTemplateNetworkPerformanceConfigToTerraform;
function dataGoogleComputeInstanceTemplateNetworkPerformanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateNetworkPerformanceConfigToHclTerraform = dataGoogleComputeInstanceTemplateNetworkPerformanceConfigToHclTerraform;
class DataGoogleComputeInstanceTemplateNetworkPerformanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // total_egress_bandwidth_tier - computed: true, optional: false, required: false
    get totalEgressBandwidthTier() {
        return this.getStringAttribute('total_egress_bandwidth_tier');
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkPerformanceConfigOutputReference = DataGoogleComputeInstanceTemplateNetworkPerformanceConfigOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkPerformanceConfigOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkPerformanceConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateNetworkPerformanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateNetworkPerformanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateNetworkPerformanceConfigList = DataGoogleComputeInstanceTemplateNetworkPerformanceConfigList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateNetworkPerformanceConfigList[_z] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateNetworkPerformanceConfigList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToTerraform = dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToTerraform;
function dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToHclTerraform = dataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationToHclTerraform;
class DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return this.getListAttribute('values');
    }
}
exports.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference = DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList = DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList[_1] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateReservationAffinityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateReservationAffinityToTerraform = dataGoogleComputeInstanceTemplateReservationAffinityToTerraform;
function dataGoogleComputeInstanceTemplateReservationAffinityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateReservationAffinityToHclTerraform = dataGoogleComputeInstanceTemplateReservationAffinityToHclTerraform;
class DataGoogleComputeInstanceTemplateReservationAffinityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // specific_reservation - computed: true, optional: false, required: false
        this._specificReservation = new DataGoogleComputeInstanceTemplateReservationAffinitySpecificReservationList(this, "specific_reservation", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get specificReservation() {
        return this._specificReservation;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceTemplateReservationAffinityOutputReference = DataGoogleComputeInstanceTemplateReservationAffinityOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateReservationAffinityOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateReservationAffinityOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateReservationAffinityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateReservationAffinityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateReservationAffinityList = DataGoogleComputeInstanceTemplateReservationAffinityList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateReservationAffinityList[_3] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateReservationAffinityList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToTerraform = dataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToTerraform;
function dataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToHclTerraform = dataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToHclTerraform;
class DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // nanos - computed: true, optional: false, required: false
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    // seconds - computed: true, optional: false, required: false
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference = DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList = DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList[_5] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateSchedulingMaxRunDurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateSchedulingMaxRunDurationToTerraform = dataGoogleComputeInstanceTemplateSchedulingMaxRunDurationToTerraform;
function dataGoogleComputeInstanceTemplateSchedulingMaxRunDurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateSchedulingMaxRunDurationToHclTerraform = dataGoogleComputeInstanceTemplateSchedulingMaxRunDurationToHclTerraform;
class DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // nanos - computed: true, optional: false, required: false
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    // seconds - computed: true, optional: false, required: false
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationOutputReference = DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationList = DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationList[_7] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToTerraform = dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToTerraform;
function dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToHclTerraform = dataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesToHclTerraform;
class DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // operator - computed: true, optional: false, required: false
    get operator() {
        return this.getStringAttribute('operator');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference = DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList = DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList[_9] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionToTerraform = dataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionToTerraform;
function dataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionToHclTerraform = dataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionToHclTerraform;
class DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // discard_local_ssd - computed: true, optional: false, required: false
    get discardLocalSsd() {
        return this.getBooleanAttribute('discard_local_ssd');
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionOutputReference = DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionOutputReference[_10] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionList = DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionList;
_11 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionList[_11] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateSchedulingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateSchedulingToTerraform = dataGoogleComputeInstanceTemplateSchedulingToTerraform;
function dataGoogleComputeInstanceTemplateSchedulingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateSchedulingToHclTerraform = dataGoogleComputeInstanceTemplateSchedulingToHclTerraform;
class DataGoogleComputeInstanceTemplateSchedulingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // local_ssd_recovery_timeout - computed: true, optional: false, required: false
        this._localSsdRecoveryTimeout = new DataGoogleComputeInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList(this, "local_ssd_recovery_timeout", false);
        // max_run_duration - computed: true, optional: false, required: false
        this._maxRunDuration = new DataGoogleComputeInstanceTemplateSchedulingMaxRunDurationList(this, "max_run_duration", false);
        // node_affinities - computed: true, optional: false, required: false
        this._nodeAffinities = new DataGoogleComputeInstanceTemplateSchedulingNodeAffinitiesList(this, "node_affinities", true);
        // on_instance_stop_action - computed: true, optional: false, required: false
        this._onInstanceStopAction = new DataGoogleComputeInstanceTemplateSchedulingOnInstanceStopActionList(this, "on_instance_stop_action", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // automatic_restart - computed: true, optional: false, required: false
    get automaticRestart() {
        return this.getBooleanAttribute('automatic_restart');
    }
    // instance_termination_action - computed: true, optional: false, required: false
    get instanceTerminationAction() {
        return this.getStringAttribute('instance_termination_action');
    }
    get localSsdRecoveryTimeout() {
        return this._localSsdRecoveryTimeout;
    }
    get maxRunDuration() {
        return this._maxRunDuration;
    }
    // min_node_cpus - computed: true, optional: false, required: false
    get minNodeCpus() {
        return this.getNumberAttribute('min_node_cpus');
    }
    get nodeAffinities() {
        return this._nodeAffinities;
    }
    // on_host_maintenance - computed: true, optional: false, required: false
    get onHostMaintenance() {
        return this.getStringAttribute('on_host_maintenance');
    }
    get onInstanceStopAction() {
        return this._onInstanceStopAction;
    }
    // preemptible - computed: true, optional: false, required: false
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
    // provisioning_model - computed: true, optional: false, required: false
    get provisioningModel() {
        return this.getStringAttribute('provisioning_model');
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingOutputReference = DataGoogleComputeInstanceTemplateSchedulingOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingOutputReference[_12] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateSchedulingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateSchedulingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateSchedulingList = DataGoogleComputeInstanceTemplateSchedulingList;
_13 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateSchedulingList[_13] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateSchedulingList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateServiceAccountToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateServiceAccountToTerraform = dataGoogleComputeInstanceTemplateServiceAccountToTerraform;
function dataGoogleComputeInstanceTemplateServiceAccountToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateServiceAccountToHclTerraform = dataGoogleComputeInstanceTemplateServiceAccountToHclTerraform;
class DataGoogleComputeInstanceTemplateServiceAccountOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    // scopes - computed: true, optional: false, required: false
    get scopes() {
        return cdktf.Fn.tolist(this.getListAttribute('scopes'));
    }
}
exports.DataGoogleComputeInstanceTemplateServiceAccountOutputReference = DataGoogleComputeInstanceTemplateServiceAccountOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateServiceAccountOutputReference[_14] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateServiceAccountOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateServiceAccountList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateServiceAccountOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateServiceAccountList = DataGoogleComputeInstanceTemplateServiceAccountList;
_15 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateServiceAccountList[_15] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateServiceAccountList", version: "14.13.1" };
function dataGoogleComputeInstanceTemplateShieldedInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceTemplateShieldedInstanceConfigToTerraform = dataGoogleComputeInstanceTemplateShieldedInstanceConfigToTerraform;
function dataGoogleComputeInstanceTemplateShieldedInstanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceTemplateShieldedInstanceConfigToHclTerraform = dataGoogleComputeInstanceTemplateShieldedInstanceConfigToHclTerraform;
class DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_integrity_monitoring - computed: true, optional: false, required: false
    get enableIntegrityMonitoring() {
        return this.getBooleanAttribute('enable_integrity_monitoring');
    }
    // enable_secure_boot - computed: true, optional: false, required: false
    get enableSecureBoot() {
        return this.getBooleanAttribute('enable_secure_boot');
    }
    // enable_vtpm - computed: true, optional: false, required: false
    get enableVtpm() {
        return this.getBooleanAttribute('enable_vtpm');
    }
}
exports.DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference = DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference[_16] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceTemplateShieldedInstanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceTemplateShieldedInstanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceTemplateShieldedInstanceConfigList = DataGoogleComputeInstanceTemplateShieldedInstanceConfigList;
_17 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplateShieldedInstanceConfigList[_17] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplateShieldedInstanceConfigList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_instance_template google_compute_instance_template}
*/
class DataGoogleComputeInstanceTemplate extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeInstanceTemplate resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeInstanceTemplate to import
    * @param importFromId The id of the existing DataGoogleComputeInstanceTemplate that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_instance_template#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeInstanceTemplate to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_instance_template", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_instance_template google_compute_instance_template} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeInstanceTemplateConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_instance_template',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // advanced_machine_features - computed: true, optional: false, required: false
        this._advancedMachineFeatures = new DataGoogleComputeInstanceTemplateAdvancedMachineFeaturesList(this, "advanced_machine_features", false);
        // confidential_instance_config - computed: true, optional: false, required: false
        this._confidentialInstanceConfig = new DataGoogleComputeInstanceTemplateConfidentialInstanceConfigList(this, "confidential_instance_config", false);
        // disk - computed: true, optional: false, required: false
        this._disk = new DataGoogleComputeInstanceTemplateDiskList(this, "disk", false);
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // guest_accelerator - computed: true, optional: false, required: false
        this._guestAccelerator = new DataGoogleComputeInstanceTemplateGuestAcceleratorList(this, "guest_accelerator", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // metadata - computed: true, optional: false, required: false
        this._metadata = new cdktf.StringMap(this, "metadata");
        // network_interface - computed: true, optional: false, required: false
        this._networkInterface = new DataGoogleComputeInstanceTemplateNetworkInterfaceList(this, "network_interface", false);
        // network_performance_config - computed: true, optional: false, required: false
        this._networkPerformanceConfig = new DataGoogleComputeInstanceTemplateNetworkPerformanceConfigList(this, "network_performance_config", false);
        // reservation_affinity - computed: true, optional: false, required: false
        this._reservationAffinity = new DataGoogleComputeInstanceTemplateReservationAffinityList(this, "reservation_affinity", false);
        // resource_manager_tags - computed: true, optional: false, required: false
        this._resourceManagerTags = new cdktf.StringMap(this, "resource_manager_tags");
        // scheduling - computed: true, optional: false, required: false
        this._scheduling = new DataGoogleComputeInstanceTemplateSchedulingList(this, "scheduling", false);
        // service_account - computed: true, optional: false, required: false
        this._serviceAccount = new DataGoogleComputeInstanceTemplateServiceAccountList(this, "service_account", false);
        // shielded_instance_config - computed: true, optional: false, required: false
        this._shieldedInstanceConfig = new DataGoogleComputeInstanceTemplateShieldedInstanceConfigList(this, "shielded_instance_config", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._filter = config.filter;
        this._id = config.id;
        this._mostRecent = config.mostRecent;
        this._name = config.name;
        this._project = config.project;
        this._selfLinkUnique = config.selfLinkUnique;
    }
    get advancedMachineFeatures() {
        return this._advancedMachineFeatures;
    }
    // can_ip_forward - computed: true, optional: false, required: false
    get canIpForward() {
        return this.getBooleanAttribute('can_ip_forward');
    }
    get confidentialInstanceConfig() {
        return this._confidentialInstanceConfig;
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get disk() {
        return this._disk;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get guestAccelerator() {
        return this._guestAccelerator;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_description - computed: true, optional: false, required: false
    get instanceDescription() {
        return this.getStringAttribute('instance_description');
    }
    // key_revocation_action_type - computed: true, optional: false, required: false
    get keyRevocationActionType() {
        return this.getStringAttribute('key_revocation_action_type');
    }
    get labels() {
        return this._labels;
    }
    // machine_type - computed: true, optional: false, required: false
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    get metadata() {
        return this._metadata;
    }
    // metadata_fingerprint - computed: true, optional: false, required: false
    get metadataFingerprint() {
        return this.getStringAttribute('metadata_fingerprint');
    }
    // metadata_startup_script - computed: true, optional: false, required: false
    get metadataStartupScript() {
        return this.getStringAttribute('metadata_startup_script');
    }
    // min_cpu_platform - computed: true, optional: false, required: false
    get minCpuPlatform() {
        return this.getStringAttribute('min_cpu_platform');
    }
    get mostRecent() {
        return this.getBooleanAttribute('most_recent');
    }
    set mostRecent(value) {
        this._mostRecent = value;
    }
    resetMostRecent() {
        this._mostRecent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mostRecentInput() {
        return this._mostRecent;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // name_prefix - computed: true, optional: false, required: false
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    get networkInterface() {
        return this._networkInterface;
    }
    get networkPerformanceConfig() {
        return this._networkPerformanceConfig;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
    get reservationAffinity() {
        return this._reservationAffinity;
    }
    get resourceManagerTags() {
        return this._resourceManagerTags;
    }
    // resource_policies - computed: true, optional: false, required: false
    get resourcePolicies() {
        return this.getListAttribute('resource_policies');
    }
    get scheduling() {
        return this._scheduling;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get selfLinkUnique() {
        return this.getStringAttribute('self_link_unique');
    }
    set selfLinkUnique(value) {
        this._selfLinkUnique = value;
    }
    resetSelfLinkUnique() {
        this._selfLinkUnique = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfLinkUniqueInput() {
        return this._selfLinkUnique;
    }
    get serviceAccount() {
        return this._serviceAccount;
    }
    get shieldedInstanceConfig() {
        return this._shieldedInstanceConfig;
    }
    // tags - computed: true, optional: false, required: false
    get tags() {
        return cdktf.Fn.tolist(this.getListAttribute('tags'));
    }
    // tags_fingerprint - computed: true, optional: false, required: false
    get tagsFingerprint() {
        return this.getStringAttribute('tags_fingerprint');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filter: cdktf.stringToTerraform(this._filter),
            id: cdktf.stringToTerraform(this._id),
            most_recent: cdktf.booleanToTerraform(this._mostRecent),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            self_link_unique: cdktf.stringToTerraform(this._selfLinkUnique),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            filter: {
                value: cdktf.stringToHclTerraform(this._filter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            most_recent: {
                value: cdktf.booleanToHclTerraform(this._mostRecent),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            self_link_unique: {
                value: cdktf.stringToHclTerraform(this._selfLinkUnique),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeInstanceTemplate = DataGoogleComputeInstanceTemplate;
_18 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceTemplate[_18] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceTemplate.DataGoogleComputeInstanceTemplate", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeInstanceTemplate.tfResourceType = "google_compute_instance_template";
