// @cdktf/provider-google/lib/data-google-compute-instance-group-manager/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleComputeInstanceGroupManagerUpdatePolicyToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerUpdatePolicyToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusList = exports.DataGoogleComputeInstanceGroupManagerStatusOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerStatusToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusVersionTargetList = exports.DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusVersionTargetToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerStatusVersionTargetToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusStatefulList = exports.DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusStatefulToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerStatusStatefulToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList = exports.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigList = exports.DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatefulInternalIpList = exports.DataGoogleComputeInstanceGroupManagerStatefulInternalIpOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatefulInternalIpToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerStatefulInternalIpToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatefulExternalIpList = exports.DataGoogleComputeInstanceGroupManagerStatefulExternalIpOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatefulExternalIpToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerStatefulExternalIpToTerraform = exports.DataGoogleComputeInstanceGroupManagerStatefulDiskList = exports.DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference = exports.dataGoogleComputeInstanceGroupManagerStatefulDiskToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerStatefulDiskToTerraform = exports.DataGoogleComputeInstanceGroupManagerNamedPortList = exports.DataGoogleComputeInstanceGroupManagerNamedPortOutputReference = exports.dataGoogleComputeInstanceGroupManagerNamedPortToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerNamedPortToTerraform = exports.DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyList = exports.DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference = exports.dataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyToTerraform = exports.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList = exports.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference = exports.dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToTerraform = exports.DataGoogleComputeInstanceGroupManagerAllInstancesConfigList = exports.DataGoogleComputeInstanceGroupManagerAllInstancesConfigOutputReference = exports.dataGoogleComputeInstanceGroupManagerAllInstancesConfigToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerAllInstancesConfigToTerraform = void 0;
exports.DataGoogleComputeInstanceGroupManager = exports.DataGoogleComputeInstanceGroupManagerVersionList = exports.DataGoogleComputeInstanceGroupManagerVersionOutputReference = exports.dataGoogleComputeInstanceGroupManagerVersionToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerVersionToTerraform = exports.DataGoogleComputeInstanceGroupManagerVersionTargetSizeList = exports.DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference = exports.dataGoogleComputeInstanceGroupManagerVersionTargetSizeToHclTerraform = exports.dataGoogleComputeInstanceGroupManagerVersionTargetSizeToTerraform = exports.DataGoogleComputeInstanceGroupManagerUpdatePolicyList = exports.DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeInstanceGroupManagerAllInstancesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerAllInstancesConfigToTerraform = dataGoogleComputeInstanceGroupManagerAllInstancesConfigToTerraform;
function dataGoogleComputeInstanceGroupManagerAllInstancesConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerAllInstancesConfigToHclTerraform = dataGoogleComputeInstanceGroupManagerAllInstancesConfigToHclTerraform;
class DataGoogleComputeInstanceGroupManagerAllInstancesConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // metadata - computed: true, optional: false, required: false
        this._metadata = new cdktf.StringMap(this, "metadata");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get labels() {
        return this._labels;
    }
    get metadata() {
        return this._metadata;
    }
}
exports.DataGoogleComputeInstanceGroupManagerAllInstancesConfigOutputReference = DataGoogleComputeInstanceGroupManagerAllInstancesConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerAllInstancesConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerAllInstancesConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerAllInstancesConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerAllInstancesConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerAllInstancesConfigList = DataGoogleComputeInstanceGroupManagerAllInstancesConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerAllInstancesConfigList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerAllInstancesConfigList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToTerraform = dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToTerraform;
function dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToHclTerraform = dataGoogleComputeInstanceGroupManagerAutoHealingPoliciesToHclTerraform;
class DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // health_check - computed: true, optional: false, required: false
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    // initial_delay_sec - computed: true, optional: false, required: false
    get initialDelaySec() {
        return this.getNumberAttribute('initial_delay_sec');
    }
}
exports.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference = DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList = DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyToTerraform = dataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyToTerraform;
function dataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform = dataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform;
class DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // default_action_on_failure - computed: true, optional: false, required: false
    get defaultActionOnFailure() {
        return this.getStringAttribute('default_action_on_failure');
    }
    // force_update_on_repair - computed: true, optional: false, required: false
    get forceUpdateOnRepair() {
        return this.getStringAttribute('force_update_on_repair');
    }
}
exports.DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference = DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyList = DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerNamedPortToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerNamedPortToTerraform = dataGoogleComputeInstanceGroupManagerNamedPortToTerraform;
function dataGoogleComputeInstanceGroupManagerNamedPortToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerNamedPortToHclTerraform = dataGoogleComputeInstanceGroupManagerNamedPortToHclTerraform;
class DataGoogleComputeInstanceGroupManagerNamedPortOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataGoogleComputeInstanceGroupManagerNamedPortOutputReference = DataGoogleComputeInstanceGroupManagerNamedPortOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerNamedPortOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerNamedPortOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerNamedPortList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerNamedPortOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerNamedPortList = DataGoogleComputeInstanceGroupManagerNamedPortList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerNamedPortList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerNamedPortList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerStatefulDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatefulDiskToTerraform = dataGoogleComputeInstanceGroupManagerStatefulDiskToTerraform;
function dataGoogleComputeInstanceGroupManagerStatefulDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerStatefulDiskToHclTerraform = dataGoogleComputeInstanceGroupManagerStatefulDiskToHclTerraform;
class DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_rule - computed: true, optional: false, required: false
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference = DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerStatefulDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatefulDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatefulDiskList = DataGoogleComputeInstanceGroupManagerStatefulDiskList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatefulDiskList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatefulDiskList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerStatefulExternalIpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatefulExternalIpToTerraform = dataGoogleComputeInstanceGroupManagerStatefulExternalIpToTerraform;
function dataGoogleComputeInstanceGroupManagerStatefulExternalIpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerStatefulExternalIpToHclTerraform = dataGoogleComputeInstanceGroupManagerStatefulExternalIpToHclTerraform;
class DataGoogleComputeInstanceGroupManagerStatefulExternalIpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_rule - computed: true, optional: false, required: false
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    // interface_name - computed: true, optional: false, required: false
    get interfaceName() {
        return this.getStringAttribute('interface_name');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatefulExternalIpOutputReference = DataGoogleComputeInstanceGroupManagerStatefulExternalIpOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatefulExternalIpOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatefulExternalIpOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerStatefulExternalIpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatefulExternalIpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatefulExternalIpList = DataGoogleComputeInstanceGroupManagerStatefulExternalIpList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatefulExternalIpList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatefulExternalIpList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerStatefulInternalIpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatefulInternalIpToTerraform = dataGoogleComputeInstanceGroupManagerStatefulInternalIpToTerraform;
function dataGoogleComputeInstanceGroupManagerStatefulInternalIpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerStatefulInternalIpToHclTerraform = dataGoogleComputeInstanceGroupManagerStatefulInternalIpToHclTerraform;
class DataGoogleComputeInstanceGroupManagerStatefulInternalIpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_rule - computed: true, optional: false, required: false
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    // interface_name - computed: true, optional: false, required: false
    get interfaceName() {
        return this.getStringAttribute('interface_name');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatefulInternalIpOutputReference = DataGoogleComputeInstanceGroupManagerStatefulInternalIpOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatefulInternalIpOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatefulInternalIpOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerStatefulInternalIpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatefulInternalIpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatefulInternalIpList = DataGoogleComputeInstanceGroupManagerStatefulInternalIpList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatefulInternalIpList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatefulInternalIpList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigToTerraform = dataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigToTerraform;
function dataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigToHclTerraform = dataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigToHclTerraform;
class DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // current_revision - computed: true, optional: false, required: false
    get currentRevision() {
        return this.getStringAttribute('current_revision');
    }
    // effective - computed: true, optional: false, required: false
    get effective() {
        return this.getBooleanAttribute('effective');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference = DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigList = DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigList[_r] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform;
function dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform = dataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform;
class DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all_effective - computed: true, optional: false, required: false
    get allEffective() {
        return this.getBooleanAttribute('all_effective');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList = DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList[_t] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerStatusStatefulToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusStatefulToTerraform = dataGoogleComputeInstanceGroupManagerStatusStatefulToTerraform;
function dataGoogleComputeInstanceGroupManagerStatusStatefulToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerStatusStatefulToHclTerraform = dataGoogleComputeInstanceGroupManagerStatusStatefulToHclTerraform;
class DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // per_instance_configs - computed: true, optional: false, required: false
        this._perInstanceConfigs = new DataGoogleComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList(this, "per_instance_configs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // has_stateful_config - computed: true, optional: false, required: false
    get hasStatefulConfig() {
        return this.getBooleanAttribute('has_stateful_config');
    }
    get perInstanceConfigs() {
        return this._perInstanceConfigs;
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference = DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerStatusStatefulList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusStatefulOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusStatefulList = DataGoogleComputeInstanceGroupManagerStatusStatefulList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusStatefulList[_v] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusStatefulList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerStatusVersionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusVersionTargetToTerraform = dataGoogleComputeInstanceGroupManagerStatusVersionTargetToTerraform;
function dataGoogleComputeInstanceGroupManagerStatusVersionTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerStatusVersionTargetToHclTerraform = dataGoogleComputeInstanceGroupManagerStatusVersionTargetToHclTerraform;
class DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_reached - computed: true, optional: false, required: false
    get isReached() {
        return this.getBooleanAttribute('is_reached');
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference = DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerStatusVersionTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusVersionTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusVersionTargetList = DataGoogleComputeInstanceGroupManagerStatusVersionTargetList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusVersionTargetList[_x] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusVersionTargetList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerStatusToTerraform = dataGoogleComputeInstanceGroupManagerStatusToTerraform;
function dataGoogleComputeInstanceGroupManagerStatusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerStatusToHclTerraform = dataGoogleComputeInstanceGroupManagerStatusToHclTerraform;
class DataGoogleComputeInstanceGroupManagerStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // all_instances_config - computed: true, optional: false, required: false
        this._allInstancesConfig = new DataGoogleComputeInstanceGroupManagerStatusAllInstancesConfigList(this, "all_instances_config", false);
        // stateful - computed: true, optional: false, required: false
        this._stateful = new DataGoogleComputeInstanceGroupManagerStatusStatefulList(this, "stateful", false);
        // version_target - computed: true, optional: false, required: false
        this._versionTarget = new DataGoogleComputeInstanceGroupManagerStatusVersionTargetList(this, "version_target", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get allInstancesConfig() {
        return this._allInstancesConfig;
    }
    // is_stable - computed: true, optional: false, required: false
    get isStable() {
        return this.getBooleanAttribute('is_stable');
    }
    get stateful() {
        return this._stateful;
    }
    get versionTarget() {
        return this._versionTarget;
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusOutputReference = DataGoogleComputeInstanceGroupManagerStatusOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerStatusList = DataGoogleComputeInstanceGroupManagerStatusList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerStatusList[_z] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerStatusList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerUpdatePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerUpdatePolicyToTerraform = dataGoogleComputeInstanceGroupManagerUpdatePolicyToTerraform;
function dataGoogleComputeInstanceGroupManagerUpdatePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerUpdatePolicyToHclTerraform = dataGoogleComputeInstanceGroupManagerUpdatePolicyToHclTerraform;
class DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_surge_fixed - computed: true, optional: false, required: false
    get maxSurgeFixed() {
        return this.getNumberAttribute('max_surge_fixed');
    }
    // max_surge_percent - computed: true, optional: false, required: false
    get maxSurgePercent() {
        return this.getNumberAttribute('max_surge_percent');
    }
    // max_unavailable_fixed - computed: true, optional: false, required: false
    get maxUnavailableFixed() {
        return this.getNumberAttribute('max_unavailable_fixed');
    }
    // max_unavailable_percent - computed: true, optional: false, required: false
    get maxUnavailablePercent() {
        return this.getNumberAttribute('max_unavailable_percent');
    }
    // minimal_action - computed: true, optional: false, required: false
    get minimalAction() {
        return this.getStringAttribute('minimal_action');
    }
    // most_disruptive_allowed_action - computed: true, optional: false, required: false
    get mostDisruptiveAllowedAction() {
        return this.getStringAttribute('most_disruptive_allowed_action');
    }
    // replacement_method - computed: true, optional: false, required: false
    get replacementMethod() {
        return this.getStringAttribute('replacement_method');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference = DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerUpdatePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerUpdatePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerUpdatePolicyList = DataGoogleComputeInstanceGroupManagerUpdatePolicyList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerUpdatePolicyList[_1] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerUpdatePolicyList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerVersionTargetSizeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerVersionTargetSizeToTerraform = dataGoogleComputeInstanceGroupManagerVersionTargetSizeToTerraform;
function dataGoogleComputeInstanceGroupManagerVersionTargetSizeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerVersionTargetSizeToHclTerraform = dataGoogleComputeInstanceGroupManagerVersionTargetSizeToHclTerraform;
class DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // fixed - computed: true, optional: false, required: false
    get fixed() {
        return this.getNumberAttribute('fixed');
    }
    // percent - computed: true, optional: false, required: false
    get percent() {
        return this.getNumberAttribute('percent');
    }
}
exports.DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference = DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerVersionTargetSizeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerVersionTargetSizeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerVersionTargetSizeList = DataGoogleComputeInstanceGroupManagerVersionTargetSizeList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerVersionTargetSizeList[_3] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerVersionTargetSizeList", version: "14.13.1" };
function dataGoogleComputeInstanceGroupManagerVersionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeInstanceGroupManagerVersionToTerraform = dataGoogleComputeInstanceGroupManagerVersionToTerraform;
function dataGoogleComputeInstanceGroupManagerVersionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeInstanceGroupManagerVersionToHclTerraform = dataGoogleComputeInstanceGroupManagerVersionToHclTerraform;
class DataGoogleComputeInstanceGroupManagerVersionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // target_size - computed: true, optional: false, required: false
        this._targetSize = new DataGoogleComputeInstanceGroupManagerVersionTargetSizeList(this, "target_size", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // instance_template - computed: true, optional: false, required: false
    get instanceTemplate() {
        return this.getStringAttribute('instance_template');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get targetSize() {
        return this._targetSize;
    }
}
exports.DataGoogleComputeInstanceGroupManagerVersionOutputReference = DataGoogleComputeInstanceGroupManagerVersionOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerVersionOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerVersionOutputReference", version: "14.13.1" };
class DataGoogleComputeInstanceGroupManagerVersionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeInstanceGroupManagerVersionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeInstanceGroupManagerVersionList = DataGoogleComputeInstanceGroupManagerVersionList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManagerVersionList[_5] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManagerVersionList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_instance_group_manager google_compute_instance_group_manager}
*/
class DataGoogleComputeInstanceGroupManager extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeInstanceGroupManager resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeInstanceGroupManager to import
    * @param importFromId The id of the existing DataGoogleComputeInstanceGroupManager that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_instance_group_manager#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeInstanceGroupManager to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_instance_group_manager", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_instance_group_manager google_compute_instance_group_manager} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeInstanceGroupManagerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_instance_group_manager',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // all_instances_config - computed: true, optional: false, required: false
        this._allInstancesConfig = new DataGoogleComputeInstanceGroupManagerAllInstancesConfigList(this, "all_instances_config", false);
        // auto_healing_policies - computed: true, optional: false, required: false
        this._autoHealingPolicies = new DataGoogleComputeInstanceGroupManagerAutoHealingPoliciesList(this, "auto_healing_policies", false);
        // instance_lifecycle_policy - computed: true, optional: false, required: false
        this._instanceLifecyclePolicy = new DataGoogleComputeInstanceGroupManagerInstanceLifecyclePolicyList(this, "instance_lifecycle_policy", false);
        // named_port - computed: true, optional: false, required: false
        this._namedPort = new DataGoogleComputeInstanceGroupManagerNamedPortList(this, "named_port", true);
        // stateful_disk - computed: true, optional: false, required: false
        this._statefulDisk = new DataGoogleComputeInstanceGroupManagerStatefulDiskList(this, "stateful_disk", true);
        // stateful_external_ip - computed: true, optional: false, required: false
        this._statefulExternalIp = new DataGoogleComputeInstanceGroupManagerStatefulExternalIpList(this, "stateful_external_ip", false);
        // stateful_internal_ip - computed: true, optional: false, required: false
        this._statefulInternalIp = new DataGoogleComputeInstanceGroupManagerStatefulInternalIpList(this, "stateful_internal_ip", false);
        // status - computed: true, optional: false, required: false
        this._status = new DataGoogleComputeInstanceGroupManagerStatusList(this, "status", false);
        // update_policy - computed: true, optional: false, required: false
        this._updatePolicy = new DataGoogleComputeInstanceGroupManagerUpdatePolicyList(this, "update_policy", false);
        // version - computed: true, optional: false, required: false
        this._version = new DataGoogleComputeInstanceGroupManagerVersionList(this, "version", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._selfLink = config.selfLink;
        this._zone = config.zone;
    }
    get allInstancesConfig() {
        return this._allInstancesConfig;
    }
    get autoHealingPolicies() {
        return this._autoHealingPolicies;
    }
    // base_instance_name - computed: true, optional: false, required: false
    get baseInstanceName() {
        return this.getStringAttribute('base_instance_name');
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // fingerprint - computed: true, optional: false, required: false
    get fingerprint() {
        return this.getStringAttribute('fingerprint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_group - computed: true, optional: false, required: false
    get instanceGroup() {
        return this.getStringAttribute('instance_group');
    }
    // instance_group_manager_id - computed: true, optional: false, required: false
    get instanceGroupManagerId() {
        return this.getNumberAttribute('instance_group_manager_id');
    }
    get instanceLifecyclePolicy() {
        return this._instanceLifecyclePolicy;
    }
    // list_managed_instances_results - computed: true, optional: false, required: false
    get listManagedInstancesResults() {
        return this.getStringAttribute('list_managed_instances_results');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namedPort() {
        return this._namedPort;
    }
    // operation - computed: true, optional: false, required: false
    get operation() {
        return this.getStringAttribute('operation');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    set selfLink(value) {
        this._selfLink = value;
    }
    resetSelfLink() {
        this._selfLink = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfLinkInput() {
        return this._selfLink;
    }
    get statefulDisk() {
        return this._statefulDisk;
    }
    get statefulExternalIp() {
        return this._statefulExternalIp;
    }
    get statefulInternalIp() {
        return this._statefulInternalIp;
    }
    get status() {
        return this._status;
    }
    // target_pools - computed: true, optional: false, required: false
    get targetPools() {
        return cdktf.Fn.tolist(this.getListAttribute('target_pools'));
    }
    // target_size - computed: true, optional: false, required: false
    get targetSize() {
        return this.getNumberAttribute('target_size');
    }
    get updatePolicy() {
        return this._updatePolicy;
    }
    get version() {
        return this._version;
    }
    // wait_for_instances - computed: true, optional: false, required: false
    get waitForInstances() {
        return this.getBooleanAttribute('wait_for_instances');
    }
    // wait_for_instances_status - computed: true, optional: false, required: false
    get waitForInstancesStatus() {
        return this.getStringAttribute('wait_for_instances_status');
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            self_link: cdktf.stringToTerraform(this._selfLink),
            zone: cdktf.stringToTerraform(this._zone),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            self_link: {
                value: cdktf.stringToHclTerraform(this._selfLink),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            zone: {
                value: cdktf.stringToHclTerraform(this._zone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeInstanceGroupManager = DataGoogleComputeInstanceGroupManager;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeInstanceGroupManager[_6] = { fqn: "@cdktf/provider-google.dataGoogleComputeInstanceGroupManager.DataGoogleComputeInstanceGroupManager", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeInstanceGroupManager.tfResourceType = "google_compute_instance_group_manager";
