// @cdktf/provider-google/lib/data-google-backup-dr-backup-vault/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleBackupDrBackupVault = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/backup_dr_backup_vault google_backup_dr_backup_vault}
*/
class DataGoogleBackupDrBackupVault extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleBackupDrBackupVault resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleBackupDrBackupVault to import
    * @param importFromId The id of the existing DataGoogleBackupDrBackupVault that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/backup_dr_backup_vault#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleBackupDrBackupVault to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_backup_dr_backup_vault", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/backup_dr_backup_vault google_backup_dr_backup_vault} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleBackupDrBackupVaultConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_backup_dr_backup_vault',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // annotations - computed: true, optional: false, required: false
        this._annotations = new cdktf.StringMap(this, "annotations");
        // effective_annotations - computed: true, optional: false, required: false
        this._effectiveAnnotations = new cdktf.StringMap(this, "effective_annotations");
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._backupVaultId = config.backupVaultId;
        this._id = config.id;
        this._location = config.location;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // access_restriction - computed: true, optional: false, required: false
    get accessRestriction() {
        return this.getStringAttribute('access_restriction');
    }
    // allow_missing - computed: true, optional: false, required: false
    get allowMissing() {
        return this.getBooleanAttribute('allow_missing');
    }
    get annotations() {
        return this._annotations;
    }
    // backup_count - computed: true, optional: false, required: false
    get backupCount() {
        return this.getStringAttribute('backup_count');
    }
    // backup_minimum_enforced_retention_duration - computed: true, optional: false, required: false
    get backupMinimumEnforcedRetentionDuration() {
        return this.getStringAttribute('backup_minimum_enforced_retention_duration');
    }
    get backupVaultId() {
        return this.getStringAttribute('backup_vault_id');
    }
    set backupVaultId(value) {
        this._backupVaultId = value;
    }
    // Temporarily expose input value. Use with caution.
    get backupVaultIdInput() {
        return this._backupVaultId;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // deletable - computed: true, optional: false, required: false
    get deletable() {
        return this.getBooleanAttribute('deletable');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get effectiveAnnotations() {
        return this._effectiveAnnotations;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // effective_time - computed: true, optional: false, required: false
    get effectiveTime() {
        return this.getStringAttribute('effective_time');
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    // force_delete - computed: true, optional: false, required: false
    get forceDelete() {
        return this.getBooleanAttribute('force_delete');
    }
    // force_update - computed: true, optional: false, required: false
    get forceUpdate() {
        return this.getBooleanAttribute('force_update');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // ignore_backup_plan_references - computed: true, optional: false, required: false
    get ignoreBackupPlanReferences() {
        return this.getBooleanAttribute('ignore_backup_plan_references');
    }
    // ignore_inactive_datasources - computed: true, optional: false, required: false
    get ignoreInactiveDatasources() {
        return this.getBooleanAttribute('ignore_inactive_datasources');
    }
    get labels() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // service_account - computed: true, optional: false, required: false
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // total_stored_bytes - computed: true, optional: false, required: false
    get totalStoredBytes() {
        return this.getStringAttribute('total_stored_bytes');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            backup_vault_id: cdktf.stringToTerraform(this._backupVaultId),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            backup_vault_id: {
                value: cdktf.stringToHclTerraform(this._backupVaultId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleBackupDrBackupVault = DataGoogleBackupDrBackupVault;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleBackupDrBackupVault[_a] = { fqn: "@cdktf/provider-google.dataGoogleBackupDrBackupVault.DataGoogleBackupDrBackupVault", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleBackupDrBackupVault.tfResourceType = "google_backup_dr_backup_vault";
