// @cdktf/provider-google/lib/access-context-manager-service-perimeter-dry-run-ingress-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessContextManagerServicePerimeterDryRunIngressPolicy = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsOutputReference = exports.accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToHclTerraform = exports.accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToTerraform = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOutputReference = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToHclTerraform = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToTerraform = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsList = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsOutputReference = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToHclTerraform = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToTerraform = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsList = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsOutputReference = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToHclTerraform = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToTerraform = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromOutputReference = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToHclTerraform = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToTerraform = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesList = exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesOutputReference = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToHclTerraform = exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access_level: cdktf.stringToTerraform(struct.accessLevel),
        resource: cdktf.stringToTerraform(struct.resource),
    };
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToTerraform;
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        access_level: {
            value: cdktf.stringToHclTerraform(struct.accessLevel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        resource: {
            value: cdktf.stringToHclTerraform(struct.resource),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToHclTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToHclTerraform;
class AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accessLevel !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessLevel = this._accessLevel;
        }
        if (this._resource !== undefined) {
            hasAnyValues = true;
            internalValueResult.resource = this._resource;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._accessLevel = undefined;
            this._resource = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._accessLevel = value.accessLevel;
            this._resource = value.resource;
        }
    }
    get accessLevel() {
        return this.getStringAttribute('access_level');
    }
    set accessLevel(value) {
        this._accessLevel = value;
    }
    resetAccessLevel() {
        this._accessLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessLevelInput() {
        return this._accessLevel;
    }
    get resource() {
        return this.getStringAttribute('resource');
    }
    set resource(value) {
        this._resource = value;
    }
    resetResource() {
        this._resource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceInput() {
        return this._resource;
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesOutputReference = AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesOutputReference[_a] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesOutputReference", version: "14.13.1" };
class AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesList = AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesList;
_b = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesList[_b] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesList", version: "14.13.1" };
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        identities: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.identities),
        identity_type: cdktf.stringToTerraform(struct.identityType),
        sources: cdktf.listMapper(accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToTerraform, true)(struct.sources),
    };
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToTerraform;
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        identities: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.identities),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        identity_type: {
            value: cdktf.stringToHclTerraform(struct.identityType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        sources: {
            value: cdktf.listMapperHcl(accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesToHclTerraform, true)(struct.sources),
            isBlock: true,
            type: "list",
            storageClassType: "AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToHclTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToHclTerraform;
class AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // sources - computed: false, optional: true, required: false
        this._sources = new AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromSourcesList(this, "sources", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._identities !== undefined) {
            hasAnyValues = true;
            internalValueResult.identities = this._identities;
        }
        if (this._identityType !== undefined) {
            hasAnyValues = true;
            internalValueResult.identityType = this._identityType;
        }
        if (this._sources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sources = this._sources?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._identities = undefined;
            this._identityType = undefined;
            this._sources.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._identities = value.identities;
            this._identityType = value.identityType;
            this._sources.internalValue = value.sources;
        }
    }
    get identities() {
        return this.getListAttribute('identities');
    }
    set identities(value) {
        this._identities = value;
    }
    resetIdentities() {
        this._identities = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get identitiesInput() {
        return this._identities;
    }
    get identityType() {
        return this.getStringAttribute('identity_type');
    }
    set identityType(value) {
        this._identityType = value;
    }
    resetIdentityType() {
        this._identityType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get identityTypeInput() {
        return this._identityType;
    }
    get sources() {
        return this._sources;
    }
    putSources(value) {
        this._sources.internalValue = value;
    }
    resetSources() {
        this._sources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourcesInput() {
        return this._sources.internalValue;
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromOutputReference = AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromOutputReference;
_c = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromOutputReference[_c] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromOutputReference", version: "14.13.1" };
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        method: cdktf.stringToTerraform(struct.method),
        permission: cdktf.stringToTerraform(struct.permission),
    };
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToTerraform;
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        method: {
            value: cdktf.stringToHclTerraform(struct.method),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        permission: {
            value: cdktf.stringToHclTerraform(struct.permission),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToHclTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToHclTerraform;
class AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._method !== undefined) {
            hasAnyValues = true;
            internalValueResult.method = this._method;
        }
        if (this._permission !== undefined) {
            hasAnyValues = true;
            internalValueResult.permission = this._permission;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._method = undefined;
            this._permission = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._method = value.method;
            this._permission = value.permission;
        }
    }
    get method() {
        return this.getStringAttribute('method');
    }
    set method(value) {
        this._method = value;
    }
    resetMethod() {
        this._method = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get methodInput() {
        return this._method;
    }
    get permission() {
        return this.getStringAttribute('permission');
    }
    set permission(value) {
        this._permission = value;
    }
    resetPermission() {
        this._permission = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission;
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsOutputReference = AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsOutputReference[_d] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsOutputReference", version: "14.13.1" };
class AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsList = AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsList;
_e = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsList[_e] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsList", version: "14.13.1" };
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        service_name: cdktf.stringToTerraform(struct.serviceName),
        method_selectors: cdktf.listMapper(accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToTerraform, true)(struct.methodSelectors),
    };
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToTerraform;
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        service_name: {
            value: cdktf.stringToHclTerraform(struct.serviceName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        method_selectors: {
            value: cdktf.listMapperHcl(accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsToHclTerraform, true)(struct.methodSelectors),
            isBlock: true,
            type: "list",
            storageClassType: "AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToHclTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToHclTerraform;
class AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // method_selectors - computed: false, optional: true, required: false
        this._methodSelectors = new AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsMethodSelectorsList(this, "method_selectors", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._serviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceName = this._serviceName;
        }
        if (this._methodSelectors?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.methodSelectors = this._methodSelectors?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._serviceName = undefined;
            this._methodSelectors.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._serviceName = value.serviceName;
            this._methodSelectors.internalValue = value.methodSelectors;
        }
    }
    get serviceName() {
        return this.getStringAttribute('service_name');
    }
    set serviceName(value) {
        this._serviceName = value;
    }
    resetServiceName() {
        this._serviceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceNameInput() {
        return this._serviceName;
    }
    get methodSelectors() {
        return this._methodSelectors;
    }
    putMethodSelectors(value) {
        this._methodSelectors.internalValue = value;
    }
    resetMethodSelectors() {
        this._methodSelectors.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get methodSelectorsInput() {
        return this._methodSelectors.internalValue;
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsOutputReference = AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsOutputReference[_f] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsOutputReference", version: "14.13.1" };
class AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsList = AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsList;
_g = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsList[_g] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsList", version: "14.13.1" };
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resources: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.resources),
        operations: cdktf.listMapper(accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToTerraform, true)(struct.operations),
    };
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToTerraform;
function accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        resources: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.resources),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        operations: {
            value: cdktf.listMapperHcl(accessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsToHclTerraform, true)(struct.operations),
            isBlock: true,
            type: "list",
            storageClassType: "AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToHclTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToHclTerraform;
class AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // operations - computed: false, optional: true, required: false
        this._operations = new AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOperationsList(this, "operations", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resources !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources;
        }
        if (this._operations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.operations = this._operations?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resources = undefined;
            this._operations.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resources = value.resources;
            this._operations.internalValue = value.operations;
        }
    }
    get resources() {
        return this.getListAttribute('resources');
    }
    set resources(value) {
        this._resources = value;
    }
    resetResources() {
        this._resources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources;
    }
    get operations() {
        return this._operations;
    }
    putOperations(value) {
        this._operations.internalValue = value;
    }
    resetOperations() {
        this._operations.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get operationsInput() {
        return this._operations.internalValue;
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOutputReference = AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOutputReference;
_h = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOutputReference[_h] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOutputReference", version: "14.13.1" };
function accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToTerraform;
function accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToHclTerraform = accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToHclTerraform;
class AccessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsOutputReference = AccessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsOutputReference[_j] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/access_context_manager_service_perimeter_dry_run_ingress_policy google_access_context_manager_service_perimeter_dry_run_ingress_policy}
*/
class AccessContextManagerServicePerimeterDryRunIngressPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a AccessContextManagerServicePerimeterDryRunIngressPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the AccessContextManagerServicePerimeterDryRunIngressPolicy to import
    * @param importFromId The id of the existing AccessContextManagerServicePerimeterDryRunIngressPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/access_context_manager_service_perimeter_dry_run_ingress_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the AccessContextManagerServicePerimeterDryRunIngressPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_access_context_manager_service_perimeter_dry_run_ingress_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/access_context_manager_service_perimeter_dry_run_ingress_policy google_access_context_manager_service_perimeter_dry_run_ingress_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options AccessContextManagerServicePerimeterDryRunIngressPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_access_context_manager_service_perimeter_dry_run_ingress_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ingress_from - computed: false, optional: true, required: false
        this._ingressFrom = new AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromOutputReference(this, "ingress_from");
        // ingress_to - computed: false, optional: true, required: false
        this._ingressTo = new AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToOutputReference(this, "ingress_to");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new AccessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._perimeter = config.perimeter;
        this._ingressFrom.internalValue = config.ingressFrom;
        this._ingressTo.internalValue = config.ingressTo;
        this._timeouts.internalValue = config.timeouts;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get perimeter() {
        return this.getStringAttribute('perimeter');
    }
    set perimeter(value) {
        this._perimeter = value;
    }
    // Temporarily expose input value. Use with caution.
    get perimeterInput() {
        return this._perimeter;
    }
    get ingressFrom() {
        return this._ingressFrom;
    }
    putIngressFrom(value) {
        this._ingressFrom.internalValue = value;
    }
    resetIngressFrom() {
        this._ingressFrom.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressFromInput() {
        return this._ingressFrom.internalValue;
    }
    get ingressTo() {
        return this._ingressTo;
    }
    putIngressTo(value) {
        this._ingressTo.internalValue = value;
    }
    resetIngressTo() {
        this._ingressTo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressToInput() {
        return this._ingressTo.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            perimeter: cdktf.stringToTerraform(this._perimeter),
            ingress_from: accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToTerraform(this._ingressFrom.internalValue),
            ingress_to: accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToTerraform(this._ingressTo.internalValue),
            timeouts: accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            perimeter: {
                value: cdktf.stringToHclTerraform(this._perimeter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ingress_from: {
                value: accessContextManagerServicePerimeterDryRunIngressPolicyIngressFromToHclTerraform(this._ingressFrom.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AccessContextManagerServicePerimeterDryRunIngressPolicyIngressFromList",
            },
            ingress_to: {
                value: accessContextManagerServicePerimeterDryRunIngressPolicyIngressToToHclTerraform(this._ingressTo.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "AccessContextManagerServicePerimeterDryRunIngressPolicyIngressToList",
            },
            timeouts: {
                value: accessContextManagerServicePerimeterDryRunIngressPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "AccessContextManagerServicePerimeterDryRunIngressPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.AccessContextManagerServicePerimeterDryRunIngressPolicy = AccessContextManagerServicePerimeterDryRunIngressPolicy;
_k = JSII_RTTI_SYMBOL_1;
AccessContextManagerServicePerimeterDryRunIngressPolicy[_k] = { fqn: "@cdktf/provider-google.accessContextManagerServicePerimeterDryRunIngressPolicy.AccessContextManagerServicePerimeterDryRunIngressPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
AccessContextManagerServicePerimeterDryRunIngressPolicy.tfResourceType = "google_access_context_manager_service_perimeter_dry_run_ingress_policy";
