// @cdktf/provider-generator/lib/__tests__/edge-provider-schema/cli.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
const _1 = require(".");
const __1 = require("../..");
// This script builds the schema for the edge provider into the given directory for all languages.
const args = process.argv.slice(2);
const targetPath = args[0];
if (!targetPath) {
    console.error("No target path specified");
    process.exit(1);
}
// Clear / Create path
fs.mkdirpSync(targetPath);
fs.emptyDirSync(targetPath);
const deps = ["@types/node", "constructs", "cdktf"];
(async () => {
    await (0, __1.generateProviderBindingsFromSchema)(targetPath, _1.edgeSchema, {
        entrypoint: path.join("providers", "edge", "index.ts"),
        deps: deps.map((dep) => path.dirname(require.resolve(`${dep}/package.json`))),
        moduleKey: "edge",
        python: {
            outdir: path.resolve(targetPath, "python"),
            moduleName: "edge",
        },
        java: {
            outdir: path.resolve(targetPath, "java"),
            package: `imports.edgeprovider`,
        },
        csharp: {
            outdir: path.resolve(targetPath, "csharp"),
            namespace: "Providers.Edge",
        },
        golang: {
            outdir: path.resolve(targetPath, "go"),
            moduleName: "cdk.tf/go/stack/generated/hashicorp",
            packageName: "edge",
        },
    });
})();
