// @cdktf/provider-generator/lib/__tests__/edge-provider-schema/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.edgeSchema = void 0;
const builder_1 = require("./builder");
const required_attribute_resource = new builder_1.SchemaBuilder()
    .addAllPrimitiveTypes({ required: true, computed: false })
    .addAllPrimitiveListTypes({ required: true, computed: false })
    .build();
const optional_attribute_resource = new builder_1.SchemaBuilder()
    .addAllPrimitiveTypes({ required: false, computed: false })
    .addAllPrimitiveListTypes({ required: false, computed: false })
    .build();
const optional_computed_attribute_resource = new builder_1.SchemaBuilder()
    .addAllPrimitiveTypes({ required: false, computed: true })
    .addAllPrimitiveListTypes({ required: false, computed: true })
    .build();
const list_block_resource = new builder_1.SchemaBuilder()
    .listBlock({
    name: "opt",
    block: new builder_1.SchemaBuilder().addAllPrimitivePermutations().asBlock(),
    minItems: 0,
    maxItems: 42,
})
    .listBlock({
    name: "req",
    block: new builder_1.SchemaBuilder().addAllPrimitivePermutations().asBlock(),
    minItems: 1,
    maxItems: 42,
})
    .listBlock({
    name: "singleopt",
    block: new builder_1.SchemaBuilder().addAllPrimitivePermutations().asBlock(),
    minItems: 0,
    maxItems: 1,
})
    .listBlock({
    name: "singlereq",
    block: new builder_1.SchemaBuilder().addAllPrimitivePermutations().asBlock(),
    minItems: 1,
    maxItems: 1,
})
    .listBlock({
    name: "singleComputedBlock",
    block: new builder_1.SchemaBuilder()
        .attribute({
        computed: true,
        name: "computed",
        type: "string",
        required: false,
        optional: false,
    })
        .attribute({
        computed: false,
        name: "configured",
        type: "string",
        required: false,
    })
        .asBlock(),
    minItems: 0,
    maxItems: 1,
})
    .attribute({
    name: "computedListOfObject",
    type: [
        "list",
        [
            "object",
            {
                str: "string",
            },
        ],
    ],
    computed: true,
    optional: false,
})
    .attribute({
    name: "computedListOfMapOfObject",
    type: [
        "list",
        [
            "map",
            [
                "object",
                {
                    str: "string",
                    other: "string",
                },
            ],
        ],
    ],
    computed: true,
    optional: false,
})
    .build();
const map_list_resource = new builder_1.SchemaBuilder()
    .attribute({
    name: "mapListOfObject",
    type: [
        "map",
        [
            "list",
            [
                "object",
                {
                    hello: "string",
                },
            ],
        ],
    ],
    computed: false,
    optional: false,
})
    .build();
const map_resource = new builder_1.SchemaBuilder()
    .attribute({
    name: "optMap",
    type: ["map", "string"],
    required: false,
    computed: false,
})
    .attribute({
    name: "reqMap",
    type: ["map", "bool"],
    required: true,
    computed: false,
})
    .attribute({
    name: "computedMap",
    type: ["map", "number"],
    required: false,
    computed: true,
})
    .build();
const set_block_resource = new builder_1.SchemaBuilder()
    .setBlock({
    name: "set",
    block: new builder_1.SchemaBuilder().addAllPrimitivePermutations().asBlock(),
})
    .build();
exports.edgeSchema = (0, builder_1.schema)({
    name: "edge",
    provider: new builder_1.SchemaBuilder().addAllPrimitivePermutations().build(),
    resources: {
        required_attribute_resource,
        optional_attribute_resource,
        optional_computed_attribute_resource,
        list_block_resource,
        map_resource,
        set_block_resource,
        map_list_resource,
    },
    dataSources: {},
});
