// @cdktf/provider-google/lib/data-google-compute-disk/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeDisk = exports.DataGoogleComputeDiskSourceSnapshotEncryptionKeyList = exports.DataGoogleComputeDiskSourceSnapshotEncryptionKeyOutputReference = exports.dataGoogleComputeDiskSourceSnapshotEncryptionKeyToHclTerraform = exports.dataGoogleComputeDiskSourceSnapshotEncryptionKeyToTerraform = exports.DataGoogleComputeDiskSourceImageEncryptionKeyList = exports.DataGoogleComputeDiskSourceImageEncryptionKeyOutputReference = exports.dataGoogleComputeDiskSourceImageEncryptionKeyToHclTerraform = exports.dataGoogleComputeDiskSourceImageEncryptionKeyToTerraform = exports.DataGoogleComputeDiskGuestOsFeaturesList = exports.DataGoogleComputeDiskGuestOsFeaturesOutputReference = exports.dataGoogleComputeDiskGuestOsFeaturesToHclTerraform = exports.dataGoogleComputeDiskGuestOsFeaturesToTerraform = exports.DataGoogleComputeDiskDiskEncryptionKeyList = exports.DataGoogleComputeDiskDiskEncryptionKeyOutputReference = exports.dataGoogleComputeDiskDiskEncryptionKeyToHclTerraform = exports.dataGoogleComputeDiskDiskEncryptionKeyToTerraform = exports.DataGoogleComputeDiskAsyncPrimaryDiskList = exports.DataGoogleComputeDiskAsyncPrimaryDiskOutputReference = exports.dataGoogleComputeDiskAsyncPrimaryDiskToHclTerraform = exports.dataGoogleComputeDiskAsyncPrimaryDiskToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeDiskAsyncPrimaryDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeDiskAsyncPrimaryDiskToTerraform = dataGoogleComputeDiskAsyncPrimaryDiskToTerraform;
function dataGoogleComputeDiskAsyncPrimaryDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeDiskAsyncPrimaryDiskToHclTerraform = dataGoogleComputeDiskAsyncPrimaryDiskToHclTerraform;
class DataGoogleComputeDiskAsyncPrimaryDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // disk - computed: true, optional: false, required: false
    get disk() {
        return this.getStringAttribute('disk');
    }
}
exports.DataGoogleComputeDiskAsyncPrimaryDiskOutputReference = DataGoogleComputeDiskAsyncPrimaryDiskOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskAsyncPrimaryDiskOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskAsyncPrimaryDiskOutputReference", version: "14.13.1" };
class DataGoogleComputeDiskAsyncPrimaryDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeDiskAsyncPrimaryDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeDiskAsyncPrimaryDiskList = DataGoogleComputeDiskAsyncPrimaryDiskList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskAsyncPrimaryDiskList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskAsyncPrimaryDiskList", version: "14.13.1" };
function dataGoogleComputeDiskDiskEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeDiskDiskEncryptionKeyToTerraform = dataGoogleComputeDiskDiskEncryptionKeyToTerraform;
function dataGoogleComputeDiskDiskEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeDiskDiskEncryptionKeyToHclTerraform = dataGoogleComputeDiskDiskEncryptionKeyToHclTerraform;
class DataGoogleComputeDiskDiskEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // kms_key_service_account - computed: true, optional: false, required: false
    get kmsKeyServiceAccount() {
        return this.getStringAttribute('kms_key_service_account');
    }
    // raw_key - computed: true, optional: false, required: false
    get rawKey() {
        return this.getStringAttribute('raw_key');
    }
    // rsa_encrypted_key - computed: true, optional: false, required: false
    get rsaEncryptedKey() {
        return this.getStringAttribute('rsa_encrypted_key');
    }
    // sha256 - computed: true, optional: false, required: false
    get sha256() {
        return this.getStringAttribute('sha256');
    }
}
exports.DataGoogleComputeDiskDiskEncryptionKeyOutputReference = DataGoogleComputeDiskDiskEncryptionKeyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskDiskEncryptionKeyOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskDiskEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeDiskDiskEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeDiskDiskEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeDiskDiskEncryptionKeyList = DataGoogleComputeDiskDiskEncryptionKeyList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskDiskEncryptionKeyList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskDiskEncryptionKeyList", version: "14.13.1" };
function dataGoogleComputeDiskGuestOsFeaturesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeDiskGuestOsFeaturesToTerraform = dataGoogleComputeDiskGuestOsFeaturesToTerraform;
function dataGoogleComputeDiskGuestOsFeaturesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeDiskGuestOsFeaturesToHclTerraform = dataGoogleComputeDiskGuestOsFeaturesToHclTerraform;
class DataGoogleComputeDiskGuestOsFeaturesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeDiskGuestOsFeaturesOutputReference = DataGoogleComputeDiskGuestOsFeaturesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskGuestOsFeaturesOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskGuestOsFeaturesOutputReference", version: "14.13.1" };
class DataGoogleComputeDiskGuestOsFeaturesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeDiskGuestOsFeaturesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeDiskGuestOsFeaturesList = DataGoogleComputeDiskGuestOsFeaturesList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskGuestOsFeaturesList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskGuestOsFeaturesList", version: "14.13.1" };
function dataGoogleComputeDiskSourceImageEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeDiskSourceImageEncryptionKeyToTerraform = dataGoogleComputeDiskSourceImageEncryptionKeyToTerraform;
function dataGoogleComputeDiskSourceImageEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeDiskSourceImageEncryptionKeyToHclTerraform = dataGoogleComputeDiskSourceImageEncryptionKeyToHclTerraform;
class DataGoogleComputeDiskSourceImageEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // kms_key_service_account - computed: true, optional: false, required: false
    get kmsKeyServiceAccount() {
        return this.getStringAttribute('kms_key_service_account');
    }
    // raw_key - computed: true, optional: false, required: false
    get rawKey() {
        return this.getStringAttribute('raw_key');
    }
    // sha256 - computed: true, optional: false, required: false
    get sha256() {
        return this.getStringAttribute('sha256');
    }
}
exports.DataGoogleComputeDiskSourceImageEncryptionKeyOutputReference = DataGoogleComputeDiskSourceImageEncryptionKeyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskSourceImageEncryptionKeyOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskSourceImageEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeDiskSourceImageEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeDiskSourceImageEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeDiskSourceImageEncryptionKeyList = DataGoogleComputeDiskSourceImageEncryptionKeyList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskSourceImageEncryptionKeyList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskSourceImageEncryptionKeyList", version: "14.13.1" };
function dataGoogleComputeDiskSourceSnapshotEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeDiskSourceSnapshotEncryptionKeyToTerraform = dataGoogleComputeDiskSourceSnapshotEncryptionKeyToTerraform;
function dataGoogleComputeDiskSourceSnapshotEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeDiskSourceSnapshotEncryptionKeyToHclTerraform = dataGoogleComputeDiskSourceSnapshotEncryptionKeyToHclTerraform;
class DataGoogleComputeDiskSourceSnapshotEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // kms_key_service_account - computed: true, optional: false, required: false
    get kmsKeyServiceAccount() {
        return this.getStringAttribute('kms_key_service_account');
    }
    // raw_key - computed: true, optional: false, required: false
    get rawKey() {
        return this.getStringAttribute('raw_key');
    }
    // sha256 - computed: true, optional: false, required: false
    get sha256() {
        return this.getStringAttribute('sha256');
    }
}
exports.DataGoogleComputeDiskSourceSnapshotEncryptionKeyOutputReference = DataGoogleComputeDiskSourceSnapshotEncryptionKeyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskSourceSnapshotEncryptionKeyOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskSourceSnapshotEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeDiskSourceSnapshotEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeDiskSourceSnapshotEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeDiskSourceSnapshotEncryptionKeyList = DataGoogleComputeDiskSourceSnapshotEncryptionKeyList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDiskSourceSnapshotEncryptionKeyList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDiskSourceSnapshotEncryptionKeyList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_disk google_compute_disk}
*/
class DataGoogleComputeDisk extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeDisk resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeDisk to import
    * @param importFromId The id of the existing DataGoogleComputeDisk that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_disk#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeDisk to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_disk", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_disk google_compute_disk} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeDiskConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_disk',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // async_primary_disk - computed: true, optional: false, required: false
        this._asyncPrimaryDisk = new DataGoogleComputeDiskAsyncPrimaryDiskList(this, "async_primary_disk", false);
        // disk_encryption_key - computed: true, optional: false, required: false
        this._diskEncryptionKey = new DataGoogleComputeDiskDiskEncryptionKeyList(this, "disk_encryption_key", false);
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // guest_os_features - computed: true, optional: false, required: false
        this._guestOsFeatures = new DataGoogleComputeDiskGuestOsFeaturesList(this, "guest_os_features", true);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // source_image_encryption_key - computed: true, optional: false, required: false
        this._sourceImageEncryptionKey = new DataGoogleComputeDiskSourceImageEncryptionKeyList(this, "source_image_encryption_key", false);
        // source_snapshot_encryption_key - computed: true, optional: false, required: false
        this._sourceSnapshotEncryptionKey = new DataGoogleComputeDiskSourceSnapshotEncryptionKeyList(this, "source_snapshot_encryption_key", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._zone = config.zone;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // access_mode - computed: true, optional: false, required: false
    get accessMode() {
        return this.getStringAttribute('access_mode');
    }
    get asyncPrimaryDisk() {
        return this._asyncPrimaryDisk;
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get diskEncryptionKey() {
        return this._diskEncryptionKey;
    }
    // disk_id - computed: true, optional: false, required: false
    get diskId() {
        return this.getStringAttribute('disk_id');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // enable_confidential_compute - computed: true, optional: false, required: false
    get enableConfidentialCompute() {
        return this.getBooleanAttribute('enable_confidential_compute');
    }
    get guestOsFeatures() {
        return this._guestOsFeatures;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // image - computed: true, optional: false, required: false
    get image() {
        return this.getStringAttribute('image');
    }
    // label_fingerprint - computed: true, optional: false, required: false
    get labelFingerprint() {
        return this.getStringAttribute('label_fingerprint');
    }
    get labels() {
        return this._labels;
    }
    // last_attach_timestamp - computed: true, optional: false, required: false
    get lastAttachTimestamp() {
        return this.getStringAttribute('last_attach_timestamp');
    }
    // last_detach_timestamp - computed: true, optional: false, required: false
    get lastDetachTimestamp() {
        return this.getStringAttribute('last_detach_timestamp');
    }
    // licenses - computed: true, optional: false, required: false
    get licenses() {
        return this.getListAttribute('licenses');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // physical_block_size_bytes - computed: true, optional: false, required: false
    get physicalBlockSizeBytes() {
        return this.getNumberAttribute('physical_block_size_bytes');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // provisioned_iops - computed: true, optional: false, required: false
    get provisionedIops() {
        return this.getNumberAttribute('provisioned_iops');
    }
    // provisioned_throughput - computed: true, optional: false, required: false
    get provisionedThroughput() {
        return this.getNumberAttribute('provisioned_throughput');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
    // snapshot - computed: true, optional: false, required: false
    get snapshot() {
        return this.getStringAttribute('snapshot');
    }
    // source_disk - computed: true, optional: false, required: false
    get sourceDisk() {
        return this.getStringAttribute('source_disk');
    }
    // source_disk_id - computed: true, optional: false, required: false
    get sourceDiskId() {
        return this.getStringAttribute('source_disk_id');
    }
    get sourceImageEncryptionKey() {
        return this._sourceImageEncryptionKey;
    }
    // source_image_id - computed: true, optional: false, required: false
    get sourceImageId() {
        return this.getStringAttribute('source_image_id');
    }
    get sourceSnapshotEncryptionKey() {
        return this._sourceSnapshotEncryptionKey;
    }
    // source_snapshot_id - computed: true, optional: false, required: false
    get sourceSnapshotId() {
        return this.getStringAttribute('source_snapshot_id');
    }
    // storage_pool - computed: true, optional: false, required: false
    get storagePool() {
        return this.getStringAttribute('storage_pool');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // users - computed: true, optional: false, required: false
    get users() {
        return this.getListAttribute('users');
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            zone: cdktf.stringToTerraform(this._zone),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            zone: {
                value: cdktf.stringToHclTerraform(this._zone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeDisk = DataGoogleComputeDisk;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeDisk[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeDisk.DataGoogleComputeDisk", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeDisk.tfResourceType = "google_compute_disk";
