// @cdktf/provider-generator/lib/get/__tests__/util.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.expectModuleToMatchSnapshot = void 0;
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const fs = __importStar(require("fs"));
const commons_1 = require("@cdktf/commons");
const constructs_maker_1 = require("../constructs-maker");
const path = __importStar(require("path"));
function expectModuleToMatchSnapshot(testName, testCategory, fixtureNames) {
    test(testName, async () => {
        await (0, commons_1.withTempDir)(`${testName.replace(/\s*/, "-")}.test`, async () => {
            const curdir = process.cwd();
            fs.mkdirSync("module");
            fixtureNames.forEach((fixtureName) => {
                fs.copyFileSync(path.join(__dirname, testCategory, "fixtures", fixtureName), path.join(curdir, "module", fixtureName));
            });
            const constraint = new commons_1.TerraformModuleConstraint({
                source: "./module",
                name: "module",
                fqn: "module",
            });
            fs.mkdirSync("work");
            const workdir = path.join(curdir, "work");
            const maker = new constructs_maker_1.ConstructsMaker({
                codeMakerOutput: workdir,
                targetLanguage: commons_1.Language.TYPESCRIPT,
            }, process.env.CDKTF_EXPERIMENTAL_PROVIDER_SCHEMA_CACHE_PATH);
            await maker.generate([constraint]);
            const output = fs.readFileSync(path.join(workdir, "modules/module.ts"), "utf-8");
            expect(output).toMatchSnapshot();
        });
    });
}
exports.expectModuleToMatchSnapshot = expectModuleToMatchSnapshot;
