// @cdktf/provider-google/lib/spanner-backup-schedule/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpannerBackupSchedule = exports.SpannerBackupScheduleTimeoutsOutputReference = exports.spannerBackupScheduleTimeoutsToHclTerraform = exports.spannerBackupScheduleTimeoutsToTerraform = exports.SpannerBackupScheduleSpecOutputReference = exports.spannerBackupScheduleSpecToHclTerraform = exports.spannerBackupScheduleSpecToTerraform = exports.SpannerBackupScheduleSpecCronSpecOutputReference = exports.spannerBackupScheduleSpecCronSpecToHclTerraform = exports.spannerBackupScheduleSpecCronSpecToTerraform = exports.SpannerBackupScheduleIncrementalBackupSpecOutputReference = exports.spannerBackupScheduleIncrementalBackupSpecToHclTerraform = exports.spannerBackupScheduleIncrementalBackupSpecToTerraform = exports.SpannerBackupScheduleFullBackupSpecOutputReference = exports.spannerBackupScheduleFullBackupSpecToHclTerraform = exports.spannerBackupScheduleFullBackupSpecToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function spannerBackupScheduleFullBackupSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.spannerBackupScheduleFullBackupSpecToTerraform = spannerBackupScheduleFullBackupSpecToTerraform;
function spannerBackupScheduleFullBackupSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.spannerBackupScheduleFullBackupSpecToHclTerraform = spannerBackupScheduleFullBackupSpecToHclTerraform;
class SpannerBackupScheduleFullBackupSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.SpannerBackupScheduleFullBackupSpecOutputReference = SpannerBackupScheduleFullBackupSpecOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SpannerBackupScheduleFullBackupSpecOutputReference[_a] = { fqn: "@cdktf/provider-google.spannerBackupSchedule.SpannerBackupScheduleFullBackupSpecOutputReference", version: "14.13.1" };
function spannerBackupScheduleIncrementalBackupSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.spannerBackupScheduleIncrementalBackupSpecToTerraform = spannerBackupScheduleIncrementalBackupSpecToTerraform;
function spannerBackupScheduleIncrementalBackupSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.spannerBackupScheduleIncrementalBackupSpecToHclTerraform = spannerBackupScheduleIncrementalBackupSpecToHclTerraform;
class SpannerBackupScheduleIncrementalBackupSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.SpannerBackupScheduleIncrementalBackupSpecOutputReference = SpannerBackupScheduleIncrementalBackupSpecOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SpannerBackupScheduleIncrementalBackupSpecOutputReference[_b] = { fqn: "@cdktf/provider-google.spannerBackupSchedule.SpannerBackupScheduleIncrementalBackupSpecOutputReference", version: "14.13.1" };
function spannerBackupScheduleSpecCronSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        text: cdktf.stringToTerraform(struct.text),
    };
}
exports.spannerBackupScheduleSpecCronSpecToTerraform = spannerBackupScheduleSpecCronSpecToTerraform;
function spannerBackupScheduleSpecCronSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        text: {
            value: cdktf.stringToHclTerraform(struct.text),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.spannerBackupScheduleSpecCronSpecToHclTerraform = spannerBackupScheduleSpecCronSpecToHclTerraform;
class SpannerBackupScheduleSpecCronSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._text !== undefined) {
            hasAnyValues = true;
            internalValueResult.text = this._text;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._text = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._text = value.text;
        }
    }
    get text() {
        return this.getStringAttribute('text');
    }
    set text(value) {
        this._text = value;
    }
    resetText() {
        this._text = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get textInput() {
        return this._text;
    }
}
exports.SpannerBackupScheduleSpecCronSpecOutputReference = SpannerBackupScheduleSpecCronSpecOutputReference;
_c = JSII_RTTI_SYMBOL_1;
SpannerBackupScheduleSpecCronSpecOutputReference[_c] = { fqn: "@cdktf/provider-google.spannerBackupSchedule.SpannerBackupScheduleSpecCronSpecOutputReference", version: "14.13.1" };
function spannerBackupScheduleSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cron_spec: spannerBackupScheduleSpecCronSpecToTerraform(struct.cronSpec),
    };
}
exports.spannerBackupScheduleSpecToTerraform = spannerBackupScheduleSpecToTerraform;
function spannerBackupScheduleSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cron_spec: {
            value: spannerBackupScheduleSpecCronSpecToHclTerraform(struct.cronSpec),
            isBlock: true,
            type: "list",
            storageClassType: "SpannerBackupScheduleSpecCronSpecList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.spannerBackupScheduleSpecToHclTerraform = spannerBackupScheduleSpecToHclTerraform;
class SpannerBackupScheduleSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cron_spec - computed: false, optional: true, required: false
        this._cronSpec = new SpannerBackupScheduleSpecCronSpecOutputReference(this, "cron_spec");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cronSpec?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cronSpec = this._cronSpec?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cronSpec.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cronSpec.internalValue = value.cronSpec;
        }
    }
    get cronSpec() {
        return this._cronSpec;
    }
    putCronSpec(value) {
        this._cronSpec.internalValue = value;
    }
    resetCronSpec() {
        this._cronSpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cronSpecInput() {
        return this._cronSpec.internalValue;
    }
}
exports.SpannerBackupScheduleSpecOutputReference = SpannerBackupScheduleSpecOutputReference;
_d = JSII_RTTI_SYMBOL_1;
SpannerBackupScheduleSpecOutputReference[_d] = { fqn: "@cdktf/provider-google.spannerBackupSchedule.SpannerBackupScheduleSpecOutputReference", version: "14.13.1" };
function spannerBackupScheduleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.spannerBackupScheduleTimeoutsToTerraform = spannerBackupScheduleTimeoutsToTerraform;
function spannerBackupScheduleTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.spannerBackupScheduleTimeoutsToHclTerraform = spannerBackupScheduleTimeoutsToHclTerraform;
class SpannerBackupScheduleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.SpannerBackupScheduleTimeoutsOutputReference = SpannerBackupScheduleTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SpannerBackupScheduleTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-google.spannerBackupSchedule.SpannerBackupScheduleTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/spanner_backup_schedule google_spanner_backup_schedule}
*/
class SpannerBackupSchedule extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SpannerBackupSchedule resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SpannerBackupSchedule to import
    * @param importFromId The id of the existing SpannerBackupSchedule that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/spanner_backup_schedule#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SpannerBackupSchedule to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_spanner_backup_schedule", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/spanner_backup_schedule google_spanner_backup_schedule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SpannerBackupScheduleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_spanner_backup_schedule',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // full_backup_spec - computed: false, optional: true, required: false
        this._fullBackupSpec = new SpannerBackupScheduleFullBackupSpecOutputReference(this, "full_backup_spec");
        // incremental_backup_spec - computed: false, optional: true, required: false
        this._incrementalBackupSpec = new SpannerBackupScheduleIncrementalBackupSpecOutputReference(this, "incremental_backup_spec");
        // spec - computed: false, optional: true, required: false
        this._spec = new SpannerBackupScheduleSpecOutputReference(this, "spec");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new SpannerBackupScheduleTimeoutsOutputReference(this, "timeouts");
        this._database = config.database;
        this._id = config.id;
        this._instance = config.instance;
        this._name = config.name;
        this._project = config.project;
        this._retentionDuration = config.retentionDuration;
        this._fullBackupSpec.internalValue = config.fullBackupSpec;
        this._incrementalBackupSpec.internalValue = config.incrementalBackupSpec;
        this._spec.internalValue = config.spec;
        this._timeouts.internalValue = config.timeouts;
    }
    get database() {
        return this.getStringAttribute('database');
    }
    set database(value) {
        this._database = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instance() {
        return this.getStringAttribute('instance');
    }
    set instance(value) {
        this._instance = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceInput() {
        return this._instance;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get retentionDuration() {
        return this.getStringAttribute('retention_duration');
    }
    set retentionDuration(value) {
        this._retentionDuration = value;
    }
    // Temporarily expose input value. Use with caution.
    get retentionDurationInput() {
        return this._retentionDuration;
    }
    get fullBackupSpec() {
        return this._fullBackupSpec;
    }
    putFullBackupSpec(value) {
        this._fullBackupSpec.internalValue = value;
    }
    resetFullBackupSpec() {
        this._fullBackupSpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fullBackupSpecInput() {
        return this._fullBackupSpec.internalValue;
    }
    get incrementalBackupSpec() {
        return this._incrementalBackupSpec;
    }
    putIncrementalBackupSpec(value) {
        this._incrementalBackupSpec.internalValue = value;
    }
    resetIncrementalBackupSpec() {
        this._incrementalBackupSpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get incrementalBackupSpecInput() {
        return this._incrementalBackupSpec.internalValue;
    }
    get spec() {
        return this._spec;
    }
    putSpec(value) {
        this._spec.internalValue = value;
    }
    resetSpec() {
        this._spec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get specInput() {
        return this._spec.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            database: cdktf.stringToTerraform(this._database),
            id: cdktf.stringToTerraform(this._id),
            instance: cdktf.stringToTerraform(this._instance),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            retention_duration: cdktf.stringToTerraform(this._retentionDuration),
            full_backup_spec: spannerBackupScheduleFullBackupSpecToTerraform(this._fullBackupSpec.internalValue),
            incremental_backup_spec: spannerBackupScheduleIncrementalBackupSpecToTerraform(this._incrementalBackupSpec.internalValue),
            spec: spannerBackupScheduleSpecToTerraform(this._spec.internalValue),
            timeouts: spannerBackupScheduleTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            database: {
                value: cdktf.stringToHclTerraform(this._database),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance: {
                value: cdktf.stringToHclTerraform(this._instance),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            retention_duration: {
                value: cdktf.stringToHclTerraform(this._retentionDuration),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            full_backup_spec: {
                value: spannerBackupScheduleFullBackupSpecToHclTerraform(this._fullBackupSpec.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SpannerBackupScheduleFullBackupSpecList",
            },
            incremental_backup_spec: {
                value: spannerBackupScheduleIncrementalBackupSpecToHclTerraform(this._incrementalBackupSpec.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SpannerBackupScheduleIncrementalBackupSpecList",
            },
            spec: {
                value: spannerBackupScheduleSpecToHclTerraform(this._spec.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SpannerBackupScheduleSpecList",
            },
            timeouts: {
                value: spannerBackupScheduleTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "SpannerBackupScheduleTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SpannerBackupSchedule = SpannerBackupSchedule;
_f = JSII_RTTI_SYMBOL_1;
SpannerBackupSchedule[_f] = { fqn: "@cdktf/provider-google.spannerBackupSchedule.SpannerBackupSchedule", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
SpannerBackupSchedule.tfResourceType = "google_spanner_backup_schedule";
