// @cdktf/provider-google/lib/data-google-sql-database-instance/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleSqlDatabaseInstanceSettingsInsightsConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsInsightsConfigToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodList = exports.DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsList = exports.DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigList = exports.DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsDataCacheConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsDataCacheConfigToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationList = exports.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsList = exports.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesList = exports.DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigList = exports.DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigToTerraform = exports.DataGoogleSqlDatabaseInstanceServerCaCertList = exports.DataGoogleSqlDatabaseInstanceServerCaCertOutputReference = exports.dataGoogleSqlDatabaseInstanceServerCaCertToHclTerraform = exports.dataGoogleSqlDatabaseInstanceServerCaCertToTerraform = exports.DataGoogleSqlDatabaseInstanceRestoreBackupContextList = exports.DataGoogleSqlDatabaseInstanceRestoreBackupContextOutputReference = exports.dataGoogleSqlDatabaseInstanceRestoreBackupContextToHclTerraform = exports.dataGoogleSqlDatabaseInstanceRestoreBackupContextToTerraform = exports.DataGoogleSqlDatabaseInstanceReplicaConfigurationList = exports.DataGoogleSqlDatabaseInstanceReplicaConfigurationOutputReference = exports.dataGoogleSqlDatabaseInstanceReplicaConfigurationToHclTerraform = exports.dataGoogleSqlDatabaseInstanceReplicaConfigurationToTerraform = exports.DataGoogleSqlDatabaseInstanceIpAddressList = exports.DataGoogleSqlDatabaseInstanceIpAddressOutputReference = exports.dataGoogleSqlDatabaseInstanceIpAddressToHclTerraform = exports.dataGoogleSqlDatabaseInstanceIpAddressToTerraform = exports.DataGoogleSqlDatabaseInstanceCloneList = exports.DataGoogleSqlDatabaseInstanceCloneOutputReference = exports.dataGoogleSqlDatabaseInstanceCloneToHclTerraform = exports.dataGoogleSqlDatabaseInstanceCloneToTerraform = void 0;
exports.DataGoogleSqlDatabaseInstance = exports.DataGoogleSqlDatabaseInstanceSettingsList = exports.DataGoogleSqlDatabaseInstanceSettingsOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigList = exports.DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyList = exports.DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowList = exports.DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceList = exports.DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsLocationPreferenceToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsLocationPreferenceToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationList = exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigList = exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsList = exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksList = exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksOutputReference = exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksToHclTerraform = exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksToTerraform = exports.DataGoogleSqlDatabaseInstanceSettingsInsightsConfigList = exports.DataGoogleSqlDatabaseInstanceSettingsInsightsConfigOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleSqlDatabaseInstanceCloneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceCloneToTerraform = dataGoogleSqlDatabaseInstanceCloneToTerraform;
function dataGoogleSqlDatabaseInstanceCloneToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceCloneToHclTerraform = dataGoogleSqlDatabaseInstanceCloneToHclTerraform;
class DataGoogleSqlDatabaseInstanceCloneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allocated_ip_range - computed: true, optional: false, required: false
    get allocatedIpRange() {
        return this.getStringAttribute('allocated_ip_range');
    }
    // database_names - computed: true, optional: false, required: false
    get databaseNames() {
        return this.getListAttribute('database_names');
    }
    // point_in_time - computed: true, optional: false, required: false
    get pointInTime() {
        return this.getStringAttribute('point_in_time');
    }
    // preferred_zone - computed: true, optional: false, required: false
    get preferredZone() {
        return this.getStringAttribute('preferred_zone');
    }
    // source_instance_name - computed: true, optional: false, required: false
    get sourceInstanceName() {
        return this.getStringAttribute('source_instance_name');
    }
}
exports.DataGoogleSqlDatabaseInstanceCloneOutputReference = DataGoogleSqlDatabaseInstanceCloneOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceCloneOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceCloneOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceCloneList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceCloneOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceCloneList = DataGoogleSqlDatabaseInstanceCloneList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceCloneList[_b] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceCloneList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceIpAddressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceIpAddressToTerraform = dataGoogleSqlDatabaseInstanceIpAddressToTerraform;
function dataGoogleSqlDatabaseInstanceIpAddressToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceIpAddressToHclTerraform = dataGoogleSqlDatabaseInstanceIpAddressToHclTerraform;
class DataGoogleSqlDatabaseInstanceIpAddressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_address - computed: true, optional: false, required: false
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    // time_to_retire - computed: true, optional: false, required: false
    get timeToRetire() {
        return this.getStringAttribute('time_to_retire');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleSqlDatabaseInstanceIpAddressOutputReference = DataGoogleSqlDatabaseInstanceIpAddressOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceIpAddressOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceIpAddressOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceIpAddressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceIpAddressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceIpAddressList = DataGoogleSqlDatabaseInstanceIpAddressList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceIpAddressList[_d] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceIpAddressList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceReplicaConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceReplicaConfigurationToTerraform = dataGoogleSqlDatabaseInstanceReplicaConfigurationToTerraform;
function dataGoogleSqlDatabaseInstanceReplicaConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceReplicaConfigurationToHclTerraform = dataGoogleSqlDatabaseInstanceReplicaConfigurationToHclTerraform;
class DataGoogleSqlDatabaseInstanceReplicaConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ca_certificate - computed: true, optional: false, required: false
    get caCertificate() {
        return this.getStringAttribute('ca_certificate');
    }
    // cascadable_replica - computed: true, optional: false, required: false
    get cascadableReplica() {
        return this.getBooleanAttribute('cascadable_replica');
    }
    // client_certificate - computed: true, optional: false, required: false
    get clientCertificate() {
        return this.getStringAttribute('client_certificate');
    }
    // client_key - computed: true, optional: false, required: false
    get clientKey() {
        return this.getStringAttribute('client_key');
    }
    // connect_retry_interval - computed: true, optional: false, required: false
    get connectRetryInterval() {
        return this.getNumberAttribute('connect_retry_interval');
    }
    // dump_file_path - computed: true, optional: false, required: false
    get dumpFilePath() {
        return this.getStringAttribute('dump_file_path');
    }
    // failover_target - computed: true, optional: false, required: false
    get failoverTarget() {
        return this.getBooleanAttribute('failover_target');
    }
    // master_heartbeat_period - computed: true, optional: false, required: false
    get masterHeartbeatPeriod() {
        return this.getNumberAttribute('master_heartbeat_period');
    }
    // password - computed: true, optional: false, required: false
    get password() {
        return this.getStringAttribute('password');
    }
    // ssl_cipher - computed: true, optional: false, required: false
    get sslCipher() {
        return this.getStringAttribute('ssl_cipher');
    }
    // username - computed: true, optional: false, required: false
    get username() {
        return this.getStringAttribute('username');
    }
    // verify_server_certificate - computed: true, optional: false, required: false
    get verifyServerCertificate() {
        return this.getBooleanAttribute('verify_server_certificate');
    }
}
exports.DataGoogleSqlDatabaseInstanceReplicaConfigurationOutputReference = DataGoogleSqlDatabaseInstanceReplicaConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceReplicaConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceReplicaConfigurationOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceReplicaConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceReplicaConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceReplicaConfigurationList = DataGoogleSqlDatabaseInstanceReplicaConfigurationList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceReplicaConfigurationList[_f] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceReplicaConfigurationList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceRestoreBackupContextToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceRestoreBackupContextToTerraform = dataGoogleSqlDatabaseInstanceRestoreBackupContextToTerraform;
function dataGoogleSqlDatabaseInstanceRestoreBackupContextToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceRestoreBackupContextToHclTerraform = dataGoogleSqlDatabaseInstanceRestoreBackupContextToHclTerraform;
class DataGoogleSqlDatabaseInstanceRestoreBackupContextOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // backup_run_id - computed: true, optional: false, required: false
    get backupRunId() {
        return this.getNumberAttribute('backup_run_id');
    }
    // instance_id - computed: true, optional: false, required: false
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
}
exports.DataGoogleSqlDatabaseInstanceRestoreBackupContextOutputReference = DataGoogleSqlDatabaseInstanceRestoreBackupContextOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceRestoreBackupContextOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceRestoreBackupContextOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceRestoreBackupContextList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceRestoreBackupContextOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceRestoreBackupContextList = DataGoogleSqlDatabaseInstanceRestoreBackupContextList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceRestoreBackupContextList[_h] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceRestoreBackupContextList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceServerCaCertToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceServerCaCertToTerraform = dataGoogleSqlDatabaseInstanceServerCaCertToTerraform;
function dataGoogleSqlDatabaseInstanceServerCaCertToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceServerCaCertToHclTerraform = dataGoogleSqlDatabaseInstanceServerCaCertToHclTerraform;
class DataGoogleSqlDatabaseInstanceServerCaCertOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cert - computed: true, optional: false, required: false
    get cert() {
        return this.getStringAttribute('cert');
    }
    // common_name - computed: true, optional: false, required: false
    get commonName() {
        return this.getStringAttribute('common_name');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // expiration_time - computed: true, optional: false, required: false
    get expirationTime() {
        return this.getStringAttribute('expiration_time');
    }
    // sha1_fingerprint - computed: true, optional: false, required: false
    get sha1Fingerprint() {
        return this.getStringAttribute('sha1_fingerprint');
    }
}
exports.DataGoogleSqlDatabaseInstanceServerCaCertOutputReference = DataGoogleSqlDatabaseInstanceServerCaCertOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceServerCaCertOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceServerCaCertOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceServerCaCertList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceServerCaCertOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceServerCaCertList = DataGoogleSqlDatabaseInstanceServerCaCertList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceServerCaCertList[_k] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceServerCaCertList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigToTerraform = dataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigOutputReference = DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigList = DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigList[_m] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesToTerraform = dataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // threads_per_core - computed: true, optional: false, required: false
    get threadsPerCore() {
        return this.getNumberAttribute('threads_per_core');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesOutputReference = DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesList = DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesList[_p] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsToTerraform = dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // retained_backups - computed: true, optional: false, required: false
    get retainedBackups() {
        return this.getNumberAttribute('retained_backups');
    }
    // retention_unit - computed: true, optional: false, required: false
    get retentionUnit() {
        return this.getStringAttribute('retention_unit');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsOutputReference = DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsList = DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsList[_r] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationToTerraform = dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsBackupConfigurationToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // backup_retention_settings - computed: true, optional: false, required: false
        this._backupRetentionSettings = new DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationBackupRetentionSettingsList(this, "backup_retention_settings", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get backupRetentionSettings() {
        return this._backupRetentionSettings;
    }
    // binary_log_enabled - computed: true, optional: false, required: false
    get binaryLogEnabled() {
        return this.getBooleanAttribute('binary_log_enabled');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // point_in_time_recovery_enabled - computed: true, optional: false, required: false
    get pointInTimeRecoveryEnabled() {
        return this.getBooleanAttribute('point_in_time_recovery_enabled');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    // transaction_log_retention_days - computed: true, optional: false, required: false
    get transactionLogRetentionDays() {
        return this.getNumberAttribute('transaction_log_retention_days');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationOutputReference = DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationList = DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationList[_t] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsDataCacheConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsDataCacheConfigToTerraform = dataGoogleSqlDatabaseInstanceSettingsDataCacheConfigToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsDataCacheConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsDataCacheConfigToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsDataCacheConfigToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_cache_enabled - computed: true, optional: false, required: false
    get dataCacheEnabled() {
        return this.getBooleanAttribute('data_cache_enabled');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigOutputReference = DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigList = DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigList[_v] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsToTerraform = dataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsOutputReference = DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsList = DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsList[_x] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodToTerraform = dataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // end_date - computed: true, optional: false, required: false
    get endDate() {
        return this.getStringAttribute('end_date');
    }
    // start_date - computed: true, optional: false, required: false
    get startDate() {
        return this.getStringAttribute('start_date');
    }
    // time - computed: true, optional: false, required: false
    get time() {
        return this.getStringAttribute('time');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodOutputReference = DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodList = DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodList[_z] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsInsightsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsInsightsConfigToTerraform = dataGoogleSqlDatabaseInstanceSettingsInsightsConfigToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsInsightsConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsInsightsConfigToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsInsightsConfigToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsInsightsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // query_insights_enabled - computed: true, optional: false, required: false
    get queryInsightsEnabled() {
        return this.getBooleanAttribute('query_insights_enabled');
    }
    // query_plans_per_minute - computed: true, optional: false, required: false
    get queryPlansPerMinute() {
        return this.getNumberAttribute('query_plans_per_minute');
    }
    // query_string_length - computed: true, optional: false, required: false
    get queryStringLength() {
        return this.getNumberAttribute('query_string_length');
    }
    // record_application_tags - computed: true, optional: false, required: false
    get recordApplicationTags() {
        return this.getBooleanAttribute('record_application_tags');
    }
    // record_client_address - computed: true, optional: false, required: false
    get recordClientAddress() {
        return this.getBooleanAttribute('record_client_address');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsInsightsConfigOutputReference = DataGoogleSqlDatabaseInstanceSettingsInsightsConfigOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsInsightsConfigOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsInsightsConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsInsightsConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsInsightsConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsInsightsConfigList = DataGoogleSqlDatabaseInstanceSettingsInsightsConfigList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsInsightsConfigList[_1] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsInsightsConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksToTerraform = dataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // expiration_time - computed: true, optional: false, required: false
    get expirationTime() {
        return this.getStringAttribute('expiration_time');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksOutputReference = DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksList = DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksList[_3] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsToTerraform = dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // consumer_network - computed: true, optional: false, required: false
    get consumerNetwork() {
        return this.getStringAttribute('consumer_network');
    }
    // consumer_service_project_id - computed: true, optional: false, required: false
    get consumerServiceProjectId() {
        return this.getStringAttribute('consumer_service_project_id');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference = DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsList = DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsList[_5] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigToTerraform = dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // psc_auto_connections - computed: true, optional: false, required: false
        this._pscAutoConnections = new DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigPscAutoConnectionsList(this, "psc_auto_connections", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allowed_consumer_projects - computed: true, optional: false, required: false
    get allowedConsumerProjects() {
        return cdktf.Fn.tolist(this.getListAttribute('allowed_consumer_projects'));
    }
    get pscAutoConnections() {
        return this._pscAutoConnections;
    }
    // psc_enabled - computed: true, optional: false, required: false
    get pscEnabled() {
        return this.getBooleanAttribute('psc_enabled');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigOutputReference = DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigList = DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigList[_7] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsIpConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationToTerraform = dataGoogleSqlDatabaseInstanceSettingsIpConfigurationToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsIpConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsIpConfigurationToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsIpConfigurationToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsIpConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authorized_networks - computed: true, optional: false, required: false
        this._authorizedNetworks = new DataGoogleSqlDatabaseInstanceSettingsIpConfigurationAuthorizedNetworksList(this, "authorized_networks", true);
        // psc_config - computed: true, optional: false, required: false
        this._pscConfig = new DataGoogleSqlDatabaseInstanceSettingsIpConfigurationPscConfigList(this, "psc_config", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allocated_ip_range - computed: true, optional: false, required: false
    get allocatedIpRange() {
        return this.getStringAttribute('allocated_ip_range');
    }
    get authorizedNetworks() {
        return this._authorizedNetworks;
    }
    // enable_private_path_for_google_cloud_services - computed: true, optional: false, required: false
    get enablePrivatePathForGoogleCloudServices() {
        return this.getBooleanAttribute('enable_private_path_for_google_cloud_services');
    }
    // ipv4_enabled - computed: true, optional: false, required: false
    get ipv4Enabled() {
        return this.getBooleanAttribute('ipv4_enabled');
    }
    // private_network - computed: true, optional: false, required: false
    get privateNetwork() {
        return this.getStringAttribute('private_network');
    }
    get pscConfig() {
        return this._pscConfig;
    }
    // server_ca_mode - computed: true, optional: false, required: false
    get serverCaMode() {
        return this.getStringAttribute('server_ca_mode');
    }
    // ssl_mode - computed: true, optional: false, required: false
    get sslMode() {
        return this.getStringAttribute('ssl_mode');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationOutputReference = DataGoogleSqlDatabaseInstanceSettingsIpConfigurationOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsIpConfigurationOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsIpConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsIpConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationList = DataGoogleSqlDatabaseInstanceSettingsIpConfigurationList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsIpConfigurationList[_9] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsIpConfigurationList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsLocationPreferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsLocationPreferenceToTerraform = dataGoogleSqlDatabaseInstanceSettingsLocationPreferenceToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsLocationPreferenceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsLocationPreferenceToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsLocationPreferenceToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // follow_gae_application - computed: true, optional: false, required: false
    get followGaeApplication() {
        return this.getStringAttribute('follow_gae_application');
    }
    // secondary_zone - computed: true, optional: false, required: false
    get secondaryZone() {
        return this.getStringAttribute('secondary_zone');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceOutputReference = DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceOutputReference[_10] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceList = DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceList;
_11 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceList[_11] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowToTerraform = dataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day - computed: true, optional: false, required: false
    get day() {
        return this.getNumberAttribute('day');
    }
    // hour - computed: true, optional: false, required: false
    get hour() {
        return this.getNumberAttribute('hour');
    }
    // update_track - computed: true, optional: false, required: false
    get updateTrack() {
        return this.getStringAttribute('update_track');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowOutputReference = DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowOutputReference[_12] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowList = DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowList;
_13 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowList[_13] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyToTerraform = dataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // complexity - computed: true, optional: false, required: false
    get complexity() {
        return this.getStringAttribute('complexity');
    }
    // disallow_username_substring - computed: true, optional: false, required: false
    get disallowUsernameSubstring() {
        return this.getBooleanAttribute('disallow_username_substring');
    }
    // enable_password_policy - computed: true, optional: false, required: false
    get enablePasswordPolicy() {
        return this.getBooleanAttribute('enable_password_policy');
    }
    // min_length - computed: true, optional: false, required: false
    get minLength() {
        return this.getNumberAttribute('min_length');
    }
    // password_change_interval - computed: true, optional: false, required: false
    get passwordChangeInterval() {
        return this.getStringAttribute('password_change_interval');
    }
    // reuse_interval - computed: true, optional: false, required: false
    get reuseInterval() {
        return this.getNumberAttribute('reuse_interval');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyOutputReference = DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyOutputReference[_14] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyList = DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyList;
_15 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyList[_15] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigToTerraform = dataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // retention_interval - computed: true, optional: false, required: false
    get retentionInterval() {
        return this.getStringAttribute('retention_interval');
    }
    // upload_interval - computed: true, optional: false, required: false
    get uploadInterval() {
        return this.getStringAttribute('upload_interval');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigOutputReference = DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigOutputReference[_16] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigList = DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigList;
_17 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigList[_17] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstanceSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstanceSettingsToTerraform = dataGoogleSqlDatabaseInstanceSettingsToTerraform;
function dataGoogleSqlDatabaseInstanceSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstanceSettingsToHclTerraform = dataGoogleSqlDatabaseInstanceSettingsToHclTerraform;
class DataGoogleSqlDatabaseInstanceSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // active_directory_config - computed: true, optional: false, required: false
        this._activeDirectoryConfig = new DataGoogleSqlDatabaseInstanceSettingsActiveDirectoryConfigList(this, "active_directory_config", false);
        // advanced_machine_features - computed: true, optional: false, required: false
        this._advancedMachineFeatures = new DataGoogleSqlDatabaseInstanceSettingsAdvancedMachineFeaturesList(this, "advanced_machine_features", false);
        // backup_configuration - computed: true, optional: false, required: false
        this._backupConfiguration = new DataGoogleSqlDatabaseInstanceSettingsBackupConfigurationList(this, "backup_configuration", false);
        // data_cache_config - computed: true, optional: false, required: false
        this._dataCacheConfig = new DataGoogleSqlDatabaseInstanceSettingsDataCacheConfigList(this, "data_cache_config", false);
        // database_flags - computed: true, optional: false, required: false
        this._databaseFlags = new DataGoogleSqlDatabaseInstanceSettingsDatabaseFlagsList(this, "database_flags", true);
        // deny_maintenance_period - computed: true, optional: false, required: false
        this._denyMaintenancePeriod = new DataGoogleSqlDatabaseInstanceSettingsDenyMaintenancePeriodList(this, "deny_maintenance_period", false);
        // insights_config - computed: true, optional: false, required: false
        this._insightsConfig = new DataGoogleSqlDatabaseInstanceSettingsInsightsConfigList(this, "insights_config", false);
        // ip_configuration - computed: true, optional: false, required: false
        this._ipConfiguration = new DataGoogleSqlDatabaseInstanceSettingsIpConfigurationList(this, "ip_configuration", false);
        // location_preference - computed: true, optional: false, required: false
        this._locationPreference = new DataGoogleSqlDatabaseInstanceSettingsLocationPreferenceList(this, "location_preference", false);
        // maintenance_window - computed: true, optional: false, required: false
        this._maintenanceWindow = new DataGoogleSqlDatabaseInstanceSettingsMaintenanceWindowList(this, "maintenance_window", false);
        // password_validation_policy - computed: true, optional: false, required: false
        this._passwordValidationPolicy = new DataGoogleSqlDatabaseInstanceSettingsPasswordValidationPolicyList(this, "password_validation_policy", false);
        // sql_server_audit_config - computed: true, optional: false, required: false
        this._sqlServerAuditConfig = new DataGoogleSqlDatabaseInstanceSettingsSqlServerAuditConfigList(this, "sql_server_audit_config", false);
        // user_labels - computed: true, optional: false, required: false
        this._userLabels = new cdktf.StringMap(this, "user_labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // activation_policy - computed: true, optional: false, required: false
    get activationPolicy() {
        return this.getStringAttribute('activation_policy');
    }
    get activeDirectoryConfig() {
        return this._activeDirectoryConfig;
    }
    get advancedMachineFeatures() {
        return this._advancedMachineFeatures;
    }
    // availability_type - computed: true, optional: false, required: false
    get availabilityType() {
        return this.getStringAttribute('availability_type');
    }
    get backupConfiguration() {
        return this._backupConfiguration;
    }
    // collation - computed: true, optional: false, required: false
    get collation() {
        return this.getStringAttribute('collation');
    }
    // connector_enforcement - computed: true, optional: false, required: false
    get connectorEnforcement() {
        return this.getStringAttribute('connector_enforcement');
    }
    get dataCacheConfig() {
        return this._dataCacheConfig;
    }
    get databaseFlags() {
        return this._databaseFlags;
    }
    // deletion_protection_enabled - computed: true, optional: false, required: false
    get deletionProtectionEnabled() {
        return this.getBooleanAttribute('deletion_protection_enabled');
    }
    get denyMaintenancePeriod() {
        return this._denyMaintenancePeriod;
    }
    // disk_autoresize - computed: true, optional: false, required: false
    get diskAutoresize() {
        return this.getBooleanAttribute('disk_autoresize');
    }
    // disk_autoresize_limit - computed: true, optional: false, required: false
    get diskAutoresizeLimit() {
        return this.getNumberAttribute('disk_autoresize_limit');
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    // disk_type - computed: true, optional: false, required: false
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    // edition - computed: true, optional: false, required: false
    get edition() {
        return this.getStringAttribute('edition');
    }
    // enable_dataplex_integration - computed: true, optional: false, required: false
    get enableDataplexIntegration() {
        return this.getBooleanAttribute('enable_dataplex_integration');
    }
    // enable_google_ml_integration - computed: true, optional: false, required: false
    get enableGoogleMlIntegration() {
        return this.getBooleanAttribute('enable_google_ml_integration');
    }
    get insightsConfig() {
        return this._insightsConfig;
    }
    get ipConfiguration() {
        return this._ipConfiguration;
    }
    get locationPreference() {
        return this._locationPreference;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    get passwordValidationPolicy() {
        return this._passwordValidationPolicy;
    }
    // pricing_plan - computed: true, optional: false, required: false
    get pricingPlan() {
        return this.getStringAttribute('pricing_plan');
    }
    get sqlServerAuditConfig() {
        return this._sqlServerAuditConfig;
    }
    // tier - computed: true, optional: false, required: false
    get tier() {
        return this.getStringAttribute('tier');
    }
    // time_zone - computed: true, optional: false, required: false
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    get userLabels() {
        return this._userLabels;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsOutputReference = DataGoogleSqlDatabaseInstanceSettingsOutputReference;
_18 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsOutputReference[_18] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstanceSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstanceSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstanceSettingsList = DataGoogleSqlDatabaseInstanceSettingsList;
_19 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstanceSettingsList[_19] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstanceSettingsList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/sql_database_instance google_sql_database_instance}
*/
class DataGoogleSqlDatabaseInstance extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleSqlDatabaseInstance resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleSqlDatabaseInstance to import
    * @param importFromId The id of the existing DataGoogleSqlDatabaseInstance that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/sql_database_instance#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleSqlDatabaseInstance to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_sql_database_instance", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/sql_database_instance google_sql_database_instance} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleSqlDatabaseInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_sql_database_instance',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // clone - computed: true, optional: false, required: false
        this._clone = new DataGoogleSqlDatabaseInstanceCloneList(this, "clone", false);
        // ip_address - computed: true, optional: false, required: false
        this._ipAddress = new DataGoogleSqlDatabaseInstanceIpAddressList(this, "ip_address", false);
        // replica_configuration - computed: true, optional: false, required: false
        this._replicaConfiguration = new DataGoogleSqlDatabaseInstanceReplicaConfigurationList(this, "replica_configuration", false);
        // restore_backup_context - computed: true, optional: false, required: false
        this._restoreBackupContext = new DataGoogleSqlDatabaseInstanceRestoreBackupContextList(this, "restore_backup_context", false);
        // server_ca_cert - computed: true, optional: false, required: false
        this._serverCaCert = new DataGoogleSqlDatabaseInstanceServerCaCertList(this, "server_ca_cert", false);
        // settings - computed: true, optional: false, required: false
        this._settings = new DataGoogleSqlDatabaseInstanceSettingsList(this, "settings", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // available_maintenance_versions - computed: true, optional: false, required: false
    get availableMaintenanceVersions() {
        return this.getListAttribute('available_maintenance_versions');
    }
    get clone() {
        return this._clone;
    }
    // connection_name - computed: true, optional: false, required: false
    get connectionName() {
        return this.getStringAttribute('connection_name');
    }
    // database_version - computed: true, optional: false, required: false
    get databaseVersion() {
        return this.getStringAttribute('database_version');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // encryption_key_name - computed: true, optional: false, required: false
    get encryptionKeyName() {
        return this.getStringAttribute('encryption_key_name');
    }
    // first_ip_address - computed: true, optional: false, required: false
    get firstIpAddress() {
        return this.getStringAttribute('first_ip_address');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_type - computed: true, optional: false, required: false
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    get ipAddress() {
        return this._ipAddress;
    }
    // maintenance_version - computed: true, optional: false, required: false
    get maintenanceVersion() {
        return this.getStringAttribute('maintenance_version');
    }
    // master_instance_name - computed: true, optional: false, required: false
    get masterInstanceName() {
        return this.getStringAttribute('master_instance_name');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // private_ip_address - computed: true, optional: false, required: false
    get privateIpAddress() {
        return this.getStringAttribute('private_ip_address');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // psc_service_attachment_link - computed: true, optional: false, required: false
    get pscServiceAttachmentLink() {
        return this.getStringAttribute('psc_service_attachment_link');
    }
    // public_ip_address - computed: true, optional: false, required: false
    get publicIpAddress() {
        return this.getStringAttribute('public_ip_address');
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
    get replicaConfiguration() {
        return this._replicaConfiguration;
    }
    // replica_names - computed: true, optional: false, required: false
    get replicaNames() {
        return this.getListAttribute('replica_names');
    }
    get restoreBackupContext() {
        return this._restoreBackupContext;
    }
    // root_password - computed: true, optional: false, required: false
    get rootPassword() {
        return this.getStringAttribute('root_password');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get serverCaCert() {
        return this._serverCaCert;
    }
    // service_account_email_address - computed: true, optional: false, required: false
    get serviceAccountEmailAddress() {
        return this.getStringAttribute('service_account_email_address');
    }
    get settings() {
        return this._settings;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleSqlDatabaseInstance = DataGoogleSqlDatabaseInstance;
_20 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstance[_20] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstance.DataGoogleSqlDatabaseInstance", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleSqlDatabaseInstance.tfResourceType = "google_sql_database_instance";
