// @cdktf/provider-google/lib/vertex-ai-feature-group/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VertexAiFeatureGroup = exports.VertexAiFeatureGroupTimeoutsOutputReference = exports.vertexAiFeatureGroupTimeoutsToHclTerraform = exports.vertexAiFeatureGroupTimeoutsToTerraform = exports.VertexAiFeatureGroupBigQueryOutputReference = exports.vertexAiFeatureGroupBigQueryToHclTerraform = exports.vertexAiFeatureGroupBigQueryToTerraform = exports.VertexAiFeatureGroupBigQueryBigQuerySourceOutputReference = exports.vertexAiFeatureGroupBigQueryBigQuerySourceToHclTerraform = exports.vertexAiFeatureGroupBigQueryBigQuerySourceToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vertexAiFeatureGroupBigQueryBigQuerySourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        input_uri: cdktf.stringToTerraform(struct.inputUri),
    };
}
exports.vertexAiFeatureGroupBigQueryBigQuerySourceToTerraform = vertexAiFeatureGroupBigQueryBigQuerySourceToTerraform;
function vertexAiFeatureGroupBigQueryBigQuerySourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        input_uri: {
            value: cdktf.stringToHclTerraform(struct.inputUri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiFeatureGroupBigQueryBigQuerySourceToHclTerraform = vertexAiFeatureGroupBigQueryBigQuerySourceToHclTerraform;
class VertexAiFeatureGroupBigQueryBigQuerySourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._inputUri !== undefined) {
            hasAnyValues = true;
            internalValueResult.inputUri = this._inputUri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._inputUri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._inputUri = value.inputUri;
        }
    }
    get inputUri() {
        return this.getStringAttribute('input_uri');
    }
    set inputUri(value) {
        this._inputUri = value;
    }
    // Temporarily expose input value. Use with caution.
    get inputUriInput() {
        return this._inputUri;
    }
}
exports.VertexAiFeatureGroupBigQueryBigQuerySourceOutputReference = VertexAiFeatureGroupBigQueryBigQuerySourceOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VertexAiFeatureGroupBigQueryBigQuerySourceOutputReference[_a] = { fqn: "@cdktf/provider-google.vertexAiFeatureGroup.VertexAiFeatureGroupBigQueryBigQuerySourceOutputReference", version: "14.13.1" };
function vertexAiFeatureGroupBigQueryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        entity_id_columns: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.entityIdColumns),
        big_query_source: vertexAiFeatureGroupBigQueryBigQuerySourceToTerraform(struct.bigQuerySource),
    };
}
exports.vertexAiFeatureGroupBigQueryToTerraform = vertexAiFeatureGroupBigQueryToTerraform;
function vertexAiFeatureGroupBigQueryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        entity_id_columns: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.entityIdColumns),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        big_query_source: {
            value: vertexAiFeatureGroupBigQueryBigQuerySourceToHclTerraform(struct.bigQuerySource),
            isBlock: true,
            type: "list",
            storageClassType: "VertexAiFeatureGroupBigQueryBigQuerySourceList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiFeatureGroupBigQueryToHclTerraform = vertexAiFeatureGroupBigQueryToHclTerraform;
class VertexAiFeatureGroupBigQueryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // big_query_source - computed: false, optional: false, required: true
        this._bigQuerySource = new VertexAiFeatureGroupBigQueryBigQuerySourceOutputReference(this, "big_query_source");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._entityIdColumns !== undefined) {
            hasAnyValues = true;
            internalValueResult.entityIdColumns = this._entityIdColumns;
        }
        if (this._bigQuerySource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.bigQuerySource = this._bigQuerySource?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._entityIdColumns = undefined;
            this._bigQuerySource.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._entityIdColumns = value.entityIdColumns;
            this._bigQuerySource.internalValue = value.bigQuerySource;
        }
    }
    get entityIdColumns() {
        return this.getListAttribute('entity_id_columns');
    }
    set entityIdColumns(value) {
        this._entityIdColumns = value;
    }
    resetEntityIdColumns() {
        this._entityIdColumns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get entityIdColumnsInput() {
        return this._entityIdColumns;
    }
    get bigQuerySource() {
        return this._bigQuerySource;
    }
    putBigQuerySource(value) {
        this._bigQuerySource.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get bigQuerySourceInput() {
        return this._bigQuerySource.internalValue;
    }
}
exports.VertexAiFeatureGroupBigQueryOutputReference = VertexAiFeatureGroupBigQueryOutputReference;
_b = JSII_RTTI_SYMBOL_1;
VertexAiFeatureGroupBigQueryOutputReference[_b] = { fqn: "@cdktf/provider-google.vertexAiFeatureGroup.VertexAiFeatureGroupBigQueryOutputReference", version: "14.13.1" };
function vertexAiFeatureGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.vertexAiFeatureGroupTimeoutsToTerraform = vertexAiFeatureGroupTimeoutsToTerraform;
function vertexAiFeatureGroupTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiFeatureGroupTimeoutsToHclTerraform = vertexAiFeatureGroupTimeoutsToHclTerraform;
class VertexAiFeatureGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VertexAiFeatureGroupTimeoutsOutputReference = VertexAiFeatureGroupTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
VertexAiFeatureGroupTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-google.vertexAiFeatureGroup.VertexAiFeatureGroupTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_feature_group google_vertex_ai_feature_group}
*/
class VertexAiFeatureGroup extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VertexAiFeatureGroup resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VertexAiFeatureGroup to import
    * @param importFromId The id of the existing VertexAiFeatureGroup that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_feature_group#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VertexAiFeatureGroup to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_vertex_ai_feature_group", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_feature_group google_vertex_ai_feature_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VertexAiFeatureGroupConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_vertex_ai_feature_group',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // big_query - computed: false, optional: true, required: false
        this._bigQuery = new VertexAiFeatureGroupBigQueryOutputReference(this, "big_query");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VertexAiFeatureGroupTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
        this._bigQuery.internalValue = config.bigQuery;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get bigQuery() {
        return this._bigQuery;
    }
    putBigQuery(value) {
        this._bigQuery.internalValue = value;
    }
    resetBigQuery() {
        this._bigQuery.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigQueryInput() {
        return this._bigQuery.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            big_query: vertexAiFeatureGroupBigQueryToTerraform(this._bigQuery.internalValue),
            timeouts: vertexAiFeatureGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            big_query: {
                value: vertexAiFeatureGroupBigQueryToHclTerraform(this._bigQuery.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VertexAiFeatureGroupBigQueryList",
            },
            timeouts: {
                value: vertexAiFeatureGroupTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "VertexAiFeatureGroupTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VertexAiFeatureGroup = VertexAiFeatureGroup;
_d = JSII_RTTI_SYMBOL_1;
VertexAiFeatureGroup[_d] = { fqn: "@cdktf/provider-google.vertexAiFeatureGroup.VertexAiFeatureGroup", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
VertexAiFeatureGroup.tfResourceType = "google_vertex_ai_feature_group";
