// @cdktf/provider-google/lib/data-google-compute-region-instance-group-manager/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulList = exports.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerStatusStatefulToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerStatusStatefulToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList = exports.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigList = exports.DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpList = exports.DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpList = exports.DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerStatefulDiskList = exports.DataGoogleComputeRegionInstanceGroupManagerStatefulDiskOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerStatefulDiskToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerStatefulDiskToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerNamedPortList = exports.DataGoogleComputeRegionInstanceGroupManagerNamedPortOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerNamedPortToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerNamedPortToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyList = exports.DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyList = exports.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsList = exports.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesList = exports.DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigList = exports.DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigToTerraform = void 0;
exports.DataGoogleComputeRegionInstanceGroupManager = exports.DataGoogleComputeRegionInstanceGroupManagerVersionList = exports.DataGoogleComputeRegionInstanceGroupManagerVersionOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerVersionToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerVersionToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeList = exports.DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyList = exports.DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerUpdatePolicyToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerUpdatePolicyToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerStatusList = exports.DataGoogleComputeRegionInstanceGroupManagerStatusOutputReference = exports.dataGoogleComputeRegionInstanceGroupManagerStatusToHclTerraform = exports.dataGoogleComputeRegionInstanceGroupManagerStatusToTerraform = exports.DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetList = exports.DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigToTerraform = dataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // metadata - computed: true, optional: false, required: false
        this._metadata = new cdktf.StringMap(this, "metadata");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get labels() {
        return this._labels;
    }
    get metadata() {
        return this._metadata;
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigOutputReference = DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigList = DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform = dataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // health_check - computed: true, optional: false, required: false
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    // initial_delay_sec - computed: true, optional: false, required: false
    get initialDelaySec() {
        return this.getNumberAttribute('initial_delay_sec');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference = DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesList = DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsToTerraform = dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // machine_types - computed: true, optional: false, required: false
    get machineTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('machine_types'));
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // rank - computed: true, optional: false, required: false
    get rank() {
        return this.getNumberAttribute('rank');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsOutputReference = DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsList = DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyToTerraform = dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // instance_selections - computed: true, optional: false, required: false
        this._instanceSelections = new DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyInstanceSelectionsList(this, "instance_selections", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get instanceSelections() {
        return this._instanceSelections;
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyOutputReference = DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyList = DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyToTerraform = dataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // default_action_on_failure - computed: true, optional: false, required: false
    get defaultActionOnFailure() {
        return this.getStringAttribute('default_action_on_failure');
    }
    // force_update_on_repair - computed: true, optional: false, required: false
    get forceUpdateOnRepair() {
        return this.getStringAttribute('force_update_on_repair');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyOutputReference = DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyList = DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerNamedPortToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerNamedPortToTerraform = dataGoogleComputeRegionInstanceGroupManagerNamedPortToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerNamedPortToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerNamedPortToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerNamedPortToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerNamedPortOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerNamedPortOutputReference = DataGoogleComputeRegionInstanceGroupManagerNamedPortOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerNamedPortOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerNamedPortOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerNamedPortList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerNamedPortOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerNamedPortList = DataGoogleComputeRegionInstanceGroupManagerNamedPortList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerNamedPortList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerNamedPortList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerStatefulDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatefulDiskToTerraform = dataGoogleComputeRegionInstanceGroupManagerStatefulDiskToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerStatefulDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatefulDiskToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerStatefulDiskToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerStatefulDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_rule - computed: true, optional: false, required: false
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatefulDiskOutputReference = DataGoogleComputeRegionInstanceGroupManagerStatefulDiskOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatefulDiskOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatefulDiskOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerStatefulDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerStatefulDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatefulDiskList = DataGoogleComputeRegionInstanceGroupManagerStatefulDiskList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatefulDiskList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatefulDiskList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpToTerraform = dataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_rule - computed: true, optional: false, required: false
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    // interface_name - computed: true, optional: false, required: false
    get interfaceName() {
        return this.getStringAttribute('interface_name');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpOutputReference = DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpList = DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpList[_r] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpToTerraform = dataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delete_rule - computed: true, optional: false, required: false
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    // interface_name - computed: true, optional: false, required: false
    get interfaceName() {
        return this.getStringAttribute('interface_name');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpOutputReference = DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpList = DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpList[_t] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigToTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // current_revision - computed: true, optional: false, required: false
    get currentRevision() {
        return this.getStringAttribute('current_revision');
    }
    // effective - computed: true, optional: false, required: false
    get effective() {
        return this.getBooleanAttribute('effective');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigOutputReference = DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigList = DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigList[_v] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all_effective - computed: true, optional: false, required: false
    get allEffective() {
        return this.getBooleanAttribute('all_effective');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList = DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList[_x] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerStatusStatefulToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusStatefulToTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusStatefulToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerStatusStatefulToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusStatefulToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusStatefulToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerStatusStatefulOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // per_instance_configs - computed: true, optional: false, required: false
        this._perInstanceConfigs = new DataGoogleComputeRegionInstanceGroupManagerStatusStatefulPerInstanceConfigsList(this, "per_instance_configs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // has_stateful_config - computed: true, optional: false, required: false
    get hasStatefulConfig() {
        return this.getBooleanAttribute('has_stateful_config');
    }
    get perInstanceConfigs() {
        return this._perInstanceConfigs;
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulOutputReference = DataGoogleComputeRegionInstanceGroupManagerStatusStatefulOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusStatefulOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerStatusStatefulList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerStatusStatefulOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulList = DataGoogleComputeRegionInstanceGroupManagerStatusStatefulList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusStatefulList[_z] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusStatefulList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetToTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_reached - computed: true, optional: false, required: false
    get isReached() {
        return this.getBooleanAttribute('is_reached');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference = DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetList = DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetList[_1] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusToTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerStatusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerStatusToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerStatusToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // all_instances_config - computed: true, optional: false, required: false
        this._allInstancesConfig = new DataGoogleComputeRegionInstanceGroupManagerStatusAllInstancesConfigList(this, "all_instances_config", false);
        // stateful - computed: true, optional: false, required: false
        this._stateful = new DataGoogleComputeRegionInstanceGroupManagerStatusStatefulList(this, "stateful", false);
        // version_target - computed: true, optional: false, required: false
        this._versionTarget = new DataGoogleComputeRegionInstanceGroupManagerStatusVersionTargetList(this, "version_target", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get allInstancesConfig() {
        return this._allInstancesConfig;
    }
    // is_stable - computed: true, optional: false, required: false
    get isStable() {
        return this.getBooleanAttribute('is_stable');
    }
    get stateful() {
        return this._stateful;
    }
    get versionTarget() {
        return this._versionTarget;
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusOutputReference = DataGoogleComputeRegionInstanceGroupManagerStatusOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerStatusList = DataGoogleComputeRegionInstanceGroupManagerStatusList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerStatusList[_3] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerStatusList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerUpdatePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerUpdatePolicyToTerraform = dataGoogleComputeRegionInstanceGroupManagerUpdatePolicyToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerUpdatePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerUpdatePolicyToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerUpdatePolicyToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // instance_redistribution_type - computed: true, optional: false, required: false
    get instanceRedistributionType() {
        return this.getStringAttribute('instance_redistribution_type');
    }
    // max_surge_fixed - computed: true, optional: false, required: false
    get maxSurgeFixed() {
        return this.getNumberAttribute('max_surge_fixed');
    }
    // max_surge_percent - computed: true, optional: false, required: false
    get maxSurgePercent() {
        return this.getNumberAttribute('max_surge_percent');
    }
    // max_unavailable_fixed - computed: true, optional: false, required: false
    get maxUnavailableFixed() {
        return this.getNumberAttribute('max_unavailable_fixed');
    }
    // max_unavailable_percent - computed: true, optional: false, required: false
    get maxUnavailablePercent() {
        return this.getNumberAttribute('max_unavailable_percent');
    }
    // minimal_action - computed: true, optional: false, required: false
    get minimalAction() {
        return this.getStringAttribute('minimal_action');
    }
    // most_disruptive_allowed_action - computed: true, optional: false, required: false
    get mostDisruptiveAllowedAction() {
        return this.getStringAttribute('most_disruptive_allowed_action');
    }
    // replacement_method - computed: true, optional: false, required: false
    get replacementMethod() {
        return this.getStringAttribute('replacement_method');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyOutputReference = DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyList = DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyList[_5] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeToTerraform = dataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // fixed - computed: true, optional: false, required: false
    get fixed() {
        return this.getNumberAttribute('fixed');
    }
    // percent - computed: true, optional: false, required: false
    get percent() {
        return this.getNumberAttribute('percent');
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference = DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeList = DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeList[_7] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceGroupManagerVersionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceGroupManagerVersionToTerraform = dataGoogleComputeRegionInstanceGroupManagerVersionToTerraform;
function dataGoogleComputeRegionInstanceGroupManagerVersionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceGroupManagerVersionToHclTerraform = dataGoogleComputeRegionInstanceGroupManagerVersionToHclTerraform;
class DataGoogleComputeRegionInstanceGroupManagerVersionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // target_size - computed: true, optional: false, required: false
        this._targetSize = new DataGoogleComputeRegionInstanceGroupManagerVersionTargetSizeList(this, "target_size", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // instance_template - computed: true, optional: false, required: false
    get instanceTemplate() {
        return this.getStringAttribute('instance_template');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get targetSize() {
        return this._targetSize;
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerVersionOutputReference = DataGoogleComputeRegionInstanceGroupManagerVersionOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerVersionOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerVersionOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceGroupManagerVersionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceGroupManagerVersionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceGroupManagerVersionList = DataGoogleComputeRegionInstanceGroupManagerVersionList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManagerVersionList[_9] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManagerVersionList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_region_instance_group_manager google_compute_region_instance_group_manager}
*/
class DataGoogleComputeRegionInstanceGroupManager extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeRegionInstanceGroupManager resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeRegionInstanceGroupManager to import
    * @param importFromId The id of the existing DataGoogleComputeRegionInstanceGroupManager that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_region_instance_group_manager#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeRegionInstanceGroupManager to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_region_instance_group_manager", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_region_instance_group_manager google_compute_region_instance_group_manager} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeRegionInstanceGroupManagerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_region_instance_group_manager',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // all_instances_config - computed: true, optional: false, required: false
        this._allInstancesConfig = new DataGoogleComputeRegionInstanceGroupManagerAllInstancesConfigList(this, "all_instances_config", false);
        // auto_healing_policies - computed: true, optional: false, required: false
        this._autoHealingPolicies = new DataGoogleComputeRegionInstanceGroupManagerAutoHealingPoliciesList(this, "auto_healing_policies", false);
        // instance_flexibility_policy - computed: true, optional: false, required: false
        this._instanceFlexibilityPolicy = new DataGoogleComputeRegionInstanceGroupManagerInstanceFlexibilityPolicyList(this, "instance_flexibility_policy", false);
        // instance_lifecycle_policy - computed: true, optional: false, required: false
        this._instanceLifecyclePolicy = new DataGoogleComputeRegionInstanceGroupManagerInstanceLifecyclePolicyList(this, "instance_lifecycle_policy", false);
        // named_port - computed: true, optional: false, required: false
        this._namedPort = new DataGoogleComputeRegionInstanceGroupManagerNamedPortList(this, "named_port", true);
        // stateful_disk - computed: true, optional: false, required: false
        this._statefulDisk = new DataGoogleComputeRegionInstanceGroupManagerStatefulDiskList(this, "stateful_disk", true);
        // stateful_external_ip - computed: true, optional: false, required: false
        this._statefulExternalIp = new DataGoogleComputeRegionInstanceGroupManagerStatefulExternalIpList(this, "stateful_external_ip", false);
        // stateful_internal_ip - computed: true, optional: false, required: false
        this._statefulInternalIp = new DataGoogleComputeRegionInstanceGroupManagerStatefulInternalIpList(this, "stateful_internal_ip", false);
        // status - computed: true, optional: false, required: false
        this._status = new DataGoogleComputeRegionInstanceGroupManagerStatusList(this, "status", false);
        // update_policy - computed: true, optional: false, required: false
        this._updatePolicy = new DataGoogleComputeRegionInstanceGroupManagerUpdatePolicyList(this, "update_policy", false);
        // version - computed: true, optional: false, required: false
        this._version = new DataGoogleComputeRegionInstanceGroupManagerVersionList(this, "version", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
        this._selfLink = config.selfLink;
    }
    get allInstancesConfig() {
        return this._allInstancesConfig;
    }
    get autoHealingPolicies() {
        return this._autoHealingPolicies;
    }
    // base_instance_name - computed: true, optional: false, required: false
    get baseInstanceName() {
        return this.getStringAttribute('base_instance_name');
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // distribution_policy_target_shape - computed: true, optional: false, required: false
    get distributionPolicyTargetShape() {
        return this.getStringAttribute('distribution_policy_target_shape');
    }
    // distribution_policy_zones - computed: true, optional: false, required: false
    get distributionPolicyZones() {
        return cdktf.Fn.tolist(this.getListAttribute('distribution_policy_zones'));
    }
    // fingerprint - computed: true, optional: false, required: false
    get fingerprint() {
        return this.getStringAttribute('fingerprint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceFlexibilityPolicy() {
        return this._instanceFlexibilityPolicy;
    }
    // instance_group - computed: true, optional: false, required: false
    get instanceGroup() {
        return this.getStringAttribute('instance_group');
    }
    // instance_group_manager_id - computed: true, optional: false, required: false
    get instanceGroupManagerId() {
        return this.getNumberAttribute('instance_group_manager_id');
    }
    get instanceLifecyclePolicy() {
        return this._instanceLifecyclePolicy;
    }
    // list_managed_instances_results - computed: true, optional: false, required: false
    get listManagedInstancesResults() {
        return this.getStringAttribute('list_managed_instances_results');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namedPort() {
        return this._namedPort;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    set selfLink(value) {
        this._selfLink = value;
    }
    resetSelfLink() {
        this._selfLink = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selfLinkInput() {
        return this._selfLink;
    }
    get statefulDisk() {
        return this._statefulDisk;
    }
    get statefulExternalIp() {
        return this._statefulExternalIp;
    }
    get statefulInternalIp() {
        return this._statefulInternalIp;
    }
    get status() {
        return this._status;
    }
    // target_pools - computed: true, optional: false, required: false
    get targetPools() {
        return cdktf.Fn.tolist(this.getListAttribute('target_pools'));
    }
    // target_size - computed: true, optional: false, required: false
    get targetSize() {
        return this.getNumberAttribute('target_size');
    }
    get updatePolicy() {
        return this._updatePolicy;
    }
    get version() {
        return this._version;
    }
    // wait_for_instances - computed: true, optional: false, required: false
    get waitForInstances() {
        return this.getBooleanAttribute('wait_for_instances');
    }
    // wait_for_instances_status - computed: true, optional: false, required: false
    get waitForInstancesStatus() {
        return this.getStringAttribute('wait_for_instances_status');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            self_link: cdktf.stringToTerraform(this._selfLink),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            self_link: {
                value: cdktf.stringToHclTerraform(this._selfLink),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeRegionInstanceGroupManager = DataGoogleComputeRegionInstanceGroupManager;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceGroupManager[_10] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceGroupManager.DataGoogleComputeRegionInstanceGroupManager", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeRegionInstanceGroupManager.tfResourceType = "google_compute_region_instance_group_manager";
