// @cdktf/provider-google/lib/compute-security-policy-rule/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3;
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdAOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigAOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigAToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigAToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToTerraform = exports.ComputeSecurityPolicyRuleMatchAOutputReference = exports.computeSecurityPolicyRuleMatchAToHclTerraform = exports.computeSecurityPolicyRuleMatchAToTerraform = exports.ComputeSecurityPolicyRuleMatchExprOptionsAOutputReference = exports.computeSecurityPolicyRuleMatchExprOptionsAToHclTerraform = exports.computeSecurityPolicyRuleMatchExprOptionsAToTerraform = exports.ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAOutputReference = exports.computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToHclTerraform = exports.computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToTerraform = exports.ComputeSecurityPolicyRuleMatchExprAOutputReference = exports.computeSecurityPolicyRuleMatchExprAToHclTerraform = exports.computeSecurityPolicyRuleMatchExprAToTerraform = exports.ComputeSecurityPolicyRuleMatchConfigAOutputReference = exports.computeSecurityPolicyRuleMatchConfigAToHclTerraform = exports.computeSecurityPolicyRuleMatchConfigAToTerraform = exports.ComputeSecurityPolicyRuleHeaderActionAOutputReference = exports.computeSecurityPolicyRuleHeaderActionAToHclTerraform = exports.computeSecurityPolicyRuleHeaderActionAToTerraform = exports.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAList = exports.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAOutputReference = exports.computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToHclTerraform = exports.computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToTerraform = void 0;
exports.ComputeSecurityPolicyRuleA = exports.ComputeSecurityPolicyRuleTimeoutsOutputReference = exports.computeSecurityPolicyRuleTimeoutsToHclTerraform = exports.computeSecurityPolicyRuleTimeoutsToTerraform = exports.ComputeSecurityPolicyRuleRedirectOptionsAOutputReference = exports.computeSecurityPolicyRuleRedirectOptionsAToHclTerraform = exports.computeSecurityPolicyRuleRedirectOptionsAToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsAOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsAToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsAToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAList = exports.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        header_name: cdktf.stringToTerraform(struct.headerName),
        header_value: cdktf.stringToTerraform(struct.headerValue),
    };
}
exports.computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToTerraform = computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToTerraform;
function computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        header_name: {
            value: cdktf.stringToHclTerraform(struct.headerName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        header_value: {
            value: cdktf.stringToHclTerraform(struct.headerValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToHclTerraform = computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToHclTerraform;
class ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._headerName !== undefined) {
            hasAnyValues = true;
            internalValueResult.headerName = this._headerName;
        }
        if (this._headerValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.headerValue = this._headerValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._headerName = undefined;
            this._headerValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._headerName = value.headerName;
            this._headerValue = value.headerValue;
        }
    }
    get headerName() {
        return this.getStringAttribute('header_name');
    }
    set headerName(value) {
        this._headerName = value;
    }
    resetHeaderName() {
        this._headerName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerNameInput() {
        return this._headerName;
    }
    get headerValue() {
        return this.getStringAttribute('header_value');
    }
    set headerValue(value) {
        this._headerValue = value;
    }
    resetHeaderValue() {
        this._headerValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerValueInput() {
        return this._headerValue;
    }
}
exports.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAOutputReference = ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAOutputReference[_a] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAList = ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAList;
_b = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAList[_b] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAList", version: "14.13.1" };
function computeSecurityPolicyRuleHeaderActionAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        request_headers_to_adds: cdktf.listMapper(computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToTerraform, true)(struct.requestHeadersToAdds),
    };
}
exports.computeSecurityPolicyRuleHeaderActionAToTerraform = computeSecurityPolicyRuleHeaderActionAToTerraform;
function computeSecurityPolicyRuleHeaderActionAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        request_headers_to_adds: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAToHclTerraform, true)(struct.requestHeadersToAdds),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleHeaderActionAToHclTerraform = computeSecurityPolicyRuleHeaderActionAToHclTerraform;
class ComputeSecurityPolicyRuleHeaderActionAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // request_headers_to_adds - computed: false, optional: true, required: false
        this._requestHeadersToAdds = new ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsAList(this, "request_headers_to_adds", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._requestHeadersToAdds?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestHeadersToAdds = this._requestHeadersToAdds?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._requestHeadersToAdds.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._requestHeadersToAdds.internalValue = value.requestHeadersToAdds;
        }
    }
    get requestHeadersToAdds() {
        return this._requestHeadersToAdds;
    }
    putRequestHeadersToAdds(value) {
        this._requestHeadersToAdds.internalValue = value;
    }
    resetRequestHeadersToAdds() {
        this._requestHeadersToAdds.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestHeadersToAddsInput() {
        return this._requestHeadersToAdds.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleHeaderActionAOutputReference = ComputeSecurityPolicyRuleHeaderActionAOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleHeaderActionAOutputReference[_c] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleHeaderActionAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        src_ip_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.srcIpRanges),
    };
}
exports.computeSecurityPolicyRuleMatchConfigAToTerraform = computeSecurityPolicyRuleMatchConfigAToTerraform;
function computeSecurityPolicyRuleMatchConfigAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        src_ip_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.srcIpRanges),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchConfigAToHclTerraform = computeSecurityPolicyRuleMatchConfigAToHclTerraform;
class ComputeSecurityPolicyRuleMatchConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._srcIpRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.srcIpRanges = this._srcIpRanges;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._srcIpRanges = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._srcIpRanges = value.srcIpRanges;
        }
    }
    get srcIpRanges() {
        return this.getListAttribute('src_ip_ranges');
    }
    set srcIpRanges(value) {
        this._srcIpRanges = value;
    }
    resetSrcIpRanges() {
        this._srcIpRanges = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get srcIpRangesInput() {
        return this._srcIpRanges;
    }
}
exports.ComputeSecurityPolicyRuleMatchConfigAOutputReference = ComputeSecurityPolicyRuleMatchConfigAOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchConfigAOutputReference[_d] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleMatchConfigAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchExprAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expression: cdktf.stringToTerraform(struct.expression),
    };
}
exports.computeSecurityPolicyRuleMatchExprAToTerraform = computeSecurityPolicyRuleMatchExprAToTerraform;
function computeSecurityPolicyRuleMatchExprAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchExprAToHclTerraform = computeSecurityPolicyRuleMatchExprAToHclTerraform;
class ComputeSecurityPolicyRuleMatchExprAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._expression = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._expression = value.expression;
        }
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
}
exports.ComputeSecurityPolicyRuleMatchExprAOutputReference = ComputeSecurityPolicyRuleMatchExprAOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchExprAOutputReference[_e] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleMatchExprAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action_token_site_keys: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.actionTokenSiteKeys),
        session_token_site_keys: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.sessionTokenSiteKeys),
    };
}
exports.computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToTerraform = computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToTerraform;
function computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action_token_site_keys: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.actionTokenSiteKeys),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        session_token_site_keys: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.sessionTokenSiteKeys),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToHclTerraform = computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToHclTerraform;
class ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._actionTokenSiteKeys !== undefined) {
            hasAnyValues = true;
            internalValueResult.actionTokenSiteKeys = this._actionTokenSiteKeys;
        }
        if (this._sessionTokenSiteKeys !== undefined) {
            hasAnyValues = true;
            internalValueResult.sessionTokenSiteKeys = this._sessionTokenSiteKeys;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._actionTokenSiteKeys = undefined;
            this._sessionTokenSiteKeys = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._actionTokenSiteKeys = value.actionTokenSiteKeys;
            this._sessionTokenSiteKeys = value.sessionTokenSiteKeys;
        }
    }
    get actionTokenSiteKeys() {
        return this.getListAttribute('action_token_site_keys');
    }
    set actionTokenSiteKeys(value) {
        this._actionTokenSiteKeys = value;
    }
    resetActionTokenSiteKeys() {
        this._actionTokenSiteKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get actionTokenSiteKeysInput() {
        return this._actionTokenSiteKeys;
    }
    get sessionTokenSiteKeys() {
        return this.getListAttribute('session_token_site_keys');
    }
    set sessionTokenSiteKeys(value) {
        this._sessionTokenSiteKeys = value;
    }
    resetSessionTokenSiteKeys() {
        this._sessionTokenSiteKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionTokenSiteKeysInput() {
        return this._sessionTokenSiteKeys;
    }
}
exports.ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAOutputReference = ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAOutputReference[_f] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchExprOptionsAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        recaptcha_options: computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToTerraform(struct.recaptchaOptions),
    };
}
exports.computeSecurityPolicyRuleMatchExprOptionsAToTerraform = computeSecurityPolicyRuleMatchExprOptionsAToTerraform;
function computeSecurityPolicyRuleMatchExprOptionsAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        recaptcha_options: {
            value: computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAToHclTerraform(struct.recaptchaOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchExprOptionsAToHclTerraform = computeSecurityPolicyRuleMatchExprOptionsAToHclTerraform;
class ComputeSecurityPolicyRuleMatchExprOptionsAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // recaptcha_options - computed: false, optional: false, required: true
        this._recaptchaOptions = new ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsAOutputReference(this, "recaptcha_options");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recaptchaOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.recaptchaOptions = this._recaptchaOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recaptchaOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recaptchaOptions.internalValue = value.recaptchaOptions;
        }
    }
    get recaptchaOptions() {
        return this._recaptchaOptions;
    }
    putRecaptchaOptions(value) {
        this._recaptchaOptions.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get recaptchaOptionsInput() {
        return this._recaptchaOptions.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleMatchExprOptionsAOutputReference = ComputeSecurityPolicyRuleMatchExprOptionsAOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchExprOptionsAOutputReference[_g] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleMatchExprOptionsAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        versioned_expr: cdktf.stringToTerraform(struct.versionedExpr),
        config: computeSecurityPolicyRuleMatchConfigAToTerraform(struct.config),
        expr: computeSecurityPolicyRuleMatchExprAToTerraform(struct.expr),
        expr_options: computeSecurityPolicyRuleMatchExprOptionsAToTerraform(struct.exprOptions),
    };
}
exports.computeSecurityPolicyRuleMatchAToTerraform = computeSecurityPolicyRuleMatchAToTerraform;
function computeSecurityPolicyRuleMatchAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        versioned_expr: {
            value: cdktf.stringToHclTerraform(struct.versionedExpr),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        config: {
            value: computeSecurityPolicyRuleMatchConfigAToHclTerraform(struct.config),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchConfigAList",
        },
        expr: {
            value: computeSecurityPolicyRuleMatchExprAToHclTerraform(struct.expr),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchExprAList",
        },
        expr_options: {
            value: computeSecurityPolicyRuleMatchExprOptionsAToHclTerraform(struct.exprOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchExprOptionsAList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchAToHclTerraform = computeSecurityPolicyRuleMatchAToHclTerraform;
class ComputeSecurityPolicyRuleMatchAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // config - computed: false, optional: true, required: false
        this._config = new ComputeSecurityPolicyRuleMatchConfigAOutputReference(this, "config");
        // expr - computed: false, optional: true, required: false
        this._expr = new ComputeSecurityPolicyRuleMatchExprAOutputReference(this, "expr");
        // expr_options - computed: false, optional: true, required: false
        this._exprOptions = new ComputeSecurityPolicyRuleMatchExprOptionsAOutputReference(this, "expr_options");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._versionedExpr !== undefined) {
            hasAnyValues = true;
            internalValueResult.versionedExpr = this._versionedExpr;
        }
        if (this._config?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.config = this._config?.internalValue;
        }
        if (this._expr?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.expr = this._expr?.internalValue;
        }
        if (this._exprOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.exprOptions = this._exprOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._versionedExpr = undefined;
            this._config.internalValue = undefined;
            this._expr.internalValue = undefined;
            this._exprOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._versionedExpr = value.versionedExpr;
            this._config.internalValue = value.config;
            this._expr.internalValue = value.expr;
            this._exprOptions.internalValue = value.exprOptions;
        }
    }
    get versionedExpr() {
        return this.getStringAttribute('versioned_expr');
    }
    set versionedExpr(value) {
        this._versionedExpr = value;
    }
    resetVersionedExpr() {
        this._versionedExpr = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionedExprInput() {
        return this._versionedExpr;
    }
    get config() {
        return this._config;
    }
    putConfig(value) {
        this._config.internalValue = value;
    }
    resetConfig() {
        this._config.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configInput() {
        return this._config.internalValue;
    }
    get expr() {
        return this._expr;
    }
    putExpr(value) {
        this._expr.internalValue = value;
    }
    resetExpr() {
        this._expr.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exprInput() {
        return this._expr.internalValue;
    }
    get exprOptions() {
        return this._exprOptions;
    }
    putExprOptions(value) {
        this._exprOptions.internalValue = value;
    }
    resetExprOptions() {
        this._exprOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exprOptionsInput() {
        return this._exprOptions.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleMatchAOutputReference = ComputeSecurityPolicyRuleMatchAOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchAOutputReference[_h] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleMatchAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operator: cdktf.stringToTerraform(struct.operator),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._operator = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._operator = value.operator;
            this._value = value.value;
        }
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAOutputReference[_j] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAList;
_k = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAList[_k] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operator: cdktf.stringToTerraform(struct.operator),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._operator = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._operator = value.operator;
            this._value = value.value;
        }
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAOutputReference[_l] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAList;
_m = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAList[_m] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operator: cdktf.stringToTerraform(struct.operator),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._operator = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._operator = value.operator;
            this._value = value.value;
        }
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAOutputReference[_o] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAList;
_p = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAList[_p] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operator: cdktf.stringToTerraform(struct.operator),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._operator = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._operator = value.operator;
            this._value = value.value;
        }
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAOutputReference[_q] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAList;
_r = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAList[_r] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target_rule_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.targetRuleIds),
        target_rule_set: cdktf.stringToTerraform(struct.targetRuleSet),
        request_cookie: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToTerraform, true)(struct.requestCookie),
        request_header: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToTerraform, true)(struct.requestHeader),
        request_query_param: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToTerraform, true)(struct.requestQueryParam),
        request_uri: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToTerraform, true)(struct.requestUri),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target_rule_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.targetRuleIds),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        target_rule_set: {
            value: cdktf.stringToHclTerraform(struct.targetRuleSet),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        request_cookie: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAToHclTerraform, true)(struct.requestCookie),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAList",
        },
        request_header: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAToHclTerraform, true)(struct.requestHeader),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAList",
        },
        request_query_param: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAToHclTerraform, true)(struct.requestQueryParam),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAList",
        },
        request_uri: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAToHclTerraform, true)(struct.requestUri),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // request_cookie - computed: false, optional: true, required: false
        this._requestCookie = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieAList(this, "request_cookie", false);
        // request_header - computed: false, optional: true, required: false
        this._requestHeader = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderAList(this, "request_header", false);
        // request_query_param - computed: false, optional: true, required: false
        this._requestQueryParam = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamAList(this, "request_query_param", false);
        // request_uri - computed: false, optional: true, required: false
        this._requestUri = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriAList(this, "request_uri", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._targetRuleIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetRuleIds = this._targetRuleIds;
        }
        if (this._targetRuleSet !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetRuleSet = this._targetRuleSet;
        }
        if (this._requestCookie?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestCookie = this._requestCookie?.internalValue;
        }
        if (this._requestHeader?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestHeader = this._requestHeader?.internalValue;
        }
        if (this._requestQueryParam?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestQueryParam = this._requestQueryParam?.internalValue;
        }
        if (this._requestUri?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestUri = this._requestUri?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._targetRuleIds = undefined;
            this._targetRuleSet = undefined;
            this._requestCookie.internalValue = undefined;
            this._requestHeader.internalValue = undefined;
            this._requestQueryParam.internalValue = undefined;
            this._requestUri.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._targetRuleIds = value.targetRuleIds;
            this._targetRuleSet = value.targetRuleSet;
            this._requestCookie.internalValue = value.requestCookie;
            this._requestHeader.internalValue = value.requestHeader;
            this._requestQueryParam.internalValue = value.requestQueryParam;
            this._requestUri.internalValue = value.requestUri;
        }
    }
    get targetRuleIds() {
        return this.getListAttribute('target_rule_ids');
    }
    set targetRuleIds(value) {
        this._targetRuleIds = value;
    }
    resetTargetRuleIds() {
        this._targetRuleIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetRuleIdsInput() {
        return this._targetRuleIds;
    }
    get targetRuleSet() {
        return this.getStringAttribute('target_rule_set');
    }
    set targetRuleSet(value) {
        this._targetRuleSet = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetRuleSetInput() {
        return this._targetRuleSet;
    }
    get requestCookie() {
        return this._requestCookie;
    }
    putRequestCookie(value) {
        this._requestCookie.internalValue = value;
    }
    resetRequestCookie() {
        this._requestCookie.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestCookieInput() {
        return this._requestCookie.internalValue;
    }
    get requestHeader() {
        return this._requestHeader;
    }
    putRequestHeader(value) {
        this._requestHeader.internalValue = value;
    }
    resetRequestHeader() {
        this._requestHeader.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestHeaderInput() {
        return this._requestHeader.internalValue;
    }
    get requestQueryParam() {
        return this._requestQueryParam;
    }
    putRequestQueryParam(value) {
        this._requestQueryParam.internalValue = value;
    }
    resetRequestQueryParam() {
        this._requestQueryParam.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestQueryParamInput() {
        return this._requestQueryParam.internalValue;
    }
    get requestUri() {
        return this._requestUri;
    }
    putRequestUri(value) {
        this._requestUri.internalValue = value;
    }
    resetRequestUri() {
        this._requestUri.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestUriInput() {
        return this._requestUri.internalValue;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAOutputReference;
_s = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAOutputReference[_s] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAList;
_t = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAList[_t] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exclusion: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToTerraform, true)(struct.exclusion),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigAToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigAToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        exclusion: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionAToHclTerraform, true)(struct.exclusion),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigAToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigAToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // exclusion - computed: false, optional: true, required: false
        this._exclusion = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionAList(this, "exclusion", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exclusion?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.exclusion = this._exclusion?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exclusion.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exclusion.internalValue = value.exclusion;
        }
    }
    get exclusion() {
        return this._exclusion;
    }
    putExclusion(value) {
        this._exclusion.internalValue = value;
    }
    resetExclusion() {
        this._exclusion.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exclusionInput() {
        return this._exclusion.internalValue;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigAOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigAOutputReference;
_u = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigAOutputReference[_u] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRulePreconfiguredWafConfigAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        count: cdktf.numberToTerraform(struct.count),
        interval_sec: cdktf.numberToTerraform(struct.intervalSec),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToTerraform = computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        count: {
            value: cdktf.numberToHclTerraform(struct.count),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        interval_sec: {
            value: cdktf.numberToHclTerraform(struct.intervalSec),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._count !== undefined) {
            hasAnyValues = true;
            internalValueResult.count = this._count;
        }
        if (this._intervalSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.intervalSec = this._intervalSec;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._count = undefined;
            this._intervalSec = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._count = value.count;
            this._intervalSec = value.intervalSec;
        }
    }
    get count() {
        return this.getNumberAttribute('count');
    }
    set count(value) {
        this._count = value;
    }
    resetCount() {
        this._count = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countInput() {
        return this._count;
    }
    get intervalSec() {
        return this.getNumberAttribute('interval_sec');
    }
    set intervalSec(value) {
        this._intervalSec = value;
    }
    resetIntervalSec() {
        this._intervalSec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get intervalSecInput() {
        return this._intervalSec;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdAOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdAOutputReference;
_v = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdAOutputReference[_v] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enforce_on_key_name: cdktf.stringToTerraform(struct.enforceOnKeyName),
        enforce_on_key_type: cdktf.stringToTerraform(struct.enforceOnKeyType),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToTerraform = computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enforce_on_key_name: {
            value: cdktf.stringToHclTerraform(struct.enforceOnKeyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enforce_on_key_type: {
            value: cdktf.stringToHclTerraform(struct.enforceOnKeyType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enforceOnKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKeyName = this._enforceOnKeyName;
        }
        if (this._enforceOnKeyType !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKeyType = this._enforceOnKeyType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._enforceOnKeyName = undefined;
            this._enforceOnKeyType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._enforceOnKeyName = value.enforceOnKeyName;
            this._enforceOnKeyType = value.enforceOnKeyType;
        }
    }
    get enforceOnKeyName() {
        return this.getStringAttribute('enforce_on_key_name');
    }
    set enforceOnKeyName(value) {
        this._enforceOnKeyName = value;
    }
    resetEnforceOnKeyName() {
        this._enforceOnKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyNameInput() {
        return this._enforceOnKeyName;
    }
    get enforceOnKeyType() {
        return this.getStringAttribute('enforce_on_key_type');
    }
    set enforceOnKeyType(value) {
        this._enforceOnKeyType = value;
    }
    resetEnforceOnKeyType() {
        this._enforceOnKeyType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyTypeInput() {
        return this._enforceOnKeyType;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAOutputReference;
_w = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAOutputReference[_w] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAList = ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAList;
_x = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAList[_x] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAList", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target: cdktf.stringToTerraform(struct.target),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToTerraform = computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target: {
            value: cdktf.stringToHclTerraform(struct.target),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._target = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._target = value.target;
            this._type = value.type;
        }
    }
    get target() {
        return this.getStringAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAOutputReference;
_y = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAOutputReference[_y] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        count: cdktf.numberToTerraform(struct.count),
        interval_sec: cdktf.numberToTerraform(struct.intervalSec),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToTerraform = computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        count: {
            value: cdktf.numberToHclTerraform(struct.count),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        interval_sec: {
            value: cdktf.numberToHclTerraform(struct.intervalSec),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._count !== undefined) {
            hasAnyValues = true;
            internalValueResult.count = this._count;
        }
        if (this._intervalSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.intervalSec = this._intervalSec;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._count = undefined;
            this._intervalSec = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._count = value.count;
            this._intervalSec = value.intervalSec;
        }
    }
    get count() {
        return this.getNumberAttribute('count');
    }
    set count(value) {
        this._count = value;
    }
    resetCount() {
        this._count = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get countInput() {
        return this._count;
    }
    get intervalSec() {
        return this.getNumberAttribute('interval_sec');
    }
    set intervalSec(value) {
        this._intervalSec = value;
    }
    resetIntervalSec() {
        this._intervalSec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get intervalSecInput() {
        return this._intervalSec;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAOutputReference;
_z = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAOutputReference[_z] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ban_duration_sec: cdktf.numberToTerraform(struct.banDurationSec),
        conform_action: cdktf.stringToTerraform(struct.conformAction),
        enforce_on_key: cdktf.stringToTerraform(struct.enforceOnKey),
        enforce_on_key_name: cdktf.stringToTerraform(struct.enforceOnKeyName),
        exceed_action: cdktf.stringToTerraform(struct.exceedAction),
        ban_threshold: computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToTerraform(struct.banThreshold),
        enforce_on_key_configs: cdktf.listMapper(computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToTerraform, true)(struct.enforceOnKeyConfigs),
        exceed_redirect_options: computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToTerraform(struct.exceedRedirectOptions),
        rate_limit_threshold: computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToTerraform(struct.rateLimitThreshold),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsAToTerraform = computeSecurityPolicyRuleRateLimitOptionsAToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ban_duration_sec: {
            value: cdktf.numberToHclTerraform(struct.banDurationSec),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        conform_action: {
            value: cdktf.stringToHclTerraform(struct.conformAction),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enforce_on_key: {
            value: cdktf.stringToHclTerraform(struct.enforceOnKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enforce_on_key_name: {
            value: cdktf.stringToHclTerraform(struct.enforceOnKeyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exceed_action: {
            value: cdktf.stringToHclTerraform(struct.exceedAction),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ban_threshold: {
            value: computeSecurityPolicyRuleRateLimitOptionsBanThresholdAToHclTerraform(struct.banThreshold),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdAList",
        },
        enforce_on_key_configs: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAToHclTerraform, true)(struct.enforceOnKeyConfigs),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAList",
        },
        exceed_redirect_options: {
            value: computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAToHclTerraform(struct.exceedRedirectOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAList",
        },
        rate_limit_threshold: {
            value: computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAToHclTerraform(struct.rateLimitThreshold),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsAToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsAToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // ban_threshold - computed: false, optional: true, required: false
        this._banThreshold = new ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdAOutputReference(this, "ban_threshold");
        // enforce_on_key_configs - computed: false, optional: true, required: false
        this._enforceOnKeyConfigs = new ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsAList(this, "enforce_on_key_configs", false);
        // exceed_redirect_options - computed: false, optional: true, required: false
        this._exceedRedirectOptions = new ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsAOutputReference(this, "exceed_redirect_options");
        // rate_limit_threshold - computed: false, optional: true, required: false
        this._rateLimitThreshold = new ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdAOutputReference(this, "rate_limit_threshold");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._banDurationSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.banDurationSec = this._banDurationSec;
        }
        if (this._conformAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.conformAction = this._conformAction;
        }
        if (this._enforceOnKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKey = this._enforceOnKey;
        }
        if (this._enforceOnKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKeyName = this._enforceOnKeyName;
        }
        if (this._exceedAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.exceedAction = this._exceedAction;
        }
        if (this._banThreshold?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.banThreshold = this._banThreshold?.internalValue;
        }
        if (this._enforceOnKeyConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKeyConfigs = this._enforceOnKeyConfigs?.internalValue;
        }
        if (this._exceedRedirectOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.exceedRedirectOptions = this._exceedRedirectOptions?.internalValue;
        }
        if (this._rateLimitThreshold?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rateLimitThreshold = this._rateLimitThreshold?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._banDurationSec = undefined;
            this._conformAction = undefined;
            this._enforceOnKey = undefined;
            this._enforceOnKeyName = undefined;
            this._exceedAction = undefined;
            this._banThreshold.internalValue = undefined;
            this._enforceOnKeyConfigs.internalValue = undefined;
            this._exceedRedirectOptions.internalValue = undefined;
            this._rateLimitThreshold.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._banDurationSec = value.banDurationSec;
            this._conformAction = value.conformAction;
            this._enforceOnKey = value.enforceOnKey;
            this._enforceOnKeyName = value.enforceOnKeyName;
            this._exceedAction = value.exceedAction;
            this._banThreshold.internalValue = value.banThreshold;
            this._enforceOnKeyConfigs.internalValue = value.enforceOnKeyConfigs;
            this._exceedRedirectOptions.internalValue = value.exceedRedirectOptions;
            this._rateLimitThreshold.internalValue = value.rateLimitThreshold;
        }
    }
    get banDurationSec() {
        return this.getNumberAttribute('ban_duration_sec');
    }
    set banDurationSec(value) {
        this._banDurationSec = value;
    }
    resetBanDurationSec() {
        this._banDurationSec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get banDurationSecInput() {
        return this._banDurationSec;
    }
    get conformAction() {
        return this.getStringAttribute('conform_action');
    }
    set conformAction(value) {
        this._conformAction = value;
    }
    resetConformAction() {
        this._conformAction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get conformActionInput() {
        return this._conformAction;
    }
    get enforceOnKey() {
        return this.getStringAttribute('enforce_on_key');
    }
    set enforceOnKey(value) {
        this._enforceOnKey = value;
    }
    resetEnforceOnKey() {
        this._enforceOnKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyInput() {
        return this._enforceOnKey;
    }
    get enforceOnKeyName() {
        return this.getStringAttribute('enforce_on_key_name');
    }
    set enforceOnKeyName(value) {
        this._enforceOnKeyName = value;
    }
    resetEnforceOnKeyName() {
        this._enforceOnKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyNameInput() {
        return this._enforceOnKeyName;
    }
    get exceedAction() {
        return this.getStringAttribute('exceed_action');
    }
    set exceedAction(value) {
        this._exceedAction = value;
    }
    resetExceedAction() {
        this._exceedAction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exceedActionInput() {
        return this._exceedAction;
    }
    get banThreshold() {
        return this._banThreshold;
    }
    putBanThreshold(value) {
        this._banThreshold.internalValue = value;
    }
    resetBanThreshold() {
        this._banThreshold.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get banThresholdInput() {
        return this._banThreshold.internalValue;
    }
    get enforceOnKeyConfigs() {
        return this._enforceOnKeyConfigs;
    }
    putEnforceOnKeyConfigs(value) {
        this._enforceOnKeyConfigs.internalValue = value;
    }
    resetEnforceOnKeyConfigs() {
        this._enforceOnKeyConfigs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyConfigsInput() {
        return this._enforceOnKeyConfigs.internalValue;
    }
    get exceedRedirectOptions() {
        return this._exceedRedirectOptions;
    }
    putExceedRedirectOptions(value) {
        this._exceedRedirectOptions.internalValue = value;
    }
    resetExceedRedirectOptions() {
        this._exceedRedirectOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exceedRedirectOptionsInput() {
        return this._exceedRedirectOptions.internalValue;
    }
    get rateLimitThreshold() {
        return this._rateLimitThreshold;
    }
    putRateLimitThreshold(value) {
        this._rateLimitThreshold.internalValue = value;
    }
    resetRateLimitThreshold() {
        this._rateLimitThreshold.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rateLimitThresholdInput() {
        return this._rateLimitThreshold.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsAOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsAOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsAOutputReference[_0] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleRateLimitOptionsAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRedirectOptionsAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target: cdktf.stringToTerraform(struct.target),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.computeSecurityPolicyRuleRedirectOptionsAToTerraform = computeSecurityPolicyRuleRedirectOptionsAToTerraform;
function computeSecurityPolicyRuleRedirectOptionsAToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target: {
            value: cdktf.stringToHclTerraform(struct.target),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRedirectOptionsAToHclTerraform = computeSecurityPolicyRuleRedirectOptionsAToHclTerraform;
class ComputeSecurityPolicyRuleRedirectOptionsAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._target = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._target = value.target;
            this._type = value.type;
        }
    }
    get target() {
        return this.getStringAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComputeSecurityPolicyRuleRedirectOptionsAOutputReference = ComputeSecurityPolicyRuleRedirectOptionsAOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRedirectOptionsAOutputReference[_1] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleRedirectOptionsAOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.computeSecurityPolicyRuleTimeoutsToTerraform = computeSecurityPolicyRuleTimeoutsToTerraform;
function computeSecurityPolicyRuleTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleTimeoutsToHclTerraform = computeSecurityPolicyRuleTimeoutsToHclTerraform;
class ComputeSecurityPolicyRuleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ComputeSecurityPolicyRuleTimeoutsOutputReference = ComputeSecurityPolicyRuleTimeoutsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleTimeoutsOutputReference[_2] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_security_policy_rule google_compute_security_policy_rule}
*/
class ComputeSecurityPolicyRuleA extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ComputeSecurityPolicyRuleA resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ComputeSecurityPolicyRuleA to import
    * @param importFromId The id of the existing ComputeSecurityPolicyRuleA that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_security_policy_rule#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ComputeSecurityPolicyRuleA to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_security_policy_rule", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_security_policy_rule google_compute_security_policy_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeSecurityPolicyRuleAConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_security_policy_rule',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // header_action - computed: false, optional: true, required: false
        this._headerAction = new ComputeSecurityPolicyRuleHeaderActionAOutputReference(this, "header_action");
        // match - computed: false, optional: true, required: false
        this._match = new ComputeSecurityPolicyRuleMatchAOutputReference(this, "match");
        // preconfigured_waf_config - computed: false, optional: true, required: false
        this._preconfiguredWafConfig = new ComputeSecurityPolicyRulePreconfiguredWafConfigAOutputReference(this, "preconfigured_waf_config");
        // rate_limit_options - computed: false, optional: true, required: false
        this._rateLimitOptions = new ComputeSecurityPolicyRuleRateLimitOptionsAOutputReference(this, "rate_limit_options");
        // redirect_options - computed: false, optional: true, required: false
        this._redirectOptions = new ComputeSecurityPolicyRuleRedirectOptionsAOutputReference(this, "redirect_options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeSecurityPolicyRuleTimeoutsOutputReference(this, "timeouts");
        this._action = config.action;
        this._description = config.description;
        this._id = config.id;
        this._preview = config.preview;
        this._priority = config.priority;
        this._project = config.project;
        this._securityPolicy = config.securityPolicy;
        this._headerAction.internalValue = config.headerAction;
        this._match.internalValue = config.match;
        this._preconfiguredWafConfig.internalValue = config.preconfiguredWafConfig;
        this._rateLimitOptions.internalValue = config.rateLimitOptions;
        this._redirectOptions.internalValue = config.redirectOptions;
        this._timeouts.internalValue = config.timeouts;
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get preview() {
        return this.getBooleanAttribute('preview');
    }
    set preview(value) {
        this._preview = value;
    }
    resetPreview() {
        this._preview = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get previewInput() {
        return this._preview;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get securityPolicy() {
        return this.getStringAttribute('security_policy');
    }
    set securityPolicy(value) {
        this._securityPolicy = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityPolicyInput() {
        return this._securityPolicy;
    }
    get headerAction() {
        return this._headerAction;
    }
    putHeaderAction(value) {
        this._headerAction.internalValue = value;
    }
    resetHeaderAction() {
        this._headerAction.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerActionInput() {
        return this._headerAction.internalValue;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    resetMatch() {
        this._match.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
    get preconfiguredWafConfig() {
        return this._preconfiguredWafConfig;
    }
    putPreconfiguredWafConfig(value) {
        this._preconfiguredWafConfig.internalValue = value;
    }
    resetPreconfiguredWafConfig() {
        this._preconfiguredWafConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preconfiguredWafConfigInput() {
        return this._preconfiguredWafConfig.internalValue;
    }
    get rateLimitOptions() {
        return this._rateLimitOptions;
    }
    putRateLimitOptions(value) {
        this._rateLimitOptions.internalValue = value;
    }
    resetRateLimitOptions() {
        this._rateLimitOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rateLimitOptionsInput() {
        return this._rateLimitOptions.internalValue;
    }
    get redirectOptions() {
        return this._redirectOptions;
    }
    putRedirectOptions(value) {
        this._redirectOptions.internalValue = value;
    }
    resetRedirectOptions() {
        this._redirectOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get redirectOptionsInput() {
        return this._redirectOptions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            action: cdktf.stringToTerraform(this._action),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            preview: cdktf.booleanToTerraform(this._preview),
            priority: cdktf.numberToTerraform(this._priority),
            project: cdktf.stringToTerraform(this._project),
            security_policy: cdktf.stringToTerraform(this._securityPolicy),
            header_action: computeSecurityPolicyRuleHeaderActionAToTerraform(this._headerAction.internalValue),
            match: computeSecurityPolicyRuleMatchAToTerraform(this._match.internalValue),
            preconfigured_waf_config: computeSecurityPolicyRulePreconfiguredWafConfigAToTerraform(this._preconfiguredWafConfig.internalValue),
            rate_limit_options: computeSecurityPolicyRuleRateLimitOptionsAToTerraform(this._rateLimitOptions.internalValue),
            redirect_options: computeSecurityPolicyRuleRedirectOptionsAToTerraform(this._redirectOptions.internalValue),
            timeouts: computeSecurityPolicyRuleTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            action: {
                value: cdktf.stringToHclTerraform(this._action),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            preview: {
                value: cdktf.booleanToHclTerraform(this._preview),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            priority: {
                value: cdktf.numberToHclTerraform(this._priority),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            security_policy: {
                value: cdktf.stringToHclTerraform(this._securityPolicy),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            header_action: {
                value: computeSecurityPolicyRuleHeaderActionAToHclTerraform(this._headerAction.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeSecurityPolicyRuleHeaderActionAList",
            },
            match: {
                value: computeSecurityPolicyRuleMatchAToHclTerraform(this._match.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeSecurityPolicyRuleMatchAList",
            },
            preconfigured_waf_config: {
                value: computeSecurityPolicyRulePreconfiguredWafConfigAToHclTerraform(this._preconfiguredWafConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigAList",
            },
            rate_limit_options: {
                value: computeSecurityPolicyRuleRateLimitOptionsAToHclTerraform(this._rateLimitOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsAList",
            },
            redirect_options: {
                value: computeSecurityPolicyRuleRedirectOptionsAToHclTerraform(this._redirectOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeSecurityPolicyRuleRedirectOptionsAList",
            },
            timeouts: {
                value: computeSecurityPolicyRuleTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ComputeSecurityPolicyRuleTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ComputeSecurityPolicyRuleA = ComputeSecurityPolicyRuleA;
_3 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleA[_3] = { fqn: "@cdktf/provider-google.computeSecurityPolicyRule.ComputeSecurityPolicyRuleA", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ComputeSecurityPolicyRuleA.tfResourceType = "google_compute_security_policy_rule";
