// @cdktf/provider-google/lib/data-google-compute-backend-bucket/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeBackendBucket = exports.DataGoogleComputeBackendBucketCdnPolicyList = exports.DataGoogleComputeBackendBucketCdnPolicyOutputReference = exports.dataGoogleComputeBackendBucketCdnPolicyToHclTerraform = exports.dataGoogleComputeBackendBucketCdnPolicyToTerraform = exports.DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyList = exports.DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyOutputReference = exports.dataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyToHclTerraform = exports.dataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyToTerraform = exports.DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyList = exports.DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyOutputReference = exports.dataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyToHclTerraform = exports.dataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyToTerraform = exports.DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersList = exports.DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersOutputReference = exports.dataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersToHclTerraform = exports.dataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersToTerraform = dataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersToTerraform;
function dataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersToHclTerraform = dataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersToHclTerraform;
class DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // header_name - computed: true, optional: false, required: false
    get headerName() {
        return this.getStringAttribute('header_name');
    }
}
exports.DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersOutputReference = DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersOutputReference", version: "14.13.1" };
class DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersList = DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersList", version: "14.13.1" };
function dataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyToTerraform = dataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyToTerraform;
function dataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyToHclTerraform = dataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyToHclTerraform;
class DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // include_http_headers - computed: true, optional: false, required: false
    get includeHttpHeaders() {
        return this.getListAttribute('include_http_headers');
    }
    // query_string_whitelist - computed: true, optional: false, required: false
    get queryStringWhitelist() {
        return this.getListAttribute('query_string_whitelist');
    }
}
exports.DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyOutputReference = DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyList = DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyList", version: "14.13.1" };
function dataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyToTerraform = dataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyToTerraform;
function dataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyToHclTerraform = dataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyToHclTerraform;
class DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // code - computed: true, optional: false, required: false
    get code() {
        return this.getNumberAttribute('code');
    }
    // ttl - computed: true, optional: false, required: false
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
}
exports.DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyOutputReference = DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyList = DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyList", version: "14.13.1" };
function dataGoogleComputeBackendBucketCdnPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeBackendBucketCdnPolicyToTerraform = dataGoogleComputeBackendBucketCdnPolicyToTerraform;
function dataGoogleComputeBackendBucketCdnPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeBackendBucketCdnPolicyToHclTerraform = dataGoogleComputeBackendBucketCdnPolicyToHclTerraform;
class DataGoogleComputeBackendBucketCdnPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // bypass_cache_on_request_headers - computed: true, optional: false, required: false
        this._bypassCacheOnRequestHeaders = new DataGoogleComputeBackendBucketCdnPolicyBypassCacheOnRequestHeadersList(this, "bypass_cache_on_request_headers", false);
        // cache_key_policy - computed: true, optional: false, required: false
        this._cacheKeyPolicy = new DataGoogleComputeBackendBucketCdnPolicyCacheKeyPolicyList(this, "cache_key_policy", false);
        // negative_caching_policy - computed: true, optional: false, required: false
        this._negativeCachingPolicy = new DataGoogleComputeBackendBucketCdnPolicyNegativeCachingPolicyList(this, "negative_caching_policy", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get bypassCacheOnRequestHeaders() {
        return this._bypassCacheOnRequestHeaders;
    }
    get cacheKeyPolicy() {
        return this._cacheKeyPolicy;
    }
    // cache_mode - computed: true, optional: false, required: false
    get cacheMode() {
        return this.getStringAttribute('cache_mode');
    }
    // client_ttl - computed: true, optional: false, required: false
    get clientTtl() {
        return this.getNumberAttribute('client_ttl');
    }
    // default_ttl - computed: true, optional: false, required: false
    get defaultTtl() {
        return this.getNumberAttribute('default_ttl');
    }
    // max_ttl - computed: true, optional: false, required: false
    get maxTtl() {
        return this.getNumberAttribute('max_ttl');
    }
    // negative_caching - computed: true, optional: false, required: false
    get negativeCaching() {
        return this.getBooleanAttribute('negative_caching');
    }
    get negativeCachingPolicy() {
        return this._negativeCachingPolicy;
    }
    // request_coalescing - computed: true, optional: false, required: false
    get requestCoalescing() {
        return this.getBooleanAttribute('request_coalescing');
    }
    // serve_while_stale - computed: true, optional: false, required: false
    get serveWhileStale() {
        return this.getNumberAttribute('serve_while_stale');
    }
    // signed_url_cache_max_age_sec - computed: true, optional: false, required: false
    get signedUrlCacheMaxAgeSec() {
        return this.getNumberAttribute('signed_url_cache_max_age_sec');
    }
}
exports.DataGoogleComputeBackendBucketCdnPolicyOutputReference = DataGoogleComputeBackendBucketCdnPolicyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucketCdnPolicyOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucketCdnPolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeBackendBucketCdnPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeBackendBucketCdnPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeBackendBucketCdnPolicyList = DataGoogleComputeBackendBucketCdnPolicyList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucketCdnPolicyList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucketCdnPolicyList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_backend_bucket google_compute_backend_bucket}
*/
class DataGoogleComputeBackendBucket extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeBackendBucket resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeBackendBucket to import
    * @param importFromId The id of the existing DataGoogleComputeBackendBucket that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_backend_bucket#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeBackendBucket to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_backend_bucket", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_backend_bucket google_compute_backend_bucket} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeBackendBucketConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_backend_bucket',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cdn_policy - computed: true, optional: false, required: false
        this._cdnPolicy = new DataGoogleComputeBackendBucketCdnPolicyList(this, "cdn_policy", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // bucket_name - computed: true, optional: false, required: false
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    get cdnPolicy() {
        return this._cdnPolicy;
    }
    // compression_mode - computed: true, optional: false, required: false
    get compressionMode() {
        return this.getStringAttribute('compression_mode');
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    // custom_response_headers - computed: true, optional: false, required: false
    get customResponseHeaders() {
        return this.getListAttribute('custom_response_headers');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // edge_security_policy - computed: true, optional: false, required: false
    get edgeSecurityPolicy() {
        return this.getStringAttribute('edge_security_policy');
    }
    // enable_cdn - computed: true, optional: false, required: false
    get enableCdn() {
        return this.getBooleanAttribute('enable_cdn');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeBackendBucket = DataGoogleComputeBackendBucket;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeBackendBucket[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeBackendBucket.DataGoogleComputeBackendBucket", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeBackendBucket.tfResourceType = "google_compute_backend_bucket";
