// @cdktf/provider-google/lib/dataplex-datascan/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataplexDatascanDataQualitySpecRulesTableConditionExpectationToTerraform = exports.DataplexDatascanDataQualitySpecRulesStatisticRangeExpectationOutputReference = exports.dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToTerraform = exports.DataplexDatascanDataQualitySpecRulesSqlAssertionOutputReference = exports.dataplexDatascanDataQualitySpecRulesSqlAssertionToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesSqlAssertionToTerraform = exports.DataplexDatascanDataQualitySpecRulesSetExpectationOutputReference = exports.dataplexDatascanDataQualitySpecRulesSetExpectationToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesSetExpectationToTerraform = exports.DataplexDatascanDataQualitySpecRulesRowConditionExpectationOutputReference = exports.dataplexDatascanDataQualitySpecRulesRowConditionExpectationToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesRowConditionExpectationToTerraform = exports.DataplexDatascanDataQualitySpecRulesRegexExpectationOutputReference = exports.dataplexDatascanDataQualitySpecRulesRegexExpectationToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesRegexExpectationToTerraform = exports.DataplexDatascanDataQualitySpecRulesRangeExpectationOutputReference = exports.dataplexDatascanDataQualitySpecRulesRangeExpectationToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesRangeExpectationToTerraform = exports.DataplexDatascanDataQualitySpecRulesNonNullExpectationOutputReference = exports.dataplexDatascanDataQualitySpecRulesNonNullExpectationToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesNonNullExpectationToTerraform = exports.DataplexDatascanDataQualitySpecPostScanActionsOutputReference = exports.dataplexDatascanDataQualitySpecPostScanActionsToHclTerraform = exports.dataplexDatascanDataQualitySpecPostScanActionsToTerraform = exports.DataplexDatascanDataQualitySpecPostScanActionsBigqueryExportOutputReference = exports.dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToHclTerraform = exports.dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToTerraform = exports.DataplexDatascanDataProfileSpecOutputReference = exports.dataplexDatascanDataProfileSpecToHclTerraform = exports.dataplexDatascanDataProfileSpecToTerraform = exports.DataplexDatascanDataProfileSpecPostScanActionsOutputReference = exports.dataplexDatascanDataProfileSpecPostScanActionsToHclTerraform = exports.dataplexDatascanDataProfileSpecPostScanActionsToTerraform = exports.DataplexDatascanDataProfileSpecPostScanActionsBigqueryExportOutputReference = exports.dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToHclTerraform = exports.dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToTerraform = exports.DataplexDatascanDataProfileSpecIncludeFieldsOutputReference = exports.dataplexDatascanDataProfileSpecIncludeFieldsToHclTerraform = exports.dataplexDatascanDataProfileSpecIncludeFieldsToTerraform = exports.DataplexDatascanDataProfileSpecExcludeFieldsOutputReference = exports.dataplexDatascanDataProfileSpecExcludeFieldsToHclTerraform = exports.dataplexDatascanDataProfileSpecExcludeFieldsToTerraform = exports.DataplexDatascanDataOutputReference = exports.dataplexDatascanDataToHclTerraform = exports.dataplexDatascanDataToTerraform = exports.DataplexDatascanExecutionStatusList = exports.DataplexDatascanExecutionStatusOutputReference = exports.dataplexDatascanExecutionStatusToHclTerraform = exports.dataplexDatascanExecutionStatusToTerraform = void 0;
exports.DataplexDatascan = exports.DataplexDatascanTimeoutsOutputReference = exports.dataplexDatascanTimeoutsToHclTerraform = exports.dataplexDatascanTimeoutsToTerraform = exports.DataplexDatascanExecutionSpecOutputReference = exports.dataplexDatascanExecutionSpecToHclTerraform = exports.dataplexDatascanExecutionSpecToTerraform = exports.DataplexDatascanExecutionSpecTriggerOutputReference = exports.dataplexDatascanExecutionSpecTriggerToHclTerraform = exports.dataplexDatascanExecutionSpecTriggerToTerraform = exports.DataplexDatascanExecutionSpecTriggerScheduleOutputReference = exports.dataplexDatascanExecutionSpecTriggerScheduleToHclTerraform = exports.dataplexDatascanExecutionSpecTriggerScheduleToTerraform = exports.DataplexDatascanExecutionSpecTriggerOnDemandOutputReference = exports.dataplexDatascanExecutionSpecTriggerOnDemandToHclTerraform = exports.dataplexDatascanExecutionSpecTriggerOnDemandToTerraform = exports.DataplexDatascanDataQualitySpecOutputReference = exports.dataplexDatascanDataQualitySpecToHclTerraform = exports.dataplexDatascanDataQualitySpecToTerraform = exports.DataplexDatascanDataQualitySpecRulesList = exports.DataplexDatascanDataQualitySpecRulesOutputReference = exports.dataplexDatascanDataQualitySpecRulesToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesToTerraform = exports.DataplexDatascanDataQualitySpecRulesUniquenessExpectationOutputReference = exports.dataplexDatascanDataQualitySpecRulesUniquenessExpectationToHclTerraform = exports.dataplexDatascanDataQualitySpecRulesUniquenessExpectationToTerraform = exports.DataplexDatascanDataQualitySpecRulesTableConditionExpectationOutputReference = exports.dataplexDatascanDataQualitySpecRulesTableConditionExpectationToHclTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataplexDatascanExecutionStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataplexDatascanExecutionStatusToTerraform = dataplexDatascanExecutionStatusToTerraform;
function dataplexDatascanExecutionStatusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataplexDatascanExecutionStatusToHclTerraform = dataplexDatascanExecutionStatusToHclTerraform;
class DataplexDatascanExecutionStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // latest_job_end_time - computed: true, optional: false, required: false
    get latestJobEndTime() {
        return this.getStringAttribute('latest_job_end_time');
    }
    // latest_job_start_time - computed: true, optional: false, required: false
    get latestJobStartTime() {
        return this.getStringAttribute('latest_job_start_time');
    }
}
exports.DataplexDatascanExecutionStatusOutputReference = DataplexDatascanExecutionStatusOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataplexDatascanExecutionStatusOutputReference[_a] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanExecutionStatusOutputReference", version: "14.13.1" };
class DataplexDatascanExecutionStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataplexDatascanExecutionStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataplexDatascanExecutionStatusList = DataplexDatascanExecutionStatusList;
_b = JSII_RTTI_SYMBOL_1;
DataplexDatascanExecutionStatusList[_b] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanExecutionStatusList", version: "14.13.1" };
function dataplexDatascanDataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        entity: cdktf.stringToTerraform(struct.entity),
        resource: cdktf.stringToTerraform(struct.resource),
    };
}
exports.dataplexDatascanDataToTerraform = dataplexDatascanDataToTerraform;
function dataplexDatascanDataToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        entity: {
            value: cdktf.stringToHclTerraform(struct.entity),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        resource: {
            value: cdktf.stringToHclTerraform(struct.resource),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataToHclTerraform = dataplexDatascanDataToHclTerraform;
class DataplexDatascanDataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._entity !== undefined) {
            hasAnyValues = true;
            internalValueResult.entity = this._entity;
        }
        if (this._resource !== undefined) {
            hasAnyValues = true;
            internalValueResult.resource = this._resource;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._entity = undefined;
            this._resource = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._entity = value.entity;
            this._resource = value.resource;
        }
    }
    get entity() {
        return this.getStringAttribute('entity');
    }
    set entity(value) {
        this._entity = value;
    }
    resetEntity() {
        this._entity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get entityInput() {
        return this._entity;
    }
    get resource() {
        return this.getStringAttribute('resource');
    }
    set resource(value) {
        this._resource = value;
    }
    resetResource() {
        this._resource = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourceInput() {
        return this._resource;
    }
}
exports.DataplexDatascanDataOutputReference = DataplexDatascanDataOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataOutputReference[_c] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataOutputReference", version: "14.13.1" };
function dataplexDatascanDataProfileSpecExcludeFieldsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field_names: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.fieldNames),
    };
}
exports.dataplexDatascanDataProfileSpecExcludeFieldsToTerraform = dataplexDatascanDataProfileSpecExcludeFieldsToTerraform;
function dataplexDatascanDataProfileSpecExcludeFieldsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        field_names: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.fieldNames),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataProfileSpecExcludeFieldsToHclTerraform = dataplexDatascanDataProfileSpecExcludeFieldsToHclTerraform;
class DataplexDatascanDataProfileSpecExcludeFieldsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fieldNames !== undefined) {
            hasAnyValues = true;
            internalValueResult.fieldNames = this._fieldNames;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fieldNames = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fieldNames = value.fieldNames;
        }
    }
    get fieldNames() {
        return this.getListAttribute('field_names');
    }
    set fieldNames(value) {
        this._fieldNames = value;
    }
    resetFieldNames() {
        this._fieldNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldNamesInput() {
        return this._fieldNames;
    }
}
exports.DataplexDatascanDataProfileSpecExcludeFieldsOutputReference = DataplexDatascanDataProfileSpecExcludeFieldsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataProfileSpecExcludeFieldsOutputReference[_d] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataProfileSpecExcludeFieldsOutputReference", version: "14.13.1" };
function dataplexDatascanDataProfileSpecIncludeFieldsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field_names: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.fieldNames),
    };
}
exports.dataplexDatascanDataProfileSpecIncludeFieldsToTerraform = dataplexDatascanDataProfileSpecIncludeFieldsToTerraform;
function dataplexDatascanDataProfileSpecIncludeFieldsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        field_names: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.fieldNames),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataProfileSpecIncludeFieldsToHclTerraform = dataplexDatascanDataProfileSpecIncludeFieldsToHclTerraform;
class DataplexDatascanDataProfileSpecIncludeFieldsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fieldNames !== undefined) {
            hasAnyValues = true;
            internalValueResult.fieldNames = this._fieldNames;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fieldNames = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fieldNames = value.fieldNames;
        }
    }
    get fieldNames() {
        return this.getListAttribute('field_names');
    }
    set fieldNames(value) {
        this._fieldNames = value;
    }
    resetFieldNames() {
        this._fieldNames = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldNamesInput() {
        return this._fieldNames;
    }
}
exports.DataplexDatascanDataProfileSpecIncludeFieldsOutputReference = DataplexDatascanDataProfileSpecIncludeFieldsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataProfileSpecIncludeFieldsOutputReference[_e] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataProfileSpecIncludeFieldsOutputReference", version: "14.13.1" };
function dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        results_table: cdktf.stringToTerraform(struct.resultsTable),
    };
}
exports.dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToTerraform = dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToTerraform;
function dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        results_table: {
            value: cdktf.stringToHclTerraform(struct.resultsTable),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToHclTerraform = dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToHclTerraform;
class DataplexDatascanDataProfileSpecPostScanActionsBigqueryExportOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resultsTable !== undefined) {
            hasAnyValues = true;
            internalValueResult.resultsTable = this._resultsTable;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resultsTable = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resultsTable = value.resultsTable;
        }
    }
    get resultsTable() {
        return this.getStringAttribute('results_table');
    }
    set resultsTable(value) {
        this._resultsTable = value;
    }
    resetResultsTable() {
        this._resultsTable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resultsTableInput() {
        return this._resultsTable;
    }
}
exports.DataplexDatascanDataProfileSpecPostScanActionsBigqueryExportOutputReference = DataplexDatascanDataProfileSpecPostScanActionsBigqueryExportOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataProfileSpecPostScanActionsBigqueryExportOutputReference[_f] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataProfileSpecPostScanActionsBigqueryExportOutputReference", version: "14.13.1" };
function dataplexDatascanDataProfileSpecPostScanActionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bigquery_export: dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToTerraform(struct.bigqueryExport),
    };
}
exports.dataplexDatascanDataProfileSpecPostScanActionsToTerraform = dataplexDatascanDataProfileSpecPostScanActionsToTerraform;
function dataplexDatascanDataProfileSpecPostScanActionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bigquery_export: {
            value: dataplexDatascanDataProfileSpecPostScanActionsBigqueryExportToHclTerraform(struct.bigqueryExport),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataProfileSpecPostScanActionsBigqueryExportList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataProfileSpecPostScanActionsToHclTerraform = dataplexDatascanDataProfileSpecPostScanActionsToHclTerraform;
class DataplexDatascanDataProfileSpecPostScanActionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // bigquery_export - computed: false, optional: true, required: false
        this._bigqueryExport = new DataplexDatascanDataProfileSpecPostScanActionsBigqueryExportOutputReference(this, "bigquery_export");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bigqueryExport?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.bigqueryExport = this._bigqueryExport?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bigqueryExport.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bigqueryExport.internalValue = value.bigqueryExport;
        }
    }
    get bigqueryExport() {
        return this._bigqueryExport;
    }
    putBigqueryExport(value) {
        this._bigqueryExport.internalValue = value;
    }
    resetBigqueryExport() {
        this._bigqueryExport.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigqueryExportInput() {
        return this._bigqueryExport.internalValue;
    }
}
exports.DataplexDatascanDataProfileSpecPostScanActionsOutputReference = DataplexDatascanDataProfileSpecPostScanActionsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataProfileSpecPostScanActionsOutputReference[_g] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataProfileSpecPostScanActionsOutputReference", version: "14.13.1" };
function dataplexDatascanDataProfileSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        row_filter: cdktf.stringToTerraform(struct.rowFilter),
        sampling_percent: cdktf.numberToTerraform(struct.samplingPercent),
        exclude_fields: dataplexDatascanDataProfileSpecExcludeFieldsToTerraform(struct.excludeFields),
        include_fields: dataplexDatascanDataProfileSpecIncludeFieldsToTerraform(struct.includeFields),
        post_scan_actions: dataplexDatascanDataProfileSpecPostScanActionsToTerraform(struct.postScanActions),
    };
}
exports.dataplexDatascanDataProfileSpecToTerraform = dataplexDatascanDataProfileSpecToTerraform;
function dataplexDatascanDataProfileSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        row_filter: {
            value: cdktf.stringToHclTerraform(struct.rowFilter),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        sampling_percent: {
            value: cdktf.numberToHclTerraform(struct.samplingPercent),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        exclude_fields: {
            value: dataplexDatascanDataProfileSpecExcludeFieldsToHclTerraform(struct.excludeFields),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataProfileSpecExcludeFieldsList",
        },
        include_fields: {
            value: dataplexDatascanDataProfileSpecIncludeFieldsToHclTerraform(struct.includeFields),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataProfileSpecIncludeFieldsList",
        },
        post_scan_actions: {
            value: dataplexDatascanDataProfileSpecPostScanActionsToHclTerraform(struct.postScanActions),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataProfileSpecPostScanActionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataProfileSpecToHclTerraform = dataplexDatascanDataProfileSpecToHclTerraform;
class DataplexDatascanDataProfileSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // exclude_fields - computed: false, optional: true, required: false
        this._excludeFields = new DataplexDatascanDataProfileSpecExcludeFieldsOutputReference(this, "exclude_fields");
        // include_fields - computed: false, optional: true, required: false
        this._includeFields = new DataplexDatascanDataProfileSpecIncludeFieldsOutputReference(this, "include_fields");
        // post_scan_actions - computed: false, optional: true, required: false
        this._postScanActions = new DataplexDatascanDataProfileSpecPostScanActionsOutputReference(this, "post_scan_actions");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rowFilter !== undefined) {
            hasAnyValues = true;
            internalValueResult.rowFilter = this._rowFilter;
        }
        if (this._samplingPercent !== undefined) {
            hasAnyValues = true;
            internalValueResult.samplingPercent = this._samplingPercent;
        }
        if (this._excludeFields?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludeFields = this._excludeFields?.internalValue;
        }
        if (this._includeFields?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeFields = this._includeFields?.internalValue;
        }
        if (this._postScanActions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.postScanActions = this._postScanActions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rowFilter = undefined;
            this._samplingPercent = undefined;
            this._excludeFields.internalValue = undefined;
            this._includeFields.internalValue = undefined;
            this._postScanActions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rowFilter = value.rowFilter;
            this._samplingPercent = value.samplingPercent;
            this._excludeFields.internalValue = value.excludeFields;
            this._includeFields.internalValue = value.includeFields;
            this._postScanActions.internalValue = value.postScanActions;
        }
    }
    get rowFilter() {
        return this.getStringAttribute('row_filter');
    }
    set rowFilter(value) {
        this._rowFilter = value;
    }
    resetRowFilter() {
        this._rowFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rowFilterInput() {
        return this._rowFilter;
    }
    get samplingPercent() {
        return this.getNumberAttribute('sampling_percent');
    }
    set samplingPercent(value) {
        this._samplingPercent = value;
    }
    resetSamplingPercent() {
        this._samplingPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get samplingPercentInput() {
        return this._samplingPercent;
    }
    get excludeFields() {
        return this._excludeFields;
    }
    putExcludeFields(value) {
        this._excludeFields.internalValue = value;
    }
    resetExcludeFields() {
        this._excludeFields.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludeFieldsInput() {
        return this._excludeFields.internalValue;
    }
    get includeFields() {
        return this._includeFields;
    }
    putIncludeFields(value) {
        this._includeFields.internalValue = value;
    }
    resetIncludeFields() {
        this._includeFields.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeFieldsInput() {
        return this._includeFields.internalValue;
    }
    get postScanActions() {
        return this._postScanActions;
    }
    putPostScanActions(value) {
        this._postScanActions.internalValue = value;
    }
    resetPostScanActions() {
        this._postScanActions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postScanActionsInput() {
        return this._postScanActions.internalValue;
    }
}
exports.DataplexDatascanDataProfileSpecOutputReference = DataplexDatascanDataProfileSpecOutputReference;
_h = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataProfileSpecOutputReference[_h] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataProfileSpecOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        results_table: cdktf.stringToTerraform(struct.resultsTable),
    };
}
exports.dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToTerraform = dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToTerraform;
function dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        results_table: {
            value: cdktf.stringToHclTerraform(struct.resultsTable),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToHclTerraform = dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToHclTerraform;
class DataplexDatascanDataQualitySpecPostScanActionsBigqueryExportOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resultsTable !== undefined) {
            hasAnyValues = true;
            internalValueResult.resultsTable = this._resultsTable;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resultsTable = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resultsTable = value.resultsTable;
        }
    }
    get resultsTable() {
        return this.getStringAttribute('results_table');
    }
    set resultsTable(value) {
        this._resultsTable = value;
    }
    resetResultsTable() {
        this._resultsTable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resultsTableInput() {
        return this._resultsTable;
    }
}
exports.DataplexDatascanDataQualitySpecPostScanActionsBigqueryExportOutputReference = DataplexDatascanDataQualitySpecPostScanActionsBigqueryExportOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecPostScanActionsBigqueryExportOutputReference[_j] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecPostScanActionsBigqueryExportOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecPostScanActionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bigquery_export: dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToTerraform(struct.bigqueryExport),
    };
}
exports.dataplexDatascanDataQualitySpecPostScanActionsToTerraform = dataplexDatascanDataQualitySpecPostScanActionsToTerraform;
function dataplexDatascanDataQualitySpecPostScanActionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bigquery_export: {
            value: dataplexDatascanDataQualitySpecPostScanActionsBigqueryExportToHclTerraform(struct.bigqueryExport),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecPostScanActionsBigqueryExportList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecPostScanActionsToHclTerraform = dataplexDatascanDataQualitySpecPostScanActionsToHclTerraform;
class DataplexDatascanDataQualitySpecPostScanActionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // bigquery_export - computed: false, optional: true, required: false
        this._bigqueryExport = new DataplexDatascanDataQualitySpecPostScanActionsBigqueryExportOutputReference(this, "bigquery_export");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bigqueryExport?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.bigqueryExport = this._bigqueryExport?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bigqueryExport.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bigqueryExport.internalValue = value.bigqueryExport;
        }
    }
    get bigqueryExport() {
        return this._bigqueryExport;
    }
    putBigqueryExport(value) {
        this._bigqueryExport.internalValue = value;
    }
    resetBigqueryExport() {
        this._bigqueryExport.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigqueryExportInput() {
        return this._bigqueryExport.internalValue;
    }
}
exports.DataplexDatascanDataQualitySpecPostScanActionsOutputReference = DataplexDatascanDataQualitySpecPostScanActionsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecPostScanActionsOutputReference[_k] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecPostScanActionsOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesNonNullExpectationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataplexDatascanDataQualitySpecRulesNonNullExpectationToTerraform = dataplexDatascanDataQualitySpecRulesNonNullExpectationToTerraform;
function dataplexDatascanDataQualitySpecRulesNonNullExpectationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataplexDatascanDataQualitySpecRulesNonNullExpectationToHclTerraform = dataplexDatascanDataQualitySpecRulesNonNullExpectationToHclTerraform;
class DataplexDatascanDataQualitySpecRulesNonNullExpectationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataplexDatascanDataQualitySpecRulesNonNullExpectationOutputReference = DataplexDatascanDataQualitySpecRulesNonNullExpectationOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesNonNullExpectationOutputReference[_l] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesNonNullExpectationOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesRangeExpectationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_value: cdktf.stringToTerraform(struct.maxValue),
        min_value: cdktf.stringToTerraform(struct.minValue),
        strict_max_enabled: cdktf.booleanToTerraform(struct.strictMaxEnabled),
        strict_min_enabled: cdktf.booleanToTerraform(struct.strictMinEnabled),
    };
}
exports.dataplexDatascanDataQualitySpecRulesRangeExpectationToTerraform = dataplexDatascanDataQualitySpecRulesRangeExpectationToTerraform;
function dataplexDatascanDataQualitySpecRulesRangeExpectationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_value: {
            value: cdktf.stringToHclTerraform(struct.maxValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        min_value: {
            value: cdktf.stringToHclTerraform(struct.minValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        strict_max_enabled: {
            value: cdktf.booleanToHclTerraform(struct.strictMaxEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        strict_min_enabled: {
            value: cdktf.booleanToHclTerraform(struct.strictMinEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecRulesRangeExpectationToHclTerraform = dataplexDatascanDataQualitySpecRulesRangeExpectationToHclTerraform;
class DataplexDatascanDataQualitySpecRulesRangeExpectationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxValue = this._maxValue;
        }
        if (this._minValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.minValue = this._minValue;
        }
        if (this._strictMaxEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.strictMaxEnabled = this._strictMaxEnabled;
        }
        if (this._strictMinEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.strictMinEnabled = this._strictMinEnabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxValue = undefined;
            this._minValue = undefined;
            this._strictMaxEnabled = undefined;
            this._strictMinEnabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxValue = value.maxValue;
            this._minValue = value.minValue;
            this._strictMaxEnabled = value.strictMaxEnabled;
            this._strictMinEnabled = value.strictMinEnabled;
        }
    }
    get maxValue() {
        return this.getStringAttribute('max_value');
    }
    set maxValue(value) {
        this._maxValue = value;
    }
    resetMaxValue() {
        this._maxValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxValueInput() {
        return this._maxValue;
    }
    get minValue() {
        return this.getStringAttribute('min_value');
    }
    set minValue(value) {
        this._minValue = value;
    }
    resetMinValue() {
        this._minValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minValueInput() {
        return this._minValue;
    }
    get strictMaxEnabled() {
        return this.getBooleanAttribute('strict_max_enabled');
    }
    set strictMaxEnabled(value) {
        this._strictMaxEnabled = value;
    }
    resetStrictMaxEnabled() {
        this._strictMaxEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get strictMaxEnabledInput() {
        return this._strictMaxEnabled;
    }
    get strictMinEnabled() {
        return this.getBooleanAttribute('strict_min_enabled');
    }
    set strictMinEnabled(value) {
        this._strictMinEnabled = value;
    }
    resetStrictMinEnabled() {
        this._strictMinEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get strictMinEnabledInput() {
        return this._strictMinEnabled;
    }
}
exports.DataplexDatascanDataQualitySpecRulesRangeExpectationOutputReference = DataplexDatascanDataQualitySpecRulesRangeExpectationOutputReference;
_m = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesRangeExpectationOutputReference[_m] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesRangeExpectationOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesRegexExpectationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        regex: cdktf.stringToTerraform(struct.regex),
    };
}
exports.dataplexDatascanDataQualitySpecRulesRegexExpectationToTerraform = dataplexDatascanDataQualitySpecRulesRegexExpectationToTerraform;
function dataplexDatascanDataQualitySpecRulesRegexExpectationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        regex: {
            value: cdktf.stringToHclTerraform(struct.regex),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecRulesRegexExpectationToHclTerraform = dataplexDatascanDataQualitySpecRulesRegexExpectationToHclTerraform;
class DataplexDatascanDataQualitySpecRulesRegexExpectationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._regex !== undefined) {
            hasAnyValues = true;
            internalValueResult.regex = this._regex;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._regex = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._regex = value.regex;
        }
    }
    get regex() {
        return this.getStringAttribute('regex');
    }
    set regex(value) {
        this._regex = value;
    }
    // Temporarily expose input value. Use with caution.
    get regexInput() {
        return this._regex;
    }
}
exports.DataplexDatascanDataQualitySpecRulesRegexExpectationOutputReference = DataplexDatascanDataQualitySpecRulesRegexExpectationOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesRegexExpectationOutputReference[_o] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesRegexExpectationOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesRowConditionExpectationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sql_expression: cdktf.stringToTerraform(struct.sqlExpression),
    };
}
exports.dataplexDatascanDataQualitySpecRulesRowConditionExpectationToTerraform = dataplexDatascanDataQualitySpecRulesRowConditionExpectationToTerraform;
function dataplexDatascanDataQualitySpecRulesRowConditionExpectationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        sql_expression: {
            value: cdktf.stringToHclTerraform(struct.sqlExpression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecRulesRowConditionExpectationToHclTerraform = dataplexDatascanDataQualitySpecRulesRowConditionExpectationToHclTerraform;
class DataplexDatascanDataQualitySpecRulesRowConditionExpectationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sqlExpression !== undefined) {
            hasAnyValues = true;
            internalValueResult.sqlExpression = this._sqlExpression;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sqlExpression = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sqlExpression = value.sqlExpression;
        }
    }
    get sqlExpression() {
        return this.getStringAttribute('sql_expression');
    }
    set sqlExpression(value) {
        this._sqlExpression = value;
    }
    // Temporarily expose input value. Use with caution.
    get sqlExpressionInput() {
        return this._sqlExpression;
    }
}
exports.DataplexDatascanDataQualitySpecRulesRowConditionExpectationOutputReference = DataplexDatascanDataQualitySpecRulesRowConditionExpectationOutputReference;
_p = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesRowConditionExpectationOutputReference[_p] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesRowConditionExpectationOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesSetExpectationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.dataplexDatascanDataQualitySpecRulesSetExpectationToTerraform = dataplexDatascanDataQualitySpecRulesSetExpectationToTerraform;
function dataplexDatascanDataQualitySpecRulesSetExpectationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        values: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.values),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecRulesSetExpectationToHclTerraform = dataplexDatascanDataQualitySpecRulesSetExpectationToHclTerraform;
class DataplexDatascanDataQualitySpecRulesSetExpectationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._values = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._values = value.values;
        }
    }
    get values() {
        return this.getListAttribute('values');
    }
    set values(value) {
        this._values = value;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.DataplexDatascanDataQualitySpecRulesSetExpectationOutputReference = DataplexDatascanDataQualitySpecRulesSetExpectationOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesSetExpectationOutputReference[_q] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesSetExpectationOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesSqlAssertionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sql_statement: cdktf.stringToTerraform(struct.sqlStatement),
    };
}
exports.dataplexDatascanDataQualitySpecRulesSqlAssertionToTerraform = dataplexDatascanDataQualitySpecRulesSqlAssertionToTerraform;
function dataplexDatascanDataQualitySpecRulesSqlAssertionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        sql_statement: {
            value: cdktf.stringToHclTerraform(struct.sqlStatement),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecRulesSqlAssertionToHclTerraform = dataplexDatascanDataQualitySpecRulesSqlAssertionToHclTerraform;
class DataplexDatascanDataQualitySpecRulesSqlAssertionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sqlStatement !== undefined) {
            hasAnyValues = true;
            internalValueResult.sqlStatement = this._sqlStatement;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sqlStatement = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sqlStatement = value.sqlStatement;
        }
    }
    get sqlStatement() {
        return this.getStringAttribute('sql_statement');
    }
    set sqlStatement(value) {
        this._sqlStatement = value;
    }
    // Temporarily expose input value. Use with caution.
    get sqlStatementInput() {
        return this._sqlStatement;
    }
}
exports.DataplexDatascanDataQualitySpecRulesSqlAssertionOutputReference = DataplexDatascanDataQualitySpecRulesSqlAssertionOutputReference;
_r = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesSqlAssertionOutputReference[_r] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesSqlAssertionOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_value: cdktf.stringToTerraform(struct.maxValue),
        min_value: cdktf.stringToTerraform(struct.minValue),
        statistic: cdktf.stringToTerraform(struct.statistic),
        strict_max_enabled: cdktf.booleanToTerraform(struct.strictMaxEnabled),
        strict_min_enabled: cdktf.booleanToTerraform(struct.strictMinEnabled),
    };
}
exports.dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToTerraform = dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToTerraform;
function dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_value: {
            value: cdktf.stringToHclTerraform(struct.maxValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        min_value: {
            value: cdktf.stringToHclTerraform(struct.minValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        statistic: {
            value: cdktf.stringToHclTerraform(struct.statistic),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        strict_max_enabled: {
            value: cdktf.booleanToHclTerraform(struct.strictMaxEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        strict_min_enabled: {
            value: cdktf.booleanToHclTerraform(struct.strictMinEnabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToHclTerraform = dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToHclTerraform;
class DataplexDatascanDataQualitySpecRulesStatisticRangeExpectationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxValue = this._maxValue;
        }
        if (this._minValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.minValue = this._minValue;
        }
        if (this._statistic !== undefined) {
            hasAnyValues = true;
            internalValueResult.statistic = this._statistic;
        }
        if (this._strictMaxEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.strictMaxEnabled = this._strictMaxEnabled;
        }
        if (this._strictMinEnabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.strictMinEnabled = this._strictMinEnabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxValue = undefined;
            this._minValue = undefined;
            this._statistic = undefined;
            this._strictMaxEnabled = undefined;
            this._strictMinEnabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxValue = value.maxValue;
            this._minValue = value.minValue;
            this._statistic = value.statistic;
            this._strictMaxEnabled = value.strictMaxEnabled;
            this._strictMinEnabled = value.strictMinEnabled;
        }
    }
    get maxValue() {
        return this.getStringAttribute('max_value');
    }
    set maxValue(value) {
        this._maxValue = value;
    }
    resetMaxValue() {
        this._maxValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxValueInput() {
        return this._maxValue;
    }
    get minValue() {
        return this.getStringAttribute('min_value');
    }
    set minValue(value) {
        this._minValue = value;
    }
    resetMinValue() {
        this._minValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minValueInput() {
        return this._minValue;
    }
    get statistic() {
        return this.getStringAttribute('statistic');
    }
    set statistic(value) {
        this._statistic = value;
    }
    // Temporarily expose input value. Use with caution.
    get statisticInput() {
        return this._statistic;
    }
    get strictMaxEnabled() {
        return this.getBooleanAttribute('strict_max_enabled');
    }
    set strictMaxEnabled(value) {
        this._strictMaxEnabled = value;
    }
    resetStrictMaxEnabled() {
        this._strictMaxEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get strictMaxEnabledInput() {
        return this._strictMaxEnabled;
    }
    get strictMinEnabled() {
        return this.getBooleanAttribute('strict_min_enabled');
    }
    set strictMinEnabled(value) {
        this._strictMinEnabled = value;
    }
    resetStrictMinEnabled() {
        this._strictMinEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get strictMinEnabledInput() {
        return this._strictMinEnabled;
    }
}
exports.DataplexDatascanDataQualitySpecRulesStatisticRangeExpectationOutputReference = DataplexDatascanDataQualitySpecRulesStatisticRangeExpectationOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesStatisticRangeExpectationOutputReference[_s] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesStatisticRangeExpectationOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesTableConditionExpectationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sql_expression: cdktf.stringToTerraform(struct.sqlExpression),
    };
}
exports.dataplexDatascanDataQualitySpecRulesTableConditionExpectationToTerraform = dataplexDatascanDataQualitySpecRulesTableConditionExpectationToTerraform;
function dataplexDatascanDataQualitySpecRulesTableConditionExpectationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        sql_expression: {
            value: cdktf.stringToHclTerraform(struct.sqlExpression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecRulesTableConditionExpectationToHclTerraform = dataplexDatascanDataQualitySpecRulesTableConditionExpectationToHclTerraform;
class DataplexDatascanDataQualitySpecRulesTableConditionExpectationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sqlExpression !== undefined) {
            hasAnyValues = true;
            internalValueResult.sqlExpression = this._sqlExpression;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sqlExpression = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sqlExpression = value.sqlExpression;
        }
    }
    get sqlExpression() {
        return this.getStringAttribute('sql_expression');
    }
    set sqlExpression(value) {
        this._sqlExpression = value;
    }
    // Temporarily expose input value. Use with caution.
    get sqlExpressionInput() {
        return this._sqlExpression;
    }
}
exports.DataplexDatascanDataQualitySpecRulesTableConditionExpectationOutputReference = DataplexDatascanDataQualitySpecRulesTableConditionExpectationOutputReference;
_t = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesTableConditionExpectationOutputReference[_t] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesTableConditionExpectationOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesUniquenessExpectationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataplexDatascanDataQualitySpecRulesUniquenessExpectationToTerraform = dataplexDatascanDataQualitySpecRulesUniquenessExpectationToTerraform;
function dataplexDatascanDataQualitySpecRulesUniquenessExpectationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataplexDatascanDataQualitySpecRulesUniquenessExpectationToHclTerraform = dataplexDatascanDataQualitySpecRulesUniquenessExpectationToHclTerraform;
class DataplexDatascanDataQualitySpecRulesUniquenessExpectationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataplexDatascanDataQualitySpecRulesUniquenessExpectationOutputReference = DataplexDatascanDataQualitySpecRulesUniquenessExpectationOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesUniquenessExpectationOutputReference[_u] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesUniquenessExpectationOutputReference", version: "14.13.1" };
function dataplexDatascanDataQualitySpecRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        column: cdktf.stringToTerraform(struct.column),
        description: cdktf.stringToTerraform(struct.description),
        dimension: cdktf.stringToTerraform(struct.dimension),
        ignore_null: cdktf.booleanToTerraform(struct.ignoreNull),
        name: cdktf.stringToTerraform(struct.name),
        threshold: cdktf.numberToTerraform(struct.threshold),
        non_null_expectation: dataplexDatascanDataQualitySpecRulesNonNullExpectationToTerraform(struct.nonNullExpectation),
        range_expectation: dataplexDatascanDataQualitySpecRulesRangeExpectationToTerraform(struct.rangeExpectation),
        regex_expectation: dataplexDatascanDataQualitySpecRulesRegexExpectationToTerraform(struct.regexExpectation),
        row_condition_expectation: dataplexDatascanDataQualitySpecRulesRowConditionExpectationToTerraform(struct.rowConditionExpectation),
        set_expectation: dataplexDatascanDataQualitySpecRulesSetExpectationToTerraform(struct.setExpectation),
        sql_assertion: dataplexDatascanDataQualitySpecRulesSqlAssertionToTerraform(struct.sqlAssertion),
        statistic_range_expectation: dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToTerraform(struct.statisticRangeExpectation),
        table_condition_expectation: dataplexDatascanDataQualitySpecRulesTableConditionExpectationToTerraform(struct.tableConditionExpectation),
        uniqueness_expectation: dataplexDatascanDataQualitySpecRulesUniquenessExpectationToTerraform(struct.uniquenessExpectation),
    };
}
exports.dataplexDatascanDataQualitySpecRulesToTerraform = dataplexDatascanDataQualitySpecRulesToTerraform;
function dataplexDatascanDataQualitySpecRulesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        column: {
            value: cdktf.stringToHclTerraform(struct.column),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        dimension: {
            value: cdktf.stringToHclTerraform(struct.dimension),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ignore_null: {
            value: cdktf.booleanToHclTerraform(struct.ignoreNull),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        threshold: {
            value: cdktf.numberToHclTerraform(struct.threshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        non_null_expectation: {
            value: dataplexDatascanDataQualitySpecRulesNonNullExpectationToHclTerraform(struct.nonNullExpectation),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesNonNullExpectationList",
        },
        range_expectation: {
            value: dataplexDatascanDataQualitySpecRulesRangeExpectationToHclTerraform(struct.rangeExpectation),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesRangeExpectationList",
        },
        regex_expectation: {
            value: dataplexDatascanDataQualitySpecRulesRegexExpectationToHclTerraform(struct.regexExpectation),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesRegexExpectationList",
        },
        row_condition_expectation: {
            value: dataplexDatascanDataQualitySpecRulesRowConditionExpectationToHclTerraform(struct.rowConditionExpectation),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesRowConditionExpectationList",
        },
        set_expectation: {
            value: dataplexDatascanDataQualitySpecRulesSetExpectationToHclTerraform(struct.setExpectation),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesSetExpectationList",
        },
        sql_assertion: {
            value: dataplexDatascanDataQualitySpecRulesSqlAssertionToHclTerraform(struct.sqlAssertion),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesSqlAssertionList",
        },
        statistic_range_expectation: {
            value: dataplexDatascanDataQualitySpecRulesStatisticRangeExpectationToHclTerraform(struct.statisticRangeExpectation),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesStatisticRangeExpectationList",
        },
        table_condition_expectation: {
            value: dataplexDatascanDataQualitySpecRulesTableConditionExpectationToHclTerraform(struct.tableConditionExpectation),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesTableConditionExpectationList",
        },
        uniqueness_expectation: {
            value: dataplexDatascanDataQualitySpecRulesUniquenessExpectationToHclTerraform(struct.uniquenessExpectation),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesUniquenessExpectationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecRulesToHclTerraform = dataplexDatascanDataQualitySpecRulesToHclTerraform;
class DataplexDatascanDataQualitySpecRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // non_null_expectation - computed: false, optional: true, required: false
        this._nonNullExpectation = new DataplexDatascanDataQualitySpecRulesNonNullExpectationOutputReference(this, "non_null_expectation");
        // range_expectation - computed: false, optional: true, required: false
        this._rangeExpectation = new DataplexDatascanDataQualitySpecRulesRangeExpectationOutputReference(this, "range_expectation");
        // regex_expectation - computed: false, optional: true, required: false
        this._regexExpectation = new DataplexDatascanDataQualitySpecRulesRegexExpectationOutputReference(this, "regex_expectation");
        // row_condition_expectation - computed: false, optional: true, required: false
        this._rowConditionExpectation = new DataplexDatascanDataQualitySpecRulesRowConditionExpectationOutputReference(this, "row_condition_expectation");
        // set_expectation - computed: false, optional: true, required: false
        this._setExpectation = new DataplexDatascanDataQualitySpecRulesSetExpectationOutputReference(this, "set_expectation");
        // sql_assertion - computed: false, optional: true, required: false
        this._sqlAssertion = new DataplexDatascanDataQualitySpecRulesSqlAssertionOutputReference(this, "sql_assertion");
        // statistic_range_expectation - computed: false, optional: true, required: false
        this._statisticRangeExpectation = new DataplexDatascanDataQualitySpecRulesStatisticRangeExpectationOutputReference(this, "statistic_range_expectation");
        // table_condition_expectation - computed: false, optional: true, required: false
        this._tableConditionExpectation = new DataplexDatascanDataQualitySpecRulesTableConditionExpectationOutputReference(this, "table_condition_expectation");
        // uniqueness_expectation - computed: false, optional: true, required: false
        this._uniquenessExpectation = new DataplexDatascanDataQualitySpecRulesUniquenessExpectationOutputReference(this, "uniqueness_expectation");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._column !== undefined) {
            hasAnyValues = true;
            internalValueResult.column = this._column;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._dimension !== undefined) {
            hasAnyValues = true;
            internalValueResult.dimension = this._dimension;
        }
        if (this._ignoreNull !== undefined) {
            hasAnyValues = true;
            internalValueResult.ignoreNull = this._ignoreNull;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._threshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.threshold = this._threshold;
        }
        if (this._nonNullExpectation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.nonNullExpectation = this._nonNullExpectation?.internalValue;
        }
        if (this._rangeExpectation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rangeExpectation = this._rangeExpectation?.internalValue;
        }
        if (this._regexExpectation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.regexExpectation = this._regexExpectation?.internalValue;
        }
        if (this._rowConditionExpectation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rowConditionExpectation = this._rowConditionExpectation?.internalValue;
        }
        if (this._setExpectation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.setExpectation = this._setExpectation?.internalValue;
        }
        if (this._sqlAssertion?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sqlAssertion = this._sqlAssertion?.internalValue;
        }
        if (this._statisticRangeExpectation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.statisticRangeExpectation = this._statisticRangeExpectation?.internalValue;
        }
        if (this._tableConditionExpectation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tableConditionExpectation = this._tableConditionExpectation?.internalValue;
        }
        if (this._uniquenessExpectation?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.uniquenessExpectation = this._uniquenessExpectation?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._column = undefined;
            this._description = undefined;
            this._dimension = undefined;
            this._ignoreNull = undefined;
            this._name = undefined;
            this._threshold = undefined;
            this._nonNullExpectation.internalValue = undefined;
            this._rangeExpectation.internalValue = undefined;
            this._regexExpectation.internalValue = undefined;
            this._rowConditionExpectation.internalValue = undefined;
            this._setExpectation.internalValue = undefined;
            this._sqlAssertion.internalValue = undefined;
            this._statisticRangeExpectation.internalValue = undefined;
            this._tableConditionExpectation.internalValue = undefined;
            this._uniquenessExpectation.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._column = value.column;
            this._description = value.description;
            this._dimension = value.dimension;
            this._ignoreNull = value.ignoreNull;
            this._name = value.name;
            this._threshold = value.threshold;
            this._nonNullExpectation.internalValue = value.nonNullExpectation;
            this._rangeExpectation.internalValue = value.rangeExpectation;
            this._regexExpectation.internalValue = value.regexExpectation;
            this._rowConditionExpectation.internalValue = value.rowConditionExpectation;
            this._setExpectation.internalValue = value.setExpectation;
            this._sqlAssertion.internalValue = value.sqlAssertion;
            this._statisticRangeExpectation.internalValue = value.statisticRangeExpectation;
            this._tableConditionExpectation.internalValue = value.tableConditionExpectation;
            this._uniquenessExpectation.internalValue = value.uniquenessExpectation;
        }
    }
    get column() {
        return this.getStringAttribute('column');
    }
    set column(value) {
        this._column = value;
    }
    resetColumn() {
        this._column = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get columnInput() {
        return this._column;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get dimension() {
        return this.getStringAttribute('dimension');
    }
    set dimension(value) {
        this._dimension = value;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionInput() {
        return this._dimension;
    }
    get ignoreNull() {
        return this.getBooleanAttribute('ignore_null');
    }
    set ignoreNull(value) {
        this._ignoreNull = value;
    }
    resetIgnoreNull() {
        this._ignoreNull = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreNullInput() {
        return this._ignoreNull;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get threshold() {
        return this.getNumberAttribute('threshold');
    }
    set threshold(value) {
        this._threshold = value;
    }
    resetThreshold() {
        this._threshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get thresholdInput() {
        return this._threshold;
    }
    get nonNullExpectation() {
        return this._nonNullExpectation;
    }
    putNonNullExpectation(value) {
        this._nonNullExpectation.internalValue = value;
    }
    resetNonNullExpectation() {
        this._nonNullExpectation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nonNullExpectationInput() {
        return this._nonNullExpectation.internalValue;
    }
    get rangeExpectation() {
        return this._rangeExpectation;
    }
    putRangeExpectation(value) {
        this._rangeExpectation.internalValue = value;
    }
    resetRangeExpectation() {
        this._rangeExpectation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rangeExpectationInput() {
        return this._rangeExpectation.internalValue;
    }
    get regexExpectation() {
        return this._regexExpectation;
    }
    putRegexExpectation(value) {
        this._regexExpectation.internalValue = value;
    }
    resetRegexExpectation() {
        this._regexExpectation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regexExpectationInput() {
        return this._regexExpectation.internalValue;
    }
    get rowConditionExpectation() {
        return this._rowConditionExpectation;
    }
    putRowConditionExpectation(value) {
        this._rowConditionExpectation.internalValue = value;
    }
    resetRowConditionExpectation() {
        this._rowConditionExpectation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rowConditionExpectationInput() {
        return this._rowConditionExpectation.internalValue;
    }
    get setExpectation() {
        return this._setExpectation;
    }
    putSetExpectation(value) {
        this._setExpectation.internalValue = value;
    }
    resetSetExpectation() {
        this._setExpectation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setExpectationInput() {
        return this._setExpectation.internalValue;
    }
    get sqlAssertion() {
        return this._sqlAssertion;
    }
    putSqlAssertion(value) {
        this._sqlAssertion.internalValue = value;
    }
    resetSqlAssertion() {
        this._sqlAssertion.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqlAssertionInput() {
        return this._sqlAssertion.internalValue;
    }
    get statisticRangeExpectation() {
        return this._statisticRangeExpectation;
    }
    putStatisticRangeExpectation(value) {
        this._statisticRangeExpectation.internalValue = value;
    }
    resetStatisticRangeExpectation() {
        this._statisticRangeExpectation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statisticRangeExpectationInput() {
        return this._statisticRangeExpectation.internalValue;
    }
    get tableConditionExpectation() {
        return this._tableConditionExpectation;
    }
    putTableConditionExpectation(value) {
        this._tableConditionExpectation.internalValue = value;
    }
    resetTableConditionExpectation() {
        this._tableConditionExpectation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tableConditionExpectationInput() {
        return this._tableConditionExpectation.internalValue;
    }
    get uniquenessExpectation() {
        return this._uniquenessExpectation;
    }
    putUniquenessExpectation(value) {
        this._uniquenessExpectation.internalValue = value;
    }
    resetUniquenessExpectation() {
        this._uniquenessExpectation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uniquenessExpectationInput() {
        return this._uniquenessExpectation.internalValue;
    }
}
exports.DataplexDatascanDataQualitySpecRulesOutputReference = DataplexDatascanDataQualitySpecRulesOutputReference;
_v = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesOutputReference[_v] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesOutputReference", version: "14.13.1" };
class DataplexDatascanDataQualitySpecRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataplexDatascanDataQualitySpecRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataplexDatascanDataQualitySpecRulesList = DataplexDatascanDataQualitySpecRulesList;
_w = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecRulesList[_w] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecRulesList", version: "14.13.1" };
function dataplexDatascanDataQualitySpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        row_filter: cdktf.stringToTerraform(struct.rowFilter),
        sampling_percent: cdktf.numberToTerraform(struct.samplingPercent),
        post_scan_actions: dataplexDatascanDataQualitySpecPostScanActionsToTerraform(struct.postScanActions),
        rules: cdktf.listMapper(dataplexDatascanDataQualitySpecRulesToTerraform, true)(struct.rules),
    };
}
exports.dataplexDatascanDataQualitySpecToTerraform = dataplexDatascanDataQualitySpecToTerraform;
function dataplexDatascanDataQualitySpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        row_filter: {
            value: cdktf.stringToHclTerraform(struct.rowFilter),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        sampling_percent: {
            value: cdktf.numberToHclTerraform(struct.samplingPercent),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        post_scan_actions: {
            value: dataplexDatascanDataQualitySpecPostScanActionsToHclTerraform(struct.postScanActions),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecPostScanActionsList",
        },
        rules: {
            value: cdktf.listMapperHcl(dataplexDatascanDataQualitySpecRulesToHclTerraform, true)(struct.rules),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanDataQualitySpecRulesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanDataQualitySpecToHclTerraform = dataplexDatascanDataQualitySpecToHclTerraform;
class DataplexDatascanDataQualitySpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // post_scan_actions - computed: false, optional: true, required: false
        this._postScanActions = new DataplexDatascanDataQualitySpecPostScanActionsOutputReference(this, "post_scan_actions");
        // rules - computed: false, optional: true, required: false
        this._rules = new DataplexDatascanDataQualitySpecRulesList(this, "rules", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rowFilter !== undefined) {
            hasAnyValues = true;
            internalValueResult.rowFilter = this._rowFilter;
        }
        if (this._samplingPercent !== undefined) {
            hasAnyValues = true;
            internalValueResult.samplingPercent = this._samplingPercent;
        }
        if (this._postScanActions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.postScanActions = this._postScanActions?.internalValue;
        }
        if (this._rules?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rules = this._rules?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rowFilter = undefined;
            this._samplingPercent = undefined;
            this._postScanActions.internalValue = undefined;
            this._rules.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rowFilter = value.rowFilter;
            this._samplingPercent = value.samplingPercent;
            this._postScanActions.internalValue = value.postScanActions;
            this._rules.internalValue = value.rules;
        }
    }
    get rowFilter() {
        return this.getStringAttribute('row_filter');
    }
    set rowFilter(value) {
        this._rowFilter = value;
    }
    resetRowFilter() {
        this._rowFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rowFilterInput() {
        return this._rowFilter;
    }
    get samplingPercent() {
        return this.getNumberAttribute('sampling_percent');
    }
    set samplingPercent(value) {
        this._samplingPercent = value;
    }
    resetSamplingPercent() {
        this._samplingPercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get samplingPercentInput() {
        return this._samplingPercent;
    }
    get postScanActions() {
        return this._postScanActions;
    }
    putPostScanActions(value) {
        this._postScanActions.internalValue = value;
    }
    resetPostScanActions() {
        this._postScanActions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postScanActionsInput() {
        return this._postScanActions.internalValue;
    }
    get rules() {
        return this._rules;
    }
    putRules(value) {
        this._rules.internalValue = value;
    }
    resetRules() {
        this._rules.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rulesInput() {
        return this._rules.internalValue;
    }
}
exports.DataplexDatascanDataQualitySpecOutputReference = DataplexDatascanDataQualitySpecOutputReference;
_x = JSII_RTTI_SYMBOL_1;
DataplexDatascanDataQualitySpecOutputReference[_x] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanDataQualitySpecOutputReference", version: "14.13.1" };
function dataplexDatascanExecutionSpecTriggerOnDemandToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataplexDatascanExecutionSpecTriggerOnDemandToTerraform = dataplexDatascanExecutionSpecTriggerOnDemandToTerraform;
function dataplexDatascanExecutionSpecTriggerOnDemandToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataplexDatascanExecutionSpecTriggerOnDemandToHclTerraform = dataplexDatascanExecutionSpecTriggerOnDemandToHclTerraform;
class DataplexDatascanExecutionSpecTriggerOnDemandOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataplexDatascanExecutionSpecTriggerOnDemandOutputReference = DataplexDatascanExecutionSpecTriggerOnDemandOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataplexDatascanExecutionSpecTriggerOnDemandOutputReference[_y] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanExecutionSpecTriggerOnDemandOutputReference", version: "14.13.1" };
function dataplexDatascanExecutionSpecTriggerScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cron: cdktf.stringToTerraform(struct.cron),
    };
}
exports.dataplexDatascanExecutionSpecTriggerScheduleToTerraform = dataplexDatascanExecutionSpecTriggerScheduleToTerraform;
function dataplexDatascanExecutionSpecTriggerScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cron: {
            value: cdktf.stringToHclTerraform(struct.cron),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanExecutionSpecTriggerScheduleToHclTerraform = dataplexDatascanExecutionSpecTriggerScheduleToHclTerraform;
class DataplexDatascanExecutionSpecTriggerScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cron !== undefined) {
            hasAnyValues = true;
            internalValueResult.cron = this._cron;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cron = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cron = value.cron;
        }
    }
    get cron() {
        return this.getStringAttribute('cron');
    }
    set cron(value) {
        this._cron = value;
    }
    // Temporarily expose input value. Use with caution.
    get cronInput() {
        return this._cron;
    }
}
exports.DataplexDatascanExecutionSpecTriggerScheduleOutputReference = DataplexDatascanExecutionSpecTriggerScheduleOutputReference;
_z = JSII_RTTI_SYMBOL_1;
DataplexDatascanExecutionSpecTriggerScheduleOutputReference[_z] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanExecutionSpecTriggerScheduleOutputReference", version: "14.13.1" };
function dataplexDatascanExecutionSpecTriggerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        on_demand: dataplexDatascanExecutionSpecTriggerOnDemandToTerraform(struct.onDemand),
        schedule: dataplexDatascanExecutionSpecTriggerScheduleToTerraform(struct.schedule),
    };
}
exports.dataplexDatascanExecutionSpecTriggerToTerraform = dataplexDatascanExecutionSpecTriggerToTerraform;
function dataplexDatascanExecutionSpecTriggerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        on_demand: {
            value: dataplexDatascanExecutionSpecTriggerOnDemandToHclTerraform(struct.onDemand),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanExecutionSpecTriggerOnDemandList",
        },
        schedule: {
            value: dataplexDatascanExecutionSpecTriggerScheduleToHclTerraform(struct.schedule),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanExecutionSpecTriggerScheduleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanExecutionSpecTriggerToHclTerraform = dataplexDatascanExecutionSpecTriggerToHclTerraform;
class DataplexDatascanExecutionSpecTriggerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // on_demand - computed: false, optional: true, required: false
        this._onDemand = new DataplexDatascanExecutionSpecTriggerOnDemandOutputReference(this, "on_demand");
        // schedule - computed: false, optional: true, required: false
        this._schedule = new DataplexDatascanExecutionSpecTriggerScheduleOutputReference(this, "schedule");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._onDemand?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.onDemand = this._onDemand?.internalValue;
        }
        if (this._schedule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.schedule = this._schedule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._onDemand.internalValue = undefined;
            this._schedule.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._onDemand.internalValue = value.onDemand;
            this._schedule.internalValue = value.schedule;
        }
    }
    get onDemand() {
        return this._onDemand;
    }
    putOnDemand(value) {
        this._onDemand.internalValue = value;
    }
    resetOnDemand() {
        this._onDemand.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get onDemandInput() {
        return this._onDemand.internalValue;
    }
    get schedule() {
        return this._schedule;
    }
    putSchedule(value) {
        this._schedule.internalValue = value;
    }
    resetSchedule() {
        this._schedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule.internalValue;
    }
}
exports.DataplexDatascanExecutionSpecTriggerOutputReference = DataplexDatascanExecutionSpecTriggerOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataplexDatascanExecutionSpecTriggerOutputReference[_0] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanExecutionSpecTriggerOutputReference", version: "14.13.1" };
function dataplexDatascanExecutionSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        field: cdktf.stringToTerraform(struct.field),
        trigger: dataplexDatascanExecutionSpecTriggerToTerraform(struct.trigger),
    };
}
exports.dataplexDatascanExecutionSpecToTerraform = dataplexDatascanExecutionSpecToTerraform;
function dataplexDatascanExecutionSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        field: {
            value: cdktf.stringToHclTerraform(struct.field),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        trigger: {
            value: dataplexDatascanExecutionSpecTriggerToHclTerraform(struct.trigger),
            isBlock: true,
            type: "list",
            storageClassType: "DataplexDatascanExecutionSpecTriggerList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanExecutionSpecToHclTerraform = dataplexDatascanExecutionSpecToHclTerraform;
class DataplexDatascanExecutionSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // trigger - computed: false, optional: false, required: true
        this._trigger = new DataplexDatascanExecutionSpecTriggerOutputReference(this, "trigger");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._field !== undefined) {
            hasAnyValues = true;
            internalValueResult.field = this._field;
        }
        if (this._trigger?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.trigger = this._trigger?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._field = undefined;
            this._trigger.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._field = value.field;
            this._trigger.internalValue = value.trigger;
        }
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    resetField() {
        this._field = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get trigger() {
        return this._trigger;
    }
    putTrigger(value) {
        this._trigger.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get triggerInput() {
        return this._trigger.internalValue;
    }
}
exports.DataplexDatascanExecutionSpecOutputReference = DataplexDatascanExecutionSpecOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
DataplexDatascanExecutionSpecOutputReference[_1] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanExecutionSpecOutputReference", version: "14.13.1" };
function dataplexDatascanTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dataplexDatascanTimeoutsToTerraform = dataplexDatascanTimeoutsToTerraform;
function dataplexDatascanTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataplexDatascanTimeoutsToHclTerraform = dataplexDatascanTimeoutsToHclTerraform;
class DataplexDatascanTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DataplexDatascanTimeoutsOutputReference = DataplexDatascanTimeoutsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataplexDatascanTimeoutsOutputReference[_2] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascanTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataplex_datascan google_dataplex_datascan}
*/
class DataplexDatascan extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataplexDatascan resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataplexDatascan to import
    * @param importFromId The id of the existing DataplexDatascan that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataplex_datascan#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataplexDatascan to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_dataplex_datascan", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataplex_datascan google_dataplex_datascan} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataplexDatascanConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_dataplex_datascan',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // execution_status - computed: true, optional: false, required: false
        this._executionStatus = new DataplexDatascanExecutionStatusList(this, "execution_status", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // data - computed: false, optional: false, required: true
        this._data = new DataplexDatascanDataOutputReference(this, "data");
        // data_profile_spec - computed: false, optional: true, required: false
        this._dataProfileSpec = new DataplexDatascanDataProfileSpecOutputReference(this, "data_profile_spec");
        // data_quality_spec - computed: false, optional: true, required: false
        this._dataQualitySpec = new DataplexDatascanDataQualitySpecOutputReference(this, "data_quality_spec");
        // execution_spec - computed: false, optional: false, required: true
        this._executionSpec = new DataplexDatascanExecutionSpecOutputReference(this, "execution_spec");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataplexDatascanTimeoutsOutputReference(this, "timeouts");
        this._dataScanId = config.dataScanId;
        this._description = config.description;
        this._displayName = config.displayName;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._project = config.project;
        this._data.internalValue = config.data;
        this._dataProfileSpec.internalValue = config.dataProfileSpec;
        this._dataQualitySpec.internalValue = config.dataQualitySpec;
        this._executionSpec.internalValue = config.executionSpec;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get dataScanId() {
        return this.getStringAttribute('data_scan_id');
    }
    set dataScanId(value) {
        this._dataScanId = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataScanIdInput() {
        return this._dataScanId;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get executionStatus() {
        return this._executionStatus;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get data() {
        return this._data;
    }
    putData(value) {
        this._data.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataInput() {
        return this._data.internalValue;
    }
    get dataProfileSpec() {
        return this._dataProfileSpec;
    }
    putDataProfileSpec(value) {
        this._dataProfileSpec.internalValue = value;
    }
    resetDataProfileSpec() {
        this._dataProfileSpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataProfileSpecInput() {
        return this._dataProfileSpec.internalValue;
    }
    get dataQualitySpec() {
        return this._dataQualitySpec;
    }
    putDataQualitySpec(value) {
        this._dataQualitySpec.internalValue = value;
    }
    resetDataQualitySpec() {
        this._dataQualitySpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataQualitySpecInput() {
        return this._dataQualitySpec.internalValue;
    }
    get executionSpec() {
        return this._executionSpec;
    }
    putExecutionSpec(value) {
        this._executionSpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get executionSpecInput() {
        return this._executionSpec.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            data_scan_id: cdktf.stringToTerraform(this._dataScanId),
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            data: dataplexDatascanDataToTerraform(this._data.internalValue),
            data_profile_spec: dataplexDatascanDataProfileSpecToTerraform(this._dataProfileSpec.internalValue),
            data_quality_spec: dataplexDatascanDataQualitySpecToTerraform(this._dataQualitySpec.internalValue),
            execution_spec: dataplexDatascanExecutionSpecToTerraform(this._executionSpec.internalValue),
            timeouts: dataplexDatascanTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            data_scan_id: {
                value: cdktf.stringToHclTerraform(this._dataScanId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data: {
                value: dataplexDatascanDataToHclTerraform(this._data.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataplexDatascanDataList",
            },
            data_profile_spec: {
                value: dataplexDatascanDataProfileSpecToHclTerraform(this._dataProfileSpec.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataplexDatascanDataProfileSpecList",
            },
            data_quality_spec: {
                value: dataplexDatascanDataQualitySpecToHclTerraform(this._dataQualitySpec.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataplexDatascanDataQualitySpecList",
            },
            execution_spec: {
                value: dataplexDatascanExecutionSpecToHclTerraform(this._executionSpec.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataplexDatascanExecutionSpecList",
            },
            timeouts: {
                value: dataplexDatascanTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataplexDatascanTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataplexDatascan = DataplexDatascan;
_3 = JSII_RTTI_SYMBOL_1;
DataplexDatascan[_3] = { fqn: "@cdktf/provider-google.dataplexDatascan.DataplexDatascan", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataplexDatascan.tfResourceType = "google_dataplex_datascan";
