// @cdktf/provider-google/lib/data-google-compute-resource-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyToHclTerraform = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyToTerraform = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesList = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToHclTerraform = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleList = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleToHclTerraform = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleToTerraform = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleList = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToHclTerraform = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToHclTerraform = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleList = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToHclTerraform = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleList = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToHclTerraform = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyList = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyToHclTerraform = exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform = exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyList = exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyOutputReference = exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyToHclTerraform = exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyToTerraform = exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleList = exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference = exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleToHclTerraform = exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform = exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleList = exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference = exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleToHclTerraform = exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform = exports.DataGoogleComputeResourcePolicyGroupPlacementPolicyList = exports.DataGoogleComputeResourcePolicyGroupPlacementPolicyOutputReference = exports.dataGoogleComputeResourcePolicyGroupPlacementPolicyToHclTerraform = exports.dataGoogleComputeResourcePolicyGroupPlacementPolicyToTerraform = exports.DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyList = exports.DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyOutputReference = exports.dataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyToHclTerraform = exports.dataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyToTerraform = void 0;
exports.DataGoogleComputeResourcePolicy = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyList = exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyToTerraform = dataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyToTerraform;
function dataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyToHclTerraform = dataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyToHclTerraform;
class DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyOutputReference = DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyList = DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyList", version: "14.13.1" };
function dataGoogleComputeResourcePolicyGroupPlacementPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicyGroupPlacementPolicyToTerraform = dataGoogleComputeResourcePolicyGroupPlacementPolicyToTerraform;
function dataGoogleComputeResourcePolicyGroupPlacementPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicyGroupPlacementPolicyToHclTerraform = dataGoogleComputeResourcePolicyGroupPlacementPolicyToHclTerraform;
class DataGoogleComputeResourcePolicyGroupPlacementPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // availability_domain_count - computed: true, optional: false, required: false
    get availabilityDomainCount() {
        return this.getNumberAttribute('availability_domain_count');
    }
    // collocation - computed: true, optional: false, required: false
    get collocation() {
        return this.getStringAttribute('collocation');
    }
    // vm_count - computed: true, optional: false, required: false
    get vmCount() {
        return this.getNumberAttribute('vm_count');
    }
}
exports.DataGoogleComputeResourcePolicyGroupPlacementPolicyOutputReference = DataGoogleComputeResourcePolicyGroupPlacementPolicyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyGroupPlacementPolicyOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyGroupPlacementPolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicyGroupPlacementPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicyGroupPlacementPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicyGroupPlacementPolicyList = DataGoogleComputeResourcePolicyGroupPlacementPolicyList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyGroupPlacementPolicyList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyGroupPlacementPolicyList", version: "14.13.1" };
function dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform = dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleToTerraform;
function dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleToHclTerraform = dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleToHclTerraform;
class DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // schedule - computed: true, optional: false, required: false
    get schedule() {
        return this.getStringAttribute('schedule');
    }
}
exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference = DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleList = DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleList", version: "14.13.1" };
function dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform = dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleToTerraform;
function dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleToHclTerraform = dataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleToHclTerraform;
class DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // schedule - computed: true, optional: false, required: false
    get schedule() {
        return this.getStringAttribute('schedule');
    }
}
exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference = DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleList = DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleList", version: "14.13.1" };
function dataGoogleComputeResourcePolicyInstanceSchedulePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyToTerraform = dataGoogleComputeResourcePolicyInstanceSchedulePolicyToTerraform;
function dataGoogleComputeResourcePolicyInstanceSchedulePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicyInstanceSchedulePolicyToHclTerraform = dataGoogleComputeResourcePolicyInstanceSchedulePolicyToHclTerraform;
class DataGoogleComputeResourcePolicyInstanceSchedulePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // vm_start_schedule - computed: true, optional: false, required: false
        this._vmStartSchedule = new DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStartScheduleList(this, "vm_start_schedule", false);
        // vm_stop_schedule - computed: true, optional: false, required: false
        this._vmStopSchedule = new DataGoogleComputeResourcePolicyInstanceSchedulePolicyVmStopScheduleList(this, "vm_stop_schedule", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // expiration_time - computed: true, optional: false, required: false
    get expirationTime() {
        return this.getStringAttribute('expiration_time');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    // time_zone - computed: true, optional: false, required: false
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    get vmStartSchedule() {
        return this._vmStartSchedule;
    }
    get vmStopSchedule() {
        return this._vmStopSchedule;
    }
}
exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyOutputReference = DataGoogleComputeResourcePolicyInstanceSchedulePolicyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyInstanceSchedulePolicyOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyInstanceSchedulePolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicyInstanceSchedulePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicyInstanceSchedulePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicyInstanceSchedulePolicyList = DataGoogleComputeResourcePolicyInstanceSchedulePolicyList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicyInstanceSchedulePolicyList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicyInstanceSchedulePolicyList", version: "14.13.1" };
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyToTerraform;
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyToHclTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyToHclTerraform;
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_retention_days - computed: true, optional: false, required: false
    get maxRetentionDays() {
        return this.getNumberAttribute('max_retention_days');
    }
    // on_source_disk_delete - computed: true, optional: false, required: false
    get onSourceDiskDelete() {
        return this.getStringAttribute('on_source_disk_delete');
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference = DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyList = DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyList", version: "14.13.1" };
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToTerraform;
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToHclTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleToHclTerraform;
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // days_in_cycle - computed: true, optional: false, required: false
    get daysInCycle() {
        return this.getNumberAttribute('days_in_cycle');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleList = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleList", version: "14.13.1" };
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToTerraform;
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToHclTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleToHclTerraform;
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hours_in_cycle - computed: true, optional: false, required: false
    get hoursInCycle() {
        return this.getNumberAttribute('hours_in_cycle');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleList = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleList[_r] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleList", version: "14.13.1" };
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToTerraform;
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToHclTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksToHclTerraform;
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day - computed: true, optional: false, required: false
    get day() {
        return this.getStringAttribute('day');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList[_t] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList", version: "14.13.1" };
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToTerraform;
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToHclTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleToHclTerraform;
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // day_of_weeks - computed: true, optional: false, required: false
        this._dayOfWeeks = new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleDayOfWeeksList(this, "day_of_weeks", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get dayOfWeeks() {
        return this._dayOfWeeks;
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleList = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleList[_v] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleList", version: "14.13.1" };
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleToTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleToTerraform;
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleToHclTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleToHclTerraform;
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // daily_schedule - computed: true, optional: false, required: false
        this._dailySchedule = new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleDailyScheduleList(this, "daily_schedule", false);
        // hourly_schedule - computed: true, optional: false, required: false
        this._hourlySchedule = new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleHourlyScheduleList(this, "hourly_schedule", false);
        // weekly_schedule - computed: true, optional: false, required: false
        this._weeklySchedule = new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleWeeklyScheduleList(this, "weekly_schedule", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get dailySchedule() {
        return this._dailySchedule;
    }
    get hourlySchedule() {
        return this._hourlySchedule;
    }
    get weeklySchedule() {
        return this._weeklySchedule;
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleList = DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleList[_x] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleList", version: "14.13.1" };
function dataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToTerraform;
function dataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToHclTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesToHclTerraform;
class DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // chain_name - computed: true, optional: false, required: false
    get chainName() {
        return this.getStringAttribute('chain_name');
    }
    // guest_flush - computed: true, optional: false, required: false
    get guestFlush() {
        return this.getBooleanAttribute('guest_flush');
    }
    get labels() {
        return this._labels;
    }
    // storage_locations - computed: true, optional: false, required: false
    get storageLocations() {
        return cdktf.Fn.tolist(this.getListAttribute('storage_locations'));
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference = DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesList = DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesList[_z] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesList", version: "14.13.1" };
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyToTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyToTerraform;
function dataGoogleComputeResourcePolicySnapshotSchedulePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeResourcePolicySnapshotSchedulePolicyToHclTerraform = dataGoogleComputeResourcePolicySnapshotSchedulePolicyToHclTerraform;
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // retention_policy - computed: true, optional: false, required: false
        this._retentionPolicy = new DataGoogleComputeResourcePolicySnapshotSchedulePolicyRetentionPolicyList(this, "retention_policy", false);
        // schedule - computed: true, optional: false, required: false
        this._schedule = new DataGoogleComputeResourcePolicySnapshotSchedulePolicyScheduleList(this, "schedule", false);
        // snapshot_properties - computed: true, optional: false, required: false
        this._snapshotProperties = new DataGoogleComputeResourcePolicySnapshotSchedulePolicySnapshotPropertiesList(this, "snapshot_properties", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get retentionPolicy() {
        return this._retentionPolicy;
    }
    get schedule() {
        return this._schedule;
    }
    get snapshotProperties() {
        return this._snapshotProperties;
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyOutputReference = DataGoogleComputeResourcePolicySnapshotSchedulePolicyOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyOutputReference", version: "14.13.1" };
class DataGoogleComputeResourcePolicySnapshotSchedulePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeResourcePolicySnapshotSchedulePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeResourcePolicySnapshotSchedulePolicyList = DataGoogleComputeResourcePolicySnapshotSchedulePolicyList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicySnapshotSchedulePolicyList[_1] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicySnapshotSchedulePolicyList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_resource_policy google_compute_resource_policy}
*/
class DataGoogleComputeResourcePolicy extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeResourcePolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeResourcePolicy to import
    * @param importFromId The id of the existing DataGoogleComputeResourcePolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_resource_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeResourcePolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_resource_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_resource_policy google_compute_resource_policy} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeResourcePolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_resource_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // disk_consistency_group_policy - computed: true, optional: false, required: false
        this._diskConsistencyGroupPolicy = new DataGoogleComputeResourcePolicyDiskConsistencyGroupPolicyList(this, "disk_consistency_group_policy", false);
        // group_placement_policy - computed: true, optional: false, required: false
        this._groupPlacementPolicy = new DataGoogleComputeResourcePolicyGroupPlacementPolicyList(this, "group_placement_policy", false);
        // instance_schedule_policy - computed: true, optional: false, required: false
        this._instanceSchedulePolicy = new DataGoogleComputeResourcePolicyInstanceSchedulePolicyList(this, "instance_schedule_policy", false);
        // snapshot_schedule_policy - computed: true, optional: false, required: false
        this._snapshotSchedulePolicy = new DataGoogleComputeResourcePolicySnapshotSchedulePolicyList(this, "snapshot_schedule_policy", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get diskConsistencyGroupPolicy() {
        return this._diskConsistencyGroupPolicy;
    }
    get groupPlacementPolicy() {
        return this._groupPlacementPolicy;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceSchedulePolicy() {
        return this._instanceSchedulePolicy;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get snapshotSchedulePolicy() {
        return this._snapshotSchedulePolicy;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeResourcePolicy = DataGoogleComputeResourcePolicy;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeResourcePolicy[_2] = { fqn: "@cdktf/provider-google.dataGoogleComputeResourcePolicy.DataGoogleComputeResourcePolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeResourcePolicy.tfResourceType = "google_compute_resource_policy";
