// @cdktf/provider-google/lib/data-google-kms-crypto-keys/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleKmsCryptoKeys = exports.DataGoogleKmsCryptoKeysKeysList = exports.DataGoogleKmsCryptoKeysKeysOutputReference = exports.dataGoogleKmsCryptoKeysKeysToHclTerraform = exports.dataGoogleKmsCryptoKeysKeysToTerraform = exports.DataGoogleKmsCryptoKeysKeysVersionTemplateList = exports.DataGoogleKmsCryptoKeysKeysVersionTemplateOutputReference = exports.dataGoogleKmsCryptoKeysKeysVersionTemplateToHclTerraform = exports.dataGoogleKmsCryptoKeysKeysVersionTemplateToTerraform = exports.DataGoogleKmsCryptoKeysKeysPrimaryList = exports.DataGoogleKmsCryptoKeysKeysPrimaryOutputReference = exports.dataGoogleKmsCryptoKeysKeysPrimaryToHclTerraform = exports.dataGoogleKmsCryptoKeysKeysPrimaryToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleKmsCryptoKeysKeysPrimaryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleKmsCryptoKeysKeysPrimaryToTerraform = dataGoogleKmsCryptoKeysKeysPrimaryToTerraform;
function dataGoogleKmsCryptoKeysKeysPrimaryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleKmsCryptoKeysKeysPrimaryToHclTerraform = dataGoogleKmsCryptoKeysKeysPrimaryToHclTerraform;
class DataGoogleKmsCryptoKeysKeysPrimaryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.DataGoogleKmsCryptoKeysKeysPrimaryOutputReference = DataGoogleKmsCryptoKeysKeysPrimaryOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleKmsCryptoKeysKeysPrimaryOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleKmsCryptoKeys.DataGoogleKmsCryptoKeysKeysPrimaryOutputReference", version: "14.13.1" };
class DataGoogleKmsCryptoKeysKeysPrimaryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleKmsCryptoKeysKeysPrimaryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleKmsCryptoKeysKeysPrimaryList = DataGoogleKmsCryptoKeysKeysPrimaryList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleKmsCryptoKeysKeysPrimaryList[_b] = { fqn: "@cdktf/provider-google.dataGoogleKmsCryptoKeys.DataGoogleKmsCryptoKeysKeysPrimaryList", version: "14.13.1" };
function dataGoogleKmsCryptoKeysKeysVersionTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleKmsCryptoKeysKeysVersionTemplateToTerraform = dataGoogleKmsCryptoKeysKeysVersionTemplateToTerraform;
function dataGoogleKmsCryptoKeysKeysVersionTemplateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleKmsCryptoKeysKeysVersionTemplateToHclTerraform = dataGoogleKmsCryptoKeysKeysVersionTemplateToHclTerraform;
class DataGoogleKmsCryptoKeysKeysVersionTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // algorithm - computed: true, optional: false, required: false
    get algorithm() {
        return this.getStringAttribute('algorithm');
    }
    // protection_level - computed: true, optional: false, required: false
    get protectionLevel() {
        return this.getStringAttribute('protection_level');
    }
}
exports.DataGoogleKmsCryptoKeysKeysVersionTemplateOutputReference = DataGoogleKmsCryptoKeysKeysVersionTemplateOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleKmsCryptoKeysKeysVersionTemplateOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleKmsCryptoKeys.DataGoogleKmsCryptoKeysKeysVersionTemplateOutputReference", version: "14.13.1" };
class DataGoogleKmsCryptoKeysKeysVersionTemplateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleKmsCryptoKeysKeysVersionTemplateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleKmsCryptoKeysKeysVersionTemplateList = DataGoogleKmsCryptoKeysKeysVersionTemplateList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleKmsCryptoKeysKeysVersionTemplateList[_d] = { fqn: "@cdktf/provider-google.dataGoogleKmsCryptoKeys.DataGoogleKmsCryptoKeysKeysVersionTemplateList", version: "14.13.1" };
function dataGoogleKmsCryptoKeysKeysToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleKmsCryptoKeysKeysToTerraform = dataGoogleKmsCryptoKeysKeysToTerraform;
function dataGoogleKmsCryptoKeysKeysToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleKmsCryptoKeysKeysToHclTerraform = dataGoogleKmsCryptoKeysKeysToHclTerraform;
class DataGoogleKmsCryptoKeysKeysOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // primary - computed: true, optional: false, required: false
        this._primary = new DataGoogleKmsCryptoKeysKeysPrimaryList(this, "primary", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // version_template - computed: true, optional: false, required: false
        this._versionTemplate = new DataGoogleKmsCryptoKeysKeysVersionTemplateList(this, "version_template", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // crypto_key_backend - computed: true, optional: false, required: false
    get cryptoKeyBackend() {
        return this.getStringAttribute('crypto_key_backend');
    }
    // destroy_scheduled_duration - computed: true, optional: false, required: false
    get destroyScheduledDuration() {
        return this.getStringAttribute('destroy_scheduled_duration');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // import_only - computed: true, optional: false, required: false
    get importOnly() {
        return this.getBooleanAttribute('import_only');
    }
    // key_ring - computed: true, optional: false, required: false
    get keyRing() {
        return this.getStringAttribute('key_ring');
    }
    get labels() {
        return this._labels;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get primary() {
        return this._primary;
    }
    // purpose - computed: true, optional: false, required: false
    get purpose() {
        return this.getStringAttribute('purpose');
    }
    // rotation_period - computed: true, optional: false, required: false
    get rotationPeriod() {
        return this.getStringAttribute('rotation_period');
    }
    // skip_initial_version_creation - computed: true, optional: false, required: false
    get skipInitialVersionCreation() {
        return this.getBooleanAttribute('skip_initial_version_creation');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    get versionTemplate() {
        return this._versionTemplate;
    }
}
exports.DataGoogleKmsCryptoKeysKeysOutputReference = DataGoogleKmsCryptoKeysKeysOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleKmsCryptoKeysKeysOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleKmsCryptoKeys.DataGoogleKmsCryptoKeysKeysOutputReference", version: "14.13.1" };
class DataGoogleKmsCryptoKeysKeysList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleKmsCryptoKeysKeysOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleKmsCryptoKeysKeysList = DataGoogleKmsCryptoKeysKeysList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleKmsCryptoKeysKeysList[_f] = { fqn: "@cdktf/provider-google.dataGoogleKmsCryptoKeys.DataGoogleKmsCryptoKeysKeysList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/kms_crypto_keys google_kms_crypto_keys}
*/
class DataGoogleKmsCryptoKeys extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleKmsCryptoKeys resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleKmsCryptoKeys to import
    * @param importFromId The id of the existing DataGoogleKmsCryptoKeys that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/kms_crypto_keys#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleKmsCryptoKeys to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_kms_crypto_keys", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/kms_crypto_keys google_kms_crypto_keys} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleKmsCryptoKeysConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_kms_crypto_keys',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // keys - computed: true, optional: false, required: false
        this._keys = new DataGoogleKmsCryptoKeysKeysList(this, "keys", false);
        this._filter = config.filter;
        this._id = config.id;
        this._keyRing = config.keyRing;
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keyRing() {
        return this.getStringAttribute('key_ring');
    }
    set keyRing(value) {
        this._keyRing = value;
    }
    // Temporarily expose input value. Use with caution.
    get keyRingInput() {
        return this._keyRing;
    }
    get keys() {
        return this._keys;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filter: cdktf.stringToTerraform(this._filter),
            id: cdktf.stringToTerraform(this._id),
            key_ring: cdktf.stringToTerraform(this._keyRing),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            filter: {
                value: cdktf.stringToHclTerraform(this._filter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            key_ring: {
                value: cdktf.stringToHclTerraform(this._keyRing),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleKmsCryptoKeys = DataGoogleKmsCryptoKeys;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleKmsCryptoKeys[_g] = { fqn: "@cdktf/provider-google.dataGoogleKmsCryptoKeys.DataGoogleKmsCryptoKeys", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleKmsCryptoKeys.tfResourceType = "google_kms_crypto_keys";
