// @cdktf/provider-google/lib/data-google-compute-router-nat/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeRouterNat = exports.DataGoogleComputeRouterNatSubnetworkList = exports.DataGoogleComputeRouterNatSubnetworkOutputReference = exports.dataGoogleComputeRouterNatSubnetworkToHclTerraform = exports.dataGoogleComputeRouterNatSubnetworkToTerraform = exports.DataGoogleComputeRouterNatRulesList = exports.DataGoogleComputeRouterNatRulesOutputReference = exports.dataGoogleComputeRouterNatRulesToHclTerraform = exports.dataGoogleComputeRouterNatRulesToTerraform = exports.DataGoogleComputeRouterNatRulesActionList = exports.DataGoogleComputeRouterNatRulesActionOutputReference = exports.dataGoogleComputeRouterNatRulesActionToHclTerraform = exports.dataGoogleComputeRouterNatRulesActionToTerraform = exports.DataGoogleComputeRouterNatLogConfigList = exports.DataGoogleComputeRouterNatLogConfigOutputReference = exports.dataGoogleComputeRouterNatLogConfigToHclTerraform = exports.dataGoogleComputeRouterNatLogConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeRouterNatLogConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRouterNatLogConfigToTerraform = dataGoogleComputeRouterNatLogConfigToTerraform;
function dataGoogleComputeRouterNatLogConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRouterNatLogConfigToHclTerraform = dataGoogleComputeRouterNatLogConfigToHclTerraform;
class DataGoogleComputeRouterNatLogConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable - computed: true, optional: false, required: false
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    // filter - computed: true, optional: false, required: false
    get filter() {
        return this.getStringAttribute('filter');
    }
}
exports.DataGoogleComputeRouterNatLogConfigOutputReference = DataGoogleComputeRouterNatLogConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNatLogConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNatLogConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeRouterNatLogConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRouterNatLogConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRouterNatLogConfigList = DataGoogleComputeRouterNatLogConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNatLogConfigList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNatLogConfigList", version: "14.13.1" };
function dataGoogleComputeRouterNatRulesActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRouterNatRulesActionToTerraform = dataGoogleComputeRouterNatRulesActionToTerraform;
function dataGoogleComputeRouterNatRulesActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRouterNatRulesActionToHclTerraform = dataGoogleComputeRouterNatRulesActionToHclTerraform;
class DataGoogleComputeRouterNatRulesActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // source_nat_active_ips - computed: true, optional: false, required: false
    get sourceNatActiveIps() {
        return cdktf.Fn.tolist(this.getListAttribute('source_nat_active_ips'));
    }
    // source_nat_drain_ips - computed: true, optional: false, required: false
    get sourceNatDrainIps() {
        return cdktf.Fn.tolist(this.getListAttribute('source_nat_drain_ips'));
    }
}
exports.DataGoogleComputeRouterNatRulesActionOutputReference = DataGoogleComputeRouterNatRulesActionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNatRulesActionOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNatRulesActionOutputReference", version: "14.13.1" };
class DataGoogleComputeRouterNatRulesActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRouterNatRulesActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRouterNatRulesActionList = DataGoogleComputeRouterNatRulesActionList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNatRulesActionList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNatRulesActionList", version: "14.13.1" };
function dataGoogleComputeRouterNatRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRouterNatRulesToTerraform = dataGoogleComputeRouterNatRulesToTerraform;
function dataGoogleComputeRouterNatRulesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRouterNatRulesToHclTerraform = dataGoogleComputeRouterNatRulesToHclTerraform;
class DataGoogleComputeRouterNatRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // action - computed: true, optional: false, required: false
        this._action = new DataGoogleComputeRouterNatRulesActionList(this, "action", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get action() {
        return this._action;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // match - computed: true, optional: false, required: false
    get match() {
        return this.getStringAttribute('match');
    }
    // rule_number - computed: true, optional: false, required: false
    get ruleNumber() {
        return this.getNumberAttribute('rule_number');
    }
}
exports.DataGoogleComputeRouterNatRulesOutputReference = DataGoogleComputeRouterNatRulesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNatRulesOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNatRulesOutputReference", version: "14.13.1" };
class DataGoogleComputeRouterNatRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRouterNatRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRouterNatRulesList = DataGoogleComputeRouterNatRulesList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNatRulesList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNatRulesList", version: "14.13.1" };
function dataGoogleComputeRouterNatSubnetworkToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRouterNatSubnetworkToTerraform = dataGoogleComputeRouterNatSubnetworkToTerraform;
function dataGoogleComputeRouterNatSubnetworkToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRouterNatSubnetworkToHclTerraform = dataGoogleComputeRouterNatSubnetworkToHclTerraform;
class DataGoogleComputeRouterNatSubnetworkOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // secondary_ip_range_names - computed: true, optional: false, required: false
    get secondaryIpRangeNames() {
        return cdktf.Fn.tolist(this.getListAttribute('secondary_ip_range_names'));
    }
    // source_ip_ranges_to_nat - computed: true, optional: false, required: false
    get sourceIpRangesToNat() {
        return cdktf.Fn.tolist(this.getListAttribute('source_ip_ranges_to_nat'));
    }
}
exports.DataGoogleComputeRouterNatSubnetworkOutputReference = DataGoogleComputeRouterNatSubnetworkOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNatSubnetworkOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNatSubnetworkOutputReference", version: "14.13.1" };
class DataGoogleComputeRouterNatSubnetworkList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRouterNatSubnetworkOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRouterNatSubnetworkList = DataGoogleComputeRouterNatSubnetworkList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNatSubnetworkList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNatSubnetworkList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_router_nat google_compute_router_nat}
*/
class DataGoogleComputeRouterNat extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeRouterNat resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeRouterNat to import
    * @param importFromId The id of the existing DataGoogleComputeRouterNat that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_router_nat#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeRouterNat to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_router_nat", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_router_nat google_compute_router_nat} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeRouterNatConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_router_nat',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // log_config - computed: true, optional: false, required: false
        this._logConfig = new DataGoogleComputeRouterNatLogConfigList(this, "log_config", false);
        // rules - computed: true, optional: false, required: false
        this._rules = new DataGoogleComputeRouterNatRulesList(this, "rules", true);
        // subnetwork - computed: true, optional: false, required: false
        this._subnetwork = new DataGoogleComputeRouterNatSubnetworkList(this, "subnetwork", true);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
        this._router = config.router;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // auto_network_tier - computed: true, optional: false, required: false
    get autoNetworkTier() {
        return this.getStringAttribute('auto_network_tier');
    }
    // drain_nat_ips - computed: true, optional: false, required: false
    get drainNatIps() {
        return cdktf.Fn.tolist(this.getListAttribute('drain_nat_ips'));
    }
    // enable_dynamic_port_allocation - computed: true, optional: false, required: false
    get enableDynamicPortAllocation() {
        return this.getBooleanAttribute('enable_dynamic_port_allocation');
    }
    // enable_endpoint_independent_mapping - computed: true, optional: false, required: false
    get enableEndpointIndependentMapping() {
        return this.getBooleanAttribute('enable_endpoint_independent_mapping');
    }
    // endpoint_types - computed: true, optional: false, required: false
    get endpointTypes() {
        return this.getListAttribute('endpoint_types');
    }
    // icmp_idle_timeout_sec - computed: true, optional: false, required: false
    get icmpIdleTimeoutSec() {
        return this.getNumberAttribute('icmp_idle_timeout_sec');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // initial_nat_ips - computed: true, optional: false, required: false
    get initialNatIps() {
        return cdktf.Fn.tolist(this.getListAttribute('initial_nat_ips'));
    }
    get logConfig() {
        return this._logConfig;
    }
    // max_ports_per_vm - computed: true, optional: false, required: false
    get maxPortsPerVm() {
        return this.getNumberAttribute('max_ports_per_vm');
    }
    // min_ports_per_vm - computed: true, optional: false, required: false
    get minPortsPerVm() {
        return this.getNumberAttribute('min_ports_per_vm');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // nat_ip_allocate_option - computed: true, optional: false, required: false
    get natIpAllocateOption() {
        return this.getStringAttribute('nat_ip_allocate_option');
    }
    // nat_ips - computed: true, optional: false, required: false
    get natIps() {
        return cdktf.Fn.tolist(this.getListAttribute('nat_ips'));
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get router() {
        return this.getStringAttribute('router');
    }
    set router(value) {
        this._router = value;
    }
    // Temporarily expose input value. Use with caution.
    get routerInput() {
        return this._router;
    }
    get rules() {
        return this._rules;
    }
    // source_subnetwork_ip_ranges_to_nat - computed: true, optional: false, required: false
    get sourceSubnetworkIpRangesToNat() {
        return this.getStringAttribute('source_subnetwork_ip_ranges_to_nat');
    }
    get subnetwork() {
        return this._subnetwork;
    }
    // tcp_established_idle_timeout_sec - computed: true, optional: false, required: false
    get tcpEstablishedIdleTimeoutSec() {
        return this.getNumberAttribute('tcp_established_idle_timeout_sec');
    }
    // tcp_time_wait_timeout_sec - computed: true, optional: false, required: false
    get tcpTimeWaitTimeoutSec() {
        return this.getNumberAttribute('tcp_time_wait_timeout_sec');
    }
    // tcp_transitory_idle_timeout_sec - computed: true, optional: false, required: false
    get tcpTransitoryIdleTimeoutSec() {
        return this.getNumberAttribute('tcp_transitory_idle_timeout_sec');
    }
    // udp_idle_timeout_sec - computed: true, optional: false, required: false
    get udpIdleTimeoutSec() {
        return this.getNumberAttribute('udp_idle_timeout_sec');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            router: cdktf.stringToTerraform(this._router),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            router: {
                value: cdktf.stringToHclTerraform(this._router),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeRouterNat = DataGoogleComputeRouterNat;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterNat[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterNat.DataGoogleComputeRouterNat", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeRouterNat.tfResourceType = "google_compute_router_nat";
