// @cdktf/provider-google/lib/data-google-cloud-asset-search-all-resources/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleCloudAssetSearchAllResources = exports.DataGoogleCloudAssetSearchAllResourcesResultsList = exports.DataGoogleCloudAssetSearchAllResourcesResultsOutputReference = exports.dataGoogleCloudAssetSearchAllResourcesResultsToHclTerraform = exports.dataGoogleCloudAssetSearchAllResourcesResultsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleCloudAssetSearchAllResourcesResultsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudAssetSearchAllResourcesResultsToTerraform = dataGoogleCloudAssetSearchAllResourcesResultsToTerraform;
function dataGoogleCloudAssetSearchAllResourcesResultsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudAssetSearchAllResourcesResultsToHclTerraform = dataGoogleCloudAssetSearchAllResourcesResultsToHclTerraform;
class DataGoogleCloudAssetSearchAllResourcesResultsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // asset_type - computed: true, optional: false, required: false
    get assetType() {
        return this.getStringAttribute('asset_type');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    // folders - computed: true, optional: false, required: false
    get folders() {
        return this.getListAttribute('folders');
    }
    // kms_keys - computed: true, optional: false, required: false
    get kmsKeys() {
        return this.getListAttribute('kms_keys');
    }
    get labels() {
        return this._labels;
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network_tags - computed: true, optional: false, required: false
    get networkTags() {
        return this.getListAttribute('network_tags');
    }
    // organization - computed: true, optional: false, required: false
    get organization() {
        return this.getStringAttribute('organization');
    }
    // parent_asset_type - computed: true, optional: false, required: false
    get parentAssetType() {
        return this.getStringAttribute('parent_asset_type');
    }
    // parent_full_resource_name - computed: true, optional: false, required: false
    get parentFullResourceName() {
        return this.getStringAttribute('parent_full_resource_name');
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
}
exports.DataGoogleCloudAssetSearchAllResourcesResultsOutputReference = DataGoogleCloudAssetSearchAllResourcesResultsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleCloudAssetSearchAllResourcesResultsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleCloudAssetSearchAllResources.DataGoogleCloudAssetSearchAllResourcesResultsOutputReference", version: "14.13.1" };
class DataGoogleCloudAssetSearchAllResourcesResultsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudAssetSearchAllResourcesResultsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudAssetSearchAllResourcesResultsList = DataGoogleCloudAssetSearchAllResourcesResultsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleCloudAssetSearchAllResourcesResultsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleCloudAssetSearchAllResources.DataGoogleCloudAssetSearchAllResourcesResultsList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_asset_search_all_resources google_cloud_asset_search_all_resources}
*/
class DataGoogleCloudAssetSearchAllResources extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleCloudAssetSearchAllResources resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleCloudAssetSearchAllResources to import
    * @param importFromId The id of the existing DataGoogleCloudAssetSearchAllResources that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_asset_search_all_resources#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleCloudAssetSearchAllResources to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_cloud_asset_search_all_resources", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloud_asset_search_all_resources google_cloud_asset_search_all_resources} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleCloudAssetSearchAllResourcesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloud_asset_search_all_resources',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // results - computed: true, optional: false, required: false
        this._results = new DataGoogleCloudAssetSearchAllResourcesResultsList(this, "results", false);
        this._assetTypes = config.assetTypes;
        this._id = config.id;
        this._query = config.query;
        this._scope = config.scope;
    }
    get assetTypes() {
        return this.getListAttribute('asset_types');
    }
    set assetTypes(value) {
        this._assetTypes = value;
    }
    resetAssetTypes() {
        this._assetTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assetTypesInput() {
        return this._assetTypes;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get query() {
        return this.getStringAttribute('query');
    }
    set query(value) {
        this._query = value;
    }
    resetQuery() {
        this._query = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queryInput() {
        return this._query;
    }
    get results() {
        return this._results;
    }
    get scope() {
        return this.getStringAttribute('scope');
    }
    set scope(value) {
        this._scope = value;
    }
    // Temporarily expose input value. Use with caution.
    get scopeInput() {
        return this._scope;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            asset_types: cdktf.listMapper(cdktf.stringToTerraform, false)(this._assetTypes),
            id: cdktf.stringToTerraform(this._id),
            query: cdktf.stringToTerraform(this._query),
            scope: cdktf.stringToTerraform(this._scope),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            asset_types: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._assetTypes),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            query: {
                value: cdktf.stringToHclTerraform(this._query),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            scope: {
                value: cdktf.stringToHclTerraform(this._scope),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleCloudAssetSearchAllResources = DataGoogleCloudAssetSearchAllResources;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleCloudAssetSearchAllResources[_c] = { fqn: "@cdktf/provider-google.dataGoogleCloudAssetSearchAllResources.DataGoogleCloudAssetSearchAllResources", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleCloudAssetSearchAllResources.tfResourceType = "google_cloud_asset_search_all_resources";
