// @cdktf/cli-core/src/lib/cdktf-project-io-handler.js
"use strict";
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CdktfProjectIOHandler = void 0;
const cdktf_project_1 = require("./cdktf-project");
/**
 * This class is used to buffer events while the user is being asked for input. It is solely consumed by the CdktfProject class,
 * and is mostly created as a separation of concerns to keep the CdktfProject class as clean as possible.
 */
class CdktfProjectIOHandler {
    constructor() {
        // Pauses all progress / status events from being forwarded to the user
        // If set from true to false, the events will be sent through the channels they came in
        // (until a waiting for approval event is sent)
        this.waitingForUserInput = false;
        this.eventBuffer = [];
    }
    awaitUserInput() {
        this.waitingForUserInput = true;
    }
    filterUserInputEventsFromBuffer() {
        this.eventBuffer = this.eventBuffer.filter((event) => event.type === "projectUpdate"
            ? !(0, cdktf_project_1.isWaitingForUserInputUpdate)(event.value) // we want to filter out the waiting for approval events
            : true // we want all other types
        );
    }
    resumeAfterUserInput(stackName) {
        // remove waiting for approval event that should be resumed
        this.eventBuffer = this.eventBuffer.filter((event) => !(event.type === "projectUpdate" &&
            (0, cdktf_project_1.isWaitingForUserInputUpdate)(event.value) &&
            event.value.stackName === stackName));
        if (this.eventBuffer.length &&
            this.eventBuffer[0].type === "projectUpdate" &&
            (0, cdktf_project_1.isWaitingForUserInputUpdate)(this.eventBuffer[0].value)) {
            // we are still waiting on approval for the current stack
            // we removed a future "waiting for approval" event for some other stack
            return;
        }
        // We first need to flush all events, we can not resume if there is a new waiting for approval update
        let event = this.eventBuffer.shift();
        while (event) {
            if (event.type === "projectUpdate") {
                event.cb(event.value);
                if ((0, cdktf_project_1.isWaitingForUserInputUpdate)(event.value)) {
                    // We have to wait for user input again, therefore we can not resume
                    return;
                }
            }
            if (event.type === "logMessage") {
                event.cb(event.value);
            }
            event = this.eventBuffer.shift();
        }
        // If we reach this point there was no waiting for user input event, so we can safely resume
        this.waitingForUserInput = false;
    }
    isWaitingForUserInput() {
        return this.waitingForUserInput;
    }
    pushEvent(event) {
        this.eventBuffer.push(event);
    }
    bufferWhileAwaitingUserInput(cb) {
        if (!cb) {
            return undefined;
        }
        return (msg) => {
            if (this.waitingForUserInput) {
                this.eventBuffer.push({
                    cb,
                    value: msg,
                    type: "logMessage",
                });
            }
            else {
                cb(msg);
            }
        };
    }
}
exports.CdktfProjectIOHandler = CdktfProjectIOHandler;
