// @cdktf/cli-core/src/lib/local-provider-constraints.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalProviderConstraints = void 0;
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
const commons_1 = require("@cdktf/commons");
const cdktf_config_1 = require("./cdktf-config");
const z = __importStar(require("zod"));
const constraintsType = z.object({
    cdktf: z.string(),
    providers: z.record(z.string()),
});
// TODO: move this to @cdktf/commons
// tracked here https://github.com/hashicorp/terraform-cdk/issues/1814
/**
 * Class to help with reading `constraints.json` file
 * published by `provider get` or `provider add` commands
 */
class LocalProviderConstraints {
    get jsonFilePath() {
        const config = cdktf_config_1.CdktfConfig.read();
        return path_1.default.resolve(config.codeMakerOutput, "constraints.json");
    }
    async readLocalConstraintsJson() {
        let constraintsJson;
        try {
            constraintsJson = await fs_extra_1.default.readFile(this.jsonFilePath, "utf8");
        }
        catch (e) {
            // If we cannot read the constraints.json file, it may not exist
            commons_1.logger.debug("constraints.json file reading error: ", e);
            this.constraints = undefined;
            return;
        }
        let providerConstraints;
        try {
            providerConstraints = constraintsType.parse(JSON.parse(constraintsJson));
            providerConstraints.providers = Object.fromEntries(Object.entries(providerConstraints.providers).map(([name, constraint]) => {
                return [name.toLowerCase(), constraint];
            }));
        }
        catch (e) {
            throw commons_1.Errors.External("constraints.json file is malformed. The root must be a JSON object.", e);
        }
        this.constraints = providerConstraints;
    }
    async constraintsForAllProviders() {
        if (this.constraints) {
            return this.constraints;
        }
        await this.readLocalConstraintsJson();
        return this.constraints;
    }
    async constraintForProvider(providerName) {
        const constraints = await this.constraintsForAllProviders();
        if (!constraints) {
            return null;
        }
        const providerConstraint = constraints.providers[providerName.toLowerCase()];
        if (!providerName) {
            return null;
        }
        return providerConstraint;
    }
}
exports.LocalProviderConstraints = LocalProviderConstraints;
