// @cdktf/cli-core/src/lib/terraform-json.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.terraformJsonSchema = void 0;
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const z = __importStar(require("zod"));
const remote = z
    .object({
    organization: z.string(),
    hostname: z.string().optional(),
    token: z.string().optional(),
    workspaces: z.object({
        name: z.string().optional(),
        prefix: z.string().optional(),
    }),
})
    .deepPartial();
exports.terraformJsonSchema = z
    .object({
    "//": z
        .object({
        metadata: z
            .object({
            version: z.string(),
            stackName: z.string(),
            backend: z.string(),
        })
            .nonstrict(),
        outputs: z.record(z.any()),
    })
        .nonstrict(),
    terraform: z.object({
        backend: z
            .object({
            // All other backends are here as well, but we don't read them right now
            remote,
        })
            .nonstrict(),
        cloud: z
            .object({
            organization: z.string(),
            hostname: z.string().optional(),
            token: z.string().optional(),
            workspaces: z.union([
                z.object({ name: z.string() }),
                z.object({ tags: z.array(z.string()) }),
            ]),
        })
            .nonstrict(),
        required_providers: z.record(z.object({ source: z.string(), version: z.string() }).nonstrict()),
        required_version: z.string(),
    }),
    data: z.record(z.any()),
    provider: z.record(z.any()),
    resource: z.record(z.any()),
})
    .deepPartial()
    .nonstrict();
