// @cdktf/cli-core/src/lib/terraform-provider-lock.js
"use strict";
/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformProviderLock = void 0;
const path_1 = __importDefault(require("path"));
const promises_1 = __importDefault(require("fs/promises"));
const dependency_manager_1 = require("./dependencies/dependency-manager");
const commons_1 = require("@cdktf/commons");
const TerraformLockFileName = ".terraform.lock.hcl";
class TerraformProviderLock {
    constructor(stackWorkingDirectory) {
        this.stackWorkingDirectory = stackWorkingDirectory;
        this._providerLockData = null;
    }
    get lockFilePath() {
        return path_1.default.join(this.stackWorkingDirectory, TerraformLockFileName);
    }
    async hasProviderLockFile() {
        try {
            await promises_1.default.stat(this.lockFilePath);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    async readProviderLockFile() {
        try {
            const lockFile = (await promises_1.default.readFile(this.lockFilePath)).toString();
            return lockFile;
        }
        catch (e) {
            commons_1.logger.debug("Unable to read provider lock file", e);
        }
        return "";
    }
    parseProviderLockFile(contents) {
        const providerLockData = {
            providers: {},
        };
        let currentProvider;
        contents.split(/\r\n|\r|\n/).forEach((line) => {
            if (currentProvider) {
                const constraintMatch = line.match(/^\s*constraints\s+=\s+"(.*)"/);
                if (constraintMatch) {
                    providerLockData.providers[currentProvider].constraints =
                        new dependency_manager_1.ProviderConstraint(currentProvider, constraintMatch[1]);
                    return;
                }
                const versionMatch = line.match(/^\s*version\s+=\s+"(.*)"/);
                if (versionMatch) {
                    providerLockData.providers[currentProvider].version = versionMatch[1];
                    return;
                }
                const endMatch = line.match(/^\s*}/);
                if (endMatch) {
                    currentProvider = undefined;
                }
                return;
            }
            const providerMatch = line.match(/provider "(.*)"/);
            if (providerMatch) {
                currentProvider = providerMatch[1];
                providerLockData.providers[currentProvider] = {
                    name: new dependency_manager_1.ProviderConstraint(currentProvider, ">=0").simplifiedName,
                };
            }
        });
        return providerLockData;
    }
    async providers(forceReread = false) {
        if (!this._providerLockData || forceReread) {
            const contents = await this.readProviderLockFile();
            this._providerLockData = this.parseProviderLockFile(contents);
        }
        return this._providerLockData.providers;
    }
    async hasMatchingProvider(constraint) {
        var _a, _b;
        const providerLockData = await this.providers();
        const lockedProvider = providerLockData[constraint.source];
        if (lockedProvider) {
            return (_a = lockedProvider.constraints) === null || _a === void 0 ? void 0 : _a.matchesVersion((_b = constraint.version) !== null && _b !== void 0 ? _b : ">0");
        }
        return false;
    }
}
exports.TerraformProviderLock = TerraformProviderLock;
