// @cdktf/provider-google/lib/network-security-authz-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11;
Object.defineProperty(exports, "__esModule", { value: true });
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesList = exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesList = exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsList = exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesList = exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesList = exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsList = exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsOutputReference = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToTerraform = exports.NetworkSecurityAuthzPolicyCustomProviderOutputReference = exports.networkSecurityAuthzPolicyCustomProviderToHclTerraform = exports.networkSecurityAuthzPolicyCustomProviderToTerraform = exports.NetworkSecurityAuthzPolicyCustomProviderCloudIapOutputReference = exports.networkSecurityAuthzPolicyCustomProviderCloudIapToHclTerraform = exports.networkSecurityAuthzPolicyCustomProviderCloudIapToTerraform = exports.NetworkSecurityAuthzPolicyCustomProviderAuthzExtensionOutputReference = exports.networkSecurityAuthzPolicyCustomProviderAuthzExtensionToHclTerraform = exports.networkSecurityAuthzPolicyCustomProviderAuthzExtensionToTerraform = void 0;
exports.NetworkSecurityAuthzPolicy = exports.NetworkSecurityAuthzPolicyTimeoutsOutputReference = exports.networkSecurityAuthzPolicyTimeoutsToHclTerraform = exports.networkSecurityAuthzPolicyTimeoutsToTerraform = exports.NetworkSecurityAuthzPolicyTargetOutputReference = exports.networkSecurityAuthzPolicyTargetToHclTerraform = exports.networkSecurityAuthzPolicyTargetToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesList = exports.NetworkSecurityAuthzPolicyHttpRulesOutputReference = exports.networkSecurityAuthzPolicyHttpRulesToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesToOutputReference = exports.networkSecurityAuthzPolicyHttpRulesToToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesToToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsList = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsOutputReference = exports.networkSecurityAuthzPolicyHttpRulesToOperationsToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesToOperationsToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsList = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsOutputReference = exports.networkSecurityAuthzPolicyHttpRulesToOperationsPathsToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesToOperationsPathsToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsList = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsOutputReference = exports.networkSecurityAuthzPolicyHttpRulesToOperationsHostsToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesToOperationsHostsToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetOutputReference = exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersList = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersOutputReference = exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToHclTerraform = exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToTerraform = exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkSecurityAuthzPolicyCustomProviderAuthzExtensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resources: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.resources),
    };
}
exports.networkSecurityAuthzPolicyCustomProviderAuthzExtensionToTerraform = networkSecurityAuthzPolicyCustomProviderAuthzExtensionToTerraform;
function networkSecurityAuthzPolicyCustomProviderAuthzExtensionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        resources: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.resources),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyCustomProviderAuthzExtensionToHclTerraform = networkSecurityAuthzPolicyCustomProviderAuthzExtensionToHclTerraform;
class NetworkSecurityAuthzPolicyCustomProviderAuthzExtensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resources !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resources = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resources = value.resources;
        }
    }
    get resources() {
        return this.getListAttribute('resources');
    }
    set resources(value) {
        this._resources = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources;
    }
}
exports.NetworkSecurityAuthzPolicyCustomProviderAuthzExtensionOutputReference = NetworkSecurityAuthzPolicyCustomProviderAuthzExtensionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyCustomProviderAuthzExtensionOutputReference[_a] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyCustomProviderAuthzExtensionOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyCustomProviderCloudIapToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
    };
}
exports.networkSecurityAuthzPolicyCustomProviderCloudIapToTerraform = networkSecurityAuthzPolicyCustomProviderCloudIapToTerraform;
function networkSecurityAuthzPolicyCustomProviderCloudIapToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyCustomProviderCloudIapToHclTerraform = networkSecurityAuthzPolicyCustomProviderCloudIapToHclTerraform;
class NetworkSecurityAuthzPolicyCustomProviderCloudIapOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
}
exports.NetworkSecurityAuthzPolicyCustomProviderCloudIapOutputReference = NetworkSecurityAuthzPolicyCustomProviderCloudIapOutputReference;
_b = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyCustomProviderCloudIapOutputReference[_b] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyCustomProviderCloudIapOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyCustomProviderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        authz_extension: networkSecurityAuthzPolicyCustomProviderAuthzExtensionToTerraform(struct.authzExtension),
        cloud_iap: networkSecurityAuthzPolicyCustomProviderCloudIapToTerraform(struct.cloudIap),
    };
}
exports.networkSecurityAuthzPolicyCustomProviderToTerraform = networkSecurityAuthzPolicyCustomProviderToTerraform;
function networkSecurityAuthzPolicyCustomProviderToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        authz_extension: {
            value: networkSecurityAuthzPolicyCustomProviderAuthzExtensionToHclTerraform(struct.authzExtension),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyCustomProviderAuthzExtensionList",
        },
        cloud_iap: {
            value: networkSecurityAuthzPolicyCustomProviderCloudIapToHclTerraform(struct.cloudIap),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyCustomProviderCloudIapList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyCustomProviderToHclTerraform = networkSecurityAuthzPolicyCustomProviderToHclTerraform;
class NetworkSecurityAuthzPolicyCustomProviderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // authz_extension - computed: false, optional: true, required: false
        this._authzExtension = new NetworkSecurityAuthzPolicyCustomProviderAuthzExtensionOutputReference(this, "authz_extension");
        // cloud_iap - computed: false, optional: true, required: false
        this._cloudIap = new NetworkSecurityAuthzPolicyCustomProviderCloudIapOutputReference(this, "cloud_iap");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authzExtension?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.authzExtension = this._authzExtension?.internalValue;
        }
        if (this._cloudIap?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cloudIap = this._cloudIap?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._authzExtension.internalValue = undefined;
            this._cloudIap.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._authzExtension.internalValue = value.authzExtension;
            this._cloudIap.internalValue = value.cloudIap;
        }
    }
    get authzExtension() {
        return this._authzExtension;
    }
    putAuthzExtension(value) {
        this._authzExtension.internalValue = value;
    }
    resetAuthzExtension() {
        this._authzExtension.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authzExtensionInput() {
        return this._authzExtension.internalValue;
    }
    get cloudIap() {
        return this._cloudIap;
    }
    putCloudIap(value) {
        this._cloudIap.internalValue = value;
    }
    resetCloudIap() {
        this._cloudIap.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudIapInput() {
        return this._cloudIap.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyCustomProviderOutputReference = NetworkSecurityAuthzPolicyCustomProviderOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyCustomProviderOutputReference[_c] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyCustomProviderOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contains: cdktf.stringToTerraform(struct.contains),
        exact: cdktf.stringToTerraform(struct.exact),
        ignore_case: cdktf.booleanToTerraform(struct.ignoreCase),
        prefix: cdktf.stringToTerraform(struct.prefix),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        contains: {
            value: cdktf.stringToHclTerraform(struct.contains),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exact: {
            value: cdktf.stringToHclTerraform(struct.exact),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ignore_case: {
            value: cdktf.booleanToHclTerraform(struct.ignoreCase),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        suffix: {
            value: cdktf.stringToHclTerraform(struct.suffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contains !== undefined) {
            hasAnyValues = true;
            internalValueResult.contains = this._contains;
        }
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._ignoreCase !== undefined) {
            hasAnyValues = true;
            internalValueResult.ignoreCase = this._ignoreCase;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._contains = undefined;
            this._exact = undefined;
            this._ignoreCase = undefined;
            this._prefix = undefined;
            this._suffix = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._contains = value.contains;
            this._exact = value.exact;
            this._ignoreCase = value.ignoreCase;
            this._prefix = value.prefix;
            this._suffix = value.suffix;
        }
    }
    get contains() {
        return this.getStringAttribute('contains');
    }
    set contains(value) {
        this._contains = value;
    }
    resetContains() {
        this._contains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containsInput() {
        return this._contains;
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get ignoreCase() {
        return this.getBooleanAttribute('ignore_case');
    }
    set ignoreCase(value) {
        this._ignoreCase = value;
    }
    resetIgnoreCase() {
        this._ignoreCase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreCaseInput() {
        return this._ignoreCase;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsOutputReference[_d] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsList = NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsList;
_e = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsList[_e] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contains: cdktf.stringToTerraform(struct.contains),
        exact: cdktf.stringToTerraform(struct.exact),
        ignore_case: cdktf.booleanToTerraform(struct.ignoreCase),
        prefix: cdktf.stringToTerraform(struct.prefix),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        contains: {
            value: cdktf.stringToHclTerraform(struct.contains),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exact: {
            value: cdktf.stringToHclTerraform(struct.exact),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ignore_case: {
            value: cdktf.booleanToHclTerraform(struct.ignoreCase),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        suffix: {
            value: cdktf.stringToHclTerraform(struct.suffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contains !== undefined) {
            hasAnyValues = true;
            internalValueResult.contains = this._contains;
        }
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._ignoreCase !== undefined) {
            hasAnyValues = true;
            internalValueResult.ignoreCase = this._ignoreCase;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._contains = undefined;
            this._exact = undefined;
            this._ignoreCase = undefined;
            this._prefix = undefined;
            this._suffix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._contains = value.contains;
            this._exact = value.exact;
            this._ignoreCase = value.ignoreCase;
            this._prefix = value.prefix;
            this._suffix = value.suffix;
        }
    }
    get contains() {
        return this.getStringAttribute('contains');
    }
    set contains(value) {
        this._contains = value;
    }
    resetContains() {
        this._contains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containsInput() {
        return this._contains;
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get ignoreCase() {
        return this.getBooleanAttribute('ignore_case');
    }
    set ignoreCase(value) {
        this._ignoreCase = value;
    }
    resetIgnoreCase() {
        this._ignoreCase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreCaseInput() {
        return this._ignoreCase;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountOutputReference;
_f = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountOutputReference[_f] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.ids),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.ids),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ids !== undefined) {
            hasAnyValues = true;
            internalValueResult.ids = this._ids;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ids = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ids = value.ids;
        }
    }
    get ids() {
        return this.getListAttribute('ids');
    }
    set ids(value) {
        this._ids = value;
    }
    resetIds() {
        this._ids = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idsInput() {
        return this._ids;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetOutputReference;
_g = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetOutputReference[_g] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam_service_account: networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToTerraform(struct.iamServiceAccount),
        tag_value_id_set: networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToTerraform(struct.tagValueIdSet),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        iam_service_account: {
            value: networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountToHclTerraform(struct.iamServiceAccount),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountList",
        },
        tag_value_id_set: {
            value: networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetToHclTerraform(struct.tagValueIdSet),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // iam_service_account - computed: false, optional: true, required: false
        this._iamServiceAccount = new NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesIamServiceAccountOutputReference(this, "iam_service_account");
        // tag_value_id_set - computed: false, optional: true, required: false
        this._tagValueIdSet = new NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesTagValueIdSetOutputReference(this, "tag_value_id_set");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iamServiceAccount?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.iamServiceAccount = this._iamServiceAccount?.internalValue;
        }
        if (this._tagValueIdSet?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagValueIdSet = this._tagValueIdSet?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._iamServiceAccount.internalValue = undefined;
            this._tagValueIdSet.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._iamServiceAccount.internalValue = value.iamServiceAccount;
            this._tagValueIdSet.internalValue = value.tagValueIdSet;
        }
    }
    get iamServiceAccount() {
        return this._iamServiceAccount;
    }
    putIamServiceAccount(value) {
        this._iamServiceAccount.internalValue = value;
    }
    resetIamServiceAccount() {
        this._iamServiceAccount.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamServiceAccountInput() {
        return this._iamServiceAccount.internalValue;
    }
    get tagValueIdSet() {
        return this._tagValueIdSet;
    }
    putTagValueIdSet(value) {
        this._tagValueIdSet.internalValue = value;
    }
    resetTagValueIdSet() {
        this._tagValueIdSet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagValueIdSetInput() {
        return this._tagValueIdSet.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesOutputReference;
_h = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesOutputReference[_h] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesList = NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesList;
_j = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesList[_j] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        principals: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToTerraform, true)(struct.principals),
        resources: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToTerraform, true)(struct.resources),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesToTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromNotSourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        principals: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsToHclTerraform, true)(struct.principals),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsList",
        },
        resources: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesToHclTerraform, true)(struct.resources),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromNotSourcesToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromNotSourcesToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // principals - computed: false, optional: true, required: false
        this._principals = new NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesPrincipalsList(this, "principals", false);
        // resources - computed: false, optional: true, required: false
        this._resources = new NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesResourcesList(this, "resources", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._principals?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.principals = this._principals?.internalValue;
        }
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._principals.internalValue = undefined;
            this._resources.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._principals.internalValue = value.principals;
            this._resources.internalValue = value.resources;
        }
    }
    get principals() {
        return this._principals;
    }
    putPrincipals(value) {
        this._principals.internalValue = value;
    }
    resetPrincipals() {
        this._principals.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get principalsInput() {
        return this._principals.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    resetResources() {
        this._resources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesOutputReference;
_k = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesOutputReference[_k] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesList = NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesList;
_l = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesList[_l] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contains: cdktf.stringToTerraform(struct.contains),
        exact: cdktf.stringToTerraform(struct.exact),
        ignore_case: cdktf.booleanToTerraform(struct.ignoreCase),
        prefix: cdktf.stringToTerraform(struct.prefix),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        contains: {
            value: cdktf.stringToHclTerraform(struct.contains),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exact: {
            value: cdktf.stringToHclTerraform(struct.exact),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ignore_case: {
            value: cdktf.booleanToHclTerraform(struct.ignoreCase),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        suffix: {
            value: cdktf.stringToHclTerraform(struct.suffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contains !== undefined) {
            hasAnyValues = true;
            internalValueResult.contains = this._contains;
        }
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._ignoreCase !== undefined) {
            hasAnyValues = true;
            internalValueResult.ignoreCase = this._ignoreCase;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._contains = undefined;
            this._exact = undefined;
            this._ignoreCase = undefined;
            this._prefix = undefined;
            this._suffix = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._contains = value.contains;
            this._exact = value.exact;
            this._ignoreCase = value.ignoreCase;
            this._prefix = value.prefix;
            this._suffix = value.suffix;
        }
    }
    get contains() {
        return this.getStringAttribute('contains');
    }
    set contains(value) {
        this._contains = value;
    }
    resetContains() {
        this._contains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containsInput() {
        return this._contains;
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get ignoreCase() {
        return this.getBooleanAttribute('ignore_case');
    }
    set ignoreCase(value) {
        this._ignoreCase = value;
    }
    resetIgnoreCase() {
        this._ignoreCase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreCaseInput() {
        return this._ignoreCase;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsOutputReference;
_m = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsOutputReference[_m] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsList = NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsList;
_o = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsList[_o] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contains: cdktf.stringToTerraform(struct.contains),
        exact: cdktf.stringToTerraform(struct.exact),
        ignore_case: cdktf.booleanToTerraform(struct.ignoreCase),
        prefix: cdktf.stringToTerraform(struct.prefix),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        contains: {
            value: cdktf.stringToHclTerraform(struct.contains),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exact: {
            value: cdktf.stringToHclTerraform(struct.exact),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ignore_case: {
            value: cdktf.booleanToHclTerraform(struct.ignoreCase),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        suffix: {
            value: cdktf.stringToHclTerraform(struct.suffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contains !== undefined) {
            hasAnyValues = true;
            internalValueResult.contains = this._contains;
        }
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._ignoreCase !== undefined) {
            hasAnyValues = true;
            internalValueResult.ignoreCase = this._ignoreCase;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._contains = undefined;
            this._exact = undefined;
            this._ignoreCase = undefined;
            this._prefix = undefined;
            this._suffix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._contains = value.contains;
            this._exact = value.exact;
            this._ignoreCase = value.ignoreCase;
            this._prefix = value.prefix;
            this._suffix = value.suffix;
        }
    }
    get contains() {
        return this.getStringAttribute('contains');
    }
    set contains(value) {
        this._contains = value;
    }
    resetContains() {
        this._contains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containsInput() {
        return this._contains;
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get ignoreCase() {
        return this.getBooleanAttribute('ignore_case');
    }
    set ignoreCase(value) {
        this._ignoreCase = value;
    }
    resetIgnoreCase() {
        this._ignoreCase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreCaseInput() {
        return this._ignoreCase;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountOutputReference;
_p = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountOutputReference[_p] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.ids),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.ids),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ids !== undefined) {
            hasAnyValues = true;
            internalValueResult.ids = this._ids;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ids = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ids = value.ids;
        }
    }
    get ids() {
        return this.getListAttribute('ids');
    }
    set ids(value) {
        this._ids = value;
    }
    resetIds() {
        this._ids = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idsInput() {
        return this._ids;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetOutputReference;
_q = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetOutputReference[_q] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam_service_account: networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToTerraform(struct.iamServiceAccount),
        tag_value_id_set: networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToTerraform(struct.tagValueIdSet),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        iam_service_account: {
            value: networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountToHclTerraform(struct.iamServiceAccount),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountList",
        },
        tag_value_id_set: {
            value: networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetToHclTerraform(struct.tagValueIdSet),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // iam_service_account - computed: false, optional: true, required: false
        this._iamServiceAccount = new NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesIamServiceAccountOutputReference(this, "iam_service_account");
        // tag_value_id_set - computed: false, optional: true, required: false
        this._tagValueIdSet = new NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesTagValueIdSetOutputReference(this, "tag_value_id_set");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iamServiceAccount?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.iamServiceAccount = this._iamServiceAccount?.internalValue;
        }
        if (this._tagValueIdSet?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tagValueIdSet = this._tagValueIdSet?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._iamServiceAccount.internalValue = undefined;
            this._tagValueIdSet.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._iamServiceAccount.internalValue = value.iamServiceAccount;
            this._tagValueIdSet.internalValue = value.tagValueIdSet;
        }
    }
    get iamServiceAccount() {
        return this._iamServiceAccount;
    }
    putIamServiceAccount(value) {
        this._iamServiceAccount.internalValue = value;
    }
    resetIamServiceAccount() {
        this._iamServiceAccount.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get iamServiceAccountInput() {
        return this._iamServiceAccount.internalValue;
    }
    get tagValueIdSet() {
        return this._tagValueIdSet;
    }
    putTagValueIdSet(value) {
        this._tagValueIdSet.internalValue = value;
    }
    resetTagValueIdSet() {
        this._tagValueIdSet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagValueIdSetInput() {
        return this._tagValueIdSet.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesOutputReference;
_r = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesOutputReference[_r] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesList = NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesList;
_s = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesList[_s] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromSourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        principals: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToTerraform, true)(struct.principals),
        resources: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToTerraform, true)(struct.resources),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesToTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromSourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        principals: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsToHclTerraform, true)(struct.principals),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsList",
        },
        resources: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesFromSourcesResourcesToHclTerraform, true)(struct.resources),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromSourcesToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromSourcesToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromSourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // principals - computed: false, optional: true, required: false
        this._principals = new NetworkSecurityAuthzPolicyHttpRulesFromSourcesPrincipalsList(this, "principals", false);
        // resources - computed: false, optional: true, required: false
        this._resources = new NetworkSecurityAuthzPolicyHttpRulesFromSourcesResourcesList(this, "resources", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._principals?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.principals = this._principals?.internalValue;
        }
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._principals.internalValue = undefined;
            this._resources.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._principals.internalValue = value.principals;
            this._resources.internalValue = value.resources;
        }
    }
    get principals() {
        return this._principals;
    }
    putPrincipals(value) {
        this._principals.internalValue = value;
    }
    resetPrincipals() {
        this._principals.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get principalsInput() {
        return this._principals.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    resetResources() {
        this._resources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromSourcesOutputReference;
_t = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromSourcesOutputReference[_t] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromSourcesOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesFromSourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesFromSourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromSourcesList = NetworkSecurityAuthzPolicyHttpRulesFromSourcesList;
_u = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromSourcesList[_u] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromSourcesList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesFromToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        not_sources: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesFromNotSourcesToTerraform, true)(struct.notSources),
        sources: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesFromSourcesToTerraform, true)(struct.sources),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesFromToTerraform = networkSecurityAuthzPolicyHttpRulesFromToTerraform;
function networkSecurityAuthzPolicyHttpRulesFromToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        not_sources: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesFromNotSourcesToHclTerraform, true)(struct.notSources),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesList",
        },
        sources: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesFromSourcesToHclTerraform, true)(struct.sources),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromSourcesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesFromToHclTerraform = networkSecurityAuthzPolicyHttpRulesFromToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesFromOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // not_sources - computed: false, optional: true, required: false
        this._notSources = new NetworkSecurityAuthzPolicyHttpRulesFromNotSourcesList(this, "not_sources", false);
        // sources - computed: false, optional: true, required: false
        this._sources = new NetworkSecurityAuthzPolicyHttpRulesFromSourcesList(this, "sources", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._notSources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.notSources = this._notSources?.internalValue;
        }
        if (this._sources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sources = this._sources?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._notSources.internalValue = undefined;
            this._sources.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._notSources.internalValue = value.notSources;
            this._sources.internalValue = value.sources;
        }
    }
    get notSources() {
        return this._notSources;
    }
    putNotSources(value) {
        this._notSources.internalValue = value;
    }
    resetNotSources() {
        this._notSources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notSourcesInput() {
        return this._notSources.internalValue;
    }
    get sources() {
        return this._sources;
    }
    putSources(value) {
        this._sources.internalValue = value;
    }
    resetSources() {
        this._sources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourcesInput() {
        return this._sources.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesFromOutputReference = NetworkSecurityAuthzPolicyHttpRulesFromOutputReference;
_v = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesFromOutputReference[_v] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesFromOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contains: cdktf.stringToTerraform(struct.contains),
        exact: cdktf.stringToTerraform(struct.exact),
        ignore_case: cdktf.booleanToTerraform(struct.ignoreCase),
        prefix: cdktf.stringToTerraform(struct.prefix),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToTerraform;
function networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        contains: {
            value: cdktf.stringToHclTerraform(struct.contains),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exact: {
            value: cdktf.stringToHclTerraform(struct.exact),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ignore_case: {
            value: cdktf.booleanToHclTerraform(struct.ignoreCase),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        suffix: {
            value: cdktf.stringToHclTerraform(struct.suffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToHclTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contains !== undefined) {
            hasAnyValues = true;
            internalValueResult.contains = this._contains;
        }
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._ignoreCase !== undefined) {
            hasAnyValues = true;
            internalValueResult.ignoreCase = this._ignoreCase;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._contains = undefined;
            this._exact = undefined;
            this._ignoreCase = undefined;
            this._prefix = undefined;
            this._suffix = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._contains = value.contains;
            this._exact = value.exact;
            this._ignoreCase = value.ignoreCase;
            this._prefix = value.prefix;
            this._suffix = value.suffix;
        }
    }
    get contains() {
        return this.getStringAttribute('contains');
    }
    set contains(value) {
        this._contains = value;
    }
    resetContains() {
        this._contains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containsInput() {
        return this._contains;
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get ignoreCase() {
        return this.getBooleanAttribute('ignore_case');
    }
    set ignoreCase(value) {
        this._ignoreCase = value;
    }
    resetIgnoreCase() {
        this._ignoreCase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreCaseInput() {
        return this._ignoreCase;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueOutputReference = NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueOutputReference;
_w = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueOutputReference[_w] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToTerraform(struct.value),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToTerraform;
function networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueToHclTerraform(struct.value),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToHclTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // value - computed: false, optional: true, required: false
        this._value = new NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersValueOutputReference(this, "value");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value.internalValue = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this._value;
    }
    putValue(value) {
        this._value.internalValue = value;
    }
    resetValue() {
        this._value.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersOutputReference = NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersOutputReference;
_x = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersOutputReference[_x] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersList = NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersList;
_y = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersList[_y] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        headers: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToTerraform, true)(struct.headers),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToTerraform;
function networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        headers: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersToHclTerraform, true)(struct.headers),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToHclTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // headers - computed: false, optional: true, required: false
        this._headers = new NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetHeadersList(this, "headers", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._headers?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.headers = this._headers?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._headers.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._headers.internalValue = value.headers;
        }
    }
    get headers() {
        return this._headers;
    }
    putHeaders(value) {
        this._headers.internalValue = value;
    }
    resetHeaders() {
        this._headers.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headersInput() {
        return this._headers.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetOutputReference = NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetOutputReference;
_z = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetOutputReference[_z] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesToOperationsHostsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contains: cdktf.stringToTerraform(struct.contains),
        exact: cdktf.stringToTerraform(struct.exact),
        ignore_case: cdktf.booleanToTerraform(struct.ignoreCase),
        prefix: cdktf.stringToTerraform(struct.prefix),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHostsToTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsHostsToTerraform;
function networkSecurityAuthzPolicyHttpRulesToOperationsHostsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        contains: {
            value: cdktf.stringToHclTerraform(struct.contains),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exact: {
            value: cdktf.stringToHclTerraform(struct.exact),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ignore_case: {
            value: cdktf.booleanToHclTerraform(struct.ignoreCase),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        suffix: {
            value: cdktf.stringToHclTerraform(struct.suffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsHostsToHclTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsHostsToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contains !== undefined) {
            hasAnyValues = true;
            internalValueResult.contains = this._contains;
        }
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._ignoreCase !== undefined) {
            hasAnyValues = true;
            internalValueResult.ignoreCase = this._ignoreCase;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._contains = undefined;
            this._exact = undefined;
            this._ignoreCase = undefined;
            this._prefix = undefined;
            this._suffix = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._contains = value.contains;
            this._exact = value.exact;
            this._ignoreCase = value.ignoreCase;
            this._prefix = value.prefix;
            this._suffix = value.suffix;
        }
    }
    get contains() {
        return this.getStringAttribute('contains');
    }
    set contains(value) {
        this._contains = value;
    }
    resetContains() {
        this._contains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containsInput() {
        return this._contains;
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get ignoreCase() {
        return this.getBooleanAttribute('ignore_case');
    }
    set ignoreCase(value) {
        this._ignoreCase = value;
    }
    resetIgnoreCase() {
        this._ignoreCase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreCaseInput() {
        return this._ignoreCase;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsOutputReference = NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsOutputReference[_0] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsList = NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsList;
_1 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsList[_1] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesToOperationsPathsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        contains: cdktf.stringToTerraform(struct.contains),
        exact: cdktf.stringToTerraform(struct.exact),
        ignore_case: cdktf.booleanToTerraform(struct.ignoreCase),
        prefix: cdktf.stringToTerraform(struct.prefix),
        suffix: cdktf.stringToTerraform(struct.suffix),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsPathsToTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsPathsToTerraform;
function networkSecurityAuthzPolicyHttpRulesToOperationsPathsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        contains: {
            value: cdktf.stringToHclTerraform(struct.contains),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exact: {
            value: cdktf.stringToHclTerraform(struct.exact),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ignore_case: {
            value: cdktf.booleanToHclTerraform(struct.ignoreCase),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        prefix: {
            value: cdktf.stringToHclTerraform(struct.prefix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        suffix: {
            value: cdktf.stringToHclTerraform(struct.suffix),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsPathsToHclTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsPathsToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contains !== undefined) {
            hasAnyValues = true;
            internalValueResult.contains = this._contains;
        }
        if (this._exact !== undefined) {
            hasAnyValues = true;
            internalValueResult.exact = this._exact;
        }
        if (this._ignoreCase !== undefined) {
            hasAnyValues = true;
            internalValueResult.ignoreCase = this._ignoreCase;
        }
        if (this._prefix !== undefined) {
            hasAnyValues = true;
            internalValueResult.prefix = this._prefix;
        }
        if (this._suffix !== undefined) {
            hasAnyValues = true;
            internalValueResult.suffix = this._suffix;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._contains = undefined;
            this._exact = undefined;
            this._ignoreCase = undefined;
            this._prefix = undefined;
            this._suffix = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._contains = value.contains;
            this._exact = value.exact;
            this._ignoreCase = value.ignoreCase;
            this._prefix = value.prefix;
            this._suffix = value.suffix;
        }
    }
    get contains() {
        return this.getStringAttribute('contains');
    }
    set contains(value) {
        this._contains = value;
    }
    resetContains() {
        this._contains = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containsInput() {
        return this._contains;
    }
    get exact() {
        return this.getStringAttribute('exact');
    }
    set exact(value) {
        this._exact = value;
    }
    resetExact() {
        this._exact = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exactInput() {
        return this._exact;
    }
    get ignoreCase() {
        return this.getBooleanAttribute('ignore_case');
    }
    set ignoreCase(value) {
        this._ignoreCase = value;
    }
    resetIgnoreCase() {
        this._ignoreCase = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreCaseInput() {
        return this._ignoreCase;
    }
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    set prefix(value) {
        this._prefix = value;
    }
    resetPrefix() {
        this._prefix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get prefixInput() {
        return this._prefix;
    }
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    set suffix(value) {
        this._suffix = value;
    }
    resetSuffix() {
        this._suffix = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get suffixInput() {
        return this._suffix;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsOutputReference = NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsOutputReference[_2] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsList = NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsList;
_3 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsList[_3] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesToOperationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        methods: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.methods),
        header_set: networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToTerraform(struct.headerSet),
        hosts: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesToOperationsHostsToTerraform, true)(struct.hosts),
        paths: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesToOperationsPathsToTerraform, true)(struct.paths),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsToTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsToTerraform;
function networkSecurityAuthzPolicyHttpRulesToOperationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        methods: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.methods),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        header_set: {
            value: networkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetToHclTerraform(struct.headerSet),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetList",
        },
        hosts: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesToOperationsHostsToHclTerraform, true)(struct.hosts),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsList",
        },
        paths: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesToOperationsPathsToHclTerraform, true)(struct.paths),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesToOperationsToHclTerraform = networkSecurityAuthzPolicyHttpRulesToOperationsToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesToOperationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // header_set - computed: false, optional: true, required: false
        this._headerSet = new NetworkSecurityAuthzPolicyHttpRulesToOperationsHeaderSetOutputReference(this, "header_set");
        // hosts - computed: false, optional: true, required: false
        this._hosts = new NetworkSecurityAuthzPolicyHttpRulesToOperationsHostsList(this, "hosts", false);
        // paths - computed: false, optional: true, required: false
        this._paths = new NetworkSecurityAuthzPolicyHttpRulesToOperationsPathsList(this, "paths", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._methods !== undefined) {
            hasAnyValues = true;
            internalValueResult.methods = this._methods;
        }
        if (this._headerSet?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.headerSet = this._headerSet?.internalValue;
        }
        if (this._hosts?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hosts = this._hosts?.internalValue;
        }
        if (this._paths?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.paths = this._paths?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._methods = undefined;
            this._headerSet.internalValue = undefined;
            this._hosts.internalValue = undefined;
            this._paths.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._methods = value.methods;
            this._headerSet.internalValue = value.headerSet;
            this._hosts.internalValue = value.hosts;
            this._paths.internalValue = value.paths;
        }
    }
    get methods() {
        return this.getListAttribute('methods');
    }
    set methods(value) {
        this._methods = value;
    }
    resetMethods() {
        this._methods = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get methodsInput() {
        return this._methods;
    }
    get headerSet() {
        return this._headerSet;
    }
    putHeaderSet(value) {
        this._headerSet.internalValue = value;
    }
    resetHeaderSet() {
        this._headerSet.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerSetInput() {
        return this._headerSet.internalValue;
    }
    get hosts() {
        return this._hosts;
    }
    putHosts(value) {
        this._hosts.internalValue = value;
    }
    resetHosts() {
        this._hosts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostsInput() {
        return this._hosts.internalValue;
    }
    get paths() {
        return this._paths;
    }
    putPaths(value) {
        this._paths.internalValue = value;
    }
    resetPaths() {
        this._paths.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathsInput() {
        return this._paths.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsOutputReference = NetworkSecurityAuthzPolicyHttpRulesToOperationsOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsOutputReference[_4] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesToOperationsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesToOperationsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOperationsList = NetworkSecurityAuthzPolicyHttpRulesToOperationsList;
_5 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOperationsList[_5] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOperationsList", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesToToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operations: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesToOperationsToTerraform, true)(struct.operations),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesToToTerraform = networkSecurityAuthzPolicyHttpRulesToToTerraform;
function networkSecurityAuthzPolicyHttpRulesToToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operations: {
            value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesToOperationsToHclTerraform, true)(struct.operations),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesToOperationsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesToToHclTerraform = networkSecurityAuthzPolicyHttpRulesToToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesToOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // operations - computed: false, optional: true, required: false
        this._operations = new NetworkSecurityAuthzPolicyHttpRulesToOperationsList(this, "operations", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operations?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.operations = this._operations?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._operations.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._operations.internalValue = value.operations;
        }
    }
    get operations() {
        return this._operations;
    }
    putOperations(value) {
        this._operations.internalValue = value;
    }
    resetOperations() {
        this._operations.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get operationsInput() {
        return this._operations.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesToOutputReference = NetworkSecurityAuthzPolicyHttpRulesToOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesToOutputReference[_6] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesToOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyHttpRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        when: cdktf.stringToTerraform(struct.when),
        from: networkSecurityAuthzPolicyHttpRulesFromToTerraform(struct.from),
        to: networkSecurityAuthzPolicyHttpRulesToToTerraform(struct.to),
    };
}
exports.networkSecurityAuthzPolicyHttpRulesToTerraform = networkSecurityAuthzPolicyHttpRulesToTerraform;
function networkSecurityAuthzPolicyHttpRulesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        when: {
            value: cdktf.stringToHclTerraform(struct.when),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        from: {
            value: networkSecurityAuthzPolicyHttpRulesFromToHclTerraform(struct.from),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesFromList",
        },
        to: {
            value: networkSecurityAuthzPolicyHttpRulesToToHclTerraform(struct.to),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecurityAuthzPolicyHttpRulesToList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyHttpRulesToHclTerraform = networkSecurityAuthzPolicyHttpRulesToHclTerraform;
class NetworkSecurityAuthzPolicyHttpRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // from - computed: false, optional: true, required: false
        this._from = new NetworkSecurityAuthzPolicyHttpRulesFromOutputReference(this, "from");
        // to - computed: false, optional: true, required: false
        this._to = new NetworkSecurityAuthzPolicyHttpRulesToOutputReference(this, "to");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._when !== undefined) {
            hasAnyValues = true;
            internalValueResult.when = this._when;
        }
        if (this._from?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.from = this._from?.internalValue;
        }
        if (this._to?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.to = this._to?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._when = undefined;
            this._from.internalValue = undefined;
            this._to.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._when = value.when;
            this._from.internalValue = value.from;
            this._to.internalValue = value.to;
        }
    }
    get when() {
        return this.getStringAttribute('when');
    }
    set when(value) {
        this._when = value;
    }
    resetWhen() {
        this._when = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get whenInput() {
        return this._when;
    }
    get from() {
        return this._from;
    }
    putFrom(value) {
        this._from.internalValue = value;
    }
    resetFrom() {
        this._from.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fromInput() {
        return this._from.internalValue;
    }
    get to() {
        return this._to;
    }
    putTo(value) {
        this._to.internalValue = value;
    }
    resetTo() {
        this._to.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get toInput() {
        return this._to.internalValue;
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesOutputReference = NetworkSecurityAuthzPolicyHttpRulesOutputReference;
_7 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesOutputReference[_7] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesOutputReference", version: "14.13.1" };
class NetworkSecurityAuthzPolicyHttpRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecurityAuthzPolicyHttpRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecurityAuthzPolicyHttpRulesList = NetworkSecurityAuthzPolicyHttpRulesList;
_8 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyHttpRulesList[_8] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyHttpRulesList", version: "14.13.1" };
function networkSecurityAuthzPolicyTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        load_balancing_scheme: cdktf.stringToTerraform(struct.loadBalancingScheme),
        resources: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.resources),
    };
}
exports.networkSecurityAuthzPolicyTargetToTerraform = networkSecurityAuthzPolicyTargetToTerraform;
function networkSecurityAuthzPolicyTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        load_balancing_scheme: {
            value: cdktf.stringToHclTerraform(struct.loadBalancingScheme),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        resources: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.resources),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyTargetToHclTerraform = networkSecurityAuthzPolicyTargetToHclTerraform;
class NetworkSecurityAuthzPolicyTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._loadBalancingScheme !== undefined) {
            hasAnyValues = true;
            internalValueResult.loadBalancingScheme = this._loadBalancingScheme;
        }
        if (this._resources !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._loadBalancingScheme = undefined;
            this._resources = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._loadBalancingScheme = value.loadBalancingScheme;
            this._resources = value.resources;
        }
    }
    get loadBalancingScheme() {
        return this.getStringAttribute('load_balancing_scheme');
    }
    set loadBalancingScheme(value) {
        this._loadBalancingScheme = value;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancingSchemeInput() {
        return this._loadBalancingScheme;
    }
    get resources() {
        return this.getListAttribute('resources');
    }
    set resources(value) {
        this._resources = value;
    }
    resetResources() {
        this._resources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources;
    }
}
exports.NetworkSecurityAuthzPolicyTargetOutputReference = NetworkSecurityAuthzPolicyTargetOutputReference;
_9 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyTargetOutputReference[_9] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyTargetOutputReference", version: "14.13.1" };
function networkSecurityAuthzPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkSecurityAuthzPolicyTimeoutsToTerraform = networkSecurityAuthzPolicyTimeoutsToTerraform;
function networkSecurityAuthzPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecurityAuthzPolicyTimeoutsToHclTerraform = networkSecurityAuthzPolicyTimeoutsToHclTerraform;
class NetworkSecurityAuthzPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkSecurityAuthzPolicyTimeoutsOutputReference = NetworkSecurityAuthzPolicyTimeoutsOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicyTimeoutsOutputReference[_10] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_authz_policy google_network_security_authz_policy}
*/
class NetworkSecurityAuthzPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkSecurityAuthzPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkSecurityAuthzPolicy to import
    * @param importFromId The id of the existing NetworkSecurityAuthzPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_authz_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkSecurityAuthzPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_network_security_authz_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_authz_policy google_network_security_authz_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkSecurityAuthzPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_network_security_authz_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // custom_provider - computed: false, optional: true, required: false
        this._customProvider = new NetworkSecurityAuthzPolicyCustomProviderOutputReference(this, "custom_provider");
        // http_rules - computed: false, optional: true, required: false
        this._httpRules = new NetworkSecurityAuthzPolicyHttpRulesList(this, "http_rules", false);
        // target - computed: false, optional: false, required: true
        this._target = new NetworkSecurityAuthzPolicyTargetOutputReference(this, "target");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkSecurityAuthzPolicyTimeoutsOutputReference(this, "timeouts");
        this._action = config.action;
        this._description = config.description;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._customProvider.internalValue = config.customProvider;
        this._httpRules.internalValue = config.httpRules;
        this._target.internalValue = config.target;
        this._timeouts.internalValue = config.timeouts;
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get customProvider() {
        return this._customProvider;
    }
    putCustomProvider(value) {
        this._customProvider.internalValue = value;
    }
    resetCustomProvider() {
        this._customProvider.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customProviderInput() {
        return this._customProvider.internalValue;
    }
    get httpRules() {
        return this._httpRules;
    }
    putHttpRules(value) {
        this._httpRules.internalValue = value;
    }
    resetHttpRules() {
        this._httpRules.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpRulesInput() {
        return this._httpRules.internalValue;
    }
    get target() {
        return this._target;
    }
    putTarget(value) {
        this._target.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            action: cdktf.stringToTerraform(this._action),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            custom_provider: networkSecurityAuthzPolicyCustomProviderToTerraform(this._customProvider.internalValue),
            http_rules: cdktf.listMapper(networkSecurityAuthzPolicyHttpRulesToTerraform, true)(this._httpRules.internalValue),
            target: networkSecurityAuthzPolicyTargetToTerraform(this._target.internalValue),
            timeouts: networkSecurityAuthzPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            action: {
                value: cdktf.stringToHclTerraform(this._action),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            custom_provider: {
                value: networkSecurityAuthzPolicyCustomProviderToHclTerraform(this._customProvider.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkSecurityAuthzPolicyCustomProviderList",
            },
            http_rules: {
                value: cdktf.listMapperHcl(networkSecurityAuthzPolicyHttpRulesToHclTerraform, true)(this._httpRules.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkSecurityAuthzPolicyHttpRulesList",
            },
            target: {
                value: networkSecurityAuthzPolicyTargetToHclTerraform(this._target.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkSecurityAuthzPolicyTargetList",
            },
            timeouts: {
                value: networkSecurityAuthzPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkSecurityAuthzPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkSecurityAuthzPolicy = NetworkSecurityAuthzPolicy;
_11 = JSII_RTTI_SYMBOL_1;
NetworkSecurityAuthzPolicy[_11] = { fqn: "@cdktf/provider-google.networkSecurityAuthzPolicy.NetworkSecurityAuthzPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetworkSecurityAuthzPolicy.tfResourceType = "google_network_security_authz_policy";
