// @cdktf/provider-google/lib/data-google-cloudbuild-trigger/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleCloudbuildTriggerBuildOptionsVolumesToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildOptionsVolumesToTerraform = exports.DataGoogleCloudbuildTriggerBuildAvailableSecretsList = exports.DataGoogleCloudbuildTriggerBuildAvailableSecretsOutputReference = exports.dataGoogleCloudbuildTriggerBuildAvailableSecretsToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildAvailableSecretsToTerraform = exports.DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerList = exports.DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference = exports.dataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform = exports.DataGoogleCloudbuildTriggerBuildArtifactsList = exports.DataGoogleCloudbuildTriggerBuildArtifactsOutputReference = exports.dataGoogleCloudbuildTriggerBuildArtifactsToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildArtifactsToTerraform = exports.DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesList = exports.DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesOutputReference = exports.dataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesToTerraform = exports.DataGoogleCloudbuildTriggerBuildArtifactsObjectsList = exports.DataGoogleCloudbuildTriggerBuildArtifactsObjectsOutputReference = exports.dataGoogleCloudbuildTriggerBuildArtifactsObjectsToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildArtifactsObjectsToTerraform = exports.DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingList = exports.DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingOutputReference = exports.dataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingToTerraform = exports.DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesList = exports.DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesOutputReference = exports.dataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesToTerraform = exports.DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsList = exports.DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsOutputReference = exports.dataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsToTerraform = exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigList = exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigOutputReference = exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigToHclTerraform = exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigToTerraform = exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushList = exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushOutputReference = exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushToHclTerraform = exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushToTerraform = exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestList = exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestOutputReference = exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestToHclTerraform = exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestToTerraform = exports.DataGoogleCloudbuildTriggerApprovalConfigList = exports.DataGoogleCloudbuildTriggerApprovalConfigOutputReference = exports.dataGoogleCloudbuildTriggerApprovalConfigToHclTerraform = exports.dataGoogleCloudbuildTriggerApprovalConfigToTerraform = void 0;
exports.DataGoogleCloudbuildTriggerGithubList = exports.DataGoogleCloudbuildTriggerGithubOutputReference = exports.dataGoogleCloudbuildTriggerGithubToHclTerraform = exports.dataGoogleCloudbuildTriggerGithubToTerraform = exports.DataGoogleCloudbuildTriggerGithubPushList = exports.DataGoogleCloudbuildTriggerGithubPushOutputReference = exports.dataGoogleCloudbuildTriggerGithubPushToHclTerraform = exports.dataGoogleCloudbuildTriggerGithubPushToTerraform = exports.DataGoogleCloudbuildTriggerGithubPullRequestList = exports.DataGoogleCloudbuildTriggerGithubPullRequestOutputReference = exports.dataGoogleCloudbuildTriggerGithubPullRequestToHclTerraform = exports.dataGoogleCloudbuildTriggerGithubPullRequestToTerraform = exports.DataGoogleCloudbuildTriggerGitFileSourceList = exports.DataGoogleCloudbuildTriggerGitFileSourceOutputReference = exports.dataGoogleCloudbuildTriggerGitFileSourceToHclTerraform = exports.dataGoogleCloudbuildTriggerGitFileSourceToTerraform = exports.DataGoogleCloudbuildTriggerBuildList = exports.DataGoogleCloudbuildTriggerBuildOutputReference = exports.dataGoogleCloudbuildTriggerBuildToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildToTerraform = exports.DataGoogleCloudbuildTriggerBuildStepList = exports.DataGoogleCloudbuildTriggerBuildStepOutputReference = exports.dataGoogleCloudbuildTriggerBuildStepToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildStepToTerraform = exports.DataGoogleCloudbuildTriggerBuildStepVolumesList = exports.DataGoogleCloudbuildTriggerBuildStepVolumesOutputReference = exports.dataGoogleCloudbuildTriggerBuildStepVolumesToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildStepVolumesToTerraform = exports.DataGoogleCloudbuildTriggerBuildSourceList = exports.DataGoogleCloudbuildTriggerBuildSourceOutputReference = exports.dataGoogleCloudbuildTriggerBuildSourceToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildSourceToTerraform = exports.DataGoogleCloudbuildTriggerBuildSourceStorageSourceList = exports.DataGoogleCloudbuildTriggerBuildSourceStorageSourceOutputReference = exports.dataGoogleCloudbuildTriggerBuildSourceStorageSourceToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildSourceStorageSourceToTerraform = exports.DataGoogleCloudbuildTriggerBuildSourceRepoSourceList = exports.DataGoogleCloudbuildTriggerBuildSourceRepoSourceOutputReference = exports.dataGoogleCloudbuildTriggerBuildSourceRepoSourceToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildSourceRepoSourceToTerraform = exports.DataGoogleCloudbuildTriggerBuildSecretList = exports.DataGoogleCloudbuildTriggerBuildSecretOutputReference = exports.dataGoogleCloudbuildTriggerBuildSecretToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildSecretToTerraform = exports.DataGoogleCloudbuildTriggerBuildOptionsList = exports.DataGoogleCloudbuildTriggerBuildOptionsOutputReference = exports.dataGoogleCloudbuildTriggerBuildOptionsToHclTerraform = exports.dataGoogleCloudbuildTriggerBuildOptionsToTerraform = exports.DataGoogleCloudbuildTriggerBuildOptionsVolumesList = exports.DataGoogleCloudbuildTriggerBuildOptionsVolumesOutputReference = void 0;
exports.DataGoogleCloudbuildTrigger = exports.DataGoogleCloudbuildTriggerWebhookConfigList = exports.DataGoogleCloudbuildTriggerWebhookConfigOutputReference = exports.dataGoogleCloudbuildTriggerWebhookConfigToHclTerraform = exports.dataGoogleCloudbuildTriggerWebhookConfigToTerraform = exports.DataGoogleCloudbuildTriggerTriggerTemplateList = exports.DataGoogleCloudbuildTriggerTriggerTemplateOutputReference = exports.dataGoogleCloudbuildTriggerTriggerTemplateToHclTerraform = exports.dataGoogleCloudbuildTriggerTriggerTemplateToTerraform = exports.DataGoogleCloudbuildTriggerSourceToBuildList = exports.DataGoogleCloudbuildTriggerSourceToBuildOutputReference = exports.dataGoogleCloudbuildTriggerSourceToBuildToHclTerraform = exports.dataGoogleCloudbuildTriggerSourceToBuildToTerraform = exports.DataGoogleCloudbuildTriggerRepositoryEventConfigList = exports.DataGoogleCloudbuildTriggerRepositoryEventConfigOutputReference = exports.dataGoogleCloudbuildTriggerRepositoryEventConfigToHclTerraform = exports.dataGoogleCloudbuildTriggerRepositoryEventConfigToTerraform = exports.DataGoogleCloudbuildTriggerRepositoryEventConfigPushList = exports.DataGoogleCloudbuildTriggerRepositoryEventConfigPushOutputReference = exports.dataGoogleCloudbuildTriggerRepositoryEventConfigPushToHclTerraform = exports.dataGoogleCloudbuildTriggerRepositoryEventConfigPushToTerraform = exports.DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestList = exports.DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestOutputReference = exports.dataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestToHclTerraform = exports.dataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestToTerraform = exports.DataGoogleCloudbuildTriggerPubsubConfigList = exports.DataGoogleCloudbuildTriggerPubsubConfigOutputReference = exports.dataGoogleCloudbuildTriggerPubsubConfigToHclTerraform = exports.dataGoogleCloudbuildTriggerPubsubConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleCloudbuildTriggerApprovalConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerApprovalConfigToTerraform = dataGoogleCloudbuildTriggerApprovalConfigToTerraform;
function dataGoogleCloudbuildTriggerApprovalConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerApprovalConfigToHclTerraform = dataGoogleCloudbuildTriggerApprovalConfigToHclTerraform;
class DataGoogleCloudbuildTriggerApprovalConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // approval_required - computed: true, optional: false, required: false
    get approvalRequired() {
        return this.getBooleanAttribute('approval_required');
    }
}
exports.DataGoogleCloudbuildTriggerApprovalConfigOutputReference = DataGoogleCloudbuildTriggerApprovalConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerApprovalConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerApprovalConfigOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerApprovalConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerApprovalConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerApprovalConfigList = DataGoogleCloudbuildTriggerApprovalConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerApprovalConfigList[_b] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerApprovalConfigList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestToTerraform = dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestToTerraform;
function dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestToHclTerraform = dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestToHclTerraform;
class DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch - computed: true, optional: false, required: false
    get branch() {
        return this.getStringAttribute('branch');
    }
    // comment_control - computed: true, optional: false, required: false
    get commentControl() {
        return this.getStringAttribute('comment_control');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
}
exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestOutputReference = DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestList = DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestList[_d] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushToTerraform = dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushToTerraform;
function dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushToHclTerraform = dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushToHclTerraform;
class DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch - computed: true, optional: false, required: false
    get branch() {
        return this.getStringAttribute('branch');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    // tag - computed: true, optional: false, required: false
    get tag() {
        return this.getStringAttribute('tag');
    }
}
exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushOutputReference = DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushList = DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushList[_f] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigToTerraform = dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigToTerraform;
function dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigToHclTerraform = dataGoogleCloudbuildTriggerBitbucketServerTriggerConfigToHclTerraform;
class DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // pull_request - computed: true, optional: false, required: false
        this._pullRequest = new DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPullRequestList(this, "pull_request", false);
        // push - computed: true, optional: false, required: false
        this._push = new DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigPushList(this, "push", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bitbucket_server_config_resource - computed: true, optional: false, required: false
    get bitbucketServerConfigResource() {
        return this.getStringAttribute('bitbucket_server_config_resource');
    }
    // project_key - computed: true, optional: false, required: false
    get projectKey() {
        return this.getStringAttribute('project_key');
    }
    get pullRequest() {
        return this._pullRequest;
    }
    get push() {
        return this._push;
    }
    // repo_slug - computed: true, optional: false, required: false
    get repoSlug() {
        return this.getStringAttribute('repo_slug');
    }
}
exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigOutputReference = DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigList = DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigList[_h] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsToTerraform = dataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsToTerraform;
function dataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsToHclTerraform = dataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsToHclTerraform;
class DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // artifact_id - computed: true, optional: false, required: false
    get artifactId() {
        return this.getStringAttribute('artifact_id');
    }
    // group_id - computed: true, optional: false, required: false
    get groupId() {
        return this.getStringAttribute('group_id');
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // repository - computed: true, optional: false, required: false
    get repository() {
        return this.getStringAttribute('repository');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsOutputReference = DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsList = DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsList[_k] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesToTerraform = dataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesToTerraform;
function dataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesToHclTerraform = dataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesToHclTerraform;
class DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // package_path - computed: true, optional: false, required: false
    get packagePath() {
        return this.getStringAttribute('package_path');
    }
    // repository - computed: true, optional: false, required: false
    get repository() {
        return this.getStringAttribute('repository');
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesOutputReference = DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesList = DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesList[_m] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingToTerraform = dataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingToTerraform;
function dataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingToHclTerraform = dataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingToHclTerraform;
class DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // end_time - computed: true, optional: false, required: false
    get endTime() {
        return this.getStringAttribute('end_time');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingOutputReference = DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingList = DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingList[_p] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildArtifactsObjectsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsObjectsToTerraform = dataGoogleCloudbuildTriggerBuildArtifactsObjectsToTerraform;
function dataGoogleCloudbuildTriggerBuildArtifactsObjectsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsObjectsToHclTerraform = dataGoogleCloudbuildTriggerBuildArtifactsObjectsToHclTerraform;
class DataGoogleCloudbuildTriggerBuildArtifactsObjectsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // timing - computed: true, optional: false, required: false
        this._timing = new DataGoogleCloudbuildTriggerBuildArtifactsObjectsTimingList(this, "timing", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // paths - computed: true, optional: false, required: false
    get paths() {
        return this.getListAttribute('paths');
    }
    get timing() {
        return this._timing;
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsObjectsOutputReference = DataGoogleCloudbuildTriggerBuildArtifactsObjectsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsObjectsOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsObjectsOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildArtifactsObjectsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildArtifactsObjectsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsObjectsList = DataGoogleCloudbuildTriggerBuildArtifactsObjectsList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsObjectsList[_r] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsObjectsList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesToTerraform = dataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesToTerraform;
function dataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesToHclTerraform = dataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesToHclTerraform;
class DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // paths - computed: true, optional: false, required: false
    get paths() {
        return this.getListAttribute('paths');
    }
    // repository - computed: true, optional: false, required: false
    get repository() {
        return this.getStringAttribute('repository');
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesOutputReference = DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesList = DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesList[_t] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildArtifactsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsToTerraform = dataGoogleCloudbuildTriggerBuildArtifactsToTerraform;
function dataGoogleCloudbuildTriggerBuildArtifactsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildArtifactsToHclTerraform = dataGoogleCloudbuildTriggerBuildArtifactsToHclTerraform;
class DataGoogleCloudbuildTriggerBuildArtifactsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // maven_artifacts - computed: true, optional: false, required: false
        this._mavenArtifacts = new DataGoogleCloudbuildTriggerBuildArtifactsMavenArtifactsList(this, "maven_artifacts", false);
        // npm_packages - computed: true, optional: false, required: false
        this._npmPackages = new DataGoogleCloudbuildTriggerBuildArtifactsNpmPackagesList(this, "npm_packages", false);
        // objects - computed: true, optional: false, required: false
        this._objects = new DataGoogleCloudbuildTriggerBuildArtifactsObjectsList(this, "objects", false);
        // python_packages - computed: true, optional: false, required: false
        this._pythonPackages = new DataGoogleCloudbuildTriggerBuildArtifactsPythonPackagesList(this, "python_packages", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // images - computed: true, optional: false, required: false
    get images() {
        return this.getListAttribute('images');
    }
    get mavenArtifacts() {
        return this._mavenArtifacts;
    }
    get npmPackages() {
        return this._npmPackages;
    }
    get objects() {
        return this._objects;
    }
    get pythonPackages() {
        return this._pythonPackages;
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsOutputReference = DataGoogleCloudbuildTriggerBuildArtifactsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildArtifactsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildArtifactsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildArtifactsList = DataGoogleCloudbuildTriggerBuildArtifactsList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildArtifactsList[_v] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildArtifactsList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform = dataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerToTerraform;
function dataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerToHclTerraform = dataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerToHclTerraform;
class DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // env - computed: true, optional: false, required: false
    get env() {
        return this.getStringAttribute('env');
    }
    // version_name - computed: true, optional: false, required: false
    get versionName() {
        return this.getStringAttribute('version_name');
    }
}
exports.DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference = DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerList = DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerList[_x] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildAvailableSecretsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildAvailableSecretsToTerraform = dataGoogleCloudbuildTriggerBuildAvailableSecretsToTerraform;
function dataGoogleCloudbuildTriggerBuildAvailableSecretsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildAvailableSecretsToHclTerraform = dataGoogleCloudbuildTriggerBuildAvailableSecretsToHclTerraform;
class DataGoogleCloudbuildTriggerBuildAvailableSecretsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // secret_manager - computed: true, optional: false, required: false
        this._secretManager = new DataGoogleCloudbuildTriggerBuildAvailableSecretsSecretManagerList(this, "secret_manager", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get secretManager() {
        return this._secretManager;
    }
}
exports.DataGoogleCloudbuildTriggerBuildAvailableSecretsOutputReference = DataGoogleCloudbuildTriggerBuildAvailableSecretsOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildAvailableSecretsOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildAvailableSecretsOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildAvailableSecretsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildAvailableSecretsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildAvailableSecretsList = DataGoogleCloudbuildTriggerBuildAvailableSecretsList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildAvailableSecretsList[_z] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildAvailableSecretsList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildOptionsVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildOptionsVolumesToTerraform = dataGoogleCloudbuildTriggerBuildOptionsVolumesToTerraform;
function dataGoogleCloudbuildTriggerBuildOptionsVolumesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildOptionsVolumesToHclTerraform = dataGoogleCloudbuildTriggerBuildOptionsVolumesToHclTerraform;
class DataGoogleCloudbuildTriggerBuildOptionsVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
}
exports.DataGoogleCloudbuildTriggerBuildOptionsVolumesOutputReference = DataGoogleCloudbuildTriggerBuildOptionsVolumesOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildOptionsVolumesOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildOptionsVolumesOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildOptionsVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildOptionsVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildOptionsVolumesList = DataGoogleCloudbuildTriggerBuildOptionsVolumesList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildOptionsVolumesList[_1] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildOptionsVolumesList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildOptionsToTerraform = dataGoogleCloudbuildTriggerBuildOptionsToTerraform;
function dataGoogleCloudbuildTriggerBuildOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildOptionsToHclTerraform = dataGoogleCloudbuildTriggerBuildOptionsToHclTerraform;
class DataGoogleCloudbuildTriggerBuildOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // volumes - computed: true, optional: false, required: false
        this._volumes = new DataGoogleCloudbuildTriggerBuildOptionsVolumesList(this, "volumes", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // disk_size_gb - computed: true, optional: false, required: false
    get diskSizeGb() {
        return this.getNumberAttribute('disk_size_gb');
    }
    // dynamic_substitutions - computed: true, optional: false, required: false
    get dynamicSubstitutions() {
        return this.getBooleanAttribute('dynamic_substitutions');
    }
    // env - computed: true, optional: false, required: false
    get env() {
        return this.getListAttribute('env');
    }
    // log_streaming_option - computed: true, optional: false, required: false
    get logStreamingOption() {
        return this.getStringAttribute('log_streaming_option');
    }
    // logging - computed: true, optional: false, required: false
    get logging() {
        return this.getStringAttribute('logging');
    }
    // machine_type - computed: true, optional: false, required: false
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    // requested_verify_option - computed: true, optional: false, required: false
    get requestedVerifyOption() {
        return this.getStringAttribute('requested_verify_option');
    }
    // secret_env - computed: true, optional: false, required: false
    get secretEnv() {
        return this.getListAttribute('secret_env');
    }
    // source_provenance_hash - computed: true, optional: false, required: false
    get sourceProvenanceHash() {
        return this.getListAttribute('source_provenance_hash');
    }
    // substitution_option - computed: true, optional: false, required: false
    get substitutionOption() {
        return this.getStringAttribute('substitution_option');
    }
    get volumes() {
        return this._volumes;
    }
    // worker_pool - computed: true, optional: false, required: false
    get workerPool() {
        return this.getStringAttribute('worker_pool');
    }
}
exports.DataGoogleCloudbuildTriggerBuildOptionsOutputReference = DataGoogleCloudbuildTriggerBuildOptionsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildOptionsOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildOptionsOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildOptionsList = DataGoogleCloudbuildTriggerBuildOptionsList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildOptionsList[_3] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildOptionsList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildSecretToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildSecretToTerraform = dataGoogleCloudbuildTriggerBuildSecretToTerraform;
function dataGoogleCloudbuildTriggerBuildSecretToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildSecretToHclTerraform = dataGoogleCloudbuildTriggerBuildSecretToHclTerraform;
class DataGoogleCloudbuildTriggerBuildSecretOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // secret_env - computed: true, optional: false, required: false
        this._secretEnv = new cdktf.StringMap(this, "secret_env");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_name - computed: true, optional: false, required: false
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    get secretEnv() {
        return this._secretEnv;
    }
}
exports.DataGoogleCloudbuildTriggerBuildSecretOutputReference = DataGoogleCloudbuildTriggerBuildSecretOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildSecretOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildSecretOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildSecretList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildSecretOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildSecretList = DataGoogleCloudbuildTriggerBuildSecretList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildSecretList[_5] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildSecretList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildSourceRepoSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildSourceRepoSourceToTerraform = dataGoogleCloudbuildTriggerBuildSourceRepoSourceToTerraform;
function dataGoogleCloudbuildTriggerBuildSourceRepoSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildSourceRepoSourceToHclTerraform = dataGoogleCloudbuildTriggerBuildSourceRepoSourceToHclTerraform;
class DataGoogleCloudbuildTriggerBuildSourceRepoSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // substitutions - computed: true, optional: false, required: false
        this._substitutions = new cdktf.StringMap(this, "substitutions");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch_name - computed: true, optional: false, required: false
    get branchName() {
        return this.getStringAttribute('branch_name');
    }
    // commit_sha - computed: true, optional: false, required: false
    get commitSha() {
        return this.getStringAttribute('commit_sha');
    }
    // dir - computed: true, optional: false, required: false
    get dir() {
        return this.getStringAttribute('dir');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    // project_id - computed: true, optional: false, required: false
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    // repo_name - computed: true, optional: false, required: false
    get repoName() {
        return this.getStringAttribute('repo_name');
    }
    get substitutions() {
        return this._substitutions;
    }
    // tag_name - computed: true, optional: false, required: false
    get tagName() {
        return this.getStringAttribute('tag_name');
    }
}
exports.DataGoogleCloudbuildTriggerBuildSourceRepoSourceOutputReference = DataGoogleCloudbuildTriggerBuildSourceRepoSourceOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildSourceRepoSourceOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildSourceRepoSourceOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildSourceRepoSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildSourceRepoSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildSourceRepoSourceList = DataGoogleCloudbuildTriggerBuildSourceRepoSourceList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildSourceRepoSourceList[_7] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildSourceRepoSourceList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildSourceStorageSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildSourceStorageSourceToTerraform = dataGoogleCloudbuildTriggerBuildSourceStorageSourceToTerraform;
function dataGoogleCloudbuildTriggerBuildSourceStorageSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildSourceStorageSourceToHclTerraform = dataGoogleCloudbuildTriggerBuildSourceStorageSourceToHclTerraform;
class DataGoogleCloudbuildTriggerBuildSourceStorageSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getStringAttribute('generation');
    }
    // object - computed: true, optional: false, required: false
    get object() {
        return this.getStringAttribute('object');
    }
}
exports.DataGoogleCloudbuildTriggerBuildSourceStorageSourceOutputReference = DataGoogleCloudbuildTriggerBuildSourceStorageSourceOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildSourceStorageSourceOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildSourceStorageSourceOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildSourceStorageSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildSourceStorageSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildSourceStorageSourceList = DataGoogleCloudbuildTriggerBuildSourceStorageSourceList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildSourceStorageSourceList[_9] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildSourceStorageSourceList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildSourceToTerraform = dataGoogleCloudbuildTriggerBuildSourceToTerraform;
function dataGoogleCloudbuildTriggerBuildSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildSourceToHclTerraform = dataGoogleCloudbuildTriggerBuildSourceToHclTerraform;
class DataGoogleCloudbuildTriggerBuildSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // repo_source - computed: true, optional: false, required: false
        this._repoSource = new DataGoogleCloudbuildTriggerBuildSourceRepoSourceList(this, "repo_source", false);
        // storage_source - computed: true, optional: false, required: false
        this._storageSource = new DataGoogleCloudbuildTriggerBuildSourceStorageSourceList(this, "storage_source", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get repoSource() {
        return this._repoSource;
    }
    get storageSource() {
        return this._storageSource;
    }
}
exports.DataGoogleCloudbuildTriggerBuildSourceOutputReference = DataGoogleCloudbuildTriggerBuildSourceOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildSourceOutputReference[_10] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildSourceOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildSourceList = DataGoogleCloudbuildTriggerBuildSourceList;
_11 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildSourceList[_11] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildSourceList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildStepVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildStepVolumesToTerraform = dataGoogleCloudbuildTriggerBuildStepVolumesToTerraform;
function dataGoogleCloudbuildTriggerBuildStepVolumesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildStepVolumesToHclTerraform = dataGoogleCloudbuildTriggerBuildStepVolumesToHclTerraform;
class DataGoogleCloudbuildTriggerBuildStepVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
}
exports.DataGoogleCloudbuildTriggerBuildStepVolumesOutputReference = DataGoogleCloudbuildTriggerBuildStepVolumesOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildStepVolumesOutputReference[_12] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildStepVolumesOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildStepVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildStepVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildStepVolumesList = DataGoogleCloudbuildTriggerBuildStepVolumesList;
_13 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildStepVolumesList[_13] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildStepVolumesList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildStepToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildStepToTerraform = dataGoogleCloudbuildTriggerBuildStepToTerraform;
function dataGoogleCloudbuildTriggerBuildStepToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildStepToHclTerraform = dataGoogleCloudbuildTriggerBuildStepToHclTerraform;
class DataGoogleCloudbuildTriggerBuildStepOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // volumes - computed: true, optional: false, required: false
        this._volumes = new DataGoogleCloudbuildTriggerBuildStepVolumesList(this, "volumes", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allow_exit_codes - computed: true, optional: false, required: false
    get allowExitCodes() {
        return this.getNumberListAttribute('allow_exit_codes');
    }
    // allow_failure - computed: true, optional: false, required: false
    get allowFailure() {
        return this.getBooleanAttribute('allow_failure');
    }
    // args - computed: true, optional: false, required: false
    get args() {
        return this.getListAttribute('args');
    }
    // dir - computed: true, optional: false, required: false
    get dir() {
        return this.getStringAttribute('dir');
    }
    // entrypoint - computed: true, optional: false, required: false
    get entrypoint() {
        return this.getStringAttribute('entrypoint');
    }
    // env - computed: true, optional: false, required: false
    get env() {
        return this.getListAttribute('env');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // script - computed: true, optional: false, required: false
    get script() {
        return this.getStringAttribute('script');
    }
    // secret_env - computed: true, optional: false, required: false
    get secretEnv() {
        return this.getListAttribute('secret_env');
    }
    // timeout - computed: true, optional: false, required: false
    get timeout() {
        return this.getStringAttribute('timeout');
    }
    // timing - computed: true, optional: false, required: false
    get timing() {
        return this.getStringAttribute('timing');
    }
    get volumes() {
        return this._volumes;
    }
    // wait_for - computed: true, optional: false, required: false
    get waitFor() {
        return this.getListAttribute('wait_for');
    }
}
exports.DataGoogleCloudbuildTriggerBuildStepOutputReference = DataGoogleCloudbuildTriggerBuildStepOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildStepOutputReference[_14] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildStepOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildStepList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildStepOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildStepList = DataGoogleCloudbuildTriggerBuildStepList;
_15 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildStepList[_15] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildStepList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerBuildToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerBuildToTerraform = dataGoogleCloudbuildTriggerBuildToTerraform;
function dataGoogleCloudbuildTriggerBuildToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerBuildToHclTerraform = dataGoogleCloudbuildTriggerBuildToHclTerraform;
class DataGoogleCloudbuildTriggerBuildOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // artifacts - computed: true, optional: false, required: false
        this._artifacts = new DataGoogleCloudbuildTriggerBuildArtifactsList(this, "artifacts", false);
        // available_secrets - computed: true, optional: false, required: false
        this._availableSecrets = new DataGoogleCloudbuildTriggerBuildAvailableSecretsList(this, "available_secrets", false);
        // options - computed: true, optional: false, required: false
        this._options = new DataGoogleCloudbuildTriggerBuildOptionsList(this, "options", false);
        // secret - computed: true, optional: false, required: false
        this._secret = new DataGoogleCloudbuildTriggerBuildSecretList(this, "secret", false);
        // source - computed: true, optional: false, required: false
        this._source = new DataGoogleCloudbuildTriggerBuildSourceList(this, "source", false);
        // step - computed: true, optional: false, required: false
        this._step = new DataGoogleCloudbuildTriggerBuildStepList(this, "step", false);
        // substitutions - computed: true, optional: false, required: false
        this._substitutions = new cdktf.StringMap(this, "substitutions");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get artifacts() {
        return this._artifacts;
    }
    get availableSecrets() {
        return this._availableSecrets;
    }
    // images - computed: true, optional: false, required: false
    get images() {
        return this.getListAttribute('images');
    }
    // logs_bucket - computed: true, optional: false, required: false
    get logsBucket() {
        return this.getStringAttribute('logs_bucket');
    }
    get options() {
        return this._options;
    }
    // queue_ttl - computed: true, optional: false, required: false
    get queueTtl() {
        return this.getStringAttribute('queue_ttl');
    }
    get secret() {
        return this._secret;
    }
    get source() {
        return this._source;
    }
    get step() {
        return this._step;
    }
    get substitutions() {
        return this._substitutions;
    }
    // tags - computed: true, optional: false, required: false
    get tags() {
        return this.getListAttribute('tags');
    }
    // timeout - computed: true, optional: false, required: false
    get timeout() {
        return this.getStringAttribute('timeout');
    }
}
exports.DataGoogleCloudbuildTriggerBuildOutputReference = DataGoogleCloudbuildTriggerBuildOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildOutputReference[_16] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerBuildList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerBuildOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerBuildList = DataGoogleCloudbuildTriggerBuildList;
_17 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerBuildList[_17] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerBuildList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerGitFileSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerGitFileSourceToTerraform = dataGoogleCloudbuildTriggerGitFileSourceToTerraform;
function dataGoogleCloudbuildTriggerGitFileSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerGitFileSourceToHclTerraform = dataGoogleCloudbuildTriggerGitFileSourceToHclTerraform;
class DataGoogleCloudbuildTriggerGitFileSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bitbucket_server_config - computed: true, optional: false, required: false
    get bitbucketServerConfig() {
        return this.getStringAttribute('bitbucket_server_config');
    }
    // github_enterprise_config - computed: true, optional: false, required: false
    get githubEnterpriseConfig() {
        return this.getStringAttribute('github_enterprise_config');
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // repo_type - computed: true, optional: false, required: false
    get repoType() {
        return this.getStringAttribute('repo_type');
    }
    // repository - computed: true, optional: false, required: false
    get repository() {
        return this.getStringAttribute('repository');
    }
    // revision - computed: true, optional: false, required: false
    get revision() {
        return this.getStringAttribute('revision');
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
}
exports.DataGoogleCloudbuildTriggerGitFileSourceOutputReference = DataGoogleCloudbuildTriggerGitFileSourceOutputReference;
_18 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerGitFileSourceOutputReference[_18] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerGitFileSourceOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerGitFileSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerGitFileSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerGitFileSourceList = DataGoogleCloudbuildTriggerGitFileSourceList;
_19 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerGitFileSourceList[_19] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerGitFileSourceList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerGithubPullRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerGithubPullRequestToTerraform = dataGoogleCloudbuildTriggerGithubPullRequestToTerraform;
function dataGoogleCloudbuildTriggerGithubPullRequestToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerGithubPullRequestToHclTerraform = dataGoogleCloudbuildTriggerGithubPullRequestToHclTerraform;
class DataGoogleCloudbuildTriggerGithubPullRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch - computed: true, optional: false, required: false
    get branch() {
        return this.getStringAttribute('branch');
    }
    // comment_control - computed: true, optional: false, required: false
    get commentControl() {
        return this.getStringAttribute('comment_control');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
}
exports.DataGoogleCloudbuildTriggerGithubPullRequestOutputReference = DataGoogleCloudbuildTriggerGithubPullRequestOutputReference;
_20 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerGithubPullRequestOutputReference[_20] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerGithubPullRequestOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerGithubPullRequestList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerGithubPullRequestOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerGithubPullRequestList = DataGoogleCloudbuildTriggerGithubPullRequestList;
_21 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerGithubPullRequestList[_21] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerGithubPullRequestList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerGithubPushToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerGithubPushToTerraform = dataGoogleCloudbuildTriggerGithubPushToTerraform;
function dataGoogleCloudbuildTriggerGithubPushToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerGithubPushToHclTerraform = dataGoogleCloudbuildTriggerGithubPushToHclTerraform;
class DataGoogleCloudbuildTriggerGithubPushOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch - computed: true, optional: false, required: false
    get branch() {
        return this.getStringAttribute('branch');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    // tag - computed: true, optional: false, required: false
    get tag() {
        return this.getStringAttribute('tag');
    }
}
exports.DataGoogleCloudbuildTriggerGithubPushOutputReference = DataGoogleCloudbuildTriggerGithubPushOutputReference;
_22 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerGithubPushOutputReference[_22] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerGithubPushOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerGithubPushList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerGithubPushOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerGithubPushList = DataGoogleCloudbuildTriggerGithubPushList;
_23 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerGithubPushList[_23] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerGithubPushList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerGithubToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerGithubToTerraform = dataGoogleCloudbuildTriggerGithubToTerraform;
function dataGoogleCloudbuildTriggerGithubToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerGithubToHclTerraform = dataGoogleCloudbuildTriggerGithubToHclTerraform;
class DataGoogleCloudbuildTriggerGithubOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // pull_request - computed: true, optional: false, required: false
        this._pullRequest = new DataGoogleCloudbuildTriggerGithubPullRequestList(this, "pull_request", false);
        // push - computed: true, optional: false, required: false
        this._push = new DataGoogleCloudbuildTriggerGithubPushList(this, "push", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enterprise_config_resource_name - computed: true, optional: false, required: false
    get enterpriseConfigResourceName() {
        return this.getStringAttribute('enterprise_config_resource_name');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // owner - computed: true, optional: false, required: false
    get owner() {
        return this.getStringAttribute('owner');
    }
    get pullRequest() {
        return this._pullRequest;
    }
    get push() {
        return this._push;
    }
}
exports.DataGoogleCloudbuildTriggerGithubOutputReference = DataGoogleCloudbuildTriggerGithubOutputReference;
_24 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerGithubOutputReference[_24] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerGithubOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerGithubList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerGithubOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerGithubList = DataGoogleCloudbuildTriggerGithubList;
_25 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerGithubList[_25] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerGithubList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerPubsubConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerPubsubConfigToTerraform = dataGoogleCloudbuildTriggerPubsubConfigToTerraform;
function dataGoogleCloudbuildTriggerPubsubConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerPubsubConfigToHclTerraform = dataGoogleCloudbuildTriggerPubsubConfigToHclTerraform;
class DataGoogleCloudbuildTriggerPubsubConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // service_account_email - computed: true, optional: false, required: false
    get serviceAccountEmail() {
        return this.getStringAttribute('service_account_email');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // subscription - computed: true, optional: false, required: false
    get subscription() {
        return this.getStringAttribute('subscription');
    }
    // topic - computed: true, optional: false, required: false
    get topic() {
        return this.getStringAttribute('topic');
    }
}
exports.DataGoogleCloudbuildTriggerPubsubConfigOutputReference = DataGoogleCloudbuildTriggerPubsubConfigOutputReference;
_26 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerPubsubConfigOutputReference[_26] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerPubsubConfigOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerPubsubConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerPubsubConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerPubsubConfigList = DataGoogleCloudbuildTriggerPubsubConfigList;
_27 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerPubsubConfigList[_27] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerPubsubConfigList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestToTerraform = dataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestToTerraform;
function dataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestToHclTerraform = dataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestToHclTerraform;
class DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch - computed: true, optional: false, required: false
    get branch() {
        return this.getStringAttribute('branch');
    }
    // comment_control - computed: true, optional: false, required: false
    get commentControl() {
        return this.getStringAttribute('comment_control');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
}
exports.DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestOutputReference = DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestOutputReference;
_28 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestOutputReference[_28] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestList = DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestList;
_29 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestList[_29] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerRepositoryEventConfigPushToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerRepositoryEventConfigPushToTerraform = dataGoogleCloudbuildTriggerRepositoryEventConfigPushToTerraform;
function dataGoogleCloudbuildTriggerRepositoryEventConfigPushToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerRepositoryEventConfigPushToHclTerraform = dataGoogleCloudbuildTriggerRepositoryEventConfigPushToHclTerraform;
class DataGoogleCloudbuildTriggerRepositoryEventConfigPushOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch - computed: true, optional: false, required: false
    get branch() {
        return this.getStringAttribute('branch');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    // tag - computed: true, optional: false, required: false
    get tag() {
        return this.getStringAttribute('tag');
    }
}
exports.DataGoogleCloudbuildTriggerRepositoryEventConfigPushOutputReference = DataGoogleCloudbuildTriggerRepositoryEventConfigPushOutputReference;
_30 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerRepositoryEventConfigPushOutputReference[_30] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerRepositoryEventConfigPushOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerRepositoryEventConfigPushList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerRepositoryEventConfigPushOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerRepositoryEventConfigPushList = DataGoogleCloudbuildTriggerRepositoryEventConfigPushList;
_31 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerRepositoryEventConfigPushList[_31] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerRepositoryEventConfigPushList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerRepositoryEventConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerRepositoryEventConfigToTerraform = dataGoogleCloudbuildTriggerRepositoryEventConfigToTerraform;
function dataGoogleCloudbuildTriggerRepositoryEventConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerRepositoryEventConfigToHclTerraform = dataGoogleCloudbuildTriggerRepositoryEventConfigToHclTerraform;
class DataGoogleCloudbuildTriggerRepositoryEventConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // pull_request - computed: true, optional: false, required: false
        this._pullRequest = new DataGoogleCloudbuildTriggerRepositoryEventConfigPullRequestList(this, "pull_request", false);
        // push - computed: true, optional: false, required: false
        this._push = new DataGoogleCloudbuildTriggerRepositoryEventConfigPushList(this, "push", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get pullRequest() {
        return this._pullRequest;
    }
    get push() {
        return this._push;
    }
    // repository - computed: true, optional: false, required: false
    get repository() {
        return this.getStringAttribute('repository');
    }
}
exports.DataGoogleCloudbuildTriggerRepositoryEventConfigOutputReference = DataGoogleCloudbuildTriggerRepositoryEventConfigOutputReference;
_32 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerRepositoryEventConfigOutputReference[_32] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerRepositoryEventConfigOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerRepositoryEventConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerRepositoryEventConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerRepositoryEventConfigList = DataGoogleCloudbuildTriggerRepositoryEventConfigList;
_33 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerRepositoryEventConfigList[_33] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerRepositoryEventConfigList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerSourceToBuildToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerSourceToBuildToTerraform = dataGoogleCloudbuildTriggerSourceToBuildToTerraform;
function dataGoogleCloudbuildTriggerSourceToBuildToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerSourceToBuildToHclTerraform = dataGoogleCloudbuildTriggerSourceToBuildToHclTerraform;
class DataGoogleCloudbuildTriggerSourceToBuildOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bitbucket_server_config - computed: true, optional: false, required: false
    get bitbucketServerConfig() {
        return this.getStringAttribute('bitbucket_server_config');
    }
    // github_enterprise_config - computed: true, optional: false, required: false
    get githubEnterpriseConfig() {
        return this.getStringAttribute('github_enterprise_config');
    }
    // ref - computed: true, optional: false, required: false
    get ref() {
        return this.getStringAttribute('ref');
    }
    // repo_type - computed: true, optional: false, required: false
    get repoType() {
        return this.getStringAttribute('repo_type');
    }
    // repository - computed: true, optional: false, required: false
    get repository() {
        return this.getStringAttribute('repository');
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
}
exports.DataGoogleCloudbuildTriggerSourceToBuildOutputReference = DataGoogleCloudbuildTriggerSourceToBuildOutputReference;
_34 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerSourceToBuildOutputReference[_34] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerSourceToBuildOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerSourceToBuildList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerSourceToBuildOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerSourceToBuildList = DataGoogleCloudbuildTriggerSourceToBuildList;
_35 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerSourceToBuildList[_35] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerSourceToBuildList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerTriggerTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerTriggerTemplateToTerraform = dataGoogleCloudbuildTriggerTriggerTemplateToTerraform;
function dataGoogleCloudbuildTriggerTriggerTemplateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerTriggerTemplateToHclTerraform = dataGoogleCloudbuildTriggerTriggerTemplateToHclTerraform;
class DataGoogleCloudbuildTriggerTriggerTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch_name - computed: true, optional: false, required: false
    get branchName() {
        return this.getStringAttribute('branch_name');
    }
    // commit_sha - computed: true, optional: false, required: false
    get commitSha() {
        return this.getStringAttribute('commit_sha');
    }
    // dir - computed: true, optional: false, required: false
    get dir() {
        return this.getStringAttribute('dir');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    // project_id - computed: true, optional: false, required: false
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    // repo_name - computed: true, optional: false, required: false
    get repoName() {
        return this.getStringAttribute('repo_name');
    }
    // tag_name - computed: true, optional: false, required: false
    get tagName() {
        return this.getStringAttribute('tag_name');
    }
}
exports.DataGoogleCloudbuildTriggerTriggerTemplateOutputReference = DataGoogleCloudbuildTriggerTriggerTemplateOutputReference;
_36 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerTriggerTemplateOutputReference[_36] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerTriggerTemplateOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerTriggerTemplateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerTriggerTemplateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerTriggerTemplateList = DataGoogleCloudbuildTriggerTriggerTemplateList;
_37 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerTriggerTemplateList[_37] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerTriggerTemplateList", version: "14.13.1" };
function dataGoogleCloudbuildTriggerWebhookConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudbuildTriggerWebhookConfigToTerraform = dataGoogleCloudbuildTriggerWebhookConfigToTerraform;
function dataGoogleCloudbuildTriggerWebhookConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudbuildTriggerWebhookConfigToHclTerraform = dataGoogleCloudbuildTriggerWebhookConfigToHclTerraform;
class DataGoogleCloudbuildTriggerWebhookConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // secret - computed: true, optional: false, required: false
    get secret() {
        return this.getStringAttribute('secret');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.DataGoogleCloudbuildTriggerWebhookConfigOutputReference = DataGoogleCloudbuildTriggerWebhookConfigOutputReference;
_38 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerWebhookConfigOutputReference[_38] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerWebhookConfigOutputReference", version: "14.13.1" };
class DataGoogleCloudbuildTriggerWebhookConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudbuildTriggerWebhookConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudbuildTriggerWebhookConfigList = DataGoogleCloudbuildTriggerWebhookConfigList;
_39 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTriggerWebhookConfigList[_39] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTriggerWebhookConfigList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloudbuild_trigger google_cloudbuild_trigger}
*/
class DataGoogleCloudbuildTrigger extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleCloudbuildTrigger resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleCloudbuildTrigger to import
    * @param importFromId The id of the existing DataGoogleCloudbuildTrigger that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloudbuild_trigger#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleCloudbuildTrigger to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_cloudbuild_trigger", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloudbuild_trigger google_cloudbuild_trigger} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleCloudbuildTriggerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloudbuild_trigger',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // approval_config - computed: true, optional: false, required: false
        this._approvalConfig = new DataGoogleCloudbuildTriggerApprovalConfigList(this, "approval_config", false);
        // bitbucket_server_trigger_config - computed: true, optional: false, required: false
        this._bitbucketServerTriggerConfig = new DataGoogleCloudbuildTriggerBitbucketServerTriggerConfigList(this, "bitbucket_server_trigger_config", false);
        // build - computed: true, optional: false, required: false
        this._build = new DataGoogleCloudbuildTriggerBuildList(this, "build", false);
        // git_file_source - computed: true, optional: false, required: false
        this._gitFileSource = new DataGoogleCloudbuildTriggerGitFileSourceList(this, "git_file_source", false);
        // github - computed: true, optional: false, required: false
        this._github = new DataGoogleCloudbuildTriggerGithubList(this, "github", false);
        // pubsub_config - computed: true, optional: false, required: false
        this._pubsubConfig = new DataGoogleCloudbuildTriggerPubsubConfigList(this, "pubsub_config", false);
        // repository_event_config - computed: true, optional: false, required: false
        this._repositoryEventConfig = new DataGoogleCloudbuildTriggerRepositoryEventConfigList(this, "repository_event_config", false);
        // source_to_build - computed: true, optional: false, required: false
        this._sourceToBuild = new DataGoogleCloudbuildTriggerSourceToBuildList(this, "source_to_build", false);
        // substitutions - computed: true, optional: false, required: false
        this._substitutions = new cdktf.StringMap(this, "substitutions");
        // trigger_template - computed: true, optional: false, required: false
        this._triggerTemplate = new DataGoogleCloudbuildTriggerTriggerTemplateList(this, "trigger_template", false);
        // webhook_config - computed: true, optional: false, required: false
        this._webhookConfig = new DataGoogleCloudbuildTriggerWebhookConfigList(this, "webhook_config", false);
        this._id = config.id;
        this._location = config.location;
        this._project = config.project;
        this._triggerId = config.triggerId;
    }
    get approvalConfig() {
        return this._approvalConfig;
    }
    get bitbucketServerTriggerConfig() {
        return this._bitbucketServerTriggerConfig;
    }
    get buildAttribute() {
        return this._build;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // disabled - computed: true, optional: false, required: false
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    // filename - computed: true, optional: false, required: false
    get filename() {
        return this.getStringAttribute('filename');
    }
    // filter - computed: true, optional: false, required: false
    get filter() {
        return this.getStringAttribute('filter');
    }
    get gitFileSource() {
        return this._gitFileSource;
    }
    get github() {
        return this._github;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // ignored_files - computed: true, optional: false, required: false
    get ignoredFiles() {
        return this.getListAttribute('ignored_files');
    }
    // include_build_logs - computed: true, optional: false, required: false
    get includeBuildLogs() {
        return this.getStringAttribute('include_build_logs');
    }
    // included_files - computed: true, optional: false, required: false
    get includedFiles() {
        return this.getListAttribute('included_files');
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get pubsubConfig() {
        return this._pubsubConfig;
    }
    get repositoryEventConfig() {
        return this._repositoryEventConfig;
    }
    // service_account - computed: true, optional: false, required: false
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    get sourceToBuild() {
        return this._sourceToBuild;
    }
    get substitutions() {
        return this._substitutions;
    }
    // tags - computed: true, optional: false, required: false
    get tags() {
        return this.getListAttribute('tags');
    }
    get triggerId() {
        return this.getStringAttribute('trigger_id');
    }
    set triggerId(value) {
        this._triggerId = value;
    }
    // Temporarily expose input value. Use with caution.
    get triggerIdInput() {
        return this._triggerId;
    }
    get triggerTemplate() {
        return this._triggerTemplate;
    }
    get webhookConfig() {
        return this._webhookConfig;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            trigger_id: cdktf.stringToTerraform(this._triggerId),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            trigger_id: {
                value: cdktf.stringToHclTerraform(this._triggerId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleCloudbuildTrigger = DataGoogleCloudbuildTrigger;
_40 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudbuildTrigger[_40] = { fqn: "@cdktf/provider-google.dataGoogleCloudbuildTrigger.DataGoogleCloudbuildTrigger", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleCloudbuildTrigger.tfResourceType = "google_cloudbuild_trigger";
