// @cdktf/provider-google/lib/apphub-service/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApphubService = exports.ApphubServiceTimeoutsOutputReference = exports.apphubServiceTimeoutsToHclTerraform = exports.apphubServiceTimeoutsToTerraform = exports.ApphubServiceAttributesOutputReference = exports.apphubServiceAttributesToHclTerraform = exports.apphubServiceAttributesToTerraform = exports.ApphubServiceAttributesOperatorOwnersList = exports.ApphubServiceAttributesOperatorOwnersOutputReference = exports.apphubServiceAttributesOperatorOwnersToHclTerraform = exports.apphubServiceAttributesOperatorOwnersToTerraform = exports.ApphubServiceAttributesEnvironmentOutputReference = exports.apphubServiceAttributesEnvironmentToHclTerraform = exports.apphubServiceAttributesEnvironmentToTerraform = exports.ApphubServiceAttributesDeveloperOwnersList = exports.ApphubServiceAttributesDeveloperOwnersOutputReference = exports.apphubServiceAttributesDeveloperOwnersToHclTerraform = exports.apphubServiceAttributesDeveloperOwnersToTerraform = exports.ApphubServiceAttributesCriticalityOutputReference = exports.apphubServiceAttributesCriticalityToHclTerraform = exports.apphubServiceAttributesCriticalityToTerraform = exports.ApphubServiceAttributesBusinessOwnersList = exports.ApphubServiceAttributesBusinessOwnersOutputReference = exports.apphubServiceAttributesBusinessOwnersToHclTerraform = exports.apphubServiceAttributesBusinessOwnersToTerraform = exports.ApphubServiceServiceReferenceList = exports.ApphubServiceServiceReferenceOutputReference = exports.apphubServiceServiceReferenceToHclTerraform = exports.apphubServiceServiceReferenceToTerraform = exports.ApphubServiceServicePropertiesList = exports.ApphubServiceServicePropertiesOutputReference = exports.apphubServiceServicePropertiesToHclTerraform = exports.apphubServiceServicePropertiesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apphubServiceServicePropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.apphubServiceServicePropertiesToTerraform = apphubServiceServicePropertiesToTerraform;
function apphubServiceServicePropertiesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.apphubServiceServicePropertiesToHclTerraform = apphubServiceServicePropertiesToHclTerraform;
class ApphubServiceServicePropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // gcp_project - computed: true, optional: false, required: false
    get gcpProject() {
        return this.getStringAttribute('gcp_project');
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
}
exports.ApphubServiceServicePropertiesOutputReference = ApphubServiceServicePropertiesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApphubServiceServicePropertiesOutputReference[_a] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceServicePropertiesOutputReference", version: "14.13.1" };
class ApphubServiceServicePropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApphubServiceServicePropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApphubServiceServicePropertiesList = ApphubServiceServicePropertiesList;
_b = JSII_RTTI_SYMBOL_1;
ApphubServiceServicePropertiesList[_b] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceServicePropertiesList", version: "14.13.1" };
function apphubServiceServiceReferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.apphubServiceServiceReferenceToTerraform = apphubServiceServiceReferenceToTerraform;
function apphubServiceServiceReferenceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.apphubServiceServiceReferenceToHclTerraform = apphubServiceServiceReferenceToHclTerraform;
class ApphubServiceServiceReferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
}
exports.ApphubServiceServiceReferenceOutputReference = ApphubServiceServiceReferenceOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ApphubServiceServiceReferenceOutputReference[_c] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceServiceReferenceOutputReference", version: "14.13.1" };
class ApphubServiceServiceReferenceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApphubServiceServiceReferenceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApphubServiceServiceReferenceList = ApphubServiceServiceReferenceList;
_d = JSII_RTTI_SYMBOL_1;
ApphubServiceServiceReferenceList[_d] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceServiceReferenceList", version: "14.13.1" };
function apphubServiceAttributesBusinessOwnersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        display_name: cdktf.stringToTerraform(struct.displayName),
        email: cdktf.stringToTerraform(struct.email),
    };
}
exports.apphubServiceAttributesBusinessOwnersToTerraform = apphubServiceAttributesBusinessOwnersToTerraform;
function apphubServiceAttributesBusinessOwnersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        display_name: {
            value: cdktf.stringToHclTerraform(struct.displayName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        email: {
            value: cdktf.stringToHclTerraform(struct.email),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apphubServiceAttributesBusinessOwnersToHclTerraform = apphubServiceAttributesBusinessOwnersToHclTerraform;
class ApphubServiceAttributesBusinessOwnersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._displayName !== undefined) {
            hasAnyValues = true;
            internalValueResult.displayName = this._displayName;
        }
        if (this._email !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._displayName = undefined;
            this._email = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._displayName = value.displayName;
            this._email = value.email;
        }
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
}
exports.ApphubServiceAttributesBusinessOwnersOutputReference = ApphubServiceAttributesBusinessOwnersOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesBusinessOwnersOutputReference[_e] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesBusinessOwnersOutputReference", version: "14.13.1" };
class ApphubServiceAttributesBusinessOwnersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApphubServiceAttributesBusinessOwnersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApphubServiceAttributesBusinessOwnersList = ApphubServiceAttributesBusinessOwnersList;
_f = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesBusinessOwnersList[_f] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesBusinessOwnersList", version: "14.13.1" };
function apphubServiceAttributesCriticalityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.apphubServiceAttributesCriticalityToTerraform = apphubServiceAttributesCriticalityToTerraform;
function apphubServiceAttributesCriticalityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apphubServiceAttributesCriticalityToHclTerraform = apphubServiceAttributesCriticalityToHclTerraform;
class ApphubServiceAttributesCriticalityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ApphubServiceAttributesCriticalityOutputReference = ApphubServiceAttributesCriticalityOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesCriticalityOutputReference[_g] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesCriticalityOutputReference", version: "14.13.1" };
function apphubServiceAttributesDeveloperOwnersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        display_name: cdktf.stringToTerraform(struct.displayName),
        email: cdktf.stringToTerraform(struct.email),
    };
}
exports.apphubServiceAttributesDeveloperOwnersToTerraform = apphubServiceAttributesDeveloperOwnersToTerraform;
function apphubServiceAttributesDeveloperOwnersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        display_name: {
            value: cdktf.stringToHclTerraform(struct.displayName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        email: {
            value: cdktf.stringToHclTerraform(struct.email),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apphubServiceAttributesDeveloperOwnersToHclTerraform = apphubServiceAttributesDeveloperOwnersToHclTerraform;
class ApphubServiceAttributesDeveloperOwnersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._displayName !== undefined) {
            hasAnyValues = true;
            internalValueResult.displayName = this._displayName;
        }
        if (this._email !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._displayName = undefined;
            this._email = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._displayName = value.displayName;
            this._email = value.email;
        }
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
}
exports.ApphubServiceAttributesDeveloperOwnersOutputReference = ApphubServiceAttributesDeveloperOwnersOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesDeveloperOwnersOutputReference[_h] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesDeveloperOwnersOutputReference", version: "14.13.1" };
class ApphubServiceAttributesDeveloperOwnersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApphubServiceAttributesDeveloperOwnersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApphubServiceAttributesDeveloperOwnersList = ApphubServiceAttributesDeveloperOwnersList;
_j = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesDeveloperOwnersList[_j] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesDeveloperOwnersList", version: "14.13.1" };
function apphubServiceAttributesEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.apphubServiceAttributesEnvironmentToTerraform = apphubServiceAttributesEnvironmentToTerraform;
function apphubServiceAttributesEnvironmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apphubServiceAttributesEnvironmentToHclTerraform = apphubServiceAttributesEnvironmentToHclTerraform;
class ApphubServiceAttributesEnvironmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._type = value.type;
        }
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ApphubServiceAttributesEnvironmentOutputReference = ApphubServiceAttributesEnvironmentOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesEnvironmentOutputReference[_k] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesEnvironmentOutputReference", version: "14.13.1" };
function apphubServiceAttributesOperatorOwnersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        display_name: cdktf.stringToTerraform(struct.displayName),
        email: cdktf.stringToTerraform(struct.email),
    };
}
exports.apphubServiceAttributesOperatorOwnersToTerraform = apphubServiceAttributesOperatorOwnersToTerraform;
function apphubServiceAttributesOperatorOwnersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        display_name: {
            value: cdktf.stringToHclTerraform(struct.displayName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        email: {
            value: cdktf.stringToHclTerraform(struct.email),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apphubServiceAttributesOperatorOwnersToHclTerraform = apphubServiceAttributesOperatorOwnersToHclTerraform;
class ApphubServiceAttributesOperatorOwnersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._displayName !== undefined) {
            hasAnyValues = true;
            internalValueResult.displayName = this._displayName;
        }
        if (this._email !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._displayName = undefined;
            this._email = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._displayName = value.displayName;
            this._email = value.email;
        }
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
}
exports.ApphubServiceAttributesOperatorOwnersOutputReference = ApphubServiceAttributesOperatorOwnersOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesOperatorOwnersOutputReference[_l] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesOperatorOwnersOutputReference", version: "14.13.1" };
class ApphubServiceAttributesOperatorOwnersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApphubServiceAttributesOperatorOwnersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApphubServiceAttributesOperatorOwnersList = ApphubServiceAttributesOperatorOwnersList;
_m = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesOperatorOwnersList[_m] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesOperatorOwnersList", version: "14.13.1" };
function apphubServiceAttributesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        business_owners: cdktf.listMapper(apphubServiceAttributesBusinessOwnersToTerraform, true)(struct.businessOwners),
        criticality: apphubServiceAttributesCriticalityToTerraform(struct.criticality),
        developer_owners: cdktf.listMapper(apphubServiceAttributesDeveloperOwnersToTerraform, true)(struct.developerOwners),
        environment: apphubServiceAttributesEnvironmentToTerraform(struct.environment),
        operator_owners: cdktf.listMapper(apphubServiceAttributesOperatorOwnersToTerraform, true)(struct.operatorOwners),
    };
}
exports.apphubServiceAttributesToTerraform = apphubServiceAttributesToTerraform;
function apphubServiceAttributesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        business_owners: {
            value: cdktf.listMapperHcl(apphubServiceAttributesBusinessOwnersToHclTerraform, true)(struct.businessOwners),
            isBlock: true,
            type: "list",
            storageClassType: "ApphubServiceAttributesBusinessOwnersList",
        },
        criticality: {
            value: apphubServiceAttributesCriticalityToHclTerraform(struct.criticality),
            isBlock: true,
            type: "list",
            storageClassType: "ApphubServiceAttributesCriticalityList",
        },
        developer_owners: {
            value: cdktf.listMapperHcl(apphubServiceAttributesDeveloperOwnersToHclTerraform, true)(struct.developerOwners),
            isBlock: true,
            type: "list",
            storageClassType: "ApphubServiceAttributesDeveloperOwnersList",
        },
        environment: {
            value: apphubServiceAttributesEnvironmentToHclTerraform(struct.environment),
            isBlock: true,
            type: "list",
            storageClassType: "ApphubServiceAttributesEnvironmentList",
        },
        operator_owners: {
            value: cdktf.listMapperHcl(apphubServiceAttributesOperatorOwnersToHclTerraform, true)(struct.operatorOwners),
            isBlock: true,
            type: "list",
            storageClassType: "ApphubServiceAttributesOperatorOwnersList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apphubServiceAttributesToHclTerraform = apphubServiceAttributesToHclTerraform;
class ApphubServiceAttributesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // business_owners - computed: false, optional: true, required: false
        this._businessOwners = new ApphubServiceAttributesBusinessOwnersList(this, "business_owners", false);
        // criticality - computed: false, optional: true, required: false
        this._criticality = new ApphubServiceAttributesCriticalityOutputReference(this, "criticality");
        // developer_owners - computed: false, optional: true, required: false
        this._developerOwners = new ApphubServiceAttributesDeveloperOwnersList(this, "developer_owners", false);
        // environment - computed: false, optional: true, required: false
        this._environment = new ApphubServiceAttributesEnvironmentOutputReference(this, "environment");
        // operator_owners - computed: false, optional: true, required: false
        this._operatorOwners = new ApphubServiceAttributesOperatorOwnersList(this, "operator_owners", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._businessOwners?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.businessOwners = this._businessOwners?.internalValue;
        }
        if (this._criticality?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.criticality = this._criticality?.internalValue;
        }
        if (this._developerOwners?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.developerOwners = this._developerOwners?.internalValue;
        }
        if (this._environment?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.environment = this._environment?.internalValue;
        }
        if (this._operatorOwners?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.operatorOwners = this._operatorOwners?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._businessOwners.internalValue = undefined;
            this._criticality.internalValue = undefined;
            this._developerOwners.internalValue = undefined;
            this._environment.internalValue = undefined;
            this._operatorOwners.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._businessOwners.internalValue = value.businessOwners;
            this._criticality.internalValue = value.criticality;
            this._developerOwners.internalValue = value.developerOwners;
            this._environment.internalValue = value.environment;
            this._operatorOwners.internalValue = value.operatorOwners;
        }
    }
    get businessOwners() {
        return this._businessOwners;
    }
    putBusinessOwners(value) {
        this._businessOwners.internalValue = value;
    }
    resetBusinessOwners() {
        this._businessOwners.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get businessOwnersInput() {
        return this._businessOwners.internalValue;
    }
    get criticality() {
        return this._criticality;
    }
    putCriticality(value) {
        this._criticality.internalValue = value;
    }
    resetCriticality() {
        this._criticality.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get criticalityInput() {
        return this._criticality.internalValue;
    }
    get developerOwners() {
        return this._developerOwners;
    }
    putDeveloperOwners(value) {
        this._developerOwners.internalValue = value;
    }
    resetDeveloperOwners() {
        this._developerOwners.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get developerOwnersInput() {
        return this._developerOwners.internalValue;
    }
    get environment() {
        return this._environment;
    }
    putEnvironment(value) {
        this._environment.internalValue = value;
    }
    resetEnvironment() {
        this._environment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment.internalValue;
    }
    get operatorOwners() {
        return this._operatorOwners;
    }
    putOperatorOwners(value) {
        this._operatorOwners.internalValue = value;
    }
    resetOperatorOwners() {
        this._operatorOwners.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get operatorOwnersInput() {
        return this._operatorOwners.internalValue;
    }
}
exports.ApphubServiceAttributesOutputReference = ApphubServiceAttributesOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ApphubServiceAttributesOutputReference[_o] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceAttributesOutputReference", version: "14.13.1" };
function apphubServiceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.apphubServiceTimeoutsToTerraform = apphubServiceTimeoutsToTerraform;
function apphubServiceTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apphubServiceTimeoutsToHclTerraform = apphubServiceTimeoutsToHclTerraform;
class ApphubServiceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ApphubServiceTimeoutsOutputReference = ApphubServiceTimeoutsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ApphubServiceTimeoutsOutputReference[_p] = { fqn: "@cdktf/provider-google.apphubService.ApphubServiceTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apphub_service google_apphub_service}
*/
class ApphubService extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ApphubService resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ApphubService to import
    * @param importFromId The id of the existing ApphubService that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apphub_service#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ApphubService to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_apphub_service", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apphub_service google_apphub_service} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApphubServiceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_apphub_service',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // service_properties - computed: true, optional: false, required: false
        this._serviceProperties = new ApphubServiceServicePropertiesList(this, "service_properties", false);
        // service_reference - computed: true, optional: false, required: false
        this._serviceReference = new ApphubServiceServiceReferenceList(this, "service_reference", false);
        // attributes - computed: false, optional: true, required: false
        this._attributes = new ApphubServiceAttributesOutputReference(this, "attributes");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ApphubServiceTimeoutsOutputReference(this, "timeouts");
        this._applicationId = config.applicationId;
        this._description = config.description;
        this._discoveredService = config.discoveredService;
        this._displayName = config.displayName;
        this._id = config.id;
        this._location = config.location;
        this._project = config.project;
        this._serviceId = config.serviceId;
        this._attributes.internalValue = config.attributes;
        this._timeouts.internalValue = config.timeouts;
    }
    get applicationId() {
        return this.getStringAttribute('application_id');
    }
    set applicationId(value) {
        this._applicationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get applicationIdInput() {
        return this._applicationId;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get discoveredService() {
        return this.getStringAttribute('discovered_service');
    }
    set discoveredService(value) {
        this._discoveredService = value;
    }
    // Temporarily expose input value. Use with caution.
    get discoveredServiceInput() {
        return this._discoveredService;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get serviceId() {
        return this.getStringAttribute('service_id');
    }
    set serviceId(value) {
        this._serviceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceIdInput() {
        return this._serviceId;
    }
    get serviceProperties() {
        return this._serviceProperties;
    }
    get serviceReference() {
        return this._serviceReference;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get attributes() {
        return this._attributes;
    }
    putAttributes(value) {
        this._attributes.internalValue = value;
    }
    resetAttributes() {
        this._attributes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attributesInput() {
        return this._attributes.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            application_id: cdktf.stringToTerraform(this._applicationId),
            description: cdktf.stringToTerraform(this._description),
            discovered_service: cdktf.stringToTerraform(this._discoveredService),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            service_id: cdktf.stringToTerraform(this._serviceId),
            attributes: apphubServiceAttributesToTerraform(this._attributes.internalValue),
            timeouts: apphubServiceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            application_id: {
                value: cdktf.stringToHclTerraform(this._applicationId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            discovered_service: {
                value: cdktf.stringToHclTerraform(this._discoveredService),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_id: {
                value: cdktf.stringToHclTerraform(this._serviceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            attributes: {
                value: apphubServiceAttributesToHclTerraform(this._attributes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ApphubServiceAttributesList",
            },
            timeouts: {
                value: apphubServiceTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ApphubServiceTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ApphubService = ApphubService;
_q = JSII_RTTI_SYMBOL_1;
ApphubService[_q] = { fqn: "@cdktf/provider-google.apphubService.ApphubService", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ApphubService.tfResourceType = "google_apphub_service";
