// @cdktf/provider-google/lib/bigquery-routine/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigqueryRoutine = exports.BigqueryRoutineTimeoutsOutputReference = exports.bigqueryRoutineTimeoutsToHclTerraform = exports.bigqueryRoutineTimeoutsToTerraform = exports.BigqueryRoutineSparkOptionsOutputReference = exports.bigqueryRoutineSparkOptionsToHclTerraform = exports.bigqueryRoutineSparkOptionsToTerraform = exports.BigqueryRoutineRemoteFunctionOptionsOutputReference = exports.bigqueryRoutineRemoteFunctionOptionsToHclTerraform = exports.bigqueryRoutineRemoteFunctionOptionsToTerraform = exports.BigqueryRoutineArgumentsList = exports.BigqueryRoutineArgumentsOutputReference = exports.bigqueryRoutineArgumentsToHclTerraform = exports.bigqueryRoutineArgumentsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function bigqueryRoutineArgumentsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        argument_kind: cdktf.stringToTerraform(struct.argumentKind),
        data_type: cdktf.stringToTerraform(struct.dataType),
        mode: cdktf.stringToTerraform(struct.mode),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.bigqueryRoutineArgumentsToTerraform = bigqueryRoutineArgumentsToTerraform;
function bigqueryRoutineArgumentsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        argument_kind: {
            value: cdktf.stringToHclTerraform(struct.argumentKind),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        data_type: {
            value: cdktf.stringToHclTerraform(struct.dataType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryRoutineArgumentsToHclTerraform = bigqueryRoutineArgumentsToHclTerraform;
class BigqueryRoutineArgumentsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._argumentKind !== undefined) {
            hasAnyValues = true;
            internalValueResult.argumentKind = this._argumentKind;
        }
        if (this._dataType !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataType = this._dataType;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._argumentKind = undefined;
            this._dataType = undefined;
            this._mode = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._argumentKind = value.argumentKind;
            this._dataType = value.dataType;
            this._mode = value.mode;
            this._name = value.name;
        }
    }
    get argumentKind() {
        return this.getStringAttribute('argument_kind');
    }
    set argumentKind(value) {
        this._argumentKind = value;
    }
    resetArgumentKind() {
        this._argumentKind = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argumentKindInput() {
        return this._argumentKind;
    }
    get dataType() {
        return this.getStringAttribute('data_type');
    }
    set dataType(value) {
        this._dataType = value;
    }
    resetDataType() {
        this._dataType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataTypeInput() {
        return this._dataType;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.BigqueryRoutineArgumentsOutputReference = BigqueryRoutineArgumentsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BigqueryRoutineArgumentsOutputReference[_a] = { fqn: "@cdktf/provider-google.bigqueryRoutine.BigqueryRoutineArgumentsOutputReference", version: "14.13.1" };
class BigqueryRoutineArgumentsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new BigqueryRoutineArgumentsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.BigqueryRoutineArgumentsList = BigqueryRoutineArgumentsList;
_b = JSII_RTTI_SYMBOL_1;
BigqueryRoutineArgumentsList[_b] = { fqn: "@cdktf/provider-google.bigqueryRoutine.BigqueryRoutineArgumentsList", version: "14.13.1" };
function bigqueryRoutineRemoteFunctionOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        connection: cdktf.stringToTerraform(struct.connection),
        endpoint: cdktf.stringToTerraform(struct.endpoint),
        max_batching_rows: cdktf.stringToTerraform(struct.maxBatchingRows),
        user_defined_context: cdktf.hashMapper(cdktf.stringToTerraform)(struct.userDefinedContext),
    };
}
exports.bigqueryRoutineRemoteFunctionOptionsToTerraform = bigqueryRoutineRemoteFunctionOptionsToTerraform;
function bigqueryRoutineRemoteFunctionOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        connection: {
            value: cdktf.stringToHclTerraform(struct.connection),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        endpoint: {
            value: cdktf.stringToHclTerraform(struct.endpoint),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        max_batching_rows: {
            value: cdktf.stringToHclTerraform(struct.maxBatchingRows),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_defined_context: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.userDefinedContext),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryRoutineRemoteFunctionOptionsToHclTerraform = bigqueryRoutineRemoteFunctionOptionsToHclTerraform;
class BigqueryRoutineRemoteFunctionOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connection !== undefined) {
            hasAnyValues = true;
            internalValueResult.connection = this._connection;
        }
        if (this._endpoint !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpoint = this._endpoint;
        }
        if (this._maxBatchingRows !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxBatchingRows = this._maxBatchingRows;
        }
        if (this._userDefinedContext !== undefined) {
            hasAnyValues = true;
            internalValueResult.userDefinedContext = this._userDefinedContext;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._connection = undefined;
            this._endpoint = undefined;
            this._maxBatchingRows = undefined;
            this._userDefinedContext = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._connection = value.connection;
            this._endpoint = value.endpoint;
            this._maxBatchingRows = value.maxBatchingRows;
            this._userDefinedContext = value.userDefinedContext;
        }
    }
    get connection() {
        return this.getStringAttribute('connection');
    }
    set connection(value) {
        this._connection = value;
    }
    resetConnection() {
        this._connection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionInput() {
        return this._connection;
    }
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    set endpoint(value) {
        this._endpoint = value;
    }
    resetEndpoint() {
        this._endpoint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointInput() {
        return this._endpoint;
    }
    get maxBatchingRows() {
        return this.getStringAttribute('max_batching_rows');
    }
    set maxBatchingRows(value) {
        this._maxBatchingRows = value;
    }
    resetMaxBatchingRows() {
        this._maxBatchingRows = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxBatchingRowsInput() {
        return this._maxBatchingRows;
    }
    get userDefinedContext() {
        return this.getStringMapAttribute('user_defined_context');
    }
    set userDefinedContext(value) {
        this._userDefinedContext = value;
    }
    resetUserDefinedContext() {
        this._userDefinedContext = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userDefinedContextInput() {
        return this._userDefinedContext;
    }
}
exports.BigqueryRoutineRemoteFunctionOptionsOutputReference = BigqueryRoutineRemoteFunctionOptionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
BigqueryRoutineRemoteFunctionOptionsOutputReference[_c] = { fqn: "@cdktf/provider-google.bigqueryRoutine.BigqueryRoutineRemoteFunctionOptionsOutputReference", version: "14.13.1" };
function bigqueryRoutineSparkOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        archive_uris: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.archiveUris),
        connection: cdktf.stringToTerraform(struct.connection),
        container_image: cdktf.stringToTerraform(struct.containerImage),
        file_uris: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.fileUris),
        jar_uris: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.jarUris),
        main_class: cdktf.stringToTerraform(struct.mainClass),
        main_file_uri: cdktf.stringToTerraform(struct.mainFileUri),
        properties: cdktf.hashMapper(cdktf.stringToTerraform)(struct.properties),
        py_file_uris: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.pyFileUris),
        runtime_version: cdktf.stringToTerraform(struct.runtimeVersion),
    };
}
exports.bigqueryRoutineSparkOptionsToTerraform = bigqueryRoutineSparkOptionsToTerraform;
function bigqueryRoutineSparkOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        archive_uris: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.archiveUris),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        connection: {
            value: cdktf.stringToHclTerraform(struct.connection),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        container_image: {
            value: cdktf.stringToHclTerraform(struct.containerImage),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        file_uris: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.fileUris),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        jar_uris: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.jarUris),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        main_class: {
            value: cdktf.stringToHclTerraform(struct.mainClass),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        main_file_uri: {
            value: cdktf.stringToHclTerraform(struct.mainFileUri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        properties: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.properties),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        py_file_uris: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.pyFileUris),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        runtime_version: {
            value: cdktf.stringToHclTerraform(struct.runtimeVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryRoutineSparkOptionsToHclTerraform = bigqueryRoutineSparkOptionsToHclTerraform;
class BigqueryRoutineSparkOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._archiveUris !== undefined) {
            hasAnyValues = true;
            internalValueResult.archiveUris = this._archiveUris;
        }
        if (this._connection !== undefined) {
            hasAnyValues = true;
            internalValueResult.connection = this._connection;
        }
        if (this._containerImage !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerImage = this._containerImage;
        }
        if (this._fileUris !== undefined) {
            hasAnyValues = true;
            internalValueResult.fileUris = this._fileUris;
        }
        if (this._jarUris !== undefined) {
            hasAnyValues = true;
            internalValueResult.jarUris = this._jarUris;
        }
        if (this._mainClass !== undefined) {
            hasAnyValues = true;
            internalValueResult.mainClass = this._mainClass;
        }
        if (this._mainFileUri !== undefined) {
            hasAnyValues = true;
            internalValueResult.mainFileUri = this._mainFileUri;
        }
        if (this._properties !== undefined) {
            hasAnyValues = true;
            internalValueResult.properties = this._properties;
        }
        if (this._pyFileUris !== undefined) {
            hasAnyValues = true;
            internalValueResult.pyFileUris = this._pyFileUris;
        }
        if (this._runtimeVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtimeVersion = this._runtimeVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._archiveUris = undefined;
            this._connection = undefined;
            this._containerImage = undefined;
            this._fileUris = undefined;
            this._jarUris = undefined;
            this._mainClass = undefined;
            this._mainFileUri = undefined;
            this._properties = undefined;
            this._pyFileUris = undefined;
            this._runtimeVersion = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._archiveUris = value.archiveUris;
            this._connection = value.connection;
            this._containerImage = value.containerImage;
            this._fileUris = value.fileUris;
            this._jarUris = value.jarUris;
            this._mainClass = value.mainClass;
            this._mainFileUri = value.mainFileUri;
            this._properties = value.properties;
            this._pyFileUris = value.pyFileUris;
            this._runtimeVersion = value.runtimeVersion;
        }
    }
    get archiveUris() {
        return this.getListAttribute('archive_uris');
    }
    set archiveUris(value) {
        this._archiveUris = value;
    }
    resetArchiveUris() {
        this._archiveUris = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get archiveUrisInput() {
        return this._archiveUris;
    }
    get connection() {
        return this.getStringAttribute('connection');
    }
    set connection(value) {
        this._connection = value;
    }
    resetConnection() {
        this._connection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionInput() {
        return this._connection;
    }
    get containerImage() {
        return this.getStringAttribute('container_image');
    }
    set containerImage(value) {
        this._containerImage = value;
    }
    resetContainerImage() {
        this._containerImage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerImageInput() {
        return this._containerImage;
    }
    get fileUris() {
        return this.getListAttribute('file_uris');
    }
    set fileUris(value) {
        this._fileUris = value;
    }
    resetFileUris() {
        this._fileUris = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fileUrisInput() {
        return this._fileUris;
    }
    get jarUris() {
        return this.getListAttribute('jar_uris');
    }
    set jarUris(value) {
        this._jarUris = value;
    }
    resetJarUris() {
        this._jarUris = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jarUrisInput() {
        return this._jarUris;
    }
    get mainClass() {
        return this.getStringAttribute('main_class');
    }
    set mainClass(value) {
        this._mainClass = value;
    }
    resetMainClass() {
        this._mainClass = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mainClassInput() {
        return this._mainClass;
    }
    get mainFileUri() {
        return this.getStringAttribute('main_file_uri');
    }
    set mainFileUri(value) {
        this._mainFileUri = value;
    }
    resetMainFileUri() {
        this._mainFileUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mainFileUriInput() {
        return this._mainFileUri;
    }
    get properties() {
        return this.getStringMapAttribute('properties');
    }
    set properties(value) {
        this._properties = value;
    }
    resetProperties() {
        this._properties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propertiesInput() {
        return this._properties;
    }
    get pyFileUris() {
        return this.getListAttribute('py_file_uris');
    }
    set pyFileUris(value) {
        this._pyFileUris = value;
    }
    resetPyFileUris() {
        this._pyFileUris = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pyFileUrisInput() {
        return this._pyFileUris;
    }
    get runtimeVersion() {
        return this.getStringAttribute('runtime_version');
    }
    set runtimeVersion(value) {
        this._runtimeVersion = value;
    }
    resetRuntimeVersion() {
        this._runtimeVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeVersionInput() {
        return this._runtimeVersion;
    }
}
exports.BigqueryRoutineSparkOptionsOutputReference = BigqueryRoutineSparkOptionsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
BigqueryRoutineSparkOptionsOutputReference[_d] = { fqn: "@cdktf/provider-google.bigqueryRoutine.BigqueryRoutineSparkOptionsOutputReference", version: "14.13.1" };
function bigqueryRoutineTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.bigqueryRoutineTimeoutsToTerraform = bigqueryRoutineTimeoutsToTerraform;
function bigqueryRoutineTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryRoutineTimeoutsToHclTerraform = bigqueryRoutineTimeoutsToHclTerraform;
class BigqueryRoutineTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.BigqueryRoutineTimeoutsOutputReference = BigqueryRoutineTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
BigqueryRoutineTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-google.bigqueryRoutine.BigqueryRoutineTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_routine google_bigquery_routine}
*/
class BigqueryRoutine extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a BigqueryRoutine resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the BigqueryRoutine to import
    * @param importFromId The id of the existing BigqueryRoutine that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_routine#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the BigqueryRoutine to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_bigquery_routine", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_routine google_bigquery_routine} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BigqueryRoutineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_bigquery_routine',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // arguments - computed: false, optional: true, required: false
        this._arguments = new BigqueryRoutineArgumentsList(this, "arguments", false);
        // remote_function_options - computed: false, optional: true, required: false
        this._remoteFunctionOptions = new BigqueryRoutineRemoteFunctionOptionsOutputReference(this, "remote_function_options");
        // spark_options - computed: false, optional: true, required: false
        this._sparkOptions = new BigqueryRoutineSparkOptionsOutputReference(this, "spark_options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new BigqueryRoutineTimeoutsOutputReference(this, "timeouts");
        this._dataGovernanceType = config.dataGovernanceType;
        this._datasetId = config.datasetId;
        this._definitionBody = config.definitionBody;
        this._description = config.description;
        this._determinismLevel = config.determinismLevel;
        this._id = config.id;
        this._importedLibraries = config.importedLibraries;
        this._language = config.language;
        this._project = config.project;
        this._returnTableType = config.returnTableType;
        this._returnType = config.returnType;
        this._routineId = config.routineId;
        this._routineType = config.routineType;
        this._arguments.internalValue = config.arguments;
        this._remoteFunctionOptions.internalValue = config.remoteFunctionOptions;
        this._sparkOptions.internalValue = config.sparkOptions;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getNumberAttribute('creation_time');
    }
    get dataGovernanceType() {
        return this.getStringAttribute('data_governance_type');
    }
    set dataGovernanceType(value) {
        this._dataGovernanceType = value;
    }
    resetDataGovernanceType() {
        this._dataGovernanceType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataGovernanceTypeInput() {
        return this._dataGovernanceType;
    }
    get datasetId() {
        return this.getStringAttribute('dataset_id');
    }
    set datasetId(value) {
        this._datasetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get datasetIdInput() {
        return this._datasetId;
    }
    get definitionBody() {
        return this.getStringAttribute('definition_body');
    }
    set definitionBody(value) {
        this._definitionBody = value;
    }
    // Temporarily expose input value. Use with caution.
    get definitionBodyInput() {
        return this._definitionBody;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get determinismLevel() {
        return this.getStringAttribute('determinism_level');
    }
    set determinismLevel(value) {
        this._determinismLevel = value;
    }
    resetDeterminismLevel() {
        this._determinismLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get determinismLevelInput() {
        return this._determinismLevel;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get importedLibraries() {
        return this.getListAttribute('imported_libraries');
    }
    set importedLibraries(value) {
        this._importedLibraries = value;
    }
    resetImportedLibraries() {
        this._importedLibraries = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get importedLibrariesInput() {
        return this._importedLibraries;
    }
    get language() {
        return this.getStringAttribute('language');
    }
    set language(value) {
        this._language = value;
    }
    resetLanguage() {
        this._language = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get languageInput() {
        return this._language;
    }
    // last_modified_time - computed: true, optional: false, required: false
    get lastModifiedTime() {
        return this.getNumberAttribute('last_modified_time');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get returnTableType() {
        return this.getStringAttribute('return_table_type');
    }
    set returnTableType(value) {
        this._returnTableType = value;
    }
    resetReturnTableType() {
        this._returnTableType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get returnTableTypeInput() {
        return this._returnTableType;
    }
    get returnType() {
        return this.getStringAttribute('return_type');
    }
    set returnType(value) {
        this._returnType = value;
    }
    resetReturnType() {
        this._returnType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get returnTypeInput() {
        return this._returnType;
    }
    get routineId() {
        return this.getStringAttribute('routine_id');
    }
    set routineId(value) {
        this._routineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get routineIdInput() {
        return this._routineId;
    }
    get routineType() {
        return this.getStringAttribute('routine_type');
    }
    set routineType(value) {
        this._routineType = value;
    }
    // Temporarily expose input value. Use with caution.
    get routineTypeInput() {
        return this._routineType;
    }
    get arguments() {
        return this._arguments;
    }
    putArguments(value) {
        this._arguments.internalValue = value;
    }
    resetArguments() {
        this._arguments.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argumentsInput() {
        return this._arguments.internalValue;
    }
    get remoteFunctionOptions() {
        return this._remoteFunctionOptions;
    }
    putRemoteFunctionOptions(value) {
        this._remoteFunctionOptions.internalValue = value;
    }
    resetRemoteFunctionOptions() {
        this._remoteFunctionOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get remoteFunctionOptionsInput() {
        return this._remoteFunctionOptions.internalValue;
    }
    get sparkOptions() {
        return this._sparkOptions;
    }
    putSparkOptions(value) {
        this._sparkOptions.internalValue = value;
    }
    resetSparkOptions() {
        this._sparkOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sparkOptionsInput() {
        return this._sparkOptions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            data_governance_type: cdktf.stringToTerraform(this._dataGovernanceType),
            dataset_id: cdktf.stringToTerraform(this._datasetId),
            definition_body: cdktf.stringToTerraform(this._definitionBody),
            description: cdktf.stringToTerraform(this._description),
            determinism_level: cdktf.stringToTerraform(this._determinismLevel),
            id: cdktf.stringToTerraform(this._id),
            imported_libraries: cdktf.listMapper(cdktf.stringToTerraform, false)(this._importedLibraries),
            language: cdktf.stringToTerraform(this._language),
            project: cdktf.stringToTerraform(this._project),
            return_table_type: cdktf.stringToTerraform(this._returnTableType),
            return_type: cdktf.stringToTerraform(this._returnType),
            routine_id: cdktf.stringToTerraform(this._routineId),
            routine_type: cdktf.stringToTerraform(this._routineType),
            arguments: cdktf.listMapper(bigqueryRoutineArgumentsToTerraform, true)(this._arguments.internalValue),
            remote_function_options: bigqueryRoutineRemoteFunctionOptionsToTerraform(this._remoteFunctionOptions.internalValue),
            spark_options: bigqueryRoutineSparkOptionsToTerraform(this._sparkOptions.internalValue),
            timeouts: bigqueryRoutineTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            data_governance_type: {
                value: cdktf.stringToHclTerraform(this._dataGovernanceType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dataset_id: {
                value: cdktf.stringToHclTerraform(this._datasetId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            definition_body: {
                value: cdktf.stringToHclTerraform(this._definitionBody),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            determinism_level: {
                value: cdktf.stringToHclTerraform(this._determinismLevel),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            imported_libraries: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._importedLibraries),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            language: {
                value: cdktf.stringToHclTerraform(this._language),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            return_table_type: {
                value: cdktf.stringToHclTerraform(this._returnTableType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            return_type: {
                value: cdktf.stringToHclTerraform(this._returnType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            routine_id: {
                value: cdktf.stringToHclTerraform(this._routineId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            routine_type: {
                value: cdktf.stringToHclTerraform(this._routineType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            arguments: {
                value: cdktf.listMapperHcl(bigqueryRoutineArgumentsToHclTerraform, true)(this._arguments.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryRoutineArgumentsList",
            },
            remote_function_options: {
                value: bigqueryRoutineRemoteFunctionOptionsToHclTerraform(this._remoteFunctionOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryRoutineRemoteFunctionOptionsList",
            },
            spark_options: {
                value: bigqueryRoutineSparkOptionsToHclTerraform(this._sparkOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryRoutineSparkOptionsList",
            },
            timeouts: {
                value: bigqueryRoutineTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "BigqueryRoutineTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.BigqueryRoutine = BigqueryRoutine;
_f = JSII_RTTI_SYMBOL_1;
BigqueryRoutine[_f] = { fqn: "@cdktf/provider-google.bigqueryRoutine.BigqueryRoutine", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
BigqueryRoutine.tfResourceType = "google_bigquery_routine";
