// @cdktf/provider-google/lib/workbench-instance/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkbenchInstance = exports.WorkbenchInstanceTimeoutsOutputReference = exports.workbenchInstanceTimeoutsToHclTerraform = exports.workbenchInstanceTimeoutsToTerraform = exports.WorkbenchInstanceGceSetupOutputReference = exports.workbenchInstanceGceSetupToHclTerraform = exports.workbenchInstanceGceSetupToTerraform = exports.WorkbenchInstanceGceSetupVmImageOutputReference = exports.workbenchInstanceGceSetupVmImageToHclTerraform = exports.workbenchInstanceGceSetupVmImageToTerraform = exports.WorkbenchInstanceGceSetupShieldedInstanceConfigOutputReference = exports.workbenchInstanceGceSetupShieldedInstanceConfigToHclTerraform = exports.workbenchInstanceGceSetupShieldedInstanceConfigToTerraform = exports.WorkbenchInstanceGceSetupServiceAccountsList = exports.WorkbenchInstanceGceSetupServiceAccountsOutputReference = exports.workbenchInstanceGceSetupServiceAccountsToHclTerraform = exports.workbenchInstanceGceSetupServiceAccountsToTerraform = exports.WorkbenchInstanceGceSetupNetworkInterfacesList = exports.WorkbenchInstanceGceSetupNetworkInterfacesOutputReference = exports.workbenchInstanceGceSetupNetworkInterfacesToHclTerraform = exports.workbenchInstanceGceSetupNetworkInterfacesToTerraform = exports.WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsList = exports.WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsOutputReference = exports.workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToHclTerraform = exports.workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToTerraform = exports.WorkbenchInstanceGceSetupDataDisksOutputReference = exports.workbenchInstanceGceSetupDataDisksToHclTerraform = exports.workbenchInstanceGceSetupDataDisksToTerraform = exports.WorkbenchInstanceGceSetupContainerImageOutputReference = exports.workbenchInstanceGceSetupContainerImageToHclTerraform = exports.workbenchInstanceGceSetupContainerImageToTerraform = exports.WorkbenchInstanceGceSetupBootDiskOutputReference = exports.workbenchInstanceGceSetupBootDiskToHclTerraform = exports.workbenchInstanceGceSetupBootDiskToTerraform = exports.WorkbenchInstanceGceSetupAcceleratorConfigsList = exports.WorkbenchInstanceGceSetupAcceleratorConfigsOutputReference = exports.workbenchInstanceGceSetupAcceleratorConfigsToHclTerraform = exports.workbenchInstanceGceSetupAcceleratorConfigsToTerraform = exports.WorkbenchInstanceUpgradeHistoryList = exports.WorkbenchInstanceUpgradeHistoryOutputReference = exports.workbenchInstanceUpgradeHistoryToHclTerraform = exports.workbenchInstanceUpgradeHistoryToTerraform = exports.WorkbenchInstanceHealthInfoList = exports.WorkbenchInstanceHealthInfoOutputReference = exports.workbenchInstanceHealthInfoToHclTerraform = exports.workbenchInstanceHealthInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function workbenchInstanceHealthInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.workbenchInstanceHealthInfoToTerraform = workbenchInstanceHealthInfoToTerraform;
function workbenchInstanceHealthInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.workbenchInstanceHealthInfoToHclTerraform = workbenchInstanceHealthInfoToHclTerraform;
class WorkbenchInstanceHealthInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.WorkbenchInstanceHealthInfoOutputReference = WorkbenchInstanceHealthInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceHealthInfoOutputReference[_a] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceHealthInfoOutputReference", version: "14.13.1" };
class WorkbenchInstanceHealthInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WorkbenchInstanceHealthInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WorkbenchInstanceHealthInfoList = WorkbenchInstanceHealthInfoList;
_b = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceHealthInfoList[_b] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceHealthInfoList", version: "14.13.1" };
function workbenchInstanceUpgradeHistoryToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.workbenchInstanceUpgradeHistoryToTerraform = workbenchInstanceUpgradeHistoryToTerraform;
function workbenchInstanceUpgradeHistoryToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.workbenchInstanceUpgradeHistoryToHclTerraform = workbenchInstanceUpgradeHistoryToHclTerraform;
class WorkbenchInstanceUpgradeHistoryOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // action - computed: true, optional: false, required: false
    get action() {
        return this.getStringAttribute('action');
    }
    // container_image - computed: true, optional: false, required: false
    get containerImage() {
        return this.getStringAttribute('container_image');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // framework - computed: true, optional: false, required: false
    get framework() {
        return this.getStringAttribute('framework');
    }
    // snapshot - computed: true, optional: false, required: false
    get snapshot() {
        return this.getStringAttribute('snapshot');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // target_version - computed: true, optional: false, required: false
    get targetVersion() {
        return this.getStringAttribute('target_version');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // vm_image - computed: true, optional: false, required: false
    get vmImage() {
        return this.getStringAttribute('vm_image');
    }
}
exports.WorkbenchInstanceUpgradeHistoryOutputReference = WorkbenchInstanceUpgradeHistoryOutputReference;
_c = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceUpgradeHistoryOutputReference[_c] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceUpgradeHistoryOutputReference", version: "14.13.1" };
class WorkbenchInstanceUpgradeHistoryList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WorkbenchInstanceUpgradeHistoryOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WorkbenchInstanceUpgradeHistoryList = WorkbenchInstanceUpgradeHistoryList;
_d = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceUpgradeHistoryList[_d] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceUpgradeHistoryList", version: "14.13.1" };
function workbenchInstanceGceSetupAcceleratorConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        core_count: cdktf.stringToTerraform(struct.coreCount),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.workbenchInstanceGceSetupAcceleratorConfigsToTerraform = workbenchInstanceGceSetupAcceleratorConfigsToTerraform;
function workbenchInstanceGceSetupAcceleratorConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        core_count: {
            value: cdktf.stringToHclTerraform(struct.coreCount),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupAcceleratorConfigsToHclTerraform = workbenchInstanceGceSetupAcceleratorConfigsToHclTerraform;
class WorkbenchInstanceGceSetupAcceleratorConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coreCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.coreCount = this._coreCount;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._coreCount = undefined;
            this._type = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._coreCount = value.coreCount;
            this._type = value.type;
        }
    }
    get coreCount() {
        return this.getStringAttribute('core_count');
    }
    set coreCount(value) {
        this._coreCount = value;
    }
    resetCoreCount() {
        this._coreCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreCountInput() {
        return this._coreCount;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.WorkbenchInstanceGceSetupAcceleratorConfigsOutputReference = WorkbenchInstanceGceSetupAcceleratorConfigsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupAcceleratorConfigsOutputReference[_e] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupAcceleratorConfigsOutputReference", version: "14.13.1" };
class WorkbenchInstanceGceSetupAcceleratorConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WorkbenchInstanceGceSetupAcceleratorConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WorkbenchInstanceGceSetupAcceleratorConfigsList = WorkbenchInstanceGceSetupAcceleratorConfigsList;
_f = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupAcceleratorConfigsList[_f] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupAcceleratorConfigsList", version: "14.13.1" };
function workbenchInstanceGceSetupBootDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_encryption: cdktf.stringToTerraform(struct.diskEncryption),
        disk_size_gb: cdktf.stringToTerraform(struct.diskSizeGb),
        disk_type: cdktf.stringToTerraform(struct.diskType),
        kms_key: cdktf.stringToTerraform(struct.kmsKey),
    };
}
exports.workbenchInstanceGceSetupBootDiskToTerraform = workbenchInstanceGceSetupBootDiskToTerraform;
function workbenchInstanceGceSetupBootDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        disk_encryption: {
            value: cdktf.stringToHclTerraform(struct.diskEncryption),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        disk_size_gb: {
            value: cdktf.stringToHclTerraform(struct.diskSizeGb),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        disk_type: {
            value: cdktf.stringToHclTerraform(struct.diskType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kms_key: {
            value: cdktf.stringToHclTerraform(struct.kmsKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupBootDiskToHclTerraform = workbenchInstanceGceSetupBootDiskToHclTerraform;
class WorkbenchInstanceGceSetupBootDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskEncryption !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskEncryption = this._diskEncryption;
        }
        if (this._diskSizeGb !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSizeGb = this._diskSizeGb;
        }
        if (this._diskType !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskType = this._diskType;
        }
        if (this._kmsKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKey = this._kmsKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskEncryption = undefined;
            this._diskSizeGb = undefined;
            this._diskType = undefined;
            this._kmsKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskEncryption = value.diskEncryption;
            this._diskSizeGb = value.diskSizeGb;
            this._diskType = value.diskType;
            this._kmsKey = value.kmsKey;
        }
    }
    get diskEncryption() {
        return this.getStringAttribute('disk_encryption');
    }
    set diskEncryption(value) {
        this._diskEncryption = value;
    }
    resetDiskEncryption() {
        this._diskEncryption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskEncryptionInput() {
        return this._diskEncryption;
    }
    get diskSizeGb() {
        return this.getStringAttribute('disk_size_gb');
    }
    set diskSizeGb(value) {
        this._diskSizeGb = value;
    }
    resetDiskSizeGb() {
        this._diskSizeGb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeGbInput() {
        return this._diskSizeGb;
    }
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    set diskType(value) {
        this._diskType = value;
    }
    resetDiskType() {
        this._diskType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeInput() {
        return this._diskType;
    }
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    set kmsKey(value) {
        this._kmsKey = value;
    }
    resetKmsKey() {
        this._kmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyInput() {
        return this._kmsKey;
    }
}
exports.WorkbenchInstanceGceSetupBootDiskOutputReference = WorkbenchInstanceGceSetupBootDiskOutputReference;
_g = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupBootDiskOutputReference[_g] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupBootDiskOutputReference", version: "14.13.1" };
function workbenchInstanceGceSetupContainerImageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repository: cdktf.stringToTerraform(struct.repository),
        tag: cdktf.stringToTerraform(struct.tag),
    };
}
exports.workbenchInstanceGceSetupContainerImageToTerraform = workbenchInstanceGceSetupContainerImageToTerraform;
function workbenchInstanceGceSetupContainerImageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repository: {
            value: cdktf.stringToHclTerraform(struct.repository),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        tag: {
            value: cdktf.stringToHclTerraform(struct.tag),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupContainerImageToHclTerraform = workbenchInstanceGceSetupContainerImageToHclTerraform;
class WorkbenchInstanceGceSetupContainerImageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repository !== undefined) {
            hasAnyValues = true;
            internalValueResult.repository = this._repository;
        }
        if (this._tag !== undefined) {
            hasAnyValues = true;
            internalValueResult.tag = this._tag;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repository = undefined;
            this._tag = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repository = value.repository;
            this._tag = value.tag;
        }
    }
    get repository() {
        return this.getStringAttribute('repository');
    }
    set repository(value) {
        this._repository = value;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryInput() {
        return this._repository;
    }
    get tag() {
        return this.getStringAttribute('tag');
    }
    set tag(value) {
        this._tag = value;
    }
    resetTag() {
        this._tag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagInput() {
        return this._tag;
    }
}
exports.WorkbenchInstanceGceSetupContainerImageOutputReference = WorkbenchInstanceGceSetupContainerImageOutputReference;
_h = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupContainerImageOutputReference[_h] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupContainerImageOutputReference", version: "14.13.1" };
function workbenchInstanceGceSetupDataDisksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_encryption: cdktf.stringToTerraform(struct.diskEncryption),
        disk_size_gb: cdktf.stringToTerraform(struct.diskSizeGb),
        disk_type: cdktf.stringToTerraform(struct.diskType),
        kms_key: cdktf.stringToTerraform(struct.kmsKey),
    };
}
exports.workbenchInstanceGceSetupDataDisksToTerraform = workbenchInstanceGceSetupDataDisksToTerraform;
function workbenchInstanceGceSetupDataDisksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        disk_encryption: {
            value: cdktf.stringToHclTerraform(struct.diskEncryption),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        disk_size_gb: {
            value: cdktf.stringToHclTerraform(struct.diskSizeGb),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        disk_type: {
            value: cdktf.stringToHclTerraform(struct.diskType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kms_key: {
            value: cdktf.stringToHclTerraform(struct.kmsKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupDataDisksToHclTerraform = workbenchInstanceGceSetupDataDisksToHclTerraform;
class WorkbenchInstanceGceSetupDataDisksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskEncryption !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskEncryption = this._diskEncryption;
        }
        if (this._diskSizeGb !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSizeGb = this._diskSizeGb;
        }
        if (this._diskType !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskType = this._diskType;
        }
        if (this._kmsKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKey = this._kmsKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskEncryption = undefined;
            this._diskSizeGb = undefined;
            this._diskType = undefined;
            this._kmsKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskEncryption = value.diskEncryption;
            this._diskSizeGb = value.diskSizeGb;
            this._diskType = value.diskType;
            this._kmsKey = value.kmsKey;
        }
    }
    get diskEncryption() {
        return this.getStringAttribute('disk_encryption');
    }
    set diskEncryption(value) {
        this._diskEncryption = value;
    }
    resetDiskEncryption() {
        this._diskEncryption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskEncryptionInput() {
        return this._diskEncryption;
    }
    get diskSizeGb() {
        return this.getStringAttribute('disk_size_gb');
    }
    set diskSizeGb(value) {
        this._diskSizeGb = value;
    }
    resetDiskSizeGb() {
        this._diskSizeGb = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeGbInput() {
        return this._diskSizeGb;
    }
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    set diskType(value) {
        this._diskType = value;
    }
    resetDiskType() {
        this._diskType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeInput() {
        return this._diskType;
    }
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    set kmsKey(value) {
        this._kmsKey = value;
    }
    resetKmsKey() {
        this._kmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyInput() {
        return this._kmsKey;
    }
}
exports.WorkbenchInstanceGceSetupDataDisksOutputReference = WorkbenchInstanceGceSetupDataDisksOutputReference;
_j = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupDataDisksOutputReference[_j] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupDataDisksOutputReference", version: "14.13.1" };
function workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        external_ip: cdktf.stringToTerraform(struct.externalIp),
    };
}
exports.workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToTerraform = workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToTerraform;
function workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        external_ip: {
            value: cdktf.stringToHclTerraform(struct.externalIp),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToHclTerraform = workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToHclTerraform;
class WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._externalIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.externalIp = this._externalIp;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._externalIp = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._externalIp = value.externalIp;
        }
    }
    get externalIp() {
        return this.getStringAttribute('external_ip');
    }
    set externalIp(value) {
        this._externalIp = value;
    }
    // Temporarily expose input value. Use with caution.
    get externalIpInput() {
        return this._externalIp;
    }
}
exports.WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsOutputReference = WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsOutputReference[_k] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsOutputReference", version: "14.13.1" };
class WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsList = WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsList;
_l = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsList[_l] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsList", version: "14.13.1" };
function workbenchInstanceGceSetupNetworkInterfacesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        network: cdktf.stringToTerraform(struct.network),
        nic_type: cdktf.stringToTerraform(struct.nicType),
        subnet: cdktf.stringToTerraform(struct.subnet),
        access_configs: cdktf.listMapper(workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToTerraform, true)(struct.accessConfigs),
    };
}
exports.workbenchInstanceGceSetupNetworkInterfacesToTerraform = workbenchInstanceGceSetupNetworkInterfacesToTerraform;
function workbenchInstanceGceSetupNetworkInterfacesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        network: {
            value: cdktf.stringToHclTerraform(struct.network),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        nic_type: {
            value: cdktf.stringToHclTerraform(struct.nicType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subnet: {
            value: cdktf.stringToHclTerraform(struct.subnet),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        access_configs: {
            value: cdktf.listMapperHcl(workbenchInstanceGceSetupNetworkInterfacesAccessConfigsToHclTerraform, true)(struct.accessConfigs),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupNetworkInterfacesToHclTerraform = workbenchInstanceGceSetupNetworkInterfacesToHclTerraform;
class WorkbenchInstanceGceSetupNetworkInterfacesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // access_configs - computed: false, optional: true, required: false
        this._accessConfigs = new WorkbenchInstanceGceSetupNetworkInterfacesAccessConfigsList(this, "access_configs", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._network !== undefined) {
            hasAnyValues = true;
            internalValueResult.network = this._network;
        }
        if (this._nicType !== undefined) {
            hasAnyValues = true;
            internalValueResult.nicType = this._nicType;
        }
        if (this._subnet !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnet = this._subnet;
        }
        if (this._accessConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessConfigs = this._accessConfigs?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._network = undefined;
            this._nicType = undefined;
            this._subnet = undefined;
            this._accessConfigs.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._network = value.network;
            this._nicType = value.nicType;
            this._subnet = value.subnet;
            this._accessConfigs.internalValue = value.accessConfigs;
        }
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    resetNetwork() {
        this._network = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get nicType() {
        return this.getStringAttribute('nic_type');
    }
    set nicType(value) {
        this._nicType = value;
    }
    resetNicType() {
        this._nicType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nicTypeInput() {
        return this._nicType;
    }
    get subnet() {
        return this.getStringAttribute('subnet');
    }
    set subnet(value) {
        this._subnet = value;
    }
    resetSubnet() {
        this._subnet = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetInput() {
        return this._subnet;
    }
    get accessConfigs() {
        return this._accessConfigs;
    }
    putAccessConfigs(value) {
        this._accessConfigs.internalValue = value;
    }
    resetAccessConfigs() {
        this._accessConfigs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessConfigsInput() {
        return this._accessConfigs.internalValue;
    }
}
exports.WorkbenchInstanceGceSetupNetworkInterfacesOutputReference = WorkbenchInstanceGceSetupNetworkInterfacesOutputReference;
_m = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupNetworkInterfacesOutputReference[_m] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupNetworkInterfacesOutputReference", version: "14.13.1" };
class WorkbenchInstanceGceSetupNetworkInterfacesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WorkbenchInstanceGceSetupNetworkInterfacesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WorkbenchInstanceGceSetupNetworkInterfacesList = WorkbenchInstanceGceSetupNetworkInterfacesList;
_o = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupNetworkInterfacesList[_o] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupNetworkInterfacesList", version: "14.13.1" };
function workbenchInstanceGceSetupServiceAccountsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        email: cdktf.stringToTerraform(struct.email),
    };
}
exports.workbenchInstanceGceSetupServiceAccountsToTerraform = workbenchInstanceGceSetupServiceAccountsToTerraform;
function workbenchInstanceGceSetupServiceAccountsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        email: {
            value: cdktf.stringToHclTerraform(struct.email),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupServiceAccountsToHclTerraform = workbenchInstanceGceSetupServiceAccountsToHclTerraform;
class WorkbenchInstanceGceSetupServiceAccountsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._email !== undefined) {
            hasAnyValues = true;
            internalValueResult.email = this._email;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._email = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._email = value.email;
        }
    }
    get email() {
        return this.getStringAttribute('email');
    }
    set email(value) {
        this._email = value;
    }
    resetEmail() {
        this._email = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get emailInput() {
        return this._email;
    }
    // scopes - computed: true, optional: false, required: false
    get scopes() {
        return this.getListAttribute('scopes');
    }
}
exports.WorkbenchInstanceGceSetupServiceAccountsOutputReference = WorkbenchInstanceGceSetupServiceAccountsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupServiceAccountsOutputReference[_p] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupServiceAccountsOutputReference", version: "14.13.1" };
class WorkbenchInstanceGceSetupServiceAccountsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new WorkbenchInstanceGceSetupServiceAccountsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.WorkbenchInstanceGceSetupServiceAccountsList = WorkbenchInstanceGceSetupServiceAccountsList;
_q = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupServiceAccountsList[_q] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupServiceAccountsList", version: "14.13.1" };
function workbenchInstanceGceSetupShieldedInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_integrity_monitoring: cdktf.booleanToTerraform(struct.enableIntegrityMonitoring),
        enable_secure_boot: cdktf.booleanToTerraform(struct.enableSecureBoot),
        enable_vtpm: cdktf.booleanToTerraform(struct.enableVtpm),
    };
}
exports.workbenchInstanceGceSetupShieldedInstanceConfigToTerraform = workbenchInstanceGceSetupShieldedInstanceConfigToTerraform;
function workbenchInstanceGceSetupShieldedInstanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_integrity_monitoring: {
            value: cdktf.booleanToHclTerraform(struct.enableIntegrityMonitoring),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_secure_boot: {
            value: cdktf.booleanToHclTerraform(struct.enableSecureBoot),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_vtpm: {
            value: cdktf.booleanToHclTerraform(struct.enableVtpm),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupShieldedInstanceConfigToHclTerraform = workbenchInstanceGceSetupShieldedInstanceConfigToHclTerraform;
class WorkbenchInstanceGceSetupShieldedInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableIntegrityMonitoring !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableIntegrityMonitoring = this._enableIntegrityMonitoring;
        }
        if (this._enableSecureBoot !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableSecureBoot = this._enableSecureBoot;
        }
        if (this._enableVtpm !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableVtpm = this._enableVtpm;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableIntegrityMonitoring = undefined;
            this._enableSecureBoot = undefined;
            this._enableVtpm = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableIntegrityMonitoring = value.enableIntegrityMonitoring;
            this._enableSecureBoot = value.enableSecureBoot;
            this._enableVtpm = value.enableVtpm;
        }
    }
    get enableIntegrityMonitoring() {
        return this.getBooleanAttribute('enable_integrity_monitoring');
    }
    set enableIntegrityMonitoring(value) {
        this._enableIntegrityMonitoring = value;
    }
    resetEnableIntegrityMonitoring() {
        this._enableIntegrityMonitoring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableIntegrityMonitoringInput() {
        return this._enableIntegrityMonitoring;
    }
    get enableSecureBoot() {
        return this.getBooleanAttribute('enable_secure_boot');
    }
    set enableSecureBoot(value) {
        this._enableSecureBoot = value;
    }
    resetEnableSecureBoot() {
        this._enableSecureBoot = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableSecureBootInput() {
        return this._enableSecureBoot;
    }
    get enableVtpm() {
        return this.getBooleanAttribute('enable_vtpm');
    }
    set enableVtpm(value) {
        this._enableVtpm = value;
    }
    resetEnableVtpm() {
        this._enableVtpm = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableVtpmInput() {
        return this._enableVtpm;
    }
}
exports.WorkbenchInstanceGceSetupShieldedInstanceConfigOutputReference = WorkbenchInstanceGceSetupShieldedInstanceConfigOutputReference;
_r = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupShieldedInstanceConfigOutputReference[_r] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupShieldedInstanceConfigOutputReference", version: "14.13.1" };
function workbenchInstanceGceSetupVmImageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        family: cdktf.stringToTerraform(struct.family),
        name: cdktf.stringToTerraform(struct.name),
        project: cdktf.stringToTerraform(struct.project),
    };
}
exports.workbenchInstanceGceSetupVmImageToTerraform = workbenchInstanceGceSetupVmImageToTerraform;
function workbenchInstanceGceSetupVmImageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        family: {
            value: cdktf.stringToHclTerraform(struct.family),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        project: {
            value: cdktf.stringToHclTerraform(struct.project),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupVmImageToHclTerraform = workbenchInstanceGceSetupVmImageToHclTerraform;
class WorkbenchInstanceGceSetupVmImageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._family !== undefined) {
            hasAnyValues = true;
            internalValueResult.family = this._family;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._project !== undefined) {
            hasAnyValues = true;
            internalValueResult.project = this._project;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._family = undefined;
            this._name = undefined;
            this._project = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._family = value.family;
            this._name = value.name;
            this._project = value.project;
        }
    }
    get family() {
        return this.getStringAttribute('family');
    }
    set family(value) {
        this._family = value;
    }
    resetFamily() {
        this._family = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get familyInput() {
        return this._family;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
}
exports.WorkbenchInstanceGceSetupVmImageOutputReference = WorkbenchInstanceGceSetupVmImageOutputReference;
_s = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupVmImageOutputReference[_s] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupVmImageOutputReference", version: "14.13.1" };
function workbenchInstanceGceSetupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disable_public_ip: cdktf.booleanToTerraform(struct.disablePublicIp),
        enable_ip_forwarding: cdktf.booleanToTerraform(struct.enableIpForwarding),
        machine_type: cdktf.stringToTerraform(struct.machineType),
        metadata: cdktf.hashMapper(cdktf.stringToTerraform)(struct.metadata),
        tags: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tags),
        accelerator_configs: cdktf.listMapper(workbenchInstanceGceSetupAcceleratorConfigsToTerraform, true)(struct.acceleratorConfigs),
        boot_disk: workbenchInstanceGceSetupBootDiskToTerraform(struct.bootDisk),
        container_image: workbenchInstanceGceSetupContainerImageToTerraform(struct.containerImage),
        data_disks: workbenchInstanceGceSetupDataDisksToTerraform(struct.dataDisks),
        network_interfaces: cdktf.listMapper(workbenchInstanceGceSetupNetworkInterfacesToTerraform, true)(struct.networkInterfaces),
        service_accounts: cdktf.listMapper(workbenchInstanceGceSetupServiceAccountsToTerraform, true)(struct.serviceAccounts),
        shielded_instance_config: workbenchInstanceGceSetupShieldedInstanceConfigToTerraform(struct.shieldedInstanceConfig),
        vm_image: workbenchInstanceGceSetupVmImageToTerraform(struct.vmImage),
    };
}
exports.workbenchInstanceGceSetupToTerraform = workbenchInstanceGceSetupToTerraform;
function workbenchInstanceGceSetupToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        disable_public_ip: {
            value: cdktf.booleanToHclTerraform(struct.disablePublicIp),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_ip_forwarding: {
            value: cdktf.booleanToHclTerraform(struct.enableIpForwarding),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        machine_type: {
            value: cdktf.stringToHclTerraform(struct.machineType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        metadata: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.metadata),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        tags: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.tags),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        accelerator_configs: {
            value: cdktf.listMapperHcl(workbenchInstanceGceSetupAcceleratorConfigsToHclTerraform, true)(struct.acceleratorConfigs),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupAcceleratorConfigsList",
        },
        boot_disk: {
            value: workbenchInstanceGceSetupBootDiskToHclTerraform(struct.bootDisk),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupBootDiskList",
        },
        container_image: {
            value: workbenchInstanceGceSetupContainerImageToHclTerraform(struct.containerImage),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupContainerImageList",
        },
        data_disks: {
            value: workbenchInstanceGceSetupDataDisksToHclTerraform(struct.dataDisks),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupDataDisksList",
        },
        network_interfaces: {
            value: cdktf.listMapperHcl(workbenchInstanceGceSetupNetworkInterfacesToHclTerraform, true)(struct.networkInterfaces),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupNetworkInterfacesList",
        },
        service_accounts: {
            value: cdktf.listMapperHcl(workbenchInstanceGceSetupServiceAccountsToHclTerraform, true)(struct.serviceAccounts),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupServiceAccountsList",
        },
        shielded_instance_config: {
            value: workbenchInstanceGceSetupShieldedInstanceConfigToHclTerraform(struct.shieldedInstanceConfig),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupShieldedInstanceConfigList",
        },
        vm_image: {
            value: workbenchInstanceGceSetupVmImageToHclTerraform(struct.vmImage),
            isBlock: true,
            type: "list",
            storageClassType: "WorkbenchInstanceGceSetupVmImageList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceGceSetupToHclTerraform = workbenchInstanceGceSetupToHclTerraform;
class WorkbenchInstanceGceSetupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // accelerator_configs - computed: false, optional: true, required: false
        this._acceleratorConfigs = new WorkbenchInstanceGceSetupAcceleratorConfigsList(this, "accelerator_configs", false);
        // boot_disk - computed: false, optional: true, required: false
        this._bootDisk = new WorkbenchInstanceGceSetupBootDiskOutputReference(this, "boot_disk");
        // container_image - computed: false, optional: true, required: false
        this._containerImage = new WorkbenchInstanceGceSetupContainerImageOutputReference(this, "container_image");
        // data_disks - computed: false, optional: true, required: false
        this._dataDisks = new WorkbenchInstanceGceSetupDataDisksOutputReference(this, "data_disks");
        // network_interfaces - computed: false, optional: true, required: false
        this._networkInterfaces = new WorkbenchInstanceGceSetupNetworkInterfacesList(this, "network_interfaces", false);
        // service_accounts - computed: false, optional: true, required: false
        this._serviceAccounts = new WorkbenchInstanceGceSetupServiceAccountsList(this, "service_accounts", false);
        // shielded_instance_config - computed: false, optional: true, required: false
        this._shieldedInstanceConfig = new WorkbenchInstanceGceSetupShieldedInstanceConfigOutputReference(this, "shielded_instance_config");
        // vm_image - computed: false, optional: true, required: false
        this._vmImage = new WorkbenchInstanceGceSetupVmImageOutputReference(this, "vm_image");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._disablePublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.disablePublicIp = this._disablePublicIp;
        }
        if (this._enableIpForwarding !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableIpForwarding = this._enableIpForwarding;
        }
        if (this._machineType !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineType = this._machineType;
        }
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        if (this._acceleratorConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.acceleratorConfigs = this._acceleratorConfigs?.internalValue;
        }
        if (this._bootDisk?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.bootDisk = this._bootDisk?.internalValue;
        }
        if (this._containerImage?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerImage = this._containerImage?.internalValue;
        }
        if (this._dataDisks?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataDisks = this._dataDisks?.internalValue;
        }
        if (this._networkInterfaces?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.networkInterfaces = this._networkInterfaces?.internalValue;
        }
        if (this._serviceAccounts?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccounts = this._serviceAccounts?.internalValue;
        }
        if (this._shieldedInstanceConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.shieldedInstanceConfig = this._shieldedInstanceConfig?.internalValue;
        }
        if (this._vmImage?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.vmImage = this._vmImage?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._disablePublicIp = undefined;
            this._enableIpForwarding = undefined;
            this._machineType = undefined;
            this._metadata = undefined;
            this._tags = undefined;
            this._acceleratorConfigs.internalValue = undefined;
            this._bootDisk.internalValue = undefined;
            this._containerImage.internalValue = undefined;
            this._dataDisks.internalValue = undefined;
            this._networkInterfaces.internalValue = undefined;
            this._serviceAccounts.internalValue = undefined;
            this._shieldedInstanceConfig.internalValue = undefined;
            this._vmImage.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._disablePublicIp = value.disablePublicIp;
            this._enableIpForwarding = value.enableIpForwarding;
            this._machineType = value.machineType;
            this._metadata = value.metadata;
            this._tags = value.tags;
            this._acceleratorConfigs.internalValue = value.acceleratorConfigs;
            this._bootDisk.internalValue = value.bootDisk;
            this._containerImage.internalValue = value.containerImage;
            this._dataDisks.internalValue = value.dataDisks;
            this._networkInterfaces.internalValue = value.networkInterfaces;
            this._serviceAccounts.internalValue = value.serviceAccounts;
            this._shieldedInstanceConfig.internalValue = value.shieldedInstanceConfig;
            this._vmImage.internalValue = value.vmImage;
        }
    }
    get disablePublicIp() {
        return this.getBooleanAttribute('disable_public_ip');
    }
    set disablePublicIp(value) {
        this._disablePublicIp = value;
    }
    resetDisablePublicIp() {
        this._disablePublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disablePublicIpInput() {
        return this._disablePublicIp;
    }
    get enableIpForwarding() {
        return this.getBooleanAttribute('enable_ip_forwarding');
    }
    set enableIpForwarding(value) {
        this._enableIpForwarding = value;
    }
    resetEnableIpForwarding() {
        this._enableIpForwarding = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableIpForwardingInput() {
        return this._enableIpForwarding;
    }
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    set machineType(value) {
        this._machineType = value;
    }
    resetMachineType() {
        this._machineType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get machineTypeInput() {
        return this._machineType;
    }
    get metadata() {
        return this.getStringMapAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
    get tags() {
        return this.getListAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    resetTags() {
        this._tags = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
    get acceleratorConfigs() {
        return this._acceleratorConfigs;
    }
    putAcceleratorConfigs(value) {
        this._acceleratorConfigs.internalValue = value;
    }
    resetAcceleratorConfigs() {
        this._acceleratorConfigs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acceleratorConfigsInput() {
        return this._acceleratorConfigs.internalValue;
    }
    get bootDisk() {
        return this._bootDisk;
    }
    putBootDisk(value) {
        this._bootDisk.internalValue = value;
    }
    resetBootDisk() {
        this._bootDisk.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bootDiskInput() {
        return this._bootDisk.internalValue;
    }
    get containerImage() {
        return this._containerImage;
    }
    putContainerImage(value) {
        this._containerImage.internalValue = value;
    }
    resetContainerImage() {
        this._containerImage.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerImageInput() {
        return this._containerImage.internalValue;
    }
    get dataDisks() {
        return this._dataDisks;
    }
    putDataDisks(value) {
        this._dataDisks.internalValue = value;
    }
    resetDataDisks() {
        this._dataDisks.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataDisksInput() {
        return this._dataDisks.internalValue;
    }
    get networkInterfaces() {
        return this._networkInterfaces;
    }
    putNetworkInterfaces(value) {
        this._networkInterfaces.internalValue = value;
    }
    resetNetworkInterfaces() {
        this._networkInterfaces.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInterfacesInput() {
        return this._networkInterfaces.internalValue;
    }
    get serviceAccounts() {
        return this._serviceAccounts;
    }
    putServiceAccounts(value) {
        this._serviceAccounts.internalValue = value;
    }
    resetServiceAccounts() {
        this._serviceAccounts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountsInput() {
        return this._serviceAccounts.internalValue;
    }
    get shieldedInstanceConfig() {
        return this._shieldedInstanceConfig;
    }
    putShieldedInstanceConfig(value) {
        this._shieldedInstanceConfig.internalValue = value;
    }
    resetShieldedInstanceConfig() {
        this._shieldedInstanceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shieldedInstanceConfigInput() {
        return this._shieldedInstanceConfig.internalValue;
    }
    get vmImage() {
        return this._vmImage;
    }
    putVmImage(value) {
        this._vmImage.internalValue = value;
    }
    resetVmImage() {
        this._vmImage.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get vmImageInput() {
        return this._vmImage.internalValue;
    }
}
exports.WorkbenchInstanceGceSetupOutputReference = WorkbenchInstanceGceSetupOutputReference;
_t = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceGceSetupOutputReference[_t] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceGceSetupOutputReference", version: "14.13.1" };
function workbenchInstanceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.workbenchInstanceTimeoutsToTerraform = workbenchInstanceTimeoutsToTerraform;
function workbenchInstanceTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.workbenchInstanceTimeoutsToHclTerraform = workbenchInstanceTimeoutsToHclTerraform;
class WorkbenchInstanceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.WorkbenchInstanceTimeoutsOutputReference = WorkbenchInstanceTimeoutsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
WorkbenchInstanceTimeoutsOutputReference[_u] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstanceTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/workbench_instance google_workbench_instance}
*/
class WorkbenchInstance extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a WorkbenchInstance resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the WorkbenchInstance to import
    * @param importFromId The id of the existing WorkbenchInstance that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/workbench_instance#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the WorkbenchInstance to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_workbench_instance", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/workbench_instance google_workbench_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options WorkbenchInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_workbench_instance',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // health_info - computed: true, optional: false, required: false
        this._healthInfo = new WorkbenchInstanceHealthInfoList(this, "health_info", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // upgrade_history - computed: true, optional: false, required: false
        this._upgradeHistory = new WorkbenchInstanceUpgradeHistoryList(this, "upgrade_history", false);
        // gce_setup - computed: false, optional: true, required: false
        this._gceSetup = new WorkbenchInstanceGceSetupOutputReference(this, "gce_setup");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new WorkbenchInstanceTimeoutsOutputReference(this, "timeouts");
        this._desiredState = config.desiredState;
        this._disableProxyAccess = config.disableProxyAccess;
        this._id = config.id;
        this._instanceId = config.instanceId;
        this._instanceOwners = config.instanceOwners;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._gceSetup.internalValue = config.gceSetup;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // creator - computed: true, optional: false, required: false
    get creator() {
        return this.getStringAttribute('creator');
    }
    get desiredState() {
        return this.getStringAttribute('desired_state');
    }
    set desiredState(value) {
        this._desiredState = value;
    }
    resetDesiredState() {
        this._desiredState = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get desiredStateInput() {
        return this._desiredState;
    }
    get disableProxyAccess() {
        return this.getBooleanAttribute('disable_proxy_access');
    }
    set disableProxyAccess(value) {
        this._disableProxyAccess = value;
    }
    resetDisableProxyAccess() {
        this._disableProxyAccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableProxyAccessInput() {
        return this._disableProxyAccess;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get healthInfo() {
        return this._healthInfo;
    }
    // health_state - computed: true, optional: false, required: false
    get healthState() {
        return this.getStringAttribute('health_state');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    resetInstanceId() {
        this._instanceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    get instanceOwners() {
        return this.getListAttribute('instance_owners');
    }
    set instanceOwners(value) {
        this._instanceOwners = value;
    }
    resetInstanceOwners() {
        this._instanceOwners = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceOwnersInput() {
        return this._instanceOwners;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // proxy_uri - computed: true, optional: false, required: false
    get proxyUri() {
        return this.getStringAttribute('proxy_uri');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get upgradeHistory() {
        return this._upgradeHistory;
    }
    get gceSetup() {
        return this._gceSetup;
    }
    putGceSetup(value) {
        this._gceSetup.internalValue = value;
    }
    resetGceSetup() {
        this._gceSetup.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gceSetupInput() {
        return this._gceSetup.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            desired_state: cdktf.stringToTerraform(this._desiredState),
            disable_proxy_access: cdktf.booleanToTerraform(this._disableProxyAccess),
            id: cdktf.stringToTerraform(this._id),
            instance_id: cdktf.stringToTerraform(this._instanceId),
            instance_owners: cdktf.listMapper(cdktf.stringToTerraform, false)(this._instanceOwners),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            gce_setup: workbenchInstanceGceSetupToTerraform(this._gceSetup.internalValue),
            timeouts: workbenchInstanceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            desired_state: {
                value: cdktf.stringToHclTerraform(this._desiredState),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            disable_proxy_access: {
                value: cdktf.booleanToHclTerraform(this._disableProxyAccess),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_id: {
                value: cdktf.stringToHclTerraform(this._instanceId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            instance_owners: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._instanceOwners),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            gce_setup: {
                value: workbenchInstanceGceSetupToHclTerraform(this._gceSetup.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "WorkbenchInstanceGceSetupList",
            },
            timeouts: {
                value: workbenchInstanceTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "WorkbenchInstanceTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.WorkbenchInstance = WorkbenchInstance;
_v = JSII_RTTI_SYMBOL_1;
WorkbenchInstance[_v] = { fqn: "@cdktf/provider-google.workbenchInstance.WorkbenchInstance", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
WorkbenchInstance.tfResourceType = "google_workbench_instance";
