// @cdktf/provider-google/lib/iam-principal-access-boundary-policy/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamPrincipalAccessBoundaryPolicy = exports.IamPrincipalAccessBoundaryPolicyTimeoutsOutputReference = exports.iamPrincipalAccessBoundaryPolicyTimeoutsToHclTerraform = exports.iamPrincipalAccessBoundaryPolicyTimeoutsToTerraform = exports.IamPrincipalAccessBoundaryPolicyDetailsOutputReference = exports.iamPrincipalAccessBoundaryPolicyDetailsToHclTerraform = exports.iamPrincipalAccessBoundaryPolicyDetailsToTerraform = exports.IamPrincipalAccessBoundaryPolicyDetailsRulesList = exports.IamPrincipalAccessBoundaryPolicyDetailsRulesOutputReference = exports.iamPrincipalAccessBoundaryPolicyDetailsRulesToHclTerraform = exports.iamPrincipalAccessBoundaryPolicyDetailsRulesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iamPrincipalAccessBoundaryPolicyDetailsRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        effect: cdktf.stringToTerraform(struct.effect),
        resources: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.resources),
    };
}
exports.iamPrincipalAccessBoundaryPolicyDetailsRulesToTerraform = iamPrincipalAccessBoundaryPolicyDetailsRulesToTerraform;
function iamPrincipalAccessBoundaryPolicyDetailsRulesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        effect: {
            value: cdktf.stringToHclTerraform(struct.effect),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        resources: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.resources),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamPrincipalAccessBoundaryPolicyDetailsRulesToHclTerraform = iamPrincipalAccessBoundaryPolicyDetailsRulesToHclTerraform;
class IamPrincipalAccessBoundaryPolicyDetailsRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._effect !== undefined) {
            hasAnyValues = true;
            internalValueResult.effect = this._effect;
        }
        if (this._resources !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._effect = undefined;
            this._resources = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._effect = value.effect;
            this._resources = value.resources;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effect() {
        return this.getStringAttribute('effect');
    }
    set effect(value) {
        this._effect = value;
    }
    // Temporarily expose input value. Use with caution.
    get effectInput() {
        return this._effect;
    }
    get resources() {
        return this.getListAttribute('resources');
    }
    set resources(value) {
        this._resources = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources;
    }
}
exports.IamPrincipalAccessBoundaryPolicyDetailsRulesOutputReference = IamPrincipalAccessBoundaryPolicyDetailsRulesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IamPrincipalAccessBoundaryPolicyDetailsRulesOutputReference[_a] = { fqn: "@cdktf/provider-google.iamPrincipalAccessBoundaryPolicy.IamPrincipalAccessBoundaryPolicyDetailsRulesOutputReference", version: "14.13.1" };
class IamPrincipalAccessBoundaryPolicyDetailsRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IamPrincipalAccessBoundaryPolicyDetailsRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IamPrincipalAccessBoundaryPolicyDetailsRulesList = IamPrincipalAccessBoundaryPolicyDetailsRulesList;
_b = JSII_RTTI_SYMBOL_1;
IamPrincipalAccessBoundaryPolicyDetailsRulesList[_b] = { fqn: "@cdktf/provider-google.iamPrincipalAccessBoundaryPolicy.IamPrincipalAccessBoundaryPolicyDetailsRulesList", version: "14.13.1" };
function iamPrincipalAccessBoundaryPolicyDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enforcement_version: cdktf.stringToTerraform(struct.enforcementVersion),
        rules: cdktf.listMapper(iamPrincipalAccessBoundaryPolicyDetailsRulesToTerraform, true)(struct.rules),
    };
}
exports.iamPrincipalAccessBoundaryPolicyDetailsToTerraform = iamPrincipalAccessBoundaryPolicyDetailsToTerraform;
function iamPrincipalAccessBoundaryPolicyDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enforcement_version: {
            value: cdktf.stringToHclTerraform(struct.enforcementVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        rules: {
            value: cdktf.listMapperHcl(iamPrincipalAccessBoundaryPolicyDetailsRulesToHclTerraform, true)(struct.rules),
            isBlock: true,
            type: "list",
            storageClassType: "IamPrincipalAccessBoundaryPolicyDetailsRulesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamPrincipalAccessBoundaryPolicyDetailsToHclTerraform = iamPrincipalAccessBoundaryPolicyDetailsToHclTerraform;
class IamPrincipalAccessBoundaryPolicyDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // rules - computed: false, optional: false, required: true
        this._rules = new IamPrincipalAccessBoundaryPolicyDetailsRulesList(this, "rules", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enforcementVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforcementVersion = this._enforcementVersion;
        }
        if (this._rules?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rules = this._rules?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enforcementVersion = undefined;
            this._rules.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enforcementVersion = value.enforcementVersion;
            this._rules.internalValue = value.rules;
        }
    }
    get enforcementVersion() {
        return this.getStringAttribute('enforcement_version');
    }
    set enforcementVersion(value) {
        this._enforcementVersion = value;
    }
    resetEnforcementVersion() {
        this._enforcementVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforcementVersionInput() {
        return this._enforcementVersion;
    }
    get rules() {
        return this._rules;
    }
    putRules(value) {
        this._rules.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get rulesInput() {
        return this._rules.internalValue;
    }
}
exports.IamPrincipalAccessBoundaryPolicyDetailsOutputReference = IamPrincipalAccessBoundaryPolicyDetailsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IamPrincipalAccessBoundaryPolicyDetailsOutputReference[_c] = { fqn: "@cdktf/provider-google.iamPrincipalAccessBoundaryPolicy.IamPrincipalAccessBoundaryPolicyDetailsOutputReference", version: "14.13.1" };
function iamPrincipalAccessBoundaryPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.iamPrincipalAccessBoundaryPolicyTimeoutsToTerraform = iamPrincipalAccessBoundaryPolicyTimeoutsToTerraform;
function iamPrincipalAccessBoundaryPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamPrincipalAccessBoundaryPolicyTimeoutsToHclTerraform = iamPrincipalAccessBoundaryPolicyTimeoutsToHclTerraform;
class IamPrincipalAccessBoundaryPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.IamPrincipalAccessBoundaryPolicyTimeoutsOutputReference = IamPrincipalAccessBoundaryPolicyTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
IamPrincipalAccessBoundaryPolicyTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-google.iamPrincipalAccessBoundaryPolicy.IamPrincipalAccessBoundaryPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_principal_access_boundary_policy google_iam_principal_access_boundary_policy}
*/
class IamPrincipalAccessBoundaryPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IamPrincipalAccessBoundaryPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IamPrincipalAccessBoundaryPolicy to import
    * @param importFromId The id of the existing IamPrincipalAccessBoundaryPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_principal_access_boundary_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IamPrincipalAccessBoundaryPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_iam_principal_access_boundary_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_principal_access_boundary_policy google_iam_principal_access_boundary_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamPrincipalAccessBoundaryPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_iam_principal_access_boundary_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_annotations - computed: true, optional: false, required: false
        this._effectiveAnnotations = new cdktf.StringMap(this, "effective_annotations");
        // details - computed: false, optional: true, required: false
        this._details = new IamPrincipalAccessBoundaryPolicyDetailsOutputReference(this, "details");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new IamPrincipalAccessBoundaryPolicyTimeoutsOutputReference(this, "timeouts");
        this._annotations = config.annotations;
        this._displayName = config.displayName;
        this._id = config.id;
        this._location = config.location;
        this._organization = config.organization;
        this._principalAccessBoundaryPolicyId = config.principalAccessBoundaryPolicyId;
        this._details.internalValue = config.details;
        this._timeouts.internalValue = config.timeouts;
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get effectiveAnnotations() {
        return this._effectiveAnnotations;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get organization() {
        return this.getStringAttribute('organization');
    }
    set organization(value) {
        this._organization = value;
    }
    // Temporarily expose input value. Use with caution.
    get organizationInput() {
        return this._organization;
    }
    get principalAccessBoundaryPolicyId() {
        return this.getStringAttribute('principal_access_boundary_policy_id');
    }
    set principalAccessBoundaryPolicyId(value) {
        this._principalAccessBoundaryPolicyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get principalAccessBoundaryPolicyIdInput() {
        return this._principalAccessBoundaryPolicyId;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get details() {
        return this._details;
    }
    putDetails(value) {
        this._details.internalValue = value;
    }
    resetDetails() {
        this._details.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get detailsInput() {
        return this._details.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            annotations: cdktf.hashMapper(cdktf.stringToTerraform)(this._annotations),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            organization: cdktf.stringToTerraform(this._organization),
            principal_access_boundary_policy_id: cdktf.stringToTerraform(this._principalAccessBoundaryPolicyId),
            details: iamPrincipalAccessBoundaryPolicyDetailsToTerraform(this._details.internalValue),
            timeouts: iamPrincipalAccessBoundaryPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            annotations: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._annotations),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            organization: {
                value: cdktf.stringToHclTerraform(this._organization),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            principal_access_boundary_policy_id: {
                value: cdktf.stringToHclTerraform(this._principalAccessBoundaryPolicyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            details: {
                value: iamPrincipalAccessBoundaryPolicyDetailsToHclTerraform(this._details.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IamPrincipalAccessBoundaryPolicyDetailsList",
            },
            timeouts: {
                value: iamPrincipalAccessBoundaryPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "IamPrincipalAccessBoundaryPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IamPrincipalAccessBoundaryPolicy = IamPrincipalAccessBoundaryPolicy;
_e = JSII_RTTI_SYMBOL_1;
IamPrincipalAccessBoundaryPolicy[_e] = { fqn: "@cdktf/provider-google.iamPrincipalAccessBoundaryPolicy.IamPrincipalAccessBoundaryPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
IamPrincipalAccessBoundaryPolicy.tfResourceType = "google_iam_principal_access_boundary_policy";
