// @cdktf/provider-google/lib/discovery-engine-search-engine/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscoveryEngineSearchEngine = exports.DiscoveryEngineSearchEngineTimeoutsOutputReference = exports.discoveryEngineSearchEngineTimeoutsToHclTerraform = exports.discoveryEngineSearchEngineTimeoutsToTerraform = exports.DiscoveryEngineSearchEngineSearchEngineConfigOutputReference = exports.discoveryEngineSearchEngineSearchEngineConfigToHclTerraform = exports.discoveryEngineSearchEngineSearchEngineConfigToTerraform = exports.DiscoveryEngineSearchEngineCommonConfigOutputReference = exports.discoveryEngineSearchEngineCommonConfigToHclTerraform = exports.discoveryEngineSearchEngineCommonConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function discoveryEngineSearchEngineCommonConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        company_name: cdktf.stringToTerraform(struct.companyName),
    };
}
exports.discoveryEngineSearchEngineCommonConfigToTerraform = discoveryEngineSearchEngineCommonConfigToTerraform;
function discoveryEngineSearchEngineCommonConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        company_name: {
            value: cdktf.stringToHclTerraform(struct.companyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.discoveryEngineSearchEngineCommonConfigToHclTerraform = discoveryEngineSearchEngineCommonConfigToHclTerraform;
class DiscoveryEngineSearchEngineCommonConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._companyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.companyName = this._companyName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._companyName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._companyName = value.companyName;
        }
    }
    get companyName() {
        return this.getStringAttribute('company_name');
    }
    set companyName(value) {
        this._companyName = value;
    }
    resetCompanyName() {
        this._companyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get companyNameInput() {
        return this._companyName;
    }
}
exports.DiscoveryEngineSearchEngineCommonConfigOutputReference = DiscoveryEngineSearchEngineCommonConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DiscoveryEngineSearchEngineCommonConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.discoveryEngineSearchEngine.DiscoveryEngineSearchEngineCommonConfigOutputReference", version: "14.13.1" };
function discoveryEngineSearchEngineSearchEngineConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        search_add_ons: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.searchAddOns),
        search_tier: cdktf.stringToTerraform(struct.searchTier),
    };
}
exports.discoveryEngineSearchEngineSearchEngineConfigToTerraform = discoveryEngineSearchEngineSearchEngineConfigToTerraform;
function discoveryEngineSearchEngineSearchEngineConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        search_add_ons: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.searchAddOns),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        search_tier: {
            value: cdktf.stringToHclTerraform(struct.searchTier),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.discoveryEngineSearchEngineSearchEngineConfigToHclTerraform = discoveryEngineSearchEngineSearchEngineConfigToHclTerraform;
class DiscoveryEngineSearchEngineSearchEngineConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._searchAddOns !== undefined) {
            hasAnyValues = true;
            internalValueResult.searchAddOns = this._searchAddOns;
        }
        if (this._searchTier !== undefined) {
            hasAnyValues = true;
            internalValueResult.searchTier = this._searchTier;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._searchAddOns = undefined;
            this._searchTier = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._searchAddOns = value.searchAddOns;
            this._searchTier = value.searchTier;
        }
    }
    get searchAddOns() {
        return this.getListAttribute('search_add_ons');
    }
    set searchAddOns(value) {
        this._searchAddOns = value;
    }
    resetSearchAddOns() {
        this._searchAddOns = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get searchAddOnsInput() {
        return this._searchAddOns;
    }
    get searchTier() {
        return this.getStringAttribute('search_tier');
    }
    set searchTier(value) {
        this._searchTier = value;
    }
    resetSearchTier() {
        this._searchTier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get searchTierInput() {
        return this._searchTier;
    }
}
exports.DiscoveryEngineSearchEngineSearchEngineConfigOutputReference = DiscoveryEngineSearchEngineSearchEngineConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DiscoveryEngineSearchEngineSearchEngineConfigOutputReference[_b] = { fqn: "@cdktf/provider-google.discoveryEngineSearchEngine.DiscoveryEngineSearchEngineSearchEngineConfigOutputReference", version: "14.13.1" };
function discoveryEngineSearchEngineTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.discoveryEngineSearchEngineTimeoutsToTerraform = discoveryEngineSearchEngineTimeoutsToTerraform;
function discoveryEngineSearchEngineTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.discoveryEngineSearchEngineTimeoutsToHclTerraform = discoveryEngineSearchEngineTimeoutsToHclTerraform;
class DiscoveryEngineSearchEngineTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DiscoveryEngineSearchEngineTimeoutsOutputReference = DiscoveryEngineSearchEngineTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DiscoveryEngineSearchEngineTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-google.discoveryEngineSearchEngine.DiscoveryEngineSearchEngineTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/discovery_engine_search_engine google_discovery_engine_search_engine}
*/
class DiscoveryEngineSearchEngine extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DiscoveryEngineSearchEngine resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DiscoveryEngineSearchEngine to import
    * @param importFromId The id of the existing DiscoveryEngineSearchEngine that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/discovery_engine_search_engine#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DiscoveryEngineSearchEngine to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_discovery_engine_search_engine", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/discovery_engine_search_engine google_discovery_engine_search_engine} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DiscoveryEngineSearchEngineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_discovery_engine_search_engine',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // common_config - computed: false, optional: true, required: false
        this._commonConfig = new DiscoveryEngineSearchEngineCommonConfigOutputReference(this, "common_config");
        // search_engine_config - computed: false, optional: false, required: true
        this._searchEngineConfig = new DiscoveryEngineSearchEngineSearchEngineConfigOutputReference(this, "search_engine_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DiscoveryEngineSearchEngineTimeoutsOutputReference(this, "timeouts");
        this._collectionId = config.collectionId;
        this._dataStoreIds = config.dataStoreIds;
        this._displayName = config.displayName;
        this._engineId = config.engineId;
        this._id = config.id;
        this._industryVertical = config.industryVertical;
        this._location = config.location;
        this._project = config.project;
        this._commonConfig.internalValue = config.commonConfig;
        this._searchEngineConfig.internalValue = config.searchEngineConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    get collectionId() {
        return this.getStringAttribute('collection_id');
    }
    set collectionId(value) {
        this._collectionId = value;
    }
    // Temporarily expose input value. Use with caution.
    get collectionIdInput() {
        return this._collectionId;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get dataStoreIds() {
        return this.getListAttribute('data_store_ids');
    }
    set dataStoreIds(value) {
        this._dataStoreIds = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataStoreIdsInput() {
        return this._dataStoreIds;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get engineId() {
        return this.getStringAttribute('engine_id');
    }
    set engineId(value) {
        this._engineId = value;
    }
    // Temporarily expose input value. Use with caution.
    get engineIdInput() {
        return this._engineId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get industryVertical() {
        return this.getStringAttribute('industry_vertical');
    }
    set industryVertical(value) {
        this._industryVertical = value;
    }
    resetIndustryVertical() {
        this._industryVertical = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get industryVerticalInput() {
        return this._industryVertical;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get commonConfig() {
        return this._commonConfig;
    }
    putCommonConfig(value) {
        this._commonConfig.internalValue = value;
    }
    resetCommonConfig() {
        this._commonConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get commonConfigInput() {
        return this._commonConfig.internalValue;
    }
    get searchEngineConfig() {
        return this._searchEngineConfig;
    }
    putSearchEngineConfig(value) {
        this._searchEngineConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get searchEngineConfigInput() {
        return this._searchEngineConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            collection_id: cdktf.stringToTerraform(this._collectionId),
            data_store_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._dataStoreIds),
            display_name: cdktf.stringToTerraform(this._displayName),
            engine_id: cdktf.stringToTerraform(this._engineId),
            id: cdktf.stringToTerraform(this._id),
            industry_vertical: cdktf.stringToTerraform(this._industryVertical),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            common_config: discoveryEngineSearchEngineCommonConfigToTerraform(this._commonConfig.internalValue),
            search_engine_config: discoveryEngineSearchEngineSearchEngineConfigToTerraform(this._searchEngineConfig.internalValue),
            timeouts: discoveryEngineSearchEngineTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            collection_id: {
                value: cdktf.stringToHclTerraform(this._collectionId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_store_ids: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._dataStoreIds),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            engine_id: {
                value: cdktf.stringToHclTerraform(this._engineId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            industry_vertical: {
                value: cdktf.stringToHclTerraform(this._industryVertical),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            common_config: {
                value: discoveryEngineSearchEngineCommonConfigToHclTerraform(this._commonConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DiscoveryEngineSearchEngineCommonConfigList",
            },
            search_engine_config: {
                value: discoveryEngineSearchEngineSearchEngineConfigToHclTerraform(this._searchEngineConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DiscoveryEngineSearchEngineSearchEngineConfigList",
            },
            timeouts: {
                value: discoveryEngineSearchEngineTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DiscoveryEngineSearchEngineTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DiscoveryEngineSearchEngine = DiscoveryEngineSearchEngine;
_d = JSII_RTTI_SYMBOL_1;
DiscoveryEngineSearchEngine[_d] = { fqn: "@cdktf/provider-google.discoveryEngineSearchEngine.DiscoveryEngineSearchEngine", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DiscoveryEngineSearchEngine.tfResourceType = "google_discovery_engine_search_engine";
