// @cdktf/cli-core/src/lib/models/terraform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractTerraformPlan = exports.isTerraformOutput = exports.DeployingResourceApplyState = exports.PlannedResourceAction = void 0;
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
var PlannedResourceAction;
(function (PlannedResourceAction) {
    PlannedResourceAction["CREATE"] = "create";
    PlannedResourceAction["UPDATE"] = "update";
    PlannedResourceAction["DELETE"] = "delete";
    PlannedResourceAction["READ"] = "read";
    PlannedResourceAction["NO_OP"] = "no-op";
})(PlannedResourceAction || (exports.PlannedResourceAction = PlannedResourceAction = {}));
var DeployingResourceApplyState;
(function (DeployingResourceApplyState) {
    DeployingResourceApplyState["WAITING"] = "waiting";
    DeployingResourceApplyState["UPDATING"] = "updating";
    DeployingResourceApplyState["UPDATED"] = "updated";
    DeployingResourceApplyState["CREATING"] = "creating";
    DeployingResourceApplyState["CREATED"] = "created";
    DeployingResourceApplyState["DESTROYING"] = "destroying";
    DeployingResourceApplyState["DESTROYED"] = "destroyed";
    DeployingResourceApplyState["SUCCESS"] = "success";
    DeployingResourceApplyState["ERROR"] = "error";
})(DeployingResourceApplyState || (exports.DeployingResourceApplyState = DeployingResourceApplyState = {}));
function isTerraformOutput(output) {
    return (typeof output === "object" &&
        typeof output.sensitive === "boolean" &&
        (typeof output.type === "string" || Array.isArray(output.type)));
}
exports.isTerraformOutput = isTerraformOutput;
function filterChangingResources(resources) {
    const applyActions = [
        PlannedResourceAction.UPDATE,
        PlannedResourceAction.CREATE,
        PlannedResourceAction.DELETE,
        PlannedResourceAction.READ,
    ];
    return resources.filter((resource) => applyActions.includes(resource.action));
}
class AbstractTerraformPlan {
    constructor(planFile, resourceChanges, outputChanges) {
        this.planFile = planFile;
        this.resourceChanges = resourceChanges;
        this.outputChanges = outputChanges;
    }
    get resources() {
        return (this.resourceChanges || []).map((resource) => {
            return {
                id: resource.address,
                action: resource.change.actions[0],
            };
        });
    }
    get applyableResources() {
        return filterChangingResources(this.resources);
    }
    get outputs() {
        return Object.entries(this.outputChanges || {}).map(([key, value]) => ({
            id: `output.${key}`,
            action: value.actions[0],
        }));
    }
    get changingOutputs() {
        return filterChangingResources(this.outputs);
    }
    get needsApply() {
        return (this.applyableResources.length > 0 || this.changingOutputs.length > 0);
    }
}
exports.AbstractTerraformPlan = AbstractTerraformPlan;
