// @cdktf/provider-google/lib/cloud-quotas-quota-preference/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudQuotasQuotaPreference = exports.CloudQuotasQuotaPreferenceTimeoutsOutputReference = exports.cloudQuotasQuotaPreferenceTimeoutsToHclTerraform = exports.cloudQuotasQuotaPreferenceTimeoutsToTerraform = exports.CloudQuotasQuotaPreferenceQuotaConfigOutputReference = exports.cloudQuotasQuotaPreferenceQuotaConfigToHclTerraform = exports.cloudQuotasQuotaPreferenceQuotaConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cloudQuotasQuotaPreferenceQuotaConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        annotations: cdktf.hashMapper(cdktf.stringToTerraform)(struct.annotations),
        preferred_value: cdktf.stringToTerraform(struct.preferredValue),
    };
}
exports.cloudQuotasQuotaPreferenceQuotaConfigToTerraform = cloudQuotasQuotaPreferenceQuotaConfigToTerraform;
function cloudQuotasQuotaPreferenceQuotaConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        annotations: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.annotations),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        preferred_value: {
            value: cdktf.stringToHclTerraform(struct.preferredValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudQuotasQuotaPreferenceQuotaConfigToHclTerraform = cloudQuotasQuotaPreferenceQuotaConfigToHclTerraform;
class CloudQuotasQuotaPreferenceQuotaConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._annotations !== undefined) {
            hasAnyValues = true;
            internalValueResult.annotations = this._annotations;
        }
        if (this._preferredValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.preferredValue = this._preferredValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._annotations = undefined;
            this._preferredValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._annotations = value.annotations;
            this._preferredValue = value.preferredValue;
        }
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    // granted_value - computed: true, optional: false, required: false
    get grantedValue() {
        return this.getStringAttribute('granted_value');
    }
    get preferredValue() {
        return this.getStringAttribute('preferred_value');
    }
    set preferredValue(value) {
        this._preferredValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get preferredValueInput() {
        return this._preferredValue;
    }
    // request_origin - computed: true, optional: false, required: false
    get requestOrigin() {
        return this.getStringAttribute('request_origin');
    }
    // state_detail - computed: true, optional: false, required: false
    get stateDetail() {
        return this.getStringAttribute('state_detail');
    }
    // trace_id - computed: true, optional: false, required: false
    get traceId() {
        return this.getStringAttribute('trace_id');
    }
}
exports.CloudQuotasQuotaPreferenceQuotaConfigOutputReference = CloudQuotasQuotaPreferenceQuotaConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CloudQuotasQuotaPreferenceQuotaConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.cloudQuotasQuotaPreference.CloudQuotasQuotaPreferenceQuotaConfigOutputReference", version: "14.13.1" };
function cloudQuotasQuotaPreferenceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cloudQuotasQuotaPreferenceTimeoutsToTerraform = cloudQuotasQuotaPreferenceTimeoutsToTerraform;
function cloudQuotasQuotaPreferenceTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.cloudQuotasQuotaPreferenceTimeoutsToHclTerraform = cloudQuotasQuotaPreferenceTimeoutsToHclTerraform;
class CloudQuotasQuotaPreferenceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CloudQuotasQuotaPreferenceTimeoutsOutputReference = CloudQuotasQuotaPreferenceTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
CloudQuotasQuotaPreferenceTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-google.cloudQuotasQuotaPreference.CloudQuotasQuotaPreferenceTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/cloud_quotas_quota_preference google_cloud_quotas_quota_preference}
*/
class CloudQuotasQuotaPreference extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a CloudQuotasQuotaPreference resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the CloudQuotasQuotaPreference to import
    * @param importFromId The id of the existing CloudQuotasQuotaPreference that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/cloud_quotas_quota_preference#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the CloudQuotasQuotaPreference to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_cloud_quotas_quota_preference", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/cloud_quotas_quota_preference google_cloud_quotas_quota_preference} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CloudQuotasQuotaPreferenceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloud_quotas_quota_preference',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // quota_config - computed: false, optional: false, required: true
        this._quotaConfig = new CloudQuotasQuotaPreferenceQuotaConfigOutputReference(this, "quota_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CloudQuotasQuotaPreferenceTimeoutsOutputReference(this, "timeouts");
        this._contactEmail = config.contactEmail;
        this._dimensions = config.dimensions;
        this._id = config.id;
        this._ignoreSafetyChecks = config.ignoreSafetyChecks;
        this._justification = config.justification;
        this._name = config.name;
        this._parent = config.parent;
        this._quotaId = config.quotaId;
        this._service = config.service;
        this._quotaConfig.internalValue = config.quotaConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    get contactEmail() {
        return this.getStringAttribute('contact_email');
    }
    set contactEmail(value) {
        this._contactEmail = value;
    }
    resetContactEmail() {
        this._contactEmail = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get contactEmailInput() {
        return this._contactEmail;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get dimensions() {
        return this.getStringMapAttribute('dimensions');
    }
    set dimensions(value) {
        this._dimensions = value;
    }
    resetDimensions() {
        this._dimensions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dimensionsInput() {
        return this._dimensions;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ignoreSafetyChecks() {
        return this.getStringAttribute('ignore_safety_checks');
    }
    set ignoreSafetyChecks(value) {
        this._ignoreSafetyChecks = value;
    }
    resetIgnoreSafetyChecks() {
        this._ignoreSafetyChecks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ignoreSafetyChecksInput() {
        return this._ignoreSafetyChecks;
    }
    get justification() {
        return this.getStringAttribute('justification');
    }
    set justification(value) {
        this._justification = value;
    }
    resetJustification() {
        this._justification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get justificationInput() {
        return this._justification;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    resetParent() {
        this._parent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    get quotaId() {
        return this.getStringAttribute('quota_id');
    }
    set quotaId(value) {
        this._quotaId = value;
    }
    resetQuotaId() {
        this._quotaId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get quotaIdInput() {
        return this._quotaId;
    }
    // reconciling - computed: true, optional: false, required: false
    get reconciling() {
        return this.getBooleanAttribute('reconciling');
    }
    get service() {
        return this.getStringAttribute('service');
    }
    set service(value) {
        this._service = value;
    }
    resetService() {
        this._service = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceInput() {
        return this._service;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get quotaConfig() {
        return this._quotaConfig;
    }
    putQuotaConfig(value) {
        this._quotaConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get quotaConfigInput() {
        return this._quotaConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            contact_email: cdktf.stringToTerraform(this._contactEmail),
            dimensions: cdktf.hashMapper(cdktf.stringToTerraform)(this._dimensions),
            id: cdktf.stringToTerraform(this._id),
            ignore_safety_checks: cdktf.stringToTerraform(this._ignoreSafetyChecks),
            justification: cdktf.stringToTerraform(this._justification),
            name: cdktf.stringToTerraform(this._name),
            parent: cdktf.stringToTerraform(this._parent),
            quota_id: cdktf.stringToTerraform(this._quotaId),
            service: cdktf.stringToTerraform(this._service),
            quota_config: cloudQuotasQuotaPreferenceQuotaConfigToTerraform(this._quotaConfig.internalValue),
            timeouts: cloudQuotasQuotaPreferenceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            contact_email: {
                value: cdktf.stringToHclTerraform(this._contactEmail),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dimensions: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._dimensions),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            ignore_safety_checks: {
                value: cdktf.stringToHclTerraform(this._ignoreSafetyChecks),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            justification: {
                value: cdktf.stringToHclTerraform(this._justification),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent: {
                value: cdktf.stringToHclTerraform(this._parent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            quota_id: {
                value: cdktf.stringToHclTerraform(this._quotaId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service: {
                value: cdktf.stringToHclTerraform(this._service),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            quota_config: {
                value: cloudQuotasQuotaPreferenceQuotaConfigToHclTerraform(this._quotaConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "CloudQuotasQuotaPreferenceQuotaConfigList",
            },
            timeouts: {
                value: cloudQuotasQuotaPreferenceTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "CloudQuotasQuotaPreferenceTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.CloudQuotasQuotaPreference = CloudQuotasQuotaPreference;
_c = JSII_RTTI_SYMBOL_1;
CloudQuotasQuotaPreference[_c] = { fqn: "@cdktf/provider-google.cloudQuotasQuotaPreference.CloudQuotasQuotaPreference", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
CloudQuotasQuotaPreference.tfResourceType = "google_cloud_quotas_quota_preference";
