// @cdktf/provider-google/lib/gke-backup-backup-plan/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GkeBackupBackupPlan = exports.GkeBackupBackupPlanTimeoutsOutputReference = exports.gkeBackupBackupPlanTimeoutsToHclTerraform = exports.gkeBackupBackupPlanTimeoutsToTerraform = exports.GkeBackupBackupPlanRetentionPolicyOutputReference = exports.gkeBackupBackupPlanRetentionPolicyToHclTerraform = exports.gkeBackupBackupPlanRetentionPolicyToTerraform = exports.GkeBackupBackupPlanBackupScheduleOutputReference = exports.gkeBackupBackupPlanBackupScheduleToHclTerraform = exports.gkeBackupBackupPlanBackupScheduleToTerraform = exports.GkeBackupBackupPlanBackupScheduleRpoConfigOutputReference = exports.gkeBackupBackupPlanBackupScheduleRpoConfigToHclTerraform = exports.gkeBackupBackupPlanBackupScheduleRpoConfigToTerraform = exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsList = exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsOutputReference = exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToHclTerraform = exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToTerraform = exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeOutputReference = exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToHclTerraform = exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToTerraform = exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateOutputReference = exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToHclTerraform = exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToTerraform = exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekOutputReference = exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToHclTerraform = exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToTerraform = exports.GkeBackupBackupPlanBackupConfigOutputReference = exports.gkeBackupBackupPlanBackupConfigToHclTerraform = exports.gkeBackupBackupPlanBackupConfigToTerraform = exports.GkeBackupBackupPlanBackupConfigSelectedNamespacesOutputReference = exports.gkeBackupBackupPlanBackupConfigSelectedNamespacesToHclTerraform = exports.gkeBackupBackupPlanBackupConfigSelectedNamespacesToTerraform = exports.GkeBackupBackupPlanBackupConfigSelectedApplicationsOutputReference = exports.gkeBackupBackupPlanBackupConfigSelectedApplicationsToHclTerraform = exports.gkeBackupBackupPlanBackupConfigSelectedApplicationsToTerraform = exports.GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesList = exports.GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesOutputReference = exports.gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToHclTerraform = exports.gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToTerraform = exports.GkeBackupBackupPlanBackupConfigEncryptionKeyOutputReference = exports.gkeBackupBackupPlanBackupConfigEncryptionKeyToHclTerraform = exports.gkeBackupBackupPlanBackupConfigEncryptionKeyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function gkeBackupBackupPlanBackupConfigEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        gcp_kms_encryption_key: cdktf.stringToTerraform(struct.gcpKmsEncryptionKey),
    };
}
exports.gkeBackupBackupPlanBackupConfigEncryptionKeyToTerraform = gkeBackupBackupPlanBackupConfigEncryptionKeyToTerraform;
function gkeBackupBackupPlanBackupConfigEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        gcp_kms_encryption_key: {
            value: cdktf.stringToHclTerraform(struct.gcpKmsEncryptionKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupConfigEncryptionKeyToHclTerraform = gkeBackupBackupPlanBackupConfigEncryptionKeyToHclTerraform;
class GkeBackupBackupPlanBackupConfigEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._gcpKmsEncryptionKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.gcpKmsEncryptionKey = this._gcpKmsEncryptionKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._gcpKmsEncryptionKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._gcpKmsEncryptionKey = value.gcpKmsEncryptionKey;
        }
    }
    get gcpKmsEncryptionKey() {
        return this.getStringAttribute('gcp_kms_encryption_key');
    }
    set gcpKmsEncryptionKey(value) {
        this._gcpKmsEncryptionKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get gcpKmsEncryptionKeyInput() {
        return this._gcpKmsEncryptionKey;
    }
}
exports.GkeBackupBackupPlanBackupConfigEncryptionKeyOutputReference = GkeBackupBackupPlanBackupConfigEncryptionKeyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupConfigEncryptionKeyOutputReference[_a] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupConfigEncryptionKeyOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        namespace: cdktf.stringToTerraform(struct.namespace),
    };
}
exports.gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToTerraform = gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToTerraform;
function gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        namespace: {
            value: cdktf.stringToHclTerraform(struct.namespace),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToHclTerraform = gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToHclTerraform;
class GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._namespace !== undefined) {
            hasAnyValues = true;
            internalValueResult.namespace = this._namespace;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._namespace = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._namespace = value.namespace;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
}
exports.GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesOutputReference = GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesOutputReference[_b] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesOutputReference", version: "14.13.1" };
class GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesList = GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesList;
_c = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesList[_c] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesList", version: "14.13.1" };
function gkeBackupBackupPlanBackupConfigSelectedApplicationsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        namespaced_names: cdktf.listMapper(gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToTerraform, true)(struct.namespacedNames),
    };
}
exports.gkeBackupBackupPlanBackupConfigSelectedApplicationsToTerraform = gkeBackupBackupPlanBackupConfigSelectedApplicationsToTerraform;
function gkeBackupBackupPlanBackupConfigSelectedApplicationsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        namespaced_names: {
            value: cdktf.listMapperHcl(gkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesToHclTerraform, true)(struct.namespacedNames),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupConfigSelectedApplicationsToHclTerraform = gkeBackupBackupPlanBackupConfigSelectedApplicationsToHclTerraform;
class GkeBackupBackupPlanBackupConfigSelectedApplicationsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // namespaced_names - computed: false, optional: false, required: true
        this._namespacedNames = new GkeBackupBackupPlanBackupConfigSelectedApplicationsNamespacedNamesList(this, "namespaced_names", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._namespacedNames?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.namespacedNames = this._namespacedNames?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._namespacedNames.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._namespacedNames.internalValue = value.namespacedNames;
        }
    }
    get namespacedNames() {
        return this._namespacedNames;
    }
    putNamespacedNames(value) {
        this._namespacedNames.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespacedNamesInput() {
        return this._namespacedNames.internalValue;
    }
}
exports.GkeBackupBackupPlanBackupConfigSelectedApplicationsOutputReference = GkeBackupBackupPlanBackupConfigSelectedApplicationsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupConfigSelectedApplicationsOutputReference[_d] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupConfigSelectedApplicationsOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanBackupConfigSelectedNamespacesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        namespaces: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.namespaces),
    };
}
exports.gkeBackupBackupPlanBackupConfigSelectedNamespacesToTerraform = gkeBackupBackupPlanBackupConfigSelectedNamespacesToTerraform;
function gkeBackupBackupPlanBackupConfigSelectedNamespacesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        namespaces: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.namespaces),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupConfigSelectedNamespacesToHclTerraform = gkeBackupBackupPlanBackupConfigSelectedNamespacesToHclTerraform;
class GkeBackupBackupPlanBackupConfigSelectedNamespacesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._namespaces !== undefined) {
            hasAnyValues = true;
            internalValueResult.namespaces = this._namespaces;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._namespaces = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._namespaces = value.namespaces;
        }
    }
    get namespaces() {
        return this.getListAttribute('namespaces');
    }
    set namespaces(value) {
        this._namespaces = value;
    }
    // Temporarily expose input value. Use with caution.
    get namespacesInput() {
        return this._namespaces;
    }
}
exports.GkeBackupBackupPlanBackupConfigSelectedNamespacesOutputReference = GkeBackupBackupPlanBackupConfigSelectedNamespacesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupConfigSelectedNamespacesOutputReference[_e] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupConfigSelectedNamespacesOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanBackupConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        all_namespaces: cdktf.booleanToTerraform(struct.allNamespaces),
        include_secrets: cdktf.booleanToTerraform(struct.includeSecrets),
        include_volume_data: cdktf.booleanToTerraform(struct.includeVolumeData),
        permissive_mode: cdktf.booleanToTerraform(struct.permissiveMode),
        encryption_key: gkeBackupBackupPlanBackupConfigEncryptionKeyToTerraform(struct.encryptionKey),
        selected_applications: gkeBackupBackupPlanBackupConfigSelectedApplicationsToTerraform(struct.selectedApplications),
        selected_namespaces: gkeBackupBackupPlanBackupConfigSelectedNamespacesToTerraform(struct.selectedNamespaces),
    };
}
exports.gkeBackupBackupPlanBackupConfigToTerraform = gkeBackupBackupPlanBackupConfigToTerraform;
function gkeBackupBackupPlanBackupConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        all_namespaces: {
            value: cdktf.booleanToHclTerraform(struct.allNamespaces),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        include_secrets: {
            value: cdktf.booleanToHclTerraform(struct.includeSecrets),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        include_volume_data: {
            value: cdktf.booleanToHclTerraform(struct.includeVolumeData),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        permissive_mode: {
            value: cdktf.booleanToHclTerraform(struct.permissiveMode),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        encryption_key: {
            value: gkeBackupBackupPlanBackupConfigEncryptionKeyToHclTerraform(struct.encryptionKey),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupConfigEncryptionKeyList",
        },
        selected_applications: {
            value: gkeBackupBackupPlanBackupConfigSelectedApplicationsToHclTerraform(struct.selectedApplications),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupConfigSelectedApplicationsList",
        },
        selected_namespaces: {
            value: gkeBackupBackupPlanBackupConfigSelectedNamespacesToHclTerraform(struct.selectedNamespaces),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupConfigSelectedNamespacesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupConfigToHclTerraform = gkeBackupBackupPlanBackupConfigToHclTerraform;
class GkeBackupBackupPlanBackupConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // encryption_key - computed: false, optional: true, required: false
        this._encryptionKey = new GkeBackupBackupPlanBackupConfigEncryptionKeyOutputReference(this, "encryption_key");
        // selected_applications - computed: false, optional: true, required: false
        this._selectedApplications = new GkeBackupBackupPlanBackupConfigSelectedApplicationsOutputReference(this, "selected_applications");
        // selected_namespaces - computed: false, optional: true, required: false
        this._selectedNamespaces = new GkeBackupBackupPlanBackupConfigSelectedNamespacesOutputReference(this, "selected_namespaces");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._allNamespaces !== undefined) {
            hasAnyValues = true;
            internalValueResult.allNamespaces = this._allNamespaces;
        }
        if (this._includeSecrets !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeSecrets = this._includeSecrets;
        }
        if (this._includeVolumeData !== undefined) {
            hasAnyValues = true;
            internalValueResult.includeVolumeData = this._includeVolumeData;
        }
        if (this._permissiveMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.permissiveMode = this._permissiveMode;
        }
        if (this._encryptionKey?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptionKey = this._encryptionKey?.internalValue;
        }
        if (this._selectedApplications?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.selectedApplications = this._selectedApplications?.internalValue;
        }
        if (this._selectedNamespaces?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.selectedNamespaces = this._selectedNamespaces?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._allNamespaces = undefined;
            this._includeSecrets = undefined;
            this._includeVolumeData = undefined;
            this._permissiveMode = undefined;
            this._encryptionKey.internalValue = undefined;
            this._selectedApplications.internalValue = undefined;
            this._selectedNamespaces.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._allNamespaces = value.allNamespaces;
            this._includeSecrets = value.includeSecrets;
            this._includeVolumeData = value.includeVolumeData;
            this._permissiveMode = value.permissiveMode;
            this._encryptionKey.internalValue = value.encryptionKey;
            this._selectedApplications.internalValue = value.selectedApplications;
            this._selectedNamespaces.internalValue = value.selectedNamespaces;
        }
    }
    get allNamespaces() {
        return this.getBooleanAttribute('all_namespaces');
    }
    set allNamespaces(value) {
        this._allNamespaces = value;
    }
    resetAllNamespaces() {
        this._allNamespaces = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allNamespacesInput() {
        return this._allNamespaces;
    }
    get includeSecrets() {
        return this.getBooleanAttribute('include_secrets');
    }
    set includeSecrets(value) {
        this._includeSecrets = value;
    }
    resetIncludeSecrets() {
        this._includeSecrets = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeSecretsInput() {
        return this._includeSecrets;
    }
    get includeVolumeData() {
        return this.getBooleanAttribute('include_volume_data');
    }
    set includeVolumeData(value) {
        this._includeVolumeData = value;
    }
    resetIncludeVolumeData() {
        this._includeVolumeData = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeVolumeDataInput() {
        return this._includeVolumeData;
    }
    get permissiveMode() {
        return this.getBooleanAttribute('permissive_mode');
    }
    set permissiveMode(value) {
        this._permissiveMode = value;
    }
    resetPermissiveMode() {
        this._permissiveMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissiveModeInput() {
        return this._permissiveMode;
    }
    get encryptionKey() {
        return this._encryptionKey;
    }
    putEncryptionKey(value) {
        this._encryptionKey.internalValue = value;
    }
    resetEncryptionKey() {
        this._encryptionKey.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionKeyInput() {
        return this._encryptionKey.internalValue;
    }
    get selectedApplications() {
        return this._selectedApplications;
    }
    putSelectedApplications(value) {
        this._selectedApplications.internalValue = value;
    }
    resetSelectedApplications() {
        this._selectedApplications.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selectedApplicationsInput() {
        return this._selectedApplications.internalValue;
    }
    get selectedNamespaces() {
        return this._selectedNamespaces;
    }
    putSelectedNamespaces(value) {
        this._selectedNamespaces.internalValue = value;
    }
    resetSelectedNamespaces() {
        this._selectedNamespaces.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get selectedNamespacesInput() {
        return this._selectedNamespaces.internalValue;
    }
}
exports.GkeBackupBackupPlanBackupConfigOutputReference = GkeBackupBackupPlanBackupConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupConfigOutputReference[_f] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupConfigOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        days_of_week: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.daysOfWeek),
    };
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToTerraform;
function gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        days_of_week: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.daysOfWeek),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToHclTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToHclTerraform;
class GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._daysOfWeek !== undefined) {
            hasAnyValues = true;
            internalValueResult.daysOfWeek = this._daysOfWeek;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._daysOfWeek = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._daysOfWeek = value.daysOfWeek;
        }
    }
    get daysOfWeek() {
        return this.getListAttribute('days_of_week');
    }
    set daysOfWeek(value) {
        this._daysOfWeek = value;
    }
    resetDaysOfWeek() {
        this._daysOfWeek = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get daysOfWeekInput() {
        return this._daysOfWeek;
    }
}
exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekOutputReference = GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekOutputReference;
_g = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekOutputReference[_g] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.numberToTerraform(struct.day),
        month: cdktf.numberToTerraform(struct.month),
        year: cdktf.numberToTerraform(struct.year),
    };
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToTerraform;
function gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        day: {
            value: cdktf.numberToHclTerraform(struct.day),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        month: {
            value: cdktf.numberToHclTerraform(struct.month),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        year: {
            value: cdktf.numberToHclTerraform(struct.year),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToHclTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToHclTerraform;
class GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._month !== undefined) {
            hasAnyValues = true;
            internalValueResult.month = this._month;
        }
        if (this._year !== undefined) {
            hasAnyValues = true;
            internalValueResult.year = this._year;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._month = undefined;
            this._year = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._month = value.month;
            this._year = value.year;
        }
    }
    get day() {
        return this.getNumberAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    resetDay() {
        this._day = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get month() {
        return this.getNumberAttribute('month');
    }
    set month(value) {
        this._month = value;
    }
    resetMonth() {
        this._month = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get monthInput() {
        return this._month;
    }
    get year() {
        return this.getNumberAttribute('year');
    }
    set year(value) {
        this._year = value;
    }
    resetYear() {
        this._year = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get yearInput() {
        return this._year;
    }
}
exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateOutputReference = GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateOutputReference;
_h = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateOutputReference[_h] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hours: cdktf.numberToTerraform(struct.hours),
        minutes: cdktf.numberToTerraform(struct.minutes),
        nanos: cdktf.numberToTerraform(struct.nanos),
        seconds: cdktf.numberToTerraform(struct.seconds),
    };
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToTerraform;
function gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        hours: {
            value: cdktf.numberToHclTerraform(struct.hours),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minutes: {
            value: cdktf.numberToHclTerraform(struct.minutes),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        nanos: {
            value: cdktf.numberToHclTerraform(struct.nanos),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        seconds: {
            value: cdktf.numberToHclTerraform(struct.seconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToHclTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToHclTerraform;
class GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hours !== undefined) {
            hasAnyValues = true;
            internalValueResult.hours = this._hours;
        }
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        if (this._nanos !== undefined) {
            hasAnyValues = true;
            internalValueResult.nanos = this._nanos;
        }
        if (this._seconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.seconds = this._seconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hours = undefined;
            this._minutes = undefined;
            this._nanos = undefined;
            this._seconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hours = value.hours;
            this._minutes = value.minutes;
            this._nanos = value.nanos;
            this._seconds = value.seconds;
        }
    }
    get hours() {
        return this.getNumberAttribute('hours');
    }
    set hours(value) {
        this._hours = value;
    }
    resetHours() {
        this._hours = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hoursInput() {
        return this._hours;
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    resetMinutes() {
        this._minutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    set nanos(value) {
        this._nanos = value;
    }
    resetNanos() {
        this._nanos = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nanosInput() {
        return this._nanos;
    }
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
    set seconds(value) {
        this._seconds = value;
    }
    resetSeconds() {
        this._seconds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secondsInput() {
        return this._seconds;
    }
}
exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeOutputReference = GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeOutputReference;
_j = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeOutputReference[_j] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        daily: cdktf.booleanToTerraform(struct.daily),
        duration: cdktf.stringToTerraform(struct.duration),
        days_of_week: gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToTerraform(struct.daysOfWeek),
        single_occurrence_date: gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToTerraform(struct.singleOccurrenceDate),
        start_time: gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToTerraform(struct.startTime),
    };
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToTerraform;
function gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        daily: {
            value: cdktf.booleanToHclTerraform(struct.daily),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        duration: {
            value: cdktf.stringToHclTerraform(struct.duration),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        days_of_week: {
            value: gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekToHclTerraform(struct.daysOfWeek),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekList",
        },
        single_occurrence_date: {
            value: gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateToHclTerraform(struct.singleOccurrenceDate),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateList",
        },
        start_time: {
            value: gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeToHclTerraform(struct.startTime),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToHclTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToHclTerraform;
class GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // days_of_week - computed: false, optional: true, required: false
        this._daysOfWeek = new GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsDaysOfWeekOutputReference(this, "days_of_week");
        // single_occurrence_date - computed: false, optional: true, required: false
        this._singleOccurrenceDate = new GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsSingleOccurrenceDateOutputReference(this, "single_occurrence_date");
        // start_time - computed: false, optional: false, required: true
        this._startTime = new GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsStartTimeOutputReference(this, "start_time");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._daily !== undefined) {
            hasAnyValues = true;
            internalValueResult.daily = this._daily;
        }
        if (this._duration !== undefined) {
            hasAnyValues = true;
            internalValueResult.duration = this._duration;
        }
        if (this._daysOfWeek?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.daysOfWeek = this._daysOfWeek?.internalValue;
        }
        if (this._singleOccurrenceDate?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.singleOccurrenceDate = this._singleOccurrenceDate?.internalValue;
        }
        if (this._startTime?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._daily = undefined;
            this._duration = undefined;
            this._daysOfWeek.internalValue = undefined;
            this._singleOccurrenceDate.internalValue = undefined;
            this._startTime.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._daily = value.daily;
            this._duration = value.duration;
            this._daysOfWeek.internalValue = value.daysOfWeek;
            this._singleOccurrenceDate.internalValue = value.singleOccurrenceDate;
            this._startTime.internalValue = value.startTime;
        }
    }
    get daily() {
        return this.getBooleanAttribute('daily');
    }
    set daily(value) {
        this._daily = value;
    }
    resetDaily() {
        this._daily = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dailyInput() {
        return this._daily;
    }
    get duration() {
        return this.getStringAttribute('duration');
    }
    set duration(value) {
        this._duration = value;
    }
    // Temporarily expose input value. Use with caution.
    get durationInput() {
        return this._duration;
    }
    get daysOfWeek() {
        return this._daysOfWeek;
    }
    putDaysOfWeek(value) {
        this._daysOfWeek.internalValue = value;
    }
    resetDaysOfWeek() {
        this._daysOfWeek.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get daysOfWeekInput() {
        return this._daysOfWeek.internalValue;
    }
    get singleOccurrenceDate() {
        return this._singleOccurrenceDate;
    }
    putSingleOccurrenceDate(value) {
        this._singleOccurrenceDate.internalValue = value;
    }
    resetSingleOccurrenceDate() {
        this._singleOccurrenceDate.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get singleOccurrenceDateInput() {
        return this._singleOccurrenceDate.internalValue;
    }
    get startTime() {
        return this._startTime;
    }
    putStartTime(value) {
        this._startTime.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime.internalValue;
    }
}
exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsOutputReference = GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsOutputReference[_k] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsOutputReference", version: "14.13.1" };
class GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsList = GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsList;
_l = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsList[_l] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsList", version: "14.13.1" };
function gkeBackupBackupPlanBackupScheduleRpoConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target_rpo_minutes: cdktf.numberToTerraform(struct.targetRpoMinutes),
        exclusion_windows: cdktf.listMapper(gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToTerraform, true)(struct.exclusionWindows),
    };
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigToTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigToTerraform;
function gkeBackupBackupPlanBackupScheduleRpoConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target_rpo_minutes: {
            value: cdktf.numberToHclTerraform(struct.targetRpoMinutes),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        exclusion_windows: {
            value: cdktf.listMapperHcl(gkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsToHclTerraform, true)(struct.exclusionWindows),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupScheduleRpoConfigToHclTerraform = gkeBackupBackupPlanBackupScheduleRpoConfigToHclTerraform;
class GkeBackupBackupPlanBackupScheduleRpoConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // exclusion_windows - computed: false, optional: true, required: false
        this._exclusionWindows = new GkeBackupBackupPlanBackupScheduleRpoConfigExclusionWindowsList(this, "exclusion_windows", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._targetRpoMinutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetRpoMinutes = this._targetRpoMinutes;
        }
        if (this._exclusionWindows?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.exclusionWindows = this._exclusionWindows?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._targetRpoMinutes = undefined;
            this._exclusionWindows.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._targetRpoMinutes = value.targetRpoMinutes;
            this._exclusionWindows.internalValue = value.exclusionWindows;
        }
    }
    get targetRpoMinutes() {
        return this.getNumberAttribute('target_rpo_minutes');
    }
    set targetRpoMinutes(value) {
        this._targetRpoMinutes = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetRpoMinutesInput() {
        return this._targetRpoMinutes;
    }
    get exclusionWindows() {
        return this._exclusionWindows;
    }
    putExclusionWindows(value) {
        this._exclusionWindows.internalValue = value;
    }
    resetExclusionWindows() {
        this._exclusionWindows.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exclusionWindowsInput() {
        return this._exclusionWindows.internalValue;
    }
}
exports.GkeBackupBackupPlanBackupScheduleRpoConfigOutputReference = GkeBackupBackupPlanBackupScheduleRpoConfigOutputReference;
_m = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupScheduleRpoConfigOutputReference[_m] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupScheduleRpoConfigOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanBackupScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cron_schedule: cdktf.stringToTerraform(struct.cronSchedule),
        paused: cdktf.booleanToTerraform(struct.paused),
        rpo_config: gkeBackupBackupPlanBackupScheduleRpoConfigToTerraform(struct.rpoConfig),
    };
}
exports.gkeBackupBackupPlanBackupScheduleToTerraform = gkeBackupBackupPlanBackupScheduleToTerraform;
function gkeBackupBackupPlanBackupScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cron_schedule: {
            value: cdktf.stringToHclTerraform(struct.cronSchedule),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        paused: {
            value: cdktf.booleanToHclTerraform(struct.paused),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        rpo_config: {
            value: gkeBackupBackupPlanBackupScheduleRpoConfigToHclTerraform(struct.rpoConfig),
            isBlock: true,
            type: "list",
            storageClassType: "GkeBackupBackupPlanBackupScheduleRpoConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanBackupScheduleToHclTerraform = gkeBackupBackupPlanBackupScheduleToHclTerraform;
class GkeBackupBackupPlanBackupScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // rpo_config - computed: false, optional: true, required: false
        this._rpoConfig = new GkeBackupBackupPlanBackupScheduleRpoConfigOutputReference(this, "rpo_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cronSchedule !== undefined) {
            hasAnyValues = true;
            internalValueResult.cronSchedule = this._cronSchedule;
        }
        if (this._paused !== undefined) {
            hasAnyValues = true;
            internalValueResult.paused = this._paused;
        }
        if (this._rpoConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rpoConfig = this._rpoConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cronSchedule = undefined;
            this._paused = undefined;
            this._rpoConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cronSchedule = value.cronSchedule;
            this._paused = value.paused;
            this._rpoConfig.internalValue = value.rpoConfig;
        }
    }
    get cronSchedule() {
        return this.getStringAttribute('cron_schedule');
    }
    set cronSchedule(value) {
        this._cronSchedule = value;
    }
    resetCronSchedule() {
        this._cronSchedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cronScheduleInput() {
        return this._cronSchedule;
    }
    get paused() {
        return this.getBooleanAttribute('paused');
    }
    set paused(value) {
        this._paused = value;
    }
    resetPaused() {
        this._paused = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pausedInput() {
        return this._paused;
    }
    get rpoConfig() {
        return this._rpoConfig;
    }
    putRpoConfig(value) {
        this._rpoConfig.internalValue = value;
    }
    resetRpoConfig() {
        this._rpoConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rpoConfigInput() {
        return this._rpoConfig.internalValue;
    }
}
exports.GkeBackupBackupPlanBackupScheduleOutputReference = GkeBackupBackupPlanBackupScheduleOutputReference;
_o = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanBackupScheduleOutputReference[_o] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanBackupScheduleOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanRetentionPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        backup_delete_lock_days: cdktf.numberToTerraform(struct.backupDeleteLockDays),
        backup_retain_days: cdktf.numberToTerraform(struct.backupRetainDays),
        locked: cdktf.booleanToTerraform(struct.locked),
    };
}
exports.gkeBackupBackupPlanRetentionPolicyToTerraform = gkeBackupBackupPlanRetentionPolicyToTerraform;
function gkeBackupBackupPlanRetentionPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        backup_delete_lock_days: {
            value: cdktf.numberToHclTerraform(struct.backupDeleteLockDays),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        backup_retain_days: {
            value: cdktf.numberToHclTerraform(struct.backupRetainDays),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        locked: {
            value: cdktf.booleanToHclTerraform(struct.locked),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanRetentionPolicyToHclTerraform = gkeBackupBackupPlanRetentionPolicyToHclTerraform;
class GkeBackupBackupPlanRetentionPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._backupDeleteLockDays !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupDeleteLockDays = this._backupDeleteLockDays;
        }
        if (this._backupRetainDays !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupRetainDays = this._backupRetainDays;
        }
        if (this._locked !== undefined) {
            hasAnyValues = true;
            internalValueResult.locked = this._locked;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._backupDeleteLockDays = undefined;
            this._backupRetainDays = undefined;
            this._locked = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._backupDeleteLockDays = value.backupDeleteLockDays;
            this._backupRetainDays = value.backupRetainDays;
            this._locked = value.locked;
        }
    }
    get backupDeleteLockDays() {
        return this.getNumberAttribute('backup_delete_lock_days');
    }
    set backupDeleteLockDays(value) {
        this._backupDeleteLockDays = value;
    }
    resetBackupDeleteLockDays() {
        this._backupDeleteLockDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupDeleteLockDaysInput() {
        return this._backupDeleteLockDays;
    }
    get backupRetainDays() {
        return this.getNumberAttribute('backup_retain_days');
    }
    set backupRetainDays(value) {
        this._backupRetainDays = value;
    }
    resetBackupRetainDays() {
        this._backupRetainDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupRetainDaysInput() {
        return this._backupRetainDays;
    }
    get locked() {
        return this.getBooleanAttribute('locked');
    }
    set locked(value) {
        this._locked = value;
    }
    resetLocked() {
        this._locked = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lockedInput() {
        return this._locked;
    }
}
exports.GkeBackupBackupPlanRetentionPolicyOutputReference = GkeBackupBackupPlanRetentionPolicyOutputReference;
_p = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanRetentionPolicyOutputReference[_p] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanRetentionPolicyOutputReference", version: "14.13.1" };
function gkeBackupBackupPlanTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.gkeBackupBackupPlanTimeoutsToTerraform = gkeBackupBackupPlanTimeoutsToTerraform;
function gkeBackupBackupPlanTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.gkeBackupBackupPlanTimeoutsToHclTerraform = gkeBackupBackupPlanTimeoutsToHclTerraform;
class GkeBackupBackupPlanTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.GkeBackupBackupPlanTimeoutsOutputReference = GkeBackupBackupPlanTimeoutsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlanTimeoutsOutputReference[_q] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlanTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/gke_backup_backup_plan google_gke_backup_backup_plan}
*/
class GkeBackupBackupPlan extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a GkeBackupBackupPlan resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the GkeBackupBackupPlan to import
    * @param importFromId The id of the existing GkeBackupBackupPlan that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/gke_backup_backup_plan#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the GkeBackupBackupPlan to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_gke_backup_backup_plan", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/gke_backup_backup_plan google_gke_backup_backup_plan} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options GkeBackupBackupPlanConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_gke_backup_backup_plan',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // backup_config - computed: false, optional: true, required: false
        this._backupConfig = new GkeBackupBackupPlanBackupConfigOutputReference(this, "backup_config");
        // backup_schedule - computed: false, optional: true, required: false
        this._backupSchedule = new GkeBackupBackupPlanBackupScheduleOutputReference(this, "backup_schedule");
        // retention_policy - computed: false, optional: true, required: false
        this._retentionPolicy = new GkeBackupBackupPlanRetentionPolicyOutputReference(this, "retention_policy");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new GkeBackupBackupPlanTimeoutsOutputReference(this, "timeouts");
        this._cluster = config.cluster;
        this._deactivated = config.deactivated;
        this._description = config.description;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._backupConfig.internalValue = config.backupConfig;
        this._backupSchedule.internalValue = config.backupSchedule;
        this._retentionPolicy.internalValue = config.retentionPolicy;
        this._timeouts.internalValue = config.timeouts;
    }
    get cluster() {
        return this.getStringAttribute('cluster');
    }
    set cluster(value) {
        this._cluster = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterInput() {
        return this._cluster;
    }
    get deactivated() {
        return this.getBooleanAttribute('deactivated');
    }
    set deactivated(value) {
        this._deactivated = value;
    }
    resetDeactivated() {
        this._deactivated = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deactivatedInput() {
        return this._deactivated;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // protected_pod_count - computed: true, optional: false, required: false
    get protectedPodCount() {
        return this.getNumberAttribute('protected_pod_count');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // state_reason - computed: true, optional: false, required: false
    get stateReason() {
        return this.getStringAttribute('state_reason');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    get backupConfig() {
        return this._backupConfig;
    }
    putBackupConfig(value) {
        this._backupConfig.internalValue = value;
    }
    resetBackupConfig() {
        this._backupConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupConfigInput() {
        return this._backupConfig.internalValue;
    }
    get backupSchedule() {
        return this._backupSchedule;
    }
    putBackupSchedule(value) {
        this._backupSchedule.internalValue = value;
    }
    resetBackupSchedule() {
        this._backupSchedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupScheduleInput() {
        return this._backupSchedule.internalValue;
    }
    get retentionPolicy() {
        return this._retentionPolicy;
    }
    putRetentionPolicy(value) {
        this._retentionPolicy.internalValue = value;
    }
    resetRetentionPolicy() {
        this._retentionPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get retentionPolicyInput() {
        return this._retentionPolicy.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster: cdktf.stringToTerraform(this._cluster),
            deactivated: cdktf.booleanToTerraform(this._deactivated),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            backup_config: gkeBackupBackupPlanBackupConfigToTerraform(this._backupConfig.internalValue),
            backup_schedule: gkeBackupBackupPlanBackupScheduleToTerraform(this._backupSchedule.internalValue),
            retention_policy: gkeBackupBackupPlanRetentionPolicyToTerraform(this._retentionPolicy.internalValue),
            timeouts: gkeBackupBackupPlanTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cluster: {
                value: cdktf.stringToHclTerraform(this._cluster),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            deactivated: {
                value: cdktf.booleanToHclTerraform(this._deactivated),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            backup_config: {
                value: gkeBackupBackupPlanBackupConfigToHclTerraform(this._backupConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GkeBackupBackupPlanBackupConfigList",
            },
            backup_schedule: {
                value: gkeBackupBackupPlanBackupScheduleToHclTerraform(this._backupSchedule.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GkeBackupBackupPlanBackupScheduleList",
            },
            retention_policy: {
                value: gkeBackupBackupPlanRetentionPolicyToHclTerraform(this._retentionPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "GkeBackupBackupPlanRetentionPolicyList",
            },
            timeouts: {
                value: gkeBackupBackupPlanTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "GkeBackupBackupPlanTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.GkeBackupBackupPlan = GkeBackupBackupPlan;
_r = JSII_RTTI_SYMBOL_1;
GkeBackupBackupPlan[_r] = { fqn: "@cdktf/provider-google.gkeBackupBackupPlan.GkeBackupBackupPlan", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
GkeBackupBackupPlan.tfResourceType = "google_gke_backup_backup_plan";
