// @cdktf/provider-google/lib/data-google-dns-keys/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleDnsKeys = exports.DataGoogleDnsKeysZoneSigningKeysList = exports.DataGoogleDnsKeysZoneSigningKeysOutputReference = exports.dataGoogleDnsKeysZoneSigningKeysToHclTerraform = exports.dataGoogleDnsKeysZoneSigningKeysToTerraform = exports.DataGoogleDnsKeysZoneSigningKeysDigestsList = exports.DataGoogleDnsKeysZoneSigningKeysDigestsOutputReference = exports.dataGoogleDnsKeysZoneSigningKeysDigestsToHclTerraform = exports.dataGoogleDnsKeysZoneSigningKeysDigestsToTerraform = exports.DataGoogleDnsKeysKeySigningKeysList = exports.DataGoogleDnsKeysKeySigningKeysOutputReference = exports.dataGoogleDnsKeysKeySigningKeysToHclTerraform = exports.dataGoogleDnsKeysKeySigningKeysToTerraform = exports.DataGoogleDnsKeysKeySigningKeysDigestsList = exports.DataGoogleDnsKeysKeySigningKeysDigestsOutputReference = exports.dataGoogleDnsKeysKeySigningKeysDigestsToHclTerraform = exports.dataGoogleDnsKeysKeySigningKeysDigestsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleDnsKeysKeySigningKeysDigestsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDnsKeysKeySigningKeysDigestsToTerraform = dataGoogleDnsKeysKeySigningKeysDigestsToTerraform;
function dataGoogleDnsKeysKeySigningKeysDigestsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDnsKeysKeySigningKeysDigestsToHclTerraform = dataGoogleDnsKeysKeySigningKeysDigestsToHclTerraform;
class DataGoogleDnsKeysKeySigningKeysDigestsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // digest - computed: true, optional: false, required: false
    get digest() {
        return this.getStringAttribute('digest');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleDnsKeysKeySigningKeysDigestsOutputReference = DataGoogleDnsKeysKeySigningKeysDigestsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeysKeySigningKeysDigestsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeysKeySigningKeysDigestsOutputReference", version: "14.13.1" };
class DataGoogleDnsKeysKeySigningKeysDigestsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDnsKeysKeySigningKeysDigestsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDnsKeysKeySigningKeysDigestsList = DataGoogleDnsKeysKeySigningKeysDigestsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeysKeySigningKeysDigestsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeysKeySigningKeysDigestsList", version: "14.13.1" };
function dataGoogleDnsKeysKeySigningKeysToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDnsKeysKeySigningKeysToTerraform = dataGoogleDnsKeysKeySigningKeysToTerraform;
function dataGoogleDnsKeysKeySigningKeysToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDnsKeysKeySigningKeysToHclTerraform = dataGoogleDnsKeysKeySigningKeysToHclTerraform;
class DataGoogleDnsKeysKeySigningKeysOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // digests - computed: true, optional: false, required: false
        this._digests = new DataGoogleDnsKeysKeySigningKeysDigestsList(this, "digests", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // algorithm - computed: true, optional: false, required: false
    get algorithm() {
        return this.getStringAttribute('algorithm');
    }
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get digests() {
        return this._digests;
    }
    // ds_record - computed: true, optional: false, required: false
    get dsRecord() {
        return this.getStringAttribute('ds_record');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // is_active - computed: true, optional: false, required: false
    get isActive() {
        return this.getBooleanAttribute('is_active');
    }
    // key_length - computed: true, optional: false, required: false
    get keyLength() {
        return this.getNumberAttribute('key_length');
    }
    // key_tag - computed: true, optional: false, required: false
    get keyTag() {
        return this.getNumberAttribute('key_tag');
    }
    // public_key - computed: true, optional: false, required: false
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
}
exports.DataGoogleDnsKeysKeySigningKeysOutputReference = DataGoogleDnsKeysKeySigningKeysOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeysKeySigningKeysOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeysKeySigningKeysOutputReference", version: "14.13.1" };
class DataGoogleDnsKeysKeySigningKeysList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDnsKeysKeySigningKeysOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDnsKeysKeySigningKeysList = DataGoogleDnsKeysKeySigningKeysList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeysKeySigningKeysList[_d] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeysKeySigningKeysList", version: "14.13.1" };
function dataGoogleDnsKeysZoneSigningKeysDigestsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDnsKeysZoneSigningKeysDigestsToTerraform = dataGoogleDnsKeysZoneSigningKeysDigestsToTerraform;
function dataGoogleDnsKeysZoneSigningKeysDigestsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDnsKeysZoneSigningKeysDigestsToHclTerraform = dataGoogleDnsKeysZoneSigningKeysDigestsToHclTerraform;
class DataGoogleDnsKeysZoneSigningKeysDigestsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // digest - computed: true, optional: false, required: false
    get digest() {
        return this.getStringAttribute('digest');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleDnsKeysZoneSigningKeysDigestsOutputReference = DataGoogleDnsKeysZoneSigningKeysDigestsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeysZoneSigningKeysDigestsOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeysZoneSigningKeysDigestsOutputReference", version: "14.13.1" };
class DataGoogleDnsKeysZoneSigningKeysDigestsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDnsKeysZoneSigningKeysDigestsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDnsKeysZoneSigningKeysDigestsList = DataGoogleDnsKeysZoneSigningKeysDigestsList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeysZoneSigningKeysDigestsList[_f] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeysZoneSigningKeysDigestsList", version: "14.13.1" };
function dataGoogleDnsKeysZoneSigningKeysToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleDnsKeysZoneSigningKeysToTerraform = dataGoogleDnsKeysZoneSigningKeysToTerraform;
function dataGoogleDnsKeysZoneSigningKeysToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleDnsKeysZoneSigningKeysToHclTerraform = dataGoogleDnsKeysZoneSigningKeysToHclTerraform;
class DataGoogleDnsKeysZoneSigningKeysOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // digests - computed: true, optional: false, required: false
        this._digests = new DataGoogleDnsKeysZoneSigningKeysDigestsList(this, "digests", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // algorithm - computed: true, optional: false, required: false
    get algorithm() {
        return this.getStringAttribute('algorithm');
    }
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get digests() {
        return this._digests;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // is_active - computed: true, optional: false, required: false
    get isActive() {
        return this.getBooleanAttribute('is_active');
    }
    // key_length - computed: true, optional: false, required: false
    get keyLength() {
        return this.getNumberAttribute('key_length');
    }
    // key_tag - computed: true, optional: false, required: false
    get keyTag() {
        return this.getNumberAttribute('key_tag');
    }
    // public_key - computed: true, optional: false, required: false
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
}
exports.DataGoogleDnsKeysZoneSigningKeysOutputReference = DataGoogleDnsKeysZoneSigningKeysOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeysZoneSigningKeysOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeysZoneSigningKeysOutputReference", version: "14.13.1" };
class DataGoogleDnsKeysZoneSigningKeysList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleDnsKeysZoneSigningKeysOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleDnsKeysZoneSigningKeysList = DataGoogleDnsKeysZoneSigningKeysList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeysZoneSigningKeysList[_h] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeysZoneSigningKeysList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/dns_keys google_dns_keys}
*/
class DataGoogleDnsKeys extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleDnsKeys resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleDnsKeys to import
    * @param importFromId The id of the existing DataGoogleDnsKeys that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/dns_keys#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleDnsKeys to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_dns_keys", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/dns_keys google_dns_keys} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleDnsKeysConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_dns_keys',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // key_signing_keys - computed: true, optional: false, required: false
        this._keySigningKeys = new DataGoogleDnsKeysKeySigningKeysList(this, "key_signing_keys", false);
        // zone_signing_keys - computed: true, optional: false, required: false
        this._zoneSigningKeys = new DataGoogleDnsKeysZoneSigningKeysList(this, "zone_signing_keys", false);
        this._id = config.id;
        this._managedZone = config.managedZone;
        this._project = config.project;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keySigningKeys() {
        return this._keySigningKeys;
    }
    get managedZone() {
        return this.getStringAttribute('managed_zone');
    }
    set managedZone(value) {
        this._managedZone = value;
    }
    // Temporarily expose input value. Use with caution.
    get managedZoneInput() {
        return this._managedZone;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get zoneSigningKeys() {
        return this._zoneSigningKeys;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            managed_zone: cdktf.stringToTerraform(this._managedZone),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            managed_zone: {
                value: cdktf.stringToHclTerraform(this._managedZone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleDnsKeys = DataGoogleDnsKeys;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleDnsKeys[_j] = { fqn: "@cdktf/provider-google.dataGoogleDnsKeys.DataGoogleDnsKeys", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleDnsKeys.tfResourceType = "google_dns_keys";
