// @cdktf/provider-google/lib/iam-workforce-pool/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamWorkforcePool = exports.IamWorkforcePoolTimeoutsOutputReference = exports.iamWorkforcePoolTimeoutsToHclTerraform = exports.iamWorkforcePoolTimeoutsToTerraform = exports.IamWorkforcePoolAccessRestrictionsOutputReference = exports.iamWorkforcePoolAccessRestrictionsToHclTerraform = exports.iamWorkforcePoolAccessRestrictionsToTerraform = exports.IamWorkforcePoolAccessRestrictionsAllowedServicesList = exports.IamWorkforcePoolAccessRestrictionsAllowedServicesOutputReference = exports.iamWorkforcePoolAccessRestrictionsAllowedServicesToHclTerraform = exports.iamWorkforcePoolAccessRestrictionsAllowedServicesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iamWorkforcePoolAccessRestrictionsAllowedServicesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        domain: cdktf.stringToTerraform(struct.domain),
    };
}
exports.iamWorkforcePoolAccessRestrictionsAllowedServicesToTerraform = iamWorkforcePoolAccessRestrictionsAllowedServicesToTerraform;
function iamWorkforcePoolAccessRestrictionsAllowedServicesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        domain: {
            value: cdktf.stringToHclTerraform(struct.domain),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkforcePoolAccessRestrictionsAllowedServicesToHclTerraform = iamWorkforcePoolAccessRestrictionsAllowedServicesToHclTerraform;
class IamWorkforcePoolAccessRestrictionsAllowedServicesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._domain !== undefined) {
            hasAnyValues = true;
            internalValueResult.domain = this._domain;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._domain = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._domain = value.domain;
        }
    }
    get domain() {
        return this.getStringAttribute('domain');
    }
    set domain(value) {
        this._domain = value;
    }
    resetDomain() {
        this._domain = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get domainInput() {
        return this._domain;
    }
}
exports.IamWorkforcePoolAccessRestrictionsAllowedServicesOutputReference = IamWorkforcePoolAccessRestrictionsAllowedServicesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IamWorkforcePoolAccessRestrictionsAllowedServicesOutputReference[_a] = { fqn: "@cdktf/provider-google.iamWorkforcePool.IamWorkforcePoolAccessRestrictionsAllowedServicesOutputReference", version: "14.13.1" };
class IamWorkforcePoolAccessRestrictionsAllowedServicesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IamWorkforcePoolAccessRestrictionsAllowedServicesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IamWorkforcePoolAccessRestrictionsAllowedServicesList = IamWorkforcePoolAccessRestrictionsAllowedServicesList;
_b = JSII_RTTI_SYMBOL_1;
IamWorkforcePoolAccessRestrictionsAllowedServicesList[_b] = { fqn: "@cdktf/provider-google.iamWorkforcePool.IamWorkforcePoolAccessRestrictionsAllowedServicesList", version: "14.13.1" };
function iamWorkforcePoolAccessRestrictionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disable_programmatic_signin: cdktf.booleanToTerraform(struct.disableProgrammaticSignin),
        allowed_services: cdktf.listMapper(iamWorkforcePoolAccessRestrictionsAllowedServicesToTerraform, true)(struct.allowedServices),
    };
}
exports.iamWorkforcePoolAccessRestrictionsToTerraform = iamWorkforcePoolAccessRestrictionsToTerraform;
function iamWorkforcePoolAccessRestrictionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        disable_programmatic_signin: {
            value: cdktf.booleanToHclTerraform(struct.disableProgrammaticSignin),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        allowed_services: {
            value: cdktf.listMapperHcl(iamWorkforcePoolAccessRestrictionsAllowedServicesToHclTerraform, true)(struct.allowedServices),
            isBlock: true,
            type: "list",
            storageClassType: "IamWorkforcePoolAccessRestrictionsAllowedServicesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkforcePoolAccessRestrictionsToHclTerraform = iamWorkforcePoolAccessRestrictionsToHclTerraform;
class IamWorkforcePoolAccessRestrictionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // allowed_services - computed: false, optional: true, required: false
        this._allowedServices = new IamWorkforcePoolAccessRestrictionsAllowedServicesList(this, "allowed_services", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._disableProgrammaticSignin !== undefined) {
            hasAnyValues = true;
            internalValueResult.disableProgrammaticSignin = this._disableProgrammaticSignin;
        }
        if (this._allowedServices?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.allowedServices = this._allowedServices?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._disableProgrammaticSignin = undefined;
            this._allowedServices.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._disableProgrammaticSignin = value.disableProgrammaticSignin;
            this._allowedServices.internalValue = value.allowedServices;
        }
    }
    get disableProgrammaticSignin() {
        return this.getBooleanAttribute('disable_programmatic_signin');
    }
    set disableProgrammaticSignin(value) {
        this._disableProgrammaticSignin = value;
    }
    resetDisableProgrammaticSignin() {
        this._disableProgrammaticSignin = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableProgrammaticSigninInput() {
        return this._disableProgrammaticSignin;
    }
    get allowedServices() {
        return this._allowedServices;
    }
    putAllowedServices(value) {
        this._allowedServices.internalValue = value;
    }
    resetAllowedServices() {
        this._allowedServices.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowedServicesInput() {
        return this._allowedServices.internalValue;
    }
}
exports.IamWorkforcePoolAccessRestrictionsOutputReference = IamWorkforcePoolAccessRestrictionsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IamWorkforcePoolAccessRestrictionsOutputReference[_c] = { fqn: "@cdktf/provider-google.iamWorkforcePool.IamWorkforcePoolAccessRestrictionsOutputReference", version: "14.13.1" };
function iamWorkforcePoolTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.iamWorkforcePoolTimeoutsToTerraform = iamWorkforcePoolTimeoutsToTerraform;
function iamWorkforcePoolTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamWorkforcePoolTimeoutsToHclTerraform = iamWorkforcePoolTimeoutsToHclTerraform;
class IamWorkforcePoolTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.IamWorkforcePoolTimeoutsOutputReference = IamWorkforcePoolTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
IamWorkforcePoolTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-google.iamWorkforcePool.IamWorkforcePoolTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_workforce_pool google_iam_workforce_pool}
*/
class IamWorkforcePool extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IamWorkforcePool resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IamWorkforcePool to import
    * @param importFromId The id of the existing IamWorkforcePool that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_workforce_pool#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IamWorkforcePool to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_iam_workforce_pool", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_workforce_pool google_iam_workforce_pool} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamWorkforcePoolConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_iam_workforce_pool',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // access_restrictions - computed: false, optional: true, required: false
        this._accessRestrictions = new IamWorkforcePoolAccessRestrictionsOutputReference(this, "access_restrictions");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new IamWorkforcePoolTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._disabled = config.disabled;
        this._displayName = config.displayName;
        this._id = config.id;
        this._location = config.location;
        this._parent = config.parent;
        this._sessionDuration = config.sessionDuration;
        this._workforcePoolId = config.workforcePoolId;
        this._accessRestrictions.internalValue = config.accessRestrictions;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    set disabled(value) {
        this._disabled = value;
    }
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disabledInput() {
        return this._disabled;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    get sessionDuration() {
        return this.getStringAttribute('session_duration');
    }
    set sessionDuration(value) {
        this._sessionDuration = value;
    }
    resetSessionDuration() {
        this._sessionDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionDurationInput() {
        return this._sessionDuration;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get workforcePoolId() {
        return this.getStringAttribute('workforce_pool_id');
    }
    set workforcePoolId(value) {
        this._workforcePoolId = value;
    }
    // Temporarily expose input value. Use with caution.
    get workforcePoolIdInput() {
        return this._workforcePoolId;
    }
    get accessRestrictions() {
        return this._accessRestrictions;
    }
    putAccessRestrictions(value) {
        this._accessRestrictions.internalValue = value;
    }
    resetAccessRestrictions() {
        this._accessRestrictions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessRestrictionsInput() {
        return this._accessRestrictions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            disabled: cdktf.booleanToTerraform(this._disabled),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            parent: cdktf.stringToTerraform(this._parent),
            session_duration: cdktf.stringToTerraform(this._sessionDuration),
            workforce_pool_id: cdktf.stringToTerraform(this._workforcePoolId),
            access_restrictions: iamWorkforcePoolAccessRestrictionsToTerraform(this._accessRestrictions.internalValue),
            timeouts: iamWorkforcePoolTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            disabled: {
                value: cdktf.booleanToHclTerraform(this._disabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent: {
                value: cdktf.stringToHclTerraform(this._parent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            session_duration: {
                value: cdktf.stringToHclTerraform(this._sessionDuration),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            workforce_pool_id: {
                value: cdktf.stringToHclTerraform(this._workforcePoolId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            access_restrictions: {
                value: iamWorkforcePoolAccessRestrictionsToHclTerraform(this._accessRestrictions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IamWorkforcePoolAccessRestrictionsList",
            },
            timeouts: {
                value: iamWorkforcePoolTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "IamWorkforcePoolTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IamWorkforcePool = IamWorkforcePool;
_e = JSII_RTTI_SYMBOL_1;
IamWorkforcePool[_e] = { fqn: "@cdktf/provider-google.iamWorkforcePool.IamWorkforcePool", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
IamWorkforcePool.tfResourceType = "google_iam_workforce_pool";
