// @cdktf/provider-google/lib/data-google-compute-region-instance-template/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationToTerraform = exports.DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigList = exports.DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigOutputReference = exports.dataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigToTerraform = exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceList = exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceOutputReference = exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceToTerraform = exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigList = exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference = exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform = exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeList = exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference = exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform = exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigList = exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigOutputReference = exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigToTerraform = exports.DataGoogleComputeRegionInstanceTemplateGuestAcceleratorList = exports.DataGoogleComputeRegionInstanceTemplateGuestAcceleratorOutputReference = exports.dataGoogleComputeRegionInstanceTemplateGuestAcceleratorToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateGuestAcceleratorToTerraform = exports.DataGoogleComputeRegionInstanceTemplateDiskList = exports.DataGoogleComputeRegionInstanceTemplateDiskOutputReference = exports.dataGoogleComputeRegionInstanceTemplateDiskToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateDiskToTerraform = exports.DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyList = exports.DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference = exports.dataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyToTerraform = exports.DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyList = exports.DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyOutputReference = exports.dataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyToTerraform = exports.DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyList = exports.DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyOutputReference = exports.dataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyToTerraform = exports.DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigList = exports.DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigOutputReference = exports.dataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigToTerraform = exports.DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesList = exports.DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesOutputReference = exports.dataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesToTerraform = void 0;
exports.DataGoogleComputeRegionInstanceTemplate = exports.DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigList = exports.DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigOutputReference = exports.dataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigToTerraform = exports.DataGoogleComputeRegionInstanceTemplateServiceAccountList = exports.DataGoogleComputeRegionInstanceTemplateServiceAccountOutputReference = exports.dataGoogleComputeRegionInstanceTemplateServiceAccountToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateServiceAccountToTerraform = exports.DataGoogleComputeRegionInstanceTemplateSchedulingList = exports.DataGoogleComputeRegionInstanceTemplateSchedulingOutputReference = exports.dataGoogleComputeRegionInstanceTemplateSchedulingToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateSchedulingToTerraform = exports.DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionList = exports.DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionOutputReference = exports.dataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionToTerraform = exports.DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesList = exports.DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesOutputReference = exports.dataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesToTerraform = exports.DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationList = exports.DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationOutputReference = exports.dataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationToTerraform = exports.DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList = exports.DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference = exports.dataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToTerraform = exports.DataGoogleComputeRegionInstanceTemplateReservationAffinityList = exports.DataGoogleComputeRegionInstanceTemplateReservationAffinityOutputReference = exports.dataGoogleComputeRegionInstanceTemplateReservationAffinityToHclTerraform = exports.dataGoogleComputeRegionInstanceTemplateReservationAffinityToTerraform = exports.DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationList = exports.DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesToTerraform = dataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesToTerraform;
function dataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesToHclTerraform = dataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_nested_virtualization - computed: true, optional: false, required: false
    get enableNestedVirtualization() {
        return this.getBooleanAttribute('enable_nested_virtualization');
    }
    // enable_uefi_networking - computed: true, optional: false, required: false
    get enableUefiNetworking() {
        return this.getBooleanAttribute('enable_uefi_networking');
    }
    // performance_monitoring_unit - computed: true, optional: false, required: false
    get performanceMonitoringUnit() {
        return this.getStringAttribute('performance_monitoring_unit');
    }
    // threads_per_core - computed: true, optional: false, required: false
    get threadsPerCore() {
        return this.getNumberAttribute('threads_per_core');
    }
    // turbo_mode - computed: true, optional: false, required: false
    get turboMode() {
        return this.getStringAttribute('turbo_mode');
    }
    // visible_core_count - computed: true, optional: false, required: false
    get visibleCoreCount() {
        return this.getNumberAttribute('visible_core_count');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesOutputReference = DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesList = DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigToTerraform = dataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigToTerraform;
function dataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigToHclTerraform = dataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // confidential_instance_type - computed: true, optional: false, required: false
    get confidentialInstanceType() {
        return this.getStringAttribute('confidential_instance_type');
    }
    // enable_confidential_compute - computed: true, optional: false, required: false
    get enableConfidentialCompute() {
        return this.getBooleanAttribute('enable_confidential_compute');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigOutputReference = DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigList = DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyToTerraform = dataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyToTerraform;
function dataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyToHclTerraform = dataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyOutputReference = DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyList = DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyToTerraform = dataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyToTerraform;
function dataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyToHclTerraform = dataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // kms_key_service_account - computed: true, optional: false, required: false
    get kmsKeyServiceAccount() {
        return this.getStringAttribute('kms_key_service_account');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyOutputReference = DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyList = DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyList[_h] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyToTerraform = dataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyToTerraform;
function dataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyToHclTerraform = dataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // kms_key_self_link - computed: true, optional: false, required: false
    get kmsKeySelfLink() {
        return this.getStringAttribute('kms_key_self_link');
    }
    // kms_key_service_account - computed: true, optional: false, required: false
    get kmsKeyServiceAccount() {
        return this.getStringAttribute('kms_key_service_account');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference = DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyList = DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyList[_k] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateDiskToTerraform = dataGoogleComputeRegionInstanceTemplateDiskToTerraform;
function dataGoogleComputeRegionInstanceTemplateDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateDiskToHclTerraform = dataGoogleComputeRegionInstanceTemplateDiskToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // disk_encryption_key - computed: true, optional: false, required: false
        this._diskEncryptionKey = new DataGoogleComputeRegionInstanceTemplateDiskDiskEncryptionKeyList(this, "disk_encryption_key", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // resource_manager_tags - computed: true, optional: false, required: false
        this._resourceManagerTags = new cdktf.StringMap(this, "resource_manager_tags");
        // source_image_encryption_key - computed: true, optional: false, required: false
        this._sourceImageEncryptionKey = new DataGoogleComputeRegionInstanceTemplateDiskSourceImageEncryptionKeyList(this, "source_image_encryption_key", false);
        // source_snapshot_encryption_key - computed: true, optional: false, required: false
        this._sourceSnapshotEncryptionKey = new DataGoogleComputeRegionInstanceTemplateDiskSourceSnapshotEncryptionKeyList(this, "source_snapshot_encryption_key", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // auto_delete - computed: true, optional: false, required: false
    get autoDelete() {
        return this.getBooleanAttribute('auto_delete');
    }
    // boot - computed: true, optional: false, required: false
    get boot() {
        return this.getBooleanAttribute('boot');
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    get diskEncryptionKey() {
        return this._diskEncryptionKey;
    }
    // disk_name - computed: true, optional: false, required: false
    get diskName() {
        return this.getStringAttribute('disk_name');
    }
    // disk_size_gb - computed: true, optional: false, required: false
    get diskSizeGb() {
        return this.getNumberAttribute('disk_size_gb');
    }
    // disk_type - computed: true, optional: false, required: false
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    // interface - computed: true, optional: false, required: false
    get interface() {
        return this.getStringAttribute('interface');
    }
    get labels() {
        return this._labels;
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
    // provisioned_iops - computed: true, optional: false, required: false
    get provisionedIops() {
        return this.getNumberAttribute('provisioned_iops');
    }
    // provisioned_throughput - computed: true, optional: false, required: false
    get provisionedThroughput() {
        return this.getNumberAttribute('provisioned_throughput');
    }
    get resourceManagerTags() {
        return this._resourceManagerTags;
    }
    // resource_policies - computed: true, optional: false, required: false
    get resourcePolicies() {
        return this.getListAttribute('resource_policies');
    }
    // source - computed: true, optional: false, required: false
    get source() {
        return this.getStringAttribute('source');
    }
    // source_image - computed: true, optional: false, required: false
    get sourceImage() {
        return this.getStringAttribute('source_image');
    }
    get sourceImageEncryptionKey() {
        return this._sourceImageEncryptionKey;
    }
    // source_snapshot - computed: true, optional: false, required: false
    get sourceSnapshot() {
        return this.getStringAttribute('source_snapshot');
    }
    get sourceSnapshotEncryptionKey() {
        return this._sourceSnapshotEncryptionKey;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateDiskOutputReference = DataGoogleComputeRegionInstanceTemplateDiskOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateDiskOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateDiskOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateDiskList = DataGoogleComputeRegionInstanceTemplateDiskList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateDiskList[_m] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateDiskList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateGuestAcceleratorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateGuestAcceleratorToTerraform = dataGoogleComputeRegionInstanceTemplateGuestAcceleratorToTerraform;
function dataGoogleComputeRegionInstanceTemplateGuestAcceleratorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateGuestAcceleratorToHclTerraform = dataGoogleComputeRegionInstanceTemplateGuestAcceleratorToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateGuestAcceleratorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // count - computed: true, optional: false, required: false
    get count() {
        return this.getNumberAttribute('count');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateGuestAcceleratorOutputReference = DataGoogleComputeRegionInstanceTemplateGuestAcceleratorOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateGuestAcceleratorOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateGuestAcceleratorOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateGuestAcceleratorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateGuestAcceleratorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateGuestAcceleratorList = DataGoogleComputeRegionInstanceTemplateGuestAcceleratorList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateGuestAcceleratorList[_p] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateGuestAcceleratorList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigToTerraform = dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigToTerraform;
function dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigToHclTerraform = dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // nat_ip - computed: true, optional: false, required: false
    get natIp() {
        return this.getStringAttribute('nat_ip');
    }
    // network_tier - computed: true, optional: false, required: false
    get networkTier() {
        return this.getStringAttribute('network_tier');
    }
    // public_ptr_domain_name - computed: true, optional: false, required: false
    get publicPtrDomainName() {
        return this.getStringAttribute('public_ptr_domain_name');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigOutputReference = DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigList = DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigList[_r] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform = dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeToTerraform;
function dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeToHclTerraform = dataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_cidr_range - computed: true, optional: false, required: false
    get ipCidrRange() {
        return this.getStringAttribute('ip_cidr_range');
    }
    // subnetwork_range_name - computed: true, optional: false, required: false
    get subnetworkRangeName() {
        return this.getStringAttribute('subnetwork_range_name');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference = DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeList = DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeList[_t] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform = dataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigToTerraform;
function dataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigToHclTerraform = dataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // external_ipv6 - computed: true, optional: false, required: false
    get externalIpv6() {
        return this.getStringAttribute('external_ipv6');
    }
    // external_ipv6_prefix_length - computed: true, optional: false, required: false
    get externalIpv6PrefixLength() {
        return this.getStringAttribute('external_ipv6_prefix_length');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network_tier - computed: true, optional: false, required: false
    get networkTier() {
        return this.getStringAttribute('network_tier');
    }
    // public_ptr_domain_name - computed: true, optional: false, required: false
    get publicPtrDomainName() {
        return this.getStringAttribute('public_ptr_domain_name');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference = DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigList = DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigList[_v] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateNetworkInterfaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceToTerraform = dataGoogleComputeRegionInstanceTemplateNetworkInterfaceToTerraform;
function dataGoogleComputeRegionInstanceTemplateNetworkInterfaceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkInterfaceToHclTerraform = dataGoogleComputeRegionInstanceTemplateNetworkInterfaceToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateNetworkInterfaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // access_config - computed: true, optional: false, required: false
        this._accessConfig = new DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAccessConfigList(this, "access_config", false);
        // alias_ip_range - computed: true, optional: false, required: false
        this._aliasIpRange = new DataGoogleComputeRegionInstanceTemplateNetworkInterfaceAliasIpRangeList(this, "alias_ip_range", false);
        // ipv6_access_config - computed: true, optional: false, required: false
        this._ipv6AccessConfig = new DataGoogleComputeRegionInstanceTemplateNetworkInterfaceIpv6AccessConfigList(this, "ipv6_access_config", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get accessConfig() {
        return this._accessConfig;
    }
    get aliasIpRange() {
        return this._aliasIpRange;
    }
    // internal_ipv6_prefix_length - computed: true, optional: false, required: false
    get internalIpv6PrefixLength() {
        return this.getNumberAttribute('internal_ipv6_prefix_length');
    }
    get ipv6AccessConfig() {
        return this._ipv6AccessConfig;
    }
    // ipv6_access_type - computed: true, optional: false, required: false
    get ipv6AccessType() {
        return this.getStringAttribute('ipv6_access_type');
    }
    // ipv6_address - computed: true, optional: false, required: false
    get ipv6Address() {
        return this.getStringAttribute('ipv6_address');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    // network_ip - computed: true, optional: false, required: false
    get networkIp() {
        return this.getStringAttribute('network_ip');
    }
    // nic_type - computed: true, optional: false, required: false
    get nicType() {
        return this.getStringAttribute('nic_type');
    }
    // queue_count - computed: true, optional: false, required: false
    get queueCount() {
        return this.getNumberAttribute('queue_count');
    }
    // stack_type - computed: true, optional: false, required: false
    get stackType() {
        return this.getStringAttribute('stack_type');
    }
    // subnetwork - computed: true, optional: false, required: false
    get subnetwork() {
        return this.getStringAttribute('subnetwork');
    }
    // subnetwork_project - computed: true, optional: false, required: false
    get subnetworkProject() {
        return this.getStringAttribute('subnetwork_project');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceOutputReference = DataGoogleComputeRegionInstanceTemplateNetworkInterfaceOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkInterfaceOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateNetworkInterfaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateNetworkInterfaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceList = DataGoogleComputeRegionInstanceTemplateNetworkInterfaceList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkInterfaceList[_x] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkInterfaceList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigToTerraform = dataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigToTerraform;
function dataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigToHclTerraform = dataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // total_egress_bandwidth_tier - computed: true, optional: false, required: false
    get totalEgressBandwidthTier() {
        return this.getStringAttribute('total_egress_bandwidth_tier');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigOutputReference = DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigList = DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigList[_z] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationToTerraform = dataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationToTerraform;
function dataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationToHclTerraform = dataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return this.getListAttribute('values');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationOutputReference = DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationList = DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationList[_1] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateReservationAffinityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateReservationAffinityToTerraform = dataGoogleComputeRegionInstanceTemplateReservationAffinityToTerraform;
function dataGoogleComputeRegionInstanceTemplateReservationAffinityToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateReservationAffinityToHclTerraform = dataGoogleComputeRegionInstanceTemplateReservationAffinityToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateReservationAffinityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // specific_reservation - computed: true, optional: false, required: false
        this._specificReservation = new DataGoogleComputeRegionInstanceTemplateReservationAffinitySpecificReservationList(this, "specific_reservation", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get specificReservation() {
        return this._specificReservation;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateReservationAffinityOutputReference = DataGoogleComputeRegionInstanceTemplateReservationAffinityOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateReservationAffinityOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateReservationAffinityOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateReservationAffinityList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateReservationAffinityOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateReservationAffinityList = DataGoogleComputeRegionInstanceTemplateReservationAffinityList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateReservationAffinityList[_3] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateReservationAffinityList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToTerraform;
function dataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToHclTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // nanos - computed: true, optional: false, required: false
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    // seconds - computed: true, optional: false, required: false
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference = DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList = DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList[_5] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationToTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationToTerraform;
function dataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationToHclTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // nanos - computed: true, optional: false, required: false
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    // seconds - computed: true, optional: false, required: false
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationOutputReference = DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationList = DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationList[_7] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesToTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesToTerraform;
function dataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesToHclTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // operator - computed: true, optional: false, required: false
    get operator() {
        return this.getStringAttribute('operator');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesOutputReference = DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesList = DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesList[_9] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionToTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionToTerraform;
function dataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionToHclTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // discard_local_ssd - computed: true, optional: false, required: false
    get discardLocalSsd() {
        return this.getBooleanAttribute('discard_local_ssd');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionOutputReference = DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionOutputReference[_10] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionList = DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionList;
_11 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionList[_11] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateSchedulingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingToTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingToTerraform;
function dataGoogleComputeRegionInstanceTemplateSchedulingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateSchedulingToHclTerraform = dataGoogleComputeRegionInstanceTemplateSchedulingToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateSchedulingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // local_ssd_recovery_timeout - computed: true, optional: false, required: false
        this._localSsdRecoveryTimeout = new DataGoogleComputeRegionInstanceTemplateSchedulingLocalSsdRecoveryTimeoutList(this, "local_ssd_recovery_timeout", false);
        // max_run_duration - computed: true, optional: false, required: false
        this._maxRunDuration = new DataGoogleComputeRegionInstanceTemplateSchedulingMaxRunDurationList(this, "max_run_duration", false);
        // node_affinities - computed: true, optional: false, required: false
        this._nodeAffinities = new DataGoogleComputeRegionInstanceTemplateSchedulingNodeAffinitiesList(this, "node_affinities", true);
        // on_instance_stop_action - computed: true, optional: false, required: false
        this._onInstanceStopAction = new DataGoogleComputeRegionInstanceTemplateSchedulingOnInstanceStopActionList(this, "on_instance_stop_action", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // automatic_restart - computed: true, optional: false, required: false
    get automaticRestart() {
        return this.getBooleanAttribute('automatic_restart');
    }
    // instance_termination_action - computed: true, optional: false, required: false
    get instanceTerminationAction() {
        return this.getStringAttribute('instance_termination_action');
    }
    get localSsdRecoveryTimeout() {
        return this._localSsdRecoveryTimeout;
    }
    get maxRunDuration() {
        return this._maxRunDuration;
    }
    // min_node_cpus - computed: true, optional: false, required: false
    get minNodeCpus() {
        return this.getNumberAttribute('min_node_cpus');
    }
    get nodeAffinities() {
        return this._nodeAffinities;
    }
    // on_host_maintenance - computed: true, optional: false, required: false
    get onHostMaintenance() {
        return this.getStringAttribute('on_host_maintenance');
    }
    get onInstanceStopAction() {
        return this._onInstanceStopAction;
    }
    // preemptible - computed: true, optional: false, required: false
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
    // provisioning_model - computed: true, optional: false, required: false
    get provisioningModel() {
        return this.getStringAttribute('provisioning_model');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingOutputReference = DataGoogleComputeRegionInstanceTemplateSchedulingOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingOutputReference[_12] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateSchedulingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateSchedulingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateSchedulingList = DataGoogleComputeRegionInstanceTemplateSchedulingList;
_13 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateSchedulingList[_13] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateSchedulingList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateServiceAccountToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateServiceAccountToTerraform = dataGoogleComputeRegionInstanceTemplateServiceAccountToTerraform;
function dataGoogleComputeRegionInstanceTemplateServiceAccountToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateServiceAccountToHclTerraform = dataGoogleComputeRegionInstanceTemplateServiceAccountToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateServiceAccountOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
    // scopes - computed: true, optional: false, required: false
    get scopes() {
        return cdktf.Fn.tolist(this.getListAttribute('scopes'));
    }
}
exports.DataGoogleComputeRegionInstanceTemplateServiceAccountOutputReference = DataGoogleComputeRegionInstanceTemplateServiceAccountOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateServiceAccountOutputReference[_14] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateServiceAccountOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateServiceAccountList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateServiceAccountOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateServiceAccountList = DataGoogleComputeRegionInstanceTemplateServiceAccountList;
_15 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateServiceAccountList[_15] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateServiceAccountList", version: "14.13.1" };
function dataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigToTerraform = dataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigToTerraform;
function dataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigToHclTerraform = dataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigToHclTerraform;
class DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enable_integrity_monitoring - computed: true, optional: false, required: false
    get enableIntegrityMonitoring() {
        return this.getBooleanAttribute('enable_integrity_monitoring');
    }
    // enable_secure_boot - computed: true, optional: false, required: false
    get enableSecureBoot() {
        return this.getBooleanAttribute('enable_secure_boot');
    }
    // enable_vtpm - computed: true, optional: false, required: false
    get enableVtpm() {
        return this.getBooleanAttribute('enable_vtpm');
    }
}
exports.DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigOutputReference = DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigOutputReference[_16] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigOutputReference", version: "14.13.1" };
class DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigList = DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigList;
_17 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigList[_17] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_region_instance_template google_compute_region_instance_template}
*/
class DataGoogleComputeRegionInstanceTemplate extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeRegionInstanceTemplate resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeRegionInstanceTemplate to import
    * @param importFromId The id of the existing DataGoogleComputeRegionInstanceTemplate that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_region_instance_template#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeRegionInstanceTemplate to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_region_instance_template", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_region_instance_template google_compute_region_instance_template} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeRegionInstanceTemplateConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_region_instance_template',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // advanced_machine_features - computed: true, optional: false, required: false
        this._advancedMachineFeatures = new DataGoogleComputeRegionInstanceTemplateAdvancedMachineFeaturesList(this, "advanced_machine_features", false);
        // confidential_instance_config - computed: true, optional: false, required: false
        this._confidentialInstanceConfig = new DataGoogleComputeRegionInstanceTemplateConfidentialInstanceConfigList(this, "confidential_instance_config", false);
        // disk - computed: true, optional: false, required: false
        this._disk = new DataGoogleComputeRegionInstanceTemplateDiskList(this, "disk", false);
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // guest_accelerator - computed: true, optional: false, required: false
        this._guestAccelerator = new DataGoogleComputeRegionInstanceTemplateGuestAcceleratorList(this, "guest_accelerator", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // metadata - computed: true, optional: false, required: false
        this._metadata = new cdktf.StringMap(this, "metadata");
        // network_interface - computed: true, optional: false, required: false
        this._networkInterface = new DataGoogleComputeRegionInstanceTemplateNetworkInterfaceList(this, "network_interface", false);
        // network_performance_config - computed: true, optional: false, required: false
        this._networkPerformanceConfig = new DataGoogleComputeRegionInstanceTemplateNetworkPerformanceConfigList(this, "network_performance_config", false);
        // reservation_affinity - computed: true, optional: false, required: false
        this._reservationAffinity = new DataGoogleComputeRegionInstanceTemplateReservationAffinityList(this, "reservation_affinity", false);
        // resource_manager_tags - computed: true, optional: false, required: false
        this._resourceManagerTags = new cdktf.StringMap(this, "resource_manager_tags");
        // scheduling - computed: true, optional: false, required: false
        this._scheduling = new DataGoogleComputeRegionInstanceTemplateSchedulingList(this, "scheduling", false);
        // service_account - computed: true, optional: false, required: false
        this._serviceAccount = new DataGoogleComputeRegionInstanceTemplateServiceAccountList(this, "service_account", false);
        // shielded_instance_config - computed: true, optional: false, required: false
        this._shieldedInstanceConfig = new DataGoogleComputeRegionInstanceTemplateShieldedInstanceConfigList(this, "shielded_instance_config", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._filter = config.filter;
        this._id = config.id;
        this._mostRecent = config.mostRecent;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
    }
    get advancedMachineFeatures() {
        return this._advancedMachineFeatures;
    }
    // can_ip_forward - computed: true, optional: false, required: false
    get canIpForward() {
        return this.getBooleanAttribute('can_ip_forward');
    }
    get confidentialInstanceConfig() {
        return this._confidentialInstanceConfig;
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get disk() {
        return this._disk;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get guestAccelerator() {
        return this._guestAccelerator;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_description - computed: true, optional: false, required: false
    get instanceDescription() {
        return this.getStringAttribute('instance_description');
    }
    // key_revocation_action_type - computed: true, optional: false, required: false
    get keyRevocationActionType() {
        return this.getStringAttribute('key_revocation_action_type');
    }
    get labels() {
        return this._labels;
    }
    // machine_type - computed: true, optional: false, required: false
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    get metadata() {
        return this._metadata;
    }
    // metadata_fingerprint - computed: true, optional: false, required: false
    get metadataFingerprint() {
        return this.getStringAttribute('metadata_fingerprint');
    }
    // metadata_startup_script - computed: true, optional: false, required: false
    get metadataStartupScript() {
        return this.getStringAttribute('metadata_startup_script');
    }
    // min_cpu_platform - computed: true, optional: false, required: false
    get minCpuPlatform() {
        return this.getStringAttribute('min_cpu_platform');
    }
    get mostRecent() {
        return this.getBooleanAttribute('most_recent');
    }
    set mostRecent(value) {
        this._mostRecent = value;
    }
    resetMostRecent() {
        this._mostRecent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mostRecentInput() {
        return this._mostRecent;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // name_prefix - computed: true, optional: false, required: false
    get namePrefix() {
        return this.getStringAttribute('name_prefix');
    }
    get networkInterface() {
        return this._networkInterface;
    }
    get networkPerformanceConfig() {
        return this._networkPerformanceConfig;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get reservationAffinity() {
        return this._reservationAffinity;
    }
    get resourceManagerTags() {
        return this._resourceManagerTags;
    }
    // resource_policies - computed: true, optional: false, required: false
    get resourcePolicies() {
        return this.getListAttribute('resource_policies');
    }
    get scheduling() {
        return this._scheduling;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get serviceAccount() {
        return this._serviceAccount;
    }
    get shieldedInstanceConfig() {
        return this._shieldedInstanceConfig;
    }
    // tags - computed: true, optional: false, required: false
    get tags() {
        return cdktf.Fn.tolist(this.getListAttribute('tags'));
    }
    // tags_fingerprint - computed: true, optional: false, required: false
    get tagsFingerprint() {
        return this.getStringAttribute('tags_fingerprint');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filter: cdktf.stringToTerraform(this._filter),
            id: cdktf.stringToTerraform(this._id),
            most_recent: cdktf.booleanToTerraform(this._mostRecent),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            filter: {
                value: cdktf.stringToHclTerraform(this._filter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            most_recent: {
                value: cdktf.booleanToHclTerraform(this._mostRecent),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeRegionInstanceTemplate = DataGoogleComputeRegionInstanceTemplate;
_18 = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRegionInstanceTemplate[_18] = { fqn: "@cdktf/provider-google.dataGoogleComputeRegionInstanceTemplate.DataGoogleComputeRegionInstanceTemplate", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeRegionInstanceTemplate.tfResourceType = "google_compute_region_instance_template";
