// @cdktf/provider-google/lib/data-google-cloudfunctions2-function/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleCloudfunctions2Function = exports.DataGoogleCloudfunctions2FunctionServiceConfigList = exports.DataGoogleCloudfunctions2FunctionServiceConfigOutputReference = exports.dataGoogleCloudfunctions2FunctionServiceConfigToHclTerraform = exports.dataGoogleCloudfunctions2FunctionServiceConfigToTerraform = exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesList = exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesOutputReference = exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesToHclTerraform = exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesToTerraform = exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsList = exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsOutputReference = exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsToHclTerraform = exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsToTerraform = exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesList = exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesOutputReference = exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesToHclTerraform = exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesToTerraform = exports.DataGoogleCloudfunctions2FunctionEventTriggerList = exports.DataGoogleCloudfunctions2FunctionEventTriggerOutputReference = exports.dataGoogleCloudfunctions2FunctionEventTriggerToHclTerraform = exports.dataGoogleCloudfunctions2FunctionEventTriggerToTerraform = exports.DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersList = exports.DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersOutputReference = exports.dataGoogleCloudfunctions2FunctionEventTriggerEventFiltersToHclTerraform = exports.dataGoogleCloudfunctions2FunctionEventTriggerEventFiltersToTerraform = exports.DataGoogleCloudfunctions2FunctionBuildConfigList = exports.DataGoogleCloudfunctions2FunctionBuildConfigOutputReference = exports.dataGoogleCloudfunctions2FunctionBuildConfigToHclTerraform = exports.dataGoogleCloudfunctions2FunctionBuildConfigToTerraform = exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceList = exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceOutputReference = exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceToHclTerraform = exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceToTerraform = exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceList = exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceOutputReference = exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceToHclTerraform = exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceToTerraform = exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceList = exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceOutputReference = exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceToHclTerraform = exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceToTerraform = exports.DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyList = exports.DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyOutputReference = exports.dataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyToHclTerraform = exports.dataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyToTerraform = exports.DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyList = exports.DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyOutputReference = exports.dataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyToHclTerraform = exports.dataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyToTerraform = dataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyToTerraform;
function dataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyToHclTerraform = dataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyToHclTerraform;
class DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyOutputReference = DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyList = DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyList[_b] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyToTerraform = dataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyToTerraform;
function dataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyToHclTerraform = dataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyToHclTerraform;
class DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // runtime_version - computed: true, optional: false, required: false
    get runtimeVersion() {
        return this.getStringAttribute('runtime_version');
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyOutputReference = DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyList = DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyList[_d] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceToTerraform = dataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceToTerraform;
function dataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceToHclTerraform = dataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceToHclTerraform;
class DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // branch_name - computed: true, optional: false, required: false
    get branchName() {
        return this.getStringAttribute('branch_name');
    }
    // commit_sha - computed: true, optional: false, required: false
    get commitSha() {
        return this.getStringAttribute('commit_sha');
    }
    // dir - computed: true, optional: false, required: false
    get dir() {
        return this.getStringAttribute('dir');
    }
    // invert_regex - computed: true, optional: false, required: false
    get invertRegex() {
        return this.getBooleanAttribute('invert_regex');
    }
    // project_id - computed: true, optional: false, required: false
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    // repo_name - computed: true, optional: false, required: false
    get repoName() {
        return this.getStringAttribute('repo_name');
    }
    // tag_name - computed: true, optional: false, required: false
    get tagName() {
        return this.getStringAttribute('tag_name');
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceOutputReference = DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceList = DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceList[_f] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceToTerraform = dataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceToTerraform;
function dataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceToHclTerraform = dataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceToHclTerraform;
class DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // generation - computed: true, optional: false, required: false
    get generation() {
        return this.getNumberAttribute('generation');
    }
    // object - computed: true, optional: false, required: false
    get object() {
        return this.getStringAttribute('object');
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceOutputReference = DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceList = DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceList[_h] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionBuildConfigSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceToTerraform = dataGoogleCloudfunctions2FunctionBuildConfigSourceToTerraform;
function dataGoogleCloudfunctions2FunctionBuildConfigSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigSourceToHclTerraform = dataGoogleCloudfunctions2FunctionBuildConfigSourceToHclTerraform;
class DataGoogleCloudfunctions2FunctionBuildConfigSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // repo_source - computed: true, optional: false, required: false
        this._repoSource = new DataGoogleCloudfunctions2FunctionBuildConfigSourceRepoSourceList(this, "repo_source", false);
        // storage_source - computed: true, optional: false, required: false
        this._storageSource = new DataGoogleCloudfunctions2FunctionBuildConfigSourceStorageSourceList(this, "storage_source", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get repoSource() {
        return this._repoSource;
    }
    get storageSource() {
        return this._storageSource;
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceOutputReference = DataGoogleCloudfunctions2FunctionBuildConfigSourceOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigSourceOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigSourceOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionBuildConfigSourceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionBuildConfigSourceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigSourceList = DataGoogleCloudfunctions2FunctionBuildConfigSourceList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigSourceList[_k] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigSourceList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionBuildConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigToTerraform = dataGoogleCloudfunctions2FunctionBuildConfigToTerraform;
function dataGoogleCloudfunctions2FunctionBuildConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionBuildConfigToHclTerraform = dataGoogleCloudfunctions2FunctionBuildConfigToHclTerraform;
class DataGoogleCloudfunctions2FunctionBuildConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // automatic_update_policy - computed: true, optional: false, required: false
        this._automaticUpdatePolicy = new DataGoogleCloudfunctions2FunctionBuildConfigAutomaticUpdatePolicyList(this, "automatic_update_policy", false);
        // environment_variables - computed: true, optional: false, required: false
        this._environmentVariables = new cdktf.StringMap(this, "environment_variables");
        // on_deploy_update_policy - computed: true, optional: false, required: false
        this._onDeployUpdatePolicy = new DataGoogleCloudfunctions2FunctionBuildConfigOnDeployUpdatePolicyList(this, "on_deploy_update_policy", false);
        // source - computed: true, optional: false, required: false
        this._source = new DataGoogleCloudfunctions2FunctionBuildConfigSourceList(this, "source", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get automaticUpdatePolicy() {
        return this._automaticUpdatePolicy;
    }
    // build - computed: true, optional: false, required: false
    get buildAttribute() {
        return this.getStringAttribute('build');
    }
    // docker_repository - computed: true, optional: false, required: false
    get dockerRepository() {
        return this.getStringAttribute('docker_repository');
    }
    // entry_point - computed: true, optional: false, required: false
    get entryPoint() {
        return this.getStringAttribute('entry_point');
    }
    get environmentVariables() {
        return this._environmentVariables;
    }
    get onDeployUpdatePolicy() {
        return this._onDeployUpdatePolicy;
    }
    // runtime - computed: true, optional: false, required: false
    get runtime() {
        return this.getStringAttribute('runtime');
    }
    // service_account - computed: true, optional: false, required: false
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    get source() {
        return this._source;
    }
    // worker_pool - computed: true, optional: false, required: false
    get workerPool() {
        return this.getStringAttribute('worker_pool');
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigOutputReference = DataGoogleCloudfunctions2FunctionBuildConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionBuildConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionBuildConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionBuildConfigList = DataGoogleCloudfunctions2FunctionBuildConfigList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionBuildConfigList[_m] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionBuildConfigList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionEventTriggerEventFiltersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionEventTriggerEventFiltersToTerraform = dataGoogleCloudfunctions2FunctionEventTriggerEventFiltersToTerraform;
function dataGoogleCloudfunctions2FunctionEventTriggerEventFiltersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionEventTriggerEventFiltersToHclTerraform = dataGoogleCloudfunctions2FunctionEventTriggerEventFiltersToHclTerraform;
class DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // attribute - computed: true, optional: false, required: false
    get attribute() {
        return this.getStringAttribute('attribute');
    }
    // operator - computed: true, optional: false, required: false
    get operator() {
        return this.getStringAttribute('operator');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersOutputReference = DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersList = DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersList[_p] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionEventTriggerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionEventTriggerToTerraform = dataGoogleCloudfunctions2FunctionEventTriggerToTerraform;
function dataGoogleCloudfunctions2FunctionEventTriggerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionEventTriggerToHclTerraform = dataGoogleCloudfunctions2FunctionEventTriggerToHclTerraform;
class DataGoogleCloudfunctions2FunctionEventTriggerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // event_filters - computed: true, optional: false, required: false
        this._eventFilters = new DataGoogleCloudfunctions2FunctionEventTriggerEventFiltersList(this, "event_filters", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get eventFilters() {
        return this._eventFilters;
    }
    // event_type - computed: true, optional: false, required: false
    get eventType() {
        return this.getStringAttribute('event_type');
    }
    // pubsub_topic - computed: true, optional: false, required: false
    get pubsubTopic() {
        return this.getStringAttribute('pubsub_topic');
    }
    // retry_policy - computed: true, optional: false, required: false
    get retryPolicy() {
        return this.getStringAttribute('retry_policy');
    }
    // service_account_email - computed: true, optional: false, required: false
    get serviceAccountEmail() {
        return this.getStringAttribute('service_account_email');
    }
    // trigger - computed: true, optional: false, required: false
    get trigger() {
        return this.getStringAttribute('trigger');
    }
    // trigger_region - computed: true, optional: false, required: false
    get triggerRegion() {
        return this.getStringAttribute('trigger_region');
    }
}
exports.DataGoogleCloudfunctions2FunctionEventTriggerOutputReference = DataGoogleCloudfunctions2FunctionEventTriggerOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionEventTriggerOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionEventTriggerOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionEventTriggerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionEventTriggerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionEventTriggerList = DataGoogleCloudfunctions2FunctionEventTriggerList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionEventTriggerList[_r] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionEventTriggerList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesToTerraform = dataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesToTerraform;
function dataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesToHclTerraform = dataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesToHclTerraform;
class DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // key - computed: true, optional: false, required: false
    get key() {
        return this.getStringAttribute('key');
    }
    // project_id - computed: true, optional: false, required: false
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    // secret - computed: true, optional: false, required: false
    get secret() {
        return this.getStringAttribute('secret');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesOutputReference = DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesList = DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesList[_t] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsToTerraform = dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsToTerraform;
function dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsToHclTerraform = dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsToHclTerraform;
class DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsOutputReference = DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsList = DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsList[_v] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesToTerraform = dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesToTerraform;
function dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesToHclTerraform = dataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesToHclTerraform;
class DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // versions - computed: true, optional: false, required: false
        this._versions = new DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesVersionsList(this, "versions", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // mount_path - computed: true, optional: false, required: false
    get mountPath() {
        return this.getStringAttribute('mount_path');
    }
    // project_id - computed: true, optional: false, required: false
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    // secret - computed: true, optional: false, required: false
    get secret() {
        return this.getStringAttribute('secret');
    }
    get versions() {
        return this._versions;
    }
}
exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesOutputReference = DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesList = DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesList[_x] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesList", version: "14.13.1" };
function dataGoogleCloudfunctions2FunctionServiceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCloudfunctions2FunctionServiceConfigToTerraform = dataGoogleCloudfunctions2FunctionServiceConfigToTerraform;
function dataGoogleCloudfunctions2FunctionServiceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCloudfunctions2FunctionServiceConfigToHclTerraform = dataGoogleCloudfunctions2FunctionServiceConfigToHclTerraform;
class DataGoogleCloudfunctions2FunctionServiceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // environment_variables - computed: true, optional: false, required: false
        this._environmentVariables = new cdktf.StringMap(this, "environment_variables");
        // secret_environment_variables - computed: true, optional: false, required: false
        this._secretEnvironmentVariables = new DataGoogleCloudfunctions2FunctionServiceConfigSecretEnvironmentVariablesList(this, "secret_environment_variables", false);
        // secret_volumes - computed: true, optional: false, required: false
        this._secretVolumes = new DataGoogleCloudfunctions2FunctionServiceConfigSecretVolumesList(this, "secret_volumes", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all_traffic_on_latest_revision - computed: true, optional: false, required: false
    get allTrafficOnLatestRevision() {
        return this.getBooleanAttribute('all_traffic_on_latest_revision');
    }
    // available_cpu - computed: true, optional: false, required: false
    get availableCpu() {
        return this.getStringAttribute('available_cpu');
    }
    // available_memory - computed: true, optional: false, required: false
    get availableMemory() {
        return this.getStringAttribute('available_memory');
    }
    get environmentVariables() {
        return this._environmentVariables;
    }
    // gcf_uri - computed: true, optional: false, required: false
    get gcfUri() {
        return this.getStringAttribute('gcf_uri');
    }
    // ingress_settings - computed: true, optional: false, required: false
    get ingressSettings() {
        return this.getStringAttribute('ingress_settings');
    }
    // max_instance_count - computed: true, optional: false, required: false
    get maxInstanceCount() {
        return this.getNumberAttribute('max_instance_count');
    }
    // max_instance_request_concurrency - computed: true, optional: false, required: false
    get maxInstanceRequestConcurrency() {
        return this.getNumberAttribute('max_instance_request_concurrency');
    }
    // min_instance_count - computed: true, optional: false, required: false
    get minInstanceCount() {
        return this.getNumberAttribute('min_instance_count');
    }
    get secretEnvironmentVariables() {
        return this._secretEnvironmentVariables;
    }
    get secretVolumes() {
        return this._secretVolumes;
    }
    // service - computed: true, optional: false, required: false
    get service() {
        return this.getStringAttribute('service');
    }
    // service_account_email - computed: true, optional: false, required: false
    get serviceAccountEmail() {
        return this.getStringAttribute('service_account_email');
    }
    // timeout_seconds - computed: true, optional: false, required: false
    get timeoutSeconds() {
        return this.getNumberAttribute('timeout_seconds');
    }
    // uri - computed: true, optional: false, required: false
    get uri() {
        return this.getStringAttribute('uri');
    }
    // vpc_connector - computed: true, optional: false, required: false
    get vpcConnector() {
        return this.getStringAttribute('vpc_connector');
    }
    // vpc_connector_egress_settings - computed: true, optional: false, required: false
    get vpcConnectorEgressSettings() {
        return this.getStringAttribute('vpc_connector_egress_settings');
    }
}
exports.DataGoogleCloudfunctions2FunctionServiceConfigOutputReference = DataGoogleCloudfunctions2FunctionServiceConfigOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionServiceConfigOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionServiceConfigOutputReference", version: "14.13.1" };
class DataGoogleCloudfunctions2FunctionServiceConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCloudfunctions2FunctionServiceConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCloudfunctions2FunctionServiceConfigList = DataGoogleCloudfunctions2FunctionServiceConfigList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2FunctionServiceConfigList[_z] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2FunctionServiceConfigList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloudfunctions2_function google_cloudfunctions2_function}
*/
class DataGoogleCloudfunctions2Function extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleCloudfunctions2Function resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleCloudfunctions2Function to import
    * @param importFromId The id of the existing DataGoogleCloudfunctions2Function that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloudfunctions2_function#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleCloudfunctions2Function to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_cloudfunctions2_function", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/cloudfunctions2_function google_cloudfunctions2_function} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleCloudfunctions2FunctionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_cloudfunctions2_function',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // build_config - computed: true, optional: false, required: false
        this._buildConfig = new DataGoogleCloudfunctions2FunctionBuildConfigList(this, "build_config", false);
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // event_trigger - computed: true, optional: false, required: false
        this._eventTrigger = new DataGoogleCloudfunctions2FunctionEventTriggerList(this, "event_trigger", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // service_config - computed: true, optional: false, required: false
        this._serviceConfig = new DataGoogleCloudfunctions2FunctionServiceConfigList(this, "service_config", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._id = config.id;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
    }
    get buildConfig() {
        return this._buildConfig;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // environment - computed: true, optional: false, required: false
    get environment() {
        return this.getStringAttribute('environment');
    }
    get eventTrigger() {
        return this._eventTrigger;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kms_key_name - computed: true, optional: false, required: false
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    get labels() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get serviceConfig() {
        return this._serviceConfig;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleCloudfunctions2Function = DataGoogleCloudfunctions2Function;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleCloudfunctions2Function[_0] = { fqn: "@cdktf/provider-google.dataGoogleCloudfunctions2Function.DataGoogleCloudfunctions2Function", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleCloudfunctions2Function.tfResourceType = "google_cloudfunctions2_function";
