// @cdktf/provider-google/lib/network-connectivity-service-connection-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkConnectivityServiceConnectionPolicy = exports.NetworkConnectivityServiceConnectionPolicyTimeoutsOutputReference = exports.networkConnectivityServiceConnectionPolicyTimeoutsToHclTerraform = exports.networkConnectivityServiceConnectionPolicyTimeoutsToTerraform = exports.NetworkConnectivityServiceConnectionPolicyPscConfigOutputReference = exports.networkConnectivityServiceConnectionPolicyPscConfigToHclTerraform = exports.networkConnectivityServiceConnectionPolicyPscConfigToTerraform = exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsList = exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsOutputReference = exports.networkConnectivityServiceConnectionPolicyPscConnectionsToHclTerraform = exports.networkConnectivityServiceConnectionPolicyPscConnectionsToTerraform = exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoList = exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoOutputReference = exports.networkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoToHclTerraform = exports.networkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoToTerraform = exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorList = exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorOutputReference = exports.networkConnectivityServiceConnectionPolicyPscConnectionsErrorToHclTerraform = exports.networkConnectivityServiceConnectionPolicyPscConnectionsErrorToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkConnectivityServiceConnectionPolicyPscConnectionsErrorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkConnectivityServiceConnectionPolicyPscConnectionsErrorToTerraform = networkConnectivityServiceConnectionPolicyPscConnectionsErrorToTerraform;
function networkConnectivityServiceConnectionPolicyPscConnectionsErrorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.networkConnectivityServiceConnectionPolicyPscConnectionsErrorToHclTerraform = networkConnectivityServiceConnectionPolicyPscConnectionsErrorToHclTerraform;
class NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // details - computed: true, optional: false, required: false
        this._details = new cdktf.StringMapList(this, "details", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // code - computed: true, optional: false, required: false
    get code() {
        return this.getNumberAttribute('code');
    }
    get details() {
        return this._details;
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
}
exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorOutputReference = NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorOutputReference[_a] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorOutputReference", version: "14.13.1" };
class NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorList = NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorList;
_b = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorList[_b] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorList", version: "14.13.1" };
function networkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoToTerraform = networkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoToTerraform;
function networkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.networkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoToHclTerraform = networkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoToHclTerraform;
class NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // metadata - computed: true, optional: false, required: false
        this._metadata = new cdktf.StringMap(this, "metadata");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
    get metadata() {
        return this._metadata;
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
}
exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoOutputReference = NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoOutputReference[_c] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoOutputReference", version: "14.13.1" };
class NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoList = NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoList;
_d = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoList[_d] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoList", version: "14.13.1" };
function networkConnectivityServiceConnectionPolicyPscConnectionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkConnectivityServiceConnectionPolicyPscConnectionsToTerraform = networkConnectivityServiceConnectionPolicyPscConnectionsToTerraform;
function networkConnectivityServiceConnectionPolicyPscConnectionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.networkConnectivityServiceConnectionPolicyPscConnectionsToHclTerraform = networkConnectivityServiceConnectionPolicyPscConnectionsToHclTerraform;
class NetworkConnectivityServiceConnectionPolicyPscConnectionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // error - computed: true, optional: false, required: false
        this._error = new NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorList(this, "error", false);
        // error_info - computed: true, optional: false, required: false
        this._errorInfo = new NetworkConnectivityServiceConnectionPolicyPscConnectionsErrorInfoList(this, "error_info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // consumer_address - computed: true, optional: false, required: false
    get consumerAddress() {
        return this.getStringAttribute('consumer_address');
    }
    // consumer_forwarding_rule - computed: true, optional: false, required: false
    get consumerForwardingRule() {
        return this.getStringAttribute('consumer_forwarding_rule');
    }
    // consumer_target_project - computed: true, optional: false, required: false
    get consumerTargetProject() {
        return this.getStringAttribute('consumer_target_project');
    }
    get error() {
        return this._error;
    }
    get errorInfo() {
        return this._errorInfo;
    }
    // error_type - computed: true, optional: false, required: false
    get errorType() {
        return this.getStringAttribute('error_type');
    }
    // gce_operation - computed: true, optional: false, required: false
    get gceOperation() {
        return this.getStringAttribute('gce_operation');
    }
    // psc_connection_id - computed: true, optional: false, required: false
    get pscConnectionId() {
        return this.getStringAttribute('psc_connection_id');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsOutputReference = NetworkConnectivityServiceConnectionPolicyPscConnectionsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicyPscConnectionsOutputReference[_e] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicyPscConnectionsOutputReference", version: "14.13.1" };
class NetworkConnectivityServiceConnectionPolicyPscConnectionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkConnectivityServiceConnectionPolicyPscConnectionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkConnectivityServiceConnectionPolicyPscConnectionsList = NetworkConnectivityServiceConnectionPolicyPscConnectionsList;
_f = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicyPscConnectionsList[_f] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicyPscConnectionsList", version: "14.13.1" };
function networkConnectivityServiceConnectionPolicyPscConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        limit: cdktf.stringToTerraform(struct.limit),
        subnetworks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.subnetworks),
    };
}
exports.networkConnectivityServiceConnectionPolicyPscConfigToTerraform = networkConnectivityServiceConnectionPolicyPscConfigToTerraform;
function networkConnectivityServiceConnectionPolicyPscConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        limit: {
            value: cdktf.stringToHclTerraform(struct.limit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subnetworks: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.subnetworks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivityServiceConnectionPolicyPscConfigToHclTerraform = networkConnectivityServiceConnectionPolicyPscConfigToHclTerraform;
class NetworkConnectivityServiceConnectionPolicyPscConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._limit !== undefined) {
            hasAnyValues = true;
            internalValueResult.limit = this._limit;
        }
        if (this._subnetworks !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetworks = this._subnetworks;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._limit = undefined;
            this._subnetworks = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._limit = value.limit;
            this._subnetworks = value.subnetworks;
        }
    }
    get limit() {
        return this.getStringAttribute('limit');
    }
    set limit(value) {
        this._limit = value;
    }
    resetLimit() {
        this._limit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get limitInput() {
        return this._limit;
    }
    get subnetworks() {
        return this.getListAttribute('subnetworks');
    }
    set subnetworks(value) {
        this._subnetworks = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetworksInput() {
        return this._subnetworks;
    }
}
exports.NetworkConnectivityServiceConnectionPolicyPscConfigOutputReference = NetworkConnectivityServiceConnectionPolicyPscConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicyPscConfigOutputReference[_g] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicyPscConfigOutputReference", version: "14.13.1" };
function networkConnectivityServiceConnectionPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkConnectivityServiceConnectionPolicyTimeoutsToTerraform = networkConnectivityServiceConnectionPolicyTimeoutsToTerraform;
function networkConnectivityServiceConnectionPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivityServiceConnectionPolicyTimeoutsToHclTerraform = networkConnectivityServiceConnectionPolicyTimeoutsToHclTerraform;
class NetworkConnectivityServiceConnectionPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkConnectivityServiceConnectionPolicyTimeoutsOutputReference = NetworkConnectivityServiceConnectionPolicyTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicyTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_service_connection_policy google_network_connectivity_service_connection_policy}
*/
class NetworkConnectivityServiceConnectionPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkConnectivityServiceConnectionPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkConnectivityServiceConnectionPolicy to import
    * @param importFromId The id of the existing NetworkConnectivityServiceConnectionPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_service_connection_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkConnectivityServiceConnectionPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_network_connectivity_service_connection_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_service_connection_policy google_network_connectivity_service_connection_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkConnectivityServiceConnectionPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_network_connectivity_service_connection_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // psc_connections - computed: true, optional: false, required: false
        this._pscConnections = new NetworkConnectivityServiceConnectionPolicyPscConnectionsList(this, "psc_connections", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // psc_config - computed: false, optional: true, required: false
        this._pscConfig = new NetworkConnectivityServiceConnectionPolicyPscConfigOutputReference(this, "psc_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkConnectivityServiceConnectionPolicyTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._network = config.network;
        this._project = config.project;
        this._serviceClass = config.serviceClass;
        this._pscConfig.internalValue = config.pscConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // infrastructure - computed: true, optional: false, required: false
    get infrastructure() {
        return this.getStringAttribute('infrastructure');
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get pscConnections() {
        return this._pscConnections;
    }
    get serviceClass() {
        return this.getStringAttribute('service_class');
    }
    set serviceClass(value) {
        this._serviceClass = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceClassInput() {
        return this._serviceClass;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get pscConfig() {
        return this._pscConfig;
    }
    putPscConfig(value) {
        this._pscConfig.internalValue = value;
    }
    resetPscConfig() {
        this._pscConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pscConfigInput() {
        return this._pscConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            network: cdktf.stringToTerraform(this._network),
            project: cdktf.stringToTerraform(this._project),
            service_class: cdktf.stringToTerraform(this._serviceClass),
            psc_config: networkConnectivityServiceConnectionPolicyPscConfigToTerraform(this._pscConfig.internalValue),
            timeouts: networkConnectivityServiceConnectionPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            network: {
                value: cdktf.stringToHclTerraform(this._network),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            service_class: {
                value: cdktf.stringToHclTerraform(this._serviceClass),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            psc_config: {
                value: networkConnectivityServiceConnectionPolicyPscConfigToHclTerraform(this._pscConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivityServiceConnectionPolicyPscConfigList",
            },
            timeouts: {
                value: networkConnectivityServiceConnectionPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkConnectivityServiceConnectionPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkConnectivityServiceConnectionPolicy = NetworkConnectivityServiceConnectionPolicy;
_j = JSII_RTTI_SYMBOL_1;
NetworkConnectivityServiceConnectionPolicy[_j] = { fqn: "@cdktf/provider-google.networkConnectivityServiceConnectionPolicy.NetworkConnectivityServiceConnectionPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetworkConnectivityServiceConnectionPolicy.tfResourceType = "google_network_connectivity_service_connection_policy";
