// @cdktf/provider-google/lib/data-google-compute-machine-types/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeMachineTypes = exports.DataGoogleComputeMachineTypesMachineTypesList = exports.DataGoogleComputeMachineTypesMachineTypesOutputReference = exports.dataGoogleComputeMachineTypesMachineTypesToHclTerraform = exports.dataGoogleComputeMachineTypesMachineTypesToTerraform = exports.DataGoogleComputeMachineTypesMachineTypesDeprecatedList = exports.DataGoogleComputeMachineTypesMachineTypesDeprecatedOutputReference = exports.dataGoogleComputeMachineTypesMachineTypesDeprecatedToHclTerraform = exports.dataGoogleComputeMachineTypesMachineTypesDeprecatedToTerraform = exports.DataGoogleComputeMachineTypesMachineTypesAcceleratorsList = exports.DataGoogleComputeMachineTypesMachineTypesAcceleratorsOutputReference = exports.dataGoogleComputeMachineTypesMachineTypesAcceleratorsToHclTerraform = exports.dataGoogleComputeMachineTypesMachineTypesAcceleratorsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeMachineTypesMachineTypesAcceleratorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeMachineTypesMachineTypesAcceleratorsToTerraform = dataGoogleComputeMachineTypesMachineTypesAcceleratorsToTerraform;
function dataGoogleComputeMachineTypesMachineTypesAcceleratorsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeMachineTypesMachineTypesAcceleratorsToHclTerraform = dataGoogleComputeMachineTypesMachineTypesAcceleratorsToHclTerraform;
class DataGoogleComputeMachineTypesMachineTypesAcceleratorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // guest_accelerator_count - computed: true, optional: false, required: false
    get guestAcceleratorCount() {
        return this.getNumberAttribute('guest_accelerator_count');
    }
    // guest_accelerator_type - computed: true, optional: false, required: false
    get guestAcceleratorType() {
        return this.getStringAttribute('guest_accelerator_type');
    }
}
exports.DataGoogleComputeMachineTypesMachineTypesAcceleratorsOutputReference = DataGoogleComputeMachineTypesMachineTypesAcceleratorsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeMachineTypesMachineTypesAcceleratorsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeMachineTypes.DataGoogleComputeMachineTypesMachineTypesAcceleratorsOutputReference", version: "14.13.1" };
class DataGoogleComputeMachineTypesMachineTypesAcceleratorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeMachineTypesMachineTypesAcceleratorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeMachineTypesMachineTypesAcceleratorsList = DataGoogleComputeMachineTypesMachineTypesAcceleratorsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeMachineTypesMachineTypesAcceleratorsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeMachineTypes.DataGoogleComputeMachineTypesMachineTypesAcceleratorsList", version: "14.13.1" };
function dataGoogleComputeMachineTypesMachineTypesDeprecatedToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeMachineTypesMachineTypesDeprecatedToTerraform = dataGoogleComputeMachineTypesMachineTypesDeprecatedToTerraform;
function dataGoogleComputeMachineTypesMachineTypesDeprecatedToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeMachineTypesMachineTypesDeprecatedToHclTerraform = dataGoogleComputeMachineTypesMachineTypesDeprecatedToHclTerraform;
class DataGoogleComputeMachineTypesMachineTypesDeprecatedOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // replacement - computed: true, optional: false, required: false
    get replacement() {
        return this.getStringAttribute('replacement');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.DataGoogleComputeMachineTypesMachineTypesDeprecatedOutputReference = DataGoogleComputeMachineTypesMachineTypesDeprecatedOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeMachineTypesMachineTypesDeprecatedOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeMachineTypes.DataGoogleComputeMachineTypesMachineTypesDeprecatedOutputReference", version: "14.13.1" };
class DataGoogleComputeMachineTypesMachineTypesDeprecatedList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeMachineTypesMachineTypesDeprecatedOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeMachineTypesMachineTypesDeprecatedList = DataGoogleComputeMachineTypesMachineTypesDeprecatedList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeMachineTypesMachineTypesDeprecatedList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeMachineTypes.DataGoogleComputeMachineTypesMachineTypesDeprecatedList", version: "14.13.1" };
function dataGoogleComputeMachineTypesMachineTypesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeMachineTypesMachineTypesToTerraform = dataGoogleComputeMachineTypesMachineTypesToTerraform;
function dataGoogleComputeMachineTypesMachineTypesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeMachineTypesMachineTypesToHclTerraform = dataGoogleComputeMachineTypesMachineTypesToHclTerraform;
class DataGoogleComputeMachineTypesMachineTypesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // accelerators - computed: true, optional: false, required: false
        this._accelerators = new DataGoogleComputeMachineTypesMachineTypesAcceleratorsList(this, "accelerators", false);
        // deprecated - computed: true, optional: false, required: false
        this._deprecated = new DataGoogleComputeMachineTypesMachineTypesDeprecatedList(this, "deprecated", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get accelerators() {
        return this._accelerators;
    }
    get deprecated() {
        return this._deprecated;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // guest_cpus - computed: true, optional: false, required: false
    get guestCpus() {
        return this.getNumberAttribute('guest_cpus');
    }
    // is_shared_cpus - computed: true, optional: false, required: false
    get isSharedCpus() {
        return this.getBooleanAttribute('is_shared_cpus');
    }
    // maximum_persistent_disks - computed: true, optional: false, required: false
    get maximumPersistentDisks() {
        return this.getNumberAttribute('maximum_persistent_disks');
    }
    // maximum_persistent_disks_size_gb - computed: true, optional: false, required: false
    get maximumPersistentDisksSizeGb() {
        return this.getNumberAttribute('maximum_persistent_disks_size_gb');
    }
    // memory_mb - computed: true, optional: false, required: false
    get memoryMb() {
        return this.getNumberAttribute('memory_mb');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
}
exports.DataGoogleComputeMachineTypesMachineTypesOutputReference = DataGoogleComputeMachineTypesMachineTypesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeMachineTypesMachineTypesOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeMachineTypes.DataGoogleComputeMachineTypesMachineTypesOutputReference", version: "14.13.1" };
class DataGoogleComputeMachineTypesMachineTypesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeMachineTypesMachineTypesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeMachineTypesMachineTypesList = DataGoogleComputeMachineTypesMachineTypesList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleComputeMachineTypesMachineTypesList[_f] = { fqn: "@cdktf/provider-google.dataGoogleComputeMachineTypes.DataGoogleComputeMachineTypesMachineTypesList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_machine_types google_compute_machine_types}
*/
class DataGoogleComputeMachineTypes extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeMachineTypes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeMachineTypes to import
    * @param importFromId The id of the existing DataGoogleComputeMachineTypes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_machine_types#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeMachineTypes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_machine_types", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_machine_types google_compute_machine_types} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeMachineTypesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_compute_machine_types',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // machine_types - computed: true, optional: false, required: false
        this._machineTypes = new DataGoogleComputeMachineTypesMachineTypesList(this, "machine_types", false);
        this._filter = config.filter;
        this._id = config.id;
        this._project = config.project;
        this._zone = config.zone;
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get machineTypes() {
        return this._machineTypes;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filter: cdktf.stringToTerraform(this._filter),
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
            zone: cdktf.stringToTerraform(this._zone),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            filter: {
                value: cdktf.stringToHclTerraform(this._filter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            zone: {
                value: cdktf.stringToHclTerraform(this._zone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeMachineTypes = DataGoogleComputeMachineTypes;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleComputeMachineTypes[_g] = { fqn: "@cdktf/provider-google.dataGoogleComputeMachineTypes.DataGoogleComputeMachineTypes", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeMachineTypes.tfResourceType = "google_compute_machine_types";
