// @cdktf/provider-google/lib/compute-instance-group-manager/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2;
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeInstanceGroupManagerUpdatePolicyToHclTerraform = exports.computeInstanceGroupManagerUpdatePolicyToTerraform = exports.ComputeInstanceGroupManagerTimeoutsOutputReference = exports.computeInstanceGroupManagerTimeoutsToHclTerraform = exports.computeInstanceGroupManagerTimeoutsToTerraform = exports.ComputeInstanceGroupManagerStatefulInternalIpList = exports.ComputeInstanceGroupManagerStatefulInternalIpOutputReference = exports.computeInstanceGroupManagerStatefulInternalIpToHclTerraform = exports.computeInstanceGroupManagerStatefulInternalIpToTerraform = exports.ComputeInstanceGroupManagerStatefulExternalIpList = exports.ComputeInstanceGroupManagerStatefulExternalIpOutputReference = exports.computeInstanceGroupManagerStatefulExternalIpToHclTerraform = exports.computeInstanceGroupManagerStatefulExternalIpToTerraform = exports.ComputeInstanceGroupManagerStatefulDiskList = exports.ComputeInstanceGroupManagerStatefulDiskOutputReference = exports.computeInstanceGroupManagerStatefulDiskToHclTerraform = exports.computeInstanceGroupManagerStatefulDiskToTerraform = exports.ComputeInstanceGroupManagerNamedPortList = exports.ComputeInstanceGroupManagerNamedPortOutputReference = exports.computeInstanceGroupManagerNamedPortToHclTerraform = exports.computeInstanceGroupManagerNamedPortToTerraform = exports.ComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference = exports.computeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform = exports.computeInstanceGroupManagerInstanceLifecyclePolicyToTerraform = exports.ComputeInstanceGroupManagerAutoHealingPoliciesOutputReference = exports.computeInstanceGroupManagerAutoHealingPoliciesToHclTerraform = exports.computeInstanceGroupManagerAutoHealingPoliciesToTerraform = exports.ComputeInstanceGroupManagerAllInstancesConfigOutputReference = exports.computeInstanceGroupManagerAllInstancesConfigToHclTerraform = exports.computeInstanceGroupManagerAllInstancesConfigToTerraform = exports.ComputeInstanceGroupManagerStatusList = exports.ComputeInstanceGroupManagerStatusOutputReference = exports.computeInstanceGroupManagerStatusToHclTerraform = exports.computeInstanceGroupManagerStatusToTerraform = exports.ComputeInstanceGroupManagerStatusVersionTargetList = exports.ComputeInstanceGroupManagerStatusVersionTargetOutputReference = exports.computeInstanceGroupManagerStatusVersionTargetToHclTerraform = exports.computeInstanceGroupManagerStatusVersionTargetToTerraform = exports.ComputeInstanceGroupManagerStatusStatefulList = exports.ComputeInstanceGroupManagerStatusStatefulOutputReference = exports.computeInstanceGroupManagerStatusStatefulToHclTerraform = exports.computeInstanceGroupManagerStatusStatefulToTerraform = exports.ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList = exports.ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = exports.computeInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform = exports.computeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = exports.ComputeInstanceGroupManagerStatusAllInstancesConfigList = exports.ComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference = exports.computeInstanceGroupManagerStatusAllInstancesConfigToHclTerraform = exports.computeInstanceGroupManagerStatusAllInstancesConfigToTerraform = void 0;
exports.ComputeInstanceGroupManager = exports.ComputeInstanceGroupManagerVersionList = exports.ComputeInstanceGroupManagerVersionOutputReference = exports.computeInstanceGroupManagerVersionToHclTerraform = exports.computeInstanceGroupManagerVersionToTerraform = exports.ComputeInstanceGroupManagerVersionTargetSizeOutputReference = exports.computeInstanceGroupManagerVersionTargetSizeToHclTerraform = exports.computeInstanceGroupManagerVersionTargetSizeToTerraform = exports.ComputeInstanceGroupManagerUpdatePolicyOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeInstanceGroupManagerStatusAllInstancesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeInstanceGroupManagerStatusAllInstancesConfigToTerraform = computeInstanceGroupManagerStatusAllInstancesConfigToTerraform;
function computeInstanceGroupManagerStatusAllInstancesConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeInstanceGroupManagerStatusAllInstancesConfigToHclTerraform = computeInstanceGroupManagerStatusAllInstancesConfigToHclTerraform;
class ComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // current_revision - computed: true, optional: false, required: false
    get currentRevision() {
        return this.getStringAttribute('current_revision');
    }
    // effective - computed: true, optional: false, required: false
    get effective() {
        return this.getBooleanAttribute('effective');
    }
}
exports.ComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference = ComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerStatusAllInstancesConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerStatusAllInstancesConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerStatusAllInstancesConfigList = ComputeInstanceGroupManagerStatusAllInstancesConfigList;
_b = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusAllInstancesConfigList[_b] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusAllInstancesConfigList", version: "14.13.1" };
function computeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform = computeInstanceGroupManagerStatusStatefulPerInstanceConfigsToTerraform;
function computeInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform = computeInstanceGroupManagerStatusStatefulPerInstanceConfigsToHclTerraform;
class ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all_effective - computed: true, optional: false, required: false
    get allEffective() {
        return this.getBooleanAttribute('all_effective');
    }
}
exports.ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference = ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference[_c] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList = ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList;
_d = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList[_d] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList", version: "14.13.1" };
function computeInstanceGroupManagerStatusStatefulToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeInstanceGroupManagerStatusStatefulToTerraform = computeInstanceGroupManagerStatusStatefulToTerraform;
function computeInstanceGroupManagerStatusStatefulToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeInstanceGroupManagerStatusStatefulToHclTerraform = computeInstanceGroupManagerStatusStatefulToHclTerraform;
class ComputeInstanceGroupManagerStatusStatefulOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // per_instance_configs - computed: true, optional: false, required: false
        this._perInstanceConfigs = new ComputeInstanceGroupManagerStatusStatefulPerInstanceConfigsList(this, "per_instance_configs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // has_stateful_config - computed: true, optional: false, required: false
    get hasStatefulConfig() {
        return this.getBooleanAttribute('has_stateful_config');
    }
    get perInstanceConfigs() {
        return this._perInstanceConfigs;
    }
}
exports.ComputeInstanceGroupManagerStatusStatefulOutputReference = ComputeInstanceGroupManagerStatusStatefulOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusStatefulOutputReference[_e] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusStatefulOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerStatusStatefulList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerStatusStatefulOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerStatusStatefulList = ComputeInstanceGroupManagerStatusStatefulList;
_f = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusStatefulList[_f] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusStatefulList", version: "14.13.1" };
function computeInstanceGroupManagerStatusVersionTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeInstanceGroupManagerStatusVersionTargetToTerraform = computeInstanceGroupManagerStatusVersionTargetToTerraform;
function computeInstanceGroupManagerStatusVersionTargetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeInstanceGroupManagerStatusVersionTargetToHclTerraform = computeInstanceGroupManagerStatusVersionTargetToHclTerraform;
class ComputeInstanceGroupManagerStatusVersionTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_reached - computed: true, optional: false, required: false
    get isReached() {
        return this.getBooleanAttribute('is_reached');
    }
}
exports.ComputeInstanceGroupManagerStatusVersionTargetOutputReference = ComputeInstanceGroupManagerStatusVersionTargetOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusVersionTargetOutputReference[_g] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusVersionTargetOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerStatusVersionTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerStatusVersionTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerStatusVersionTargetList = ComputeInstanceGroupManagerStatusVersionTargetList;
_h = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusVersionTargetList[_h] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusVersionTargetList", version: "14.13.1" };
function computeInstanceGroupManagerStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.computeInstanceGroupManagerStatusToTerraform = computeInstanceGroupManagerStatusToTerraform;
function computeInstanceGroupManagerStatusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.computeInstanceGroupManagerStatusToHclTerraform = computeInstanceGroupManagerStatusToHclTerraform;
class ComputeInstanceGroupManagerStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // all_instances_config - computed: true, optional: false, required: false
        this._allInstancesConfig = new ComputeInstanceGroupManagerStatusAllInstancesConfigList(this, "all_instances_config", false);
        // stateful - computed: true, optional: false, required: false
        this._stateful = new ComputeInstanceGroupManagerStatusStatefulList(this, "stateful", false);
        // version_target - computed: true, optional: false, required: false
        this._versionTarget = new ComputeInstanceGroupManagerStatusVersionTargetList(this, "version_target", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get allInstancesConfig() {
        return this._allInstancesConfig;
    }
    // is_stable - computed: true, optional: false, required: false
    get isStable() {
        return this.getBooleanAttribute('is_stable');
    }
    get stateful() {
        return this._stateful;
    }
    get versionTarget() {
        return this._versionTarget;
    }
}
exports.ComputeInstanceGroupManagerStatusOutputReference = ComputeInstanceGroupManagerStatusOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusOutputReference[_j] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerStatusList = ComputeInstanceGroupManagerStatusList;
_k = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatusList[_k] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatusList", version: "14.13.1" };
function computeInstanceGroupManagerAllInstancesConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        metadata: cdktf.hashMapper(cdktf.stringToTerraform)(struct.metadata),
    };
}
exports.computeInstanceGroupManagerAllInstancesConfigToTerraform = computeInstanceGroupManagerAllInstancesConfigToTerraform;
function computeInstanceGroupManagerAllInstancesConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        labels: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.labels),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        metadata: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.metadata),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerAllInstancesConfigToHclTerraform = computeInstanceGroupManagerAllInstancesConfigToHclTerraform;
class ComputeInstanceGroupManagerAllInstancesConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._labels = undefined;
            this._metadata = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._labels = value.labels;
            this._metadata = value.metadata;
        }
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get metadata() {
        return this.getStringMapAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
}
exports.ComputeInstanceGroupManagerAllInstancesConfigOutputReference = ComputeInstanceGroupManagerAllInstancesConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerAllInstancesConfigOutputReference[_l] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerAllInstancesConfigOutputReference", version: "14.13.1" };
function computeInstanceGroupManagerAutoHealingPoliciesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        health_check: cdktf.stringToTerraform(struct.healthCheck),
        initial_delay_sec: cdktf.numberToTerraform(struct.initialDelaySec),
    };
}
exports.computeInstanceGroupManagerAutoHealingPoliciesToTerraform = computeInstanceGroupManagerAutoHealingPoliciesToTerraform;
function computeInstanceGroupManagerAutoHealingPoliciesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        health_check: {
            value: cdktf.stringToHclTerraform(struct.healthCheck),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        initial_delay_sec: {
            value: cdktf.numberToHclTerraform(struct.initialDelaySec),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerAutoHealingPoliciesToHclTerraform = computeInstanceGroupManagerAutoHealingPoliciesToHclTerraform;
class ComputeInstanceGroupManagerAutoHealingPoliciesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._healthCheck !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthCheck = this._healthCheck;
        }
        if (this._initialDelaySec !== undefined) {
            hasAnyValues = true;
            internalValueResult.initialDelaySec = this._initialDelaySec;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._healthCheck = undefined;
            this._initialDelaySec = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._healthCheck = value.healthCheck;
            this._initialDelaySec = value.initialDelaySec;
        }
    }
    get healthCheck() {
        return this.getStringAttribute('health_check');
    }
    set healthCheck(value) {
        this._healthCheck = value;
    }
    // Temporarily expose input value. Use with caution.
    get healthCheckInput() {
        return this._healthCheck;
    }
    get initialDelaySec() {
        return this.getNumberAttribute('initial_delay_sec');
    }
    set initialDelaySec(value) {
        this._initialDelaySec = value;
    }
    // Temporarily expose input value. Use with caution.
    get initialDelaySecInput() {
        return this._initialDelaySec;
    }
}
exports.ComputeInstanceGroupManagerAutoHealingPoliciesOutputReference = ComputeInstanceGroupManagerAutoHealingPoliciesOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerAutoHealingPoliciesOutputReference[_m] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerAutoHealingPoliciesOutputReference", version: "14.13.1" };
function computeInstanceGroupManagerInstanceLifecyclePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_action_on_failure: cdktf.stringToTerraform(struct.defaultActionOnFailure),
        force_update_on_repair: cdktf.stringToTerraform(struct.forceUpdateOnRepair),
    };
}
exports.computeInstanceGroupManagerInstanceLifecyclePolicyToTerraform = computeInstanceGroupManagerInstanceLifecyclePolicyToTerraform;
function computeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        default_action_on_failure: {
            value: cdktf.stringToHclTerraform(struct.defaultActionOnFailure),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        force_update_on_repair: {
            value: cdktf.stringToHclTerraform(struct.forceUpdateOnRepair),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform = computeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform;
class ComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultActionOnFailure !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultActionOnFailure = this._defaultActionOnFailure;
        }
        if (this._forceUpdateOnRepair !== undefined) {
            hasAnyValues = true;
            internalValueResult.forceUpdateOnRepair = this._forceUpdateOnRepair;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultActionOnFailure = undefined;
            this._forceUpdateOnRepair = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultActionOnFailure = value.defaultActionOnFailure;
            this._forceUpdateOnRepair = value.forceUpdateOnRepair;
        }
    }
    get defaultActionOnFailure() {
        return this.getStringAttribute('default_action_on_failure');
    }
    set defaultActionOnFailure(value) {
        this._defaultActionOnFailure = value;
    }
    resetDefaultActionOnFailure() {
        this._defaultActionOnFailure = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultActionOnFailureInput() {
        return this._defaultActionOnFailure;
    }
    get forceUpdateOnRepair() {
        return this.getStringAttribute('force_update_on_repair');
    }
    set forceUpdateOnRepair(value) {
        this._forceUpdateOnRepair = value;
    }
    resetForceUpdateOnRepair() {
        this._forceUpdateOnRepair = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forceUpdateOnRepairInput() {
        return this._forceUpdateOnRepair;
    }
}
exports.ComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference = ComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference[_o] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference", version: "14.13.1" };
function computeInstanceGroupManagerNamedPortToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.computeInstanceGroupManagerNamedPortToTerraform = computeInstanceGroupManagerNamedPortToTerraform;
function computeInstanceGroupManagerNamedPortToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        port: {
            value: cdktf.numberToHclTerraform(struct.port),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerNamedPortToHclTerraform = computeInstanceGroupManagerNamedPortToHclTerraform;
class ComputeInstanceGroupManagerNamedPortOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._port = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._port = value.port;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.ComputeInstanceGroupManagerNamedPortOutputReference = ComputeInstanceGroupManagerNamedPortOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerNamedPortOutputReference[_p] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerNamedPortOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerNamedPortList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerNamedPortOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerNamedPortList = ComputeInstanceGroupManagerNamedPortList;
_q = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerNamedPortList[_q] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerNamedPortList", version: "14.13.1" };
function computeInstanceGroupManagerStatefulDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_rule: cdktf.stringToTerraform(struct.deleteRule),
        device_name: cdktf.stringToTerraform(struct.deviceName),
    };
}
exports.computeInstanceGroupManagerStatefulDiskToTerraform = computeInstanceGroupManagerStatefulDiskToTerraform;
function computeInstanceGroupManagerStatefulDiskToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        delete_rule: {
            value: cdktf.stringToHclTerraform(struct.deleteRule),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        device_name: {
            value: cdktf.stringToHclTerraform(struct.deviceName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerStatefulDiskToHclTerraform = computeInstanceGroupManagerStatefulDiskToHclTerraform;
class ComputeInstanceGroupManagerStatefulDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteRule !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteRule = this._deleteRule;
        }
        if (this._deviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.deviceName = this._deviceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._deleteRule = undefined;
            this._deviceName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._deleteRule = value.deleteRule;
            this._deviceName = value.deviceName;
        }
    }
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    set deleteRule(value) {
        this._deleteRule = value;
    }
    resetDeleteRule() {
        this._deleteRule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteRuleInput() {
        return this._deleteRule;
    }
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    set deviceName(value) {
        this._deviceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get deviceNameInput() {
        return this._deviceName;
    }
}
exports.ComputeInstanceGroupManagerStatefulDiskOutputReference = ComputeInstanceGroupManagerStatefulDiskOutputReference;
_r = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatefulDiskOutputReference[_r] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatefulDiskOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerStatefulDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerStatefulDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerStatefulDiskList = ComputeInstanceGroupManagerStatefulDiskList;
_s = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatefulDiskList[_s] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatefulDiskList", version: "14.13.1" };
function computeInstanceGroupManagerStatefulExternalIpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_rule: cdktf.stringToTerraform(struct.deleteRule),
        interface_name: cdktf.stringToTerraform(struct.interfaceName),
    };
}
exports.computeInstanceGroupManagerStatefulExternalIpToTerraform = computeInstanceGroupManagerStatefulExternalIpToTerraform;
function computeInstanceGroupManagerStatefulExternalIpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        delete_rule: {
            value: cdktf.stringToHclTerraform(struct.deleteRule),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        interface_name: {
            value: cdktf.stringToHclTerraform(struct.interfaceName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerStatefulExternalIpToHclTerraform = computeInstanceGroupManagerStatefulExternalIpToHclTerraform;
class ComputeInstanceGroupManagerStatefulExternalIpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteRule !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteRule = this._deleteRule;
        }
        if (this._interfaceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.interfaceName = this._interfaceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._deleteRule = undefined;
            this._interfaceName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._deleteRule = value.deleteRule;
            this._interfaceName = value.interfaceName;
        }
    }
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    set deleteRule(value) {
        this._deleteRule = value;
    }
    resetDeleteRule() {
        this._deleteRule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteRuleInput() {
        return this._deleteRule;
    }
    get interfaceName() {
        return this.getStringAttribute('interface_name');
    }
    set interfaceName(value) {
        this._interfaceName = value;
    }
    resetInterfaceName() {
        this._interfaceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get interfaceNameInput() {
        return this._interfaceName;
    }
}
exports.ComputeInstanceGroupManagerStatefulExternalIpOutputReference = ComputeInstanceGroupManagerStatefulExternalIpOutputReference;
_t = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatefulExternalIpOutputReference[_t] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatefulExternalIpOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerStatefulExternalIpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerStatefulExternalIpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerStatefulExternalIpList = ComputeInstanceGroupManagerStatefulExternalIpList;
_u = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatefulExternalIpList[_u] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatefulExternalIpList", version: "14.13.1" };
function computeInstanceGroupManagerStatefulInternalIpToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_rule: cdktf.stringToTerraform(struct.deleteRule),
        interface_name: cdktf.stringToTerraform(struct.interfaceName),
    };
}
exports.computeInstanceGroupManagerStatefulInternalIpToTerraform = computeInstanceGroupManagerStatefulInternalIpToTerraform;
function computeInstanceGroupManagerStatefulInternalIpToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        delete_rule: {
            value: cdktf.stringToHclTerraform(struct.deleteRule),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        interface_name: {
            value: cdktf.stringToHclTerraform(struct.interfaceName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerStatefulInternalIpToHclTerraform = computeInstanceGroupManagerStatefulInternalIpToHclTerraform;
class ComputeInstanceGroupManagerStatefulInternalIpOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteRule !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteRule = this._deleteRule;
        }
        if (this._interfaceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.interfaceName = this._interfaceName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._deleteRule = undefined;
            this._interfaceName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._deleteRule = value.deleteRule;
            this._interfaceName = value.interfaceName;
        }
    }
    get deleteRule() {
        return this.getStringAttribute('delete_rule');
    }
    set deleteRule(value) {
        this._deleteRule = value;
    }
    resetDeleteRule() {
        this._deleteRule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteRuleInput() {
        return this._deleteRule;
    }
    get interfaceName() {
        return this.getStringAttribute('interface_name');
    }
    set interfaceName(value) {
        this._interfaceName = value;
    }
    resetInterfaceName() {
        this._interfaceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get interfaceNameInput() {
        return this._interfaceName;
    }
}
exports.ComputeInstanceGroupManagerStatefulInternalIpOutputReference = ComputeInstanceGroupManagerStatefulInternalIpOutputReference;
_v = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatefulInternalIpOutputReference[_v] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatefulInternalIpOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerStatefulInternalIpList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerStatefulInternalIpOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerStatefulInternalIpList = ComputeInstanceGroupManagerStatefulInternalIpList;
_w = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerStatefulInternalIpList[_w] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerStatefulInternalIpList", version: "14.13.1" };
function computeInstanceGroupManagerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.computeInstanceGroupManagerTimeoutsToTerraform = computeInstanceGroupManagerTimeoutsToTerraform;
function computeInstanceGroupManagerTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerTimeoutsToHclTerraform = computeInstanceGroupManagerTimeoutsToHclTerraform;
class ComputeInstanceGroupManagerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ComputeInstanceGroupManagerTimeoutsOutputReference = ComputeInstanceGroupManagerTimeoutsOutputReference;
_x = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerTimeoutsOutputReference[_x] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerTimeoutsOutputReference", version: "14.13.1" };
function computeInstanceGroupManagerUpdatePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_surge_fixed: cdktf.numberToTerraform(struct.maxSurgeFixed),
        max_surge_percent: cdktf.numberToTerraform(struct.maxSurgePercent),
        max_unavailable_fixed: cdktf.numberToTerraform(struct.maxUnavailableFixed),
        max_unavailable_percent: cdktf.numberToTerraform(struct.maxUnavailablePercent),
        minimal_action: cdktf.stringToTerraform(struct.minimalAction),
        most_disruptive_allowed_action: cdktf.stringToTerraform(struct.mostDisruptiveAllowedAction),
        replacement_method: cdktf.stringToTerraform(struct.replacementMethod),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.computeInstanceGroupManagerUpdatePolicyToTerraform = computeInstanceGroupManagerUpdatePolicyToTerraform;
function computeInstanceGroupManagerUpdatePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_surge_fixed: {
            value: cdktf.numberToHclTerraform(struct.maxSurgeFixed),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        max_surge_percent: {
            value: cdktf.numberToHclTerraform(struct.maxSurgePercent),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        max_unavailable_fixed: {
            value: cdktf.numberToHclTerraform(struct.maxUnavailableFixed),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        max_unavailable_percent: {
            value: cdktf.numberToHclTerraform(struct.maxUnavailablePercent),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minimal_action: {
            value: cdktf.stringToHclTerraform(struct.minimalAction),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        most_disruptive_allowed_action: {
            value: cdktf.stringToHclTerraform(struct.mostDisruptiveAllowedAction),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        replacement_method: {
            value: cdktf.stringToHclTerraform(struct.replacementMethod),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerUpdatePolicyToHclTerraform = computeInstanceGroupManagerUpdatePolicyToHclTerraform;
class ComputeInstanceGroupManagerUpdatePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxSurgeFixed !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxSurgeFixed = this._maxSurgeFixed;
        }
        if (this._maxSurgePercent !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxSurgePercent = this._maxSurgePercent;
        }
        if (this._maxUnavailableFixed !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxUnavailableFixed = this._maxUnavailableFixed;
        }
        if (this._maxUnavailablePercent !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxUnavailablePercent = this._maxUnavailablePercent;
        }
        if (this._minimalAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.minimalAction = this._minimalAction;
        }
        if (this._mostDisruptiveAllowedAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.mostDisruptiveAllowedAction = this._mostDisruptiveAllowedAction;
        }
        if (this._replacementMethod !== undefined) {
            hasAnyValues = true;
            internalValueResult.replacementMethod = this._replacementMethod;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxSurgeFixed = undefined;
            this._maxSurgePercent = undefined;
            this._maxUnavailableFixed = undefined;
            this._maxUnavailablePercent = undefined;
            this._minimalAction = undefined;
            this._mostDisruptiveAllowedAction = undefined;
            this._replacementMethod = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxSurgeFixed = value.maxSurgeFixed;
            this._maxSurgePercent = value.maxSurgePercent;
            this._maxUnavailableFixed = value.maxUnavailableFixed;
            this._maxUnavailablePercent = value.maxUnavailablePercent;
            this._minimalAction = value.minimalAction;
            this._mostDisruptiveAllowedAction = value.mostDisruptiveAllowedAction;
            this._replacementMethod = value.replacementMethod;
            this._type = value.type;
        }
    }
    get maxSurgeFixed() {
        return this.getNumberAttribute('max_surge_fixed');
    }
    set maxSurgeFixed(value) {
        this._maxSurgeFixed = value;
    }
    resetMaxSurgeFixed() {
        this._maxSurgeFixed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxSurgeFixedInput() {
        return this._maxSurgeFixed;
    }
    get maxSurgePercent() {
        return this.getNumberAttribute('max_surge_percent');
    }
    set maxSurgePercent(value) {
        this._maxSurgePercent = value;
    }
    resetMaxSurgePercent() {
        this._maxSurgePercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxSurgePercentInput() {
        return this._maxSurgePercent;
    }
    get maxUnavailableFixed() {
        return this.getNumberAttribute('max_unavailable_fixed');
    }
    set maxUnavailableFixed(value) {
        this._maxUnavailableFixed = value;
    }
    resetMaxUnavailableFixed() {
        this._maxUnavailableFixed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUnavailableFixedInput() {
        return this._maxUnavailableFixed;
    }
    get maxUnavailablePercent() {
        return this.getNumberAttribute('max_unavailable_percent');
    }
    set maxUnavailablePercent(value) {
        this._maxUnavailablePercent = value;
    }
    resetMaxUnavailablePercent() {
        this._maxUnavailablePercent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUnavailablePercentInput() {
        return this._maxUnavailablePercent;
    }
    get minimalAction() {
        return this.getStringAttribute('minimal_action');
    }
    set minimalAction(value) {
        this._minimalAction = value;
    }
    // Temporarily expose input value. Use with caution.
    get minimalActionInput() {
        return this._minimalAction;
    }
    get mostDisruptiveAllowedAction() {
        return this.getStringAttribute('most_disruptive_allowed_action');
    }
    set mostDisruptiveAllowedAction(value) {
        this._mostDisruptiveAllowedAction = value;
    }
    resetMostDisruptiveAllowedAction() {
        this._mostDisruptiveAllowedAction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mostDisruptiveAllowedActionInput() {
        return this._mostDisruptiveAllowedAction;
    }
    get replacementMethod() {
        return this.getStringAttribute('replacement_method');
    }
    set replacementMethod(value) {
        this._replacementMethod = value;
    }
    resetReplacementMethod() {
        this._replacementMethod = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replacementMethodInput() {
        return this._replacementMethod;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComputeInstanceGroupManagerUpdatePolicyOutputReference = ComputeInstanceGroupManagerUpdatePolicyOutputReference;
_y = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerUpdatePolicyOutputReference[_y] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerUpdatePolicyOutputReference", version: "14.13.1" };
function computeInstanceGroupManagerVersionTargetSizeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        fixed: cdktf.numberToTerraform(struct.fixed),
        percent: cdktf.numberToTerraform(struct.percent),
    };
}
exports.computeInstanceGroupManagerVersionTargetSizeToTerraform = computeInstanceGroupManagerVersionTargetSizeToTerraform;
function computeInstanceGroupManagerVersionTargetSizeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        fixed: {
            value: cdktf.numberToHclTerraform(struct.fixed),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        percent: {
            value: cdktf.numberToHclTerraform(struct.percent),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerVersionTargetSizeToHclTerraform = computeInstanceGroupManagerVersionTargetSizeToHclTerraform;
class ComputeInstanceGroupManagerVersionTargetSizeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._fixed !== undefined) {
            hasAnyValues = true;
            internalValueResult.fixed = this._fixed;
        }
        if (this._percent !== undefined) {
            hasAnyValues = true;
            internalValueResult.percent = this._percent;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._fixed = undefined;
            this._percent = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._fixed = value.fixed;
            this._percent = value.percent;
        }
    }
    get fixed() {
        return this.getNumberAttribute('fixed');
    }
    set fixed(value) {
        this._fixed = value;
    }
    resetFixed() {
        this._fixed = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fixedInput() {
        return this._fixed;
    }
    get percent() {
        return this.getNumberAttribute('percent');
    }
    set percent(value) {
        this._percent = value;
    }
    resetPercent() {
        this._percent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get percentInput() {
        return this._percent;
    }
}
exports.ComputeInstanceGroupManagerVersionTargetSizeOutputReference = ComputeInstanceGroupManagerVersionTargetSizeOutputReference;
_z = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerVersionTargetSizeOutputReference[_z] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerVersionTargetSizeOutputReference", version: "14.13.1" };
function computeInstanceGroupManagerVersionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        instance_template: cdktf.stringToTerraform(struct.instanceTemplate),
        name: cdktf.stringToTerraform(struct.name),
        target_size: computeInstanceGroupManagerVersionTargetSizeToTerraform(struct.targetSize),
    };
}
exports.computeInstanceGroupManagerVersionToTerraform = computeInstanceGroupManagerVersionToTerraform;
function computeInstanceGroupManagerVersionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        instance_template: {
            value: cdktf.stringToHclTerraform(struct.instanceTemplate),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        target_size: {
            value: computeInstanceGroupManagerVersionTargetSizeToHclTerraform(struct.targetSize),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeInstanceGroupManagerVersionTargetSizeList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeInstanceGroupManagerVersionToHclTerraform = computeInstanceGroupManagerVersionToHclTerraform;
class ComputeInstanceGroupManagerVersionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // target_size - computed: false, optional: true, required: false
        this._targetSize = new ComputeInstanceGroupManagerVersionTargetSizeOutputReference(this, "target_size");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._instanceTemplate !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceTemplate = this._instanceTemplate;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._targetSize?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetSize = this._targetSize?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._instanceTemplate = undefined;
            this._name = undefined;
            this._targetSize.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._instanceTemplate = value.instanceTemplate;
            this._name = value.name;
            this._targetSize.internalValue = value.targetSize;
        }
    }
    get instanceTemplate() {
        return this.getStringAttribute('instance_template');
    }
    set instanceTemplate(value) {
        this._instanceTemplate = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceTemplateInput() {
        return this._instanceTemplate;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get targetSize() {
        return this._targetSize;
    }
    putTargetSize(value) {
        this._targetSize.internalValue = value;
    }
    resetTargetSize() {
        this._targetSize.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetSizeInput() {
        return this._targetSize.internalValue;
    }
}
exports.ComputeInstanceGroupManagerVersionOutputReference = ComputeInstanceGroupManagerVersionOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerVersionOutputReference[_0] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerVersionOutputReference", version: "14.13.1" };
class ComputeInstanceGroupManagerVersionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceGroupManagerVersionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceGroupManagerVersionList = ComputeInstanceGroupManagerVersionList;
_1 = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManagerVersionList[_1] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManagerVersionList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_instance_group_manager google_compute_instance_group_manager}
*/
class ComputeInstanceGroupManager extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ComputeInstanceGroupManager resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ComputeInstanceGroupManager to import
    * @param importFromId The id of the existing ComputeInstanceGroupManager that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_instance_group_manager#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ComputeInstanceGroupManager to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_instance_group_manager", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_instance_group_manager google_compute_instance_group_manager} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeInstanceGroupManagerConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_instance_group_manager',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // status - computed: true, optional: false, required: false
        this._status = new ComputeInstanceGroupManagerStatusList(this, "status", false);
        // all_instances_config - computed: false, optional: true, required: false
        this._allInstancesConfig = new ComputeInstanceGroupManagerAllInstancesConfigOutputReference(this, "all_instances_config");
        // auto_healing_policies - computed: false, optional: true, required: false
        this._autoHealingPolicies = new ComputeInstanceGroupManagerAutoHealingPoliciesOutputReference(this, "auto_healing_policies");
        // instance_lifecycle_policy - computed: false, optional: true, required: false
        this._instanceLifecyclePolicy = new ComputeInstanceGroupManagerInstanceLifecyclePolicyOutputReference(this, "instance_lifecycle_policy");
        // named_port - computed: false, optional: true, required: false
        this._namedPort = new ComputeInstanceGroupManagerNamedPortList(this, "named_port", true);
        // stateful_disk - computed: false, optional: true, required: false
        this._statefulDisk = new ComputeInstanceGroupManagerStatefulDiskList(this, "stateful_disk", true);
        // stateful_external_ip - computed: false, optional: true, required: false
        this._statefulExternalIp = new ComputeInstanceGroupManagerStatefulExternalIpList(this, "stateful_external_ip", false);
        // stateful_internal_ip - computed: false, optional: true, required: false
        this._statefulInternalIp = new ComputeInstanceGroupManagerStatefulInternalIpList(this, "stateful_internal_ip", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeInstanceGroupManagerTimeoutsOutputReference(this, "timeouts");
        // update_policy - computed: false, optional: true, required: false
        this._updatePolicy = new ComputeInstanceGroupManagerUpdatePolicyOutputReference(this, "update_policy");
        // version - computed: false, optional: false, required: true
        this._version = new ComputeInstanceGroupManagerVersionList(this, "version", false);
        this._baseInstanceName = config.baseInstanceName;
        this._description = config.description;
        this._id = config.id;
        this._listManagedInstancesResults = config.listManagedInstancesResults;
        this._name = config.name;
        this._project = config.project;
        this._targetPools = config.targetPools;
        this._targetSize = config.targetSize;
        this._waitForInstances = config.waitForInstances;
        this._waitForInstancesStatus = config.waitForInstancesStatus;
        this._zone = config.zone;
        this._allInstancesConfig.internalValue = config.allInstancesConfig;
        this._autoHealingPolicies.internalValue = config.autoHealingPolicies;
        this._instanceLifecyclePolicy.internalValue = config.instanceLifecyclePolicy;
        this._namedPort.internalValue = config.namedPort;
        this._statefulDisk.internalValue = config.statefulDisk;
        this._statefulExternalIp.internalValue = config.statefulExternalIp;
        this._statefulInternalIp.internalValue = config.statefulInternalIp;
        this._timeouts.internalValue = config.timeouts;
        this._updatePolicy.internalValue = config.updatePolicy;
        this._version.internalValue = config.version;
    }
    get baseInstanceName() {
        return this.getStringAttribute('base_instance_name');
    }
    set baseInstanceName(value) {
        this._baseInstanceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get baseInstanceNameInput() {
        return this._baseInstanceName;
    }
    // creation_timestamp - computed: true, optional: false, required: false
    get creationTimestamp() {
        return this.getStringAttribute('creation_timestamp');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // fingerprint - computed: true, optional: false, required: false
    get fingerprint() {
        return this.getStringAttribute('fingerprint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // instance_group - computed: true, optional: false, required: false
    get instanceGroup() {
        return this.getStringAttribute('instance_group');
    }
    // instance_group_manager_id - computed: true, optional: false, required: false
    get instanceGroupManagerId() {
        return this.getNumberAttribute('instance_group_manager_id');
    }
    get listManagedInstancesResults() {
        return this.getStringAttribute('list_managed_instances_results');
    }
    set listManagedInstancesResults(value) {
        this._listManagedInstancesResults = value;
    }
    resetListManagedInstancesResults() {
        this._listManagedInstancesResults = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get listManagedInstancesResultsInput() {
        return this._listManagedInstancesResults;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // operation - computed: true, optional: false, required: false
    get operation() {
        return this.getStringAttribute('operation');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get status() {
        return this._status;
    }
    get targetPools() {
        return cdktf.Fn.tolist(this.getListAttribute('target_pools'));
    }
    set targetPools(value) {
        this._targetPools = value;
    }
    resetTargetPools() {
        this._targetPools = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetPoolsInput() {
        return this._targetPools;
    }
    get targetSize() {
        return this.getNumberAttribute('target_size');
    }
    set targetSize(value) {
        this._targetSize = value;
    }
    resetTargetSize() {
        this._targetSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetSizeInput() {
        return this._targetSize;
    }
    get waitForInstances() {
        return this.getBooleanAttribute('wait_for_instances');
    }
    set waitForInstances(value) {
        this._waitForInstances = value;
    }
    resetWaitForInstances() {
        this._waitForInstances = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForInstancesInput() {
        return this._waitForInstances;
    }
    get waitForInstancesStatus() {
        return this.getStringAttribute('wait_for_instances_status');
    }
    set waitForInstancesStatus(value) {
        this._waitForInstancesStatus = value;
    }
    resetWaitForInstancesStatus() {
        this._waitForInstancesStatus = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForInstancesStatusInput() {
        return this._waitForInstancesStatus;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    get allInstancesConfig() {
        return this._allInstancesConfig;
    }
    putAllInstancesConfig(value) {
        this._allInstancesConfig.internalValue = value;
    }
    resetAllInstancesConfig() {
        this._allInstancesConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allInstancesConfigInput() {
        return this._allInstancesConfig.internalValue;
    }
    get autoHealingPolicies() {
        return this._autoHealingPolicies;
    }
    putAutoHealingPolicies(value) {
        this._autoHealingPolicies.internalValue = value;
    }
    resetAutoHealingPolicies() {
        this._autoHealingPolicies.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoHealingPoliciesInput() {
        return this._autoHealingPolicies.internalValue;
    }
    get instanceLifecyclePolicy() {
        return this._instanceLifecyclePolicy;
    }
    putInstanceLifecyclePolicy(value) {
        this._instanceLifecyclePolicy.internalValue = value;
    }
    resetInstanceLifecyclePolicy() {
        this._instanceLifecyclePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get instanceLifecyclePolicyInput() {
        return this._instanceLifecyclePolicy.internalValue;
    }
    get namedPort() {
        return this._namedPort;
    }
    putNamedPort(value) {
        this._namedPort.internalValue = value;
    }
    resetNamedPort() {
        this._namedPort.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namedPortInput() {
        return this._namedPort.internalValue;
    }
    get statefulDisk() {
        return this._statefulDisk;
    }
    putStatefulDisk(value) {
        this._statefulDisk.internalValue = value;
    }
    resetStatefulDisk() {
        this._statefulDisk.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statefulDiskInput() {
        return this._statefulDisk.internalValue;
    }
    get statefulExternalIp() {
        return this._statefulExternalIp;
    }
    putStatefulExternalIp(value) {
        this._statefulExternalIp.internalValue = value;
    }
    resetStatefulExternalIp() {
        this._statefulExternalIp.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statefulExternalIpInput() {
        return this._statefulExternalIp.internalValue;
    }
    get statefulInternalIp() {
        return this._statefulInternalIp;
    }
    putStatefulInternalIp(value) {
        this._statefulInternalIp.internalValue = value;
    }
    resetStatefulInternalIp() {
        this._statefulInternalIp.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statefulInternalIpInput() {
        return this._statefulInternalIp.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get updatePolicy() {
        return this._updatePolicy;
    }
    putUpdatePolicy(value) {
        this._updatePolicy.internalValue = value;
    }
    resetUpdatePolicy() {
        this._updatePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updatePolicyInput() {
        return this._updatePolicy.internalValue;
    }
    get version() {
        return this._version;
    }
    putVersion(value) {
        this._version.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            base_instance_name: cdktf.stringToTerraform(this._baseInstanceName),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            list_managed_instances_results: cdktf.stringToTerraform(this._listManagedInstancesResults),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            target_pools: cdktf.listMapper(cdktf.stringToTerraform, false)(this._targetPools),
            target_size: cdktf.numberToTerraform(this._targetSize),
            wait_for_instances: cdktf.booleanToTerraform(this._waitForInstances),
            wait_for_instances_status: cdktf.stringToTerraform(this._waitForInstancesStatus),
            zone: cdktf.stringToTerraform(this._zone),
            all_instances_config: computeInstanceGroupManagerAllInstancesConfigToTerraform(this._allInstancesConfig.internalValue),
            auto_healing_policies: computeInstanceGroupManagerAutoHealingPoliciesToTerraform(this._autoHealingPolicies.internalValue),
            instance_lifecycle_policy: computeInstanceGroupManagerInstanceLifecyclePolicyToTerraform(this._instanceLifecyclePolicy.internalValue),
            named_port: cdktf.listMapper(computeInstanceGroupManagerNamedPortToTerraform, true)(this._namedPort.internalValue),
            stateful_disk: cdktf.listMapper(computeInstanceGroupManagerStatefulDiskToTerraform, true)(this._statefulDisk.internalValue),
            stateful_external_ip: cdktf.listMapper(computeInstanceGroupManagerStatefulExternalIpToTerraform, true)(this._statefulExternalIp.internalValue),
            stateful_internal_ip: cdktf.listMapper(computeInstanceGroupManagerStatefulInternalIpToTerraform, true)(this._statefulInternalIp.internalValue),
            timeouts: computeInstanceGroupManagerTimeoutsToTerraform(this._timeouts.internalValue),
            update_policy: computeInstanceGroupManagerUpdatePolicyToTerraform(this._updatePolicy.internalValue),
            version: cdktf.listMapper(computeInstanceGroupManagerVersionToTerraform, true)(this._version.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            base_instance_name: {
                value: cdktf.stringToHclTerraform(this._baseInstanceName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            list_managed_instances_results: {
                value: cdktf.stringToHclTerraform(this._listManagedInstancesResults),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            target_pools: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._targetPools),
                isBlock: false,
                type: "set",
                storageClassType: "stringList",
            },
            target_size: {
                value: cdktf.numberToHclTerraform(this._targetSize),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            wait_for_instances: {
                value: cdktf.booleanToHclTerraform(this._waitForInstances),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            wait_for_instances_status: {
                value: cdktf.stringToHclTerraform(this._waitForInstancesStatus),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            zone: {
                value: cdktf.stringToHclTerraform(this._zone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            all_instances_config: {
                value: computeInstanceGroupManagerAllInstancesConfigToHclTerraform(this._allInstancesConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeInstanceGroupManagerAllInstancesConfigList",
            },
            auto_healing_policies: {
                value: computeInstanceGroupManagerAutoHealingPoliciesToHclTerraform(this._autoHealingPolicies.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeInstanceGroupManagerAutoHealingPoliciesList",
            },
            instance_lifecycle_policy: {
                value: computeInstanceGroupManagerInstanceLifecyclePolicyToHclTerraform(this._instanceLifecyclePolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeInstanceGroupManagerInstanceLifecyclePolicyList",
            },
            named_port: {
                value: cdktf.listMapperHcl(computeInstanceGroupManagerNamedPortToHclTerraform, true)(this._namedPort.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "ComputeInstanceGroupManagerNamedPortList",
            },
            stateful_disk: {
                value: cdktf.listMapperHcl(computeInstanceGroupManagerStatefulDiskToHclTerraform, true)(this._statefulDisk.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "ComputeInstanceGroupManagerStatefulDiskList",
            },
            stateful_external_ip: {
                value: cdktf.listMapperHcl(computeInstanceGroupManagerStatefulExternalIpToHclTerraform, true)(this._statefulExternalIp.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeInstanceGroupManagerStatefulExternalIpList",
            },
            stateful_internal_ip: {
                value: cdktf.listMapperHcl(computeInstanceGroupManagerStatefulInternalIpToHclTerraform, true)(this._statefulInternalIp.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeInstanceGroupManagerStatefulInternalIpList",
            },
            timeouts: {
                value: computeInstanceGroupManagerTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ComputeInstanceGroupManagerTimeouts",
            },
            update_policy: {
                value: computeInstanceGroupManagerUpdatePolicyToHclTerraform(this._updatePolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeInstanceGroupManagerUpdatePolicyList",
            },
            version: {
                value: cdktf.listMapperHcl(computeInstanceGroupManagerVersionToHclTerraform, true)(this._version.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeInstanceGroupManagerVersionList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ComputeInstanceGroupManager = ComputeInstanceGroupManager;
_2 = JSII_RTTI_SYMBOL_1;
ComputeInstanceGroupManager[_2] = { fqn: "@cdktf/provider-google.computeInstanceGroupManager.ComputeInstanceGroupManager", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ComputeInstanceGroupManager.tfResourceType = "google_compute_instance_group_manager";
