// @cdktf/provider-google/lib/vertex-ai-endpoint/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VertexAiEndpoint = exports.VertexAiEndpointTimeoutsOutputReference = exports.vertexAiEndpointTimeoutsToHclTerraform = exports.vertexAiEndpointTimeoutsToTerraform = exports.VertexAiEndpointPrivateServiceConnectConfigOutputReference = exports.vertexAiEndpointPrivateServiceConnectConfigToHclTerraform = exports.vertexAiEndpointPrivateServiceConnectConfigToTerraform = exports.VertexAiEndpointPredictRequestResponseLoggingConfigOutputReference = exports.vertexAiEndpointPredictRequestResponseLoggingConfigToHclTerraform = exports.vertexAiEndpointPredictRequestResponseLoggingConfigToTerraform = exports.VertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationOutputReference = exports.vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToHclTerraform = exports.vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToTerraform = exports.VertexAiEndpointEncryptionSpecOutputReference = exports.vertexAiEndpointEncryptionSpecToHclTerraform = exports.vertexAiEndpointEncryptionSpecToTerraform = exports.VertexAiEndpointDeployedModelsList = exports.VertexAiEndpointDeployedModelsOutputReference = exports.vertexAiEndpointDeployedModelsToHclTerraform = exports.vertexAiEndpointDeployedModelsToTerraform = exports.VertexAiEndpointDeployedModelsPrivateEndpointsList = exports.VertexAiEndpointDeployedModelsPrivateEndpointsOutputReference = exports.vertexAiEndpointDeployedModelsPrivateEndpointsToHclTerraform = exports.vertexAiEndpointDeployedModelsPrivateEndpointsToTerraform = exports.VertexAiEndpointDeployedModelsDedicatedResourcesList = exports.VertexAiEndpointDeployedModelsDedicatedResourcesOutputReference = exports.vertexAiEndpointDeployedModelsDedicatedResourcesToHclTerraform = exports.vertexAiEndpointDeployedModelsDedicatedResourcesToTerraform = exports.VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecList = exports.VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecOutputReference = exports.vertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecToHclTerraform = exports.vertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecToTerraform = exports.VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsList = exports.VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsOutputReference = exports.vertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsToHclTerraform = exports.vertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsToTerraform = exports.VertexAiEndpointDeployedModelsAutomaticResourcesList = exports.VertexAiEndpointDeployedModelsAutomaticResourcesOutputReference = exports.vertexAiEndpointDeployedModelsAutomaticResourcesToHclTerraform = exports.vertexAiEndpointDeployedModelsAutomaticResourcesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vertexAiEndpointDeployedModelsAutomaticResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.vertexAiEndpointDeployedModelsAutomaticResourcesToTerraform = vertexAiEndpointDeployedModelsAutomaticResourcesToTerraform;
function vertexAiEndpointDeployedModelsAutomaticResourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.vertexAiEndpointDeployedModelsAutomaticResourcesToHclTerraform = vertexAiEndpointDeployedModelsAutomaticResourcesToHclTerraform;
class VertexAiEndpointDeployedModelsAutomaticResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_replica_count - computed: true, optional: false, required: false
    get maxReplicaCount() {
        return this.getNumberAttribute('max_replica_count');
    }
    // min_replica_count - computed: true, optional: false, required: false
    get minReplicaCount() {
        return this.getNumberAttribute('min_replica_count');
    }
}
exports.VertexAiEndpointDeployedModelsAutomaticResourcesOutputReference = VertexAiEndpointDeployedModelsAutomaticResourcesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsAutomaticResourcesOutputReference[_a] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsAutomaticResourcesOutputReference", version: "14.13.1" };
class VertexAiEndpointDeployedModelsAutomaticResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VertexAiEndpointDeployedModelsAutomaticResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VertexAiEndpointDeployedModelsAutomaticResourcesList = VertexAiEndpointDeployedModelsAutomaticResourcesList;
_b = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsAutomaticResourcesList[_b] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsAutomaticResourcesList", version: "14.13.1" };
function vertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.vertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsToTerraform = vertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsToTerraform;
function vertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.vertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsToHclTerraform = vertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsToHclTerraform;
class VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // metric_name - computed: true, optional: false, required: false
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    // target - computed: true, optional: false, required: false
    get target() {
        return this.getNumberAttribute('target');
    }
}
exports.VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsOutputReference = VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsOutputReference[_c] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsOutputReference", version: "14.13.1" };
class VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsList = VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsList;
_d = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsList[_d] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsList", version: "14.13.1" };
function vertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.vertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecToTerraform = vertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecToTerraform;
function vertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.vertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecToHclTerraform = vertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecToHclTerraform;
class VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // accelerator_count - computed: true, optional: false, required: false
    get acceleratorCount() {
        return this.getNumberAttribute('accelerator_count');
    }
    // accelerator_type - computed: true, optional: false, required: false
    get acceleratorType() {
        return this.getStringAttribute('accelerator_type');
    }
    // machine_type - computed: true, optional: false, required: false
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
}
exports.VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecOutputReference = VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecOutputReference;
_e = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecOutputReference[_e] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecOutputReference", version: "14.13.1" };
class VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecList = VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecList;
_f = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecList[_f] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecList", version: "14.13.1" };
function vertexAiEndpointDeployedModelsDedicatedResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.vertexAiEndpointDeployedModelsDedicatedResourcesToTerraform = vertexAiEndpointDeployedModelsDedicatedResourcesToTerraform;
function vertexAiEndpointDeployedModelsDedicatedResourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.vertexAiEndpointDeployedModelsDedicatedResourcesToHclTerraform = vertexAiEndpointDeployedModelsDedicatedResourcesToHclTerraform;
class VertexAiEndpointDeployedModelsDedicatedResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // autoscaling_metric_specs - computed: true, optional: false, required: false
        this._autoscalingMetricSpecs = new VertexAiEndpointDeployedModelsDedicatedResourcesAutoscalingMetricSpecsList(this, "autoscaling_metric_specs", false);
        // machine_spec - computed: true, optional: false, required: false
        this._machineSpec = new VertexAiEndpointDeployedModelsDedicatedResourcesMachineSpecList(this, "machine_spec", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get autoscalingMetricSpecs() {
        return this._autoscalingMetricSpecs;
    }
    get machineSpec() {
        return this._machineSpec;
    }
    // max_replica_count - computed: true, optional: false, required: false
    get maxReplicaCount() {
        return this.getNumberAttribute('max_replica_count');
    }
    // min_replica_count - computed: true, optional: false, required: false
    get minReplicaCount() {
        return this.getNumberAttribute('min_replica_count');
    }
}
exports.VertexAiEndpointDeployedModelsDedicatedResourcesOutputReference = VertexAiEndpointDeployedModelsDedicatedResourcesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsDedicatedResourcesOutputReference[_g] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsDedicatedResourcesOutputReference", version: "14.13.1" };
class VertexAiEndpointDeployedModelsDedicatedResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VertexAiEndpointDeployedModelsDedicatedResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VertexAiEndpointDeployedModelsDedicatedResourcesList = VertexAiEndpointDeployedModelsDedicatedResourcesList;
_h = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsDedicatedResourcesList[_h] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsDedicatedResourcesList", version: "14.13.1" };
function vertexAiEndpointDeployedModelsPrivateEndpointsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.vertexAiEndpointDeployedModelsPrivateEndpointsToTerraform = vertexAiEndpointDeployedModelsPrivateEndpointsToTerraform;
function vertexAiEndpointDeployedModelsPrivateEndpointsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.vertexAiEndpointDeployedModelsPrivateEndpointsToHclTerraform = vertexAiEndpointDeployedModelsPrivateEndpointsToHclTerraform;
class VertexAiEndpointDeployedModelsPrivateEndpointsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // explain_http_uri - computed: true, optional: false, required: false
    get explainHttpUri() {
        return this.getStringAttribute('explain_http_uri');
    }
    // health_http_uri - computed: true, optional: false, required: false
    get healthHttpUri() {
        return this.getStringAttribute('health_http_uri');
    }
    // predict_http_uri - computed: true, optional: false, required: false
    get predictHttpUri() {
        return this.getStringAttribute('predict_http_uri');
    }
    // service_attachment - computed: true, optional: false, required: false
    get serviceAttachment() {
        return this.getStringAttribute('service_attachment');
    }
}
exports.VertexAiEndpointDeployedModelsPrivateEndpointsOutputReference = VertexAiEndpointDeployedModelsPrivateEndpointsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsPrivateEndpointsOutputReference[_j] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsPrivateEndpointsOutputReference", version: "14.13.1" };
class VertexAiEndpointDeployedModelsPrivateEndpointsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VertexAiEndpointDeployedModelsPrivateEndpointsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VertexAiEndpointDeployedModelsPrivateEndpointsList = VertexAiEndpointDeployedModelsPrivateEndpointsList;
_k = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsPrivateEndpointsList[_k] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsPrivateEndpointsList", version: "14.13.1" };
function vertexAiEndpointDeployedModelsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.vertexAiEndpointDeployedModelsToTerraform = vertexAiEndpointDeployedModelsToTerraform;
function vertexAiEndpointDeployedModelsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.vertexAiEndpointDeployedModelsToHclTerraform = vertexAiEndpointDeployedModelsToHclTerraform;
class VertexAiEndpointDeployedModelsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // automatic_resources - computed: true, optional: false, required: false
        this._automaticResources = new VertexAiEndpointDeployedModelsAutomaticResourcesList(this, "automatic_resources", false);
        // dedicated_resources - computed: true, optional: false, required: false
        this._dedicatedResources = new VertexAiEndpointDeployedModelsDedicatedResourcesList(this, "dedicated_resources", false);
        // private_endpoints - computed: true, optional: false, required: false
        this._privateEndpoints = new VertexAiEndpointDeployedModelsPrivateEndpointsList(this, "private_endpoints", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get automaticResources() {
        return this._automaticResources;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get dedicatedResources() {
        return this._dedicatedResources;
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    // enable_access_logging - computed: true, optional: false, required: false
    get enableAccessLogging() {
        return this.getBooleanAttribute('enable_access_logging');
    }
    // enable_container_logging - computed: true, optional: false, required: false
    get enableContainerLogging() {
        return this.getBooleanAttribute('enable_container_logging');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // model - computed: true, optional: false, required: false
    get model() {
        return this.getStringAttribute('model');
    }
    // model_version_id - computed: true, optional: false, required: false
    get modelVersionId() {
        return this.getStringAttribute('model_version_id');
    }
    get privateEndpoints() {
        return this._privateEndpoints;
    }
    // service_account - computed: true, optional: false, required: false
    get serviceAccount() {
        return this.getStringAttribute('service_account');
    }
    // shared_resources - computed: true, optional: false, required: false
    get sharedResources() {
        return this.getStringAttribute('shared_resources');
    }
}
exports.VertexAiEndpointDeployedModelsOutputReference = VertexAiEndpointDeployedModelsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsOutputReference[_l] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsOutputReference", version: "14.13.1" };
class VertexAiEndpointDeployedModelsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VertexAiEndpointDeployedModelsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VertexAiEndpointDeployedModelsList = VertexAiEndpointDeployedModelsList;
_m = JSII_RTTI_SYMBOL_1;
VertexAiEndpointDeployedModelsList[_m] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointDeployedModelsList", version: "14.13.1" };
function vertexAiEndpointEncryptionSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key_name: cdktf.stringToTerraform(struct.kmsKeyName),
    };
}
exports.vertexAiEndpointEncryptionSpecToTerraform = vertexAiEndpointEncryptionSpecToTerraform;
function vertexAiEndpointEncryptionSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key_name: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiEndpointEncryptionSpecToHclTerraform = vertexAiEndpointEncryptionSpecToHclTerraform;
class VertexAiEndpointEncryptionSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyName = this._kmsKeyName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKeyName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKeyName = value.kmsKeyName;
        }
    }
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    set kmsKeyName(value) {
        this._kmsKeyName = value;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyNameInput() {
        return this._kmsKeyName;
    }
}
exports.VertexAiEndpointEncryptionSpecOutputReference = VertexAiEndpointEncryptionSpecOutputReference;
_o = JSII_RTTI_SYMBOL_1;
VertexAiEndpointEncryptionSpecOutputReference[_o] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointEncryptionSpecOutputReference", version: "14.13.1" };
function vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        output_uri: cdktf.stringToTerraform(struct.outputUri),
    };
}
exports.vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToTerraform = vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToTerraform;
function vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        output_uri: {
            value: cdktf.stringToHclTerraform(struct.outputUri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToHclTerraform = vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToHclTerraform;
class VertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._outputUri !== undefined) {
            hasAnyValues = true;
            internalValueResult.outputUri = this._outputUri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._outputUri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._outputUri = value.outputUri;
        }
    }
    get outputUri() {
        return this.getStringAttribute('output_uri');
    }
    set outputUri(value) {
        this._outputUri = value;
    }
    resetOutputUri() {
        this._outputUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputUriInput() {
        return this._outputUri;
    }
}
exports.VertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationOutputReference = VertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationOutputReference;
_p = JSII_RTTI_SYMBOL_1;
VertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationOutputReference[_p] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationOutputReference", version: "14.13.1" };
function vertexAiEndpointPredictRequestResponseLoggingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        sampling_rate: cdktf.numberToTerraform(struct.samplingRate),
        bigquery_destination: vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToTerraform(struct.bigqueryDestination),
    };
}
exports.vertexAiEndpointPredictRequestResponseLoggingConfigToTerraform = vertexAiEndpointPredictRequestResponseLoggingConfigToTerraform;
function vertexAiEndpointPredictRequestResponseLoggingConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        sampling_rate: {
            value: cdktf.numberToHclTerraform(struct.samplingRate),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        bigquery_destination: {
            value: vertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationToHclTerraform(struct.bigqueryDestination),
            isBlock: true,
            type: "list",
            storageClassType: "VertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiEndpointPredictRequestResponseLoggingConfigToHclTerraform = vertexAiEndpointPredictRequestResponseLoggingConfigToHclTerraform;
class VertexAiEndpointPredictRequestResponseLoggingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // bigquery_destination - computed: false, optional: true, required: false
        this._bigqueryDestination = new VertexAiEndpointPredictRequestResponseLoggingConfigBigqueryDestinationOutputReference(this, "bigquery_destination");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._samplingRate !== undefined) {
            hasAnyValues = true;
            internalValueResult.samplingRate = this._samplingRate;
        }
        if (this._bigqueryDestination?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.bigqueryDestination = this._bigqueryDestination?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._samplingRate = undefined;
            this._bigqueryDestination.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._samplingRate = value.samplingRate;
            this._bigqueryDestination.internalValue = value.bigqueryDestination;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get samplingRate() {
        return this.getNumberAttribute('sampling_rate');
    }
    set samplingRate(value) {
        this._samplingRate = value;
    }
    resetSamplingRate() {
        this._samplingRate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get samplingRateInput() {
        return this._samplingRate;
    }
    get bigqueryDestination() {
        return this._bigqueryDestination;
    }
    putBigqueryDestination(value) {
        this._bigqueryDestination.internalValue = value;
    }
    resetBigqueryDestination() {
        this._bigqueryDestination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bigqueryDestinationInput() {
        return this._bigqueryDestination.internalValue;
    }
}
exports.VertexAiEndpointPredictRequestResponseLoggingConfigOutputReference = VertexAiEndpointPredictRequestResponseLoggingConfigOutputReference;
_q = JSII_RTTI_SYMBOL_1;
VertexAiEndpointPredictRequestResponseLoggingConfigOutputReference[_q] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointPredictRequestResponseLoggingConfigOutputReference", version: "14.13.1" };
function vertexAiEndpointPrivateServiceConnectConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_private_service_connect: cdktf.booleanToTerraform(struct.enablePrivateServiceConnect),
        enable_secure_private_service_connect: cdktf.booleanToTerraform(struct.enableSecurePrivateServiceConnect),
        project_allowlist: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.projectAllowlist),
    };
}
exports.vertexAiEndpointPrivateServiceConnectConfigToTerraform = vertexAiEndpointPrivateServiceConnectConfigToTerraform;
function vertexAiEndpointPrivateServiceConnectConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_private_service_connect: {
            value: cdktf.booleanToHclTerraform(struct.enablePrivateServiceConnect),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_secure_private_service_connect: {
            value: cdktf.booleanToHclTerraform(struct.enableSecurePrivateServiceConnect),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        project_allowlist: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.projectAllowlist),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiEndpointPrivateServiceConnectConfigToHclTerraform = vertexAiEndpointPrivateServiceConnectConfigToHclTerraform;
class VertexAiEndpointPrivateServiceConnectConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enablePrivateServiceConnect !== undefined) {
            hasAnyValues = true;
            internalValueResult.enablePrivateServiceConnect = this._enablePrivateServiceConnect;
        }
        if (this._enableSecurePrivateServiceConnect !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableSecurePrivateServiceConnect = this._enableSecurePrivateServiceConnect;
        }
        if (this._projectAllowlist !== undefined) {
            hasAnyValues = true;
            internalValueResult.projectAllowlist = this._projectAllowlist;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enablePrivateServiceConnect = undefined;
            this._enableSecurePrivateServiceConnect = undefined;
            this._projectAllowlist = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enablePrivateServiceConnect = value.enablePrivateServiceConnect;
            this._enableSecurePrivateServiceConnect = value.enableSecurePrivateServiceConnect;
            this._projectAllowlist = value.projectAllowlist;
        }
    }
    get enablePrivateServiceConnect() {
        return this.getBooleanAttribute('enable_private_service_connect');
    }
    set enablePrivateServiceConnect(value) {
        this._enablePrivateServiceConnect = value;
    }
    // Temporarily expose input value. Use with caution.
    get enablePrivateServiceConnectInput() {
        return this._enablePrivateServiceConnect;
    }
    get enableSecurePrivateServiceConnect() {
        return this.getBooleanAttribute('enable_secure_private_service_connect');
    }
    set enableSecurePrivateServiceConnect(value) {
        this._enableSecurePrivateServiceConnect = value;
    }
    resetEnableSecurePrivateServiceConnect() {
        this._enableSecurePrivateServiceConnect = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableSecurePrivateServiceConnectInput() {
        return this._enableSecurePrivateServiceConnect;
    }
    get projectAllowlist() {
        return this.getListAttribute('project_allowlist');
    }
    set projectAllowlist(value) {
        this._projectAllowlist = value;
    }
    resetProjectAllowlist() {
        this._projectAllowlist = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectAllowlistInput() {
        return this._projectAllowlist;
    }
}
exports.VertexAiEndpointPrivateServiceConnectConfigOutputReference = VertexAiEndpointPrivateServiceConnectConfigOutputReference;
_r = JSII_RTTI_SYMBOL_1;
VertexAiEndpointPrivateServiceConnectConfigOutputReference[_r] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointPrivateServiceConnectConfigOutputReference", version: "14.13.1" };
function vertexAiEndpointTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.vertexAiEndpointTimeoutsToTerraform = vertexAiEndpointTimeoutsToTerraform;
function vertexAiEndpointTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiEndpointTimeoutsToHclTerraform = vertexAiEndpointTimeoutsToHclTerraform;
class VertexAiEndpointTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VertexAiEndpointTimeoutsOutputReference = VertexAiEndpointTimeoutsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
VertexAiEndpointTimeoutsOutputReference[_s] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpointTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_endpoint google_vertex_ai_endpoint}
*/
class VertexAiEndpoint extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VertexAiEndpoint resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VertexAiEndpoint to import
    * @param importFromId The id of the existing VertexAiEndpoint that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_endpoint#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VertexAiEndpoint to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_vertex_ai_endpoint", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_endpoint google_vertex_ai_endpoint} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VertexAiEndpointConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_vertex_ai_endpoint',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // deployed_models - computed: true, optional: false, required: false
        this._deployedModels = new VertexAiEndpointDeployedModelsList(this, "deployed_models", false);
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // encryption_spec - computed: false, optional: true, required: false
        this._encryptionSpec = new VertexAiEndpointEncryptionSpecOutputReference(this, "encryption_spec");
        // predict_request_response_logging_config - computed: false, optional: true, required: false
        this._predictRequestResponseLoggingConfig = new VertexAiEndpointPredictRequestResponseLoggingConfigOutputReference(this, "predict_request_response_logging_config");
        // private_service_connect_config - computed: false, optional: true, required: false
        this._privateServiceConnectConfig = new VertexAiEndpointPrivateServiceConnectConfigOutputReference(this, "private_service_connect_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VertexAiEndpointTimeoutsOutputReference(this, "timeouts");
        this._dedicatedEndpointEnabled = config.dedicatedEndpointEnabled;
        this._description = config.description;
        this._displayName = config.displayName;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._network = config.network;
        this._project = config.project;
        this._region = config.region;
        this._trafficSplit = config.trafficSplit;
        this._encryptionSpec.internalValue = config.encryptionSpec;
        this._predictRequestResponseLoggingConfig.internalValue = config.predictRequestResponseLoggingConfig;
        this._privateServiceConnectConfig.internalValue = config.privateServiceConnectConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // dedicated_endpoint_dns - computed: true, optional: false, required: false
    get dedicatedEndpointDns() {
        return this.getStringAttribute('dedicated_endpoint_dns');
    }
    get dedicatedEndpointEnabled() {
        return this.getBooleanAttribute('dedicated_endpoint_enabled');
    }
    set dedicatedEndpointEnabled(value) {
        this._dedicatedEndpointEnabled = value;
    }
    resetDedicatedEndpointEnabled() {
        this._dedicatedEndpointEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dedicatedEndpointEnabledInput() {
        return this._dedicatedEndpointEnabled;
    }
    get deployedModels() {
        return this._deployedModels;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // model_deployment_monitoring_job - computed: true, optional: false, required: false
    get modelDeploymentMonitoringJob() {
        return this.getStringAttribute('model_deployment_monitoring_job');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    resetNetwork() {
        this._network = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    get trafficSplit() {
        return this.getStringAttribute('traffic_split');
    }
    set trafficSplit(value) {
        this._trafficSplit = value;
    }
    resetTrafficSplit() {
        this._trafficSplit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trafficSplitInput() {
        return this._trafficSplit;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get encryptionSpec() {
        return this._encryptionSpec;
    }
    putEncryptionSpec(value) {
        this._encryptionSpec.internalValue = value;
    }
    resetEncryptionSpec() {
        this._encryptionSpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get encryptionSpecInput() {
        return this._encryptionSpec.internalValue;
    }
    get predictRequestResponseLoggingConfig() {
        return this._predictRequestResponseLoggingConfig;
    }
    putPredictRequestResponseLoggingConfig(value) {
        this._predictRequestResponseLoggingConfig.internalValue = value;
    }
    resetPredictRequestResponseLoggingConfig() {
        this._predictRequestResponseLoggingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get predictRequestResponseLoggingConfigInput() {
        return this._predictRequestResponseLoggingConfig.internalValue;
    }
    get privateServiceConnectConfig() {
        return this._privateServiceConnectConfig;
    }
    putPrivateServiceConnectConfig(value) {
        this._privateServiceConnectConfig.internalValue = value;
    }
    resetPrivateServiceConnectConfig() {
        this._privateServiceConnectConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get privateServiceConnectConfigInput() {
        return this._privateServiceConnectConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            dedicated_endpoint_enabled: cdktf.booleanToTerraform(this._dedicatedEndpointEnabled),
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            network: cdktf.stringToTerraform(this._network),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            traffic_split: cdktf.stringToTerraform(this._trafficSplit),
            encryption_spec: vertexAiEndpointEncryptionSpecToTerraform(this._encryptionSpec.internalValue),
            predict_request_response_logging_config: vertexAiEndpointPredictRequestResponseLoggingConfigToTerraform(this._predictRequestResponseLoggingConfig.internalValue),
            private_service_connect_config: vertexAiEndpointPrivateServiceConnectConfigToTerraform(this._privateServiceConnectConfig.internalValue),
            timeouts: vertexAiEndpointTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            dedicated_endpoint_enabled: {
                value: cdktf.booleanToHclTerraform(this._dedicatedEndpointEnabled),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            network: {
                value: cdktf.stringToHclTerraform(this._network),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            traffic_split: {
                value: cdktf.stringToHclTerraform(this._trafficSplit),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            encryption_spec: {
                value: vertexAiEndpointEncryptionSpecToHclTerraform(this._encryptionSpec.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VertexAiEndpointEncryptionSpecList",
            },
            predict_request_response_logging_config: {
                value: vertexAiEndpointPredictRequestResponseLoggingConfigToHclTerraform(this._predictRequestResponseLoggingConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VertexAiEndpointPredictRequestResponseLoggingConfigList",
            },
            private_service_connect_config: {
                value: vertexAiEndpointPrivateServiceConnectConfigToHclTerraform(this._privateServiceConnectConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VertexAiEndpointPrivateServiceConnectConfigList",
            },
            timeouts: {
                value: vertexAiEndpointTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "VertexAiEndpointTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VertexAiEndpoint = VertexAiEndpoint;
_t = JSII_RTTI_SYMBOL_1;
VertexAiEndpoint[_t] = { fqn: "@cdktf/provider-google.vertexAiEndpoint.VertexAiEndpoint", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
VertexAiEndpoint.tfResourceType = "google_vertex_ai_endpoint";
