// @cdktf/provider-google/lib/data-google-pubsub-topic/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGooglePubsubTopic = exports.DataGooglePubsubTopicSchemaSettingsList = exports.DataGooglePubsubTopicSchemaSettingsOutputReference = exports.dataGooglePubsubTopicSchemaSettingsToHclTerraform = exports.dataGooglePubsubTopicSchemaSettingsToTerraform = exports.DataGooglePubsubTopicMessageStoragePolicyList = exports.DataGooglePubsubTopicMessageStoragePolicyOutputReference = exports.dataGooglePubsubTopicMessageStoragePolicyToHclTerraform = exports.dataGooglePubsubTopicMessageStoragePolicyToTerraform = exports.DataGooglePubsubTopicIngestionDataSourceSettingsList = exports.DataGooglePubsubTopicIngestionDataSourceSettingsOutputReference = exports.dataGooglePubsubTopicIngestionDataSourceSettingsToHclTerraform = exports.dataGooglePubsubTopicIngestionDataSourceSettingsToTerraform = exports.DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsList = exports.DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsOutputReference = exports.dataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsToHclTerraform = exports.dataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsToTerraform = exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageList = exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageOutputReference = exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageToHclTerraform = exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageToTerraform = exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatList = exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatOutputReference = exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatToHclTerraform = exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatToTerraform = exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatList = exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatOutputReference = exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatToHclTerraform = exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatToTerraform = exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatList = exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatOutputReference = exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatToHclTerraform = exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatToTerraform = exports.DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisList = exports.DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisOutputReference = exports.dataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisToHclTerraform = exports.dataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisToTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisToTerraform;
function dataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisToHclTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisToHclTerraform;
class DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // aws_role_arn - computed: true, optional: false, required: false
    get awsRoleArn() {
        return this.getStringAttribute('aws_role_arn');
    }
    // consumer_arn - computed: true, optional: false, required: false
    get consumerArn() {
        return this.getStringAttribute('consumer_arn');
    }
    // gcp_service_account - computed: true, optional: false, required: false
    get gcpServiceAccount() {
        return this.getStringAttribute('gcp_service_account');
    }
    // stream_arn - computed: true, optional: false, required: false
    get streamArn() {
        return this.getStringAttribute('stream_arn');
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisOutputReference = DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisList = DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisList;
_b = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisList[_b] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisList", version: "14.13.1" };
function dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatToTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatToTerraform;
function dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatToHclTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatToHclTerraform;
class DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatOutputReference = DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatList = DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatList;
_d = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatList[_d] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatList", version: "14.13.1" };
function dataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatToTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatToTerraform;
function dataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatToHclTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatToHclTerraform;
class DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatOutputReference = DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatList = DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatList;
_f = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatList[_f] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatList", version: "14.13.1" };
function dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatToTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatToTerraform;
function dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatToHclTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatToHclTerraform;
class DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // delimiter - computed: true, optional: false, required: false
    get delimiter() {
        return this.getStringAttribute('delimiter');
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatOutputReference = DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatList = DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatList;
_h = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatList[_h] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatList", version: "14.13.1" };
function dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageToTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageToTerraform;
function dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageToHclTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageToHclTerraform;
class DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // avro_format - computed: true, optional: false, required: false
        this._avroFormat = new DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageAvroFormatList(this, "avro_format", false);
        // pubsub_avro_format - computed: true, optional: false, required: false
        this._pubsubAvroFormat = new DataGooglePubsubTopicIngestionDataSourceSettingsCloudStoragePubsubAvroFormatList(this, "pubsub_avro_format", false);
        // text_format - computed: true, optional: false, required: false
        this._textFormat = new DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageTextFormatList(this, "text_format", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get avroFormat() {
        return this._avroFormat;
    }
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // match_glob - computed: true, optional: false, required: false
    get matchGlob() {
        return this.getStringAttribute('match_glob');
    }
    // minimum_object_create_time - computed: true, optional: false, required: false
    get minimumObjectCreateTime() {
        return this.getStringAttribute('minimum_object_create_time');
    }
    get pubsubAvroFormat() {
        return this._pubsubAvroFormat;
    }
    get textFormat() {
        return this._textFormat;
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageOutputReference = DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageList = DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageList;
_k = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageList[_k] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageList", version: "14.13.1" };
function dataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsToTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsToTerraform;
function dataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsToHclTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsToHclTerraform;
class DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // severity - computed: true, optional: false, required: false
    get severity() {
        return this.getStringAttribute('severity');
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsOutputReference = DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsList = DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsList;
_m = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsList[_m] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsList", version: "14.13.1" };
function dataGooglePubsubTopicIngestionDataSourceSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsToTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsToTerraform;
function dataGooglePubsubTopicIngestionDataSourceSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicIngestionDataSourceSettingsToHclTerraform = dataGooglePubsubTopicIngestionDataSourceSettingsToHclTerraform;
class DataGooglePubsubTopicIngestionDataSourceSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // aws_kinesis - computed: true, optional: false, required: false
        this._awsKinesis = new DataGooglePubsubTopicIngestionDataSourceSettingsAwsKinesisList(this, "aws_kinesis", false);
        // cloud_storage - computed: true, optional: false, required: false
        this._cloudStorage = new DataGooglePubsubTopicIngestionDataSourceSettingsCloudStorageList(this, "cloud_storage", false);
        // platform_logs_settings - computed: true, optional: false, required: false
        this._platformLogsSettings = new DataGooglePubsubTopicIngestionDataSourceSettingsPlatformLogsSettingsList(this, "platform_logs_settings", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get awsKinesis() {
        return this._awsKinesis;
    }
    get cloudStorage() {
        return this._cloudStorage;
    }
    get platformLogsSettings() {
        return this._platformLogsSettings;
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsOutputReference = DataGooglePubsubTopicIngestionDataSourceSettingsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicIngestionDataSourceSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicIngestionDataSourceSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicIngestionDataSourceSettingsList = DataGooglePubsubTopicIngestionDataSourceSettingsList;
_p = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicIngestionDataSourceSettingsList[_p] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicIngestionDataSourceSettingsList", version: "14.13.1" };
function dataGooglePubsubTopicMessageStoragePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicMessageStoragePolicyToTerraform = dataGooglePubsubTopicMessageStoragePolicyToTerraform;
function dataGooglePubsubTopicMessageStoragePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicMessageStoragePolicyToHclTerraform = dataGooglePubsubTopicMessageStoragePolicyToHclTerraform;
class DataGooglePubsubTopicMessageStoragePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allowed_persistence_regions - computed: true, optional: false, required: false
    get allowedPersistenceRegions() {
        return this.getListAttribute('allowed_persistence_regions');
    }
}
exports.DataGooglePubsubTopicMessageStoragePolicyOutputReference = DataGooglePubsubTopicMessageStoragePolicyOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicMessageStoragePolicyOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicMessageStoragePolicyOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicMessageStoragePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicMessageStoragePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicMessageStoragePolicyList = DataGooglePubsubTopicMessageStoragePolicyList;
_r = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicMessageStoragePolicyList[_r] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicMessageStoragePolicyList", version: "14.13.1" };
function dataGooglePubsubTopicSchemaSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGooglePubsubTopicSchemaSettingsToTerraform = dataGooglePubsubTopicSchemaSettingsToTerraform;
function dataGooglePubsubTopicSchemaSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGooglePubsubTopicSchemaSettingsToHclTerraform = dataGooglePubsubTopicSchemaSettingsToHclTerraform;
class DataGooglePubsubTopicSchemaSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // encoding - computed: true, optional: false, required: false
    get encoding() {
        return this.getStringAttribute('encoding');
    }
    // schema - computed: true, optional: false, required: false
    get schema() {
        return this.getStringAttribute('schema');
    }
}
exports.DataGooglePubsubTopicSchemaSettingsOutputReference = DataGooglePubsubTopicSchemaSettingsOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicSchemaSettingsOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicSchemaSettingsOutputReference", version: "14.13.1" };
class DataGooglePubsubTopicSchemaSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGooglePubsubTopicSchemaSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGooglePubsubTopicSchemaSettingsList = DataGooglePubsubTopicSchemaSettingsList;
_t = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopicSchemaSettingsList[_t] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopicSchemaSettingsList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/pubsub_topic google_pubsub_topic}
*/
class DataGooglePubsubTopic extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGooglePubsubTopic resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGooglePubsubTopic to import
    * @param importFromId The id of the existing DataGooglePubsubTopic that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/pubsub_topic#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGooglePubsubTopic to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_pubsub_topic", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/pubsub_topic google_pubsub_topic} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGooglePubsubTopicConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_pubsub_topic',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // ingestion_data_source_settings - computed: true, optional: false, required: false
        this._ingestionDataSourceSettings = new DataGooglePubsubTopicIngestionDataSourceSettingsList(this, "ingestion_data_source_settings", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // message_storage_policy - computed: true, optional: false, required: false
        this._messageStoragePolicy = new DataGooglePubsubTopicMessageStoragePolicyList(this, "message_storage_policy", false);
        // schema_settings - computed: true, optional: false, required: false
        this._schemaSettings = new DataGooglePubsubTopicSchemaSettingsList(this, "schema_settings", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get ingestionDataSourceSettings() {
        return this._ingestionDataSourceSettings;
    }
    // kms_key_name - computed: true, optional: false, required: false
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    get labels() {
        return this._labels;
    }
    // message_retention_duration - computed: true, optional: false, required: false
    get messageRetentionDuration() {
        return this.getStringAttribute('message_retention_duration');
    }
    get messageStoragePolicy() {
        return this._messageStoragePolicy;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get schemaSettings() {
        return this._schemaSettings;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGooglePubsubTopic = DataGooglePubsubTopic;
_u = JSII_RTTI_SYMBOL_1;
DataGooglePubsubTopic[_u] = { fqn: "@cdktf/provider-google.dataGooglePubsubTopic.DataGooglePubsubTopic", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGooglePubsubTopic.tfResourceType = "google_pubsub_topic";
