// @cdktf/provider-google/lib/managed-kafka-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManagedKafkaCluster = exports.ManagedKafkaClusterTimeoutsOutputReference = exports.managedKafkaClusterTimeoutsToHclTerraform = exports.managedKafkaClusterTimeoutsToTerraform = exports.ManagedKafkaClusterRebalanceConfigOutputReference = exports.managedKafkaClusterRebalanceConfigToHclTerraform = exports.managedKafkaClusterRebalanceConfigToTerraform = exports.ManagedKafkaClusterGcpConfigOutputReference = exports.managedKafkaClusterGcpConfigToHclTerraform = exports.managedKafkaClusterGcpConfigToTerraform = exports.ManagedKafkaClusterGcpConfigAccessConfigOutputReference = exports.managedKafkaClusterGcpConfigAccessConfigToHclTerraform = exports.managedKafkaClusterGcpConfigAccessConfigToTerraform = exports.ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsList = exports.ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsOutputReference = exports.managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToHclTerraform = exports.managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToTerraform = exports.ManagedKafkaClusterCapacityConfigOutputReference = exports.managedKafkaClusterCapacityConfigToHclTerraform = exports.managedKafkaClusterCapacityConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function managedKafkaClusterCapacityConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        memory_bytes: cdktf.stringToTerraform(struct.memoryBytes),
        vcpu_count: cdktf.stringToTerraform(struct.vcpuCount),
    };
}
exports.managedKafkaClusterCapacityConfigToTerraform = managedKafkaClusterCapacityConfigToTerraform;
function managedKafkaClusterCapacityConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        memory_bytes: {
            value: cdktf.stringToHclTerraform(struct.memoryBytes),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        vcpu_count: {
            value: cdktf.stringToHclTerraform(struct.vcpuCount),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.managedKafkaClusterCapacityConfigToHclTerraform = managedKafkaClusterCapacityConfigToHclTerraform;
class ManagedKafkaClusterCapacityConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._memoryBytes !== undefined) {
            hasAnyValues = true;
            internalValueResult.memoryBytes = this._memoryBytes;
        }
        if (this._vcpuCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.vcpuCount = this._vcpuCount;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._memoryBytes = undefined;
            this._vcpuCount = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._memoryBytes = value.memoryBytes;
            this._vcpuCount = value.vcpuCount;
        }
    }
    get memoryBytes() {
        return this.getStringAttribute('memory_bytes');
    }
    set memoryBytes(value) {
        this._memoryBytes = value;
    }
    // Temporarily expose input value. Use with caution.
    get memoryBytesInput() {
        return this._memoryBytes;
    }
    get vcpuCount() {
        return this.getStringAttribute('vcpu_count');
    }
    set vcpuCount(value) {
        this._vcpuCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get vcpuCountInput() {
        return this._vcpuCount;
    }
}
exports.ManagedKafkaClusterCapacityConfigOutputReference = ManagedKafkaClusterCapacityConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ManagedKafkaClusterCapacityConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.managedKafkaCluster.ManagedKafkaClusterCapacityConfigOutputReference", version: "14.13.1" };
function managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subnet: cdktf.stringToTerraform(struct.subnet),
    };
}
exports.managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToTerraform = managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToTerraform;
function managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        subnet: {
            value: cdktf.stringToHclTerraform(struct.subnet),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToHclTerraform = managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToHclTerraform;
class ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._subnet !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnet = this._subnet;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._subnet = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._subnet = value.subnet;
        }
    }
    get subnet() {
        return this.getStringAttribute('subnet');
    }
    set subnet(value) {
        this._subnet = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetInput() {
        return this._subnet;
    }
}
exports.ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsOutputReference = ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsOutputReference[_b] = { fqn: "@cdktf/provider-google.managedKafkaCluster.ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsOutputReference", version: "14.13.1" };
class ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsList = ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsList;
_c = JSII_RTTI_SYMBOL_1;
ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsList[_c] = { fqn: "@cdktf/provider-google.managedKafkaCluster.ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsList", version: "14.13.1" };
function managedKafkaClusterGcpConfigAccessConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        network_configs: cdktf.listMapper(managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToTerraform, true)(struct.networkConfigs),
    };
}
exports.managedKafkaClusterGcpConfigAccessConfigToTerraform = managedKafkaClusterGcpConfigAccessConfigToTerraform;
function managedKafkaClusterGcpConfigAccessConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        network_configs: {
            value: cdktf.listMapperHcl(managedKafkaClusterGcpConfigAccessConfigNetworkConfigsToHclTerraform, true)(struct.networkConfigs),
            isBlock: true,
            type: "list",
            storageClassType: "ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.managedKafkaClusterGcpConfigAccessConfigToHclTerraform = managedKafkaClusterGcpConfigAccessConfigToHclTerraform;
class ManagedKafkaClusterGcpConfigAccessConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // network_configs - computed: false, optional: false, required: true
        this._networkConfigs = new ManagedKafkaClusterGcpConfigAccessConfigNetworkConfigsList(this, "network_configs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._networkConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.networkConfigs = this._networkConfigs?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._networkConfigs.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._networkConfigs.internalValue = value.networkConfigs;
        }
    }
    get networkConfigs() {
        return this._networkConfigs;
    }
    putNetworkConfigs(value) {
        this._networkConfigs.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkConfigsInput() {
        return this._networkConfigs.internalValue;
    }
}
exports.ManagedKafkaClusterGcpConfigAccessConfigOutputReference = ManagedKafkaClusterGcpConfigAccessConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ManagedKafkaClusterGcpConfigAccessConfigOutputReference[_d] = { fqn: "@cdktf/provider-google.managedKafkaCluster.ManagedKafkaClusterGcpConfigAccessConfigOutputReference", version: "14.13.1" };
function managedKafkaClusterGcpConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key: cdktf.stringToTerraform(struct.kmsKey),
        access_config: managedKafkaClusterGcpConfigAccessConfigToTerraform(struct.accessConfig),
    };
}
exports.managedKafkaClusterGcpConfigToTerraform = managedKafkaClusterGcpConfigToTerraform;
function managedKafkaClusterGcpConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key: {
            value: cdktf.stringToHclTerraform(struct.kmsKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        access_config: {
            value: managedKafkaClusterGcpConfigAccessConfigToHclTerraform(struct.accessConfig),
            isBlock: true,
            type: "list",
            storageClassType: "ManagedKafkaClusterGcpConfigAccessConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.managedKafkaClusterGcpConfigToHclTerraform = managedKafkaClusterGcpConfigToHclTerraform;
class ManagedKafkaClusterGcpConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // access_config - computed: false, optional: false, required: true
        this._accessConfig = new ManagedKafkaClusterGcpConfigAccessConfigOutputReference(this, "access_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKey = this._kmsKey;
        }
        if (this._accessConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessConfig = this._accessConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKey = undefined;
            this._accessConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKey = value.kmsKey;
            this._accessConfig.internalValue = value.accessConfig;
        }
    }
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    set kmsKey(value) {
        this._kmsKey = value;
    }
    resetKmsKey() {
        this._kmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyInput() {
        return this._kmsKey;
    }
    get accessConfig() {
        return this._accessConfig;
    }
    putAccessConfig(value) {
        this._accessConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessConfigInput() {
        return this._accessConfig.internalValue;
    }
}
exports.ManagedKafkaClusterGcpConfigOutputReference = ManagedKafkaClusterGcpConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ManagedKafkaClusterGcpConfigOutputReference[_e] = { fqn: "@cdktf/provider-google.managedKafkaCluster.ManagedKafkaClusterGcpConfigOutputReference", version: "14.13.1" };
function managedKafkaClusterRebalanceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.managedKafkaClusterRebalanceConfigToTerraform = managedKafkaClusterRebalanceConfigToTerraform;
function managedKafkaClusterRebalanceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        mode: {
            value: cdktf.stringToHclTerraform(struct.mode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.managedKafkaClusterRebalanceConfigToHclTerraform = managedKafkaClusterRebalanceConfigToHclTerraform;
class ManagedKafkaClusterRebalanceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._mode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._mode = value.mode;
        }
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.ManagedKafkaClusterRebalanceConfigOutputReference = ManagedKafkaClusterRebalanceConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ManagedKafkaClusterRebalanceConfigOutputReference[_f] = { fqn: "@cdktf/provider-google.managedKafkaCluster.ManagedKafkaClusterRebalanceConfigOutputReference", version: "14.13.1" };
function managedKafkaClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.managedKafkaClusterTimeoutsToTerraform = managedKafkaClusterTimeoutsToTerraform;
function managedKafkaClusterTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.managedKafkaClusterTimeoutsToHclTerraform = managedKafkaClusterTimeoutsToHclTerraform;
class ManagedKafkaClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ManagedKafkaClusterTimeoutsOutputReference = ManagedKafkaClusterTimeoutsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ManagedKafkaClusterTimeoutsOutputReference[_g] = { fqn: "@cdktf/provider-google.managedKafkaCluster.ManagedKafkaClusterTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/managed_kafka_cluster google_managed_kafka_cluster}
*/
class ManagedKafkaCluster extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ManagedKafkaCluster resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ManagedKafkaCluster to import
    * @param importFromId The id of the existing ManagedKafkaCluster that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/managed_kafka_cluster#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ManagedKafkaCluster to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_managed_kafka_cluster", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/managed_kafka_cluster google_managed_kafka_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ManagedKafkaClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_managed_kafka_cluster',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // capacity_config - computed: false, optional: false, required: true
        this._capacityConfig = new ManagedKafkaClusterCapacityConfigOutputReference(this, "capacity_config");
        // gcp_config - computed: false, optional: false, required: true
        this._gcpConfig = new ManagedKafkaClusterGcpConfigOutputReference(this, "gcp_config");
        // rebalance_config - computed: false, optional: true, required: false
        this._rebalanceConfig = new ManagedKafkaClusterRebalanceConfigOutputReference(this, "rebalance_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ManagedKafkaClusterTimeoutsOutputReference(this, "timeouts");
        this._clusterId = config.clusterId;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._project = config.project;
        this._capacityConfig.internalValue = config.capacityConfig;
        this._gcpConfig.internalValue = config.gcpConfig;
        this._rebalanceConfig.internalValue = config.rebalanceConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get capacityConfig() {
        return this._capacityConfig;
    }
    putCapacityConfig(value) {
        this._capacityConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get capacityConfigInput() {
        return this._capacityConfig.internalValue;
    }
    get gcpConfig() {
        return this._gcpConfig;
    }
    putGcpConfig(value) {
        this._gcpConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get gcpConfigInput() {
        return this._gcpConfig.internalValue;
    }
    get rebalanceConfig() {
        return this._rebalanceConfig;
    }
    putRebalanceConfig(value) {
        this._rebalanceConfig.internalValue = value;
    }
    resetRebalanceConfig() {
        this._rebalanceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rebalanceConfigInput() {
        return this._rebalanceConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            capacity_config: managedKafkaClusterCapacityConfigToTerraform(this._capacityConfig.internalValue),
            gcp_config: managedKafkaClusterGcpConfigToTerraform(this._gcpConfig.internalValue),
            rebalance_config: managedKafkaClusterRebalanceConfigToTerraform(this._rebalanceConfig.internalValue),
            timeouts: managedKafkaClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cluster_id: {
                value: cdktf.stringToHclTerraform(this._clusterId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            capacity_config: {
                value: managedKafkaClusterCapacityConfigToHclTerraform(this._capacityConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ManagedKafkaClusterCapacityConfigList",
            },
            gcp_config: {
                value: managedKafkaClusterGcpConfigToHclTerraform(this._gcpConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ManagedKafkaClusterGcpConfigList",
            },
            rebalance_config: {
                value: managedKafkaClusterRebalanceConfigToHclTerraform(this._rebalanceConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ManagedKafkaClusterRebalanceConfigList",
            },
            timeouts: {
                value: managedKafkaClusterTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ManagedKafkaClusterTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ManagedKafkaCluster = ManagedKafkaCluster;
_h = JSII_RTTI_SYMBOL_1;
ManagedKafkaCluster[_h] = { fqn: "@cdktf/provider-google.managedKafkaCluster.ManagedKafkaCluster", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ManagedKafkaCluster.tfResourceType = "google_managed_kafka_cluster";
