// @cdktf/provider-google/lib/biglake-table/index.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BiglakeTable = exports.BiglakeTableTimeoutsOutputReference = exports.biglakeTableTimeoutsToHclTerraform = exports.biglakeTableTimeoutsToTerraform = exports.BiglakeTableHiveOptionsOutputReference = exports.biglakeTableHiveOptionsToHclTerraform = exports.biglakeTableHiveOptionsToTerraform = exports.BiglakeTableHiveOptionsStorageDescriptorOutputReference = exports.biglakeTableHiveOptionsStorageDescriptorToHclTerraform = exports.biglakeTableHiveOptionsStorageDescriptorToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function biglakeTableHiveOptionsStorageDescriptorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        input_format: cdktf.stringToTerraform(struct.inputFormat),
        location_uri: cdktf.stringToTerraform(struct.locationUri),
        output_format: cdktf.stringToTerraform(struct.outputFormat),
    };
}
exports.biglakeTableHiveOptionsStorageDescriptorToTerraform = biglakeTableHiveOptionsStorageDescriptorToTerraform;
function biglakeTableHiveOptionsStorageDescriptorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        input_format: {
            value: cdktf.stringToHclTerraform(struct.inputFormat),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        location_uri: {
            value: cdktf.stringToHclTerraform(struct.locationUri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        output_format: {
            value: cdktf.stringToHclTerraform(struct.outputFormat),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.biglakeTableHiveOptionsStorageDescriptorToHclTerraform = biglakeTableHiveOptionsStorageDescriptorToHclTerraform;
class BiglakeTableHiveOptionsStorageDescriptorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._inputFormat !== undefined) {
            hasAnyValues = true;
            internalValueResult.inputFormat = this._inputFormat;
        }
        if (this._locationUri !== undefined) {
            hasAnyValues = true;
            internalValueResult.locationUri = this._locationUri;
        }
        if (this._outputFormat !== undefined) {
            hasAnyValues = true;
            internalValueResult.outputFormat = this._outputFormat;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._inputFormat = undefined;
            this._locationUri = undefined;
            this._outputFormat = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._inputFormat = value.inputFormat;
            this._locationUri = value.locationUri;
            this._outputFormat = value.outputFormat;
        }
    }
    get inputFormat() {
        return this.getStringAttribute('input_format');
    }
    set inputFormat(value) {
        this._inputFormat = value;
    }
    resetInputFormat() {
        this._inputFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get inputFormatInput() {
        return this._inputFormat;
    }
    get locationUri() {
        return this.getStringAttribute('location_uri');
    }
    set locationUri(value) {
        this._locationUri = value;
    }
    resetLocationUri() {
        this._locationUri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationUriInput() {
        return this._locationUri;
    }
    get outputFormat() {
        return this.getStringAttribute('output_format');
    }
    set outputFormat(value) {
        this._outputFormat = value;
    }
    resetOutputFormat() {
        this._outputFormat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputFormatInput() {
        return this._outputFormat;
    }
}
exports.BiglakeTableHiveOptionsStorageDescriptorOutputReference = BiglakeTableHiveOptionsStorageDescriptorOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BiglakeTableHiveOptionsStorageDescriptorOutputReference[_a] = { fqn: "@cdktf/provider-google.biglakeTable.BiglakeTableHiveOptionsStorageDescriptorOutputReference", version: "14.13.1" };
function biglakeTableHiveOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        parameters: cdktf.hashMapper(cdktf.stringToTerraform)(struct.parameters),
        table_type: cdktf.stringToTerraform(struct.tableType),
        storage_descriptor: biglakeTableHiveOptionsStorageDescriptorToTerraform(struct.storageDescriptor),
    };
}
exports.biglakeTableHiveOptionsToTerraform = biglakeTableHiveOptionsToTerraform;
function biglakeTableHiveOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        parameters: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.parameters),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        table_type: {
            value: cdktf.stringToHclTerraform(struct.tableType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        storage_descriptor: {
            value: biglakeTableHiveOptionsStorageDescriptorToHclTerraform(struct.storageDescriptor),
            isBlock: true,
            type: "list",
            storageClassType: "BiglakeTableHiveOptionsStorageDescriptorList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.biglakeTableHiveOptionsToHclTerraform = biglakeTableHiveOptionsToHclTerraform;
class BiglakeTableHiveOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // storage_descriptor - computed: false, optional: true, required: false
        this._storageDescriptor = new BiglakeTableHiveOptionsStorageDescriptorOutputReference(this, "storage_descriptor");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._parameters !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameters = this._parameters;
        }
        if (this._tableType !== undefined) {
            hasAnyValues = true;
            internalValueResult.tableType = this._tableType;
        }
        if (this._storageDescriptor?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageDescriptor = this._storageDescriptor?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._parameters = undefined;
            this._tableType = undefined;
            this._storageDescriptor.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._parameters = value.parameters;
            this._tableType = value.tableType;
            this._storageDescriptor.internalValue = value.storageDescriptor;
        }
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get tableType() {
        return this.getStringAttribute('table_type');
    }
    set tableType(value) {
        this._tableType = value;
    }
    resetTableType() {
        this._tableType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tableTypeInput() {
        return this._tableType;
    }
    get storageDescriptor() {
        return this._storageDescriptor;
    }
    putStorageDescriptor(value) {
        this._storageDescriptor.internalValue = value;
    }
    resetStorageDescriptor() {
        this._storageDescriptor.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get storageDescriptorInput() {
        return this._storageDescriptor.internalValue;
    }
}
exports.BiglakeTableHiveOptionsOutputReference = BiglakeTableHiveOptionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
BiglakeTableHiveOptionsOutputReference[_b] = { fqn: "@cdktf/provider-google.biglakeTable.BiglakeTableHiveOptionsOutputReference", version: "14.13.1" };
function biglakeTableTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.biglakeTableTimeoutsToTerraform = biglakeTableTimeoutsToTerraform;
function biglakeTableTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.biglakeTableTimeoutsToHclTerraform = biglakeTableTimeoutsToHclTerraform;
class BiglakeTableTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.BiglakeTableTimeoutsOutputReference = BiglakeTableTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
BiglakeTableTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-google.biglakeTable.BiglakeTableTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/biglake_table google_biglake_table}
*/
class BiglakeTable extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a BiglakeTable resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the BiglakeTable to import
    * @param importFromId The id of the existing BiglakeTable that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/biglake_table#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the BiglakeTable to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_biglake_table", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/biglake_table google_biglake_table} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BiglakeTableConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_biglake_table',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // hive_options - computed: false, optional: true, required: false
        this._hiveOptions = new BiglakeTableHiveOptionsOutputReference(this, "hive_options");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new BiglakeTableTimeoutsOutputReference(this, "timeouts");
        this._database = config.database;
        this._id = config.id;
        this._name = config.name;
        this._type = config.type;
        this._hiveOptions.internalValue = config.hiveOptions;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get database() {
        return this.getStringAttribute('database');
    }
    set database(value) {
        this._database = value;
    }
    resetDatabase() {
        this._database = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database;
    }
    // delete_time - computed: true, optional: false, required: false
    get deleteTime() {
        return this.getStringAttribute('delete_time');
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    // expire_time - computed: true, optional: false, required: false
    get expireTime() {
        return this.getStringAttribute('expire_time');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get hiveOptions() {
        return this._hiveOptions;
    }
    putHiveOptions(value) {
        this._hiveOptions.internalValue = value;
    }
    resetHiveOptions() {
        this._hiveOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hiveOptionsInput() {
        return this._hiveOptions.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            database: cdktf.stringToTerraform(this._database),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            type: cdktf.stringToTerraform(this._type),
            hive_options: biglakeTableHiveOptionsToTerraform(this._hiveOptions.internalValue),
            timeouts: biglakeTableTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            database: {
                value: cdktf.stringToHclTerraform(this._database),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            hive_options: {
                value: biglakeTableHiveOptionsToHclTerraform(this._hiveOptions.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BiglakeTableHiveOptionsList",
            },
            timeouts: {
                value: biglakeTableTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "BiglakeTableTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.BiglakeTable = BiglakeTable;
_d = JSII_RTTI_SYMBOL_1;
BiglakeTable[_d] = { fqn: "@cdktf/provider-google.biglakeTable.BiglakeTable", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
BiglakeTable.tfResourceType = "google_biglake_table";
