// @cdktf/provider-google/lib/data-google-certificate-manager-certificates/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleCertificateManagerCertificates = exports.DataGoogleCertificateManagerCertificatesCertificatesList = exports.DataGoogleCertificateManagerCertificatesCertificatesOutputReference = exports.dataGoogleCertificateManagerCertificatesCertificatesToHclTerraform = exports.dataGoogleCertificateManagerCertificatesCertificatesToTerraform = exports.DataGoogleCertificateManagerCertificatesCertificatesManagedList = exports.DataGoogleCertificateManagerCertificatesCertificatesManagedOutputReference = exports.dataGoogleCertificateManagerCertificatesCertificatesManagedToHclTerraform = exports.dataGoogleCertificateManagerCertificatesCertificatesManagedToTerraform = exports.DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueList = exports.DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueOutputReference = exports.dataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueToHclTerraform = exports.dataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueToTerraform = exports.DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoList = exports.DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoOutputReference = exports.dataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoToHclTerraform = exports.dataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoToTerraform = dataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoToTerraform;
function dataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoToHclTerraform = dataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoToHclTerraform;
class DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // details - computed: true, optional: false, required: false
    get details() {
        return this.getStringAttribute('details');
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
    // failure_reason - computed: true, optional: false, required: false
    get failureReason() {
        return this.getStringAttribute('failure_reason');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoOutputReference = DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoOutputReference", version: "14.13.1" };
class DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoList = DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoList[_b] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoList", version: "14.13.1" };
function dataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueToTerraform = dataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueToTerraform;
function dataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueToHclTerraform = dataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueToHclTerraform;
class DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // details - computed: true, optional: false, required: false
    get details() {
        return this.getStringAttribute('details');
    }
    // reason - computed: true, optional: false, required: false
    get reason() {
        return this.getStringAttribute('reason');
    }
}
exports.DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueOutputReference = DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueOutputReference", version: "14.13.1" };
class DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueList = DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueList[_d] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueList", version: "14.13.1" };
function dataGoogleCertificateManagerCertificatesCertificatesManagedToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCertificateManagerCertificatesCertificatesManagedToTerraform = dataGoogleCertificateManagerCertificatesCertificatesManagedToTerraform;
function dataGoogleCertificateManagerCertificatesCertificatesManagedToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCertificateManagerCertificatesCertificatesManagedToHclTerraform = dataGoogleCertificateManagerCertificatesCertificatesManagedToHclTerraform;
class DataGoogleCertificateManagerCertificatesCertificatesManagedOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authorization_attempt_info - computed: true, optional: false, required: false
        this._authorizationAttemptInfo = new DataGoogleCertificateManagerCertificatesCertificatesManagedAuthorizationAttemptInfoList(this, "authorization_attempt_info", false);
        // provisioning_issue - computed: true, optional: false, required: false
        this._provisioningIssue = new DataGoogleCertificateManagerCertificatesCertificatesManagedProvisioningIssueList(this, "provisioning_issue", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get authorizationAttemptInfo() {
        return this._authorizationAttemptInfo;
    }
    // dns_authorizations - computed: true, optional: false, required: false
    get dnsAuthorizations() {
        return this.getListAttribute('dns_authorizations');
    }
    // domains - computed: true, optional: false, required: false
    get domains() {
        return this.getListAttribute('domains');
    }
    // issuance_config - computed: true, optional: false, required: false
    get issuanceConfig() {
        return this.getStringAttribute('issuance_config');
    }
    get provisioningIssue() {
        return this._provisioningIssue;
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.DataGoogleCertificateManagerCertificatesCertificatesManagedOutputReference = DataGoogleCertificateManagerCertificatesCertificatesManagedOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificatesCertificatesManagedOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificatesCertificatesManagedOutputReference", version: "14.13.1" };
class DataGoogleCertificateManagerCertificatesCertificatesManagedList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCertificateManagerCertificatesCertificatesManagedOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCertificateManagerCertificatesCertificatesManagedList = DataGoogleCertificateManagerCertificatesCertificatesManagedList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificatesCertificatesManagedList[_f] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificatesCertificatesManagedList", version: "14.13.1" };
function dataGoogleCertificateManagerCertificatesCertificatesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleCertificateManagerCertificatesCertificatesToTerraform = dataGoogleCertificateManagerCertificatesCertificatesToTerraform;
function dataGoogleCertificateManagerCertificatesCertificatesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleCertificateManagerCertificatesCertificatesToHclTerraform = dataGoogleCertificateManagerCertificatesCertificatesToHclTerraform;
class DataGoogleCertificateManagerCertificatesCertificatesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // managed - computed: true, optional: false, required: false
        this._managed = new DataGoogleCertificateManagerCertificatesCertificatesManagedList(this, "managed", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get labels() {
        return this._labels;
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    get managed() {
        return this._managed;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
    // san_dnsnames - computed: true, optional: false, required: false
    get sanDnsnames() {
        return this.getListAttribute('san_dnsnames');
    }
    // scope - computed: true, optional: false, required: false
    get scope() {
        return this.getStringAttribute('scope');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
}
exports.DataGoogleCertificateManagerCertificatesCertificatesOutputReference = DataGoogleCertificateManagerCertificatesCertificatesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificatesCertificatesOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificatesCertificatesOutputReference", version: "14.13.1" };
class DataGoogleCertificateManagerCertificatesCertificatesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleCertificateManagerCertificatesCertificatesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleCertificateManagerCertificatesCertificatesList = DataGoogleCertificateManagerCertificatesCertificatesList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificatesCertificatesList[_h] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificatesCertificatesList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/certificate_manager_certificates google_certificate_manager_certificates}
*/
class DataGoogleCertificateManagerCertificates extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleCertificateManagerCertificates resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleCertificateManagerCertificates to import
    * @param importFromId The id of the existing DataGoogleCertificateManagerCertificates that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/certificate_manager_certificates#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleCertificateManagerCertificates to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_certificate_manager_certificates", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/certificate_manager_certificates google_certificate_manager_certificates} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleCertificateManagerCertificatesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_certificate_manager_certificates',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // certificates - computed: true, optional: false, required: false
        this._certificates = new DataGoogleCertificateManagerCertificatesCertificatesList(this, "certificates", false);
        this._filter = config.filter;
        this._id = config.id;
        this._region = config.region;
    }
    get certificates() {
        return this._certificates;
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            filter: cdktf.stringToTerraform(this._filter),
            id: cdktf.stringToTerraform(this._id),
            region: cdktf.stringToTerraform(this._region),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            filter: {
                value: cdktf.stringToHclTerraform(this._filter),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleCertificateManagerCertificates = DataGoogleCertificateManagerCertificates;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleCertificateManagerCertificates[_j] = { fqn: "@cdktf/provider-google.dataGoogleCertificateManagerCertificates.DataGoogleCertificateManagerCertificates", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleCertificateManagerCertificates.tfResourceType = "google_certificate_manager_certificates";
