// @cdktf/provider-google/lib/vertex-ai-deployment-resource-pool/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VertexAiDeploymentResourcePool = exports.VertexAiDeploymentResourcePoolTimeoutsOutputReference = exports.vertexAiDeploymentResourcePoolTimeoutsToHclTerraform = exports.vertexAiDeploymentResourcePoolTimeoutsToTerraform = exports.VertexAiDeploymentResourcePoolDedicatedResourcesOutputReference = exports.vertexAiDeploymentResourcePoolDedicatedResourcesToHclTerraform = exports.vertexAiDeploymentResourcePoolDedicatedResourcesToTerraform = exports.VertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecOutputReference = exports.vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToHclTerraform = exports.vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToTerraform = exports.VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsList = exports.VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsOutputReference = exports.vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToHclTerraform = exports.vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metric_name: cdktf.stringToTerraform(struct.metricName),
        target: cdktf.numberToTerraform(struct.target),
    };
}
exports.vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToTerraform = vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToTerraform;
function vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        metric_name: {
            value: cdktf.stringToHclTerraform(struct.metricName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        target: {
            value: cdktf.numberToHclTerraform(struct.target),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToHclTerraform = vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToHclTerraform;
class VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metricName !== undefined) {
            hasAnyValues = true;
            internalValueResult.metricName = this._metricName;
        }
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._metricName = undefined;
            this._target = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._metricName = value.metricName;
            this._target = value.target;
        }
    }
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    set metricName(value) {
        this._metricName = value;
    }
    // Temporarily expose input value. Use with caution.
    get metricNameInput() {
        return this._metricName;
    }
    get target() {
        return this.getNumberAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
}
exports.VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsOutputReference = VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsOutputReference[_a] = { fqn: "@cdktf/provider-google.vertexAiDeploymentResourcePool.VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsOutputReference", version: "14.13.1" };
class VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsList = VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsList;
_b = JSII_RTTI_SYMBOL_1;
VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsList[_b] = { fqn: "@cdktf/provider-google.vertexAiDeploymentResourcePool.VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsList", version: "14.13.1" };
function vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        accelerator_count: cdktf.numberToTerraform(struct.acceleratorCount),
        accelerator_type: cdktf.stringToTerraform(struct.acceleratorType),
        machine_type: cdktf.stringToTerraform(struct.machineType),
    };
}
exports.vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToTerraform = vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToTerraform;
function vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        accelerator_count: {
            value: cdktf.numberToHclTerraform(struct.acceleratorCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        accelerator_type: {
            value: cdktf.stringToHclTerraform(struct.acceleratorType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        machine_type: {
            value: cdktf.stringToHclTerraform(struct.machineType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToHclTerraform = vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToHclTerraform;
class VertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._acceleratorCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.acceleratorCount = this._acceleratorCount;
        }
        if (this._acceleratorType !== undefined) {
            hasAnyValues = true;
            internalValueResult.acceleratorType = this._acceleratorType;
        }
        if (this._machineType !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineType = this._machineType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._acceleratorCount = undefined;
            this._acceleratorType = undefined;
            this._machineType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._acceleratorCount = value.acceleratorCount;
            this._acceleratorType = value.acceleratorType;
            this._machineType = value.machineType;
        }
    }
    get acceleratorCount() {
        return this.getNumberAttribute('accelerator_count');
    }
    set acceleratorCount(value) {
        this._acceleratorCount = value;
    }
    resetAcceleratorCount() {
        this._acceleratorCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acceleratorCountInput() {
        return this._acceleratorCount;
    }
    get acceleratorType() {
        return this.getStringAttribute('accelerator_type');
    }
    set acceleratorType(value) {
        this._acceleratorType = value;
    }
    resetAcceleratorType() {
        this._acceleratorType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get acceleratorTypeInput() {
        return this._acceleratorType;
    }
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    set machineType(value) {
        this._machineType = value;
    }
    resetMachineType() {
        this._machineType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get machineTypeInput() {
        return this._machineType;
    }
}
exports.VertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecOutputReference = VertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecOutputReference;
_c = JSII_RTTI_SYMBOL_1;
VertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecOutputReference[_c] = { fqn: "@cdktf/provider-google.vertexAiDeploymentResourcePool.VertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecOutputReference", version: "14.13.1" };
function vertexAiDeploymentResourcePoolDedicatedResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_replica_count: cdktf.numberToTerraform(struct.maxReplicaCount),
        min_replica_count: cdktf.numberToTerraform(struct.minReplicaCount),
        autoscaling_metric_specs: cdktf.listMapper(vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToTerraform, true)(struct.autoscalingMetricSpecs),
        machine_spec: vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToTerraform(struct.machineSpec),
    };
}
exports.vertexAiDeploymentResourcePoolDedicatedResourcesToTerraform = vertexAiDeploymentResourcePoolDedicatedResourcesToTerraform;
function vertexAiDeploymentResourcePoolDedicatedResourcesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        max_replica_count: {
            value: cdktf.numberToHclTerraform(struct.maxReplicaCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        min_replica_count: {
            value: cdktf.numberToHclTerraform(struct.minReplicaCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        autoscaling_metric_specs: {
            value: cdktf.listMapperHcl(vertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsToHclTerraform, true)(struct.autoscalingMetricSpecs),
            isBlock: true,
            type: "list",
            storageClassType: "VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsList",
        },
        machine_spec: {
            value: vertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecToHclTerraform(struct.machineSpec),
            isBlock: true,
            type: "list",
            storageClassType: "VertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiDeploymentResourcePoolDedicatedResourcesToHclTerraform = vertexAiDeploymentResourcePoolDedicatedResourcesToHclTerraform;
class VertexAiDeploymentResourcePoolDedicatedResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // autoscaling_metric_specs - computed: false, optional: true, required: false
        this._autoscalingMetricSpecs = new VertexAiDeploymentResourcePoolDedicatedResourcesAutoscalingMetricSpecsList(this, "autoscaling_metric_specs", false);
        // machine_spec - computed: false, optional: false, required: true
        this._machineSpec = new VertexAiDeploymentResourcePoolDedicatedResourcesMachineSpecOutputReference(this, "machine_spec");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxReplicaCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxReplicaCount = this._maxReplicaCount;
        }
        if (this._minReplicaCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.minReplicaCount = this._minReplicaCount;
        }
        if (this._autoscalingMetricSpecs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoscalingMetricSpecs = this._autoscalingMetricSpecs?.internalValue;
        }
        if (this._machineSpec?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineSpec = this._machineSpec?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxReplicaCount = undefined;
            this._minReplicaCount = undefined;
            this._autoscalingMetricSpecs.internalValue = undefined;
            this._machineSpec.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxReplicaCount = value.maxReplicaCount;
            this._minReplicaCount = value.minReplicaCount;
            this._autoscalingMetricSpecs.internalValue = value.autoscalingMetricSpecs;
            this._machineSpec.internalValue = value.machineSpec;
        }
    }
    get maxReplicaCount() {
        return this.getNumberAttribute('max_replica_count');
    }
    set maxReplicaCount(value) {
        this._maxReplicaCount = value;
    }
    resetMaxReplicaCount() {
        this._maxReplicaCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxReplicaCountInput() {
        return this._maxReplicaCount;
    }
    get minReplicaCount() {
        return this.getNumberAttribute('min_replica_count');
    }
    set minReplicaCount(value) {
        this._minReplicaCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get minReplicaCountInput() {
        return this._minReplicaCount;
    }
    get autoscalingMetricSpecs() {
        return this._autoscalingMetricSpecs;
    }
    putAutoscalingMetricSpecs(value) {
        this._autoscalingMetricSpecs.internalValue = value;
    }
    resetAutoscalingMetricSpecs() {
        this._autoscalingMetricSpecs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingMetricSpecsInput() {
        return this._autoscalingMetricSpecs.internalValue;
    }
    get machineSpec() {
        return this._machineSpec;
    }
    putMachineSpec(value) {
        this._machineSpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get machineSpecInput() {
        return this._machineSpec.internalValue;
    }
}
exports.VertexAiDeploymentResourcePoolDedicatedResourcesOutputReference = VertexAiDeploymentResourcePoolDedicatedResourcesOutputReference;
_d = JSII_RTTI_SYMBOL_1;
VertexAiDeploymentResourcePoolDedicatedResourcesOutputReference[_d] = { fqn: "@cdktf/provider-google.vertexAiDeploymentResourcePool.VertexAiDeploymentResourcePoolDedicatedResourcesOutputReference", version: "14.13.1" };
function vertexAiDeploymentResourcePoolTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.vertexAiDeploymentResourcePoolTimeoutsToTerraform = vertexAiDeploymentResourcePoolTimeoutsToTerraform;
function vertexAiDeploymentResourcePoolTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.vertexAiDeploymentResourcePoolTimeoutsToHclTerraform = vertexAiDeploymentResourcePoolTimeoutsToHclTerraform;
class VertexAiDeploymentResourcePoolTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.VertexAiDeploymentResourcePoolTimeoutsOutputReference = VertexAiDeploymentResourcePoolTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
VertexAiDeploymentResourcePoolTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-google.vertexAiDeploymentResourcePool.VertexAiDeploymentResourcePoolTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_deployment_resource_pool google_vertex_ai_deployment_resource_pool}
*/
class VertexAiDeploymentResourcePool extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a VertexAiDeploymentResourcePool resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the VertexAiDeploymentResourcePool to import
    * @param importFromId The id of the existing VertexAiDeploymentResourcePool that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_deployment_resource_pool#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the VertexAiDeploymentResourcePool to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_vertex_ai_deployment_resource_pool", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/vertex_ai_deployment_resource_pool google_vertex_ai_deployment_resource_pool} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VertexAiDeploymentResourcePoolConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_vertex_ai_deployment_resource_pool',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dedicated_resources - computed: false, optional: true, required: false
        this._dedicatedResources = new VertexAiDeploymentResourcePoolDedicatedResourcesOutputReference(this, "dedicated_resources");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VertexAiDeploymentResourcePoolTimeoutsOutputReference(this, "timeouts");
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
        this._dedicatedResources.internalValue = config.dedicatedResources;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get dedicatedResources() {
        return this._dedicatedResources;
    }
    putDedicatedResources(value) {
        this._dedicatedResources.internalValue = value;
    }
    resetDedicatedResources() {
        this._dedicatedResources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dedicatedResourcesInput() {
        return this._dedicatedResources.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            dedicated_resources: vertexAiDeploymentResourcePoolDedicatedResourcesToTerraform(this._dedicatedResources.internalValue),
            timeouts: vertexAiDeploymentResourcePoolTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            dedicated_resources: {
                value: vertexAiDeploymentResourcePoolDedicatedResourcesToHclTerraform(this._dedicatedResources.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "VertexAiDeploymentResourcePoolDedicatedResourcesList",
            },
            timeouts: {
                value: vertexAiDeploymentResourcePoolTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "VertexAiDeploymentResourcePoolTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.VertexAiDeploymentResourcePool = VertexAiDeploymentResourcePool;
_f = JSII_RTTI_SYMBOL_1;
VertexAiDeploymentResourcePool[_f] = { fqn: "@cdktf/provider-google.vertexAiDeploymentResourcePool.VertexAiDeploymentResourcePool", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
VertexAiDeploymentResourcePool.tfResourceType = "google_vertex_ai_deployment_resource_pool";
