// @cdktf/provider-google/lib/data-google-project-organization-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleProjectOrganizationPolicy = exports.DataGoogleProjectOrganizationPolicyRestorePolicyList = exports.DataGoogleProjectOrganizationPolicyRestorePolicyOutputReference = exports.dataGoogleProjectOrganizationPolicyRestorePolicyToHclTerraform = exports.dataGoogleProjectOrganizationPolicyRestorePolicyToTerraform = exports.DataGoogleProjectOrganizationPolicyListPolicyList = exports.DataGoogleProjectOrganizationPolicyListPolicyOutputReference = exports.dataGoogleProjectOrganizationPolicyListPolicyToHclTerraform = exports.dataGoogleProjectOrganizationPolicyListPolicyToTerraform = exports.DataGoogleProjectOrganizationPolicyListPolicyDenyList = exports.DataGoogleProjectOrganizationPolicyListPolicyDenyOutputReference = exports.dataGoogleProjectOrganizationPolicyListPolicyDenyToHclTerraform = exports.dataGoogleProjectOrganizationPolicyListPolicyDenyToTerraform = exports.DataGoogleProjectOrganizationPolicyListPolicyAllowList = exports.DataGoogleProjectOrganizationPolicyListPolicyAllowOutputReference = exports.dataGoogleProjectOrganizationPolicyListPolicyAllowToHclTerraform = exports.dataGoogleProjectOrganizationPolicyListPolicyAllowToTerraform = exports.DataGoogleProjectOrganizationPolicyBooleanPolicyList = exports.DataGoogleProjectOrganizationPolicyBooleanPolicyOutputReference = exports.dataGoogleProjectOrganizationPolicyBooleanPolicyToHclTerraform = exports.dataGoogleProjectOrganizationPolicyBooleanPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleProjectOrganizationPolicyBooleanPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleProjectOrganizationPolicyBooleanPolicyToTerraform = dataGoogleProjectOrganizationPolicyBooleanPolicyToTerraform;
function dataGoogleProjectOrganizationPolicyBooleanPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleProjectOrganizationPolicyBooleanPolicyToHclTerraform = dataGoogleProjectOrganizationPolicyBooleanPolicyToHclTerraform;
class DataGoogleProjectOrganizationPolicyBooleanPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enforced - computed: true, optional: false, required: false
    get enforced() {
        return this.getBooleanAttribute('enforced');
    }
}
exports.DataGoogleProjectOrganizationPolicyBooleanPolicyOutputReference = DataGoogleProjectOrganizationPolicyBooleanPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyBooleanPolicyOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyBooleanPolicyOutputReference", version: "14.13.1" };
class DataGoogleProjectOrganizationPolicyBooleanPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleProjectOrganizationPolicyBooleanPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleProjectOrganizationPolicyBooleanPolicyList = DataGoogleProjectOrganizationPolicyBooleanPolicyList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyBooleanPolicyList[_b] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyBooleanPolicyList", version: "14.13.1" };
function dataGoogleProjectOrganizationPolicyListPolicyAllowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleProjectOrganizationPolicyListPolicyAllowToTerraform = dataGoogleProjectOrganizationPolicyListPolicyAllowToTerraform;
function dataGoogleProjectOrganizationPolicyListPolicyAllowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleProjectOrganizationPolicyListPolicyAllowToHclTerraform = dataGoogleProjectOrganizationPolicyListPolicyAllowToHclTerraform;
class DataGoogleProjectOrganizationPolicyListPolicyAllowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all - computed: true, optional: false, required: false
    get all() {
        return this.getBooleanAttribute('all');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataGoogleProjectOrganizationPolicyListPolicyAllowOutputReference = DataGoogleProjectOrganizationPolicyListPolicyAllowOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyListPolicyAllowOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyListPolicyAllowOutputReference", version: "14.13.1" };
class DataGoogleProjectOrganizationPolicyListPolicyAllowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleProjectOrganizationPolicyListPolicyAllowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleProjectOrganizationPolicyListPolicyAllowList = DataGoogleProjectOrganizationPolicyListPolicyAllowList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyListPolicyAllowList[_d] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyListPolicyAllowList", version: "14.13.1" };
function dataGoogleProjectOrganizationPolicyListPolicyDenyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleProjectOrganizationPolicyListPolicyDenyToTerraform = dataGoogleProjectOrganizationPolicyListPolicyDenyToTerraform;
function dataGoogleProjectOrganizationPolicyListPolicyDenyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleProjectOrganizationPolicyListPolicyDenyToHclTerraform = dataGoogleProjectOrganizationPolicyListPolicyDenyToHclTerraform;
class DataGoogleProjectOrganizationPolicyListPolicyDenyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // all - computed: true, optional: false, required: false
    get all() {
        return this.getBooleanAttribute('all');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return cdktf.Fn.tolist(this.getListAttribute('values'));
    }
}
exports.DataGoogleProjectOrganizationPolicyListPolicyDenyOutputReference = DataGoogleProjectOrganizationPolicyListPolicyDenyOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyListPolicyDenyOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyListPolicyDenyOutputReference", version: "14.13.1" };
class DataGoogleProjectOrganizationPolicyListPolicyDenyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleProjectOrganizationPolicyListPolicyDenyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleProjectOrganizationPolicyListPolicyDenyList = DataGoogleProjectOrganizationPolicyListPolicyDenyList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyListPolicyDenyList[_f] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyListPolicyDenyList", version: "14.13.1" };
function dataGoogleProjectOrganizationPolicyListPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleProjectOrganizationPolicyListPolicyToTerraform = dataGoogleProjectOrganizationPolicyListPolicyToTerraform;
function dataGoogleProjectOrganizationPolicyListPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleProjectOrganizationPolicyListPolicyToHclTerraform = dataGoogleProjectOrganizationPolicyListPolicyToHclTerraform;
class DataGoogleProjectOrganizationPolicyListPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // allow - computed: true, optional: false, required: false
        this._allow = new DataGoogleProjectOrganizationPolicyListPolicyAllowList(this, "allow", false);
        // deny - computed: true, optional: false, required: false
        this._deny = new DataGoogleProjectOrganizationPolicyListPolicyDenyList(this, "deny", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get allow() {
        return this._allow;
    }
    get deny() {
        return this._deny;
    }
    // inherit_from_parent - computed: true, optional: false, required: false
    get inheritFromParent() {
        return this.getBooleanAttribute('inherit_from_parent');
    }
    // suggested_value - computed: true, optional: false, required: false
    get suggestedValue() {
        return this.getStringAttribute('suggested_value');
    }
}
exports.DataGoogleProjectOrganizationPolicyListPolicyOutputReference = DataGoogleProjectOrganizationPolicyListPolicyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyListPolicyOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyListPolicyOutputReference", version: "14.13.1" };
class DataGoogleProjectOrganizationPolicyListPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleProjectOrganizationPolicyListPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleProjectOrganizationPolicyListPolicyList = DataGoogleProjectOrganizationPolicyListPolicyList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyListPolicyList[_h] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyListPolicyList", version: "14.13.1" };
function dataGoogleProjectOrganizationPolicyRestorePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleProjectOrganizationPolicyRestorePolicyToTerraform = dataGoogleProjectOrganizationPolicyRestorePolicyToTerraform;
function dataGoogleProjectOrganizationPolicyRestorePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleProjectOrganizationPolicyRestorePolicyToHclTerraform = dataGoogleProjectOrganizationPolicyRestorePolicyToHclTerraform;
class DataGoogleProjectOrganizationPolicyRestorePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // default - computed: true, optional: false, required: false
    get default() {
        return this.getBooleanAttribute('default');
    }
}
exports.DataGoogleProjectOrganizationPolicyRestorePolicyOutputReference = DataGoogleProjectOrganizationPolicyRestorePolicyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyRestorePolicyOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyRestorePolicyOutputReference", version: "14.13.1" };
class DataGoogleProjectOrganizationPolicyRestorePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleProjectOrganizationPolicyRestorePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleProjectOrganizationPolicyRestorePolicyList = DataGoogleProjectOrganizationPolicyRestorePolicyList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicyRestorePolicyList[_k] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicyRestorePolicyList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/project_organization_policy google_project_organization_policy}
*/
class DataGoogleProjectOrganizationPolicy extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleProjectOrganizationPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleProjectOrganizationPolicy to import
    * @param importFromId The id of the existing DataGoogleProjectOrganizationPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/project_organization_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleProjectOrganizationPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_project_organization_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/project_organization_policy google_project_organization_policy} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleProjectOrganizationPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_project_organization_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // boolean_policy - computed: true, optional: false, required: false
        this._booleanPolicy = new DataGoogleProjectOrganizationPolicyBooleanPolicyList(this, "boolean_policy", false);
        // list_policy - computed: true, optional: false, required: false
        this._listPolicy = new DataGoogleProjectOrganizationPolicyListPolicyList(this, "list_policy", false);
        // restore_policy - computed: true, optional: false, required: false
        this._restorePolicy = new DataGoogleProjectOrganizationPolicyRestorePolicyList(this, "restore_policy", false);
        this._constraint = config.constraint;
        this._id = config.id;
        this._project = config.project;
    }
    get booleanPolicy() {
        return this._booleanPolicy;
    }
    get constraint() {
        return this.getStringAttribute('constraint');
    }
    set constraint(value) {
        this._constraint = value;
    }
    // Temporarily expose input value. Use with caution.
    get constraintInput() {
        return this._constraint;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get listPolicy() {
        return this._listPolicy;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get restorePolicy() {
        return this._restorePolicy;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            constraint: cdktf.stringToTerraform(this._constraint),
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            constraint: {
                value: cdktf.stringToHclTerraform(this._constraint),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleProjectOrganizationPolicy = DataGoogleProjectOrganizationPolicy;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleProjectOrganizationPolicy[_l] = { fqn: "@cdktf/provider-google.dataGoogleProjectOrganizationPolicy.DataGoogleProjectOrganizationPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleProjectOrganizationPolicy.tfResourceType = "google_project_organization_policy";
