// @cdktf/provider-google/lib/dataproc-gdc-application-environment/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataprocGdcApplicationEnvironment = exports.DataprocGdcApplicationEnvironmentTimeoutsOutputReference = exports.dataprocGdcApplicationEnvironmentTimeoutsToHclTerraform = exports.dataprocGdcApplicationEnvironmentTimeoutsToTerraform = exports.DataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigOutputReference = exports.dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToHclTerraform = exports.dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default_properties: cdktf.hashMapper(cdktf.stringToTerraform)(struct.defaultProperties),
        default_version: cdktf.stringToTerraform(struct.defaultVersion),
    };
}
exports.dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToTerraform = dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToTerraform;
function dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        default_properties: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.defaultProperties),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        default_version: {
            value: cdktf.stringToHclTerraform(struct.defaultVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToHclTerraform = dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToHclTerraform;
class DataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._defaultProperties !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultProperties = this._defaultProperties;
        }
        if (this._defaultVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.defaultVersion = this._defaultVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._defaultProperties = undefined;
            this._defaultVersion = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._defaultProperties = value.defaultProperties;
            this._defaultVersion = value.defaultVersion;
        }
    }
    get defaultProperties() {
        return this.getStringMapAttribute('default_properties');
    }
    set defaultProperties(value) {
        this._defaultProperties = value;
    }
    resetDefaultProperties() {
        this._defaultProperties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultPropertiesInput() {
        return this._defaultProperties;
    }
    get defaultVersion() {
        return this.getStringAttribute('default_version');
    }
    set defaultVersion(value) {
        this._defaultVersion = value;
    }
    resetDefaultVersion() {
        this._defaultVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultVersionInput() {
        return this._defaultVersion;
    }
}
exports.DataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigOutputReference = DataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigOutputReference[_a] = { fqn: "@cdktf/provider-google.dataprocGdcApplicationEnvironment.DataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigOutputReference", version: "14.13.1" };
function dataprocGdcApplicationEnvironmentTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dataprocGdcApplicationEnvironmentTimeoutsToTerraform = dataprocGdcApplicationEnvironmentTimeoutsToTerraform;
function dataprocGdcApplicationEnvironmentTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataprocGdcApplicationEnvironmentTimeoutsToHclTerraform = dataprocGdcApplicationEnvironmentTimeoutsToHclTerraform;
class DataprocGdcApplicationEnvironmentTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DataprocGdcApplicationEnvironmentTimeoutsOutputReference = DataprocGdcApplicationEnvironmentTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DataprocGdcApplicationEnvironmentTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-google.dataprocGdcApplicationEnvironment.DataprocGdcApplicationEnvironmentTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataproc_gdc_application_environment google_dataproc_gdc_application_environment}
*/
class DataprocGdcApplicationEnvironment extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataprocGdcApplicationEnvironment resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataprocGdcApplicationEnvironment to import
    * @param importFromId The id of the existing DataprocGdcApplicationEnvironment that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataproc_gdc_application_environment#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataprocGdcApplicationEnvironment to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_dataproc_gdc_application_environment", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dataproc_gdc_application_environment google_dataproc_gdc_application_environment} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataprocGdcApplicationEnvironmentConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_dataproc_gdc_application_environment',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_annotations - computed: true, optional: false, required: false
        this._effectiveAnnotations = new cdktf.StringMap(this, "effective_annotations");
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // spark_application_environment_config - computed: false, optional: true, required: false
        this._sparkApplicationEnvironmentConfig = new DataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigOutputReference(this, "spark_application_environment_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataprocGdcApplicationEnvironmentTimeoutsOutputReference(this, "timeouts");
        this._annotations = config.annotations;
        this._applicationEnvironmentId = config.applicationEnvironmentId;
        this._displayName = config.displayName;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._namespace = config.namespace;
        this._project = config.project;
        this._serviceinstance = config.serviceinstance;
        this._sparkApplicationEnvironmentConfig.internalValue = config.sparkApplicationEnvironmentConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    get annotations() {
        return this.getStringMapAttribute('annotations');
    }
    set annotations(value) {
        this._annotations = value;
    }
    resetAnnotations() {
        this._annotations = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get annotationsInput() {
        return this._annotations;
    }
    get applicationEnvironmentId() {
        return this.getStringAttribute('application_environment_id');
    }
    set applicationEnvironmentId(value) {
        this._applicationEnvironmentId = value;
    }
    resetApplicationEnvironmentId() {
        this._applicationEnvironmentId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get applicationEnvironmentIdInput() {
        return this._applicationEnvironmentId;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get effectiveAnnotations() {
        return this._effectiveAnnotations;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    resetNamespace() {
        this._namespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get serviceinstance() {
        return this.getStringAttribute('serviceinstance');
    }
    set serviceinstance(value) {
        this._serviceinstance = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceinstanceInput() {
        return this._serviceinstance;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // uid - computed: true, optional: false, required: false
    get uid() {
        return this.getStringAttribute('uid');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get sparkApplicationEnvironmentConfig() {
        return this._sparkApplicationEnvironmentConfig;
    }
    putSparkApplicationEnvironmentConfig(value) {
        this._sparkApplicationEnvironmentConfig.internalValue = value;
    }
    resetSparkApplicationEnvironmentConfig() {
        this._sparkApplicationEnvironmentConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sparkApplicationEnvironmentConfigInput() {
        return this._sparkApplicationEnvironmentConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            annotations: cdktf.hashMapper(cdktf.stringToTerraform)(this._annotations),
            application_environment_id: cdktf.stringToTerraform(this._applicationEnvironmentId),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            namespace: cdktf.stringToTerraform(this._namespace),
            project: cdktf.stringToTerraform(this._project),
            serviceinstance: cdktf.stringToTerraform(this._serviceinstance),
            spark_application_environment_config: dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToTerraform(this._sparkApplicationEnvironmentConfig.internalValue),
            timeouts: dataprocGdcApplicationEnvironmentTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            annotations: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._annotations),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            application_environment_id: {
                value: cdktf.stringToHclTerraform(this._applicationEnvironmentId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            namespace: {
                value: cdktf.stringToHclTerraform(this._namespace),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            serviceinstance: {
                value: cdktf.stringToHclTerraform(this._serviceinstance),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            spark_application_environment_config: {
                value: dataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigToHclTerraform(this._sparkApplicationEnvironmentConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataprocGdcApplicationEnvironmentSparkApplicationEnvironmentConfigList",
            },
            timeouts: {
                value: dataprocGdcApplicationEnvironmentTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataprocGdcApplicationEnvironmentTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataprocGdcApplicationEnvironment = DataprocGdcApplicationEnvironment;
_c = JSII_RTTI_SYMBOL_1;
DataprocGdcApplicationEnvironment[_c] = { fqn: "@cdktf/provider-google.dataprocGdcApplicationEnvironment.DataprocGdcApplicationEnvironment", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataprocGdcApplicationEnvironment.tfResourceType = "google_dataproc_gdc_application_environment";
