// @cdktf/provider-google/lib/data-google-oracle-database-cloud-vm-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleOracleDatabaseCloudVmCluster = exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesList = exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesOutputReference = exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesToHclTerraform = exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesToTerraform = exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneList = exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneOutputReference = exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneToHclTerraform = exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneToTerraform = exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsList = exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsOutputReference = exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsToHclTerraform = exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsToTerraform = dataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsToTerraform;
function dataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsToHclTerraform = dataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsToHclTerraform;
class DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // diagnostics_events_enabled - computed: true, optional: false, required: false
    get diagnosticsEventsEnabled() {
        return this.getBooleanAttribute('diagnostics_events_enabled');
    }
    // health_monitoring_enabled - computed: true, optional: false, required: false
    get healthMonitoringEnabled() {
        return this.getBooleanAttribute('health_monitoring_enabled');
    }
    // incident_logs_enabled - computed: true, optional: false, required: false
    get incidentLogsEnabled() {
        return this.getBooleanAttribute('incident_logs_enabled');
    }
}
exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsOutputReference = DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseCloudVmCluster.DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsList = DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseCloudVmCluster.DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsList", version: "14.13.1" };
function dataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneToTerraform = dataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneToTerraform;
function dataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneToHclTerraform = dataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneToHclTerraform;
class DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
}
exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneOutputReference = DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseCloudVmCluster.DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneList = DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneList[_d] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseCloudVmCluster.DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneList", version: "14.13.1" };
function dataGoogleOracleDatabaseCloudVmClusterPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesToTerraform = dataGoogleOracleDatabaseCloudVmClusterPropertiesToTerraform;
function dataGoogleOracleDatabaseCloudVmClusterPropertiesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseCloudVmClusterPropertiesToHclTerraform = dataGoogleOracleDatabaseCloudVmClusterPropertiesToHclTerraform;
class DataGoogleOracleDatabaseCloudVmClusterPropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // diagnostics_data_collection_options - computed: true, optional: false, required: false
        this._diagnosticsDataCollectionOptions = new DataGoogleOracleDatabaseCloudVmClusterPropertiesDiagnosticsDataCollectionOptionsList(this, "diagnostics_data_collection_options", false);
        // time_zone - computed: true, optional: false, required: false
        this._timeZone = new DataGoogleOracleDatabaseCloudVmClusterPropertiesTimeZoneList(this, "time_zone", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cluster_name - computed: true, optional: false, required: false
    get clusterName() {
        return this.getStringAttribute('cluster_name');
    }
    // compartment_id - computed: true, optional: false, required: false
    get compartmentId() {
        return this.getStringAttribute('compartment_id');
    }
    // cpu_core_count - computed: true, optional: false, required: false
    get cpuCoreCount() {
        return this.getNumberAttribute('cpu_core_count');
    }
    // data_storage_size_tb - computed: true, optional: false, required: false
    get dataStorageSizeTb() {
        return this.getNumberAttribute('data_storage_size_tb');
    }
    // db_node_storage_size_gb - computed: true, optional: false, required: false
    get dbNodeStorageSizeGb() {
        return this.getNumberAttribute('db_node_storage_size_gb');
    }
    // db_server_ocids - computed: true, optional: false, required: false
    get dbServerOcids() {
        return this.getListAttribute('db_server_ocids');
    }
    get diagnosticsDataCollectionOptions() {
        return this._diagnosticsDataCollectionOptions;
    }
    // disk_redundancy - computed: true, optional: false, required: false
    get diskRedundancy() {
        return this.getStringAttribute('disk_redundancy');
    }
    // dns_listener_ip - computed: true, optional: false, required: false
    get dnsListenerIp() {
        return this.getStringAttribute('dns_listener_ip');
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
    // gi_version - computed: true, optional: false, required: false
    get giVersion() {
        return this.getStringAttribute('gi_version');
    }
    // hostname - computed: true, optional: false, required: false
    get hostname() {
        return this.getStringAttribute('hostname');
    }
    // hostname_prefix - computed: true, optional: false, required: false
    get hostnamePrefix() {
        return this.getStringAttribute('hostname_prefix');
    }
    // license_type - computed: true, optional: false, required: false
    get licenseType() {
        return this.getStringAttribute('license_type');
    }
    // local_backup_enabled - computed: true, optional: false, required: false
    get localBackupEnabled() {
        return this.getBooleanAttribute('local_backup_enabled');
    }
    // memory_size_gb - computed: true, optional: false, required: false
    get memorySizeGb() {
        return this.getNumberAttribute('memory_size_gb');
    }
    // node_count - computed: true, optional: false, required: false
    get nodeCount() {
        return this.getNumberAttribute('node_count');
    }
    // oci_url - computed: true, optional: false, required: false
    get ociUrl() {
        return this.getStringAttribute('oci_url');
    }
    // ocid - computed: true, optional: false, required: false
    get ocid() {
        return this.getStringAttribute('ocid');
    }
    // ocpu_count - computed: true, optional: false, required: false
    get ocpuCount() {
        return this.getNumberAttribute('ocpu_count');
    }
    // scan_dns - computed: true, optional: false, required: false
    get scanDns() {
        return this.getStringAttribute('scan_dns');
    }
    // scan_dns_record_id - computed: true, optional: false, required: false
    get scanDnsRecordId() {
        return this.getStringAttribute('scan_dns_record_id');
    }
    // scan_ip_ids - computed: true, optional: false, required: false
    get scanIpIds() {
        return this.getListAttribute('scan_ip_ids');
    }
    // scan_listener_port_tcp - computed: true, optional: false, required: false
    get scanListenerPortTcp() {
        return this.getNumberAttribute('scan_listener_port_tcp');
    }
    // scan_listener_port_tcp_ssl - computed: true, optional: false, required: false
    get scanListenerPortTcpSsl() {
        return this.getNumberAttribute('scan_listener_port_tcp_ssl');
    }
    // shape - computed: true, optional: false, required: false
    get shape() {
        return this.getStringAttribute('shape');
    }
    // sparse_diskgroup_enabled - computed: true, optional: false, required: false
    get sparseDiskgroupEnabled() {
        return this.getBooleanAttribute('sparse_diskgroup_enabled');
    }
    // ssh_public_keys - computed: true, optional: false, required: false
    get sshPublicKeys() {
        return this.getListAttribute('ssh_public_keys');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // storage_size_gb - computed: true, optional: false, required: false
    get storageSizeGb() {
        return this.getNumberAttribute('storage_size_gb');
    }
    // system_version - computed: true, optional: false, required: false
    get systemVersion() {
        return this.getStringAttribute('system_version');
    }
    get timeZone() {
        return this._timeZone;
    }
}
exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesOutputReference = DataGoogleOracleDatabaseCloudVmClusterPropertiesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseCloudVmClusterPropertiesOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseCloudVmCluster.DataGoogleOracleDatabaseCloudVmClusterPropertiesOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseCloudVmClusterPropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseCloudVmClusterPropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseCloudVmClusterPropertiesList = DataGoogleOracleDatabaseCloudVmClusterPropertiesList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseCloudVmClusterPropertiesList[_f] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseCloudVmCluster.DataGoogleOracleDatabaseCloudVmClusterPropertiesList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_cloud_vm_cluster google_oracle_database_cloud_vm_cluster}
*/
class DataGoogleOracleDatabaseCloudVmCluster extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleOracleDatabaseCloudVmCluster resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleOracleDatabaseCloudVmCluster to import
    * @param importFromId The id of the existing DataGoogleOracleDatabaseCloudVmCluster that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_cloud_vm_cluster#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleOracleDatabaseCloudVmCluster to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_oracle_database_cloud_vm_cluster", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_cloud_vm_cluster google_oracle_database_cloud_vm_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleOracleDatabaseCloudVmClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_oracle_database_cloud_vm_cluster',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // properties - computed: true, optional: false, required: false
        this._properties = new DataGoogleOracleDatabaseCloudVmClusterPropertiesList(this, "properties", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._cloudVmClusterId = config.cloudVmClusterId;
        this._id = config.id;
        this._location = config.location;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // backup_subnet_cidr - computed: true, optional: false, required: false
    get backupSubnetCidr() {
        return this.getStringAttribute('backup_subnet_cidr');
    }
    // cidr - computed: true, optional: false, required: false
    get cidr() {
        return this.getStringAttribute('cidr');
    }
    get cloudVmClusterId() {
        return this.getStringAttribute('cloud_vm_cluster_id');
    }
    set cloudVmClusterId(value) {
        this._cloudVmClusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get cloudVmClusterIdInput() {
        return this._cloudVmClusterId;
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // exadata_infrastructure - computed: true, optional: false, required: false
    get exadataInfrastructure() {
        return this.getStringAttribute('exadata_infrastructure');
    }
    // gcp_oracle_zone - computed: true, optional: false, required: false
    get gcpOracleZone() {
        return this.getStringAttribute('gcp_oracle_zone');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get properties() {
        return this._properties;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cloud_vm_cluster_id: cdktf.stringToTerraform(this._cloudVmClusterId),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cloud_vm_cluster_id: {
                value: cdktf.stringToHclTerraform(this._cloudVmClusterId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleOracleDatabaseCloudVmCluster = DataGoogleOracleDatabaseCloudVmCluster;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseCloudVmCluster[_g] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseCloudVmCluster.DataGoogleOracleDatabaseCloudVmCluster", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleOracleDatabaseCloudVmCluster.tfResourceType = "google_oracle_database_cloud_vm_cluster";
