// @cdktf/provider-google/lib/scc-project-custom-module/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SccProjectCustomModule = exports.SccProjectCustomModuleTimeoutsOutputReference = exports.sccProjectCustomModuleTimeoutsToHclTerraform = exports.sccProjectCustomModuleTimeoutsToTerraform = exports.SccProjectCustomModuleCustomConfigOutputReference = exports.sccProjectCustomModuleCustomConfigToHclTerraform = exports.sccProjectCustomModuleCustomConfigToTerraform = exports.SccProjectCustomModuleCustomConfigResourceSelectorOutputReference = exports.sccProjectCustomModuleCustomConfigResourceSelectorToHclTerraform = exports.sccProjectCustomModuleCustomConfigResourceSelectorToTerraform = exports.SccProjectCustomModuleCustomConfigPredicateOutputReference = exports.sccProjectCustomModuleCustomConfigPredicateToHclTerraform = exports.sccProjectCustomModuleCustomConfigPredicateToTerraform = exports.SccProjectCustomModuleCustomConfigCustomOutputOutputReference = exports.sccProjectCustomModuleCustomConfigCustomOutputToHclTerraform = exports.sccProjectCustomModuleCustomConfigCustomOutputToTerraform = exports.SccProjectCustomModuleCustomConfigCustomOutputPropertiesList = exports.SccProjectCustomModuleCustomConfigCustomOutputPropertiesOutputReference = exports.sccProjectCustomModuleCustomConfigCustomOutputPropertiesToHclTerraform = exports.sccProjectCustomModuleCustomConfigCustomOutputPropertiesToTerraform = exports.SccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionOutputReference = exports.sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToHclTerraform = exports.sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        expression: cdktf.stringToTerraform(struct.expression),
        location: cdktf.stringToTerraform(struct.location),
        title: cdktf.stringToTerraform(struct.title),
    };
}
exports.sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToTerraform = sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToTerraform;
function sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        location: {
            value: cdktf.stringToHclTerraform(struct.location),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        title: {
            value: cdktf.stringToHclTerraform(struct.title),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToHclTerraform = sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToHclTerraform;
class SccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._title !== undefined) {
            hasAnyValues = true;
            internalValueResult.title = this._title;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._expression = undefined;
            this._location = undefined;
            this._title = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._expression = value.expression;
            this._location = value.location;
            this._title = value.title;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get title() {
        return this.getStringAttribute('title');
    }
    set title(value) {
        this._title = value;
    }
    resetTitle() {
        this._title = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get titleInput() {
        return this._title;
    }
}
exports.SccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionOutputReference = SccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
SccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionOutputReference[_a] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionOutputReference", version: "14.13.1" };
function sccProjectCustomModuleCustomConfigCustomOutputPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value_expression: sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToTerraform(struct.valueExpression),
    };
}
exports.sccProjectCustomModuleCustomConfigCustomOutputPropertiesToTerraform = sccProjectCustomModuleCustomConfigCustomOutputPropertiesToTerraform;
function sccProjectCustomModuleCustomConfigCustomOutputPropertiesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value_expression: {
            value: sccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionToHclTerraform(struct.valueExpression),
            isBlock: true,
            type: "list",
            storageClassType: "SccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sccProjectCustomModuleCustomConfigCustomOutputPropertiesToHclTerraform = sccProjectCustomModuleCustomConfigCustomOutputPropertiesToHclTerraform;
class SccProjectCustomModuleCustomConfigCustomOutputPropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // value_expression - computed: false, optional: true, required: false
        this._valueExpression = new SccProjectCustomModuleCustomConfigCustomOutputPropertiesValueExpressionOutputReference(this, "value_expression");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._valueExpression?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.valueExpression = this._valueExpression?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._valueExpression.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._valueExpression.internalValue = value.valueExpression;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get valueExpression() {
        return this._valueExpression;
    }
    putValueExpression(value) {
        this._valueExpression.internalValue = value;
    }
    resetValueExpression() {
        this._valueExpression.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueExpressionInput() {
        return this._valueExpression.internalValue;
    }
}
exports.SccProjectCustomModuleCustomConfigCustomOutputPropertiesOutputReference = SccProjectCustomModuleCustomConfigCustomOutputPropertiesOutputReference;
_b = JSII_RTTI_SYMBOL_1;
SccProjectCustomModuleCustomConfigCustomOutputPropertiesOutputReference[_b] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModuleCustomConfigCustomOutputPropertiesOutputReference", version: "14.13.1" };
class SccProjectCustomModuleCustomConfigCustomOutputPropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new SccProjectCustomModuleCustomConfigCustomOutputPropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.SccProjectCustomModuleCustomConfigCustomOutputPropertiesList = SccProjectCustomModuleCustomConfigCustomOutputPropertiesList;
_c = JSII_RTTI_SYMBOL_1;
SccProjectCustomModuleCustomConfigCustomOutputPropertiesList[_c] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModuleCustomConfigCustomOutputPropertiesList", version: "14.13.1" };
function sccProjectCustomModuleCustomConfigCustomOutputToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        properties: cdktf.listMapper(sccProjectCustomModuleCustomConfigCustomOutputPropertiesToTerraform, true)(struct.properties),
    };
}
exports.sccProjectCustomModuleCustomConfigCustomOutputToTerraform = sccProjectCustomModuleCustomConfigCustomOutputToTerraform;
function sccProjectCustomModuleCustomConfigCustomOutputToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        properties: {
            value: cdktf.listMapperHcl(sccProjectCustomModuleCustomConfigCustomOutputPropertiesToHclTerraform, true)(struct.properties),
            isBlock: true,
            type: "list",
            storageClassType: "SccProjectCustomModuleCustomConfigCustomOutputPropertiesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sccProjectCustomModuleCustomConfigCustomOutputToHclTerraform = sccProjectCustomModuleCustomConfigCustomOutputToHclTerraform;
class SccProjectCustomModuleCustomConfigCustomOutputOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // properties - computed: false, optional: true, required: false
        this._properties = new SccProjectCustomModuleCustomConfigCustomOutputPropertiesList(this, "properties", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._properties?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.properties = this._properties?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._properties.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._properties.internalValue = value.properties;
        }
    }
    get properties() {
        return this._properties;
    }
    putProperties(value) {
        this._properties.internalValue = value;
    }
    resetProperties() {
        this._properties.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propertiesInput() {
        return this._properties.internalValue;
    }
}
exports.SccProjectCustomModuleCustomConfigCustomOutputOutputReference = SccProjectCustomModuleCustomConfigCustomOutputOutputReference;
_d = JSII_RTTI_SYMBOL_1;
SccProjectCustomModuleCustomConfigCustomOutputOutputReference[_d] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModuleCustomConfigCustomOutputOutputReference", version: "14.13.1" };
function sccProjectCustomModuleCustomConfigPredicateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        expression: cdktf.stringToTerraform(struct.expression),
        location: cdktf.stringToTerraform(struct.location),
        title: cdktf.stringToTerraform(struct.title),
    };
}
exports.sccProjectCustomModuleCustomConfigPredicateToTerraform = sccProjectCustomModuleCustomConfigPredicateToTerraform;
function sccProjectCustomModuleCustomConfigPredicateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        location: {
            value: cdktf.stringToHclTerraform(struct.location),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        title: {
            value: cdktf.stringToHclTerraform(struct.title),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sccProjectCustomModuleCustomConfigPredicateToHclTerraform = sccProjectCustomModuleCustomConfigPredicateToHclTerraform;
class SccProjectCustomModuleCustomConfigPredicateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._title !== undefined) {
            hasAnyValues = true;
            internalValueResult.title = this._title;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._expression = undefined;
            this._location = undefined;
            this._title = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._expression = value.expression;
            this._location = value.location;
            this._title = value.title;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get title() {
        return this.getStringAttribute('title');
    }
    set title(value) {
        this._title = value;
    }
    resetTitle() {
        this._title = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get titleInput() {
        return this._title;
    }
}
exports.SccProjectCustomModuleCustomConfigPredicateOutputReference = SccProjectCustomModuleCustomConfigPredicateOutputReference;
_e = JSII_RTTI_SYMBOL_1;
SccProjectCustomModuleCustomConfigPredicateOutputReference[_e] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModuleCustomConfigPredicateOutputReference", version: "14.13.1" };
function sccProjectCustomModuleCustomConfigResourceSelectorToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resource_types: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.resourceTypes),
    };
}
exports.sccProjectCustomModuleCustomConfigResourceSelectorToTerraform = sccProjectCustomModuleCustomConfigResourceSelectorToTerraform;
function sccProjectCustomModuleCustomConfigResourceSelectorToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        resource_types: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.resourceTypes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sccProjectCustomModuleCustomConfigResourceSelectorToHclTerraform = sccProjectCustomModuleCustomConfigResourceSelectorToHclTerraform;
class SccProjectCustomModuleCustomConfigResourceSelectorOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resourceTypes !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceTypes = this._resourceTypes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resourceTypes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resourceTypes = value.resourceTypes;
        }
    }
    get resourceTypes() {
        return this.getListAttribute('resource_types');
    }
    set resourceTypes(value) {
        this._resourceTypes = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceTypesInput() {
        return this._resourceTypes;
    }
}
exports.SccProjectCustomModuleCustomConfigResourceSelectorOutputReference = SccProjectCustomModuleCustomConfigResourceSelectorOutputReference;
_f = JSII_RTTI_SYMBOL_1;
SccProjectCustomModuleCustomConfigResourceSelectorOutputReference[_f] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModuleCustomConfigResourceSelectorOutputReference", version: "14.13.1" };
function sccProjectCustomModuleCustomConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        recommendation: cdktf.stringToTerraform(struct.recommendation),
        severity: cdktf.stringToTerraform(struct.severity),
        custom_output: sccProjectCustomModuleCustomConfigCustomOutputToTerraform(struct.customOutput),
        predicate: sccProjectCustomModuleCustomConfigPredicateToTerraform(struct.predicate),
        resource_selector: sccProjectCustomModuleCustomConfigResourceSelectorToTerraform(struct.resourceSelector),
    };
}
exports.sccProjectCustomModuleCustomConfigToTerraform = sccProjectCustomModuleCustomConfigToTerraform;
function sccProjectCustomModuleCustomConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        recommendation: {
            value: cdktf.stringToHclTerraform(struct.recommendation),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        severity: {
            value: cdktf.stringToHclTerraform(struct.severity),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        custom_output: {
            value: sccProjectCustomModuleCustomConfigCustomOutputToHclTerraform(struct.customOutput),
            isBlock: true,
            type: "list",
            storageClassType: "SccProjectCustomModuleCustomConfigCustomOutputList",
        },
        predicate: {
            value: sccProjectCustomModuleCustomConfigPredicateToHclTerraform(struct.predicate),
            isBlock: true,
            type: "list",
            storageClassType: "SccProjectCustomModuleCustomConfigPredicateList",
        },
        resource_selector: {
            value: sccProjectCustomModuleCustomConfigResourceSelectorToHclTerraform(struct.resourceSelector),
            isBlock: true,
            type: "list",
            storageClassType: "SccProjectCustomModuleCustomConfigResourceSelectorList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sccProjectCustomModuleCustomConfigToHclTerraform = sccProjectCustomModuleCustomConfigToHclTerraform;
class SccProjectCustomModuleCustomConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // custom_output - computed: false, optional: true, required: false
        this._customOutput = new SccProjectCustomModuleCustomConfigCustomOutputOutputReference(this, "custom_output");
        // predicate - computed: false, optional: false, required: true
        this._predicate = new SccProjectCustomModuleCustomConfigPredicateOutputReference(this, "predicate");
        // resource_selector - computed: false, optional: false, required: true
        this._resourceSelector = new SccProjectCustomModuleCustomConfigResourceSelectorOutputReference(this, "resource_selector");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._recommendation !== undefined) {
            hasAnyValues = true;
            internalValueResult.recommendation = this._recommendation;
        }
        if (this._severity !== undefined) {
            hasAnyValues = true;
            internalValueResult.severity = this._severity;
        }
        if (this._customOutput?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.customOutput = this._customOutput?.internalValue;
        }
        if (this._predicate?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.predicate = this._predicate?.internalValue;
        }
        if (this._resourceSelector?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourceSelector = this._resourceSelector?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._recommendation = undefined;
            this._severity = undefined;
            this._customOutput.internalValue = undefined;
            this._predicate.internalValue = undefined;
            this._resourceSelector.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._recommendation = value.recommendation;
            this._severity = value.severity;
            this._customOutput.internalValue = value.customOutput;
            this._predicate.internalValue = value.predicate;
            this._resourceSelector.internalValue = value.resourceSelector;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get recommendation() {
        return this.getStringAttribute('recommendation');
    }
    set recommendation(value) {
        this._recommendation = value;
    }
    // Temporarily expose input value. Use with caution.
    get recommendationInput() {
        return this._recommendation;
    }
    get severity() {
        return this.getStringAttribute('severity');
    }
    set severity(value) {
        this._severity = value;
    }
    // Temporarily expose input value. Use with caution.
    get severityInput() {
        return this._severity;
    }
    get customOutput() {
        return this._customOutput;
    }
    putCustomOutput(value) {
        this._customOutput.internalValue = value;
    }
    resetCustomOutput() {
        this._customOutput.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customOutputInput() {
        return this._customOutput.internalValue;
    }
    get predicate() {
        return this._predicate;
    }
    putPredicate(value) {
        this._predicate.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get predicateInput() {
        return this._predicate.internalValue;
    }
    get resourceSelector() {
        return this._resourceSelector;
    }
    putResourceSelector(value) {
        this._resourceSelector.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourceSelectorInput() {
        return this._resourceSelector.internalValue;
    }
}
exports.SccProjectCustomModuleCustomConfigOutputReference = SccProjectCustomModuleCustomConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
SccProjectCustomModuleCustomConfigOutputReference[_g] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModuleCustomConfigOutputReference", version: "14.13.1" };
function sccProjectCustomModuleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.sccProjectCustomModuleTimeoutsToTerraform = sccProjectCustomModuleTimeoutsToTerraform;
function sccProjectCustomModuleTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.sccProjectCustomModuleTimeoutsToHclTerraform = sccProjectCustomModuleTimeoutsToHclTerraform;
class SccProjectCustomModuleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.SccProjectCustomModuleTimeoutsOutputReference = SccProjectCustomModuleTimeoutsOutputReference;
_h = JSII_RTTI_SYMBOL_1;
SccProjectCustomModuleTimeoutsOutputReference[_h] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModuleTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/scc_project_custom_module google_scc_project_custom_module}
*/
class SccProjectCustomModule extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a SccProjectCustomModule resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the SccProjectCustomModule to import
    * @param importFromId The id of the existing SccProjectCustomModule that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/scc_project_custom_module#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the SccProjectCustomModule to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_scc_project_custom_module", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/scc_project_custom_module google_scc_project_custom_module} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options SccProjectCustomModuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_scc_project_custom_module',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // custom_config - computed: false, optional: false, required: true
        this._customConfig = new SccProjectCustomModuleCustomConfigOutputReference(this, "custom_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new SccProjectCustomModuleTimeoutsOutputReference(this, "timeouts");
        this._displayName = config.displayName;
        this._enablementState = config.enablementState;
        this._id = config.id;
        this._project = config.project;
        this._customConfig.internalValue = config.customConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // ancestor_module - computed: true, optional: false, required: false
    get ancestorModule() {
        return this.getStringAttribute('ancestor_module');
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get enablementState() {
        return this.getStringAttribute('enablement_state');
    }
    set enablementState(value) {
        this._enablementState = value;
    }
    // Temporarily expose input value. Use with caution.
    get enablementStateInput() {
        return this._enablementState;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_editor - computed: true, optional: false, required: false
    get lastEditor() {
        return this.getStringAttribute('last_editor');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get customConfig() {
        return this._customConfig;
    }
    putCustomConfig(value) {
        this._customConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get customConfigInput() {
        return this._customConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            display_name: cdktf.stringToTerraform(this._displayName),
            enablement_state: cdktf.stringToTerraform(this._enablementState),
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
            custom_config: sccProjectCustomModuleCustomConfigToTerraform(this._customConfig.internalValue),
            timeouts: sccProjectCustomModuleTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            enablement_state: {
                value: cdktf.stringToHclTerraform(this._enablementState),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            custom_config: {
                value: sccProjectCustomModuleCustomConfigToHclTerraform(this._customConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "SccProjectCustomModuleCustomConfigList",
            },
            timeouts: {
                value: sccProjectCustomModuleTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "SccProjectCustomModuleTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.SccProjectCustomModule = SccProjectCustomModule;
_j = JSII_RTTI_SYMBOL_1;
SccProjectCustomModule[_j] = { fqn: "@cdktf/provider-google.sccProjectCustomModule.SccProjectCustomModule", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
SccProjectCustomModule.tfResourceType = "google_scc_project_custom_module";
