// @cdktf/provider-google/lib/firestore-field/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirestoreField = exports.FirestoreFieldTtlConfigOutputReference = exports.firestoreFieldTtlConfigToHclTerraform = exports.firestoreFieldTtlConfigToTerraform = exports.FirestoreFieldTimeoutsOutputReference = exports.firestoreFieldTimeoutsToHclTerraform = exports.firestoreFieldTimeoutsToTerraform = exports.FirestoreFieldIndexConfigOutputReference = exports.firestoreFieldIndexConfigToHclTerraform = exports.firestoreFieldIndexConfigToTerraform = exports.FirestoreFieldIndexConfigIndexesList = exports.FirestoreFieldIndexConfigIndexesOutputReference = exports.firestoreFieldIndexConfigIndexesToHclTerraform = exports.firestoreFieldIndexConfigIndexesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function firestoreFieldIndexConfigIndexesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        array_config: cdktf.stringToTerraform(struct.arrayConfig),
        order: cdktf.stringToTerraform(struct.order),
        query_scope: cdktf.stringToTerraform(struct.queryScope),
    };
}
exports.firestoreFieldIndexConfigIndexesToTerraform = firestoreFieldIndexConfigIndexesToTerraform;
function firestoreFieldIndexConfigIndexesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        array_config: {
            value: cdktf.stringToHclTerraform(struct.arrayConfig),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        order: {
            value: cdktf.stringToHclTerraform(struct.order),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        query_scope: {
            value: cdktf.stringToHclTerraform(struct.queryScope),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.firestoreFieldIndexConfigIndexesToHclTerraform = firestoreFieldIndexConfigIndexesToHclTerraform;
class FirestoreFieldIndexConfigIndexesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._arrayConfig !== undefined) {
            hasAnyValues = true;
            internalValueResult.arrayConfig = this._arrayConfig;
        }
        if (this._order !== undefined) {
            hasAnyValues = true;
            internalValueResult.order = this._order;
        }
        if (this._queryScope !== undefined) {
            hasAnyValues = true;
            internalValueResult.queryScope = this._queryScope;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._arrayConfig = undefined;
            this._order = undefined;
            this._queryScope = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._arrayConfig = value.arrayConfig;
            this._order = value.order;
            this._queryScope = value.queryScope;
        }
    }
    get arrayConfig() {
        return this.getStringAttribute('array_config');
    }
    set arrayConfig(value) {
        this._arrayConfig = value;
    }
    resetArrayConfig() {
        this._arrayConfig = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get arrayConfigInput() {
        return this._arrayConfig;
    }
    get order() {
        return this.getStringAttribute('order');
    }
    set order(value) {
        this._order = value;
    }
    resetOrder() {
        this._order = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get orderInput() {
        return this._order;
    }
    get queryScope() {
        return this.getStringAttribute('query_scope');
    }
    set queryScope(value) {
        this._queryScope = value;
    }
    resetQueryScope() {
        this._queryScope = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get queryScopeInput() {
        return this._queryScope;
    }
}
exports.FirestoreFieldIndexConfigIndexesOutputReference = FirestoreFieldIndexConfigIndexesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
FirestoreFieldIndexConfigIndexesOutputReference[_a] = { fqn: "@cdktf/provider-google.firestoreField.FirestoreFieldIndexConfigIndexesOutputReference", version: "14.13.1" };
class FirestoreFieldIndexConfigIndexesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new FirestoreFieldIndexConfigIndexesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.FirestoreFieldIndexConfigIndexesList = FirestoreFieldIndexConfigIndexesList;
_b = JSII_RTTI_SYMBOL_1;
FirestoreFieldIndexConfigIndexesList[_b] = { fqn: "@cdktf/provider-google.firestoreField.FirestoreFieldIndexConfigIndexesList", version: "14.13.1" };
function firestoreFieldIndexConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        indexes: cdktf.listMapper(firestoreFieldIndexConfigIndexesToTerraform, true)(struct.indexes),
    };
}
exports.firestoreFieldIndexConfigToTerraform = firestoreFieldIndexConfigToTerraform;
function firestoreFieldIndexConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        indexes: {
            value: cdktf.listMapperHcl(firestoreFieldIndexConfigIndexesToHclTerraform, true)(struct.indexes),
            isBlock: true,
            type: "set",
            storageClassType: "FirestoreFieldIndexConfigIndexesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.firestoreFieldIndexConfigToHclTerraform = firestoreFieldIndexConfigToHclTerraform;
class FirestoreFieldIndexConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // indexes - computed: false, optional: true, required: false
        this._indexes = new FirestoreFieldIndexConfigIndexesList(this, "indexes", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._indexes?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.indexes = this._indexes?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._indexes.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._indexes.internalValue = value.indexes;
        }
    }
    get indexes() {
        return this._indexes;
    }
    putIndexes(value) {
        this._indexes.internalValue = value;
    }
    resetIndexes() {
        this._indexes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get indexesInput() {
        return this._indexes.internalValue;
    }
}
exports.FirestoreFieldIndexConfigOutputReference = FirestoreFieldIndexConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
FirestoreFieldIndexConfigOutputReference[_c] = { fqn: "@cdktf/provider-google.firestoreField.FirestoreFieldIndexConfigOutputReference", version: "14.13.1" };
function firestoreFieldTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.firestoreFieldTimeoutsToTerraform = firestoreFieldTimeoutsToTerraform;
function firestoreFieldTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.firestoreFieldTimeoutsToHclTerraform = firestoreFieldTimeoutsToHclTerraform;
class FirestoreFieldTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.FirestoreFieldTimeoutsOutputReference = FirestoreFieldTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
FirestoreFieldTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-google.firestoreField.FirestoreFieldTimeoutsOutputReference", version: "14.13.1" };
function firestoreFieldTtlConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.firestoreFieldTtlConfigToTerraform = firestoreFieldTtlConfigToTerraform;
function firestoreFieldTtlConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.firestoreFieldTtlConfigToHclTerraform = firestoreFieldTtlConfigToHclTerraform;
class FirestoreFieldTtlConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.FirestoreFieldTtlConfigOutputReference = FirestoreFieldTtlConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
FirestoreFieldTtlConfigOutputReference[_e] = { fqn: "@cdktf/provider-google.firestoreField.FirestoreFieldTtlConfigOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/firestore_field google_firestore_field}
*/
class FirestoreField extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a FirestoreField resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the FirestoreField to import
    * @param importFromId The id of the existing FirestoreField that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/firestore_field#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the FirestoreField to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_firestore_field", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/firestore_field google_firestore_field} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options FirestoreFieldConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_firestore_field',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // index_config - computed: false, optional: true, required: false
        this._indexConfig = new FirestoreFieldIndexConfigOutputReference(this, "index_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new FirestoreFieldTimeoutsOutputReference(this, "timeouts");
        // ttl_config - computed: false, optional: true, required: false
        this._ttlConfig = new FirestoreFieldTtlConfigOutputReference(this, "ttl_config");
        this._collection = config.collection;
        this._database = config.database;
        this._field = config.field;
        this._id = config.id;
        this._project = config.project;
        this._indexConfig.internalValue = config.indexConfig;
        this._timeouts.internalValue = config.timeouts;
        this._ttlConfig.internalValue = config.ttlConfig;
    }
    get collection() {
        return this.getStringAttribute('collection');
    }
    set collection(value) {
        this._collection = value;
    }
    // Temporarily expose input value. Use with caution.
    get collectionInput() {
        return this._collection;
    }
    get database() {
        return this.getStringAttribute('database');
    }
    set database(value) {
        this._database = value;
    }
    resetDatabase() {
        this._database = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database;
    }
    get field() {
        return this.getStringAttribute('field');
    }
    set field(value) {
        this._field = value;
    }
    // Temporarily expose input value. Use with caution.
    get fieldInput() {
        return this._field;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get indexConfig() {
        return this._indexConfig;
    }
    putIndexConfig(value) {
        this._indexConfig.internalValue = value;
    }
    resetIndexConfig() {
        this._indexConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get indexConfigInput() {
        return this._indexConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get ttlConfig() {
        return this._ttlConfig;
    }
    putTtlConfig(value) {
        this._ttlConfig.internalValue = value;
    }
    resetTtlConfig() {
        this._ttlConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ttlConfigInput() {
        return this._ttlConfig.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            collection: cdktf.stringToTerraform(this._collection),
            database: cdktf.stringToTerraform(this._database),
            field: cdktf.stringToTerraform(this._field),
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
            index_config: firestoreFieldIndexConfigToTerraform(this._indexConfig.internalValue),
            timeouts: firestoreFieldTimeoutsToTerraform(this._timeouts.internalValue),
            ttl_config: firestoreFieldTtlConfigToTerraform(this._ttlConfig.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            collection: {
                value: cdktf.stringToHclTerraform(this._collection),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            database: {
                value: cdktf.stringToHclTerraform(this._database),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            field: {
                value: cdktf.stringToHclTerraform(this._field),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            index_config: {
                value: firestoreFieldIndexConfigToHclTerraform(this._indexConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FirestoreFieldIndexConfigList",
            },
            timeouts: {
                value: firestoreFieldTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "FirestoreFieldTimeouts",
            },
            ttl_config: {
                value: firestoreFieldTtlConfigToHclTerraform(this._ttlConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "FirestoreFieldTtlConfigList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.FirestoreField = FirestoreField;
_f = JSII_RTTI_SYMBOL_1;
FirestoreField[_f] = { fqn: "@cdktf/provider-google.firestoreField.FirestoreField", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
FirestoreField.tfResourceType = "google_firestore_field";
