// @cdktf/provider-google/lib/data-google-iam-testable-permissions/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleIamTestablePermissions = exports.DataGoogleIamTestablePermissionsPermissionsList = exports.DataGoogleIamTestablePermissionsPermissionsOutputReference = exports.dataGoogleIamTestablePermissionsPermissionsToHclTerraform = exports.dataGoogleIamTestablePermissionsPermissionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleIamTestablePermissionsPermissionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleIamTestablePermissionsPermissionsToTerraform = dataGoogleIamTestablePermissionsPermissionsToTerraform;
function dataGoogleIamTestablePermissionsPermissionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleIamTestablePermissionsPermissionsToHclTerraform = dataGoogleIamTestablePermissionsPermissionsToHclTerraform;
class DataGoogleIamTestablePermissionsPermissionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // api_disabled - computed: true, optional: false, required: false
    get apiDisabled() {
        return this.getBooleanAttribute('api_disabled');
    }
    // custom_support_level - computed: true, optional: false, required: false
    get customSupportLevel() {
        return this.getStringAttribute('custom_support_level');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // stage - computed: true, optional: false, required: false
    get stage() {
        return this.getStringAttribute('stage');
    }
    // title - computed: true, optional: false, required: false
    get title() {
        return this.getStringAttribute('title');
    }
}
exports.DataGoogleIamTestablePermissionsPermissionsOutputReference = DataGoogleIamTestablePermissionsPermissionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleIamTestablePermissionsPermissionsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleIamTestablePermissions.DataGoogleIamTestablePermissionsPermissionsOutputReference", version: "14.13.1" };
class DataGoogleIamTestablePermissionsPermissionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleIamTestablePermissionsPermissionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleIamTestablePermissionsPermissionsList = DataGoogleIamTestablePermissionsPermissionsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleIamTestablePermissionsPermissionsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleIamTestablePermissions.DataGoogleIamTestablePermissionsPermissionsList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/iam_testable_permissions google_iam_testable_permissions}
*/
class DataGoogleIamTestablePermissions extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleIamTestablePermissions resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleIamTestablePermissions to import
    * @param importFromId The id of the existing DataGoogleIamTestablePermissions that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/iam_testable_permissions#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleIamTestablePermissions to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_iam_testable_permissions", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/iam_testable_permissions google_iam_testable_permissions} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleIamTestablePermissionsConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_iam_testable_permissions',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // permissions - computed: true, optional: false, required: false
        this._permissions = new DataGoogleIamTestablePermissionsPermissionsList(this, "permissions", false);
        this._customSupportLevel = config.customSupportLevel;
        this._fullResourceName = config.fullResourceName;
        this._id = config.id;
        this._stages = config.stages;
    }
    get customSupportLevel() {
        return this.getStringAttribute('custom_support_level');
    }
    set customSupportLevel(value) {
        this._customSupportLevel = value;
    }
    resetCustomSupportLevel() {
        this._customSupportLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get customSupportLevelInput() {
        return this._customSupportLevel;
    }
    get fullResourceName() {
        return this.getStringAttribute('full_resource_name');
    }
    set fullResourceName(value) {
        this._fullResourceName = value;
    }
    // Temporarily expose input value. Use with caution.
    get fullResourceNameInput() {
        return this._fullResourceName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get permissions() {
        return this._permissions;
    }
    get stages() {
        return this.getListAttribute('stages');
    }
    set stages(value) {
        this._stages = value;
    }
    resetStages() {
        this._stages = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stagesInput() {
        return this._stages;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            custom_support_level: cdktf.stringToTerraform(this._customSupportLevel),
            full_resource_name: cdktf.stringToTerraform(this._fullResourceName),
            id: cdktf.stringToTerraform(this._id),
            stages: cdktf.listMapper(cdktf.stringToTerraform, false)(this._stages),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            custom_support_level: {
                value: cdktf.stringToHclTerraform(this._customSupportLevel),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            full_resource_name: {
                value: cdktf.stringToHclTerraform(this._fullResourceName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            stages: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._stages),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleIamTestablePermissions = DataGoogleIamTestablePermissions;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleIamTestablePermissions[_c] = { fqn: "@cdktf/provider-google.dataGoogleIamTestablePermissions.DataGoogleIamTestablePermissions", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleIamTestablePermissions.tfResourceType = "google_iam_testable_permissions";
