// @cdktf/provider-google/lib/data-google-oracle-database-db-nodes/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleOracleDatabaseDbNodes = exports.DataGoogleOracleDatabaseDbNodesDbNodesList = exports.DataGoogleOracleDatabaseDbNodesDbNodesOutputReference = exports.dataGoogleOracleDatabaseDbNodesDbNodesToHclTerraform = exports.dataGoogleOracleDatabaseDbNodesDbNodesToTerraform = exports.DataGoogleOracleDatabaseDbNodesDbNodesPropertiesList = exports.DataGoogleOracleDatabaseDbNodesDbNodesPropertiesOutputReference = exports.dataGoogleOracleDatabaseDbNodesDbNodesPropertiesToHclTerraform = exports.dataGoogleOracleDatabaseDbNodesDbNodesPropertiesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleOracleDatabaseDbNodesDbNodesPropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseDbNodesDbNodesPropertiesToTerraform = dataGoogleOracleDatabaseDbNodesDbNodesPropertiesToTerraform;
function dataGoogleOracleDatabaseDbNodesDbNodesPropertiesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseDbNodesDbNodesPropertiesToHclTerraform = dataGoogleOracleDatabaseDbNodesDbNodesPropertiesToHclTerraform;
class DataGoogleOracleDatabaseDbNodesDbNodesPropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // db_node_storage_size_gb - computed: true, optional: false, required: false
    get dbNodeStorageSizeGb() {
        return this.getNumberAttribute('db_node_storage_size_gb');
    }
    // db_server_ocid - computed: true, optional: false, required: false
    get dbServerOcid() {
        return this.getStringAttribute('db_server_ocid');
    }
    // hostname - computed: true, optional: false, required: false
    get hostname() {
        return this.getStringAttribute('hostname');
    }
    // memory_size_gb - computed: true, optional: false, required: false
    get memorySizeGb() {
        return this.getNumberAttribute('memory_size_gb');
    }
    // ocid - computed: true, optional: false, required: false
    get ocid() {
        return this.getStringAttribute('ocid');
    }
    // ocpu_count - computed: true, optional: false, required: false
    get ocpuCount() {
        return this.getNumberAttribute('ocpu_count');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // total_cpu_core_count - computed: true, optional: false, required: false
    get totalCpuCoreCount() {
        return this.getNumberAttribute('total_cpu_core_count');
    }
}
exports.DataGoogleOracleDatabaseDbNodesDbNodesPropertiesOutputReference = DataGoogleOracleDatabaseDbNodesDbNodesPropertiesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseDbNodesDbNodesPropertiesOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseDbNodes.DataGoogleOracleDatabaseDbNodesDbNodesPropertiesOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseDbNodesDbNodesPropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseDbNodesDbNodesPropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseDbNodesDbNodesPropertiesList = DataGoogleOracleDatabaseDbNodesDbNodesPropertiesList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseDbNodesDbNodesPropertiesList[_b] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseDbNodes.DataGoogleOracleDatabaseDbNodesDbNodesPropertiesList", version: "14.13.1" };
function dataGoogleOracleDatabaseDbNodesDbNodesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseDbNodesDbNodesToTerraform = dataGoogleOracleDatabaseDbNodesDbNodesToTerraform;
function dataGoogleOracleDatabaseDbNodesDbNodesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseDbNodesDbNodesToHclTerraform = dataGoogleOracleDatabaseDbNodesDbNodesToHclTerraform;
class DataGoogleOracleDatabaseDbNodesDbNodesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // properties - computed: true, optional: false, required: false
        this._properties = new DataGoogleOracleDatabaseDbNodesDbNodesPropertiesList(this, "properties", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get properties() {
        return this._properties;
    }
}
exports.DataGoogleOracleDatabaseDbNodesDbNodesOutputReference = DataGoogleOracleDatabaseDbNodesDbNodesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseDbNodesDbNodesOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseDbNodes.DataGoogleOracleDatabaseDbNodesDbNodesOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseDbNodesDbNodesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseDbNodesDbNodesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseDbNodesDbNodesList = DataGoogleOracleDatabaseDbNodesDbNodesList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseDbNodesDbNodesList[_d] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseDbNodes.DataGoogleOracleDatabaseDbNodesDbNodesList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_db_nodes google_oracle_database_db_nodes}
*/
class DataGoogleOracleDatabaseDbNodes extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleOracleDatabaseDbNodes resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleOracleDatabaseDbNodes to import
    * @param importFromId The id of the existing DataGoogleOracleDatabaseDbNodes that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_db_nodes#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleOracleDatabaseDbNodes to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_oracle_database_db_nodes", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_db_nodes google_oracle_database_db_nodes} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleOracleDatabaseDbNodesConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_oracle_database_db_nodes',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // db_nodes - computed: true, optional: false, required: false
        this._dbNodes = new DataGoogleOracleDatabaseDbNodesDbNodesList(this, "db_nodes", false);
        this._cloudVmCluster = config.cloudVmCluster;
        this._id = config.id;
        this._location = config.location;
        this._project = config.project;
    }
    get cloudVmCluster() {
        return this.getStringAttribute('cloud_vm_cluster');
    }
    set cloudVmCluster(value) {
        this._cloudVmCluster = value;
    }
    // Temporarily expose input value. Use with caution.
    get cloudVmClusterInput() {
        return this._cloudVmCluster;
    }
    get dbNodes() {
        return this._dbNodes;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cloud_vm_cluster: cdktf.stringToTerraform(this._cloudVmCluster),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            cloud_vm_cluster: {
                value: cdktf.stringToHclTerraform(this._cloudVmCluster),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleOracleDatabaseDbNodes = DataGoogleOracleDatabaseDbNodes;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseDbNodes[_e] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseDbNodes.DataGoogleOracleDatabaseDbNodes", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleOracleDatabaseDbNodes.tfResourceType = "google_oracle_database_db_nodes";
