// @cdktf/provider-google/lib/data-pipeline-pipeline/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataPipelinePipeline = exports.DataPipelinePipelineWorkloadOutputReference = exports.dataPipelinePipelineWorkloadToHclTerraform = exports.dataPipelinePipelineWorkloadToTerraform = exports.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestOutputReference = exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToHclTerraform = exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToTerraform = exports.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersOutputReference = exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToHclTerraform = exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToTerraform = exports.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentOutputReference = exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToHclTerraform = exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToTerraform = exports.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestOutputReference = exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToHclTerraform = exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToTerraform = exports.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterOutputReference = exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToHclTerraform = exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToTerraform = exports.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentOutputReference = exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToHclTerraform = exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToTerraform = exports.DataPipelinePipelineTimeoutsOutputReference = exports.dataPipelinePipelineTimeoutsToHclTerraform = exports.dataPipelinePipelineTimeoutsToTerraform = exports.DataPipelinePipelineScheduleInfoOutputReference = exports.dataPipelinePipelineScheduleInfoToHclTerraform = exports.dataPipelinePipelineScheduleInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataPipelinePipelineScheduleInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        schedule: cdktf.stringToTerraform(struct.schedule),
        time_zone: cdktf.stringToTerraform(struct.timeZone),
    };
}
exports.dataPipelinePipelineScheduleInfoToTerraform = dataPipelinePipelineScheduleInfoToTerraform;
function dataPipelinePipelineScheduleInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        schedule: {
            value: cdktf.stringToHclTerraform(struct.schedule),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        time_zone: {
            value: cdktf.stringToHclTerraform(struct.timeZone),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineScheduleInfoToHclTerraform = dataPipelinePipelineScheduleInfoToHclTerraform;
class DataPipelinePipelineScheduleInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._schedule !== undefined) {
            hasAnyValues = true;
            internalValueResult.schedule = this._schedule;
        }
        if (this._timeZone !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeZone = this._timeZone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._schedule = undefined;
            this._timeZone = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._schedule = value.schedule;
            this._timeZone = value.timeZone;
        }
    }
    // next_job_time - computed: true, optional: false, required: false
    get nextJobTime() {
        return this.getStringAttribute('next_job_time');
    }
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    set schedule(value) {
        this._schedule = value;
    }
    resetSchedule() {
        this._schedule = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule;
    }
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    set timeZone(value) {
        this._timeZone = value;
    }
    resetTimeZone() {
        this._timeZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeZoneInput() {
        return this._timeZone;
    }
}
exports.DataPipelinePipelineScheduleInfoOutputReference = DataPipelinePipelineScheduleInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineScheduleInfoOutputReference[_a] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineScheduleInfoOutputReference", version: "14.13.1" };
function dataPipelinePipelineTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dataPipelinePipelineTimeoutsToTerraform = dataPipelinePipelineTimeoutsToTerraform;
function dataPipelinePipelineTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineTimeoutsToHclTerraform = dataPipelinePipelineTimeoutsToHclTerraform;
class DataPipelinePipelineTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DataPipelinePipelineTimeoutsOutputReference = DataPipelinePipelineTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineTimeoutsOutputReference", version: "14.13.1" };
function dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        additional_experiments: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.additionalExperiments),
        additional_user_labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.additionalUserLabels),
        enable_streaming_engine: cdktf.booleanToTerraform(struct.enableStreamingEngine),
        flexrs_goal: cdktf.stringToTerraform(struct.flexrsGoal),
        ip_configuration: cdktf.stringToTerraform(struct.ipConfiguration),
        kms_key_name: cdktf.stringToTerraform(struct.kmsKeyName),
        machine_type: cdktf.stringToTerraform(struct.machineType),
        max_workers: cdktf.numberToTerraform(struct.maxWorkers),
        network: cdktf.stringToTerraform(struct.network),
        num_workers: cdktf.numberToTerraform(struct.numWorkers),
        service_account_email: cdktf.stringToTerraform(struct.serviceAccountEmail),
        subnetwork: cdktf.stringToTerraform(struct.subnetwork),
        temp_location: cdktf.stringToTerraform(struct.tempLocation),
        worker_region: cdktf.stringToTerraform(struct.workerRegion),
        worker_zone: cdktf.stringToTerraform(struct.workerZone),
        zone: cdktf.stringToTerraform(struct.zone),
    };
}
exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToTerraform = dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToTerraform;
function dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        additional_experiments: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.additionalExperiments),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        additional_user_labels: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.additionalUserLabels),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        enable_streaming_engine: {
            value: cdktf.booleanToHclTerraform(struct.enableStreamingEngine),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        flexrs_goal: {
            value: cdktf.stringToHclTerraform(struct.flexrsGoal),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ip_configuration: {
            value: cdktf.stringToHclTerraform(struct.ipConfiguration),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kms_key_name: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        machine_type: {
            value: cdktf.stringToHclTerraform(struct.machineType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        max_workers: {
            value: cdktf.numberToHclTerraform(struct.maxWorkers),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        network: {
            value: cdktf.stringToHclTerraform(struct.network),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        num_workers: {
            value: cdktf.numberToHclTerraform(struct.numWorkers),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        service_account_email: {
            value: cdktf.stringToHclTerraform(struct.serviceAccountEmail),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subnetwork: {
            value: cdktf.stringToHclTerraform(struct.subnetwork),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        temp_location: {
            value: cdktf.stringToHclTerraform(struct.tempLocation),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        worker_region: {
            value: cdktf.stringToHclTerraform(struct.workerRegion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        worker_zone: {
            value: cdktf.stringToHclTerraform(struct.workerZone),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        zone: {
            value: cdktf.stringToHclTerraform(struct.zone),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToHclTerraform = dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToHclTerraform;
class DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._additionalExperiments !== undefined) {
            hasAnyValues = true;
            internalValueResult.additionalExperiments = this._additionalExperiments;
        }
        if (this._additionalUserLabels !== undefined) {
            hasAnyValues = true;
            internalValueResult.additionalUserLabels = this._additionalUserLabels;
        }
        if (this._enableStreamingEngine !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableStreamingEngine = this._enableStreamingEngine;
        }
        if (this._flexrsGoal !== undefined) {
            hasAnyValues = true;
            internalValueResult.flexrsGoal = this._flexrsGoal;
        }
        if (this._ipConfiguration !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipConfiguration = this._ipConfiguration;
        }
        if (this._kmsKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyName = this._kmsKeyName;
        }
        if (this._machineType !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineType = this._machineType;
        }
        if (this._maxWorkers !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxWorkers = this._maxWorkers;
        }
        if (this._network !== undefined) {
            hasAnyValues = true;
            internalValueResult.network = this._network;
        }
        if (this._numWorkers !== undefined) {
            hasAnyValues = true;
            internalValueResult.numWorkers = this._numWorkers;
        }
        if (this._serviceAccountEmail !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccountEmail = this._serviceAccountEmail;
        }
        if (this._subnetwork !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetwork = this._subnetwork;
        }
        if (this._tempLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.tempLocation = this._tempLocation;
        }
        if (this._workerRegion !== undefined) {
            hasAnyValues = true;
            internalValueResult.workerRegion = this._workerRegion;
        }
        if (this._workerZone !== undefined) {
            hasAnyValues = true;
            internalValueResult.workerZone = this._workerZone;
        }
        if (this._zone !== undefined) {
            hasAnyValues = true;
            internalValueResult.zone = this._zone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._additionalExperiments = undefined;
            this._additionalUserLabels = undefined;
            this._enableStreamingEngine = undefined;
            this._flexrsGoal = undefined;
            this._ipConfiguration = undefined;
            this._kmsKeyName = undefined;
            this._machineType = undefined;
            this._maxWorkers = undefined;
            this._network = undefined;
            this._numWorkers = undefined;
            this._serviceAccountEmail = undefined;
            this._subnetwork = undefined;
            this._tempLocation = undefined;
            this._workerRegion = undefined;
            this._workerZone = undefined;
            this._zone = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._additionalExperiments = value.additionalExperiments;
            this._additionalUserLabels = value.additionalUserLabels;
            this._enableStreamingEngine = value.enableStreamingEngine;
            this._flexrsGoal = value.flexrsGoal;
            this._ipConfiguration = value.ipConfiguration;
            this._kmsKeyName = value.kmsKeyName;
            this._machineType = value.machineType;
            this._maxWorkers = value.maxWorkers;
            this._network = value.network;
            this._numWorkers = value.numWorkers;
            this._serviceAccountEmail = value.serviceAccountEmail;
            this._subnetwork = value.subnetwork;
            this._tempLocation = value.tempLocation;
            this._workerRegion = value.workerRegion;
            this._workerZone = value.workerZone;
            this._zone = value.zone;
        }
    }
    get additionalExperiments() {
        return this.getListAttribute('additional_experiments');
    }
    set additionalExperiments(value) {
        this._additionalExperiments = value;
    }
    resetAdditionalExperiments() {
        this._additionalExperiments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalExperimentsInput() {
        return this._additionalExperiments;
    }
    get additionalUserLabels() {
        return this.getStringMapAttribute('additional_user_labels');
    }
    set additionalUserLabels(value) {
        this._additionalUserLabels = value;
    }
    resetAdditionalUserLabels() {
        this._additionalUserLabels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalUserLabelsInput() {
        return this._additionalUserLabels;
    }
    get enableStreamingEngine() {
        return this.getBooleanAttribute('enable_streaming_engine');
    }
    set enableStreamingEngine(value) {
        this._enableStreamingEngine = value;
    }
    resetEnableStreamingEngine() {
        this._enableStreamingEngine = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableStreamingEngineInput() {
        return this._enableStreamingEngine;
    }
    get flexrsGoal() {
        return this.getStringAttribute('flexrs_goal');
    }
    set flexrsGoal(value) {
        this._flexrsGoal = value;
    }
    resetFlexrsGoal() {
        this._flexrsGoal = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get flexrsGoalInput() {
        return this._flexrsGoal;
    }
    get ipConfiguration() {
        return this.getStringAttribute('ip_configuration');
    }
    set ipConfiguration(value) {
        this._ipConfiguration = value;
    }
    resetIpConfiguration() {
        this._ipConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipConfigurationInput() {
        return this._ipConfiguration;
    }
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    set kmsKeyName(value) {
        this._kmsKeyName = value;
    }
    resetKmsKeyName() {
        this._kmsKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyNameInput() {
        return this._kmsKeyName;
    }
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    set machineType(value) {
        this._machineType = value;
    }
    resetMachineType() {
        this._machineType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get machineTypeInput() {
        return this._machineType;
    }
    get maxWorkers() {
        return this.getNumberAttribute('max_workers');
    }
    set maxWorkers(value) {
        this._maxWorkers = value;
    }
    resetMaxWorkers() {
        this._maxWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxWorkersInput() {
        return this._maxWorkers;
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    resetNetwork() {
        this._network = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get numWorkers() {
        return this.getNumberAttribute('num_workers');
    }
    set numWorkers(value) {
        this._numWorkers = value;
    }
    resetNumWorkers() {
        this._numWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get numWorkersInput() {
        return this._numWorkers;
    }
    get serviceAccountEmail() {
        return this.getStringAttribute('service_account_email');
    }
    set serviceAccountEmail(value) {
        this._serviceAccountEmail = value;
    }
    resetServiceAccountEmail() {
        this._serviceAccountEmail = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountEmailInput() {
        return this._serviceAccountEmail;
    }
    get subnetwork() {
        return this.getStringAttribute('subnetwork');
    }
    set subnetwork(value) {
        this._subnetwork = value;
    }
    resetSubnetwork() {
        this._subnetwork = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetworkInput() {
        return this._subnetwork;
    }
    get tempLocation() {
        return this.getStringAttribute('temp_location');
    }
    set tempLocation(value) {
        this._tempLocation = value;
    }
    resetTempLocation() {
        this._tempLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tempLocationInput() {
        return this._tempLocation;
    }
    get workerRegion() {
        return this.getStringAttribute('worker_region');
    }
    set workerRegion(value) {
        this._workerRegion = value;
    }
    resetWorkerRegion() {
        this._workerRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workerRegionInput() {
        return this._workerRegion;
    }
    get workerZone() {
        return this.getStringAttribute('worker_zone');
    }
    set workerZone(value) {
        this._workerZone = value;
    }
    resetWorkerZone() {
        this._workerZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workerZoneInput() {
        return this._workerZone;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
}
exports.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentOutputReference = DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentOutputReference[_c] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentOutputReference", version: "14.13.1" };
function dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container_spec_gcs_path: cdktf.stringToTerraform(struct.containerSpecGcsPath),
        job_name: cdktf.stringToTerraform(struct.jobName),
        launch_options: cdktf.hashMapper(cdktf.stringToTerraform)(struct.launchOptions),
        parameters: cdktf.hashMapper(cdktf.stringToTerraform)(struct.parameters),
        transform_name_mappings: cdktf.hashMapper(cdktf.stringToTerraform)(struct.transformNameMappings),
        update: cdktf.booleanToTerraform(struct.update),
        environment: dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToTerraform(struct.environment),
    };
}
exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToTerraform = dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToTerraform;
function dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        container_spec_gcs_path: {
            value: cdktf.stringToHclTerraform(struct.containerSpecGcsPath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        job_name: {
            value: cdktf.stringToHclTerraform(struct.jobName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        launch_options: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.launchOptions),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        parameters: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.parameters),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        transform_name_mappings: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.transformNameMappings),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        update: {
            value: cdktf.booleanToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        environment: {
            value: dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentToHclTerraform(struct.environment),
            isBlock: true,
            type: "list",
            storageClassType: "DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToHclTerraform = dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToHclTerraform;
class DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // environment - computed: false, optional: true, required: false
        this._environment = new DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterEnvironmentOutputReference(this, "environment");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._containerSpecGcsPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.containerSpecGcsPath = this._containerSpecGcsPath;
        }
        if (this._jobName !== undefined) {
            hasAnyValues = true;
            internalValueResult.jobName = this._jobName;
        }
        if (this._launchOptions !== undefined) {
            hasAnyValues = true;
            internalValueResult.launchOptions = this._launchOptions;
        }
        if (this._parameters !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameters = this._parameters;
        }
        if (this._transformNameMappings !== undefined) {
            hasAnyValues = true;
            internalValueResult.transformNameMappings = this._transformNameMappings;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        if (this._environment?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.environment = this._environment?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._containerSpecGcsPath = undefined;
            this._jobName = undefined;
            this._launchOptions = undefined;
            this._parameters = undefined;
            this._transformNameMappings = undefined;
            this._update = undefined;
            this._environment.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._containerSpecGcsPath = value.containerSpecGcsPath;
            this._jobName = value.jobName;
            this._launchOptions = value.launchOptions;
            this._parameters = value.parameters;
            this._transformNameMappings = value.transformNameMappings;
            this._update = value.update;
            this._environment.internalValue = value.environment;
        }
    }
    get containerSpecGcsPath() {
        return this.getStringAttribute('container_spec_gcs_path');
    }
    set containerSpecGcsPath(value) {
        this._containerSpecGcsPath = value;
    }
    resetContainerSpecGcsPath() {
        this._containerSpecGcsPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get containerSpecGcsPathInput() {
        return this._containerSpecGcsPath;
    }
    get jobName() {
        return this.getStringAttribute('job_name');
    }
    set jobName(value) {
        this._jobName = value;
    }
    // Temporarily expose input value. Use with caution.
    get jobNameInput() {
        return this._jobName;
    }
    get launchOptions() {
        return this.getStringMapAttribute('launch_options');
    }
    set launchOptions(value) {
        this._launchOptions = value;
    }
    resetLaunchOptions() {
        this._launchOptions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchOptionsInput() {
        return this._launchOptions;
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get transformNameMappings() {
        return this.getStringMapAttribute('transform_name_mappings');
    }
    set transformNameMappings(value) {
        this._transformNameMappings = value;
    }
    resetTransformNameMappings() {
        this._transformNameMappings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transformNameMappingsInput() {
        return this._transformNameMappings;
    }
    get update() {
        return this.getBooleanAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
    get environment() {
        return this._environment;
    }
    putEnvironment(value) {
        this._environment.internalValue = value;
    }
    resetEnvironment() {
        this._environment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment.internalValue;
    }
}
exports.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterOutputReference = DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterOutputReference[_d] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterOutputReference", version: "14.13.1" };
function dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        location: cdktf.stringToTerraform(struct.location),
        project_id: cdktf.stringToTerraform(struct.projectId),
        validate_only: cdktf.booleanToTerraform(struct.validateOnly),
        launch_parameter: dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToTerraform(struct.launchParameter),
    };
}
exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToTerraform = dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToTerraform;
function dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        location: {
            value: cdktf.stringToHclTerraform(struct.location),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        project_id: {
            value: cdktf.stringToHclTerraform(struct.projectId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        validate_only: {
            value: cdktf.booleanToHclTerraform(struct.validateOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        launch_parameter: {
            value: dataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterToHclTerraform(struct.launchParameter),
            isBlock: true,
            type: "list",
            storageClassType: "DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToHclTerraform = dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToHclTerraform;
class DataPipelinePipelineWorkloadDataflowFlexTemplateRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // launch_parameter - computed: false, optional: false, required: true
        this._launchParameter = new DataPipelinePipelineWorkloadDataflowFlexTemplateRequestLaunchParameterOutputReference(this, "launch_parameter");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._projectId !== undefined) {
            hasAnyValues = true;
            internalValueResult.projectId = this._projectId;
        }
        if (this._validateOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.validateOnly = this._validateOnly;
        }
        if (this._launchParameter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.launchParameter = this._launchParameter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._location = undefined;
            this._projectId = undefined;
            this._validateOnly = undefined;
            this._launchParameter.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._location = value.location;
            this._projectId = value.projectId;
            this._validateOnly = value.validateOnly;
            this._launchParameter.internalValue = value.launchParameter;
        }
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    set projectId(value) {
        this._projectId = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectIdInput() {
        return this._projectId;
    }
    get validateOnly() {
        return this.getBooleanAttribute('validate_only');
    }
    set validateOnly(value) {
        this._validateOnly = value;
    }
    resetValidateOnly() {
        this._validateOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validateOnlyInput() {
        return this._validateOnly;
    }
    get launchParameter() {
        return this._launchParameter;
    }
    putLaunchParameter(value) {
        this._launchParameter.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get launchParameterInput() {
        return this._launchParameter.internalValue;
    }
}
exports.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestOutputReference = DataPipelinePipelineWorkloadDataflowFlexTemplateRequestOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineWorkloadDataflowFlexTemplateRequestOutputReference[_e] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineWorkloadDataflowFlexTemplateRequestOutputReference", version: "14.13.1" };
function dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        additional_experiments: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.additionalExperiments),
        additional_user_labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.additionalUserLabels),
        bypass_temp_dir_validation: cdktf.booleanToTerraform(struct.bypassTempDirValidation),
        enable_streaming_engine: cdktf.booleanToTerraform(struct.enableStreamingEngine),
        ip_configuration: cdktf.stringToTerraform(struct.ipConfiguration),
        kms_key_name: cdktf.stringToTerraform(struct.kmsKeyName),
        machine_type: cdktf.stringToTerraform(struct.machineType),
        max_workers: cdktf.numberToTerraform(struct.maxWorkers),
        network: cdktf.stringToTerraform(struct.network),
        num_workers: cdktf.numberToTerraform(struct.numWorkers),
        service_account_email: cdktf.stringToTerraform(struct.serviceAccountEmail),
        subnetwork: cdktf.stringToTerraform(struct.subnetwork),
        temp_location: cdktf.stringToTerraform(struct.tempLocation),
        worker_region: cdktf.stringToTerraform(struct.workerRegion),
        worker_zone: cdktf.stringToTerraform(struct.workerZone),
        zone: cdktf.stringToTerraform(struct.zone),
    };
}
exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToTerraform = dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToTerraform;
function dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        additional_experiments: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.additionalExperiments),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        additional_user_labels: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.additionalUserLabels),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        bypass_temp_dir_validation: {
            value: cdktf.booleanToHclTerraform(struct.bypassTempDirValidation),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_streaming_engine: {
            value: cdktf.booleanToHclTerraform(struct.enableStreamingEngine),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        ip_configuration: {
            value: cdktf.stringToHclTerraform(struct.ipConfiguration),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        kms_key_name: {
            value: cdktf.stringToHclTerraform(struct.kmsKeyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        machine_type: {
            value: cdktf.stringToHclTerraform(struct.machineType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        max_workers: {
            value: cdktf.numberToHclTerraform(struct.maxWorkers),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        network: {
            value: cdktf.stringToHclTerraform(struct.network),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        num_workers: {
            value: cdktf.numberToHclTerraform(struct.numWorkers),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        service_account_email: {
            value: cdktf.stringToHclTerraform(struct.serviceAccountEmail),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        subnetwork: {
            value: cdktf.stringToHclTerraform(struct.subnetwork),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        temp_location: {
            value: cdktf.stringToHclTerraform(struct.tempLocation),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        worker_region: {
            value: cdktf.stringToHclTerraform(struct.workerRegion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        worker_zone: {
            value: cdktf.stringToHclTerraform(struct.workerZone),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        zone: {
            value: cdktf.stringToHclTerraform(struct.zone),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToHclTerraform = dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToHclTerraform;
class DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._additionalExperiments !== undefined) {
            hasAnyValues = true;
            internalValueResult.additionalExperiments = this._additionalExperiments;
        }
        if (this._additionalUserLabels !== undefined) {
            hasAnyValues = true;
            internalValueResult.additionalUserLabels = this._additionalUserLabels;
        }
        if (this._bypassTempDirValidation !== undefined) {
            hasAnyValues = true;
            internalValueResult.bypassTempDirValidation = this._bypassTempDirValidation;
        }
        if (this._enableStreamingEngine !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableStreamingEngine = this._enableStreamingEngine;
        }
        if (this._ipConfiguration !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipConfiguration = this._ipConfiguration;
        }
        if (this._kmsKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKeyName = this._kmsKeyName;
        }
        if (this._machineType !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineType = this._machineType;
        }
        if (this._maxWorkers !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxWorkers = this._maxWorkers;
        }
        if (this._network !== undefined) {
            hasAnyValues = true;
            internalValueResult.network = this._network;
        }
        if (this._numWorkers !== undefined) {
            hasAnyValues = true;
            internalValueResult.numWorkers = this._numWorkers;
        }
        if (this._serviceAccountEmail !== undefined) {
            hasAnyValues = true;
            internalValueResult.serviceAccountEmail = this._serviceAccountEmail;
        }
        if (this._subnetwork !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetwork = this._subnetwork;
        }
        if (this._tempLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.tempLocation = this._tempLocation;
        }
        if (this._workerRegion !== undefined) {
            hasAnyValues = true;
            internalValueResult.workerRegion = this._workerRegion;
        }
        if (this._workerZone !== undefined) {
            hasAnyValues = true;
            internalValueResult.workerZone = this._workerZone;
        }
        if (this._zone !== undefined) {
            hasAnyValues = true;
            internalValueResult.zone = this._zone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._additionalExperiments = undefined;
            this._additionalUserLabels = undefined;
            this._bypassTempDirValidation = undefined;
            this._enableStreamingEngine = undefined;
            this._ipConfiguration = undefined;
            this._kmsKeyName = undefined;
            this._machineType = undefined;
            this._maxWorkers = undefined;
            this._network = undefined;
            this._numWorkers = undefined;
            this._serviceAccountEmail = undefined;
            this._subnetwork = undefined;
            this._tempLocation = undefined;
            this._workerRegion = undefined;
            this._workerZone = undefined;
            this._zone = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._additionalExperiments = value.additionalExperiments;
            this._additionalUserLabels = value.additionalUserLabels;
            this._bypassTempDirValidation = value.bypassTempDirValidation;
            this._enableStreamingEngine = value.enableStreamingEngine;
            this._ipConfiguration = value.ipConfiguration;
            this._kmsKeyName = value.kmsKeyName;
            this._machineType = value.machineType;
            this._maxWorkers = value.maxWorkers;
            this._network = value.network;
            this._numWorkers = value.numWorkers;
            this._serviceAccountEmail = value.serviceAccountEmail;
            this._subnetwork = value.subnetwork;
            this._tempLocation = value.tempLocation;
            this._workerRegion = value.workerRegion;
            this._workerZone = value.workerZone;
            this._zone = value.zone;
        }
    }
    get additionalExperiments() {
        return this.getListAttribute('additional_experiments');
    }
    set additionalExperiments(value) {
        this._additionalExperiments = value;
    }
    resetAdditionalExperiments() {
        this._additionalExperiments = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalExperimentsInput() {
        return this._additionalExperiments;
    }
    get additionalUserLabels() {
        return this.getStringMapAttribute('additional_user_labels');
    }
    set additionalUserLabels(value) {
        this._additionalUserLabels = value;
    }
    resetAdditionalUserLabels() {
        this._additionalUserLabels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get additionalUserLabelsInput() {
        return this._additionalUserLabels;
    }
    get bypassTempDirValidation() {
        return this.getBooleanAttribute('bypass_temp_dir_validation');
    }
    set bypassTempDirValidation(value) {
        this._bypassTempDirValidation = value;
    }
    resetBypassTempDirValidation() {
        this._bypassTempDirValidation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bypassTempDirValidationInput() {
        return this._bypassTempDirValidation;
    }
    get enableStreamingEngine() {
        return this.getBooleanAttribute('enable_streaming_engine');
    }
    set enableStreamingEngine(value) {
        this._enableStreamingEngine = value;
    }
    resetEnableStreamingEngine() {
        this._enableStreamingEngine = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableStreamingEngineInput() {
        return this._enableStreamingEngine;
    }
    get ipConfiguration() {
        return this.getStringAttribute('ip_configuration');
    }
    set ipConfiguration(value) {
        this._ipConfiguration = value;
    }
    resetIpConfiguration() {
        this._ipConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipConfigurationInput() {
        return this._ipConfiguration;
    }
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    set kmsKeyName(value) {
        this._kmsKeyName = value;
    }
    resetKmsKeyName() {
        this._kmsKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyNameInput() {
        return this._kmsKeyName;
    }
    get machineType() {
        return this.getStringAttribute('machine_type');
    }
    set machineType(value) {
        this._machineType = value;
    }
    resetMachineType() {
        this._machineType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get machineTypeInput() {
        return this._machineType;
    }
    get maxWorkers() {
        return this.getNumberAttribute('max_workers');
    }
    set maxWorkers(value) {
        this._maxWorkers = value;
    }
    resetMaxWorkers() {
        this._maxWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxWorkersInput() {
        return this._maxWorkers;
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    resetNetwork() {
        this._network = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get numWorkers() {
        return this.getNumberAttribute('num_workers');
    }
    set numWorkers(value) {
        this._numWorkers = value;
    }
    resetNumWorkers() {
        this._numWorkers = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get numWorkersInput() {
        return this._numWorkers;
    }
    get serviceAccountEmail() {
        return this.getStringAttribute('service_account_email');
    }
    set serviceAccountEmail(value) {
        this._serviceAccountEmail = value;
    }
    resetServiceAccountEmail() {
        this._serviceAccountEmail = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountEmailInput() {
        return this._serviceAccountEmail;
    }
    get subnetwork() {
        return this.getStringAttribute('subnetwork');
    }
    set subnetwork(value) {
        this._subnetwork = value;
    }
    resetSubnetwork() {
        this._subnetwork = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetworkInput() {
        return this._subnetwork;
    }
    get tempLocation() {
        return this.getStringAttribute('temp_location');
    }
    set tempLocation(value) {
        this._tempLocation = value;
    }
    resetTempLocation() {
        this._tempLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tempLocationInput() {
        return this._tempLocation;
    }
    get workerRegion() {
        return this.getStringAttribute('worker_region');
    }
    set workerRegion(value) {
        this._workerRegion = value;
    }
    resetWorkerRegion() {
        this._workerRegion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workerRegionInput() {
        return this._workerRegion;
    }
    get workerZone() {
        return this.getStringAttribute('worker_zone');
    }
    set workerZone(value) {
        this._workerZone = value;
    }
    resetWorkerZone() {
        this._workerZone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workerZoneInput() {
        return this._workerZone;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
}
exports.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentOutputReference = DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentOutputReference[_f] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentOutputReference", version: "14.13.1" };
function dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        job_name: cdktf.stringToTerraform(struct.jobName),
        parameters: cdktf.hashMapper(cdktf.stringToTerraform)(struct.parameters),
        transform_name_mapping: cdktf.hashMapper(cdktf.stringToTerraform)(struct.transformNameMapping),
        update: cdktf.booleanToTerraform(struct.update),
        environment: dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToTerraform(struct.environment),
    };
}
exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToTerraform = dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToTerraform;
function dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        job_name: {
            value: cdktf.stringToHclTerraform(struct.jobName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        parameters: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.parameters),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        transform_name_mapping: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.transformNameMapping),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        update: {
            value: cdktf.booleanToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        environment: {
            value: dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentToHclTerraform(struct.environment),
            isBlock: true,
            type: "list",
            storageClassType: "DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToHclTerraform = dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToHclTerraform;
class DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // environment - computed: false, optional: true, required: false
        this._environment = new DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersEnvironmentOutputReference(this, "environment");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._jobName !== undefined) {
            hasAnyValues = true;
            internalValueResult.jobName = this._jobName;
        }
        if (this._parameters !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameters = this._parameters;
        }
        if (this._transformNameMapping !== undefined) {
            hasAnyValues = true;
            internalValueResult.transformNameMapping = this._transformNameMapping;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        if (this._environment?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.environment = this._environment?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._jobName = undefined;
            this._parameters = undefined;
            this._transformNameMapping = undefined;
            this._update = undefined;
            this._environment.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._jobName = value.jobName;
            this._parameters = value.parameters;
            this._transformNameMapping = value.transformNameMapping;
            this._update = value.update;
            this._environment.internalValue = value.environment;
        }
    }
    get jobName() {
        return this.getStringAttribute('job_name');
    }
    set jobName(value) {
        this._jobName = value;
    }
    // Temporarily expose input value. Use with caution.
    get jobNameInput() {
        return this._jobName;
    }
    get parameters() {
        return this.getStringMapAttribute('parameters');
    }
    set parameters(value) {
        this._parameters = value;
    }
    resetParameters() {
        this._parameters = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parametersInput() {
        return this._parameters;
    }
    get transformNameMapping() {
        return this.getStringMapAttribute('transform_name_mapping');
    }
    set transformNameMapping(value) {
        this._transformNameMapping = value;
    }
    resetTransformNameMapping() {
        this._transformNameMapping = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transformNameMappingInput() {
        return this._transformNameMapping;
    }
    get update() {
        return this.getBooleanAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
    get environment() {
        return this._environment;
    }
    putEnvironment(value) {
        this._environment.internalValue = value;
    }
    resetEnvironment() {
        this._environment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment.internalValue;
    }
}
exports.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersOutputReference = DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersOutputReference[_g] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersOutputReference", version: "14.13.1" };
function dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        gcs_path: cdktf.stringToTerraform(struct.gcsPath),
        location: cdktf.stringToTerraform(struct.location),
        project_id: cdktf.stringToTerraform(struct.projectId),
        validate_only: cdktf.booleanToTerraform(struct.validateOnly),
        launch_parameters: dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToTerraform(struct.launchParameters),
    };
}
exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToTerraform = dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToTerraform;
function dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        gcs_path: {
            value: cdktf.stringToHclTerraform(struct.gcsPath),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        location: {
            value: cdktf.stringToHclTerraform(struct.location),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        project_id: {
            value: cdktf.stringToHclTerraform(struct.projectId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        validate_only: {
            value: cdktf.booleanToHclTerraform(struct.validateOnly),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        launch_parameters: {
            value: dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersToHclTerraform(struct.launchParameters),
            isBlock: true,
            type: "list",
            storageClassType: "DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToHclTerraform = dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToHclTerraform;
class DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // launch_parameters - computed: false, optional: true, required: false
        this._launchParameters = new DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestLaunchParametersOutputReference(this, "launch_parameters");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._gcsPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.gcsPath = this._gcsPath;
        }
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._projectId !== undefined) {
            hasAnyValues = true;
            internalValueResult.projectId = this._projectId;
        }
        if (this._validateOnly !== undefined) {
            hasAnyValues = true;
            internalValueResult.validateOnly = this._validateOnly;
        }
        if (this._launchParameters?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.launchParameters = this._launchParameters?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._gcsPath = undefined;
            this._location = undefined;
            this._projectId = undefined;
            this._validateOnly = undefined;
            this._launchParameters.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._gcsPath = value.gcsPath;
            this._location = value.location;
            this._projectId = value.projectId;
            this._validateOnly = value.validateOnly;
            this._launchParameters.internalValue = value.launchParameters;
        }
    }
    get gcsPath() {
        return this.getStringAttribute('gcs_path');
    }
    set gcsPath(value) {
        this._gcsPath = value;
    }
    resetGcsPath() {
        this._gcsPath = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gcsPathInput() {
        return this._gcsPath;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get projectId() {
        return this.getStringAttribute('project_id');
    }
    set projectId(value) {
        this._projectId = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectIdInput() {
        return this._projectId;
    }
    get validateOnly() {
        return this.getBooleanAttribute('validate_only');
    }
    set validateOnly(value) {
        this._validateOnly = value;
    }
    resetValidateOnly() {
        this._validateOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validateOnlyInput() {
        return this._validateOnly;
    }
    get launchParameters() {
        return this._launchParameters;
    }
    putLaunchParameters(value) {
        this._launchParameters.internalValue = value;
    }
    resetLaunchParameters() {
        this._launchParameters.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get launchParametersInput() {
        return this._launchParameters.internalValue;
    }
}
exports.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestOutputReference = DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestOutputReference;
_h = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestOutputReference[_h] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestOutputReference", version: "14.13.1" };
function dataPipelinePipelineWorkloadToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dataflow_flex_template_request: dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToTerraform(struct.dataflowFlexTemplateRequest),
        dataflow_launch_template_request: dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToTerraform(struct.dataflowLaunchTemplateRequest),
    };
}
exports.dataPipelinePipelineWorkloadToTerraform = dataPipelinePipelineWorkloadToTerraform;
function dataPipelinePipelineWorkloadToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        dataflow_flex_template_request: {
            value: dataPipelinePipelineWorkloadDataflowFlexTemplateRequestToHclTerraform(struct.dataflowFlexTemplateRequest),
            isBlock: true,
            type: "list",
            storageClassType: "DataPipelinePipelineWorkloadDataflowFlexTemplateRequestList",
        },
        dataflow_launch_template_request: {
            value: dataPipelinePipelineWorkloadDataflowLaunchTemplateRequestToHclTerraform(struct.dataflowLaunchTemplateRequest),
            isBlock: true,
            type: "list",
            storageClassType: "DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dataPipelinePipelineWorkloadToHclTerraform = dataPipelinePipelineWorkloadToHclTerraform;
class DataPipelinePipelineWorkloadOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // dataflow_flex_template_request - computed: false, optional: true, required: false
        this._dataflowFlexTemplateRequest = new DataPipelinePipelineWorkloadDataflowFlexTemplateRequestOutputReference(this, "dataflow_flex_template_request");
        // dataflow_launch_template_request - computed: false, optional: true, required: false
        this._dataflowLaunchTemplateRequest = new DataPipelinePipelineWorkloadDataflowLaunchTemplateRequestOutputReference(this, "dataflow_launch_template_request");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataflowFlexTemplateRequest?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataflowFlexTemplateRequest = this._dataflowFlexTemplateRequest?.internalValue;
        }
        if (this._dataflowLaunchTemplateRequest?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataflowLaunchTemplateRequest = this._dataflowLaunchTemplateRequest?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataflowFlexTemplateRequest.internalValue = undefined;
            this._dataflowLaunchTemplateRequest.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataflowFlexTemplateRequest.internalValue = value.dataflowFlexTemplateRequest;
            this._dataflowLaunchTemplateRequest.internalValue = value.dataflowLaunchTemplateRequest;
        }
    }
    get dataflowFlexTemplateRequest() {
        return this._dataflowFlexTemplateRequest;
    }
    putDataflowFlexTemplateRequest(value) {
        this._dataflowFlexTemplateRequest.internalValue = value;
    }
    resetDataflowFlexTemplateRequest() {
        this._dataflowFlexTemplateRequest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataflowFlexTemplateRequestInput() {
        return this._dataflowFlexTemplateRequest.internalValue;
    }
    get dataflowLaunchTemplateRequest() {
        return this._dataflowLaunchTemplateRequest;
    }
    putDataflowLaunchTemplateRequest(value) {
        this._dataflowLaunchTemplateRequest.internalValue = value;
    }
    resetDataflowLaunchTemplateRequest() {
        this._dataflowLaunchTemplateRequest.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataflowLaunchTemplateRequestInput() {
        return this._dataflowLaunchTemplateRequest.internalValue;
    }
}
exports.DataPipelinePipelineWorkloadOutputReference = DataPipelinePipelineWorkloadOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataPipelinePipelineWorkloadOutputReference[_j] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipelineWorkloadOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/data_pipeline_pipeline google_data_pipeline_pipeline}
*/
class DataPipelinePipeline extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataPipelinePipeline resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataPipelinePipeline to import
    * @param importFromId The id of the existing DataPipelinePipeline that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/data_pipeline_pipeline#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataPipelinePipeline to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_data_pipeline_pipeline", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/data_pipeline_pipeline google_data_pipeline_pipeline} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataPipelinePipelineConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_data_pipeline_pipeline',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // schedule_info - computed: false, optional: true, required: false
        this._scheduleInfo = new DataPipelinePipelineScheduleInfoOutputReference(this, "schedule_info");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataPipelinePipelineTimeoutsOutputReference(this, "timeouts");
        // workload - computed: false, optional: true, required: false
        this._workload = new DataPipelinePipelineWorkloadOutputReference(this, "workload");
        this._displayName = config.displayName;
        this._id = config.id;
        this._name = config.name;
        this._pipelineSources = config.pipelineSources;
        this._project = config.project;
        this._region = config.region;
        this._schedulerServiceAccountEmail = config.schedulerServiceAccountEmail;
        this._state = config.state;
        this._type = config.type;
        this._scheduleInfo.internalValue = config.scheduleInfo;
        this._timeouts.internalValue = config.timeouts;
        this._workload.internalValue = config.workload;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // job_count - computed: true, optional: false, required: false
    get jobCount() {
        return this.getNumberAttribute('job_count');
    }
    // last_update_time - computed: true, optional: false, required: false
    get lastUpdateTime() {
        return this.getStringAttribute('last_update_time');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get pipelineSources() {
        return this.getStringMapAttribute('pipeline_sources');
    }
    set pipelineSources(value) {
        this._pipelineSources = value;
    }
    resetPipelineSources() {
        this._pipelineSources = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pipelineSourcesInput() {
        return this._pipelineSources;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get schedulerServiceAccountEmail() {
        return this.getStringAttribute('scheduler_service_account_email');
    }
    set schedulerServiceAccountEmail(value) {
        this._schedulerServiceAccountEmail = value;
    }
    resetSchedulerServiceAccountEmail() {
        this._schedulerServiceAccountEmail = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schedulerServiceAccountEmailInput() {
        return this._schedulerServiceAccountEmail;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get scheduleInfo() {
        return this._scheduleInfo;
    }
    putScheduleInfo(value) {
        this._scheduleInfo.internalValue = value;
    }
    resetScheduleInfo() {
        this._scheduleInfo.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInfoInput() {
        return this._scheduleInfo.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get workload() {
        return this._workload;
    }
    putWorkload(value) {
        this._workload.internalValue = value;
    }
    resetWorkload() {
        this._workload.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get workloadInput() {
        return this._workload.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            pipeline_sources: cdktf.hashMapper(cdktf.stringToTerraform)(this._pipelineSources),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            scheduler_service_account_email: cdktf.stringToTerraform(this._schedulerServiceAccountEmail),
            state: cdktf.stringToTerraform(this._state),
            type: cdktf.stringToTerraform(this._type),
            schedule_info: dataPipelinePipelineScheduleInfoToTerraform(this._scheduleInfo.internalValue),
            timeouts: dataPipelinePipelineTimeoutsToTerraform(this._timeouts.internalValue),
            workload: dataPipelinePipelineWorkloadToTerraform(this._workload.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            pipeline_sources: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._pipelineSources),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            scheduler_service_account_email: {
                value: cdktf.stringToHclTerraform(this._schedulerServiceAccountEmail),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            state: {
                value: cdktf.stringToHclTerraform(this._state),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            schedule_info: {
                value: dataPipelinePipelineScheduleInfoToHclTerraform(this._scheduleInfo.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataPipelinePipelineScheduleInfoList",
            },
            timeouts: {
                value: dataPipelinePipelineTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DataPipelinePipelineTimeouts",
            },
            workload: {
                value: dataPipelinePipelineWorkloadToHclTerraform(this._workload.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DataPipelinePipelineWorkloadList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataPipelinePipeline = DataPipelinePipeline;
_k = JSII_RTTI_SYMBOL_1;
DataPipelinePipeline[_k] = { fqn: "@cdktf/provider-google.dataPipelinePipeline.DataPipelinePipeline", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataPipelinePipeline.tfResourceType = "google_data_pipeline_pipeline";
