// @cdktf/provider-google/lib/network-services-lb-traffic-extension/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkServicesLbTrafficExtension = exports.NetworkServicesLbTrafficExtensionTimeoutsOutputReference = exports.networkServicesLbTrafficExtensionTimeoutsToHclTerraform = exports.networkServicesLbTrafficExtensionTimeoutsToTerraform = exports.NetworkServicesLbTrafficExtensionExtensionChainsList = exports.NetworkServicesLbTrafficExtensionExtensionChainsOutputReference = exports.networkServicesLbTrafficExtensionExtensionChainsToHclTerraform = exports.networkServicesLbTrafficExtensionExtensionChainsToTerraform = exports.NetworkServicesLbTrafficExtensionExtensionChainsMatchConditionOutputReference = exports.networkServicesLbTrafficExtensionExtensionChainsMatchConditionToHclTerraform = exports.networkServicesLbTrafficExtensionExtensionChainsMatchConditionToTerraform = exports.NetworkServicesLbTrafficExtensionExtensionChainsExtensionsList = exports.NetworkServicesLbTrafficExtensionExtensionChainsExtensionsOutputReference = exports.networkServicesLbTrafficExtensionExtensionChainsExtensionsToHclTerraform = exports.networkServicesLbTrafficExtensionExtensionChainsExtensionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkServicesLbTrafficExtensionExtensionChainsExtensionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        authority: cdktf.stringToTerraform(struct.authority),
        fail_open: cdktf.booleanToTerraform(struct.failOpen),
        forward_headers: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.forwardHeaders),
        name: cdktf.stringToTerraform(struct.name),
        service: cdktf.stringToTerraform(struct.service),
        supported_events: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.supportedEvents),
        timeout: cdktf.stringToTerraform(struct.timeout),
    };
}
exports.networkServicesLbTrafficExtensionExtensionChainsExtensionsToTerraform = networkServicesLbTrafficExtensionExtensionChainsExtensionsToTerraform;
function networkServicesLbTrafficExtensionExtensionChainsExtensionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        authority: {
            value: cdktf.stringToHclTerraform(struct.authority),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        fail_open: {
            value: cdktf.booleanToHclTerraform(struct.failOpen),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        forward_headers: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.forwardHeaders),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        service: {
            value: cdktf.stringToHclTerraform(struct.service),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        supported_events: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.supportedEvents),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        timeout: {
            value: cdktf.stringToHclTerraform(struct.timeout),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkServicesLbTrafficExtensionExtensionChainsExtensionsToHclTerraform = networkServicesLbTrafficExtensionExtensionChainsExtensionsToHclTerraform;
class NetworkServicesLbTrafficExtensionExtensionChainsExtensionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._authority !== undefined) {
            hasAnyValues = true;
            internalValueResult.authority = this._authority;
        }
        if (this._failOpen !== undefined) {
            hasAnyValues = true;
            internalValueResult.failOpen = this._failOpen;
        }
        if (this._forwardHeaders !== undefined) {
            hasAnyValues = true;
            internalValueResult.forwardHeaders = this._forwardHeaders;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._service !== undefined) {
            hasAnyValues = true;
            internalValueResult.service = this._service;
        }
        if (this._supportedEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.supportedEvents = this._supportedEvents;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._authority = undefined;
            this._failOpen = undefined;
            this._forwardHeaders = undefined;
            this._name = undefined;
            this._service = undefined;
            this._supportedEvents = undefined;
            this._timeout = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._authority = value.authority;
            this._failOpen = value.failOpen;
            this._forwardHeaders = value.forwardHeaders;
            this._name = value.name;
            this._service = value.service;
            this._supportedEvents = value.supportedEvents;
            this._timeout = value.timeout;
        }
    }
    get authority() {
        return this.getStringAttribute('authority');
    }
    set authority(value) {
        this._authority = value;
    }
    resetAuthority() {
        this._authority = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authorityInput() {
        return this._authority;
    }
    get failOpen() {
        return this.getBooleanAttribute('fail_open');
    }
    set failOpen(value) {
        this._failOpen = value;
    }
    resetFailOpen() {
        this._failOpen = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get failOpenInput() {
        return this._failOpen;
    }
    get forwardHeaders() {
        return this.getListAttribute('forward_headers');
    }
    set forwardHeaders(value) {
        this._forwardHeaders = value;
    }
    resetForwardHeaders() {
        this._forwardHeaders = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forwardHeadersInput() {
        return this._forwardHeaders;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get service() {
        return this.getStringAttribute('service');
    }
    set service(value) {
        this._service = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceInput() {
        return this._service;
    }
    get supportedEvents() {
        return this.getListAttribute('supported_events');
    }
    set supportedEvents(value) {
        this._supportedEvents = value;
    }
    resetSupportedEvents() {
        this._supportedEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get supportedEventsInput() {
        return this._supportedEvents;
    }
    get timeout() {
        return this.getStringAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
}
exports.NetworkServicesLbTrafficExtensionExtensionChainsExtensionsOutputReference = NetworkServicesLbTrafficExtensionExtensionChainsExtensionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkServicesLbTrafficExtensionExtensionChainsExtensionsOutputReference[_a] = { fqn: "@cdktf/provider-google.networkServicesLbTrafficExtension.NetworkServicesLbTrafficExtensionExtensionChainsExtensionsOutputReference", version: "14.13.1" };
class NetworkServicesLbTrafficExtensionExtensionChainsExtensionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkServicesLbTrafficExtensionExtensionChainsExtensionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkServicesLbTrafficExtensionExtensionChainsExtensionsList = NetworkServicesLbTrafficExtensionExtensionChainsExtensionsList;
_b = JSII_RTTI_SYMBOL_1;
NetworkServicesLbTrafficExtensionExtensionChainsExtensionsList[_b] = { fqn: "@cdktf/provider-google.networkServicesLbTrafficExtension.NetworkServicesLbTrafficExtensionExtensionChainsExtensionsList", version: "14.13.1" };
function networkServicesLbTrafficExtensionExtensionChainsMatchConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cel_expression: cdktf.stringToTerraform(struct.celExpression),
    };
}
exports.networkServicesLbTrafficExtensionExtensionChainsMatchConditionToTerraform = networkServicesLbTrafficExtensionExtensionChainsMatchConditionToTerraform;
function networkServicesLbTrafficExtensionExtensionChainsMatchConditionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cel_expression: {
            value: cdktf.stringToHclTerraform(struct.celExpression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkServicesLbTrafficExtensionExtensionChainsMatchConditionToHclTerraform = networkServicesLbTrafficExtensionExtensionChainsMatchConditionToHclTerraform;
class NetworkServicesLbTrafficExtensionExtensionChainsMatchConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._celExpression !== undefined) {
            hasAnyValues = true;
            internalValueResult.celExpression = this._celExpression;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._celExpression = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._celExpression = value.celExpression;
        }
    }
    get celExpression() {
        return this.getStringAttribute('cel_expression');
    }
    set celExpression(value) {
        this._celExpression = value;
    }
    // Temporarily expose input value. Use with caution.
    get celExpressionInput() {
        return this._celExpression;
    }
}
exports.NetworkServicesLbTrafficExtensionExtensionChainsMatchConditionOutputReference = NetworkServicesLbTrafficExtensionExtensionChainsMatchConditionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetworkServicesLbTrafficExtensionExtensionChainsMatchConditionOutputReference[_c] = { fqn: "@cdktf/provider-google.networkServicesLbTrafficExtension.NetworkServicesLbTrafficExtensionExtensionChainsMatchConditionOutputReference", version: "14.13.1" };
function networkServicesLbTrafficExtensionExtensionChainsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        extensions: cdktf.listMapper(networkServicesLbTrafficExtensionExtensionChainsExtensionsToTerraform, true)(struct.extensions),
        match_condition: networkServicesLbTrafficExtensionExtensionChainsMatchConditionToTerraform(struct.matchCondition),
    };
}
exports.networkServicesLbTrafficExtensionExtensionChainsToTerraform = networkServicesLbTrafficExtensionExtensionChainsToTerraform;
function networkServicesLbTrafficExtensionExtensionChainsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        extensions: {
            value: cdktf.listMapperHcl(networkServicesLbTrafficExtensionExtensionChainsExtensionsToHclTerraform, true)(struct.extensions),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkServicesLbTrafficExtensionExtensionChainsExtensionsList",
        },
        match_condition: {
            value: networkServicesLbTrafficExtensionExtensionChainsMatchConditionToHclTerraform(struct.matchCondition),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkServicesLbTrafficExtensionExtensionChainsMatchConditionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkServicesLbTrafficExtensionExtensionChainsToHclTerraform = networkServicesLbTrafficExtensionExtensionChainsToHclTerraform;
class NetworkServicesLbTrafficExtensionExtensionChainsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // extensions - computed: false, optional: false, required: true
        this._extensions = new NetworkServicesLbTrafficExtensionExtensionChainsExtensionsList(this, "extensions", false);
        // match_condition - computed: false, optional: false, required: true
        this._matchCondition = new NetworkServicesLbTrafficExtensionExtensionChainsMatchConditionOutputReference(this, "match_condition");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._extensions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.extensions = this._extensions?.internalValue;
        }
        if (this._matchCondition?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.matchCondition = this._matchCondition?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._extensions.internalValue = undefined;
            this._matchCondition.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._extensions.internalValue = value.extensions;
            this._matchCondition.internalValue = value.matchCondition;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get extensions() {
        return this._extensions;
    }
    putExtensions(value) {
        this._extensions.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get extensionsInput() {
        return this._extensions.internalValue;
    }
    get matchCondition() {
        return this._matchCondition;
    }
    putMatchCondition(value) {
        this._matchCondition.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchConditionInput() {
        return this._matchCondition.internalValue;
    }
}
exports.NetworkServicesLbTrafficExtensionExtensionChainsOutputReference = NetworkServicesLbTrafficExtensionExtensionChainsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
NetworkServicesLbTrafficExtensionExtensionChainsOutputReference[_d] = { fqn: "@cdktf/provider-google.networkServicesLbTrafficExtension.NetworkServicesLbTrafficExtensionExtensionChainsOutputReference", version: "14.13.1" };
class NetworkServicesLbTrafficExtensionExtensionChainsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkServicesLbTrafficExtensionExtensionChainsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkServicesLbTrafficExtensionExtensionChainsList = NetworkServicesLbTrafficExtensionExtensionChainsList;
_e = JSII_RTTI_SYMBOL_1;
NetworkServicesLbTrafficExtensionExtensionChainsList[_e] = { fqn: "@cdktf/provider-google.networkServicesLbTrafficExtension.NetworkServicesLbTrafficExtensionExtensionChainsList", version: "14.13.1" };
function networkServicesLbTrafficExtensionTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkServicesLbTrafficExtensionTimeoutsToTerraform = networkServicesLbTrafficExtensionTimeoutsToTerraform;
function networkServicesLbTrafficExtensionTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkServicesLbTrafficExtensionTimeoutsToHclTerraform = networkServicesLbTrafficExtensionTimeoutsToHclTerraform;
class NetworkServicesLbTrafficExtensionTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkServicesLbTrafficExtensionTimeoutsOutputReference = NetworkServicesLbTrafficExtensionTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
NetworkServicesLbTrafficExtensionTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-google.networkServicesLbTrafficExtension.NetworkServicesLbTrafficExtensionTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_services_lb_traffic_extension google_network_services_lb_traffic_extension}
*/
class NetworkServicesLbTrafficExtension extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkServicesLbTrafficExtension resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkServicesLbTrafficExtension to import
    * @param importFromId The id of the existing NetworkServicesLbTrafficExtension that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_services_lb_traffic_extension#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkServicesLbTrafficExtension to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_network_services_lb_traffic_extension", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_services_lb_traffic_extension google_network_services_lb_traffic_extension} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkServicesLbTrafficExtensionConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_network_services_lb_traffic_extension',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // extension_chains - computed: false, optional: false, required: true
        this._extensionChains = new NetworkServicesLbTrafficExtensionExtensionChainsList(this, "extension_chains", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkServicesLbTrafficExtensionTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._forwardingRules = config.forwardingRules;
        this._id = config.id;
        this._labels = config.labels;
        this._loadBalancingScheme = config.loadBalancingScheme;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._extensionChains.internalValue = config.extensionChains;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get forwardingRules() {
        return this.getListAttribute('forwarding_rules');
    }
    set forwardingRules(value) {
        this._forwardingRules = value;
    }
    // Temporarily expose input value. Use with caution.
    get forwardingRulesInput() {
        return this._forwardingRules;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get loadBalancingScheme() {
        return this.getStringAttribute('load_balancing_scheme');
    }
    set loadBalancingScheme(value) {
        this._loadBalancingScheme = value;
    }
    resetLoadBalancingScheme() {
        this._loadBalancingScheme = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loadBalancingSchemeInput() {
        return this._loadBalancingScheme;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    get extensionChains() {
        return this._extensionChains;
    }
    putExtensionChains(value) {
        this._extensionChains.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get extensionChainsInput() {
        return this._extensionChains.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            forwarding_rules: cdktf.listMapper(cdktf.stringToTerraform, false)(this._forwardingRules),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            load_balancing_scheme: cdktf.stringToTerraform(this._loadBalancingScheme),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            extension_chains: cdktf.listMapper(networkServicesLbTrafficExtensionExtensionChainsToTerraform, true)(this._extensionChains.internalValue),
            timeouts: networkServicesLbTrafficExtensionTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            forwarding_rules: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._forwardingRules),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            load_balancing_scheme: {
                value: cdktf.stringToHclTerraform(this._loadBalancingScheme),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            extension_chains: {
                value: cdktf.listMapperHcl(networkServicesLbTrafficExtensionExtensionChainsToHclTerraform, true)(this._extensionChains.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkServicesLbTrafficExtensionExtensionChainsList",
            },
            timeouts: {
                value: networkServicesLbTrafficExtensionTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkServicesLbTrafficExtensionTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkServicesLbTrafficExtension = NetworkServicesLbTrafficExtension;
_g = JSII_RTTI_SYMBOL_1;
NetworkServicesLbTrafficExtension[_g] = { fqn: "@cdktf/provider-google.networkServicesLbTrafficExtension.NetworkServicesLbTrafficExtension", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetworkServicesLbTrafficExtension.tfResourceType = "google_network_services_lb_traffic_extension";
