// @cdktf/provider-google/lib/data-google-compute-router-status/index.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleComputeRouterStatus = exports.DataGoogleComputeRouterStatusBestRoutesForRouterList = exports.DataGoogleComputeRouterStatusBestRoutesForRouterOutputReference = exports.dataGoogleComputeRouterStatusBestRoutesForRouterToHclTerraform = exports.dataGoogleComputeRouterStatusBestRoutesForRouterToTerraform = exports.DataGoogleComputeRouterStatusBestRoutesList = exports.DataGoogleComputeRouterStatusBestRoutesOutputReference = exports.dataGoogleComputeRouterStatusBestRoutesToHclTerraform = exports.dataGoogleComputeRouterStatusBestRoutesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleComputeRouterStatusBestRoutesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRouterStatusBestRoutesToTerraform = dataGoogleComputeRouterStatusBestRoutesToTerraform;
function dataGoogleComputeRouterStatusBestRoutesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRouterStatusBestRoutesToHclTerraform = dataGoogleComputeRouterStatusBestRoutesToHclTerraform;
class DataGoogleComputeRouterStatusBestRoutesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // dest_range - computed: true, optional: false, required: false
    get destRange() {
        return this.getStringAttribute('dest_range');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    // next_hop_gateway - computed: true, optional: false, required: false
    get nextHopGateway() {
        return this.getStringAttribute('next_hop_gateway');
    }
    // next_hop_ilb - computed: true, optional: false, required: false
    get nextHopIlb() {
        return this.getStringAttribute('next_hop_ilb');
    }
    // next_hop_instance - computed: true, optional: false, required: false
    get nextHopInstance() {
        return this.getStringAttribute('next_hop_instance');
    }
    // next_hop_instance_zone - computed: true, optional: false, required: false
    get nextHopInstanceZone() {
        return this.getStringAttribute('next_hop_instance_zone');
    }
    // next_hop_ip - computed: true, optional: false, required: false
    get nextHopIp() {
        return this.getStringAttribute('next_hop_ip');
    }
    // next_hop_network - computed: true, optional: false, required: false
    get nextHopNetwork() {
        return this.getStringAttribute('next_hop_network');
    }
    // next_hop_vpn_tunnel - computed: true, optional: false, required: false
    get nextHopVpnTunnel() {
        return this.getStringAttribute('next_hop_vpn_tunnel');
    }
    // priority - computed: true, optional: false, required: false
    get priority() {
        return this.getNumberAttribute('priority');
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    // tags - computed: true, optional: false, required: false
    get tags() {
        return cdktf.Fn.tolist(this.getListAttribute('tags'));
    }
}
exports.DataGoogleComputeRouterStatusBestRoutesOutputReference = DataGoogleComputeRouterStatusBestRoutesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterStatusBestRoutesOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterStatus.DataGoogleComputeRouterStatusBestRoutesOutputReference", version: "14.13.1" };
class DataGoogleComputeRouterStatusBestRoutesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRouterStatusBestRoutesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRouterStatusBestRoutesList = DataGoogleComputeRouterStatusBestRoutesList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterStatusBestRoutesList[_b] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterStatus.DataGoogleComputeRouterStatusBestRoutesList", version: "14.13.1" };
function dataGoogleComputeRouterStatusBestRoutesForRouterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleComputeRouterStatusBestRoutesForRouterToTerraform = dataGoogleComputeRouterStatusBestRoutesForRouterToTerraform;
function dataGoogleComputeRouterStatusBestRoutesForRouterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleComputeRouterStatusBestRoutesForRouterToHclTerraform = dataGoogleComputeRouterStatusBestRoutesForRouterToHclTerraform;
class DataGoogleComputeRouterStatusBestRoutesForRouterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // dest_range - computed: true, optional: false, required: false
    get destRange() {
        return this.getStringAttribute('dest_range');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    // next_hop_gateway - computed: true, optional: false, required: false
    get nextHopGateway() {
        return this.getStringAttribute('next_hop_gateway');
    }
    // next_hop_ilb - computed: true, optional: false, required: false
    get nextHopIlb() {
        return this.getStringAttribute('next_hop_ilb');
    }
    // next_hop_instance - computed: true, optional: false, required: false
    get nextHopInstance() {
        return this.getStringAttribute('next_hop_instance');
    }
    // next_hop_instance_zone - computed: true, optional: false, required: false
    get nextHopInstanceZone() {
        return this.getStringAttribute('next_hop_instance_zone');
    }
    // next_hop_ip - computed: true, optional: false, required: false
    get nextHopIp() {
        return this.getStringAttribute('next_hop_ip');
    }
    // next_hop_network - computed: true, optional: false, required: false
    get nextHopNetwork() {
        return this.getStringAttribute('next_hop_network');
    }
    // next_hop_vpn_tunnel - computed: true, optional: false, required: false
    get nextHopVpnTunnel() {
        return this.getStringAttribute('next_hop_vpn_tunnel');
    }
    // priority - computed: true, optional: false, required: false
    get priority() {
        return this.getNumberAttribute('priority');
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    // tags - computed: true, optional: false, required: false
    get tags() {
        return cdktf.Fn.tolist(this.getListAttribute('tags'));
    }
}
exports.DataGoogleComputeRouterStatusBestRoutesForRouterOutputReference = DataGoogleComputeRouterStatusBestRoutesForRouterOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterStatusBestRoutesForRouterOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterStatus.DataGoogleComputeRouterStatusBestRoutesForRouterOutputReference", version: "14.13.1" };
class DataGoogleComputeRouterStatusBestRoutesForRouterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleComputeRouterStatusBestRoutesForRouterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleComputeRouterStatusBestRoutesForRouterList = DataGoogleComputeRouterStatusBestRoutesForRouterList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterStatusBestRoutesForRouterList[_d] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterStatus.DataGoogleComputeRouterStatusBestRoutesForRouterList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_router_status google_compute_router_status}
*/
class DataGoogleComputeRouterStatus extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleComputeRouterStatus resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleComputeRouterStatus to import
    * @param importFromId The id of the existing DataGoogleComputeRouterStatus that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_router_status#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleComputeRouterStatus to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_router_status", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/compute_router_status google_compute_router_status} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleComputeRouterStatusConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_router_status',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // best_routes - computed: true, optional: false, required: false
        this._bestRoutes = new DataGoogleComputeRouterStatusBestRoutesList(this, "best_routes", false);
        // best_routes_for_router - computed: true, optional: false, required: false
        this._bestRoutesForRouter = new DataGoogleComputeRouterStatusBestRoutesForRouterList(this, "best_routes_for_router", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._region = config.region;
    }
    get bestRoutes() {
        return this._bestRoutes;
    }
    get bestRoutesForRouter() {
        return this._bestRoutesForRouter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleComputeRouterStatus = DataGoogleComputeRouterStatus;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleComputeRouterStatus[_e] = { fqn: "@cdktf/provider-google.dataGoogleComputeRouterStatus.DataGoogleComputeRouterStatus", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleComputeRouterStatus.tfResourceType = "google_compute_router_status";
