// @cdktf/hcl2cdk/lib/__tests__/expressions.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.referenceToAst = void 0;
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const generator_1 = __importDefault(require("@babel/generator"));
const t = __importStar(require("@babel/types"));
const expressions_1 = require("../expressions");
const references_1 = require("../references");
const hcl2json_1 = require("@cdktf/hcl2json");
const utils_1 = require("../utils");
const nodeIds = [
    "var.input",
    "module.vpc",
    "data.aws_s3_bucket.examplebucket",
    "aws_s3_bucket.examplebucket",
    "aws_s3_bucket.otherbucket",
    "var.users",
    "aws_kms_key.key",
    "local.service_name",
];
// Transforms a path with segments into literals describing the path
function getPropertyAccessPath(input) {
    return input
        .split(/(\[|\]|\.)/g)
        .filter((p) => p.length > 0 && p !== "." && p !== "[" && p !== "]")
        .map((p) => (p.startsWith(`"`) && p.endsWith(`"`) ? p.slice(1, -1) : p));
}
function referenceToAst(scope, ref) {
    const [resource, , ...selector] = ref.referencee.full.split(".");
    const variableReference = t.identifier((0, utils_1.camelCase)((0, references_1.referenceToVariableName)(scope, ref)));
    if (resource === "data") {
        selector.shift(); // remove the data part so that the name is not used in the selector
    }
    const accessor = selector.reduce((carry, member, index) => t.memberExpression(carry, t.identifier(index === 0 && resource === "module"
        ? (0, utils_1.camelCase)(member + "Output")
        : (0, utils_1.camelCase)(member))), variableReference);
    if (ref.useFqn) {
        return t.memberExpression(accessor, t.identifier("fqn"));
    }
    if (ref.isVariable) {
        return t.memberExpression(accessor, t.identifier("value"));
    }
    return accessor;
}
exports.referenceToAst = referenceToAst;
describe("expressions", () => {
    describe("#extractReferencesFromExpression", () => {
        it("finds no references in literals", () => {
            return expect((0, references_1.extractReferencesFromExpression)("nothingtobeseen", nodeIds)).resolves.toEqual([]);
        });
        it("finds no references in literals with functions", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${foo(nothingtobeseen)}", nodeIds)).resolves.toEqual([]);
        });
        it("finds no references in literals with functions and artihmetics", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${foo(nothingtobeseen - 2) + 3}", nodeIds)).resolves.toEqual([]);
        });
        it("finds plain var reference", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${var.input}", nodeIds)).resolves.toEqual([
                {
                    referencee: { id: "var.input", full: "var.input" },
                    useFqn: false,
                    isVariable: true,
                    start: 2,
                    end: 11,
                },
            ]);
        });
        it("finds plain module reference", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${module.vpc.public_subnets}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "module.vpc",
                        full: "module.vpc.public_subnets",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 2,
                    end: 27,
                },
            ]);
        });
        it("finds plain data reference", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${data.aws_s3_bucket.examplebucket.arn}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "data.aws_s3_bucket.examplebucket",
                        full: "data.aws_s3_bucket.examplebucket.arn",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 2,
                    end: 38,
                },
            ]);
        });
        it("finds plain local reference", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${local.service_name}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "local.service_name",
                        full: "local.service_name",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 2,
                    end: 20,
                },
            ]);
        });
        it("finds plain resource reference", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${aws_s3_bucket.examplebucket.id}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket.id",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 2,
                    end: 32,
                },
            ]);
        });
        it("finds plain resource references in artihmetics", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${aws_s3_bucket.examplebucket.count + aws_s3_bucket.otherbucket.count }", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket.count",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 2,
                    end: 35,
                },
                {
                    referencee: {
                        id: "aws_s3_bucket.otherbucket",
                        full: "aws_s3_bucket.otherbucket.count",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 38,
                    end: 69,
                },
            ]);
        });
        it("use fqn for splat reference", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${aws_s3_bucket.examplebucket.*.id}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket",
                    },
                    useFqn: true,
                    isVariable: false,
                    start: 2,
                    end: 29,
                },
            ]);
        });
        it("use no fqn if property is present on numeric access", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${aws_s3_bucket.examplebucket.network_interface.0.access_config.0.assigned_nat_ip}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket",
                    },
                    useFqn: true,
                    isVariable: false,
                    start: 2,
                    end: 81,
                },
            ]);
        });
        it("detect splat reference within function", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${toset(aws_s3_bucket.examplebucket.*)}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket",
                    },
                    useFqn: true,
                    isVariable: false,
                    start: 8,
                    end: 35,
                },
            ]);
        });
        it("finds all resources in conditional", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${aws_kms_key.key.deletion_window_in_days > 3 ? aws_s3_bucket.examplebucket.id : []}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_kms_key.key",
                        full: "aws_kms_key.key.deletion_window_in_days",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 2,
                    end: 41,
                },
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket.id",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 48,
                    end: 78,
                },
            ]);
        });
        it("finds all resources in functions", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${element(aws_s3_bucket.examplebucket, 0).id}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 10,
                    end: 37,
                },
            ]);
        });
        it("finds all resources in functions with splat", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${element(aws_s3_bucket.examplebucket.*.id, 0)}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket",
                    },
                    useFqn: true,
                    isVariable: false,
                    start: 10,
                    end: 37,
                },
            ]);
        });
        it("finds all resources in for loops", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${{ for name, user in var.users : user.role => name...}}", nodeIds)).resolves.toEqual([
                {
                    referencee: { id: "var.users", full: "var.users" },
                    useFqn: false,
                    isVariable: true,
                    start: 22,
                    end: 31,
                },
            ]);
        });
        it("finds resources with property access", () => {
            return expect((0, references_1.extractReferencesFromExpression)("${aws_s3_bucket.examplebucket[0].id}", nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket",
                    },
                    useFqn: true,
                    isVariable: false,
                    start: 2,
                    end: 35,
                },
            ]);
        });
        it("finds references within functions that use arrays and comments", () => {
            return expect((0, references_1.extractReferencesFromExpression)(`\${compact([
            # The example "bucket"
            aws_s3_bucket.examplebucket,
        
            # The "Learn" single page application. This is not configured in all environments.
            var.input,
          ])}`, nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "aws_s3_bucket.examplebucket",
                        full: "aws_s3_bucket.examplebucket",
                    },
                    useFqn: false,
                    isVariable: false,
                    start: 59,
                    end: 86,
                },
                {
                    referencee: {
                        id: "var.input",
                        full: "var.input",
                    },
                    useFqn: false,
                    isVariable: true,
                    start: 204,
                    end: 213,
                },
            ]);
        });
        it("finds references for same referencees", () => {
            return expect((0, references_1.extractReferencesFromExpression)(`\${var.input == "test" ? "azure-ad-int" : "azure-ad-\${var.input}"}`, nodeIds)).resolves.toEqual([
                {
                    referencee: {
                        id: "var.input",
                        full: "var.input",
                    },
                    useFqn: false,
                    isVariable: true,
                    start: 2,
                    end: 11,
                },
                {
                    referencee: {
                        id: "var.input",
                        full: "var.input",
                    },
                    useFqn: false,
                    isVariable: true,
                    start: 53,
                    end: 62,
                },
            ]);
        });
    });
    describe("#referenceToAst", () => {
        it("property access", () => {
            const scope = {
                providerSchema: { format_version: "0.1" },
                providerGenerator: {},
                constructs: new Set(),
                variables: {},
                hasTokenBasedTypeCoercion: false,
                nodeIds: [],
                importables: [],
                topLevelConfig: {},
            };
            expect((0, generator_1.default)(t.program([
                t.expressionStatement(referenceToAst(scope, {
                    start: 0,
                    end: 0,
                    useFqn: false,
                    isVariable: false,
                    referencee: {
                        id: "aws_kms_key.key",
                        full: "aws_kms_key.key.deletion_window_in_days",
                    },
                })),
            ])).code).toMatchInlineSnapshot(`"key.deletionWindowInDays;"`);
        });
    });
    describe("#iteratorVariableToAst", () => {
        async function run(value) {
            const scope = {
                providerSchema: { format_version: "0.1" },
                providerGenerator: {},
                constructs: new Set(),
                variables: {},
                hasTokenBasedTypeCoercion: false,
                nodeIds: [],
                importables: [],
                topLevelConfig: {},
            };
            const ast = await (0, hcl2json_1.getExpressionAst)("main.tf", value);
            return (0, generator_1.default)(t.program([
                t.expressionStatement((0, expressions_1.dynamicVariableToAst)(scope, ast.children[0], "myIterator")),
            ])).code;
        }
        it("should convert iterator key accessor", async () => {
            expect(await run('"${each.key}"')).toMatchInlineSnapshot(`"myIterator.key;"`);
        });
        it("should convert iterator value accessor", async () => {
            expect(await run('"${each.value}"')).toMatchInlineSnapshot(`"myIterator.value;"`);
        });
        it("should convert iterator value deep accessor", async () => {
            expect(await run('"${each.value.list.map.name}"')).toMatchInlineSnapshot(`"Fn.lookupNested(myIterator.value, ["list", "map", "name"]);"`);
        });
        it("should convert iterator value with map access", async () => {
            expect(await run('"${each.value[0]["map"]["name"]}"')).toMatchInlineSnapshot(`"Fn.lookupNested(myIterator.value, ["[0]", "[\\"map\\"]", "[\\"name\\"]"]);"`);
        });
    });
    describe("#getPropertyAccessPath", () => {
        it.each([
            [".list.map.name", ["list", "map", "name"]],
            [`[0]["map"]["name"]`, ["0", "map", "name"]],
            [`[0].map["name"]`, ["0", "map", "name"]],
        ])("should return the correct path for %s", (input, expected) => {
            expect(getPropertyAccessPath(input)).toEqual(expected);
        });
    });
});
