// @cdktf/provider-google/lib/apigee-keystores-aliases-pkcs12/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApigeeKeystoresAliasesPkcs12 = exports.ApigeeKeystoresAliasesPkcs12TimeoutsOutputReference = exports.apigeeKeystoresAliasesPkcs12TimeoutsToHclTerraform = exports.apigeeKeystoresAliasesPkcs12TimeoutsToTerraform = exports.ApigeeKeystoresAliasesPkcs12CertsInfoList = exports.ApigeeKeystoresAliasesPkcs12CertsInfoOutputReference = exports.apigeeKeystoresAliasesPkcs12CertsInfoToHclTerraform = exports.apigeeKeystoresAliasesPkcs12CertsInfoToTerraform = exports.ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoList = exports.ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoOutputReference = exports.apigeeKeystoresAliasesPkcs12CertsInfoCertInfoToHclTerraform = exports.apigeeKeystoresAliasesPkcs12CertsInfoCertInfoToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function apigeeKeystoresAliasesPkcs12CertsInfoCertInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.apigeeKeystoresAliasesPkcs12CertsInfoCertInfoToTerraform = apigeeKeystoresAliasesPkcs12CertsInfoCertInfoToTerraform;
function apigeeKeystoresAliasesPkcs12CertsInfoCertInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.apigeeKeystoresAliasesPkcs12CertsInfoCertInfoToHclTerraform = apigeeKeystoresAliasesPkcs12CertsInfoCertInfoToHclTerraform;
class ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // basic_constraints - computed: true, optional: false, required: false
    get basicConstraints() {
        return this.getStringAttribute('basic_constraints');
    }
    // expiry_date - computed: true, optional: false, required: false
    get expiryDate() {
        return this.getStringAttribute('expiry_date');
    }
    // is_valid - computed: true, optional: false, required: false
    get isValid() {
        return this.getStringAttribute('is_valid');
    }
    // issuer - computed: true, optional: false, required: false
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    // public_key - computed: true, optional: false, required: false
    get publicKey() {
        return this.getStringAttribute('public_key');
    }
    // serial_number - computed: true, optional: false, required: false
    get serialNumber() {
        return this.getStringAttribute('serial_number');
    }
    // sig_alg_name - computed: true, optional: false, required: false
    get sigAlgName() {
        return this.getStringAttribute('sig_alg_name');
    }
    // subject - computed: true, optional: false, required: false
    get subject() {
        return this.getStringAttribute('subject');
    }
    // subject_alternative_names - computed: true, optional: false, required: false
    get subjectAlternativeNames() {
        return this.getListAttribute('subject_alternative_names');
    }
    // valid_from - computed: true, optional: false, required: false
    get validFrom() {
        return this.getStringAttribute('valid_from');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
}
exports.ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoOutputReference = ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoOutputReference[_a] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesPkcs12.ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoOutputReference", version: "14.13.1" };
class ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoList = ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoList;
_b = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoList[_b] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesPkcs12.ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoList", version: "14.13.1" };
function apigeeKeystoresAliasesPkcs12CertsInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.apigeeKeystoresAliasesPkcs12CertsInfoToTerraform = apigeeKeystoresAliasesPkcs12CertsInfoToTerraform;
function apigeeKeystoresAliasesPkcs12CertsInfoToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.apigeeKeystoresAliasesPkcs12CertsInfoToHclTerraform = apigeeKeystoresAliasesPkcs12CertsInfoToHclTerraform;
class ApigeeKeystoresAliasesPkcs12CertsInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // cert_info - computed: true, optional: false, required: false
        this._certInfo = new ApigeeKeystoresAliasesPkcs12CertsInfoCertInfoList(this, "cert_info", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get certInfo() {
        return this._certInfo;
    }
}
exports.ApigeeKeystoresAliasesPkcs12CertsInfoOutputReference = ApigeeKeystoresAliasesPkcs12CertsInfoOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesPkcs12CertsInfoOutputReference[_c] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesPkcs12.ApigeeKeystoresAliasesPkcs12CertsInfoOutputReference", version: "14.13.1" };
class ApigeeKeystoresAliasesPkcs12CertsInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ApigeeKeystoresAliasesPkcs12CertsInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ApigeeKeystoresAliasesPkcs12CertsInfoList = ApigeeKeystoresAliasesPkcs12CertsInfoList;
_d = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesPkcs12CertsInfoList[_d] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesPkcs12.ApigeeKeystoresAliasesPkcs12CertsInfoList", version: "14.13.1" };
function apigeeKeystoresAliasesPkcs12TimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
    };
}
exports.apigeeKeystoresAliasesPkcs12TimeoutsToTerraform = apigeeKeystoresAliasesPkcs12TimeoutsToTerraform;
function apigeeKeystoresAliasesPkcs12TimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.apigeeKeystoresAliasesPkcs12TimeoutsToHclTerraform = apigeeKeystoresAliasesPkcs12TimeoutsToHclTerraform;
class ApigeeKeystoresAliasesPkcs12TimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
}
exports.ApigeeKeystoresAliasesPkcs12TimeoutsOutputReference = ApigeeKeystoresAliasesPkcs12TimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesPkcs12TimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesPkcs12.ApigeeKeystoresAliasesPkcs12TimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apigee_keystores_aliases_pkcs12 google_apigee_keystores_aliases_pkcs12}
*/
class ApigeeKeystoresAliasesPkcs12 extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ApigeeKeystoresAliasesPkcs12 resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ApigeeKeystoresAliasesPkcs12 to import
    * @param importFromId The id of the existing ApigeeKeystoresAliasesPkcs12 that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apigee_keystores_aliases_pkcs12#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ApigeeKeystoresAliasesPkcs12 to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_apigee_keystores_aliases_pkcs12", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/apigee_keystores_aliases_pkcs12 google_apigee_keystores_aliases_pkcs12} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ApigeeKeystoresAliasesPkcs12Config
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_apigee_keystores_aliases_pkcs12',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // certs_info - computed: true, optional: false, required: false
        this._certsInfo = new ApigeeKeystoresAliasesPkcs12CertsInfoList(this, "certs_info", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ApigeeKeystoresAliasesPkcs12TimeoutsOutputReference(this, "timeouts");
        this._alias = config.alias;
        this._environment = config.environment;
        this._file = config.file;
        this._filehash = config.filehash;
        this._id = config.id;
        this._keystore = config.keystore;
        this._orgId = config.orgId;
        this._password = config.password;
        this._timeouts.internalValue = config.timeouts;
    }
    get alias() {
        return this.getStringAttribute('alias');
    }
    set alias(value) {
        this._alias = value;
    }
    // Temporarily expose input value. Use with caution.
    get aliasInput() {
        return this._alias;
    }
    get certsInfo() {
        return this._certsInfo;
    }
    get environment() {
        return this.getStringAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get file() {
        return this.getStringAttribute('file');
    }
    set file(value) {
        this._file = value;
    }
    // Temporarily expose input value. Use with caution.
    get fileInput() {
        return this._file;
    }
    get filehash() {
        return this.getStringAttribute('filehash');
    }
    set filehash(value) {
        this._filehash = value;
    }
    // Temporarily expose input value. Use with caution.
    get filehashInput() {
        return this._filehash;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keystore() {
        return this.getStringAttribute('keystore');
    }
    set keystore(value) {
        this._keystore = value;
    }
    // Temporarily expose input value. Use with caution.
    get keystoreInput() {
        return this._keystore;
    }
    get orgId() {
        return this.getStringAttribute('org_id');
    }
    set orgId(value) {
        this._orgId = value;
    }
    // Temporarily expose input value. Use with caution.
    get orgIdInput() {
        return this._orgId;
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    resetPassword() {
        this._password = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            alias: cdktf.stringToTerraform(this._alias),
            environment: cdktf.stringToTerraform(this._environment),
            file: cdktf.stringToTerraform(this._file),
            filehash: cdktf.stringToTerraform(this._filehash),
            id: cdktf.stringToTerraform(this._id),
            keystore: cdktf.stringToTerraform(this._keystore),
            org_id: cdktf.stringToTerraform(this._orgId),
            password: cdktf.stringToTerraform(this._password),
            timeouts: apigeeKeystoresAliasesPkcs12TimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            alias: {
                value: cdktf.stringToHclTerraform(this._alias),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            environment: {
                value: cdktf.stringToHclTerraform(this._environment),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            file: {
                value: cdktf.stringToHclTerraform(this._file),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            filehash: {
                value: cdktf.stringToHclTerraform(this._filehash),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            keystore: {
                value: cdktf.stringToHclTerraform(this._keystore),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            org_id: {
                value: cdktf.stringToHclTerraform(this._orgId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            password: {
                value: cdktf.stringToHclTerraform(this._password),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            timeouts: {
                value: apigeeKeystoresAliasesPkcs12TimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ApigeeKeystoresAliasesPkcs12Timeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ApigeeKeystoresAliasesPkcs12 = ApigeeKeystoresAliasesPkcs12;
_f = JSII_RTTI_SYMBOL_1;
ApigeeKeystoresAliasesPkcs12[_f] = { fqn: "@cdktf/provider-google.apigeeKeystoresAliasesPkcs12.ApigeeKeystoresAliasesPkcs12", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ApigeeKeystoresAliasesPkcs12.tfResourceType = "google_apigee_keystores_aliases_pkcs12";
