// @cdktf/provider-google/lib/dialogflow-cx-flow/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dialogflowCxFlowEventHandlersToHclTerraform = exports.dialogflowCxFlowEventHandlersToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsList = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesList = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToTerraform = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesList = exports.DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesOutputReference = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToHclTerraform = exports.dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToTerraform = exports.DialogflowCxFlowAdvancedSettingsOutputReference = exports.dialogflowCxFlowAdvancedSettingsToHclTerraform = exports.dialogflowCxFlowAdvancedSettingsToTerraform = exports.DialogflowCxFlowAdvancedSettingsSpeechSettingsOutputReference = exports.dialogflowCxFlowAdvancedSettingsSpeechSettingsToHclTerraform = exports.dialogflowCxFlowAdvancedSettingsSpeechSettingsToTerraform = exports.DialogflowCxFlowAdvancedSettingsLoggingSettingsOutputReference = exports.dialogflowCxFlowAdvancedSettingsLoggingSettingsToHclTerraform = exports.dialogflowCxFlowAdvancedSettingsLoggingSettingsToTerraform = exports.DialogflowCxFlowAdvancedSettingsDtmfSettingsOutputReference = exports.dialogflowCxFlowAdvancedSettingsDtmfSettingsToHclTerraform = exports.dialogflowCxFlowAdvancedSettingsDtmfSettingsToTerraform = exports.DialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationOutputReference = exports.dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToHclTerraform = exports.dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToTerraform = void 0;
exports.DialogflowCxFlow = exports.DialogflowCxFlowTransitionRoutesList = exports.DialogflowCxFlowTransitionRoutesOutputReference = exports.dialogflowCxFlowTransitionRoutesToHclTerraform = exports.dialogflowCxFlowTransitionRoutesToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsList = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesList = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToTerraform = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesList = exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesOutputReference = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToHclTerraform = exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToTerraform = exports.DialogflowCxFlowTimeoutsOutputReference = exports.dialogflowCxFlowTimeoutsToHclTerraform = exports.dialogflowCxFlowTimeoutsToTerraform = exports.DialogflowCxFlowNluSettingsOutputReference = exports.dialogflowCxFlowNluSettingsToHclTerraform = exports.dialogflowCxFlowNluSettingsToTerraform = exports.DialogflowCxFlowEventHandlersList = exports.DialogflowCxFlowEventHandlersOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        uri: cdktf.stringToTerraform(struct.uri),
    };
}
exports.dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToTerraform = dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToTerraform;
function dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        uri: {
            value: cdktf.stringToHclTerraform(struct.uri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToHclTerraform = dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToHclTerraform;
class DialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._uri !== undefined) {
            hasAnyValues = true;
            internalValueResult.uri = this._uri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._uri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._uri = value.uri;
        }
    }
    get uri() {
        return this.getStringAttribute('uri');
    }
    set uri(value) {
        this._uri = value;
    }
    resetUri() {
        this._uri = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uriInput() {
        return this._uri;
    }
}
exports.DialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationOutputReference = DialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationOutputReference[_a] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationOutputReference", version: "14.13.1" };
function dialogflowCxFlowAdvancedSettingsDtmfSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        finish_digit: cdktf.stringToTerraform(struct.finishDigit),
        max_digits: cdktf.numberToTerraform(struct.maxDigits),
    };
}
exports.dialogflowCxFlowAdvancedSettingsDtmfSettingsToTerraform = dialogflowCxFlowAdvancedSettingsDtmfSettingsToTerraform;
function dialogflowCxFlowAdvancedSettingsDtmfSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enabled: {
            value: cdktf.booleanToHclTerraform(struct.enabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        finish_digit: {
            value: cdktf.stringToHclTerraform(struct.finishDigit),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        max_digits: {
            value: cdktf.numberToHclTerraform(struct.maxDigits),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowAdvancedSettingsDtmfSettingsToHclTerraform = dialogflowCxFlowAdvancedSettingsDtmfSettingsToHclTerraform;
class DialogflowCxFlowAdvancedSettingsDtmfSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._finishDigit !== undefined) {
            hasAnyValues = true;
            internalValueResult.finishDigit = this._finishDigit;
        }
        if (this._maxDigits !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxDigits = this._maxDigits;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._finishDigit = undefined;
            this._maxDigits = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._finishDigit = value.finishDigit;
            this._maxDigits = value.maxDigits;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get finishDigit() {
        return this.getStringAttribute('finish_digit');
    }
    set finishDigit(value) {
        this._finishDigit = value;
    }
    resetFinishDigit() {
        this._finishDigit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get finishDigitInput() {
        return this._finishDigit;
    }
    get maxDigits() {
        return this.getNumberAttribute('max_digits');
    }
    set maxDigits(value) {
        this._maxDigits = value;
    }
    resetMaxDigits() {
        this._maxDigits = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxDigitsInput() {
        return this._maxDigits;
    }
}
exports.DialogflowCxFlowAdvancedSettingsDtmfSettingsOutputReference = DialogflowCxFlowAdvancedSettingsDtmfSettingsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowAdvancedSettingsDtmfSettingsOutputReference[_b] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowAdvancedSettingsDtmfSettingsOutputReference", version: "14.13.1" };
function dialogflowCxFlowAdvancedSettingsLoggingSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_consent_based_redaction: cdktf.booleanToTerraform(struct.enableConsentBasedRedaction),
        enable_interaction_logging: cdktf.booleanToTerraform(struct.enableInteractionLogging),
        enable_stackdriver_logging: cdktf.booleanToTerraform(struct.enableStackdriverLogging),
    };
}
exports.dialogflowCxFlowAdvancedSettingsLoggingSettingsToTerraform = dialogflowCxFlowAdvancedSettingsLoggingSettingsToTerraform;
function dialogflowCxFlowAdvancedSettingsLoggingSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_consent_based_redaction: {
            value: cdktf.booleanToHclTerraform(struct.enableConsentBasedRedaction),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_interaction_logging: {
            value: cdktf.booleanToHclTerraform(struct.enableInteractionLogging),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        enable_stackdriver_logging: {
            value: cdktf.booleanToHclTerraform(struct.enableStackdriverLogging),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowAdvancedSettingsLoggingSettingsToHclTerraform = dialogflowCxFlowAdvancedSettingsLoggingSettingsToHclTerraform;
class DialogflowCxFlowAdvancedSettingsLoggingSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableConsentBasedRedaction !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableConsentBasedRedaction = this._enableConsentBasedRedaction;
        }
        if (this._enableInteractionLogging !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableInteractionLogging = this._enableInteractionLogging;
        }
        if (this._enableStackdriverLogging !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableStackdriverLogging = this._enableStackdriverLogging;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableConsentBasedRedaction = undefined;
            this._enableInteractionLogging = undefined;
            this._enableStackdriverLogging = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableConsentBasedRedaction = value.enableConsentBasedRedaction;
            this._enableInteractionLogging = value.enableInteractionLogging;
            this._enableStackdriverLogging = value.enableStackdriverLogging;
        }
    }
    get enableConsentBasedRedaction() {
        return this.getBooleanAttribute('enable_consent_based_redaction');
    }
    set enableConsentBasedRedaction(value) {
        this._enableConsentBasedRedaction = value;
    }
    resetEnableConsentBasedRedaction() {
        this._enableConsentBasedRedaction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableConsentBasedRedactionInput() {
        return this._enableConsentBasedRedaction;
    }
    get enableInteractionLogging() {
        return this.getBooleanAttribute('enable_interaction_logging');
    }
    set enableInteractionLogging(value) {
        this._enableInteractionLogging = value;
    }
    resetEnableInteractionLogging() {
        this._enableInteractionLogging = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableInteractionLoggingInput() {
        return this._enableInteractionLogging;
    }
    get enableStackdriverLogging() {
        return this.getBooleanAttribute('enable_stackdriver_logging');
    }
    set enableStackdriverLogging(value) {
        this._enableStackdriverLogging = value;
    }
    resetEnableStackdriverLogging() {
        this._enableStackdriverLogging = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableStackdriverLoggingInput() {
        return this._enableStackdriverLogging;
    }
}
exports.DialogflowCxFlowAdvancedSettingsLoggingSettingsOutputReference = DialogflowCxFlowAdvancedSettingsLoggingSettingsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowAdvancedSettingsLoggingSettingsOutputReference[_c] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowAdvancedSettingsLoggingSettingsOutputReference", version: "14.13.1" };
function dialogflowCxFlowAdvancedSettingsSpeechSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        endpointer_sensitivity: cdktf.numberToTerraform(struct.endpointerSensitivity),
        models: cdktf.hashMapper(cdktf.stringToTerraform)(struct.models),
        no_speech_timeout: cdktf.stringToTerraform(struct.noSpeechTimeout),
        use_timeout_based_endpointing: cdktf.booleanToTerraform(struct.useTimeoutBasedEndpointing),
    };
}
exports.dialogflowCxFlowAdvancedSettingsSpeechSettingsToTerraform = dialogflowCxFlowAdvancedSettingsSpeechSettingsToTerraform;
function dialogflowCxFlowAdvancedSettingsSpeechSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        endpointer_sensitivity: {
            value: cdktf.numberToHclTerraform(struct.endpointerSensitivity),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        models: {
            value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(struct.models),
            isBlock: false,
            type: "map",
            storageClassType: "stringMap",
        },
        no_speech_timeout: {
            value: cdktf.stringToHclTerraform(struct.noSpeechTimeout),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        use_timeout_based_endpointing: {
            value: cdktf.booleanToHclTerraform(struct.useTimeoutBasedEndpointing),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowAdvancedSettingsSpeechSettingsToHclTerraform = dialogflowCxFlowAdvancedSettingsSpeechSettingsToHclTerraform;
class DialogflowCxFlowAdvancedSettingsSpeechSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endpointerSensitivity !== undefined) {
            hasAnyValues = true;
            internalValueResult.endpointerSensitivity = this._endpointerSensitivity;
        }
        if (this._models !== undefined) {
            hasAnyValues = true;
            internalValueResult.models = this._models;
        }
        if (this._noSpeechTimeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.noSpeechTimeout = this._noSpeechTimeout;
        }
        if (this._useTimeoutBasedEndpointing !== undefined) {
            hasAnyValues = true;
            internalValueResult.useTimeoutBasedEndpointing = this._useTimeoutBasedEndpointing;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endpointerSensitivity = undefined;
            this._models = undefined;
            this._noSpeechTimeout = undefined;
            this._useTimeoutBasedEndpointing = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endpointerSensitivity = value.endpointerSensitivity;
            this._models = value.models;
            this._noSpeechTimeout = value.noSpeechTimeout;
            this._useTimeoutBasedEndpointing = value.useTimeoutBasedEndpointing;
        }
    }
    get endpointerSensitivity() {
        return this.getNumberAttribute('endpointer_sensitivity');
    }
    set endpointerSensitivity(value) {
        this._endpointerSensitivity = value;
    }
    resetEndpointerSensitivity() {
        this._endpointerSensitivity = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endpointerSensitivityInput() {
        return this._endpointerSensitivity;
    }
    get models() {
        return this.getStringMapAttribute('models');
    }
    set models(value) {
        this._models = value;
    }
    resetModels() {
        this._models = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelsInput() {
        return this._models;
    }
    get noSpeechTimeout() {
        return this.getStringAttribute('no_speech_timeout');
    }
    set noSpeechTimeout(value) {
        this._noSpeechTimeout = value;
    }
    resetNoSpeechTimeout() {
        this._noSpeechTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get noSpeechTimeoutInput() {
        return this._noSpeechTimeout;
    }
    get useTimeoutBasedEndpointing() {
        return this.getBooleanAttribute('use_timeout_based_endpointing');
    }
    set useTimeoutBasedEndpointing(value) {
        this._useTimeoutBasedEndpointing = value;
    }
    resetUseTimeoutBasedEndpointing() {
        this._useTimeoutBasedEndpointing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useTimeoutBasedEndpointingInput() {
        return this._useTimeoutBasedEndpointing;
    }
}
exports.DialogflowCxFlowAdvancedSettingsSpeechSettingsOutputReference = DialogflowCxFlowAdvancedSettingsSpeechSettingsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowAdvancedSettingsSpeechSettingsOutputReference[_d] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowAdvancedSettingsSpeechSettingsOutputReference", version: "14.13.1" };
function dialogflowCxFlowAdvancedSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audio_export_gcs_destination: dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToTerraform(struct.audioExportGcsDestination),
        dtmf_settings: dialogflowCxFlowAdvancedSettingsDtmfSettingsToTerraform(struct.dtmfSettings),
        logging_settings: dialogflowCxFlowAdvancedSettingsLoggingSettingsToTerraform(struct.loggingSettings),
        speech_settings: dialogflowCxFlowAdvancedSettingsSpeechSettingsToTerraform(struct.speechSettings),
    };
}
exports.dialogflowCxFlowAdvancedSettingsToTerraform = dialogflowCxFlowAdvancedSettingsToTerraform;
function dialogflowCxFlowAdvancedSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        audio_export_gcs_destination: {
            value: dialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationToHclTerraform(struct.audioExportGcsDestination),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationList",
        },
        dtmf_settings: {
            value: dialogflowCxFlowAdvancedSettingsDtmfSettingsToHclTerraform(struct.dtmfSettings),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowAdvancedSettingsDtmfSettingsList",
        },
        logging_settings: {
            value: dialogflowCxFlowAdvancedSettingsLoggingSettingsToHclTerraform(struct.loggingSettings),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowAdvancedSettingsLoggingSettingsList",
        },
        speech_settings: {
            value: dialogflowCxFlowAdvancedSettingsSpeechSettingsToHclTerraform(struct.speechSettings),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowAdvancedSettingsSpeechSettingsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowAdvancedSettingsToHclTerraform = dialogflowCxFlowAdvancedSettingsToHclTerraform;
class DialogflowCxFlowAdvancedSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // audio_export_gcs_destination - computed: false, optional: true, required: false
        this._audioExportGcsDestination = new DialogflowCxFlowAdvancedSettingsAudioExportGcsDestinationOutputReference(this, "audio_export_gcs_destination");
        // dtmf_settings - computed: false, optional: true, required: false
        this._dtmfSettings = new DialogflowCxFlowAdvancedSettingsDtmfSettingsOutputReference(this, "dtmf_settings");
        // logging_settings - computed: false, optional: true, required: false
        this._loggingSettings = new DialogflowCxFlowAdvancedSettingsLoggingSettingsOutputReference(this, "logging_settings");
        // speech_settings - computed: false, optional: true, required: false
        this._speechSettings = new DialogflowCxFlowAdvancedSettingsSpeechSettingsOutputReference(this, "speech_settings");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._audioExportGcsDestination?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.audioExportGcsDestination = this._audioExportGcsDestination?.internalValue;
        }
        if (this._dtmfSettings?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dtmfSettings = this._dtmfSettings?.internalValue;
        }
        if (this._loggingSettings?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.loggingSettings = this._loggingSettings?.internalValue;
        }
        if (this._speechSettings?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.speechSettings = this._speechSettings?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._audioExportGcsDestination.internalValue = undefined;
            this._dtmfSettings.internalValue = undefined;
            this._loggingSettings.internalValue = undefined;
            this._speechSettings.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._audioExportGcsDestination.internalValue = value.audioExportGcsDestination;
            this._dtmfSettings.internalValue = value.dtmfSettings;
            this._loggingSettings.internalValue = value.loggingSettings;
            this._speechSettings.internalValue = value.speechSettings;
        }
    }
    get audioExportGcsDestination() {
        return this._audioExportGcsDestination;
    }
    putAudioExportGcsDestination(value) {
        this._audioExportGcsDestination.internalValue = value;
    }
    resetAudioExportGcsDestination() {
        this._audioExportGcsDestination.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get audioExportGcsDestinationInput() {
        return this._audioExportGcsDestination.internalValue;
    }
    get dtmfSettings() {
        return this._dtmfSettings;
    }
    putDtmfSettings(value) {
        this._dtmfSettings.internalValue = value;
    }
    resetDtmfSettings() {
        this._dtmfSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dtmfSettingsInput() {
        return this._dtmfSettings.internalValue;
    }
    get loggingSettings() {
        return this._loggingSettings;
    }
    putLoggingSettings(value) {
        this._loggingSettings.internalValue = value;
    }
    resetLoggingSettings() {
        this._loggingSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loggingSettingsInput() {
        return this._loggingSettings.internalValue;
    }
    get speechSettings() {
        return this._speechSettings;
    }
    putSpeechSettings(value) {
        this._speechSettings.internalValue = value;
    }
    resetSpeechSettings() {
        this._speechSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get speechSettingsInput() {
        return this._speechSettings.internalValue;
    }
}
exports.DialogflowCxFlowAdvancedSettingsOutputReference = DialogflowCxFlowAdvancedSettingsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowAdvancedSettingsOutputReference[_e] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowAdvancedSettingsOutputReference", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cases: cdktf.stringToTerraform(struct.cases),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cases: {
            value: cdktf.stringToHclTerraform(struct.cases),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cases !== undefined) {
            hasAnyValues = true;
            internalValueResult.cases = this._cases;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cases = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cases = value.cases;
        }
    }
    get cases() {
        return this.getStringAttribute('cases');
    }
    set cases(value) {
        this._cases = value;
    }
    resetCases() {
        this._cases = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get casesInput() {
        return this._cases;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesOutputReference[_f] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesOutputReference", version: "14.13.1" };
class DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesList = DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesList;
_g = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesList[_g] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesList", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metadata: cdktf.stringToTerraform(struct.metadata),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        metadata: {
            value: cdktf.stringToHclTerraform(struct.metadata),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._metadata = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._metadata = value.metadata;
        }
    }
    get metadata() {
        return this.getStringAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessOutputReference;
_h = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessOutputReference[_h] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessOutputReference", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metadata: cdktf.stringToTerraform(struct.metadata),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        metadata: {
            value: cdktf.stringToHclTerraform(struct.metadata),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._metadata = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._metadata = value.metadata;
        }
    }
    get metadata() {
        return this.getStringAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffOutputReference[_j] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffOutputReference", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ssml: cdktf.stringToTerraform(struct.ssml),
        text: cdktf.stringToTerraform(struct.text),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ssml: {
            value: cdktf.stringToHclTerraform(struct.ssml),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        text: {
            value: cdktf.stringToHclTerraform(struct.text),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ssml !== undefined) {
            hasAnyValues = true;
            internalValueResult.ssml = this._ssml;
        }
        if (this._text !== undefined) {
            hasAnyValues = true;
            internalValueResult.text = this._text;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ssml = undefined;
            this._text = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ssml = value.ssml;
            this._text = value.text;
        }
    }
    // allow_playback_interruption - computed: true, optional: false, required: false
    get allowPlaybackInterruption() {
        return this.getBooleanAttribute('allow_playback_interruption');
    }
    get ssml() {
        return this.getStringAttribute('ssml');
    }
    set ssml(value) {
        this._ssml = value;
    }
    resetSsml() {
        this._ssml = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ssmlInput() {
        return this._ssml;
    }
    get text() {
        return this.getStringAttribute('text');
    }
    set text(value) {
        this._text = value;
    }
    resetText() {
        this._text = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get textInput() {
        return this._text;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextOutputReference;
_k = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextOutputReference[_k] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextOutputReference", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audio_uri: cdktf.stringToTerraform(struct.audioUri),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        audio_uri: {
            value: cdktf.stringToHclTerraform(struct.audioUri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._audioUri !== undefined) {
            hasAnyValues = true;
            internalValueResult.audioUri = this._audioUri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._audioUri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._audioUri = value.audioUri;
        }
    }
    // allow_playback_interruption - computed: true, optional: false, required: false
    get allowPlaybackInterruption() {
        return this.getBooleanAttribute('allow_playback_interruption');
    }
    get audioUri() {
        return this.getStringAttribute('audio_uri');
    }
    set audioUri(value) {
        this._audioUri = value;
    }
    // Temporarily expose input value. Use with caution.
    get audioUriInput() {
        return this._audioUri;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioOutputReference[_l] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioOutputReference", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        phone_number: cdktf.stringToTerraform(struct.phoneNumber),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        phone_number: {
            value: cdktf.stringToHclTerraform(struct.phoneNumber),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._phoneNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneNumber = this._phoneNumber;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._phoneNumber = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._phoneNumber = value.phoneNumber;
        }
    }
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
    set phoneNumber(value) {
        this._phoneNumber = value;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallOutputReference;
_m = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallOutputReference[_m] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallOutputReference", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        text: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.text),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        text: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.text),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._text !== undefined) {
            hasAnyValues = true;
            internalValueResult.text = this._text;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._text = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._text = value.text;
        }
    }
    // allow_playback_interruption - computed: true, optional: false, required: false
    get allowPlaybackInterruption() {
        return this.getBooleanAttribute('allow_playback_interruption');
    }
    get text() {
        return this.getListAttribute('text');
    }
    set text(value) {
        this._text = value;
    }
    resetText() {
        this._text = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get textInput() {
        return this._text;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextOutputReference[_o] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextOutputReference", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        channel: cdktf.stringToTerraform(struct.channel),
        payload: cdktf.stringToTerraform(struct.payload),
        conversation_success: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToTerraform(struct.conversationSuccess),
        live_agent_handoff: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToTerraform(struct.liveAgentHandoff),
        output_audio_text: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToTerraform(struct.outputAudioText),
        play_audio: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToTerraform(struct.playAudio),
        telephony_transfer_call: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToTerraform(struct.telephonyTransferCall),
        text: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToTerraform(struct.text),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        channel: {
            value: cdktf.stringToHclTerraform(struct.channel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        payload: {
            value: cdktf.stringToHclTerraform(struct.payload),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        conversation_success: {
            value: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessToHclTerraform(struct.conversationSuccess),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessList",
        },
        live_agent_handoff: {
            value: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform(struct.liveAgentHandoff),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffList",
        },
        output_audio_text: {
            value: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextToHclTerraform(struct.outputAudioText),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextList",
        },
        play_audio: {
            value: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioToHclTerraform(struct.playAudio),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioList",
        },
        telephony_transfer_call: {
            value: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform(struct.telephonyTransferCall),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallList",
        },
        text: {
            value: dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextToHclTerraform(struct.text),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // conversation_success - computed: false, optional: true, required: false
        this._conversationSuccess = new DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesConversationSuccessOutputReference(this, "conversation_success");
        // live_agent_handoff - computed: false, optional: true, required: false
        this._liveAgentHandoff = new DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesLiveAgentHandoffOutputReference(this, "live_agent_handoff");
        // output_audio_text - computed: false, optional: true, required: false
        this._outputAudioText = new DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputAudioTextOutputReference(this, "output_audio_text");
        // play_audio - computed: false, optional: true, required: false
        this._playAudio = new DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesPlayAudioOutputReference(this, "play_audio");
        // telephony_transfer_call - computed: false, optional: true, required: false
        this._telephonyTransferCall = new DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTelephonyTransferCallOutputReference(this, "telephony_transfer_call");
        // text - computed: false, optional: true, required: false
        this._text = new DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesTextOutputReference(this, "text");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._channel !== undefined) {
            hasAnyValues = true;
            internalValueResult.channel = this._channel;
        }
        if (this._payload !== undefined) {
            hasAnyValues = true;
            internalValueResult.payload = this._payload;
        }
        if (this._conversationSuccess?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.conversationSuccess = this._conversationSuccess?.internalValue;
        }
        if (this._liveAgentHandoff?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.liveAgentHandoff = this._liveAgentHandoff?.internalValue;
        }
        if (this._outputAudioText?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.outputAudioText = this._outputAudioText?.internalValue;
        }
        if (this._playAudio?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.playAudio = this._playAudio?.internalValue;
        }
        if (this._telephonyTransferCall?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.telephonyTransferCall = this._telephonyTransferCall?.internalValue;
        }
        if (this._text?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.text = this._text?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._channel = undefined;
            this._payload = undefined;
            this._conversationSuccess.internalValue = undefined;
            this._liveAgentHandoff.internalValue = undefined;
            this._outputAudioText.internalValue = undefined;
            this._playAudio.internalValue = undefined;
            this._telephonyTransferCall.internalValue = undefined;
            this._text.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._channel = value.channel;
            this._payload = value.payload;
            this._conversationSuccess.internalValue = value.conversationSuccess;
            this._liveAgentHandoff.internalValue = value.liveAgentHandoff;
            this._outputAudioText.internalValue = value.outputAudioText;
            this._playAudio.internalValue = value.playAudio;
            this._telephonyTransferCall.internalValue = value.telephonyTransferCall;
            this._text.internalValue = value.text;
        }
    }
    get channel() {
        return this.getStringAttribute('channel');
    }
    set channel(value) {
        this._channel = value;
    }
    resetChannel() {
        this._channel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get channelInput() {
        return this._channel;
    }
    get payload() {
        return this.getStringAttribute('payload');
    }
    set payload(value) {
        this._payload = value;
    }
    resetPayload() {
        this._payload = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get payloadInput() {
        return this._payload;
    }
    get conversationSuccess() {
        return this._conversationSuccess;
    }
    putConversationSuccess(value) {
        this._conversationSuccess.internalValue = value;
    }
    resetConversationSuccess() {
        this._conversationSuccess.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get conversationSuccessInput() {
        return this._conversationSuccess.internalValue;
    }
    get liveAgentHandoff() {
        return this._liveAgentHandoff;
    }
    putLiveAgentHandoff(value) {
        this._liveAgentHandoff.internalValue = value;
    }
    resetLiveAgentHandoff() {
        this._liveAgentHandoff.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get liveAgentHandoffInput() {
        return this._liveAgentHandoff.internalValue;
    }
    get outputAudioText() {
        return this._outputAudioText;
    }
    putOutputAudioText(value) {
        this._outputAudioText.internalValue = value;
    }
    resetOutputAudioText() {
        this._outputAudioText.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputAudioTextInput() {
        return this._outputAudioText.internalValue;
    }
    get playAudio() {
        return this._playAudio;
    }
    putPlayAudio(value) {
        this._playAudio.internalValue = value;
    }
    resetPlayAudio() {
        this._playAudio.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get playAudioInput() {
        return this._playAudio.internalValue;
    }
    get telephonyTransferCall() {
        return this._telephonyTransferCall;
    }
    putTelephonyTransferCall(value) {
        this._telephonyTransferCall.internalValue = value;
    }
    resetTelephonyTransferCall() {
        this._telephonyTransferCall.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get telephonyTransferCallInput() {
        return this._telephonyTransferCall.internalValue;
    }
    get text() {
        return this._text;
    }
    putText(value) {
        this._text.internalValue = value;
    }
    resetText() {
        this._text.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get textInput() {
        return this._text.internalValue;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputReference;
_p = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputReference[_p] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputReference", version: "14.13.1" };
class DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesList = DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesList;
_q = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesList[_q] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesList", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        parameter: cdktf.stringToTerraform(struct.parameter),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        parameter: {
            value: cdktf.stringToHclTerraform(struct.parameter),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._parameter !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameter = this._parameter;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._parameter = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._parameter = value.parameter;
            this._value = value.value;
        }
    }
    get parameter() {
        return this.getStringAttribute('parameter');
    }
    set parameter(value) {
        this._parameter = value;
    }
    resetParameter() {
        this._parameter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterInput() {
        return this._parameter;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsOutputReference[_r] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsOutputReference", version: "14.13.1" };
class DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsList = DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsList;
_s = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsList[_s] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsList", version: "14.13.1" };
function dialogflowCxFlowEventHandlersTriggerFulfillmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        return_partial_responses: cdktf.booleanToTerraform(struct.returnPartialResponses),
        tag: cdktf.stringToTerraform(struct.tag),
        webhook: cdktf.stringToTerraform(struct.webhook),
        conditional_cases: cdktf.listMapper(dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToTerraform, true)(struct.conditionalCases),
        messages: cdktf.listMapper(dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToTerraform, true)(struct.messages),
        set_parameter_actions: cdktf.listMapper(dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToTerraform, true)(struct.setParameterActions),
    };
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentToTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentToTerraform;
function dialogflowCxFlowEventHandlersTriggerFulfillmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        return_partial_responses: {
            value: cdktf.booleanToHclTerraform(struct.returnPartialResponses),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        tag: {
            value: cdktf.stringToHclTerraform(struct.tag),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        webhook: {
            value: cdktf.stringToHclTerraform(struct.webhook),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        conditional_cases: {
            value: cdktf.listMapperHcl(dialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesToHclTerraform, true)(struct.conditionalCases),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesList",
        },
        messages: {
            value: cdktf.listMapperHcl(dialogflowCxFlowEventHandlersTriggerFulfillmentMessagesToHclTerraform, true)(struct.messages),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesList",
        },
        set_parameter_actions: {
            value: cdktf.listMapperHcl(dialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsToHclTerraform, true)(struct.setParameterActions),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersTriggerFulfillmentToHclTerraform = dialogflowCxFlowEventHandlersTriggerFulfillmentToHclTerraform;
class DialogflowCxFlowEventHandlersTriggerFulfillmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // conditional_cases - computed: false, optional: true, required: false
        this._conditionalCases = new DialogflowCxFlowEventHandlersTriggerFulfillmentConditionalCasesList(this, "conditional_cases", false);
        // messages - computed: false, optional: true, required: false
        this._messages = new DialogflowCxFlowEventHandlersTriggerFulfillmentMessagesList(this, "messages", false);
        // set_parameter_actions - computed: false, optional: true, required: false
        this._setParameterActions = new DialogflowCxFlowEventHandlersTriggerFulfillmentSetParameterActionsList(this, "set_parameter_actions", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._returnPartialResponses !== undefined) {
            hasAnyValues = true;
            internalValueResult.returnPartialResponses = this._returnPartialResponses;
        }
        if (this._tag !== undefined) {
            hasAnyValues = true;
            internalValueResult.tag = this._tag;
        }
        if (this._webhook !== undefined) {
            hasAnyValues = true;
            internalValueResult.webhook = this._webhook;
        }
        if (this._conditionalCases?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.conditionalCases = this._conditionalCases?.internalValue;
        }
        if (this._messages?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.messages = this._messages?.internalValue;
        }
        if (this._setParameterActions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.setParameterActions = this._setParameterActions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._returnPartialResponses = undefined;
            this._tag = undefined;
            this._webhook = undefined;
            this._conditionalCases.internalValue = undefined;
            this._messages.internalValue = undefined;
            this._setParameterActions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._returnPartialResponses = value.returnPartialResponses;
            this._tag = value.tag;
            this._webhook = value.webhook;
            this._conditionalCases.internalValue = value.conditionalCases;
            this._messages.internalValue = value.messages;
            this._setParameterActions.internalValue = value.setParameterActions;
        }
    }
    get returnPartialResponses() {
        return this.getBooleanAttribute('return_partial_responses');
    }
    set returnPartialResponses(value) {
        this._returnPartialResponses = value;
    }
    resetReturnPartialResponses() {
        this._returnPartialResponses = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get returnPartialResponsesInput() {
        return this._returnPartialResponses;
    }
    get tag() {
        return this.getStringAttribute('tag');
    }
    set tag(value) {
        this._tag = value;
    }
    resetTag() {
        this._tag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagInput() {
        return this._tag;
    }
    get webhook() {
        return this.getStringAttribute('webhook');
    }
    set webhook(value) {
        this._webhook = value;
    }
    resetWebhook() {
        this._webhook = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get webhookInput() {
        return this._webhook;
    }
    get conditionalCases() {
        return this._conditionalCases;
    }
    putConditionalCases(value) {
        this._conditionalCases.internalValue = value;
    }
    resetConditionalCases() {
        this._conditionalCases.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get conditionalCasesInput() {
        return this._conditionalCases.internalValue;
    }
    get messages() {
        return this._messages;
    }
    putMessages(value) {
        this._messages.internalValue = value;
    }
    resetMessages() {
        this._messages.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get messagesInput() {
        return this._messages.internalValue;
    }
    get setParameterActions() {
        return this._setParameterActions;
    }
    putSetParameterActions(value) {
        this._setParameterActions.internalValue = value;
    }
    resetSetParameterActions() {
        this._setParameterActions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setParameterActionsInput() {
        return this._setParameterActions.internalValue;
    }
}
exports.DialogflowCxFlowEventHandlersTriggerFulfillmentOutputReference = DialogflowCxFlowEventHandlersTriggerFulfillmentOutputReference;
_t = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersTriggerFulfillmentOutputReference[_t] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersTriggerFulfillmentOutputReference", version: "14.13.1" };
function dialogflowCxFlowEventHandlersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event: cdktf.stringToTerraform(struct.event),
        target_flow: cdktf.stringToTerraform(struct.targetFlow),
        target_page: cdktf.stringToTerraform(struct.targetPage),
        trigger_fulfillment: dialogflowCxFlowEventHandlersTriggerFulfillmentToTerraform(struct.triggerFulfillment),
    };
}
exports.dialogflowCxFlowEventHandlersToTerraform = dialogflowCxFlowEventHandlersToTerraform;
function dialogflowCxFlowEventHandlersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        event: {
            value: cdktf.stringToHclTerraform(struct.event),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        target_flow: {
            value: cdktf.stringToHclTerraform(struct.targetFlow),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        target_page: {
            value: cdktf.stringToHclTerraform(struct.targetPage),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        trigger_fulfillment: {
            value: dialogflowCxFlowEventHandlersTriggerFulfillmentToHclTerraform(struct.triggerFulfillment),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowEventHandlersTriggerFulfillmentList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowEventHandlersToHclTerraform = dialogflowCxFlowEventHandlersToHclTerraform;
class DialogflowCxFlowEventHandlersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // trigger_fulfillment - computed: false, optional: true, required: false
        this._triggerFulfillment = new DialogflowCxFlowEventHandlersTriggerFulfillmentOutputReference(this, "trigger_fulfillment");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._event !== undefined) {
            hasAnyValues = true;
            internalValueResult.event = this._event;
        }
        if (this._targetFlow !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetFlow = this._targetFlow;
        }
        if (this._targetPage !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetPage = this._targetPage;
        }
        if (this._triggerFulfillment?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.triggerFulfillment = this._triggerFulfillment?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._event = undefined;
            this._targetFlow = undefined;
            this._targetPage = undefined;
            this._triggerFulfillment.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._event = value.event;
            this._targetFlow = value.targetFlow;
            this._targetPage = value.targetPage;
            this._triggerFulfillment.internalValue = value.triggerFulfillment;
        }
    }
    get event() {
        return this.getStringAttribute('event');
    }
    set event(value) {
        this._event = value;
    }
    resetEvent() {
        this._event = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventInput() {
        return this._event;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get targetFlow() {
        return this.getStringAttribute('target_flow');
    }
    set targetFlow(value) {
        this._targetFlow = value;
    }
    resetTargetFlow() {
        this._targetFlow = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetFlowInput() {
        return this._targetFlow;
    }
    get targetPage() {
        return this.getStringAttribute('target_page');
    }
    set targetPage(value) {
        this._targetPage = value;
    }
    resetTargetPage() {
        this._targetPage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetPageInput() {
        return this._targetPage;
    }
    get triggerFulfillment() {
        return this._triggerFulfillment;
    }
    putTriggerFulfillment(value) {
        this._triggerFulfillment.internalValue = value;
    }
    resetTriggerFulfillment() {
        this._triggerFulfillment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get triggerFulfillmentInput() {
        return this._triggerFulfillment.internalValue;
    }
}
exports.DialogflowCxFlowEventHandlersOutputReference = DialogflowCxFlowEventHandlersOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersOutputReference[_u] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersOutputReference", version: "14.13.1" };
class DialogflowCxFlowEventHandlersList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DialogflowCxFlowEventHandlersOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DialogflowCxFlowEventHandlersList = DialogflowCxFlowEventHandlersList;
_v = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowEventHandlersList[_v] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowEventHandlersList", version: "14.13.1" };
function dialogflowCxFlowNluSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        classification_threshold: cdktf.numberToTerraform(struct.classificationThreshold),
        model_training_mode: cdktf.stringToTerraform(struct.modelTrainingMode),
        model_type: cdktf.stringToTerraform(struct.modelType),
    };
}
exports.dialogflowCxFlowNluSettingsToTerraform = dialogflowCxFlowNluSettingsToTerraform;
function dialogflowCxFlowNluSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        classification_threshold: {
            value: cdktf.numberToHclTerraform(struct.classificationThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        model_training_mode: {
            value: cdktf.stringToHclTerraform(struct.modelTrainingMode),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        model_type: {
            value: cdktf.stringToHclTerraform(struct.modelType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowNluSettingsToHclTerraform = dialogflowCxFlowNluSettingsToHclTerraform;
class DialogflowCxFlowNluSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._classificationThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.classificationThreshold = this._classificationThreshold;
        }
        if (this._modelTrainingMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelTrainingMode = this._modelTrainingMode;
        }
        if (this._modelType !== undefined) {
            hasAnyValues = true;
            internalValueResult.modelType = this._modelType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._classificationThreshold = undefined;
            this._modelTrainingMode = undefined;
            this._modelType = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._classificationThreshold = value.classificationThreshold;
            this._modelTrainingMode = value.modelTrainingMode;
            this._modelType = value.modelType;
        }
    }
    get classificationThreshold() {
        return this.getNumberAttribute('classification_threshold');
    }
    set classificationThreshold(value) {
        this._classificationThreshold = value;
    }
    resetClassificationThreshold() {
        this._classificationThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get classificationThresholdInput() {
        return this._classificationThreshold;
    }
    get modelTrainingMode() {
        return this.getStringAttribute('model_training_mode');
    }
    set modelTrainingMode(value) {
        this._modelTrainingMode = value;
    }
    resetModelTrainingMode() {
        this._modelTrainingMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelTrainingModeInput() {
        return this._modelTrainingMode;
    }
    get modelType() {
        return this.getStringAttribute('model_type');
    }
    set modelType(value) {
        this._modelType = value;
    }
    resetModelType() {
        this._modelType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modelTypeInput() {
        return this._modelType;
    }
}
exports.DialogflowCxFlowNluSettingsOutputReference = DialogflowCxFlowNluSettingsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowNluSettingsOutputReference[_w] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowNluSettingsOutputReference", version: "14.13.1" };
function dialogflowCxFlowTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dialogflowCxFlowTimeoutsToTerraform = dialogflowCxFlowTimeoutsToTerraform;
function dialogflowCxFlowTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTimeoutsToHclTerraform = dialogflowCxFlowTimeoutsToHclTerraform;
class DialogflowCxFlowTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DialogflowCxFlowTimeoutsOutputReference = DialogflowCxFlowTimeoutsOutputReference;
_x = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTimeoutsOutputReference[_x] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTimeoutsOutputReference", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cases: cdktf.stringToTerraform(struct.cases),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cases: {
            value: cdktf.stringToHclTerraform(struct.cases),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cases !== undefined) {
            hasAnyValues = true;
            internalValueResult.cases = this._cases;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._cases = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._cases = value.cases;
        }
    }
    get cases() {
        return this.getStringAttribute('cases');
    }
    set cases(value) {
        this._cases = value;
    }
    resetCases() {
        this._cases = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get casesInput() {
        return this._cases;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesOutputReference[_y] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesOutputReference", version: "14.13.1" };
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesList = DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesList;
_z = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesList[_z] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesList", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metadata: cdktf.stringToTerraform(struct.metadata),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        metadata: {
            value: cdktf.stringToHclTerraform(struct.metadata),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._metadata = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._metadata = value.metadata;
        }
    }
    get metadata() {
        return this.getStringAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessOutputReference[_0] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessOutputReference", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metadata: cdktf.stringToTerraform(struct.metadata),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        metadata: {
            value: cdktf.stringToHclTerraform(struct.metadata),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metadata !== undefined) {
            hasAnyValues = true;
            internalValueResult.metadata = this._metadata;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._metadata = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._metadata = value.metadata;
        }
    }
    get metadata() {
        return this.getStringAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffOutputReference[_1] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffOutputReference", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ssml: cdktf.stringToTerraform(struct.ssml),
        text: cdktf.stringToTerraform(struct.text),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ssml: {
            value: cdktf.stringToHclTerraform(struct.ssml),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        text: {
            value: cdktf.stringToHclTerraform(struct.text),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ssml !== undefined) {
            hasAnyValues = true;
            internalValueResult.ssml = this._ssml;
        }
        if (this._text !== undefined) {
            hasAnyValues = true;
            internalValueResult.text = this._text;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ssml = undefined;
            this._text = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ssml = value.ssml;
            this._text = value.text;
        }
    }
    // allow_playback_interruption - computed: true, optional: false, required: false
    get allowPlaybackInterruption() {
        return this.getBooleanAttribute('allow_playback_interruption');
    }
    get ssml() {
        return this.getStringAttribute('ssml');
    }
    set ssml(value) {
        this._ssml = value;
    }
    resetSsml() {
        this._ssml = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ssmlInput() {
        return this._ssml;
    }
    get text() {
        return this.getStringAttribute('text');
    }
    set text(value) {
        this._text = value;
    }
    resetText() {
        this._text = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get textInput() {
        return this._text;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextOutputReference[_2] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextOutputReference", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audio_uri: cdktf.stringToTerraform(struct.audioUri),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        audio_uri: {
            value: cdktf.stringToHclTerraform(struct.audioUri),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._audioUri !== undefined) {
            hasAnyValues = true;
            internalValueResult.audioUri = this._audioUri;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._audioUri = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._audioUri = value.audioUri;
        }
    }
    // allow_playback_interruption - computed: true, optional: false, required: false
    get allowPlaybackInterruption() {
        return this.getBooleanAttribute('allow_playback_interruption');
    }
    get audioUri() {
        return this.getStringAttribute('audio_uri');
    }
    set audioUri(value) {
        this._audioUri = value;
    }
    // Temporarily expose input value. Use with caution.
    get audioUriInput() {
        return this._audioUri;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioOutputReference;
_3 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioOutputReference[_3] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioOutputReference", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        phone_number: cdktf.stringToTerraform(struct.phoneNumber),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        phone_number: {
            value: cdktf.stringToHclTerraform(struct.phoneNumber),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._phoneNumber !== undefined) {
            hasAnyValues = true;
            internalValueResult.phoneNumber = this._phoneNumber;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._phoneNumber = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._phoneNumber = value.phoneNumber;
        }
    }
    get phoneNumber() {
        return this.getStringAttribute('phone_number');
    }
    set phoneNumber(value) {
        this._phoneNumber = value;
    }
    // Temporarily expose input value. Use with caution.
    get phoneNumberInput() {
        return this._phoneNumber;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallOutputReference[_4] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallOutputReference", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        text: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.text),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        text: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.text),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._text !== undefined) {
            hasAnyValues = true;
            internalValueResult.text = this._text;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._text = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._text = value.text;
        }
    }
    // allow_playback_interruption - computed: true, optional: false, required: false
    get allowPlaybackInterruption() {
        return this.getBooleanAttribute('allow_playback_interruption');
    }
    get text() {
        return this.getListAttribute('text');
    }
    set text(value) {
        this._text = value;
    }
    resetText() {
        this._text = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get textInput() {
        return this._text;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextOutputReference[_5] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextOutputReference", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        channel: cdktf.stringToTerraform(struct.channel),
        payload: cdktf.stringToTerraform(struct.payload),
        conversation_success: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToTerraform(struct.conversationSuccess),
        live_agent_handoff: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToTerraform(struct.liveAgentHandoff),
        output_audio_text: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToTerraform(struct.outputAudioText),
        play_audio: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToTerraform(struct.playAudio),
        telephony_transfer_call: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToTerraform(struct.telephonyTransferCall),
        text: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToTerraform(struct.text),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        channel: {
            value: cdktf.stringToHclTerraform(struct.channel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        payload: {
            value: cdktf.stringToHclTerraform(struct.payload),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        conversation_success: {
            value: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessToHclTerraform(struct.conversationSuccess),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessList",
        },
        live_agent_handoff: {
            value: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffToHclTerraform(struct.liveAgentHandoff),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffList",
        },
        output_audio_text: {
            value: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextToHclTerraform(struct.outputAudioText),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextList",
        },
        play_audio: {
            value: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioToHclTerraform(struct.playAudio),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioList",
        },
        telephony_transfer_call: {
            value: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallToHclTerraform(struct.telephonyTransferCall),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallList",
        },
        text: {
            value: dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextToHclTerraform(struct.text),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // conversation_success - computed: false, optional: true, required: false
        this._conversationSuccess = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesConversationSuccessOutputReference(this, "conversation_success");
        // live_agent_handoff - computed: false, optional: true, required: false
        this._liveAgentHandoff = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesLiveAgentHandoffOutputReference(this, "live_agent_handoff");
        // output_audio_text - computed: false, optional: true, required: false
        this._outputAudioText = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputAudioTextOutputReference(this, "output_audio_text");
        // play_audio - computed: false, optional: true, required: false
        this._playAudio = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesPlayAudioOutputReference(this, "play_audio");
        // telephony_transfer_call - computed: false, optional: true, required: false
        this._telephonyTransferCall = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTelephonyTransferCallOutputReference(this, "telephony_transfer_call");
        // text - computed: false, optional: true, required: false
        this._text = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesTextOutputReference(this, "text");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._channel !== undefined) {
            hasAnyValues = true;
            internalValueResult.channel = this._channel;
        }
        if (this._payload !== undefined) {
            hasAnyValues = true;
            internalValueResult.payload = this._payload;
        }
        if (this._conversationSuccess?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.conversationSuccess = this._conversationSuccess?.internalValue;
        }
        if (this._liveAgentHandoff?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.liveAgentHandoff = this._liveAgentHandoff?.internalValue;
        }
        if (this._outputAudioText?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.outputAudioText = this._outputAudioText?.internalValue;
        }
        if (this._playAudio?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.playAudio = this._playAudio?.internalValue;
        }
        if (this._telephonyTransferCall?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.telephonyTransferCall = this._telephonyTransferCall?.internalValue;
        }
        if (this._text?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.text = this._text?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._channel = undefined;
            this._payload = undefined;
            this._conversationSuccess.internalValue = undefined;
            this._liveAgentHandoff.internalValue = undefined;
            this._outputAudioText.internalValue = undefined;
            this._playAudio.internalValue = undefined;
            this._telephonyTransferCall.internalValue = undefined;
            this._text.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._channel = value.channel;
            this._payload = value.payload;
            this._conversationSuccess.internalValue = value.conversationSuccess;
            this._liveAgentHandoff.internalValue = value.liveAgentHandoff;
            this._outputAudioText.internalValue = value.outputAudioText;
            this._playAudio.internalValue = value.playAudio;
            this._telephonyTransferCall.internalValue = value.telephonyTransferCall;
            this._text.internalValue = value.text;
        }
    }
    get channel() {
        return this.getStringAttribute('channel');
    }
    set channel(value) {
        this._channel = value;
    }
    resetChannel() {
        this._channel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get channelInput() {
        return this._channel;
    }
    get payload() {
        return this.getStringAttribute('payload');
    }
    set payload(value) {
        this._payload = value;
    }
    resetPayload() {
        this._payload = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get payloadInput() {
        return this._payload;
    }
    get conversationSuccess() {
        return this._conversationSuccess;
    }
    putConversationSuccess(value) {
        this._conversationSuccess.internalValue = value;
    }
    resetConversationSuccess() {
        this._conversationSuccess.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get conversationSuccessInput() {
        return this._conversationSuccess.internalValue;
    }
    get liveAgentHandoff() {
        return this._liveAgentHandoff;
    }
    putLiveAgentHandoff(value) {
        this._liveAgentHandoff.internalValue = value;
    }
    resetLiveAgentHandoff() {
        this._liveAgentHandoff.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get liveAgentHandoffInput() {
        return this._liveAgentHandoff.internalValue;
    }
    get outputAudioText() {
        return this._outputAudioText;
    }
    putOutputAudioText(value) {
        this._outputAudioText.internalValue = value;
    }
    resetOutputAudioText() {
        this._outputAudioText.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outputAudioTextInput() {
        return this._outputAudioText.internalValue;
    }
    get playAudio() {
        return this._playAudio;
    }
    putPlayAudio(value) {
        this._playAudio.internalValue = value;
    }
    resetPlayAudio() {
        this._playAudio.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get playAudioInput() {
        return this._playAudio.internalValue;
    }
    get telephonyTransferCall() {
        return this._telephonyTransferCall;
    }
    putTelephonyTransferCall(value) {
        this._telephonyTransferCall.internalValue = value;
    }
    resetTelephonyTransferCall() {
        this._telephonyTransferCall.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get telephonyTransferCallInput() {
        return this._telephonyTransferCall.internalValue;
    }
    get text() {
        return this._text;
    }
    putText(value) {
        this._text.internalValue = value;
    }
    resetText() {
        this._text.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get textInput() {
        return this._text.internalValue;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputReference[_6] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputReference", version: "14.13.1" };
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesList = DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesList;
_7 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesList[_7] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesList", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        parameter: cdktf.stringToTerraform(struct.parameter),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        parameter: {
            value: cdktf.stringToHclTerraform(struct.parameter),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._parameter !== undefined) {
            hasAnyValues = true;
            internalValueResult.parameter = this._parameter;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._parameter = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._parameter = value.parameter;
            this._value = value.value;
        }
    }
    get parameter() {
        return this.getStringAttribute('parameter');
    }
    set parameter(value) {
        this._parameter = value;
    }
    resetParameter() {
        this._parameter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parameterInput() {
        return this._parameter;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsOutputReference[_8] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsOutputReference", version: "14.13.1" };
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsList = DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsList;
_9 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsList[_9] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsList", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        return_partial_responses: cdktf.booleanToTerraform(struct.returnPartialResponses),
        tag: cdktf.stringToTerraform(struct.tag),
        webhook: cdktf.stringToTerraform(struct.webhook),
        conditional_cases: cdktf.listMapper(dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToTerraform, true)(struct.conditionalCases),
        messages: cdktf.listMapper(dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToTerraform, true)(struct.messages),
        set_parameter_actions: cdktf.listMapper(dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToTerraform, true)(struct.setParameterActions),
    };
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentToTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentToTerraform;
function dialogflowCxFlowTransitionRoutesTriggerFulfillmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        return_partial_responses: {
            value: cdktf.booleanToHclTerraform(struct.returnPartialResponses),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        tag: {
            value: cdktf.stringToHclTerraform(struct.tag),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        webhook: {
            value: cdktf.stringToHclTerraform(struct.webhook),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        conditional_cases: {
            value: cdktf.listMapperHcl(dialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesToHclTerraform, true)(struct.conditionalCases),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesList",
        },
        messages: {
            value: cdktf.listMapperHcl(dialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesToHclTerraform, true)(struct.messages),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesList",
        },
        set_parameter_actions: {
            value: cdktf.listMapperHcl(dialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsToHclTerraform, true)(struct.setParameterActions),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesTriggerFulfillmentToHclTerraform = dialogflowCxFlowTransitionRoutesTriggerFulfillmentToHclTerraform;
class DialogflowCxFlowTransitionRoutesTriggerFulfillmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // conditional_cases - computed: false, optional: true, required: false
        this._conditionalCases = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentConditionalCasesList(this, "conditional_cases", false);
        // messages - computed: false, optional: true, required: false
        this._messages = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentMessagesList(this, "messages", false);
        // set_parameter_actions - computed: false, optional: true, required: false
        this._setParameterActions = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentSetParameterActionsList(this, "set_parameter_actions", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._returnPartialResponses !== undefined) {
            hasAnyValues = true;
            internalValueResult.returnPartialResponses = this._returnPartialResponses;
        }
        if (this._tag !== undefined) {
            hasAnyValues = true;
            internalValueResult.tag = this._tag;
        }
        if (this._webhook !== undefined) {
            hasAnyValues = true;
            internalValueResult.webhook = this._webhook;
        }
        if (this._conditionalCases?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.conditionalCases = this._conditionalCases?.internalValue;
        }
        if (this._messages?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.messages = this._messages?.internalValue;
        }
        if (this._setParameterActions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.setParameterActions = this._setParameterActions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._returnPartialResponses = undefined;
            this._tag = undefined;
            this._webhook = undefined;
            this._conditionalCases.internalValue = undefined;
            this._messages.internalValue = undefined;
            this._setParameterActions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._returnPartialResponses = value.returnPartialResponses;
            this._tag = value.tag;
            this._webhook = value.webhook;
            this._conditionalCases.internalValue = value.conditionalCases;
            this._messages.internalValue = value.messages;
            this._setParameterActions.internalValue = value.setParameterActions;
        }
    }
    get returnPartialResponses() {
        return this.getBooleanAttribute('return_partial_responses');
    }
    set returnPartialResponses(value) {
        this._returnPartialResponses = value;
    }
    resetReturnPartialResponses() {
        this._returnPartialResponses = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get returnPartialResponsesInput() {
        return this._returnPartialResponses;
    }
    get tag() {
        return this.getStringAttribute('tag');
    }
    set tag(value) {
        this._tag = value;
    }
    resetTag() {
        this._tag = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tagInput() {
        return this._tag;
    }
    get webhook() {
        return this.getStringAttribute('webhook');
    }
    set webhook(value) {
        this._webhook = value;
    }
    resetWebhook() {
        this._webhook = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get webhookInput() {
        return this._webhook;
    }
    get conditionalCases() {
        return this._conditionalCases;
    }
    putConditionalCases(value) {
        this._conditionalCases.internalValue = value;
    }
    resetConditionalCases() {
        this._conditionalCases.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get conditionalCasesInput() {
        return this._conditionalCases.internalValue;
    }
    get messages() {
        return this._messages;
    }
    putMessages(value) {
        this._messages.internalValue = value;
    }
    resetMessages() {
        this._messages.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get messagesInput() {
        return this._messages.internalValue;
    }
    get setParameterActions() {
        return this._setParameterActions;
    }
    putSetParameterActions(value) {
        this._setParameterActions.internalValue = value;
    }
    resetSetParameterActions() {
        this._setParameterActions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setParameterActionsInput() {
        return this._setParameterActions.internalValue;
    }
}
exports.DialogflowCxFlowTransitionRoutesTriggerFulfillmentOutputReference = DialogflowCxFlowTransitionRoutesTriggerFulfillmentOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesTriggerFulfillmentOutputReference[_10] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesTriggerFulfillmentOutputReference", version: "14.13.1" };
function dialogflowCxFlowTransitionRoutesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        condition: cdktf.stringToTerraform(struct.condition),
        intent: cdktf.stringToTerraform(struct.intent),
        target_flow: cdktf.stringToTerraform(struct.targetFlow),
        target_page: cdktf.stringToTerraform(struct.targetPage),
        trigger_fulfillment: dialogflowCxFlowTransitionRoutesTriggerFulfillmentToTerraform(struct.triggerFulfillment),
    };
}
exports.dialogflowCxFlowTransitionRoutesToTerraform = dialogflowCxFlowTransitionRoutesToTerraform;
function dialogflowCxFlowTransitionRoutesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        condition: {
            value: cdktf.stringToHclTerraform(struct.condition),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        intent: {
            value: cdktf.stringToHclTerraform(struct.intent),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        target_flow: {
            value: cdktf.stringToHclTerraform(struct.targetFlow),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        target_page: {
            value: cdktf.stringToHclTerraform(struct.targetPage),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        trigger_fulfillment: {
            value: dialogflowCxFlowTransitionRoutesTriggerFulfillmentToHclTerraform(struct.triggerFulfillment),
            isBlock: true,
            type: "list",
            storageClassType: "DialogflowCxFlowTransitionRoutesTriggerFulfillmentList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.dialogflowCxFlowTransitionRoutesToHclTerraform = dialogflowCxFlowTransitionRoutesToHclTerraform;
class DialogflowCxFlowTransitionRoutesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // trigger_fulfillment - computed: false, optional: true, required: false
        this._triggerFulfillment = new DialogflowCxFlowTransitionRoutesTriggerFulfillmentOutputReference(this, "trigger_fulfillment");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._condition !== undefined) {
            hasAnyValues = true;
            internalValueResult.condition = this._condition;
        }
        if (this._intent !== undefined) {
            hasAnyValues = true;
            internalValueResult.intent = this._intent;
        }
        if (this._targetFlow !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetFlow = this._targetFlow;
        }
        if (this._targetPage !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetPage = this._targetPage;
        }
        if (this._triggerFulfillment?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.triggerFulfillment = this._triggerFulfillment?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._condition = undefined;
            this._intent = undefined;
            this._targetFlow = undefined;
            this._targetPage = undefined;
            this._triggerFulfillment.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._condition = value.condition;
            this._intent = value.intent;
            this._targetFlow = value.targetFlow;
            this._targetPage = value.targetPage;
            this._triggerFulfillment.internalValue = value.triggerFulfillment;
        }
    }
    get condition() {
        return this.getStringAttribute('condition');
    }
    set condition(value) {
        this._condition = value;
    }
    resetCondition() {
        this._condition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get conditionInput() {
        return this._condition;
    }
    get intent() {
        return this.getStringAttribute('intent');
    }
    set intent(value) {
        this._intent = value;
    }
    resetIntent() {
        this._intent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get intentInput() {
        return this._intent;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get targetFlow() {
        return this.getStringAttribute('target_flow');
    }
    set targetFlow(value) {
        this._targetFlow = value;
    }
    resetTargetFlow() {
        this._targetFlow = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetFlowInput() {
        return this._targetFlow;
    }
    get targetPage() {
        return this.getStringAttribute('target_page');
    }
    set targetPage(value) {
        this._targetPage = value;
    }
    resetTargetPage() {
        this._targetPage = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetPageInput() {
        return this._targetPage;
    }
    get triggerFulfillment() {
        return this._triggerFulfillment;
    }
    putTriggerFulfillment(value) {
        this._triggerFulfillment.internalValue = value;
    }
    resetTriggerFulfillment() {
        this._triggerFulfillment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get triggerFulfillmentInput() {
        return this._triggerFulfillment.internalValue;
    }
}
exports.DialogflowCxFlowTransitionRoutesOutputReference = DialogflowCxFlowTransitionRoutesOutputReference;
_11 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesOutputReference[_11] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesOutputReference", version: "14.13.1" };
class DialogflowCxFlowTransitionRoutesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DialogflowCxFlowTransitionRoutesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DialogflowCxFlowTransitionRoutesList = DialogflowCxFlowTransitionRoutesList;
_12 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlowTransitionRoutesList[_12] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlowTransitionRoutesList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dialogflow_cx_flow google_dialogflow_cx_flow}
*/
class DialogflowCxFlow extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DialogflowCxFlow resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DialogflowCxFlow to import
    * @param importFromId The id of the existing DialogflowCxFlow that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dialogflow_cx_flow#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DialogflowCxFlow to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_dialogflow_cx_flow", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/dialogflow_cx_flow google_dialogflow_cx_flow} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DialogflowCxFlowConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_dialogflow_cx_flow',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // advanced_settings - computed: false, optional: true, required: false
        this._advancedSettings = new DialogflowCxFlowAdvancedSettingsOutputReference(this, "advanced_settings");
        // event_handlers - computed: false, optional: true, required: false
        this._eventHandlers = new DialogflowCxFlowEventHandlersList(this, "event_handlers", false);
        // nlu_settings - computed: false, optional: true, required: false
        this._nluSettings = new DialogflowCxFlowNluSettingsOutputReference(this, "nlu_settings");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DialogflowCxFlowTimeoutsOutputReference(this, "timeouts");
        // transition_routes - computed: false, optional: true, required: false
        this._transitionRoutes = new DialogflowCxFlowTransitionRoutesList(this, "transition_routes", false);
        this._description = config.description;
        this._displayName = config.displayName;
        this._id = config.id;
        this._isDefaultStartFlow = config.isDefaultStartFlow;
        this._languageCode = config.languageCode;
        this._parent = config.parent;
        this._transitionRouteGroups = config.transitionRouteGroups;
        this._advancedSettings.internalValue = config.advancedSettings;
        this._eventHandlers.internalValue = config.eventHandlers;
        this._nluSettings.internalValue = config.nluSettings;
        this._timeouts.internalValue = config.timeouts;
        this._transitionRoutes.internalValue = config.transitionRoutes;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get isDefaultStartFlow() {
        return this.getBooleanAttribute('is_default_start_flow');
    }
    set isDefaultStartFlow(value) {
        this._isDefaultStartFlow = value;
    }
    resetIsDefaultStartFlow() {
        this._isDefaultStartFlow = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isDefaultStartFlowInput() {
        return this._isDefaultStartFlow;
    }
    get languageCode() {
        return this.getStringAttribute('language_code');
    }
    set languageCode(value) {
        this._languageCode = value;
    }
    resetLanguageCode() {
        this._languageCode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get languageCodeInput() {
        return this._languageCode;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    resetParent() {
        this._parent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    get transitionRouteGroups() {
        return this.getListAttribute('transition_route_groups');
    }
    set transitionRouteGroups(value) {
        this._transitionRouteGroups = value;
    }
    resetTransitionRouteGroups() {
        this._transitionRouteGroups = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transitionRouteGroupsInput() {
        return this._transitionRouteGroups;
    }
    get advancedSettings() {
        return this._advancedSettings;
    }
    putAdvancedSettings(value) {
        this._advancedSettings.internalValue = value;
    }
    resetAdvancedSettings() {
        this._advancedSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get advancedSettingsInput() {
        return this._advancedSettings.internalValue;
    }
    get eventHandlers() {
        return this._eventHandlers;
    }
    putEventHandlers(value) {
        this._eventHandlers.internalValue = value;
    }
    resetEventHandlers() {
        this._eventHandlers.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventHandlersInput() {
        return this._eventHandlers.internalValue;
    }
    get nluSettings() {
        return this._nluSettings;
    }
    putNluSettings(value) {
        this._nluSettings.internalValue = value;
    }
    resetNluSettings() {
        this._nluSettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nluSettingsInput() {
        return this._nluSettings.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get transitionRoutes() {
        return this._transitionRoutes;
    }
    putTransitionRoutes(value) {
        this._transitionRoutes.internalValue = value;
    }
    resetTransitionRoutes() {
        this._transitionRoutes.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transitionRoutesInput() {
        return this._transitionRoutes.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            is_default_start_flow: cdktf.booleanToTerraform(this._isDefaultStartFlow),
            language_code: cdktf.stringToTerraform(this._languageCode),
            parent: cdktf.stringToTerraform(this._parent),
            transition_route_groups: cdktf.listMapper(cdktf.stringToTerraform, false)(this._transitionRouteGroups),
            advanced_settings: dialogflowCxFlowAdvancedSettingsToTerraform(this._advancedSettings.internalValue),
            event_handlers: cdktf.listMapper(dialogflowCxFlowEventHandlersToTerraform, true)(this._eventHandlers.internalValue),
            nlu_settings: dialogflowCxFlowNluSettingsToTerraform(this._nluSettings.internalValue),
            timeouts: dialogflowCxFlowTimeoutsToTerraform(this._timeouts.internalValue),
            transition_routes: cdktf.listMapper(dialogflowCxFlowTransitionRoutesToTerraform, true)(this._transitionRoutes.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            is_default_start_flow: {
                value: cdktf.booleanToHclTerraform(this._isDefaultStartFlow),
                isBlock: false,
                type: "simple",
                storageClassType: "boolean",
            },
            language_code: {
                value: cdktf.stringToHclTerraform(this._languageCode),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent: {
                value: cdktf.stringToHclTerraform(this._parent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            transition_route_groups: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._transitionRouteGroups),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            advanced_settings: {
                value: dialogflowCxFlowAdvancedSettingsToHclTerraform(this._advancedSettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DialogflowCxFlowAdvancedSettingsList",
            },
            event_handlers: {
                value: cdktf.listMapperHcl(dialogflowCxFlowEventHandlersToHclTerraform, true)(this._eventHandlers.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DialogflowCxFlowEventHandlersList",
            },
            nlu_settings: {
                value: dialogflowCxFlowNluSettingsToHclTerraform(this._nluSettings.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DialogflowCxFlowNluSettingsList",
            },
            timeouts: {
                value: dialogflowCxFlowTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "DialogflowCxFlowTimeouts",
            },
            transition_routes: {
                value: cdktf.listMapperHcl(dialogflowCxFlowTransitionRoutesToHclTerraform, true)(this._transitionRoutes.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "DialogflowCxFlowTransitionRoutesList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DialogflowCxFlow = DialogflowCxFlow;
_13 = JSII_RTTI_SYMBOL_1;
DialogflowCxFlow[_13] = { fqn: "@cdktf/provider-google.dialogflowCxFlow.DialogflowCxFlow", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DialogflowCxFlow.tfResourceType = "google_dialogflow_cx_flow";
