// @cdktf/provider-google/lib/bigquery-datapolicy-data-policy/index.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigqueryDatapolicyDataPolicy = exports.BigqueryDatapolicyDataPolicyTimeoutsOutputReference = exports.bigqueryDatapolicyDataPolicyTimeoutsToHclTerraform = exports.bigqueryDatapolicyDataPolicyTimeoutsToTerraform = exports.BigqueryDatapolicyDataPolicyDataMaskingPolicyOutputReference = exports.bigqueryDatapolicyDataPolicyDataMaskingPolicyToHclTerraform = exports.bigqueryDatapolicyDataPolicyDataMaskingPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function bigqueryDatapolicyDataPolicyDataMaskingPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        predefined_expression: cdktf.stringToTerraform(struct.predefinedExpression),
        routine: cdktf.stringToTerraform(struct.routine),
    };
}
exports.bigqueryDatapolicyDataPolicyDataMaskingPolicyToTerraform = bigqueryDatapolicyDataPolicyDataMaskingPolicyToTerraform;
function bigqueryDatapolicyDataPolicyDataMaskingPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        predefined_expression: {
            value: cdktf.stringToHclTerraform(struct.predefinedExpression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        routine: {
            value: cdktf.stringToHclTerraform(struct.routine),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryDatapolicyDataPolicyDataMaskingPolicyToHclTerraform = bigqueryDatapolicyDataPolicyDataMaskingPolicyToHclTerraform;
class BigqueryDatapolicyDataPolicyDataMaskingPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._predefinedExpression !== undefined) {
            hasAnyValues = true;
            internalValueResult.predefinedExpression = this._predefinedExpression;
        }
        if (this._routine !== undefined) {
            hasAnyValues = true;
            internalValueResult.routine = this._routine;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._predefinedExpression = undefined;
            this._routine = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._predefinedExpression = value.predefinedExpression;
            this._routine = value.routine;
        }
    }
    get predefinedExpression() {
        return this.getStringAttribute('predefined_expression');
    }
    set predefinedExpression(value) {
        this._predefinedExpression = value;
    }
    resetPredefinedExpression() {
        this._predefinedExpression = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get predefinedExpressionInput() {
        return this._predefinedExpression;
    }
    get routine() {
        return this.getStringAttribute('routine');
    }
    set routine(value) {
        this._routine = value;
    }
    resetRoutine() {
        this._routine = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get routineInput() {
        return this._routine;
    }
}
exports.BigqueryDatapolicyDataPolicyDataMaskingPolicyOutputReference = BigqueryDatapolicyDataPolicyDataMaskingPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BigqueryDatapolicyDataPolicyDataMaskingPolicyOutputReference[_a] = { fqn: "@cdktf/provider-google.bigqueryDatapolicyDataPolicy.BigqueryDatapolicyDataPolicyDataMaskingPolicyOutputReference", version: "14.13.1" };
function bigqueryDatapolicyDataPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.bigqueryDatapolicyDataPolicyTimeoutsToTerraform = bigqueryDatapolicyDataPolicyTimeoutsToTerraform;
function bigqueryDatapolicyDataPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryDatapolicyDataPolicyTimeoutsToHclTerraform = bigqueryDatapolicyDataPolicyTimeoutsToHclTerraform;
class BigqueryDatapolicyDataPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.BigqueryDatapolicyDataPolicyTimeoutsOutputReference = BigqueryDatapolicyDataPolicyTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
BigqueryDatapolicyDataPolicyTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-google.bigqueryDatapolicyDataPolicy.BigqueryDatapolicyDataPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_datapolicy_data_policy google_bigquery_datapolicy_data_policy}
*/
class BigqueryDatapolicyDataPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a BigqueryDatapolicyDataPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the BigqueryDatapolicyDataPolicy to import
    * @param importFromId The id of the existing BigqueryDatapolicyDataPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_datapolicy_data_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the BigqueryDatapolicyDataPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_bigquery_datapolicy_data_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_datapolicy_data_policy google_bigquery_datapolicy_data_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BigqueryDatapolicyDataPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_bigquery_datapolicy_data_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // data_masking_policy - computed: false, optional: true, required: false
        this._dataMaskingPolicy = new BigqueryDatapolicyDataPolicyDataMaskingPolicyOutputReference(this, "data_masking_policy");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new BigqueryDatapolicyDataPolicyTimeoutsOutputReference(this, "timeouts");
        this._dataPolicyId = config.dataPolicyId;
        this._dataPolicyType = config.dataPolicyType;
        this._id = config.id;
        this._location = config.location;
        this._policyTag = config.policyTag;
        this._project = config.project;
        this._dataMaskingPolicy.internalValue = config.dataMaskingPolicy;
        this._timeouts.internalValue = config.timeouts;
    }
    get dataPolicyId() {
        return this.getStringAttribute('data_policy_id');
    }
    set dataPolicyId(value) {
        this._dataPolicyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataPolicyIdInput() {
        return this._dataPolicyId;
    }
    get dataPolicyType() {
        return this.getStringAttribute('data_policy_type');
    }
    set dataPolicyType(value) {
        this._dataPolicyType = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataPolicyTypeInput() {
        return this._dataPolicyType;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get policyTag() {
        return this.getStringAttribute('policy_tag');
    }
    set policyTag(value) {
        this._policyTag = value;
    }
    // Temporarily expose input value. Use with caution.
    get policyTagInput() {
        return this._policyTag;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get dataMaskingPolicy() {
        return this._dataMaskingPolicy;
    }
    putDataMaskingPolicy(value) {
        this._dataMaskingPolicy.internalValue = value;
    }
    resetDataMaskingPolicy() {
        this._dataMaskingPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataMaskingPolicyInput() {
        return this._dataMaskingPolicy.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            data_policy_id: cdktf.stringToTerraform(this._dataPolicyId),
            data_policy_type: cdktf.stringToTerraform(this._dataPolicyType),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            policy_tag: cdktf.stringToTerraform(this._policyTag),
            project: cdktf.stringToTerraform(this._project),
            data_masking_policy: bigqueryDatapolicyDataPolicyDataMaskingPolicyToTerraform(this._dataMaskingPolicy.internalValue),
            timeouts: bigqueryDatapolicyDataPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            data_policy_id: {
                value: cdktf.stringToHclTerraform(this._dataPolicyId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_policy_type: {
                value: cdktf.stringToHclTerraform(this._dataPolicyType),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            policy_tag: {
                value: cdktf.stringToHclTerraform(this._policyTag),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            data_masking_policy: {
                value: bigqueryDatapolicyDataPolicyDataMaskingPolicyToHclTerraform(this._dataMaskingPolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryDatapolicyDataPolicyDataMaskingPolicyList",
            },
            timeouts: {
                value: bigqueryDatapolicyDataPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "BigqueryDatapolicyDataPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.BigqueryDatapolicyDataPolicy = BigqueryDatapolicyDataPolicy;
_c = JSII_RTTI_SYMBOL_1;
BigqueryDatapolicyDataPolicy[_c] = { fqn: "@cdktf/provider-google.bigqueryDatapolicyDataPolicy.BigqueryDatapolicyDataPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
BigqueryDatapolicyDataPolicy.tfResourceType = "google_bigquery_datapolicy_data_policy";
