// @cdktf/provider-google/lib/edgecontainer-cluster/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y;
Object.defineProperty(exports, "__esModule", { value: true });
exports.edgecontainerClusterNetworkingToHclTerraform = exports.edgecontainerClusterNetworkingToTerraform = exports.EdgecontainerClusterMaintenancePolicyOutputReference = exports.edgecontainerClusterMaintenancePolicyToHclTerraform = exports.edgecontainerClusterMaintenancePolicyToTerraform = exports.EdgecontainerClusterMaintenancePolicyWindowOutputReference = exports.edgecontainerClusterMaintenancePolicyWindowToHclTerraform = exports.edgecontainerClusterMaintenancePolicyWindowToTerraform = exports.EdgecontainerClusterMaintenancePolicyWindowRecurringWindowOutputReference = exports.edgecontainerClusterMaintenancePolicyWindowRecurringWindowToHclTerraform = exports.edgecontainerClusterMaintenancePolicyWindowRecurringWindowToTerraform = exports.EdgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowOutputReference = exports.edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToHclTerraform = exports.edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToTerraform = exports.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsList = exports.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsOutputReference = exports.edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToHclTerraform = exports.edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToTerraform = exports.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowOutputReference = exports.edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToHclTerraform = exports.edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToTerraform = exports.EdgecontainerClusterFleetOutputReference = exports.edgecontainerClusterFleetToHclTerraform = exports.edgecontainerClusterFleetToTerraform = exports.EdgecontainerClusterControlPlaneEncryptionOutputReference = exports.edgecontainerClusterControlPlaneEncryptionToHclTerraform = exports.edgecontainerClusterControlPlaneEncryptionToTerraform = exports.EdgecontainerClusterControlPlaneEncryptionKmsStatusList = exports.EdgecontainerClusterControlPlaneEncryptionKmsStatusOutputReference = exports.edgecontainerClusterControlPlaneEncryptionKmsStatusToHclTerraform = exports.edgecontainerClusterControlPlaneEncryptionKmsStatusToTerraform = exports.EdgecontainerClusterControlPlaneOutputReference = exports.edgecontainerClusterControlPlaneToHclTerraform = exports.edgecontainerClusterControlPlaneToTerraform = exports.EdgecontainerClusterControlPlaneRemoteOutputReference = exports.edgecontainerClusterControlPlaneRemoteToHclTerraform = exports.edgecontainerClusterControlPlaneRemoteToTerraform = exports.EdgecontainerClusterControlPlaneLocalOutputReference = exports.edgecontainerClusterControlPlaneLocalToHclTerraform = exports.edgecontainerClusterControlPlaneLocalToTerraform = exports.EdgecontainerClusterAuthorizationOutputReference = exports.edgecontainerClusterAuthorizationToHclTerraform = exports.edgecontainerClusterAuthorizationToTerraform = exports.EdgecontainerClusterAuthorizationAdminUsersOutputReference = exports.edgecontainerClusterAuthorizationAdminUsersToHclTerraform = exports.edgecontainerClusterAuthorizationAdminUsersToTerraform = exports.EdgecontainerClusterMaintenanceEventsList = exports.EdgecontainerClusterMaintenanceEventsOutputReference = exports.edgecontainerClusterMaintenanceEventsToHclTerraform = exports.edgecontainerClusterMaintenanceEventsToTerraform = void 0;
exports.EdgecontainerCluster = exports.EdgecontainerClusterTimeoutsOutputReference = exports.edgecontainerClusterTimeoutsToHclTerraform = exports.edgecontainerClusterTimeoutsToTerraform = exports.EdgecontainerClusterSystemAddonsConfigOutputReference = exports.edgecontainerClusterSystemAddonsConfigToHclTerraform = exports.edgecontainerClusterSystemAddonsConfigToTerraform = exports.EdgecontainerClusterSystemAddonsConfigIngressOutputReference = exports.edgecontainerClusterSystemAddonsConfigIngressToHclTerraform = exports.edgecontainerClusterSystemAddonsConfigIngressToTerraform = exports.EdgecontainerClusterNetworkingOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function edgecontainerClusterMaintenanceEventsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.edgecontainerClusterMaintenanceEventsToTerraform = edgecontainerClusterMaintenanceEventsToTerraform;
function edgecontainerClusterMaintenanceEventsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.edgecontainerClusterMaintenanceEventsToHclTerraform = edgecontainerClusterMaintenanceEventsToHclTerraform;
class EdgecontainerClusterMaintenanceEventsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // end_time - computed: true, optional: false, required: false
    get endTime() {
        return this.getStringAttribute('end_time');
    }
    // operation - computed: true, optional: false, required: false
    get operation() {
        return this.getStringAttribute('operation');
    }
    // schedule - computed: true, optional: false, required: false
    get schedule() {
        return this.getStringAttribute('schedule');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // target_version - computed: true, optional: false, required: false
    get targetVersion() {
        return this.getStringAttribute('target_version');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    // uuid - computed: true, optional: false, required: false
    get uuid() {
        return this.getStringAttribute('uuid');
    }
}
exports.EdgecontainerClusterMaintenanceEventsOutputReference = EdgecontainerClusterMaintenanceEventsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenanceEventsOutputReference[_a] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenanceEventsOutputReference", version: "14.13.1" };
class EdgecontainerClusterMaintenanceEventsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EdgecontainerClusterMaintenanceEventsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EdgecontainerClusterMaintenanceEventsList = EdgecontainerClusterMaintenanceEventsList;
_b = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenanceEventsList[_b] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenanceEventsList", version: "14.13.1" };
function edgecontainerClusterAuthorizationAdminUsersToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.edgecontainerClusterAuthorizationAdminUsersToTerraform = edgecontainerClusterAuthorizationAdminUsersToTerraform;
function edgecontainerClusterAuthorizationAdminUsersToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        username: {
            value: cdktf.stringToHclTerraform(struct.username),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterAuthorizationAdminUsersToHclTerraform = edgecontainerClusterAuthorizationAdminUsersToHclTerraform;
class EdgecontainerClusterAuthorizationAdminUsersOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._username !== undefined) {
            hasAnyValues = true;
            internalValueResult.username = this._username;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._username = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._username = value.username;
        }
    }
    get username() {
        return this.getStringAttribute('username');
    }
    set username(value) {
        this._username = value;
    }
    // Temporarily expose input value. Use with caution.
    get usernameInput() {
        return this._username;
    }
}
exports.EdgecontainerClusterAuthorizationAdminUsersOutputReference = EdgecontainerClusterAuthorizationAdminUsersOutputReference;
_c = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterAuthorizationAdminUsersOutputReference[_c] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterAuthorizationAdminUsersOutputReference", version: "14.13.1" };
function edgecontainerClusterAuthorizationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        admin_users: edgecontainerClusterAuthorizationAdminUsersToTerraform(struct.adminUsers),
    };
}
exports.edgecontainerClusterAuthorizationToTerraform = edgecontainerClusterAuthorizationToTerraform;
function edgecontainerClusterAuthorizationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        admin_users: {
            value: edgecontainerClusterAuthorizationAdminUsersToHclTerraform(struct.adminUsers),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterAuthorizationAdminUsersList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterAuthorizationToHclTerraform = edgecontainerClusterAuthorizationToHclTerraform;
class EdgecontainerClusterAuthorizationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // admin_users - computed: false, optional: false, required: true
        this._adminUsers = new EdgecontainerClusterAuthorizationAdminUsersOutputReference(this, "admin_users");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._adminUsers?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.adminUsers = this._adminUsers?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._adminUsers.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._adminUsers.internalValue = value.adminUsers;
        }
    }
    get adminUsers() {
        return this._adminUsers;
    }
    putAdminUsers(value) {
        this._adminUsers.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get adminUsersInput() {
        return this._adminUsers.internalValue;
    }
}
exports.EdgecontainerClusterAuthorizationOutputReference = EdgecontainerClusterAuthorizationOutputReference;
_d = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterAuthorizationOutputReference[_d] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterAuthorizationOutputReference", version: "14.13.1" };
function edgecontainerClusterControlPlaneLocalToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        machine_filter: cdktf.stringToTerraform(struct.machineFilter),
        node_count: cdktf.numberToTerraform(struct.nodeCount),
        node_location: cdktf.stringToTerraform(struct.nodeLocation),
        shared_deployment_policy: cdktf.stringToTerraform(struct.sharedDeploymentPolicy),
    };
}
exports.edgecontainerClusterControlPlaneLocalToTerraform = edgecontainerClusterControlPlaneLocalToTerraform;
function edgecontainerClusterControlPlaneLocalToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        machine_filter: {
            value: cdktf.stringToHclTerraform(struct.machineFilter),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        node_count: {
            value: cdktf.numberToHclTerraform(struct.nodeCount),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        node_location: {
            value: cdktf.stringToHclTerraform(struct.nodeLocation),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        shared_deployment_policy: {
            value: cdktf.stringToHclTerraform(struct.sharedDeploymentPolicy),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterControlPlaneLocalToHclTerraform = edgecontainerClusterControlPlaneLocalToHclTerraform;
class EdgecontainerClusterControlPlaneLocalOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._machineFilter !== undefined) {
            hasAnyValues = true;
            internalValueResult.machineFilter = this._machineFilter;
        }
        if (this._nodeCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.nodeCount = this._nodeCount;
        }
        if (this._nodeLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.nodeLocation = this._nodeLocation;
        }
        if (this._sharedDeploymentPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.sharedDeploymentPolicy = this._sharedDeploymentPolicy;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._machineFilter = undefined;
            this._nodeCount = undefined;
            this._nodeLocation = undefined;
            this._sharedDeploymentPolicy = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._machineFilter = value.machineFilter;
            this._nodeCount = value.nodeCount;
            this._nodeLocation = value.nodeLocation;
            this._sharedDeploymentPolicy = value.sharedDeploymentPolicy;
        }
    }
    get machineFilter() {
        return this.getStringAttribute('machine_filter');
    }
    set machineFilter(value) {
        this._machineFilter = value;
    }
    resetMachineFilter() {
        this._machineFilter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get machineFilterInput() {
        return this._machineFilter;
    }
    get nodeCount() {
        return this.getNumberAttribute('node_count');
    }
    set nodeCount(value) {
        this._nodeCount = value;
    }
    resetNodeCount() {
        this._nodeCount = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeCountInput() {
        return this._nodeCount;
    }
    get nodeLocation() {
        return this.getStringAttribute('node_location');
    }
    set nodeLocation(value) {
        this._nodeLocation = value;
    }
    resetNodeLocation() {
        this._nodeLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeLocationInput() {
        return this._nodeLocation;
    }
    get sharedDeploymentPolicy() {
        return this.getStringAttribute('shared_deployment_policy');
    }
    set sharedDeploymentPolicy(value) {
        this._sharedDeploymentPolicy = value;
    }
    resetSharedDeploymentPolicy() {
        this._sharedDeploymentPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sharedDeploymentPolicyInput() {
        return this._sharedDeploymentPolicy;
    }
}
exports.EdgecontainerClusterControlPlaneLocalOutputReference = EdgecontainerClusterControlPlaneLocalOutputReference;
_e = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterControlPlaneLocalOutputReference[_e] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterControlPlaneLocalOutputReference", version: "14.13.1" };
function edgecontainerClusterControlPlaneRemoteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        node_location: cdktf.stringToTerraform(struct.nodeLocation),
    };
}
exports.edgecontainerClusterControlPlaneRemoteToTerraform = edgecontainerClusterControlPlaneRemoteToTerraform;
function edgecontainerClusterControlPlaneRemoteToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        node_location: {
            value: cdktf.stringToHclTerraform(struct.nodeLocation),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterControlPlaneRemoteToHclTerraform = edgecontainerClusterControlPlaneRemoteToHclTerraform;
class EdgecontainerClusterControlPlaneRemoteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._nodeLocation !== undefined) {
            hasAnyValues = true;
            internalValueResult.nodeLocation = this._nodeLocation;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._nodeLocation = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._nodeLocation = value.nodeLocation;
        }
    }
    get nodeLocation() {
        return this.getStringAttribute('node_location');
    }
    set nodeLocation(value) {
        this._nodeLocation = value;
    }
    resetNodeLocation() {
        this._nodeLocation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeLocationInput() {
        return this._nodeLocation;
    }
}
exports.EdgecontainerClusterControlPlaneRemoteOutputReference = EdgecontainerClusterControlPlaneRemoteOutputReference;
_f = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterControlPlaneRemoteOutputReference[_f] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterControlPlaneRemoteOutputReference", version: "14.13.1" };
function edgecontainerClusterControlPlaneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        local: edgecontainerClusterControlPlaneLocalToTerraform(struct.local),
        remote: edgecontainerClusterControlPlaneRemoteToTerraform(struct.remote),
    };
}
exports.edgecontainerClusterControlPlaneToTerraform = edgecontainerClusterControlPlaneToTerraform;
function edgecontainerClusterControlPlaneToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        local: {
            value: edgecontainerClusterControlPlaneLocalToHclTerraform(struct.local),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterControlPlaneLocalList",
        },
        remote: {
            value: edgecontainerClusterControlPlaneRemoteToHclTerraform(struct.remote),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterControlPlaneRemoteList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterControlPlaneToHclTerraform = edgecontainerClusterControlPlaneToHclTerraform;
class EdgecontainerClusterControlPlaneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // local - computed: false, optional: true, required: false
        this._local = new EdgecontainerClusterControlPlaneLocalOutputReference(this, "local");
        // remote - computed: false, optional: true, required: false
        this._remote = new EdgecontainerClusterControlPlaneRemoteOutputReference(this, "remote");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._local?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.local = this._local?.internalValue;
        }
        if (this._remote?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.remote = this._remote?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._local.internalValue = undefined;
            this._remote.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._local.internalValue = value.local;
            this._remote.internalValue = value.remote;
        }
    }
    get local() {
        return this._local;
    }
    putLocal(value) {
        this._local.internalValue = value;
    }
    resetLocal() {
        this._local.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get localInput() {
        return this._local.internalValue;
    }
    get remote() {
        return this._remote;
    }
    putRemote(value) {
        this._remote.internalValue = value;
    }
    resetRemote() {
        this._remote.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get remoteInput() {
        return this._remote.internalValue;
    }
}
exports.EdgecontainerClusterControlPlaneOutputReference = EdgecontainerClusterControlPlaneOutputReference;
_g = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterControlPlaneOutputReference[_g] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterControlPlaneOutputReference", version: "14.13.1" };
function edgecontainerClusterControlPlaneEncryptionKmsStatusToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.edgecontainerClusterControlPlaneEncryptionKmsStatusToTerraform = edgecontainerClusterControlPlaneEncryptionKmsStatusToTerraform;
function edgecontainerClusterControlPlaneEncryptionKmsStatusToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.edgecontainerClusterControlPlaneEncryptionKmsStatusToHclTerraform = edgecontainerClusterControlPlaneEncryptionKmsStatusToHclTerraform;
class EdgecontainerClusterControlPlaneEncryptionKmsStatusOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // code - computed: true, optional: false, required: false
    get code() {
        return this.getNumberAttribute('code');
    }
    // message - computed: true, optional: false, required: false
    get message() {
        return this.getStringAttribute('message');
    }
}
exports.EdgecontainerClusterControlPlaneEncryptionKmsStatusOutputReference = EdgecontainerClusterControlPlaneEncryptionKmsStatusOutputReference;
_h = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterControlPlaneEncryptionKmsStatusOutputReference[_h] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterControlPlaneEncryptionKmsStatusOutputReference", version: "14.13.1" };
class EdgecontainerClusterControlPlaneEncryptionKmsStatusList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EdgecontainerClusterControlPlaneEncryptionKmsStatusOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EdgecontainerClusterControlPlaneEncryptionKmsStatusList = EdgecontainerClusterControlPlaneEncryptionKmsStatusList;
_j = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterControlPlaneEncryptionKmsStatusList[_j] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterControlPlaneEncryptionKmsStatusList", version: "14.13.1" };
function edgecontainerClusterControlPlaneEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        kms_key: cdktf.stringToTerraform(struct.kmsKey),
    };
}
exports.edgecontainerClusterControlPlaneEncryptionToTerraform = edgecontainerClusterControlPlaneEncryptionToTerraform;
function edgecontainerClusterControlPlaneEncryptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        kms_key: {
            value: cdktf.stringToHclTerraform(struct.kmsKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterControlPlaneEncryptionToHclTerraform = edgecontainerClusterControlPlaneEncryptionToHclTerraform;
class EdgecontainerClusterControlPlaneEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // kms_status - computed: true, optional: false, required: false
        this._kmsStatus = new EdgecontainerClusterControlPlaneEncryptionKmsStatusList(this, "kms_status", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._kmsKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmsKey = this._kmsKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._kmsKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._kmsKey = value.kmsKey;
        }
    }
    get kmsKey() {
        return this.getStringAttribute('kms_key');
    }
    set kmsKey(value) {
        this._kmsKey = value;
    }
    resetKmsKey() {
        this._kmsKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyInput() {
        return this._kmsKey;
    }
    // kms_key_active_version - computed: true, optional: false, required: false
    get kmsKeyActiveVersion() {
        return this.getStringAttribute('kms_key_active_version');
    }
    // kms_key_state - computed: true, optional: false, required: false
    get kmsKeyState() {
        return this.getStringAttribute('kms_key_state');
    }
    get kmsStatus() {
        return this._kmsStatus;
    }
}
exports.EdgecontainerClusterControlPlaneEncryptionOutputReference = EdgecontainerClusterControlPlaneEncryptionOutputReference;
_k = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterControlPlaneEncryptionOutputReference[_k] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterControlPlaneEncryptionOutputReference", version: "14.13.1" };
function edgecontainerClusterFleetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        project: cdktf.stringToTerraform(struct.project),
    };
}
exports.edgecontainerClusterFleetToTerraform = edgecontainerClusterFleetToTerraform;
function edgecontainerClusterFleetToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        project: {
            value: cdktf.stringToHclTerraform(struct.project),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterFleetToHclTerraform = edgecontainerClusterFleetToHclTerraform;
class EdgecontainerClusterFleetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._project !== undefined) {
            hasAnyValues = true;
            internalValueResult.project = this._project;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._project = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._project = value.project;
        }
    }
    // membership - computed: true, optional: false, required: false
    get membership() {
        return this.getStringAttribute('membership');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
}
exports.EdgecontainerClusterFleetOutputReference = EdgecontainerClusterFleetOutputReference;
_l = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterFleetOutputReference[_l] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterFleetOutputReference", version: "14.13.1" };
function edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        end_time: cdktf.stringToTerraform(struct.endTime),
        start_time: cdktf.stringToTerraform(struct.startTime),
    };
}
exports.edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToTerraform = edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToTerraform;
function edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        end_time: {
            value: cdktf.stringToHclTerraform(struct.endTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start_time: {
            value: cdktf.stringToHclTerraform(struct.startTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToHclTerraform = edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToHclTerraform;
class EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.endTime = this._endTime;
        }
        if (this._startTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endTime = undefined;
            this._startTime = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endTime = value.endTime;
            this._startTime = value.startTime;
        }
    }
    get endTime() {
        return this.getStringAttribute('end_time');
    }
    set endTime(value) {
        this._endTime = value;
    }
    resetEndTime() {
        this._endTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endTimeInput() {
        return this._endTime;
    }
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    set startTime(value) {
        this._startTime = value;
    }
    resetStartTime() {
        this._startTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime;
    }
}
exports.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowOutputReference = EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowOutputReference;
_m = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowOutputReference[_m] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowOutputReference", version: "14.13.1" };
function edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        id: cdktf.stringToTerraform(struct.id),
        window: edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToTerraform(struct.window),
    };
}
exports.edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToTerraform = edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToTerraform;
function edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        id: {
            value: cdktf.stringToHclTerraform(struct.id),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        window: {
            value: edgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowToHclTerraform(struct.window),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToHclTerraform = edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToHclTerraform;
class EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // window - computed: false, optional: true, required: false
        this._window = new EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsWindowOutputReference(this, "window");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._id !== undefined) {
            hasAnyValues = true;
            internalValueResult.id = this._id;
        }
        if (this._window?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.window = this._window?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._id = undefined;
            this._window.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._id = value.id;
            this._window.internalValue = value.window;
        }
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get window() {
        return this._window;
    }
    putWindow(value) {
        this._window.internalValue = value;
    }
    resetWindow() {
        this._window.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get windowInput() {
        return this._window.internalValue;
    }
}
exports.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsOutputReference = EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsOutputReference[_o] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsOutputReference", version: "14.13.1" };
class EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsList = EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsList;
_p = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsList[_p] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsList", version: "14.13.1" };
function edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        end_time: cdktf.stringToTerraform(struct.endTime),
        start_time: cdktf.stringToTerraform(struct.startTime),
    };
}
exports.edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToTerraform = edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToTerraform;
function edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        end_time: {
            value: cdktf.stringToHclTerraform(struct.endTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        start_time: {
            value: cdktf.stringToHclTerraform(struct.startTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToHclTerraform = edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToHclTerraform;
class EdgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._endTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.endTime = this._endTime;
        }
        if (this._startTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._endTime = undefined;
            this._startTime = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._endTime = value.endTime;
            this._startTime = value.startTime;
        }
    }
    get endTime() {
        return this.getStringAttribute('end_time');
    }
    set endTime(value) {
        this._endTime = value;
    }
    resetEndTime() {
        this._endTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get endTimeInput() {
        return this._endTime;
    }
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    set startTime(value) {
        this._startTime = value;
    }
    resetStartTime() {
        this._startTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime;
    }
}
exports.EdgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowOutputReference = EdgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowOutputReference;
_q = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowOutputReference[_q] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowOutputReference", version: "14.13.1" };
function edgecontainerClusterMaintenancePolicyWindowRecurringWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        recurrence: cdktf.stringToTerraform(struct.recurrence),
        window: edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToTerraform(struct.window),
    };
}
exports.edgecontainerClusterMaintenancePolicyWindowRecurringWindowToTerraform = edgecontainerClusterMaintenancePolicyWindowRecurringWindowToTerraform;
function edgecontainerClusterMaintenancePolicyWindowRecurringWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        recurrence: {
            value: cdktf.stringToHclTerraform(struct.recurrence),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        window: {
            value: edgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowToHclTerraform(struct.window),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterMaintenancePolicyWindowRecurringWindowToHclTerraform = edgecontainerClusterMaintenancePolicyWindowRecurringWindowToHclTerraform;
class EdgecontainerClusterMaintenancePolicyWindowRecurringWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // window - computed: false, optional: true, required: false
        this._window = new EdgecontainerClusterMaintenancePolicyWindowRecurringWindowWindowOutputReference(this, "window");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recurrence !== undefined) {
            hasAnyValues = true;
            internalValueResult.recurrence = this._recurrence;
        }
        if (this._window?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.window = this._window?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recurrence = undefined;
            this._window.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recurrence = value.recurrence;
            this._window.internalValue = value.window;
        }
    }
    get recurrence() {
        return this.getStringAttribute('recurrence');
    }
    set recurrence(value) {
        this._recurrence = value;
    }
    resetRecurrence() {
        this._recurrence = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recurrenceInput() {
        return this._recurrence;
    }
    get window() {
        return this._window;
    }
    putWindow(value) {
        this._window.internalValue = value;
    }
    resetWindow() {
        this._window.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get windowInput() {
        return this._window.internalValue;
    }
}
exports.EdgecontainerClusterMaintenancePolicyWindowRecurringWindowOutputReference = EdgecontainerClusterMaintenancePolicyWindowRecurringWindowOutputReference;
_r = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenancePolicyWindowRecurringWindowOutputReference[_r] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenancePolicyWindowRecurringWindowOutputReference", version: "14.13.1" };
function edgecontainerClusterMaintenancePolicyWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        recurring_window: edgecontainerClusterMaintenancePolicyWindowRecurringWindowToTerraform(struct.recurringWindow),
    };
}
exports.edgecontainerClusterMaintenancePolicyWindowToTerraform = edgecontainerClusterMaintenancePolicyWindowToTerraform;
function edgecontainerClusterMaintenancePolicyWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        recurring_window: {
            value: edgecontainerClusterMaintenancePolicyWindowRecurringWindowToHclTerraform(struct.recurringWindow),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterMaintenancePolicyWindowRecurringWindowList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterMaintenancePolicyWindowToHclTerraform = edgecontainerClusterMaintenancePolicyWindowToHclTerraform;
class EdgecontainerClusterMaintenancePolicyWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // recurring_window - computed: false, optional: false, required: true
        this._recurringWindow = new EdgecontainerClusterMaintenancePolicyWindowRecurringWindowOutputReference(this, "recurring_window");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recurringWindow?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.recurringWindow = this._recurringWindow?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recurringWindow.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recurringWindow.internalValue = value.recurringWindow;
        }
    }
    get recurringWindow() {
        return this._recurringWindow;
    }
    putRecurringWindow(value) {
        this._recurringWindow.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get recurringWindowInput() {
        return this._recurringWindow.internalValue;
    }
}
exports.EdgecontainerClusterMaintenancePolicyWindowOutputReference = EdgecontainerClusterMaintenancePolicyWindowOutputReference;
_s = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenancePolicyWindowOutputReference[_s] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenancePolicyWindowOutputReference", version: "14.13.1" };
function edgecontainerClusterMaintenancePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        maintenance_exclusions: cdktf.listMapper(edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToTerraform, true)(struct.maintenanceExclusions),
        window: edgecontainerClusterMaintenancePolicyWindowToTerraform(struct.window),
    };
}
exports.edgecontainerClusterMaintenancePolicyToTerraform = edgecontainerClusterMaintenancePolicyToTerraform;
function edgecontainerClusterMaintenancePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        maintenance_exclusions: {
            value: cdktf.listMapperHcl(edgecontainerClusterMaintenancePolicyMaintenanceExclusionsToHclTerraform, true)(struct.maintenanceExclusions),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsList",
        },
        window: {
            value: edgecontainerClusterMaintenancePolicyWindowToHclTerraform(struct.window),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterMaintenancePolicyWindowList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterMaintenancePolicyToHclTerraform = edgecontainerClusterMaintenancePolicyToHclTerraform;
class EdgecontainerClusterMaintenancePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // maintenance_exclusions - computed: false, optional: true, required: false
        this._maintenanceExclusions = new EdgecontainerClusterMaintenancePolicyMaintenanceExclusionsList(this, "maintenance_exclusions", false);
        // window - computed: false, optional: false, required: true
        this._window = new EdgecontainerClusterMaintenancePolicyWindowOutputReference(this, "window");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maintenanceExclusions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maintenanceExclusions = this._maintenanceExclusions?.internalValue;
        }
        if (this._window?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.window = this._window?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maintenanceExclusions.internalValue = undefined;
            this._window.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maintenanceExclusions.internalValue = value.maintenanceExclusions;
            this._window.internalValue = value.window;
        }
    }
    get maintenanceExclusions() {
        return this._maintenanceExclusions;
    }
    putMaintenanceExclusions(value) {
        this._maintenanceExclusions.internalValue = value;
    }
    resetMaintenanceExclusions() {
        this._maintenanceExclusions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceExclusionsInput() {
        return this._maintenanceExclusions.internalValue;
    }
    get window() {
        return this._window;
    }
    putWindow(value) {
        this._window.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get windowInput() {
        return this._window.internalValue;
    }
}
exports.EdgecontainerClusterMaintenancePolicyOutputReference = EdgecontainerClusterMaintenancePolicyOutputReference;
_t = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterMaintenancePolicyOutputReference[_t] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterMaintenancePolicyOutputReference", version: "14.13.1" };
function edgecontainerClusterNetworkingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cluster_ipv4_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.clusterIpv4CidrBlocks),
        cluster_ipv6_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.clusterIpv6CidrBlocks),
        services_ipv4_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.servicesIpv4CidrBlocks),
        services_ipv6_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.servicesIpv6CidrBlocks),
    };
}
exports.edgecontainerClusterNetworkingToTerraform = edgecontainerClusterNetworkingToTerraform;
function edgecontainerClusterNetworkingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        cluster_ipv4_cidr_blocks: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.clusterIpv4CidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        cluster_ipv6_cidr_blocks: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.clusterIpv6CidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        services_ipv4_cidr_blocks: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.servicesIpv4CidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        services_ipv6_cidr_blocks: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.servicesIpv6CidrBlocks),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterNetworkingToHclTerraform = edgecontainerClusterNetworkingToHclTerraform;
class EdgecontainerClusterNetworkingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clusterIpv4CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.clusterIpv4CidrBlocks = this._clusterIpv4CidrBlocks;
        }
        if (this._clusterIpv6CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.clusterIpv6CidrBlocks = this._clusterIpv6CidrBlocks;
        }
        if (this._servicesIpv4CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.servicesIpv4CidrBlocks = this._servicesIpv4CidrBlocks;
        }
        if (this._servicesIpv6CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.servicesIpv6CidrBlocks = this._servicesIpv6CidrBlocks;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clusterIpv4CidrBlocks = undefined;
            this._clusterIpv6CidrBlocks = undefined;
            this._servicesIpv4CidrBlocks = undefined;
            this._servicesIpv6CidrBlocks = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clusterIpv4CidrBlocks = value.clusterIpv4CidrBlocks;
            this._clusterIpv6CidrBlocks = value.clusterIpv6CidrBlocks;
            this._servicesIpv4CidrBlocks = value.servicesIpv4CidrBlocks;
            this._servicesIpv6CidrBlocks = value.servicesIpv6CidrBlocks;
        }
    }
    get clusterIpv4CidrBlocks() {
        return this.getListAttribute('cluster_ipv4_cidr_blocks');
    }
    set clusterIpv4CidrBlocks(value) {
        this._clusterIpv4CidrBlocks = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIpv4CidrBlocksInput() {
        return this._clusterIpv4CidrBlocks;
    }
    get clusterIpv6CidrBlocks() {
        return this.getListAttribute('cluster_ipv6_cidr_blocks');
    }
    set clusterIpv6CidrBlocks(value) {
        this._clusterIpv6CidrBlocks = value;
    }
    resetClusterIpv6CidrBlocks() {
        this._clusterIpv6CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIpv6CidrBlocksInput() {
        return this._clusterIpv6CidrBlocks;
    }
    // network_type - computed: true, optional: false, required: false
    get networkType() {
        return this.getStringAttribute('network_type');
    }
    get servicesIpv4CidrBlocks() {
        return this.getListAttribute('services_ipv4_cidr_blocks');
    }
    set servicesIpv4CidrBlocks(value) {
        this._servicesIpv4CidrBlocks = value;
    }
    // Temporarily expose input value. Use with caution.
    get servicesIpv4CidrBlocksInput() {
        return this._servicesIpv4CidrBlocks;
    }
    get servicesIpv6CidrBlocks() {
        return this.getListAttribute('services_ipv6_cidr_blocks');
    }
    set servicesIpv6CidrBlocks(value) {
        this._servicesIpv6CidrBlocks = value;
    }
    resetServicesIpv6CidrBlocks() {
        this._servicesIpv6CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get servicesIpv6CidrBlocksInput() {
        return this._servicesIpv6CidrBlocks;
    }
}
exports.EdgecontainerClusterNetworkingOutputReference = EdgecontainerClusterNetworkingOutputReference;
_u = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterNetworkingOutputReference[_u] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterNetworkingOutputReference", version: "14.13.1" };
function edgecontainerClusterSystemAddonsConfigIngressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disabled: cdktf.booleanToTerraform(struct.disabled),
        ipv4_vip: cdktf.stringToTerraform(struct.ipv4Vip),
    };
}
exports.edgecontainerClusterSystemAddonsConfigIngressToTerraform = edgecontainerClusterSystemAddonsConfigIngressToTerraform;
function edgecontainerClusterSystemAddonsConfigIngressToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        disabled: {
            value: cdktf.booleanToHclTerraform(struct.disabled),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        ipv4_vip: {
            value: cdktf.stringToHclTerraform(struct.ipv4Vip),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterSystemAddonsConfigIngressToHclTerraform = edgecontainerClusterSystemAddonsConfigIngressToHclTerraform;
class EdgecontainerClusterSystemAddonsConfigIngressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._disabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.disabled = this._disabled;
        }
        if (this._ipv4Vip !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv4Vip = this._ipv4Vip;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._disabled = undefined;
            this._ipv4Vip = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._disabled = value.disabled;
            this._ipv4Vip = value.ipv4Vip;
        }
    }
    get disabled() {
        return this.getBooleanAttribute('disabled');
    }
    set disabled(value) {
        this._disabled = value;
    }
    resetDisabled() {
        this._disabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disabledInput() {
        return this._disabled;
    }
    get ipv4Vip() {
        return this.getStringAttribute('ipv4_vip');
    }
    set ipv4Vip(value) {
        this._ipv4Vip = value;
    }
    resetIpv4Vip() {
        this._ipv4Vip = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv4VipInput() {
        return this._ipv4Vip;
    }
}
exports.EdgecontainerClusterSystemAddonsConfigIngressOutputReference = EdgecontainerClusterSystemAddonsConfigIngressOutputReference;
_v = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterSystemAddonsConfigIngressOutputReference[_v] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterSystemAddonsConfigIngressOutputReference", version: "14.13.1" };
function edgecontainerClusterSystemAddonsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ingress: edgecontainerClusterSystemAddonsConfigIngressToTerraform(struct.ingress),
    };
}
exports.edgecontainerClusterSystemAddonsConfigToTerraform = edgecontainerClusterSystemAddonsConfigToTerraform;
function edgecontainerClusterSystemAddonsConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ingress: {
            value: edgecontainerClusterSystemAddonsConfigIngressToHclTerraform(struct.ingress),
            isBlock: true,
            type: "list",
            storageClassType: "EdgecontainerClusterSystemAddonsConfigIngressList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterSystemAddonsConfigToHclTerraform = edgecontainerClusterSystemAddonsConfigToHclTerraform;
class EdgecontainerClusterSystemAddonsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // ingress - computed: false, optional: true, required: false
        this._ingress = new EdgecontainerClusterSystemAddonsConfigIngressOutputReference(this, "ingress");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ingress?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ingress = this._ingress?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ingress.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ingress.internalValue = value.ingress;
        }
    }
    get ingress() {
        return this._ingress;
    }
    putIngress(value) {
        this._ingress.internalValue = value;
    }
    resetIngress() {
        this._ingress.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressInput() {
        return this._ingress.internalValue;
    }
}
exports.EdgecontainerClusterSystemAddonsConfigOutputReference = EdgecontainerClusterSystemAddonsConfigOutputReference;
_w = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterSystemAddonsConfigOutputReference[_w] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterSystemAddonsConfigOutputReference", version: "14.13.1" };
function edgecontainerClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.edgecontainerClusterTimeoutsToTerraform = edgecontainerClusterTimeoutsToTerraform;
function edgecontainerClusterTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.edgecontainerClusterTimeoutsToHclTerraform = edgecontainerClusterTimeoutsToHclTerraform;
class EdgecontainerClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.EdgecontainerClusterTimeoutsOutputReference = EdgecontainerClusterTimeoutsOutputReference;
_x = JSII_RTTI_SYMBOL_1;
EdgecontainerClusterTimeoutsOutputReference[_x] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerClusterTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/edgecontainer_cluster google_edgecontainer_cluster}
*/
class EdgecontainerCluster extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a EdgecontainerCluster resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the EdgecontainerCluster to import
    * @param importFromId The id of the existing EdgecontainerCluster that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/edgecontainer_cluster#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the EdgecontainerCluster to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_edgecontainer_cluster", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/edgecontainer_cluster google_edgecontainer_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options EdgecontainerClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_edgecontainer_cluster',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // maintenance_events - computed: true, optional: false, required: false
        this._maintenanceEvents = new EdgecontainerClusterMaintenanceEventsList(this, "maintenance_events", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // authorization - computed: false, optional: false, required: true
        this._authorization = new EdgecontainerClusterAuthorizationOutputReference(this, "authorization");
        // control_plane - computed: false, optional: true, required: false
        this._controlPlane = new EdgecontainerClusterControlPlaneOutputReference(this, "control_plane");
        // control_plane_encryption - computed: false, optional: true, required: false
        this._controlPlaneEncryption = new EdgecontainerClusterControlPlaneEncryptionOutputReference(this, "control_plane_encryption");
        // fleet - computed: false, optional: false, required: true
        this._fleet = new EdgecontainerClusterFleetOutputReference(this, "fleet");
        // maintenance_policy - computed: false, optional: true, required: false
        this._maintenancePolicy = new EdgecontainerClusterMaintenancePolicyOutputReference(this, "maintenance_policy");
        // networking - computed: false, optional: false, required: true
        this._networking = new EdgecontainerClusterNetworkingOutputReference(this, "networking");
        // system_addons_config - computed: false, optional: true, required: false
        this._systemAddonsConfig = new EdgecontainerClusterSystemAddonsConfigOutputReference(this, "system_addons_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new EdgecontainerClusterTimeoutsOutputReference(this, "timeouts");
        this._defaultMaxPodsPerNode = config.defaultMaxPodsPerNode;
        this._externalLoadBalancerIpv4AddressPools = config.externalLoadBalancerIpv4AddressPools;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._project = config.project;
        this._releaseChannel = config.releaseChannel;
        this._targetVersion = config.targetVersion;
        this._authorization.internalValue = config.authorization;
        this._controlPlane.internalValue = config.controlPlane;
        this._controlPlaneEncryption.internalValue = config.controlPlaneEncryption;
        this._fleet.internalValue = config.fleet;
        this._maintenancePolicy.internalValue = config.maintenancePolicy;
        this._networking.internalValue = config.networking;
        this._systemAddonsConfig.internalValue = config.systemAddonsConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // cluster_ca_certificate - computed: true, optional: false, required: false
    get clusterCaCertificate() {
        return this.getStringAttribute('cluster_ca_certificate');
    }
    // control_plane_version - computed: true, optional: false, required: false
    get controlPlaneVersion() {
        return this.getStringAttribute('control_plane_version');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get defaultMaxPodsPerNode() {
        return this.getNumberAttribute('default_max_pods_per_node');
    }
    set defaultMaxPodsPerNode(value) {
        this._defaultMaxPodsPerNode = value;
    }
    resetDefaultMaxPodsPerNode() {
        this._defaultMaxPodsPerNode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultMaxPodsPerNodeInput() {
        return this._defaultMaxPodsPerNode;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // endpoint - computed: true, optional: false, required: false
    get endpoint() {
        return this.getStringAttribute('endpoint');
    }
    get externalLoadBalancerIpv4AddressPools() {
        return this.getListAttribute('external_load_balancer_ipv4_address_pools');
    }
    set externalLoadBalancerIpv4AddressPools(value) {
        this._externalLoadBalancerIpv4AddressPools = value;
    }
    resetExternalLoadBalancerIpv4AddressPools() {
        this._externalLoadBalancerIpv4AddressPools = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get externalLoadBalancerIpv4AddressPoolsInput() {
        return this._externalLoadBalancerIpv4AddressPools;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get maintenanceEvents() {
        return this._maintenanceEvents;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // node_version - computed: true, optional: false, required: false
    get nodeVersion() {
        return this.getStringAttribute('node_version');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get releaseChannel() {
        return this.getStringAttribute('release_channel');
    }
    set releaseChannel(value) {
        this._releaseChannel = value;
    }
    resetReleaseChannel() {
        this._releaseChannel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get releaseChannelInput() {
        return this._releaseChannel;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get targetVersion() {
        return this.getStringAttribute('target_version');
    }
    set targetVersion(value) {
        this._targetVersion = value;
    }
    resetTargetVersion() {
        this._targetVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetVersionInput() {
        return this._targetVersion;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get authorization() {
        return this._authorization;
    }
    putAuthorization(value) {
        this._authorization.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get authorizationInput() {
        return this._authorization.internalValue;
    }
    get controlPlane() {
        return this._controlPlane;
    }
    putControlPlane(value) {
        this._controlPlane.internalValue = value;
    }
    resetControlPlane() {
        this._controlPlane.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get controlPlaneInput() {
        return this._controlPlane.internalValue;
    }
    get controlPlaneEncryption() {
        return this._controlPlaneEncryption;
    }
    putControlPlaneEncryption(value) {
        this._controlPlaneEncryption.internalValue = value;
    }
    resetControlPlaneEncryption() {
        this._controlPlaneEncryption.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get controlPlaneEncryptionInput() {
        return this._controlPlaneEncryption.internalValue;
    }
    get fleet() {
        return this._fleet;
    }
    putFleet(value) {
        this._fleet.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get fleetInput() {
        return this._fleet.internalValue;
    }
    get maintenancePolicy() {
        return this._maintenancePolicy;
    }
    putMaintenancePolicy(value) {
        this._maintenancePolicy.internalValue = value;
    }
    resetMaintenancePolicy() {
        this._maintenancePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenancePolicyInput() {
        return this._maintenancePolicy.internalValue;
    }
    get networking() {
        return this._networking;
    }
    putNetworking(value) {
        this._networking.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkingInput() {
        return this._networking.internalValue;
    }
    get systemAddonsConfig() {
        return this._systemAddonsConfig;
    }
    putSystemAddonsConfig(value) {
        this._systemAddonsConfig.internalValue = value;
    }
    resetSystemAddonsConfig() {
        this._systemAddonsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get systemAddonsConfigInput() {
        return this._systemAddonsConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            default_max_pods_per_node: cdktf.numberToTerraform(this._defaultMaxPodsPerNode),
            external_load_balancer_ipv4_address_pools: cdktf.listMapper(cdktf.stringToTerraform, false)(this._externalLoadBalancerIpv4AddressPools),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            release_channel: cdktf.stringToTerraform(this._releaseChannel),
            target_version: cdktf.stringToTerraform(this._targetVersion),
            authorization: edgecontainerClusterAuthorizationToTerraform(this._authorization.internalValue),
            control_plane: edgecontainerClusterControlPlaneToTerraform(this._controlPlane.internalValue),
            control_plane_encryption: edgecontainerClusterControlPlaneEncryptionToTerraform(this._controlPlaneEncryption.internalValue),
            fleet: edgecontainerClusterFleetToTerraform(this._fleet.internalValue),
            maintenance_policy: edgecontainerClusterMaintenancePolicyToTerraform(this._maintenancePolicy.internalValue),
            networking: edgecontainerClusterNetworkingToTerraform(this._networking.internalValue),
            system_addons_config: edgecontainerClusterSystemAddonsConfigToTerraform(this._systemAddonsConfig.internalValue),
            timeouts: edgecontainerClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            default_max_pods_per_node: {
                value: cdktf.numberToHclTerraform(this._defaultMaxPodsPerNode),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            external_load_balancer_ipv4_address_pools: {
                value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(this._externalLoadBalancerIpv4AddressPools),
                isBlock: false,
                type: "list",
                storageClassType: "stringList",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            release_channel: {
                value: cdktf.stringToHclTerraform(this._releaseChannel),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            target_version: {
                value: cdktf.stringToHclTerraform(this._targetVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            authorization: {
                value: edgecontainerClusterAuthorizationToHclTerraform(this._authorization.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EdgecontainerClusterAuthorizationList",
            },
            control_plane: {
                value: edgecontainerClusterControlPlaneToHclTerraform(this._controlPlane.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EdgecontainerClusterControlPlaneList",
            },
            control_plane_encryption: {
                value: edgecontainerClusterControlPlaneEncryptionToHclTerraform(this._controlPlaneEncryption.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EdgecontainerClusterControlPlaneEncryptionList",
            },
            fleet: {
                value: edgecontainerClusterFleetToHclTerraform(this._fleet.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EdgecontainerClusterFleetList",
            },
            maintenance_policy: {
                value: edgecontainerClusterMaintenancePolicyToHclTerraform(this._maintenancePolicy.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EdgecontainerClusterMaintenancePolicyList",
            },
            networking: {
                value: edgecontainerClusterNetworkingToHclTerraform(this._networking.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EdgecontainerClusterNetworkingList",
            },
            system_addons_config: {
                value: edgecontainerClusterSystemAddonsConfigToHclTerraform(this._systemAddonsConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "EdgecontainerClusterSystemAddonsConfigList",
            },
            timeouts: {
                value: edgecontainerClusterTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "EdgecontainerClusterTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.EdgecontainerCluster = EdgecontainerCluster;
_y = JSII_RTTI_SYMBOL_1;
EdgecontainerCluster[_y] = { fqn: "@cdktf/provider-google.edgecontainerCluster.EdgecontainerCluster", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
EdgecontainerCluster.tfResourceType = "google_edgecontainer_cluster";
