// @cdktf/provider-google/lib/data-google-sql-database-instances/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesServerCaCertList = exports.DataGoogleSqlDatabaseInstancesInstancesServerCaCertOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesServerCaCertToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesServerCaCertToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextList = exports.DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationList = exports.DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesIpAddressList = exports.DataGoogleSqlDatabaseInstancesInstancesIpAddressOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesIpAddressToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesIpAddressToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesCloneList = exports.DataGoogleSqlDatabaseInstancesInstancesCloneOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesCloneToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesCloneToTerraform = void 0;
exports.DataGoogleSqlDatabaseInstances = exports.DataGoogleSqlDatabaseInstancesInstancesList = exports.DataGoogleSqlDatabaseInstancesInstancesOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksOutputReference = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksToHclTerraform = exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksToTerraform = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigList = exports.DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleSqlDatabaseInstancesInstancesCloneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesCloneToTerraform = dataGoogleSqlDatabaseInstancesInstancesCloneToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesCloneToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesCloneToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesCloneToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesCloneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allocated_ip_range - computed: true, optional: false, required: false
    get allocatedIpRange() {
        return this.getStringAttribute('allocated_ip_range');
    }
    // database_names - computed: true, optional: false, required: false
    get databaseNames() {
        return this.getListAttribute('database_names');
    }
    // point_in_time - computed: true, optional: false, required: false
    get pointInTime() {
        return this.getStringAttribute('point_in_time');
    }
    // preferred_zone - computed: true, optional: false, required: false
    get preferredZone() {
        return this.getStringAttribute('preferred_zone');
    }
    // source_instance_name - computed: true, optional: false, required: false
    get sourceInstanceName() {
        return this.getStringAttribute('source_instance_name');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesCloneOutputReference = DataGoogleSqlDatabaseInstancesInstancesCloneOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesCloneOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesCloneOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesCloneList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesCloneOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesCloneList = DataGoogleSqlDatabaseInstancesInstancesCloneList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesCloneList[_b] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesCloneList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesIpAddressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesIpAddressToTerraform = dataGoogleSqlDatabaseInstancesInstancesIpAddressToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesIpAddressToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesIpAddressToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesIpAddressToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesIpAddressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ip_address - computed: true, optional: false, required: false
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    // time_to_retire - computed: true, optional: false, required: false
    get timeToRetire() {
        return this.getStringAttribute('time_to_retire');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesIpAddressOutputReference = DataGoogleSqlDatabaseInstancesInstancesIpAddressOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesIpAddressOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesIpAddressOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesIpAddressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesIpAddressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesIpAddressList = DataGoogleSqlDatabaseInstancesInstancesIpAddressList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesIpAddressList[_d] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesIpAddressList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationToTerraform = dataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // ca_certificate - computed: true, optional: false, required: false
    get caCertificate() {
        return this.getStringAttribute('ca_certificate');
    }
    // cascadable_replica - computed: true, optional: false, required: false
    get cascadableReplica() {
        return this.getBooleanAttribute('cascadable_replica');
    }
    // client_certificate - computed: true, optional: false, required: false
    get clientCertificate() {
        return this.getStringAttribute('client_certificate');
    }
    // client_key - computed: true, optional: false, required: false
    get clientKey() {
        return this.getStringAttribute('client_key');
    }
    // connect_retry_interval - computed: true, optional: false, required: false
    get connectRetryInterval() {
        return this.getNumberAttribute('connect_retry_interval');
    }
    // dump_file_path - computed: true, optional: false, required: false
    get dumpFilePath() {
        return this.getStringAttribute('dump_file_path');
    }
    // failover_target - computed: true, optional: false, required: false
    get failoverTarget() {
        return this.getBooleanAttribute('failover_target');
    }
    // master_heartbeat_period - computed: true, optional: false, required: false
    get masterHeartbeatPeriod() {
        return this.getNumberAttribute('master_heartbeat_period');
    }
    // password - computed: true, optional: false, required: false
    get password() {
        return this.getStringAttribute('password');
    }
    // ssl_cipher - computed: true, optional: false, required: false
    get sslCipher() {
        return this.getStringAttribute('ssl_cipher');
    }
    // username - computed: true, optional: false, required: false
    get username() {
        return this.getStringAttribute('username');
    }
    // verify_server_certificate - computed: true, optional: false, required: false
    get verifyServerCertificate() {
        return this.getBooleanAttribute('verify_server_certificate');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationOutputReference = DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationList = DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationList[_f] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextToTerraform = dataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // backup_run_id - computed: true, optional: false, required: false
    get backupRunId() {
        return this.getNumberAttribute('backup_run_id');
    }
    // instance_id - computed: true, optional: false, required: false
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextOutputReference = DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextList = DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextList[_h] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesServerCaCertToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesServerCaCertToTerraform = dataGoogleSqlDatabaseInstancesInstancesServerCaCertToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesServerCaCertToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesServerCaCertToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesServerCaCertToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesServerCaCertOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cert - computed: true, optional: false, required: false
    get cert() {
        return this.getStringAttribute('cert');
    }
    // common_name - computed: true, optional: false, required: false
    get commonName() {
        return this.getStringAttribute('common_name');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // expiration_time - computed: true, optional: false, required: false
    get expirationTime() {
        return this.getStringAttribute('expiration_time');
    }
    // sha1_fingerprint - computed: true, optional: false, required: false
    get sha1Fingerprint() {
        return this.getStringAttribute('sha1_fingerprint');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesServerCaCertOutputReference = DataGoogleSqlDatabaseInstancesInstancesServerCaCertOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesServerCaCertOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesServerCaCertOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesServerCaCertList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesServerCaCertOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesServerCaCertList = DataGoogleSqlDatabaseInstancesInstancesServerCaCertList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesServerCaCertList[_k] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesServerCaCertList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigList = DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigList[_m] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // threads_per_core - computed: true, optional: false, required: false
    get threadsPerCore() {
        return this.getNumberAttribute('threads_per_core');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesList = DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesList[_p] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // retained_backups - computed: true, optional: false, required: false
    get retainedBackups() {
        return this.getNumberAttribute('retained_backups');
    }
    // retention_unit - computed: true, optional: false, required: false
    get retentionUnit() {
        return this.getStringAttribute('retention_unit');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsList = DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsList[_r] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // backup_retention_settings - computed: true, optional: false, required: false
        this._backupRetentionSettings = new DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationBackupRetentionSettingsList(this, "backup_retention_settings", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get backupRetentionSettings() {
        return this._backupRetentionSettings;
    }
    // binary_log_enabled - computed: true, optional: false, required: false
    get binaryLogEnabled() {
        return this.getBooleanAttribute('binary_log_enabled');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    // point_in_time_recovery_enabled - computed: true, optional: false, required: false
    get pointInTimeRecoveryEnabled() {
        return this.getBooleanAttribute('point_in_time_recovery_enabled');
    }
    // start_time - computed: true, optional: false, required: false
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    // transaction_log_retention_days - computed: true, optional: false, required: false
    get transactionLogRetentionDays() {
        return this.getNumberAttribute('transaction_log_retention_days');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationList = DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationList[_t] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_cache_enabled - computed: true, optional: false, required: false
    get dataCacheEnabled() {
        return this.getBooleanAttribute('data_cache_enabled');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigList = DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigList[_v] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsList = DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsList[_x] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // end_date - computed: true, optional: false, required: false
    get endDate() {
        return this.getStringAttribute('end_date');
    }
    // start_date - computed: true, optional: false, required: false
    get startDate() {
        return this.getStringAttribute('start_date');
    }
    // time - computed: true, optional: false, required: false
    get time() {
        return this.getStringAttribute('time');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodList = DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodList[_z] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // query_insights_enabled - computed: true, optional: false, required: false
    get queryInsightsEnabled() {
        return this.getBooleanAttribute('query_insights_enabled');
    }
    // query_plans_per_minute - computed: true, optional: false, required: false
    get queryPlansPerMinute() {
        return this.getNumberAttribute('query_plans_per_minute');
    }
    // query_string_length - computed: true, optional: false, required: false
    get queryStringLength() {
        return this.getNumberAttribute('query_string_length');
    }
    // record_application_tags - computed: true, optional: false, required: false
    get recordApplicationTags() {
        return this.getBooleanAttribute('record_application_tags');
    }
    // record_client_address - computed: true, optional: false, required: false
    get recordClientAddress() {
        return this.getBooleanAttribute('record_client_address');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigList = DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigList[_1] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // expiration_time - computed: true, optional: false, required: false
    get expirationTime() {
        return this.getStringAttribute('expiration_time');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksOutputReference[_2] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksList = DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksList;
_3 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksList[_3] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // consumer_network - computed: true, optional: false, required: false
    get consumerNetwork() {
        return this.getStringAttribute('consumer_network');
    }
    // consumer_service_project_id - computed: true, optional: false, required: false
    get consumerServiceProjectId() {
        return this.getStringAttribute('consumer_service_project_id');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference[_4] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsList = DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsList;
_5 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsList[_5] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // psc_auto_connections - computed: true, optional: false, required: false
        this._pscAutoConnections = new DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigPscAutoConnectionsList(this, "psc_auto_connections", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allowed_consumer_projects - computed: true, optional: false, required: false
    get allowedConsumerProjects() {
        return cdktf.Fn.tolist(this.getListAttribute('allowed_consumer_projects'));
    }
    get pscAutoConnections() {
        return this._pscAutoConnections;
    }
    // psc_enabled - computed: true, optional: false, required: false
    get pscEnabled() {
        return this.getBooleanAttribute('psc_enabled');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigOutputReference[_6] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigList = DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigList;
_7 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigList[_7] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // authorized_networks - computed: true, optional: false, required: false
        this._authorizedNetworks = new DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationAuthorizedNetworksList(this, "authorized_networks", true);
        // psc_config - computed: true, optional: false, required: false
        this._pscConfig = new DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationPscConfigList(this, "psc_config", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // allocated_ip_range - computed: true, optional: false, required: false
    get allocatedIpRange() {
        return this.getStringAttribute('allocated_ip_range');
    }
    get authorizedNetworks() {
        return this._authorizedNetworks;
    }
    // enable_private_path_for_google_cloud_services - computed: true, optional: false, required: false
    get enablePrivatePathForGoogleCloudServices() {
        return this.getBooleanAttribute('enable_private_path_for_google_cloud_services');
    }
    // ipv4_enabled - computed: true, optional: false, required: false
    get ipv4Enabled() {
        return this.getBooleanAttribute('ipv4_enabled');
    }
    // private_network - computed: true, optional: false, required: false
    get privateNetwork() {
        return this.getStringAttribute('private_network');
    }
    get pscConfig() {
        return this._pscConfig;
    }
    // server_ca_mode - computed: true, optional: false, required: false
    get serverCaMode() {
        return this.getStringAttribute('server_ca_mode');
    }
    // ssl_mode - computed: true, optional: false, required: false
    get sslMode() {
        return this.getStringAttribute('ssl_mode');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationOutputReference[_8] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationList = DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationList;
_9 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationList[_9] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // follow_gae_application - computed: true, optional: false, required: false
    get followGaeApplication() {
        return this.getStringAttribute('follow_gae_application');
    }
    // secondary_zone - computed: true, optional: false, required: false
    get secondaryZone() {
        return this.getStringAttribute('secondary_zone');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceOutputReference[_10] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceList = DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceList;
_11 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceList[_11] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day - computed: true, optional: false, required: false
    get day() {
        return this.getNumberAttribute('day');
    }
    // hour - computed: true, optional: false, required: false
    get hour() {
        return this.getNumberAttribute('hour');
    }
    // update_track - computed: true, optional: false, required: false
    get updateTrack() {
        return this.getStringAttribute('update_track');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowOutputReference[_12] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowList = DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowList;
_13 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowList[_13] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // complexity - computed: true, optional: false, required: false
    get complexity() {
        return this.getStringAttribute('complexity');
    }
    // disallow_username_substring - computed: true, optional: false, required: false
    get disallowUsernameSubstring() {
        return this.getBooleanAttribute('disallow_username_substring');
    }
    // enable_password_policy - computed: true, optional: false, required: false
    get enablePasswordPolicy() {
        return this.getBooleanAttribute('enable_password_policy');
    }
    // min_length - computed: true, optional: false, required: false
    get minLength() {
        return this.getNumberAttribute('min_length');
    }
    // password_change_interval - computed: true, optional: false, required: false
    get passwordChangeInterval() {
        return this.getStringAttribute('password_change_interval');
    }
    // reuse_interval - computed: true, optional: false, required: false
    get reuseInterval() {
        return this.getNumberAttribute('reuse_interval');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyOutputReference[_14] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyList = DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyList;
_15 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyList[_15] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    // retention_interval - computed: true, optional: false, required: false
    get retentionInterval() {
        return this.getStringAttribute('retention_interval');
    }
    // upload_interval - computed: true, optional: false, required: false
    get uploadInterval() {
        return this.getStringAttribute('upload_interval');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigOutputReference[_16] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigList = DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigList;
_17 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigList[_17] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsToTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesSettingsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesSettingsToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesSettingsToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // active_directory_config - computed: true, optional: false, required: false
        this._activeDirectoryConfig = new DataGoogleSqlDatabaseInstancesInstancesSettingsActiveDirectoryConfigList(this, "active_directory_config", false);
        // advanced_machine_features - computed: true, optional: false, required: false
        this._advancedMachineFeatures = new DataGoogleSqlDatabaseInstancesInstancesSettingsAdvancedMachineFeaturesList(this, "advanced_machine_features", false);
        // backup_configuration - computed: true, optional: false, required: false
        this._backupConfiguration = new DataGoogleSqlDatabaseInstancesInstancesSettingsBackupConfigurationList(this, "backup_configuration", false);
        // data_cache_config - computed: true, optional: false, required: false
        this._dataCacheConfig = new DataGoogleSqlDatabaseInstancesInstancesSettingsDataCacheConfigList(this, "data_cache_config", false);
        // database_flags - computed: true, optional: false, required: false
        this._databaseFlags = new DataGoogleSqlDatabaseInstancesInstancesSettingsDatabaseFlagsList(this, "database_flags", true);
        // deny_maintenance_period - computed: true, optional: false, required: false
        this._denyMaintenancePeriod = new DataGoogleSqlDatabaseInstancesInstancesSettingsDenyMaintenancePeriodList(this, "deny_maintenance_period", false);
        // insights_config - computed: true, optional: false, required: false
        this._insightsConfig = new DataGoogleSqlDatabaseInstancesInstancesSettingsInsightsConfigList(this, "insights_config", false);
        // ip_configuration - computed: true, optional: false, required: false
        this._ipConfiguration = new DataGoogleSqlDatabaseInstancesInstancesSettingsIpConfigurationList(this, "ip_configuration", false);
        // location_preference - computed: true, optional: false, required: false
        this._locationPreference = new DataGoogleSqlDatabaseInstancesInstancesSettingsLocationPreferenceList(this, "location_preference", false);
        // maintenance_window - computed: true, optional: false, required: false
        this._maintenanceWindow = new DataGoogleSqlDatabaseInstancesInstancesSettingsMaintenanceWindowList(this, "maintenance_window", false);
        // password_validation_policy - computed: true, optional: false, required: false
        this._passwordValidationPolicy = new DataGoogleSqlDatabaseInstancesInstancesSettingsPasswordValidationPolicyList(this, "password_validation_policy", false);
        // sql_server_audit_config - computed: true, optional: false, required: false
        this._sqlServerAuditConfig = new DataGoogleSqlDatabaseInstancesInstancesSettingsSqlServerAuditConfigList(this, "sql_server_audit_config", false);
        // user_labels - computed: true, optional: false, required: false
        this._userLabels = new cdktf.StringMap(this, "user_labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // activation_policy - computed: true, optional: false, required: false
    get activationPolicy() {
        return this.getStringAttribute('activation_policy');
    }
    get activeDirectoryConfig() {
        return this._activeDirectoryConfig;
    }
    get advancedMachineFeatures() {
        return this._advancedMachineFeatures;
    }
    // availability_type - computed: true, optional: false, required: false
    get availabilityType() {
        return this.getStringAttribute('availability_type');
    }
    get backupConfiguration() {
        return this._backupConfiguration;
    }
    // collation - computed: true, optional: false, required: false
    get collation() {
        return this.getStringAttribute('collation');
    }
    // connector_enforcement - computed: true, optional: false, required: false
    get connectorEnforcement() {
        return this.getStringAttribute('connector_enforcement');
    }
    get dataCacheConfig() {
        return this._dataCacheConfig;
    }
    get databaseFlags() {
        return this._databaseFlags;
    }
    // deletion_protection_enabled - computed: true, optional: false, required: false
    get deletionProtectionEnabled() {
        return this.getBooleanAttribute('deletion_protection_enabled');
    }
    get denyMaintenancePeriod() {
        return this._denyMaintenancePeriod;
    }
    // disk_autoresize - computed: true, optional: false, required: false
    get diskAutoresize() {
        return this.getBooleanAttribute('disk_autoresize');
    }
    // disk_autoresize_limit - computed: true, optional: false, required: false
    get diskAutoresizeLimit() {
        return this.getNumberAttribute('disk_autoresize_limit');
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    // disk_type - computed: true, optional: false, required: false
    get diskType() {
        return this.getStringAttribute('disk_type');
    }
    // edition - computed: true, optional: false, required: false
    get edition() {
        return this.getStringAttribute('edition');
    }
    // enable_dataplex_integration - computed: true, optional: false, required: false
    get enableDataplexIntegration() {
        return this.getBooleanAttribute('enable_dataplex_integration');
    }
    // enable_google_ml_integration - computed: true, optional: false, required: false
    get enableGoogleMlIntegration() {
        return this.getBooleanAttribute('enable_google_ml_integration');
    }
    get insightsConfig() {
        return this._insightsConfig;
    }
    get ipConfiguration() {
        return this._ipConfiguration;
    }
    get locationPreference() {
        return this._locationPreference;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    get passwordValidationPolicy() {
        return this._passwordValidationPolicy;
    }
    // pricing_plan - computed: true, optional: false, required: false
    get pricingPlan() {
        return this.getStringAttribute('pricing_plan');
    }
    get sqlServerAuditConfig() {
        return this._sqlServerAuditConfig;
    }
    // tier - computed: true, optional: false, required: false
    get tier() {
        return this.getStringAttribute('tier');
    }
    // time_zone - computed: true, optional: false, required: false
    get timeZone() {
        return this.getStringAttribute('time_zone');
    }
    get userLabels() {
        return this._userLabels;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getNumberAttribute('version');
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsOutputReference = DataGoogleSqlDatabaseInstancesInstancesSettingsOutputReference;
_18 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsOutputReference[_18] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesSettingsList = DataGoogleSqlDatabaseInstancesInstancesSettingsList;
_19 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesSettingsList[_19] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesSettingsList", version: "14.13.1" };
function dataGoogleSqlDatabaseInstancesInstancesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleSqlDatabaseInstancesInstancesToTerraform = dataGoogleSqlDatabaseInstancesInstancesToTerraform;
function dataGoogleSqlDatabaseInstancesInstancesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleSqlDatabaseInstancesInstancesToHclTerraform = dataGoogleSqlDatabaseInstancesInstancesToHclTerraform;
class DataGoogleSqlDatabaseInstancesInstancesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // clone - computed: true, optional: false, required: false
        this._clone = new DataGoogleSqlDatabaseInstancesInstancesCloneList(this, "clone", false);
        // ip_address - computed: true, optional: false, required: false
        this._ipAddress = new DataGoogleSqlDatabaseInstancesInstancesIpAddressList(this, "ip_address", false);
        // replica_configuration - computed: true, optional: false, required: false
        this._replicaConfiguration = new DataGoogleSqlDatabaseInstancesInstancesReplicaConfigurationList(this, "replica_configuration", false);
        // restore_backup_context - computed: true, optional: false, required: false
        this._restoreBackupContext = new DataGoogleSqlDatabaseInstancesInstancesRestoreBackupContextList(this, "restore_backup_context", false);
        // server_ca_cert - computed: true, optional: false, required: false
        this._serverCaCert = new DataGoogleSqlDatabaseInstancesInstancesServerCaCertList(this, "server_ca_cert", false);
        // settings - computed: true, optional: false, required: false
        this._settings = new DataGoogleSqlDatabaseInstancesInstancesSettingsList(this, "settings", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // available_maintenance_versions - computed: true, optional: false, required: false
    get availableMaintenanceVersions() {
        return this.getListAttribute('available_maintenance_versions');
    }
    get clone() {
        return this._clone;
    }
    // connection_name - computed: true, optional: false, required: false
    get connectionName() {
        return this.getStringAttribute('connection_name');
    }
    // database_version - computed: true, optional: false, required: false
    get databaseVersion() {
        return this.getStringAttribute('database_version');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    // dns_name - computed: true, optional: false, required: false
    get dnsName() {
        return this.getStringAttribute('dns_name');
    }
    // encryption_key_name - computed: true, optional: false, required: false
    get encryptionKeyName() {
        return this.getStringAttribute('encryption_key_name');
    }
    // first_ip_address - computed: true, optional: false, required: false
    get firstIpAddress() {
        return this.getStringAttribute('first_ip_address');
    }
    // instance_type - computed: true, optional: false, required: false
    get instanceType() {
        return this.getStringAttribute('instance_type');
    }
    get ipAddress() {
        return this._ipAddress;
    }
    // maintenance_version - computed: true, optional: false, required: false
    get maintenanceVersion() {
        return this.getStringAttribute('maintenance_version');
    }
    // master_instance_name - computed: true, optional: false, required: false
    get masterInstanceName() {
        return this.getStringAttribute('master_instance_name');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // private_ip_address - computed: true, optional: false, required: false
    get privateIpAddress() {
        return this.getStringAttribute('private_ip_address');
    }
    // project - computed: true, optional: false, required: false
    get project() {
        return this.getStringAttribute('project');
    }
    // psc_service_attachment_link - computed: true, optional: false, required: false
    get pscServiceAttachmentLink() {
        return this.getStringAttribute('psc_service_attachment_link');
    }
    // public_ip_address - computed: true, optional: false, required: false
    get publicIpAddress() {
        return this.getStringAttribute('public_ip_address');
    }
    // region - computed: true, optional: false, required: false
    get region() {
        return this.getStringAttribute('region');
    }
    get replicaConfiguration() {
        return this._replicaConfiguration;
    }
    // replica_names - computed: true, optional: false, required: false
    get replicaNames() {
        return this.getListAttribute('replica_names');
    }
    get restoreBackupContext() {
        return this._restoreBackupContext;
    }
    // root_password - computed: true, optional: false, required: false
    get rootPassword() {
        return this.getStringAttribute('root_password');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get serverCaCert() {
        return this._serverCaCert;
    }
    // service_account_email_address - computed: true, optional: false, required: false
    get serviceAccountEmailAddress() {
        return this.getStringAttribute('service_account_email_address');
    }
    get settings() {
        return this._settings;
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesOutputReference = DataGoogleSqlDatabaseInstancesInstancesOutputReference;
_20 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesOutputReference[_20] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesOutputReference", version: "14.13.1" };
class DataGoogleSqlDatabaseInstancesInstancesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleSqlDatabaseInstancesInstancesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleSqlDatabaseInstancesInstancesList = DataGoogleSqlDatabaseInstancesInstancesList;
_21 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstancesInstancesList[_21] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstancesInstancesList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/sql_database_instances google_sql_database_instances}
*/
class DataGoogleSqlDatabaseInstances extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleSqlDatabaseInstances resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleSqlDatabaseInstances to import
    * @param importFromId The id of the existing DataGoogleSqlDatabaseInstances that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/sql_database_instances#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleSqlDatabaseInstances to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_sql_database_instances", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/sql_database_instances google_sql_database_instances} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleSqlDatabaseInstancesConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_sql_database_instances',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // instances - computed: true, optional: false, required: false
        this._instances = new DataGoogleSqlDatabaseInstancesInstancesList(this, "instances", false);
        this._databaseVersion = config.databaseVersion;
        this._id = config.id;
        this._project = config.project;
        this._region = config.region;
        this._state = config.state;
        this._tier = config.tier;
        this._zone = config.zone;
    }
    get databaseVersion() {
        return this.getStringAttribute('database_version');
    }
    set databaseVersion(value) {
        this._databaseVersion = value;
    }
    resetDatabaseVersion() {
        this._databaseVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseVersionInput() {
        return this._databaseVersion;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instances() {
        return this._instances;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    resetRegion() {
        this._region = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get state() {
        return this.getStringAttribute('state');
    }
    set state(value) {
        this._state = value;
    }
    resetState() {
        this._state = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stateInput() {
        return this._state;
    }
    get tier() {
        return this.getStringAttribute('tier');
    }
    set tier(value) {
        this._tier = value;
    }
    resetTier() {
        this._tier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tierInput() {
        return this._tier;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            database_version: cdktf.stringToTerraform(this._databaseVersion),
            id: cdktf.stringToTerraform(this._id),
            project: cdktf.stringToTerraform(this._project),
            region: cdktf.stringToTerraform(this._region),
            state: cdktf.stringToTerraform(this._state),
            tier: cdktf.stringToTerraform(this._tier),
            zone: cdktf.stringToTerraform(this._zone),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            database_version: {
                value: cdktf.stringToHclTerraform(this._databaseVersion),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            region: {
                value: cdktf.stringToHclTerraform(this._region),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            state: {
                value: cdktf.stringToHclTerraform(this._state),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            tier: {
                value: cdktf.stringToHclTerraform(this._tier),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            zone: {
                value: cdktf.stringToHclTerraform(this._zone),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleSqlDatabaseInstances = DataGoogleSqlDatabaseInstances;
_22 = JSII_RTTI_SYMBOL_1;
DataGoogleSqlDatabaseInstances[_22] = { fqn: "@cdktf/provider-google.dataGoogleSqlDatabaseInstances.DataGoogleSqlDatabaseInstances", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleSqlDatabaseInstances.tfResourceType = "google_sql_database_instances";
