// @cdktf/provider-google/lib/compute-security-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14;
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToTerraform = exports.ComputeSecurityPolicyRuleMatchOutputReference = exports.computeSecurityPolicyRuleMatchToHclTerraform = exports.computeSecurityPolicyRuleMatchToTerraform = exports.ComputeSecurityPolicyRuleMatchExprOptionsOutputReference = exports.computeSecurityPolicyRuleMatchExprOptionsToHclTerraform = exports.computeSecurityPolicyRuleMatchExprOptionsToTerraform = exports.ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsOutputReference = exports.computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToHclTerraform = exports.computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToTerraform = exports.ComputeSecurityPolicyRuleMatchExprOutputReference = exports.computeSecurityPolicyRuleMatchExprToHclTerraform = exports.computeSecurityPolicyRuleMatchExprToTerraform = exports.ComputeSecurityPolicyRuleMatchConfigOutputReference = exports.computeSecurityPolicyRuleMatchConfigToHclTerraform = exports.computeSecurityPolicyRuleMatchConfigToTerraform = exports.ComputeSecurityPolicyRuleHeaderActionOutputReference = exports.computeSecurityPolicyRuleHeaderActionToHclTerraform = exports.computeSecurityPolicyRuleHeaderActionToTerraform = exports.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsList = exports.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsOutputReference = exports.computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToHclTerraform = exports.computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToTerraform = exports.ComputeSecurityPolicyRecaptchaOptionsConfigOutputReference = exports.computeSecurityPolicyRecaptchaOptionsConfigToHclTerraform = exports.computeSecurityPolicyRecaptchaOptionsConfigToTerraform = exports.ComputeSecurityPolicyAdvancedOptionsConfigOutputReference = exports.computeSecurityPolicyAdvancedOptionsConfigToHclTerraform = exports.computeSecurityPolicyAdvancedOptionsConfigToTerraform = exports.ComputeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigOutputReference = exports.computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToHclTerraform = exports.computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToTerraform = exports.ComputeSecurityPolicyAdaptiveProtectionConfigOutputReference = exports.computeSecurityPolicyAdaptiveProtectionConfigToHclTerraform = exports.computeSecurityPolicyAdaptiveProtectionConfigToTerraform = exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigOutputReference = exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToHclTerraform = exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToTerraform = exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsList = exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsOutputReference = exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToHclTerraform = exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToTerraform = exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsList = exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsOutputReference = exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToHclTerraform = exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToTerraform = void 0;
exports.ComputeSecurityPolicy = exports.ComputeSecurityPolicyTimeoutsOutputReference = exports.computeSecurityPolicyTimeoutsToHclTerraform = exports.computeSecurityPolicyTimeoutsToTerraform = exports.ComputeSecurityPolicyRuleList = exports.ComputeSecurityPolicyRuleOutputReference = exports.computeSecurityPolicyRuleToHclTerraform = exports.computeSecurityPolicyRuleToTerraform = exports.ComputeSecurityPolicyRuleRedirectOptionsOutputReference = exports.computeSecurityPolicyRuleRedirectOptionsToHclTerraform = exports.computeSecurityPolicyRuleRedirectOptionsToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsList = exports.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToTerraform = exports.ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdOutputReference = exports.computeSecurityPolicyRuleRateLimitOptionsBanThresholdToHclTerraform = exports.computeSecurityPolicyRuleRateLimitOptionsBanThresholdToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToHclTerraform = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToTerraform = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderList = exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderOutputReference = exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToHclTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_each_unique_value: cdktf.booleanToTerraform(struct.enableEachUniqueValue),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToTerraform = computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToTerraform;
function computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable_each_unique_value: {
            value: cdktf.booleanToHclTerraform(struct.enableEachUniqueValue),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToHclTerraform = computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToHclTerraform;
class ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableEachUniqueValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableEachUniqueValue = this._enableEachUniqueValue;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._enableEachUniqueValue = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._enableEachUniqueValue = value.enableEachUniqueValue;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get enableEachUniqueValue() {
        return this.getBooleanAttribute('enable_each_unique_value');
    }
    set enableEachUniqueValue(value) {
        this._enableEachUniqueValue = value;
    }
    resetEnableEachUniqueValue() {
        this._enableEachUniqueValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableEachUniqueValueInput() {
        return this._enableEachUniqueValue;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsOutputReference = ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsOutputReference[_a] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsList = ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsList;
_b = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsList[_b] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsList", version: "14.13.1" };
function computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_deploy_confidence_threshold: cdktf.numberToTerraform(struct.autoDeployConfidenceThreshold),
        auto_deploy_expiration_sec: cdktf.numberToTerraform(struct.autoDeployExpirationSec),
        auto_deploy_impacted_baseline_threshold: cdktf.numberToTerraform(struct.autoDeployImpactedBaselineThreshold),
        auto_deploy_load_threshold: cdktf.numberToTerraform(struct.autoDeployLoadThreshold),
        detection_absolute_qps: cdktf.numberToTerraform(struct.detectionAbsoluteQps),
        detection_load_threshold: cdktf.numberToTerraform(struct.detectionLoadThreshold),
        detection_relative_to_baseline_qps: cdktf.numberToTerraform(struct.detectionRelativeToBaselineQps),
        name: cdktf.stringToTerraform(struct.name),
        traffic_granularity_configs: cdktf.listMapper(computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToTerraform, true)(struct.trafficGranularityConfigs),
    };
}
exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToTerraform = computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToTerraform;
function computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        auto_deploy_confidence_threshold: {
            value: cdktf.numberToHclTerraform(struct.autoDeployConfidenceThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        auto_deploy_expiration_sec: {
            value: cdktf.numberToHclTerraform(struct.autoDeployExpirationSec),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        auto_deploy_impacted_baseline_threshold: {
            value: cdktf.numberToHclTerraform(struct.autoDeployImpactedBaselineThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        auto_deploy_load_threshold: {
            value: cdktf.numberToHclTerraform(struct.autoDeployLoadThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        detection_absolute_qps: {
            value: cdktf.numberToHclTerraform(struct.detectionAbsoluteQps),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        detection_load_threshold: {
            value: cdktf.numberToHclTerraform(struct.detectionLoadThreshold),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        detection_relative_to_baseline_qps: {
            value: cdktf.numberToHclTerraform(struct.detectionRelativeToBaselineQps),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        traffic_granularity_configs: {
            value: cdktf.listMapperHcl(computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsToHclTerraform, true)(struct.trafficGranularityConfigs),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToHclTerraform = computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToHclTerraform;
class ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // traffic_granularity_configs - computed: false, optional: true, required: false
        this._trafficGranularityConfigs = new ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsTrafficGranularityConfigsList(this, "traffic_granularity_configs", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoDeployConfidenceThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoDeployConfidenceThreshold = this._autoDeployConfidenceThreshold;
        }
        if (this._autoDeployExpirationSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoDeployExpirationSec = this._autoDeployExpirationSec;
        }
        if (this._autoDeployImpactedBaselineThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoDeployImpactedBaselineThreshold = this._autoDeployImpactedBaselineThreshold;
        }
        if (this._autoDeployLoadThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoDeployLoadThreshold = this._autoDeployLoadThreshold;
        }
        if (this._detectionAbsoluteQps !== undefined) {
            hasAnyValues = true;
            internalValueResult.detectionAbsoluteQps = this._detectionAbsoluteQps;
        }
        if (this._detectionLoadThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.detectionLoadThreshold = this._detectionLoadThreshold;
        }
        if (this._detectionRelativeToBaselineQps !== undefined) {
            hasAnyValues = true;
            internalValueResult.detectionRelativeToBaselineQps = this._detectionRelativeToBaselineQps;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._trafficGranularityConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.trafficGranularityConfigs = this._trafficGranularityConfigs?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._autoDeployConfidenceThreshold = undefined;
            this._autoDeployExpirationSec = undefined;
            this._autoDeployImpactedBaselineThreshold = undefined;
            this._autoDeployLoadThreshold = undefined;
            this._detectionAbsoluteQps = undefined;
            this._detectionLoadThreshold = undefined;
            this._detectionRelativeToBaselineQps = undefined;
            this._name = undefined;
            this._trafficGranularityConfigs.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._autoDeployConfidenceThreshold = value.autoDeployConfidenceThreshold;
            this._autoDeployExpirationSec = value.autoDeployExpirationSec;
            this._autoDeployImpactedBaselineThreshold = value.autoDeployImpactedBaselineThreshold;
            this._autoDeployLoadThreshold = value.autoDeployLoadThreshold;
            this._detectionAbsoluteQps = value.detectionAbsoluteQps;
            this._detectionLoadThreshold = value.detectionLoadThreshold;
            this._detectionRelativeToBaselineQps = value.detectionRelativeToBaselineQps;
            this._name = value.name;
            this._trafficGranularityConfigs.internalValue = value.trafficGranularityConfigs;
        }
    }
    get autoDeployConfidenceThreshold() {
        return this.getNumberAttribute('auto_deploy_confidence_threshold');
    }
    set autoDeployConfidenceThreshold(value) {
        this._autoDeployConfidenceThreshold = value;
    }
    resetAutoDeployConfidenceThreshold() {
        this._autoDeployConfidenceThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoDeployConfidenceThresholdInput() {
        return this._autoDeployConfidenceThreshold;
    }
    get autoDeployExpirationSec() {
        return this.getNumberAttribute('auto_deploy_expiration_sec');
    }
    set autoDeployExpirationSec(value) {
        this._autoDeployExpirationSec = value;
    }
    resetAutoDeployExpirationSec() {
        this._autoDeployExpirationSec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoDeployExpirationSecInput() {
        return this._autoDeployExpirationSec;
    }
    get autoDeployImpactedBaselineThreshold() {
        return this.getNumberAttribute('auto_deploy_impacted_baseline_threshold');
    }
    set autoDeployImpactedBaselineThreshold(value) {
        this._autoDeployImpactedBaselineThreshold = value;
    }
    resetAutoDeployImpactedBaselineThreshold() {
        this._autoDeployImpactedBaselineThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoDeployImpactedBaselineThresholdInput() {
        return this._autoDeployImpactedBaselineThreshold;
    }
    get autoDeployLoadThreshold() {
        return this.getNumberAttribute('auto_deploy_load_threshold');
    }
    set autoDeployLoadThreshold(value) {
        this._autoDeployLoadThreshold = value;
    }
    resetAutoDeployLoadThreshold() {
        this._autoDeployLoadThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoDeployLoadThresholdInput() {
        return this._autoDeployLoadThreshold;
    }
    get detectionAbsoluteQps() {
        return this.getNumberAttribute('detection_absolute_qps');
    }
    set detectionAbsoluteQps(value) {
        this._detectionAbsoluteQps = value;
    }
    resetDetectionAbsoluteQps() {
        this._detectionAbsoluteQps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get detectionAbsoluteQpsInput() {
        return this._detectionAbsoluteQps;
    }
    get detectionLoadThreshold() {
        return this.getNumberAttribute('detection_load_threshold');
    }
    set detectionLoadThreshold(value) {
        this._detectionLoadThreshold = value;
    }
    resetDetectionLoadThreshold() {
        this._detectionLoadThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get detectionLoadThresholdInput() {
        return this._detectionLoadThreshold;
    }
    get detectionRelativeToBaselineQps() {
        return this.getNumberAttribute('detection_relative_to_baseline_qps');
    }
    set detectionRelativeToBaselineQps(value) {
        this._detectionRelativeToBaselineQps = value;
    }
    resetDetectionRelativeToBaselineQps() {
        this._detectionRelativeToBaselineQps = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get detectionRelativeToBaselineQpsInput() {
        return this._detectionRelativeToBaselineQps;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get trafficGranularityConfigs() {
        return this._trafficGranularityConfigs;
    }
    putTrafficGranularityConfigs(value) {
        this._trafficGranularityConfigs.internalValue = value;
    }
    resetTrafficGranularityConfigs() {
        this._trafficGranularityConfigs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get trafficGranularityConfigsInput() {
        return this._trafficGranularityConfigs.internalValue;
    }
}
exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsOutputReference = ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsOutputReference[_c] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsList = ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsList;
_d = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsList[_d] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsList", version: "14.13.1" };
function computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable: cdktf.booleanToTerraform(struct.enable),
        rule_visibility: cdktf.stringToTerraform(struct.ruleVisibility),
        threshold_configs: cdktf.listMapper(computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToTerraform, true)(struct.thresholdConfigs),
    };
}
exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToTerraform = computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToTerraform;
function computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enable: {
            value: cdktf.booleanToHclTerraform(struct.enable),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        rule_visibility: {
            value: cdktf.stringToHclTerraform(struct.ruleVisibility),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        threshold_configs: {
            value: cdktf.listMapperHcl(computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsToHclTerraform, true)(struct.thresholdConfigs),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToHclTerraform = computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToHclTerraform;
class ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // threshold_configs - computed: false, optional: true, required: false
        this._thresholdConfigs = new ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigThresholdConfigsList(this, "threshold_configs", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enable !== undefined) {
            hasAnyValues = true;
            internalValueResult.enable = this._enable;
        }
        if (this._ruleVisibility !== undefined) {
            hasAnyValues = true;
            internalValueResult.ruleVisibility = this._ruleVisibility;
        }
        if (this._thresholdConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.thresholdConfigs = this._thresholdConfigs?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enable = undefined;
            this._ruleVisibility = undefined;
            this._thresholdConfigs.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enable = value.enable;
            this._ruleVisibility = value.ruleVisibility;
            this._thresholdConfigs.internalValue = value.thresholdConfigs;
        }
    }
    get enable() {
        return this.getBooleanAttribute('enable');
    }
    set enable(value) {
        this._enable = value;
    }
    resetEnable() {
        this._enable = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableInput() {
        return this._enable;
    }
    get ruleVisibility() {
        return this.getStringAttribute('rule_visibility');
    }
    set ruleVisibility(value) {
        this._ruleVisibility = value;
    }
    resetRuleVisibility() {
        this._ruleVisibility = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleVisibilityInput() {
        return this._ruleVisibility;
    }
    get thresholdConfigs() {
        return this._thresholdConfigs;
    }
    putThresholdConfigs(value) {
        this._thresholdConfigs.internalValue = value;
    }
    resetThresholdConfigs() {
        this._thresholdConfigs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get thresholdConfigsInput() {
        return this._thresholdConfigs.internalValue;
    }
}
exports.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigOutputReference = ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigOutputReference[_e] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigOutputReference", version: "14.13.1" };
function computeSecurityPolicyAdaptiveProtectionConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        layer_7_ddos_defense_config: computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToTerraform(struct.layer7DdosDefenseConfig),
    };
}
exports.computeSecurityPolicyAdaptiveProtectionConfigToTerraform = computeSecurityPolicyAdaptiveProtectionConfigToTerraform;
function computeSecurityPolicyAdaptiveProtectionConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        layer_7_ddos_defense_config: {
            value: computeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigToHclTerraform(struct.layer7DdosDefenseConfig),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyAdaptiveProtectionConfigToHclTerraform = computeSecurityPolicyAdaptiveProtectionConfigToHclTerraform;
class ComputeSecurityPolicyAdaptiveProtectionConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // layer_7_ddos_defense_config - computed: false, optional: true, required: false
        this._layer7DdosDefenseConfig = new ComputeSecurityPolicyAdaptiveProtectionConfigLayer7DdosDefenseConfigOutputReference(this, "layer_7_ddos_defense_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._layer7DdosDefenseConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.layer7DdosDefenseConfig = this._layer7DdosDefenseConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._layer7DdosDefenseConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._layer7DdosDefenseConfig.internalValue = value.layer7DdosDefenseConfig;
        }
    }
    get layer7DdosDefenseConfig() {
        return this._layer7DdosDefenseConfig;
    }
    putLayer7DdosDefenseConfig(value) {
        this._layer7DdosDefenseConfig.internalValue = value;
    }
    resetLayer7DdosDefenseConfig() {
        this._layer7DdosDefenseConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get layer7DdosDefenseConfigInput() {
        return this._layer7DdosDefenseConfig.internalValue;
    }
}
exports.ComputeSecurityPolicyAdaptiveProtectionConfigOutputReference = ComputeSecurityPolicyAdaptiveProtectionConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyAdaptiveProtectionConfigOutputReference[_f] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyAdaptiveProtectionConfigOutputReference", version: "14.13.1" };
function computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        content_types: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.contentTypes),
    };
}
exports.computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToTerraform = computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToTerraform;
function computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        content_types: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.contentTypes),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToHclTerraform = computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToHclTerraform;
class ComputeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._contentTypes !== undefined) {
            hasAnyValues = true;
            internalValueResult.contentTypes = this._contentTypes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._contentTypes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._contentTypes = value.contentTypes;
        }
    }
    get contentTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('content_types'));
    }
    set contentTypes(value) {
        this._contentTypes = value;
    }
    // Temporarily expose input value. Use with caution.
    get contentTypesInput() {
        return this._contentTypes;
    }
}
exports.ComputeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigOutputReference = ComputeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigOutputReference[_g] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigOutputReference", version: "14.13.1" };
function computeSecurityPolicyAdvancedOptionsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        json_parsing: cdktf.stringToTerraform(struct.jsonParsing),
        log_level: cdktf.stringToTerraform(struct.logLevel),
        user_ip_request_headers: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.userIpRequestHeaders),
        json_custom_config: computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToTerraform(struct.jsonCustomConfig),
    };
}
exports.computeSecurityPolicyAdvancedOptionsConfigToTerraform = computeSecurityPolicyAdvancedOptionsConfigToTerraform;
function computeSecurityPolicyAdvancedOptionsConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        json_parsing: {
            value: cdktf.stringToHclTerraform(struct.jsonParsing),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        log_level: {
            value: cdktf.stringToHclTerraform(struct.logLevel),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        user_ip_request_headers: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.userIpRequestHeaders),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        json_custom_config: {
            value: computeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigToHclTerraform(struct.jsonCustomConfig),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyAdvancedOptionsConfigToHclTerraform = computeSecurityPolicyAdvancedOptionsConfigToHclTerraform;
class ComputeSecurityPolicyAdvancedOptionsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // json_custom_config - computed: false, optional: true, required: false
        this._jsonCustomConfig = new ComputeSecurityPolicyAdvancedOptionsConfigJsonCustomConfigOutputReference(this, "json_custom_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._jsonParsing !== undefined) {
            hasAnyValues = true;
            internalValueResult.jsonParsing = this._jsonParsing;
        }
        if (this._logLevel !== undefined) {
            hasAnyValues = true;
            internalValueResult.logLevel = this._logLevel;
        }
        if (this._userIpRequestHeaders !== undefined) {
            hasAnyValues = true;
            internalValueResult.userIpRequestHeaders = this._userIpRequestHeaders;
        }
        if (this._jsonCustomConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.jsonCustomConfig = this._jsonCustomConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._jsonParsing = undefined;
            this._logLevel = undefined;
            this._userIpRequestHeaders = undefined;
            this._jsonCustomConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._jsonParsing = value.jsonParsing;
            this._logLevel = value.logLevel;
            this._userIpRequestHeaders = value.userIpRequestHeaders;
            this._jsonCustomConfig.internalValue = value.jsonCustomConfig;
        }
    }
    get jsonParsing() {
        return this.getStringAttribute('json_parsing');
    }
    set jsonParsing(value) {
        this._jsonParsing = value;
    }
    resetJsonParsing() {
        this._jsonParsing = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jsonParsingInput() {
        return this._jsonParsing;
    }
    get logLevel() {
        return this.getStringAttribute('log_level');
    }
    set logLevel(value) {
        this._logLevel = value;
    }
    resetLogLevel() {
        this._logLevel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get logLevelInput() {
        return this._logLevel;
    }
    get userIpRequestHeaders() {
        return cdktf.Fn.tolist(this.getListAttribute('user_ip_request_headers'));
    }
    set userIpRequestHeaders(value) {
        this._userIpRequestHeaders = value;
    }
    resetUserIpRequestHeaders() {
        this._userIpRequestHeaders = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userIpRequestHeadersInput() {
        return this._userIpRequestHeaders;
    }
    get jsonCustomConfig() {
        return this._jsonCustomConfig;
    }
    putJsonCustomConfig(value) {
        this._jsonCustomConfig.internalValue = value;
    }
    resetJsonCustomConfig() {
        this._jsonCustomConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get jsonCustomConfigInput() {
        return this._jsonCustomConfig.internalValue;
    }
}
exports.ComputeSecurityPolicyAdvancedOptionsConfigOutputReference = ComputeSecurityPolicyAdvancedOptionsConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyAdvancedOptionsConfigOutputReference[_h] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyAdvancedOptionsConfigOutputReference", version: "14.13.1" };
function computeSecurityPolicyRecaptchaOptionsConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        redirect_site_key: cdktf.stringToTerraform(struct.redirectSiteKey),
    };
}
exports.computeSecurityPolicyRecaptchaOptionsConfigToTerraform = computeSecurityPolicyRecaptchaOptionsConfigToTerraform;
function computeSecurityPolicyRecaptchaOptionsConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        redirect_site_key: {
            value: cdktf.stringToHclTerraform(struct.redirectSiteKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRecaptchaOptionsConfigToHclTerraform = computeSecurityPolicyRecaptchaOptionsConfigToHclTerraform;
class ComputeSecurityPolicyRecaptchaOptionsConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._redirectSiteKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.redirectSiteKey = this._redirectSiteKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._redirectSiteKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._redirectSiteKey = value.redirectSiteKey;
        }
    }
    get redirectSiteKey() {
        return this.getStringAttribute('redirect_site_key');
    }
    set redirectSiteKey(value) {
        this._redirectSiteKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get redirectSiteKeyInput() {
        return this._redirectSiteKey;
    }
}
exports.ComputeSecurityPolicyRecaptchaOptionsConfigOutputReference = ComputeSecurityPolicyRecaptchaOptionsConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRecaptchaOptionsConfigOutputReference[_j] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRecaptchaOptionsConfigOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        header_name: cdktf.stringToTerraform(struct.headerName),
        header_value: cdktf.stringToTerraform(struct.headerValue),
    };
}
exports.computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToTerraform = computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToTerraform;
function computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        header_name: {
            value: cdktf.stringToHclTerraform(struct.headerName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        header_value: {
            value: cdktf.stringToHclTerraform(struct.headerValue),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToHclTerraform = computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToHclTerraform;
class ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._headerName !== undefined) {
            hasAnyValues = true;
            internalValueResult.headerName = this._headerName;
        }
        if (this._headerValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.headerValue = this._headerValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._headerName = undefined;
            this._headerValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._headerName = value.headerName;
            this._headerValue = value.headerValue;
        }
    }
    get headerName() {
        return this.getStringAttribute('header_name');
    }
    set headerName(value) {
        this._headerName = value;
    }
    // Temporarily expose input value. Use with caution.
    get headerNameInput() {
        return this._headerName;
    }
    get headerValue() {
        return this.getStringAttribute('header_value');
    }
    set headerValue(value) {
        this._headerValue = value;
    }
    resetHeaderValue() {
        this._headerValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerValueInput() {
        return this._headerValue;
    }
}
exports.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsOutputReference = ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsOutputReference;
_k = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsOutputReference[_k] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsList = ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsList;
_l = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsList[_l] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsList", version: "14.13.1" };
function computeSecurityPolicyRuleHeaderActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        request_headers_to_adds: cdktf.listMapper(computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToTerraform, true)(struct.requestHeadersToAdds),
    };
}
exports.computeSecurityPolicyRuleHeaderActionToTerraform = computeSecurityPolicyRuleHeaderActionToTerraform;
function computeSecurityPolicyRuleHeaderActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        request_headers_to_adds: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRuleHeaderActionRequestHeadersToAddsToHclTerraform, true)(struct.requestHeadersToAdds),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleHeaderActionToHclTerraform = computeSecurityPolicyRuleHeaderActionToHclTerraform;
class ComputeSecurityPolicyRuleHeaderActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // request_headers_to_adds - computed: false, optional: false, required: true
        this._requestHeadersToAdds = new ComputeSecurityPolicyRuleHeaderActionRequestHeadersToAddsList(this, "request_headers_to_adds", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._requestHeadersToAdds?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestHeadersToAdds = this._requestHeadersToAdds?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._requestHeadersToAdds.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._requestHeadersToAdds.internalValue = value.requestHeadersToAdds;
        }
    }
    get requestHeadersToAdds() {
        return this._requestHeadersToAdds;
    }
    putRequestHeadersToAdds(value) {
        this._requestHeadersToAdds.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get requestHeadersToAddsInput() {
        return this._requestHeadersToAdds.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleHeaderActionOutputReference = ComputeSecurityPolicyRuleHeaderActionOutputReference;
_m = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleHeaderActionOutputReference[_m] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleHeaderActionOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        src_ip_ranges: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.srcIpRanges),
    };
}
exports.computeSecurityPolicyRuleMatchConfigToTerraform = computeSecurityPolicyRuleMatchConfigToTerraform;
function computeSecurityPolicyRuleMatchConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        src_ip_ranges: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.srcIpRanges),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchConfigToHclTerraform = computeSecurityPolicyRuleMatchConfigToHclTerraform;
class ComputeSecurityPolicyRuleMatchConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._srcIpRanges !== undefined) {
            hasAnyValues = true;
            internalValueResult.srcIpRanges = this._srcIpRanges;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._srcIpRanges = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._srcIpRanges = value.srcIpRanges;
        }
    }
    get srcIpRanges() {
        return cdktf.Fn.tolist(this.getListAttribute('src_ip_ranges'));
    }
    set srcIpRanges(value) {
        this._srcIpRanges = value;
    }
    // Temporarily expose input value. Use with caution.
    get srcIpRangesInput() {
        return this._srcIpRanges;
    }
}
exports.ComputeSecurityPolicyRuleMatchConfigOutputReference = ComputeSecurityPolicyRuleMatchConfigOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchConfigOutputReference[_o] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleMatchConfigOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchExprToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        expression: cdktf.stringToTerraform(struct.expression),
    };
}
exports.computeSecurityPolicyRuleMatchExprToTerraform = computeSecurityPolicyRuleMatchExprToTerraform;
function computeSecurityPolicyRuleMatchExprToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchExprToHclTerraform = computeSecurityPolicyRuleMatchExprToHclTerraform;
class ComputeSecurityPolicyRuleMatchExprOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._expression = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._expression = value.expression;
        }
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
}
exports.ComputeSecurityPolicyRuleMatchExprOutputReference = ComputeSecurityPolicyRuleMatchExprOutputReference;
_p = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchExprOutputReference[_p] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleMatchExprOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action_token_site_keys: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.actionTokenSiteKeys),
        session_token_site_keys: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.sessionTokenSiteKeys),
    };
}
exports.computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToTerraform = computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToTerraform;
function computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action_token_site_keys: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.actionTokenSiteKeys),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        session_token_site_keys: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.sessionTokenSiteKeys),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToHclTerraform = computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToHclTerraform;
class ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._actionTokenSiteKeys !== undefined) {
            hasAnyValues = true;
            internalValueResult.actionTokenSiteKeys = this._actionTokenSiteKeys;
        }
        if (this._sessionTokenSiteKeys !== undefined) {
            hasAnyValues = true;
            internalValueResult.sessionTokenSiteKeys = this._sessionTokenSiteKeys;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._actionTokenSiteKeys = undefined;
            this._sessionTokenSiteKeys = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._actionTokenSiteKeys = value.actionTokenSiteKeys;
            this._sessionTokenSiteKeys = value.sessionTokenSiteKeys;
        }
    }
    get actionTokenSiteKeys() {
        return this.getListAttribute('action_token_site_keys');
    }
    set actionTokenSiteKeys(value) {
        this._actionTokenSiteKeys = value;
    }
    resetActionTokenSiteKeys() {
        this._actionTokenSiteKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get actionTokenSiteKeysInput() {
        return this._actionTokenSiteKeys;
    }
    get sessionTokenSiteKeys() {
        return this.getListAttribute('session_token_site_keys');
    }
    set sessionTokenSiteKeys(value) {
        this._sessionTokenSiteKeys = value;
    }
    resetSessionTokenSiteKeys() {
        this._sessionTokenSiteKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sessionTokenSiteKeysInput() {
        return this._sessionTokenSiteKeys;
    }
}
exports.ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsOutputReference = ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsOutputReference[_q] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchExprOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        recaptcha_options: computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToTerraform(struct.recaptchaOptions),
    };
}
exports.computeSecurityPolicyRuleMatchExprOptionsToTerraform = computeSecurityPolicyRuleMatchExprOptionsToTerraform;
function computeSecurityPolicyRuleMatchExprOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        recaptcha_options: {
            value: computeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsToHclTerraform(struct.recaptchaOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchExprOptionsToHclTerraform = computeSecurityPolicyRuleMatchExprOptionsToHclTerraform;
class ComputeSecurityPolicyRuleMatchExprOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // recaptcha_options - computed: false, optional: false, required: true
        this._recaptchaOptions = new ComputeSecurityPolicyRuleMatchExprOptionsRecaptchaOptionsOutputReference(this, "recaptcha_options");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._recaptchaOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.recaptchaOptions = this._recaptchaOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._recaptchaOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._recaptchaOptions.internalValue = value.recaptchaOptions;
        }
    }
    get recaptchaOptions() {
        return this._recaptchaOptions;
    }
    putRecaptchaOptions(value) {
        this._recaptchaOptions.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get recaptchaOptionsInput() {
        return this._recaptchaOptions.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleMatchExprOptionsOutputReference = ComputeSecurityPolicyRuleMatchExprOptionsOutputReference;
_r = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchExprOptionsOutputReference[_r] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleMatchExprOptionsOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleMatchToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        versioned_expr: cdktf.stringToTerraform(struct.versionedExpr),
        config: computeSecurityPolicyRuleMatchConfigToTerraform(struct.config),
        expr: computeSecurityPolicyRuleMatchExprToTerraform(struct.expr),
        expr_options: computeSecurityPolicyRuleMatchExprOptionsToTerraform(struct.exprOptions),
    };
}
exports.computeSecurityPolicyRuleMatchToTerraform = computeSecurityPolicyRuleMatchToTerraform;
function computeSecurityPolicyRuleMatchToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        versioned_expr: {
            value: cdktf.stringToHclTerraform(struct.versionedExpr),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        config: {
            value: computeSecurityPolicyRuleMatchConfigToHclTerraform(struct.config),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchConfigList",
        },
        expr: {
            value: computeSecurityPolicyRuleMatchExprToHclTerraform(struct.expr),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchExprList",
        },
        expr_options: {
            value: computeSecurityPolicyRuleMatchExprOptionsToHclTerraform(struct.exprOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchExprOptionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleMatchToHclTerraform = computeSecurityPolicyRuleMatchToHclTerraform;
class ComputeSecurityPolicyRuleMatchOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // config - computed: false, optional: true, required: false
        this._config = new ComputeSecurityPolicyRuleMatchConfigOutputReference(this, "config");
        // expr - computed: false, optional: true, required: false
        this._expr = new ComputeSecurityPolicyRuleMatchExprOutputReference(this, "expr");
        // expr_options - computed: false, optional: true, required: false
        this._exprOptions = new ComputeSecurityPolicyRuleMatchExprOptionsOutputReference(this, "expr_options");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._versionedExpr !== undefined) {
            hasAnyValues = true;
            internalValueResult.versionedExpr = this._versionedExpr;
        }
        if (this._config?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.config = this._config?.internalValue;
        }
        if (this._expr?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.expr = this._expr?.internalValue;
        }
        if (this._exprOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.exprOptions = this._exprOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._versionedExpr = undefined;
            this._config.internalValue = undefined;
            this._expr.internalValue = undefined;
            this._exprOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._versionedExpr = value.versionedExpr;
            this._config.internalValue = value.config;
            this._expr.internalValue = value.expr;
            this._exprOptions.internalValue = value.exprOptions;
        }
    }
    get versionedExpr() {
        return this.getStringAttribute('versioned_expr');
    }
    set versionedExpr(value) {
        this._versionedExpr = value;
    }
    resetVersionedExpr() {
        this._versionedExpr = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionedExprInput() {
        return this._versionedExpr;
    }
    get config() {
        return this._config;
    }
    putConfig(value) {
        this._config.internalValue = value;
    }
    resetConfig() {
        this._config.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get configInput() {
        return this._config.internalValue;
    }
    get expr() {
        return this._expr;
    }
    putExpr(value) {
        this._expr.internalValue = value;
    }
    resetExpr() {
        this._expr.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exprInput() {
        return this._expr.internalValue;
    }
    get exprOptions() {
        return this._exprOptions;
    }
    putExprOptions(value) {
        this._exprOptions.internalValue = value;
    }
    resetExprOptions() {
        this._exprOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exprOptionsInput() {
        return this._exprOptions.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleMatchOutputReference = ComputeSecurityPolicyRuleMatchOutputReference;
_s = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleMatchOutputReference[_s] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleMatchOutputReference", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operator: cdktf.stringToTerraform(struct.operator),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._operator = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._operator = value.operator;
            this._value = value.value;
        }
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieOutputReference;
_t = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieOutputReference[_t] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieList;
_u = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieList[_u] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operator: cdktf.stringToTerraform(struct.operator),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._operator = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._operator = value.operator;
            this._value = value.value;
        }
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderOutputReference;
_v = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderOutputReference[_v] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderList;
_w = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderList[_w] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operator: cdktf.stringToTerraform(struct.operator),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._operator = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._operator = value.operator;
            this._value = value.value;
        }
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamOutputReference;
_x = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamOutputReference[_x] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamList;
_y = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamList[_y] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        operator: cdktf.stringToTerraform(struct.operator),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        operator: {
            value: cdktf.stringToHclTerraform(struct.operator),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        value: {
            value: cdktf.stringToHclTerraform(struct.value),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._operator !== undefined) {
            hasAnyValues = true;
            internalValueResult.operator = this._operator;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._operator = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._operator = value.operator;
            this._value = value.value;
        }
    }
    get operator() {
        return this.getStringAttribute('operator');
    }
    set operator(value) {
        this._operator = value;
    }
    // Temporarily expose input value. Use with caution.
    get operatorInput() {
        return this._operator;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    resetValue() {
        this._value = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriOutputReference;
_z = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriOutputReference[_z] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriList;
_0 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriList[_0] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target_rule_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.targetRuleIds),
        target_rule_set: cdktf.stringToTerraform(struct.targetRuleSet),
        request_cookie: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToTerraform, true)(struct.requestCookie),
        request_header: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToTerraform, true)(struct.requestHeader),
        request_query_param: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToTerraform, true)(struct.requestQueryParam),
        request_uri: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToTerraform, true)(struct.requestUri),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigExclusionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target_rule_ids: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.targetRuleIds),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        target_rule_set: {
            value: cdktf.stringToHclTerraform(struct.targetRuleSet),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        request_cookie: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieToHclTerraform, true)(struct.requestCookie),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieList",
        },
        request_header: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderToHclTerraform, true)(struct.requestHeader),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderList",
        },
        request_query_param: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamToHclTerraform, true)(struct.requestQueryParam),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamList",
        },
        request_uri: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriToHclTerraform, true)(struct.requestUri),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigExclusionToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigExclusionToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // request_cookie - computed: false, optional: true, required: false
        this._requestCookie = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestCookieList(this, "request_cookie", false);
        // request_header - computed: false, optional: true, required: false
        this._requestHeader = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestHeaderList(this, "request_header", false);
        // request_query_param - computed: false, optional: true, required: false
        this._requestQueryParam = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestQueryParamList(this, "request_query_param", false);
        // request_uri - computed: false, optional: true, required: false
        this._requestUri = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionRequestUriList(this, "request_uri", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._targetRuleIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetRuleIds = this._targetRuleIds;
        }
        if (this._targetRuleSet !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetRuleSet = this._targetRuleSet;
        }
        if (this._requestCookie?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestCookie = this._requestCookie?.internalValue;
        }
        if (this._requestHeader?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestHeader = this._requestHeader?.internalValue;
        }
        if (this._requestQueryParam?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestQueryParam = this._requestQueryParam?.internalValue;
        }
        if (this._requestUri?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.requestUri = this._requestUri?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._targetRuleIds = undefined;
            this._targetRuleSet = undefined;
            this._requestCookie.internalValue = undefined;
            this._requestHeader.internalValue = undefined;
            this._requestQueryParam.internalValue = undefined;
            this._requestUri.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._targetRuleIds = value.targetRuleIds;
            this._targetRuleSet = value.targetRuleSet;
            this._requestCookie.internalValue = value.requestCookie;
            this._requestHeader.internalValue = value.requestHeader;
            this._requestQueryParam.internalValue = value.requestQueryParam;
            this._requestUri.internalValue = value.requestUri;
        }
    }
    get targetRuleIds() {
        return cdktf.Fn.tolist(this.getListAttribute('target_rule_ids'));
    }
    set targetRuleIds(value) {
        this._targetRuleIds = value;
    }
    resetTargetRuleIds() {
        this._targetRuleIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetRuleIdsInput() {
        return this._targetRuleIds;
    }
    get targetRuleSet() {
        return this.getStringAttribute('target_rule_set');
    }
    set targetRuleSet(value) {
        this._targetRuleSet = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetRuleSetInput() {
        return this._targetRuleSet;
    }
    get requestCookie() {
        return this._requestCookie;
    }
    putRequestCookie(value) {
        this._requestCookie.internalValue = value;
    }
    resetRequestCookie() {
        this._requestCookie.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestCookieInput() {
        return this._requestCookie.internalValue;
    }
    get requestHeader() {
        return this._requestHeader;
    }
    putRequestHeader(value) {
        this._requestHeader.internalValue = value;
    }
    resetRequestHeader() {
        this._requestHeader.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestHeaderInput() {
        return this._requestHeader.internalValue;
    }
    get requestQueryParam() {
        return this._requestQueryParam;
    }
    putRequestQueryParam(value) {
        this._requestQueryParam.internalValue = value;
    }
    resetRequestQueryParam() {
        this._requestQueryParam.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestQueryParamInput() {
        return this._requestQueryParam.internalValue;
    }
    get requestUri() {
        return this._requestUri;
    }
    putRequestUri(value) {
        this._requestUri.internalValue = value;
    }
    resetRequestUri() {
        this._requestUri.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get requestUriInput() {
        return this._requestUri.internalValue;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionOutputReference;
_1 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionOutputReference[_1] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionList = ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionList;
_2 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionList[_2] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionList", version: "14.13.1" };
function computeSecurityPolicyRulePreconfiguredWafConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exclusion: cdktf.listMapper(computeSecurityPolicyRulePreconfiguredWafConfigExclusionToTerraform, true)(struct.exclusion),
    };
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigToTerraform = computeSecurityPolicyRulePreconfiguredWafConfigToTerraform;
function computeSecurityPolicyRulePreconfiguredWafConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        exclusion: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRulePreconfiguredWafConfigExclusionToHclTerraform, true)(struct.exclusion),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRulePreconfiguredWafConfigToHclTerraform = computeSecurityPolicyRulePreconfiguredWafConfigToHclTerraform;
class ComputeSecurityPolicyRulePreconfiguredWafConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // exclusion - computed: false, optional: true, required: false
        this._exclusion = new ComputeSecurityPolicyRulePreconfiguredWafConfigExclusionList(this, "exclusion", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._exclusion?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.exclusion = this._exclusion?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._exclusion.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._exclusion.internalValue = value.exclusion;
        }
    }
    get exclusion() {
        return this._exclusion;
    }
    putExclusion(value) {
        this._exclusion.internalValue = value;
    }
    resetExclusion() {
        this._exclusion.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exclusionInput() {
        return this._exclusion.internalValue;
    }
}
exports.ComputeSecurityPolicyRulePreconfiguredWafConfigOutputReference = ComputeSecurityPolicyRulePreconfiguredWafConfigOutputReference;
_3 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRulePreconfiguredWafConfigOutputReference[_3] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRulePreconfiguredWafConfigOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsBanThresholdToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        count: cdktf.numberToTerraform(struct.count),
        interval_sec: cdktf.numberToTerraform(struct.intervalSec),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsBanThresholdToTerraform = computeSecurityPolicyRuleRateLimitOptionsBanThresholdToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsBanThresholdToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        count: {
            value: cdktf.numberToHclTerraform(struct.count),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        interval_sec: {
            value: cdktf.numberToHclTerraform(struct.intervalSec),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsBanThresholdToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsBanThresholdToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._count !== undefined) {
            hasAnyValues = true;
            internalValueResult.count = this._count;
        }
        if (this._intervalSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.intervalSec = this._intervalSec;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._count = undefined;
            this._intervalSec = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._count = value.count;
            this._intervalSec = value.intervalSec;
        }
    }
    get count() {
        return this.getNumberAttribute('count');
    }
    set count(value) {
        this._count = value;
    }
    // Temporarily expose input value. Use with caution.
    get countInput() {
        return this._count;
    }
    get intervalSec() {
        return this.getNumberAttribute('interval_sec');
    }
    set intervalSec(value) {
        this._intervalSec = value;
    }
    // Temporarily expose input value. Use with caution.
    get intervalSecInput() {
        return this._intervalSec;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdOutputReference[_4] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enforce_on_key_name: cdktf.stringToTerraform(struct.enforceOnKeyName),
        enforce_on_key_type: cdktf.stringToTerraform(struct.enforceOnKeyType),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToTerraform = computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        enforce_on_key_name: {
            value: cdktf.stringToHclTerraform(struct.enforceOnKeyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enforce_on_key_type: {
            value: cdktf.stringToHclTerraform(struct.enforceOnKeyType),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enforceOnKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKeyName = this._enforceOnKeyName;
        }
        if (this._enforceOnKeyType !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKeyType = this._enforceOnKeyType;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._enforceOnKeyName = undefined;
            this._enforceOnKeyType = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._enforceOnKeyName = value.enforceOnKeyName;
            this._enforceOnKeyType = value.enforceOnKeyType;
        }
    }
    get enforceOnKeyName() {
        return this.getStringAttribute('enforce_on_key_name');
    }
    set enforceOnKeyName(value) {
        this._enforceOnKeyName = value;
    }
    resetEnforceOnKeyName() {
        this._enforceOnKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyNameInput() {
        return this._enforceOnKeyName;
    }
    get enforceOnKeyType() {
        return this.getStringAttribute('enforce_on_key_type');
    }
    set enforceOnKeyType(value) {
        this._enforceOnKeyType = value;
    }
    resetEnforceOnKeyType() {
        this._enforceOnKeyType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyTypeInput() {
        return this._enforceOnKeyType;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsOutputReference;
_5 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsOutputReference[_5] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsList = ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsList;
_6 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsList[_6] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsList", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target: cdktf.stringToTerraform(struct.target),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToTerraform = computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target: {
            value: cdktf.stringToHclTerraform(struct.target),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._target = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._target = value.target;
            this._type = value.type;
        }
    }
    get target() {
        return this.getStringAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsOutputReference;
_7 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsOutputReference[_7] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        count: cdktf.numberToTerraform(struct.count),
        interval_sec: cdktf.numberToTerraform(struct.intervalSec),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToTerraform = computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        count: {
            value: cdktf.numberToHclTerraform(struct.count),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        interval_sec: {
            value: cdktf.numberToHclTerraform(struct.intervalSec),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._count !== undefined) {
            hasAnyValues = true;
            internalValueResult.count = this._count;
        }
        if (this._intervalSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.intervalSec = this._intervalSec;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._count = undefined;
            this._intervalSec = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._count = value.count;
            this._intervalSec = value.intervalSec;
        }
    }
    get count() {
        return this.getNumberAttribute('count');
    }
    set count(value) {
        this._count = value;
    }
    // Temporarily expose input value. Use with caution.
    get countInput() {
        return this._count;
    }
    get intervalSec() {
        return this.getNumberAttribute('interval_sec');
    }
    set intervalSec(value) {
        this._intervalSec = value;
    }
    // Temporarily expose input value. Use with caution.
    get intervalSecInput() {
        return this._intervalSec;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdOutputReference[_8] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRateLimitOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ban_duration_sec: cdktf.numberToTerraform(struct.banDurationSec),
        conform_action: cdktf.stringToTerraform(struct.conformAction),
        enforce_on_key: cdktf.stringToTerraform(struct.enforceOnKey),
        enforce_on_key_name: cdktf.stringToTerraform(struct.enforceOnKeyName),
        exceed_action: cdktf.stringToTerraform(struct.exceedAction),
        ban_threshold: computeSecurityPolicyRuleRateLimitOptionsBanThresholdToTerraform(struct.banThreshold),
        enforce_on_key_configs: cdktf.listMapper(computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToTerraform, true)(struct.enforceOnKeyConfigs),
        exceed_redirect_options: computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToTerraform(struct.exceedRedirectOptions),
        rate_limit_threshold: computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToTerraform(struct.rateLimitThreshold),
    };
}
exports.computeSecurityPolicyRuleRateLimitOptionsToTerraform = computeSecurityPolicyRuleRateLimitOptionsToTerraform;
function computeSecurityPolicyRuleRateLimitOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        ban_duration_sec: {
            value: cdktf.numberToHclTerraform(struct.banDurationSec),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        conform_action: {
            value: cdktf.stringToHclTerraform(struct.conformAction),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enforce_on_key: {
            value: cdktf.stringToHclTerraform(struct.enforceOnKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        enforce_on_key_name: {
            value: cdktf.stringToHclTerraform(struct.enforceOnKeyName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        exceed_action: {
            value: cdktf.stringToHclTerraform(struct.exceedAction),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ban_threshold: {
            value: computeSecurityPolicyRuleRateLimitOptionsBanThresholdToHclTerraform(struct.banThreshold),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdList",
        },
        enforce_on_key_configs: {
            value: cdktf.listMapperHcl(computeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsToHclTerraform, true)(struct.enforceOnKeyConfigs),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsList",
        },
        exceed_redirect_options: {
            value: computeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsToHclTerraform(struct.exceedRedirectOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsList",
        },
        rate_limit_threshold: {
            value: computeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdToHclTerraform(struct.rateLimitThreshold),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRateLimitOptionsToHclTerraform = computeSecurityPolicyRuleRateLimitOptionsToHclTerraform;
class ComputeSecurityPolicyRuleRateLimitOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // ban_threshold - computed: false, optional: true, required: false
        this._banThreshold = new ComputeSecurityPolicyRuleRateLimitOptionsBanThresholdOutputReference(this, "ban_threshold");
        // enforce_on_key_configs - computed: false, optional: true, required: false
        this._enforceOnKeyConfigs = new ComputeSecurityPolicyRuleRateLimitOptionsEnforceOnKeyConfigsList(this, "enforce_on_key_configs", false);
        // exceed_redirect_options - computed: false, optional: true, required: false
        this._exceedRedirectOptions = new ComputeSecurityPolicyRuleRateLimitOptionsExceedRedirectOptionsOutputReference(this, "exceed_redirect_options");
        // rate_limit_threshold - computed: false, optional: false, required: true
        this._rateLimitThreshold = new ComputeSecurityPolicyRuleRateLimitOptionsRateLimitThresholdOutputReference(this, "rate_limit_threshold");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._banDurationSec !== undefined) {
            hasAnyValues = true;
            internalValueResult.banDurationSec = this._banDurationSec;
        }
        if (this._conformAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.conformAction = this._conformAction;
        }
        if (this._enforceOnKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKey = this._enforceOnKey;
        }
        if (this._enforceOnKeyName !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKeyName = this._enforceOnKeyName;
        }
        if (this._exceedAction !== undefined) {
            hasAnyValues = true;
            internalValueResult.exceedAction = this._exceedAction;
        }
        if (this._banThreshold?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.banThreshold = this._banThreshold?.internalValue;
        }
        if (this._enforceOnKeyConfigs?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.enforceOnKeyConfigs = this._enforceOnKeyConfigs?.internalValue;
        }
        if (this._exceedRedirectOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.exceedRedirectOptions = this._exceedRedirectOptions?.internalValue;
        }
        if (this._rateLimitThreshold?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rateLimitThreshold = this._rateLimitThreshold?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._banDurationSec = undefined;
            this._conformAction = undefined;
            this._enforceOnKey = undefined;
            this._enforceOnKeyName = undefined;
            this._exceedAction = undefined;
            this._banThreshold.internalValue = undefined;
            this._enforceOnKeyConfigs.internalValue = undefined;
            this._exceedRedirectOptions.internalValue = undefined;
            this._rateLimitThreshold.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._banDurationSec = value.banDurationSec;
            this._conformAction = value.conformAction;
            this._enforceOnKey = value.enforceOnKey;
            this._enforceOnKeyName = value.enforceOnKeyName;
            this._exceedAction = value.exceedAction;
            this._banThreshold.internalValue = value.banThreshold;
            this._enforceOnKeyConfigs.internalValue = value.enforceOnKeyConfigs;
            this._exceedRedirectOptions.internalValue = value.exceedRedirectOptions;
            this._rateLimitThreshold.internalValue = value.rateLimitThreshold;
        }
    }
    get banDurationSec() {
        return this.getNumberAttribute('ban_duration_sec');
    }
    set banDurationSec(value) {
        this._banDurationSec = value;
    }
    resetBanDurationSec() {
        this._banDurationSec = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get banDurationSecInput() {
        return this._banDurationSec;
    }
    get conformAction() {
        return this.getStringAttribute('conform_action');
    }
    set conformAction(value) {
        this._conformAction = value;
    }
    // Temporarily expose input value. Use with caution.
    get conformActionInput() {
        return this._conformAction;
    }
    get enforceOnKey() {
        return this.getStringAttribute('enforce_on_key');
    }
    set enforceOnKey(value) {
        this._enforceOnKey = value;
    }
    resetEnforceOnKey() {
        this._enforceOnKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyInput() {
        return this._enforceOnKey;
    }
    get enforceOnKeyName() {
        return this.getStringAttribute('enforce_on_key_name');
    }
    set enforceOnKeyName(value) {
        this._enforceOnKeyName = value;
    }
    resetEnforceOnKeyName() {
        this._enforceOnKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyNameInput() {
        return this._enforceOnKeyName;
    }
    get exceedAction() {
        return this.getStringAttribute('exceed_action');
    }
    set exceedAction(value) {
        this._exceedAction = value;
    }
    // Temporarily expose input value. Use with caution.
    get exceedActionInput() {
        return this._exceedAction;
    }
    get banThreshold() {
        return this._banThreshold;
    }
    putBanThreshold(value) {
        this._banThreshold.internalValue = value;
    }
    resetBanThreshold() {
        this._banThreshold.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get banThresholdInput() {
        return this._banThreshold.internalValue;
    }
    get enforceOnKeyConfigs() {
        return this._enforceOnKeyConfigs;
    }
    putEnforceOnKeyConfigs(value) {
        this._enforceOnKeyConfigs.internalValue = value;
    }
    resetEnforceOnKeyConfigs() {
        this._enforceOnKeyConfigs.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enforceOnKeyConfigsInput() {
        return this._enforceOnKeyConfigs.internalValue;
    }
    get exceedRedirectOptions() {
        return this._exceedRedirectOptions;
    }
    putExceedRedirectOptions(value) {
        this._exceedRedirectOptions.internalValue = value;
    }
    resetExceedRedirectOptions() {
        this._exceedRedirectOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exceedRedirectOptionsInput() {
        return this._exceedRedirectOptions.internalValue;
    }
    get rateLimitThreshold() {
        return this._rateLimitThreshold;
    }
    putRateLimitThreshold(value) {
        this._rateLimitThreshold.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get rateLimitThresholdInput() {
        return this._rateLimitThreshold.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleRateLimitOptionsOutputReference = ComputeSecurityPolicyRuleRateLimitOptionsOutputReference;
_9 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRateLimitOptionsOutputReference[_9] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleRateLimitOptionsOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleRedirectOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target: cdktf.stringToTerraform(struct.target),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.computeSecurityPolicyRuleRedirectOptionsToTerraform = computeSecurityPolicyRuleRedirectOptionsToTerraform;
function computeSecurityPolicyRuleRedirectOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        target: {
            value: cdktf.stringToHclTerraform(struct.target),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleRedirectOptionsToHclTerraform = computeSecurityPolicyRuleRedirectOptionsToHclTerraform;
class ComputeSecurityPolicyRuleRedirectOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._target !== undefined) {
            hasAnyValues = true;
            internalValueResult.target = this._target;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._target = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._target = value.target;
            this._type = value.type;
        }
    }
    get target() {
        return this.getStringAttribute('target');
    }
    set target(value) {
        this._target = value;
    }
    resetTarget() {
        this._target = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetInput() {
        return this._target;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComputeSecurityPolicyRuleRedirectOptionsOutputReference = ComputeSecurityPolicyRuleRedirectOptionsOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleRedirectOptionsOutputReference[_10] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleRedirectOptionsOutputReference", version: "14.13.1" };
function computeSecurityPolicyRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: cdktf.stringToTerraform(struct.action),
        description: cdktf.stringToTerraform(struct.description),
        preview: cdktf.booleanToTerraform(struct.preview),
        priority: cdktf.numberToTerraform(struct.priority),
        header_action: computeSecurityPolicyRuleHeaderActionToTerraform(struct.headerAction),
        match: computeSecurityPolicyRuleMatchToTerraform(struct.match),
        preconfigured_waf_config: computeSecurityPolicyRulePreconfiguredWafConfigToTerraform(struct.preconfiguredWafConfig),
        rate_limit_options: computeSecurityPolicyRuleRateLimitOptionsToTerraform(struct.rateLimitOptions),
        redirect_options: computeSecurityPolicyRuleRedirectOptionsToTerraform(struct.redirectOptions),
    };
}
exports.computeSecurityPolicyRuleToTerraform = computeSecurityPolicyRuleToTerraform;
function computeSecurityPolicyRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action: {
            value: cdktf.stringToHclTerraform(struct.action),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        preview: {
            value: cdktf.booleanToHclTerraform(struct.preview),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        priority: {
            value: cdktf.numberToHclTerraform(struct.priority),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        header_action: {
            value: computeSecurityPolicyRuleHeaderActionToHclTerraform(struct.headerAction),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleHeaderActionList",
        },
        match: {
            value: computeSecurityPolicyRuleMatchToHclTerraform(struct.match),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleMatchList",
        },
        preconfigured_waf_config: {
            value: computeSecurityPolicyRulePreconfiguredWafConfigToHclTerraform(struct.preconfiguredWafConfig),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRulePreconfiguredWafConfigList",
        },
        rate_limit_options: {
            value: computeSecurityPolicyRuleRateLimitOptionsToHclTerraform(struct.rateLimitOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRateLimitOptionsList",
        },
        redirect_options: {
            value: computeSecurityPolicyRuleRedirectOptionsToHclTerraform(struct.redirectOptions),
            isBlock: true,
            type: "list",
            storageClassType: "ComputeSecurityPolicyRuleRedirectOptionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyRuleToHclTerraform = computeSecurityPolicyRuleToHclTerraform;
class ComputeSecurityPolicyRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // header_action - computed: false, optional: true, required: false
        this._headerAction = new ComputeSecurityPolicyRuleHeaderActionOutputReference(this, "header_action");
        // match - computed: false, optional: false, required: true
        this._match = new ComputeSecurityPolicyRuleMatchOutputReference(this, "match");
        // preconfigured_waf_config - computed: false, optional: true, required: false
        this._preconfiguredWafConfig = new ComputeSecurityPolicyRulePreconfiguredWafConfigOutputReference(this, "preconfigured_waf_config");
        // rate_limit_options - computed: false, optional: true, required: false
        this._rateLimitOptions = new ComputeSecurityPolicyRuleRateLimitOptionsOutputReference(this, "rate_limit_options");
        // redirect_options - computed: false, optional: true, required: false
        this._redirectOptions = new ComputeSecurityPolicyRuleRedirectOptionsOutputReference(this, "redirect_options");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._preview !== undefined) {
            hasAnyValues = true;
            internalValueResult.preview = this._preview;
        }
        if (this._priority !== undefined) {
            hasAnyValues = true;
            internalValueResult.priority = this._priority;
        }
        if (this._headerAction?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.headerAction = this._headerAction?.internalValue;
        }
        if (this._match?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.match = this._match?.internalValue;
        }
        if (this._preconfiguredWafConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.preconfiguredWafConfig = this._preconfiguredWafConfig?.internalValue;
        }
        if (this._rateLimitOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.rateLimitOptions = this._rateLimitOptions?.internalValue;
        }
        if (this._redirectOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.redirectOptions = this._redirectOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._action = undefined;
            this._description = undefined;
            this._preview = undefined;
            this._priority = undefined;
            this._headerAction.internalValue = undefined;
            this._match.internalValue = undefined;
            this._preconfiguredWafConfig.internalValue = undefined;
            this._rateLimitOptions.internalValue = undefined;
            this._redirectOptions.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._action = value.action;
            this._description = value.description;
            this._preview = value.preview;
            this._priority = value.priority;
            this._headerAction.internalValue = value.headerAction;
            this._match.internalValue = value.match;
            this._preconfiguredWafConfig.internalValue = value.preconfiguredWafConfig;
            this._rateLimitOptions.internalValue = value.rateLimitOptions;
            this._redirectOptions.internalValue = value.redirectOptions;
        }
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get preview() {
        return this.getBooleanAttribute('preview');
    }
    set preview(value) {
        this._preview = value;
    }
    resetPreview() {
        this._preview = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get previewInput() {
        return this._preview;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get headerAction() {
        return this._headerAction;
    }
    putHeaderAction(value) {
        this._headerAction.internalValue = value;
    }
    resetHeaderAction() {
        this._headerAction.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get headerActionInput() {
        return this._headerAction.internalValue;
    }
    get match() {
        return this._match;
    }
    putMatch(value) {
        this._match.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get matchInput() {
        return this._match.internalValue;
    }
    get preconfiguredWafConfig() {
        return this._preconfiguredWafConfig;
    }
    putPreconfiguredWafConfig(value) {
        this._preconfiguredWafConfig.internalValue = value;
    }
    resetPreconfiguredWafConfig() {
        this._preconfiguredWafConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preconfiguredWafConfigInput() {
        return this._preconfiguredWafConfig.internalValue;
    }
    get rateLimitOptions() {
        return this._rateLimitOptions;
    }
    putRateLimitOptions(value) {
        this._rateLimitOptions.internalValue = value;
    }
    resetRateLimitOptions() {
        this._rateLimitOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rateLimitOptionsInput() {
        return this._rateLimitOptions.internalValue;
    }
    get redirectOptions() {
        return this._redirectOptions;
    }
    putRedirectOptions(value) {
        this._redirectOptions.internalValue = value;
    }
    resetRedirectOptions() {
        this._redirectOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get redirectOptionsInput() {
        return this._redirectOptions.internalValue;
    }
}
exports.ComputeSecurityPolicyRuleOutputReference = ComputeSecurityPolicyRuleOutputReference;
_11 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleOutputReference[_11] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleOutputReference", version: "14.13.1" };
class ComputeSecurityPolicyRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeSecurityPolicyRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeSecurityPolicyRuleList = ComputeSecurityPolicyRuleList;
_12 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyRuleList[_12] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyRuleList", version: "14.13.1" };
function computeSecurityPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.computeSecurityPolicyTimeoutsToTerraform = computeSecurityPolicyTimeoutsToTerraform;
function computeSecurityPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.computeSecurityPolicyTimeoutsToHclTerraform = computeSecurityPolicyTimeoutsToHclTerraform;
class ComputeSecurityPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ComputeSecurityPolicyTimeoutsOutputReference = ComputeSecurityPolicyTimeoutsOutputReference;
_13 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicyTimeoutsOutputReference[_13] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_security_policy google_compute_security_policy}
*/
class ComputeSecurityPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a ComputeSecurityPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the ComputeSecurityPolicy to import
    * @param importFromId The id of the existing ComputeSecurityPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_security_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the ComputeSecurityPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_compute_security_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/compute_security_policy google_compute_security_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeSecurityPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_compute_security_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // adaptive_protection_config - computed: false, optional: true, required: false
        this._adaptiveProtectionConfig = new ComputeSecurityPolicyAdaptiveProtectionConfigOutputReference(this, "adaptive_protection_config");
        // advanced_options_config - computed: false, optional: true, required: false
        this._advancedOptionsConfig = new ComputeSecurityPolicyAdvancedOptionsConfigOutputReference(this, "advanced_options_config");
        // recaptcha_options_config - computed: false, optional: true, required: false
        this._recaptchaOptionsConfig = new ComputeSecurityPolicyRecaptchaOptionsConfigOutputReference(this, "recaptcha_options_config");
        // rule - computed: false, optional: true, required: false
        this._rule = new ComputeSecurityPolicyRuleList(this, "rule", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeSecurityPolicyTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._type = config.type;
        this._adaptiveProtectionConfig.internalValue = config.adaptiveProtectionConfig;
        this._advancedOptionsConfig.internalValue = config.advancedOptionsConfig;
        this._recaptchaOptionsConfig.internalValue = config.recaptchaOptionsConfig;
        this._rule.internalValue = config.rule;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // fingerprint - computed: true, optional: false, required: false
    get fingerprint() {
        return this.getStringAttribute('fingerprint');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get adaptiveProtectionConfig() {
        return this._adaptiveProtectionConfig;
    }
    putAdaptiveProtectionConfig(value) {
        this._adaptiveProtectionConfig.internalValue = value;
    }
    resetAdaptiveProtectionConfig() {
        this._adaptiveProtectionConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get adaptiveProtectionConfigInput() {
        return this._adaptiveProtectionConfig.internalValue;
    }
    get advancedOptionsConfig() {
        return this._advancedOptionsConfig;
    }
    putAdvancedOptionsConfig(value) {
        this._advancedOptionsConfig.internalValue = value;
    }
    resetAdvancedOptionsConfig() {
        this._advancedOptionsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get advancedOptionsConfigInput() {
        return this._advancedOptionsConfig.internalValue;
    }
    get recaptchaOptionsConfig() {
        return this._recaptchaOptionsConfig;
    }
    putRecaptchaOptionsConfig(value) {
        this._recaptchaOptionsConfig.internalValue = value;
    }
    resetRecaptchaOptionsConfig() {
        this._recaptchaOptionsConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recaptchaOptionsConfigInput() {
        return this._recaptchaOptionsConfig.internalValue;
    }
    get rule() {
        return this._rule;
    }
    putRule(value) {
        this._rule.internalValue = value;
    }
    resetRule() {
        this._rule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ruleInput() {
        return this._rule.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            type: cdktf.stringToTerraform(this._type),
            adaptive_protection_config: computeSecurityPolicyAdaptiveProtectionConfigToTerraform(this._adaptiveProtectionConfig.internalValue),
            advanced_options_config: computeSecurityPolicyAdvancedOptionsConfigToTerraform(this._advancedOptionsConfig.internalValue),
            recaptcha_options_config: computeSecurityPolicyRecaptchaOptionsConfigToTerraform(this._recaptchaOptionsConfig.internalValue),
            rule: cdktf.listMapper(computeSecurityPolicyRuleToTerraform, true)(this._rule.internalValue),
            timeouts: computeSecurityPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            adaptive_protection_config: {
                value: computeSecurityPolicyAdaptiveProtectionConfigToHclTerraform(this._adaptiveProtectionConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeSecurityPolicyAdaptiveProtectionConfigList",
            },
            advanced_options_config: {
                value: computeSecurityPolicyAdvancedOptionsConfigToHclTerraform(this._advancedOptionsConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeSecurityPolicyAdvancedOptionsConfigList",
            },
            recaptcha_options_config: {
                value: computeSecurityPolicyRecaptchaOptionsConfigToHclTerraform(this._recaptchaOptionsConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "ComputeSecurityPolicyRecaptchaOptionsConfigList",
            },
            rule: {
                value: cdktf.listMapperHcl(computeSecurityPolicyRuleToHclTerraform, true)(this._rule.internalValue),
                isBlock: true,
                type: "set",
                storageClassType: "ComputeSecurityPolicyRuleList",
            },
            timeouts: {
                value: computeSecurityPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "ComputeSecurityPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.ComputeSecurityPolicy = ComputeSecurityPolicy;
_14 = JSII_RTTI_SYMBOL_1;
ComputeSecurityPolicy[_14] = { fqn: "@cdktf/provider-google.computeSecurityPolicy.ComputeSecurityPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
ComputeSecurityPolicy.tfResourceType = "google_compute_security_policy";
