// @cdktf/provider-google/lib/storage-transfer-job/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.storageTransferJobTransferSpecPosixDataSourceToHclTerraform = exports.storageTransferJobTransferSpecPosixDataSourceToTerraform = exports.StorageTransferJobTransferSpecPosixDataSinkOutputReference = exports.storageTransferJobTransferSpecPosixDataSinkToHclTerraform = exports.storageTransferJobTransferSpecPosixDataSinkToTerraform = exports.StorageTransferJobTransferSpecObjectConditionsOutputReference = exports.storageTransferJobTransferSpecObjectConditionsToHclTerraform = exports.storageTransferJobTransferSpecObjectConditionsToTerraform = exports.StorageTransferJobTransferSpecHttpDataSourceOutputReference = exports.storageTransferJobTransferSpecHttpDataSourceToHclTerraform = exports.storageTransferJobTransferSpecHttpDataSourceToTerraform = exports.StorageTransferJobTransferSpecHdfsDataSourceOutputReference = exports.storageTransferJobTransferSpecHdfsDataSourceToHclTerraform = exports.storageTransferJobTransferSpecHdfsDataSourceToTerraform = exports.StorageTransferJobTransferSpecGcsDataSourceOutputReference = exports.storageTransferJobTransferSpecGcsDataSourceToHclTerraform = exports.storageTransferJobTransferSpecGcsDataSourceToTerraform = exports.StorageTransferJobTransferSpecGcsDataSinkOutputReference = exports.storageTransferJobTransferSpecGcsDataSinkToHclTerraform = exports.storageTransferJobTransferSpecGcsDataSinkToTerraform = exports.StorageTransferJobTransferSpecAzureBlobStorageDataSourceOutputReference = exports.storageTransferJobTransferSpecAzureBlobStorageDataSourceToHclTerraform = exports.storageTransferJobTransferSpecAzureBlobStorageDataSourceToTerraform = exports.StorageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsOutputReference = exports.storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToHclTerraform = exports.storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToTerraform = exports.StorageTransferJobTransferSpecAwsS3DataSourceOutputReference = exports.storageTransferJobTransferSpecAwsS3DataSourceToHclTerraform = exports.storageTransferJobTransferSpecAwsS3DataSourceToTerraform = exports.StorageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyOutputReference = exports.storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToHclTerraform = exports.storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToTerraform = exports.StorageTransferJobScheduleOutputReference = exports.storageTransferJobScheduleToHclTerraform = exports.storageTransferJobScheduleToTerraform = exports.StorageTransferJobScheduleStartTimeOfDayOutputReference = exports.storageTransferJobScheduleStartTimeOfDayToHclTerraform = exports.storageTransferJobScheduleStartTimeOfDayToTerraform = exports.StorageTransferJobScheduleScheduleStartDateOutputReference = exports.storageTransferJobScheduleScheduleStartDateToHclTerraform = exports.storageTransferJobScheduleScheduleStartDateToTerraform = exports.StorageTransferJobScheduleScheduleEndDateOutputReference = exports.storageTransferJobScheduleScheduleEndDateToHclTerraform = exports.storageTransferJobScheduleScheduleEndDateToTerraform = exports.StorageTransferJobNotificationConfigOutputReference = exports.storageTransferJobNotificationConfigToHclTerraform = exports.storageTransferJobNotificationConfigToTerraform = exports.StorageTransferJobEventStreamOutputReference = exports.storageTransferJobEventStreamToHclTerraform = exports.storageTransferJobEventStreamToTerraform = void 0;
exports.StorageTransferJob = exports.StorageTransferJobTransferSpecOutputReference = exports.storageTransferJobTransferSpecToHclTerraform = exports.storageTransferJobTransferSpecToTerraform = exports.StorageTransferJobTransferSpecTransferOptionsOutputReference = exports.storageTransferJobTransferSpecTransferOptionsToHclTerraform = exports.storageTransferJobTransferSpecTransferOptionsToTerraform = exports.StorageTransferJobTransferSpecPosixDataSourceOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function storageTransferJobEventStreamToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event_stream_expiration_time: cdktf.stringToTerraform(struct.eventStreamExpirationTime),
        event_stream_start_time: cdktf.stringToTerraform(struct.eventStreamStartTime),
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.storageTransferJobEventStreamToTerraform = storageTransferJobEventStreamToTerraform;
function storageTransferJobEventStreamToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        event_stream_expiration_time: {
            value: cdktf.stringToHclTerraform(struct.eventStreamExpirationTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        event_stream_start_time: {
            value: cdktf.stringToHclTerraform(struct.eventStreamStartTime),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        name: {
            value: cdktf.stringToHclTerraform(struct.name),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobEventStreamToHclTerraform = storageTransferJobEventStreamToHclTerraform;
class StorageTransferJobEventStreamOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._eventStreamExpirationTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.eventStreamExpirationTime = this._eventStreamExpirationTime;
        }
        if (this._eventStreamStartTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.eventStreamStartTime = this._eventStreamStartTime;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._eventStreamExpirationTime = undefined;
            this._eventStreamStartTime = undefined;
            this._name = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._eventStreamExpirationTime = value.eventStreamExpirationTime;
            this._eventStreamStartTime = value.eventStreamStartTime;
            this._name = value.name;
        }
    }
    get eventStreamExpirationTime() {
        return this.getStringAttribute('event_stream_expiration_time');
    }
    set eventStreamExpirationTime(value) {
        this._eventStreamExpirationTime = value;
    }
    resetEventStreamExpirationTime() {
        this._eventStreamExpirationTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventStreamExpirationTimeInput() {
        return this._eventStreamExpirationTime;
    }
    get eventStreamStartTime() {
        return this.getStringAttribute('event_stream_start_time');
    }
    set eventStreamStartTime(value) {
        this._eventStreamStartTime = value;
    }
    resetEventStreamStartTime() {
        this._eventStreamStartTime = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventStreamStartTimeInput() {
        return this._eventStreamStartTime;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.StorageTransferJobEventStreamOutputReference = StorageTransferJobEventStreamOutputReference;
_a = JSII_RTTI_SYMBOL_1;
StorageTransferJobEventStreamOutputReference[_a] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobEventStreamOutputReference", version: "14.13.1" };
function storageTransferJobNotificationConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        event_types: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.eventTypes),
        payload_format: cdktf.stringToTerraform(struct.payloadFormat),
        pubsub_topic: cdktf.stringToTerraform(struct.pubsubTopic),
    };
}
exports.storageTransferJobNotificationConfigToTerraform = storageTransferJobNotificationConfigToTerraform;
function storageTransferJobNotificationConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        event_types: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.eventTypes),
            isBlock: false,
            type: "set",
            storageClassType: "stringList",
        },
        payload_format: {
            value: cdktf.stringToHclTerraform(struct.payloadFormat),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        pubsub_topic: {
            value: cdktf.stringToHclTerraform(struct.pubsubTopic),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobNotificationConfigToHclTerraform = storageTransferJobNotificationConfigToHclTerraform;
class StorageTransferJobNotificationConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._eventTypes !== undefined) {
            hasAnyValues = true;
            internalValueResult.eventTypes = this._eventTypes;
        }
        if (this._payloadFormat !== undefined) {
            hasAnyValues = true;
            internalValueResult.payloadFormat = this._payloadFormat;
        }
        if (this._pubsubTopic !== undefined) {
            hasAnyValues = true;
            internalValueResult.pubsubTopic = this._pubsubTopic;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._eventTypes = undefined;
            this._payloadFormat = undefined;
            this._pubsubTopic = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._eventTypes = value.eventTypes;
            this._payloadFormat = value.payloadFormat;
            this._pubsubTopic = value.pubsubTopic;
        }
    }
    get eventTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('event_types'));
    }
    set eventTypes(value) {
        this._eventTypes = value;
    }
    resetEventTypes() {
        this._eventTypes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventTypesInput() {
        return this._eventTypes;
    }
    get payloadFormat() {
        return this.getStringAttribute('payload_format');
    }
    set payloadFormat(value) {
        this._payloadFormat = value;
    }
    // Temporarily expose input value. Use with caution.
    get payloadFormatInput() {
        return this._payloadFormat;
    }
    get pubsubTopic() {
        return this.getStringAttribute('pubsub_topic');
    }
    set pubsubTopic(value) {
        this._pubsubTopic = value;
    }
    // Temporarily expose input value. Use with caution.
    get pubsubTopicInput() {
        return this._pubsubTopic;
    }
}
exports.StorageTransferJobNotificationConfigOutputReference = StorageTransferJobNotificationConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
StorageTransferJobNotificationConfigOutputReference[_b] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobNotificationConfigOutputReference", version: "14.13.1" };
function storageTransferJobScheduleScheduleEndDateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.numberToTerraform(struct.day),
        month: cdktf.numberToTerraform(struct.month),
        year: cdktf.numberToTerraform(struct.year),
    };
}
exports.storageTransferJobScheduleScheduleEndDateToTerraform = storageTransferJobScheduleScheduleEndDateToTerraform;
function storageTransferJobScheduleScheduleEndDateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        day: {
            value: cdktf.numberToHclTerraform(struct.day),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        month: {
            value: cdktf.numberToHclTerraform(struct.month),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        year: {
            value: cdktf.numberToHclTerraform(struct.year),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobScheduleScheduleEndDateToHclTerraform = storageTransferJobScheduleScheduleEndDateToHclTerraform;
class StorageTransferJobScheduleScheduleEndDateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._month !== undefined) {
            hasAnyValues = true;
            internalValueResult.month = this._month;
        }
        if (this._year !== undefined) {
            hasAnyValues = true;
            internalValueResult.year = this._year;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._month = undefined;
            this._year = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._month = value.month;
            this._year = value.year;
        }
    }
    get day() {
        return this.getNumberAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get month() {
        return this.getNumberAttribute('month');
    }
    set month(value) {
        this._month = value;
    }
    // Temporarily expose input value. Use with caution.
    get monthInput() {
        return this._month;
    }
    get year() {
        return this.getNumberAttribute('year');
    }
    set year(value) {
        this._year = value;
    }
    // Temporarily expose input value. Use with caution.
    get yearInput() {
        return this._year;
    }
}
exports.StorageTransferJobScheduleScheduleEndDateOutputReference = StorageTransferJobScheduleScheduleEndDateOutputReference;
_c = JSII_RTTI_SYMBOL_1;
StorageTransferJobScheduleScheduleEndDateOutputReference[_c] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobScheduleScheduleEndDateOutputReference", version: "14.13.1" };
function storageTransferJobScheduleScheduleStartDateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.numberToTerraform(struct.day),
        month: cdktf.numberToTerraform(struct.month),
        year: cdktf.numberToTerraform(struct.year),
    };
}
exports.storageTransferJobScheduleScheduleStartDateToTerraform = storageTransferJobScheduleScheduleStartDateToTerraform;
function storageTransferJobScheduleScheduleStartDateToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        day: {
            value: cdktf.numberToHclTerraform(struct.day),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        month: {
            value: cdktf.numberToHclTerraform(struct.month),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        year: {
            value: cdktf.numberToHclTerraform(struct.year),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobScheduleScheduleStartDateToHclTerraform = storageTransferJobScheduleScheduleStartDateToHclTerraform;
class StorageTransferJobScheduleScheduleStartDateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._month !== undefined) {
            hasAnyValues = true;
            internalValueResult.month = this._month;
        }
        if (this._year !== undefined) {
            hasAnyValues = true;
            internalValueResult.year = this._year;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._month = undefined;
            this._year = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._month = value.month;
            this._year = value.year;
        }
    }
    get day() {
        return this.getNumberAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get month() {
        return this.getNumberAttribute('month');
    }
    set month(value) {
        this._month = value;
    }
    // Temporarily expose input value. Use with caution.
    get monthInput() {
        return this._month;
    }
    get year() {
        return this.getNumberAttribute('year');
    }
    set year(value) {
        this._year = value;
    }
    // Temporarily expose input value. Use with caution.
    get yearInput() {
        return this._year;
    }
}
exports.StorageTransferJobScheduleScheduleStartDateOutputReference = StorageTransferJobScheduleScheduleStartDateOutputReference;
_d = JSII_RTTI_SYMBOL_1;
StorageTransferJobScheduleScheduleStartDateOutputReference[_d] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobScheduleScheduleStartDateOutputReference", version: "14.13.1" };
function storageTransferJobScheduleStartTimeOfDayToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hours: cdktf.numberToTerraform(struct.hours),
        minutes: cdktf.numberToTerraform(struct.minutes),
        nanos: cdktf.numberToTerraform(struct.nanos),
        seconds: cdktf.numberToTerraform(struct.seconds),
    };
}
exports.storageTransferJobScheduleStartTimeOfDayToTerraform = storageTransferJobScheduleStartTimeOfDayToTerraform;
function storageTransferJobScheduleStartTimeOfDayToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        hours: {
            value: cdktf.numberToHclTerraform(struct.hours),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        minutes: {
            value: cdktf.numberToHclTerraform(struct.minutes),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        nanos: {
            value: cdktf.numberToHclTerraform(struct.nanos),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        seconds: {
            value: cdktf.numberToHclTerraform(struct.seconds),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobScheduleStartTimeOfDayToHclTerraform = storageTransferJobScheduleStartTimeOfDayToHclTerraform;
class StorageTransferJobScheduleStartTimeOfDayOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hours !== undefined) {
            hasAnyValues = true;
            internalValueResult.hours = this._hours;
        }
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        if (this._nanos !== undefined) {
            hasAnyValues = true;
            internalValueResult.nanos = this._nanos;
        }
        if (this._seconds !== undefined) {
            hasAnyValues = true;
            internalValueResult.seconds = this._seconds;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hours = undefined;
            this._minutes = undefined;
            this._nanos = undefined;
            this._seconds = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hours = value.hours;
            this._minutes = value.minutes;
            this._nanos = value.nanos;
            this._seconds = value.seconds;
        }
    }
    get hours() {
        return this.getNumberAttribute('hours');
    }
    set hours(value) {
        this._hours = value;
    }
    // Temporarily expose input value. Use with caution.
    get hoursInput() {
        return this._hours;
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    set nanos(value) {
        this._nanos = value;
    }
    // Temporarily expose input value. Use with caution.
    get nanosInput() {
        return this._nanos;
    }
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
    set seconds(value) {
        this._seconds = value;
    }
    // Temporarily expose input value. Use with caution.
    get secondsInput() {
        return this._seconds;
    }
}
exports.StorageTransferJobScheduleStartTimeOfDayOutputReference = StorageTransferJobScheduleStartTimeOfDayOutputReference;
_e = JSII_RTTI_SYMBOL_1;
StorageTransferJobScheduleStartTimeOfDayOutputReference[_e] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobScheduleStartTimeOfDayOutputReference", version: "14.13.1" };
function storageTransferJobScheduleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        repeat_interval: cdktf.stringToTerraform(struct.repeatInterval),
        schedule_end_date: storageTransferJobScheduleScheduleEndDateToTerraform(struct.scheduleEndDate),
        schedule_start_date: storageTransferJobScheduleScheduleStartDateToTerraform(struct.scheduleStartDate),
        start_time_of_day: storageTransferJobScheduleStartTimeOfDayToTerraform(struct.startTimeOfDay),
    };
}
exports.storageTransferJobScheduleToTerraform = storageTransferJobScheduleToTerraform;
function storageTransferJobScheduleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        repeat_interval: {
            value: cdktf.stringToHclTerraform(struct.repeatInterval),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        schedule_end_date: {
            value: storageTransferJobScheduleScheduleEndDateToHclTerraform(struct.scheduleEndDate),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobScheduleScheduleEndDateList",
        },
        schedule_start_date: {
            value: storageTransferJobScheduleScheduleStartDateToHclTerraform(struct.scheduleStartDate),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobScheduleScheduleStartDateList",
        },
        start_time_of_day: {
            value: storageTransferJobScheduleStartTimeOfDayToHclTerraform(struct.startTimeOfDay),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobScheduleStartTimeOfDayList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobScheduleToHclTerraform = storageTransferJobScheduleToHclTerraform;
class StorageTransferJobScheduleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // schedule_end_date - computed: false, optional: true, required: false
        this._scheduleEndDate = new StorageTransferJobScheduleScheduleEndDateOutputReference(this, "schedule_end_date");
        // schedule_start_date - computed: false, optional: false, required: true
        this._scheduleStartDate = new StorageTransferJobScheduleScheduleStartDateOutputReference(this, "schedule_start_date");
        // start_time_of_day - computed: false, optional: true, required: false
        this._startTimeOfDay = new StorageTransferJobScheduleStartTimeOfDayOutputReference(this, "start_time_of_day");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._repeatInterval !== undefined) {
            hasAnyValues = true;
            internalValueResult.repeatInterval = this._repeatInterval;
        }
        if (this._scheduleEndDate?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.scheduleEndDate = this._scheduleEndDate?.internalValue;
        }
        if (this._scheduleStartDate?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.scheduleStartDate = this._scheduleStartDate?.internalValue;
        }
        if (this._startTimeOfDay?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTimeOfDay = this._startTimeOfDay?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._repeatInterval = undefined;
            this._scheduleEndDate.internalValue = undefined;
            this._scheduleStartDate.internalValue = undefined;
            this._startTimeOfDay.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._repeatInterval = value.repeatInterval;
            this._scheduleEndDate.internalValue = value.scheduleEndDate;
            this._scheduleStartDate.internalValue = value.scheduleStartDate;
            this._startTimeOfDay.internalValue = value.startTimeOfDay;
        }
    }
    get repeatInterval() {
        return this.getStringAttribute('repeat_interval');
    }
    set repeatInterval(value) {
        this._repeatInterval = value;
    }
    resetRepeatInterval() {
        this._repeatInterval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repeatIntervalInput() {
        return this._repeatInterval;
    }
    get scheduleEndDate() {
        return this._scheduleEndDate;
    }
    putScheduleEndDate(value) {
        this._scheduleEndDate.internalValue = value;
    }
    resetScheduleEndDate() {
        this._scheduleEndDate.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleEndDateInput() {
        return this._scheduleEndDate.internalValue;
    }
    get scheduleStartDate() {
        return this._scheduleStartDate;
    }
    putScheduleStartDate(value) {
        this._scheduleStartDate.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleStartDateInput() {
        return this._scheduleStartDate.internalValue;
    }
    get startTimeOfDay() {
        return this._startTimeOfDay;
    }
    putStartTimeOfDay(value) {
        this._startTimeOfDay.internalValue = value;
    }
    resetStartTimeOfDay() {
        this._startTimeOfDay.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeOfDayInput() {
        return this._startTimeOfDay.internalValue;
    }
}
exports.StorageTransferJobScheduleOutputReference = StorageTransferJobScheduleOutputReference;
_f = JSII_RTTI_SYMBOL_1;
StorageTransferJobScheduleOutputReference[_f] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobScheduleOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access_key_id: cdktf.stringToTerraform(struct.accessKeyId),
        secret_access_key: cdktf.stringToTerraform(struct.secretAccessKey),
    };
}
exports.storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToTerraform = storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToTerraform;
function storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        access_key_id: {
            value: cdktf.stringToHclTerraform(struct.accessKeyId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        secret_access_key: {
            value: cdktf.stringToHclTerraform(struct.secretAccessKey),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToHclTerraform = storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToHclTerraform;
class StorageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accessKeyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessKeyId = this._accessKeyId;
        }
        if (this._secretAccessKey !== undefined) {
            hasAnyValues = true;
            internalValueResult.secretAccessKey = this._secretAccessKey;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accessKeyId = undefined;
            this._secretAccessKey = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accessKeyId = value.accessKeyId;
            this._secretAccessKey = value.secretAccessKey;
        }
    }
    get accessKeyId() {
        return this.getStringAttribute('access_key_id');
    }
    set accessKeyId(value) {
        this._accessKeyId = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessKeyIdInput() {
        return this._accessKeyId;
    }
    get secretAccessKey() {
        return this.getStringAttribute('secret_access_key');
    }
    set secretAccessKey(value) {
        this._secretAccessKey = value;
    }
    // Temporarily expose input value. Use with caution.
    get secretAccessKeyInput() {
        return this._secretAccessKey;
    }
}
exports.StorageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyOutputReference = StorageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyOutputReference[_g] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecAwsS3DataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
        path: cdktf.stringToTerraform(struct.path),
        role_arn: cdktf.stringToTerraform(struct.roleArn),
        aws_access_key: storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToTerraform(struct.awsAccessKey),
    };
}
exports.storageTransferJobTransferSpecAwsS3DataSourceToTerraform = storageTransferJobTransferSpecAwsS3DataSourceToTerraform;
function storageTransferJobTransferSpecAwsS3DataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket_name: {
            value: cdktf.stringToHclTerraform(struct.bucketName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        role_arn: {
            value: cdktf.stringToHclTerraform(struct.roleArn),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        aws_access_key: {
            value: storageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyToHclTerraform(struct.awsAccessKey),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecAwsS3DataSourceToHclTerraform = storageTransferJobTransferSpecAwsS3DataSourceToHclTerraform;
class StorageTransferJobTransferSpecAwsS3DataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // aws_access_key - computed: false, optional: true, required: false
        this._awsAccessKey = new StorageTransferJobTransferSpecAwsS3DataSourceAwsAccessKeyOutputReference(this, "aws_access_key");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketName = this._bucketName;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._roleArn !== undefined) {
            hasAnyValues = true;
            internalValueResult.roleArn = this._roleArn;
        }
        if (this._awsAccessKey?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.awsAccessKey = this._awsAccessKey?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketName = undefined;
            this._path = undefined;
            this._roleArn = undefined;
            this._awsAccessKey.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketName = value.bucketName;
            this._path = value.path;
            this._roleArn = value.roleArn;
            this._awsAccessKey.internalValue = value.awsAccessKey;
        }
    }
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    set bucketName(value) {
        this._bucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketNameInput() {
        return this._bucketName;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get roleArn() {
        return this.getStringAttribute('role_arn');
    }
    set roleArn(value) {
        this._roleArn = value;
    }
    resetRoleArn() {
        this._roleArn = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleArnInput() {
        return this._roleArn;
    }
    get awsAccessKey() {
        return this._awsAccessKey;
    }
    putAwsAccessKey(value) {
        this._awsAccessKey.internalValue = value;
    }
    resetAwsAccessKey() {
        this._awsAccessKey.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsAccessKeyInput() {
        return this._awsAccessKey.internalValue;
    }
}
exports.StorageTransferJobTransferSpecAwsS3DataSourceOutputReference = StorageTransferJobTransferSpecAwsS3DataSourceOutputReference;
_h = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecAwsS3DataSourceOutputReference[_h] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecAwsS3DataSourceOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sas_token: cdktf.stringToTerraform(struct.sasToken),
    };
}
exports.storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToTerraform = storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToTerraform;
function storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        sas_token: {
            value: cdktf.stringToHclTerraform(struct.sasToken),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToHclTerraform = storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToHclTerraform;
class StorageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sasToken !== undefined) {
            hasAnyValues = true;
            internalValueResult.sasToken = this._sasToken;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sasToken = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sasToken = value.sasToken;
        }
    }
    get sasToken() {
        return this.getStringAttribute('sas_token');
    }
    set sasToken(value) {
        this._sasToken = value;
    }
    // Temporarily expose input value. Use with caution.
    get sasTokenInput() {
        return this._sasToken;
    }
}
exports.StorageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsOutputReference = StorageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsOutputReference[_j] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecAzureBlobStorageDataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        container: cdktf.stringToTerraform(struct.container),
        path: cdktf.stringToTerraform(struct.path),
        storage_account: cdktf.stringToTerraform(struct.storageAccount),
        azure_credentials: storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToTerraform(struct.azureCredentials),
    };
}
exports.storageTransferJobTransferSpecAzureBlobStorageDataSourceToTerraform = storageTransferJobTransferSpecAzureBlobStorageDataSourceToTerraform;
function storageTransferJobTransferSpecAzureBlobStorageDataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        container: {
            value: cdktf.stringToHclTerraform(struct.container),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        storage_account: {
            value: cdktf.stringToHclTerraform(struct.storageAccount),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        azure_credentials: {
            value: storageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsToHclTerraform(struct.azureCredentials),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecAzureBlobStorageDataSourceToHclTerraform = storageTransferJobTransferSpecAzureBlobStorageDataSourceToHclTerraform;
class StorageTransferJobTransferSpecAzureBlobStorageDataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // azure_credentials - computed: false, optional: false, required: true
        this._azureCredentials = new StorageTransferJobTransferSpecAzureBlobStorageDataSourceAzureCredentialsOutputReference(this, "azure_credentials");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._container !== undefined) {
            hasAnyValues = true;
            internalValueResult.container = this._container;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._storageAccount !== undefined) {
            hasAnyValues = true;
            internalValueResult.storageAccount = this._storageAccount;
        }
        if (this._azureCredentials?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.azureCredentials = this._azureCredentials?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._container = undefined;
            this._path = undefined;
            this._storageAccount = undefined;
            this._azureCredentials.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._container = value.container;
            this._path = value.path;
            this._storageAccount = value.storageAccount;
            this._azureCredentials.internalValue = value.azureCredentials;
        }
    }
    get container() {
        return this.getStringAttribute('container');
    }
    set container(value) {
        this._container = value;
    }
    // Temporarily expose input value. Use with caution.
    get containerInput() {
        return this._container;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get storageAccount() {
        return this.getStringAttribute('storage_account');
    }
    set storageAccount(value) {
        this._storageAccount = value;
    }
    // Temporarily expose input value. Use with caution.
    get storageAccountInput() {
        return this._storageAccount;
    }
    get azureCredentials() {
        return this._azureCredentials;
    }
    putAzureCredentials(value) {
        this._azureCredentials.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get azureCredentialsInput() {
        return this._azureCredentials.internalValue;
    }
}
exports.StorageTransferJobTransferSpecAzureBlobStorageDataSourceOutputReference = StorageTransferJobTransferSpecAzureBlobStorageDataSourceOutputReference;
_k = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecAzureBlobStorageDataSourceOutputReference[_k] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecAzureBlobStorageDataSourceOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecGcsDataSinkToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.storageTransferJobTransferSpecGcsDataSinkToTerraform = storageTransferJobTransferSpecGcsDataSinkToTerraform;
function storageTransferJobTransferSpecGcsDataSinkToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket_name: {
            value: cdktf.stringToHclTerraform(struct.bucketName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecGcsDataSinkToHclTerraform = storageTransferJobTransferSpecGcsDataSinkToHclTerraform;
class StorageTransferJobTransferSpecGcsDataSinkOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketName = this._bucketName;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketName = undefined;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketName = value.bucketName;
            this._path = value.path;
        }
    }
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    set bucketName(value) {
        this._bucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketNameInput() {
        return this._bucketName;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.StorageTransferJobTransferSpecGcsDataSinkOutputReference = StorageTransferJobTransferSpecGcsDataSinkOutputReference;
_l = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecGcsDataSinkOutputReference[_l] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecGcsDataSinkOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecGcsDataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        bucket_name: cdktf.stringToTerraform(struct.bucketName),
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.storageTransferJobTransferSpecGcsDataSourceToTerraform = storageTransferJobTransferSpecGcsDataSourceToTerraform;
function storageTransferJobTransferSpecGcsDataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        bucket_name: {
            value: cdktf.stringToHclTerraform(struct.bucketName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecGcsDataSourceToHclTerraform = storageTransferJobTransferSpecGcsDataSourceToHclTerraform;
class StorageTransferJobTransferSpecGcsDataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._bucketName !== undefined) {
            hasAnyValues = true;
            internalValueResult.bucketName = this._bucketName;
        }
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._bucketName = undefined;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._bucketName = value.bucketName;
            this._path = value.path;
        }
    }
    get bucketName() {
        return this.getStringAttribute('bucket_name');
    }
    set bucketName(value) {
        this._bucketName = value;
    }
    // Temporarily expose input value. Use with caution.
    get bucketNameInput() {
        return this._bucketName;
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.StorageTransferJobTransferSpecGcsDataSourceOutputReference = StorageTransferJobTransferSpecGcsDataSourceOutputReference;
_m = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecGcsDataSourceOutputReference[_m] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecGcsDataSourceOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecHdfsDataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
    };
}
exports.storageTransferJobTransferSpecHdfsDataSourceToTerraform = storageTransferJobTransferSpecHdfsDataSourceToTerraform;
function storageTransferJobTransferSpecHdfsDataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        path: {
            value: cdktf.stringToHclTerraform(struct.path),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecHdfsDataSourceToHclTerraform = storageTransferJobTransferSpecHdfsDataSourceToHclTerraform;
class StorageTransferJobTransferSpecHdfsDataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
}
exports.StorageTransferJobTransferSpecHdfsDataSourceOutputReference = StorageTransferJobTransferSpecHdfsDataSourceOutputReference;
_o = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecHdfsDataSourceOutputReference[_o] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecHdfsDataSourceOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecHttpDataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        list_url: cdktf.stringToTerraform(struct.listUrl),
    };
}
exports.storageTransferJobTransferSpecHttpDataSourceToTerraform = storageTransferJobTransferSpecHttpDataSourceToTerraform;
function storageTransferJobTransferSpecHttpDataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        list_url: {
            value: cdktf.stringToHclTerraform(struct.listUrl),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecHttpDataSourceToHclTerraform = storageTransferJobTransferSpecHttpDataSourceToHclTerraform;
class StorageTransferJobTransferSpecHttpDataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._listUrl !== undefined) {
            hasAnyValues = true;
            internalValueResult.listUrl = this._listUrl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._listUrl = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._listUrl = value.listUrl;
        }
    }
    get listUrl() {
        return this.getStringAttribute('list_url');
    }
    set listUrl(value) {
        this._listUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get listUrlInput() {
        return this._listUrl;
    }
}
exports.StorageTransferJobTransferSpecHttpDataSourceOutputReference = StorageTransferJobTransferSpecHttpDataSourceOutputReference;
_p = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecHttpDataSourceOutputReference[_p] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecHttpDataSourceOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecObjectConditionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        exclude_prefixes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.excludePrefixes),
        include_prefixes: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.includePrefixes),
        last_modified_before: cdktf.stringToTerraform(struct.lastModifiedBefore),
        last_modified_since: cdktf.stringToTerraform(struct.lastModifiedSince),
        max_time_elapsed_since_last_modification: cdktf.stringToTerraform(struct.maxTimeElapsedSinceLastModification),
        min_time_elapsed_since_last_modification: cdktf.stringToTerraform(struct.minTimeElapsedSinceLastModification),
    };
}
exports.storageTransferJobTransferSpecObjectConditionsToTerraform = storageTransferJobTransferSpecObjectConditionsToTerraform;
function storageTransferJobTransferSpecObjectConditionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        exclude_prefixes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.excludePrefixes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        include_prefixes: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.includePrefixes),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        last_modified_before: {
            value: cdktf.stringToHclTerraform(struct.lastModifiedBefore),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        last_modified_since: {
            value: cdktf.stringToHclTerraform(struct.lastModifiedSince),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        max_time_elapsed_since_last_modification: {
            value: cdktf.stringToHclTerraform(struct.maxTimeElapsedSinceLastModification),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        min_time_elapsed_since_last_modification: {
            value: cdktf.stringToHclTerraform(struct.minTimeElapsedSinceLastModification),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecObjectConditionsToHclTerraform = storageTransferJobTransferSpecObjectConditionsToHclTerraform;
class StorageTransferJobTransferSpecObjectConditionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._excludePrefixes !== undefined) {
            hasAnyValues = true;
            internalValueResult.excludePrefixes = this._excludePrefixes;
        }
        if (this._includePrefixes !== undefined) {
            hasAnyValues = true;
            internalValueResult.includePrefixes = this._includePrefixes;
        }
        if (this._lastModifiedBefore !== undefined) {
            hasAnyValues = true;
            internalValueResult.lastModifiedBefore = this._lastModifiedBefore;
        }
        if (this._lastModifiedSince !== undefined) {
            hasAnyValues = true;
            internalValueResult.lastModifiedSince = this._lastModifiedSince;
        }
        if (this._maxTimeElapsedSinceLastModification !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxTimeElapsedSinceLastModification = this._maxTimeElapsedSinceLastModification;
        }
        if (this._minTimeElapsedSinceLastModification !== undefined) {
            hasAnyValues = true;
            internalValueResult.minTimeElapsedSinceLastModification = this._minTimeElapsedSinceLastModification;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._excludePrefixes = undefined;
            this._includePrefixes = undefined;
            this._lastModifiedBefore = undefined;
            this._lastModifiedSince = undefined;
            this._maxTimeElapsedSinceLastModification = undefined;
            this._minTimeElapsedSinceLastModification = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._excludePrefixes = value.excludePrefixes;
            this._includePrefixes = value.includePrefixes;
            this._lastModifiedBefore = value.lastModifiedBefore;
            this._lastModifiedSince = value.lastModifiedSince;
            this._maxTimeElapsedSinceLastModification = value.maxTimeElapsedSinceLastModification;
            this._minTimeElapsedSinceLastModification = value.minTimeElapsedSinceLastModification;
        }
    }
    get excludePrefixes() {
        return this.getListAttribute('exclude_prefixes');
    }
    set excludePrefixes(value) {
        this._excludePrefixes = value;
    }
    resetExcludePrefixes() {
        this._excludePrefixes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get excludePrefixesInput() {
        return this._excludePrefixes;
    }
    get includePrefixes() {
        return this.getListAttribute('include_prefixes');
    }
    set includePrefixes(value) {
        this._includePrefixes = value;
    }
    resetIncludePrefixes() {
        this._includePrefixes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includePrefixesInput() {
        return this._includePrefixes;
    }
    get lastModifiedBefore() {
        return this.getStringAttribute('last_modified_before');
    }
    set lastModifiedBefore(value) {
        this._lastModifiedBefore = value;
    }
    resetLastModifiedBefore() {
        this._lastModifiedBefore = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lastModifiedBeforeInput() {
        return this._lastModifiedBefore;
    }
    get lastModifiedSince() {
        return this.getStringAttribute('last_modified_since');
    }
    set lastModifiedSince(value) {
        this._lastModifiedSince = value;
    }
    resetLastModifiedSince() {
        this._lastModifiedSince = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lastModifiedSinceInput() {
        return this._lastModifiedSince;
    }
    get maxTimeElapsedSinceLastModification() {
        return this.getStringAttribute('max_time_elapsed_since_last_modification');
    }
    set maxTimeElapsedSinceLastModification(value) {
        this._maxTimeElapsedSinceLastModification = value;
    }
    resetMaxTimeElapsedSinceLastModification() {
        this._maxTimeElapsedSinceLastModification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxTimeElapsedSinceLastModificationInput() {
        return this._maxTimeElapsedSinceLastModification;
    }
    get minTimeElapsedSinceLastModification() {
        return this.getStringAttribute('min_time_elapsed_since_last_modification');
    }
    set minTimeElapsedSinceLastModification(value) {
        this._minTimeElapsedSinceLastModification = value;
    }
    resetMinTimeElapsedSinceLastModification() {
        this._minTimeElapsedSinceLastModification = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minTimeElapsedSinceLastModificationInput() {
        return this._minTimeElapsedSinceLastModification;
    }
}
exports.StorageTransferJobTransferSpecObjectConditionsOutputReference = StorageTransferJobTransferSpecObjectConditionsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecObjectConditionsOutputReference[_q] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecObjectConditionsOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecPosixDataSinkToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        root_directory: cdktf.stringToTerraform(struct.rootDirectory),
    };
}
exports.storageTransferJobTransferSpecPosixDataSinkToTerraform = storageTransferJobTransferSpecPosixDataSinkToTerraform;
function storageTransferJobTransferSpecPosixDataSinkToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        root_directory: {
            value: cdktf.stringToHclTerraform(struct.rootDirectory),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecPosixDataSinkToHclTerraform = storageTransferJobTransferSpecPosixDataSinkToHclTerraform;
class StorageTransferJobTransferSpecPosixDataSinkOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rootDirectory !== undefined) {
            hasAnyValues = true;
            internalValueResult.rootDirectory = this._rootDirectory;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rootDirectory = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rootDirectory = value.rootDirectory;
        }
    }
    get rootDirectory() {
        return this.getStringAttribute('root_directory');
    }
    set rootDirectory(value) {
        this._rootDirectory = value;
    }
    // Temporarily expose input value. Use with caution.
    get rootDirectoryInput() {
        return this._rootDirectory;
    }
}
exports.StorageTransferJobTransferSpecPosixDataSinkOutputReference = StorageTransferJobTransferSpecPosixDataSinkOutputReference;
_r = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecPosixDataSinkOutputReference[_r] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecPosixDataSinkOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecPosixDataSourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        root_directory: cdktf.stringToTerraform(struct.rootDirectory),
    };
}
exports.storageTransferJobTransferSpecPosixDataSourceToTerraform = storageTransferJobTransferSpecPosixDataSourceToTerraform;
function storageTransferJobTransferSpecPosixDataSourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        root_directory: {
            value: cdktf.stringToHclTerraform(struct.rootDirectory),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecPosixDataSourceToHclTerraform = storageTransferJobTransferSpecPosixDataSourceToHclTerraform;
class StorageTransferJobTransferSpecPosixDataSourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._rootDirectory !== undefined) {
            hasAnyValues = true;
            internalValueResult.rootDirectory = this._rootDirectory;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._rootDirectory = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._rootDirectory = value.rootDirectory;
        }
    }
    get rootDirectory() {
        return this.getStringAttribute('root_directory');
    }
    set rootDirectory(value) {
        this._rootDirectory = value;
    }
    // Temporarily expose input value. Use with caution.
    get rootDirectoryInput() {
        return this._rootDirectory;
    }
}
exports.StorageTransferJobTransferSpecPosixDataSourceOutputReference = StorageTransferJobTransferSpecPosixDataSourceOutputReference;
_s = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecPosixDataSourceOutputReference[_s] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecPosixDataSourceOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecTransferOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        delete_objects_from_source_after_transfer: cdktf.booleanToTerraform(struct.deleteObjectsFromSourceAfterTransfer),
        delete_objects_unique_in_sink: cdktf.booleanToTerraform(struct.deleteObjectsUniqueInSink),
        overwrite_objects_already_existing_in_sink: cdktf.booleanToTerraform(struct.overwriteObjectsAlreadyExistingInSink),
        overwrite_when: cdktf.stringToTerraform(struct.overwriteWhen),
    };
}
exports.storageTransferJobTransferSpecTransferOptionsToTerraform = storageTransferJobTransferSpecTransferOptionsToTerraform;
function storageTransferJobTransferSpecTransferOptionsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        delete_objects_from_source_after_transfer: {
            value: cdktf.booleanToHclTerraform(struct.deleteObjectsFromSourceAfterTransfer),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        delete_objects_unique_in_sink: {
            value: cdktf.booleanToHclTerraform(struct.deleteObjectsUniqueInSink),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        overwrite_objects_already_existing_in_sink: {
            value: cdktf.booleanToHclTerraform(struct.overwriteObjectsAlreadyExistingInSink),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        overwrite_when: {
            value: cdktf.stringToHclTerraform(struct.overwriteWhen),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecTransferOptionsToHclTerraform = storageTransferJobTransferSpecTransferOptionsToHclTerraform;
class StorageTransferJobTransferSpecTransferOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deleteObjectsFromSourceAfterTransfer !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteObjectsFromSourceAfterTransfer = this._deleteObjectsFromSourceAfterTransfer;
        }
        if (this._deleteObjectsUniqueInSink !== undefined) {
            hasAnyValues = true;
            internalValueResult.deleteObjectsUniqueInSink = this._deleteObjectsUniqueInSink;
        }
        if (this._overwriteObjectsAlreadyExistingInSink !== undefined) {
            hasAnyValues = true;
            internalValueResult.overwriteObjectsAlreadyExistingInSink = this._overwriteObjectsAlreadyExistingInSink;
        }
        if (this._overwriteWhen !== undefined) {
            hasAnyValues = true;
            internalValueResult.overwriteWhen = this._overwriteWhen;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deleteObjectsFromSourceAfterTransfer = undefined;
            this._deleteObjectsUniqueInSink = undefined;
            this._overwriteObjectsAlreadyExistingInSink = undefined;
            this._overwriteWhen = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deleteObjectsFromSourceAfterTransfer = value.deleteObjectsFromSourceAfterTransfer;
            this._deleteObjectsUniqueInSink = value.deleteObjectsUniqueInSink;
            this._overwriteObjectsAlreadyExistingInSink = value.overwriteObjectsAlreadyExistingInSink;
            this._overwriteWhen = value.overwriteWhen;
        }
    }
    get deleteObjectsFromSourceAfterTransfer() {
        return this.getBooleanAttribute('delete_objects_from_source_after_transfer');
    }
    set deleteObjectsFromSourceAfterTransfer(value) {
        this._deleteObjectsFromSourceAfterTransfer = value;
    }
    resetDeleteObjectsFromSourceAfterTransfer() {
        this._deleteObjectsFromSourceAfterTransfer = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteObjectsFromSourceAfterTransferInput() {
        return this._deleteObjectsFromSourceAfterTransfer;
    }
    get deleteObjectsUniqueInSink() {
        return this.getBooleanAttribute('delete_objects_unique_in_sink');
    }
    set deleteObjectsUniqueInSink(value) {
        this._deleteObjectsUniqueInSink = value;
    }
    resetDeleteObjectsUniqueInSink() {
        this._deleteObjectsUniqueInSink = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteObjectsUniqueInSinkInput() {
        return this._deleteObjectsUniqueInSink;
    }
    get overwriteObjectsAlreadyExistingInSink() {
        return this.getBooleanAttribute('overwrite_objects_already_existing_in_sink');
    }
    set overwriteObjectsAlreadyExistingInSink(value) {
        this._overwriteObjectsAlreadyExistingInSink = value;
    }
    resetOverwriteObjectsAlreadyExistingInSink() {
        this._overwriteObjectsAlreadyExistingInSink = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overwriteObjectsAlreadyExistingInSinkInput() {
        return this._overwriteObjectsAlreadyExistingInSink;
    }
    get overwriteWhen() {
        return this.getStringAttribute('overwrite_when');
    }
    set overwriteWhen(value) {
        this._overwriteWhen = value;
    }
    resetOverwriteWhen() {
        this._overwriteWhen = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get overwriteWhenInput() {
        return this._overwriteWhen;
    }
}
exports.StorageTransferJobTransferSpecTransferOptionsOutputReference = StorageTransferJobTransferSpecTransferOptionsOutputReference;
_t = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecTransferOptionsOutputReference[_t] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecTransferOptionsOutputReference", version: "14.13.1" };
function storageTransferJobTransferSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        sink_agent_pool_name: cdktf.stringToTerraform(struct.sinkAgentPoolName),
        source_agent_pool_name: cdktf.stringToTerraform(struct.sourceAgentPoolName),
        aws_s3_data_source: storageTransferJobTransferSpecAwsS3DataSourceToTerraform(struct.awsS3DataSource),
        azure_blob_storage_data_source: storageTransferJobTransferSpecAzureBlobStorageDataSourceToTerraform(struct.azureBlobStorageDataSource),
        gcs_data_sink: storageTransferJobTransferSpecGcsDataSinkToTerraform(struct.gcsDataSink),
        gcs_data_source: storageTransferJobTransferSpecGcsDataSourceToTerraform(struct.gcsDataSource),
        hdfs_data_source: storageTransferJobTransferSpecHdfsDataSourceToTerraform(struct.hdfsDataSource),
        http_data_source: storageTransferJobTransferSpecHttpDataSourceToTerraform(struct.httpDataSource),
        object_conditions: storageTransferJobTransferSpecObjectConditionsToTerraform(struct.objectConditions),
        posix_data_sink: storageTransferJobTransferSpecPosixDataSinkToTerraform(struct.posixDataSink),
        posix_data_source: storageTransferJobTransferSpecPosixDataSourceToTerraform(struct.posixDataSource),
        transfer_options: storageTransferJobTransferSpecTransferOptionsToTerraform(struct.transferOptions),
    };
}
exports.storageTransferJobTransferSpecToTerraform = storageTransferJobTransferSpecToTerraform;
function storageTransferJobTransferSpecToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        sink_agent_pool_name: {
            value: cdktf.stringToHclTerraform(struct.sinkAgentPoolName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        source_agent_pool_name: {
            value: cdktf.stringToHclTerraform(struct.sourceAgentPoolName),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        aws_s3_data_source: {
            value: storageTransferJobTransferSpecAwsS3DataSourceToHclTerraform(struct.awsS3DataSource),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecAwsS3DataSourceList",
        },
        azure_blob_storage_data_source: {
            value: storageTransferJobTransferSpecAzureBlobStorageDataSourceToHclTerraform(struct.azureBlobStorageDataSource),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecAzureBlobStorageDataSourceList",
        },
        gcs_data_sink: {
            value: storageTransferJobTransferSpecGcsDataSinkToHclTerraform(struct.gcsDataSink),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecGcsDataSinkList",
        },
        gcs_data_source: {
            value: storageTransferJobTransferSpecGcsDataSourceToHclTerraform(struct.gcsDataSource),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecGcsDataSourceList",
        },
        hdfs_data_source: {
            value: storageTransferJobTransferSpecHdfsDataSourceToHclTerraform(struct.hdfsDataSource),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecHdfsDataSourceList",
        },
        http_data_source: {
            value: storageTransferJobTransferSpecHttpDataSourceToHclTerraform(struct.httpDataSource),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecHttpDataSourceList",
        },
        object_conditions: {
            value: storageTransferJobTransferSpecObjectConditionsToHclTerraform(struct.objectConditions),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecObjectConditionsList",
        },
        posix_data_sink: {
            value: storageTransferJobTransferSpecPosixDataSinkToHclTerraform(struct.posixDataSink),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecPosixDataSinkList",
        },
        posix_data_source: {
            value: storageTransferJobTransferSpecPosixDataSourceToHclTerraform(struct.posixDataSource),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecPosixDataSourceList",
        },
        transfer_options: {
            value: storageTransferJobTransferSpecTransferOptionsToHclTerraform(struct.transferOptions),
            isBlock: true,
            type: "list",
            storageClassType: "StorageTransferJobTransferSpecTransferOptionsList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.storageTransferJobTransferSpecToHclTerraform = storageTransferJobTransferSpecToHclTerraform;
class StorageTransferJobTransferSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // aws_s3_data_source - computed: false, optional: true, required: false
        this._awsS3DataSource = new StorageTransferJobTransferSpecAwsS3DataSourceOutputReference(this, "aws_s3_data_source");
        // azure_blob_storage_data_source - computed: false, optional: true, required: false
        this._azureBlobStorageDataSource = new StorageTransferJobTransferSpecAzureBlobStorageDataSourceOutputReference(this, "azure_blob_storage_data_source");
        // gcs_data_sink - computed: false, optional: true, required: false
        this._gcsDataSink = new StorageTransferJobTransferSpecGcsDataSinkOutputReference(this, "gcs_data_sink");
        // gcs_data_source - computed: false, optional: true, required: false
        this._gcsDataSource = new StorageTransferJobTransferSpecGcsDataSourceOutputReference(this, "gcs_data_source");
        // hdfs_data_source - computed: false, optional: true, required: false
        this._hdfsDataSource = new StorageTransferJobTransferSpecHdfsDataSourceOutputReference(this, "hdfs_data_source");
        // http_data_source - computed: false, optional: true, required: false
        this._httpDataSource = new StorageTransferJobTransferSpecHttpDataSourceOutputReference(this, "http_data_source");
        // object_conditions - computed: false, optional: true, required: false
        this._objectConditions = new StorageTransferJobTransferSpecObjectConditionsOutputReference(this, "object_conditions");
        // posix_data_sink - computed: false, optional: true, required: false
        this._posixDataSink = new StorageTransferJobTransferSpecPosixDataSinkOutputReference(this, "posix_data_sink");
        // posix_data_source - computed: false, optional: true, required: false
        this._posixDataSource = new StorageTransferJobTransferSpecPosixDataSourceOutputReference(this, "posix_data_source");
        // transfer_options - computed: false, optional: true, required: false
        this._transferOptions = new StorageTransferJobTransferSpecTransferOptionsOutputReference(this, "transfer_options");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sinkAgentPoolName !== undefined) {
            hasAnyValues = true;
            internalValueResult.sinkAgentPoolName = this._sinkAgentPoolName;
        }
        if (this._sourceAgentPoolName !== undefined) {
            hasAnyValues = true;
            internalValueResult.sourceAgentPoolName = this._sourceAgentPoolName;
        }
        if (this._awsS3DataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.awsS3DataSource = this._awsS3DataSource?.internalValue;
        }
        if (this._azureBlobStorageDataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.azureBlobStorageDataSource = this._azureBlobStorageDataSource?.internalValue;
        }
        if (this._gcsDataSink?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.gcsDataSink = this._gcsDataSink?.internalValue;
        }
        if (this._gcsDataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.gcsDataSource = this._gcsDataSource?.internalValue;
        }
        if (this._hdfsDataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hdfsDataSource = this._hdfsDataSource?.internalValue;
        }
        if (this._httpDataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpDataSource = this._httpDataSource?.internalValue;
        }
        if (this._objectConditions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.objectConditions = this._objectConditions?.internalValue;
        }
        if (this._posixDataSink?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.posixDataSink = this._posixDataSink?.internalValue;
        }
        if (this._posixDataSource?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.posixDataSource = this._posixDataSource?.internalValue;
        }
        if (this._transferOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.transferOptions = this._transferOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._sinkAgentPoolName = undefined;
            this._sourceAgentPoolName = undefined;
            this._awsS3DataSource.internalValue = undefined;
            this._azureBlobStorageDataSource.internalValue = undefined;
            this._gcsDataSink.internalValue = undefined;
            this._gcsDataSource.internalValue = undefined;
            this._hdfsDataSource.internalValue = undefined;
            this._httpDataSource.internalValue = undefined;
            this._objectConditions.internalValue = undefined;
            this._posixDataSink.internalValue = undefined;
            this._posixDataSource.internalValue = undefined;
            this._transferOptions.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._sinkAgentPoolName = value.sinkAgentPoolName;
            this._sourceAgentPoolName = value.sourceAgentPoolName;
            this._awsS3DataSource.internalValue = value.awsS3DataSource;
            this._azureBlobStorageDataSource.internalValue = value.azureBlobStorageDataSource;
            this._gcsDataSink.internalValue = value.gcsDataSink;
            this._gcsDataSource.internalValue = value.gcsDataSource;
            this._hdfsDataSource.internalValue = value.hdfsDataSource;
            this._httpDataSource.internalValue = value.httpDataSource;
            this._objectConditions.internalValue = value.objectConditions;
            this._posixDataSink.internalValue = value.posixDataSink;
            this._posixDataSource.internalValue = value.posixDataSource;
            this._transferOptions.internalValue = value.transferOptions;
        }
    }
    get sinkAgentPoolName() {
        return this.getStringAttribute('sink_agent_pool_name');
    }
    set sinkAgentPoolName(value) {
        this._sinkAgentPoolName = value;
    }
    resetSinkAgentPoolName() {
        this._sinkAgentPoolName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sinkAgentPoolNameInput() {
        return this._sinkAgentPoolName;
    }
    get sourceAgentPoolName() {
        return this.getStringAttribute('source_agent_pool_name');
    }
    set sourceAgentPoolName(value) {
        this._sourceAgentPoolName = value;
    }
    resetSourceAgentPoolName() {
        this._sourceAgentPoolName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sourceAgentPoolNameInput() {
        return this._sourceAgentPoolName;
    }
    get awsS3DataSource() {
        return this._awsS3DataSource;
    }
    putAwsS3DataSource(value) {
        this._awsS3DataSource.internalValue = value;
    }
    resetAwsS3DataSource() {
        this._awsS3DataSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsS3DataSourceInput() {
        return this._awsS3DataSource.internalValue;
    }
    get azureBlobStorageDataSource() {
        return this._azureBlobStorageDataSource;
    }
    putAzureBlobStorageDataSource(value) {
        this._azureBlobStorageDataSource.internalValue = value;
    }
    resetAzureBlobStorageDataSource() {
        this._azureBlobStorageDataSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get azureBlobStorageDataSourceInput() {
        return this._azureBlobStorageDataSource.internalValue;
    }
    get gcsDataSink() {
        return this._gcsDataSink;
    }
    putGcsDataSink(value) {
        this._gcsDataSink.internalValue = value;
    }
    resetGcsDataSink() {
        this._gcsDataSink.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gcsDataSinkInput() {
        return this._gcsDataSink.internalValue;
    }
    get gcsDataSource() {
        return this._gcsDataSource;
    }
    putGcsDataSource(value) {
        this._gcsDataSource.internalValue = value;
    }
    resetGcsDataSource() {
        this._gcsDataSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gcsDataSourceInput() {
        return this._gcsDataSource.internalValue;
    }
    get hdfsDataSource() {
        return this._hdfsDataSource;
    }
    putHdfsDataSource(value) {
        this._hdfsDataSource.internalValue = value;
    }
    resetHdfsDataSource() {
        this._hdfsDataSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hdfsDataSourceInput() {
        return this._hdfsDataSource.internalValue;
    }
    get httpDataSource() {
        return this._httpDataSource;
    }
    putHttpDataSource(value) {
        this._httpDataSource.internalValue = value;
    }
    resetHttpDataSource() {
        this._httpDataSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpDataSourceInput() {
        return this._httpDataSource.internalValue;
    }
    get objectConditions() {
        return this._objectConditions;
    }
    putObjectConditions(value) {
        this._objectConditions.internalValue = value;
    }
    resetObjectConditions() {
        this._objectConditions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get objectConditionsInput() {
        return this._objectConditions.internalValue;
    }
    get posixDataSink() {
        return this._posixDataSink;
    }
    putPosixDataSink(value) {
        this._posixDataSink.internalValue = value;
    }
    resetPosixDataSink() {
        this._posixDataSink.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get posixDataSinkInput() {
        return this._posixDataSink.internalValue;
    }
    get posixDataSource() {
        return this._posixDataSource;
    }
    putPosixDataSource(value) {
        this._posixDataSource.internalValue = value;
    }
    resetPosixDataSource() {
        this._posixDataSource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get posixDataSourceInput() {
        return this._posixDataSource.internalValue;
    }
    get transferOptions() {
        return this._transferOptions;
    }
    putTransferOptions(value) {
        this._transferOptions.internalValue = value;
    }
    resetTransferOptions() {
        this._transferOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get transferOptionsInput() {
        return this._transferOptions.internalValue;
    }
}
exports.StorageTransferJobTransferSpecOutputReference = StorageTransferJobTransferSpecOutputReference;
_u = JSII_RTTI_SYMBOL_1;
StorageTransferJobTransferSpecOutputReference[_u] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJobTransferSpecOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/storage_transfer_job google_storage_transfer_job}
*/
class StorageTransferJob extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a StorageTransferJob resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the StorageTransferJob to import
    * @param importFromId The id of the existing StorageTransferJob that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/storage_transfer_job#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the StorageTransferJob to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_storage_transfer_job", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/storage_transfer_job google_storage_transfer_job} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options StorageTransferJobConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_storage_transfer_job',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // event_stream - computed: false, optional: true, required: false
        this._eventStream = new StorageTransferJobEventStreamOutputReference(this, "event_stream");
        // notification_config - computed: false, optional: true, required: false
        this._notificationConfig = new StorageTransferJobNotificationConfigOutputReference(this, "notification_config");
        // schedule - computed: false, optional: true, required: false
        this._schedule = new StorageTransferJobScheduleOutputReference(this, "schedule");
        // transfer_spec - computed: false, optional: false, required: true
        this._transferSpec = new StorageTransferJobTransferSpecOutputReference(this, "transfer_spec");
        this._description = config.description;
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
        this._status = config.status;
        this._eventStream.internalValue = config.eventStream;
        this._notificationConfig.internalValue = config.notificationConfig;
        this._schedule.internalValue = config.schedule;
        this._transferSpec.internalValue = config.transferSpec;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // creation_time - computed: true, optional: false, required: false
    get creationTime() {
        return this.getStringAttribute('creation_time');
    }
    // deletion_time - computed: true, optional: false, required: false
    get deletionTime() {
        return this.getStringAttribute('deletion_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // last_modification_time - computed: true, optional: false, required: false
    get lastModificationTime() {
        return this.getStringAttribute('last_modification_time');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get eventStream() {
        return this._eventStream;
    }
    putEventStream(value) {
        this._eventStream.internalValue = value;
    }
    resetEventStream() {
        this._eventStream.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get eventStreamInput() {
        return this._eventStream.internalValue;
    }
    get notificationConfig() {
        return this._notificationConfig;
    }
    putNotificationConfig(value) {
        this._notificationConfig.internalValue = value;
    }
    resetNotificationConfig() {
        this._notificationConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notificationConfigInput() {
        return this._notificationConfig.internalValue;
    }
    get schedule() {
        return this._schedule;
    }
    putSchedule(value) {
        this._schedule.internalValue = value;
    }
    resetSchedule() {
        this._schedule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get scheduleInput() {
        return this._schedule.internalValue;
    }
    get transferSpec() {
        return this._transferSpec;
    }
    putTransferSpec(value) {
        this._transferSpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get transferSpecInput() {
        return this._transferSpec.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
            status: cdktf.stringToTerraform(this._status),
            event_stream: storageTransferJobEventStreamToTerraform(this._eventStream.internalValue),
            notification_config: storageTransferJobNotificationConfigToTerraform(this._notificationConfig.internalValue),
            schedule: storageTransferJobScheduleToTerraform(this._schedule.internalValue),
            transfer_spec: storageTransferJobTransferSpecToTerraform(this._transferSpec.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            status: {
                value: cdktf.stringToHclTerraform(this._status),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            event_stream: {
                value: storageTransferJobEventStreamToHclTerraform(this._eventStream.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "StorageTransferJobEventStreamList",
            },
            notification_config: {
                value: storageTransferJobNotificationConfigToHclTerraform(this._notificationConfig.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "StorageTransferJobNotificationConfigList",
            },
            schedule: {
                value: storageTransferJobScheduleToHclTerraform(this._schedule.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "StorageTransferJobScheduleList",
            },
            transfer_spec: {
                value: storageTransferJobTransferSpecToHclTerraform(this._transferSpec.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "StorageTransferJobTransferSpecList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.StorageTransferJob = StorageTransferJob;
_v = JSII_RTTI_SYMBOL_1;
StorageTransferJob[_v] = { fqn: "@cdktf/provider-google.storageTransferJob.StorageTransferJob", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
StorageTransferJob.tfResourceType = "google_storage_transfer_job";
