// @cdktf/provider-google/lib/network-connectivity-policy-based-route/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkConnectivityPolicyBasedRoute = exports.NetworkConnectivityPolicyBasedRouteVirtualMachineOutputReference = exports.networkConnectivityPolicyBasedRouteVirtualMachineToHclTerraform = exports.networkConnectivityPolicyBasedRouteVirtualMachineToTerraform = exports.NetworkConnectivityPolicyBasedRouteTimeoutsOutputReference = exports.networkConnectivityPolicyBasedRouteTimeoutsToHclTerraform = exports.networkConnectivityPolicyBasedRouteTimeoutsToTerraform = exports.NetworkConnectivityPolicyBasedRouteInterconnectAttachmentOutputReference = exports.networkConnectivityPolicyBasedRouteInterconnectAttachmentToHclTerraform = exports.networkConnectivityPolicyBasedRouteInterconnectAttachmentToTerraform = exports.NetworkConnectivityPolicyBasedRouteFilterOutputReference = exports.networkConnectivityPolicyBasedRouteFilterToHclTerraform = exports.networkConnectivityPolicyBasedRouteFilterToTerraform = exports.NetworkConnectivityPolicyBasedRouteWarningsList = exports.NetworkConnectivityPolicyBasedRouteWarningsOutputReference = exports.networkConnectivityPolicyBasedRouteWarningsToHclTerraform = exports.networkConnectivityPolicyBasedRouteWarningsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkConnectivityPolicyBasedRouteWarningsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.networkConnectivityPolicyBasedRouteWarningsToTerraform = networkConnectivityPolicyBasedRouteWarningsToTerraform;
function networkConnectivityPolicyBasedRouteWarningsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.networkConnectivityPolicyBasedRouteWarningsToHclTerraform = networkConnectivityPolicyBasedRouteWarningsToHclTerraform;
class NetworkConnectivityPolicyBasedRouteWarningsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // data - computed: true, optional: false, required: false
        this._data = new cdktf.StringMap(this, "data");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // code - computed: true, optional: false, required: false
    get code() {
        return this.getStringAttribute('code');
    }
    get data() {
        return this._data;
    }
    // warning_message - computed: true, optional: false, required: false
    get warningMessage() {
        return this.getStringAttribute('warning_message');
    }
}
exports.NetworkConnectivityPolicyBasedRouteWarningsOutputReference = NetworkConnectivityPolicyBasedRouteWarningsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkConnectivityPolicyBasedRouteWarningsOutputReference[_a] = { fqn: "@cdktf/provider-google.networkConnectivityPolicyBasedRoute.NetworkConnectivityPolicyBasedRouteWarningsOutputReference", version: "14.13.1" };
class NetworkConnectivityPolicyBasedRouteWarningsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkConnectivityPolicyBasedRouteWarningsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkConnectivityPolicyBasedRouteWarningsList = NetworkConnectivityPolicyBasedRouteWarningsList;
_b = JSII_RTTI_SYMBOL_1;
NetworkConnectivityPolicyBasedRouteWarningsList[_b] = { fqn: "@cdktf/provider-google.networkConnectivityPolicyBasedRoute.NetworkConnectivityPolicyBasedRouteWarningsList", version: "14.13.1" };
function networkConnectivityPolicyBasedRouteFilterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dest_range: cdktf.stringToTerraform(struct.destRange),
        ip_protocol: cdktf.stringToTerraform(struct.ipProtocol),
        protocol_version: cdktf.stringToTerraform(struct.protocolVersion),
        src_range: cdktf.stringToTerraform(struct.srcRange),
    };
}
exports.networkConnectivityPolicyBasedRouteFilterToTerraform = networkConnectivityPolicyBasedRouteFilterToTerraform;
function networkConnectivityPolicyBasedRouteFilterToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        dest_range: {
            value: cdktf.stringToHclTerraform(struct.destRange),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        ip_protocol: {
            value: cdktf.stringToHclTerraform(struct.ipProtocol),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        protocol_version: {
            value: cdktf.stringToHclTerraform(struct.protocolVersion),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        src_range: {
            value: cdktf.stringToHclTerraform(struct.srcRange),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivityPolicyBasedRouteFilterToHclTerraform = networkConnectivityPolicyBasedRouteFilterToHclTerraform;
class NetworkConnectivityPolicyBasedRouteFilterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._destRange !== undefined) {
            hasAnyValues = true;
            internalValueResult.destRange = this._destRange;
        }
        if (this._ipProtocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipProtocol = this._ipProtocol;
        }
        if (this._protocolVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocolVersion = this._protocolVersion;
        }
        if (this._srcRange !== undefined) {
            hasAnyValues = true;
            internalValueResult.srcRange = this._srcRange;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._destRange = undefined;
            this._ipProtocol = undefined;
            this._protocolVersion = undefined;
            this._srcRange = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._destRange = value.destRange;
            this._ipProtocol = value.ipProtocol;
            this._protocolVersion = value.protocolVersion;
            this._srcRange = value.srcRange;
        }
    }
    get destRange() {
        return this.getStringAttribute('dest_range');
    }
    set destRange(value) {
        this._destRange = value;
    }
    resetDestRange() {
        this._destRange = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get destRangeInput() {
        return this._destRange;
    }
    get ipProtocol() {
        return this.getStringAttribute('ip_protocol');
    }
    set ipProtocol(value) {
        this._ipProtocol = value;
    }
    resetIpProtocol() {
        this._ipProtocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipProtocolInput() {
        return this._ipProtocol;
    }
    get protocolVersion() {
        return this.getStringAttribute('protocol_version');
    }
    set protocolVersion(value) {
        this._protocolVersion = value;
    }
    // Temporarily expose input value. Use with caution.
    get protocolVersionInput() {
        return this._protocolVersion;
    }
    get srcRange() {
        return this.getStringAttribute('src_range');
    }
    set srcRange(value) {
        this._srcRange = value;
    }
    resetSrcRange() {
        this._srcRange = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get srcRangeInput() {
        return this._srcRange;
    }
}
exports.NetworkConnectivityPolicyBasedRouteFilterOutputReference = NetworkConnectivityPolicyBasedRouteFilterOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetworkConnectivityPolicyBasedRouteFilterOutputReference[_c] = { fqn: "@cdktf/provider-google.networkConnectivityPolicyBasedRoute.NetworkConnectivityPolicyBasedRouteFilterOutputReference", version: "14.13.1" };
function networkConnectivityPolicyBasedRouteInterconnectAttachmentToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        region: cdktf.stringToTerraform(struct.region),
    };
}
exports.networkConnectivityPolicyBasedRouteInterconnectAttachmentToTerraform = networkConnectivityPolicyBasedRouteInterconnectAttachmentToTerraform;
function networkConnectivityPolicyBasedRouteInterconnectAttachmentToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        region: {
            value: cdktf.stringToHclTerraform(struct.region),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivityPolicyBasedRouteInterconnectAttachmentToHclTerraform = networkConnectivityPolicyBasedRouteInterconnectAttachmentToHclTerraform;
class NetworkConnectivityPolicyBasedRouteInterconnectAttachmentOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._region = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._region = value.region;
        }
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
}
exports.NetworkConnectivityPolicyBasedRouteInterconnectAttachmentOutputReference = NetworkConnectivityPolicyBasedRouteInterconnectAttachmentOutputReference;
_d = JSII_RTTI_SYMBOL_1;
NetworkConnectivityPolicyBasedRouteInterconnectAttachmentOutputReference[_d] = { fqn: "@cdktf/provider-google.networkConnectivityPolicyBasedRoute.NetworkConnectivityPolicyBasedRouteInterconnectAttachmentOutputReference", version: "14.13.1" };
function networkConnectivityPolicyBasedRouteTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkConnectivityPolicyBasedRouteTimeoutsToTerraform = networkConnectivityPolicyBasedRouteTimeoutsToTerraform;
function networkConnectivityPolicyBasedRouteTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivityPolicyBasedRouteTimeoutsToHclTerraform = networkConnectivityPolicyBasedRouteTimeoutsToHclTerraform;
class NetworkConnectivityPolicyBasedRouteTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkConnectivityPolicyBasedRouteTimeoutsOutputReference = NetworkConnectivityPolicyBasedRouteTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
NetworkConnectivityPolicyBasedRouteTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-google.networkConnectivityPolicyBasedRoute.NetworkConnectivityPolicyBasedRouteTimeoutsOutputReference", version: "14.13.1" };
function networkConnectivityPolicyBasedRouteVirtualMachineToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        tags: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.tags),
    };
}
exports.networkConnectivityPolicyBasedRouteVirtualMachineToTerraform = networkConnectivityPolicyBasedRouteVirtualMachineToTerraform;
function networkConnectivityPolicyBasedRouteVirtualMachineToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        tags: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.tags),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkConnectivityPolicyBasedRouteVirtualMachineToHclTerraform = networkConnectivityPolicyBasedRouteVirtualMachineToHclTerraform;
class NetworkConnectivityPolicyBasedRouteVirtualMachineOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._tags !== undefined) {
            hasAnyValues = true;
            internalValueResult.tags = this._tags;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._tags = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._tags = value.tags;
        }
    }
    get tags() {
        return this.getListAttribute('tags');
    }
    set tags(value) {
        this._tags = value;
    }
    // Temporarily expose input value. Use with caution.
    get tagsInput() {
        return this._tags;
    }
}
exports.NetworkConnectivityPolicyBasedRouteVirtualMachineOutputReference = NetworkConnectivityPolicyBasedRouteVirtualMachineOutputReference;
_f = JSII_RTTI_SYMBOL_1;
NetworkConnectivityPolicyBasedRouteVirtualMachineOutputReference[_f] = { fqn: "@cdktf/provider-google.networkConnectivityPolicyBasedRoute.NetworkConnectivityPolicyBasedRouteVirtualMachineOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_policy_based_route google_network_connectivity_policy_based_route}
*/
class NetworkConnectivityPolicyBasedRoute extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkConnectivityPolicyBasedRoute resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkConnectivityPolicyBasedRoute to import
    * @param importFromId The id of the existing NetworkConnectivityPolicyBasedRoute that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_policy_based_route#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkConnectivityPolicyBasedRoute to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_network_connectivity_policy_based_route", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_connectivity_policy_based_route google_network_connectivity_policy_based_route} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkConnectivityPolicyBasedRouteConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_network_connectivity_policy_based_route',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // warnings - computed: true, optional: false, required: false
        this._warnings = new NetworkConnectivityPolicyBasedRouteWarningsList(this, "warnings", false);
        // filter - computed: false, optional: false, required: true
        this._filter = new NetworkConnectivityPolicyBasedRouteFilterOutputReference(this, "filter");
        // interconnect_attachment - computed: false, optional: true, required: false
        this._interconnectAttachment = new NetworkConnectivityPolicyBasedRouteInterconnectAttachmentOutputReference(this, "interconnect_attachment");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkConnectivityPolicyBasedRouteTimeoutsOutputReference(this, "timeouts");
        // virtual_machine - computed: false, optional: true, required: false
        this._virtualMachine = new NetworkConnectivityPolicyBasedRouteVirtualMachineOutputReference(this, "virtual_machine");
        this._description = config.description;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._network = config.network;
        this._nextHopIlbIp = config.nextHopIlbIp;
        this._nextHopOtherRoutes = config.nextHopOtherRoutes;
        this._priority = config.priority;
        this._project = config.project;
        this._filter.internalValue = config.filter;
        this._interconnectAttachment.internalValue = config.interconnectAttachment;
        this._timeouts.internalValue = config.timeouts;
        this._virtualMachine.internalValue = config.virtualMachine;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // kind - computed: true, optional: false, required: false
    get kind() {
        return this.getStringAttribute('kind');
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get network() {
        return this.getStringAttribute('network');
    }
    set network(value) {
        this._network = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkInput() {
        return this._network;
    }
    get nextHopIlbIp() {
        return this.getStringAttribute('next_hop_ilb_ip');
    }
    set nextHopIlbIp(value) {
        this._nextHopIlbIp = value;
    }
    resetNextHopIlbIp() {
        this._nextHopIlbIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nextHopIlbIpInput() {
        return this._nextHopIlbIp;
    }
    get nextHopOtherRoutes() {
        return this.getStringAttribute('next_hop_other_routes');
    }
    set nextHopOtherRoutes(value) {
        this._nextHopOtherRoutes = value;
    }
    resetNextHopOtherRoutes() {
        this._nextHopOtherRoutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nextHopOtherRoutesInput() {
        return this._nextHopOtherRoutes;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    resetPriority() {
        this._priority = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get warnings() {
        return this._warnings;
    }
    get filter() {
        return this._filter;
    }
    putFilter(value) {
        this._filter.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter.internalValue;
    }
    get interconnectAttachment() {
        return this._interconnectAttachment;
    }
    putInterconnectAttachment(value) {
        this._interconnectAttachment.internalValue = value;
    }
    resetInterconnectAttachment() {
        this._interconnectAttachment.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get interconnectAttachmentInput() {
        return this._interconnectAttachment.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get virtualMachine() {
        return this._virtualMachine;
    }
    putVirtualMachine(value) {
        this._virtualMachine.internalValue = value;
    }
    resetVirtualMachine() {
        this._virtualMachine.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get virtualMachineInput() {
        return this._virtualMachine.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            network: cdktf.stringToTerraform(this._network),
            next_hop_ilb_ip: cdktf.stringToTerraform(this._nextHopIlbIp),
            next_hop_other_routes: cdktf.stringToTerraform(this._nextHopOtherRoutes),
            priority: cdktf.numberToTerraform(this._priority),
            project: cdktf.stringToTerraform(this._project),
            filter: networkConnectivityPolicyBasedRouteFilterToTerraform(this._filter.internalValue),
            interconnect_attachment: networkConnectivityPolicyBasedRouteInterconnectAttachmentToTerraform(this._interconnectAttachment.internalValue),
            timeouts: networkConnectivityPolicyBasedRouteTimeoutsToTerraform(this._timeouts.internalValue),
            virtual_machine: networkConnectivityPolicyBasedRouteVirtualMachineToTerraform(this._virtualMachine.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            network: {
                value: cdktf.stringToHclTerraform(this._network),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            next_hop_ilb_ip: {
                value: cdktf.stringToHclTerraform(this._nextHopIlbIp),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            next_hop_other_routes: {
                value: cdktf.stringToHclTerraform(this._nextHopOtherRoutes),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            priority: {
                value: cdktf.numberToHclTerraform(this._priority),
                isBlock: false,
                type: "simple",
                storageClassType: "number",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            filter: {
                value: networkConnectivityPolicyBasedRouteFilterToHclTerraform(this._filter.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivityPolicyBasedRouteFilterList",
            },
            interconnect_attachment: {
                value: networkConnectivityPolicyBasedRouteInterconnectAttachmentToHclTerraform(this._interconnectAttachment.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivityPolicyBasedRouteInterconnectAttachmentList",
            },
            timeouts: {
                value: networkConnectivityPolicyBasedRouteTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkConnectivityPolicyBasedRouteTimeouts",
            },
            virtual_machine: {
                value: networkConnectivityPolicyBasedRouteVirtualMachineToHclTerraform(this._virtualMachine.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkConnectivityPolicyBasedRouteVirtualMachineList",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkConnectivityPolicyBasedRoute = NetworkConnectivityPolicyBasedRoute;
_g = JSII_RTTI_SYMBOL_1;
NetworkConnectivityPolicyBasedRoute[_g] = { fqn: "@cdktf/provider-google.networkConnectivityPolicyBasedRoute.NetworkConnectivityPolicyBasedRoute", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetworkConnectivityPolicyBasedRoute.tfResourceType = "google_network_connectivity_policy_based_route";
