// @cdktf/provider-google/lib/iam-deny-policy/index.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamDenyPolicy = exports.IamDenyPolicyTimeoutsOutputReference = exports.iamDenyPolicyTimeoutsToHclTerraform = exports.iamDenyPolicyTimeoutsToTerraform = exports.IamDenyPolicyRulesList = exports.IamDenyPolicyRulesOutputReference = exports.iamDenyPolicyRulesToHclTerraform = exports.iamDenyPolicyRulesToTerraform = exports.IamDenyPolicyRulesDenyRuleOutputReference = exports.iamDenyPolicyRulesDenyRuleToHclTerraform = exports.iamDenyPolicyRulesDenyRuleToTerraform = exports.IamDenyPolicyRulesDenyRuleDenialConditionOutputReference = exports.iamDenyPolicyRulesDenyRuleDenialConditionToHclTerraform = exports.iamDenyPolicyRulesDenyRuleDenialConditionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function iamDenyPolicyRulesDenyRuleDenialConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        expression: cdktf.stringToTerraform(struct.expression),
        location: cdktf.stringToTerraform(struct.location),
        title: cdktf.stringToTerraform(struct.title),
    };
}
exports.iamDenyPolicyRulesDenyRuleDenialConditionToTerraform = iamDenyPolicyRulesDenyRuleDenialConditionToTerraform;
function iamDenyPolicyRulesDenyRuleDenialConditionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        expression: {
            value: cdktf.stringToHclTerraform(struct.expression),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        location: {
            value: cdktf.stringToHclTerraform(struct.location),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        title: {
            value: cdktf.stringToHclTerraform(struct.title),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamDenyPolicyRulesDenyRuleDenialConditionToHclTerraform = iamDenyPolicyRulesDenyRuleDenialConditionToHclTerraform;
class IamDenyPolicyRulesDenyRuleDenialConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._expression !== undefined) {
            hasAnyValues = true;
            internalValueResult.expression = this._expression;
        }
        if (this._location !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location;
        }
        if (this._title !== undefined) {
            hasAnyValues = true;
            internalValueResult.title = this._title;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._description = undefined;
            this._expression = undefined;
            this._location = undefined;
            this._title = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._description = value.description;
            this._expression = value.expression;
            this._location = value.location;
            this._title = value.title;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get expression() {
        return this.getStringAttribute('expression');
    }
    set expression(value) {
        this._expression = value;
    }
    // Temporarily expose input value. Use with caution.
    get expressionInput() {
        return this._expression;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get title() {
        return this.getStringAttribute('title');
    }
    set title(value) {
        this._title = value;
    }
    resetTitle() {
        this._title = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get titleInput() {
        return this._title;
    }
}
exports.IamDenyPolicyRulesDenyRuleDenialConditionOutputReference = IamDenyPolicyRulesDenyRuleDenialConditionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
IamDenyPolicyRulesDenyRuleDenialConditionOutputReference[_a] = { fqn: "@cdktf/provider-google.iamDenyPolicy.IamDenyPolicyRulesDenyRuleDenialConditionOutputReference", version: "14.13.1" };
function iamDenyPolicyRulesDenyRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        denied_permissions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.deniedPermissions),
        denied_principals: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.deniedPrincipals),
        exception_permissions: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.exceptionPermissions),
        exception_principals: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.exceptionPrincipals),
        denial_condition: iamDenyPolicyRulesDenyRuleDenialConditionToTerraform(struct.denialCondition),
    };
}
exports.iamDenyPolicyRulesDenyRuleToTerraform = iamDenyPolicyRulesDenyRuleToTerraform;
function iamDenyPolicyRulesDenyRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        denied_permissions: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.deniedPermissions),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        denied_principals: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.deniedPrincipals),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        exception_permissions: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.exceptionPermissions),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        exception_principals: {
            value: cdktf.listMapperHcl(cdktf.stringToHclTerraform, false)(struct.exceptionPrincipals),
            isBlock: false,
            type: "list",
            storageClassType: "stringList",
        },
        denial_condition: {
            value: iamDenyPolicyRulesDenyRuleDenialConditionToHclTerraform(struct.denialCondition),
            isBlock: true,
            type: "list",
            storageClassType: "IamDenyPolicyRulesDenyRuleDenialConditionList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamDenyPolicyRulesDenyRuleToHclTerraform = iamDenyPolicyRulesDenyRuleToHclTerraform;
class IamDenyPolicyRulesDenyRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // denial_condition - computed: false, optional: true, required: false
        this._denialCondition = new IamDenyPolicyRulesDenyRuleDenialConditionOutputReference(this, "denial_condition");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._deniedPermissions !== undefined) {
            hasAnyValues = true;
            internalValueResult.deniedPermissions = this._deniedPermissions;
        }
        if (this._deniedPrincipals !== undefined) {
            hasAnyValues = true;
            internalValueResult.deniedPrincipals = this._deniedPrincipals;
        }
        if (this._exceptionPermissions !== undefined) {
            hasAnyValues = true;
            internalValueResult.exceptionPermissions = this._exceptionPermissions;
        }
        if (this._exceptionPrincipals !== undefined) {
            hasAnyValues = true;
            internalValueResult.exceptionPrincipals = this._exceptionPrincipals;
        }
        if (this._denialCondition?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.denialCondition = this._denialCondition?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._deniedPermissions = undefined;
            this._deniedPrincipals = undefined;
            this._exceptionPermissions = undefined;
            this._exceptionPrincipals = undefined;
            this._denialCondition.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._deniedPermissions = value.deniedPermissions;
            this._deniedPrincipals = value.deniedPrincipals;
            this._exceptionPermissions = value.exceptionPermissions;
            this._exceptionPrincipals = value.exceptionPrincipals;
            this._denialCondition.internalValue = value.denialCondition;
        }
    }
    get deniedPermissions() {
        return this.getListAttribute('denied_permissions');
    }
    set deniedPermissions(value) {
        this._deniedPermissions = value;
    }
    resetDeniedPermissions() {
        this._deniedPermissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deniedPermissionsInput() {
        return this._deniedPermissions;
    }
    get deniedPrincipals() {
        return this.getListAttribute('denied_principals');
    }
    set deniedPrincipals(value) {
        this._deniedPrincipals = value;
    }
    resetDeniedPrincipals() {
        this._deniedPrincipals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deniedPrincipalsInput() {
        return this._deniedPrincipals;
    }
    get exceptionPermissions() {
        return this.getListAttribute('exception_permissions');
    }
    set exceptionPermissions(value) {
        this._exceptionPermissions = value;
    }
    resetExceptionPermissions() {
        this._exceptionPermissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exceptionPermissionsInput() {
        return this._exceptionPermissions;
    }
    get exceptionPrincipals() {
        return this.getListAttribute('exception_principals');
    }
    set exceptionPrincipals(value) {
        this._exceptionPrincipals = value;
    }
    resetExceptionPrincipals() {
        this._exceptionPrincipals = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get exceptionPrincipalsInput() {
        return this._exceptionPrincipals;
    }
    get denialCondition() {
        return this._denialCondition;
    }
    putDenialCondition(value) {
        this._denialCondition.internalValue = value;
    }
    resetDenialCondition() {
        this._denialCondition.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get denialConditionInput() {
        return this._denialCondition.internalValue;
    }
}
exports.IamDenyPolicyRulesDenyRuleOutputReference = IamDenyPolicyRulesDenyRuleOutputReference;
_b = JSII_RTTI_SYMBOL_1;
IamDenyPolicyRulesDenyRuleOutputReference[_b] = { fqn: "@cdktf/provider-google.iamDenyPolicy.IamDenyPolicyRulesDenyRuleOutputReference", version: "14.13.1" };
function iamDenyPolicyRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        deny_rule: iamDenyPolicyRulesDenyRuleToTerraform(struct.denyRule),
    };
}
exports.iamDenyPolicyRulesToTerraform = iamDenyPolicyRulesToTerraform;
function iamDenyPolicyRulesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        description: {
            value: cdktf.stringToHclTerraform(struct.description),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        deny_rule: {
            value: iamDenyPolicyRulesDenyRuleToHclTerraform(struct.denyRule),
            isBlock: true,
            type: "list",
            storageClassType: "IamDenyPolicyRulesDenyRuleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamDenyPolicyRulesToHclTerraform = iamDenyPolicyRulesToHclTerraform;
class IamDenyPolicyRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // deny_rule - computed: false, optional: true, required: false
        this._denyRule = new IamDenyPolicyRulesDenyRuleOutputReference(this, "deny_rule");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._denyRule?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.denyRule = this._denyRule?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._denyRule.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._denyRule.internalValue = value.denyRule;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get denyRule() {
        return this._denyRule;
    }
    putDenyRule(value) {
        this._denyRule.internalValue = value;
    }
    resetDenyRule() {
        this._denyRule.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get denyRuleInput() {
        return this._denyRule.internalValue;
    }
}
exports.IamDenyPolicyRulesOutputReference = IamDenyPolicyRulesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
IamDenyPolicyRulesOutputReference[_c] = { fqn: "@cdktf/provider-google.iamDenyPolicy.IamDenyPolicyRulesOutputReference", version: "14.13.1" };
class IamDenyPolicyRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new IamDenyPolicyRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.IamDenyPolicyRulesList = IamDenyPolicyRulesList;
_d = JSII_RTTI_SYMBOL_1;
IamDenyPolicyRulesList[_d] = { fqn: "@cdktf/provider-google.iamDenyPolicy.IamDenyPolicyRulesList", version: "14.13.1" };
function iamDenyPolicyTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.iamDenyPolicyTimeoutsToTerraform = iamDenyPolicyTimeoutsToTerraform;
function iamDenyPolicyTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.iamDenyPolicyTimeoutsToHclTerraform = iamDenyPolicyTimeoutsToHclTerraform;
class IamDenyPolicyTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.IamDenyPolicyTimeoutsOutputReference = IamDenyPolicyTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
IamDenyPolicyTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-google.iamDenyPolicy.IamDenyPolicyTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_deny_policy google_iam_deny_policy}
*/
class IamDenyPolicy extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a IamDenyPolicy resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the IamDenyPolicy to import
    * @param importFromId The id of the existing IamDenyPolicy that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_deny_policy#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the IamDenyPolicy to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_iam_deny_policy", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/iam_deny_policy google_iam_deny_policy} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamDenyPolicyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_iam_deny_policy',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // rules - computed: false, optional: false, required: true
        this._rules = new IamDenyPolicyRulesList(this, "rules", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new IamDenyPolicyTimeoutsOutputReference(this, "timeouts");
        this._displayName = config.displayName;
        this._id = config.id;
        this._name = config.name;
        this._parent = config.parent;
        this._rules.internalValue = config.rules;
        this._timeouts.internalValue = config.timeouts;
    }
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    set displayName(value) {
        this._displayName = value;
    }
    resetDisplayName() {
        this._displayName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get displayNameInput() {
        return this._displayName;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    get rules() {
        return this._rules;
    }
    putRules(value) {
        this._rules.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get rulesInput() {
        return this._rules.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            display_name: cdktf.stringToTerraform(this._displayName),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            parent: cdktf.stringToTerraform(this._parent),
            rules: cdktf.listMapper(iamDenyPolicyRulesToTerraform, true)(this._rules.internalValue),
            timeouts: iamDenyPolicyTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            display_name: {
                value: cdktf.stringToHclTerraform(this._displayName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent: {
                value: cdktf.stringToHclTerraform(this._parent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            rules: {
                value: cdktf.listMapperHcl(iamDenyPolicyRulesToHclTerraform, true)(this._rules.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "IamDenyPolicyRulesList",
            },
            timeouts: {
                value: iamDenyPolicyTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "IamDenyPolicyTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.IamDenyPolicy = IamDenyPolicy;
_f = JSII_RTTI_SYMBOL_1;
IamDenyPolicy[_f] = { fqn: "@cdktf/provider-google.iamDenyPolicy.IamDenyPolicy", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
IamDenyPolicy.tfResourceType = "google_iam_deny_policy";
