// @cdktf/provider-google/lib/bigquery-connection/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BigqueryConnection = exports.BigqueryConnectionTimeoutsOutputReference = exports.bigqueryConnectionTimeoutsToHclTerraform = exports.bigqueryConnectionTimeoutsToTerraform = exports.BigqueryConnectionSparkOutputReference = exports.bigqueryConnectionSparkToHclTerraform = exports.bigqueryConnectionSparkToTerraform = exports.BigqueryConnectionSparkSparkHistoryServerConfigOutputReference = exports.bigqueryConnectionSparkSparkHistoryServerConfigToHclTerraform = exports.bigqueryConnectionSparkSparkHistoryServerConfigToTerraform = exports.BigqueryConnectionSparkMetastoreServiceConfigOutputReference = exports.bigqueryConnectionSparkMetastoreServiceConfigToHclTerraform = exports.bigqueryConnectionSparkMetastoreServiceConfigToTerraform = exports.BigqueryConnectionCloudSqlOutputReference = exports.bigqueryConnectionCloudSqlToHclTerraform = exports.bigqueryConnectionCloudSqlToTerraform = exports.BigqueryConnectionCloudSqlCredentialOutputReference = exports.bigqueryConnectionCloudSqlCredentialToHclTerraform = exports.bigqueryConnectionCloudSqlCredentialToTerraform = exports.BigqueryConnectionCloudSpannerOutputReference = exports.bigqueryConnectionCloudSpannerToHclTerraform = exports.bigqueryConnectionCloudSpannerToTerraform = exports.BigqueryConnectionCloudResourceOutputReference = exports.bigqueryConnectionCloudResourceToHclTerraform = exports.bigqueryConnectionCloudResourceToTerraform = exports.BigqueryConnectionAzureOutputReference = exports.bigqueryConnectionAzureToHclTerraform = exports.bigqueryConnectionAzureToTerraform = exports.BigqueryConnectionAwsOutputReference = exports.bigqueryConnectionAwsToHclTerraform = exports.bigqueryConnectionAwsToTerraform = exports.BigqueryConnectionAwsAccessRoleOutputReference = exports.bigqueryConnectionAwsAccessRoleToHclTerraform = exports.bigqueryConnectionAwsAccessRoleToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function bigqueryConnectionAwsAccessRoleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        iam_role_id: cdktf.stringToTerraform(struct.iamRoleId),
    };
}
exports.bigqueryConnectionAwsAccessRoleToTerraform = bigqueryConnectionAwsAccessRoleToTerraform;
function bigqueryConnectionAwsAccessRoleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        iam_role_id: {
            value: cdktf.stringToHclTerraform(struct.iamRoleId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionAwsAccessRoleToHclTerraform = bigqueryConnectionAwsAccessRoleToHclTerraform;
class BigqueryConnectionAwsAccessRoleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._iamRoleId !== undefined) {
            hasAnyValues = true;
            internalValueResult.iamRoleId = this._iamRoleId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._iamRoleId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._iamRoleId = value.iamRoleId;
        }
    }
    get iamRoleId() {
        return this.getStringAttribute('iam_role_id');
    }
    set iamRoleId(value) {
        this._iamRoleId = value;
    }
    // Temporarily expose input value. Use with caution.
    get iamRoleIdInput() {
        return this._iamRoleId;
    }
    // identity - computed: true, optional: false, required: false
    get identity() {
        return this.getStringAttribute('identity');
    }
}
exports.BigqueryConnectionAwsAccessRoleOutputReference = BigqueryConnectionAwsAccessRoleOutputReference;
_a = JSII_RTTI_SYMBOL_1;
BigqueryConnectionAwsAccessRoleOutputReference[_a] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionAwsAccessRoleOutputReference", version: "14.13.1" };
function bigqueryConnectionAwsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        access_role: bigqueryConnectionAwsAccessRoleToTerraform(struct.accessRole),
    };
}
exports.bigqueryConnectionAwsToTerraform = bigqueryConnectionAwsToTerraform;
function bigqueryConnectionAwsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        access_role: {
            value: bigqueryConnectionAwsAccessRoleToHclTerraform(struct.accessRole),
            isBlock: true,
            type: "list",
            storageClassType: "BigqueryConnectionAwsAccessRoleList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionAwsToHclTerraform = bigqueryConnectionAwsToHclTerraform;
class BigqueryConnectionAwsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // access_role - computed: false, optional: false, required: true
        this._accessRole = new BigqueryConnectionAwsAccessRoleOutputReference(this, "access_role");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._accessRole?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.accessRole = this._accessRole?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._accessRole.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._accessRole.internalValue = value.accessRole;
        }
    }
    get accessRole() {
        return this._accessRole;
    }
    putAccessRole(value) {
        this._accessRole.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get accessRoleInput() {
        return this._accessRole.internalValue;
    }
}
exports.BigqueryConnectionAwsOutputReference = BigqueryConnectionAwsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
BigqueryConnectionAwsOutputReference[_b] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionAwsOutputReference", version: "14.13.1" };
function bigqueryConnectionAzureToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        customer_tenant_id: cdktf.stringToTerraform(struct.customerTenantId),
        federated_application_client_id: cdktf.stringToTerraform(struct.federatedApplicationClientId),
    };
}
exports.bigqueryConnectionAzureToTerraform = bigqueryConnectionAzureToTerraform;
function bigqueryConnectionAzureToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        customer_tenant_id: {
            value: cdktf.stringToHclTerraform(struct.customerTenantId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        federated_application_client_id: {
            value: cdktf.stringToHclTerraform(struct.federatedApplicationClientId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionAzureToHclTerraform = bigqueryConnectionAzureToHclTerraform;
class BigqueryConnectionAzureOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._customerTenantId !== undefined) {
            hasAnyValues = true;
            internalValueResult.customerTenantId = this._customerTenantId;
        }
        if (this._federatedApplicationClientId !== undefined) {
            hasAnyValues = true;
            internalValueResult.federatedApplicationClientId = this._federatedApplicationClientId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._customerTenantId = undefined;
            this._federatedApplicationClientId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._customerTenantId = value.customerTenantId;
            this._federatedApplicationClientId = value.federatedApplicationClientId;
        }
    }
    // application - computed: true, optional: false, required: false
    get application() {
        return this.getStringAttribute('application');
    }
    // client_id - computed: true, optional: false, required: false
    get clientId() {
        return this.getStringAttribute('client_id');
    }
    get customerTenantId() {
        return this.getStringAttribute('customer_tenant_id');
    }
    set customerTenantId(value) {
        this._customerTenantId = value;
    }
    // Temporarily expose input value. Use with caution.
    get customerTenantIdInput() {
        return this._customerTenantId;
    }
    get federatedApplicationClientId() {
        return this.getStringAttribute('federated_application_client_id');
    }
    set federatedApplicationClientId(value) {
        this._federatedApplicationClientId = value;
    }
    resetFederatedApplicationClientId() {
        this._federatedApplicationClientId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get federatedApplicationClientIdInput() {
        return this._federatedApplicationClientId;
    }
    // identity - computed: true, optional: false, required: false
    get identity() {
        return this.getStringAttribute('identity');
    }
    // object_id - computed: true, optional: false, required: false
    get objectId() {
        return this.getStringAttribute('object_id');
    }
    // redirect_uri - computed: true, optional: false, required: false
    get redirectUri() {
        return this.getStringAttribute('redirect_uri');
    }
}
exports.BigqueryConnectionAzureOutputReference = BigqueryConnectionAzureOutputReference;
_c = JSII_RTTI_SYMBOL_1;
BigqueryConnectionAzureOutputReference[_c] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionAzureOutputReference", version: "14.13.1" };
function bigqueryConnectionCloudResourceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.bigqueryConnectionCloudResourceToTerraform = bigqueryConnectionCloudResourceToTerraform;
function bigqueryConnectionCloudResourceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.bigqueryConnectionCloudResourceToHclTerraform = bigqueryConnectionCloudResourceToHclTerraform;
class BigqueryConnectionCloudResourceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
}
exports.BigqueryConnectionCloudResourceOutputReference = BigqueryConnectionCloudResourceOutputReference;
_d = JSII_RTTI_SYMBOL_1;
BigqueryConnectionCloudResourceOutputReference[_d] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionCloudResourceOutputReference", version: "14.13.1" };
function bigqueryConnectionCloudSpannerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database: cdktf.stringToTerraform(struct.database),
        database_role: cdktf.stringToTerraform(struct.databaseRole),
        max_parallelism: cdktf.numberToTerraform(struct.maxParallelism),
        use_data_boost: cdktf.booleanToTerraform(struct.useDataBoost),
        use_parallelism: cdktf.booleanToTerraform(struct.useParallelism),
        use_serverless_analytics: cdktf.booleanToTerraform(struct.useServerlessAnalytics),
    };
}
exports.bigqueryConnectionCloudSpannerToTerraform = bigqueryConnectionCloudSpannerToTerraform;
function bigqueryConnectionCloudSpannerToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        database: {
            value: cdktf.stringToHclTerraform(struct.database),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        database_role: {
            value: cdktf.stringToHclTerraform(struct.databaseRole),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        max_parallelism: {
            value: cdktf.numberToHclTerraform(struct.maxParallelism),
            isBlock: false,
            type: "simple",
            storageClassType: "number",
        },
        use_data_boost: {
            value: cdktf.booleanToHclTerraform(struct.useDataBoost),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        use_parallelism: {
            value: cdktf.booleanToHclTerraform(struct.useParallelism),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
        use_serverless_analytics: {
            value: cdktf.booleanToHclTerraform(struct.useServerlessAnalytics),
            isBlock: false,
            type: "simple",
            storageClassType: "boolean",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionCloudSpannerToHclTerraform = bigqueryConnectionCloudSpannerToHclTerraform;
class BigqueryConnectionCloudSpannerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._database !== undefined) {
            hasAnyValues = true;
            internalValueResult.database = this._database;
        }
        if (this._databaseRole !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseRole = this._databaseRole;
        }
        if (this._maxParallelism !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxParallelism = this._maxParallelism;
        }
        if (this._useDataBoost !== undefined) {
            hasAnyValues = true;
            internalValueResult.useDataBoost = this._useDataBoost;
        }
        if (this._useParallelism !== undefined) {
            hasAnyValues = true;
            internalValueResult.useParallelism = this._useParallelism;
        }
        if (this._useServerlessAnalytics !== undefined) {
            hasAnyValues = true;
            internalValueResult.useServerlessAnalytics = this._useServerlessAnalytics;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._database = undefined;
            this._databaseRole = undefined;
            this._maxParallelism = undefined;
            this._useDataBoost = undefined;
            this._useParallelism = undefined;
            this._useServerlessAnalytics = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._database = value.database;
            this._databaseRole = value.databaseRole;
            this._maxParallelism = value.maxParallelism;
            this._useDataBoost = value.useDataBoost;
            this._useParallelism = value.useParallelism;
            this._useServerlessAnalytics = value.useServerlessAnalytics;
        }
    }
    get database() {
        return this.getStringAttribute('database');
    }
    set database(value) {
        this._database = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database;
    }
    get databaseRole() {
        return this.getStringAttribute('database_role');
    }
    set databaseRole(value) {
        this._databaseRole = value;
    }
    resetDatabaseRole() {
        this._databaseRole = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseRoleInput() {
        return this._databaseRole;
    }
    get maxParallelism() {
        return this.getNumberAttribute('max_parallelism');
    }
    set maxParallelism(value) {
        this._maxParallelism = value;
    }
    resetMaxParallelism() {
        this._maxParallelism = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxParallelismInput() {
        return this._maxParallelism;
    }
    get useDataBoost() {
        return this.getBooleanAttribute('use_data_boost');
    }
    set useDataBoost(value) {
        this._useDataBoost = value;
    }
    resetUseDataBoost() {
        this._useDataBoost = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useDataBoostInput() {
        return this._useDataBoost;
    }
    get useParallelism() {
        return this.getBooleanAttribute('use_parallelism');
    }
    set useParallelism(value) {
        this._useParallelism = value;
    }
    resetUseParallelism() {
        this._useParallelism = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useParallelismInput() {
        return this._useParallelism;
    }
    get useServerlessAnalytics() {
        return this.getBooleanAttribute('use_serverless_analytics');
    }
    set useServerlessAnalytics(value) {
        this._useServerlessAnalytics = value;
    }
    resetUseServerlessAnalytics() {
        this._useServerlessAnalytics = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useServerlessAnalyticsInput() {
        return this._useServerlessAnalytics;
    }
}
exports.BigqueryConnectionCloudSpannerOutputReference = BigqueryConnectionCloudSpannerOutputReference;
_e = JSII_RTTI_SYMBOL_1;
BigqueryConnectionCloudSpannerOutputReference[_e] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionCloudSpannerOutputReference", version: "14.13.1" };
function bigqueryConnectionCloudSqlCredentialToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        password: cdktf.stringToTerraform(struct.password),
        username: cdktf.stringToTerraform(struct.username),
    };
}
exports.bigqueryConnectionCloudSqlCredentialToTerraform = bigqueryConnectionCloudSqlCredentialToTerraform;
function bigqueryConnectionCloudSqlCredentialToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        password: {
            value: cdktf.stringToHclTerraform(struct.password),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        username: {
            value: cdktf.stringToHclTerraform(struct.username),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionCloudSqlCredentialToHclTerraform = bigqueryConnectionCloudSqlCredentialToHclTerraform;
class BigqueryConnectionCloudSqlCredentialOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._password !== undefined) {
            hasAnyValues = true;
            internalValueResult.password = this._password;
        }
        if (this._username !== undefined) {
            hasAnyValues = true;
            internalValueResult.username = this._username;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._password = undefined;
            this._username = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._password = value.password;
            this._username = value.username;
        }
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    get username() {
        return this.getStringAttribute('username');
    }
    set username(value) {
        this._username = value;
    }
    // Temporarily expose input value. Use with caution.
    get usernameInput() {
        return this._username;
    }
}
exports.BigqueryConnectionCloudSqlCredentialOutputReference = BigqueryConnectionCloudSqlCredentialOutputReference;
_f = JSII_RTTI_SYMBOL_1;
BigqueryConnectionCloudSqlCredentialOutputReference[_f] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionCloudSqlCredentialOutputReference", version: "14.13.1" };
function bigqueryConnectionCloudSqlToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database: cdktf.stringToTerraform(struct.database),
        instance_id: cdktf.stringToTerraform(struct.instanceId),
        type: cdktf.stringToTerraform(struct.type),
        credential: bigqueryConnectionCloudSqlCredentialToTerraform(struct.credential),
    };
}
exports.bigqueryConnectionCloudSqlToTerraform = bigqueryConnectionCloudSqlToTerraform;
function bigqueryConnectionCloudSqlToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        database: {
            value: cdktf.stringToHclTerraform(struct.database),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        instance_id: {
            value: cdktf.stringToHclTerraform(struct.instanceId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        type: {
            value: cdktf.stringToHclTerraform(struct.type),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        credential: {
            value: bigqueryConnectionCloudSqlCredentialToHclTerraform(struct.credential),
            isBlock: true,
            type: "list",
            storageClassType: "BigqueryConnectionCloudSqlCredentialList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionCloudSqlToHclTerraform = bigqueryConnectionCloudSqlToHclTerraform;
class BigqueryConnectionCloudSqlOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // credential - computed: false, optional: false, required: true
        this._credential = new BigqueryConnectionCloudSqlCredentialOutputReference(this, "credential");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._database !== undefined) {
            hasAnyValues = true;
            internalValueResult.database = this._database;
        }
        if (this._instanceId !== undefined) {
            hasAnyValues = true;
            internalValueResult.instanceId = this._instanceId;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._credential?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.credential = this._credential?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._database = undefined;
            this._instanceId = undefined;
            this._type = undefined;
            this._credential.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._database = value.database;
            this._instanceId = value.instanceId;
            this._type = value.type;
            this._credential.internalValue = value.credential;
        }
    }
    get database() {
        return this.getStringAttribute('database');
    }
    set database(value) {
        this._database = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database;
    }
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    set instanceId(value) {
        this._instanceId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceIdInput() {
        return this._instanceId;
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get credential() {
        return this._credential;
    }
    putCredential(value) {
        this._credential.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get credentialInput() {
        return this._credential.internalValue;
    }
}
exports.BigqueryConnectionCloudSqlOutputReference = BigqueryConnectionCloudSqlOutputReference;
_g = JSII_RTTI_SYMBOL_1;
BigqueryConnectionCloudSqlOutputReference[_g] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionCloudSqlOutputReference", version: "14.13.1" };
function bigqueryConnectionSparkMetastoreServiceConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metastore_service: cdktf.stringToTerraform(struct.metastoreService),
    };
}
exports.bigqueryConnectionSparkMetastoreServiceConfigToTerraform = bigqueryConnectionSparkMetastoreServiceConfigToTerraform;
function bigqueryConnectionSparkMetastoreServiceConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        metastore_service: {
            value: cdktf.stringToHclTerraform(struct.metastoreService),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionSparkMetastoreServiceConfigToHclTerraform = bigqueryConnectionSparkMetastoreServiceConfigToHclTerraform;
class BigqueryConnectionSparkMetastoreServiceConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metastoreService !== undefined) {
            hasAnyValues = true;
            internalValueResult.metastoreService = this._metastoreService;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._metastoreService = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._metastoreService = value.metastoreService;
        }
    }
    get metastoreService() {
        return this.getStringAttribute('metastore_service');
    }
    set metastoreService(value) {
        this._metastoreService = value;
    }
    resetMetastoreService() {
        this._metastoreService = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metastoreServiceInput() {
        return this._metastoreService;
    }
}
exports.BigqueryConnectionSparkMetastoreServiceConfigOutputReference = BigqueryConnectionSparkMetastoreServiceConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
BigqueryConnectionSparkMetastoreServiceConfigOutputReference[_h] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionSparkMetastoreServiceConfigOutputReference", version: "14.13.1" };
function bigqueryConnectionSparkSparkHistoryServerConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dataproc_cluster: cdktf.stringToTerraform(struct.dataprocCluster),
    };
}
exports.bigqueryConnectionSparkSparkHistoryServerConfigToTerraform = bigqueryConnectionSparkSparkHistoryServerConfigToTerraform;
function bigqueryConnectionSparkSparkHistoryServerConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        dataproc_cluster: {
            value: cdktf.stringToHclTerraform(struct.dataprocCluster),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionSparkSparkHistoryServerConfigToHclTerraform = bigqueryConnectionSparkSparkHistoryServerConfigToHclTerraform;
class BigqueryConnectionSparkSparkHistoryServerConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataprocCluster !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataprocCluster = this._dataprocCluster;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataprocCluster = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataprocCluster = value.dataprocCluster;
        }
    }
    get dataprocCluster() {
        return this.getStringAttribute('dataproc_cluster');
    }
    set dataprocCluster(value) {
        this._dataprocCluster = value;
    }
    resetDataprocCluster() {
        this._dataprocCluster = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataprocClusterInput() {
        return this._dataprocCluster;
    }
}
exports.BigqueryConnectionSparkSparkHistoryServerConfigOutputReference = BigqueryConnectionSparkSparkHistoryServerConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
BigqueryConnectionSparkSparkHistoryServerConfigOutputReference[_j] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionSparkSparkHistoryServerConfigOutputReference", version: "14.13.1" };
function bigqueryConnectionSparkToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        metastore_service_config: bigqueryConnectionSparkMetastoreServiceConfigToTerraform(struct.metastoreServiceConfig),
        spark_history_server_config: bigqueryConnectionSparkSparkHistoryServerConfigToTerraform(struct.sparkHistoryServerConfig),
    };
}
exports.bigqueryConnectionSparkToTerraform = bigqueryConnectionSparkToTerraform;
function bigqueryConnectionSparkToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        metastore_service_config: {
            value: bigqueryConnectionSparkMetastoreServiceConfigToHclTerraform(struct.metastoreServiceConfig),
            isBlock: true,
            type: "list",
            storageClassType: "BigqueryConnectionSparkMetastoreServiceConfigList",
        },
        spark_history_server_config: {
            value: bigqueryConnectionSparkSparkHistoryServerConfigToHclTerraform(struct.sparkHistoryServerConfig),
            isBlock: true,
            type: "list",
            storageClassType: "BigqueryConnectionSparkSparkHistoryServerConfigList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionSparkToHclTerraform = bigqueryConnectionSparkToHclTerraform;
class BigqueryConnectionSparkOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // metastore_service_config - computed: false, optional: true, required: false
        this._metastoreServiceConfig = new BigqueryConnectionSparkMetastoreServiceConfigOutputReference(this, "metastore_service_config");
        // spark_history_server_config - computed: false, optional: true, required: false
        this._sparkHistoryServerConfig = new BigqueryConnectionSparkSparkHistoryServerConfigOutputReference(this, "spark_history_server_config");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._metastoreServiceConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.metastoreServiceConfig = this._metastoreServiceConfig?.internalValue;
        }
        if (this._sparkHistoryServerConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.sparkHistoryServerConfig = this._sparkHistoryServerConfig?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._metastoreServiceConfig.internalValue = undefined;
            this._sparkHistoryServerConfig.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._metastoreServiceConfig.internalValue = value.metastoreServiceConfig;
            this._sparkHistoryServerConfig.internalValue = value.sparkHistoryServerConfig;
        }
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    get metastoreServiceConfig() {
        return this._metastoreServiceConfig;
    }
    putMetastoreServiceConfig(value) {
        this._metastoreServiceConfig.internalValue = value;
    }
    resetMetastoreServiceConfig() {
        this._metastoreServiceConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metastoreServiceConfigInput() {
        return this._metastoreServiceConfig.internalValue;
    }
    get sparkHistoryServerConfig() {
        return this._sparkHistoryServerConfig;
    }
    putSparkHistoryServerConfig(value) {
        this._sparkHistoryServerConfig.internalValue = value;
    }
    resetSparkHistoryServerConfig() {
        this._sparkHistoryServerConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sparkHistoryServerConfigInput() {
        return this._sparkHistoryServerConfig.internalValue;
    }
}
exports.BigqueryConnectionSparkOutputReference = BigqueryConnectionSparkOutputReference;
_k = JSII_RTTI_SYMBOL_1;
BigqueryConnectionSparkOutputReference[_k] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionSparkOutputReference", version: "14.13.1" };
function bigqueryConnectionTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.bigqueryConnectionTimeoutsToTerraform = bigqueryConnectionTimeoutsToTerraform;
function bigqueryConnectionTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.bigqueryConnectionTimeoutsToHclTerraform = bigqueryConnectionTimeoutsToHclTerraform;
class BigqueryConnectionTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.BigqueryConnectionTimeoutsOutputReference = BigqueryConnectionTimeoutsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
BigqueryConnectionTimeoutsOutputReference[_l] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnectionTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_connection google_bigquery_connection}
*/
class BigqueryConnection extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a BigqueryConnection resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the BigqueryConnection to import
    * @param importFromId The id of the existing BigqueryConnection that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_connection#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the BigqueryConnection to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_bigquery_connection", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/bigquery_connection google_bigquery_connection} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options BigqueryConnectionConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'google_bigquery_connection',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // aws - computed: false, optional: true, required: false
        this._aws = new BigqueryConnectionAwsOutputReference(this, "aws");
        // azure - computed: false, optional: true, required: false
        this._azure = new BigqueryConnectionAzureOutputReference(this, "azure");
        // cloud_resource - computed: false, optional: true, required: false
        this._cloudResource = new BigqueryConnectionCloudResourceOutputReference(this, "cloud_resource");
        // cloud_spanner - computed: false, optional: true, required: false
        this._cloudSpanner = new BigqueryConnectionCloudSpannerOutputReference(this, "cloud_spanner");
        // cloud_sql - computed: false, optional: true, required: false
        this._cloudSql = new BigqueryConnectionCloudSqlOutputReference(this, "cloud_sql");
        // spark - computed: false, optional: true, required: false
        this._spark = new BigqueryConnectionSparkOutputReference(this, "spark");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new BigqueryConnectionTimeoutsOutputReference(this, "timeouts");
        this._connectionId = config.connectionId;
        this._description = config.description;
        this._friendlyName = config.friendlyName;
        this._id = config.id;
        this._kmsKeyName = config.kmsKeyName;
        this._location = config.location;
        this._project = config.project;
        this._aws.internalValue = config.aws;
        this._azure.internalValue = config.azure;
        this._cloudResource.internalValue = config.cloudResource;
        this._cloudSpanner.internalValue = config.cloudSpanner;
        this._cloudSql.internalValue = config.cloudSql;
        this._spark.internalValue = config.spark;
        this._timeouts.internalValue = config.timeouts;
    }
    get connectionId() {
        return this.getStringAttribute('connection_id');
    }
    set connectionId(value) {
        this._connectionId = value;
    }
    resetConnectionId() {
        this._connectionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionIdInput() {
        return this._connectionId;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get friendlyName() {
        return this.getStringAttribute('friendly_name');
    }
    set friendlyName(value) {
        this._friendlyName = value;
    }
    resetFriendlyName() {
        this._friendlyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get friendlyNameInput() {
        return this._friendlyName;
    }
    // has_credential - computed: true, optional: false, required: false
    get hasCredential() {
        return this.getBooleanAttribute('has_credential');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get kmsKeyName() {
        return this.getStringAttribute('kms_key_name');
    }
    set kmsKeyName(value) {
        this._kmsKeyName = value;
    }
    resetKmsKeyName() {
        this._kmsKeyName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsKeyNameInput() {
        return this._kmsKeyName;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get aws() {
        return this._aws;
    }
    putAws(value) {
        this._aws.internalValue = value;
    }
    resetAws() {
        this._aws.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get awsInput() {
        return this._aws.internalValue;
    }
    get azure() {
        return this._azure;
    }
    putAzure(value) {
        this._azure.internalValue = value;
    }
    resetAzure() {
        this._azure.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get azureInput() {
        return this._azure.internalValue;
    }
    get cloudResource() {
        return this._cloudResource;
    }
    putCloudResource(value) {
        this._cloudResource.internalValue = value;
    }
    resetCloudResource() {
        this._cloudResource.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudResourceInput() {
        return this._cloudResource.internalValue;
    }
    get cloudSpanner() {
        return this._cloudSpanner;
    }
    putCloudSpanner(value) {
        this._cloudSpanner.internalValue = value;
    }
    resetCloudSpanner() {
        this._cloudSpanner.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudSpannerInput() {
        return this._cloudSpanner.internalValue;
    }
    get cloudSql() {
        return this._cloudSql;
    }
    putCloudSql(value) {
        this._cloudSql.internalValue = value;
    }
    resetCloudSql() {
        this._cloudSql.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cloudSqlInput() {
        return this._cloudSql.internalValue;
    }
    get spark() {
        return this._spark;
    }
    putSpark(value) {
        this._spark.internalValue = value;
    }
    resetSpark() {
        this._spark.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sparkInput() {
        return this._spark.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            connection_id: cdktf.stringToTerraform(this._connectionId),
            description: cdktf.stringToTerraform(this._description),
            friendly_name: cdktf.stringToTerraform(this._friendlyName),
            id: cdktf.stringToTerraform(this._id),
            kms_key_name: cdktf.stringToTerraform(this._kmsKeyName),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
            aws: bigqueryConnectionAwsToTerraform(this._aws.internalValue),
            azure: bigqueryConnectionAzureToTerraform(this._azure.internalValue),
            cloud_resource: bigqueryConnectionCloudResourceToTerraform(this._cloudResource.internalValue),
            cloud_spanner: bigqueryConnectionCloudSpannerToTerraform(this._cloudSpanner.internalValue),
            cloud_sql: bigqueryConnectionCloudSqlToTerraform(this._cloudSql.internalValue),
            spark: bigqueryConnectionSparkToTerraform(this._spark.internalValue),
            timeouts: bigqueryConnectionTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            connection_id: {
                value: cdktf.stringToHclTerraform(this._connectionId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            friendly_name: {
                value: cdktf.stringToHclTerraform(this._friendlyName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            kms_key_name: {
                value: cdktf.stringToHclTerraform(this._kmsKeyName),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            aws: {
                value: bigqueryConnectionAwsToHclTerraform(this._aws.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryConnectionAwsList",
            },
            azure: {
                value: bigqueryConnectionAzureToHclTerraform(this._azure.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryConnectionAzureList",
            },
            cloud_resource: {
                value: bigqueryConnectionCloudResourceToHclTerraform(this._cloudResource.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryConnectionCloudResourceList",
            },
            cloud_spanner: {
                value: bigqueryConnectionCloudSpannerToHclTerraform(this._cloudSpanner.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryConnectionCloudSpannerList",
            },
            cloud_sql: {
                value: bigqueryConnectionCloudSqlToHclTerraform(this._cloudSql.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryConnectionCloudSqlList",
            },
            spark: {
                value: bigqueryConnectionSparkToHclTerraform(this._spark.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "BigqueryConnectionSparkList",
            },
            timeouts: {
                value: bigqueryConnectionTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "BigqueryConnectionTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.BigqueryConnection = BigqueryConnection;
_m = JSII_RTTI_SYMBOL_1;
BigqueryConnection[_m] = { fqn: "@cdktf/provider-google.bigqueryConnection.BigqueryConnection", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
BigqueryConnection.tfResourceType = "google_bigquery_connection";
