// @cdktf/provider-google/lib/data-google-oracle-database-autonomous-database/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataGoogleOracleDatabaseAutonomousDatabase = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsToTerraform = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsList = exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsOutputReference = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsToHclTerraform = exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // apex_version - computed: true, optional: false, required: false
    get apexVersion() {
        return this.getStringAttribute('apex_version');
    }
    // ords_version - computed: true, optional: false, required: false
    get ordsVersion() {
        return this.getStringAttribute('ords_version');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsList[_b] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // high - computed: true, optional: false, required: false
    get high() {
        return this.getStringAttribute('high');
    }
    // low - computed: true, optional: false, required: false
    get low() {
        return this.getStringAttribute('low');
    }
    // medium - computed: true, optional: false, required: false
    get medium() {
        return this.getStringAttribute('medium');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsList[_d] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // consumer_group - computed: true, optional: false, required: false
    get consumerGroup() {
        return this.getStringAttribute('consumer_group');
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    // host_format - computed: true, optional: false, required: false
    get hostFormat() {
        return this.getStringAttribute('host_format');
    }
    // is_regional - computed: true, optional: false, required: false
    get isRegional() {
        return this.getBooleanAttribute('is_regional');
    }
    // protocol - computed: true, optional: false, required: false
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    // session_mode - computed: true, optional: false, required: false
    get sessionMode() {
        return this.getStringAttribute('session_mode');
    }
    // syntax_format - computed: true, optional: false, required: false
    get syntaxFormat() {
        return this.getStringAttribute('syntax_format');
    }
    // tls_authentication - computed: true, optional: false, required: false
    get tlsAuthentication() {
        return this.getStringAttribute('tls_authentication');
    }
    // value - computed: true, optional: false, required: false
    get value() {
        return this.getStringAttribute('value');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesList[_f] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // all_connection_strings - computed: true, optional: false, required: false
        this._allConnectionStrings = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsAllConnectionStringsList(this, "all_connection_strings", false);
        // profiles - computed: true, optional: false, required: false
        this._profiles = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsProfilesList(this, "profiles", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get allConnectionStrings() {
        return this._allConnectionStrings;
    }
    // dedicated - computed: true, optional: false, required: false
    get dedicated() {
        return this.getStringAttribute('dedicated');
    }
    // high - computed: true, optional: false, required: false
    get high() {
        return this.getStringAttribute('high');
    }
    // low - computed: true, optional: false, required: false
    get low() {
        return this.getStringAttribute('low');
    }
    // medium - computed: true, optional: false, required: false
    get medium() {
        return this.getStringAttribute('medium');
    }
    get profiles() {
        return this._profiles;
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsList[_h] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // apex_uri - computed: true, optional: false, required: false
    get apexUri() {
        return this.getStringAttribute('apex_uri');
    }
    // database_transforms_uri - computed: true, optional: false, required: false
    get databaseTransformsUri() {
        return this.getStringAttribute('database_transforms_uri');
    }
    // graph_studio_uri - computed: true, optional: false, required: false
    get graphStudioUri() {
        return this.getStringAttribute('graph_studio_uri');
    }
    // machine_learning_notebook_uri - computed: true, optional: false, required: false
    get machineLearningNotebookUri() {
        return this.getStringAttribute('machine_learning_notebook_uri');
    }
    // machine_learning_user_management_uri - computed: true, optional: false, required: false
    get machineLearningUserManagementUri() {
        return this.getStringAttribute('machine_learning_user_management_uri');
    }
    // mongo_db_uri - computed: true, optional: false, required: false
    get mongoDbUri() {
        return this.getStringAttribute('mongo_db_uri');
    }
    // ords_uri - computed: true, optional: false, required: false
    get ordsUri() {
        return this.getStringAttribute('ords_uri');
    }
    // sql_dev_web_uri - computed: true, optional: false, required: false
    get sqlDevWebUri() {
        return this.getStringAttribute('sql_dev_web_uri');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsList[_k] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // email - computed: true, optional: false, required: false
    get email() {
        return this.getStringAttribute('email');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsList[_m] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_guard_role_changed_time - computed: true, optional: false, required: false
    get dataGuardRoleChangedTime() {
        return this.getStringAttribute('data_guard_role_changed_time');
    }
    // disaster_recovery_role_changed_time - computed: true, optional: false, required: false
    get disasterRecoveryRoleChangedTime() {
        return this.getStringAttribute('disaster_recovery_role_changed_time');
    }
    // lag_time_duration - computed: true, optional: false, required: false
    get lagTimeDuration() {
        return this.getStringAttribute('lag_time_duration');
    }
    // lifecycle_details - computed: true, optional: false, required: false
    get lifecycleDetails() {
        return this.getStringAttribute('lifecycle_details');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbList[_p] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hours - computed: true, optional: false, required: false
    get hours() {
        return this.getNumberAttribute('hours');
    }
    // minutes - computed: true, optional: false, required: false
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    // nanos - computed: true, optional: false, required: false
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    // seconds - computed: true, optional: false, required: false
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeList[_r] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hours - computed: true, optional: false, required: false
    get hours() {
        return this.getNumberAttribute('hours');
    }
    // minutes - computed: true, optional: false, required: false
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    // nanos - computed: true, optional: false, required: false
    get nanos() {
        return this.getNumberAttribute('nanos');
    }
    // seconds - computed: true, optional: false, required: false
    get seconds() {
        return this.getNumberAttribute('seconds');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeList[_t] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // start_time - computed: true, optional: false, required: false
        this._startTime = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStartTimeList(this, "start_time", false);
        // stop_time - computed: true, optional: false, required: false
        this._stopTime = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsStopTimeList(this, "stop_time", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day_of_week - computed: true, optional: false, required: false
    get dayOfWeek() {
        return this.getStringAttribute('day_of_week');
    }
    get startTime() {
        return this._startTime;
    }
    get stopTime() {
        return this._stopTime;
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsList[_v] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsList", version: "14.13.1" };
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesToTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesToTerraform;
function dataGoogleOracleDatabaseAutonomousDatabasePropertiesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleOracleDatabaseAutonomousDatabasePropertiesToHclTerraform = dataGoogleOracleDatabaseAutonomousDatabasePropertiesToHclTerraform;
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // apex_details - computed: true, optional: false, required: false
        this._apexDetails = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesApexDetailsList(this, "apex_details", false);
        // connection_strings - computed: true, optional: false, required: false
        this._connectionStrings = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionStringsList(this, "connection_strings", false);
        // connection_urls - computed: true, optional: false, required: false
        this._connectionUrls = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesConnectionUrlsList(this, "connection_urls", false);
        // customer_contacts - computed: true, optional: false, required: false
        this._customerContacts = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesCustomerContactsList(this, "customer_contacts", false);
        // local_standby_db - computed: true, optional: false, required: false
        this._localStandbyDb = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesLocalStandbyDbList(this, "local_standby_db", false);
        // scheduled_operation_details - computed: true, optional: false, required: false
        this._scheduledOperationDetails = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesScheduledOperationDetailsList(this, "scheduled_operation_details", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // actual_used_data_storage_size_tb - computed: true, optional: false, required: false
    get actualUsedDataStorageSizeTb() {
        return this.getNumberAttribute('actual_used_data_storage_size_tb');
    }
    // allocated_storage_size_tb - computed: true, optional: false, required: false
    get allocatedStorageSizeTb() {
        return this.getNumberAttribute('allocated_storage_size_tb');
    }
    get apexDetails() {
        return this._apexDetails;
    }
    // are_primary_allowlisted_ips_used - computed: true, optional: false, required: false
    get arePrimaryAllowlistedIpsUsed() {
        return this.getBooleanAttribute('are_primary_allowlisted_ips_used');
    }
    // autonomous_container_database_id - computed: true, optional: false, required: false
    get autonomousContainerDatabaseId() {
        return this.getStringAttribute('autonomous_container_database_id');
    }
    // available_upgrade_versions - computed: true, optional: false, required: false
    get availableUpgradeVersions() {
        return this.getListAttribute('available_upgrade_versions');
    }
    // backup_retention_period_days - computed: true, optional: false, required: false
    get backupRetentionPeriodDays() {
        return this.getNumberAttribute('backup_retention_period_days');
    }
    // character_set - computed: true, optional: false, required: false
    get characterSet() {
        return this.getStringAttribute('character_set');
    }
    // compute_count - computed: true, optional: false, required: false
    get computeCount() {
        return this.getNumberAttribute('compute_count');
    }
    get connectionStrings() {
        return this._connectionStrings;
    }
    get connectionUrls() {
        return this._connectionUrls;
    }
    get customerContacts() {
        return this._customerContacts;
    }
    // data_safe_state - computed: true, optional: false, required: false
    get dataSafeState() {
        return this.getStringAttribute('data_safe_state');
    }
    // data_storage_size_gb - computed: true, optional: false, required: false
    get dataStorageSizeGb() {
        return this.getNumberAttribute('data_storage_size_gb');
    }
    // data_storage_size_tb - computed: true, optional: false, required: false
    get dataStorageSizeTb() {
        return this.getNumberAttribute('data_storage_size_tb');
    }
    // database_management_state - computed: true, optional: false, required: false
    get databaseManagementState() {
        return this.getStringAttribute('database_management_state');
    }
    // db_edition - computed: true, optional: false, required: false
    get dbEdition() {
        return this.getStringAttribute('db_edition');
    }
    // db_version - computed: true, optional: false, required: false
    get dbVersion() {
        return this.getStringAttribute('db_version');
    }
    // db_workload - computed: true, optional: false, required: false
    get dbWorkload() {
        return this.getStringAttribute('db_workload');
    }
    // failed_data_recovery_duration - computed: true, optional: false, required: false
    get failedDataRecoveryDuration() {
        return this.getStringAttribute('failed_data_recovery_duration');
    }
    // is_auto_scaling_enabled - computed: true, optional: false, required: false
    get isAutoScalingEnabled() {
        return this.getBooleanAttribute('is_auto_scaling_enabled');
    }
    // is_local_data_guard_enabled - computed: true, optional: false, required: false
    get isLocalDataGuardEnabled() {
        return this.getBooleanAttribute('is_local_data_guard_enabled');
    }
    // is_storage_auto_scaling_enabled - computed: true, optional: false, required: false
    get isStorageAutoScalingEnabled() {
        return this.getBooleanAttribute('is_storage_auto_scaling_enabled');
    }
    // license_type - computed: true, optional: false, required: false
    get licenseType() {
        return this.getStringAttribute('license_type');
    }
    // lifecycle_details - computed: true, optional: false, required: false
    get lifecycleDetails() {
        return this.getStringAttribute('lifecycle_details');
    }
    // local_adg_auto_failover_max_data_loss_limit - computed: true, optional: false, required: false
    get localAdgAutoFailoverMaxDataLossLimit() {
        return this.getNumberAttribute('local_adg_auto_failover_max_data_loss_limit');
    }
    // local_disaster_recovery_type - computed: true, optional: false, required: false
    get localDisasterRecoveryType() {
        return this.getStringAttribute('local_disaster_recovery_type');
    }
    get localStandbyDb() {
        return this._localStandbyDb;
    }
    // maintenance_begin_time - computed: true, optional: false, required: false
    get maintenanceBeginTime() {
        return this.getStringAttribute('maintenance_begin_time');
    }
    // maintenance_end_time - computed: true, optional: false, required: false
    get maintenanceEndTime() {
        return this.getStringAttribute('maintenance_end_time');
    }
    // maintenance_schedule_type - computed: true, optional: false, required: false
    get maintenanceScheduleType() {
        return this.getStringAttribute('maintenance_schedule_type');
    }
    // memory_per_oracle_compute_unit_gbs - computed: true, optional: false, required: false
    get memoryPerOracleComputeUnitGbs() {
        return this.getNumberAttribute('memory_per_oracle_compute_unit_gbs');
    }
    // memory_table_gbs - computed: true, optional: false, required: false
    get memoryTableGbs() {
        return this.getNumberAttribute('memory_table_gbs');
    }
    // mtls_connection_required - computed: true, optional: false, required: false
    get mtlsConnectionRequired() {
        return this.getBooleanAttribute('mtls_connection_required');
    }
    // n_character_set - computed: true, optional: false, required: false
    get nCharacterSet() {
        return this.getStringAttribute('n_character_set');
    }
    // next_long_term_backup_time - computed: true, optional: false, required: false
    get nextLongTermBackupTime() {
        return this.getStringAttribute('next_long_term_backup_time');
    }
    // oci_url - computed: true, optional: false, required: false
    get ociUrl() {
        return this.getStringAttribute('oci_url');
    }
    // ocid - computed: true, optional: false, required: false
    get ocid() {
        return this.getStringAttribute('ocid');
    }
    // open_mode - computed: true, optional: false, required: false
    get openMode() {
        return this.getStringAttribute('open_mode');
    }
    // operations_insights_state - computed: true, optional: false, required: false
    get operationsInsightsState() {
        return this.getStringAttribute('operations_insights_state');
    }
    // peer_db_ids - computed: true, optional: false, required: false
    get peerDbIds() {
        return this.getListAttribute('peer_db_ids');
    }
    // permission_level - computed: true, optional: false, required: false
    get permissionLevel() {
        return this.getStringAttribute('permission_level');
    }
    // private_endpoint - computed: true, optional: false, required: false
    get privateEndpoint() {
        return this.getStringAttribute('private_endpoint');
    }
    // private_endpoint_ip - computed: true, optional: false, required: false
    get privateEndpointIp() {
        return this.getStringAttribute('private_endpoint_ip');
    }
    // private_endpoint_label - computed: true, optional: false, required: false
    get privateEndpointLabel() {
        return this.getStringAttribute('private_endpoint_label');
    }
    // refreshable_mode - computed: true, optional: false, required: false
    get refreshableMode() {
        return this.getStringAttribute('refreshable_mode');
    }
    // refreshable_state - computed: true, optional: false, required: false
    get refreshableState() {
        return this.getStringAttribute('refreshable_state');
    }
    // role - computed: true, optional: false, required: false
    get role() {
        return this.getStringAttribute('role');
    }
    get scheduledOperationDetails() {
        return this._scheduledOperationDetails;
    }
    // sql_web_developer_url - computed: true, optional: false, required: false
    get sqlWebDeveloperUrl() {
        return this.getStringAttribute('sql_web_developer_url');
    }
    // state - computed: true, optional: false, required: false
    get state() {
        return this.getStringAttribute('state');
    }
    // supported_clone_regions - computed: true, optional: false, required: false
    get supportedCloneRegions() {
        return this.getListAttribute('supported_clone_regions');
    }
    // total_auto_backup_storage_size_gbs - computed: true, optional: false, required: false
    get totalAutoBackupStorageSizeGbs() {
        return this.getNumberAttribute('total_auto_backup_storage_size_gbs');
    }
    // used_data_storage_size_tbs - computed: true, optional: false, required: false
    get usedDataStorageSizeTbs() {
        return this.getNumberAttribute('used_data_storage_size_tbs');
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesOutputReference = DataGoogleOracleDatabaseAutonomousDatabasePropertiesOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesOutputReference", version: "14.13.1" };
class DataGoogleOracleDatabaseAutonomousDatabasePropertiesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleOracleDatabaseAutonomousDatabasePropertiesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabasePropertiesList = DataGoogleOracleDatabaseAutonomousDatabasePropertiesList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabasePropertiesList[_x] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabasePropertiesList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_autonomous_database google_oracle_database_autonomous_database}
*/
class DataGoogleOracleDatabaseAutonomousDatabase extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleOracleDatabaseAutonomousDatabase resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleOracleDatabaseAutonomousDatabase to import
    * @param importFromId The id of the existing DataGoogleOracleDatabaseAutonomousDatabase that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_autonomous_database#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleOracleDatabaseAutonomousDatabase to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_oracle_database_autonomous_database", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/oracle_database_autonomous_database google_oracle_database_autonomous_database} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleOracleDatabaseAutonomousDatabaseConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_oracle_database_autonomous_database',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // properties - computed: true, optional: false, required: false
        this._properties = new DataGoogleOracleDatabaseAutonomousDatabasePropertiesList(this, "properties", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        this._autonomousDatabaseId = config.autonomousDatabaseId;
        this._id = config.id;
        this._location = config.location;
        this._project = config.project;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // admin_password - computed: true, optional: false, required: false
    get adminPassword() {
        return this.getStringAttribute('admin_password');
    }
    get autonomousDatabaseId() {
        return this.getStringAttribute('autonomous_database_id');
    }
    set autonomousDatabaseId(value) {
        this._autonomousDatabaseId = value;
    }
    // Temporarily expose input value. Use with caution.
    get autonomousDatabaseIdInput() {
        return this._autonomousDatabaseId;
    }
    // cidr - computed: true, optional: false, required: false
    get cidr() {
        return this.getStringAttribute('cidr');
    }
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    // database - computed: true, optional: false, required: false
    get database() {
        return this.getStringAttribute('database');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    // display_name - computed: true, optional: false, required: false
    get displayName() {
        return this.getStringAttribute('display_name');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // entitlement_id - computed: true, optional: false, required: false
    get entitlementId() {
        return this.getStringAttribute('entitlement_id');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // network - computed: true, optional: false, required: false
    get network() {
        return this.getStringAttribute('network');
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    get properties() {
        return this._properties;
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            autonomous_database_id: cdktf.stringToTerraform(this._autonomousDatabaseId),
            id: cdktf.stringToTerraform(this._id),
            location: cdktf.stringToTerraform(this._location),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            autonomous_database_id: {
                value: cdktf.stringToHclTerraform(this._autonomousDatabaseId),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleOracleDatabaseAutonomousDatabase = DataGoogleOracleDatabaseAutonomousDatabase;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleOracleDatabaseAutonomousDatabase[_y] = { fqn: "@cdktf/provider-google.dataGoogleOracleDatabaseAutonomousDatabase.DataGoogleOracleDatabaseAutonomousDatabase", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleOracleDatabaseAutonomousDatabase.tfResourceType = "google_oracle_database_autonomous_database";
