// @cdktf/provider-google/lib/network-security-security-profile/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkSecuritySecurityProfile = exports.NetworkSecuritySecurityProfileTimeoutsOutputReference = exports.networkSecuritySecurityProfileTimeoutsToHclTerraform = exports.networkSecuritySecurityProfileTimeoutsToTerraform = exports.NetworkSecuritySecurityProfileThreatPreventionProfileOutputReference = exports.networkSecuritySecurityProfileThreatPreventionProfileToHclTerraform = exports.networkSecuritySecurityProfileThreatPreventionProfileToTerraform = exports.NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesList = exports.NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesOutputReference = exports.networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToHclTerraform = exports.networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToTerraform = exports.NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesList = exports.NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesOutputReference = exports.networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToHclTerraform = exports.networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: cdktf.stringToTerraform(struct.action),
        severity: cdktf.stringToTerraform(struct.severity),
    };
}
exports.networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToTerraform = networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToTerraform;
function networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action: {
            value: cdktf.stringToHclTerraform(struct.action),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        severity: {
            value: cdktf.stringToHclTerraform(struct.severity),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToHclTerraform = networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToHclTerraform;
class NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action;
        }
        if (this._severity !== undefined) {
            hasAnyValues = true;
            internalValueResult.severity = this._severity;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._action = undefined;
            this._severity = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._action = value.action;
            this._severity = value.severity;
        }
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get severity() {
        return this.getStringAttribute('severity');
    }
    set severity(value) {
        this._severity = value;
    }
    // Temporarily expose input value. Use with caution.
    get severityInput() {
        return this._severity;
    }
}
exports.NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesOutputReference = NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesOutputReference[_a] = { fqn: "@cdktf/provider-google.networkSecuritySecurityProfile.NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesOutputReference", version: "14.13.1" };
class NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesList = NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesList;
_b = JSII_RTTI_SYMBOL_1;
NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesList[_b] = { fqn: "@cdktf/provider-google.networkSecuritySecurityProfile.NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesList", version: "14.13.1" };
function networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        action: cdktf.stringToTerraform(struct.action),
        threat_id: cdktf.stringToTerraform(struct.threatId),
    };
}
exports.networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToTerraform = networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToTerraform;
function networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        action: {
            value: cdktf.stringToHclTerraform(struct.action),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        threat_id: {
            value: cdktf.stringToHclTerraform(struct.threatId),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToHclTerraform = networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToHclTerraform;
class NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._action !== undefined) {
            hasAnyValues = true;
            internalValueResult.action = this._action;
        }
        if (this._threatId !== undefined) {
            hasAnyValues = true;
            internalValueResult.threatId = this._threatId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._action = undefined;
            this._threatId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._action = value.action;
            this._threatId = value.threatId;
        }
    }
    get action() {
        return this.getStringAttribute('action');
    }
    set action(value) {
        this._action = value;
    }
    // Temporarily expose input value. Use with caution.
    get actionInput() {
        return this._action;
    }
    get threatId() {
        return this.getStringAttribute('threat_id');
    }
    set threatId(value) {
        this._threatId = value;
    }
    // Temporarily expose input value. Use with caution.
    get threatIdInput() {
        return this._threatId;
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesOutputReference = NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesOutputReference[_c] = { fqn: "@cdktf/provider-google.networkSecuritySecurityProfile.NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesOutputReference", version: "14.13.1" };
class NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesList = NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesList;
_d = JSII_RTTI_SYMBOL_1;
NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesList[_d] = { fqn: "@cdktf/provider-google.networkSecuritySecurityProfile.NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesList", version: "14.13.1" };
function networkSecuritySecurityProfileThreatPreventionProfileToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        severity_overrides: cdktf.listMapper(networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToTerraform, true)(struct.severityOverrides),
        threat_overrides: cdktf.listMapper(networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToTerraform, true)(struct.threatOverrides),
    };
}
exports.networkSecuritySecurityProfileThreatPreventionProfileToTerraform = networkSecuritySecurityProfileThreatPreventionProfileToTerraform;
function networkSecuritySecurityProfileThreatPreventionProfileToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        severity_overrides: {
            value: cdktf.listMapperHcl(networkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesToHclTerraform, true)(struct.severityOverrides),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesList",
        },
        threat_overrides: {
            value: cdktf.listMapperHcl(networkSecuritySecurityProfileThreatPreventionProfileThreatOverridesToHclTerraform, true)(struct.threatOverrides),
            isBlock: true,
            type: "list",
            storageClassType: "NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesList",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecuritySecurityProfileThreatPreventionProfileToHclTerraform = networkSecuritySecurityProfileThreatPreventionProfileToHclTerraform;
class NetworkSecuritySecurityProfileThreatPreventionProfileOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // severity_overrides - computed: false, optional: true, required: false
        this._severityOverrides = new NetworkSecuritySecurityProfileThreatPreventionProfileSeverityOverridesList(this, "severity_overrides", false);
        // threat_overrides - computed: false, optional: true, required: false
        this._threatOverrides = new NetworkSecuritySecurityProfileThreatPreventionProfileThreatOverridesList(this, "threat_overrides", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._severityOverrides?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.severityOverrides = this._severityOverrides?.internalValue;
        }
        if (this._threatOverrides?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.threatOverrides = this._threatOverrides?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._severityOverrides.internalValue = undefined;
            this._threatOverrides.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._severityOverrides.internalValue = value.severityOverrides;
            this._threatOverrides.internalValue = value.threatOverrides;
        }
    }
    get severityOverrides() {
        return this._severityOverrides;
    }
    putSeverityOverrides(value) {
        this._severityOverrides.internalValue = value;
    }
    resetSeverityOverrides() {
        this._severityOverrides.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get severityOverridesInput() {
        return this._severityOverrides.internalValue;
    }
    get threatOverrides() {
        return this._threatOverrides;
    }
    putThreatOverrides(value) {
        this._threatOverrides.internalValue = value;
    }
    resetThreatOverrides() {
        this._threatOverrides.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get threatOverridesInput() {
        return this._threatOverrides.internalValue;
    }
}
exports.NetworkSecuritySecurityProfileThreatPreventionProfileOutputReference = NetworkSecuritySecurityProfileThreatPreventionProfileOutputReference;
_e = JSII_RTTI_SYMBOL_1;
NetworkSecuritySecurityProfileThreatPreventionProfileOutputReference[_e] = { fqn: "@cdktf/provider-google.networkSecuritySecurityProfile.NetworkSecuritySecurityProfileThreatPreventionProfileOutputReference", version: "14.13.1" };
function networkSecuritySecurityProfileTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.networkSecuritySecurityProfileTimeoutsToTerraform = networkSecuritySecurityProfileTimeoutsToTerraform;
function networkSecuritySecurityProfileTimeoutsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {
        create: {
            value: cdktf.stringToHclTerraform(struct.create),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        delete: {
            value: cdktf.stringToHclTerraform(struct.delete),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
        update: {
            value: cdktf.stringToHclTerraform(struct.update),
            isBlock: false,
            type: "simple",
            storageClassType: "string",
        },
    };
    // remove undefined attributes
    return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
}
exports.networkSecuritySecurityProfileTimeoutsToHclTerraform = networkSecuritySecurityProfileTimeoutsToHclTerraform;
class NetworkSecuritySecurityProfileTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.NetworkSecuritySecurityProfileTimeoutsOutputReference = NetworkSecuritySecurityProfileTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
NetworkSecuritySecurityProfileTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-google.networkSecuritySecurityProfile.NetworkSecuritySecurityProfileTimeoutsOutputReference", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_security_profile google_network_security_security_profile}
*/
class NetworkSecuritySecurityProfile extends cdktf.TerraformResource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a NetworkSecuritySecurityProfile resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the NetworkSecuritySecurityProfile to import
    * @param importFromId The id of the existing NetworkSecuritySecurityProfile that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_security_profile#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the NetworkSecuritySecurityProfile to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_network_security_security_profile", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/resources/network_security_security_profile google_network_security_security_profile} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options NetworkSecuritySecurityProfileConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_network_security_security_profile',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // threat_prevention_profile - computed: false, optional: true, required: false
        this._threatPreventionProfile = new NetworkSecuritySecurityProfileThreatPreventionProfileOutputReference(this, "threat_prevention_profile");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new NetworkSecuritySecurityProfileTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._id = config.id;
        this._labels = config.labels;
        this._location = config.location;
        this._name = config.name;
        this._parent = config.parent;
        this._type = config.type;
        this._threatPreventionProfile.internalValue = config.threatPreventionProfile;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // create_time - computed: true, optional: false, required: false
    get createTime() {
        return this.getStringAttribute('create_time');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // etag - computed: true, optional: false, required: false
    get etag() {
        return this.getStringAttribute('etag');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get location() {
        return this.getStringAttribute('location');
    }
    set location(value) {
        this._location = value;
    }
    resetLocation() {
        this._location = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get parent() {
        return this.getStringAttribute('parent');
    }
    set parent(value) {
        this._parent = value;
    }
    resetParent() {
        this._parent = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get parentInput() {
        return this._parent;
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    // update_time - computed: true, optional: false, required: false
    get updateTime() {
        return this.getStringAttribute('update_time');
    }
    get threatPreventionProfile() {
        return this._threatPreventionProfile;
    }
    putThreatPreventionProfile(value) {
        this._threatPreventionProfile.internalValue = value;
    }
    resetThreatPreventionProfile() {
        this._threatPreventionProfile.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get threatPreventionProfileInput() {
        return this._threatPreventionProfile.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            location: cdktf.stringToTerraform(this._location),
            name: cdktf.stringToTerraform(this._name),
            parent: cdktf.stringToTerraform(this._parent),
            type: cdktf.stringToTerraform(this._type),
            threat_prevention_profile: networkSecuritySecurityProfileThreatPreventionProfileToTerraform(this._threatPreventionProfile.internalValue),
            timeouts: networkSecuritySecurityProfileTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            description: {
                value: cdktf.stringToHclTerraform(this._description),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            labels: {
                value: cdktf.hashMapperHcl(cdktf.stringToHclTerraform)(this._labels),
                isBlock: false,
                type: "map",
                storageClassType: "stringMap",
            },
            location: {
                value: cdktf.stringToHclTerraform(this._location),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            parent: {
                value: cdktf.stringToHclTerraform(this._parent),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            type: {
                value: cdktf.stringToHclTerraform(this._type),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            threat_prevention_profile: {
                value: networkSecuritySecurityProfileThreatPreventionProfileToHclTerraform(this._threatPreventionProfile.internalValue),
                isBlock: true,
                type: "list",
                storageClassType: "NetworkSecuritySecurityProfileThreatPreventionProfileList",
            },
            timeouts: {
                value: networkSecuritySecurityProfileTimeoutsToHclTerraform(this._timeouts.internalValue),
                isBlock: true,
                type: "struct",
                storageClassType: "NetworkSecuritySecurityProfileTimeouts",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.NetworkSecuritySecurityProfile = NetworkSecuritySecurityProfile;
_g = JSII_RTTI_SYMBOL_1;
NetworkSecuritySecurityProfile[_g] = { fqn: "@cdktf/provider-google.networkSecuritySecurityProfile.NetworkSecuritySecurityProfile", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
NetworkSecuritySecurityProfile.tfResourceType = "google_network_security_security_profile";
