// @cdktf/provider-google/lib/data-google-storage-bucket/index.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2;
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataGoogleStorageBucketWebsiteToHclTerraform = exports.dataGoogleStorageBucketWebsiteToTerraform = exports.DataGoogleStorageBucketVersioningList = exports.DataGoogleStorageBucketVersioningOutputReference = exports.dataGoogleStorageBucketVersioningToHclTerraform = exports.dataGoogleStorageBucketVersioningToTerraform = exports.DataGoogleStorageBucketSoftDeletePolicyList = exports.DataGoogleStorageBucketSoftDeletePolicyOutputReference = exports.dataGoogleStorageBucketSoftDeletePolicyToHclTerraform = exports.dataGoogleStorageBucketSoftDeletePolicyToTerraform = exports.DataGoogleStorageBucketRetentionPolicyList = exports.DataGoogleStorageBucketRetentionPolicyOutputReference = exports.dataGoogleStorageBucketRetentionPolicyToHclTerraform = exports.dataGoogleStorageBucketRetentionPolicyToTerraform = exports.DataGoogleStorageBucketLoggingList = exports.DataGoogleStorageBucketLoggingOutputReference = exports.dataGoogleStorageBucketLoggingToHclTerraform = exports.dataGoogleStorageBucketLoggingToTerraform = exports.DataGoogleStorageBucketLifecycleRuleList = exports.DataGoogleStorageBucketLifecycleRuleOutputReference = exports.dataGoogleStorageBucketLifecycleRuleToHclTerraform = exports.dataGoogleStorageBucketLifecycleRuleToTerraform = exports.DataGoogleStorageBucketLifecycleRuleConditionList = exports.DataGoogleStorageBucketLifecycleRuleConditionOutputReference = exports.dataGoogleStorageBucketLifecycleRuleConditionToHclTerraform = exports.dataGoogleStorageBucketLifecycleRuleConditionToTerraform = exports.DataGoogleStorageBucketLifecycleRuleActionList = exports.DataGoogleStorageBucketLifecycleRuleActionOutputReference = exports.dataGoogleStorageBucketLifecycleRuleActionToHclTerraform = exports.dataGoogleStorageBucketLifecycleRuleActionToTerraform = exports.DataGoogleStorageBucketHierarchicalNamespaceList = exports.DataGoogleStorageBucketHierarchicalNamespaceOutputReference = exports.dataGoogleStorageBucketHierarchicalNamespaceToHclTerraform = exports.dataGoogleStorageBucketHierarchicalNamespaceToTerraform = exports.DataGoogleStorageBucketEncryptionList = exports.DataGoogleStorageBucketEncryptionOutputReference = exports.dataGoogleStorageBucketEncryptionToHclTerraform = exports.dataGoogleStorageBucketEncryptionToTerraform = exports.DataGoogleStorageBucketCustomPlacementConfigList = exports.DataGoogleStorageBucketCustomPlacementConfigOutputReference = exports.dataGoogleStorageBucketCustomPlacementConfigToHclTerraform = exports.dataGoogleStorageBucketCustomPlacementConfigToTerraform = exports.DataGoogleStorageBucketCorsList = exports.DataGoogleStorageBucketCorsOutputReference = exports.dataGoogleStorageBucketCorsToHclTerraform = exports.dataGoogleStorageBucketCorsToTerraform = exports.DataGoogleStorageBucketAutoclassList = exports.DataGoogleStorageBucketAutoclassOutputReference = exports.dataGoogleStorageBucketAutoclassToHclTerraform = exports.dataGoogleStorageBucketAutoclassToTerraform = void 0;
exports.DataGoogleStorageBucket = exports.DataGoogleStorageBucketWebsiteList = exports.DataGoogleStorageBucketWebsiteOutputReference = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataGoogleStorageBucketAutoclassToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketAutoclassToTerraform = dataGoogleStorageBucketAutoclassToTerraform;
function dataGoogleStorageBucketAutoclassToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketAutoclassToHclTerraform = dataGoogleStorageBucketAutoclassToHclTerraform;
class DataGoogleStorageBucketAutoclassOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // terminal_storage_class - computed: true, optional: false, required: false
    get terminalStorageClass() {
        return this.getStringAttribute('terminal_storage_class');
    }
}
exports.DataGoogleStorageBucketAutoclassOutputReference = DataGoogleStorageBucketAutoclassOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketAutoclassOutputReference[_a] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketAutoclassOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketAutoclassList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketAutoclassOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketAutoclassList = DataGoogleStorageBucketAutoclassList;
_b = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketAutoclassList[_b] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketAutoclassList", version: "14.13.1" };
function dataGoogleStorageBucketCorsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketCorsToTerraform = dataGoogleStorageBucketCorsToTerraform;
function dataGoogleStorageBucketCorsToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketCorsToHclTerraform = dataGoogleStorageBucketCorsToHclTerraform;
class DataGoogleStorageBucketCorsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_age_seconds - computed: true, optional: false, required: false
    get maxAgeSeconds() {
        return this.getNumberAttribute('max_age_seconds');
    }
    // method - computed: true, optional: false, required: false
    get method() {
        return this.getListAttribute('method');
    }
    // origin - computed: true, optional: false, required: false
    get origin() {
        return this.getListAttribute('origin');
    }
    // response_header - computed: true, optional: false, required: false
    get responseHeader() {
        return this.getListAttribute('response_header');
    }
}
exports.DataGoogleStorageBucketCorsOutputReference = DataGoogleStorageBucketCorsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketCorsOutputReference[_c] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketCorsOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketCorsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketCorsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketCorsList = DataGoogleStorageBucketCorsList;
_d = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketCorsList[_d] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketCorsList", version: "14.13.1" };
function dataGoogleStorageBucketCustomPlacementConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketCustomPlacementConfigToTerraform = dataGoogleStorageBucketCustomPlacementConfigToTerraform;
function dataGoogleStorageBucketCustomPlacementConfigToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketCustomPlacementConfigToHclTerraform = dataGoogleStorageBucketCustomPlacementConfigToHclTerraform;
class DataGoogleStorageBucketCustomPlacementConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_locations - computed: true, optional: false, required: false
    get dataLocations() {
        return cdktf.Fn.tolist(this.getListAttribute('data_locations'));
    }
}
exports.DataGoogleStorageBucketCustomPlacementConfigOutputReference = DataGoogleStorageBucketCustomPlacementConfigOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketCustomPlacementConfigOutputReference[_e] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketCustomPlacementConfigOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketCustomPlacementConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketCustomPlacementConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketCustomPlacementConfigList = DataGoogleStorageBucketCustomPlacementConfigList;
_f = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketCustomPlacementConfigList[_f] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketCustomPlacementConfigList", version: "14.13.1" };
function dataGoogleStorageBucketEncryptionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketEncryptionToTerraform = dataGoogleStorageBucketEncryptionToTerraform;
function dataGoogleStorageBucketEncryptionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketEncryptionToHclTerraform = dataGoogleStorageBucketEncryptionToHclTerraform;
class DataGoogleStorageBucketEncryptionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // default_kms_key_name - computed: true, optional: false, required: false
    get defaultKmsKeyName() {
        return this.getStringAttribute('default_kms_key_name');
    }
}
exports.DataGoogleStorageBucketEncryptionOutputReference = DataGoogleStorageBucketEncryptionOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketEncryptionOutputReference[_g] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketEncryptionOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketEncryptionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketEncryptionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketEncryptionList = DataGoogleStorageBucketEncryptionList;
_h = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketEncryptionList[_h] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketEncryptionList", version: "14.13.1" };
function dataGoogleStorageBucketHierarchicalNamespaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketHierarchicalNamespaceToTerraform = dataGoogleStorageBucketHierarchicalNamespaceToTerraform;
function dataGoogleStorageBucketHierarchicalNamespaceToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketHierarchicalNamespaceToHclTerraform = dataGoogleStorageBucketHierarchicalNamespaceToHclTerraform;
class DataGoogleStorageBucketHierarchicalNamespaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataGoogleStorageBucketHierarchicalNamespaceOutputReference = DataGoogleStorageBucketHierarchicalNamespaceOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketHierarchicalNamespaceOutputReference[_j] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketHierarchicalNamespaceOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketHierarchicalNamespaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketHierarchicalNamespaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketHierarchicalNamespaceList = DataGoogleStorageBucketHierarchicalNamespaceList;
_k = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketHierarchicalNamespaceList[_k] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketHierarchicalNamespaceList", version: "14.13.1" };
function dataGoogleStorageBucketLifecycleRuleActionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketLifecycleRuleActionToTerraform = dataGoogleStorageBucketLifecycleRuleActionToTerraform;
function dataGoogleStorageBucketLifecycleRuleActionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketLifecycleRuleActionToHclTerraform = dataGoogleStorageBucketLifecycleRuleActionToHclTerraform;
class DataGoogleStorageBucketLifecycleRuleActionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // storage_class - computed: true, optional: false, required: false
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataGoogleStorageBucketLifecycleRuleActionOutputReference = DataGoogleStorageBucketLifecycleRuleActionOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleActionOutputReference[_l] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleActionOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketLifecycleRuleActionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketLifecycleRuleActionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketLifecycleRuleActionList = DataGoogleStorageBucketLifecycleRuleActionList;
_m = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleActionList[_m] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleActionList", version: "14.13.1" };
function dataGoogleStorageBucketLifecycleRuleConditionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketLifecycleRuleConditionToTerraform = dataGoogleStorageBucketLifecycleRuleConditionToTerraform;
function dataGoogleStorageBucketLifecycleRuleConditionToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketLifecycleRuleConditionToHclTerraform = dataGoogleStorageBucketLifecycleRuleConditionToHclTerraform;
class DataGoogleStorageBucketLifecycleRuleConditionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // age - computed: true, optional: false, required: false
    get age() {
        return this.getNumberAttribute('age');
    }
    // created_before - computed: true, optional: false, required: false
    get createdBefore() {
        return this.getStringAttribute('created_before');
    }
    // custom_time_before - computed: true, optional: false, required: false
    get customTimeBefore() {
        return this.getStringAttribute('custom_time_before');
    }
    // days_since_custom_time - computed: true, optional: false, required: false
    get daysSinceCustomTime() {
        return this.getNumberAttribute('days_since_custom_time');
    }
    // days_since_noncurrent_time - computed: true, optional: false, required: false
    get daysSinceNoncurrentTime() {
        return this.getNumberAttribute('days_since_noncurrent_time');
    }
    // matches_prefix - computed: true, optional: false, required: false
    get matchesPrefix() {
        return this.getListAttribute('matches_prefix');
    }
    // matches_storage_class - computed: true, optional: false, required: false
    get matchesStorageClass() {
        return this.getListAttribute('matches_storage_class');
    }
    // matches_suffix - computed: true, optional: false, required: false
    get matchesSuffix() {
        return this.getListAttribute('matches_suffix');
    }
    // noncurrent_time_before - computed: true, optional: false, required: false
    get noncurrentTimeBefore() {
        return this.getStringAttribute('noncurrent_time_before');
    }
    // num_newer_versions - computed: true, optional: false, required: false
    get numNewerVersions() {
        return this.getNumberAttribute('num_newer_versions');
    }
    // send_age_if_zero - computed: true, optional: false, required: false
    get sendAgeIfZero() {
        return this.getBooleanAttribute('send_age_if_zero');
    }
    // send_days_since_custom_time_if_zero - computed: true, optional: false, required: false
    get sendDaysSinceCustomTimeIfZero() {
        return this.getBooleanAttribute('send_days_since_custom_time_if_zero');
    }
    // send_days_since_noncurrent_time_if_zero - computed: true, optional: false, required: false
    get sendDaysSinceNoncurrentTimeIfZero() {
        return this.getBooleanAttribute('send_days_since_noncurrent_time_if_zero');
    }
    // send_num_newer_versions_if_zero - computed: true, optional: false, required: false
    get sendNumNewerVersionsIfZero() {
        return this.getBooleanAttribute('send_num_newer_versions_if_zero');
    }
    // with_state - computed: true, optional: false, required: false
    get withState() {
        return this.getStringAttribute('with_state');
    }
}
exports.DataGoogleStorageBucketLifecycleRuleConditionOutputReference = DataGoogleStorageBucketLifecycleRuleConditionOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleConditionOutputReference[_o] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleConditionOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketLifecycleRuleConditionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketLifecycleRuleConditionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketLifecycleRuleConditionList = DataGoogleStorageBucketLifecycleRuleConditionList;
_p = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleConditionList[_p] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleConditionList", version: "14.13.1" };
function dataGoogleStorageBucketLifecycleRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketLifecycleRuleToTerraform = dataGoogleStorageBucketLifecycleRuleToTerraform;
function dataGoogleStorageBucketLifecycleRuleToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketLifecycleRuleToHclTerraform = dataGoogleStorageBucketLifecycleRuleToHclTerraform;
class DataGoogleStorageBucketLifecycleRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // action - computed: true, optional: false, required: false
        this._action = new DataGoogleStorageBucketLifecycleRuleActionList(this, "action", true);
        // condition - computed: true, optional: false, required: false
        this._condition = new DataGoogleStorageBucketLifecycleRuleConditionList(this, "condition", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get action() {
        return this._action;
    }
    get condition() {
        return this._condition;
    }
}
exports.DataGoogleStorageBucketLifecycleRuleOutputReference = DataGoogleStorageBucketLifecycleRuleOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleOutputReference[_q] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketLifecycleRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketLifecycleRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketLifecycleRuleList = DataGoogleStorageBucketLifecycleRuleList;
_r = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLifecycleRuleList[_r] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLifecycleRuleList", version: "14.13.1" };
function dataGoogleStorageBucketLoggingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketLoggingToTerraform = dataGoogleStorageBucketLoggingToTerraform;
function dataGoogleStorageBucketLoggingToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketLoggingToHclTerraform = dataGoogleStorageBucketLoggingToHclTerraform;
class DataGoogleStorageBucketLoggingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // log_bucket - computed: true, optional: false, required: false
    get logBucket() {
        return this.getStringAttribute('log_bucket');
    }
    // log_object_prefix - computed: true, optional: false, required: false
    get logObjectPrefix() {
        return this.getStringAttribute('log_object_prefix');
    }
}
exports.DataGoogleStorageBucketLoggingOutputReference = DataGoogleStorageBucketLoggingOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLoggingOutputReference[_s] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLoggingOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketLoggingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketLoggingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketLoggingList = DataGoogleStorageBucketLoggingList;
_t = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketLoggingList[_t] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketLoggingList", version: "14.13.1" };
function dataGoogleStorageBucketRetentionPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketRetentionPolicyToTerraform = dataGoogleStorageBucketRetentionPolicyToTerraform;
function dataGoogleStorageBucketRetentionPolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketRetentionPolicyToHclTerraform = dataGoogleStorageBucketRetentionPolicyToHclTerraform;
class DataGoogleStorageBucketRetentionPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // is_locked - computed: true, optional: false, required: false
    get isLocked() {
        return this.getBooleanAttribute('is_locked');
    }
    // retention_period - computed: true, optional: false, required: false
    get retentionPeriod() {
        return this.getNumberAttribute('retention_period');
    }
}
exports.DataGoogleStorageBucketRetentionPolicyOutputReference = DataGoogleStorageBucketRetentionPolicyOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketRetentionPolicyOutputReference[_u] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketRetentionPolicyOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketRetentionPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketRetentionPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketRetentionPolicyList = DataGoogleStorageBucketRetentionPolicyList;
_v = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketRetentionPolicyList[_v] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketRetentionPolicyList", version: "14.13.1" };
function dataGoogleStorageBucketSoftDeletePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketSoftDeletePolicyToTerraform = dataGoogleStorageBucketSoftDeletePolicyToTerraform;
function dataGoogleStorageBucketSoftDeletePolicyToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketSoftDeletePolicyToHclTerraform = dataGoogleStorageBucketSoftDeletePolicyToHclTerraform;
class DataGoogleStorageBucketSoftDeletePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // effective_time - computed: true, optional: false, required: false
    get effectiveTime() {
        return this.getStringAttribute('effective_time');
    }
    // retention_duration_seconds - computed: true, optional: false, required: false
    get retentionDurationSeconds() {
        return this.getNumberAttribute('retention_duration_seconds');
    }
}
exports.DataGoogleStorageBucketSoftDeletePolicyOutputReference = DataGoogleStorageBucketSoftDeletePolicyOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketSoftDeletePolicyOutputReference[_w] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketSoftDeletePolicyOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketSoftDeletePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketSoftDeletePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketSoftDeletePolicyList = DataGoogleStorageBucketSoftDeletePolicyList;
_x = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketSoftDeletePolicyList[_x] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketSoftDeletePolicyList", version: "14.13.1" };
function dataGoogleStorageBucketVersioningToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketVersioningToTerraform = dataGoogleStorageBucketVersioningToTerraform;
function dataGoogleStorageBucketVersioningToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketVersioningToHclTerraform = dataGoogleStorageBucketVersioningToHclTerraform;
class DataGoogleStorageBucketVersioningOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
}
exports.DataGoogleStorageBucketVersioningOutputReference = DataGoogleStorageBucketVersioningOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketVersioningOutputReference[_y] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketVersioningOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketVersioningList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketVersioningOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketVersioningList = DataGoogleStorageBucketVersioningList;
_z = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketVersioningList[_z] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketVersioningList", version: "14.13.1" };
function dataGoogleStorageBucketWebsiteToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataGoogleStorageBucketWebsiteToTerraform = dataGoogleStorageBucketWebsiteToTerraform;
function dataGoogleStorageBucketWebsiteToHclTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    const attrs = {};
    return attrs;
}
exports.dataGoogleStorageBucketWebsiteToHclTerraform = dataGoogleStorageBucketWebsiteToHclTerraform;
class DataGoogleStorageBucketWebsiteOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // main_page_suffix - computed: true, optional: false, required: false
    get mainPageSuffix() {
        return this.getStringAttribute('main_page_suffix');
    }
    // not_found_page - computed: true, optional: false, required: false
    get notFoundPage() {
        return this.getStringAttribute('not_found_page');
    }
}
exports.DataGoogleStorageBucketWebsiteOutputReference = DataGoogleStorageBucketWebsiteOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketWebsiteOutputReference[_0] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketWebsiteOutputReference", version: "14.13.1" };
class DataGoogleStorageBucketWebsiteList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataGoogleStorageBucketWebsiteOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataGoogleStorageBucketWebsiteList = DataGoogleStorageBucketWebsiteList;
_1 = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucketWebsiteList[_1] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucketWebsiteList", version: "14.13.1" };
/**
* Represents a {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/storage_bucket google_storage_bucket}
*/
class DataGoogleStorageBucket extends cdktf.TerraformDataSource {
    // ==============
    // STATIC Methods
    // ==============
    /**
    * Generates CDKTF code for importing a DataGoogleStorageBucket resource upon running "cdktf plan <stack-name>"
    * @param scope The scope in which to define this construct
    * @param importToId The construct id used in the generated config for the DataGoogleStorageBucket to import
    * @param importFromId The id of the existing DataGoogleStorageBucket that should be imported. Refer to the {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/storage_bucket#import import section} in the documentation of this resource for the id to use
    * @param provider? Optional instance of the provider where the DataGoogleStorageBucket to import is found
    */
    static generateConfigForImport(scope, importToId, importFromId, provider) {
        return new cdktf.ImportableResource(scope, importToId, { terraformResourceType: "google_storage_bucket", importId: importFromId, provider });
    }
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://registry.terraform.io/providers/hashicorp/google/6.14.1/docs/data-sources/storage_bucket google_storage_bucket} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataGoogleStorageBucketConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'google_storage_bucket',
            terraformGeneratorMetadata: {
                providerName: 'google',
                providerVersion: '6.14.1',
                providerVersionConstraint: '~> 6.0'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // autoclass - computed: true, optional: false, required: false
        this._autoclass = new DataGoogleStorageBucketAutoclassList(this, "autoclass", false);
        // cors - computed: true, optional: false, required: false
        this._cors = new DataGoogleStorageBucketCorsList(this, "cors", false);
        // custom_placement_config - computed: true, optional: false, required: false
        this._customPlacementConfig = new DataGoogleStorageBucketCustomPlacementConfigList(this, "custom_placement_config", false);
        // effective_labels - computed: true, optional: false, required: false
        this._effectiveLabels = new cdktf.StringMap(this, "effective_labels");
        // encryption - computed: true, optional: false, required: false
        this._encryption = new DataGoogleStorageBucketEncryptionList(this, "encryption", false);
        // hierarchical_namespace - computed: true, optional: false, required: false
        this._hierarchicalNamespace = new DataGoogleStorageBucketHierarchicalNamespaceList(this, "hierarchical_namespace", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // lifecycle_rule - computed: true, optional: false, required: false
        this._lifecycleRule = new DataGoogleStorageBucketLifecycleRuleList(this, "lifecycle_rule", false);
        // logging - computed: true, optional: false, required: false
        this._logging = new DataGoogleStorageBucketLoggingList(this, "logging", false);
        // retention_policy - computed: true, optional: false, required: false
        this._retentionPolicy = new DataGoogleStorageBucketRetentionPolicyList(this, "retention_policy", false);
        // soft_delete_policy - computed: true, optional: false, required: false
        this._softDeletePolicy = new DataGoogleStorageBucketSoftDeletePolicyList(this, "soft_delete_policy", false);
        // terraform_labels - computed: true, optional: false, required: false
        this._terraformLabels = new cdktf.StringMap(this, "terraform_labels");
        // versioning - computed: true, optional: false, required: false
        this._versioning = new DataGoogleStorageBucketVersioningList(this, "versioning", false);
        // website - computed: true, optional: false, required: false
        this._website = new DataGoogleStorageBucketWebsiteList(this, "website", false);
        this._id = config.id;
        this._name = config.name;
        this._project = config.project;
    }
    get autoclass() {
        return this._autoclass;
    }
    get cors() {
        return this._cors;
    }
    get customPlacementConfig() {
        return this._customPlacementConfig;
    }
    // default_event_based_hold - computed: true, optional: false, required: false
    get defaultEventBasedHold() {
        return this.getBooleanAttribute('default_event_based_hold');
    }
    get effectiveLabels() {
        return this._effectiveLabels;
    }
    // enable_object_retention - computed: true, optional: false, required: false
    get enableObjectRetention() {
        return this.getBooleanAttribute('enable_object_retention');
    }
    get encryption() {
        return this._encryption;
    }
    // force_destroy - computed: true, optional: false, required: false
    get forceDestroy() {
        return this.getBooleanAttribute('force_destroy');
    }
    get hierarchicalNamespace() {
        return this._hierarchicalNamespace;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get lifecycleRule() {
        return this._lifecycleRule;
    }
    // location - computed: true, optional: false, required: false
    get location() {
        return this.getStringAttribute('location');
    }
    get logging() {
        return this._logging;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get project() {
        return this.getStringAttribute('project');
    }
    set project(value) {
        this._project = value;
    }
    resetProject() {
        this._project = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get projectInput() {
        return this._project;
    }
    // project_number - computed: true, optional: false, required: false
    get projectNumber() {
        return this.getNumberAttribute('project_number');
    }
    // public_access_prevention - computed: true, optional: false, required: false
    get publicAccessPrevention() {
        return this.getStringAttribute('public_access_prevention');
    }
    // requester_pays - computed: true, optional: false, required: false
    get requesterPays() {
        return this.getBooleanAttribute('requester_pays');
    }
    get retentionPolicy() {
        return this._retentionPolicy;
    }
    // rpo - computed: true, optional: false, required: false
    get rpo() {
        return this.getStringAttribute('rpo');
    }
    // self_link - computed: true, optional: false, required: false
    get selfLink() {
        return this.getStringAttribute('self_link');
    }
    get softDeletePolicy() {
        return this._softDeletePolicy;
    }
    // storage_class - computed: true, optional: false, required: false
    get storageClass() {
        return this.getStringAttribute('storage_class');
    }
    get terraformLabels() {
        return this._terraformLabels;
    }
    // uniform_bucket_level_access - computed: true, optional: false, required: false
    get uniformBucketLevelAccess() {
        return this.getBooleanAttribute('uniform_bucket_level_access');
    }
    // url - computed: true, optional: false, required: false
    get url() {
        return this.getStringAttribute('url');
    }
    get versioning() {
        return this._versioning;
    }
    get website() {
        return this._website;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            project: cdktf.stringToTerraform(this._project),
        };
    }
    synthesizeHclAttributes() {
        const attrs = {
            id: {
                value: cdktf.stringToHclTerraform(this._id),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            name: {
                value: cdktf.stringToHclTerraform(this._name),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
            project: {
                value: cdktf.stringToHclTerraform(this._project),
                isBlock: false,
                type: "simple",
                storageClassType: "string",
            },
        };
        // remove undefined attributes
        return Object.fromEntries(Object.entries(attrs).filter(([_, value]) => value !== undefined && value.value !== undefined));
    }
}
exports.DataGoogleStorageBucket = DataGoogleStorageBucket;
_2 = JSII_RTTI_SYMBOL_1;
DataGoogleStorageBucket[_2] = { fqn: "@cdktf/provider-google.dataGoogleStorageBucket.DataGoogleStorageBucket", version: "14.13.1" };
// =================
// STATIC PROPERTIES
// =================
DataGoogleStorageBucket.tfResourceType = "google_storage_bucket";
